--- a/net/minecraft/entity/projectile/ArrowEntity.java
+++ b/net/minecraft/entity/projectile/ArrowEntity.java
@@ -17,17 +_,31 @@
 import net.minecraft.potion.Potion;
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.potion.Potions;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ArrowEntity extends AbstractArrowEntity {
    private static final DataParameter<Integer> field_184559_f = EntityDataManager.func_187226_a(ArrowEntity.class, DataSerializers.field_187192_b);
    private Potion field_184560_g = Potions.field_185229_a;
-   private final Set<EffectInstance> field_184561_h = Sets.newHashSet();
+   public final Set<EffectInstance> field_184561_h = Sets.newHashSet();
    private boolean field_191509_at;
 
+   // Spigot Start
+   @Override
+   public void inactiveTick()
+   {
+      if ( this.field_70254_i )
+      {
+         this.field_70252_j += 1;
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    public ArrowEntity(EntityType<? extends ArrowEntity> p_i50172_1_, World p_i50172_2_) {
       super(p_i50172_1_, p_i50172_2_);
    }
@@ -122,11 +_,30 @@
       }
    }
 
+   // CraftBukkit start
+   public void refreshEffects() {
+      this.func_184212_Q().func_187227_b(ArrowEntity.field_184559_f, PotionUtils.func_185181_a(PotionUtils.func_185186_a(this.field_184560_g, this.field_184561_h)));
+   }
+
+   public String getTypeCB() {
+      return Registry.field_212621_j.func_177774_c(this.field_184560_g).toString();
+   }
+
+   public void setType(final String string) {
+      this.field_184560_g = Registry.field_212621_j.func_82594_a(new ResourceLocation(string));
+      this.func_184212_Q().func_187227_b(ArrowEntity.field_184559_f, PotionUtils.func_185181_a(PotionUtils.func_185186_a(this.field_184560_g, this.field_184561_h)));
+   }
+
+   public boolean isTipped() {
+      return !this.field_184561_h.isEmpty() || this.field_184560_g != Potions.field_185229_a;
+   }
+   // CraftBukkit end
+
    public int func_184557_n() {
       return this.field_70180_af.func_187225_a(field_184559_f);
    }
 
-   private void func_191507_d(int p_191507_1_) {
+   public void func_191507_d(int p_191507_1_) {
       this.field_191509_at = true;
       this.field_70180_af.func_187227_b(field_184559_f, p_191507_1_);
    }
@@ -175,12 +_,12 @@
       super.func_184548_a(p_184548_1_);
 
       for(EffectInstance effectinstance : this.field_184560_g.func_185170_a()) {
-         p_184548_1_.func_195064_c(new EffectInstance(effectinstance.func_188419_a(), Math.max(effectinstance.func_76459_b() / 8, 1), effectinstance.func_76458_c(), effectinstance.func_82720_e(), effectinstance.func_188418_e()));
+         p_184548_1_.addEffect(new EffectInstance(effectinstance.func_188419_a(), Math.max(effectinstance.func_76459_b() / 8, 1), effectinstance.func_76458_c(), effectinstance.func_82720_e(), effectinstance.func_188418_e()), Cause.ARROW); // CraftBukkit
       }
 
       if (!this.field_184561_h.isEmpty()) {
          for(EffectInstance effectinstance1 : this.field_184561_h) {
-            p_184548_1_.func_195064_c(effectinstance1);
+            p_184548_1_.addEffect(effectinstance1, Cause.ARROW); // CraftBukkit
          }
       }
 
