--- a/net/minecraft/entity/projectile/DamagingProjectileEntity.java
+++ b/net/minecraft/entity/projectile/DamagingProjectileEntity.java
@@ -16,11 +_,14 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public abstract class DamagingProjectileEntity extends ProjectileEntity {
    public double field_70232_b;
    public double field_70233_c;
    public double field_70230_d;
+   public float bukkitYield = 1; // CraftBukkit
+   public boolean isIncendiary = true; // CraftBukkit
 
    protected DamagingProjectileEntity(EntityType<? extends DamagingProjectileEntity> p_i50173_1_, World p_i50173_2_) {
       super(p_i50173_1_, p_i50173_2_);
@@ -37,7 +_,22 @@
          this.field_70230_d = p_i50174_12_ / d0 * 0.1D;
       }
 
-   }
+      // CraftBukkit start - Added setDirection method
+      this.setDirection(p_i50174_8_, p_i50174_10_, p_i50174_12_);
+   }
+
+   // CraftBukkit - start
+   public void setDirection(double x, double y, double z) {
+      // CraftBukkit end
+      x += this.field_70146_Z.nextGaussian() * 0.4;
+      y += this.field_70146_Z.nextGaussian() * 0.4;
+      z += this.field_70146_Z.nextGaussian() * 0.4;
+      final double d3 = MathHelper.func_76133_a(x * x + y * y + z * z);
+      this.field_70232_b = x / d3 * 0.1;
+      this.field_70233_c = y / d3 * 0.1;
+      this.field_70230_d = z / d3 * 0.1;
+   }
+   // CraftBukkit - end
 
    public DamagingProjectileEntity(EntityType<? extends DamagingProjectileEntity> p_i50175_1_, LivingEntity p_i50175_2_, double p_i50175_3_, double p_i50175_5_, double p_i50175_7_, World p_i50175_9_) {
       this(p_i50175_1_, p_i50175_2_.func_226277_ct_(), p_i50175_2_.func_226278_cu_(), p_i50175_2_.func_226281_cx_(), p_i50175_3_, p_i50175_5_, p_i50175_7_, p_i50175_9_);
@@ -68,8 +_,14 @@
          }
 
          RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_);
-         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS) {
-            this.func_70227_a(raytraceresult);
+         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
+            this.preOnHit(raytraceresult);
+
+            // CraftBukkit start - Fire ProjectileHitEvent
+            if (this.field_70128_L) {
+               CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
+            }
+            // CraftBukkit end
          }
 
          this.func_145775_I();
@@ -145,6 +_,11 @@
          this.func_70018_K();
          Entity entity = p_70097_1_.func_76346_g();
          if (entity != null) {
+            // CraftBukkit start
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             Vector3d vector3d = entity.func_70040_Z();
             this.func_213317_d(vector3d);
             this.field_70232_b = vector3d.field_72450_a * 0.1D;
