--- a/net/minecraft/entity/projectile/EggEntity.java
+++ b/net/minecraft/entity/projectile/EggEntity.java
@@ -1,8 +_,11 @@
 package net.minecraft.entity.projectile;
 
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.passive.ChickenEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.item.Item;
 import net.minecraft.item.Items;
 import net.minecraft.particles.ItemParticleData;
@@ -11,8 +_,15 @@
 import net.minecraft.util.math.EntityRayTraceResult;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
+import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Location;
+import org.bukkit.entity.Ageable;
+import org.bukkit.entity.Egg;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
+import org.bukkit.event.player.PlayerEggThrowEvent;
 
 public class EggEntity extends ProjectileItemEntity {
    public EggEntity(EntityType<? extends EggEntity> p_i50154_1_, World p_i50154_2_) {
@@ -46,19 +_,51 @@
 
    protected void func_70227_a(RayTraceResult p_70227_1_) {
       super.func_70227_a(p_70227_1_);
+
+      int dummy1 = 0; int dummy2 = 0; ChickenEntity hatchingEntity;
+      dummy1++; // do not allow the compiler to cut these fields off
+      dummy2++;
+
+
+
       if (!this.field_70170_p.field_72995_K) {
-         if (this.field_70146_Z.nextInt(8) == 0) {
-            int i = 1;
+         boolean hatching = this.field_70146_Z.nextInt(8) == 0; // CraftBukkit
+         if (true) {
+            byte i = 1;
             if (this.field_70146_Z.nextInt(32) == 0) {
                i = 4;
             }
 
-            for(int j = 0; j < i; ++j) {
-               ChickenEntity chickenentity = EntityType.field_200795_i.func_200721_a(this.field_70170_p);
-               chickenentity.func_70873_a(-24000);
-               chickenentity.func_70012_b(this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), this.field_70177_z, 0.0F);
-               this.field_70170_p.func_217376_c(chickenentity);
-            }
+            // CraftBukkit start
+            if (!hatching) {
+               i = 0;
+            }
+            org.bukkit.entity.EntityType hatchingType = org.bukkit.entity.EntityType.CHICKEN;
+            Entity thrower = this.func_234616_v_();
+            if (thrower instanceof ServerPlayerEntity) {
+               PlayerEggThrowEvent event = new PlayerEggThrowEvent((Player) thrower.getBukkitEntity(), (Egg) this.getBukkitEntity(), hatching, i, hatchingType);
+               this.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+
+               i = event.getNumHatches();
+               hatching = event.isHatching();
+               hatchingType = event.getHatchingType();
+            }
+
+            if (hatching) {
+               for (int j = 0; j < i; j++) {
+                  hatchingEntity = (ChickenEntity) field_70170_p.getWorld().createEntity(new Location(field_70170_p.getWorld(), this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), this.field_70177_z, 0), hatchingType.getEntityClass());
+                  if (hatchingEntity.getBukkitEntity() instanceof Ageable) {
+                     ((Ageable) hatchingEntity.getBukkitEntity()).setBaby();
+                  }
+                  // Fixed Peukui Startup
+                  if (this.field_70170_p instanceof ServerWorld){
+                     ((ServerWorld) this.field_70170_p).pehkuiSpawnReason = SpawnReason.EGG;
+                     field_70170_p.func_217376_c(hatchingEntity);
+                     ((ServerWorld) this.field_70170_p).pehkuiSpawnReason = SpawnReason.DEFAULT;
+                  }
+               }
+            }
+            // CraftBukkit end
          }
 
          this.field_70170_p.func_72960_a(this, (byte)3);
