--- a/net/minecraft/entity/projectile/EvokerFangsEntity.java
+++ b/net/minecraft/entity/projectile/EvokerFangsEntity.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class EvokerFangsEntity extends Entity {
    private int field_190553_a;
@@ -92,7 +_,9 @@
       } else if (--this.field_190553_a < 0) {
          if (this.field_190553_a == -8) {
             for(LivingEntity livingentity : this.field_70170_p.func_217357_a(LivingEntity.class, this.func_174813_aQ().func_72314_b(0.2D, 0.0D, 0.2D))) {
+               CraftEventFactory.entityDamage = this; // CraftBukkit
                this.func_190551_c(livingentity);
+               CraftEventFactory.entityDamage = null; // CraftBukkit
             }
          }
 
