--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -30,17 +_,18 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 @OnlyIn(
    value = Dist.CLIENT,
    _interface = IRendersAsItem.class
 )
 public class FireworkRocketEntity extends ProjectileEntity implements IRendersAsItem {
-   private static final DataParameter<ItemStack> field_184566_a = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187196_f);
+   public static final DataParameter<ItemStack> field_184566_a = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187196_f);
    private static final DataParameter<OptionalInt> field_191512_b = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_218814_r);
-   private static final DataParameter<Boolean> field_213895_d = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187198_h);
+   public static final DataParameter<Boolean> field_213895_d = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187198_h);
    private int field_92056_a;
-   private int field_92055_b;
+   public int field_92055_b;
    private LivingEntity field_191513_e;
 
    public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> p_i50164_1_, World p_i50164_2_) {
@@ -82,6 +_,21 @@
       this.func_212361_a(p_i231582_3_);
    }
 
+   // Spigot Start - copied from tick
+   @Override
+   public void inactiveTick() {
+      this.field_92056_a += 1;
+      if (!this.field_70170_p.field_72995_K && this.field_92056_a > this.field_92055_b) {
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.func_191510_k();
+         }
+         // CraftBukkit end
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected void func_70088_a() {
       this.field_70180_af.func_187214_a(field_184566_a, ItemStack.field_190927_a);
       this.field_70180_af.func_187214_a(field_191512_b, OptionalInt.empty());
@@ -136,7 +_,7 @@
 
       RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_);
       if (!this.field_70145_X) {
-         this.func_70227_a(raytraceresult);
+         this.preOnHit(raytraceresult);
          this.field_70160_al = true;
       }
 
@@ -151,9 +_,19 @@
       }
 
       if (!this.field_70170_p.field_72995_K && this.field_92056_a > this.field_92055_b) {
-         this.func_213893_k();
+         // CraftBukkit start
+         if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.func_213893_k();
+         }
       }
+         // CraftBukkit end
+   }
 
+   @Override
+   protected void func_70227_a(RayTraceResult result) {
+      if (result.func_216346_c() == RayTraceResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.func_70227_a(result);
+      }
    }
 
    private void func_213893_k() {
@@ -165,7 +_,11 @@
    protected void func_213868_a(EntityRayTraceResult p_213868_1_) {
       super.func_213868_a(p_213868_1_);
       if (!this.field_70170_p.field_72995_K) {
-         this.func_213893_k();
+         // CraftBukkit start
+         if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.func_213893_k();
+         }
+         // CraftBukkit end
       }
    }
 
@@ -173,7 +_,11 @@
       BlockPos blockpos = new BlockPos(p_230299_1_.func_216350_a());
       this.field_70170_p.func_180495_p(blockpos).func_196950_a(this.field_70170_p, blockpos, this);
       if (!this.field_70170_p.func_201670_d() && this.func_213894_l()) {
-         this.func_213893_k();
+         // CraftBukkit start
+         if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.func_213893_k();
+         }
+         // CraftBukkit end
       }
 
       super.func_230299_a_(p_230299_1_);
@@ -197,7 +_,9 @@
 
       if (f > 0.0F) {
          if (this.field_191513_e != null) {
+            CraftEventFactory.entityDamage = this; // CraftBukkit
             this.field_191513_e.func_70097_a(DamageSource.func_233548_a_(this, this.func_234616_v_()), 5.0F + (float)(listnbt.size() * 2));
+            CraftEventFactory.entityDamage = null; // CraftBukkit
          }
 
          double d0 = 5.0D;
@@ -218,7 +_,9 @@
 
                if (flag) {
                   float f1 = f * (float)Math.sqrt((5.0D - (double)this.func_70032_d(livingentity)) / 5.0D);
+                  CraftEventFactory.entityDamage = this; // CraftBukkit
                   livingentity.func_70097_a(DamageSource.func_233548_a_(this, this.func_234616_v_()), f1);
+                  CraftEventFactory.entityDamage = null; // CraftBukkit
                }
             }
          }
