--- a/net/minecraft/entity/projectile/ProjectileEntity.java
+++ b/net/minecraft/entity/projectile/ProjectileEntity.java
@@ -2,6 +_,8 @@
 
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import com.ibm.icu.text.RawCollationKey;
 import net.minecraft.block.BlockState;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
@@ -15,12 +_,17 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.projectiles.ProjectileSource;
 
 public abstract class ProjectileEntity extends Entity {
    private UUID field_234609_b_;
    private int field_234610_c_;
    private boolean field_234611_d_;
 
+   // CraftBukkit Start
+   private boolean hitCancelled = false;
+   // CraftBukkit end
+
    ProjectileEntity(EntityType<? extends ProjectileEntity> p_i231584_1_, World p_i231584_2_) {
       super(p_i231584_1_, p_i231584_2_);
    }
@@ -29,7 +_,9 @@
       if (p_212361_1_ != null) {
          this.field_234609_b_ = p_212361_1_.func_110124_au();
          this.field_234610_c_ = p_212361_1_.func_145782_y();
+
       }
+      this.projectileSource = (p_212361_1_ != null && p_212361_1_.getBukkitEntity() instanceof ProjectileSource) ? (ProjectileSource) p_212361_1_.getBukkitEntity() : null; // CraftBukkit
 
    }
 
@@ -103,7 +_,20 @@
       this.func_213317_d(this.func_213322_ci().func_72441_c(vector3d.field_72450_a, p_234612_1_.func_233570_aj_() ? 0.0D : vector3d.field_72448_b, vector3d.field_72449_c));
    }
 
+   // CraftBukkit Start - call projectile hit event
+   protected void preOnHit(RayTraceResult movingobjectposition)
+   {
+      org.bukkit.event.entity.ProjectileHitEvent event = org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition);
+      this.hitCancelled = event != null && event.isCancelled();
+      if(movingobjectposition.func_216346_c() == RayTraceResult.Type.BLOCK || !this.hitCancelled)
+      {
+         this.func_70227_a(movingobjectposition); // PAIL rename onHit
+      }
+   }
+
+
    protected void func_70227_a(RayTraceResult p_70227_1_) {
+      org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callProjectileHitEvent(this, p_70227_1_); // CraftBukkit - Call event
       RayTraceResult.Type raytraceresult$type = p_70227_1_.func_216346_c();
       if (raytraceresult$type == RayTraceResult.Type.ENTITY) {
          this.func_213868_a((EntityRayTraceResult)p_70227_1_);
