--- a/net/minecraft/entity/projectile/ProjectileItemEntity.java
+++ b/net/minecraft/entity/projectile/ProjectileItemEntity.java
@@ -36,7 +_,9 @@
    public void func_213884_b(ItemStack p_213884_1_) {
       if (p_213884_1_.func_77973_b() != this.func_213885_i() || p_213884_1_.func_77942_o()) {
          this.func_184212_Q().func_187227_b(field_213886_e, Util.func_200696_a(p_213884_1_.func_77946_l(), (p_213883_0_) -> {
-            p_213883_0_.func_190920_e(1);
+            if (!p_213883_0_.func_190926_b()) {
+               p_213883_0_.func_190920_e(1); // CraftBukkit
+            }
          }));
       }
 
@@ -44,7 +_,13 @@
 
    protected abstract Item func_213885_i();
 
-   protected ItemStack func_213882_k() {
+   // CraftBukkit start
+   public Item getDefaultItemPublic() {
+      return func_213885_i();
+   }
+   // CraftBukkit end
+
+   public ItemStack func_213882_k() {
       return this.func_184212_Q().func_187225_a(field_213886_e);
    }
 
