--- a/net/minecraft/entity/projectile/SmallFireballEntity.java
+++ b/net/minecraft/entity/projectile/SmallFireballEntity.java
@@ -12,6 +_,7 @@
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
 
 public class SmallFireballEntity extends AbstractFireballEntity {
    public SmallFireballEntity(EntityType<? extends SmallFireballEntity> p_i50160_1_, World p_i50160_2_) {
@@ -20,6 +_,11 @@
 
    public SmallFireballEntity(World p_i1771_1_, LivingEntity p_i1771_2_, double p_i1771_3_, double p_i1771_5_, double p_i1771_7_) {
       super(EntityType.field_200744_aj, p_i1771_2_, p_i1771_3_, p_i1771_5_, p_i1771_7_, p_i1771_1_);
+      // CraftBukkit start
+      if (this.func_234616_v_() != null && this.func_234616_v_() instanceof MobEntity) {
+         isIncendiary = this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b);
+      }
+      // CraftBukkit end
    }
 
    public SmallFireballEntity(World p_i1772_1_, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_) {
@@ -33,7 +_,15 @@
          if (!entity.func_230279_az_()) {
             Entity entity1 = this.func_234616_v_();
             int i = entity.func_223314_ad();
-            entity.func_70015_d(5);
+            // CraftBukkit start - Entity damage by entity event + combust event
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+            entity.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               entity.putCallCombustEvent(false);
+               entity.func_70015_d(event.getDuration());
+               entity.putCallCombustEvent(true);
+            }
+            // CraftBukkit end
             boolean flag = entity.func_70097_a(DamageSource.func_233547_a_(this, entity1), 5.0F);
             if (!flag) {
                entity.func_241209_g_(i);
@@ -49,9 +_,9 @@
       super.func_230299_a_(p_230299_1_);
       if (!this.field_70170_p.field_72995_K) {
          Entity entity = this.func_234616_v_();
-         if (entity == null || !(entity instanceof MobEntity) || this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+         if (isIncendiary) { // CraftBukkit
             BlockPos blockpos = p_230299_1_.func_216350_a().func_177972_a(p_230299_1_.func_216354_b());
-            if (this.field_70170_p.func_175623_d(blockpos)) {
+            if (this.field_70170_p.func_175623_d(blockpos) && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(field_70170_p, blockpos, this).isCancelled()) { // CraftBukkit
                this.field_70170_p.func_175656_a(blockpos, AbstractFireBlock.func_235326_a_(this.field_70170_p, blockpos));
             }
          }
