--- a/net/minecraft/entity/projectile/SpectralArrowEntity.java
+++ b/net/minecraft/entity/projectile/SpectralArrowEntity.java
@@ -9,9 +_,10 @@
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.potion.Effects;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class SpectralArrowEntity extends AbstractArrowEntity {
-   private int field_184562_f = 200;
+   public int field_184562_f = 200;
 
    public SpectralArrowEntity(EntityType<? extends SpectralArrowEntity> p_i50158_1_, World p_i50158_2_) {
       super(p_i50158_1_, p_i50158_2_);
@@ -40,7 +_,7 @@
    protected void func_184548_a(LivingEntity p_184548_1_) {
       super.func_184548_a(p_184548_1_);
       EffectInstance effectinstance = new EffectInstance(Effects.field_188423_x, this.field_184562_f, 0);
-      p_184548_1_.func_195064_c(effectinstance);
+      p_184548_1_.addEffect(effectinstance, Cause.ARROW); // CraftBukkit
    }
 
    public void func_70037_a(CompoundNBT p_70037_1_) {
