--- a/net/minecraft/entity/projectile/TridentEntity.java
+++ b/net/minecraft/entity/projectile/TridentEntity.java
@@ -24,11 +_,12 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.weather.LightningStrikeEvent.Cause;
 
 public class TridentEntity extends AbstractArrowEntity {
    private static final DataParameter<Byte> field_203053_g = EntityDataManager.func_187226_a(TridentEntity.class, DataSerializers.field_187191_a);
    private static final DataParameter<Boolean> field_226571_aq_ = EntityDataManager.func_187226_a(TridentEntity.class, DataSerializers.field_187198_h);
-   private ItemStack field_203054_h = new ItemStack(Items.field_203184_eO);
+   public ItemStack field_203054_h = new ItemStack(Items.field_203184_eO);
    private boolean field_203051_au;
    public int field_203052_f;
 
@@ -148,7 +_,7 @@
             LightningBoltEntity lightningboltentity = EntityType.field_200728_aG.func_200721_a(this.field_70170_p);
             lightningboltentity.func_233576_c_(Vector3d.func_237492_c_(blockpos));
             lightningboltentity.func_204809_d(entity1 instanceof ServerPlayerEntity ? (ServerPlayerEntity)entity1 : null);
-            this.field_70170_p.func_217376_c(lightningboltentity);
+            ((ServerWorld)this.field_70170_p).strikeLightning(lightningboltentity, Cause.TRIDENT); // CraftBukkit
             soundevent = SoundEvents.field_203275_iq;
             f1 = 5.0F;
          }
