--- a/net/minecraft/world/server/ServerBossInfo.java
+++ b/net/minecraft/world/server/ServerBossInfo.java
@@ -15,7 +_,7 @@
 public class ServerBossInfo extends BossInfo {
    private final Set<ServerPlayerEntity> field_186762_h = Sets.newHashSet();
    private final Set<ServerPlayerEntity> field_186763_i = Collections.unmodifiableSet(this.field_186762_h);
-   private boolean field_186764_j = true;
+   public boolean field_186764_j = true;
 
    public ServerBossInfo(ITextComponent p_i46839_1_, BossInfo.Color p_i46839_2_, BossInfo.Overlay p_i46839_3_) {
       super(MathHelper.func_188210_a(), p_i46839_1_, p_i46839_2_, p_i46839_3_);
@@ -80,7 +_,7 @@
 
    }
 
-   private void func_186759_a(SUpdateBossInfoPacket.Operation p_186759_1_) {
+   public void func_186759_a(SUpdateBossInfoPacket.Operation p_186759_1_) {
       if (this.field_186764_j) {
          SUpdateBossInfoPacket supdatebossinfopacket = new SUpdateBossInfoPacket(p_186759_1_, this);
 
