--- a/net/minecraft/entity/player/PlayerAbilities.java
+++ b/net/minecraft/entity/player/PlayerAbilities.java
@@ -10,8 +_,8 @@
    public boolean field_75101_c;
    public boolean field_75098_d;
    public boolean field_75099_e = true;
-   private float field_75096_f = 0.05F;
-   private float field_75097_g = 0.1F;
+   public float field_75096_f = 0.05F;
+   public float field_75097_g = 0.1F;
 
    public void func_75091_a(CompoundNBT p_75091_1_) {
       CompoundNBT compoundnbt = new CompoundNBT();
