--- ../src-base/minecraft/net/minecraftforge/fml/common/asm/transformers/TerminalTransformer.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/asm/transformers/TerminalTransformer.java
@@ -61,16 +61,7 @@
         @Override
         public MethodVisitor visitMethod(int mAccess, final String mName, final String mDesc, String mSignature, String[] mExceptions)
         {
-            final boolean warn = !(clsName.equals("net/minecraft/client/Minecraft") ||
-                                   clsName.equals("net/minecraft/server/dedicated/DedicatedServer") ||
-                                   clsName.equals("net/minecraft/server/dedicated/ServerHangWatchdog") ||
-                                   clsName.equals("net/minecraft/server/dedicated/ServerHangWatchdog$1") ||
-                                   clsName.equals("net/minecraftforge/fml/common/FMLCommonHandler") ||
-                                   clsName.startsWith("com/jcraft/jogg/") ||
-                                   clsName.startsWith("scala/sys/") ||
-                                   clsName.startsWith("net/minecraft/server/gui/MinecraftServerGui") ||
-                                   clsName.startsWith("com/sun/jna/")
-                                   );
+            final boolean warn = false; //CatServer - Disable warning
 
             return new MethodVisitor(Opcodes.ASM5, super.visitMethod(mAccess, mName, mDesc, mSignature, mExceptions))
             {
