--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -16,6 +_,8 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
 import net.minecraft.world.BossEvent;
+import org.bukkit.boss.KeyedBossBar;
+import org.bukkit.craftbukkit.v1_18_R2.boss.CraftKeyedBossbar;
 
 public class CustomBossEvent extends ServerBossEvent {
    private final ResourceLocation f_136256_;
@@ -178,4 +_,15 @@
    public void m_136286_(ServerPlayer p_136287_) {
       super.m_6539_(p_136287_);
    }
+
+   // CraftBukkit start
+   private KeyedBossBar bossBar;
+
+   public KeyedBossBar getBukkitEntity() {
+      if (bossBar == null) {
+         bossBar = new CraftKeyedBossbar(this);
+      }
+      return bossBar;
+   }
+   // CraftBukkit end
 }
