--- a/net/minecraft/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/item/crafting/CampfireCookingRecipe.java
@@ -3,6 +_,11 @@
 import net.minecraft.block.Blocks;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftCampfireRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
 
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
    public CampfireCookingRecipe(ResourceLocation p_i50030_1_, String p_i50030_2_, Ingredient p_i50030_3_, ItemStack p_i50030_4_, float p_i50030_5_, int p_i50030_6_) {
@@ -16,4 +_,14 @@
    public IRecipeSerializer<?> func_199559_b() {
       return IRecipeSerializer.field_222174_r;
    }
+
+   // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftCopy(CraftItemStack.asBukkitCopy(this.func_77571_b()));
+      CraftCampfireRecipe recipe = new CraftCampfireRecipe(CraftNamespacedKey.fromMinecraft(this.field_222140_b), result, CraftRecipe.toBukkit(this.field_222142_d), this.field_222144_f, this.field_222145_g);
+      recipe.setGroup(this.field_222141_c);
+      return recipe;
+    }
+   // CraftBukkit end
 }
