--- a/net/minecraft/item/crafting/FurnaceRecipe.java
+++ b/net/minecraft/item/crafting/FurnaceRecipe.java
@@ -3,6 +_,11 @@
 import net.minecraft.block.Blocks;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftFurnaceRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
 
 public class FurnaceRecipe extends AbstractCookingRecipe {
    public FurnaceRecipe(ResourceLocation p_i48715_1_, String p_i48715_2_, Ingredient p_i48715_3_, ItemStack p_i48715_4_, float p_i48715_5_, int p_i48715_6_) {
@@ -16,4 +_,14 @@
    public IRecipeSerializer<?> func_199559_b() {
       return IRecipeSerializer.field_222171_o;
    }
+
+   // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftCopy(CraftItemStack.asBukkitCopy(this.func_77571_b()));
+      CraftFurnaceRecipe recipe = new CraftFurnaceRecipe(CraftNamespacedKey.fromMinecraft(this.field_222140_b), result, CraftRecipe.toBukkit(this.field_222142_d), this.field_222144_f, this.field_222145_g);
+      recipe.setGroup(this.field_222141_c);
+      return recipe;
+    }
+   // CraftBukkit end
 }
