--- a/net/minecraft/item/crafting/SpecialRecipe.java
+++ b/net/minecraft/item/crafting/SpecialRecipe.java
@@ -2,6 +_,8 @@
 
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftComplexRecipe;
+import org.bukkit.inventory.Recipe;
 
 public abstract class SpecialRecipe implements ICraftingRecipe {
    private final ResourceLocation field_199563_a;
@@ -14,11 +_,18 @@
       return this.field_199563_a;
    }
 
-   public boolean func_192399_d() {
+    public boolean func_192399_d() {
       return true;
    }
 
    public ItemStack func_77571_b() {
       return ItemStack.field_190927_a;
    }
+
+   // CraftBukkit start
+   @Override
+   public Recipe toBukkitRecipe() {
+      return new CraftComplexRecipe(this);
+   }
+   // CraftBukkit end
 }
