--- a/net/minecraft/item/crafting/StonecuttingRecipe.java
+++ b/net/minecraft/item/crafting/StonecuttingRecipe.java
@@ -5,6 +_,11 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftStonecuttingRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
 
 public class StonecuttingRecipe extends SingleItemRecipe {
    public StonecuttingRecipe(ResourceLocation p_i50021_1_, String p_i50021_2_, Ingredient p_i50021_3_, ItemStack p_i50021_4_) {
@@ -18,4 +_,14 @@
    public ItemStack func_222128_h() {
       return new ItemStack(Blocks.field_222430_lS);
    }
+
+    // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+        CraftItemStack result = CraftItemStack.asCraftCopy(CraftItemStack.asBukkitCopy(this.func_77571_b()));
+        CraftStonecuttingRecipe recipe = new CraftStonecuttingRecipe(CraftNamespacedKey.fromMinecraft(this.field_222133_c), result, CraftRecipe.toBukkit(this.field_222131_a));
+        recipe.setGroup(this.field_222134_d);
+        return recipe;
+    }
+    // CraftBukkit end
 }
