--- ../src-base/minecraft/net/minecraft/server/ServerEula.java
+++ ../src-work/minecraft/net/minecraft/server/ServerEula.java
@@ -22,7 +22,7 @@
     public ServerEula(File eulaFile)
     {
         this.eulaFile = eulaFile;
-        this.acceptedEULA = this.loadEULAFile(eulaFile);
+        this.acceptedEULA = true || this.loadEULAFile(eulaFile); // CatServer - auto accept EULA
     }
 
     private boolean loadEULAFile(File inFile)
