--- ../src-base/minecraft/net/minecraft/world/gen/structure/ComponentScatteredFeaturePieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/ComponentScatteredFeaturePieces.java
@@ -38,13 +38,13 @@
 {
     public static void registerScatteredFeaturePieces()
     {
-        MapGenStructureIO.registerStructureComponent(ComponentScatteredFeaturePieces.DesertPyramid.class, "TeDP");
-        MapGenStructureIO.registerStructureComponent(ComponentScatteredFeaturePieces.JunglePyramid.class, "TeJP");
-        MapGenStructureIO.registerStructureComponent(ComponentScatteredFeaturePieces.SwampHut.class, "TeSH");
-        MapGenStructureIO.registerStructureComponent(ComponentScatteredFeaturePieces.Igloo.class, "Iglu");
+        MapGenStructureIO.registerStructureComponent(DesertPyramid.class, "TeDP");
+        MapGenStructureIO.registerStructureComponent(JunglePyramid.class, "TeJP");
+        MapGenStructureIO.registerStructureComponent(SwampHut.class, "TeSH");
+        MapGenStructureIO.registerStructureComponent(Igloo.class, "Iglu");
     }
 
-    public static class DesertPyramid extends ComponentScatteredFeaturePieces.Feature
+    public static class DesertPyramid extends Feature
         {
             private final boolean[] hasPlacedChest = new boolean[4];
 
@@ -363,7 +363,7 @@
             }
         }
 
-    public static class Igloo extends ComponentScatteredFeaturePieces.Feature
+    public static class Igloo extends Feature
         {
             private static final ResourceLocation IGLOO_TOP_ID = new ResourceLocation("igloo/igloo_top");
             private static final ResourceLocation IGLOO_MIDDLE_ID = new ResourceLocation("igloo/igloo_middle");
@@ -437,13 +437,13 @@
             }
         }
 
-    public static class JunglePyramid extends ComponentScatteredFeaturePieces.Feature
+    public static class JunglePyramid extends Feature
         {
             private boolean placedMainChest;
             private boolean placedHiddenChest;
             private boolean placedTrap1;
             private boolean placedTrap2;
-            private static final ComponentScatteredFeaturePieces.JunglePyramid.Stones cobblestoneSelector = new ComponentScatteredFeaturePieces.JunglePyramid.Stones();
+            private static final Stones cobblestoneSelector = new Stones();
 
             public JunglePyramid()
             {
@@ -669,7 +669,7 @@
                 }
             }
 
-            static class Stones extends StructureComponent.BlockSelector
+            static class Stones extends BlockSelector
                 {
                     private Stones()
                     {
@@ -689,7 +689,7 @@
                 }
         }
 
-    public static class SwampHut extends ComponentScatteredFeaturePieces.Feature
+    public static class SwampHut extends Feature
         {
             private boolean hasWitch;
 
@@ -773,7 +773,7 @@
                             entitywitch.enablePersistence();
                             entitywitch.setLocationAndAngles((double)l + 0.5D, (double)i1, (double)k + 0.5D, 0.0F, 0.0F);
                             entitywitch.onInitialSpawn(worldIn.getDifficultyForLocation(new BlockPos(l, i1, k)), (IEntityLivingData)null);
-                            worldIn.spawnEntity(entitywitch);
+                            worldIn.addEntity(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN);
                         }
                     }
 
