--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenNetherBridge.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenNetherBridge.java
@@ -58,7 +58,7 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        return new MapGenNetherBridge.Start(this.world, this.rand, chunkX, chunkZ);
+        return new Start(this.world, this.rand, chunkX, chunkZ);
     }
 
     public BlockPos getNearestStructurePos(World worldIn, BlockPos pos, boolean findUnexplored)
