--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenStructure.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenStructure.java
@@ -221,7 +221,11 @@
     {
         if (this.structureData == null)
         {
-            this.structureData = (MapGenStructureData)worldIn.getPerWorldStorage().getOrLoadData(MapGenStructureData.class, this.getStructureName());
+            if(worldIn.spigotConfig.saveStructureInfo && !this.getStructureName().contains("Minecraft")){
+                this.structureData = (MapGenStructureData)worldIn.loadItemData(MapGenStructureData.class,this.getStructureName());
+            } else {
+                this.structureData = new MapGenStructureData(this.getStructureName());
+            }
 
             if (this.structureData == null)
             {
