--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
@@ -63,7 +63,7 @@
 
         int k = chunkX / this.maxDistanceBetweenScatteredFeatures;
         int l = chunkZ / this.maxDistanceBetweenScatteredFeatures;
-        Random random = this.world.setRandomSeed(k, l, 14357617);
+        Random random = this.world.setRandomSeed(k, l, this.world.spigotConfig.largeFeatureSeed); // Spigot
         k = k * this.maxDistanceBetweenScatteredFeatures;
         l = l * this.maxDistanceBetweenScatteredFeatures;
         k = k + random.nextInt(this.maxDistanceBetweenScatteredFeatures - 8);
@@ -98,14 +98,14 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        return new MapGenScatteredFeature.Start(this.world, this.rand, chunkX, chunkZ);
+        return new Start(this.world, this.rand, chunkX, chunkZ);
     }
 
     public boolean isSwampHut(BlockPos pos)
     {
         StructureStart structurestart = this.getStructureAt(pos);
 
-        if (structurestart != null && structurestart instanceof MapGenScatteredFeature.Start && !structurestart.components.isEmpty())
+        if (structurestart != null && structurestart instanceof Start && !structurestart.components.isEmpty())
         {
             StructureComponent structurecomponent = structurestart.components.get(0);
             return structurecomponent instanceof ComponentScatteredFeaturePieces.SwampHut;
