--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenMineshaft.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenMineshaft.java
@@ -75,7 +75,7 @@
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
         Biome biome = this.world.getBiome(new BlockPos((chunkX << 4) + 8, 64, (chunkZ << 4) + 8));
-        MapGenMineshaft.Type mapgenmineshaft$type = biome instanceof BiomeMesa ? MapGenMineshaft.Type.MESA : MapGenMineshaft.Type.NORMAL;
+        Type mapgenmineshaft$type = biome instanceof BiomeMesa ? Type.MESA : Type.NORMAL;
         return new StructureMineshaftStart(this.world, this.rand, chunkX, chunkZ, mapgenmineshaft$type);
     }
 
@@ -84,7 +84,7 @@
         NORMAL,
         MESA;
 
-        public static MapGenMineshaft.Type byId(int id)
+        public static Type byId(int id)
         {
             return id >= 0 && id < values().length ? values()[id] : NORMAL;
         }
