--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -29,11 +29,19 @@
 
         for (Biome biome : Biome.REGISTRY)
         {
-            if (biome != null && biome.getBaseHeight() > 0.0F)
+            if (biome != null && biome.getBaseHeight() > 0.0F && !net.minecraftforge.common.BiomeManager.strongHoldBiomesBlackList.contains(biome))
             {
                 this.allowedBiomes.add(biome);
             }
         }
+
+        for (Biome biome : net.minecraftforge.common.BiomeManager.strongHoldBiomes)
+        {
+            if (!this.allowedBiomes.contains(biome))
+            {
+                this.allowedBiomes.add(biome);
+            }
+        }
     }
 
     public MapGenStronghold(Map<String, String> p_i2068_1_)
@@ -173,9 +181,9 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        MapGenStronghold.Start mapgenstronghold$start;
+        Start mapgenstronghold$start;
 
-        for (mapgenstronghold$start = new MapGenStronghold.Start(this.world, this.rand, chunkX, chunkZ); mapgenstronghold$start.getComponents().isEmpty() || ((StructureStrongholdPieces.Stairs2)mapgenstronghold$start.getComponents().get(0)).strongholdPortalRoom == null; mapgenstronghold$start = new MapGenStronghold.Start(this.world, this.rand, chunkX, chunkZ))
+        for (mapgenstronghold$start = new Start(this.world, this.rand, chunkX, chunkZ); mapgenstronghold$start.getComponents().isEmpty() || ((StructureStrongholdPieces.Stairs2)mapgenstronghold$start.getComponents().get(0)).strongholdPortalRoom == null; mapgenstronghold$start = new Start(this.world, this.rand, chunkX, chunkZ))
         {
             ;
         }
