--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenEndCity.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenEndCity.java
@@ -59,7 +59,7 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        return new MapGenEndCity.Start(this.world, this.endProvider, this.rand, chunkX, chunkZ);
+        return new Start(this.world, this.endProvider, this.rand, chunkX, chunkZ);
     }
 
     public BlockPos getNearestStructurePos(World worldIn, BlockPos pos, boolean findUnexplored)
