--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
@@ -63,7 +63,7 @@
 
         int k = chunkX / this.maxDistanceBetweenScatteredFeatures;
         int l = chunkZ / this.maxDistanceBetweenScatteredFeatures;
-        Random random = this.worldObj.setRandomSeed(k, l, 14357617);
+        Random random = this.worldObj.setRandomSeed(k, l, this.worldObj.spigotConfig.largeFeatureSeed); // Spigot
         k = k * this.maxDistanceBetweenScatteredFeatures;
         l = l * this.maxDistanceBetweenScatteredFeatures;
         k = k + random.nextInt(this.maxDistanceBetweenScatteredFeatures - 8);
