--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureOceanMonumentPieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureOceanMonumentPieces.java
@@ -21,34 +21,34 @@
 {
     public static void registerOceanMonumentPieces()
     {
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.MonumentBuilding.class, "OMB");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.MonumentCoreRoom.class, "OMCR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.DoubleXRoom.class, "OMDXR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.DoubleXYRoom.class, "OMDXYR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.DoubleYRoom.class, "OMDYR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.DoubleYZRoom.class, "OMDYZR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.DoubleZRoom.class, "OMDZR");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.EntryRoom.class, "OMEntry");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.Penthouse.class, "OMPenthouse");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.SimpleRoom.class, "OMSimple");
-        MapGenStructureIO.registerStructureComponent(StructureOceanMonumentPieces.SimpleTopRoom.class, "OMSimpleT");
+        MapGenStructureIO.registerStructureComponent(MonumentBuilding.class, "OMB");
+        MapGenStructureIO.registerStructureComponent(MonumentCoreRoom.class, "OMCR");
+        MapGenStructureIO.registerStructureComponent(DoubleXRoom.class, "OMDXR");
+        MapGenStructureIO.registerStructureComponent(DoubleXYRoom.class, "OMDXYR");
+        MapGenStructureIO.registerStructureComponent(DoubleYRoom.class, "OMDYR");
+        MapGenStructureIO.registerStructureComponent(DoubleYZRoom.class, "OMDYZR");
+        MapGenStructureIO.registerStructureComponent(DoubleZRoom.class, "OMDZR");
+        MapGenStructureIO.registerStructureComponent(EntryRoom.class, "OMEntry");
+        MapGenStructureIO.registerStructureComponent(Penthouse.class, "OMPenthouse");
+        MapGenStructureIO.registerStructureComponent(SimpleRoom.class, "OMSimple");
+        MapGenStructureIO.registerStructureComponent(SimpleTopRoom.class, "OMSimpleT");
     }
 
-    public static class DoubleXRoom extends StructureOceanMonumentPieces.Piece
+    public static class DoubleXRoom extends Piece
         {
             public DoubleXRoom()
             {
             }
 
-            public DoubleXRoom(EnumFacing p_i45597_1_, StructureOceanMonumentPieces.RoomDefinition p_i45597_2_, Random p_i45597_3_)
+            public DoubleXRoom(EnumFacing p_i45597_1_, RoomDefinition p_i45597_2_, Random p_i45597_3_)
             {
                 super(1, p_i45597_1_, p_i45597_2_, 2, 1, 1);
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
             {
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.EAST.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.EAST.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
 
                 if (this.roomDefinition.index / 25 > 0)
                 {
@@ -118,23 +118,23 @@
             }
         }
 
-    public static class DoubleXYRoom extends StructureOceanMonumentPieces.Piece
+    public static class DoubleXYRoom extends Piece
         {
             public DoubleXYRoom()
             {
             }
 
-            public DoubleXYRoom(EnumFacing p_i45596_1_, StructureOceanMonumentPieces.RoomDefinition p_i45596_2_, Random p_i45596_3_)
+            public DoubleXYRoom(EnumFacing p_i45596_1_, RoomDefinition p_i45596_2_, Random p_i45596_3_)
             {
                 super(1, p_i45596_1_, p_i45596_2_, 2, 2, 1);
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
             {
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.EAST.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition2 = structureoceanmonumentpieces$roomdefinition1.connections[EnumFacing.UP.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition3 = structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.EAST.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition2 = structureoceanmonumentpieces$roomdefinition1.connections[EnumFacing.UP.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition3 = structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()];
 
                 if (this.roomDefinition.index / 25 > 0)
                 {
@@ -255,13 +255,13 @@
             }
         }
 
-    public static class DoubleYRoom extends StructureOceanMonumentPieces.Piece
+    public static class DoubleYRoom extends Piece
         {
             public DoubleYRoom()
             {
             }
 
-            public DoubleYRoom(EnumFacing p_i45595_1_, StructureOceanMonumentPieces.RoomDefinition p_i45595_2_, Random p_i45595_3_)
+            public DoubleYRoom(EnumFacing p_i45595_1_, RoomDefinition p_i45595_2_, Random p_i45595_3_)
             {
                 super(1, p_i45595_1_, p_i45595_2_, 1, 2, 1);
             }
@@ -273,7 +273,7 @@
                     this.generateDefaultFloor(worldIn, structureBoundingBoxIn, 0, 0, this.roomDefinition.hasOpening[EnumFacing.DOWN.getIndex()]);
                 }
 
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.UP.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.UP.getIndex()];
 
                 if (structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()] == null)
                 {
@@ -292,7 +292,7 @@
                 this.fillWithBlocks(worldIn, structureBoundingBoxIn, 1, 4, 5, 1, 4, 5, BRICKS_PRISMARINE, BRICKS_PRISMARINE, false);
                 this.fillWithBlocks(worldIn, structureBoundingBoxIn, 5, 4, 5, 5, 4, 6, BRICKS_PRISMARINE, BRICKS_PRISMARINE, false);
                 this.fillWithBlocks(worldIn, structureBoundingBoxIn, 6, 4, 5, 6, 4, 5, BRICKS_PRISMARINE, BRICKS_PRISMARINE, false);
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
 
                 for (int i = 1; i <= 5; i += 4)
                 {
@@ -359,23 +359,23 @@
             }
         }
 
-    public static class DoubleYZRoom extends StructureOceanMonumentPieces.Piece
+    public static class DoubleYZRoom extends Piece
         {
             public DoubleYZRoom()
             {
             }
 
-            public DoubleYZRoom(EnumFacing p_i45594_1_, StructureOceanMonumentPieces.RoomDefinition p_i45594_2_, Random p_i45594_3_)
+            public DoubleYZRoom(EnumFacing p_i45594_1_, RoomDefinition p_i45594_2_, Random p_i45594_3_)
             {
                 super(1, p_i45594_1_, p_i45594_2_, 1, 2, 2);
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
             {
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.NORTH.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition2 = structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition3 = structureoceanmonumentpieces$roomdefinition1.connections[EnumFacing.UP.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.NORTH.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition2 = structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition3 = structureoceanmonumentpieces$roomdefinition1.connections[EnumFacing.UP.getIndex()];
 
                 if (this.roomDefinition.index / 25 > 0)
                 {
@@ -496,21 +496,21 @@
             }
         }
 
-    public static class DoubleZRoom extends StructureOceanMonumentPieces.Piece
+    public static class DoubleZRoom extends Piece
         {
             public DoubleZRoom()
             {
             }
 
-            public DoubleZRoom(EnumFacing p_i45593_1_, StructureOceanMonumentPieces.RoomDefinition p_i45593_2_, Random p_i45593_3_)
+            public DoubleZRoom(EnumFacing p_i45593_1_, RoomDefinition p_i45593_2_, Random p_i45593_3_)
             {
                 super(1, p_i45593_1_, p_i45593_2_, 1, 1, 2);
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
             {
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.NORTH.getIndex()];
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = this.roomDefinition.connections[EnumFacing.NORTH.getIndex()];
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = this.roomDefinition;
 
                 if (this.roomDefinition.index / 25 > 0)
                 {
@@ -599,13 +599,13 @@
             }
         }
 
-    public static class EntryRoom extends StructureOceanMonumentPieces.Piece
+    public static class EntryRoom extends Piece
         {
             public EntryRoom()
             {
             }
 
-            public EntryRoom(EnumFacing p_i45592_1_, StructureOceanMonumentPieces.RoomDefinition p_i45592_2_)
+            public EntryRoom(EnumFacing p_i45592_1_, RoomDefinition p_i45592_2_)
             {
                 super(1, p_i45592_1_, p_i45592_2_, 1, 1, 1);
             }
@@ -641,47 +641,47 @@
             }
         }
 
-    static class FitSimpleRoomHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class FitSimpleRoomHelper implements MonumentRoomFitHelper
         {
             private FitSimpleRoomHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 return true;
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
-                return new StructureOceanMonumentPieces.SimpleRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new SimpleRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    static class FitSimpleRoomTopHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class FitSimpleRoomTopHelper implements MonumentRoomFitHelper
         {
             private FitSimpleRoomTopHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 return !definition.hasOpening[EnumFacing.WEST.getIndex()] && !definition.hasOpening[EnumFacing.EAST.getIndex()] && !definition.hasOpening[EnumFacing.NORTH.getIndex()] && !definition.hasOpening[EnumFacing.SOUTH.getIndex()] && !definition.hasOpening[EnumFacing.UP.getIndex()];
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
-                return new StructureOceanMonumentPieces.SimpleTopRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new SimpleTopRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    public static class MonumentBuilding extends StructureOceanMonumentPieces.Piece
+    public static class MonumentBuilding extends Piece
         {
-            private StructureOceanMonumentPieces.RoomDefinition sourceRoom;
-            private StructureOceanMonumentPieces.RoomDefinition coreRoom;
-            private final List<StructureOceanMonumentPieces.Piece> childPieces = Lists.<StructureOceanMonumentPieces.Piece>newArrayList();
+            private RoomDefinition sourceRoom;
+            private RoomDefinition coreRoom;
+            private final List<Piece> childPieces = Lists.<Piece>newArrayList();
 
             public MonumentBuilding()
             {
@@ -702,26 +702,26 @@
                     this.boundingBox = new StructureBoundingBox(p_i45599_2_, 39, p_i45599_3_, p_i45599_2_ + 58 - 1, 61, p_i45599_3_ + 58 - 1);
                 }
 
-                List<StructureOceanMonumentPieces.RoomDefinition> list = this.generateRoomGraph(p_i45599_1_);
+                List<RoomDefinition> list = this.generateRoomGraph(p_i45599_1_);
                 this.sourceRoom.claimed = true;
-                this.childPieces.add(new StructureOceanMonumentPieces.EntryRoom(enumfacing, this.sourceRoom));
-                this.childPieces.add(new StructureOceanMonumentPieces.MonumentCoreRoom(enumfacing, this.coreRoom, p_i45599_1_));
-                List<StructureOceanMonumentPieces.MonumentRoomFitHelper> list1 = Lists.<StructureOceanMonumentPieces.MonumentRoomFitHelper>newArrayList();
-                list1.add(new StructureOceanMonumentPieces.XYDoubleRoomFitHelper());
-                list1.add(new StructureOceanMonumentPieces.YZDoubleRoomFitHelper());
-                list1.add(new StructureOceanMonumentPieces.ZDoubleRoomFitHelper());
-                list1.add(new StructureOceanMonumentPieces.XDoubleRoomFitHelper());
-                list1.add(new StructureOceanMonumentPieces.YDoubleRoomFitHelper());
-                list1.add(new StructureOceanMonumentPieces.FitSimpleRoomTopHelper());
-                list1.add(new StructureOceanMonumentPieces.FitSimpleRoomHelper());
+                this.childPieces.add(new EntryRoom(enumfacing, this.sourceRoom));
+                this.childPieces.add(new MonumentCoreRoom(enumfacing, this.coreRoom, p_i45599_1_));
+                List<MonumentRoomFitHelper> list1 = Lists.<MonumentRoomFitHelper>newArrayList();
+                list1.add(new XYDoubleRoomFitHelper());
+                list1.add(new YZDoubleRoomFitHelper());
+                list1.add(new ZDoubleRoomFitHelper());
+                list1.add(new XDoubleRoomFitHelper());
+                list1.add(new YDoubleRoomFitHelper());
+                list1.add(new FitSimpleRoomTopHelper());
+                list1.add(new FitSimpleRoomHelper());
                 label47:
 
-                for (StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition : list)
+                for (RoomDefinition structureoceanmonumentpieces$roomdefinition : list)
                 {
                     if (!structureoceanmonumentpieces$roomdefinition.claimed && !structureoceanmonumentpieces$roomdefinition.isSpecial())
                     {
                         Iterator lvt_10_1_ = list1.iterator();
-                        StructureOceanMonumentPieces.MonumentRoomFitHelper structureoceanmonumentpieces$monumentroomfithelper;
+                        MonumentRoomFitHelper structureoceanmonumentpieces$monumentroomfithelper;
 
                         while (true)
                         {
@@ -730,7 +730,7 @@
                                 continue label47;
                             }
 
-                            structureoceanmonumentpieces$monumentroomfithelper = (StructureOceanMonumentPieces.MonumentRoomFitHelper)lvt_10_1_.next();
+                            structureoceanmonumentpieces$monumentroomfithelper = (MonumentRoomFitHelper)lvt_10_1_.next();
 
                             if (structureoceanmonumentpieces$monumentroomfithelper.fits(structureoceanmonumentpieces$roomdefinition))
                             {
@@ -746,7 +746,7 @@
                 int k = this.getXWithOffset(9, 22);
                 int l = this.getZWithOffset(9, 22);
 
-                for (StructureOceanMonumentPieces.Piece structureoceanmonumentpieces$piece : this.childPieces)
+                for (Piece structureoceanmonumentpieces$piece : this.childPieces)
                 {
                     structureoceanmonumentpieces$piece.getBoundingBox().offset(k, j, l);
                 }
@@ -755,14 +755,14 @@
                 StructureBoundingBox structureboundingbox2 = StructureBoundingBox.createProper(this.getXWithOffset(34, 1), this.getYWithOffset(1), this.getZWithOffset(34, 1), this.getXWithOffset(56, 21), this.getYWithOffset(8), this.getZWithOffset(56, 21));
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.createProper(this.getXWithOffset(22, 22), this.getYWithOffset(13), this.getZWithOffset(22, 22), this.getXWithOffset(35, 35), this.getYWithOffset(17), this.getZWithOffset(35, 35));
                 int i = p_i45599_1_.nextInt();
-                this.childPieces.add(new StructureOceanMonumentPieces.WingRoom(enumfacing, structureboundingbox1, i++));
-                this.childPieces.add(new StructureOceanMonumentPieces.WingRoom(enumfacing, structureboundingbox2, i++));
-                this.childPieces.add(new StructureOceanMonumentPieces.Penthouse(enumfacing, structureboundingbox));
+                this.childPieces.add(new WingRoom(enumfacing, structureboundingbox1, i++));
+                this.childPieces.add(new WingRoom(enumfacing, structureboundingbox2, i++));
+                this.childPieces.add(new Penthouse(enumfacing, structureboundingbox));
             }
 
-            private List<StructureOceanMonumentPieces.RoomDefinition> generateRoomGraph(Random p_175836_1_)
+            private List<RoomDefinition> generateRoomGraph(Random p_175836_1_)
             {
-                StructureOceanMonumentPieces.RoomDefinition[] astructureoceanmonumentpieces$roomdefinition = new StructureOceanMonumentPieces.RoomDefinition[75];
+                RoomDefinition[] astructureoceanmonumentpieces$roomdefinition = new RoomDefinition[75];
 
                 for (int i = 0; i < 5; ++i)
                 {
@@ -770,7 +770,7 @@
                     {
                         int k = 0;
                         int l = getRoomIndex(i, 0, j);
-                        astructureoceanmonumentpieces$roomdefinition[l] = new StructureOceanMonumentPieces.RoomDefinition(l);
+                        astructureoceanmonumentpieces$roomdefinition[l] = new RoomDefinition(l);
                     }
                 }
 
@@ -780,7 +780,7 @@
                     {
                         int k3 = 1;
                         int j4 = getRoomIndex(i2, 1, l2);
-                        astructureoceanmonumentpieces$roomdefinition[j4] = new StructureOceanMonumentPieces.RoomDefinition(j4);
+                        astructureoceanmonumentpieces$roomdefinition[j4] = new RoomDefinition(j4);
                     }
                 }
 
@@ -790,7 +790,7 @@
                     {
                         int l3 = 2;
                         int k4 = getRoomIndex(j2, 2, i3);
-                        astructureoceanmonumentpieces$roomdefinition[k4] = new StructureOceanMonumentPieces.RoomDefinition(k4);
+                        astructureoceanmonumentpieces$roomdefinition[k4] = new RoomDefinition(k4);
                     }
                 }
 
@@ -834,9 +834,9 @@
                     }
                 }
 
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = new StructureOceanMonumentPieces.RoomDefinition(1003);
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition1 = new StructureOceanMonumentPieces.RoomDefinition(1001);
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition2 = new StructureOceanMonumentPieces.RoomDefinition(1002);
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = new RoomDefinition(1003);
+                RoomDefinition structureoceanmonumentpieces$roomdefinition1 = new RoomDefinition(1001);
+                RoomDefinition structureoceanmonumentpieces$roomdefinition2 = new RoomDefinition(1002);
                 astructureoceanmonumentpieces$roomdefinition[GRIDROOM_TOP_CONNECT_INDEX].setConnection(EnumFacing.UP, structureoceanmonumentpieces$roomdefinition);
                 astructureoceanmonumentpieces$roomdefinition[GRIDROOM_LEFTWING_CONNECT_INDEX].setConnection(EnumFacing.SOUTH, structureoceanmonumentpieces$roomdefinition1);
                 astructureoceanmonumentpieces$roomdefinition[GRIDROOM_RIGHTWING_CONNECT_INDEX].setConnection(EnumFacing.SOUTH, structureoceanmonumentpieces$roomdefinition2);
@@ -853,9 +853,9 @@
                 this.coreRoom.connections[EnumFacing.EAST.getIndex()].connections[EnumFacing.UP.getIndex()].claimed = true;
                 this.coreRoom.connections[EnumFacing.NORTH.getIndex()].connections[EnumFacing.UP.getIndex()].claimed = true;
                 this.coreRoom.connections[EnumFacing.EAST.getIndex()].connections[EnumFacing.NORTH.getIndex()].connections[EnumFacing.UP.getIndex()].claimed = true;
-                List<StructureOceanMonumentPieces.RoomDefinition> list = Lists.<StructureOceanMonumentPieces.RoomDefinition>newArrayList();
+                List<RoomDefinition> list = Lists.<RoomDefinition>newArrayList();
 
-                for (StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition4 : astructureoceanmonumentpieces$roomdefinition)
+                for (RoomDefinition structureoceanmonumentpieces$roomdefinition4 : astructureoceanmonumentpieces$roomdefinition)
                 {
                     if (structureoceanmonumentpieces$roomdefinition4 != null)
                     {
@@ -868,7 +868,7 @@
                 Collections.shuffle(list, p_175836_1_);
                 int i5 = 1;
 
-                for (StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition3 : list)
+                for (RoomDefinition structureoceanmonumentpieces$roomdefinition3 : list)
                 {
                     int j5 = 0;
                     int k5 = 0;
@@ -958,7 +958,7 @@
                     this.generateWaterBox(worldIn, structureBoundingBoxIn, 0 - l1, 0 + l1 * 2, 58 + l1, 57 + l1, 23, 58 + l1, false);
                 }
 
-                for (StructureOceanMonumentPieces.Piece structureoceanmonumentpieces$piece : this.childPieces)
+                for (Piece structureoceanmonumentpieces$piece : this.childPieces)
                 {
                     if (structureoceanmonumentpieces$piece.getBoundingBox().intersectsWith(structureBoundingBoxIn))
                     {
@@ -1336,13 +1336,13 @@
             }
         }
 
-    public static class MonumentCoreRoom extends StructureOceanMonumentPieces.Piece
+    public static class MonumentCoreRoom extends Piece
         {
             public MonumentCoreRoom()
             {
             }
 
-            public MonumentCoreRoom(EnumFacing p_i45598_1_, StructureOceanMonumentPieces.RoomDefinition p_i45598_2_, Random p_i45598_3_)
+            public MonumentCoreRoom(EnumFacing p_i45598_1_, RoomDefinition p_i45598_2_, Random p_i45598_3_)
             {
                 super(1, p_i45598_1_, p_i45598_2_, 2, 2, 2);
             }
@@ -1425,12 +1425,12 @@
 
     interface MonumentRoomFitHelper
     {
-        boolean fits(StructureOceanMonumentPieces.RoomDefinition definition);
+        boolean fits(RoomDefinition definition);
 
-        StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_);
+        Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_);
     }
 
-    public static class Penthouse extends StructureOceanMonumentPieces.Piece
+    public static class Penthouse extends Piece
         {
             public Penthouse()
             {
@@ -1503,7 +1503,7 @@
             protected static final int GRIDROOM_TOP_CONNECT_INDEX = getRoomIndex(2, 2, 0);
             protected static final int GRIDROOM_LEFTWING_CONNECT_INDEX = getRoomIndex(0, 1, 0);
             protected static final int GRIDROOM_RIGHTWING_CONNECT_INDEX = getRoomIndex(4, 1, 0);
-            protected StructureOceanMonumentPieces.RoomDefinition roomDefinition;
+            protected RoomDefinition roomDefinition;
 
             protected static final int getRoomIndex(int p_175820_0_, int p_175820_1_, int p_175820_2_)
             {
@@ -1527,7 +1527,7 @@
                 this.boundingBox = p_i45589_2_;
             }
 
-            protected Piece(int p_i45590_1_, EnumFacing p_i45590_2_, StructureOceanMonumentPieces.RoomDefinition p_i45590_3_, int p_i45590_4_, int p_i45590_5_, int p_i45590_6_)
+            protected Piece(int p_i45590_1_, EnumFacing p_i45590_2_, RoomDefinition p_i45590_3_, int p_i45590_4_, int p_i45590_5_, int p_i45590_6_)
             {
                 super(p_i45590_1_);
                 this.setCoordBaseMode(p_i45590_2_);
@@ -1664,7 +1664,7 @@
     static class RoomDefinition
         {
             int index;
-            StructureOceanMonumentPieces.RoomDefinition[] connections = new StructureOceanMonumentPieces.RoomDefinition[6];
+            RoomDefinition[] connections = new RoomDefinition[6];
             boolean[] hasOpening = new boolean[6];
             boolean claimed;
             boolean isSource;
@@ -1675,7 +1675,7 @@
                 this.index = p_i45584_1_;
             }
 
-            public void setConnection(EnumFacing p_175957_1_, StructureOceanMonumentPieces.RoomDefinition p_175957_2_)
+            public void setConnection(EnumFacing p_175957_1_, RoomDefinition p_175957_2_)
             {
                 this.connections[p_175957_1_.getIndex()] = p_175957_2_;
                 p_175957_2_.connections[p_175957_1_.getOpposite().getIndex()] = this;
@@ -1732,7 +1732,7 @@
             }
         }
 
-    public static class SimpleRoom extends StructureOceanMonumentPieces.Piece
+    public static class SimpleRoom extends Piece
         {
             private int mainDesign;
 
@@ -1740,7 +1740,7 @@
             {
             }
 
-            public SimpleRoom(EnumFacing p_i45587_1_, StructureOceanMonumentPieces.RoomDefinition p_i45587_2_, Random p_i45587_3_)
+            public SimpleRoom(EnumFacing p_i45587_1_, RoomDefinition p_i45587_2_, Random p_i45587_3_)
             {
                 super(1, p_i45587_1_, p_i45587_2_, 1, 1, 1);
                 this.mainDesign = p_i45587_3_.nextInt(3);
@@ -1933,13 +1933,13 @@
             }
         }
 
-    public static class SimpleTopRoom extends StructureOceanMonumentPieces.Piece
+    public static class SimpleTopRoom extends Piece
         {
             public SimpleTopRoom()
             {
             }
 
-            public SimpleTopRoom(EnumFacing p_i45586_1_, StructureOceanMonumentPieces.RoomDefinition p_i45586_2_, Random p_i45586_3_)
+            public SimpleTopRoom(EnumFacing p_i45586_1_, RoomDefinition p_i45586_2_, Random p_i45586_3_)
             {
                 super(1, p_i45586_1_, p_i45586_2_, 1, 1, 1);
             }
@@ -1994,7 +1994,7 @@
             }
         }
 
-    public static class WingRoom extends StructureOceanMonumentPieces.Piece
+    public static class WingRoom extends Piece
         {
             private int mainDesign;
 
@@ -2110,36 +2110,36 @@
             }
         }
 
-    static class XDoubleRoomFitHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class XDoubleRoomFitHelper implements MonumentRoomFitHelper
         {
             private XDoubleRoomFitHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 return definition.hasOpening[EnumFacing.EAST.getIndex()] && !definition.connections[EnumFacing.EAST.getIndex()].claimed;
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
                 p_175968_2_.connections[EnumFacing.EAST.getIndex()].claimed = true;
-                return new StructureOceanMonumentPieces.DoubleXRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new DoubleXRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    static class XYDoubleRoomFitHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class XYDoubleRoomFitHelper implements MonumentRoomFitHelper
         {
             private XYDoubleRoomFitHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 if (definition.hasOpening[EnumFacing.EAST.getIndex()] && !definition.connections[EnumFacing.EAST.getIndex()].claimed && definition.hasOpening[EnumFacing.UP.getIndex()] && !definition.connections[EnumFacing.UP.getIndex()].claimed)
                 {
-                    StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = definition.connections[EnumFacing.EAST.getIndex()];
+                    RoomDefinition structureoceanmonumentpieces$roomdefinition = definition.connections[EnumFacing.EAST.getIndex()];
                     return structureoceanmonumentpieces$roomdefinition.hasOpening[EnumFacing.UP.getIndex()] && !structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()].claimed;
                 }
                 else
@@ -2148,46 +2148,46 @@
                 }
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
                 p_175968_2_.connections[EnumFacing.EAST.getIndex()].claimed = true;
                 p_175968_2_.connections[EnumFacing.UP.getIndex()].claimed = true;
                 p_175968_2_.connections[EnumFacing.EAST.getIndex()].connections[EnumFacing.UP.getIndex()].claimed = true;
-                return new StructureOceanMonumentPieces.DoubleXYRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new DoubleXYRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    static class YDoubleRoomFitHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class YDoubleRoomFitHelper implements MonumentRoomFitHelper
         {
             private YDoubleRoomFitHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 return definition.hasOpening[EnumFacing.UP.getIndex()] && !definition.connections[EnumFacing.UP.getIndex()].claimed;
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
                 p_175968_2_.connections[EnumFacing.UP.getIndex()].claimed = true;
-                return new StructureOceanMonumentPieces.DoubleYRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new DoubleYRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    static class YZDoubleRoomFitHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class YZDoubleRoomFitHelper implements MonumentRoomFitHelper
         {
             private YZDoubleRoomFitHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 if (definition.hasOpening[EnumFacing.NORTH.getIndex()] && !definition.connections[EnumFacing.NORTH.getIndex()].claimed && definition.hasOpening[EnumFacing.UP.getIndex()] && !definition.connections[EnumFacing.UP.getIndex()].claimed)
                 {
-                    StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = definition.connections[EnumFacing.NORTH.getIndex()];
+                    RoomDefinition structureoceanmonumentpieces$roomdefinition = definition.connections[EnumFacing.NORTH.getIndex()];
                     return structureoceanmonumentpieces$roomdefinition.hasOpening[EnumFacing.UP.getIndex()] && !structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.UP.getIndex()].claimed;
                 }
                 else
@@ -2196,30 +2196,30 @@
                 }
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
                 p_175968_2_.claimed = true;
                 p_175968_2_.connections[EnumFacing.NORTH.getIndex()].claimed = true;
                 p_175968_2_.connections[EnumFacing.UP.getIndex()].claimed = true;
                 p_175968_2_.connections[EnumFacing.NORTH.getIndex()].connections[EnumFacing.UP.getIndex()].claimed = true;
-                return new StructureOceanMonumentPieces.DoubleYZRoom(p_175968_1_, p_175968_2_, p_175968_3_);
+                return new DoubleYZRoom(p_175968_1_, p_175968_2_, p_175968_3_);
             }
         }
 
-    static class ZDoubleRoomFitHelper implements StructureOceanMonumentPieces.MonumentRoomFitHelper
+    static class ZDoubleRoomFitHelper implements MonumentRoomFitHelper
         {
             private ZDoubleRoomFitHelper()
             {
             }
 
-            public boolean fits(StructureOceanMonumentPieces.RoomDefinition definition)
+            public boolean fits(RoomDefinition definition)
             {
                 return definition.hasOpening[EnumFacing.NORTH.getIndex()] && !definition.connections[EnumFacing.NORTH.getIndex()].claimed;
             }
 
-            public StructureOceanMonumentPieces.Piece create(EnumFacing p_175968_1_, StructureOceanMonumentPieces.RoomDefinition p_175968_2_, Random p_175968_3_)
+            public Piece create(EnumFacing p_175968_1_, RoomDefinition p_175968_2_, Random p_175968_3_)
             {
-                StructureOceanMonumentPieces.RoomDefinition structureoceanmonumentpieces$roomdefinition = p_175968_2_;
+                RoomDefinition structureoceanmonumentpieces$roomdefinition = p_175968_2_;
 
                 if (!p_175968_2_.hasOpening[EnumFacing.NORTH.getIndex()] || p_175968_2_.connections[EnumFacing.NORTH.getIndex()].claimed)
                 {
@@ -2228,7 +2228,7 @@
 
                 structureoceanmonumentpieces$roomdefinition.claimed = true;
                 structureoceanmonumentpieces$roomdefinition.connections[EnumFacing.NORTH.getIndex()].claimed = true;
-                return new StructureOceanMonumentPieces.DoubleZRoom(p_175968_1_, structureoceanmonumentpieces$roomdefinition, p_175968_3_);
+                return new DoubleZRoom(p_175968_1_, structureoceanmonumentpieces$roomdefinition, p_175968_3_);
             }
         }
 }
