--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureStart.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureStart.java
@@ -44,7 +44,7 @@
         {
             StructureComponent structurecomponent = (StructureComponent)iterator.next();
 
-            if (structurecomponent.getBoundingBox().intersectsWith(structurebb) && !structurecomponent.addComponentParts(worldIn, rand, structurebb))
+            if ((structurecomponent == null || structurecomponent.getBoundingBox() == null ) || structurecomponent.getBoundingBox().intersectsWith(structurebb) && !structurecomponent.addComponentParts(worldIn, rand, structurebb))
             {
                 iterator.remove();
             }
