--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureOceanMonument.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureOceanMonument.java
@@ -72,7 +72,7 @@
 
         int k = chunkX / this.spacing;
         int l = chunkZ / this.spacing;
-        Random random = this.world.setRandomSeed(k, l, 10387313);
+        Random random = this.world.setRandomSeed(k, l, this.world.spigotConfig.monumentSeed); // Spigot
         k = k * this.spacing;
         l = l * this.spacing;
         k = k + (random.nextInt(this.spacing - this.separation) + random.nextInt(this.spacing - this.separation)) / 2;
@@ -104,7 +104,7 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        return new StructureOceanMonument.StartMonument(this.world, this.rand, chunkX, chunkZ);
+        return new StartMonument(this.world, this.rand, chunkX, chunkZ);
     }
 
     public List<Biome.SpawnListEntry> getMonsters()
