--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureNetherBridgePieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureNetherBridgePieces.java
@@ -21,90 +21,90 @@
 @SuppressWarnings("incomplete-switch")
 public class StructureNetherBridgePieces
 {
-    private static final StructureNetherBridgePieces.PieceWeight[] PRIMARY_COMPONENTS = new StructureNetherBridgePieces.PieceWeight[] {new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Straight.class, 30, 0, true), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Crossing3.class, 10, 4), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Crossing.class, 10, 4), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Stairs.class, 10, 3), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Throne.class, 5, 2), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Entrance.class, 5, 1)};
-    private static final StructureNetherBridgePieces.PieceWeight[] SECONDARY_COMPONENTS = new StructureNetherBridgePieces.PieceWeight[] {new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Corridor5.class, 25, 0, true), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Crossing2.class, 15, 5), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Corridor2.class, 5, 10), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Corridor.class, 5, 10), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Corridor3.class, 10, 3, true), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.Corridor4.class, 7, 2), new StructureNetherBridgePieces.PieceWeight(StructureNetherBridgePieces.NetherStalkRoom.class, 5, 2)};
+    private static final PieceWeight[] PRIMARY_COMPONENTS = new PieceWeight[] {new PieceWeight(Straight.class, 30, 0, true), new PieceWeight(Crossing3.class, 10, 4), new PieceWeight(Crossing.class, 10, 4), new PieceWeight(Stairs.class, 10, 3), new PieceWeight(Throne.class, 5, 2), new PieceWeight(Entrance.class, 5, 1)};
+    private static final PieceWeight[] SECONDARY_COMPONENTS = new PieceWeight[] {new PieceWeight(Corridor5.class, 25, 0, true), new PieceWeight(Crossing2.class, 15, 5), new PieceWeight(Corridor2.class, 5, 10), new PieceWeight(Corridor.class, 5, 10), new PieceWeight(Corridor3.class, 10, 3, true), new PieceWeight(Corridor4.class, 7, 2), new PieceWeight(NetherStalkRoom.class, 5, 2)};
 
     public static void registerNetherFortressPieces()
     {
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Crossing3.class, "NeBCr");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.End.class, "NeBEF");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Straight.class, "NeBS");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Corridor3.class, "NeCCS");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Corridor4.class, "NeCTB");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Entrance.class, "NeCE");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Crossing2.class, "NeSCSC");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Corridor.class, "NeSCLT");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Corridor5.class, "NeSC");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Corridor2.class, "NeSCRT");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.NetherStalkRoom.class, "NeCSR");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Throne.class, "NeMT");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Crossing.class, "NeRC");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Stairs.class, "NeSR");
-        MapGenStructureIO.registerStructureComponent(StructureNetherBridgePieces.Start.class, "NeStart");
+        MapGenStructureIO.registerStructureComponent(Crossing3.class, "NeBCr");
+        MapGenStructureIO.registerStructureComponent(End.class, "NeBEF");
+        MapGenStructureIO.registerStructureComponent(Straight.class, "NeBS");
+        MapGenStructureIO.registerStructureComponent(Corridor3.class, "NeCCS");
+        MapGenStructureIO.registerStructureComponent(Corridor4.class, "NeCTB");
+        MapGenStructureIO.registerStructureComponent(Entrance.class, "NeCE");
+        MapGenStructureIO.registerStructureComponent(Crossing2.class, "NeSCSC");
+        MapGenStructureIO.registerStructureComponent(Corridor.class, "NeSCLT");
+        MapGenStructureIO.registerStructureComponent(Corridor5.class, "NeSC");
+        MapGenStructureIO.registerStructureComponent(Corridor2.class, "NeSCRT");
+        MapGenStructureIO.registerStructureComponent(NetherStalkRoom.class, "NeCSR");
+        MapGenStructureIO.registerStructureComponent(Throne.class, "NeMT");
+        MapGenStructureIO.registerStructureComponent(Crossing.class, "NeRC");
+        MapGenStructureIO.registerStructureComponent(Stairs.class, "NeSR");
+        MapGenStructureIO.registerStructureComponent(Start.class, "NeStart");
     }
 
-    private static StructureNetherBridgePieces.Piece findAndCreateBridgePieceFactory(StructureNetherBridgePieces.PieceWeight p_175887_0_, List<StructureComponent> p_175887_1_, Random p_175887_2_, int p_175887_3_, int p_175887_4_, int p_175887_5_, EnumFacing p_175887_6_, int p_175887_7_)
+    private static Piece findAndCreateBridgePieceFactory(PieceWeight p_175887_0_, List<StructureComponent> p_175887_1_, Random p_175887_2_, int p_175887_3_, int p_175887_4_, int p_175887_5_, EnumFacing p_175887_6_, int p_175887_7_)
     {
-        Class <? extends StructureNetherBridgePieces.Piece > oclass = p_175887_0_.weightClass;
-        StructureNetherBridgePieces.Piece structurenetherbridgepieces$piece = null;
+        Class <? extends Piece > oclass = p_175887_0_.weightClass;
+        Piece structurenetherbridgepieces$piece = null;
 
-        if (oclass == StructureNetherBridgePieces.Straight.class)
+        if (oclass == Straight.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Straight.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Straight.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Crossing3.class)
+        else if (oclass == Crossing3.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Crossing3.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Crossing3.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Crossing.class)
+        else if (oclass == Crossing.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Crossing.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Crossing.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Stairs.class)
+        else if (oclass == Stairs.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Stairs.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_7_, p_175887_6_);
+            structurenetherbridgepieces$piece = Stairs.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_7_, p_175887_6_);
         }
-        else if (oclass == StructureNetherBridgePieces.Throne.class)
+        else if (oclass == Throne.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Throne.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_7_, p_175887_6_);
+            structurenetherbridgepieces$piece = Throne.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_7_, p_175887_6_);
         }
-        else if (oclass == StructureNetherBridgePieces.Entrance.class)
+        else if (oclass == Entrance.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Entrance.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Entrance.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Corridor5.class)
+        else if (oclass == Corridor5.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Corridor5.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Corridor5.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Corridor2.class)
+        else if (oclass == Corridor2.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Corridor2.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Corridor2.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Corridor.class)
+        else if (oclass == Corridor.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Corridor.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Corridor.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Corridor3.class)
+        else if (oclass == Corridor3.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Corridor3.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Corridor3.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Corridor4.class)
+        else if (oclass == Corridor4.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Corridor4.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Corridor4.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.Crossing2.class)
+        else if (oclass == Crossing2.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.Crossing2.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = Crossing2.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
-        else if (oclass == StructureNetherBridgePieces.NetherStalkRoom.class)
+        else if (oclass == NetherStalkRoom.class)
         {
-            structurenetherbridgepieces$piece = StructureNetherBridgePieces.NetherStalkRoom.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
+            structurenetherbridgepieces$piece = NetherStalkRoom.createPiece(p_175887_1_, p_175887_2_, p_175887_3_, p_175887_4_, p_175887_5_, p_175887_6_, p_175887_7_);
         }
 
         return structurenetherbridgepieces$piece;
     }
 
-    public static class Corridor extends StructureNetherBridgePieces.Piece
+    public static class Corridor extends Piece
         {
             private boolean chest;
 
@@ -134,13 +134,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentX((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 1, true);
+                this.getNextComponentX((Start)componentIn, listIn, rand, 0, 1, true);
             }
 
-            public static StructureNetherBridgePieces.Corridor createPiece(List<StructureComponent> p_175879_0_, Random p_175879_1_, int p_175879_2_, int p_175879_3_, int p_175879_4_, EnumFacing p_175879_5_, int p_175879_6_)
+            public static Corridor createPiece(List<StructureComponent> p_175879_0_, Random p_175879_1_, int p_175879_2_, int p_175879_3_, int p_175879_4_, EnumFacing p_175879_5_, int p_175879_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175879_2_, p_175879_3_, p_175879_4_, -1, 0, 0, 5, 7, 5, p_175879_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175879_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Corridor(p_175879_6_, p_175879_1_, structureboundingbox, p_175879_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175879_0_, structureboundingbox) == null ? new Corridor(p_175879_6_, p_175879_1_, structureboundingbox, p_175879_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -175,7 +175,7 @@
             }
         }
 
-    public static class Corridor2 extends StructureNetherBridgePieces.Piece
+    public static class Corridor2 extends Piece
         {
             private boolean chest;
 
@@ -205,13 +205,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 1, true);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 0, 1, true);
             }
 
-            public static StructureNetherBridgePieces.Corridor2 createPiece(List<StructureComponent> p_175876_0_, Random p_175876_1_, int p_175876_2_, int p_175876_3_, int p_175876_4_, EnumFacing p_175876_5_, int p_175876_6_)
+            public static Corridor2 createPiece(List<StructureComponent> p_175876_0_, Random p_175876_1_, int p_175876_2_, int p_175876_3_, int p_175876_4_, EnumFacing p_175876_5_, int p_175876_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175876_2_, p_175876_3_, p_175876_4_, -1, 0, 0, 5, 7, 5, p_175876_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175876_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Corridor2(p_175876_6_, p_175876_1_, structureboundingbox, p_175876_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175876_0_, structureboundingbox) == null ? new Corridor2(p_175876_6_, p_175876_1_, structureboundingbox, p_175876_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -246,7 +246,7 @@
             }
         }
 
-    public static class Corridor3 extends StructureNetherBridgePieces.Piece
+    public static class Corridor3 extends Piece
         {
             public Corridor3()
             {
@@ -261,13 +261,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 1, 0, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 1, 0, true);
             }
 
-            public static StructureNetherBridgePieces.Corridor3 createPiece(List<StructureComponent> p_175883_0_, Random p_175883_1_, int p_175883_2_, int p_175883_3_, int p_175883_4_, EnumFacing p_175883_5_, int p_175883_6_)
+            public static Corridor3 createPiece(List<StructureComponent> p_175883_0_, Random p_175883_1_, int p_175883_2_, int p_175883_3_, int p_175883_4_, EnumFacing p_175883_5_, int p_175883_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175883_2_, p_175883_3_, p_175883_4_, -1, -7, 0, 5, 14, 10, p_175883_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175883_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Corridor3(p_175883_6_, p_175883_1_, structureboundingbox, p_175883_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175883_0_, structureboundingbox) == null ? new Corridor3(p_175883_6_, p_175883_1_, structureboundingbox, p_175883_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -309,7 +309,7 @@
             }
         }
 
-    public static class Corridor4 extends StructureNetherBridgePieces.Piece
+    public static class Corridor4 extends Piece
         {
             public Corridor4()
             {
@@ -332,14 +332,14 @@
                     i = 5;
                 }
 
-                this.getNextComponentX((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, i, rand.nextInt(8) > 0);
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, i, rand.nextInt(8) > 0);
+                this.getNextComponentX((Start)componentIn, listIn, rand, 0, i, rand.nextInt(8) > 0);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 0, i, rand.nextInt(8) > 0);
             }
 
-            public static StructureNetherBridgePieces.Corridor4 createPiece(List<StructureComponent> p_175880_0_, Random p_175880_1_, int p_175880_2_, int p_175880_3_, int p_175880_4_, EnumFacing p_175880_5_, int p_175880_6_)
+            public static Corridor4 createPiece(List<StructureComponent> p_175880_0_, Random p_175880_1_, int p_175880_2_, int p_175880_3_, int p_175880_4_, EnumFacing p_175880_5_, int p_175880_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175880_2_, p_175880_3_, p_175880_4_, -3, 0, 0, 9, 7, 9, p_175880_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175880_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Corridor4(p_175880_6_, p_175880_1_, structureboundingbox, p_175880_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175880_0_, structureboundingbox) == null ? new Corridor4(p_175880_6_, p_175880_1_, structureboundingbox, p_175880_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -376,7 +376,7 @@
             }
         }
 
-    public static class Corridor5 extends StructureNetherBridgePieces.Piece
+    public static class Corridor5 extends Piece
         {
             public Corridor5()
             {
@@ -391,13 +391,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 1, 0, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 1, 0, true);
             }
 
-            public static StructureNetherBridgePieces.Corridor5 createPiece(List<StructureComponent> p_175877_0_, Random p_175877_1_, int p_175877_2_, int p_175877_3_, int p_175877_4_, EnumFacing p_175877_5_, int p_175877_6_)
+            public static Corridor5 createPiece(List<StructureComponent> p_175877_0_, Random p_175877_1_, int p_175877_2_, int p_175877_3_, int p_175877_4_, EnumFacing p_175877_5_, int p_175877_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175877_2_, p_175877_3_, p_175877_4_, -1, 0, 0, 5, 7, 5, p_175877_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175877_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Corridor5(p_175877_6_, p_175877_1_, structureboundingbox, p_175877_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175877_0_, structureboundingbox) == null ? new Corridor5(p_175877_6_, p_175877_1_, structureboundingbox, p_175877_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -424,7 +424,7 @@
             }
         }
 
-    public static class Crossing extends StructureNetherBridgePieces.Piece
+    public static class Crossing extends Piece
         {
             public Crossing()
             {
@@ -439,15 +439,15 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 2, 0, false);
-                this.getNextComponentX((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 2, false);
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 2, false);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 2, 0, false);
+                this.getNextComponentX((Start)componentIn, listIn, rand, 0, 2, false);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 0, 2, false);
             }
 
-            public static StructureNetherBridgePieces.Crossing createPiece(List<StructureComponent> p_175873_0_, Random p_175873_1_, int p_175873_2_, int p_175873_3_, int p_175873_4_, EnumFacing p_175873_5_, int p_175873_6_)
+            public static Crossing createPiece(List<StructureComponent> p_175873_0_, Random p_175873_1_, int p_175873_2_, int p_175873_3_, int p_175873_4_, EnumFacing p_175873_5_, int p_175873_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175873_2_, p_175873_3_, p_175873_4_, -2, 0, 0, 7, 9, 7, p_175873_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175873_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Crossing(p_175873_6_, p_175873_1_, structureboundingbox, p_175873_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175873_0_, structureboundingbox) == null ? new Crossing(p_175873_6_, p_175873_1_, structureboundingbox, p_175873_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -483,7 +483,7 @@
             }
         }
 
-    public static class Crossing2 extends StructureNetherBridgePieces.Piece
+    public static class Crossing2 extends Piece
         {
             public Crossing2()
             {
@@ -498,15 +498,15 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 1, 0, true);
-                this.getNextComponentX((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 1, true);
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 0, 1, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 1, 0, true);
+                this.getNextComponentX((Start)componentIn, listIn, rand, 0, 1, true);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 0, 1, true);
             }
 
-            public static StructureNetherBridgePieces.Crossing2 createPiece(List<StructureComponent> p_175878_0_, Random p_175878_1_, int p_175878_2_, int p_175878_3_, int p_175878_4_, EnumFacing p_175878_5_, int p_175878_6_)
+            public static Crossing2 createPiece(List<StructureComponent> p_175878_0_, Random p_175878_1_, int p_175878_2_, int p_175878_3_, int p_175878_4_, EnumFacing p_175878_5_, int p_175878_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175878_2_, p_175878_3_, p_175878_4_, -1, 0, 0, 5, 7, 5, p_175878_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175878_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Crossing2(p_175878_6_, p_175878_1_, structureboundingbox, p_175878_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175878_0_, structureboundingbox) == null ? new Crossing2(p_175878_6_, p_175878_1_, structureboundingbox, p_175878_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -531,7 +531,7 @@
             }
         }
 
-    public static class Crossing3 extends StructureNetherBridgePieces.Piece
+    public static class Crossing3 extends Piece
         {
             public Crossing3()
             {
@@ -561,15 +561,15 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 8, 3, false);
-                this.getNextComponentX((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 3, 8, false);
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 3, 8, false);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 8, 3, false);
+                this.getNextComponentX((Start)componentIn, listIn, rand, 3, 8, false);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 3, 8, false);
             }
 
-            public static StructureNetherBridgePieces.Crossing3 createPiece(List<StructureComponent> p_175885_0_, Random p_175885_1_, int p_175885_2_, int p_175885_3_, int p_175885_4_, EnumFacing p_175885_5_, int p_175885_6_)
+            public static Crossing3 createPiece(List<StructureComponent> p_175885_0_, Random p_175885_1_, int p_175885_2_, int p_175885_3_, int p_175885_4_, EnumFacing p_175885_5_, int p_175885_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175885_2_, p_175885_3_, p_175885_4_, -8, -3, 0, 19, 10, 19, p_175885_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175885_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Crossing3(p_175885_6_, p_175885_1_, structureboundingbox, p_175885_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175885_0_, structureboundingbox) == null ? new Crossing3(p_175885_6_, p_175885_1_, structureboundingbox, p_175885_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -618,7 +618,7 @@
             }
         }
 
-    public static class End extends StructureNetherBridgePieces.Piece
+    public static class End extends Piece
         {
             private int fillSeed;
 
@@ -634,10 +634,10 @@
                 this.fillSeed = p_i45621_2_.nextInt();
             }
 
-            public static StructureNetherBridgePieces.End createPiece(List<StructureComponent> p_175884_0_, Random p_175884_1_, int p_175884_2_, int p_175884_3_, int p_175884_4_, EnumFacing p_175884_5_, int p_175884_6_)
+            public static End createPiece(List<StructureComponent> p_175884_0_, Random p_175884_1_, int p_175884_2_, int p_175884_3_, int p_175884_4_, EnumFacing p_175884_5_, int p_175884_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175884_2_, p_175884_3_, p_175884_4_, -1, -3, 0, 5, 10, 8, p_175884_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175884_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.End(p_175884_6_, p_175884_1_, structureboundingbox, p_175884_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175884_0_, structureboundingbox) == null ? new End(p_175884_6_, p_175884_1_, structureboundingbox, p_175884_5_) : null;
             }
 
             protected void readStructureFromNBT(NBTTagCompound tagCompound, TemplateManager p_143011_2_)
@@ -689,7 +689,7 @@
             }
         }
 
-    public static class Entrance extends StructureNetherBridgePieces.Piece
+    public static class Entrance extends Piece
         {
             public Entrance()
             {
@@ -704,13 +704,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 5, 3, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 5, 3, true);
             }
 
-            public static StructureNetherBridgePieces.Entrance createPiece(List<StructureComponent> p_175881_0_, Random p_175881_1_, int p_175881_2_, int p_175881_3_, int p_175881_4_, EnumFacing p_175881_5_, int p_175881_6_)
+            public static Entrance createPiece(List<StructureComponent> p_175881_0_, Random p_175881_1_, int p_175881_2_, int p_175881_3_, int p_175881_4_, EnumFacing p_175881_5_, int p_175881_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175881_2_, p_175881_3_, p_175881_4_, -5, -3, 0, 13, 14, 13, p_175881_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175881_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Entrance(p_175881_6_, p_175881_1_, structureboundingbox, p_175881_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175881_0_, structureboundingbox) == null ? new Entrance(p_175881_6_, p_175881_1_, structureboundingbox, p_175881_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -796,7 +796,7 @@
             }
         }
 
-    public static class NetherStalkRoom extends StructureNetherBridgePieces.Piece
+    public static class NetherStalkRoom extends Piece
         {
             public NetherStalkRoom()
             {
@@ -811,14 +811,14 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 5, 3, true);
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 5, 11, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 5, 3, true);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 5, 11, true);
             }
 
-            public static StructureNetherBridgePieces.NetherStalkRoom createPiece(List<StructureComponent> p_175875_0_, Random p_175875_1_, int p_175875_2_, int p_175875_3_, int p_175875_4_, EnumFacing p_175875_5_, int p_175875_6_)
+            public static NetherStalkRoom createPiece(List<StructureComponent> p_175875_0_, Random p_175875_1_, int p_175875_2_, int p_175875_3_, int p_175875_4_, EnumFacing p_175875_5_, int p_175875_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175875_2_, p_175875_3_, p_175875_4_, -5, -3, 0, 13, 14, 13, p_175875_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175875_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.NetherStalkRoom(p_175875_6_, p_175875_1_, structureboundingbox, p_175875_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175875_0_, structureboundingbox) == null ? new NetherStalkRoom(p_175875_6_, p_175875_1_, structureboundingbox, p_175875_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -964,12 +964,12 @@
             {
             }
 
-            private int getTotalWeight(List<StructureNetherBridgePieces.PieceWeight> p_74960_1_)
+            private int getTotalWeight(List<PieceWeight> p_74960_1_)
             {
                 boolean flag = false;
                 int i = 0;
 
-                for (StructureNetherBridgePieces.PieceWeight structurenetherbridgepieces$pieceweight : p_74960_1_)
+                for (PieceWeight structurenetherbridgepieces$pieceweight : p_74960_1_)
                 {
                     if (structurenetherbridgepieces$pieceweight.maxPlaceCount > 0 && structurenetherbridgepieces$pieceweight.placeCount < structurenetherbridgepieces$pieceweight.maxPlaceCount)
                     {
@@ -982,7 +982,7 @@
                 return flag ? i : -1;
             }
 
-            private StructureNetherBridgePieces.Piece generatePiece(StructureNetherBridgePieces.Start p_175871_1_, List<StructureNetherBridgePieces.PieceWeight> p_175871_2_, List<StructureComponent> p_175871_3_, Random p_175871_4_, int p_175871_5_, int p_175871_6_, int p_175871_7_, EnumFacing p_175871_8_, int p_175871_9_)
+            private Piece generatePiece(Start p_175871_1_, List<PieceWeight> p_175871_2_, List<StructureComponent> p_175871_3_, Random p_175871_4_, int p_175871_5_, int p_175871_6_, int p_175871_7_, EnumFacing p_175871_8_, int p_175871_9_)
             {
                 int i = this.getTotalWeight(p_175871_2_);
                 boolean flag = i > 0 && p_175871_9_ <= 30;
@@ -993,7 +993,7 @@
                     ++j;
                     int k = p_175871_4_.nextInt(i);
 
-                    for (StructureNetherBridgePieces.PieceWeight structurenetherbridgepieces$pieceweight : p_175871_2_)
+                    for (PieceWeight structurenetherbridgepieces$pieceweight : p_175871_2_)
                     {
                         k -= structurenetherbridgepieces$pieceweight.weight;
 
@@ -1004,7 +1004,7 @@
                                 break;
                             }
 
-                            StructureNetherBridgePieces.Piece structurenetherbridgepieces$piece = StructureNetherBridgePieces.findAndCreateBridgePieceFactory(structurenetherbridgepieces$pieceweight, p_175871_3_, p_175871_4_, p_175871_5_, p_175871_6_, p_175871_7_, p_175871_8_, p_175871_9_);
+                            Piece structurenetherbridgepieces$piece = StructureNetherBridgePieces.findAndCreateBridgePieceFactory(structurenetherbridgepieces$pieceweight, p_175871_3_, p_175871_4_, p_175871_5_, p_175871_6_, p_175871_7_, p_175871_8_, p_175871_9_);
 
                             if (structurenetherbridgepieces$piece != null)
                             {
@@ -1022,14 +1022,14 @@
                     }
                 }
 
-                return StructureNetherBridgePieces.End.createPiece(p_175871_3_, p_175871_4_, p_175871_5_, p_175871_6_, p_175871_7_, p_175871_8_, p_175871_9_);
+                return End.createPiece(p_175871_3_, p_175871_4_, p_175871_5_, p_175871_6_, p_175871_7_, p_175871_8_, p_175871_9_);
             }
 
-            private StructureComponent generateAndAddPiece(StructureNetherBridgePieces.Start p_175870_1_, List<StructureComponent> p_175870_2_, Random p_175870_3_, int p_175870_4_, int p_175870_5_, int p_175870_6_, @Nullable EnumFacing p_175870_7_, int p_175870_8_, boolean p_175870_9_)
+            private StructureComponent generateAndAddPiece(Start p_175870_1_, List<StructureComponent> p_175870_2_, Random p_175870_3_, int p_175870_4_, int p_175870_5_, int p_175870_6_, @Nullable EnumFacing p_175870_7_, int p_175870_8_, boolean p_175870_9_)
             {
                 if (Math.abs(p_175870_4_ - p_175870_1_.getBoundingBox().minX) <= 112 && Math.abs(p_175870_6_ - p_175870_1_.getBoundingBox().minZ) <= 112)
                 {
-                    List<StructureNetherBridgePieces.PieceWeight> list = p_175870_1_.primaryWeights;
+                    List<PieceWeight> list = p_175870_1_.primaryWeights;
 
                     if (p_175870_9_)
                     {
@@ -1048,12 +1048,12 @@
                 }
                 else
                 {
-                    return StructureNetherBridgePieces.End.createPiece(p_175870_2_, p_175870_3_, p_175870_4_, p_175870_5_, p_175870_6_, p_175870_7_, p_175870_8_);
+                    return End.createPiece(p_175870_2_, p_175870_3_, p_175870_4_, p_175870_5_, p_175870_6_, p_175870_7_, p_175870_8_);
                 }
             }
 
             @Nullable
-            protected StructureComponent getNextComponentNormal(StructureNetherBridgePieces.Start p_74963_1_, List<StructureComponent> p_74963_2_, Random p_74963_3_, int p_74963_4_, int p_74963_5_, boolean p_74963_6_)
+            protected StructureComponent getNextComponentNormal(Start p_74963_1_, List<StructureComponent> p_74963_2_, Random p_74963_3_, int p_74963_4_, int p_74963_5_, boolean p_74963_6_)
             {
                 EnumFacing enumfacing = this.getCoordBaseMode();
 
@@ -1076,7 +1076,7 @@
             }
 
             @Nullable
-            protected StructureComponent getNextComponentX(StructureNetherBridgePieces.Start p_74961_1_, List<StructureComponent> p_74961_2_, Random p_74961_3_, int p_74961_4_, int p_74961_5_, boolean p_74961_6_)
+            protected StructureComponent getNextComponentX(Start p_74961_1_, List<StructureComponent> p_74961_2_, Random p_74961_3_, int p_74961_4_, int p_74961_5_, boolean p_74961_6_)
             {
                 EnumFacing enumfacing = this.getCoordBaseMode();
 
@@ -1099,7 +1099,7 @@
             }
 
             @Nullable
-            protected StructureComponent getNextComponentZ(StructureNetherBridgePieces.Start p_74965_1_, List<StructureComponent> p_74965_2_, Random p_74965_3_, int p_74965_4_, int p_74965_5_, boolean p_74965_6_)
+            protected StructureComponent getNextComponentZ(Start p_74965_1_, List<StructureComponent> p_74965_2_, Random p_74965_3_, int p_74965_4_, int p_74965_5_, boolean p_74965_6_)
             {
                 EnumFacing enumfacing = this.getCoordBaseMode();
 
@@ -1129,13 +1129,13 @@
 
     static class PieceWeight
         {
-            public Class <? extends StructureNetherBridgePieces.Piece > weightClass;
+            public Class <? extends Piece > weightClass;
             public final int weight;
             public int placeCount;
             public int maxPlaceCount;
             public boolean allowInRow;
 
-            public PieceWeight(Class <? extends StructureNetherBridgePieces.Piece > p_i2055_1_, int p_i2055_2_, int p_i2055_3_, boolean p_i2055_4_)
+            public PieceWeight(Class <? extends Piece > p_i2055_1_, int p_i2055_2_, int p_i2055_3_, boolean p_i2055_4_)
             {
                 this.weightClass = p_i2055_1_;
                 this.weight = p_i2055_2_;
@@ -1143,7 +1143,7 @@
                 this.allowInRow = p_i2055_4_;
             }
 
-            public PieceWeight(Class <? extends StructureNetherBridgePieces.Piece > p_i2056_1_, int p_i2056_2_, int p_i2056_3_)
+            public PieceWeight(Class <? extends Piece > p_i2056_1_, int p_i2056_2_, int p_i2056_3_)
             {
                 this(p_i2056_1_, p_i2056_2_, p_i2056_3_, false);
             }
@@ -1159,7 +1159,7 @@
             }
         }
 
-    public static class Stairs extends StructureNetherBridgePieces.Piece
+    public static class Stairs extends Piece
         {
             public Stairs()
             {
@@ -1174,13 +1174,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentZ((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 6, 2, false);
+                this.getNextComponentZ((Start)componentIn, listIn, rand, 6, 2, false);
             }
 
-            public static StructureNetherBridgePieces.Stairs createPiece(List<StructureComponent> p_175872_0_, Random p_175872_1_, int p_175872_2_, int p_175872_3_, int p_175872_4_, int p_175872_5_, EnumFacing p_175872_6_)
+            public static Stairs createPiece(List<StructureComponent> p_175872_0_, Random p_175872_1_, int p_175872_2_, int p_175872_3_, int p_175872_4_, int p_175872_5_, EnumFacing p_175872_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175872_2_, p_175872_3_, p_175872_4_, -2, 0, 0, 7, 11, 7, p_175872_6_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175872_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Stairs(p_175872_5_, p_175872_1_, structureboundingbox, p_175872_6_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175872_0_, structureboundingbox) == null ? new Stairs(p_175872_5_, p_175872_1_, structureboundingbox, p_175872_6_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -1217,11 +1217,11 @@
             }
         }
 
-    public static class Start extends StructureNetherBridgePieces.Crossing3
+    public static class Start extends Crossing3
         {
-            public StructureNetherBridgePieces.PieceWeight lastPlaced;
-            public List<StructureNetherBridgePieces.PieceWeight> primaryWeights;
-            public List<StructureNetherBridgePieces.PieceWeight> secondaryWeights;
+            public PieceWeight lastPlaced;
+            public List<PieceWeight> primaryWeights;
+            public List<PieceWeight> secondaryWeights;
             public List<StructureComponent> pendingChildren = Lists.<StructureComponent>newArrayList();
 
             public Start()
@@ -1231,17 +1231,17 @@
             public Start(Random p_i2059_1_, int p_i2059_2_, int p_i2059_3_)
             {
                 super(p_i2059_1_, p_i2059_2_, p_i2059_3_);
-                this.primaryWeights = Lists.<StructureNetherBridgePieces.PieceWeight>newArrayList();
+                this.primaryWeights = Lists.<PieceWeight>newArrayList();
 
-                for (StructureNetherBridgePieces.PieceWeight structurenetherbridgepieces$pieceweight : StructureNetherBridgePieces.PRIMARY_COMPONENTS)
+                for (PieceWeight structurenetherbridgepieces$pieceweight : StructureNetherBridgePieces.PRIMARY_COMPONENTS)
                 {
                     structurenetherbridgepieces$pieceweight.placeCount = 0;
                     this.primaryWeights.add(structurenetherbridgepieces$pieceweight);
                 }
 
-                this.secondaryWeights = Lists.<StructureNetherBridgePieces.PieceWeight>newArrayList();
+                this.secondaryWeights = Lists.<PieceWeight>newArrayList();
 
-                for (StructureNetherBridgePieces.PieceWeight structurenetherbridgepieces$pieceweight1 : StructureNetherBridgePieces.SECONDARY_COMPONENTS)
+                for (PieceWeight structurenetherbridgepieces$pieceweight1 : StructureNetherBridgePieces.SECONDARY_COMPONENTS)
                 {
                     structurenetherbridgepieces$pieceweight1.placeCount = 0;
                     this.secondaryWeights.add(structurenetherbridgepieces$pieceweight1);
@@ -1249,7 +1249,7 @@
             }
         }
 
-    public static class Straight extends StructureNetherBridgePieces.Piece
+    public static class Straight extends Piece
         {
             public Straight()
             {
@@ -1264,13 +1264,13 @@
 
             public void buildComponent(StructureComponent componentIn, List<StructureComponent> listIn, Random rand)
             {
-                this.getNextComponentNormal((StructureNetherBridgePieces.Start)componentIn, listIn, rand, 1, 3, false);
+                this.getNextComponentNormal((Start)componentIn, listIn, rand, 1, 3, false);
             }
 
-            public static StructureNetherBridgePieces.Straight createPiece(List<StructureComponent> p_175882_0_, Random p_175882_1_, int p_175882_2_, int p_175882_3_, int p_175882_4_, EnumFacing p_175882_5_, int p_175882_6_)
+            public static Straight createPiece(List<StructureComponent> p_175882_0_, Random p_175882_1_, int p_175882_2_, int p_175882_3_, int p_175882_4_, EnumFacing p_175882_5_, int p_175882_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175882_2_, p_175882_3_, p_175882_4_, -1, -3, 0, 5, 10, 19, p_175882_5_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175882_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Straight(p_175882_6_, p_175882_1_, structureboundingbox, p_175882_5_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175882_0_, structureboundingbox) == null ? new Straight(p_175882_6_, p_175882_1_, structureboundingbox, p_175882_5_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
@@ -1305,7 +1305,7 @@
             }
         }
 
-    public static class Throne extends StructureNetherBridgePieces.Piece
+    public static class Throne extends Piece
         {
             private boolean hasSpawner;
 
@@ -1332,10 +1332,10 @@
                 tagCompound.setBoolean("Mob", this.hasSpawner);
             }
 
-            public static StructureNetherBridgePieces.Throne createPiece(List<StructureComponent> p_175874_0_, Random p_175874_1_, int p_175874_2_, int p_175874_3_, int p_175874_4_, int p_175874_5_, EnumFacing p_175874_6_)
+            public static Throne createPiece(List<StructureComponent> p_175874_0_, Random p_175874_1_, int p_175874_2_, int p_175874_3_, int p_175874_4_, int p_175874_5_, EnumFacing p_175874_6_)
             {
                 StructureBoundingBox structureboundingbox = StructureBoundingBox.getComponentToAddBoundingBox(p_175874_2_, p_175874_3_, p_175874_4_, -2, 0, 0, 7, 8, 9, p_175874_6_);
-                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175874_0_, structureboundingbox) == null ? new StructureNetherBridgePieces.Throne(p_175874_5_, p_175874_1_, structureboundingbox, p_175874_6_) : null;
+                return isAboveGround(structureboundingbox) && StructureComponent.findIntersecting(p_175874_0_, structureboundingbox) == null ? new Throne(p_175874_5_, p_175874_1_, structureboundingbox, p_175874_6_) : null;
             }
 
             public boolean addComponentParts(World worldIn, Random randomIn, StructureBoundingBox structureBoundingBoxIn)
