--- ../src-base/minecraft/net/minecraft/world/gen/structure/WoodlandMansion.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/WoodlandMansion.java
@@ -76,7 +76,7 @@
 
     protected StructureStart getStructureStart(int chunkX, int chunkZ)
     {
-        return new WoodlandMansion.Start(this.world, this.provider, this.rand, chunkX, chunkZ);
+        return new Start(this.world, this.provider, this.rand, chunkX, chunkZ);
     }
 
     public static class Start extends StructureStart
