--- ../src-base/minecraft/net/minecraft/world/gen/structure/WoodlandMansionPieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/WoodlandMansionPieces.java
@@ -29,17 +29,17 @@
 {
     public static void registerWoodlandMansionPieces()
     {
-        MapGenStructureIO.registerStructureComponent(WoodlandMansionPieces.MansionTemplate.class, "WMP");
+        MapGenStructureIO.registerStructureComponent(MansionTemplate.class, "WMP");
     }
 
-    public static void generateMansion(TemplateManager p_191152_0_, BlockPos p_191152_1_, Rotation p_191152_2_, List<WoodlandMansionPieces.MansionTemplate> p_191152_3_, Random p_191152_4_)
+    public static void generateMansion(TemplateManager p_191152_0_, BlockPos p_191152_1_, Rotation p_191152_2_, List<MansionTemplate> p_191152_3_, Random p_191152_4_)
     {
-        WoodlandMansionPieces.Grid woodlandmansionpieces$grid = new WoodlandMansionPieces.Grid(p_191152_4_);
-        WoodlandMansionPieces.Placer woodlandmansionpieces$placer = new WoodlandMansionPieces.Placer(p_191152_0_, p_191152_4_);
+        Grid woodlandmansionpieces$grid = new Grid(p_191152_4_);
+        Placer woodlandmansionpieces$placer = new Placer(p_191152_0_, p_191152_4_);
         woodlandmansionpieces$placer.createMansion(p_191152_1_, p_191152_2_, p_191152_3_, woodlandmansionpieces$grid);
     }
 
-    static class FirstFloor extends WoodlandMansionPieces.RoomCollection
+    static class FirstFloor extends RoomCollection
         {
             private FirstFloor()
             {
@@ -84,9 +84,9 @@
     static class Grid
         {
             private final Random random;
-            private final WoodlandMansionPieces.SimpleGrid baseGrid;
-            private final WoodlandMansionPieces.SimpleGrid thirdFloorGrid;
-            private final WoodlandMansionPieces.SimpleGrid[] floorRooms;
+            private final SimpleGrid baseGrid;
+            private final SimpleGrid thirdFloorGrid;
+            private final SimpleGrid[] floorRooms;
             private final int entranceX;
             private final int entranceY;
 
@@ -96,7 +96,7 @@
                 int i = 11;
                 this.entranceX = 7;
                 this.entranceY = 4;
-                this.baseGrid = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+                this.baseGrid = new SimpleGrid(11, 11, 5);
                 this.baseGrid.set(this.entranceX, this.entranceY, this.entranceX + 1, this.entranceY + 1, 3);
                 this.baseGrid.set(this.entranceX - 1, this.entranceY, this.entranceX - 1, this.entranceY + 1, 2);
                 this.baseGrid.set(this.entranceX + 2, this.entranceY - 2, this.entranceX + 3, this.entranceY + 3, 5);
@@ -116,32 +116,32 @@
                     ;
                 }
 
-                this.floorRooms = new WoodlandMansionPieces.SimpleGrid[3];
-                this.floorRooms[0] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-                this.floorRooms[1] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-                this.floorRooms[2] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+                this.floorRooms = new SimpleGrid[3];
+                this.floorRooms[0] = new SimpleGrid(11, 11, 5);
+                this.floorRooms[1] = new SimpleGrid(11, 11, 5);
+                this.floorRooms[2] = new SimpleGrid(11, 11, 5);
                 this.identifyRooms(this.baseGrid, this.floorRooms[0]);
                 this.identifyRooms(this.baseGrid, this.floorRooms[1]);
                 this.floorRooms[0].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
                 this.floorRooms[1].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
-                this.thirdFloorGrid = new WoodlandMansionPieces.SimpleGrid(this.baseGrid.width, this.baseGrid.height, 5);
+                this.thirdFloorGrid = new SimpleGrid(this.baseGrid.width, this.baseGrid.height, 5);
                 this.setupThirdFloor();
                 this.identifyRooms(this.thirdFloorGrid, this.floorRooms[2]);
             }
 
-            public static boolean isHouse(WoodlandMansionPieces.SimpleGrid p_191109_0_, int p_191109_1_, int p_191109_2_)
+            public static boolean isHouse(SimpleGrid p_191109_0_, int p_191109_1_, int p_191109_2_)
             {
                 int i = p_191109_0_.get(p_191109_1_, p_191109_2_);
                 return i == 1 || i == 2 || i == 3 || i == 4;
             }
 
-            public boolean isRoomId(WoodlandMansionPieces.SimpleGrid p_191114_1_, int p_191114_2_, int p_191114_3_, int p_191114_4_, int p_191114_5_)
+            public boolean isRoomId(SimpleGrid p_191114_1_, int p_191114_2_, int p_191114_3_, int p_191114_4_, int p_191114_5_)
             {
                 return (this.floorRooms[p_191114_4_].get(p_191114_2_, p_191114_3_) & 65535) == p_191114_5_;
             }
 
             @Nullable
-            public EnumFacing get1x2RoomDirection(WoodlandMansionPieces.SimpleGrid p_191113_1_, int p_191113_2_, int p_191113_3_, int p_191113_4_, int p_191113_5_)
+            public EnumFacing get1x2RoomDirection(SimpleGrid p_191113_1_, int p_191113_2_, int p_191113_3_, int p_191113_4_, int p_191113_5_)
             {
                 for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL.facings())
                 {
@@ -154,7 +154,7 @@
                 return null;
             }
 
-            private void recursiveCorridor(WoodlandMansionPieces.SimpleGrid p_191110_1_, int p_191110_2_, int p_191110_3_, EnumFacing p_191110_4_, int p_191110_5_)
+            private void recursiveCorridor(SimpleGrid p_191110_1_, int p_191110_2_, int p_191110_3_, EnumFacing p_191110_4_, int p_191110_5_)
             {
                 if (p_191110_5_ > 0)
                 {
@@ -190,7 +190,7 @@
                 }
             }
 
-            private boolean cleanEdges(WoodlandMansionPieces.SimpleGrid p_191111_1_)
+            private boolean cleanEdges(SimpleGrid p_191111_1_)
             {
                 boolean flag = false;
 
@@ -235,7 +235,7 @@
             private void setupThirdFloor()
             {
                 List<Tuple<Integer, Integer>> list = Lists.<Tuple<Integer, Integer>>newArrayList();
-                WoodlandMansionPieces.SimpleGrid woodlandmansionpieces$simplegrid = this.floorRooms[1];
+                SimpleGrid woodlandmansionpieces$simplegrid = this.floorRooms[1];
 
                 for (int i = 0; i < this.thirdFloorGrid.height; ++i)
                 {
@@ -312,7 +312,7 @@
                 }
             }
 
-            private void identifyRooms(WoodlandMansionPieces.SimpleGrid p_191116_1_, WoodlandMansionPieces.SimpleGrid p_191116_2_)
+            private void identifyRooms(SimpleGrid p_191116_1_, SimpleGrid p_191116_2_)
             {
                 List<Tuple<Integer, Integer>> list = Lists.<Tuple<Integer, Integer>>newArrayList();
 
@@ -550,13 +550,13 @@
                 this.random = p_i47361_2_;
             }
 
-            public void createMansion(BlockPos p_191125_1_, Rotation p_191125_2_, List<WoodlandMansionPieces.MansionTemplate> p_191125_3_, WoodlandMansionPieces.Grid p_191125_4_)
+            public void createMansion(BlockPos p_191125_1_, Rotation p_191125_2_, List<MansionTemplate> p_191125_3_, Grid p_191125_4_)
             {
-                WoodlandMansionPieces.PlacementData woodlandmansionpieces$placementdata = new WoodlandMansionPieces.PlacementData();
+                PlacementData woodlandmansionpieces$placementdata = new PlacementData();
                 woodlandmansionpieces$placementdata.position = p_191125_1_;
                 woodlandmansionpieces$placementdata.rotation = p_191125_2_;
                 woodlandmansionpieces$placementdata.wallType = "wall_flat";
-                WoodlandMansionPieces.PlacementData woodlandmansionpieces$placementdata1 = new WoodlandMansionPieces.PlacementData();
+                PlacementData woodlandmansionpieces$placementdata1 = new PlacementData();
                 this.entrance(p_191125_3_, woodlandmansionpieces$placementdata);
                 woodlandmansionpieces$placementdata1.position = woodlandmansionpieces$placementdata.position.up(8);
                 woodlandmansionpieces$placementdata1.rotation = woodlandmansionpieces$placementdata.rotation;
@@ -567,15 +567,15 @@
                     ;
                 }
 
-                WoodlandMansionPieces.SimpleGrid woodlandmansionpieces$simplegrid = p_191125_4_.baseGrid;
-                WoodlandMansionPieces.SimpleGrid woodlandmansionpieces$simplegrid1 = p_191125_4_.thirdFloorGrid;
+                SimpleGrid woodlandmansionpieces$simplegrid = p_191125_4_.baseGrid;
+                SimpleGrid woodlandmansionpieces$simplegrid1 = p_191125_4_.thirdFloorGrid;
                 this.startX = p_191125_4_.entranceX + 1;
                 this.startY = p_191125_4_.entranceY + 1;
                 int i = p_191125_4_.entranceX + 1;
                 int j = p_191125_4_.entranceY;
                 this.traverseOuterWalls(p_191125_3_, woodlandmansionpieces$placementdata, woodlandmansionpieces$simplegrid, EnumFacing.SOUTH, this.startX, this.startY, i, j);
                 this.traverseOuterWalls(p_191125_3_, woodlandmansionpieces$placementdata1, woodlandmansionpieces$simplegrid, EnumFacing.SOUTH, this.startX, this.startY, i, j);
-                WoodlandMansionPieces.PlacementData woodlandmansionpieces$placementdata2 = new WoodlandMansionPieces.PlacementData();
+                PlacementData woodlandmansionpieces$placementdata2 = new PlacementData();
                 woodlandmansionpieces$placementdata2.position = woodlandmansionpieces$placementdata.position.up(19);
                 woodlandmansionpieces$placementdata2.rotation = woodlandmansionpieces$placementdata.rotation;
                 woodlandmansionpieces$placementdata2.wallType = "wall_window";
@@ -585,7 +585,7 @@
                 {
                     for (int l = woodlandmansionpieces$simplegrid1.width - 1; l >= 0 && !flag; --l)
                     {
-                        if (WoodlandMansionPieces.Grid.isHouse(woodlandmansionpieces$simplegrid1, l, k))
+                        if (Grid.isHouse(woodlandmansionpieces$simplegrid1, l, k))
                         {
                             woodlandmansionpieces$placementdata2.position = woodlandmansionpieces$placementdata2.position.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 8 + (k - this.startY) * 8);
                             woodlandmansionpieces$placementdata2.position = woodlandmansionpieces$placementdata2.position.offset(p_191125_2_.rotate(EnumFacing.EAST), (l - this.startX) * 8);
@@ -597,23 +597,23 @@
                 }
 
                 this.createRoof(p_191125_3_, p_191125_1_.up(16), p_191125_2_, woodlandmansionpieces$simplegrid, woodlandmansionpieces$simplegrid1);
-                this.createRoof(p_191125_3_, p_191125_1_.up(27), p_191125_2_, woodlandmansionpieces$simplegrid1, (WoodlandMansionPieces.SimpleGrid)null);
+                this.createRoof(p_191125_3_, p_191125_1_.up(27), p_191125_2_, woodlandmansionpieces$simplegrid1, (SimpleGrid)null);
 
                 if (!p_191125_3_.isEmpty())
                 {
                     ;
                 }
 
-                WoodlandMansionPieces.RoomCollection[] awoodlandmansionpieces$roomcollection = new WoodlandMansionPieces.RoomCollection[3];
-                awoodlandmansionpieces$roomcollection[0] = new WoodlandMansionPieces.FirstFloor();
-                awoodlandmansionpieces$roomcollection[1] = new WoodlandMansionPieces.SecondFloor();
-                awoodlandmansionpieces$roomcollection[2] = new WoodlandMansionPieces.ThirdFloor();
+                RoomCollection[] awoodlandmansionpieces$roomcollection = new RoomCollection[3];
+                awoodlandmansionpieces$roomcollection[0] = new FirstFloor();
+                awoodlandmansionpieces$roomcollection[1] = new SecondFloor();
+                awoodlandmansionpieces$roomcollection[2] = new ThirdFloor();
 
                 for (int l2 = 0; l2 < 3; ++l2)
                 {
                     BlockPos blockpos = p_191125_1_.up(8 * l2 + (l2 == 2 ? 3 : 0));
-                    WoodlandMansionPieces.SimpleGrid woodlandmansionpieces$simplegrid2 = p_191125_4_.floorRooms[l2];
-                    WoodlandMansionPieces.SimpleGrid woodlandmansionpieces$simplegrid3 = l2 == 2 ? woodlandmansionpieces$simplegrid1 : woodlandmansionpieces$simplegrid;
+                    SimpleGrid woodlandmansionpieces$simplegrid2 = p_191125_4_.floorRooms[l2];
+                    SimpleGrid woodlandmansionpieces$simplegrid3 = l2 == 2 ? woodlandmansionpieces$simplegrid1 : woodlandmansionpieces$simplegrid;
                     String s = l2 == 0 ? "carpet_south" : "carpet_south_2";
                     String s1 = l2 == 0 ? "carpet_west" : "carpet_west_2";
 
@@ -625,26 +625,26 @@
                             {
                                 BlockPos blockpos1 = blockpos.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 8 + (i1 - this.startY) * 8);
                                 blockpos1 = blockpos1.offset(p_191125_2_.rotate(EnumFacing.EAST), (j1 - this.startX) * 8);
-                                p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "corridor_floor", blockpos1, p_191125_2_));
+                                p_191125_3_.add(new MansionTemplate(this.templateManager, "corridor_floor", blockpos1, p_191125_2_));
 
                                 if (woodlandmansionpieces$simplegrid3.get(j1, i1 - 1) == 1 || (woodlandmansionpieces$simplegrid2.get(j1, i1 - 1) & 8388608) == 8388608)
                                 {
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "carpet_north", blockpos1.offset(p_191125_2_.rotate(EnumFacing.EAST), 1).up(), p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, "carpet_north", blockpos1.offset(p_191125_2_.rotate(EnumFacing.EAST), 1).up(), p_191125_2_));
                                 }
 
                                 if (woodlandmansionpieces$simplegrid3.get(j1 + 1, i1) == 1 || (woodlandmansionpieces$simplegrid2.get(j1 + 1, i1) & 8388608) == 8388608)
                                 {
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "carpet_east", blockpos1.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 1).offset(p_191125_2_.rotate(EnumFacing.EAST), 5).up(), p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, "carpet_east", blockpos1.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 1).offset(p_191125_2_.rotate(EnumFacing.EAST), 5).up(), p_191125_2_));
                                 }
 
                                 if (woodlandmansionpieces$simplegrid3.get(j1, i1 + 1) == 1 || (woodlandmansionpieces$simplegrid2.get(j1, i1 + 1) & 8388608) == 8388608)
                                 {
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, s, blockpos1.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 5).offset(p_191125_2_.rotate(EnumFacing.WEST), 1), p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, s, blockpos1.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 5).offset(p_191125_2_.rotate(EnumFacing.WEST), 1), p_191125_2_));
                                 }
 
                                 if (woodlandmansionpieces$simplegrid3.get(j1 - 1, i1) == 1 || (woodlandmansionpieces$simplegrid2.get(j1 - 1, i1) & 8388608) == 8388608)
                                 {
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, s1, blockpos1.offset(p_191125_2_.rotate(EnumFacing.WEST), 1).offset(p_191125_2_.rotate(EnumFacing.NORTH), 1), p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, s1, blockpos1.offset(p_191125_2_.rotate(EnumFacing.WEST), 1).offset(p_191125_2_.rotate(EnumFacing.NORTH), 1), p_191125_2_));
                                 }
                             }
                         }
@@ -693,29 +693,29 @@
                                 BlockPos blockpos2 = blockpos.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 8 + (k1 - this.startY) * 8);
                                 blockpos2 = blockpos2.offset(p_191125_2_.rotate(EnumFacing.EAST), -1 + (l1 - this.startX) * 8);
 
-                                if (WoodlandMansionPieces.Grid.isHouse(woodlandmansionpieces$simplegrid3, l1 - 1, k1) && !p_191125_4_.isRoomId(woodlandmansionpieces$simplegrid3, l1 - 1, k1, l2, k2))
+                                if (Grid.isHouse(woodlandmansionpieces$simplegrid3, l1 - 1, k1) && !p_191125_4_.isRoomId(woodlandmansionpieces$simplegrid3, l1 - 1, k1, l2, k2))
                                 {
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.WEST ? s3 : s2, blockpos2, p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.WEST ? s3 : s2, blockpos2, p_191125_2_));
                                 }
 
                                 if (woodlandmansionpieces$simplegrid3.get(l1 + 1, k1) == 1 && !flag1)
                                 {
                                     BlockPos blockpos3 = blockpos2.offset(p_191125_2_.rotate(EnumFacing.EAST), 8);
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.EAST ? s3 : s2, blockpos3, p_191125_2_));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.EAST ? s3 : s2, blockpos3, p_191125_2_));
                                 }
 
-                                if (WoodlandMansionPieces.Grid.isHouse(woodlandmansionpieces$simplegrid3, l1, k1 + 1) && !p_191125_4_.isRoomId(woodlandmansionpieces$simplegrid3, l1, k1 + 1, l2, k2))
+                                if (Grid.isHouse(woodlandmansionpieces$simplegrid3, l1, k1 + 1) && !p_191125_4_.isRoomId(woodlandmansionpieces$simplegrid3, l1, k1 + 1, l2, k2))
                                 {
                                     BlockPos blockpos4 = blockpos2.offset(p_191125_2_.rotate(EnumFacing.SOUTH), 7);
                                     blockpos4 = blockpos4.offset(p_191125_2_.rotate(EnumFacing.EAST), 7);
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.SOUTH ? s3 : s2, blockpos4, p_191125_2_.add(Rotation.CLOCKWISE_90)));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.SOUTH ? s3 : s2, blockpos4, p_191125_2_.add(Rotation.CLOCKWISE_90)));
                                 }
 
                                 if (woodlandmansionpieces$simplegrid3.get(l1, k1 - 1) == 1 && !flag1)
                                 {
                                     BlockPos blockpos5 = blockpos2.offset(p_191125_2_.rotate(EnumFacing.NORTH), 1);
                                     blockpos5 = blockpos5.offset(p_191125_2_.rotate(EnumFacing.EAST), 7);
-                                    p_191125_3_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.NORTH ? s3 : s2, blockpos5, p_191125_2_.add(Rotation.CLOCKWISE_90)));
+                                    p_191125_3_.add(new MansionTemplate(this.templateManager, enumfacing1 == EnumFacing.NORTH ? s3 : s2, blockpos5, p_191125_2_.add(Rotation.CLOCKWISE_90)));
                                 }
 
                                 if (j2 == 65536)
@@ -749,7 +749,7 @@
                 }
             }
 
-            private void traverseOuterWalls(List<WoodlandMansionPieces.MansionTemplate> p_191130_1_, WoodlandMansionPieces.PlacementData p_191130_2_, WoodlandMansionPieces.SimpleGrid p_191130_3_, EnumFacing p_191130_4_, int p_191130_5_, int p_191130_6_, int p_191130_7_, int p_191130_8_)
+            private void traverseOuterWalls(List<MansionTemplate> p_191130_1_, PlacementData p_191130_2_, SimpleGrid p_191130_3_, EnumFacing p_191130_4_, int p_191130_5_, int p_191130_6_, int p_191130_7_, int p_191130_8_)
             {
                 int i = p_191130_5_;
                 int j = p_191130_6_;
@@ -757,7 +757,7 @@
 
                 while (true)
                 {
-                    if (!WoodlandMansionPieces.Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ()))
+                    if (!Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ()))
                     {
                         this.traverseTurn(p_191130_1_, p_191130_2_);
                         p_191130_4_ = p_191130_4_.rotateY();
@@ -767,7 +767,7 @@
                             this.traverseWallPiece(p_191130_1_, p_191130_2_);
                         }
                     }
-                    else if (WoodlandMansionPieces.Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ()) && WoodlandMansionPieces.Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX() + p_191130_4_.rotateYCCW().getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ() + p_191130_4_.rotateYCCW().getFrontOffsetZ()))
+                    else if (Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ()) && Grid.isHouse(p_191130_3_, i + p_191130_4_.getFrontOffsetX() + p_191130_4_.rotateYCCW().getFrontOffsetX(), j + p_191130_4_.getFrontOffsetZ() + p_191130_4_.rotateYCCW().getFrontOffsetZ()))
                     {
                         this.traverseInnerTurn(p_191130_1_, p_191130_2_);
                         i += p_191130_4_.getFrontOffsetX();
@@ -792,7 +792,7 @@
                 }
             }
 
-            private void createRoof(List<WoodlandMansionPieces.MansionTemplate> p_191123_1_, BlockPos p_191123_2_, Rotation p_191123_3_, WoodlandMansionPieces.SimpleGrid p_191123_4_, @Nullable WoodlandMansionPieces.SimpleGrid p_191123_5_)
+            private void createRoof(List<MansionTemplate> p_191123_1_, BlockPos p_191123_2_, Rotation p_191123_3_, SimpleGrid p_191123_4_, @Nullable WoodlandMansionPieces.SimpleGrid p_191123_5_)
             {
                 for (int i = 0; i < p_191123_4_.height; ++i)
                 {
@@ -800,36 +800,36 @@
                     {
                         BlockPos lvt_8_3_ = p_191123_2_.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 8 + (i - this.startY) * 8);
                         lvt_8_3_ = lvt_8_3_.offset(p_191123_3_.rotate(EnumFacing.EAST), (j - this.startX) * 8);
-                        boolean flag = p_191123_5_ != null && WoodlandMansionPieces.Grid.isHouse(p_191123_5_, j, i);
+                        boolean flag = p_191123_5_ != null && Grid.isHouse(p_191123_5_, j, i);
 
-                        if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j, i) && !flag)
+                        if (Grid.isHouse(p_191123_4_, j, i) && !flag)
                         {
-                            p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof", lvt_8_3_.up(3), p_191123_3_));
+                            p_191123_1_.add(new MansionTemplate(this.templateManager, "roof", lvt_8_3_.up(3), p_191123_3_));
 
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j + 1, i))
+                            if (!Grid.isHouse(p_191123_4_, j + 1, i))
                             {
                                 BlockPos blockpos1 = lvt_8_3_.offset(p_191123_3_.rotate(EnumFacing.EAST), 6);
-                                p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_front", blockpos1, p_191123_3_));
+                                p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_front", blockpos1, p_191123_3_));
                             }
 
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j - 1, i))
+                            if (!Grid.isHouse(p_191123_4_, j - 1, i))
                             {
                                 BlockPos blockpos5 = lvt_8_3_.offset(p_191123_3_.rotate(EnumFacing.EAST), 0);
                                 blockpos5 = blockpos5.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 7);
-                                p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_front", blockpos5, p_191123_3_.add(Rotation.CLOCKWISE_180)));
+                                p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_front", blockpos5, p_191123_3_.add(Rotation.CLOCKWISE_180)));
                             }
 
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j, i - 1))
+                            if (!Grid.isHouse(p_191123_4_, j, i - 1))
                             {
                                 BlockPos blockpos6 = lvt_8_3_.offset(p_191123_3_.rotate(EnumFacing.WEST), 1);
-                                p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_front", blockpos6, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                                p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_front", blockpos6, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                             }
 
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j, i + 1))
+                            if (!Grid.isHouse(p_191123_4_, j, i + 1))
                             {
                                 BlockPos blockpos7 = lvt_8_3_.offset(p_191123_3_.rotate(EnumFacing.EAST), 6);
                                 blockpos7 = blockpos7.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 6);
-                                p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_front", blockpos7, p_191123_3_.add(Rotation.CLOCKWISE_90)));
+                                p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_front", blockpos7, p_191123_3_.add(Rotation.CLOCKWISE_90)));
                             }
                         }
                     }
@@ -843,68 +843,68 @@
                         {
                             BlockPos blockpos3 = p_191123_2_.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 8 + (k - this.startY) * 8);
                             blockpos3 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.EAST), (i1 - this.startX) * 8);
-                            boolean flag1 = WoodlandMansionPieces.Grid.isHouse(p_191123_5_, i1, k);
+                            boolean flag1 = Grid.isHouse(p_191123_5_, i1, k);
 
-                            if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k) && flag1)
+                            if (Grid.isHouse(p_191123_4_, i1, k) && flag1)
                             {
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1 + 1, k))
+                                if (!Grid.isHouse(p_191123_4_, i1 + 1, k))
                                 {
                                     BlockPos blockpos8 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.EAST), 7);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall", blockpos8, p_191123_3_));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall", blockpos8, p_191123_3_));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1 - 1, k))
+                                if (!Grid.isHouse(p_191123_4_, i1 - 1, k))
                                 {
                                     BlockPos blockpos9 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.WEST), 1);
                                     blockpos9 = blockpos9.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 6);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall", blockpos9, p_191123_3_.add(Rotation.CLOCKWISE_180)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall", blockpos9, p_191123_3_.add(Rotation.CLOCKWISE_180)));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k - 1))
+                                if (!Grid.isHouse(p_191123_4_, i1, k - 1))
                                 {
                                     BlockPos blockpos10 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.WEST), 0);
                                     blockpos10 = blockpos10.offset(p_191123_3_.rotate(EnumFacing.NORTH), 1);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall", blockpos10, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall", blockpos10, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k + 1))
+                                if (!Grid.isHouse(p_191123_4_, i1, k + 1))
                                 {
                                     BlockPos blockpos11 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.EAST), 6);
                                     blockpos11 = blockpos11.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 7);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall", blockpos11, p_191123_3_.add(Rotation.CLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall", blockpos11, p_191123_3_.add(Rotation.CLOCKWISE_90)));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1 + 1, k))
+                                if (!Grid.isHouse(p_191123_4_, i1 + 1, k))
                                 {
-                                    if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k - 1))
+                                    if (!Grid.isHouse(p_191123_4_, i1, k - 1))
                                     {
                                         BlockPos blockpos12 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.EAST), 7);
                                         blockpos12 = blockpos12.offset(p_191123_3_.rotate(EnumFacing.NORTH), 2);
-                                        p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall_corner", blockpos12, p_191123_3_));
+                                        p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall_corner", blockpos12, p_191123_3_));
                                     }
 
-                                    if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k + 1))
+                                    if (!Grid.isHouse(p_191123_4_, i1, k + 1))
                                     {
                                         BlockPos blockpos13 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.EAST), 8);
                                         blockpos13 = blockpos13.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 7);
-                                        p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall_corner", blockpos13, p_191123_3_.add(Rotation.CLOCKWISE_90)));
+                                        p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall_corner", blockpos13, p_191123_3_.add(Rotation.CLOCKWISE_90)));
                                     }
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1 - 1, k))
+                                if (!Grid.isHouse(p_191123_4_, i1 - 1, k))
                                 {
-                                    if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k - 1))
+                                    if (!Grid.isHouse(p_191123_4_, i1, k - 1))
                                     {
                                         BlockPos blockpos14 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.WEST), 2);
                                         blockpos14 = blockpos14.offset(p_191123_3_.rotate(EnumFacing.NORTH), 1);
-                                        p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall_corner", blockpos14, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                                        p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall_corner", blockpos14, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                                     }
 
-                                    if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, i1, k + 1))
+                                    if (!Grid.isHouse(p_191123_4_, i1, k + 1))
                                     {
                                         BlockPos blockpos15 = blockpos3.offset(p_191123_3_.rotate(EnumFacing.WEST), 1);
                                         blockpos15 = blockpos15.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 8);
-                                        p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "small_wall_corner", blockpos15, p_191123_3_.add(Rotation.CLOCKWISE_180)));
+                                        p_191123_1_.add(new MansionTemplate(this.templateManager, "small_wall_corner", blockpos15, p_191123_3_.add(Rotation.CLOCKWISE_180)));
                                     }
                                 }
                             }
@@ -918,62 +918,62 @@
                     {
                         BlockPos blockpos4 = p_191123_2_.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 8 + (l - this.startY) * 8);
                         blockpos4 = blockpos4.offset(p_191123_3_.rotate(EnumFacing.EAST), (j1 - this.startX) * 8);
-                        boolean flag2 = p_191123_5_ != null && WoodlandMansionPieces.Grid.isHouse(p_191123_5_, j1, l);
+                        boolean flag2 = p_191123_5_ != null && Grid.isHouse(p_191123_5_, j1, l);
 
-                        if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1, l) && !flag2)
+                        if (Grid.isHouse(p_191123_4_, j1, l) && !flag2)
                         {
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 + 1, l))
+                            if (!Grid.isHouse(p_191123_4_, j1 + 1, l))
                             {
                                 BlockPos blockpos16 = blockpos4.offset(p_191123_3_.rotate(EnumFacing.EAST), 6);
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1, l + 1))
+                                if (!Grid.isHouse(p_191123_4_, j1, l + 1))
                                 {
                                     BlockPos blockpos2 = blockpos16.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 6);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_corner", blockpos2, p_191123_3_));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_corner", blockpos2, p_191123_3_));
                                 }
-                                else if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 + 1, l + 1))
+                                else if (Grid.isHouse(p_191123_4_, j1 + 1, l + 1))
                                 {
                                     BlockPos blockpos18 = blockpos16.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 5);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_inner_corner", blockpos18, p_191123_3_));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_inner_corner", blockpos18, p_191123_3_));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1, l - 1))
+                                if (!Grid.isHouse(p_191123_4_, j1, l - 1))
                                 {
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_corner", blockpos16, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_corner", blockpos16, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                                 }
-                                else if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 + 1, l - 1))
+                                else if (Grid.isHouse(p_191123_4_, j1 + 1, l - 1))
                                 {
                                     BlockPos blockpos19 = blockpos4.offset(p_191123_3_.rotate(EnumFacing.EAST), 9);
                                     blockpos19 = blockpos19.offset(p_191123_3_.rotate(EnumFacing.NORTH), 2);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_inner_corner", blockpos19, p_191123_3_.add(Rotation.CLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_inner_corner", blockpos19, p_191123_3_.add(Rotation.CLOCKWISE_90)));
                                 }
                             }
 
-                            if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 - 1, l))
+                            if (!Grid.isHouse(p_191123_4_, j1 - 1, l))
                             {
                                 BlockPos blockpos17 = blockpos4.offset(p_191123_3_.rotate(EnumFacing.EAST), 0);
                                 blockpos17 = blockpos17.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 0);
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1, l + 1))
+                                if (!Grid.isHouse(p_191123_4_, j1, l + 1))
                                 {
                                     BlockPos blockpos20 = blockpos17.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 6);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_corner", blockpos20, p_191123_3_.add(Rotation.CLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_corner", blockpos20, p_191123_3_.add(Rotation.CLOCKWISE_90)));
                                 }
-                                else if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 - 1, l + 1))
+                                else if (Grid.isHouse(p_191123_4_, j1 - 1, l + 1))
                                 {
                                     BlockPos blockpos21 = blockpos17.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 8);
                                     blockpos21 = blockpos21.offset(p_191123_3_.rotate(EnumFacing.WEST), 3);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_inner_corner", blockpos21, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_inner_corner", blockpos21, p_191123_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                                 }
 
-                                if (!WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1, l - 1))
+                                if (!Grid.isHouse(p_191123_4_, j1, l - 1))
                                 {
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_corner", blockpos17, p_191123_3_.add(Rotation.CLOCKWISE_180)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_corner", blockpos17, p_191123_3_.add(Rotation.CLOCKWISE_180)));
                                 }
-                                else if (WoodlandMansionPieces.Grid.isHouse(p_191123_4_, j1 - 1, l - 1))
+                                else if (Grid.isHouse(p_191123_4_, j1 - 1, l - 1))
                                 {
                                     BlockPos blockpos22 = blockpos17.offset(p_191123_3_.rotate(EnumFacing.SOUTH), 1);
-                                    p_191123_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "roof_inner_corner", blockpos22, p_191123_3_.add(Rotation.CLOCKWISE_180)));
+                                    p_191123_1_.add(new MansionTemplate(this.templateManager, "roof_inner_corner", blockpos22, p_191123_3_.add(Rotation.CLOCKWISE_180)));
                                 }
                             }
                         }
@@ -981,36 +981,36 @@
                 }
             }
 
-            private void entrance(List<WoodlandMansionPieces.MansionTemplate> p_191133_1_, WoodlandMansionPieces.PlacementData p_191133_2_)
+            private void entrance(List<MansionTemplate> p_191133_1_, PlacementData p_191133_2_)
             {
                 EnumFacing enumfacing = p_191133_2_.rotation.rotate(EnumFacing.WEST);
-                p_191133_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "entrance", p_191133_2_.position.offset(enumfacing, 9), p_191133_2_.rotation));
+                p_191133_1_.add(new MansionTemplate(this.templateManager, "entrance", p_191133_2_.position.offset(enumfacing, 9), p_191133_2_.rotation));
                 p_191133_2_.position = p_191133_2_.position.offset(p_191133_2_.rotation.rotate(EnumFacing.SOUTH), 16);
             }
 
-            private void traverseWallPiece(List<WoodlandMansionPieces.MansionTemplate> p_191131_1_, WoodlandMansionPieces.PlacementData p_191131_2_)
+            private void traverseWallPiece(List<MansionTemplate> p_191131_1_, PlacementData p_191131_2_)
             {
-                p_191131_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191131_2_.wallType, p_191131_2_.position.offset(p_191131_2_.rotation.rotate(EnumFacing.EAST), 7), p_191131_2_.rotation));
+                p_191131_1_.add(new MansionTemplate(this.templateManager, p_191131_2_.wallType, p_191131_2_.position.offset(p_191131_2_.rotation.rotate(EnumFacing.EAST), 7), p_191131_2_.rotation));
                 p_191131_2_.position = p_191131_2_.position.offset(p_191131_2_.rotation.rotate(EnumFacing.SOUTH), 8);
             }
 
-            private void traverseTurn(List<WoodlandMansionPieces.MansionTemplate> p_191124_1_, WoodlandMansionPieces.PlacementData p_191124_2_)
+            private void traverseTurn(List<MansionTemplate> p_191124_1_, PlacementData p_191124_2_)
             {
                 p_191124_2_.position = p_191124_2_.position.offset(p_191124_2_.rotation.rotate(EnumFacing.SOUTH), -1);
-                p_191124_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, "wall_corner", p_191124_2_.position, p_191124_2_.rotation));
+                p_191124_1_.add(new MansionTemplate(this.templateManager, "wall_corner", p_191124_2_.position, p_191124_2_.rotation));
                 p_191124_2_.position = p_191124_2_.position.offset(p_191124_2_.rotation.rotate(EnumFacing.SOUTH), -7);
                 p_191124_2_.position = p_191124_2_.position.offset(p_191124_2_.rotation.rotate(EnumFacing.WEST), -6);
                 p_191124_2_.rotation = p_191124_2_.rotation.add(Rotation.CLOCKWISE_90);
             }
 
-            private void traverseInnerTurn(List<WoodlandMansionPieces.MansionTemplate> p_191126_1_, WoodlandMansionPieces.PlacementData p_191126_2_)
+            private void traverseInnerTurn(List<MansionTemplate> p_191126_1_, PlacementData p_191126_2_)
             {
                 p_191126_2_.position = p_191126_2_.position.offset(p_191126_2_.rotation.rotate(EnumFacing.SOUTH), 6);
                 p_191126_2_.position = p_191126_2_.position.offset(p_191126_2_.rotation.rotate(EnumFacing.EAST), 8);
                 p_191126_2_.rotation = p_191126_2_.rotation.add(Rotation.COUNTERCLOCKWISE_90);
             }
 
-            private void addRoom1x1(List<WoodlandMansionPieces.MansionTemplate> p_191129_1_, BlockPos p_191129_2_, Rotation p_191129_3_, EnumFacing p_191129_4_, WoodlandMansionPieces.RoomCollection p_191129_5_)
+            private void addRoom1x1(List<MansionTemplate> p_191129_1_, BlockPos p_191129_2_, Rotation p_191129_3_, EnumFacing p_191129_4_, RoomCollection p_191129_5_)
             {
                 Rotation rotation = Rotation.NONE;
                 String s = p_191129_5_.get1x1(this.random);
@@ -1039,92 +1039,92 @@
                 rotation = rotation.add(p_191129_3_);
                 blockpos = blockpos.rotate(p_191129_3_);
                 BlockPos blockpos1 = p_191129_2_.add(blockpos.getX(), 0, blockpos.getZ());
-                p_191129_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, s, blockpos1, rotation));
+                p_191129_1_.add(new MansionTemplate(this.templateManager, s, blockpos1, rotation));
             }
 
-            private void addRoom1x2(List<WoodlandMansionPieces.MansionTemplate> p_191132_1_, BlockPos p_191132_2_, Rotation p_191132_3_, EnumFacing p_191132_4_, EnumFacing p_191132_5_, WoodlandMansionPieces.RoomCollection p_191132_6_, boolean p_191132_7_)
+            private void addRoom1x2(List<MansionTemplate> p_191132_1_, BlockPos p_191132_2_, Rotation p_191132_3_, EnumFacing p_191132_4_, EnumFacing p_191132_5_, RoomCollection p_191132_6_, boolean p_191132_7_)
             {
                 if (p_191132_5_ == EnumFacing.EAST && p_191132_4_ == EnumFacing.SOUTH)
                 {
                     BlockPos blockpos13 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos13, p_191132_3_));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos13, p_191132_3_));
                 }
                 else if (p_191132_5_ == EnumFacing.EAST && p_191132_4_ == EnumFacing.NORTH)
                 {
                     BlockPos blockpos12 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
                     blockpos12 = blockpos12.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 6);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos12, p_191132_3_, Mirror.LEFT_RIGHT));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos12, p_191132_3_, Mirror.LEFT_RIGHT));
                 }
                 else if (p_191132_5_ == EnumFacing.WEST && p_191132_4_ == EnumFacing.NORTH)
                 {
                     BlockPos blockpos11 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 7);
                     blockpos11 = blockpos11.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 6);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos11, p_191132_3_.add(Rotation.CLOCKWISE_180)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos11, p_191132_3_.add(Rotation.CLOCKWISE_180)));
                 }
                 else if (p_191132_5_ == EnumFacing.WEST && p_191132_4_ == EnumFacing.SOUTH)
                 {
                     BlockPos blockpos10 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 7);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos10, p_191132_3_, Mirror.FRONT_BACK));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos10, p_191132_3_, Mirror.FRONT_BACK));
                 }
                 else if (p_191132_5_ == EnumFacing.SOUTH && p_191132_4_ == EnumFacing.EAST)
                 {
                     BlockPos blockpos9 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos9, p_191132_3_.add(Rotation.CLOCKWISE_90), Mirror.LEFT_RIGHT));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos9, p_191132_3_.add(Rotation.CLOCKWISE_90), Mirror.LEFT_RIGHT));
                 }
                 else if (p_191132_5_ == EnumFacing.SOUTH && p_191132_4_ == EnumFacing.WEST)
                 {
                     BlockPos blockpos8 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 7);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos8, p_191132_3_.add(Rotation.CLOCKWISE_90)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos8, p_191132_3_.add(Rotation.CLOCKWISE_90)));
                 }
                 else if (p_191132_5_ == EnumFacing.NORTH && p_191132_4_ == EnumFacing.WEST)
                 {
                     BlockPos blockpos7 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 7);
                     blockpos7 = blockpos7.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 6);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos7, p_191132_3_.add(Rotation.CLOCKWISE_90), Mirror.FRONT_BACK));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos7, p_191132_3_.add(Rotation.CLOCKWISE_90), Mirror.FRONT_BACK));
                 }
                 else if (p_191132_5_ == EnumFacing.NORTH && p_191132_4_ == EnumFacing.EAST)
                 {
                     BlockPos blockpos6 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
                     blockpos6 = blockpos6.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 6);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos6, p_191132_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2SideEntrance(this.random, p_191132_7_), blockpos6, p_191132_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                 }
                 else if (p_191132_5_ == EnumFacing.SOUTH && p_191132_4_ == EnumFacing.NORTH)
                 {
                     BlockPos blockpos5 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
                     blockpos5 = blockpos5.offset(p_191132_3_.rotate(EnumFacing.NORTH), 8);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos5, p_191132_3_));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos5, p_191132_3_));
                 }
                 else if (p_191132_5_ == EnumFacing.NORTH && p_191132_4_ == EnumFacing.SOUTH)
                 {
                     BlockPos blockpos4 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 7);
                     blockpos4 = blockpos4.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 14);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos4, p_191132_3_.add(Rotation.CLOCKWISE_180)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos4, p_191132_3_.add(Rotation.CLOCKWISE_180)));
                 }
                 else if (p_191132_5_ == EnumFacing.WEST && p_191132_4_ == EnumFacing.EAST)
                 {
                     BlockPos blockpos3 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 15);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos3, p_191132_3_.add(Rotation.CLOCKWISE_90)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos3, p_191132_3_.add(Rotation.CLOCKWISE_90)));
                 }
                 else if (p_191132_5_ == EnumFacing.EAST && p_191132_4_ == EnumFacing.WEST)
                 {
                     BlockPos blockpos2 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.WEST), 7);
                     blockpos2 = blockpos2.offset(p_191132_3_.rotate(EnumFacing.SOUTH), 6);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos2, p_191132_3_.add(Rotation.COUNTERCLOCKWISE_90)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2FrontEntrance(this.random, p_191132_7_), blockpos2, p_191132_3_.add(Rotation.COUNTERCLOCKWISE_90)));
                 }
                 else if (p_191132_5_ == EnumFacing.UP && p_191132_4_ == EnumFacing.EAST)
                 {
                     BlockPos blockpos1 = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 15);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2Secret(this.random), blockpos1, p_191132_3_.add(Rotation.CLOCKWISE_90)));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2Secret(this.random), blockpos1, p_191132_3_.add(Rotation.CLOCKWISE_90)));
                 }
                 else if (p_191132_5_ == EnumFacing.UP && p_191132_4_ == EnumFacing.SOUTH)
                 {
                     BlockPos blockpos = p_191132_2_.offset(p_191132_3_.rotate(EnumFacing.EAST), 1);
                     blockpos = blockpos.offset(p_191132_3_.rotate(EnumFacing.NORTH), 0);
-                    p_191132_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191132_6_.get1x2Secret(this.random), blockpos, p_191132_3_));
+                    p_191132_1_.add(new MansionTemplate(this.templateManager, p_191132_6_.get1x2Secret(this.random), blockpos, p_191132_3_));
                 }
             }
 
-            private void addRoom2x2(List<WoodlandMansionPieces.MansionTemplate> p_191127_1_, BlockPos p_191127_2_, Rotation p_191127_3_, EnumFacing p_191127_4_, EnumFacing p_191127_5_, WoodlandMansionPieces.RoomCollection p_191127_6_)
+            private void addRoom2x2(List<MansionTemplate> p_191127_1_, BlockPos p_191127_2_, Rotation p_191127_3_, EnumFacing p_191127_4_, EnumFacing p_191127_5_, RoomCollection p_191127_6_)
             {
                 int i = 0;
                 int j = 0;
@@ -1181,13 +1181,13 @@
 
                 BlockPos blockpos = p_191127_2_.offset(p_191127_3_.rotate(EnumFacing.EAST), i);
                 blockpos = blockpos.offset(p_191127_3_.rotate(EnumFacing.SOUTH), j);
-                p_191127_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191127_6_.get2x2(this.random), blockpos, rotation, mirror));
+                p_191127_1_.add(new MansionTemplate(this.templateManager, p_191127_6_.get2x2(this.random), blockpos, rotation, mirror));
             }
 
-            private void addRoom2x2Secret(List<WoodlandMansionPieces.MansionTemplate> p_191128_1_, BlockPos p_191128_2_, Rotation p_191128_3_, WoodlandMansionPieces.RoomCollection p_191128_4_)
+            private void addRoom2x2Secret(List<MansionTemplate> p_191128_1_, BlockPos p_191128_2_, Rotation p_191128_3_, RoomCollection p_191128_4_)
             {
                 BlockPos blockpos = p_191128_2_.offset(p_191128_3_.rotate(EnumFacing.EAST), 1);
-                p_191128_1_.add(new WoodlandMansionPieces.MansionTemplate(this.templateManager, p_191128_4_.get2x2Secret(this.random), blockpos, p_191128_3_, Mirror.NONE));
+                p_191128_1_.add(new MansionTemplate(this.templateManager, p_191128_4_.get2x2Secret(this.random), blockpos, p_191128_3_, Mirror.NONE));
             }
         }
 
@@ -1212,7 +1212,7 @@
             public abstract String get2x2Secret(Random p_191103_1_);
         }
 
-    static class SecondFloor extends WoodlandMansionPieces.RoomCollection
+    static class SecondFloor extends RoomCollection
         {
             private SecondFloor()
             {
@@ -1307,7 +1307,7 @@
             }
         }
 
-    static class ThirdFloor extends WoodlandMansionPieces.SecondFloor
+    static class ThirdFloor extends SecondFloor
         {
             private ThirdFloor()
             {
