--- ../src-base/minecraft/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
@@ -85,7 +85,8 @@
     {
         if (this.getModifier(modifier.getID()) != null)
         {
-            throw new IllegalArgumentException("Modifier is already applied on this attribute!");
+            new IllegalArgumentException("Modifier is already applied on this attribute!").printStackTrace();
+            // throw new IllegalArgumentException("Modifier is already applied on this attribute!");
         }
         else
         {
