--- ../src-base/minecraft/net/minecraft/entity/ai/attributes/RangedAttribute.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/attributes/RangedAttribute.java
@@ -6,7 +6,7 @@
 public class RangedAttribute extends BaseAttribute
 {
     private final double minimumValue;
-    private final double maximumValue;
+    public double maximumValue;
     private String description;
 
     public RangedAttribute(@Nullable IAttribute parentIn, String unlocalizedNameIn, double defaultValue, double minimumValueIn, double maximumValueIn)
