--- ../src-base/minecraft/net/minecraft/stats/StatisticsManager.java
+++ ../src-work/minecraft/net/minecraft/stats/StatisticsManager.java
@@ -26,6 +26,12 @@
     {
         if (!stat.isAchievement() || this.canUnlockAchievement((Achievement)stat))
         {
+            // CraftBukkit start - fire Statistic events
+            org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.handleStatisticsIncrease(player, stat, this.readStat(stat), amount);
+            if (cancellable != null && cancellable.isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             this.unlockAchievement(player, stat, this.readStat(stat) + amount);
         }
     }
