--- a/net/minecraft/ChatFormatting.java
+++ b/net/minecraft/ChatFormatting.java
@@ -42,7 +_,7 @@
    }));
    private static final Pattern f_126620_ = Pattern.compile("(?i)\u00a7[0-9A-FK-OR]");
    private final String f_126621_;
-   private final char f_126622_;
+   public final char f_126622_;
    private final boolean f_126592_;
    private final String f_126593_;
    private final int f_126594_;
