--- a/net/minecraft/ChatFormatting.java
+++ b/net/minecraft/ChatFormatting.java
@@ -45,7 +_,7 @@
    }));
    private static final Pattern f_126620_ = Pattern.compile("(?i)\u00a7[0-9A-FK-OR]");
    private final String f_126621_;
-   private final char f_126622_;
+   public final char f_126622_; // CatServer - private -> public
    private final boolean f_126592_;
    private final String f_126593_;
    private final int f_126594_;
@@ -111,6 +_,18 @@
    public static ChatFormatting m_126657_(@Nullable String p_126658_) {
       return p_126658_ == null ? null : f_126619_.get(m_126662_(p_126658_));
    }
+
+   // Paper start
+   @Nullable public static ChatFormatting getByHexValue(int i) {
+      for (ChatFormatting value : values()) {
+         if (value.m_126665_() != null && value.m_126665_() == i) {
+            return value;
+         }
+      }
+
+      return null;
+   }
+   // Paper end
 
    @Nullable
    public static ChatFormatting m_126647_(int p_126648_) {
