--- a/net/minecraft/server/network/LegacyQueryHandler.java
+++ b/net/minecraft/server/network/LegacyQueryHandler.java
@@ -30,10 +_,11 @@
             InetSocketAddress inetsocketaddress = (InetSocketAddress)p_9686_.channel().remoteAddress();
             MinecraftServer minecraftserver = this.f_9676_.m_9722_();
             int i = bytebuf.readableBytes();
+            org.bukkit.event.server.ServerListPingEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callServerListPingEvent(minecraftserver.server, inetsocketaddress.getAddress(), minecraftserver.m_129916_(), minecraftserver.m_7416_(), minecraftserver.m_7418_()); // CraftBukkit
             switch(i) {
             case 0:
                f_9675_.debug("Ping: (<1.3.x) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s2 = String.format("%s\u00a7%d\u00a7%d", minecraftserver.m_129916_(), minecraftserver.m_7416_(), minecraftserver.m_7418_());
+               String s2 = String.format("%s\u00a7%d\u00a7%d", event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                this.m_9680_(p_9686_, this.m_9683_(s2));
                break;
             case 1:
@@ -42,7 +_,7 @@
                }
 
                f_9675_.debug("Ping: (1.4-1.5.x) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.m_7630_(), minecraftserver.m_129916_(), minecraftserver.m_7416_(), minecraftserver.m_7418_());
+               String s = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.m_7630_(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                this.m_9680_(p_9686_, this.m_9683_(s));
                break;
             default:
@@ -59,7 +_,7 @@
                }
 
                f_9675_.debug("Ping: (1.6) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s1 = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.m_7630_(), minecraftserver.m_129916_(), minecraftserver.m_7416_(), minecraftserver.m_7418_());
+               String s1 = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.m_7630_(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                ByteBuf bytebuf1 = this.m_9683_(s1);
 
                try {
