--- a/net/minecraft/util/CooldownTracker.java
+++ b/net/minecraft/util/CooldownTracker.java
@@ -10,8 +_,8 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class CooldownTracker {
-   private final Map<Item, CooldownTracker.Cooldown> field_185147_a = Maps.newHashMap();
-   private int field_185148_b;
+   public final Map<Item, CooldownTracker.Cooldown> field_185147_a = Maps.newHashMap();
+   public int field_185148_b;
 
    public boolean func_185141_a(Item p_185141_1_) {
       return this.func_185143_a(p_185141_1_, 0.0F) > 0.0F;
@@ -61,9 +_,9 @@
    protected void func_185146_c(Item p_185146_1_) {
    }
 
-   class Cooldown {
+   public class Cooldown {
       private final int field_185137_a;
-      private final int field_185138_b;
+      public final int field_185138_b;
 
       private Cooldown(int p_i47037_2_, int p_i47037_3_) {
          this.field_185137_a = p_i47037_2_;
