--- a/net/minecraft/util/DamageSource.java
+++ b/net/minecraft/util/DamageSource.java
@@ -46,6 +_,19 @@
    private boolean field_76378_k;
    public final String field_76373_n;
 
+   // CraftBukkit start
+   private boolean sweep;
+
+   public boolean isSweep() {
+      return sweep;
+   }
+
+   public DamageSource sweep() {
+      this.sweep = true;
+      return this;
+   }
+   // CraftBukkit end
+
    public static DamageSource func_226252_a_(LivingEntity p_226252_0_) {
       return new EntityDamageSource("sting", p_226252_0_);
    }
