--- a/net/minecraft/util/EntityPredicates.java
+++ b/net/minecraft/util/EntityPredicates.java
@@ -41,7 +_,7 @@
       Team team = p_200823_0_.func_96124_cp();
       Team.CollisionRule team$collisionrule = team == null ? Team.CollisionRule.ALWAYS : team.func_186681_k();
       return (Predicate<Entity>)(team$collisionrule == Team.CollisionRule.NEVER ? Predicates.alwaysFalse() : field_180132_d.and((p_233586_3_) -> {
-         if (!p_233586_3_.func_70104_M()) {
+         if (!p_233586_3_.canCollideWithCB(p_200823_0_) || !p_200823_0_.canCollideWithCB(p_233586_3_)) { // CraftBukkit - collidable API
             return false;
          } else if (!p_200823_0_.field_70170_p.field_72995_K || p_233586_3_ instanceof PlayerEntity && ((PlayerEntity)p_233586_3_).func_175144_cb()) {
             Team team1 = p_233586_3_.func_96124_cp();
