--- a/net/minecraft/util/IndirectEntityDamageSource.java
+++ b/net/minecraft/util/IndirectEntityDamageSource.java
@@ -32,4 +_,11 @@
       String s1 = s + ".item";
       return !itemstack.func_190926_b() && itemstack.func_82837_s() ? new TranslationTextComponent(s1, p_151519_1_.func_145748_c_(), itextcomponent, itemstack.func_151000_E()) : new TranslationTextComponent(s, p_151519_1_.func_145748_c_(), itextcomponent);
    }
+
+   // CraftBukkit start
+   public Entity getProximateDamageSource() {
+      return super.func_76346_g();
+   }
+   // CraftBukkit end
+
 }
