--- a/net/minecraft/util/ServerCooldownTracker.java
+++ b/net/minecraft/util/ServerCooldownTracker.java
@@ -1,8 +_,10 @@
 package net.minecraft.util;
 
+import io.papermc.paper.event.player.PlayerItemCooldownEvent;
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.item.Item;
 import net.minecraft.network.play.server.SCooldownPacket;
+import org.bukkit.Bukkit;
 
 public class ServerCooldownTracker extends CooldownTracker {
    private final ServerPlayerEntity field_185149_a;
@@ -15,6 +_,21 @@
       super.func_185140_b(p_185140_1_, p_185140_2_);
       this.field_185149_a.field_71135_a.func_147359_a(new SCooldownPacket(p_185140_1_, p_185140_2_));
    }
+
+   // Paper start
+   @Override
+   public void func_185145_a(Item p_185145_1_, int p_185145_2_) {
+      PlayerItemCooldownEvent event = new PlayerItemCooldownEvent(
+              field_185149_a.getBukkitEntity(),
+              org.bukkit.craftbukkit.v1_16_R3.util.CraftMagicNumbers.getMaterial(p_185145_1_),
+              p_185145_2_
+      );
+      Bukkit.getServer().getPluginManager().callEvent(event);
+      if (!event.isCancelled()) {
+         super.func_185145_a(p_185145_1_, p_185145_2_);
+      }
+   }
+   // Paper end
 
    protected void func_185146_c(Item p_185146_1_) {
       super.func_185146_c(p_185146_1_);
