--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -17,6 +_,17 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SpawnUtil {
+   // CatServer start - 1.20.1
+   private static final ThreadLocal<org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason> trySpawnMobReason = ThreadLocal.withInitial(() -> org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   public static <T extends Mob> Optional<T> trySpawnMob(EntityType<T> p_216404_, MobSpawnType p_216405_, ServerLevel p_216406_, BlockPos p_216407_, int p_216408_, int p_216409_, int p_216410_, SpawnUtil.Strategy p_216411_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      try {
+         SpawnUtil.trySpawnMobReason.set(reason);
+         return m_216403_(p_216404_, p_216405_, p_216406_, p_216407_, p_216408_, p_216409_, p_216410_, p_216411_);
+      } finally {
+         SpawnUtil.trySpawnMobReason.remove();
+      }
+   }
+
    public static <T extends Mob> Optional<T> m_216403_(EntityType<T> p_216404_, MobSpawnType p_216405_, ServerLevel p_216406_, BlockPos p_216407_, int p_216408_, int p_216409_, int p_216410_, SpawnUtil.Strategy p_216411_) {
       BlockPos.MutableBlockPos blockpos$mutableblockpos = p_216407_.m_122032_();
 
@@ -27,8 +_,10 @@
          if (p_216406_.m_6857_().m_61937_(blockpos$mutableblockpos) && m_216398_(p_216406_, p_216410_, blockpos$mutableblockpos, p_216411_)) {
             T t = p_216404_.m_262451_(p_216406_, (CompoundTag)null, (Consumer<T>)null, blockpos$mutableblockpos, p_216405_, false, false);
             if (t != null) {
-               if (t.m_5545_(p_216406_, p_216405_) && t.m_6914_(p_216406_)) {
-                  p_216406_.m_47205_(t);
+               if (net.minecraftforge.event.ForgeEventFactory.checkSpawnPosition(t, p_216406_, p_216405_)) {
+                  org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason = SpawnUtil.trySpawnMobReason.get(); // CatServer
+                  SpawnUtil.trySpawnMobReason.remove(); // CatServer
+                  p_216406_.addFreshEntityWithPassengers(t, reason); // CraftBukkit
                   return Optional.of(t);
                }
 
@@ -39,6 +_,7 @@
 
       return Optional.empty();
    }
+   // CatServer end
 
    private static boolean m_216398_(ServerLevel p_216399_, int p_216400_, BlockPos.MutableBlockPos p_216401_, SpawnUtil.Strategy p_216402_) {
       BlockPos.MutableBlockPos blockpos$mutableblockpos = (new BlockPos.MutableBlockPos()).m_122190_(p_216401_);
