--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -77,8 +_,9 @@
 import net.minecraft.world.phys.Vec3;
 import org.joml.Quaternionf;
 import org.joml.Vector3f;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack; // CraftBukkit
 
-public class FriendlyByteBuf extends ByteBuf {
+public class FriendlyByteBuf extends ByteBuf implements net.minecraftforge.common.extensions.IForgeFriendlyByteBuf {
    private static final int f_178335_ = 5;
    private static final int f_178336_ = 10;
    public static final int f_178337_ = 2097152;
@@ -586,7 +_,7 @@
       } else {
          try {
             NbtIo.m_128941_(p_130080_, new ByteBufOutputStream(this));
-         } catch (IOException ioexception) {
+         } catch (Exception ioexception) { // CraftBukkit - IOException -> Exception
             throw new EncoderException(ioexception);
          }
       }
@@ -622,7 +_,16 @@
    }
 
    public FriendlyByteBuf m_130055_(ItemStack p_130056_) {
-      if (p_130056_.m_41619_()) {
+      return writeItemStack(p_130056_, true);
+   }
+
+   /**
+    * Most ItemStack serialization is Server to Client,and doesn't need to know the FULL tag details.
+    * One exception is items from the creative menu, which must be sent from Client to Server with their full NBT.
+    * If you want to send the FULL tag set limitedTag to false
+    */
+   public FriendlyByteBuf writeItemStack(ItemStack p_130056_, boolean limitedTag) {
+      if (p_130056_.m_41619_() || p_130056_.m_41720_() == null) { // CraftBukkit - NPE fix itemstack.getItem()
          this.writeBoolean(false);
       } else {
          this.writeBoolean(true);
@@ -630,8 +_,12 @@
          this.m_236818_(BuiltInRegistries.f_257033_, item);
          this.writeByte(p_130056_.m_41613_());
          CompoundTag compoundtag = null;
-         if (item.m_41465_() || item.m_41468_()) {
-            compoundtag = p_130056_.m_41783_();
+         if (item.isDamageable(p_130056_) || item.m_41468_()) {
+            // Spigot start - filter
+            p_130056_ = p_130056_.m_41777_();
+            CraftItemStack.setItemMeta(p_130056_, CraftItemStack.getItemMeta(p_130056_));
+            // Spigot end
+            compoundtag = limitedTag ? p_130056_.getShareTag() : p_130056_.m_41783_();
          }
 
          this.m_130079_(compoundtag);
@@ -647,7 +_,12 @@
          Item item = this.m_236816_(BuiltInRegistries.f_257033_);
          int i = this.readByte();
          ItemStack itemstack = new ItemStack(item, i);
-         itemstack.m_41751_(this.m_130260_());
+         itemstack.readShareTag(this.m_130260_());
+         // CraftBukkit start
+         if (itemstack.m_41783_() != null) {
+            CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
+         }
+         // CraftBukkit end
          return itemstack;
       }
    }
