--- a/net/minecraft/network/LegacyPingHandler.java
+++ b/net/minecraft/network/LegacyPingHandler.java
@@ -10,6 +_,8 @@
 import net.minecraft.server.MinecraftServer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.server.ServerListPingEvent;
 
 public class LegacyPingHandler extends ChannelInboundHandlerAdapter {
    private static final Logger field_151258_a = LogManager.getLogger();
@@ -29,10 +_,11 @@
             InetSocketAddress inetsocketaddress = (InetSocketAddress)p_channelRead_1_.channel().remoteAddress();
             MinecraftServer minecraftserver = this.field_151257_b.func_151267_d();
             int i = bytebuf.readableBytes();
+            ServerListPingEvent event = CraftEventFactory.callServerListPingEvent(minecraftserver.server, inetsocketaddress.getAddress(), minecraftserver.func_71273_Y(), minecraftserver.func_71233_x(), minecraftserver.func_71275_y()); // CraftBukkit
             switch(i) {
             case 0:
                field_151258_a.debug("Ping: (<1.3.x) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s2 = String.format("%s\u00a7%d\u00a7%d", minecraftserver.func_71273_Y(), minecraftserver.func_71233_x(), minecraftserver.func_71275_y());
+               String s2 = String.format("%s\u00a7%d\u00a7%d", event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                this.func_151256_a(p_channelRead_1_, this.func_151255_a(s2));
                break;
             case 1:
@@ -41,7 +_,7 @@
                }
 
                field_151258_a.debug("Ping: (1.4-1.5.x) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.func_71249_w(), minecraftserver.func_71273_Y(), minecraftserver.func_71233_x(), minecraftserver.func_71275_y());
+               String s = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.func_71249_w(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                this.func_151256_a(p_channelRead_1_, this.func_151255_a(s));
                break;
             default:
@@ -58,7 +_,7 @@
                }
 
                field_151258_a.debug("Ping: (1.6) from {}:{}", inetsocketaddress.getAddress(), inetsocketaddress.getPort());
-               String s1 = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.func_71249_w(), minecraftserver.func_71273_Y(), minecraftserver.func_71233_x(), minecraftserver.func_71275_y());
+               String s1 = String.format("\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, minecraftserver.func_71249_w(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()); // CraftBukkit
                ByteBuf bytebuf1 = this.func_151255_a(s1);
 
                try {
