--- a/net/minecraft/network/NettyCompressionEncoder.java
+++ b/net/minecraft/network/NettyCompressionEncoder.java
@@ -3,12 +_,14 @@
 import io.netty.buffer.ByteBuf;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.MessageToByteEncoder;
+
 import java.util.zip.Deflater;
 
 public class NettyCompressionEncoder extends MessageToByteEncoder<ByteBuf> {
    private final byte[] field_179302_a = new byte[8192];
    private final Deflater field_179300_b;
    private int field_179301_c;
+   private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
 
    public NettyCompressionEncoder(int p_i46005_1_) {
       this.field_179301_c = p_i46005_1_;
@@ -22,6 +_,12 @@
          packetbuffer.func_150787_b(0);
          packetbuffer.writeBytes(p_encode_2_);
       } else {
+         if (i > 2097152){
+            p_encode_2_.markReaderIndex();
+            LOGGER.error("Attempted to send packet over maximum protocol size: {} > 2097152\nData:\n{}", i,
+                    net.minecraftforge.fml.common.network.ByteBufUtils.getContentDump(p_encode_2_));
+            p_encode_2_.resetReaderIndex();
+         }
          byte[] abyte = new byte[i];
          p_encode_2_.readBytes(abyte);
          packetbuffer.func_150787_b(abyte.length);
