--- a/net/minecraft/network/PacketThreadUtil.java
+++ b/net/minecraft/network/PacketThreadUtil.java
@@ -1,5 +_,7 @@
 package net.minecraft.network;
 
+import net.minecraft.network.play.ServerPlayNetHandler;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.concurrent.ThreadTaskExecutor;
 import net.minecraft.world.server.ServerWorld;
 import org.apache.logging.log4j.LogManager;
@@ -15,6 +_,7 @@
    public static <T extends INetHandler> void func_218797_a(IPacket<T> p_218797_0_, T p_218797_1_, ThreadTaskExecutor<?> p_218797_2_) throws ThreadQuickExitException {
       if (!p_218797_2_.func_213162_bc()) {
          p_218797_2_.execute(() -> {
+            if (MinecraftServer.getServer().hasStopped() || (p_218797_1_ instanceof ServerPlayNetHandler && ((ServerPlayNetHandler) p_218797_1_).isDisconnected())) return; // CraftBukkit, MC-142590
             if (p_218797_1_.func_147298_b().func_150724_d()) {
                p_218797_0_.func_148833_a(p_218797_1_);
             } else {
@@ -24,5 +_,10 @@
          });
          throw ThreadQuickExitException.field_179886_a;
       }
+      // CraftBukkit start - SPIGOT-5477, MC-142590
+      else if (MinecraftServer.getServer().hasStopped() || (p_218797_1_ instanceof ServerPlayNetHandler && ((ServerPlayNetHandler)p_218797_1_).isDisconnected())) {
+         throw ThreadQuickExitException.field_179886_a;
+      }
+      // CraftBukkit end
    }
 }
