--- a/net/minecraft/world/chunk/storage/RegionFileCache.java
+++ b/net/minecraft/world/chunk/storage/RegionFileCache.java
@@ -21,7 +_,11 @@
    }
 
    private RegionFile func_219098_a(ChunkPos p_219098_1_) throws IOException {
-      long i = ChunkPos.func_77272_a(p_219098_1_.func_222241_h(), p_219098_1_.func_222242_i());
+      return loadFile(p_219098_1_, false);
+   }
+
+   private RegionFile loadFile(ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit
+      long i = ChunkPos.func_77272_a(pos.func_222241_h(), pos.func_222242_i());
       RegionFile regionfile = this.field_219102_c.getAndMoveToFirst(i);
       if (regionfile != null) {
          return regionfile;
@@ -34,7 +_,8 @@
             this.field_219101_a.mkdirs();
          }
 
-         File file1 = new File(this.field_219101_a, "r." + p_219098_1_.func_222241_h() + "." + p_219098_1_.func_222242_i() + ".mca");
+         File file1 = new File(this.field_219101_a, "r." + pos.func_222241_h() + "." + pos.func_222242_i() + ".mca");
+         if (existingOnly && !file1.exists()) return null; // CraftBukkit
          RegionFile regionfile1 = new RegionFile(file1, this.field_219101_a, this.field_235986_c_);
          this.field_219102_c.putAndMoveToFirst(i, regionfile1);
          return regionfile1;
@@ -43,7 +_,12 @@
 
    @Nullable
    public CompoundNBT func_219099_e(ChunkPos p_219099_1_) throws IOException {
-      RegionFile regionfile = this.func_219098_a(p_219099_1_);
+      // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
+      RegionFile regionfile = this.loadFile(p_219099_1_, true);
+      if (regionfile == null) {
+         return null;
+      }
+      // CraftBukkit end
 
       Object object;
       try (DataInputStream datainputstream = regionfile.func_222666_a(p_219099_1_)) {
@@ -58,7 +_,7 @@
    }
 
    protected void func_219100_a(ChunkPos p_219100_1_, CompoundNBT p_219100_2_) throws IOException {
-      RegionFile regionfile = this.func_219098_a(p_219100_1_);
+      RegionFile regionfile = this.loadFile(p_219100_1_, false); // CraftBukkit
 
       try (DataOutputStream dataoutputstream = regionfile.func_222661_c(p_219100_1_)) {
          CompressedStreamTools.func_74800_a(p_219100_2_, dataoutputstream);
