--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -8,6 +_,8 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.DispenserBlock;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class BoatDispenseItemBehavior extends DefaultDispenseItemBehavior {
    private final DefaultDispenseItemBehavior f_123368_ = new DefaultDispenseItemBehavior();
@@ -35,11 +_,36 @@
          d3 = 0.0D;
       }
 
+      // Boat entityboat = new Boat(worldserver, d0, d1 + d3, d2);
+      // CraftBukkit start
+      ItemStack itemstack1 = p_123376_.m_41620_(1);
+      org.bukkit.block.Block block = level.getWorld().getBlockAt(p_123375_.m_7961_().m_123341_(), p_123375_.m_7961_().m_123342_(), p_123375_.m_7961_().m_123343_());
+      CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+
+      BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2));
+      if (!DispenserBlock.eventFired) {
+         level.getCraftServer().getPluginManager().callEvent(event);
+      }
+
+      if (event.isCancelled()) {
+         p_123376_.m_41769_(1);
+         return p_123376_;
+      }
+      if (!event.getItem().equals(craftItem)) {
+         p_123376_.m_41769_(1);
+         // Chain to handler for new item
+         ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+         DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.f_52661_.get(eventStack.m_41720_());
+         if (idispensebehavior != DispenseItemBehavior.f_123393_ && idispensebehavior != this) {
+            idispensebehavior.m_6115_(p_123375_, eventStack);
+            return p_123376_;
+         }
+      }
+      Boat entityboat = new Boat(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());// CraftBukkit end
       Boat boat = new Boat(level, d0, d1 + d3, d2);
       boat.m_38332_(this.f_123369_);
       boat.m_146922_(direction.m_122435_());
-      level.m_7967_(boat);
-      p_123376_.m_41774_(1);
+      if (!level.m_7967_(entityboat)) p_123376_.m_41769_(1); // CraftBukkit
       return p_123376_;
    }
 
