--- a/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java
@@ -8,7 +_,25 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.DispenserBlock;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftVector;
+import org.bukkit.event.block.BlockDispenseEvent;
+
+import javax.annotation.Nullable;
+// CraftBukkit end
+
 public class DefaultDispenseItemBehavior implements DispenseItemBehavior {
+   // CraftBukkit start
+   private boolean dropper;
+
+   public DefaultDispenseItemBehavior(boolean dropper) {
+      this.dropper = dropper;
+   }
+   // CraftBukkit end
+
+   public DefaultDispenseItemBehavior() {}
+
    public final ItemStack m_6115_(BlockSource p_123391_, ItemStack p_123392_) {
       ItemStack itemstack = this.m_7498_(p_123391_, p_123392_);
       this.m_6823_(p_123391_);
@@ -16,15 +_,59 @@
       return itemstack;
    }
 
+   // CatServer start - fix mixin conflict with Lychee
+   private static @Nullable BlockSource spawnItemBlockSource;
+
+   private static void spawnItemBlockSource(BlockSource source) {
+      spawnItemBlockSource = source;
+   }
+
+   private static @Nullable BlockSource spawnItemBlockSource() {
+      return spawnItemBlockSource;
+   }
+
+   private static boolean spawnItemReturnable;
+
+   private static void spawnItemReturnable(boolean b) {
+      spawnItemReturnable = b;
+   }
+
+   private static boolean spawnItemReturnable() {
+      return spawnItemReturnable;
+   }
+
+   private static boolean static_dropper = false;
+
+   private static void static_dropper(boolean b) {
+      static_dropper = b;
+   }
+
+   private static boolean static_dropper() {
+      return static_dropper;
+   }
+
    protected ItemStack m_7498_(BlockSource p_123385_, ItemStack p_123386_) {
       Direction direction = p_123385_.m_6414_().m_61143_(DispenserBlock.f_52659_);
       Position position = DispenserBlock.m_52720_(p_123385_);
       ItemStack itemstack = p_123386_.m_41620_(1);
+      // CatServer start - fix mixin conflict with Lychee
+      spawnItemBlockSource(p_123385_);
+      static_dropper(this.dropper);
       m_123378_(p_123385_.m_7727_(), itemstack, 6, direction, position);
+      if (!spawnItemReturnable()) {
+         itemstack.m_41769_(1);
+      }
+      spawnItemBlockSource(null); // CatServer - reset
+      static_dropper(false); // CatServer - reset
+      // CatServer end
       return p_123386_;
    }
 
    public static void m_123378_(Level p_123379_, ItemStack p_123380_, int p_123381_, Direction p_123382_, Position p_123383_) {
+      spawnItemReturnable = true;
+
+      if (p_123380_.m_41619_()) return;
+      // CraftBukkit end
       double d0 = p_123383_.m_7096_();
       double d1 = p_123383_.m_7098_();
       double d2 = p_123383_.m_7094_();
@@ -37,7 +_,45 @@
       ItemEntity itementity = new ItemEntity(p_123379_, d0, d1, d2, p_123380_);
       double d3 = p_123379_.f_46441_.m_188500_() * 0.1D + 0.2D;
       itementity.m_20334_(p_123379_.f_46441_.m_216328_((double)p_123382_.m_122429_() * d3, 0.0172275D * (double)p_123381_), p_123379_.f_46441_.m_216328_(0.2D, 0.0172275D * (double)p_123381_), p_123379_.f_46441_.m_216328_((double)p_123382_.m_122431_() * d3, 0.0172275D * (double)p_123381_));
-      p_123379_.m_7967_(itementity);
+
+      // CatServer start - fix mixin conflict with Lychee
+      var blockSource = spawnItemBlockSource();
+      if (blockSource != null) {
+         org.bukkit.block.Block block = p_123379_.getWorld().getBlockAt(blockSource.m_7961_().m_123341_(), blockSource.m_7961_().m_123342_(), blockSource.m_7961_().m_123343_());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_123380_);
+
+         BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), CraftVector.toBukkit(itementity.m_20184_()));
+         if (!DispenserBlock.eventFired) {
+            p_123379_.getCraftServer().getPluginManager().callEvent(event);
+         }
+
+         if (event.isCancelled()) {
+            spawnItemReturnable(false);
+            return;
+         }
+
+         itementity.m_32045_(CraftItemStack.asNMSCopy(event.getItem()));
+         itementity.m_20256_(CraftVector.toNMS(event.getVelocity()));
+
+         if (!DefaultDispenseItemBehavior.static_dropper() && !event.getItem().getType().equals(craftItem.getType())) {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.f_52661_.get(eventStack.m_41720_());
+            if (idispensebehavior != DispenseItemBehavior.f_123393_ && idispensebehavior.getClass() != DefaultDispenseItemBehavior.class) {
+               idispensebehavior.m_6115_(blockSource, eventStack);
+            } else {
+               p_123379_.m_7967_(itementity);
+            }
+            spawnItemReturnable(false);
+            return;
+         }
+
+         p_123379_.m_7967_(itementity);
+
+         spawnItemReturnable(true);
+         // CraftBukkit end
+      }
+      // CatServer end
    }
 
    protected void m_6823_(BlockSource p_123384_) {
