--- a/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
@@ -19,13 +_,37 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.block.Block;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class ShearsDispenseItemBehavior extends OptionalDispenseItemBehavior {
    protected ItemStack m_7498_(BlockSource p_123580_, ItemStack p_123581_) {
       ServerLevel serverlevel = p_123580_.m_7727_();
+      // CraftBukkit start
+      org.bukkit.block.Block bukkitBlock = serverlevel.getWorld().getBlockAt(p_123580_.m_7961_().m_123341_(), p_123580_.m_7961_().m_123342_(), p_123580_.m_7961_().m_123343_());
+      CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_123581_);
+      BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+      if (!DispenserBlock.eventFired) {
+         serverlevel.getCraftServer().getPluginManager().callEvent(event);
+      }
+      if (event.isCancelled()) {
+         return p_123581_;
+      }
+      if (!event.getItem().equals(craftItem)) {
+         // Chain to handler for new item
+         ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+         DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.f_52661_.get(eventStack.m_41720_());
+         if (idispensebehavior != DispenseItemBehavior.f_123393_ && idispensebehavior != this) {
+            idispensebehavior.m_6115_(p_123580_, eventStack);
+            return p_123581_;
+         }
+      }
+      // CraftBukkit end
       if (!serverlevel.m_5776_()) {
          BlockPos blockpos = p_123580_.m_7961_().m_121945_(p_123580_.m_6414_().m_61143_(DispenserBlock.f_52659_));
-         this.m_123573_(m_123576_(serverlevel, blockpos) || m_123582_(serverlevel, blockpos));
+         this.m_123573_(m_123576_(serverlevel, blockpos) || tryShearLivingEntity(serverlevel, blockpos, bukkitBlock, craftItem)); // CraftBukkit
          if (this.m_123570_() && p_123581_.m_220157_(1, serverlevel.m_213780_(), (ServerPlayer)null)) {
             p_123581_.m_41764_(0);
          }
@@ -52,12 +_,34 @@
       return false;
    }
 
+   // CatServer start
+   private static boolean tryShearLivingEntity(ServerLevel p_123583_, BlockPos p_123584_, org.bukkit.block.Block bukkitBlock, CraftItemStack craftItem) { // CraftBukkit - add args
+      try {
+         catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$bukkitBlock.push(null, bukkitBlock);
+         catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$craftItemStack.push(null, craftItem);
+         return m_123582_(p_123583_, p_123584_);
+      } finally {
+         catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$bukkitBlock.pop(null);
+         catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$craftItemStack.pop(null);
+      }
+   }
+
    private static boolean m_123582_(ServerLevel p_123583_, BlockPos p_123584_) {
       for(LivingEntity livingentity : p_123583_.m_6443_(LivingEntity.class, new AABB(p_123584_), EntitySelector.f_20408_)) {
          if (livingentity instanceof Shearable shearable) {
             if (shearable.m_6220_()) {
+               {
+                  // CraftBukkit start
+                  var bukkitBlock = catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$bukkitBlock.get(null);
+                  var craftItem = catserver.server.CatServerCaptures.getCatServerCaptures().ShearsDispenseItemBehavior$tryShearLivingEntity$craftItemStack.get(null);
+                  if (CraftEventFactory.callBlockShearEntityEvent(livingentity, bukkitBlock, craftItem).isCancelled()) {
+                     continue;
+                  }
+                  // CraftBukkit end
+               }
                shearable.m_5851_(SoundSource.BLOCKS);
                p_123583_.m_142346_((Entity)null, GameEvent.f_157781_, p_123584_);
+
                return true;
             }
          }
@@ -65,4 +_,5 @@
 
       return false;
    }
+   // CatServer end
 }
