--- a/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
@@ -10,6 +_,8 @@
 import net.minecraft.world.item.context.DirectionalPlaceContext;
 import net.minecraft.world.level.block.DispenserBlock;
 import org.slf4j.Logger;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class ShulkerBoxDispenseBehavior extends OptionalDispenseItemBehavior {
    private static final Logger f_175749_ = LogUtils.getLogger();
@@ -21,6 +_,30 @@
          Direction direction = p_123587_.m_6414_().m_61143_(DispenserBlock.f_52659_);
          BlockPos blockpos = p_123587_.m_7961_().m_142300_(direction);
          Direction direction1 = p_123587_.m_7727_().m_46859_(blockpos.m_7495_()) ? direction : Direction.UP;
+
+         // CraftBukkit start
+         org.bukkit.block.Block bukkitBlock = p_123587_.m_7727_().getWorld().getBlockAt(p_123587_.m_7961_().m_123341_(), p_123587_.m_7961_().m_123342_(), p_123587_.m_7961_().m_123343_());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_123588_);
+
+         BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.m_123341_(), blockpos.m_123342_(), blockpos.m_123343_()));
+         if (!DispenserBlock.eventFired) {
+            p_123587_.m_7727_().getCraftServer().getPluginManager().callEvent(event);
+         }
+
+         if (event.isCancelled()) {
+            return p_123588_;
+         }
+
+         if (!event.getItem().equals(craftItem)) {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.f_52661_.get(eventStack.m_41720_());
+            if (idispensebehavior != DispenseItemBehavior.f_123393_ && idispensebehavior != this) {
+               idispensebehavior.m_6115_(p_123587_, eventStack);
+               return p_123588_;
+            }
+         }
+         // CraftBukkit end
 
          try {
             this.m_123573_(((BlockItem)item).m_40576_(new DirectionalPlaceContext(p_123587_.m_7727_(), blockpos, direction, p_123588_, direction1)).m_19077_());
