--- a/net/minecraft/entity/item/EnderCrystalEntity.java
+++ b/net/minecraft/entity/item/EnderCrystalEntity.java
@@ -21,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class EnderCrystalEntity extends Entity {
    private static final DataParameter<Optional<BlockPos>> field_184521_b = EntityDataManager.func_187226_a(EnderCrystalEntity.class, DataSerializers.field_187201_k);
@@ -52,7 +_,11 @@
       if (this.field_70170_p instanceof ServerWorld) {
          BlockPos blockpos = this.func_233580_cy_();
          if (((ServerWorld)this.field_70170_p).func_241110_C_() != null && this.field_70170_p.func_180495_p(blockpos).func_196958_f()) {
-            this.field_70170_p.func_175656_a(blockpos, AbstractFireBlock.func_235326_a_(this.field_70170_p, blockpos));
+            // CraftBukkit start
+            if (!CraftEventFactory.callBlockIgniteEvent(this.field_70170_p, blockpos, this).isCancelled()) {
+               this.field_70170_p.func_175656_a(blockpos, AbstractFireBlock.func_235326_a_(this.field_70170_p, blockpos));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -88,9 +_,24 @@
          return false;
       } else {
          if (!this.field_70128_L && !this.field_70170_p.field_72995_K) {
+            // CraftBukkit start - All non-living entities need this
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             this.func_70106_y();
             if (!p_70097_1_.func_94541_c()) {
-               this.field_70170_p.func_217385_a((Entity)null, this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), 6.0F, Explosion.Mode.DESTROY);
+               // CraftBukkit start
+               ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+               this.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  this.field_70128_L = false;
+                  return false;
+               }
+               boolean oldFire = this.field_70170_p.pehkuiExplosionFire;
+               this.field_70170_p.pehkuiExplosionFire = event.getFire();
+               this.field_70170_p.func_217385_a((Entity)null, this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), event.getRadius(), Explosion.Mode.DESTROY);
+               this.field_70170_p.pehkuiExplosionFire = oldFire;
             }
 
             this.func_184519_a(p_70097_1_);
