--- a/net/minecraft/entity/item/ExperienceBottleEntity.java
+++ b/net/minecraft/entity/item/ExperienceBottleEntity.java
@@ -9,6 +_,7 @@
 import net.minecraft.potion.Potions;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ExperienceBottleEntity extends ProjectileItemEntity {
    public ExperienceBottleEntity(EntityType<? extends ExperienceBottleEntity> p_i50152_1_, World p_i50152_2_) {
@@ -34,8 +_,17 @@
    protected void func_70227_a(RayTraceResult p_70227_1_) {
       super.func_70227_a(p_70227_1_);
       if (!this.field_70170_p.field_72995_K) {
-         this.field_70170_p.func_217379_c(2002, this.func_233580_cy_(), PotionUtils.func_185183_a(Potions.field_185230_b));
+         // CraftBukkit - moved to after event
+         // this.world.playEvent(2002, this.getPosition(), PotionUtils.getPotionColor(Potions.WATER));
          int i = 3 + this.field_70170_p.field_73012_v.nextInt(5) + this.field_70170_p.field_73012_v.nextInt(5);
+
+         // CraftBukkit start
+         org.bukkit.event.entity.ExpBottleEvent event = CraftEventFactory.callExpBottleEvent(this, i);
+         i = event.getExperience();
+         if (event.getShowEffect()) {
+            this.field_70170_p.func_217379_c(2002, this.func_233580_cy_(), PotionUtils.func_185183_a(Potions.field_185230_b));
+         }
+         // CraftBukkit end
 
          while(i > 0) {
             int j = ExperienceOrbEntity.func_70527_a(i);
