--- a/net/minecraft/entity/item/ItemFrameEntity.java
+++ b/net/minecraft/entity/item/ItemFrameEntity.java
@@ -35,13 +_,14 @@
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ItemFrameEntity extends HangingEntity {
    private static final Logger field_195052_c = LogManager.getLogger();
    private static final DataParameter<ItemStack> field_184525_c = EntityDataManager.func_187226_a(ItemFrameEntity.class, DataSerializers.field_187196_f);
    private static final DataParameter<Integer> field_184526_d = EntityDataManager.func_187226_a(ItemFrameEntity.class, DataSerializers.field_187192_b);
-   private float field_82337_e = 1.0F;
-   private boolean field_234259_an_;
+   public float field_82337_e = 1.0F;
+   public boolean field_234259_an_;
 
    public ItemFrameEntity(EntityType<? extends ItemFrameEntity> p_i50224_1_, World p_i50224_2_) {
       super(p_i50224_1_, p_i50224_2_);
@@ -61,7 +_,7 @@
       this.func_184212_Q().func_187214_a(field_184526_d, 0);
    }
 
-   protected void func_174859_a(Direction p_174859_1_) {
+   public void func_174859_a(Direction p_174859_1_) { // protected->public CraftBukkit
       Validate.notNull(p_174859_1_);
       this.field_174860_b = p_174859_1_;
       if (p_174859_1_.func_176740_k().func_176722_c()) {
@@ -79,16 +_,25 @@
 
    protected void func_174856_o() {
       if (this.field_174860_b != null) {
-         double d0 = 0.46875D;
-         double d1 = (double)this.field_174861_a.func_177958_n() + 0.5D - (double)this.field_174860_b.func_82601_c() * 0.46875D;
-         double d2 = (double)this.field_174861_a.func_177956_o() + 0.5D - (double)this.field_174860_b.func_96559_d() * 0.46875D;
-         double d3 = (double)this.field_174861_a.func_177952_p() + 0.5D - (double)this.field_174860_b.func_82599_e() * 0.46875D;
-         this.func_226288_n_(d1, d2, d3);
-         double d4 = (double)this.func_82329_d();
-         double d5 = (double)this.func_82330_g();
-         double d6 = (double)this.func_82329_d();
-         Direction.Axis direction$axis = this.field_174860_b.func_176740_k();
-         switch(direction$axis) {
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.func_174826_a(calculateBoundingBox(this, this.field_174861_a, this.field_174860_b, this.func_82329_d(), this.func_82330_g()));
+      }
+   }
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AxisAlignedBB calculateBoundingBox(@Nullable Entity entity, BlockPos blockPosition, Direction direction, int width, int height) {
+      final double d0 = 0.46875D;
+      final double d1 = blockPosition.func_177958_n() + 0.5D - direction.func_82601_c() * d0;
+      final double d2 = blockPosition.func_177956_o() + 0.5D - direction.func_96559_d() * d0;
+      final double d3 = blockPosition.func_177952_p() + 0.5D - direction.func_82599_e() * d0;
+      if (entity != null) {
+         entity.func_226288_n_(d1, d2, d3);
+      }
+      double d4 = width;
+      double d5 = height;
+      double d6 = width;
+      final Direction.Axis direction$axis = direction.func_176740_k();
+      switch (direction$axis) {
          case X:
             d4 = 1.0D;
             break;
@@ -97,14 +_,14 @@
             break;
          case Z:
             d6 = 1.0D;
-         }
-
-         d4 = d4 / 32.0D;
-         d5 = d5 / 32.0D;
-         d6 = d6 / 32.0D;
-         this.func_174826_a(new AxisAlignedBB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6));
       }
+
+      d4 = d4 / 32.0D;
+      d5 = d5 / 32.0D;
+      d6 = d6 / 32.0D;
+      return new AxisAlignedBB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6);
    }
+   // CraftBukkit end
 
    public boolean func_70518_d() {
       if (this.field_234259_an_) {
@@ -147,6 +_,11 @@
          return false;
       } else if (!p_70097_1_.func_94541_c() && !this.func_82335_i().func_190926_b()) {
          if (!this.field_70170_p.field_72995_K) {
+            // CraftBukkit start - fire EntityDamageEvent
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false) || this.field_70128_L) {
+               return true;
+            }
+            // CraftBukkit end
             this.func_146065_b(p_70097_1_.func_76346_g(), false);
             this.func_184185_a(SoundEvents.field_187629_cO, 1.0F, 1.0F);
          }
@@ -234,14 +_,19 @@
    }
 
    public void func_174864_a(ItemStack p_174864_1_, boolean p_174864_2_) {
-      if (!p_174864_1_.func_190926_b()) {
-         p_174864_1_ = p_174864_1_.func_77946_l();
-         p_174864_1_.func_190920_e(1);
-         p_174864_1_.func_234695_a_(this);
+      this.setItem(p_174864_1_, p_174864_2_, true);
+   }
+
+   public void setItem(ItemStack stack, boolean p_174864_2_, boolean playSound) {
+      // CraftBukkit end
+      if (!stack.func_190926_b()) {
+         stack = stack.func_77946_l();
+         stack.func_190920_e(1);
+         stack.func_234695_a_(this);
       }
 
-      this.func_184212_Q().func_187227_b(field_184525_c, p_174864_1_);
-      if (!p_174864_1_.func_190926_b()) {
+      this.func_184212_Q().func_187227_b(field_184525_c, stack);
+      if (!stack.func_190926_b() && p_174864_2_ && playSound) { // CraftBukkit
          this.func_184185_a(SoundEvents.field_187620_cL, 1.0F, 1.0F);
       }
 
