--- a/net/minecraft/entity/item/PaintingEntity.java
+++ b/net/minecraft/entity/item/PaintingEntity.java
@@ -26,6 +_,10 @@
 
    public PaintingEntity(EntityType<? extends PaintingEntity> p_i50221_1_, World p_i50221_2_) {
       super(p_i50221_1_, p_i50221_2_);
+      // CraftBukkit start - generate a non-null painting
+      List<PaintingType> list = Lists.newArrayList(PaintingType.field_200843_b);
+      this.field_70522_e = list.get(this.field_70146_Z.nextInt(list.size()));
+      // CraftBukkit end
    }
 
    public PaintingEntity(World p_i45849_1_, BlockPos p_i45849_2_, Direction p_i45849_3_) {
