--- a/net/minecraft/entity/item/TNTEntity.java
+++ b/net/minecraft/entity/item/TNTEntity.java
@@ -16,12 +_,15 @@
 import net.minecraft.particles.ParticleTypes;
 import net.minecraft.world.Explosion;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class TNTEntity extends Entity {
    private static final DataParameter<Integer> field_184537_a = EntityDataManager.func_187226_a(TNTEntity.class, DataSerializers.field_187192_b);
    @Nullable
-   private LivingEntity field_94084_b;
+   public LivingEntity field_94084_b;
    private int field_70516_a = 80;
+   public float yield = 4; // CraftBukkit - add field
+   public boolean isIncendiary = false; // CraftBukkit - add field
 
    public TNTEntity(EntityType<? extends TNTEntity> p_i50216_1_, World p_i50216_2_) {
       super(p_i50216_1_, p_i50216_2_);
@@ -53,6 +_,7 @@
    }
 
    public void func_70071_h_() {
+      if (field_70170_p.spigotConfig.currentPrimedTnt++ > field_70170_p.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
       if (!this.func_189652_ae()) {
          this.func_213317_d(this.func_213322_ci().func_72441_c(0.0D, -0.04D, 0.0D));
       }
@@ -65,10 +_,13 @@
 
       --this.field_70516_a;
       if (this.field_70516_a <= 0) {
-         this.func_70106_y();
+         // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
+    //    this.remove();
          if (!this.field_70170_p.field_72995_K) {
             this.func_70515_d();
          }
+         this.func_70106_y();
+         // CraftBukkit end
       } else {
          this.func_233566_aG_();
          if (this.field_70170_p.field_72995_K) {
@@ -79,8 +_,13 @@
    }
 
    protected void func_70515_d() {
-      float f = 4.0F;
-      this.field_70170_p.func_217385_a(this, this.func_226277_ct_(), this.func_226283_e_(0.0625D), this.func_226281_cx_(), 4.0F, Explosion.Mode.BREAK);
+      // CraftBukkit start
+      ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+      this.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+      if (!event.isCancelled()) {
+         this.field_70170_p.func_217398_a(this, this.func_226277_ct_(), this.func_226283_e_(0.0625D), this.func_226281_cx_(), event.getRadius(), event.getFire(), Explosion.Mode.BREAK);
+      }
+      // CraftBukkit end
    }
 
    protected void func_213281_b(CompoundNBT p_213281_1_) {
