--- a/net/minecraft/world/level/block/entity/BannerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BannerBlockEntity.java
@@ -25,9 +_,9 @@
    public static final String f_155033_ = "Color";
    @Nullable
    private Component f_58473_;
-   private DyeColor f_58474_;
+   public DyeColor f_58474_;
    @Nullable
-   private ListTag f_58475_;
+   public ListTag f_58475_;
    @Nullable
    private List<Pair<BannerPattern, DyeColor>> f_58477_;
 
@@ -95,6 +_,11 @@
       }
 
       this.f_58475_ = p_155042_.m_128437_("Patterns", 10);
+      // CraftBukkit start
+      while (this.f_58475_.size() > 20) {
+         this.f_58475_.remove(20);
+      }
+      // CraftBukkit end
       this.f_58477_ = null;
    }
 
