--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -36,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.potion.PotionEffect;
 
 public class BeaconBlockEntity extends BlockEntity implements MenuProvider {
    private static final int f_155085_ = 4;
@@ -51,9 +_,11 @@
    public int f_58650_;
    private int f_58651_;
    @Nullable
-   public MobEffect f_58652_;
+   public
+   MobEffect f_58652_;
    @Nullable
-   public MobEffect f_58653_;
+   public
+   MobEffect f_58653_;
    @Nullable
    public Component f_58654_;
    public LockCode f_58655_ = LockCode.f_19102_;
@@ -98,6 +_,27 @@
       super(BlockEntityType.f_58930_, p_155088_, p_155089_);
    }
 
+   // CraftBukkit start - add fields and methods
+   public PotionEffect getPrimaryEffect() {
+      return null;
+   }
+
+   public PotionEffect getSecondaryEffect() {
+      return null;
+   }
+
+   public static List getHumansInRange(Level world, BlockPos blockposition, int i) {
+      {
+         double d0 = (double) (i * 10 + 10);
+
+         AABB aabb = (new AABB(blockposition)).m_82400_(d0).m_82363_(0.0D, (double)world.m_141928_(), 0.0D);
+         List<Player> list = world.m_45976_(Player.class, aabb);
+
+         return list;
+      }
+   }
+   // CraftBukkit end
+
    public static void m_155107_(Level p_155108_, BlockPos p_155109_, BlockState p_155110_, BeaconBlockEntity p_155111_) {
       int i = p_155109_.m_123341_();
       int j = p_155109_.m_123342_();
@@ -117,8 +_,8 @@
       for(int i1 = 0; i1 < 10 && blockpos.m_123342_() <= l; ++i1) {
          BlockState blockstate = p_155108_.m_8055_(blockpos);
          Block block = blockstate.m_60734_();
-         if (block instanceof BeaconBeamBlock) {
-            float[] afloat = ((BeaconBeamBlock)block).m_7988_().m_41068_();
+         float[] afloat = blockstate.getBeaconColorMultiplier(p_155108_, blockpos, p_155109_);
+         if (afloat != null) {
             if (p_155111_.f_58649_.size() <= 1) {
                beaconblockentity$beaconbeamsection = new BeaconBlockEntity.BeaconBeamSection(afloat);
                p_155111_.f_58649_.add(beaconblockentity$beaconbeamsection);
