--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -3,6 +_,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
@@ -36,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.v1_18_R2.potion.CraftPotionUtil;
+import org.bukkit.potion.PotionEffect;
 
 public class BeaconBlockEntity extends BlockEntity implements MenuProvider {
    private static final int f_155085_ = 4;
@@ -48,15 +_,27 @@
    private static final int f_155086_ = 10;
    List<BeaconBlockEntity.BeaconBeamSection> f_58648_ = Lists.newArrayList();
    private List<BeaconBlockEntity.BeaconBeamSection> f_58649_ = Lists.newArrayList();
-   int f_58650_;
+   public int f_58650_;
    private int f_58651_;
    @Nullable
+   public
    MobEffect f_58652_;
    @Nullable
+   public
    MobEffect f_58653_;
    @Nullable
-   private Component f_58654_;
-   private LockCode f_58655_ = LockCode.f_19102_;
+   public Component f_58654_;
+   public LockCode f_58655_ = LockCode.f_19102_;
+
+   // CraftBukkit start - add fields and methods
+   public PotionEffect getPrimaryEffect() {
+      return (this.f_58652_ != null) ? CraftPotionUtil.toBukkit(new MobEffectInstance(this.f_58652_, getLevel(this.f_58650_), getAmplification(f_58650_, f_58652_, f_58653_), true, true)) : null;
+   }
+   public PotionEffect getSecondaryEffect() {
+      return (hasSecondaryEffect(f_58650_, f_58652_, f_58653_)) ? CraftPotionUtil.toBukkit(new MobEffectInstance(this.f_58653_, getLevel(this.f_58650_), getAmplification(f_58650_, f_58652_, f_58653_), true, true)) : null;
+   }
+   // CraftBukkit end
+
    private final ContainerData f_58656_ = new ContainerData() {
       public int m_6413_(int p_58711_) {
          switch(p_58711_) {
@@ -117,8 +_,8 @@
       for(int i1 = 0; i1 < 10 && blockpos.m_123342_() <= l; ++i1) {
          BlockState blockstate = p_155108_.m_8055_(blockpos);
          Block block = blockstate.m_60734_();
-         if (block instanceof BeaconBeamBlock) {
-            float[] afloat = ((BeaconBeamBlock)block).m_7988_().m_41068_();
+         float[] afloat = blockstate.getBeaconColorMultiplier(p_155108_, blockpos, p_155109_);
+         if (afloat != null) {
             if (p_155111_.f_58649_.size() <= 1) {
                beaconblockentity$beaconbeamsection = new BeaconBlockEntity.BeaconBeamSection(afloat);
                p_155111_.f_58649_.add(beaconblockentity$beaconbeamsection);
@@ -209,6 +_,53 @@
       super.m_7651_();
    }
 
+   // CraftBukkit start - split into components
+   private static byte getAmplification(int p_155100_, @Nullable MobEffect p_155101_, @Nullable MobEffect p_155102_) {
+      {
+         byte i = 0;
+         if (p_155100_ >= 4 && p_155101_ == p_155102_) {
+            i = 1;
+         }
+
+         return i;
+      }
+   }
+
+   private static int getLevel(int p_155100_) {
+      {
+
+         int j = (9 + p_155100_ * 2) * 20;
+         return j;
+      }
+   }
+
+   public static List getHumansInRange(Level world, BlockPos blockposition, int i) {
+      {
+         double d0 = (double) (i * 10 + 10);
+         AABB aabb = (new AABB(blockposition)).m_82400_(d0).m_82363_(0.0D, (double) world.m_141928_(), 0.0D);
+         List<Player> list = world.m_45976_(Player.class, aabb);
+         return list;
+      }
+   }
+
+   private static void applyEffect(List<Player> list, MobEffect p_155101_, int j, int i) {
+      {
+         for (Player player : list) {
+            player.m_7292_(new MobEffectInstance(p_155101_, j, i, true, true));
+         }
+      }
+   }
+
+   private static boolean hasSecondaryEffect(int p_155100_, @Nullable MobEffect mobeffectlist, @Nullable MobEffect mobeffectlist1) {
+      {
+         if (p_155100_ >= 4 && mobeffectlist != mobeffectlist1 && mobeffectlist1 != null) {
+            return true;
+         }
+
+      }
+      return false;
+   }
+
    private static void m_155097_(Level p_155098_, BlockPos p_155099_, int p_155100_, @Nullable MobEffect p_155101_, @Nullable MobEffect p_155102_) {
       if (!p_155098_.f_46443_ && p_155101_ != null) {
          double d0 = (double)(p_155100_ * 10 + 10);
@@ -233,6 +_,7 @@
 
       }
    }
+   // CraftBukkit end
 
    public static void m_155103_(Level p_155104_, BlockPos p_155105_, SoundEvent p_155106_) {
       p_155104_.m_5594_((Player)null, p_155105_, p_155106_, SoundSource.BLOCKS, 1.0F, 1.0F);
@@ -258,8 +_,11 @@
 
    public void m_142466_(CompoundTag p_155113_) {
       super.m_142466_(p_155113_);
-      this.f_58652_ = m_58686_(p_155113_.m_128451_("Primary"));
-      this.f_58653_ = m_58686_(p_155113_.m_128451_("Secondary"));
+      // CraftBukkit start - persist manually set non-default beacon effects (SPIGOT-3598)
+      this.f_58652_ = MobEffect.m_19453_(p_155113_.m_128451_("Primary"));
+      this.f_58653_ = MobEffect.m_19453_(p_155113_.m_128451_("Secondary"));
+      this.f_58650_ = p_155113_.m_128451_("Levels"); // SPIGOT-5053, use where available
+      // CraftBukkit end
       if (p_155113_.m_128425_("CustomName", 8)) {
          this.f_58654_ = Component.Serializer.m_130701_(p_155113_.m_128461_("CustomName"));
       }
