--- a/net/minecraft/world/level/block/entity/BellBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BellBlockEntity.java
@@ -33,7 +_,7 @@
    public boolean f_58814_;
    public Direction f_58815_;
    private List<LivingEntity> f_58817_;
-   private boolean f_58818_;
+   public boolean f_58818_; // CatServer - private -> public
    private int f_58819_;
 
    public BellBlockEntity(BlockPos p_155173_, BlockState p_155174_) {
@@ -128,9 +_,14 @@
    }
 
    private static void m_155186_(Level p_155187_, BlockPos p_155188_, List<LivingEntity> p_155189_) {
+      // CraftBukkit start - 1.20.1
+      List<org.bukkit.entity.LivingEntity> entities =
       p_155189_.stream().filter((p_155219_) -> {
          return m_155196_(p_155188_, p_155219_);
-      }).forEach(BellBlockEntity::m_58840_);
+      }).map((entity) -> (org.bukkit.entity.LivingEntity) entity.getBukkitEntity()).collect(java.util.stream.Collectors.toCollection(java.util.ArrayList::new));
+
+      org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBellResonateEvent(p_155187_, p_155188_, entities).forEach(BellBlockEntity::m_58840_);
+      // CraftBukkit end
    }
 
    private static void m_155207_(Level p_155208_, BlockPos p_155209_, List<LivingEntity> p_155210_) {
