--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -28,6 +_,12 @@
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Arrays;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class BrushableBlockEntity extends BlockEntity {
    private static final Logger f_276497_ = LogUtils.getLogger();
    private static final String f_276647_ = "LootTable";
@@ -40,12 +_,12 @@
    private int f_276531_;
    private long f_276679_;
    private long f_276557_;
-   private ItemStack f_276563_ = ItemStack.f_41583_;
+   public ItemStack f_276563_ = ItemStack.f_41583_; // CatServer - private -> public
    @Nullable
    private Direction f_276638_;
    @Nullable
-   private ResourceLocation f_276466_;
-   private long f_276487_;
+   public ResourceLocation f_276466_; // CatServer - private -> public
+   public long f_276487_; // CatServer - private -> public
 
    public BrushableBlockEntity(BlockPos p_277558_, BlockState p_278093_) {
       super(BlockEntityType.f_271323_, p_277558_, p_278093_);
@@ -141,7 +_,10 @@
             double d5 = (double)blockpos.m_123343_() + 0.5D * d1 + d2;
             ItemEntity itementity = new ItemEntity(this.f_58857_, d3, d4, d5, this.f_276563_.m_41620_(this.f_58857_.f_46441_.m_188503_(21) + 10));
             itementity.m_20256_(Vec3.f_82478_);
-            this.f_58857_.m_7967_(itementity);
+            // CraftBukkit start - 1.20.1
+            org.bukkit.block.Block bblock = CraftBlock.at(this.f_58857_, this.f_58858_);
+            CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), (ServerPlayer) p_278006_, Arrays.asList(itementity));
+            // CraftBukkit end
             this.f_276563_ = ItemStack.f_41583_;
          }
 
@@ -211,6 +_,7 @@
    }
 
    public void m_142466_(CompoundTag p_277597_) {
+      super.m_142466_(p_277597_); // CraftBukkit - 1.20.1 SPIGOT-7393: Load super Bukkit data
       if (!this.m_276996_(p_277597_) && p_277597_.m_128441_("item")) {
          this.f_276563_ = ItemStack.m_41712_(p_277597_.m_128469_("item"));
       }
