--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -25,12 +_,20 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockCookEvent;
+import org.bukkit.event.block.CampfireStartEvent;
+import org.bukkit.inventory.CampfireRecipe;
+// CraftBukkit end
+
 public class CampfireBlockEntity extends BlockEntity implements Clearable {
    private static final int f_155298_ = 2;
    private static final int f_155299_ = 4;
    private final NonNullList<ItemStack> f_59042_ = NonNullList.m_122780_(4, ItemStack.f_41583_);
-   private final int[] f_59043_ = new int[4];
-   private final int[] f_59044_ = new int[4];
+   public final int[] f_59043_ = new int[4]; // CatServer - private -> public
+   public final int[] f_59044_ = new int[4]; // CatServer - private -> public
    private final RecipeManager.CachedCheck<Container, CampfireCookingRecipe> f_222760_ = RecipeManager.m_220267_(RecipeType.f_44111_);
 
    public CampfireBlockEntity(BlockPos p_155301_, BlockState p_155302_) {
@@ -51,6 +_,17 @@
                   return p_270054_.m_5874_(container, p_155307_.m_9598_());
                }).orElse(itemstack);
                if (itemstack1.m_246617_(p_155307_.m_246046_())) {
+                  // CraftBukkit start - fire BlockCookEvent
+                  CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+                  org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+                  BlockCookEvent blockCookEvent = new BlockCookEvent(CraftBlock.at(p_155307_, p_155308_), source, result);
+                  p_155307_.getCraftServer().getPluginManager().callEvent(blockCookEvent);
+                  if (blockCookEvent.isCancelled()) {
+                     return;
+                  }
+                  result = blockCookEvent.getResult();
+                  itemstack1 = CraftItemStack.asNMSCopy(result);
+                  // CraftBukkit end
                   Containers.m_18992_(p_155307_, (double)p_155308_.m_123341_(), (double)p_155308_.m_123342_(), (double)p_155308_.m_123343_(), itemstack1);
                   p_155310_.f_59042_.set(i, ItemStack.f_41583_);
                   p_155307_.m_7260_(p_155308_, p_155309_, p_155309_, 3);
@@ -153,7 +_,11 @@
       for(int i = 0; i < this.f_59042_.size(); ++i) {
          ItemStack itemstack = this.f_59042_.get(i);
          if (itemstack.m_41619_()) {
-            this.f_59044_[i] = p_238287_;
+            // CraftBukkit start
+            CampfireStartEvent event = new CampfireStartEvent(CraftBlock.at(this.f_58857_, this.f_58858_), CraftItemStack.asCraftMirror(p_238286_), (CampfireRecipe) m_59051_(p_238286_).get().toBukkitRecipe());
+            this.f_58857_.getCraftServer().getPluginManager().callEvent(event);
+            this.f_59044_[i] = event.getTotalCookTime(); // p_238287_ -> event.getTotalCookTime()
+            // CraftBukkit end
             this.f_59043_[i] = 0;
             this.f_59042_.set(i, p_238286_.m_41620_(1));
             this.f_58857_.m_220407_(GameEvent.f_157792_, this.m_58899_(), GameEvent.Context.m_223719_(p_238285_, this.m_58900_()));
