--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -20,6 +_,9 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.CampfireBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockCookEvent;
 
 public class CampfireBlockEntity extends BlockEntity implements Clearable {
    private static final int f_155298_ = 2;
@@ -45,6 +_,17 @@
                ItemStack itemstack1 = p_155307_.m_7465_().m_44015_(RecipeType.f_44111_, container, p_155307_).map((p_155305_) -> {
                   return p_155305_.m_5874_(container);
                }).orElse(itemstack);
+               // CraftBukkit start - fire BlockCookEvent
+               CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+               org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+               BlockCookEvent blockCookEvent = new BlockCookEvent(CraftBlock.at(p_155307_, p_155308_), source, result);
+               p_155307_.getCraftServer().getPluginManager().callEvent(blockCookEvent);
+               if (blockCookEvent.isCancelled()) {
+                  return;
+               }
+               result = blockCookEvent.getResult();
+               itemstack1 = CraftItemStack.asNMSCopy(result);
+               // CraftBukkit end
                Containers.m_18992_(p_155307_, (double)p_155308_.m_123341_(), (double)p_155308_.m_123342_(), (double)p_155308_.m_123343_(), itemstack1);
                p_155310_.f_59042_.set(i, ItemStack.f_41583_);
                p_155307_.m_7260_(p_155308_, p_155309_, p_155309_, 3);
