--- a/net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -19,6 +_,13 @@
    private boolean f_59124_;
    private boolean f_59125_;
    private final BaseCommandBlock f_59127_ = new BaseCommandBlock() {
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return new org.bukkit.craftbukkit.v1_20_R1.command.CraftBlockCommandSender(wrapper, CommandBlockEntity.this);
+      }
+      // CraftBukkit end
+
       public void m_6590_(String p_59157_) {
          super.m_6590_(p_59157_);
          CommandBlockEntity.this.m_6596_();
