--- a/net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -12,12 +_,18 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
 
 public class CommandBlockEntity extends BlockEntity {
    private boolean f_59123_;
    private boolean f_59124_;
    private boolean f_59125_;
    private final BaseCommandBlock f_59127_ = new BaseCommandBlock() {
+      @Override
+      public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return new org.bukkit.craftbukkit.v1_18_R2.command.CraftBlockCommandSender(wrapper, CommandBlockEntity.this);
+      }
+
       public void m_6590_(String p_59157_) {
          super.m_6590_(p_59157_);
          CommandBlockEntity.this.m_6596_();
