--- a/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -26,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class ConduitBlockEntity extends BlockEntity {
    private static final int f_155390_ = 2;
@@ -153,10 +_,8 @@
                   BlockPos blockpos1 = p_155416_.m_142082_(j1, k1, l1);
                   BlockState blockstate = p_155415_.m_8055_(blockpos1);
 
-                  for(Block block : f_59184_) {
-                     if (blockstate.m_60713_(block)) {
-                        p_155417_.add(blockpos1);
-                     }
+                  if (blockstate.isConduitFrame(p_155415_, blockpos1, p_155416_)) {
+                     p_155417_.add(blockpos1);
                   }
                }
             }
@@ -177,7 +_,7 @@
       if (!list.isEmpty()) {
          for(Player player : list) {
             if (p_155445_.m_123314_(player.m_142538_(), (double)j) && player.m_20070_()) {
-               player.m_7292_(new MobEffectInstance(MobEffects.f_19592_, 260, 0, true, true));
+               player.addEffect(new MobEffectInstance(MobEffects.f_19592_, 260, 0, true, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONDUIT); // CraftBukkit
             }
          }
 
@@ -204,8 +_,13 @@
       }
 
       if (p_155413_.f_59189_ != null) {
-         p_155409_.m_6263_((Player)null, p_155413_.f_59189_.m_20185_(), p_155413_.f_59189_.m_20186_(), p_155413_.f_59189_.m_20189_(), SoundEvents.f_11823_, SoundSource.BLOCKS, 1.0F, 1.0F);
-         p_155413_.f_59189_.m_6469_(DamageSource.f_19319_, 4.0F);
+         // CraftBukkit start
+         CraftEventFactory.blockDamage = CraftBlock.at(p_155409_, p_155410_);
+         if (p_155413_.f_59189_.m_6469_(DamageSource.f_19319_, 4.0F)) {
+            p_155409_.m_6263_((Player) null, p_155413_.f_59189_.m_20185_(), p_155413_.f_59189_.m_20186_(), p_155413_.f_59189_.m_20189_(), SoundEvents.f_11823_, SoundSource.BLOCKS, 1.0F, 1.0F);
+         }
+         CraftEventFactory.blockDamage = null;
+         // CraftBukkit end
       }
 
       if (livingentity != p_155413_.f_59189_) {
