--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -12,6 +_,7 @@
 public abstract class ContainerOpenersCounter {
    private static final int f_155447_ = 5;
    private int f_155448_;
+   public boolean opened; // CraftBukkit
 
    protected abstract void m_142292_(Level p_155460_, BlockPos p_155461_, BlockState p_155462_);
 
@@ -22,7 +_,16 @@
    protected abstract boolean m_142718_(Player p_155451_);
 
    public void m_155452_(Player p_155453_, Level p_155454_, BlockPos p_155455_, BlockState p_155456_) {
+      int oldPower = Math.max(0, Math.min(15, this.f_155448_)); // CraftBukkit - Get power before new viewer is added
       int i = this.f_155448_++;
+      // CraftBukkit start - Call redstone event
+      if (p_155454_.m_8055_(p_155455_).m_60713_(net.minecraft.world.level.block.Blocks.f_50325_)) {
+         int newPower = Math.max(0, Math.min(15, this.f_155448_));
+         if (oldPower != newPower) {
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRedstoneChange(p_155454_, p_155455_, oldPower, newPower);
+         }
+      }
+      // CraftBukkit end
       if (i == 0) {
          this.m_142292_(p_155454_, p_155455_, p_155456_);
          p_155454_.m_142346_(p_155453_, GameEvent.f_157803_, p_155455_);
@@ -33,7 +_,17 @@
    }
 
    public void m_155468_(Player p_155469_, Level p_155470_, BlockPos p_155471_, BlockState p_155472_) {
+      int oldPower = Math.max(0, Math.min(15, this.f_155448_)); // CraftBukkit - Get power before new viewer is added
       int i = this.f_155448_--;
+      // CraftBukkit start - Call redstone event
+      if (p_155470_.m_8055_(p_155471_).m_60713_(net.minecraft.world.level.block.Blocks.f_50325_)) {
+         int newPower = Math.max(0, Math.min(15, this.f_155448_));
+
+         if (oldPower != newPower) {
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRedstoneChange(p_155470_, p_155471_, oldPower, newPower);
+         }
+      }
+      // CraftBukkit end
       if (this.f_155448_ == 0) {
          this.m_142289_(p_155470_, p_155471_, p_155472_);
          p_155470_.m_142346_(p_155469_, GameEvent.f_157802_, p_155471_);
@@ -53,6 +_,7 @@
 
    public void m_155476_(Level p_155477_, BlockPos p_155478_, BlockState p_155479_) {
       int i = this.m_155457_(p_155477_, p_155478_);
+      if (opened) i++; // CraftBukkit - add dummy count from API
       int j = this.f_155448_;
       if (j != i) {
          boolean flag = i != 0;
