--- a/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
@@ -20,7 +_,7 @@
 
 public class DecoratedPotBlockEntity extends BlockEntity {
    public static final String f_271111_ = "sherds";
-   private DecoratedPotBlockEntity.Decorations f_283890_ = DecoratedPotBlockEntity.Decorations.f_283770_;
+   public DecoratedPotBlockEntity.Decorations f_283890_ = DecoratedPotBlockEntity.Decorations.f_283770_; // CatServer - private -> public
 
    public DecoratedPotBlockEntity(BlockPos p_273660_, BlockState p_272831_) {
       super(BlockEntityType.f_271291_, p_273660_, p_272831_);
