--- a/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java
@@ -11,7 +_,7 @@
 
 public class EnderChestBlockEntity extends BlockEntity implements LidBlockEntity {
    private final ChestLidController f_155510_ = new ChestLidController();
-   private final ContainerOpenersCounter f_155511_ = new ContainerOpenersCounter() {
+   public final ContainerOpenersCounter f_155511_ = new ContainerOpenersCounter() { // CatServer - private -> public
       protected void m_142292_(Level p_155531_, BlockPos p_155532_, BlockState p_155533_) {
          p_155531_.m_6263_((Player)null, (double)p_155532_.m_123341_() + 0.5D, (double)p_155532_.m_123342_() + 0.5D, (double)p_155532_.m_123343_() + 0.5D, SoundEvents.f_11889_, SoundSource.BLOCKS, 0.5F, p_155531_.f_46441_.m_188501_() * 0.1F + 0.9F);
       }
