--- a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -24,13 +_,56 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.ticks.ContainerSingleItem;
 
+// CraftBukkit start
+import java.util.List;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
 public class JukeboxBlockEntity extends BlockEntity implements Clearable, ContainerSingleItem {
    private static final int f_271500_ = 20;
    private final NonNullList<ItemStack> f_271436_ = NonNullList.m_122780_(this.m_6643_(), ItemStack.f_41583_);
    private int f_238796_;
-   private long f_238695_;
-   private long f_238572_;
-   private boolean f_238637_;
+   public long f_238695_; // CatServer - private -> public
+   public long f_238572_; // CatServer - private -> public
+   public boolean f_238637_; // CatServer - private -> public
+   // CraftBukkit start - 1.20.1 add fields and methods
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private int maxStack = MAX_STACK;
+   public boolean opened;
+
+   @Override
+   public List<ItemStack> getContents() {
+      return this.f_271436_;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+   }
+
+   @Override
+   public Location getLocation() {
+      if (f_58857_ == null) return null;
+      return new org.bukkit.Location(f_58857_.getWorld(), f_58858_.m_123341_(), f_58858_.m_123342_(), f_58858_.m_123343_());
+   }
+   // CraftBukkit end
 
    public JukeboxBlockEntity(BlockPos p_155613_, BlockState p_155614_) {
       super(BlockEntityType.f_58921_, p_155613_, p_155614_);
@@ -139,7 +_,7 @@
    }
 
    public int m_6893_() {
-      return 1;
+      return maxStack; // CraftBukkit - 1.20.1
    }
 
    public boolean m_6542_(Player p_273466_) {
@@ -185,7 +_,11 @@
    @VisibleForTesting
    public void m_272139_(ItemStack p_272693_) {
       this.f_271436_.set(0, p_272693_);
-      this.f_58857_.m_46672_(this.m_58899_(), this.m_58900_().m_60734_());
+      // CraftBukkit start - 1.20.1 add null check for level
+      if (f_58857_ != null) {
+         this.f_58857_.m_46672_(this.m_58899_(), this.m_58900_().m_60734_());
+      }
+      // CraftBukkit end
       this.m_6596_();
    }
 }
