--- a/net/minecraft/world/level/block/entity/LecternBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/LecternBlockEntity.java
@@ -1,5 +_,8 @@
 package net.minecraft.world.level.block.entity;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.commands.CommandSource;
 import net.minecraft.commands.CommandSourceStack;
@@ -25,13 +_,48 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
-
-public class LecternBlockEntity extends BlockEntity implements Clearable, MenuProvider {
+import org.bukkit.Location;
+import org.bukkit.block.Lectern;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+
+import java.util.List;
+
+public class LecternBlockEntity extends BlockEntity implements Clearable, MenuProvider, CommandSource {
    public static final int f_155617_ = 0;
    public static final int f_155618_ = 1;
    public static final int f_155619_ = 0;
    public static final int f_155620_ = 1;
-   private final Container f_59525_ = new Container() {
+   public final Container f_59525_ = new LecternInventory();
+
+   @Override
+   public void m_6352_(Component p_80166_, UUID p_80167_) {
+
+   }
+
+   @Override
+   public boolean m_6999_() {
+      return false;
+   }
+
+   @Override
+   public boolean m_7028_() {
+      return false;
+   }
+
+   @Override
+   public boolean m_6102_() {
+      return false;
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return wrapper.m_81373_() != null ? wrapper.m_81373_().getBukkitSender(wrapper) : new org.bukkit.craftbukkit.v1_18_R2.command.CraftBlockCommandSender(wrapper, this);
+   }
+
+   public class LecternInventory implements Container {
       public int m_6643_() {
          return 1;
       }
@@ -69,10 +_,17 @@
       }
 
       public void m_6836_(int p_59585_, ItemStack p_59586_) {
+         if (p_59585_ == 0) {
+            LecternBlockEntity.this.m_59536_(p_59586_);
+            if (LecternBlockEntity.this.m_58904_() != null) {
+               LecternBlock.m_54497_(LecternBlockEntity.this.m_58904_(), LecternBlockEntity.this.m_58899_(), LecternBlockEntity.this.m_58900_(), LecternBlockEntity.this.m_59567_());
+            }
+         }
+         // CraftBukkit end
       }
 
       public int m_6893_() {
-         return 1;
+         return maxStack;
       }
 
       public void m_6596_() {
@@ -91,8 +_,50 @@
          return false;
       }
 
+      public List<HumanEntity> transaction = new ArrayList<>();
+      private int maxStack = 1;
+
+      @Override
+      public List<ItemStack> getContents() {
+         return Arrays.asList(f_59527_);
+      }
+
+      @Override
+      public void onOpen(CraftHumanEntity who) {
+         transaction.add(who);
+      }
+
+      @Override
+      public void onClose(CraftHumanEntity who) {
+         transaction.remove(who);
+      }
+
+      @Override
+      public List<HumanEntity> getViewers() {
+         return transaction;
+      }
+
+      @Override
+      public InventoryHolder getOwner() {
+         return (Lectern) LecternBlockEntity.this.getOwner();
+      }
+
+      @Override
+      public void setMaxStackSize(int size) {
+         maxStack = size;
+      }
+
+      @Override
+      public Location getLocation() {
+         return new Location(f_58857_.getWorld(), f_58858_.m_123341_(), f_58858_.m_123342_(), f_58858_.m_123343_());
+      }
+
       public void m_6211_() {
       }
+
+      public MenuProvider getLectern() {
+         return LecternBlockEntity.this;
+      }
    };
    private final ContainerData f_59526_ = new ContainerData() {
       public int m_6413_(int p_59600_) {
@@ -143,12 +_,12 @@
       this.m_6596_();
    }
 
-   void m_59532_(int p_59533_) {
+   public void m_59532_(int p_59533_) {
       int i = Mth.m_14045_(p_59533_, 0, this.f_59529_ - 1);
       if (i != this.f_59528_) {
          this.f_59528_ = i;
          this.m_6596_();
-         LecternBlock.m_54488_(this.m_58904_(), this.m_58899_(), this.m_58900_());
+         if (this.f_58857_ != null) LecternBlock.m_54488_(this.m_58904_(), this.m_58899_(), this.m_58900_());
       }
 
    }
@@ -182,7 +_,7 @@
       }
 
       Vec3 vec3 = Vec3.m_82512_(this.f_58858_);
-      return new CommandSourceStack(CommandSource.f_80164_, vec3, Vec2.f_82462_, (ServerLevel)this.f_58857_, 2, s, component, this.f_58857_.m_142572_(), p_59535_);
+      return new CommandSourceStack(this, vec3, Vec2.f_82462_, (ServerLevel)this.f_58857_, 2, s, component, this.f_58857_.m_142572_(), p_59535_);
    }
 
    public boolean m_6326_() {
@@ -215,7 +_,7 @@
    }
 
    public AbstractContainerMenu m_7208_(int p_59562_, Inventory p_59563_, Player p_59564_) {
-      return new LecternMenu(p_59562_, this.f_59525_, this.f_59526_);
+      return new LecternMenu(p_59562_, this.f_59525_, this.f_59526_, p_59563_);
    }
 
    public Component m_5446_() {
