--- a/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
@@ -26,8 +_,8 @@
    public static final String f_155626_ = "LootTable";
    public static final String f_155627_ = "LootTableSeed";
    @Nullable
-   protected ResourceLocation f_59605_;
-   protected long f_59606_;
+   public ResourceLocation f_59605_;
+   public long f_59606_;
 
    protected RandomizableContainerBlockEntity(BlockEntityType<?> p_155629_, BlockPos p_155630_, BlockState p_155631_) {
       super(p_155629_, p_155630_, p_155631_);
