--- a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -55,7 +_,7 @@
       p_222866_.put(3, SoundEvents.f_215768_);
       p_222866_.put(4, SoundEvents.f_215765_);
    });
-   private int f_222831_;
+   public int f_222831_; // CatServer - private -> public
    private final VibrationSystem.User f_279594_ = new SculkShriekerBlockEntity.VibrationUser();
    private VibrationSystem.Data f_279562_ = new VibrationSystem.Data();
    private final VibrationSystem.Listener f_279640_ = new VibrationSystem.Listener(this);
