--- a/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
@@ -15,7 +_,7 @@
 
 public class SculkSensorBlockEntity extends BlockEntity implements VibrationListener.VibrationListenerConfig {
    private final VibrationListener f_155632_;
-   private int f_155633_;
+   public int f_155633_;
 
    public SculkSensorBlockEntity(BlockPos p_155635_, BlockState p_155636_) {
       super(BlockEntityType.f_155257_, p_155635_, p_155636_);
