--- a/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
@@ -23,7 +_,7 @@
    private VibrationSystem.Data f_279553_;
    private final VibrationSystem.Listener f_279626_;
    private final VibrationSystem.User f_279572_ = this.m_280175_();
-   private int f_155633_;
+   public int f_155633_; // CatServer - private -> public
 
    protected SculkSensorBlockEntity(BlockEntityType<?> p_277405_, BlockPos p_277502_, BlockState p_277699_) {
       super(p_277405_, p_277502_, p_277699_);
