--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -31,6 +_,8 @@
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity implements WorldlyContainer {
    public static final int f_155657_ = 9;
@@ -50,6 +_,42 @@
    @Nullable
    private final DyeColor f_59650_;
 
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private int maxStack = MAX_STACK;
+   public boolean opened;
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   @Override
+   public List<ItemStack> getContents() {
+      return f_59645_;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+      this.transaction.add(who);
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+      this.transaction.remove(who);
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return this.transaction;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+      this.maxStack = size;
+   }
+   // CraftBukkit end
+
    public ShulkerBoxBlockEntity(@Nullable DyeColor p_155666_, BlockPos p_155667_, BlockState p_155668_) {
       super(BlockEntityType.f_58939_, p_155667_, p_155668_);
       this.f_59650_ = p_155666_;
@@ -153,6 +_,7 @@
          }
 
          ++this.f_59646_;
+         if (opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
          this.f_58857_.m_7696_(this.f_58858_, this.m_58900_().m_60734_(), 1, this.f_59646_);
          if (this.f_59646_ == 1) {
             this.f_58857_.m_142346_(p_59692_, GameEvent.f_157803_, this.f_58858_);
@@ -165,6 +_,7 @@
    public void m_5785_(Player p_59688_) {
       if (!p_59688_.m_5833_()) {
          --this.f_59646_;
+         if (opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
          this.f_58857_.m_7696_(this.f_58858_, this.m_58900_().m_60734_(), 1, this.f_59646_);
          if (this.f_59646_ <= 0) {
             this.f_58857_.m_142346_(p_59688_, GameEvent.f_157802_, this.f_58858_);
@@ -234,6 +_,11 @@
 
    public boolean m_59702_() {
       return this.f_59647_ == ShulkerBoxBlockEntity.AnimationStatus.CLOSED;
+   }
+
+   @Override
+   protected net.minecraftforge.items.IItemHandler createUnSidedHandler() {
+      return new net.minecraftforge.items.wrapper.SidedInvWrapper(this, Direction.UP);
    }
 
    public static enum AnimationStatus {
