--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -22,8 +_,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
 
-public class SignBlockEntity extends BlockEntity {
+public class SignBlockEntity extends BlockEntity implements CommandSource{
    public static final int f_155692_ = 4;
    private static final String[] f_155693_ = new String[]{"Text1", "Text2", "Text3", "Text4"};
    private static final String[] f_155694_ = new String[]{"FilteredText1", "FilteredText2", "FilteredText3", "FilteredText4"};
@@ -55,6 +_,12 @@
          }
       }
 
+      // CraftBukkit start
+      if (Boolean.getBoolean("convertLegacySigns")) {
+         p_187515_.m_128379_("Bukkit.isConverted", true);
+      }
+      // CraftBukkit end
+
       p_187515_.m_128359_("Color", this.f_59724_.m_41065_());
       p_187515_.m_128379_("GlowingText", this.f_155697_);
    }
@@ -64,8 +_,24 @@
       super.m_142466_(p_155716_);
       this.f_59724_ = DyeColor.m_41057_(p_155716_.m_128461_("Color"), DyeColor.BLACK);
 
+      // CraftBukkit start - Add an option to convert signs correctly
+      // This is done with a flag instead of all the time because
+      // we have no way to tell whether a sign is from 1.7.10 or 1.8
+      boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !p_155716_.m_128471_("Bukkit.isConverted");
+      // CraftBukkit end
+
       for(int i = 0; i < 4; ++i) {
          String s = p_155716_.m_128461_(f_155693_[i]);
+         // CraftBukkit start
+         if (s != null && s.length() > 2048) {
+            s = "\"\"";
+         }
+
+         if (oldSign) {
+            f_59720_[i] = org.bukkit.craftbukkit.v1_18_R2.util.CraftChatMessage.fromString(s)[0];
+            continue;
+         }
+         // CraftBukkit end
          Component component = this.m_155711_(s);
          this.f_59720_[i] = component;
          String s1 = f_155694_[i];
@@ -98,6 +_,10 @@
          if (component != null) {
             return component;
          }
+         // CraftBukkit start
+      } catch (com.google.gson.JsonParseException jsonparseexception) {
+         return new TextComponent(p_155721_);
+         // CraftBukkit end
       } catch (Exception exception) {
       }
 
@@ -183,7 +_,7 @@
    public CommandSourceStack m_59735_(@Nullable ServerPlayer p_59736_) {
       String s = p_59736_ == null ? "Sign" : p_59736_.m_7755_().getString();
       Component component = (Component)(p_59736_ == null ? new TextComponent("Sign") : p_59736_.m_5446_());
-      return new CommandSourceStack(CommandSource.f_80164_, Vec3.m_82512_(this.f_58858_), Vec2.f_82462_, (ServerLevel)this.f_58857_, 2, s, component, this.f_58857_.m_142572_(), p_59736_);
+      return new CommandSourceStack(this, Vec3.m_82512_(this.f_58858_), Vec2.f_82462_, (ServerLevel)this.f_58857_, 2, s, component, this.f_58857_.m_142572_(), p_59736_);
    }
 
    public DyeColor m_59753_() {
@@ -216,6 +_,33 @@
 
    private void m_155728_() {
       this.m_6596_();
-      this.f_58857_.m_7260_(this.m_58899_(), this.m_58900_(), this.m_58900_(), 3);
-   }
+      if (this.f_58857_ != null) this.f_58857_.m_7260_(this.m_58899_(), this.m_58900_(), this.m_58900_(), 3);
+   }
+
+   // CraftBukkit start
+   @Override
+   public void m_6352_(Component p_80166_, UUID p_80167_) {
+
+   }
+
+   @Override
+   public boolean m_6999_() {
+      return false;
+   }
+
+   @Override
+   public boolean m_7028_() {
+      return false;
+   }
+
+   @Override
+   public boolean m_6102_() {
+      return false;
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return wrapper.m_81373_() != null ? wrapper.m_81373_().getBukkitSender(wrapper) : new org.bukkit.craftbukkit.v1_18_R2.command.CraftBlockCommandSender(wrapper, this);
+   }
+   // CraftBukkit end
 }
