--- a/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -30,9 +_,9 @@
    @Nullable
    private static Executor f_182457_;
    @Nullable
-   private GameProfile f_59757_;
+   public GameProfile f_59757_; // CatServer - private -> public
    @Nullable
-   private ResourceLocation f_262250_;
+   public ResourceLocation f_262250_; // CatServer - private -> public
    private int f_260504_;
    private boolean f_260642_;
 
