--- a/net/minecraft/world/level/block/entity/StructureBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/StructureBlockEntity.java
@@ -36,7 +_,7 @@
    public static final int f_155775_ = 48;
    public static final String f_155776_ = "author";
    private ResourceLocation f_59812_;
-   private String f_59813_ = "";
+   public String f_59813_ = "";
    private String f_59814_ = "";
    private BlockPos f_59815_ = new BlockPos(0, 1, 0);
    private Vec3i f_59816_ = Vec3i.f_123288_;
@@ -45,8 +_,8 @@
    private StructureMode f_59819_;
    private boolean f_59820_ = true;
    private boolean f_59821_;
-   private boolean f_59822_;
-   private boolean f_59823_ = true;
+   public boolean f_59822_;
+   public boolean f_59823_ = true;
    private float f_59824_ = 1.0F;
    private long f_59825_;
 
