--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -28,6 +_,10 @@
 import net.minecraft.world.level.levelgen.feature.configurations.EndGatewayConfiguration;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerTeleportEvent;
 import org.slf4j.Logger;
 
 public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
@@ -37,11 +_,11 @@
    private static final int f_155809_ = 2400;
    private static final int f_155810_ = 1;
    private static final int f_155811_ = 10;
-   private long f_59926_;
+   public long f_59926_;
    private int f_59927_;
    @Nullable
-   private BlockPos f_59928_;
-   private boolean f_59929_;
+   public BlockPos f_59928_;
+   public boolean f_59929_;
 
    public TheEndGatewayBlockEntity(BlockPos p_155813_, BlockState p_155814_) {
       super(BlockEntityType.f_58937_, p_155813_, p_155814_);
@@ -180,6 +_,24 @@
             } else {
                entity = p_155832_.m_20201_();
             }
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (entity instanceof ServerPlayer) {
+               org.bukkit.craftbukkit.v1_18_R2.entity.CraftPlayer player = (CraftPlayer) entity.getBukkitEntity();
+               org.bukkit.Location location = new Location(p_155829_.getWorld(), (double) blockpos1.m_123341_() + 0.5D, (double) blockpos1.m_123342_() + 0.5D, (double) blockpos1.m_123343_() + 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+               PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+               entity.m_20091_();
+               ((ServerPlayer) entity).f_8906_.teleport(teleEvent.getTo());
+               m_155849_(p_155829_, p_155830_, p_155831_, p_155833_); // CraftBukkit - call at end of method
+               return;
+            }
+            // CraftBukkit end
 
             entity.m_20091_();
             entity.m_20324_((double)blockpos1.m_123341_() + 0.5D, (double)blockpos1.m_123342_(), (double)blockpos1.m_123343_() + 0.5D);
