--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
 
 public class RconConsoleSource implements CommandSource {
    private static final String f_144022_ = "Rcon";
@@ -35,7 +_,7 @@
    }
 
    public void m_6352_(Component p_11509_, UUID p_11510_) {
-      this.f_11501_.append(p_11509_.getString());
+      this.f_11501_.append(p_11509_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean m_6999_() {
@@ -48,5 +_,10 @@
 
    public boolean m_6102_() {
       return this.f_11502_.m_6983_();
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return null;
    }
 }
