--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -9,15 +_,24 @@
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import java.net.SocketAddress;
+import org.bukkit.craftbukkit.v1_20_R1.command.CraftRemoteConsoleCommandSender;
+// CraftBukkit end
+
 public class RconConsoleSource implements CommandSource {
    private static final String f_144022_ = "Rcon";
    private static final Component f_11500_ = Component.m_237113_("Rcon");
    private final StringBuffer f_11501_ = new StringBuffer();
    private final MinecraftServer f_11502_;
+   // CraftBukkit start - 1.20.1
+   public SocketAddress socketAddress; // CatServer - 1.20.1 public final -> public
+   private final CraftRemoteConsoleCommandSender remoteConsole = new CraftRemoteConsoleCommandSender(this);
 
    public RconConsoleSource(MinecraftServer p_11505_) {
       this.f_11502_ = p_11505_;
    }
+   // CatServer end
 
    public void m_11512_() {
       this.f_11501_.setLength(0);
@@ -32,8 +_,19 @@
       return new CommandSourceStack(this, Vec3.m_82528_(serverlevel.m_220360_()), Vec2.f_82462_, serverlevel, 4, "Rcon", f_11500_, this.f_11502_, (Entity)null);
    }
 
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.f_11501_.append(message);
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return this.remoteConsole;
+   }
+   // CraftBukkit end
+
    public void m_213846_(Component p_215653_) {
-      this.f_11501_.append(p_215653_.getString());
+      this.f_11501_.append(p_215653_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean m_6999_() {
