--- a/net/minecraft/world/entity/npc/ClientSideMerchant.java
+++ b/net/minecraft/world/entity/npc/ClientSideMerchant.java
@@ -44,6 +_,13 @@
       return this.f_35340_.m_9236_().f_46443_;
    }
 
+   // CatServer start
+   @Override
+   public org.bukkit.craftbukkit.v1_20_R1.inventory.CraftMerchant getCraftMerchant() {
+      return null;
+   }
+   // CatServer end
+
    public int m_7809_() {
       return this.f_35342_;
    }
