--- a/net/minecraft/world/entity/npc/ClientSideMerchant.java
+++ b/net/minecraft/world/entity/npc/ClientSideMerchant.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchant;
 
 public class ClientSideMerchant implements Merchant {
    private final Player f_35340_;
@@ -42,6 +_,11 @@
 
    public boolean m_183595_() {
       return this.f_35340_.m_183503_().f_46443_;
+   }
+
+   @Override
+   public CraftMerchant getCraftMerchant() {
+      return null;
    }
 
    public int m_7809_() {
