--- a/net/minecraft/world/entity/npc/InventoryCarrier.java
+++ b/net/minecraft/world/entity/npc/InventoryCarrier.java
@@ -20,6 +_,13 @@
             return;
          }
 
+         // CraftBukkit start - 1.20.1
+         ItemStack remaining = new SimpleContainer(simplecontainer).m_19173_(itemstack);
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityPickupItemEvent(p_219612_, p_219614_, remaining.m_41613_(), false).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
+
          p_219612_.m_21053_(p_219614_);
          int i = itemstack.m_41613_();
          ItemStack itemstack1 = simplecontainer.m_19173_(itemstack);
