--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -86,6 +_,10 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.VillagerReplenishTradeEvent;
 import org.slf4j.Logger;
 
 public class Villager extends AbstractVillager implements ReputationEventHandler, VillagerDataHolder {
@@ -203,6 +_,17 @@
       return this.f_35366_;
    }
 
+   // Spigot Start
+   @Override
+   public void inactiveTick() {
+      // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+      if (f_19853_.spigotConfig.tickInactiveVillagers && this.m_6142_()) {
+         this.m_8024_();
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected void m_8024_() {
       this.f_19853_.m_46473_().m_6180_("villagerBrain");
       this.m_6274_().m_21865_((ServerLevel)this.f_19853_, this);
@@ -219,7 +_,7 @@
                this.f_35374_ = false;
             }
 
-            this.m_7292_(new MobEffectInstance(MobEffects.f_19605_, 200, 0));
+            this.addEffect(new MobEffectInstance(MobEffects.f_19605_, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
          }
       }
 
@@ -254,7 +_,7 @@
 
    public InteractionResult m_6071_(Player p_35472_, InteractionHand p_35473_) {
       ItemStack itemstack = p_35472_.m_21120_(p_35473_);
-      if (!itemstack.m_150930_(Items.f_42601_) && this.m_6084_() && !this.m_35306_() && !this.m_5803_()) {
+      if (itemstack.m_41720_() != Items.f_42601_ && this.m_6084_() && !this.m_35306_() && !this.m_5803_() && !p_35472_.m_36341_()) {
          if (this.m_6162_()) {
             this.m_35518_();
             return InteractionResult.m_19078_(this.f_19853_.f_46443_);
@@ -283,7 +_,7 @@
       }
    }
 
-   private void m_35518_() {
+   public void m_35518_() {
       this.m_35319_(40);
       if (!this.f_19853_.m_5776_()) {
          this.m_5496_(SoundEvents.f_12507_, this.m_6121_(), this.m_6100_());
@@ -330,7 +_,13 @@
       this.m_35523_();
 
       for(MerchantOffer merchantoffer : this.m_6616_()) {
-         merchantoffer.m_45372_();
+         // CraftBukkit start
+         VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), merchantoffer.asBukkit());
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            merchantoffer.m_45372_();
+         }
+         // CraftBukkit end
       }
 
       this.f_35363_ = this.f_19853_.m_46467_();
@@ -375,7 +_,13 @@
       int i = 2 - this.f_35364_;
       if (i > 0) {
          for(MerchantOffer merchantoffer : this.m_6616_()) {
-            merchantoffer.m_45372_();
+            // CraftBukkit start
+            VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), merchantoffer.asBukkit());
+            Bukkit.getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               merchantoffer.m_45372_();
+            }
+            // CraftBukkit end
          }
       }
 
@@ -548,7 +_,7 @@
    }
 
    public void m_6667_(DamageSource p_35419_) {
-      f_201976_.info("Villager {} died, message: '{}'", this, p_35419_.m_6157_(this).getString());
+      if (org.spigotmc.SpigotConfig.logVillagerDeaths) f_201976_.info("Villager {} died, message: '{}'", this, p_35419_.m_6157_(this).getString()); // Spigot
       Entity entity = p_35419_.m_7639_();
       if (entity != null) {
          this.m_35420_(entity);
@@ -558,7 +_,7 @@
       super.m_6667_(p_35419_);
    }
 
-   private void m_35524_() {
+   public void m_35524_() {
       this.m_35428_(MemoryModuleType.f_26359_);
       this.m_35428_(MemoryModuleType.f_26360_);
       this.m_35428_(MemoryModuleType.f_26361_);
@@ -658,7 +_,8 @@
    }
 
    protected Component m_5677_() {
-      return new TranslatableComponent(this.m_6095_().m_20675_() + "." + Registry.f_122869_.m_7981_(this.m_7141_().m_35571_()).m_135815_());
+      net.minecraft.resources.ResourceLocation profName = this.m_7141_().m_35571_().getRegistryName();
+      return new TranslatableComponent(this.m_6095_().m_20675_() + '.' + (!"minecraft".equals(profName.m_135827_()) ? profName.m_135827_() + '.' : "") + profName.m_135815_());
    }
 
    public void m_7822_(byte p_35391_) {
@@ -710,7 +_,7 @@
    }
 
    public void m_8038_(ServerLevel p_35409_, LightningBolt p_35410_) {
-      if (p_35409_.m_46791_() != Difficulty.PEACEFUL) {
+      if (p_35409_.m_46791_() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.f_20495_, (timer) -> {})) {
          f_201976_.info("Villager {} was struck by lightning {}.", this, p_35410_);
          Witch witch = EntityType.f_20495_.m_20615_(p_35409_);
          witch.m_7678_(this.m_20185_(), this.m_20186_(), this.m_20189_(), this.m_146908_(), this.m_146909_());
@@ -722,7 +_,13 @@
          }
 
          witch.m_21530_();
-         p_35409_.m_47205_(witch);
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, witch);
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityTransformEvent(this, witch, EntityTransformEvent.TransformReason.LIGHTNING).isCancelled()) {
+            return;
+         }
+         p_35409_.addFreshEntityWithPassengers(witch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+         // CraftBukkit end
          this.m_35524_();
          this.m_146870_();
       } else {
@@ -740,6 +_,13 @@
             return;
          }
 
+         // CraftBukkit start
+         ItemStack remaining = new SimpleContainer(simplecontainer).m_19173_(itemstack);
+         if (CraftEventFactory.callEntityPickupItemEvent(this, p_35467_, remaining.m_41613_(), false).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
+
          this.m_21053_(p_35467_);
          this.m_7938_(p_35467_, itemstack.m_41613_());
          ItemStack itemstack1 = simplecontainer.m_19173_(itemstack);
@@ -843,7 +_,7 @@
             IronGolem irongolem = EntityType.f_20460_.m_20655_(p_35491_, (CompoundTag)null, (Component)null, (Player)null, blockpos1, MobSpawnType.MOB_SUMMONED, false, false);
             if (irongolem != null) {
                if (irongolem.m_5545_(p_35491_, MobSpawnType.MOB_SUMMONED) && irongolem.m_6914_(p_35491_)) {
-                  p_35491_.m_47205_(irongolem);
+                  p_35491_.addFreshEntityWithPassengers(irongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE); // CraftBukkit
                   return irongolem;
                }
 
