--- a/net/minecraft/world/entity/npc/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/WanderingTraderSpawner.java
@@ -104,14 +_,14 @@
                return false;
             }
 
-            WanderingTrader wanderingtrader = EntityType.f_20494_.m_20600_(p_35916_, (CompoundTag)null, (Component)null, (Player)null, blockpos2, MobSpawnType.EVENT, false, false);
+            WanderingTrader wanderingtrader = EntityType.f_20494_.spawn(p_35916_, (CompoundTag)null, (Component)null, (Player)null, blockpos2, MobSpawnType.EVENT, false, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
             if (wanderingtrader != null) {
                for(int j = 0; j < 2; ++j) {
                   this.m_35917_(p_35916_, wanderingtrader, 4);
                }
 
                this.f_35909_.m_8115_(wanderingtrader.m_142081_());
-               wanderingtrader.m_35891_(48000);
+               // wanderingtrader.setDespawnDelay(48000); // CraftBukkit - moved to EntityVillagerTrader constructor. This lets the value be modified by plugins on CreatureSpawnEvent
                wanderingtrader.m_35883_(blockpos1);
                wanderingtrader.m_21446_(blockpos1, 16);
                return true;
@@ -125,7 +_,7 @@
    private void m_35917_(ServerLevel p_35918_, WanderingTrader p_35919_, int p_35920_) {
       BlockPos blockpos = this.m_35928_(p_35918_, p_35919_.m_142538_(), p_35920_);
       if (blockpos != null) {
-         TraderLlama traderllama = EntityType.f_20488_.m_20600_(p_35918_, (CompoundTag)null, (Component)null, (Player)null, blockpos, MobSpawnType.EVENT, false, false);
+         TraderLlama traderllama = EntityType.f_20488_.spawn(p_35918_, (CompoundTag)null, (Component)null, (Player)null, blockpos, MobSpawnType.EVENT, false, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
          if (traderllama != null) {
             traderllama.m_21463_(p_35919_, true);
          }
