--- a/net/minecraft/world/entity/npc/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/WanderingTrader.java
@@ -43,6 +_,9 @@
 import net.minecraft.world.item.trading.MerchantOffers;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchantRecipe;
+import org.bukkit.event.entity.VillagerAcquireTradeEvent;
 
 public class WanderingTrader extends AbstractVillager {
    private static final int f_150044_ = 5;
@@ -52,6 +_,7 @@
 
    public WanderingTrader(EntityType<? extends WanderingTrader> p_35843_, Level p_35844_) {
       super(p_35843_, p_35844_);
+      this.m_35891_(48000); // CraftBukkit - set default from MobSpawnerTrader
    }
 
    protected void m_8099_() {
@@ -121,6 +_,16 @@
          MerchantOffer merchantoffer = villagertrades$itemlisting.m_5670_(this, this.f_19796_);
          if (merchantoffer != null) {
             merchantoffers.add(merchantoffer);
+            // CraftBukkit start
+            VillagerAcquireTradeEvent event = new VillagerAcquireTradeEvent((org.bukkit.entity.AbstractVillager) getBukkitEntity(), merchantoffer.asBukkit());
+            // Suppress during worldgen
+            if (this.valid) {
+               Bukkit.getPluginManager().callEvent(event);
+            }
+            if (!event.isCancelled()) {
+               merchantoffers.add(CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+            }
+            // CraftBukkit end
          }
 
       }
