--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -10,6 +_,13 @@
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R1.CraftGameEvent;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation;
+import org.bukkit.event.world.GenericGameEvent;
+// CraftBukkit end
+
 public class GameEventDispatcher {
    private final ServerLevel f_243917_;
 
@@ -20,6 +_,14 @@
    public void m_245905_(GameEvent p_251754_, Vec3 p_250613_, GameEvent.Context p_251777_) {
       int i = p_251754_.m_157827_();
       BlockPos blockpos = BlockPos.m_274446_(p_250613_);
+      // CraftBukkit start - 1.20.1
+      GenericGameEvent event = new GenericGameEvent(CraftGameEvent.minecraftToBukkit(p_251754_), CraftLocation.toBukkit(blockpos, f_243917_.getWorld()), (p_251777_.f_223711_() == null) ? null : p_251777_.f_223711_().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+      f_243917_.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      i = event.getRadius();
+      // CraftBukkit end
       int j = SectionPos.m_123171_(blockpos.m_123341_() - i);
       int k = SectionPos.m_123171_(blockpos.m_123342_() - i);
       int l = SectionPos.m_123171_(blockpos.m_123343_() - i);
