--- a/net/minecraft/command/arguments/DimensionArgument.java
+++ b/net/minecraft/command/arguments/DimensionArgument.java
@@ -17,6 +_,7 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 
