--- a/net/minecraft/command/arguments/EntityArgument.java
+++ b/net/minecraft/command/arguments/EntityArgument.java
@@ -87,9 +_,15 @@
    }
 
    public EntitySelector parse(StringReader p_parse_1_) throws CommandSyntaxException {
+      // CraftBukkit start
+      return parse(p_parse_1_, false);
+   }
+
+   public EntitySelector parse(StringReader p_parse_1_, boolean overridePermissions) throws CommandSyntaxException {
+      // CraftBukkit end
       int i = 0;
       EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_parse_1_);
-      EntitySelector entityselector = entityselectorparser.func_201345_m();
+      EntitySelector entityselector = entityselectorparser.parse(overridePermissions);
       if (entityselector.func_197346_a() > 1 && this.field_197103_f) {
          if (this.field_197104_g) {
             p_parse_1_.setCursor(0);
