--- a/net/minecraft/network/chat/TextColor.java
+++ b/net/minecraft/network/chat/TextColor.java
@@ -11,7 +_,7 @@
 public final class TextColor {
    private static final String f_178538_ = "#";
    private static final Map<ChatFormatting, TextColor> f_131255_ = Stream.of(ChatFormatting.values()).filter(ChatFormatting::m_126664_).collect(ImmutableMap.toImmutableMap(Function.identity(), (p_131276_) -> {
-      return new TextColor(p_131276_.m_126665_(), p_131276_.m_126666_());
+      return new TextColor(p_131276_.m_126665_(), p_131276_.m_126666_(), p_131276_); // CraftBukkit
    }));
    private static final Map<String, TextColor> f_131256_ = f_131255_.values().stream().collect(ImmutableMap.toImmutableMap((p_131273_) -> {
       return p_131273_.f_131258_;
@@ -20,14 +_,26 @@
    @Nullable
    private final String f_131258_;
 
+   // CraftBukkit start
+   @Nullable
+   public final ChatFormatting format;
+
+   private TextColor(int p_131263_, String p_131264_, ChatFormatting format) {
+      this.f_131257_ = p_131263_;
+      this.f_131258_ = p_131264_;
+      this.format = format;
+   }
+
    private TextColor(int p_131263_, String p_131264_) {
       this.f_131257_ = p_131263_;
       this.f_131258_ = p_131264_;
+      this.format = null;
    }
 
    private TextColor(int p_131261_) {
       this.f_131257_ = p_131261_;
       this.f_131258_ = null;
+      this.format = null;
    }
 
    public int m_131265_() {
