--- a/net/minecraft/world/raid/RaidManager.java
+++ b/net/minecraft/world/raid/RaidManager.java
@@ -24,9 +_,10 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.storage.WorldSavedData;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class RaidManager extends WorldSavedData {
-   private final Map<Integer, Raid> field_215175_a = Maps.newHashMap();
+   public final Map<Integer, Raid> field_215175_a = Maps.newHashMap();
    private final ServerWorld field_215176_b;
    private int field_215177_c;
    private int field_215178_d;
@@ -108,19 +_,32 @@
             Raid raid = this.func_215168_a(p_215170_1_.func_71121_q(), blockpos1);
             boolean flag = false;
             if (!raid.func_221301_k()) {
-               if (!this.field_215175_a.containsKey(raid.func_221325_u())) {
-                  this.field_215175_a.put(raid.func_221325_u(), raid);
+               /* CraftBukkit - moved down
+               if (!this.byId.containsKey(raid.getId())) {
+                  this.byId.put(raid.getId(), raid);
                }
+               */
 
                flag = true;
-            } else if (raid.func_221291_n() < raid.func_221327_m()) {
+               // CraftBukkit start - fixed a bug with raid: players could add up Bad Omen level even when the raid had finished
+            } else if (raid.isInProgress() && raid.func_221291_n() < raid.func_221327_m()) {
                flag = true;
+               // CraftBukkit end
             } else {
                p_215170_1_.func_195063_d(Effects.field_220309_E);
                p_215170_1_.field_71135_a.func_147359_a(new SEntityStatusPacket(p_215170_1_, (byte)43));
             }
 
             if (flag) {
+               // CraftBukkit start
+               if (!CraftEventFactory.callRaidTriggerEvent(raid, p_215170_1_)) {
+                  p_215170_1_.func_195063_d(Effects.field_220309_E);
+                  return null;
+               }
+               if (!this.field_215175_a.containsKey(raid.func_221325_u())) {
+                  this.field_215175_a.put(raid.func_221325_u(), raid);
+               }
+               // CraftBukkit end
                raid.func_221309_a(p_215170_1_);
                p_215170_1_.field_71135_a.func_147359_a(new SEntityStatusPacket(p_215170_1_, (byte)43));
                if (!raid.func_221297_c()) {
