--- a/net/minecraft/world/raid/Raid.java
+++ b/net/minecraft/world/raid/Raid.java
@@ -65,21 +_,21 @@
    private static final ITextComponent field_221342_f = field_221338_b.func_230532_e_().func_240702_b_(" - ").func_230529_a_(field_221340_d);
    private final Map<Integer, AbstractRaiderEntity> field_221343_g = Maps.newHashMap();
    private final Map<Integer, Set<AbstractRaiderEntity>> field_221344_h = Maps.newHashMap();
-   private final Set<UUID> field_221345_i = Sets.newHashSet();
-   private long field_221346_j;
+   public final Set<UUID> field_221345_i = Sets.newHashSet();
+   public long field_221346_j;
    private BlockPos field_221347_k;
    private final ServerWorld field_221348_l;
    private boolean field_221349_m;
    private final int field_221350_n;
    private float field_221351_o;
-   private int field_221352_p;
+   public int field_221352_p;
    private boolean field_221353_q;
    private int field_221354_r;
    private final ServerBossInfo field_221355_s = new ServerBossInfo(field_221338_b, BossInfo.Color.RED, BossInfo.Overlay.NOTCHED_10);
    private int field_221356_t;
    private int field_221357_u;
    private final Random field_221358_v = new Random();
-   private final int field_221359_w;
+   public final int field_221359_w;
    private Raid.Status field_221360_x;
    private int field_221361_y;
    private Optional<BlockPos> field_221362_z = Optional.empty();
@@ -144,6 +_,12 @@
       return this.field_221360_x == Raid.Status.LOSS;
    }
 
+   // CraftBukkit start
+   public boolean isInProgress() {
+      return this.field_221360_x == Status.ONGOING;
+   }
+   // CraftBukkit end
+
    public World func_221316_i() {
       return this.field_221348_l;
    }
@@ -210,6 +_,7 @@
             boolean flag = this.field_221353_q;
             this.field_221353_q = this.field_221348_l.func_175667_e(this.field_221347_k);
             if (this.field_221348_l.func_175659_aa() == Difficulty.PEACEFUL) {
+               org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                this.func_221295_o();
                return;
             }
@@ -229,13 +_,16 @@
             if (!this.field_221348_l.func_217483_b_(this.field_221347_k)) {
                if (this.field_221354_r > 0) {
                   this.field_221360_x = Raid.Status.LOSS;
+                  org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidFinishEvent(this, new java.util.ArrayList<>()); // CraftBukkit
                } else {
+                  org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.NOT_IN_VILLAGE); // CraftBukkit
                   this.func_221295_o();
                }
             }
 
             ++this.field_221346_j;
             if (this.field_221346_j >= 48000L) {
+               org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.TIMEOUT); // CraftBukkit
                this.func_221295_o();
                return;
             }
@@ -306,6 +_,7 @@
                }
 
                if (k > 3) {
+                  org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.UNSPAWNABLE);  // CraftBukkit
                   this.func_221295_o();
                   break;
                }
@@ -316,7 +_,7 @@
                   ++this.field_221356_t;
                } else {
                   this.field_221360_x = Raid.Status.VICTORY;
-
+                  List<org.bukkit.entity.Player> winners = new java.util.ArrayList<>(); // CraftBukkit
                   for(UUID uuid : this.field_221345_i) {
                      Entity entity = this.field_221348_l.func_217461_a(uuid);
                      if (entity instanceof LivingEntity && !entity.func_175149_v()) {
@@ -326,9 +_,11 @@
                            ServerPlayerEntity serverplayerentity = (ServerPlayerEntity)livingentity;
                            serverplayerentity.func_195066_a(Stats.field_219742_az);
                            CriteriaTriggers.field_215101_H.func_192215_a(serverplayerentity);
+                           winners.add(serverplayerentity.getBukkitEntity()); // CraftBukkit
                         }
                      }
                   }
+                  org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidFinishEvent(this, winners); // CraftBukkit
                }
             }
 
@@ -336,6 +_,7 @@
          } else if (this.func_221319_a()) {
             ++this.field_221361_y;
             if (this.field_221361_y >= 600) {
+               org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.FINISHED); // CraftBukkit
                this.func_221295_o();
                return;
             }
@@ -457,6 +_,10 @@
       DifficultyInstance difficultyinstance = this.field_221348_l.func_175649_E(p_221294_1_);
       boolean flag1 = this.func_221305_D();
 
+      // CraftBukkit start
+      AbstractRaiderEntity leader = null;
+      List<AbstractRaiderEntity> raiders = new java.util.ArrayList<>();
+      // CraftBukkit end
       for(Raid.WaveMember raid$wavemember : Raid.WaveMember.field_221284_f) {
          int j = this.func_221330_a(raid$wavemember, i, flag1) + this.func_221335_a(raid$wavemember, this.field_221358_v, i, difficultyinstance, flag1);
          int k = 0;
@@ -467,9 +_,11 @@
                abstractraiderentity.func_213635_r(true);
                this.func_221324_a(i, abstractraiderentity);
                flag = true;
+               leader = abstractraiderentity; // CraftBukkit
             }
 
             this.func_221317_a(i, abstractraiderentity, p_221294_1_, false);
+            raiders.add(abstractraiderentity); // CraftBukkit
             if (raid$wavemember.field_221285_g == EntityType.field_220352_aU) {
                AbstractRaiderEntity abstractraiderentity1 = null;
                if (i == this.func_221306_a(Difficulty.NORMAL)) {
@@ -487,6 +_,7 @@
                   this.func_221317_a(i, abstractraiderentity1, p_221294_1_, false);
                   abstractraiderentity1.func_174828_a(p_221294_1_, 0.0F, 0.0F);
                   abstractraiderentity1.func_184220_m(abstractraiderentity);
+                  raiders.add(abstractraiderentity); // CraftBukkit
                }
             }
          }
@@ -496,6 +_,7 @@
       ++this.field_221354_r;
       this.func_221320_q();
       this.func_221331_G();
+      org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callRaidSpawnWaveEvent(this, leader, raiders); // CraftBukkit
    }
 
    public void func_221317_a(int p_221317_1_, AbstractRaiderEntity p_221317_2_, @Nullable BlockPos p_221317_3_, boolean p_221317_4_) {
@@ -510,7 +_,7 @@
             p_221317_2_.func_213386_a(this.field_221348_l, this.field_221348_l.func_175649_E(p_221317_3_), SpawnReason.EVENT, (ILivingEntityData)null, (CompoundNBT)null);
             p_221317_2_.func_213660_a(p_221317_1_, false);
             p_221317_2_.func_230245_c_(true);
-            this.field_221348_l.func_242417_l(p_221317_2_);
+            this.field_221348_l.addAllEntities(p_221317_2_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.RAID); // CraftBukkit
          }
       }
 
@@ -769,7 +_,13 @@
       }
    }
 
-   static enum WaveMember {
+   // CraftBukkit start - a method to get all raiders
+   public java.util.Collection<AbstractRaiderEntity> getRaiders() {
+      return this.field_221344_h.values().stream().flatMap(Set::stream).collect(java.util.stream.Collectors.toSet());
+   }
+   // CraftBukkit end
+
+   public static enum WaveMember implements net.minecraftforge.common.IExtensibleEnum {
       VINDICATOR(EntityType.field_200758_ax, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
       EVOKER(EntityType.field_200806_t, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
       PILLAGER(EntityType.field_220350_aJ, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
@@ -783,6 +_,20 @@
       private WaveMember(EntityType<? extends AbstractRaiderEntity> p_i50602_3_, int[] p_i50602_4_) {
          this.field_221285_g = p_i50602_3_;
          this.field_221286_h = p_i50602_4_;
+      }
+      
+      /**
+       * The waveCountsIn integer decides how many entities of the EntityType defined in typeIn will spawn in each wave.
+       * For example, one ravager will always spawn in wave 3.
+       */
+      public static WaveMember create(String name, EntityType<? extends AbstractRaiderEntity> typeIn, int[] waveCountsIn) {
+         throw new IllegalStateException("Enum not extended");
+      }
+      
+      @Override
+      @Deprecated
+      public void init() {
+         field_221284_f = values();
       }
    }
 }
