--- ../src-base/minecraft/net/minecraft/client/gui/toasts/AdvancementToast.java
+++ ../src-work/minecraft/net/minecraft/client/gui/toasts/AdvancementToast.java
@@ -25,7 +25,7 @@
         this.advancement = advancementIn;
     }
 
-    public IToast.Visibility draw(GuiToast toastGui, long delta)
+    public Visibility draw(GuiToast toastGui, long delta)
     {
         toastGui.getMinecraft().getTextureManager().bindTexture(TEXTURE_TOASTS);
         GlStateManager.color(1.0F, 1.0F, 1.0F);
@@ -77,11 +77,11 @@
 
             RenderHelper.enableGUIStandardItemLighting();
             toastGui.getMinecraft().getRenderItem().renderItemAndEffectIntoGUI((EntityLivingBase)null, displayinfo.getIcon(), 8, 8);
-            return delta >= 5000L ? IToast.Visibility.HIDE : IToast.Visibility.SHOW;
+            return delta >= 5000L ? Visibility.HIDE : Visibility.SHOW;
         }
         else
         {
-            return IToast.Visibility.HIDE;
+            return Visibility.HIDE;
         }
     }
 }
