--- ../src-base/minecraft/net/minecraft/client/gui/toasts/GuiToast.java
+++ ../src-work/minecraft/net/minecraft/client/gui/toasts/GuiToast.java
@@ -17,7 +17,7 @@
 public class GuiToast extends Gui
 {
     private final Minecraft mc;
-    private final GuiToast.ToastInstance<?>[] visible = new GuiToast.ToastInstance[5];
+    private final ToastInstance<?>[] visible = new ToastInstance[5];
     private final Deque<IToast> toastsQueue = Queues.<IToast>newArrayDeque();
 
     public GuiToast(Minecraft mcIn)
@@ -33,7 +33,7 @@
 
             for (int i = 0; i < this.visible.length; ++i)
             {
-                GuiToast.ToastInstance<?> toastinstance = this.visible[i];
+                ToastInstance<?> toastinstance = this.visible[i];
 
                 if (toastinstance != null && toastinstance.render(resolution.getScaledWidth(), i))
                 {
@@ -42,7 +42,7 @@
 
                 if (this.visible[i] == null && !this.toastsQueue.isEmpty())
                 {
-                    this.visible[i] = new GuiToast.ToastInstance(this.toastsQueue.removeFirst());
+                    this.visible[i] = new ToastInstance(this.toastsQueue.removeFirst());
                 }
             }
         }
@@ -51,7 +51,7 @@
     @Nullable
     public <T extends IToast> T getToast(Class <? extends T > p_192990_1_, Object p_192990_2_)
     {
-        for (GuiToast.ToastInstance<?> toastinstance : this.visible)
+        for (ToastInstance<?> toastinstance : this.visible)
         {
             if (toastinstance != null && p_192990_1_.isAssignableFrom(toastinstance.getToast().getClass()) && toastinstance.getToast().getType().equals(p_192990_2_))
             {
