--- ../src-base/minecraft/net/minecraft/client/gui/toasts/IToast.java
+++ ../src-work/minecraft/net/minecraft/client/gui/toasts/IToast.java
@@ -14,7 +14,7 @@
     ResourceLocation TEXTURE_TOASTS = new ResourceLocation("textures/gui/toasts.png");
     Object NO_TOKEN = new Object();
 
-    IToast.Visibility draw(GuiToast toastGui, long delta);
+    Visibility draw(GuiToast toastGui, long delta);
 
 default Object getType()
     {
