--- ../src-base/minecraft/net/minecraft/client/gui/toasts/SystemToast.java
+++ ../src-work/minecraft/net/minecraft/client/gui/toasts/SystemToast.java
@@ -9,20 +9,20 @@
 @SideOnly(Side.CLIENT)
 public class SystemToast implements IToast
 {
-    private final SystemToast.Type type;
+    private final Type type;
     private String title;
     private String subtitle;
     private long firstDrawTime;
     private boolean newDisplay;
 
-    public SystemToast(SystemToast.Type typeIn, ITextComponent titleComponent, @Nullable ITextComponent subtitleComponent)
+    public SystemToast(Type typeIn, ITextComponent titleComponent, @Nullable ITextComponent subtitleComponent)
     {
         this.type = typeIn;
         this.title = titleComponent.getUnformattedText();
         this.subtitle = subtitleComponent == null ? null : subtitleComponent.getUnformattedText();
     }
 
-    public IToast.Visibility draw(GuiToast toastGui, long delta)
+    public Visibility draw(GuiToast toastGui, long delta)
     {
         if (this.newDisplay)
         {
@@ -44,7 +44,7 @@
             toastGui.getMinecraft().fontRenderer.drawString(this.subtitle, 18, 18, -1);
         }
 
-        return delta - this.firstDrawTime < 5000L ? IToast.Visibility.SHOW : IToast.Visibility.HIDE;
+        return delta - this.firstDrawTime < 5000L ? Visibility.SHOW : Visibility.HIDE;
     }
 
     public void setDisplayedText(ITextComponent titleComponent, @Nullable ITextComponent subtitleComponent)
@@ -54,12 +54,12 @@
         this.newDisplay = true;
     }
 
-    public SystemToast.Type getType()
+    public Type getType()
     {
         return this.type;
     }
 
-    public static void addOrUpdate(GuiToast p_193657_0_, SystemToast.Type p_193657_1_, ITextComponent p_193657_2_, @Nullable ITextComponent p_193657_3_)
+    public static void addOrUpdate(GuiToast p_193657_0_, Type p_193657_1_, ITextComponent p_193657_2_, @Nullable ITextComponent p_193657_3_)
     {
         SystemToast systemtoast = (SystemToast)p_193657_0_.getToast(SystemToast.class, p_193657_1_);
 
