--- ../src-base/minecraft/net/minecraft/client/gui/toasts/TutorialToast.java
+++ ../src-work/minecraft/net/minecraft/client/gui/toasts/TutorialToast.java
@@ -11,16 +11,16 @@
 @SideOnly(Side.CLIENT)
 public class TutorialToast implements IToast
 {
-    private final TutorialToast.Icons icon;
+    private final Icons icon;
     private final String title;
     private final String subtitle;
-    private IToast.Visibility visibility = IToast.Visibility.SHOW;
+    private Visibility visibility = Visibility.SHOW;
     private long lastDelta;
     private float displayedProgress;
     private float currentProgress;
     private final boolean hasProgressBar;
 
-    public TutorialToast(TutorialToast.Icons iconIn, ITextComponent titleComponent, @Nullable ITextComponent subtitleComponent, boolean drawProgressBar)
+    public TutorialToast(Icons iconIn, ITextComponent titleComponent, @Nullable ITextComponent subtitleComponent, boolean drawProgressBar)
     {
         this.icon = iconIn;
         this.title = titleComponent.getFormattedText();
@@ -28,7 +28,7 @@
         this.hasProgressBar = drawProgressBar;
     }
 
-    public IToast.Visibility draw(GuiToast toastGui, long delta)
+    public Visibility draw(GuiToast toastGui, long delta)
     {
         toastGui.getMinecraft().getTextureManager().bindTexture(TEXTURE_TOASTS);
         GlStateManager.color(1.0F, 1.0F, 1.0F);
@@ -70,7 +70,7 @@
 
     public void hide()
     {
-        this.visibility = IToast.Visibility.HIDE;
+        this.visibility = Visibility.HIDE;
     }
 
     public void setProgress(float progress)
