--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -69,8 +_,8 @@
    private static final int f_156742_ = 96;
    public static final int f_156736_ = 128;
    private final Predicate<Entity> f_286991_;
-   private final ServerBossEvent f_64060_ = (ServerBossEvent)(new ServerBossEvent(Component.m_237115_("entity.minecraft.ender_dragon"), BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS)).m_7005_(true).m_7006_(true);
-   private final ServerLevel f_64061_;
+   public final ServerBossEvent f_64060_ = (ServerBossEvent)(new ServerBossEvent(Component.m_237115_("entity.minecraft.ender_dragon"), BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS)).m_7005_(true).m_7006_(true); // CatServer - private -> public
+   public final ServerLevel f_64061_; // CatServer - private -> public
    private final BlockPos f_286985_;
    private final ObjectArrayList<Integer> f_64062_ = new ObjectArrayList<>();
    private final BlockPattern f_64063_;
@@ -82,12 +_,12 @@
    private boolean f_64069_;
    private boolean f_286982_ = false;
    @Nullable
-   private UUID f_64070_;
+   public UUID f_64070_; // CatServer - private -> public
    private boolean f_64071_ = true;
    @Nullable
-   private BlockPos f_64072_;
+   public BlockPos f_64072_; // CatServer - private -> public
    @Nullable
-   private DragonRespawnAnimation f_64073_;
+   public DragonRespawnAnimation f_64073_; // CatServer - private -> public
    private int f_64074_;
    @Nullable
    private List<EndCrystal> f_64075_;
@@ -216,7 +_,7 @@
 
    }
 
-   protected void m_64087_(DragonRespawnAnimation p_64088_) {
+   public void m_64087_(DragonRespawnAnimation p_64088_) { // CatServer - protected -> public
       if (this.f_64073_ == null) {
          throw new IllegalStateException("Dragon respawn isn't in progress, can't skip ahead in the animation.");
       } else {
@@ -254,7 +_,7 @@
    }
 
    @Nullable
-   private BlockPattern.BlockPatternMatch m_64105_() {
+   public BlockPattern.BlockPatternMatch m_64105_() { // CatServer - private -> public
       ChunkPos chunkpos = new ChunkPos(this.f_286985_);
 
       for(int i = -8 + chunkpos.f_45578_; i <= 8 + chunkpos.f_45578_; ++i) {
@@ -387,7 +_,7 @@
       });
    }
 
-   private void m_64093_(boolean p_64094_) {
+   public void m_64093_(boolean p_64094_) { // CatServer - private -> public
       EndPodiumFeature endpodiumfeature = new EndPodiumFeature(p_64094_);
       if (this.f_64072_ == null) {
          for(this.f_64072_ = this.f_64061_.m_5452_(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.m_287210_(this.f_286985_)).m_7495_(); this.f_64061_.m_8055_(this.f_64072_).m_60713_(Blocks.f_50752_) && this.f_64072_.m_123342_() > this.f_64061_.m_5736_(); this.f_64072_ = this.f_64072_.m_7495_()) {
@@ -449,6 +_,17 @@
       return this.f_64069_;
    }
 
+   // CatServer start - 1.20.1
+   private boolean catserver$tryRespawnResult = false;
+   public boolean tryRespawnCB() { // CraftBukkit - return boolean
+      try {
+         this.m_64100_();
+         return this.catserver$tryRespawnResult;
+      } finally {
+         this.catserver$tryRespawnResult = false;
+      }
+   }
+
    public void m_64100_() {
       if (this.f_64068_ && this.f_64073_ == null) {
          BlockPos blockpos = this.f_64072_;
@@ -471,6 +_,7 @@
          for(Direction direction : Direction.Plane.HORIZONTAL) {
             List<EndCrystal> list = this.f_64061_.m_45976_(EndCrystal.class, new AABB(blockpos1.m_5484_(direction, 2)));
             if (list.isEmpty()) {
+               this.catserver$tryRespawnResult = false; // CraftBukkit - 1.20.1 return value
                return;
             }
 
@@ -478,9 +_,21 @@
          }
 
          f_64058_.debug("Found all crystals, respawning dragon.");
-         this.m_64091_(list1);
+         this.catserver$tryRespawnResult = this.respawnDragonCB(list1); // CraftBukkit - 1.20.1 return value
+         return; // CatServer - 1.20.1
       }
+      this.catserver$tryRespawnResult = false; // CraftBukkit - 1.20.1 return value
+      return; // CatServer - 1.20.1
+   }
 
+   private boolean catserver$respawnDragonResult = false;
+   public boolean respawnDragonCB(List<EndCrystal> p_64092_) { // PAIL private -> public // CraftBukkit - 1.20.1 return boolean
+      try {
+         this.m_64091_(p_64092_);
+         return this.catserver$respawnDragonResult;
+      } finally {
+         this.catserver$respawnDragonResult = false;
+      }
    }
 
    private void m_64091_(List<EndCrystal> p_64092_) {
@@ -502,9 +_,13 @@
          this.f_64074_ = 0;
          this.m_64093_(false);
          this.f_64075_ = p_64092_;
+         this.catserver$respawnDragonResult = true; // CraftBukkit - 1.20.1 return value
+         return; // CatServer - 1.20.1
       }
-
+      this.catserver$respawnDragonResult = false; // CraftBukkit - 1.20.1 return value
+      return; // CatServer - 1.20.1
    }
+   // CatServer end
 
    public void m_64101_() {
       for(SpikeFeature.EndSpike spikefeature$endspike : SpikeFeature.m_66858_(this.f_64061_)) {
@@ -513,7 +_,14 @@
             endcrystal.m_31052_((BlockPos)null);
          }
       }
-
+   }
+
+   public void addPlayer(ServerPlayer player) {
+      this.f_64060_.m_6543_(player);
+   }
+
+   public void removePlayer(ServerPlayer player) {
+      this.f_64060_.m_6539_(player);
    }
 
    @Nullable
