--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -77,7 +77,27 @@
         this.target = targetIn;
         this.direction = EnumFacing.UP;
         this.selectNextMoveDirection(p_i46772_4_);
+        projectileSource = (org.bukkit.entity.LivingEntity) ownerIn.getBukkitEntity(); // CraftBukkit
     }
+    // CraftBukkit start
+    public EntityLivingBase getShooter() {
+    	return this.owner;
+    }
+    
+    public void setShooter(EntityLivingBase e) {
+    	this.owner = e;
+    }
+    
+    public Entity getTarget() {
+    	return this.target;
+    }
+    
+    public void setTarget(Entity e) {
+    	this.target = e;
+    	this.direction = EnumFacing.UP;
+    	this.selectNextMoveDirection(EnumFacing.Axis.X);
+    }
+    // CraftBukkit end
 
     protected void writeEntityToNBT(NBTTagCompound compound)
     {
