--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
@@ -31,8 +31,8 @@
     private Block inTile;
     protected boolean inGround;
     public int throwableShake;
-    private EntityLivingBase thrower;
-    private String throwerName;
+    public EntityLivingBase thrower;
+    public String throwerName;
     private int ticksInGround;
     private int ticksInAir;
     public Entity ignoreEntity;
