--- ../src-base/minecraft/net/minecraft/advancements/AdvancementList.java
+++ ../src-work/minecraft/net/minecraft/advancements/AdvancementList.java
@@ -18,10 +18,10 @@
 public class AdvancementList
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private final Map<ResourceLocation, Advancement> advancements = Maps.<ResourceLocation, Advancement>newHashMap();
+    public final Map<ResourceLocation, Advancement> advancements = Maps.<ResourceLocation, Advancement>newHashMap(); // CatServer - private -> public
     private final Set<Advancement> roots = Sets.<Advancement>newLinkedHashSet();
     private final Set<Advancement> nonRoots = Sets.<Advancement>newLinkedHashSet();
-    private AdvancementList.Listener listener;
+    private Listener listener;
 
     @SideOnly(Side.CLIENT)
     private void remove(Advancement advancementIn)
