--- ../src-base/minecraft/net/minecraft/advancements/ICriterionTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/ICriterionTrigger.java
@@ -8,9 +8,9 @@
 {
     ResourceLocation getId();
 
-    void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<T> listener);
+    void addListener(PlayerAdvancements playerAdvancementsIn, Listener<T> listener);
 
-    void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<T> listener);
+    void removeListener(PlayerAdvancements playerAdvancementsIn, Listener<T> listener);
 
     void removeAllListeners(PlayerAdvancements playerAdvancementsIn);
 
@@ -47,7 +47,7 @@
                 }
                 else if (p_equals_1_ != null && this.getClass() == p_equals_1_.getClass())
                 {
-                    ICriterionTrigger.Listener<?> listener = (ICriterionTrigger.Listener)p_equals_1_;
+                    Listener<?> listener = (Listener)p_equals_1_;
 
                     if (!this.criterionInstance.equals(listener.criterionInstance))
                     {
