--- a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -29,7 +_,8 @@
       if (f != null) {
          RandomSource randomsource = p_81659_.m_230907_();
          float f1 = 1.0F / f;
-         return randomsource.m_188501_() <= f1;
+         // CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+         return randomsource.m_188501_() < f1;
       } else {
          return true;
       }
