--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
@@ -5,6 +_,7 @@
 import com.google.gson.JsonSerializationContext;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class LootItemRandomChanceCondition implements LootItemCondition {
    final float f_81921_;
@@ -18,7 +_,13 @@
    }
 
    public boolean test(LootContext p_81930_) {
-      return p_81930_.m_78933_().nextFloat() < this.f_81921_;
+      int i = p_81930_.getLootingModifier();
+      // CraftBukkit start - only use lootingModifier if set by Bukkit
+      if (p_81930_.m_78936_(LootContextParams.LOOTING_MOD)) {
+         i = p_81930_.m_78953_(LootContextParams.LOOTING_MOD);
+      }
+      // CraftBukkit end
+      return p_81930_.m_78933_().nextFloat() < this.f_81921_ + (float) i;
    }
 
    public static LootItemCondition.Builder m_81927_(float p_81928_) {
