--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -20,6 +_,11 @@
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
+
 public class EndCrystal extends Entity {
    private static final EntityDataAccessor<Optional<BlockPos>> f_31033_ = SynchedEntityData.m_135353_(EndCrystal.class, EntityDataSerializers.f_135039_);
    private static final EntityDataAccessor<Boolean> f_31034_ = SynchedEntityData.m_135353_(EndCrystal.class, EntityDataSerializers.f_135035_);
@@ -50,7 +_,11 @@
       if (this.m_9236_() instanceof ServerLevel) {
          BlockPos blockpos = this.m_20183_();
          if (((ServerLevel)this.m_9236_()).m_8586_() != null && this.m_9236_().m_8055_(blockpos).m_60795_()) {
-            this.m_9236_().m_46597_(blockpos, BaseFireBlock.m_49245_(this.m_9236_(), blockpos));
+            // CraftBukkit start
+            if (!CraftEventFactory.callBlockIgniteEvent(this.m_9236_(), blockpos, this).isCancelled()) {
+               this.m_9236_().m_46597_(blockpos, BaseFireBlock.m_49245_(this.m_9236_(), blockpos));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -86,10 +_,22 @@
          return false;
       } else {
          if (!this.m_213877_() && !this.m_9236_().f_46443_) {
+            // CraftBukkit start - All non-living entities need this
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31050_, p_31051_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             this.m_142687_(Entity.RemovalReason.KILLED);
             if (!p_31050_.m_269533_(DamageTypeTags.f_268415_)) {
                DamageSource damagesource = p_31050_.m_7639_() != null ? this.m_269291_().m_269036_(this, p_31050_.m_7639_()) : null;
-               this.m_9236_().m_254877_(this, damagesource, (ExplosionDamageCalculator)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), 6.0F, false, Level.ExplosionInteraction.BLOCK);
+               // CraftBukkit start
+               ExplosionPrimeEvent event = CraftEventFactory.callExplosionPrimeEvent(this, 6.0F, false);
+               if (event.isCancelled()) {
+                  this.m_146912_();
+                  return false;
+               }
+               this.m_9236_().m_254877_(this, damagesource, (ExplosionDamageCalculator)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.BLOCK);
+               // CraftBukkit end
             }
 
             this.m_31047_(p_31050_);
