--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -20,6 +_,8 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class EndCrystal extends Entity {
    private static final EntityDataAccessor<Optional<BlockPos>> f_31033_ = SynchedEntityData.m_135353_(EndCrystal.class, EntityDataSerializers.f_135039_);
@@ -51,7 +_,11 @@
       if (this.f_19853_ instanceof ServerLevel) {
          BlockPos blockpos = this.m_142538_();
          if (((ServerLevel)this.f_19853_).m_8586_() != null && this.f_19853_.m_8055_(blockpos).m_60795_()) {
-            this.f_19853_.m_46597_(blockpos, BaseFireBlock.m_49245_(this.f_19853_, blockpos));
+            // CraftBukkit start
+            if (!CraftEventFactory.callBlockIgniteEvent(this.f_19853_, blockpos, this).isCancelled()) {
+               this.f_19853_.m_46597_(blockpos, BaseFireBlock.m_49245_(this.f_19853_, blockpos));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -87,16 +_,28 @@
          return false;
       } else {
          if (!this.m_146910_() && !this.f_19853_.f_46443_) {
+            // CraftBukkit start - All non-living entities need this
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31050_, p_31051_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             this.m_142687_(Entity.RemovalReason.KILLED);
             if (!p_31050_.m_19372_()) {
-               this.f_19853_.m_46511_((Entity)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), 6.0F, Explosion.BlockInteraction.DESTROY);
+               // CraftBukkit start
+               ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+               this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  this.m_146912_();
+                  return false;
+               }
+               this.f_19853_.m_46518_(this, this.m_20185_(), this.m_20186_(), this.m_20189_(), event.getRadius(), event.getFire(), Explosion.BlockInteraction.DESTROY);
+               // CraftBukkit end
             }
-
-            this.m_31047_(p_31050_);
          }
 
-         return true;
+         this.m_31047_(p_31050_);
       }
+      return true;
    }
 
    public void m_6074_() {
