--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -36,8 +_,12 @@
 import net.minecraft.world.entity.boss.enderdragon.phases.EnderDragonPhaseManager;
 import net.minecraft.world.entity.monster.Enemy;
 import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -46,9 +_,14 @@
 import net.minecraft.world.level.pathfinder.BinaryHeap;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.event.entity.EntityExplodeEvent;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
 
 public class EnderDragon extends Mob implements Enemy {
    private static final Logger f_31087_ = LogUtils.getLogger();
@@ -85,6 +_,7 @@
    private final Node[] f_31077_ = new Node[24];
    private final int[] f_31078_ = new int[24];
    private final BinaryHeap f_31079_ = new BinaryHeap();
+   private Explosion explosionSource = new Explosion(null, this, null, null, Double.NaN, Double.NaN, Double.NaN, Float.NaN, true, Explosion.BlockInteraction.DESTROY); // CraftBukkit - reusable source for CraftTNTPrimed.getSource()
 
    public EnderDragon(EntityType<? extends EnderDragon> p_31096_, Level p_31097_) {
       super(EntityType.f_20565_, p_31097_);
@@ -150,7 +_,11 @@
       return adouble;
    }
 
+   @javax.annotation.Nullable private Player unlimitedLastHurtByPlayer = null;
    public void m_8107_() {
+      // lastHurtByPlayer is cleared after 100 ticks, capture it indefinitely in unlimitedLastHurtByPlayer for LivingExperienceDropEvent
+      if (this.f_20888_ != null) this.unlimitedLastHurtByPlayer = f_20888_;
+      if (this.unlimitedLastHurtByPlayer != null && this.unlimitedLastHurtByPlayer.m_146910_()) this.unlimitedLastHurtByPlayer = null;
       this.m_146874_();
       if (this.f_19853_.f_46443_) {
          this.m_21153_(this.m_21223_());
@@ -349,7 +_,14 @@
          if (this.f_31086_.m_146910_()) {
             this.f_31086_ = null;
          } else if (this.f_19797_ % 10 == 0 && this.m_21223_() < this.m_21233_()) {
-            this.m_21153_(this.m_21223_() + 1.0F);
+            // CraftBukkit start
+            EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), 1.0F, EntityRegainHealthEvent.RegainReason.ENDER_CRYSTAL);
+            this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+               this.m_21153_((float) (this.m_21223_() + event.getAmount()));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -414,14 +_,22 @@
       boolean flag = false;
       boolean flag1 = false;
 
+      // CraftBukkit start - Create a list to hold all the destroyed blocks
+      List<org.bukkit.block.Block> destroyedBlocks = new java.util.ArrayList<org.bukkit.block.Block>();
+      // CraftBukkit end
+
       for(int k1 = i; k1 <= l; ++k1) {
          for(int l1 = j; l1 <= i1; ++l1) {
             for(int i2 = k; i2 <= j1; ++i2) {
                BlockPos blockpos = new BlockPos(k1, l1, i2);
                BlockState blockstate = this.f_19853_.m_8055_(blockpos);
                if (!blockstate.m_60795_() && blockstate.m_60767_() != Material.f_76309_) {
-                  if (this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_) && !blockstate.m_204336_(BlockTags.f_13069_)) {
-                     flag1 = this.f_19853_.m_7471_(blockpos, false) || flag1;
+                  if (net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.f_19853_, blockpos, this) && !blockstate.m_204336_(BlockTags.f_13069_)) {
+                     // CraftBukkit start - Add blocks to list rather than destroying them
+                     // flag1 = this.level.removeBlock(blockpos, false) || flag1;
+                     flag1 = true;
+                     destroyedBlocks.add( CraftBlock.at(f_19853_, blockpos));
+                     // CraftBukkit end
                   } else {
                      flag = true;
                   }
@@ -430,6 +_,52 @@
          }
       }
 
+      // CraftBukkit start - Set off an EntityExplodeEvent for the dragon exploding all these blocks
+      // SPIGOT-4882: don't fire event if nothing hit
+      if (!flag1) {
+         return flag;
+      }
+
+      org.bukkit.entity.Entity bukkitEntity = this.getBukkitEntity();
+      EntityExplodeEvent event = new EntityExplodeEvent(bukkitEntity, bukkitEntity.getLocation(), destroyedBlocks, 0F);
+      bukkitEntity.getServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         // This flag literally means 'Dragon hit something hard' (Obsidian, White Stone or Bedrock) and will cause the dragon to slow down.
+         // We should consider adding an event extension for it, or perhaps returning true if the event is cancelled.
+         return flag;
+      } else if (event.getYield() == 0F) {
+         // Yield zero ==> no drops
+         for (org.bukkit.block.Block block : event.blockList()) {
+            this.f_19853_.m_7471_(new BlockPos(block.getX(), block.getY(), block.getZ()), false);
+         }
+      } else {
+         for (org.bukkit.block.Block block : event.blockList()) {
+            org.bukkit.Material blockId = block.getType();
+            if (blockId.isAir()) {
+               continue;
+            }
+
+            CraftBlock craftBlock = ((CraftBlock) block);
+            BlockPos blockposition = craftBlock.getPosition();
+
+            Block nmsBlock = craftBlock.getNMS().m_60734_();
+            if (nmsBlock.m_6903_(explosionSource)) {
+               BlockEntity tileentity = craftBlock.getNMS().m_155947_() ? this.f_19853_.m_7702_(blockposition) : null;
+               LootContext.Builder loottableinfo_builder = (new LootContext.Builder((ServerLevel ) this.f_19853_)).m_78977_(this.f_19853_.f_46441_).m_78972_( LootContextParams.f_81460_, Vec3.m_82512_(blockposition)).m_78972_(LootContextParams.f_81463_, ItemStack.f_41583_).m_78972_(LootContextParams.f_81464_, 1.0F / event.getYield()).m_78984_(LootContextParams.f_81462_, tileentity);
+
+               craftBlock.getNMS().m_60724_(loottableinfo_builder).forEach((itemstack) -> {
+                  Block.m_49840_(f_19853_, blockposition, itemstack);
+               });
+               craftBlock.getNMS().m_60612_((ServerLevel ) f_19853_, blockposition, ItemStack.f_41583_);
+            }
+            nmsBlock.m_7592_(f_19853_, blockposition, explosionSource);
+
+            this.f_19853_.m_7471_(blockposition, false);
+         }
+      }
+      // CraftBukkit end
+
+
       if (flag1) {
          BlockPos blockpos1 = new BlockPos(i + this.f_19796_.nextInt(l - i + 1), j + this.f_19796_.nextInt(i1 - j + 1), k + this.f_19796_.nextInt(j1 - k + 1));
          this.f_19853_.m_46796_(2008, blockpos1, 0);
@@ -514,11 +_,30 @@
 
       if (this.f_19853_ instanceof ServerLevel) {
          if (this.f_31084_ > 150 && this.f_31084_ % 5 == 0 && flag) {
-            ExperienceOrb.m_147082_((ServerLevel)this.f_19853_, this.m_20182_(), Mth.m_14143_((float)i * 0.08F));
+            int award = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.unlimitedLastHurtByPlayer, Mth.m_14143_((float)i * 0.08F));
+            ExperienceOrb.m_147082_((ServerLevel)this.f_19853_, this.m_20182_(), award);
          }
 
          if (this.f_31084_ == 1 && !this.m_20067_()) {
-            this.f_19853_.m_6798_(1028, this.m_142538_(), 0);
+            //this.level.globalLevelEvent(1028, this.blockPosition(), 0);
+            // CraftBukkit start - Use relative location for far away sounds
+            // this.world.b(1028, this.getChunkCoordinates(), 0);
+            int viewDistance = ((ServerLevel) this.f_19853_).getCraftServer().getViewDistance() * 16;
+            for (net.minecraft.server.level.ServerPlayer player : this.f_19853_.m_142572_().m_6846_().f_11196_) {
+               double deltaX = this.m_20185_() - player.m_20185_();
+               double deltaZ = this.m_20189_() - player.m_20189_();
+               double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+               if ( f_19853_.spigotConfig.dragonDeathSoundRadius > 0 && distanceSquared > f_19853_.spigotConfig.dragonDeathSoundRadius * f_19853_.spigotConfig.dragonDeathSoundRadius ) continue; // Spigot
+               if (distanceSquared > viewDistance * viewDistance) {
+                  double deltaLength = Math.sqrt(distanceSquared);
+                  double relativeX = player.m_20185_() + (deltaX / deltaLength) * viewDistance;
+                  double relativeZ = player.m_20189_() + (deltaZ / deltaLength) * viewDistance;
+                  player.f_8906_.m_141995_(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1028, new BlockPos((int) relativeX, (int) this.m_20186_(), (int) relativeZ), 0, true));
+               } else {
+                  player.f_8906_.m_141995_(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1028, new BlockPos((int) this.m_20185_(), (int) this.m_20186_(), (int) this.m_20189_()), 0, true));
+               }
+            }
+            // CraftBukkit end
          }
       }
 
@@ -527,7 +_,8 @@
       this.f_20883_ = this.m_146908_();
       if (this.f_31084_ == 200 && this.f_19853_ instanceof ServerLevel) {
          if (flag) {
-            ExperienceOrb.m_147082_((ServerLevel)this.f_19853_, this.m_20182_(), Mth.m_14143_((float)i * 0.2F));
+            int award = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.unlimitedLastHurtByPlayer, Mth.m_14143_((float)i * 0.2F));
+            ExperienceOrb.m_147082_((ServerLevel)this.f_19853_, this.m_20182_(), award);
          }
 
          if (this.f_31073_ != null) {
@@ -847,6 +_,16 @@
 
    public boolean m_6072_() {
       return false;
+   }
+
+   @Override
+   public boolean isMultipartEntity() {
+      return true;
+   }
+
+   @Override
+   public net.minecraftforge.entity.PartEntity<?>[] getParts() {
+      return this.f_31089_;
    }
 
    public void m_142223_(ClientboundAddMobPacket p_149572_) {
