--- ../src-base/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java
+++ ../src-work/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java
@@ -37,6 +37,7 @@
     private int maxY = Integer.MIN_VALUE;
     private float fade;
     private boolean centered;
+    private int page;
 
     public GuiAdvancementTab(Minecraft p_i47589_1_, GuiScreenAdvancements p_i47589_2_, AdvancementTabType p_i47589_3_, int p_i47589_4_, Advancement p_i47589_5_, DisplayInfo p_i47589_6_)
     {
@@ -160,9 +161,9 @@
         {
             for (AdvancementTabType advancementtabtype : AdvancementTabType.values())
             {
-                if (p_193936_2_ < advancementtabtype.getMax())
+                if ((p_193936_2_ % AdvancementTabType.MAX_TABS) < advancementtabtype.getMax())
                 {
-                    return new GuiAdvancementTab(p_193936_0_, p_193936_1_, advancementtabtype, p_193936_2_, p_193936_3_, p_193936_3_.getDisplay());
+                    return new GuiAdvancementTab(p_193936_0_, p_193936_1_, advancementtabtype, p_193936_2_ % AdvancementTabType.MAX_TABS, p_193936_2_ / AdvancementTabType.MAX_TABS, p_193936_3_, p_193936_3_.getDisplay());
                 }
 
                 p_193936_2_ -= advancementtabtype.getMax();
@@ -222,4 +223,17 @@
     {
         return this.screen;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    public int getPage()
+    {
+        return this.page;
+    }
+
+    public GuiAdvancementTab(Minecraft p_i47589_1_, GuiScreenAdvancements p_i47589_2_, AdvancementTabType p_i47589_3_, int p_i47589_4_, int page, Advancement p_i47589_5_, DisplayInfo p_i47589_6_)
+    {
+        this(p_i47589_1_, p_i47589_2_, p_i47589_3_, p_i47589_4_, p_i47589_5_, p_i47589_6_);
+        this.page = page;
+    }
+    /* ======================================== FORGE END =====================================*/
 }
