--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -111,6 +_,26 @@
       }
    }
 
+   public EntitySelector parse(StringReader p_91451_, boolean overridePermissions) throws CommandSyntaxException {
+      int i = 0;
+      EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
+      EntitySelector entityselector = entityselectorparser.parse(overridePermissions);
+      if (entityselector.m_121138_() > 1 && this.f_91443_) {
+         if (this.f_91444_) {
+            p_91451_.setCursor(0);
+            throw f_91437_.createWithContext(p_91451_);
+         } else {
+            p_91451_.setCursor(0);
+            throw f_91436_.createWithContext(p_91451_);
+         }
+      } else if (entityselector.m_121159_() && this.f_91444_ && !entityselector.m_121162_()) {
+         p_91451_.setCursor(0);
+         throw f_91438_.createWithContext(p_91451_);
+      } else {
+         return entityselector;
+      }
+   }
+
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> p_91482_, SuggestionsBuilder p_91483_) {
       if (p_91482_.getSource() instanceof SharedSuggestionProvider) {
          StringReader stringreader = new StringReader(p_91483_.getInput());
