--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -90,10 +_,23 @@
       }
    }
 
+   // CatServer start
+   private final ThreadLocal<Boolean> catserver$isOverride = ThreadLocal.withInitial(() -> false);
+   public EntitySelector parse(StringReader p_91451_, boolean override) throws CommandSyntaxException {
+      try {
+         this.catserver$isOverride.set(override);
+         return this.parse(p_91451_);
+      } finally {
+         this.catserver$isOverride.remove();
+      }
+   }
+
    public EntitySelector parse(StringReader p_91451_) throws CommandSyntaxException {
       int i = 0;
       EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
-      EntitySelector entityselector = entityselectorparser.m_121377_();
+      var overridePermessions = this.catserver$isOverride.get(); // CatServer
+      this.catserver$isOverride.remove(); // CatServer
+      EntitySelector entityselector = entityselectorparser.parse(overridePermessions); // CatServer - entityselectorparser.parse() -> entityselectorparser.parse(overridePermessions)
       if (entityselector.m_121138_() > 1 && this.f_91443_) {
          if (this.f_91444_) {
             p_91451_.setCursor(0);
@@ -109,13 +_,14 @@
          return entityselector;
       }
    }
+   // CatServer end
 
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> p_91482_, SuggestionsBuilder p_91483_) {
       S s = p_91482_.getSource();
       if (s instanceof SharedSuggestionProvider sharedsuggestionprovider) {
          StringReader stringreader = new StringReader(p_91483_.getInput());
          stringreader.setCursor(p_91483_.getStart());
-         EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, sharedsuggestionprovider.m_6761_(2));
+         EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, net.minecraftforge.common.ForgeHooks.canUseEntitySelectors(sharedsuggestionprovider));
 
          try {
             entityselectorparser.m_121377_();
