--- a/net/minecraft/scoreboard/ServerScoreboard.java
+++ b/net/minecraft/scoreboard/ServerScoreboard.java
@@ -26,7 +_,7 @@
    public void func_96536_a(Score p_96536_1_) {
       super.func_96536_a(p_96536_1_);
       if (this.field_96553_b.contains(p_96536_1_.func_96645_d())) {
-         this.field_96555_a.func_184103_al().func_148540_a(new SUpdateScorePacket(ServerScoreboard.Action.CHANGE, p_96536_1_.func_96645_d().func_96679_b(), p_96536_1_.func_96653_e(), p_96536_1_.func_96652_c()));
+         this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.CHANGE, p_96536_1_.func_96645_d().func_96679_b(), p_96536_1_.func_96653_e(), p_96536_1_.func_96652_c())); // CraftBukkit
       }
 
       this.func_96551_b();
@@ -34,14 +_,14 @@
 
    public void func_96516_a(String p_96516_1_) {
       super.func_96516_a(p_96516_1_);
-      this.field_96555_a.func_184103_al().func_148540_a(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, (String)null, p_96516_1_, 0));
+      this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, (String)null, p_96516_1_, 0)); // CraftBukkit
       this.func_96551_b();
    }
 
    public void func_178820_a(String p_178820_1_, ScoreObjective p_178820_2_) {
       super.func_178820_a(p_178820_1_, p_178820_2_);
       if (this.field_96553_b.contains(p_178820_2_)) {
-         this.field_96555_a.func_184103_al().func_148540_a(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, p_178820_2_.func_96679_b(), p_178820_1_, 0));
+         this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, p_178820_2_.func_96679_b(), p_178820_1_, 0)); // CraftBukkit
       }
 
       this.func_96551_b();
@@ -52,7 +_,7 @@
       super.func_96530_a(p_96530_1_, p_96530_2_);
       if (scoreobjective != p_96530_2_ && scoreobjective != null) {
          if (this.func_96552_h(scoreobjective) > 0) {
-            this.field_96555_a.func_184103_al().func_148540_a(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_));
+            this.sendAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_)); // CraftBukkit
          } else {
             this.func_96546_g(scoreobjective);
          }
@@ -60,7 +_,7 @@
 
       if (p_96530_2_ != null) {
          if (this.field_96553_b.contains(p_96530_2_)) {
-            this.field_96555_a.func_184103_al().func_148540_a(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_));
+            this.sendAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_)); // CraftBukkit
          } else {
             this.func_96549_e(p_96530_2_);
          }
@@ -71,7 +_,7 @@
 
    public boolean func_197901_a(String p_197901_1_, ScorePlayerTeam p_197901_2_) {
       if (super.func_197901_a(p_197901_1_, p_197901_2_)) {
-         this.field_96555_a.func_184103_al().func_148540_a(new STeamsPacket(p_197901_2_, Arrays.asList(p_197901_1_), 3));
+         this.sendAll(new STeamsPacket(p_197901_2_, Arrays.asList(p_197901_1_), 3)); // CraftBukkit
          this.func_96551_b();
          return true;
       } else {
@@ -81,7 +_,7 @@
 
    public void func_96512_b(String p_96512_1_, ScorePlayerTeam p_96512_2_) {
       super.func_96512_b(p_96512_1_, p_96512_2_);
-      this.field_96555_a.func_184103_al().func_148540_a(new STeamsPacket(p_96512_2_, Arrays.asList(p_96512_1_), 4));
+      this.sendAll(new STeamsPacket(p_96512_2_, Arrays.asList(p_96512_1_), 4)); // CraftBukkit
       this.func_96551_b();
    }
 
@@ -93,7 +_,7 @@
    public void func_199869_b(ScoreObjective p_199869_1_) {
       super.func_199869_b(p_199869_1_);
       if (this.field_96553_b.contains(p_199869_1_)) {
-         this.field_96555_a.func_184103_al().func_148540_a(new SScoreboardObjectivePacket(p_199869_1_, 2));
+         this.sendAll(new SScoreboardObjectivePacket(p_199869_1_, 2)); // CraftBukkit
       }
 
       this.func_96551_b();
@@ -110,19 +_,19 @@
 
    public void func_96523_a(ScorePlayerTeam p_96523_1_) {
       super.func_96523_a(p_96523_1_);
-      this.field_96555_a.func_184103_al().func_148540_a(new STeamsPacket(p_96523_1_, 0));
+      this.sendAll(new STeamsPacket(p_96523_1_, 0)); // CraftBukkit
       this.func_96551_b();
    }
 
    public void func_96538_b(ScorePlayerTeam p_96538_1_) {
       super.func_96538_b(p_96538_1_);
-      this.field_96555_a.func_184103_al().func_148540_a(new STeamsPacket(p_96538_1_, 2));
+      this.sendAll(new STeamsPacket(p_96538_1_, 2)); // CraftBukkit
       this.func_96551_b();
    }
 
    public void func_96513_c(ScorePlayerTeam p_96513_1_) {
       super.func_96513_c(p_96513_1_);
-      this.field_96555_a.func_184103_al().func_148540_a(new STeamsPacket(p_96513_1_, 1));
+      this.sendAll(new STeamsPacket(p_96513_1_, 1)); // CraftBukkit
       this.func_96551_b();
    }
 
@@ -159,6 +_,7 @@
       List<IPacket<?>> list = this.func_96550_d(p_96549_1_);
 
       for(ServerPlayerEntity serverplayerentity : this.field_96555_a.func_184103_al().func_181057_v()) {
+         if (serverplayerentity.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(IPacket<?> ipacket : list) {
             serverplayerentity.field_71135_a.func_147359_a(ipacket);
          }
@@ -184,6 +_,7 @@
       List<IPacket<?>> list = this.func_96548_f(p_96546_1_);
 
       for(ServerPlayerEntity serverplayerentity : this.field_96555_a.func_184103_al().func_181057_v()) {
+         if (serverplayerentity.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(IPacket<?> ipacket : list) {
             serverplayerentity.field_71135_a.func_147359_a(ipacket);
          }
@@ -203,6 +_,16 @@
 
       return i;
    }
+
+   // CraftBukkit start - Send to players
+   private void sendAll(IPacket packet) {
+      for (ServerPlayerEntity entityplayer : this.field_96555_a.func_184103_al().field_72404_b) {
+         if (entityplayer.getBukkitEntity().getScoreboard().getHandle() == this) {
+            entityplayer.field_71135_a.func_147359_a(packet);
+         }
+      }
+   }
+   // CraftBukkit end
 
    public static enum Action {
       CHANGE,
