--- a/net/minecraft/item/BoatItem.java
+++ b/net/minecraft/item/BoatItem.java
@@ -7,9 +_,11 @@
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
+import net.minecraft.util.ActionResultType;
 import net.minecraft.util.EntityPredicates;
 import net.minecraft.util.Hand;
 import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockRayTraceResult;
 import net.minecraft.util.math.RayTraceContext;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.util.math.vector.Vector3d;
@@ -45,6 +_,14 @@
          }
 
          if (raytraceresult.func_216346_c() == RayTraceResult.Type.BLOCK) {
+            // CraftBukkit start - Boat placement
+            BlockRayTraceResult movingobjectpositionblock = (BlockRayTraceResult) raytraceresult;
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPlayerInteractEvent(p_77659_2_, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, movingobjectpositionblock.func_216350_a(), movingobjectpositionblock.func_216354_b(), itemstack, p_77659_3_);
+
+            if (event.isCancelled()) {
+               return new ActionResult<>(ActionResultType.PASS, itemstack);
+            }
+            // CraftBukkit end
             BoatEntity boatentity = new BoatEntity(p_77659_1_, raytraceresult.func_216347_e().field_72450_a, raytraceresult.func_216347_e().field_72448_b, raytraceresult.func_216347_e().field_72449_c);
             boatentity.func_184458_a(this.field_185057_a);
             boatentity.field_70177_z = p_77659_2_.field_70177_z;
@@ -52,7 +_,16 @@
                return ActionResult.func_226251_d_(itemstack);
             } else {
                if (!p_77659_1_.field_72995_K) {
-                  p_77659_1_.func_217376_c(boatentity);
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityPlaceEvent(p_77659_1_, movingobjectpositionblock.func_216350_a(), movingobjectpositionblock.func_216354_b(), p_77659_2_, boatentity).isCancelled()) {
+                     return ActionResult.func_226251_d_(itemstack);
+                  }
+
+                  if (!p_77659_1_.func_217376_c(boatentity)) {
+                     return new ActionResult<>(ActionResultType.PASS, itemstack);
+                  }
+                  // CraftBukkit end
+
                   if (!p_77659_2_.field_71075_bZ.field_75098_d) {
                      itemstack.func_190918_g(1);
                   }
