--- a/net/minecraft/item/BoneMealItem.java
+++ b/net/minecraft/item/BoneMealItem.java
@@ -27,10 +_,16 @@
    }
 
    public ActionResultType func_195939_a(ItemUseContext p_195939_1_) {
-      World world = p_195939_1_.func_195991_k();
-      BlockPos blockpos = p_195939_1_.func_195995_a();
-      BlockPos blockpos1 = blockpos.func_177972_a(p_195939_1_.func_196000_l());
-      if (func_195966_a(p_195939_1_.func_195996_i(), world, blockpos)) {
+      // CraftBukkit start - extract bonemeal application logic to separate, static method
+      return applyBonemeal(p_195939_1_);
+   }
+
+   public static ActionResultType applyBonemeal(ItemUseContext context) {
+      // CraftBukkit end
+      World world = context.func_195991_k();
+      BlockPos blockpos = context.func_195995_a();
+      BlockPos blockpos1 = blockpos.func_177972_a(context.func_196000_l());
+      if (applyBonemeal(context.func_195996_i(), world, blockpos, context.func_195999_j())) {
          if (!world.field_72995_K) {
             world.func_217379_c(2005, blockpos, 0);
          }
@@ -38,8 +_,8 @@
          return ActionResultType.func_233537_a_(world.field_72995_K);
       } else {
          BlockState blockstate = world.func_180495_p(blockpos);
-         boolean flag = blockstate.func_224755_d(world, blockpos, p_195939_1_.func_196000_l());
-         if (flag && func_203173_b(p_195939_1_.func_195996_i(), world, blockpos1, p_195939_1_.func_196000_l())) {
+         boolean flag = blockstate.func_224755_d(world, blockpos, context.func_196000_l());
+         if (flag && func_203173_b(context.func_195996_i(), world, blockpos1, context.func_196000_l())) {
             if (!world.field_72995_K) {
                world.func_217379_c(2005, blockpos1, 0);
             }
@@ -51,8 +_,17 @@
       }
    }
 
+   @Deprecated //Forge: Use Player/Hand version
    public static boolean func_195966_a(ItemStack p_195966_0_, World p_195966_1_, BlockPos p_195966_2_) {
+      if (p_195966_1_ instanceof net.minecraft.world.server.ServerWorld)
+         return applyBonemeal(p_195966_0_, p_195966_1_, p_195966_2_, net.minecraftforge.common.util.FakePlayerFactory.getMinecraft((net.minecraft.world.server.ServerWorld)p_195966_1_));
+      return false;
+   }
+
+   public static boolean applyBonemeal(ItemStack p_195966_0_, World p_195966_1_, BlockPos p_195966_2_, net.minecraft.entity.player.PlayerEntity player) {
       BlockState blockstate = p_195966_1_.func_180495_p(p_195966_2_);
+      int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, p_195966_1_, p_195966_2_, blockstate, p_195966_0_);
+      if (hook != 0) return hook > 0;
       if (blockstate.func_177230_c() instanceof IGrowable) {
          IGrowable igrowable = (IGrowable)blockstate.func_177230_c();
          if (igrowable.func_176473_a(p_195966_1_, p_195966_2_, blockstate, p_195966_1_.field_72995_K)) {
@@ -128,7 +_,7 @@
       }
 
       BlockState blockstate = p_195965_0_.func_180495_p(p_195965_1_);
-      if (!blockstate.func_196958_f()) {
+      if (!blockstate.isAir(p_195965_0_, p_195965_1_)) {
          double d0 = 0.5D;
          double d1;
          if (blockstate.func_203425_a(Blocks.field_150355_j)) {
