--- a/net/minecraft/item/DebugStickItem.java
+++ b/net/minecraft/item/DebugStickItem.java
@@ -49,7 +_,7 @@
    }
 
    private void func_195958_a(PlayerEntity p_195958_1_, BlockState p_195958_2_, IWorld p_195958_3_, BlockPos p_195958_4_, boolean p_195958_5_, ItemStack p_195958_6_) {
-      if (p_195958_1_.func_195070_dx()) {
+      if (p_195958_1_.func_195070_dx() || (p_195958_1_.field_71075_bZ.field_75098_d && p_195958_1_.getBukkitEntity().hasPermission("minecraft.debugstick")) || p_195958_1_.getBukkitEntity().hasPermission("minecraft.debugstick.always")) { // Spigot
          Block block = p_195958_2_.func_177230_c();
          StateContainer<Block, BlockState> statecontainer = block.func_176194_O();
          Collection<Property<?>> collection = statecontainer.func_177623_d();
