--- a/net/minecraft/item/DyeItem.java
+++ b/net/minecraft/item/DyeItem.java
@@ -8,6 +_,9 @@
 import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Hand;
 
+import org.bukkit.Color;
+import org.bukkit.event.entity.SheepDyeWoolEvent;
+
 public class DyeItem extends Item {
    private static final Map<DyeColor, DyeItem> field_195963_a = Maps.newEnumMap(DyeColor.class);
    private final DyeColor field_195964_b;
@@ -23,7 +_,15 @@
          SheepEntity sheepentity = (SheepEntity)p_111207_3_;
          if (sheepentity.func_70089_S() && !sheepentity.func_70892_o() && sheepentity.func_175509_cj() != this.field_195964_b) {
             if (!p_111207_2_.field_70170_p.field_72995_K) {
-               sheepentity.func_175512_b(this.field_195964_b);
+               // CraftBukkit start
+               byte bColor = (byte) this.field_195964_b.func_196059_a();
+               SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) sheepentity.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor));
+               sheepentity.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  return ActionResultType.PASS;
+               }
+               sheepentity.func_175512_b(DyeColor.func_196056_a(event.getColor().getWoolData()));
+               // CraftBukkit end
                p_111207_1_.func_190918_g(1);
             }
 
