--- a/net/minecraft/item/EggItem.java
+++ b/net/minecraft/item/EggItem.java
@@ -1,6 +_,7 @@
 package net.minecraft.item;
 
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.EggEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
@@ -16,14 +_,22 @@
 
    public ActionResult<ItemStack> func_77659_a(World p_77659_1_, PlayerEntity p_77659_2_, Hand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
-      p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187511_aA, SoundCategory.PLAYERS, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+      //worldIn.playSound((PlayerEntity)null, playerIn.getPosX(), playerIn.getPosY(), playerIn.getPosZ(), SoundEvents.ENTITY_EGG_THROW, SoundCategory.PLAYERS, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
       if (!p_77659_1_.field_72995_K) {
          EggEntity eggentity = new EggEntity(p_77659_1_, p_77659_2_);
          eggentity.func_213884_b(itemstack);
          eggentity.func_234612_a_(p_77659_2_, p_77659_2_.field_70125_A, p_77659_2_.field_70177_z, 0.0F, 1.5F, 1.0F);
-         p_77659_1_.func_217376_c(eggentity);
+         // CraftBukkit start
+         if (!p_77659_1_.func_217376_c(eggentity)) {
+            if (p_77659_2_ instanceof ServerPlayerEntity) {
+               ((ServerPlayerEntity) p_77659_2_).getBukkitEntity().updateInventory();
+            }
+            return ActionResult.func_226251_d_(itemstack);
+      }
+         // CraftBukkit end
       }
 
+      p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187511_aA, SoundCategory.PLAYERS, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F)); // CraftBukkit - from above
       p_77659_2_.func_71029_a(Stats.field_75929_E.func_199076_b(this));
       if (!p_77659_2_.field_71075_bZ.field_75098_d) {
          itemstack.func_190918_g(1);
