--- a/net/minecraft/item/EnderPearlItem.java
+++ b/net/minecraft/item/EnderPearlItem.java
@@ -2,8 +_,10 @@
 
 import net.minecraft.entity.item.EnderPearlEntity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
+import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Hand;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
@@ -16,14 +_,22 @@
 
    public ActionResult<ItemStack> func_77659_a(World p_77659_1_, PlayerEntity p_77659_2_, Hand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
-      p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187595_bc, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
-      p_77659_2_.func_184811_cZ().func_185145_a(this, 20);
+
+      // CraftBukkit start - change order
       if (!p_77659_1_.field_72995_K) {
          EnderPearlEntity enderpearlentity = new EnderPearlEntity(p_77659_1_, p_77659_2_);
          enderpearlentity.func_213884_b(itemstack);
          enderpearlentity.func_234612_a_(p_77659_2_, p_77659_2_.field_70125_A, p_77659_2_.field_70177_z, 0.0F, 1.5F, 1.0F);
-         p_77659_1_.func_217376_c(enderpearlentity);
-      }
+         if (!p_77659_1_.func_217376_c(enderpearlentity)) {
+            if (p_77659_2_ instanceof ServerPlayerEntity) {
+               ((ServerPlayerEntity)p_77659_2_).getBukkitEntity().updateInventory();
+            }
+            return new ActionResult<ItemStack>(ActionResultType.FAIL, itemstack);
+      }
+      }
+      p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187595_bc, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+      p_77659_2_.func_184811_cZ().func_185145_a(this, 20);
+      // CraftBukkit end
 
       p_77659_2_.func_71029_a(Stats.field_75929_E.func_199076_b(this));
       if (!p_77659_2_.field_71075_bZ.field_75098_d) {
