--- a/net/minecraft/item/EnderCrystalItem.java
+++ b/net/minecraft/item/EnderCrystalItem.java
@@ -38,6 +_,13 @@
                if (world instanceof ServerWorld) {
                   EnderCrystalEntity endercrystalentity = new EnderCrystalEntity(world, d0 + 0.5D, d1, d2 + 0.5D);
                   endercrystalentity.func_184517_a(false);
+
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityPlaceEvent(p_195939_1_, endercrystalentity).isCancelled()) {
+                     return ActionResultType.FAIL;
+                  }
+                  // CraftBukkit end
+
                   world.func_217376_c(endercrystalentity);
                   DragonFightManager dragonfightmanager = ((ServerWorld)world).func_241110_C_();
                   if (dragonfightmanager != null) {
