--- a/net/minecraft/item/EnderEyeItem.java
+++ b/net/minecraft/item/EnderEyeItem.java
@@ -9,6 +_,7 @@
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.EyeOfEnderEntity;
+import net.minecraft.network.play.server.SPlaySoundEventPacket;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
 import net.minecraft.util.ActionResultType;
@@ -52,7 +_,25 @@
                   }
                }
 
-               world.func_175669_a(1038, blockpos1.func_177982_a(1, 0, 1), 0);
+               // CraftBukkit start - Use relative location for far away sounds
+               //  world.playBroadcastSound(1038, blockpos1.add(1, 0, 1), 0);
+               int viewDistance = world.getCBServer().getViewDistance() * 16;
+               BlockPos soundPos = blockpos1.func_177982_a(1, 0, 1);
+               for (ServerPlayerEntity player : world.func_73046_m().getServer().func_184103_al().field_72404_b) {
+                  double deltaX = soundPos.func_177958_n() - player.func_226277_ct_();
+                  double deltaZ = soundPos.func_177952_p() - player.func_226281_cx_();
+                  double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                  if (world.spigotConfig.endPortalSoundRadius > 0 && distanceSquared > world.spigotConfig.endPortalSoundRadius * world.spigotConfig.endPortalSoundRadius) continue; // Spigot
+                  if (distanceSquared > viewDistance * viewDistance) {
+                     double deltaLength = Math.sqrt(distanceSquared);
+                     double relativeX = player.func_226277_ct_() + (deltaX / deltaLength) * viewDistance;
+                     double relativeZ = player.func_226281_cx_() + (deltaZ / deltaLength) * viewDistance;
+                     player.field_71135_a.func_147359_a(new SPlaySoundEventPacket(1038, new BlockPos((int) relativeX, (int) soundPos.func_177956_o(), (int) relativeZ), 0, true));
+                  } else {
+                     player.field_71135_a.func_147359_a(new SPlaySoundEventPacket(1038, soundPos, 0, true));
+                  }
+               }
+               // CraftBukkit end
             }
 
             return ActionResultType.CONSUME;
@@ -75,7 +_,11 @@
                EyeOfEnderEntity eyeofenderentity = new EyeOfEnderEntity(p_77659_1_, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226283_e_(0.5D), p_77659_2_.func_226281_cx_());
                eyeofenderentity.func_213863_b(itemstack);
                eyeofenderentity.func_180465_a(blockpos);
-               p_77659_1_.func_217376_c(eyeofenderentity);
+               // CraftBukkit start
+               if (!p_77659_1_.func_217376_c(eyeofenderentity)) {
+                  return new ActionResult<>(ActionResultType.FAIL, itemstack);
+               }
+               // CraftBukkit end
                if (p_77659_2_ instanceof ServerPlayerEntity) {
                   CriteriaTriggers.field_192132_l.func_192239_a((ServerPlayerEntity)p_77659_2_, blockpos);
                }
