--- a/net/minecraft/item/FilledMapItem.java
+++ b/net/minecraft/item/FilledMapItem.java
@@ -33,6 +_,8 @@
 import net.minecraft.world.storage.MapData;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Bukkit;
+import org.bukkit.event.server.MapInitializeEvent;
 
 public class FilledMapItem extends AbstractMapItem {
    public FilledMapItem(Item.Properties p_i48482_1_) {
@@ -52,6 +_,22 @@
 
    @Nullable
    public static MapData func_195950_a(ItemStack p_195950_0_, World p_195950_1_) {
+      // FORGE: Add instance method for mods to override
+      Item map = p_195950_0_.func_77973_b();
+      if (map instanceof FilledMapItem) {
+        return ((FilledMapItem)map).getCustomMapData(p_195950_0_, p_195950_1_);
+      }
+      return null;
+   }
+
+   @Nullable
+   protected MapData getCustomMapData(ItemStack p_195950_0_, World p_195950_1_) {
+      // LoliServer start
+      return getCustomMapData_static(p_195950_0_, p_195950_1_);
+   }
+
+   @Nullable
+   public static MapData getCustomMapData_static(ItemStack p_195950_0_, World p_195950_1_) {
       MapData mapdata = func_219994_a(p_195950_0_, p_195950_1_);
       if (mapdata == null && p_195950_1_ instanceof ServerWorld) {
          mapdata = func_195951_a(p_195950_0_, p_195950_1_, p_195950_1_.func_72912_H().func_76079_c(), p_195950_1_.func_72912_H().func_76074_e(), 3, false, false, p_195950_1_.func_234923_W_());
@@ -59,10 +_,11 @@
 
       return mapdata;
    }
+   // LoliServer end
 
    public static int func_195949_f(ItemStack p_195949_0_) {
       CompoundNBT compoundnbt = p_195949_0_.func_77978_p();
-      return compoundnbt != null && compoundnbt.func_150297_b("map", 99) ? compoundnbt.func_74762_e("map") : 0;
+      return compoundnbt != null && compoundnbt.func_150297_b("map", 99) ? compoundnbt.func_74762_e("map") : -1; // CraftBukkit - make new maps for no tag
    }
 
    private static MapData func_195951_a(ItemStack p_195951_0_, World p_195951_1_, int p_195951_2_, int p_195951_3_, int p_195951_4_, boolean p_195951_5_, boolean p_195951_6_, RegistryKey<World> p_195951_7_) {
@@ -71,6 +_,11 @@
       mapdata.func_237241_a_(p_195951_2_, p_195951_3_, p_195951_4_, p_195951_5_, p_195951_6_, p_195951_7_);
       p_195951_1_.func_217399_a(mapdata);
       p_195951_0_.func_196082_o().func_74768_a("map", i);
+
+      // CraftBukkit start
+      MapInitializeEvent event = new MapInitializeEvent(mapdata.mapView);
+      Bukkit.getServer().getPluginManager().callEvent(event);
+      // CraftBukkit end
       return mapdata;
    }
 
@@ -226,7 +_,7 @@
 
             for(int l = 0; l < 128 * i; ++l) {
                for(int i1 = 0; i1 < 128 * i; ++i1) {
-                  abiome[l * 128 * i + i1] = p_226642_0_.func_226691_t_(new BlockPos((j / i - 64) * i + i1, 0, (k / i - 64) * i + l));
+                  abiome[l * 128 * i + i1] = p_226642_0_.getBiomeBySeed((j / i - 64) * i + i1, 0, (k / i - 64) * i + l); // Paper
                }
             }
 
