--- a/net/minecraft/item/FireChargeItem.java
+++ b/net/minecraft/item/FireChargeItem.java
@@ -9,6 +_,8 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.block.BlockIgniteEvent;
 
 public class FireChargeItem extends Item {
    public FireChargeItem(Item.Properties p_i48499_1_) {
@@ -21,12 +_,28 @@
       BlockState blockstate = world.func_180495_p(blockpos);
       boolean flag = false;
       if (CampfireBlock.func_241470_h_(blockstate)) {
+         // CraftBukkit start - fire BlockIgniteEvent
+         if (CraftEventFactory.callBlockIgniteEvent(world, blockpos, BlockIgniteEvent.IgniteCause.FIREBALL, p_195939_1_.func_195999_j()).isCancelled()) {
+            if (!p_195939_1_.func_195999_j().field_71075_bZ.field_75098_d) {
+               p_195939_1_.func_195996_i().func_190918_g(1);
+            }
+            return ActionResultType.PASS;
+         }
+         // CraftBukkit end
          this.func_219995_a(world, blockpos);
          world.func_175656_a(blockpos, blockstate.func_206870_a(CampfireBlock.field_220101_b, Boolean.valueOf(true)));
          flag = true;
       } else {
          blockpos = blockpos.func_177972_a(p_195939_1_.func_196000_l());
          if (AbstractFireBlock.func_241465_a_(world, blockpos, p_195939_1_.func_195992_f())) {
+            // CraftBukkit start - fire BlockIgniteEvent
+            if (CraftEventFactory.callBlockIgniteEvent(world, blockpos, BlockIgniteEvent.IgniteCause.FIREBALL, p_195939_1_.func_195999_j()).isCancelled()) {
+               if (!p_195939_1_.func_195999_j().field_71075_bZ.field_75098_d) {
+                  p_195939_1_.func_195996_i().func_190918_g(1);
+               }
+               return ActionResultType.PASS;
+            }
+            // CraftBukkit end
             this.func_219995_a(world, blockpos);
             world.func_175656_a(blockpos, AbstractFireBlock.func_235326_a_(world, blockpos));
             flag = true;
