--- a/net/minecraft/item/FishingRodItem.java
+++ b/net/minecraft/item/FishingRodItem.java
@@ -6,10 +_,12 @@
 import net.minecraft.entity.projectile.FishingBobberEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
+import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Hand;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.World;
+import org.bukkit.event.player.PlayerFishEvent;
 
 public class FishingRodItem extends Item implements IVanishable {
    public FishingRodItem(Item.Properties p_i48494_1_) {
@@ -28,11 +_,22 @@
 
          p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_193780_J, SoundCategory.NEUTRAL, 1.0F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
       } else {
-         p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187612_G, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+         // worldIn.playSound((PlayerEntity)null, playerIn.getPosX(), playerIn.getPosY(), playerIn.getPosZ(), SoundEvents.ENTITY_FISHING_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F)); // CraftBukkit - moved down
          if (!p_77659_1_.field_72995_K) {
             int k = EnchantmentHelper.func_191528_c(itemstack);
             int j = EnchantmentHelper.func_191529_b(itemstack);
-            p_77659_1_.func_217376_c(new FishingBobberEntity(p_77659_2_, p_77659_1_, j, k));
+            // CraftBukkit start
+            FishingBobberEntity entityfishinghook = new FishingBobberEntity(p_77659_2_, p_77659_1_, j, k);
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) p_77659_2_.getBukkitEntity(), null, (org.bukkit.entity.FishHook) entityfishinghook.getBukkitEntity(), PlayerFishEvent.State.FISHING);
+            p_77659_1_.getCBServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+               p_77659_2_.field_71104_cf = null;
+               return new ActionResult<>(ActionResultType.PASS, itemstack);
+            }
+            p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187612_G, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+            p_77659_1_.func_217376_c(entityfishinghook);
+            // CraftBukkit end
          }
 
          p_77659_2_.func_71029_a(Stats.field_75929_E.func_199076_b(this));
