--- a/net/minecraft/item/FlintAndSteelItem.java
+++ b/net/minecraft/item/FlintAndSteelItem.java
@@ -23,7 +_,7 @@
       World world = p_195939_1_.func_195991_k();
       BlockPos blockpos = p_195939_1_.func_195995_a();
       BlockState blockstate = world.func_180495_p(blockpos);
-      if (CampfireBlock.func_241470_h_(blockstate)) {
+      if (CampfireBlock.func_241470_h_(blockstate) && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos, playerentity).isCancelled()) { // CraftBukkit
          world.func_184133_a(playerentity, blockpos, SoundEvents.field_187649_bu, SoundCategory.BLOCKS, 1.0F, field_77697_d.nextFloat() * 0.4F + 0.8F);
          world.func_180501_a(blockpos, blockstate.func_206870_a(BlockStateProperties.field_208190_q, Boolean.valueOf(true)), 11);
          if (playerentity != null) {
@@ -36,6 +_,12 @@
       } else {
          BlockPos blockpos1 = blockpos.func_177972_a(p_195939_1_.func_196000_l());
          if (AbstractFireBlock.func_241465_a_(world, blockpos1, p_195939_1_.func_195992_f())) {
+            // CraftBukkit start - Store the clicked block
+            if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos1, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, playerentity).isCancelled()) {
+               p_195939_1_.func_195996_i().func_222118_a(1, playerentity, (entityhuman1) -> entityhuman1.func_213334_d(p_195939_1_.func_221531_n()));
+               return ActionResultType.PASS;
+            }
+            // CraftBukkit end
             world.func_184133_a(playerentity, blockpos1, SoundEvents.field_187649_bu, SoundCategory.BLOCKS, 1.0F, field_77697_d.nextFloat() * 0.4F + 0.8F);
             BlockState blockstate1 = AbstractFireBlock.func_235326_a_(world, blockpos1);
             world.func_180501_a(blockpos1, blockstate1, 11);
