--- a/net/minecraft/item/HangingEntityItem.java
+++ b/net/minecraft/item/HangingEntityItem.java
@@ -10,6 +_,8 @@
 import net.minecraft.util.Direction;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.entity.Player;
+import org.bukkit.event.hanging.HangingPlaceEvent;
 
 public class HangingEntityItem extends Item {
    private final EntityType<? extends HangingEntity> field_220001_a;
@@ -47,6 +_,16 @@
 
          if (hangingentity.func_70518_d()) {
             if (!world.field_72995_K) {
+               // CraftBukkit start - fire HangingPlaceEvent
+               Player  who = (p_195939_1_.func_195999_j() == null) ? null : (Player) p_195939_1_.func_195999_j().getBukkitEntity();
+               org.bukkit.block.Block blockClicked = world.getWorld().getBlockAt(blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+               org.bukkit.block.BlockFace blockFace = org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock.notchToBlockFace(direction);
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) ((HangingEntity) hangingentity).getBukkitEntity(), who, blockClicked, blockFace);
+               world.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  return ActionResultType.FAIL;
+               }
+               // CraftBukkit end
                hangingentity.func_184523_o();
                world.func_217376_c(hangingentity);
             }
