--- a/net/minecraft/item/ItemStack.java
+++ b/net/minecraft/item/ItemStack.java
@@ -7,23 +_,18 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.Codec;
+import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.text.DecimalFormat;
 import java.text.DecimalFormatSymbols;
-import java.util.Collection;
-import java.util.List;
-import java.util.Locale;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Random;
+import java.util.*;
 import java.util.Map.Entry;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 import net.minecraft.advancements.CriteriaTriggers;
-import net.minecraft.block.Block;
-import net.minecraft.block.BlockState;
+import net.minecraft.block.*;
 import net.minecraft.client.util.ITooltipFlag;
 import net.minecraft.command.arguments.BlockPredicateArgument;
 import net.minecraft.command.arguments.BlockStateParser;
@@ -44,18 +_,14 @@
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.INBT;
 import net.minecraft.nbt.ListNBT;
+import net.minecraft.nbt.NBTDynamicOps;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.ITag;
 import net.minecraft.tags.ITagCollectionSupplier;
-import net.minecraft.util.ActionResult;
-import net.minecraft.util.ActionResultType;
-import net.minecraft.util.CachedBlockInfo;
-import net.minecraft.util.Hand;
-import net.minecraft.util.IItemProvider;
-import net.minecraft.util.ResourceLocation;
-import net.minecraft.util.SoundEvent;
-import net.minecraft.util.Util;
+import net.minecraft.util.*;
+import net.minecraft.util.datafix.TypeReferences;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.util.text.IFormattableTextComponent;
@@ -71,8 +_,11 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.util.CraftMagicNumbers;
+import org.bukkit.event.player.PlayerItemDamageEvent;
 
-public final class ItemStack {
+public final class ItemStack extends net.minecraftforge.common.capabilities.CapabilityProvider<ItemStack> implements net.minecraftforge.common.extensions.IForgeItemStack {
    public static final Codec<ItemStack> field_234691_a_ = RecordCodecBuilder.create((p_234698_0_) -> {
       return p_234698_0_.group(Registry.field_212630_s.fieldOf("id").forGetter((p_234706_0_) -> {
          return p_234706_0_.field_151002_e;
@@ -82,6 +_,9 @@
          return Optional.ofNullable(p_234704_0_.field_77990_d);
       })).apply(p_234698_0_, ItemStack::new);
    });
+   private net.minecraftforge.registries.IRegistryDelegate<Item> delegate;
+   private CompoundNBT capNBT;
+
    private static final Logger field_199558_c = LogManager.getLogger();
    public static final ItemStack field_190927_a = new ItemStack((Item)null);
    public static final DecimalFormat field_111284_a = Util.func_200696_a(new DecimalFormat("#.##"), (p_234699_0_) -> {
@@ -91,7 +_,7 @@
    private int field_77994_a;
    private int field_77992_b;
    @Deprecated
-   private final Item field_151002_e;
+   private Item field_151002_e;
    private CompoundNBT field_77990_d;
    private boolean field_190928_g;
    private Entity field_234693_k_;
@@ -109,33 +_,61 @@
       p_i231596_3_.ifPresent(this::func_77982_d);
    }
 
-   public ItemStack(IItemProvider p_i48204_1_, int p_i48204_2_) {
+   public ItemStack(IItemProvider p_i48204_1_, int p_i48204_2_) { this(p_i48204_1_, p_i48204_2_, (CompoundNBT) null); }
+   public ItemStack(IItemProvider p_i48204_1_, int p_i48204_2_, @Nullable CompoundNBT capNBT) {
+      super(ItemStack.class, true);
+      this.capNBT = capNBT;
       this.field_151002_e = p_i48204_1_ == null ? null : p_i48204_1_.func_199767_j();
+      this.delegate = p_i48204_1_ == null ? null : p_i48204_1_.func_199767_j().delegate;
       this.field_77994_a = p_i48204_2_;
-      if (this.field_151002_e != null && this.field_151002_e.func_77645_m()) {
+      this.forgeInit();
+      if (this.field_151002_e != null && this.field_151002_e.isDamageable(this)) {
          this.func_196085_b(this.func_77952_i());
       }
 
       this.func_190923_F();
    }
+
+   // CraftBukkit start - Called to run this stack through the data converter to handle older storage methods and serialized items
+   public void convertStack(int version) {
+      if (version > 0 && version < CraftMagicNumbers.INSTANCE.getDataVersion()) {
+         CompoundNBT savedStack = new CompoundNBT();
+         this.func_77955_b(savedStack);
+         savedStack = (CompoundNBT) MinecraftServer.getServer().field_184112_s.update(TypeReferences.field_211295_k, new Dynamic<>(NBTDynamicOps.field_210820_a, savedStack), version, CraftMagicNumbers.INSTANCE.getDataVersion()).getValue();
+         this.load(savedStack);
+      }
+   }
+   // CraftBukkit end
 
    private void func_190923_F() {
       this.field_190928_g = false;
       this.field_190928_g = this.func_190926_b();
    }
 
+   // CraftBukkit start - break into own method
+   private void load(CompoundNBT compound) {
+      this.capNBT = compound.func_74764_b("ForgeCaps") ? compound.func_74775_l("ForgeCaps") : null;
+      Item rawItem =
+      this.field_151002_e = Registry.field_212630_s.func_82594_a(new ResourceLocation(compound.func_74779_i("id")));
+      this.delegate = rawItem.delegate;
+      this.field_77994_a = compound.func_74771_c("Count");
+      if (compound.func_150297_b("tag", 10)) {
+         // CraftBukkit start - make defensive copy as this data may be coming from the save thread
+         this.field_77990_d = compound.func_74775_l("tag").func_74737_b();
+         this.func_77973_b().func_179215_a(this.field_77990_d);
+         // CraftBukkit end
+      }
+
+      this.forgeInit();
+      if (this.func_77973_b().func_77645_m()) {
+         this.func_196085_b(this.func_77952_i());
+      }
+   }
+
    private ItemStack(CompoundNBT p_i47263_1_) {
-      this.field_151002_e = Registry.field_212630_s.func_82594_a(new ResourceLocation(p_i47263_1_.func_74779_i("id")));
-      this.field_77994_a = p_i47263_1_.func_74771_c("Count");
-      if (p_i47263_1_.func_150297_b("tag", 10)) {
-         this.field_77990_d = p_i47263_1_.func_74775_l("tag");
-         this.func_77973_b().func_179215_a(p_i47263_1_);
-      }
-
-      if (this.func_77973_b().func_77645_m()) {
-         this.func_196085_b(this.func_77952_i());
-      }
-
+      super(ItemStack.class, true);
+      this.load(p_i47263_1_);
+      // CraftBukkit stop
       this.func_190923_F();
    }
 
@@ -167,10 +_,19 @@
    }
 
    public Item func_77973_b() {
-      return this.field_190928_g ? Items.field_190931_a : this.field_151002_e;
+      return this.field_190928_g || this.delegate == null ? Items.field_190931_a : this.delegate.get();
    }
 
    public ActionResultType func_196084_a(ItemUseContext p_196084_1_) {
+      if (!p_196084_1_.func_195991_k().field_72995_K) return net.minecraftforge.common.ForgeHooks.onPlaceItemIntoWorld(p_196084_1_);
+      return onItemUse(p_196084_1_, (c) -> func_77973_b().func_195939_a(p_196084_1_));
+   }
+
+   public ActionResultType onItemUseFirst(ItemUseContext p_196084_1_) {
+      return onItemUse(p_196084_1_, (c) -> func_77973_b().onItemUseFirst(this, p_196084_1_));
+   }
+
+   private ActionResultType onItemUse(ItemUseContext p_196084_1_, java.util.function.Function<ItemUseContext, ActionResultType> callback) {
       PlayerEntity playerentity = p_196084_1_.func_195999_j();
       BlockPos blockpos = p_196084_1_.func_195995_a();
       CachedBlockInfo cachedblockinfo = new CachedBlockInfo(p_196084_1_.func_195991_k(), blockpos, false);
@@ -178,7 +_,7 @@
          return ActionResultType.PASS;
       } else {
          Item item = this.func_77973_b();
-         ActionResultType actionresulttype = item.func_195939_a(p_196084_1_);
+         ActionResultType actionresulttype = callback.apply(p_196084_1_);
          if (playerentity != null && actionresulttype.func_226246_a_()) {
             playerentity.func_71029_a(Stats.field_75929_E.func_199076_b(item));
          }
@@ -206,12 +_,15 @@
       if (this.field_77990_d != null) {
          p_77955_1_.func_218657_a("tag", this.field_77990_d.func_74737_b());
       }
-
+      CompoundNBT cnbt = this.serializeCaps();
+      if (cnbt != null && !cnbt.isEmpty()) {
+         p_77955_1_.func_218657_a("ForgeCaps", cnbt);
+      }
       return p_77955_1_;
    }
 
    public int func_77976_d() {
-      return this.func_77973_b().func_77639_j();
+      return this.func_77973_b().getItemStackLimit(this);
    }
 
    public boolean func_77985_e() {
@@ -219,7 +_,7 @@
    }
 
    public boolean func_77984_f() {
-      if (!this.field_190928_g && this.func_77973_b().func_77612_l() > 0) {
+      if (!this.field_190928_g && this.func_77973_b().isDamageable(this)) {
          CompoundNBT compoundnbt = this.func_77978_p();
          return compoundnbt == null || !compoundnbt.func_74767_n("Unbreakable");
       } else {
@@ -228,19 +_,19 @@
    }
 
    public boolean func_77951_h() {
-      return this.func_77984_f() && this.func_77952_i() > 0;
+      return this.func_77984_f() && func_77973_b().isDamaged(this);
    }
 
    public int func_77952_i() {
-      return this.field_77990_d == null ? 0 : this.field_77990_d.func_74762_e("Damage");
+      return this.func_77973_b().getDamage(this);
    }
 
    public void func_196085_b(int p_196085_1_) {
-      this.func_196082_o().func_74768_a("Damage", Math.max(0, p_196085_1_));
+      this.func_77973_b().setDamage(this, p_196085_1_);
    }
 
    public int func_77958_k() {
-      return this.func_77973_b().func_77612_l();
+      return this.func_77973_b().getMaxDamage(this);
    }
 
    public boolean func_96631_a(int p_96631_1_, Random p_96631_2_, @Nullable ServerPlayerEntity p_96631_3_) {
@@ -258,6 +_,19 @@
             }
 
             p_96631_1_ -= j;
+            // CraftBukkit start
+            if (p_96631_3_ != null) {
+               PlayerItemDamageEvent event = new PlayerItemDamageEvent(p_96631_3_.getBukkitEntity(), CraftItemStack.asCraftMirror(this), p_96631_1_);
+               event.getPlayer().getServer().getPluginManager().callEvent(event);
+               if (p_96631_1_ != event.getDamage() || event.isCancelled()) {
+                  event.getPlayer().updateInventory();
+               }
+               if (event.isCancelled()) {
+                  return false;
+               }
+               p_96631_1_ = event.getDamage();
+            }
+            // CraftBukkit end
             if (p_96631_1_ <= 0) {
                return false;
             }
@@ -276,9 +_,17 @@
    public <T extends LivingEntity> void func_222118_a(int p_222118_1_, T p_222118_2_, Consumer<T> p_222118_3_) {
       if (!p_222118_2_.field_70170_p.field_72995_K && (!(p_222118_2_ instanceof PlayerEntity) || !((PlayerEntity)p_222118_2_).field_71075_bZ.field_75098_d)) {
          if (this.func_77984_f()) {
+            p_222118_1_ = this.func_77973_b().damageItem(this, p_222118_1_, p_222118_2_, p_222118_3_);
             if (this.func_96631_a(p_222118_1_, p_222118_2_.func_70681_au(), p_222118_2_ instanceof ServerPlayerEntity ? (ServerPlayerEntity)p_222118_2_ : null)) {
                p_222118_3_.accept(p_222118_2_);
                Item item = this.func_77973_b();
+
+               // CraftBukkit start - Check for item breaking
+               if (this.field_77994_a == 1 && p_222118_2_ instanceof PlayerEntity) {
+                  org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPlayerItemBreakEvent((PlayerEntity) p_222118_2_, this);
+               }
+               // CraftBukkit end
+
                this.func_190918_g(1);
                if (p_222118_2_ instanceof PlayerEntity) {
                   ((PlayerEntity)p_222118_2_).func_71029_a(Stats.field_199088_e.func_199076_b(item));
@@ -308,7 +_,7 @@
    }
 
    public boolean func_150998_b(BlockState p_150998_1_) {
-      return this.func_77973_b().func_150897_b(p_150998_1_);
+      return this.func_77973_b().canHarvestBlock(this, p_150998_1_);
    }
 
    public ActionResultType func_111282_a_(PlayerEntity p_111282_1_, LivingEntity p_111282_2_, Hand p_111282_3_) {
@@ -319,7 +_,7 @@
       if (this.func_190926_b()) {
          return field_190927_a;
       } else {
-         ItemStack itemstack = new ItemStack(this.func_77973_b(), this.field_77994_a);
+         ItemStack itemstack = new ItemStack(this.func_77973_b(), this.field_77994_a, this.serializeCaps());
          itemstack.func_190915_d(this.func_190921_D());
          if (this.field_77990_d != null) {
             itemstack.field_77990_d = this.field_77990_d.func_74737_b();
@@ -336,7 +_,7 @@
          if (p_77970_0_.field_77990_d == null && p_77970_1_.field_77990_d != null) {
             return false;
          } else {
-            return p_77970_0_.field_77990_d == null || p_77970_0_.field_77990_d.equals(p_77970_1_.field_77990_d);
+            return (p_77970_0_.field_77990_d == null || p_77970_0_.field_77990_d.equals(p_77970_1_.field_77990_d)) && p_77970_0_.areCapsCompatible(p_77970_1_);
          }
       } else {
          return false;
@@ -359,7 +_,7 @@
       } else if (this.field_77990_d == null && p_77959_1_.field_77990_d != null) {
          return false;
       } else {
-         return this.field_77990_d == null || this.field_77990_d.equals(p_77959_1_.field_77990_d);
+         return (this.field_77990_d == null || this.field_77990_d.equals(p_77959_1_.field_77990_d)) && this.areCapsCompatible(p_77959_1_);
       }
    }
 
@@ -435,6 +_,16 @@
       return !this.field_190928_g && this.field_77990_d != null && !this.field_77990_d.isEmpty();
    }
 
+   // CraftBukkit start
+   @Nullable
+   private CompoundNBT getTagClone() {
+      return this.field_77990_d == null ? null : this.field_77990_d.func_74737_b();
+   }
+   private void setTagClone(@Nullable CompoundNBT nbtttagcompound) {
+      this.func_77982_d(nbtttagcompound == null ? null : nbtttagcompound.func_74737_b());
+   }
+   // CraftBukkit end
+
    @Nullable
    public CompoundNBT func_77978_p() {
       return this.field_77990_d;
@@ -479,7 +_,7 @@
 
    public void func_77982_d(@Nullable CompoundNBT p_77982_1_) {
       this.field_77990_d = p_77982_1_;
-      if (this.func_77973_b().func_77645_m()) {
+      if (this.func_77973_b().isDamageable(this)) {
          this.func_196085_b(this.func_77952_i());
       }
 
@@ -673,6 +_,7 @@
          }
       }
 
+      net.minecraftforge.event.ForgeEventFactory.onItemTooltip(this, p_82840_1_, list, p_82840_2_);
       return list;
    }
 
@@ -795,6 +_,12 @@
    }
 
    public void func_82841_c(int p_82841_1_) {
+      // CraftBukkit start - remove RepairCost tag when 0 (SPIGOT-3945)
+      if (p_82841_1_ == 0) {
+         this.func_196083_e("RepairCost");
+         return;
+      }
+      // CraftBukkit end
       this.func_196082_o().func_74768_a("RepairCost", p_82841_1_);
    }
 
@@ -817,9 +_,10 @@
             }
          }
       } else {
-         multimap = this.func_77973_b().func_111205_h(p_111283_1_);
+         multimap = this.func_77973_b().getAttributeModifiers(p_111283_1_, this);
       }
 
+      multimap = net.minecraftforge.common.ForgeHooks.getAttributeModifiers(this, p_111283_1_, multimap);
       return multimap;
    }
 
@@ -839,6 +_,14 @@
       listnbt.add(compoundnbt);
    }
 
+   // CraftBukkit start
+   @Deprecated
+   public void setItem(Item item) {
+      this.field_151002_e = item;
+      this.delegate = item.delegate;
+   }
+   // CraftBukkit end
+
    public ITextComponent func_151000_E() {
       IFormattableTextComponent iformattabletextcomponent = (new StringTextComponent("")).func_230529_a_(this.func_200301_q());
       if (this.func_82837_s()) {
@@ -952,6 +_,23 @@
 
    public boolean func_222117_E() {
       return this.func_77973_b().func_219971_r();
+   }
+
+   // FORGE START
+   public void deserializeNBT(CompoundNBT nbt) {
+      final ItemStack itemStack = ItemStack.func_199557_a(nbt);
+      getStack().func_77982_d(itemStack.func_77978_p());
+      if (itemStack.capNBT != null) deserializeCaps(itemStack.capNBT);
+   }
+
+   /**
+    * Set up forge's ItemStack additions.
+    */
+   private void forgeInit() {
+      if (this.delegate != null) {
+         this.gatherCapabilities(() -> field_151002_e.initCapabilities(this, this.capNBT));
+         if (this.capNBT != null) deserializeCaps(this.capNBT);
+      }
    }
 
    public SoundEvent func_226629_F_() {
