--- a/net/minecraft/item/LeadItem.java
+++ b/net/minecraft/item/LeadItem.java
@@ -9,6 +_,7 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.event.hanging.HangingPlaceEvent;
 
 public class LeadItem extends Item {
    public LeadItem(Item.Properties p_i48484_1_) {
@@ -43,7 +_,22 @@
          if (mobentity.func_110166_bE() == p_226641_0_) {
             if (leashknotentity == null) {
                leashknotentity = LeashKnotEntity.func_213855_a(p_226641_1_, p_226641_2_);
-            }
+
+               // CraftBukkit start - fire HangingPlaceEvent
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) leashknotentity.getBukkitEntity(), p_226641_0_ != null ? (org.bukkit.entity.Player) p_226641_0_.getBukkitEntity() : null, p_226641_1_.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF);
+               p_226641_1_.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  leashknotentity.func_70106_y();
+                  return ActionResultType.PASS;
+               }
+               // CraftBukkit end
+            }
+
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPlayerLeashEntityEvent(mobentity, leashknotentity, p_226641_0_).isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
 
             mobentity.func_110162_b(leashknotentity, true);
             flag = true;
