--- a/net/minecraft/item/MerchantOffer.java
+++ b/net/minecraft/item/MerchantOffer.java
@@ -3,18 +_,32 @@
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.util.math.MathHelper;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftMerchantRecipe;
 
 public class MerchantOffer {
-   private final ItemStack field_222223_a;
-   private final ItemStack field_222224_b;
-   private final ItemStack field_222225_c;
-   private int field_222226_d;
-   private final int field_222227_e;
-   private boolean field_222228_f = true;
+   public ItemStack field_222223_a; // private->public CraftBukkit
+   public ItemStack field_222224_b; // private->public CraftBukkit
+   public final ItemStack field_222225_c; // private->public CraftBukkit
+   public int field_222226_d; // private->public CraftBukkit
+   public int field_222227_e; // private->public CraftBukkit
+   public boolean field_222228_f = true; // private->public CraftBukkit
    private int field_222229_g;
    private int field_222230_h;
-   private float field_222231_i;
-   private int field_222232_j = 1;
+   public float field_222231_i; // private->public CraftBukkit
+   public int field_222232_j = 1; // private->public CraftBukkit
+
+   // CraftBukkit start
+   private CraftMerchantRecipe bukkitHandle; // CraftBukkit
+
+   public CraftMerchantRecipe asBukkit() {
+      return (this.bukkitHandle == null) ? (this.bukkitHandle = new CraftMerchantRecipe(this)) : this.bukkitHandle;
+   }
+
+   public MerchantOffer(ItemStack itemstack, ItemStack itemstack1, ItemStack itemstack2, int uses, int maxUses, int experience, float priceMultiplier, CraftMerchantRecipe bukkit) {
+      this(itemstack, itemstack1, itemstack2, uses, maxUses, experience, priceMultiplier);
+      this.bukkitHandle = bukkit;
+   }
+   // CraftBukkit end
 
    public MerchantOffer(CompoundNBT p_i50012_1_) {
       this.field_222223_a = ItemStack.func_199557_a(p_i50012_1_.func_74775_l("buy"));
@@ -72,9 +_,10 @@
 
    public ItemStack func_222205_b() {
       int i = this.field_222223_a.func_190916_E();
+      if (i <= 0) return ItemStack.field_190927_a; // CraftBukkit - SPIGOT-5476
       ItemStack itemstack = this.field_222223_a.func_77946_l();
       int j = Math.max(0, MathHelper.func_76141_d((float)(i * this.field_222230_h) * this.field_222231_i));
-      itemstack.func_190920_e(MathHelper.func_76125_a(i + j + this.field_222229_g, 1, this.field_222223_a.func_77973_b().func_77639_j()));
+      itemstack.func_190920_e(MathHelper.func_76125_a(i + j + this.field_222229_g, 1, this.field_222223_a.func_77976_d()));
       return itemstack;
    }
 
@@ -87,7 +_,7 @@
    }
 
    public void func_222222_e() {
-      this.field_222230_h = this.field_222230_h + this.field_222226_d - (this.field_222227_e - this.field_222226_d);
+      this.field_222230_h = Math.max(0, this.field_222230_h + this.field_222226_d - (this.field_222227_e - this.field_222226_d)); // Paper
    }
 
    public ItemStack func_222206_f() {
@@ -178,7 +_,7 @@
          return true;
       } else {
          ItemStack itemstack = p_222201_1_.func_77946_l();
-         if (itemstack.func_77973_b().func_77645_m()) {
+         if (itemstack.func_77973_b().isDamageable(itemstack)) {
             itemstack.func_196085_b(itemstack.func_77952_i());
          }
 
