--- a/net/minecraft/item/PotionItem.java
+++ b/net/minecraft/item/PotionItem.java
@@ -42,7 +_,7 @@
             if (effectinstance.func_188419_a().func_76403_b()) {
                effectinstance.func_188419_a().func_180793_a(playerentity, playerentity, p_77654_3_, effectinstance.func_76458_c(), 1.0D);
             } else {
-               p_77654_3_.func_195064_c(new EffectInstance(effectinstance));
+               p_77654_3_.addEffect(new EffectInstance(effectinstance), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
             }
          }
       }
