--- a/net/minecraft/item/SignItem.java
+++ b/net/minecraft/item/SignItem.java
@@ -9,6 +_,9 @@
 import net.minecraft.world.World;
 
 public class SignItem extends WallOrFloorItem {
+
+   public static BlockPos openSign; // CraftBukkit
+
    public SignItem(Item.Properties p_i50038_1_, Block p_i50038_2_, Block p_i50038_3_) {
       super(p_i50038_2_, p_i50038_3_, p_i50038_1_);
    }
@@ -17,6 +_,9 @@
       boolean flag = super.func_195943_a(p_195943_1_, p_195943_2_, p_195943_3_, p_195943_4_, p_195943_5_);
       if (!p_195943_2_.field_72995_K && !flag && p_195943_3_ != null) {
          p_195943_3_.func_175141_a((SignTileEntity)p_195943_2_.func_175625_s(p_195943_1_));
+         // CraftBukkit start - SPIGOT-4678
+         SignItem.openSign = p_195943_1_;
+         // CraftBukkit end
       }
 
       return flag;
