--- a/net/minecraft/item/SkullItem.java
+++ b/net/minecraft/item/SkullItem.java
@@ -4,6 +_,7 @@
 import java.util.UUID;
 import net.minecraft.block.Block;
 import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.ListNBT;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.tileentity.SkullTileEntity;
 import net.minecraft.util.text.ITextComponent;
@@ -40,10 +_,26 @@
       super.func_179215_a(p_179215_1_);
       if (p_179215_1_.func_150297_b("SkullOwner", 8) && !StringUtils.isBlank(p_179215_1_.func_74779_i("SkullOwner"))) {
          GameProfile gameprofile = new GameProfile((UUID)null, p_179215_1_.func_74779_i("SkullOwner"));
-         gameprofile = SkullTileEntity.func_174884_b(gameprofile);
-         p_179215_1_.func_218657_a("SkullOwner", NBTUtil.func_180708_a(new CompoundNBT(), gameprofile));
+         // Spigot start
+         SkullTileEntity.updateGameProfile(gameprofile, new com.google.common.base.Predicate<GameProfile>() {
+            @Override
+            public boolean apply(GameProfile gameprofile) {
+               p_179215_1_.func_218657_a("SkullOwner", NBTUtil.func_180708_a(new CompoundNBT(), gameprofile));
+               return false;
+            }
+         }, false);
+         // Spigot end
          return true;
       } else {
+         // CraftBukkit start
+         ListNBT textures = p_179215_1_.func_74775_l("SkullOwner").func_74775_l("Properties").func_150295_c("textures", 10); // Safe due to method contracts
+         for (int i = 0; i < textures.size(); i++) {
+            if (textures.get(i) instanceof CompoundNBT && !((CompoundNBT) textures.get(i)).func_150297_b("Signature", 8) && ((CompoundNBT) textures.get(i)).func_74779_i("Value").trim().isEmpty()) {
+               p_179215_1_.func_82580_o("SkullOwner");
+               break;
+            }
+         }
+         // CraftBukkit end
          return false;
       }
    }
