--- a/net/minecraft/item/SnowballItem.java
+++ b/net/minecraft/item/SnowballItem.java
@@ -1,6 +_,7 @@
 package net.minecraft.item;
 
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.SnowballEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
@@ -16,18 +_,32 @@
 
    public ActionResult<ItemStack> func_77659_a(World p_77659_1_, PlayerEntity p_77659_2_, Hand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
-      p_77659_1_.func_184148_a((PlayerEntity)null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187797_fA, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+      // CraftBukkit - moved down
+      // worldIn.playSound((PlayerEntity)null, playerIn.getPosX(), playerIn.getPosY(), playerIn.getPosZ(), SoundEvents.ENTITY_SNOWBALL_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
       if (!p_77659_1_.field_72995_K) {
          SnowballEntity snowballentity = new SnowballEntity(p_77659_1_, p_77659_2_);
          snowballentity.func_213884_b(itemstack);
          snowballentity.func_234612_a_(p_77659_2_, p_77659_2_.field_70125_A, p_77659_2_.field_70177_z, 0.0F, 1.5F, 1.0F);
-         p_77659_1_.func_217376_c(snowballentity);
+         // worldIn.addEntity(snowballentity);
+         if (p_77659_1_.func_217376_c(snowballentity)) {
+            if (!p_77659_2_.field_71075_bZ.field_75098_d) {
+               itemstack.func_190918_g(1);
+            }
+            p_77659_1_.func_184148_a((PlayerEntity) null, p_77659_2_.func_226277_ct_(), p_77659_2_.func_226278_cu_(), p_77659_2_.func_226281_cx_(), SoundEvents.field_187797_fA, SoundCategory.NEUTRAL, 0.5F, 0.4F / (field_77697_d.nextFloat() * 0.4F + 0.8F));
+         } else if (p_77659_2_ instanceof ServerPlayerEntity) {
+            ((ServerPlayerEntity) p_77659_2_).getBukkitEntity().updateInventory();
+         }
       }
+      // CraftBukkit end
 
       p_77659_2_.func_71029_a(Stats.field_75929_E.func_199076_b(this));
-      if (!p_77659_2_.field_71075_bZ.field_75098_d) {
-         itemstack.func_190918_g(1);
+
+      // CraftBukkit start - moved up
+      /*
+      if (!playerIn.abilities.isCreativeMode) {
+         itemstack.shrink(1);
       }
+       */
 
       return ActionResult.func_233538_a_(itemstack, p_77659_1_.func_201670_d());
    }
