--- a/net/minecraft/item/TridentItem.java
+++ b/net/minecraft/item/TridentItem.java
@@ -1,8 +_,8 @@
 package net.minecraft.item;
 
 import com.google.common.collect.ImmutableMultimap;
-import com.google.common.collect.Multimap;
 import com.google.common.collect.ImmutableMultimap.Builder;
+import com.google.common.collect.Multimap;
 import net.minecraft.block.BlockState;
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.enchantment.IVanishable;
@@ -12,6 +_,7 @@
 import net.minecraft.entity.ai.attributes.AttributeModifier;
 import net.minecraft.entity.ai.attributes.Attributes;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.AbstractArrowEntity;
 import net.minecraft.entity.projectile.TridentEntity;
 import net.minecraft.inventory.EquipmentSlotType;
@@ -67,7 +_,19 @@
                         tridententity.field_70251_a = AbstractArrowEntity.PickupStatus.CREATIVE_ONLY;
                      }
 
-                     p_77615_2_.func_217376_c(tridententity);
+                     // CraftBukkit start
+                     if (!p_77615_2_.func_217376_c(tridententity)) {
+                        if (playerentity instanceof ServerPlayerEntity) {
+                           ((ServerPlayerEntity) playerentity).getBukkitEntity().updateInventory();
+                        }
+                        return;
+                     }
+                     p_77615_1_.func_222118_a(1, playerentity, (p_220047_1_) -> {
+                        p_220047_1_.func_213334_d(p_77615_3_.func_184600_cs());
+                     });
+                     tridententity.field_203054_h = p_77615_1_.func_77946_l(); // SPIGOT-4511 update since damage call moved
+                     // CraftBukkit end
+
                      p_77615_2_.func_217384_a((PlayerEntity)null, tridententity, SoundEvents.field_203274_ip, SoundCategory.PLAYERS, 1.0F, 1.0F);
                      if (!playerentity.field_71075_bZ.field_75098_d) {
                         playerentity.field_71071_by.func_184437_d(p_77615_1_);
@@ -77,6 +_,10 @@
 
                playerentity.func_71029_a(Stats.field_75929_E.func_199076_b(this));
                if (j > 0) {
+                  // CraftBukkit start
+                  org.bukkit.event.player.PlayerRiptideEvent event = new org.bukkit.event.player.PlayerRiptideEvent((org.bukkit.entity.Player) playerentity.getBukkitEntity(), org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack.asCraftMirror(p_77615_1_));
+                  event.getPlayer().getServer().getPluginManager().callEvent(event);
+                  // CraftBukkit end
                   float f7 = playerentity.field_70177_z;
                   float f = playerentity.field_70125_A;
                   float f1 = -MathHelper.func_76126_a(f7 * ((float)Math.PI / 180F)) * MathHelper.func_76134_b(f * ((float)Math.PI / 180F));
