--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/FMLRelaunchLog.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/FMLRelaunchLog.java
@@ -59,9 +59,9 @@
         ThreadContext.put("side", side.name().toLowerCase(Locale.ENGLISH));
         configured = true;
         
-        FMLRelaunchLog.fine("Injecting tracing printstreams for STDOUT/STDERR.");
-        System.setOut(new TracingPrintStream(LogManager.getLogger("STDOUT"), System.out));
-        System.setErr(new TracingPrintStream(LogManager.getLogger("STDERR"), System.err));
+//        FMLRelaunchLog.fine("Injecting tracing printstreams for STDOUT/STDERR.");
+//        System.setOut(new TracingPrintStream(LogManager.getLogger("STDOUT"), System.out));
+//        System.setErr(new TracingPrintStream(LogManager.getLogger("STDERR"), System.err));
     }
 
     public static void log(String targetLog, Level level, String format, Object... data)
