--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -16,12 +_,15 @@
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class PrimedTnt extends Entity {
    private static final EntityDataAccessor<Integer> f_32071_ = SynchedEntityData.m_135353_(PrimedTnt.class, EntityDataSerializers.f_135028_);
    private static final int f_149679_ = 80;
    @Nullable
    public LivingEntity f_32072_;
+   public float yield = 4; // CraftBukkit - add field
+   public boolean isIncendiary = false; // CraftBukkit - add field
 
    public PrimedTnt(EntityType<? extends PrimedTnt> p_32076_, Level p_32077_) {
       super(p_32076_, p_32077_);
@@ -53,6 +_,7 @@
    }
 
    public void m_8119_() {
+      if (f_19853_.spigotConfig.currentPrimedTnt++ > f_19853_.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
       if (!this.m_20068_()) {
          this.m_20256_(this.m_20184_().m_82520_(0.0D, -0.04D, 0.0D));
       }
@@ -66,10 +_,10 @@
       int i = this.m_32100_() - 1;
       this.m_32085_(i);
       if (i <= 0) {
-         this.m_146870_();
          if (!this.f_19853_.f_46443_) {
             this.m_32103_();
          }
+         this.m_146870_();
       } else {
          this.m_20073_();
          if (this.f_19853_.f_46443_) {
@@ -81,7 +_,13 @@
 
    protected void m_32103_() {
       float f = 4.0F;
-      this.f_19853_.m_46511_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), 4.0F, Explosion.BlockInteraction.BREAK);
+      ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+      this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+      if (!event.isCancelled()) {
+         this.f_19853_.m_46518_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), event.getRadius(), event.getFire(), Explosion.BlockInteraction.BREAK);
+      }
+      // CraftBukkit end
    }
 
    protected void m_7380_(CompoundTag p_32097_) {
