--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -1,6 +_,5 @@
 package net.minecraft.world.entity.item;
 
-import javax.annotation.Nullable;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
@@ -8,20 +_,20 @@
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MoverType;
-import net.minecraft.world.entity.Pose;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+
+import javax.annotation.Nullable;
 
 public class PrimedTnt extends Entity {
    private static final EntityDataAccessor<Integer> f_32071_ = SynchedEntityData.m_135353_(PrimedTnt.class, EntityDataSerializers.f_135028_);
    private static final int f_149679_ = 80;
    @Nullable
    public LivingEntity f_32072_;
+   public float yield = 4; // CraftBukkit - add field
+   public boolean isIncendiary = false; // CraftBukkit - add field
 
    public PrimedTnt(EntityType<? extends PrimedTnt> p_32076_, Level p_32077_) {
       super(p_32076_, p_32077_);
@@ -53,6 +_,7 @@
    }
 
    public void m_8119_() {
+      if (f_19853_.spigotConfig.maxTntTicksPerTick > 0 && ++f_19853_.spigotConfig.currentPrimedTnt > f_19853_.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
       if (!this.m_20068_()) {
          this.m_20256_(this.m_20184_().m_82520_(0.0D, -0.04D, 0.0D));
       }
@@ -81,7 +_,13 @@
 
    protected void m_32103_() {
       float f = 4.0F;
-      this.f_19853_.m_46511_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), 4.0F, Explosion.BlockInteraction.BREAK);
+      ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+      this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+      if (!event.isCancelled()) {
+         this.f_19853_.m_46518_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), event.getRadius(), event.getFire(), Explosion.BlockInteraction.BREAK);
+      }
+      // CraftBukkit end
    }
 
    protected void m_7380_(CompoundTag p_32097_) {
