--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -16,12 +_,15 @@
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class PrimedTnt extends Entity {
    private static final EntityDataAccessor<Integer> f_32071_ = SynchedEntityData.m_135353_(PrimedTnt.class, EntityDataSerializers.f_135028_);
    private static final int f_149679_ = 80;
    @Nullable
-   private LivingEntity f_32072_;
+   public LivingEntity f_32072_;
+   public float yield = 4; // CraftBukkit - add fiele
+   public boolean isIncendiary = false; // CraftBukkit - add field
 
    public PrimedTnt(EntityType<? extends PrimedTnt> p_32076_, Level p_32077_) {
       super(p_32076_, p_32077_);
@@ -66,10 +_,13 @@
       int i = this.m_32100_() - 1;
       this.m_32085_(i);
       if (i <= 0) {
-         this.m_146870_();
+         // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
+         // this.discard();
          if (!this.f_19853_.f_46443_) {
             this.m_32103_();
          }
+         this.m_146870_();
+         // CraftBukkit end
       } else {
          this.m_20073_();
          if (this.f_19853_.f_46443_) {
@@ -80,8 +_,12 @@
    }
 
    protected void m_32103_() {
-      float f = 4.0F;
-      this.f_19853_.m_46511_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), 4.0F, Explosion.BlockInteraction.BREAK);
+      // CraftBukkit start
+      ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+      this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+      if (!event.isCancelled()) {
+         this.f_19853_.m_46518_(this, this.m_20185_(), this.m_20227_(0.0625D), this.m_20189_(), event.getRadius(), event.getFire(), Explosion.BlockInteraction.BREAK);
+      }
    }
 
    protected void m_7380_(CompoundTag p_32097_) {
