--- a/net/minecraft/world/end/DragonFightManager.java
+++ b/net/minecraft/world/end/DragonFightManager.java
@@ -58,8 +_,8 @@
 public class DragonFightManager {
    private static final Logger field_186107_a = LogManager.getLogger();
    private static final Predicate<Entity> field_186108_b = EntityPredicates.field_94557_a.and(EntityPredicates.func_188443_a(0.0D, 128.0D, 0.0D, 192.0D));
-   private final ServerBossInfo field_186109_c = (ServerBossInfo)(new ServerBossInfo(new TranslationTextComponent("entity.minecraft.ender_dragon"), BossInfo.Color.PINK, BossInfo.Overlay.PROGRESS)).func_186742_b(true).func_186743_c(true);
-   private final ServerWorld field_186110_d;
+   public final ServerBossInfo field_186109_c = (ServerBossInfo)(new ServerBossInfo(new TranslationTextComponent("entity.minecraft.ender_dragon"), BossInfo.Color.PINK, BossInfo.Overlay.PROGRESS)).func_186742_b(true).func_186743_c(true);
+   public final ServerWorld field_186110_d;
    private final List<Integer> field_186111_e = Lists.newArrayList();
    private final BlockPattern field_186112_f;
    private int field_186113_g;
@@ -68,10 +_,10 @@
    private int field_186116_j;
    private boolean field_186117_k;
    private boolean field_186118_l;
-   private UUID field_186119_m;
+   public UUID field_186119_m;
    private boolean field_186120_n = true;
-   private BlockPos field_186121_o;
-   private DragonSpawnState field_186122_p;
+   public BlockPos field_186121_o;
+   public DragonSpawnState field_186122_p;
    private int field_186123_q;
    private List<EnderCrystalEntity> field_186124_r;
 
@@ -84,6 +_,7 @@
 
          this.field_186117_k = p_i231901_4_.func_74767_n("DragonKilled");
          this.field_186118_l = p_i231901_4_.func_74767_n("PreviouslyKilled");
+         this.field_186120_n = !p_i231901_4_.func_74767_n("LegacyScanPerformed"); // Forge: fix MC-105080
          if (p_i231901_4_.func_74767_n("IsRespawning")) {
             this.field_186122_p = DragonSpawnState.START;
          }
@@ -118,6 +_,7 @@
 
       compoundnbt.func_74757_a("DragonKilled", this.field_186117_k);
       compoundnbt.func_74757_a("PreviouslyKilled", this.field_186118_l);
+      compoundnbt.func_74757_a("LegacyScanPerformed", !this.field_186120_n); // Forge: fix MC-105080
       if (this.field_186121_o != null) {
          compoundnbt.func_218657_a("ExitPortalLocation", NBTUtil.func_186859_a(this.field_186121_o));
       }
@@ -220,7 +_,7 @@
 
    }
 
-   protected void func_186095_a(DragonSpawnState p_186095_1_) {
+   public void func_186095_a(DragonSpawnState p_186095_1_) {
       if (this.field_186122_p == null) {
          throw new IllegalStateException("Dragon respawn isn't in progress, can't skip ahead in the animation.");
       } else {
@@ -257,7 +_,7 @@
    }
 
    @Nullable
-   private BlockPattern.PatternHelper func_186091_h() {
+   public BlockPattern.PatternHelper func_186091_h() {
       for(int i = -8; i <= 8; ++i) {
          for(int j = -8; j <= 8; ++j) {
             Chunk chunk = this.field_186110_d.func_212866_a_(i, j);
@@ -370,7 +_,7 @@
       Features.field_243907_c.func_242765_a(this.field_186110_d, this.field_186110_d.func_72863_F().func_201711_g(), new Random(), p_186089_1_);
    }
 
-   private void func_186094_a(boolean p_186094_1_) {
+   public void func_186094_a(boolean p_186094_1_) {
       EndPodiumFeature endpodiumfeature = new EndPodiumFeature(p_186094_1_);
       if (this.field_186121_o == null) {
          for(this.field_186121_o = this.field_186110_d.func_205770_a(Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.field_186139_a).func_177977_b(); this.field_186110_d.func_180495_p(this.field_186121_o).func_203425_a(Blocks.field_150357_h) && this.field_186121_o.func_177956_o() > this.field_186110_d.func_181545_F(); this.field_186121_o = this.field_186121_o.func_177977_b()) {
@@ -490,6 +_,13 @@
             endercrystalentity.func_184516_a((BlockPos)null);
          }
       }
-
+   }
+
+   public void addPlayer(ServerPlayerEntity player) {
+      this.field_186109_c.func_186760_a(player);
+   }
+
+   public void removePlayer(ServerPlayerEntity player) {
+      this.field_186109_c.func_186761_b(player);
    }
 }
