--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -152,7 +_,7 @@
       if (this.f_19853_ != null && !this.f_19853_.f_46443_) {
          this.f_21345_.m_25363_(this.f_32131_);
          this.f_21345_.m_25363_(this.f_32130_);
-         ItemStack itemstack = this.m_21120_(ProjectileUtil.m_37297_(this, Items.f_42411_));
+         ItemStack itemstack = this.m_21120_(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
          if (itemstack.m_150930_(Items.f_42411_)) {
             int i = 20;
             if (this.f_19853_.m_46791_() != Difficulty.HARD) {
@@ -169,15 +_,27 @@
    }
 
    public void m_6504_(LivingEntity p_32141_, float p_32142_) {
-      ItemStack itemstack = this.m_6298_(this.m_21120_(ProjectileUtil.m_37297_(this, Items.f_42411_)));
+      ItemStack itemstack = this.m_6298_(this.m_21120_(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem)));
       AbstractArrow abstractarrow = this.m_7932_(itemstack, p_32142_);
+      if (this.m_21205_().m_41720_() instanceof net.minecraft.world.item.BowItem)
+         abstractarrow = ((net.minecraft.world.item.BowItem)this.m_21205_().m_41720_()).customArrow(abstractarrow);
       double d0 = p_32141_.m_20185_() - this.m_20185_();
       double d1 = p_32141_.m_20227_(0.3333333333333333D) - abstractarrow.m_20186_();
       double d2 = p_32141_.m_20189_() - this.m_20189_();
       double d3 = Math.sqrt(d0 * d0 + d2 * d2);
       abstractarrow.m_6686_(d0, d1 + d3 * (double)0.2F, d2, 1.6F, (float)(14 - this.f_19853_.m_46791_().m_19028_() * 4));
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityShootBowEvent(this, this.m_21205_(), null, abstractarrow, net.minecraft.world.InteractionHand.MAIN_HAND, 0.8F, true);
+      if (event.isCancelled()) {
+         event.getProjectile().remove();
+         return;
+      }
+
+      if (event.getProjectile() == abstractarrow.getBukkitEntity()) {
+         f_19853_.m_7967_(abstractarrow);
+      }
+      // CraftBukkit end
       this.m_5496_(SoundEvents.f_12382_, 1.0F, 1.0F / (this.m_21187_().nextFloat() * 0.4F + 0.8F));
-      this.f_19853_.m_7967_(abstractarrow);
    }
 
    protected AbstractArrow m_7932_(ItemStack p_32156_, float p_32157_) {
