--- a/net/minecraft/world/entity/monster/Drowned.java
+++ b/net/minecraft/world/entity/monster/Drowned.java
@@ -237,7 +_,7 @@
    }
 
    public void m_6504_(LivingEntity p_32356_, float p_32357_) {
-      ThrownTrident throwntrident = new ThrownTrident(this.m_9236_(), this, new ItemStack(Items.f_42713_));
+      ThrownTrident throwntrident = new ThrownTrident(this.m_9236_(), this, this.m_21120_(net.minecraft.world.entity.projectile.ProjectileUtil.m_37297_(this, Items.f_42713_))); // CraftBukkit - Use Trident in hand like skeletons (SPIGOT-7025)
       double d0 = p_32356_.m_20185_() - this.m_20185_();
       double d1 = p_32356_.m_20227_(0.3333333333333333D) - throwntrident.m_20186_();
       double d2 = p_32356_.m_20189_() - this.m_20189_();
