--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -110,7 +_,17 @@
    }
 
    public void m_6710_(@Nullable LivingEntity p_32537_) {
-      super.m_6710_(p_32537_);
+      // CraftBukkit start - fire event
+      setTarget(p_32537_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.UNKNOWN, true);
+   }
+
+   @Override
+   public boolean setTarget(LivingEntity p_32537_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+      if (!super.setTarget(p_32537_, reason, fireEvent)) {
+         return false;
+      }
+      p_32537_ = m_5448_();
+      // CraftBukkit end
       AttributeInstance attributeinstance = this.m_21051_(Attributes.f_22279_);
       if (p_32537_ == null) {
          this.f_32477_ = 0;
@@ -125,6 +_,7 @@
          }
       }
 
+      return true;
    }
 
    protected void m_8097_() {
@@ -199,7 +_,7 @@
 
    boolean m_32534_(Player p_32535_) {
       ItemStack itemstack = p_32535_.m_150109_().f_35975_.get(3);
-      if (itemstack.m_150930_(Blocks.f_50143_.m_5456_())) {
+      if (net.minecraftforge.common.ForgeHooks.shouldSuppressEnderManAnger(this, p_32535_, itemstack)) {
          return false;
       } else {
          Vec3 vec3 = p_32535_.m_20252_(1.0F).m_82541_();
@@ -246,7 +_,7 @@
       super.m_8024_();
    }
 
-   protected boolean m_32529_() {
+   public boolean m_32529_() { // CatServer - protected -> public
       if (!this.m_9236_().m_5776_() && this.m_6084_()) {
          double d0 = this.m_20185_() + (this.f_19796_.m_188500_() - 0.5D) * 64.0D;
          double d1 = this.m_20186_() + (double)(this.f_19796_.m_188503_(64) - 32);
@@ -257,7 +_,7 @@
       }
    }
 
-   boolean m_32500_(Entity p_32501_) {
+   public boolean m_32500_(Entity p_32501_) { // CatServer - public
       Vec3 vec3 = new Vec3(this.m_20185_() - p_32501_.m_20185_(), this.m_20227_(0.5D) - p_32501_.m_20188_(), this.m_20189_() - p_32501_.m_20189_());
       vec3 = vec3.m_82541_();
       double d0 = 16.0D;
@@ -278,8 +_,10 @@
       boolean flag = blockstate.m_280555_();
       boolean flag1 = blockstate.m_60819_().m_205070_(FluidTags.f_13131_);
       if (flag && !flag1) {
+         net.minecraftforge.event.entity.EntityTeleportEvent.EnderEntity event = net.minecraftforge.event.ForgeEventFactory.onEnderTeleport(this, p_32544_, p_32545_, p_32546_);
+         if (event.isCanceled()) return false;
          Vec3 vec3 = this.m_20182_();
-         boolean flag2 = this.m_20984_(p_32544_, p_32545_, p_32546_, true);
+         boolean flag2 = this.m_20984_(event.getTargetX(), event.getTargetY(), event.getTargetZ(), true);
          if (flag2) {
             this.m_9236_().m_214171_(GameEvent.f_238175_, vec3, GameEvent.Context.m_223717_(this));
             if (!this.m_20067_()) {
@@ -419,7 +_,7 @@
       public boolean m_8036_() {
          if (this.f_32554_.m_32530_() == null) {
             return false;
-         } else if (!this.f_32554_.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_32554_.m_9236_(), this.f_32554_)) {
             return false;
          } else {
             return this.f_32554_.m_217043_().m_188503_(m_186073_(2000)) == 0;
@@ -439,17 +_,19 @@
          BlockState blockstate2 = this.f_32554_.m_32530_();
          if (blockstate2 != null) {
             blockstate2 = Block.m_49931_(blockstate2, this.f_32554_.m_9236_(), blockpos);
-            if (this.m_32558_(level, blockpos, blockstate2, blockstate, blockstate1, blockpos1)) {
-               level.m_7731_(blockpos, blockstate2, 3);
-               level.m_220407_(GameEvent.f_157797_, blockpos, GameEvent.Context.m_223719_(this.f_32554_, blockstate2));
-               this.f_32554_.m_32521_((BlockState)null);
+            if (this.m_32558_(level, blockpos, blockstate2, blockstate, blockstate1, blockpos1) && !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(f_32554_, net.minecraftforge.common.util.BlockSnapshot.create(level.m_46472_(), level, blockpos1), net.minecraft.core.Direction.UP)) {
+               if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this.f_32554_, blockpos, blockstate2)) { // CraftBukkit - Place event
+                  level.m_7731_(blockpos, blockstate2, 3);
+                  level.m_220407_(GameEvent.f_157797_, blockpos, GameEvent.Context.m_223719_(this.f_32554_, blockstate2));
+                  this.f_32554_.m_32521_((BlockState)null);
+               } // CraftBukkit
             }
 
          }
       }
 
       private boolean m_32558_(Level p_32559_, BlockPos p_32560_, BlockState p_32561_, BlockState p_32562_, BlockState p_32563_, BlockPos p_32564_) {
-         return p_32562_.m_60795_() && !p_32563_.m_60795_() && !p_32563_.m_60713_(Blocks.f_50752_) && p_32563_.m_60838_(p_32559_, p_32564_) && p_32561_.m_60710_(p_32559_, p_32560_) && p_32559_.m_45933_(this.f_32554_, AABB.m_82333_(Vec3.m_82528_(p_32560_))).isEmpty();
+         return p_32562_.m_60795_() && !p_32563_.m_60795_() && !p_32563_.m_60713_(Blocks.f_50752_) && !p_32563_.m_204336_(net.minecraftforge.common.Tags.Blocks.ENDERMAN_PLACE_ON_BLACKLIST) && p_32563_.m_60838_(p_32559_, p_32564_) && p_32561_.m_60710_(p_32559_, p_32560_) && p_32559_.m_45933_(this.f_32554_, AABB.m_82333_(Vec3.m_82528_(p_32560_))).isEmpty();
       }
    }
 
@@ -551,7 +_,7 @@
       public boolean m_8036_() {
          if (this.f_32583_.m_32530_() != null) {
             return false;
-         } else if (!this.f_32583_.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_32583_.m_9236_(), this.f_32583_)) {
             return false;
          } else {
             return this.f_32583_.m_217043_().m_188503_(m_186073_(20)) == 0;
@@ -571,9 +_,11 @@
          BlockHitResult blockhitresult = level.m_45547_(new ClipContext(vec3, vec31, ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, this.f_32583_));
          boolean flag = blockhitresult.m_82425_().equals(blockpos);
          if (blockstate.m_204336_(BlockTags.f_13046_) && flag) {
-            level.m_7471_(blockpos, false);
-            level.m_220407_(GameEvent.f_157794_, blockpos, GameEvent.Context.m_223719_(this.f_32583_, blockstate));
-            this.f_32583_.m_32521_(blockstate.m_60734_().m_49966_());
+            if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this.f_32583_, blockpos, Blocks.f_50016_.m_49966_())) { // CraftBukkit - Place event
+               level.m_7471_(blockpos, false);
+               level.m_220407_(GameEvent.f_157794_, blockpos, GameEvent.Context.m_223719_(this.f_32583_, blockstate));
+               this.f_32583_.m_32521_(blockstate.m_60734_().m_49966_());
+            } // CraftBukkit
          }
 
       }
