--- a/net/minecraft/world/entity/monster/Endermite.java
+++ b/net/minecraft/world/entity/monster/Endermite.java
@@ -15,7 +_,6 @@
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
-import net.minecraft.world.entity.ai.goal.ClimbOnTopOfPowderSnowGoal;
 import net.minecraft.world.entity.ai.goal.FloatGoal;
 import net.minecraft.world.entity.ai.goal.LookAtPlayerGoal;
 import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
@@ -39,7 +_,6 @@
 
    protected void m_8099_() {
       this.f_21345_.m_25352_(1, new FloatGoal(this));
-      this.f_21345_.m_25352_(1, new ClimbOnTopOfPowderSnowGoal(this, this.f_19853_));
       this.f_21345_.m_25352_(2, new MeleeAttackGoal(this, 1.0D, false));
       this.f_21345_.m_25352_(3, new WaterAvoidingRandomStrollGoal(this, 1.0D));
       this.f_21345_.m_25352_(7, new LookAtPlayerGoal(this, Player.class, 8.0F));
