--- a/net/minecraft/world/entity/monster/Evoker.java
+++ b/net/minecraft/world/entity/monster/Evoker.java
@@ -248,7 +_,14 @@
                vex.m_33994_(Evoker.this);
                vex.m_34033_(blockpos);
                vex.m_33987_(20 * (30 + Evoker.this.f_19796_.m_188503_(90)));
-               serverlevel.m_47205_(vex);
+               // CatServer start - compat:fabric-mod:pehkui-3.8.3+1.14.4-1.21:virtuoel.pehkui.mixin.compat116plus.SummonVexGoalMixin#pehkui$castSpell$spawnEntityAndPassengers
+               try {
+                  catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevelAccessor$addFreshEntityWithPassengers$reason.push(serverlevel, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPELL); // CraftBukkit - Add SpawnReason
+                  serverlevel.m_47205_(vex);
+               } finally {
+                  catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevelAccessor$addFreshEntityWithPassengers$reason.pop(serverlevel);
+               }
+               // CatServer end
             }
          }
 
@@ -275,7 +_,7 @@
             return false;
          } else if (Evoker.this.f_19797_ < this.f_33775_) {
             return false;
-         } else if (!Evoker.this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(Evoker.this.m_9236_(), Evoker.this)) {
             return false;
          } else {
             List<Sheep> list = Evoker.this.m_9236_().m_45971_(Sheep.class, this.f_32705_, Evoker.this, Evoker.this.m_20191_().m_82377_(16.0D, 4.0D, 16.0D));
