--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -259,6 +_,8 @@
                   }
 
                   LargeFireball largefireball = new LargeFireball(level, this.f_32774_, d2, d3, d4, this.f_32774_.m_32751_());
+                  // CraftBukkit - set bukkitYield when setting explosionpower
+                  largefireball.bukkitYield = largefireball.f_37197_ = this.f_32774_.m_32751_();
                   largefireball.m_6034_(this.f_32774_.m_20185_() + vec3.f_82479_ * 4.0D, this.f_32774_.m_20227_(0.5D) + 0.5D, largefireball.m_20189_() + vec3.f_82481_ * 4.0D);
                   level.m_7967_(largefireball);
                   this.f_32773_ = -40;
