--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -61,6 +_,7 @@
    private boolean f_32805_;
    @Nullable
    protected RandomStrollGoal f_32806_;
+   public Guardian.GuardianAttackGoal guardianAttackGoal; // CraftBukkit - 1.20.1 add field
 
    public Guardian(EntityType<? extends Guardian> p_32810_, Level p_32811_) {
       super(p_32810_, p_32811_);
@@ -74,7 +_,7 @@
    protected void m_8099_() {
       MoveTowardsRestrictionGoal movetowardsrestrictiongoal = new MoveTowardsRestrictionGoal(this, 1.0D);
       this.f_32806_ = new RandomStrollGoal(this, 1.0D, 80);
-      this.f_21345_.m_25352_(4, new Guardian.GuardianAttackGoal(this));
+      this.f_21345_.m_25352_(4, guardianAttackGoal = new Guardian.GuardianAttackGoal(this)); // CraftBukkit - 1.20.1 assign field
       this.f_21345_.m_25352_(5, movetowardsrestrictiongoal);
       this.f_21345_.m_25352_(7, this.f_32806_);
       this.f_21345_.m_25352_(8, new LookAtPlayerGoal(this, Player.class, 8.0F));
@@ -119,7 +_,7 @@
       return 80;
    }
 
-   void m_32817_(int p_32818_) {
+   public void m_32817_(int p_32818_) { // CatServer - public
       this.f_19804_.m_135381_(f_32807_, p_32818_);
    }
 
@@ -335,9 +_,9 @@
 
    }
 
-   static class GuardianAttackGoal extends Goal {
+   public static class GuardianAttackGoal extends Goal { // CatServer - public
       private final Guardian f_32867_;
-      private int f_32868_;
+      public int f_32868_; // CatServer - private -> public
       private final boolean f_32869_;
 
       public GuardianAttackGoal(Guardian p_32871_) {
