--- a/net/minecraft/world/entity/monster/Husk.java
+++ b/net/minecraft/world/entity/monster/Husk.java
@@ -49,7 +_,7 @@
       boolean flag = super.m_7327_(p_32892_);
       if (flag && this.m_21205_().m_41619_() && p_32892_ instanceof LivingEntity) {
          float f = this.m_9236_().m_6436_(this.m_20183_()).m_19056_();
-         ((LivingEntity)p_32892_).m_147207_(new MobEffectInstance(MobEffects.f_19612_, 140 * (int)f), this);
+         ((LivingEntity)p_32892_).addEffect(new MobEffectInstance(MobEffects.f_19612_, 140 * (int)f), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
       }
 
       return flag;
