--- a/net/minecraft/world/entity/monster/MagmaCube.java
+++ b/net/minecraft/world/entity/monster/MagmaCube.java
@@ -39,7 +_,7 @@
       return p_32975_.m_45784_(this) && !p_32975_.m_46855_(this.m_142469_());
    }
 
-   protected void m_7839_(int p_32972_, boolean p_32973_) {
+   public void m_7839_(int p_32972_, boolean p_32973_) {
       super.m_7839_(p_32972_, p_32973_);
       this.m_21051_(Attributes.f_22284_).m_22100_((double)(p_32972_ * 3));
    }
@@ -52,7 +_,7 @@
       return ParticleTypes.f_123744_;
    }
 
-   protected ResourceLocation m_7582_() {
+   public ResourceLocation m_7582_() {
       return this.m_33633_() ? BuiltInLootTables.f_78712_ : this.m_6095_().m_20677_();
    }
 
@@ -72,6 +_,7 @@
       Vec3 vec3 = this.m_20184_();
       this.m_20334_(vec3.f_82479_, (double)(this.m_6118_() + (float)this.m_33632_() * 0.1F), vec3.f_82481_);
       this.f_19812_ = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    protected void m_203347_(TagKey<Fluid> p_204065_) {
