--- a/net/minecraft/world/entity/monster/Phantom.java
+++ b/net/minecraft/world/entity/monster/Phantom.java
@@ -222,7 +_,7 @@
 
                for(Player player : list) {
                   if (Phantom.this.m_21040_(player, TargetingConditions.f_26872_)) {
-                     Phantom.this.m_6710_(player);
+                     Phantom.this.setTarget(player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                      return true;
                   }
                }
@@ -372,36 +_,36 @@
             this.f_33238_ = 0.1F;
          }
 
-         double d0 = Phantom.this.f_33097_.f_82479_ - Phantom.this.m_20185_();
-         double d1 = Phantom.this.f_33097_.f_82480_ - Phantom.this.m_20186_();
-         double d2 = Phantom.this.f_33097_.f_82481_ - Phantom.this.m_20189_();
-         double d3 = Math.sqrt(d0 * d0 + d2 * d2);
-         if (Math.abs(d3) > (double)1.0E-5F) {
-            double d4 = 1.0D - Math.abs(d1 * (double)0.7F) / d3;
-            d0 *= d4;
-            d2 *= d4;
-            d3 = Math.sqrt(d0 * d0 + d2 * d2);
-            double d5 = Math.sqrt(d0 * d0 + d2 * d2 + d1 * d1);
-            float f = Phantom.this.m_146908_();
-            float f1 = (float)Mth.m_14136_(d2, d0);
-            float f2 = Mth.m_14177_(Phantom.this.m_146908_() + 90.0F);
-            float f3 = Mth.m_14177_(f1 * (180F / (float)Math.PI));
-            Phantom.this.m_146922_(Mth.m_14148_(f2, f3, 4.0F) - 90.0F);
+         float f = (float)(Phantom.this.f_33097_.f_82479_ - Phantom.this.m_20185_());
+         float f1 = (float)(Phantom.this.f_33097_.f_82480_ - Phantom.this.m_20186_());
+         float f2 = (float)(Phantom.this.f_33097_.f_82481_ - Phantom.this.m_20189_());
+         double d0 = (double)Mth.m_14116_(f * f + f2 * f2);
+         if (Math.abs(d0) > (double)1.0E-5F) {
+            double d1 = 1.0D - (double)Mth.m_14154_(f1 * 0.7F) / d0;
+            f = (float)((double)f * d1);
+            f2 = (float)((double)f2 * d1);
+            d0 = (double)Mth.m_14116_(f * f + f2 * f2);
+            double d2 = (double)Mth.m_14116_(f * f + f2 * f2 + f1 * f1);
+            float f3 = Phantom.this.m_146908_();
+            float f4 = (float)Mth.m_14136_((double)f2, (double)f);
+            float f5 = Mth.m_14177_(Phantom.this.m_146908_() + 90.0F);
+            float f6 = Mth.m_14177_(f4 * (180F / (float)Math.PI));
+            Phantom.this.m_146922_(Mth.m_14148_(f5, f6, 4.0F) - 90.0F);
             Phantom.this.f_20883_ = Phantom.this.m_146908_();
-            if (Mth.m_14145_(f, Phantom.this.m_146908_()) < 3.0F) {
+            if (Mth.m_14145_(f3, Phantom.this.m_146908_()) < 3.0F) {
                this.f_33238_ = Mth.m_14121_(this.f_33238_, 1.8F, 0.005F * (1.8F / this.f_33238_));
             } else {
                this.f_33238_ = Mth.m_14121_(this.f_33238_, 0.2F, 0.025F);
             }
 
-            float f4 = (float)(-(Mth.m_14136_(-d1, d3) * (double)(180F / (float)Math.PI)));
-            Phantom.this.m_146926_(f4);
-            float f5 = Phantom.this.m_146908_() + 90.0F;
-            double d6 = (double)(this.f_33238_ * Mth.m_14089_(f5 * ((float)Math.PI / 180F))) * Math.abs(d0 / d5);
-            double d7 = (double)(this.f_33238_ * Mth.m_14031_(f5 * ((float)Math.PI / 180F))) * Math.abs(d2 / d5);
-            double d8 = (double)(this.f_33238_ * Mth.m_14031_(f4 * ((float)Math.PI / 180F))) * Math.abs(d1 / d5);
+            float f7 = (float)(-(Mth.m_14136_((double)(-f1), d0) * (double)(180F / (float)Math.PI)));
+            Phantom.this.m_146926_(f7);
+            float f8 = Phantom.this.m_146908_() + 90.0F;
+            double d3 = (double)(this.f_33238_ * Mth.m_14089_(f8 * ((float)Math.PI / 180F))) * Math.abs((double)f / d2);
+            double d4 = (double)(this.f_33238_ * Mth.m_14031_(f8 * ((float)Math.PI / 180F))) * Math.abs((double)f2 / d2);
+            double d5 = (double)(this.f_33238_ * Mth.m_14031_(f7 * ((float)Math.PI / 180F))) * Math.abs((double)f1 / d2);
             Vec3 vec3 = Phantom.this.m_20184_();
-            Phantom.this.m_20256_(vec3.m_82549_((new Vec3(d6, d8, d7)).m_82546_(vec3).m_82490_(0.2D)));
+            Phantom.this.m_20256_(vec3.m_82549_((new Vec3(d3, d5, d4)).m_82546_(vec3).m_82490_(0.2D)));
          }
 
       }
