--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -73,13 +_,13 @@
 
    }
 
-   private void m_149830_(int p_149831_) {
+   public void m_149830_(int p_149831_) {
       this.f_149828_ = p_149831_;
       this.f_19804_.m_135381_(f_149826_, true);
    }
 
    protected void m_149840_() {
-      this.m_21406_(EntityType.f_20481_, true);
+      this.m_21406_(EntityType.f_20481_, true); // FoxServer TODO
       if (!this.m_20067_()) {
          this.f_19853_.m_5898_((Player)null, 1048, this.m_142538_(), 0);
       }
