--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -14,10 +_,10 @@
 import net.minecraft.world.level.Level;
 
 public class Skeleton extends AbstractSkeleton {
-   private static final EntityDataAccessor<Boolean> f_149826_ = SynchedEntityData.m_135353_(Skeleton.class, EntityDataSerializers.f_135035_);
+   public static final EntityDataAccessor<Boolean> f_149826_ = SynchedEntityData.m_135353_(Skeleton.class, EntityDataSerializers.f_135035_);
    public static final String f_149825_ = "StrayConversionTime";
    private int f_149827_;
-   private int f_149828_;
+   public int f_149828_;
 
    public Skeleton(EntityType<? extends Skeleton> p_33570_, Level p_33571_) {
       super(p_33570_, p_33571_);
@@ -73,13 +_,13 @@
 
    }
 
-   private void m_149830_(int p_149831_) {
+   public void m_149830_(int p_149831_) {
       this.f_149828_ = p_149831_;
       this.f_19804_.m_135381_(f_149826_, true);
    }
 
    protected void m_149840_() {
-      this.m_21406_(EntityType.f_20481_, true);
+      this.convertTo(EntityType.f_20481_, true, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN); // CraftBukkit - add spawn and transform reasons
       if (!this.m_20067_()) {
          this.f_19853_.m_5898_((Player)null, 1048, this.m_142538_(), 0);
       }
