--- a/net/minecraft/world/entity/monster/Slime.java
+++ b/net/minecraft/world/entity/monster/Slime.java
@@ -42,6 +_,14 @@
 import net.minecraft.world.level.levelgen.WorldgenRandom;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import java.util.ArrayList;
+import java.util.List;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.SlimeSplitEvent;
+// CraftBukkit end
+
 public class Slime extends Mob implements Enemy {
    private static final EntityDataAccessor<Integer> f_33582_ = SynchedEntityData.m_135353_(Slime.class, EntityDataSerializers.f_135028_);
    public static final int f_149844_ = 1;
@@ -124,6 +_,8 @@
       if (this.m_20096_() && !this.f_33583_) {
          int i = this.m_33632_();
 
+         // Forge: Don't spawn particles if it's handled by the implementation itself
+         if (!spawnCustomParticles())
          for(int j = 0; j < i * 8; ++j) {
             float f = this.f_19796_.m_188501_() * ((float)Math.PI * 2F);
             float f1 = this.f_19796_.m_188501_() * 0.5F + 0.5F;
@@ -184,6 +_,18 @@
          int j = i / 2;
          int k = 2 + this.f_19796_.m_188503_(3);
 
+         // CraftBukkit start
+         SlimeSplitEvent event = new SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), k);
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled() && event.getCount() > 0) {
+            k = event.getCount();
+         } else {
+            super.m_142687_(p_149847_);
+            return;
+         }
+         List<LivingEntity> slimes = new ArrayList<>(j);
+         // CraftBukkit end
+
          for(int l = 0; l < k; ++l) {
             float f1 = ((float)(l % 2) - 0.5F) * f;
             float f2 = ((float)(l / 2) - 0.5F) * f;
@@ -198,9 +_,26 @@
                slime.m_20331_(this.m_20147_());
                slime.m_7839_(j, true);
                slime.m_7678_(this.m_20185_() + (double)f1, this.m_20186_() + 0.5D, this.m_20189_() + (double)f2, this.f_19796_.m_188501_() * 360.0F, 0.0F);
-               this.m_9236_().m_7967_(slime);
-            }
-         }
+               slimes.add(slime); // CraftBukkit
+            }
+         }
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityTransformEvent(this, slimes, EntityTransformEvent.TransformReason.SPLIT).isCancelled()) {
+            super.m_142687_(p_149847_);
+            return;
+         }
+         for (LivingEntity living : slimes) {
+            // CatServer start - Fix mixin inject
+            Level level = this.m_9236_();
+            try {
+               catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addFreshEntity$reason.push(level, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT);
+               this.m_9236_().m_7967_(living);
+            } finally {
+               catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addFreshEntity$reason.pop(level); // CatServer - Make sure reset
+            }
+            // CatServer end
+         }
+         // CraftBukkit end
       }
 
       super.m_142687_(p_149847_);
@@ -319,6 +_,12 @@
    public EntityDimensions m_6972_(Pose p_33597_) {
       return super.m_6972_(p_33597_).m_20388_(0.255F * (float)this.m_33632_());
    }
+
+   /**
+    * Called when the slime spawns particles on landing, see onUpdate.
+    * Return true to prevent the spawning of the default particles.
+    */
+   protected boolean spawnCustomParticles() { return false; }
 
    static class SlimeAttackGoal extends Goal {
       private final Slime f_33645_;
