--- a/net/minecraft/world/entity/monster/Slime.java
+++ b/net/minecraft/world/entity/monster/Slime.java
@@ -1,6 +_,10 @@
 package net.minecraft.world.entity.monster;
 
+import java.util.ArrayList;
 import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -42,6 +_,9 @@
 import net.minecraft.world.level.levelgen.WorldgenRandom;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.SlimeSplitEvent;
 
 public class Slime extends Mob implements Enemy {
    private static final EntityDataAccessor<Integer> f_33582_ = SynchedEntityData.m_135353_(Slime.class, EntityDataSerializers.f_135028_);
@@ -73,7 +_,7 @@
       this.f_19804_.m_135372_(f_33582_, 1);
    }
 
-   protected void m_7839_(int p_33594_, boolean p_33595_) {
+   public void m_7839_(int p_33594_, boolean p_33595_) {
       int i = Mth.m_14045_(p_33594_, 1, 127);
       this.f_19804_.m_135381_(f_33582_, i);
       this.m_20090_();
@@ -123,6 +_,7 @@
       if (this.f_19861_ && !this.f_33583_) {
          int i = this.m_33632_();
 
+         if (spawnCustomParticles()) i = 0; // don't spawn particles if it's handled by the implementation itself
          for(int j = 0; j < i * 8; ++j) {
             float f = this.f_19796_.nextFloat() * ((float)Math.PI * 2F);
             float f1 = this.f_19796_.nextFloat() * 0.5F + 0.5F;
@@ -182,7 +_,18 @@
          float f = (float)i / 4.0F;
          int j = i / 2;
          int k = 2 + this.f_19796_.nextInt(3);
+         // CraftBukkit start
+         SlimeSplitEvent event = new SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), k);
+         this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
 
+         if (!event.isCancelled() && event.getCount() > 0) {
+            k = event.getCount();
+         } else {
+            super.m_142687_(p_149847_);
+            return;
+         }
+         List<LivingEntity> slimes = new ArrayList<>(j);
+         // CraftBukkit end
          for(int l = 0; l < k; ++l) {
             float f1 = ((float)(l % 2) - 0.5F) * f;
             float f2 = ((float)(l / 2) - 0.5F) * f;
@@ -196,8 +_,17 @@
             slime.m_20331_(this.m_20147_());
             slime.m_7839_(j, true);
             slime.m_7678_(this.m_20185_() + (double)f1, this.m_20186_() + 0.5D, this.m_20189_() + (double)f2, this.f_19796_.nextFloat() * 360.0F, 0.0F);
-            this.f_19853_.m_7967_(slime);
-         }
+            slimes.add(slime); // CraftBukkit
+         }
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityTransformEvent(this, slimes, EntityTransformEvent.TransformReason.SPLIT).isCancelled()) {
+            super.m_142687_(p_149847_);
+            return;
+         }
+         for (LivingEntity living : slimes) {
+            this.f_19853_.addFreshEntity(living, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT); // CraftBukkit - SpawnReason
+         }
+         // CraftBukkit end
       }
 
       super.m_142687_(p_149847_);
@@ -253,13 +_,13 @@
       return this.m_33633_() ? SoundEvents.f_12470_ : SoundEvents.f_12388_;
    }
 
-   protected ResourceLocation m_7582_() {
+   public ResourceLocation m_7582_() {
       return this.m_33632_() == 1 ? this.m_6095_().m_20677_() : BuiltInLootTables.f_78712_;
    }
 
    public static boolean m_33620_(EntityType<Slime> p_33621_, LevelAccessor p_33622_, MobSpawnType p_33623_, BlockPos p_33624_, Random p_33625_) {
       if (p_33622_.m_46791_() != Difficulty.PEACEFUL) {
-         if (p_33622_.m_204166_(p_33624_).m_203565_(Biomes.f_48207_) && p_33624_.m_123342_() > 50 && p_33624_.m_123342_() < 70 && p_33625_.nextFloat() < 0.5F && p_33625_.nextFloat() < p_33622_.m_46940_() && p_33622_.m_46803_(p_33624_) <= p_33625_.nextInt(8)) {
+         if (Objects.equals(p_33622_.m_204166_(p_33624_), Optional.of(Biomes.f_48207_)) && p_33624_.m_123342_() > 50 && p_33624_.m_123342_() < 70 && p_33625_.nextFloat() < 0.5F && p_33625_.nextFloat() < p_33622_.m_46940_() && p_33622_.m_46803_(p_33624_) <= p_33625_.nextInt(8)) {
             return m_21400_(p_33621_, p_33622_, p_33623_, p_33624_, p_33625_);
          }
 
@@ -268,7 +_,7 @@
          }
 
          ChunkPos chunkpos = new ChunkPos(p_33624_);
-         boolean flag = WorldgenRandom.m_64685_(chunkpos.f_45578_, chunkpos.f_45579_, ((WorldGenLevel)p_33622_).m_7328_(), 987234911L).nextInt(10) == 0;
+         boolean flag = WorldgenRandom.m_64685_(chunkpos.f_45578_, chunkpos.f_45579_, ((WorldGenLevel)p_33622_).m_7328_(), p_33622_.getMinecraftWorld().spigotConfig.slimeSeed).nextInt(10) == 0; // Spigot
          if (p_33625_.nextInt(10) == 0 && flag && p_33624_.m_123342_() < 40) {
             return m_21400_(p_33621_, p_33622_, p_33623_, p_33624_, p_33625_);
          }
@@ -319,6 +_,12 @@
    public EntityDimensions m_6972_(Pose p_33597_) {
       return super.m_6972_(p_33597_).m_20388_(0.255F * (float)this.m_33632_());
    }
+
+   /**
+    * Called when the slime spawns particles on landing, see onUpdate.
+    * Return true to prevent the spawning of the default particles.
+    */
+   protected boolean spawnCustomParticles() { return false; }
 
    static class SlimeAttackGoal extends Goal {
       private final Slime f_33645_;
