--- a/net/minecraft/world/entity/monster/SpellcasterIllager.java
+++ b/net/minecraft/world/entity/monster/SpellcasterIllager.java
@@ -59,7 +_,7 @@
       this.f_19804_.m_135381_(f_33720_, (byte)p_33728_.f_33747_);
    }
 
-   protected SpellcasterIllager.IllagerSpell m_33737_() {
+   public SpellcasterIllager.IllagerSpell m_33737_() {
       return !this.f_19853_.f_46443_ ? this.f_33721_ : SpellcasterIllager.IllagerSpell.m_33758_(this.f_19804_.m_135370_(f_33720_));
    }
 
@@ -93,7 +_,7 @@
 
    protected abstract SoundEvent m_7894_();
 
-   protected static enum IllagerSpell {
+   public static enum IllagerSpell {
       NONE(0, 0.0D, 0.0D, 0.0D),
       SUMMON_VEX(1, 0.7D, 0.7D, 0.8D),
       FANGS(2, 0.4D, 0.3D, 0.35D),
@@ -184,6 +_,11 @@
       public void m_8037_() {
          --this.f_33774_;
          if (this.f_33774_ == 0) {
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleEntitySpellCastEvent(SpellcasterIllager.this, this.m_7269_())) {
+               return;
+            }
+            // CraftBukkit end
             this.m_8130_();
             SpellcasterIllager.this.m_5496_(SpellcasterIllager.this.m_7894_(), 1.0F, 1.0F);
          }
