--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -76,7 +_,7 @@
    private static final EntityDataAccessor<Integer> f_33854_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Boolean> f_33855_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135035_);
    private static final EntityDataAccessor<Boolean> f_33856_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135035_);
-   private final ItemBasedSteering f_33857_ = new ItemBasedSteering(this.f_19804_, f_33854_, f_33856_);
+   public final ItemBasedSteering f_33857_ = new ItemBasedSteering(this.f_19804_, f_33854_, f_33856_);
    @Nullable
    private TemptGoal f_33858_;
    @Nullable
@@ -277,7 +_,13 @@
       BlockState blockstate = this.f_19853_.m_8055_(this.m_142538_());
       BlockState blockstate1 = this.m_20075_();
       boolean flag = blockstate.m_204336_(BlockTags.f_13086_) || blockstate1.m_204336_(BlockTags.f_13086_) || this.m_204036_(FluidTags.f_13132_) > 0.0D;
-      this.m_33951_(!flag);
+      // CraftBukkit start
+      if (!flag ^ this.m_33935_()) {
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, !flag)) {
+            this.m_33951_(!flag);
+         }
+      }
+      // CraftBukkit end
       super.m_8119_();
       this.m_33940_();
       this.m_20101_();
