--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -58,7 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.LiquidBlock;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.material.FluidState;
+import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.level.pathfinder.PathFinder;
@@ -164,8 +_,8 @@
       return this.m_20202_() instanceof Strider ? ((Strider)this.m_20202_()).m_33935_() : this.f_19804_.m_135370_(f_33855_);
    }
 
-   public boolean m_203441_(FluidState p_204067_) {
-      return p_204067_.m_205070_(FluidTags.f_13132_);
+   public boolean canStandOnFluid(Fluid p_33893_) {
+      return p_33893_.m_205067_(FluidTags.f_13132_);
    }
 
    public double m_6048_() {
@@ -277,7 +_,13 @@
       BlockState blockstate = this.f_19853_.m_8055_(this.m_142538_());
       BlockState blockstate1 = this.m_20075_();
       boolean flag = blockstate.m_204336_(BlockTags.f_13086_) || blockstate1.m_204336_(BlockTags.f_13086_) || this.m_204036_(FluidTags.f_13132_) > 0.0D;
-      this.m_33951_(!flag);
+      // CraftBukkit start
+      if (!flag ^ this.m_33935_()) {
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, !flag)) {
+            this.m_33951_(!flag);
+         }
+      }
+      // CraftBukkit end
       super.m_8119_();
       this.m_33940_();
       this.m_20101_();
