--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -47,8 +_,8 @@
    Mob f_33980_;
    @Nullable
    private BlockPos f_33981_;
-   private boolean f_33978_;
-   private int f_33979_;
+   public boolean f_33978_;
+   public int f_33979_;
 
    public Vex(EntityType<? extends Vex> p_33984_, Level p_33985_) {
       super(p_33984_, p_33985_);
@@ -265,7 +_,7 @@
       }
 
       public void m_8056_() {
-         Vex.this.m_6710_(Vex.this.f_33980_.m_5448_());
+         Vex.this.setTarget(Vex.this.f_33980_.m_5448_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true);
          super.m_8056_();
       }
    }
