--- a/net/minecraft/world/entity/monster/Vindicator.java
+++ b/net/minecraft/world/entity/monster/Vindicator.java
@@ -51,7 +_,7 @@
    static final Predicate<Difficulty> f_34070_ = (p_34082_) -> {
       return p_34082_ == Difficulty.NORMAL || p_34082_ == Difficulty.HARD;
    };
-   boolean f_34071_;
+   public boolean f_34071_;
 
    public Vindicator(EntityType<? extends Vindicator> p_34074_, Level p_34075_) {
       super(p_34074_, p_34075_);
