--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -116,7 +_,7 @@
                   List<MobEffectInstance> list = PotionUtils.m_43547_(itemstack);
                   if (list != null) {
                      for(MobEffectInstance mobeffectinstance : list) {
-                        this.m_7292_(new MobEffectInstance(mobeffectinstance));
+                        this.addEffect(new MobEffectInstance(mobeffectinstance), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
                      }
                   }
                }
@@ -179,7 +_,7 @@
       }
 
       if (p_34149_.m_19387_()) {
-         p_34150_ *= 0.15F;
+         p_34150_ = (float)((double)p_34150_ * 0.15D);
       }
 
       return p_34150_;
