--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EquipmentSlot;
@@ -136,7 +_,7 @@
       }).filter((p_34456_) -> {
          return !p_34456_.m_7307_(this.m_5448_());
       }).forEach((p_34440_) -> {
-         p_34440_.m_6710_(this.m_5448_());
+         p_34440_.setTarget(this.m_5448_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -144,7 +_,8 @@
       this.m_5496_(SoundEvents.f_12611_, this.m_6121_() * 2.0F, this.m_6100_() * 1.8F);
    }
 
-   public void m_6710_(@Nullable LivingEntity p_34478_) {
+   // FoxServer TODO
+   public boolean setTarget(@Nullable LivingEntity p_34478_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) { // CraftBukkit - signature
       if (this.m_5448_() == null && p_34478_ != null) {
          this.f_34417_ = f_34424_.m_142270_(this.f_19796_);
          this.f_34422_ = f_34421_.m_142270_(this.f_19796_);
@@ -154,11 +_,21 @@
          this.m_6598_((Player)p_34478_);
       }
 
-      super.m_6710_(p_34478_);
+      return super.setTarget(p_34478_, reason, fireEvent); // CraftBukkit
    }
 
    public void m_6825_() {
       this.m_7870_(f_34418_.m_142270_(this.f_19796_));
+      // CraftBukkit start
+      Entity entity = ((ServerLevel) this.f_19853_).m_8791_(m_6120_());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), f_34418_.m_142270_(this.f_19796_));
+      this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.m_6925_(null);
+         return;
+      }
+      this.m_7870_(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean m_34449_(EntityType<ZombifiedPiglin> p_34450_, LevelAccessor p_34451_, MobSpawnType p_34452_, BlockPos p_34453_, Random p_34454_) {
