--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -14,6 +_,7 @@
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
@@ -46,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import org.slf4j.Logger;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 
 public class ZombieVillager extends Zombie implements VillagerDataHolder {
    private static final Logger f_201975_ = LogUtils.getLogger();
@@ -63,6 +_,7 @@
    @Nullable
    private CompoundTag f_34362_;
    private int f_34363_;
+   private int lastTick = MinecraftServer.currentTick; // CraftBukkit - add field
 
    public ZombieVillager(EntityType<? extends ZombieVillager> p_34368_, Level p_34369_) {
       super(p_34368_, p_34369_);
@@ -126,13 +_,18 @@
    public void m_8119_() {
       if (!this.f_19853_.f_46443_ && this.m_6084_() && this.m_34408_()) {
          int i = this.m_34410_();
+         // CraftBukkit start - Use wall time instead of ticks for villager conversion
+         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+         i *= elapsedTicks;
+         // CraftBukkit end
          this.f_34365_ -= i;
-         if (this.f_34365_ <= 0) {
+         if (this.f_34365_ <= 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.f_20492_, (timer) -> this.f_34365_ = timer)) {
             this.m_34398_((ServerLevel)this.f_19853_);
          }
       }
 
       super.m_8119_();
+      this.lastTick = MinecraftServer.currentTick; // CraftBukkit
    }
 
    public InteractionResult m_6071_(Player p_34394_, InteractionHand p_34395_) {
@@ -169,12 +_,13 @@
       return this.m_20088_().m_135370_(f_34359_);
    }
 
-   private void m_34383_(@Nullable UUID p_34384_, int p_34385_) {
+   public void m_34383_(@Nullable UUID p_34384_, int p_34385_) {
       this.f_34360_ = p_34384_;
       this.f_34365_ = p_34385_;
       this.m_20088_().m_135381_(f_34359_, true);
-      this.m_21195_(MobEffects.f_19613_);
-      this.m_7292_(new MobEffectInstance(MobEffects.f_19600_, p_34385_, Math.min(this.f_19853_.m_46791_().m_19028_() - 1, 0)));
+      this.setPersistenceRequired(true); // CraftBukkit - SPIGOT-4684 update persistence
+      this.removeEffect(MobEffects.f_19613_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+      this.addEffect(new MobEffectInstance(MobEffects.f_19600_, p_34385_, Math.min(this.f_19853_.m_46791_().m_19028_() - 1, 0)), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
       this.f_19853_.m_7605_(this, (byte)16);
    }
 
@@ -189,9 +_,9 @@
       }
    }
 
+   // FoxServer TODO
    private void m_34398_(ServerLevel p_34399_) {
       Villager villager = this.m_21406_(EntityType.f_20492_, false);
-
       for(EquipmentSlot equipmentslot : EquipmentSlot.values()) {
          ItemStack itemstack = this.m_6844_(equipmentslot);
          if (!itemstack.m_41619_()) {
@@ -200,7 +_,9 @@
             } else {
                double d0 = (double)this.m_21519_(equipmentslot);
                if (d0 > 1.0D) {
+                  this.forceDrops = true; // CraftBukkit
                   this.m_19983_(itemstack);
+                  this.forceDrops = false; // CraftBukkit
                }
             }
          }
@@ -225,11 +_,11 @@
          }
       }
 
-      villager.m_7292_(new MobEffectInstance(MobEffects.f_19604_, 200, 0));
+      villager.addEffect(new MobEffectInstance(MobEffects.f_19604_, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION); // CraftBukkit
       if (!this.m_20067_()) {
          p_34399_.m_5898_((Player)null, 1027, this.m_142538_(), 0);
       }
-
+      net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, villager);
    }
 
    private int m_34410_() {
