--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -144,7 +_,7 @@
       }).filter((p_289463_) -> {
          return !p_289463_.m_7307_(this.m_5448_());
       }).forEach((p_289464_) -> {
-         p_289464_.m_6710_(this.m_5448_());
+         p_289464_.setTarget(this.m_5448_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -152,7 +_,7 @@
       this.m_5496_(SoundEvents.f_12611_, this.m_6121_() * 2.0F, this.m_6100_() * 1.8F);
    }
 
-   public void m_6710_(@Nullable LivingEntity p_34478_) {
+   public boolean setTarget(@Nullable LivingEntity p_34478_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) { // CraftBukkit - signature
       if (this.m_5448_() == null && p_34478_ != null) {
          this.f_34417_ = f_34424_.m_214085_(this.f_19796_);
          this.f_34422_ = f_34421_.m_214085_(this.f_19796_);
@@ -162,11 +_,20 @@
          this.m_6598_((Player)p_34478_);
       }
 
-      super.m_6710_(p_34478_);
+      return super.setTarget(p_34478_, reason, fireEvent); // CraftBukkit
    }
 
    public void m_6825_() {
-      this.m_7870_(f_34418_.m_214085_(this.f_19796_));
+      // CraftBukkit start
+      net.minecraft.world.entity.Entity entity = ((ServerLevel) this.m_9236_()).m_8791_(m_6120_());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), ZombifiedPiglin.f_34418_.m_214085_(this.f_19796_));
+      this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.m_6925_(null);
+         return;
+      }
+      this.m_7870_(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean m_219173_(EntityType<ZombifiedPiglin> p_219174_, LevelAccessor p_219175_, MobSpawnType p_219176_, BlockPos p_219177_, RandomSource p_219178_) {
