--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -50,16 +_,16 @@
 public class RecipeProvider implements DataProvider {
    private static final Logger f_125968_ = LogUtils.getLogger();
    private static final Gson f_125969_ = (new GsonBuilder()).setPrettyPrinting().create();
-   protected static final ImmutableList<ItemLike> f_176505_ = ImmutableList.of(Items.f_41835_, Items.f_150963_);
-   protected static final ImmutableList<ItemLike> f_176506_ = ImmutableList.of(Items.f_41834_, Items.f_150964_, Items.f_151050_);
-   protected static final ImmutableList<ItemLike> f_176507_ = ImmutableList.of(Items.f_150965_, Items.f_150966_, Items.f_151051_);
-   protected static final ImmutableList<ItemLike> f_176508_ = ImmutableList.of(Items.f_41833_, Items.f_150967_, Items.f_41836_, Items.f_151053_);
-   protected static final ImmutableList<ItemLike> f_176509_ = ImmutableList.of(Items.f_42010_, Items.f_150994_);
-   protected static final ImmutableList<ItemLike> f_176510_ = ImmutableList.of(Items.f_41853_, Items.f_150993_);
-   protected static final ImmutableList<ItemLike> f_176511_ = ImmutableList.of(Items.f_41977_, Items.f_150968_);
-   protected static final ImmutableList<ItemLike> f_176512_ = ImmutableList.of(Items.f_42107_, Items.f_150969_);
+   private static final ImmutableList<ItemLike> f_176505_ = ImmutableList.of(Items.f_41835_, Items.f_150963_);
+   private static final ImmutableList<ItemLike> f_176506_ = ImmutableList.of(Items.f_41834_, Items.f_150964_, Items.f_151050_);
+   private static final ImmutableList<ItemLike> f_176507_ = ImmutableList.of(Items.f_150965_, Items.f_150966_, Items.f_151051_);
+   private static final ImmutableList<ItemLike> f_176508_ = ImmutableList.of(Items.f_41833_, Items.f_150967_, Items.f_41836_, Items.f_151053_);
+   private static final ImmutableList<ItemLike> f_176509_ = ImmutableList.of(Items.f_42010_, Items.f_150994_);
+   private static final ImmutableList<ItemLike> f_176510_ = ImmutableList.of(Items.f_41853_, Items.f_150993_);
+   private static final ImmutableList<ItemLike> f_176511_ = ImmutableList.of(Items.f_41977_, Items.f_150968_);
+   private static final ImmutableList<ItemLike> f_176512_ = ImmutableList.of(Items.f_42107_, Items.f_150969_);
    protected final DataGenerator f_125970_;
-   protected static final Map<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>> f_176513_ = ImmutableMap.<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>>builder().put(BlockFamily.Variant.BUTTON, (p_176733_, p_176734_) -> {
+   private static final Map<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>> f_176513_ = ImmutableMap.<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>>builder().put(BlockFamily.Variant.BUTTON, (p_176733_, p_176734_) -> {
       return m_176658_(p_176733_, Ingredient.m_43929_(p_176734_));
    }).put(BlockFamily.Variant.CHISELED, (p_176730_, p_176731_) -> {
       return m_176646_(p_176730_, Ingredient.m_43929_(p_176731_));
@@ -106,6 +_,7 @@
 
          }
       });
+      if (this.getClass() == RecipeProvider.class) //Forge: Subclasses don't need this.
       m_126013_(p_125982_, Advancement.Builder.m_138353_().m_138386_("impossible", new ImpossibleTrigger.TriggerInstance()).m_138400_(), path.resolve("data/minecraft/advancements/recipes/root.json"));
    }
 
@@ -143,7 +_,7 @@
 
    }
 
-   protected static void m_126013_(HashCache p_126014_, JsonObject p_126015_, Path p_126016_) {
+   protected void m_126013_(HashCache p_126014_, JsonObject p_126015_, Path p_126016_) {
       try {
          String s = f_125969_.toJson((JsonElement)p_126015_);
          String s1 = f_123918_.hashUnencodedChars(s).toString();
@@ -177,7 +_,7 @@
 
    }
 
-   protected static void m_176531_(Consumer<FinishedRecipe> p_176532_) {
+   protected void m_176531_(Consumer<FinishedRecipe> p_176532_) {
       BlockFamilies.m_175934_().filter(BlockFamily::m_175956_).forEach((p_176624_) -> {
          m_176580_(p_176532_, p_176624_);
       });
@@ -981,136 +_,136 @@
       m_125994_(p_176532_, Items.f_42389_, Items.f_42394_);
    }
 
-   protected static void m_176551_(Consumer<FinishedRecipe> p_176552_, ItemLike p_176553_, ItemLike p_176554_, @Nullable String p_176555_) {
+   private static void m_176551_(Consumer<FinishedRecipe> p_176552_, ItemLike p_176553_, ItemLike p_176554_, @Nullable String p_176555_) {
       m_176556_(p_176552_, p_176553_, p_176554_, p_176555_, 1);
    }
 
-   protected static void m_176556_(Consumer<FinishedRecipe> p_176557_, ItemLike p_176558_, ItemLike p_176559_, @Nullable String p_176560_, int p_176561_) {
+   private static void m_176556_(Consumer<FinishedRecipe> p_176557_, ItemLike p_176558_, ItemLike p_176559_, @Nullable String p_176560_, int p_176561_) {
       ShapelessRecipeBuilder.m_126191_(p_176558_, p_176561_).m_126209_(p_176559_).m_142409_(p_176560_).m_142284_(m_176602_(p_176559_), m_125977_(p_176559_)).m_176500_(p_176557_, m_176517_(p_176558_, p_176559_));
    }
 
-   protected static void m_176591_(Consumer<FinishedRecipe> p_176592_, List<ItemLike> p_176593_, ItemLike p_176594_, float p_176595_, int p_176596_, String p_176597_) {
+   private static void m_176591_(Consumer<FinishedRecipe> p_176592_, List<ItemLike> p_176593_, ItemLike p_176594_, float p_176595_, int p_176596_, String p_176597_) {
       m_176533_(p_176592_, RecipeSerializer.f_44091_, p_176593_, p_176594_, p_176595_, p_176596_, p_176597_, "_from_smelting");
    }
 
-   protected static void m_176625_(Consumer<FinishedRecipe> p_176626_, List<ItemLike> p_176627_, ItemLike p_176628_, float p_176629_, int p_176630_, String p_176631_) {
+   private static void m_176625_(Consumer<FinishedRecipe> p_176626_, List<ItemLike> p_176627_, ItemLike p_176628_, float p_176629_, int p_176630_, String p_176631_) {
       m_176533_(p_176626_, RecipeSerializer.f_44092_, p_176627_, p_176628_, p_176629_, p_176630_, p_176631_, "_from_blasting");
    }
 
-   protected static void m_176533_(Consumer<FinishedRecipe> p_176534_, SimpleCookingSerializer<?> p_176535_, List<ItemLike> p_176536_, ItemLike p_176537_, float p_176538_, int p_176539_, String p_176540_, String p_176541_) {
+   private static void m_176533_(Consumer<FinishedRecipe> p_176534_, SimpleCookingSerializer<?> p_176535_, List<ItemLike> p_176536_, ItemLike p_176537_, float p_176538_, int p_176539_, String p_176540_, String p_176541_) {
       for(ItemLike itemlike : p_176536_) {
          SimpleCookingRecipeBuilder.m_126248_(Ingredient.m_43929_(itemlike), p_176537_, p_176538_, p_176539_, p_176535_).m_142409_(p_176540_).m_142284_(m_176602_(itemlike), m_125977_(itemlike)).m_176500_(p_176534_, m_176632_(p_176537_) + p_176541_ + "_" + m_176632_(itemlike));
       }
 
    }
 
-   protected static void m_125994_(Consumer<FinishedRecipe> p_125995_, Item p_125996_, Item p_125997_) {
+   private static void m_125994_(Consumer<FinishedRecipe> p_125995_, Item p_125996_, Item p_125997_) {
       UpgradeRecipeBuilder.m_126385_(Ingredient.m_43929_(p_125996_), Ingredient.m_43929_(Items.f_42418_), p_125997_).m_126389_("has_netherite_ingot", m_125977_(Items.f_42418_)).m_126392_(p_125995_, m_176632_(p_125997_) + "_smithing");
    }
 
-   protected static void m_206408_(Consumer<FinishedRecipe> p_206409_, ItemLike p_206410_, TagKey<Item> p_206411_) {
+   private static void m_206408_(Consumer<FinishedRecipe> p_206409_, ItemLike p_206410_, TagKey<Item> p_206411_) {
       ShapelessRecipeBuilder.m_126191_(p_206410_, 4).m_206419_(p_206411_).m_142409_("planks").m_142284_("has_log", m_206406_(p_206411_)).m_176498_(p_206409_);
    }
 
-   protected static void m_206412_(Consumer<FinishedRecipe> p_206413_, ItemLike p_206414_, TagKey<Item> p_206415_) {
+   private static void m_206412_(Consumer<FinishedRecipe> p_206413_, ItemLike p_206414_, TagKey<Item> p_206415_) {
       ShapelessRecipeBuilder.m_126191_(p_206414_, 4).m_206419_(p_206415_).m_142409_("planks").m_142284_("has_logs", m_206406_(p_206415_)).m_176498_(p_206413_);
    }
 
-   protected static void m_126002_(Consumer<FinishedRecipe> p_126003_, ItemLike p_126004_, ItemLike p_126005_) {
+   private static void m_126002_(Consumer<FinishedRecipe> p_126003_, ItemLike p_126004_, ItemLike p_126005_) {
       ShapedRecipeBuilder.m_126118_(p_126004_, 3).m_126127_('#', p_126005_).m_126130_("##").m_126130_("##").m_142409_("bark").m_142284_("has_log", m_125977_(p_126005_)).m_176498_(p_126003_);
    }
 
-   protected static void m_126021_(Consumer<FinishedRecipe> p_126022_, ItemLike p_126023_, ItemLike p_126024_) {
+   private static void m_126021_(Consumer<FinishedRecipe> p_126022_, ItemLike p_126023_, ItemLike p_126024_) {
       ShapedRecipeBuilder.m_126116_(p_126023_).m_126127_('#', p_126024_).m_126130_("# #").m_126130_("###").m_142409_("boat").m_142284_("in_water", m_125979_(Blocks.f_49990_)).m_176498_(p_126022_);
    }
 
-   protected static RecipeBuilder m_176658_(ItemLike p_176659_, Ingredient p_176660_) {
+   private static RecipeBuilder m_176658_(ItemLike p_176659_, Ingredient p_176660_) {
       return ShapelessRecipeBuilder.m_126189_(p_176659_).m_126184_(p_176660_);
    }
 
-   protected static RecipeBuilder m_176670_(ItemLike p_176671_, Ingredient p_176672_) {
+   private static RecipeBuilder m_176670_(ItemLike p_176671_, Ingredient p_176672_) {
       return ShapedRecipeBuilder.m_126118_(p_176671_, 3).m_126124_('#', p_176672_).m_126130_("##").m_126130_("##").m_126130_("##");
    }
 
-   protected static RecipeBuilder m_176678_(ItemLike p_176679_, Ingredient p_176680_) {
+   private static RecipeBuilder m_176678_(ItemLike p_176679_, Ingredient p_176680_) {
       int i = p_176679_ == Blocks.f_50198_ ? 6 : 3;
       Item item = p_176679_ == Blocks.f_50198_ ? Items.f_42691_ : Items.f_42398_;
       return ShapedRecipeBuilder.m_126118_(p_176679_, i).m_126124_('W', p_176680_).m_126127_('#', item).m_126130_("W#W").m_126130_("W#W");
    }
 
-   protected static RecipeBuilder m_176684_(ItemLike p_176685_, Ingredient p_176686_) {
+   private static RecipeBuilder m_176684_(ItemLike p_176685_, Ingredient p_176686_) {
       return ShapedRecipeBuilder.m_126116_(p_176685_).m_126127_('#', Items.f_42398_).m_126124_('W', p_176686_).m_126130_("#W#").m_126130_("#W#");
    }
 
-   protected static void m_176690_(Consumer<FinishedRecipe> p_176691_, ItemLike p_176692_, ItemLike p_176693_) {
+   private static void m_176690_(Consumer<FinishedRecipe> p_176691_, ItemLike p_176692_, ItemLike p_176693_) {
       m_176694_(p_176692_, Ingredient.m_43929_(p_176693_)).m_142284_(m_176602_(p_176693_), m_125977_(p_176693_)).m_176498_(p_176691_);
    }
 
-   protected static RecipeBuilder m_176694_(ItemLike p_176695_, Ingredient p_176696_) {
+   private static RecipeBuilder m_176694_(ItemLike p_176695_, Ingredient p_176696_) {
       return ShapedRecipeBuilder.m_126116_(p_176695_).m_126124_('#', p_176696_).m_126130_("##");
    }
 
-   protected static void m_176700_(Consumer<FinishedRecipe> p_176701_, ItemLike p_176702_, ItemLike p_176703_) {
+   private static void m_176700_(Consumer<FinishedRecipe> p_176701_, ItemLike p_176702_, ItemLike p_176703_) {
       m_176704_(p_176702_, Ingredient.m_43929_(p_176703_)).m_142284_(m_176602_(p_176703_), m_125977_(p_176703_)).m_176498_(p_176701_);
    }
 
-   protected static RecipeBuilder m_176704_(ItemLike p_176705_, Ingredient p_176706_) {
+   private static RecipeBuilder m_176704_(ItemLike p_176705_, Ingredient p_176706_) {
       return ShapedRecipeBuilder.m_126118_(p_176705_, 6).m_126124_('#', p_176706_).m_126130_("###");
    }
 
-   protected static RecipeBuilder m_176710_(ItemLike p_176711_, Ingredient p_176712_) {
+   private static RecipeBuilder m_176710_(ItemLike p_176711_, Ingredient p_176712_) {
       return ShapedRecipeBuilder.m_126118_(p_176711_, 4).m_126124_('#', p_176712_).m_126130_("#  ").m_126130_("## ").m_126130_("###");
    }
 
-   protected static RecipeBuilder m_176720_(ItemLike p_176721_, Ingredient p_176722_) {
+   private static RecipeBuilder m_176720_(ItemLike p_176721_, Ingredient p_176722_) {
       return ShapedRecipeBuilder.m_126118_(p_176721_, 2).m_126124_('#', p_176722_).m_126130_("###").m_126130_("###");
    }
 
-   protected static RecipeBuilder m_176726_(ItemLike p_176727_, Ingredient p_176728_) {
+   private static RecipeBuilder m_176726_(ItemLike p_176727_, Ingredient p_176728_) {
       return ShapedRecipeBuilder.m_126118_(p_176727_, 3).m_142409_("sign").m_126124_('#', p_176728_).m_126127_('X', Items.f_42398_).m_126130_("###").m_126130_("###").m_126130_(" X ");
    }
 
-   protected static void m_126061_(Consumer<FinishedRecipe> p_126062_, ItemLike p_126063_, ItemLike p_126064_) {
+   private static void m_126061_(Consumer<FinishedRecipe> p_126062_, ItemLike p_126063_, ItemLike p_126064_) {
       ShapelessRecipeBuilder.m_126189_(p_126063_).m_126209_(p_126064_).m_126209_(Blocks.f_50041_).m_142409_("wool").m_142284_("has_white_wool", m_125977_(Blocks.f_50041_)).m_176498_(p_126062_);
    }
 
-   protected static void m_176716_(Consumer<FinishedRecipe> p_176717_, ItemLike p_176718_, ItemLike p_176719_) {
+   private static void m_176716_(Consumer<FinishedRecipe> p_176717_, ItemLike p_176718_, ItemLike p_176719_) {
       ShapedRecipeBuilder.m_126118_(p_176718_, 3).m_126127_('#', p_176719_).m_126130_("##").m_142409_("carpet").m_142284_(m_176602_(p_176719_), m_125977_(p_176719_)).m_176498_(p_176717_);
    }
 
-   protected static void m_126069_(Consumer<FinishedRecipe> p_126070_, ItemLike p_126071_, ItemLike p_126072_) {
+   private static void m_126069_(Consumer<FinishedRecipe> p_126070_, ItemLike p_126071_, ItemLike p_126072_) {
       ShapedRecipeBuilder.m_126118_(p_126071_, 8).m_126127_('#', Blocks.f_50336_).m_126127_('$', p_126072_).m_126130_("###").m_126130_("#$#").m_126130_("###").m_142409_("carpet").m_142284_("has_white_carpet", m_125977_(Blocks.f_50336_)).m_142284_(m_176602_(p_126072_), m_125977_(p_126072_)).m_176500_(p_126070_, m_176517_(p_126071_, Blocks.f_50336_));
    }
 
-   protected static void m_126073_(Consumer<FinishedRecipe> p_126074_, ItemLike p_126075_, ItemLike p_126076_) {
+   private static void m_126073_(Consumer<FinishedRecipe> p_126074_, ItemLike p_126075_, ItemLike p_126076_) {
       ShapedRecipeBuilder.m_126116_(p_126075_).m_126127_('#', p_126076_).m_206416_('X', ItemTags.f_13168_).m_126130_("###").m_126130_("XXX").m_142409_("bed").m_142284_(m_176602_(p_126076_), m_125977_(p_126076_)).m_176498_(p_126074_);
    }
 
-   protected static void m_126077_(Consumer<FinishedRecipe> p_126078_, ItemLike p_126079_, ItemLike p_126080_) {
+   private static void m_126077_(Consumer<FinishedRecipe> p_126078_, ItemLike p_126079_, ItemLike p_126080_) {
       ShapelessRecipeBuilder.m_126189_(p_126079_).m_126209_(Items.f_42503_).m_126209_(p_126080_).m_142409_("dyed_bed").m_142284_("has_bed", m_125977_(Items.f_42503_)).m_176500_(p_126078_, m_176517_(p_126079_, Items.f_42503_));
    }
 
-   protected static void m_126081_(Consumer<FinishedRecipe> p_126082_, ItemLike p_126083_, ItemLike p_126084_) {
+   private static void m_126081_(Consumer<FinishedRecipe> p_126082_, ItemLike p_126083_, ItemLike p_126084_) {
       ShapedRecipeBuilder.m_126116_(p_126083_).m_126127_('#', p_126084_).m_126127_('|', Items.f_42398_).m_126130_("###").m_126130_("###").m_126130_(" | ").m_142409_("banner").m_142284_(m_176602_(p_126084_), m_125977_(p_126084_)).m_176498_(p_126082_);
    }
 
-   protected static void m_126085_(Consumer<FinishedRecipe> p_126086_, ItemLike p_126087_, ItemLike p_126088_) {
+   private static void m_126085_(Consumer<FinishedRecipe> p_126086_, ItemLike p_126087_, ItemLike p_126088_) {
       ShapedRecipeBuilder.m_126118_(p_126087_, 8).m_126127_('#', Blocks.f_50058_).m_126127_('X', p_126088_).m_126130_("###").m_126130_("#X#").m_126130_("###").m_142409_("stained_glass").m_142284_("has_glass", m_125977_(Blocks.f_50058_)).m_176498_(p_126086_);
    }
 
-   protected static void m_126089_(Consumer<FinishedRecipe> p_126090_, ItemLike p_126091_, ItemLike p_126092_) {
+   private static void m_126089_(Consumer<FinishedRecipe> p_126090_, ItemLike p_126091_, ItemLike p_126092_) {
       ShapedRecipeBuilder.m_126118_(p_126091_, 16).m_126127_('#', p_126092_).m_126130_("###").m_126130_("###").m_142409_("stained_glass_pane").m_142284_("has_glass", m_125977_(p_126092_)).m_176498_(p_126090_);
    }
 
-   protected static void m_126093_(Consumer<FinishedRecipe> p_126094_, ItemLike p_126095_, ItemLike p_126096_) {
+   private static void m_126093_(Consumer<FinishedRecipe> p_126094_, ItemLike p_126095_, ItemLike p_126096_) {
       ShapedRecipeBuilder.m_126118_(p_126095_, 8).m_126127_('#', Blocks.f_50185_).m_126127_('$', p_126096_).m_126130_("###").m_126130_("#$#").m_126130_("###").m_142409_("stained_glass_pane").m_142284_("has_glass_pane", m_125977_(Blocks.f_50185_)).m_142284_(m_176602_(p_126096_), m_125977_(p_126096_)).m_176500_(p_126094_, m_176517_(p_126095_, Blocks.f_50185_));
    }
 
-   protected static void m_126097_(Consumer<FinishedRecipe> p_126098_, ItemLike p_126099_, ItemLike p_126100_) {
+   private static void m_126097_(Consumer<FinishedRecipe> p_126098_, ItemLike p_126099_, ItemLike p_126100_) {
       ShapedRecipeBuilder.m_126118_(p_126099_, 8).m_126127_('#', Blocks.f_50352_).m_126127_('X', p_126100_).m_126130_("###").m_126130_("#X#").m_126130_("###").m_142409_("stained_terracotta").m_142284_("has_terracotta", m_125977_(Blocks.f_50352_)).m_176498_(p_126098_);
    }
 
-   protected static void m_126101_(Consumer<FinishedRecipe> p_126102_, ItemLike p_126103_, ItemLike p_126104_) {
+   private static void m_126101_(Consumer<FinishedRecipe> p_126102_, ItemLike p_126103_, ItemLike p_126104_) {
       ShapelessRecipeBuilder.m_126191_(p_126103_, 8).m_126209_(p_126104_).m_126211_(Blocks.f_49992_, 4).m_126211_(Blocks.f_49994_, 4).m_142409_("concrete_powder").m_142284_("has_sand", m_125977_(Blocks.f_49992_)).m_142284_("has_gravel", m_125977_(Blocks.f_49994_)).m_176498_(p_126102_);
    }
 
@@ -1150,36 +_,36 @@
       return ShapedRecipeBuilder.m_126116_(p_176647_).m_126124_('#', p_176648_).m_126130_("#").m_126130_("#");
    }
 
-   protected static void m_176735_(Consumer<FinishedRecipe> p_176736_, ItemLike p_176737_, ItemLike p_176738_) {
+   private static void m_176735_(Consumer<FinishedRecipe> p_176736_, ItemLike p_176737_, ItemLike p_176738_) {
       m_176546_(p_176736_, p_176737_, p_176738_, 1);
    }
 
-   protected static void m_176546_(Consumer<FinishedRecipe> p_176547_, ItemLike p_176548_, ItemLike p_176549_, int p_176550_) {
+   private static void m_176546_(Consumer<FinishedRecipe> p_176547_, ItemLike p_176548_, ItemLike p_176549_, int p_176550_) {
       SingleItemRecipeBuilder.m_126316_(Ingredient.m_43929_(p_176549_), p_176548_, p_176550_).m_142284_(m_176602_(p_176549_), m_125977_(p_176549_)).m_176500_(p_176547_, m_176517_(p_176548_, p_176549_) + "_stonecutting");
    }
 
-   protected static void m_176739_(Consumer<FinishedRecipe> p_176740_, ItemLike p_176741_, ItemLike p_176742_) {
+   private static void m_176739_(Consumer<FinishedRecipe> p_176740_, ItemLike p_176741_, ItemLike p_176742_) {
       SimpleCookingRecipeBuilder.m_126272_(Ingredient.m_43929_(p_176742_), p_176741_, 0.1F, 200).m_142284_(m_176602_(p_176742_), m_125977_(p_176742_)).m_176498_(p_176740_);
    }
 
-   protected static void m_176743_(Consumer<FinishedRecipe> p_176744_, ItemLike p_176745_, ItemLike p_176746_) {
+   private static void m_176743_(Consumer<FinishedRecipe> p_176744_, ItemLike p_176745_, ItemLike p_176746_) {
       m_176568_(p_176744_, p_176745_, p_176746_, m_176644_(p_176746_), (String)null, m_176644_(p_176745_), (String)null);
    }
 
-   protected static void m_176562_(Consumer<FinishedRecipe> p_176563_, ItemLike p_176564_, ItemLike p_176565_, String p_176566_, String p_176567_) {
+   private static void m_176562_(Consumer<FinishedRecipe> p_176563_, ItemLike p_176564_, ItemLike p_176565_, String p_176566_, String p_176567_) {
       m_176568_(p_176563_, p_176564_, p_176565_, p_176566_, p_176567_, m_176644_(p_176564_), (String)null);
    }
 
-   protected static void m_176616_(Consumer<FinishedRecipe> p_176617_, ItemLike p_176618_, ItemLike p_176619_, String p_176620_, String p_176621_) {
+   private static void m_176616_(Consumer<FinishedRecipe> p_176617_, ItemLike p_176618_, ItemLike p_176619_, String p_176620_, String p_176621_) {
       m_176568_(p_176617_, p_176618_, p_176619_, m_176644_(p_176619_), (String)null, p_176620_, p_176621_);
    }
 
-   protected static void m_176568_(Consumer<FinishedRecipe> p_176569_, ItemLike p_176570_, ItemLike p_176571_, String p_176572_, @Nullable String p_176573_, String p_176574_, @Nullable String p_176575_) {
+   private static void m_176568_(Consumer<FinishedRecipe> p_176569_, ItemLike p_176570_, ItemLike p_176571_, String p_176572_, @Nullable String p_176573_, String p_176574_, @Nullable String p_176575_) {
       ShapelessRecipeBuilder.m_126191_(p_176570_, 9).m_126209_(p_176571_).m_142409_(p_176575_).m_142284_(m_176602_(p_176571_), m_125977_(p_176571_)).m_142700_(p_176569_, new ResourceLocation(p_176574_));
       ShapedRecipeBuilder.m_126116_(p_176571_).m_126127_('#', p_176570_).m_126130_("###").m_126130_("###").m_126130_("###").m_142409_(p_176573_).m_142284_(m_176602_(p_176570_), m_125977_(p_176570_)).m_142700_(p_176569_, new ResourceLocation(p_176572_));
    }
 
-   protected static void m_126006_(Consumer<FinishedRecipe> p_126007_, String p_126008_, SimpleCookingSerializer<?> p_126009_, int p_126010_) {
+   private static void m_126006_(Consumer<FinishedRecipe> p_126007_, String p_126008_, SimpleCookingSerializer<?> p_126009_, int p_126010_) {
       m_176583_(p_126007_, p_126008_, p_126009_, p_126010_, Items.f_42579_, Items.f_42580_, 0.35F);
       m_176583_(p_126007_, p_126008_, p_126009_, p_126010_, Items.f_42581_, Items.f_42582_, 0.35F);
       m_176583_(p_126007_, p_126008_, p_126009_, p_126010_, Items.f_42526_, Items.f_42530_, 0.35F);
@@ -1191,17 +_,17 @@
       m_176583_(p_126007_, p_126008_, p_126009_, p_126010_, Items.f_42697_, Items.f_42698_, 0.35F);
    }
 
-   protected static void m_176583_(Consumer<FinishedRecipe> p_176584_, String p_176585_, SimpleCookingSerializer<?> p_176586_, int p_176587_, ItemLike p_176588_, ItemLike p_176589_, float p_176590_) {
+   private static void m_176583_(Consumer<FinishedRecipe> p_176584_, String p_176585_, SimpleCookingSerializer<?> p_176586_, int p_176587_, ItemLike p_176588_, ItemLike p_176589_, float p_176590_) {
       SimpleCookingRecipeBuilder.m_126248_(Ingredient.m_43929_(p_176588_), p_176589_, p_176590_, p_176587_, p_176586_).m_142284_(m_176602_(p_176588_), m_125977_(p_176588_)).m_176500_(p_176584_, m_176632_(p_176589_) + "_from_" + p_176585_);
    }
 
-   protected static void m_176610_(Consumer<FinishedRecipe> p_176611_) {
+   private static void m_176610_(Consumer<FinishedRecipe> p_176611_) {
       HoneycombItem.f_150863_.get().forEach((p_176578_, p_176579_) -> {
          ShapelessRecipeBuilder.m_126189_(p_176579_).m_126209_(p_176578_).m_126209_(Items.f_42784_).m_142409_(m_176632_(p_176579_)).m_142284_(m_176602_(p_176578_), m_125977_(p_176578_)).m_176500_(p_176611_, m_176517_(p_176579_, Items.f_42784_));
       });
    }
 
-   protected static void m_176580_(Consumer<FinishedRecipe> p_176581_, BlockFamily p_176582_) {
+   private static void m_176580_(Consumer<FinishedRecipe> p_176581_, BlockFamily p_176582_) {
       p_176582_.m_175954_().forEach((p_176529_, p_176530_) -> {
          BiFunction<ItemLike, ItemLike, RecipeBuilder> bifunction = f_176513_.get(p_176529_);
          ItemLike itemlike = m_176523_(p_176582_, p_176529_);
@@ -1223,7 +_,7 @@
       });
    }
 
-   protected static Block m_176523_(BlockFamily p_176524_, BlockFamily.Variant p_176525_) {
+   private static Block m_176523_(BlockFamily p_176524_, BlockFamily.Variant p_176525_) {
       if (p_176525_ == BlockFamily.Variant.CHISELED) {
          if (!p_176524_.m_175954_().containsKey(BlockFamily.Variant.SLAB)) {
             throw new IllegalStateException("Slab is not defined for the family.");
@@ -1239,7 +_,7 @@
       return new EnterBlockTrigger.TriggerInstance(EntityPredicate.Composite.f_36667_, p_125980_, StatePropertiesPredicate.f_67658_);
    }
 
-   protected static InventoryChangeTrigger.TriggerInstance m_176520_(MinMaxBounds.Ints p_176521_, ItemLike p_176522_) {
+   private static InventoryChangeTrigger.TriggerInstance m_176520_(MinMaxBounds.Ints p_176521_, ItemLike p_176522_) {
       return m_126011_(ItemPredicate.Builder.m_45068_().m_151445_(p_176522_).m_151443_(p_176521_).m_45077_());
    }
 
@@ -1247,7 +_,7 @@
       return m_126011_(ItemPredicate.Builder.m_45068_().m_151445_(p_125978_).m_45077_());
    }
 
-   protected static InventoryChangeTrigger.TriggerInstance m_206406_(TagKey<Item> p_206407_) {
+   private static InventoryChangeTrigger.TriggerInstance m_206406_(TagKey<Item> p_206407_) {
       return m_126011_(ItemPredicate.Builder.m_45068_().m_204145_(p_206407_).m_45077_());
    }
 
@@ -1255,27 +_,27 @@
       return new InventoryChangeTrigger.TriggerInstance(EntityPredicate.Composite.f_36667_, MinMaxBounds.Ints.f_55364_, MinMaxBounds.Ints.f_55364_, MinMaxBounds.Ints.f_55364_, p_126012_);
    }
 
-   protected static String m_176602_(ItemLike p_176603_) {
+   private static String m_176602_(ItemLike p_176603_) {
       return "has_" + m_176632_(p_176603_);
    }
 
-   protected static String m_176632_(ItemLike p_176633_) {
+   private static String m_176632_(ItemLike p_176633_) {
       return Registry.f_122827_.m_7981_(p_176633_.m_5456_()).m_135815_();
    }
 
-   protected static String m_176644_(ItemLike p_176645_) {
+   private static String m_176644_(ItemLike p_176645_) {
       return m_176632_(p_176645_);
    }
 
-   protected static String m_176517_(ItemLike p_176518_, ItemLike p_176519_) {
+   private static String m_176517_(ItemLike p_176518_, ItemLike p_176519_) {
       return m_176632_(p_176518_) + "_from_" + m_176632_(p_176519_);
    }
 
-   protected static String m_176656_(ItemLike p_176657_) {
+   private static String m_176656_(ItemLike p_176657_) {
       return m_176632_(p_176657_) + "_from_smelting";
    }
 
-   protected static String m_176668_(ItemLike p_176669_) {
+   private static String m_176668_(ItemLike p_176669_) {
       return m_176632_(p_176669_) + "_from_blasting";
    }
 
