--- a/net/minecraft/command/impl/EffectCommand.java
+++ b/net/minecraft/command/impl/EffectCommand.java
@@ -17,6 +_,7 @@
 import net.minecraft.potion.Effect;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.util.text.TranslationTextComponent;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class EffectCommand {
    private static final SimpleCommandExceptionType field_198361_a = new SimpleCommandExceptionType(new TranslationTextComponent("commands.effect.give.failed"));
@@ -61,7 +_,7 @@
       for(Entity entity : p_198360_1_) {
          if (entity instanceof LivingEntity) {
             EffectInstance effectinstance = new EffectInstance(p_198360_2_, j, p_198360_4_, false, p_198360_5_);
-            if (((LivingEntity)entity).func_195064_c(effectinstance)) {
+            if (((LivingEntity)entity).addEffect(effectinstance, Cause.COMMAND)) { // CraftBukkit
                ++i;
             }
          }
@@ -84,7 +_,7 @@
       int i = 0;
 
       for(Entity entity : p_198354_1_) {
-         if (entity instanceof LivingEntity && ((LivingEntity)entity).func_195061_cb()) {
+         if (entity instanceof LivingEntity && ((LivingEntity)entity).clearActivePotions(Cause.COMMAND)) { // CraftBukkit
             ++i;
          }
       }
@@ -106,7 +_,7 @@
       int i = 0;
 
       for(Entity entity : p_198355_1_) {
-         if (entity instanceof LivingEntity && ((LivingEntity)entity).func_195063_d(p_198355_2_)) {
+         if (entity instanceof LivingEntity && ((LivingEntity)entity).removeEffect(p_198355_2_, Cause.COMMAND)) { // CraftBukkit
             ++i;
          }
       }
