--- a/net/minecraft/command/impl/ListCommand.java
+++ b/net/minecraft/command/impl/ListCommand.java
@@ -11,6 +_,7 @@
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextComponentUtils;
 import net.minecraft.util.text.TranslationTextComponent;
+import org.bukkit.entity.Player;
 
 public class ListCommand {
    public static void func_198522_a(CommandDispatcher<CommandSource> p_198522_0_) {
@@ -34,6 +_,14 @@
    private static int func_208200_a(CommandSource p_208200_0_, Function<ServerPlayerEntity, ITextComponent> p_208200_1_) {
       PlayerList playerlist = p_208200_0_.func_197028_i().func_184103_al();
       List<ServerPlayerEntity> list = playerlist.func_181057_v();
+      // CraftBukkit Start
+      if (p_208200_0_.getBukkitSender() instanceof org.bukkit.entity.Player)
+      {
+         org.bukkit.entity.Player sender = (org.bukkit.entity.Player)p_208200_0_.getBukkitSender();
+         list = list.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList());
+      }
+      // CraftBukkit end
+
       ITextComponent itextcomponent = TextComponentUtils.func_240649_b_(list, p_208200_1_);
       p_208200_0_.func_197030_a(new TranslationTextComponent("commands.list.players", list.size(), playerlist.func_72352_l(), itextcomponent), false);
       return list.size();
