--- a/net/minecraft/command/impl/SpreadPlayersCommand.java
+++ b/net/minecraft/command/impl/SpreadPlayersCommand.java
@@ -28,6 +_,8 @@
 import net.minecraft.util.math.vector.Vector2f;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.IBlockReader;
+import net.minecraft.world.World;
+import net.minecraft.world.server.ServerChunkProvider;
 import net.minecraft.world.server.ServerWorld;
 
 public class SpreadPlayersCommand {
@@ -162,7 +_,8 @@
             spreadplayerscommand$position = p_241072_2_[i++];
          }
 
-         entity.func_223102_j((double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198714_b) + 0.5D);
+         net.minecraftforge.event.entity.living.EntityTeleportEvent.SpreadPlayersCommand event = net.minecraftforge.event.ForgeEventFactory.onEntityTeleportSpreadPlayersCommand(entity, (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198714_b) + 0.5D);
+         if (!event.isCanceled()) entity.func_223102_j(event.getTargetX(), event.getTargetY(), event.getTargetZ());
          double d2 = Double.MAX_VALUE;
 
          for(SpreadPlayersCommand.Position spreadplayerscommand$position1 : p_241072_2_) {
@@ -238,7 +_,7 @@
 
       public int func_198710_a(IBlockReader p_198710_1_, int p_198710_2_) {
          BlockPos.Mutable blockpos$mutable = new BlockPos.Mutable(this.field_198713_a, (double)(p_198710_2_ + 1), this.field_198714_b);
-         boolean flag = p_198710_1_.func_180495_p(blockpos$mutable).func_196958_f();
+         boolean flag = getType(p_198710_1_, blockpos$mutable).func_196958_f(); // CraftBukkit
          blockpos$mutable.func_189536_c(Direction.DOWN);
 
          boolean flag2;
@@ -257,7 +_,7 @@
 
       public boolean func_241074_b_(IBlockReader p_241074_1_, int p_241074_2_) {
          BlockPos blockpos = new BlockPos(this.field_198713_a, (double)(this.func_198710_a(p_241074_1_, p_241074_2_) - 1), this.field_198714_b);
-         BlockState blockstate = p_241074_1_.func_180495_p(blockpos);
+         BlockState blockstate = getType(p_241074_1_, blockpos); // CraftBukkit
          Material material = blockstate.func_185904_a();
          return blockpos.func_177956_o() < p_241074_2_ && !material.func_76224_d() && material != Material.field_151581_o;
       }
@@ -266,5 +_,12 @@
          this.field_198713_a = MathHelper.func_82716_a(p_198711_1_, p_198711_2_, p_198711_6_);
          this.field_198714_b = MathHelper.func_82716_a(p_198711_1_, p_198711_4_, p_198711_8_);
       }
+
+      // CraftBukkit start - add a version of getType which force loads chunks
+      private static BlockState getType(IBlockReader iblockaccess, BlockPos position) {
+         ((ServerChunkProvider) ((World) iblockaccess).func_72863_F()).func_217205_a(position.func_177958_n() >> 4, position.func_177952_p() >> 4, true);
+         return iblockaccess.func_180495_p(position);
+      }
+      // CraftBukkit end
    }
 }
