--- a/net/minecraft/command/impl/SummonCommand.java
+++ b/net/minecraft/command/impl/SummonCommand.java
@@ -21,6 +_,7 @@
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class SummonCommand {
    private static final SimpleCommandExceptionType field_198741_a = new SimpleCommandExceptionType(new TranslationTextComponent("commands.summon.failed"));
@@ -58,7 +_,7 @@
                ((MobEntity)entity).func_213386_a(p_198737_0_.func_197023_e(), p_198737_0_.func_197023_e().func_175649_E(entity.func_233580_cy_()), SpawnReason.COMMAND, (ILivingEntityData)null, (CompoundNBT)null);
             }
 
-            if (!serverworld.func_242106_g(entity)) {
+            if (!serverworld.addAllEntitiesSafely(entity, CreatureSpawnEvent.SpawnReason.COMMAND)) { // CraftBukkit - pass a spawn reason of "COMMAND"
                throw field_244378_b.create();
             } else {
                p_198737_0_.func_197030_a(new TranslationTextComponent("commands.summon.success", entity.func_145748_c_()), true);
