--- a/net/minecraft/command/impl/TimeCommand.java
+++ b/net/minecraft/command/impl/TimeCommand.java
@@ -7,6 +_,8 @@
 import net.minecraft.command.arguments.TimeArgument;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.Bukkit;
+import org.bukkit.event.world.TimeSkipEvent;
 
 public class TimeCommand {
    public static void func_198823_a(CommandDispatcher<CommandSource> p_198823_0_) {
@@ -44,7 +_,13 @@
 
    public static int func_198829_a(CommandSource p_198829_0_, int p_198829_1_) {
       for(ServerWorld serverworld : p_198829_0_.func_197028_i().func_212370_w()) {
-         serverworld.func_241114_a_((long)p_198829_1_);
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverworld.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_198829_1_ - serverworld.func_72820_D());
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverworld.func_241114_a_((long) serverworld.func_72820_D() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       p_198829_0_.func_197030_a(new TranslationTextComponent("commands.time.set", p_198829_1_), true);
@@ -53,7 +_,13 @@
 
    public static int func_198826_b(CommandSource p_198826_0_, int p_198826_1_) {
       for(ServerWorld serverworld : p_198826_0_.func_197028_i().func_212370_w()) {
-         serverworld.func_241114_a_(serverworld.func_72820_D() + (long)p_198826_1_);
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverworld.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_198826_1_);
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverworld.func_241114_a_(serverworld.func_72820_D() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       int i = func_198833_a(p_198826_0_.func_197023_e());
