--- ../src-base/minecraft/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
@@ -32,7 +32,7 @@
 
         if (entity instanceof EntityLivingBase)
         {
-            int i = EnchantmentHelper.getLootingModifier((EntityLivingBase)entity);
+            int i = context.getLootingModifier();
 
             if (i == 0)
             {
