--- ../src-base/minecraft/net/minecraft/world/storage/loot/functions/SetAttributes.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/functions/SetAttributes.java
@@ -24,9 +24,9 @@
 public class SetAttributes extends LootFunction
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private final SetAttributes.Modifier[] modifiers;
+    private final Modifier[] modifiers;
 
-    public SetAttributes(LootCondition[] conditionsIn, SetAttributes.Modifier[] modifiersIn)
+    public SetAttributes(LootCondition[] conditionsIn, Modifier[] modifiersIn)
     {
         super(conditionsIn);
         this.modifiers = modifiersIn;
@@ -34,7 +34,7 @@
 
     public ItemStack apply(ItemStack stack, Random rand, LootContext context)
     {
-        for (SetAttributes.Modifier setattributes$modifier : this.modifiers)
+        for (Modifier setattributes$modifier : this.modifiers)
         {
             UUID uuid = setattributes$modifier.uuid;
 
@@ -102,7 +102,7 @@
                 return jsonobject;
             }
 
-            public static SetAttributes.Modifier deserialize(JsonObject jsonObj, JsonDeserializationContext context)
+            public static Modifier deserialize(JsonObject jsonObj, JsonDeserializationContext context)
             {
                 String s = JsonUtils.getString(jsonObj, "name");
                 String s1 = JsonUtils.getString(jsonObj, "attribute");
@@ -151,7 +151,7 @@
                     }
                 }
 
-                return new SetAttributes.Modifier(s, s1, i, randomvaluerange, aentityequipmentslot, uuid);
+                return new Modifier(s, s1, i, randomvaluerange, aentityequipmentslot, uuid);
             }
 
             private static String getOperationFromStr(int operationIn)
@@ -201,7 +201,7 @@
             {
                 JsonArray jsonarray = new JsonArray();
 
-                for (SetAttributes.Modifier setattributes$modifier : functionClazz.modifiers)
+                for (Modifier setattributes$modifier : functionClazz.modifiers)
                 {
                     jsonarray.add(setattributes$modifier.serialize(serializationContext));
                 }
@@ -212,12 +212,12 @@
             public SetAttributes deserialize(JsonObject object, JsonDeserializationContext deserializationContext, LootCondition[] conditionsIn)
             {
                 JsonArray jsonarray = JsonUtils.getJsonArray(object, "modifiers");
-                SetAttributes.Modifier[] asetattributes$modifier = new SetAttributes.Modifier[jsonarray.size()];
+                Modifier[] asetattributes$modifier = new Modifier[jsonarray.size()];
                 int i = 0;
 
                 for (JsonElement jsonelement : jsonarray)
                 {
-                    asetattributes$modifier[i++] = SetAttributes.Modifier.deserialize(JsonUtils.getJsonObject(jsonelement, "modifier"), deserializationContext);
+                    asetattributes$modifier[i++] = Modifier.deserialize(JsonUtils.getJsonObject(jsonelement, "modifier"), deserializationContext);
                 }
 
                 if (asetattributes$modifier.length == 0)
