--- a/net/minecraft/world/level/levelgen/FlatLevelSource.java
+++ b/net/minecraft/world/level/levelgen/FlatLevelSource.java
@@ -37,6 +_,13 @@
       this.f_64165_ = p_256337_;
    }
 
+   // CraftBukkit start - 1.20.1
+   public FlatLevelSource(FlatLevelGeneratorSettings p_256337_, net.minecraft.world.level.biome.BiomeSource worldchunkmanager) {
+      super(worldchunkmanager, Util.m_143827_(p_256337_::m_226294_));
+      this.f_64165_ = p_256337_;
+   }
+   // CraftBukkit end
+
    public ChunkGeneratorStructureState m_255169_(HolderLookup<StructureSet> p_256602_, RandomState p_255830_, long p_256355_) {
       Stream<Holder<StructureSet>> stream = this.f_64165_.m_209810_().map(HolderSet::m_203614_).orElseGet(() -> {
          return p_256602_.m_214062_().map((p_255579_) -> {
