--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -64,9 +_,9 @@
    private static final BlockState f_64321_ = Blocks.f_50016_.m_49966_();
    private static final BlockState[] f_158373_ = new BlockState[0];
    protected final BlockState f_64316_;
-   private final Registry<NormalNoise.NoiseParameters> f_188604_;
+   public final Registry<NormalNoise.NoiseParameters> f_188604_;
    protected final long f_64333_;
-   protected final Holder<NoiseGeneratorSettings> f_64318_;
+   public final Holder<NoiseGeneratorSettings> f_64318_;
    private final NoiseRouter f_209104_;
    protected final Climate.Sampler f_158382_;
    private final SurfaceSystem f_188605_;
