--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.food;
 
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.protocol.game.ClientboundSetHealthPacket;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.player.Player;
@@ -9,25 +_,49 @@
 import net.minecraft.world.level.GameRules;
 
 public class FoodData {
-   private int f_38696_ = 20;
-   private float f_38697_;
-   private float f_38698_;
+   public int f_38696_ = 20;
+   public float f_38697_;
+   public float f_38698_;
    private int f_38699_;
    private int f_38700_ = 20;
-
-   public FoodData() {
+   // CraftBukkit start
+   private Player entityhuman;
+   public int saturatedRegenRate = 10;
+   public int unsaturatedRegenRate = 80;
+   public int starvationRate = 80;
+   // CraftBukkit end
+
+   public FoodData() { throw new AssertionError("Whoopsie, we missed the bukkit."); } // CraftBukkit start - throw an error
+
+   // CraftBukkit start - added Player constructor
+   public FoodData(Player player) {
+      this.entityhuman = player;
       this.f_38697_ = 5.0F;
    }
+   // CraftBukkit end
 
    public void m_38707_(int p_38708_, float p_38709_) {
       this.f_38696_ = Math.min(p_38708_ + this.f_38696_, 20);
       this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
    }
 
+   // Use the LivingEntity sensitive version in favour of this.
+   @Deprecated
    public void m_38712_(Item p_38713_, ItemStack p_38714_) {
+      this.eat(p_38713_, p_38714_, null);
+   }
+
+   public void eat(Item p_38713_, ItemStack p_38714_, @javax.annotation.Nullable net.minecraft.world.entity.LivingEntity entity) {
       if (p_38713_.m_41472_()) {
-         FoodProperties foodproperties = p_38713_.m_41473_();
-         this.m_38707_(foodproperties.m_38744_(), foodproperties.m_38745_());
+         FoodProperties foodproperties = p_38714_.getFoodProperties(entity);
+         // CraftBukkit start
+         int oldFoodLevel = f_38696_;
+         org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, foodproperties.m_38744_() + oldFoodLevel, p_38714_);
+         if (!event.isCancelled()) {
+            this.m_38707_(event.getFoodLevel() - oldFoodLevel, foodproperties.m_38745_());
+         }
+         ((ServerPlayer) entityhuman).getBukkitEntity().sendHealthUpdate();
+         // CraftBukkit end
       }
 
    }
@@ -40,29 +_,37 @@
          if (this.f_38697_ > 0.0F) {
             this.f_38697_ = Math.max(this.f_38697_ - 1.0F, 0.0F);
          } else if (difficulty != Difficulty.PEACEFUL) {
-            this.f_38696_ = Math.max(this.f_38696_ - 1, 0);
+            // CraftBukkit start
+            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, Math.max(this.f_38696_ - 1, 0));
+            if (!event.isCancelled()) {
+               this.f_38696_ = event.getFoodLevel();
+            }
+            ((ServerPlayer) entityhuman).f_8906_.m_141995_(new ClientboundSetHealthPacket(((ServerPlayer) entityhuman).getBukkitEntity().getScaledHealth(), this.f_38696_, this.f_38697_));
+            // CraftBukkit end
          }
       }
 
       boolean flag = p_38711_.f_19853_.m_46469_().m_46207_(GameRules.f_46139_);
       if (flag && this.f_38697_ > 0.0F && p_38711_.m_36325_() && this.f_38696_ >= 20) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 10) {
+         if (this.f_38699_ >= this.saturatedRegenRate) { // CraftBukkit
             float f = Math.min(this.f_38697_, 6.0F);
-            p_38711_.m_5634_(f / 6.0F);
-            this.m_38703_(f);
+            p_38711_.heal(f / 6.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.SATIATED); // CraftBukkit - added RegainReason
+            //this.addExhaustion(f);
+            p_38711_.causeFoodExhaustion(f, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); // CraftBukkit - EntityExhaustionEvent
             this.f_38699_ = 0;
          }
       } else if (flag && this.f_38696_ >= 18 && p_38711_.m_36325_()) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 80) {
-            p_38711_.m_5634_(1.0F);
-            this.m_38703_(6.0F);
+         if (this.f_38699_ >= this.unsaturatedRegenRate) { // CraftBukkit - add regen rate manipulation
+            p_38711_.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.SATIATED); // CraftBukkit - added RegainReason
+            // this.addExhaustion(6.0F);
+            p_38711_.causeFoodExhaustion(6.0f, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); // CraftBukkit - EntityExhaustionEvent
             this.f_38699_ = 0;
          }
       } else if (this.f_38696_ <= 0) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 80) {
+         if (this.f_38699_ >= this.starvationRate) { // CraftBukkit - add regen rate manipulation
             if (p_38711_.m_21223_() > 10.0F || difficulty == Difficulty.HARD || p_38711_.m_21223_() > 1.0F && difficulty == Difficulty.NORMAL) {
                p_38711_.m_6469_(DamageSource.f_19313_, 1.0F);
             }
