--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -79,14 +_,23 @@
 
    public Sniffer(EntityType<? extends Animal> p_273717_, Level p_273562_) {
       super(p_273717_, p_273562_);
-      this.f_19804_.m_135372_(f_271416_, Sniffer.State.IDLING);
-      this.f_19804_.m_135372_(f_271498_, 0);
+      // this.entityData.define(DATA_STATE, Sniffer.State.IDLING); // CraftBukkit - 1.20.1 moved down to appropriate location
+      // this.entityData.define(DATA_DROP_SEED_AT_TICK, 0); // CraftBukkit - 1.20.1 moved down to appropriate location
       this.m_21573_().m_7008_(true);
       this.m_21441_(BlockPathTypes.WATER, -1.0F);
       this.m_21441_(BlockPathTypes.DANGER_POWDER_SNOW, -1.0F);
       this.m_21441_(BlockPathTypes.DAMAGE_CAUTIOUS, -1.0F);
    }
 
+   // CraftBukkit start - 1.20.1 SPIGOT-7295: moved from constructor to appropriate location
+   @Override
+   protected void m_8097_() {
+      super.m_8097_();
+      this.f_19804_.m_135372_(Sniffer.f_271416_, Sniffer.State.IDLING);
+      this.f_19804_.m_135372_(Sniffer.f_271498_, 0);
+   }
+   // CraftBukkit end
+
    protected float m_6431_(Pose p_272721_, EntityDimensions p_273353_) {
       return this.m_6972_(p_272721_).f_20378_ * 0.6F;
    }
@@ -136,7 +_,7 @@
       return this.m_20182_().m_82549_(this.m_20156_().m_82490_(2.25D));
    }
 
-   private Sniffer.State m_271917_() {
+   public Sniffer.State m_271917_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_271416_);
    }
 
@@ -229,7 +_,7 @@
       return this;
    }
 
-   Optional<BlockPos> m_271905_() {
+   public Optional<BlockPos> m_271905_() { // CatServer - public
       return IntStream.range(0, 5).mapToObj((p_273771_) -> {
          return LandRandomPos.m_148488_(this, 10 + 2 * p_273771_, 3);
       }).filter(Objects::nonNull).map(BlockPos::m_274446_).filter((p_289451_) -> {
@@ -237,7 +_,7 @@
       }).map(BlockPos::m_7495_).filter(this::m_271898_).findFirst();
    }
 
-   boolean m_272270_() {
+   public boolean m_272270_() { // CatServer - public
       return !this.m_272223_() && !this.m_278663_() && !this.m_6162_() && !this.m_20069_() && this.m_20096_() && !this.m_20159_() && this.m_271898_(this.m_284345_().m_7495_());
    }
 
@@ -257,6 +_,13 @@
 
          for(ItemStack itemstack : list) {
             ItemEntity itementity = new ItemEntity(serverlevel, (double)blockpos.m_123341_(), (double)blockpos.m_123342_(), (double)blockpos.m_123343_(), itemstack);
+            // CraftBukkit start - 1.20.1 handle EntityDropItemEvent
+            org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
             itementity.m_32060_();
             serverlevel.m_7967_(itementity);
          }
@@ -277,7 +_,7 @@
             }
 
             if (this.f_19797_ % 10 == 0) {
-               this.m_9236_().m_7785_(this.m_20185_(), this.m_20186_(), this.m_20189_(), blockstate.m_60827_().m_56778_(), this.m_5720_(), 0.5F, 0.5F, false);
+               this.m_9236_().m_7785_(this.m_20185_(), this.m_20186_(), this.m_20189_(), blockstate.getSoundType(m_9236_(), blockpos.m_7495_(), this).m_56778_(), this.m_5720_(), 0.5F, 0.5F, false);
             }
          }
       }
@@ -289,14 +_,14 @@
       return this;
    }
 
-   private Sniffer m_271874_(BlockPos p_273015_) {
+   public Sniffer m_271874_(BlockPos p_273015_) { // CatServer - private -> public
       List<GlobalPos> list = this.m_272217_().limit(20L).collect(Collectors.toList());
       list.add(0, GlobalPos.m_122643_(this.m_9236_().m_46472_(), p_273015_));
       this.m_6274_().m_21879_(MemoryModuleType.f_271415_, list);
       return this;
    }
 
-   private Stream<GlobalPos> m_272217_() {
+   public Stream<GlobalPos> m_272217_() { // CatServer - private -> public
       return this.m_6274_().m_21952_(MemoryModuleType.f_271415_).stream().flatMap(Collection::stream);
    }
 
