--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -84,6 +_,12 @@
       this.f_19804_.m_135381_(f_30745_, Math.max(1, Math.min(5, p_30841_)));
    }
 
+   // CraftBukkit start
+   public void setStrengthPublic(int i) {
+      this.m_30840_(i);
+   }
+   // CraftBukkit end
+
    private void m_218817_(RandomSource p_218818_) {
       int i = p_218818_.m_188501_() < 0.04F ? 5 : 3;
       this.m_30840_(1 + p_218818_.m_188503_(i));
