--- a/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
+++ b/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
@@ -28,6 +_,10 @@
 
    public void m_8037_() {
       ServerLevel serverlevel = (ServerLevel)this.f_30925_.m_9236_();
+      // Forge: Trigger the trap in a tick task to avoid crashes when mods add goals to skeleton horses
+      // (MC-206338/Forge PR #7509)
+      serverlevel.m_7654_().m_6937_(new net.minecraft.server.TickTask(serverlevel.m_7654_().m_129921_(), () -> {
+      if (!this.f_30925_.m_6084_()) return;
       DifficultyInstance difficultyinstance = serverlevel.m_6436_(this.f_30925_.m_20183_());
       this.f_30925_.m_30923_(false);
       this.f_30925_.m_30651_(true);
@@ -36,11 +_,11 @@
       if (lightningbolt != null) {
          lightningbolt.m_6027_(this.f_30925_.m_20185_(), this.f_30925_.m_20186_(), this.f_30925_.m_20189_());
          lightningbolt.m_20874_(true);
-         serverlevel.m_7967_(lightningbolt);
+         serverlevel.strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRAP); // CraftBukkit
          Skeleton skeleton = this.m_30931_(difficultyinstance, this.f_30925_);
          if (skeleton != null) {
             skeleton.m_20329_(this.f_30925_);
-            serverlevel.m_47205_(skeleton);
+            serverlevel.addFreshEntityWithPassengers(skeleton, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRAP); // CraftBukkit
 
             for(int i = 0; i < 3; ++i) {
                AbstractHorse abstracthorse = this.m_30929_(difficultyinstance);
@@ -49,13 +_,14 @@
                   if (skeleton1 != null) {
                      skeleton1.m_20329_(abstracthorse);
                      abstracthorse.m_5997_(this.f_30925_.m_217043_().m_216328_(0.0D, 1.1485D), 0.0D, this.f_30925_.m_217043_().m_216328_(0.0D, 1.1485D));
-                     serverlevel.m_47205_(abstracthorse);
+                     serverlevel.addFreshEntityWithPassengers(abstracthorse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit
                   }
                }
             }
 
          }
       }
+      }));
    }
 
    @Nullable
