--- a/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
+++ b/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
@@ -23,7 +_,7 @@
    private final SkeletonTrapGoal f_30890_ = new SkeletonTrapGoal(this);
    private static final int f_149551_ = 18000;
    private boolean f_30891_;
-   private int f_30892_;
+   public int f_30892_;
 
    public SkeletonHorse(EntityType<? extends SkeletonHorse> p_30894_, Level p_30895_) {
       super(p_30894_, p_30895_);
