--- a/net/minecraft/world/entity/animal/horse/TraderLlama.java
+++ b/net/minecraft/world/entity/animal/horse/TraderLlama.java
@@ -137,7 +_,7 @@
       }
 
       public void m_8056_() {
-         this.f_26135_.m_6710_(this.f_30963_);
+         this.f_26135_.setTarget(this.f_30963_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit
          Entity entity = this.f_30962_.m_21524_();
          if (entity instanceof WanderingTrader) {
             this.f_30964_ = ((WanderingTrader)entity).m_21213_();
