--- ../src-base/minecraft/net/minecraft/util/text/ITextComponent.java
+++ ../src-work/minecraft/net/minecraft/util/text/ITextComponent.java
@@ -34,7 +34,6 @@
 
     String getUnformattedText();
 
-    @SideOnly(Side.CLIENT)
     String getFormattedText();
 
     List<ITextComponent> getSiblings();
@@ -281,7 +280,7 @@
             static
             {
                 GsonBuilder gsonbuilder = new GsonBuilder();
-                gsonbuilder.registerTypeHierarchyAdapter(ITextComponent.class, new ITextComponent.Serializer());
+                gsonbuilder.registerTypeHierarchyAdapter(ITextComponent.class, new Serializer());
                 gsonbuilder.registerTypeHierarchyAdapter(Style.class, new Style.Serializer());
                 gsonbuilder.registerTypeAdapterFactory(new EnumTypeAdapterFactory());
                 GSON = gsonbuilder.create();
