--- a/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
+++ b/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
@@ -14,7 +_,7 @@
 
 public class ObjectiveCriteria {
    private static final Map<String, ObjectiveCriteria> f_166107_ = Maps.newHashMap();
-   private static final Map<String, ObjectiveCriteria> f_166108_ = Maps.newHashMap();
+   public static final Map<String, ObjectiveCriteria> f_166108_ = Maps.newHashMap();
    public static final ObjectiveCriteria f_83588_ = m_166113_("dummy");
    public static final ObjectiveCriteria f_83589_ = m_166113_("trigger");
    public static final ObjectiveCriteria f_83590_ = m_166113_("deathCount");
