--- a/net/minecraft/world/gen/feature/structure/LegacyStructureDataUtil.java
+++ b/net/minecraft/world/gen/feature/structure/LegacyStructureDataUtil.java
@@ -212,6 +_,7 @@
          List<String> list = ImmutableList.of("EndCity");
          return new LegacyStructureDataUtil(p_236992_1_, list, list);
       } else {
+         if ("minecraft".equals(p_236992_0_.func_240901_a_().func_110624_b())) return new LegacyStructureDataUtil(p_236992_1_, ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument")); // CatServer - Bukkit custom world
          throw new RuntimeException(String.format("Unknown dimension type : %s", p_236992_0_));
       }
    }
