--- a/net/minecraft/world/gen/feature/structure/StructureManager.java
+++ b/net/minecraft/world/gen/feature/structure/StructureManager.java
@@ -8,6 +_,7 @@
 import net.minecraft.util.math.SectionPos;
 import net.minecraft.world.IStructureReader;
 import net.minecraft.world.IWorld;
+import net.minecraft.world.World;
 import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.gen.WorldGenRegion;
 import net.minecraft.world.gen.settings.DimensionGeneratorSettings;
@@ -65,4 +_,10 @@
          });
       }).findFirst(), StructureStart.field_214630_a);
    }
+
+   // Spigot start
+   public World getWorld() {
+      return this.field_235003_a_.getMinecraftWorld();
+   }
+   // Spigot end
 }
