--- a/net/minecraft/world/gen/feature/structure/SwampHutPiece.java
+++ b/net/minecraft/world/gen/feature/structure/SwampHutPiece.java
@@ -94,7 +_,7 @@
                witchentity.func_110163_bv();
                witchentity.func_70012_b((double)l + 0.5D, (double)i1, (double)k + 0.5D, 0.0F, 0.0F);
                witchentity.func_213386_a(p_230383_1_, p_230383_1_.func_175649_E(new BlockPos(l, i1, k)), SpawnReason.STRUCTURE, (ILivingEntityData)null, (CompoundNBT)null);
-               p_230383_1_.func_242417_l(witchentity);
+               p_230383_1_.addAllEntities(witchentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
             }
          }
 
@@ -114,7 +_,7 @@
             catentity.func_110163_bv();
             catentity.func_70012_b((double)i + 0.5D, (double)j, (double)k + 0.5D, 0.0F, 0.0F);
             catentity.func_213386_a(p_214821_1_, p_214821_1_.func_175649_E(new BlockPos(i, j, k)), SpawnReason.STRUCTURE, (ILivingEntityData)null, (CompoundNBT)null);
-            p_214821_1_.func_242417_l(catentity);
+            p_214821_1_.addAllEntities(catentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
          }
       }
 
