--- a/net/minecraft/world/entity/player/Abilities.java
+++ b/net/minecraft/world/entity/player/Abilities.java
@@ -8,8 +_,8 @@
    public boolean f_35936_;
    public boolean f_35937_;
    public boolean f_35938_ = true;
-   private float f_35939_ = 0.05F;
-   private float f_35940_ = 0.1F;
+   public float f_35939_ = 0.05F;
+   public float f_35940_ = 0.1F;
 
    public void m_35945_(CompoundTag p_35946_) {
       CompoundTag compoundtag = new CompoundTag();
