--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.player;
 
+import catserver.server.CatServerCaptures;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
@@ -12,6 +_,8 @@
 import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.SharedConstants;
@@ -110,8 +_,18 @@
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Scoreboard;
 import net.minecraft.world.scores.Team;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftVector;
+import org.bukkit.entity.Item;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityExhaustionEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerVelocityEvent;
 
-public abstract class Player extends LivingEntity {
+public abstract class Player extends LivingEntity implements net.minecraftforge.common.extensions.IForgePlayer {
+   public static final String PERSISTED_NBT_TAG = "PlayerPersisted";
    public static final String f_150081_ = "OfflinePlayer:";
    public static final int f_150082_ = 16;
    public static final int f_150083_ = 20;
@@ -133,10 +_,10 @@
    protected static final EntityDataAccessor<CompoundTag> f_36092_ = SynchedEntityData.m_135353_(Player.class, EntityDataSerializers.f_135042_);
    private long f_36109_;
    private final Inventory f_36093_ = new Inventory(this);
-   protected PlayerEnderChestContainer f_36094_ = new PlayerEnderChestContainer();
+   protected PlayerEnderChestContainer f_36094_ = new PlayerEnderChestContainer(this); // CraftBukkit - add "this" to constructor
    public final InventoryMenu f_36095_;
    public AbstractContainerMenu f_36096_;
-   protected FoodData f_36097_ = new FoodData();
+   protected FoodData f_36097_ = new FoodData(this); // CraftBukkit - add "this" to constructor
    protected int f_36098_;
    public float f_36099_;
    public float f_36100_;
@@ -147,9 +_,9 @@
    public double f_36105_;
    public double f_36106_;
    public double f_36075_;
-   private int f_36110_;
+   public int f_36110_;
    protected boolean f_36076_;
-   private final Abilities f_36077_ = new Abilities();
+   public final Abilities f_36077_ = new Abilities();
    public int f_36078_;
    public int f_36079_;
    public float f_36080_;
@@ -162,6 +_,9 @@
    private final ItemCooldowns f_36087_ = this.m_7478_();
    @Nullable
    public FishingHook f_36083_;
+   private final java.util.Collection<MutableComponent> prefixes = new java.util.LinkedList<>();
+   private final java.util.Collection<MutableComponent> suffixes = new java.util.LinkedList<>();
+   @Nullable private Pose forcedPose;
 
    public Player(Level p_36114_, BlockPos p_36115_, float p_36116_, GameProfile p_36117_) {
       super(EntityType.f_20532_, p_36114_);
@@ -187,7 +_,7 @@
    }
 
    public static AttributeSupplier.Builder m_36340_() {
-      return LivingEntity.m_21183_().m_22268_(Attributes.f_22281_, 1.0D).m_22268_(Attributes.f_22279_, (double)0.1F).m_22266_(Attributes.f_22283_).m_22266_(Attributes.f_22286_);
+       return LivingEntity.m_21183_().m_22268_(Attributes.f_22281_, 1.0D).m_22268_(Attributes.f_22279_, (double) 0.1F).m_22266_(Attributes.f_22283_).m_22266_(Attributes.f_22286_).m_22266_(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()).m_22266_(Attributes.f_22282_).m_22266_(net.minecraftforge.common.ForgeMod.ATTACK_RANGE.get());
    }
 
    protected void m_8097_() {
@@ -201,6 +_,7 @@
    }
 
    public void m_8119_() {
+      net.minecraftforge.event.ForgeEventFactory.onPlayerPreTick(this);
       this.f_19794_ = this.m_5833_();
       if (this.m_5833_()) {
          this.f_19861_ = false;
@@ -216,7 +_,7 @@
             this.f_36110_ = 100;
          }
 
-         if (!this.f_19853_.f_46443_ && this.f_19853_.m_46461_()) {
+         if (!this.f_19853_.f_46443_ && !net.minecraftforge.event.ForgeEventFactory.fireSleepingTimeCheck(this, m_21257_())) {
             this.m_6145_(false, true);
          }
       } else if (this.f_36110_ > 0) {
@@ -271,6 +_,7 @@
       this.m_36372_();
       this.f_36087_.m_41518_();
       this.m_7594_();
+      net.minecraftforge.event.ForgeEventFactory.onPlayerPostTick(this);
    }
 
    public boolean m_36341_() {
@@ -293,7 +_,7 @@
    private void m_36372_() {
       ItemStack itemstack = this.m_6844_(EquipmentSlot.HEAD);
       if (itemstack.m_150930_(Items.f_42354_) && !this.m_204029_(FluidTags.f_13131_)) {
-         this.m_7292_(new MobEffectInstance(MobEffects.f_19608_, 200, 0, false, false, true));
+         this.addEffect(new MobEffectInstance(MobEffects.f_19608_, 200, 0, false, false, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.TURTLE_HELMET); // CraftBukkit
       }
 
    }
@@ -346,6 +_,10 @@
    }
 
    protected void m_7594_() {
+      if(forcedPose != null) {
+         this.m_20124_(forcedPose);
+         return;
+      }
       if (this.m_20175_(Pose.SWIMMING)) {
          Pose pose;
          if (this.m_21255_()) {
@@ -408,7 +_,7 @@
       return SoundSource.PLAYERS;
    }
 
-   protected int m_6101_() {
+   public int m_6101_() {
       return 20;
    }
 
@@ -469,7 +_,7 @@
 
       if (this.f_19853_.m_46791_() == Difficulty.PEACEFUL && this.f_19853_.m_46469_().m_46207_(GameRules.f_46139_)) {
          if (this.m_21223_() < this.m_21233_() && this.f_19797_ % 20 == 0) {
-            this.m_5634_(1.0F);
+            this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN);
          }
 
          if (this.f_36097_.m_38721_() && this.f_19797_ % 10 == 0) {
@@ -569,6 +_,7 @@
    }
 
    public void m_6667_(DamageSource p_36152_) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_36152_)) return;
       super.m_6667_(p_36152_);
       this.m_20090_();
       if (!this.m_5833_()) {
@@ -625,7 +_,13 @@
 
    @Nullable
    public ItemEntity m_36176_(ItemStack p_36177_, boolean p_36178_) {
-      return this.m_7197_(p_36177_, false, p_36178_);
+      return net.minecraftforge.common.ForgeHooks.onPlayerTossEvent(this, p_36177_, p_36178_);
+   }
+
+   @Nullable
+   public ItemEntity drop(ItemStack p_36179_, boolean p_36180_, boolean p_36181_, boolean callEvent) {
+      CatServerCaptures.getCatServerCaptures().captureIsCallEvent(callEvent);
+      return m_7197_(p_36179_, p_36180_, p_36181_);
    }
 
    @Nullable
@@ -659,11 +_,42 @@
             itementity.m_20334_((double)(-f3 * f2 * 0.3F) + Math.cos((double)f5) * (double)f6, (double)(-f8 * 0.3F + 0.1F + (this.f_19796_.nextFloat() - this.f_19796_.nextFloat()) * 0.1F), (double)(f4 * f2 * 0.3F) + Math.sin((double)f5) * (double)f6);
          }
 
+         // CraftBukkit start - fire PlayerDropItemEvent
+         var callEvent = CatServerCaptures.getCatServerCaptures().getCaptureIsCallEvent();
+         if (!callEvent) { // SPIGOT-2942: Add boolean to call event
+            return itementity;
+         }
+         org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.getBukkitEntity();
+         Item drop = (Item) itementity.getBukkitEntity();
+         PlayerDropItemEvent event = new PlayerDropItemEvent(player, drop);
+         this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
+            if (p_36181_ && (cur == null || cur.getAmount() == 0)) {
+               // The complete stack was dropped
+               player.getInventory().setItemInHand(drop.getItemStack());
+            } else if (p_36181_ && cur.isSimilar(drop.getItemStack()) && cur.getAmount() < cur.getMaxStackSize() && drop.getItemStack().getAmount() == 1) {
+               // Only one item is dropped
+               cur.setAmount(cur.getAmount() + 1);
+               player.getInventory().setItemInHand(cur);
+            } else {
+               // Fallback
+               player.getInventory().addItem(drop.getItemStack());
+            }
+            return null;
+         }
+         // CraftBukkit end
+
          return itementity;
       }
    }
 
+   @Deprecated //Use location sensitive version below
    public float m_36281_(BlockState p_36282_) {
+      return getDigSpeed(p_36282_, null);
+   }
+
+   public float getDigSpeed(BlockState p_36282_, @Nullable BlockPos pos) {
       float f = this.f_36093_.m_36020_(p_36282_);
       if (f > 1.0F) {
          int i = EnchantmentHelper.m_44926_(this);
@@ -705,11 +_,12 @@
          f /= 5.0F;
       }
 
+      f = net.minecraftforge.event.ForgeEventFactory.getBreakSpeed(this, p_36282_, f, pos);
       return f;
    }
 
    public boolean m_36298_(BlockState p_36299_) {
-      return !p_36299_.m_60834_() || this.f_36093_.m_36056_().m_41735_(p_36299_);
+      return net.minecraftforge.event.ForgeEventFactory.doPlayerHarvestCheck(this, p_36299_, !p_36299_.m_60834_() || this.f_36093_.m_36056_().m_41735_(p_36299_));
    }
 
    public void m_7378_(CompoundTag p_36215_) {
@@ -786,6 +_,7 @@
    }
 
    public boolean m_6469_(DamageSource p_36154_, float p_36155_) {
+      if (!net.minecraftforge.common.ForgeHooks.onPlayerAttack(this, p_36154_, p_36155_)) return false;
       if (this.m_6673_(p_36154_)) {
          return false;
       } else if (this.f_36077_.f_35934_ && !p_36154_.m_19378_()) {
@@ -796,12 +_,12 @@
             return false;
          } else {
             if (!this.f_19853_.f_46443_) {
-               this.m_36328_();
+               // this.removeEntitiesOnShoulder(); // CraftBukkit - moved down
             }
 
             if (p_36154_.m_7986_()) {
                if (this.f_19853_.m_46791_() == Difficulty.PEACEFUL) {
-                  p_36155_ = 0.0F;
+                  return false; // CraftBukkit - f = 0.0f -> return false
                }
 
                if (this.f_19853_.m_46791_() == Difficulty.EASY) {
@@ -813,14 +_,20 @@
                }
             }
 
-            return p_36155_ == 0.0F ? false : super.m_6469_(p_36154_, p_36155_);
+            // CraftBukkit start - Don't filter out 0 damage
+            boolean damaged = super.m_6469_(p_36154_, p_36155_);
+            if (damaged) {
+               this.m_36328_();
+            }
+            return damaged;
+            // CraftBukkit end
          }
       }
    }
 
    protected void m_6728_(LivingEntity p_36295_) {
       super.m_6728_(p_36295_);
-      if (p_36295_.m_21205_().m_41720_() instanceof AxeItem) {
+      if (p_36295_.m_21205_().canDisableShield(this.f_20935_, this, p_36295_)) {
          this.m_36384_(true);
       }
 
@@ -831,13 +_,27 @@
    }
 
    public boolean m_7099_(Player p_36169_) {
-      Team team = this.m_5647_();
-      Team team1 = p_36169_.m_5647_();
-      if (team == null) {
-         return true;
+      // CraftBukkit start - Change to check OTHER player's scoreboard team according to API
+      // To summarize this method's logic, it's "Can parameter hurt this"
+      org.bukkit.scoreboard.Team team;
+      if (p_36169_ instanceof ServerPlayer thatPlayer) {
+         team = thatPlayer.getBukkitEntity().getScoreboard().getPlayerTeam(thatPlayer.getBukkitEntity());
+         if (team == null || team.allowFriendlyFire()) {
+            return true;
+         }
       } else {
-         return !team.m_83536_(team1) ? true : team.m_6260_();
-      }
+         // This should never be called, but is implemented anyway
+         org.bukkit.OfflinePlayer thisPlayer = p_36169_.f_19853_.getCraftServer().getOfflinePlayer(p_36169_.m_6302_());
+         team = p_36169_.f_19853_.getCraftServer().getScoreboardManager().getMainScoreboard().getPlayerTeam(thisPlayer);
+         if (team == null || team.allowFriendlyFire()) {
+            return true;
+         }
+      }
+      if (this instanceof ServerPlayer) {
+         return !team.hasPlayer(((ServerPlayer) this).getBukkitEntity());
+      }
+      return !team.hasPlayer(this.f_19853_.getCraftServer().getOfflinePlayer(this.m_6302_()));
+      // CraftBukkit end
    }
 
    protected void m_6472_(DamageSource p_36251_, float p_36252_) {
@@ -849,7 +_,7 @@
    }
 
    protected void m_7909_(float p_36383_) {
-      if (this.f_20935_.m_150930_(Items.f_42740_)) {
+      if (this.f_20935_.canPerformAction(net.minecraftforge.common.ToolActions.SHIELD_BLOCK)) {
          if (!this.f_19853_.f_46443_) {
             this.m_36246_(Stats.f_12982_.m_12902_(this.f_20935_.m_41720_()));
          }
@@ -859,6 +_,7 @@
             InteractionHand interactionhand = this.m_7655_();
             this.f_20935_.m_41622_(i, this, (p_36149_) -> {
                p_36149_.m_21190_(interactionhand);
+               net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, this.f_20935_, interactionhand);
             });
             if (this.f_20935_.m_41619_()) {
                if (interactionhand == InteractionHand.MAIN_HAND) {
@@ -877,26 +_,58 @@
 
    protected void m_6475_(DamageSource p_36312_, float p_36313_) {
       if (!this.m_6673_(p_36312_)) {
-         p_36313_ = this.m_21161_(p_36312_, p_36313_);
-         p_36313_ = this.m_6515_(p_36312_, p_36313_);
-         float f2 = Math.max(p_36313_ - this.m_6103_(), 0.0F);
-         this.m_7911_(this.m_6103_() - (p_36313_ - f2));
-         float f = p_36313_ - f2;
-         if (f > 0.0F && f < 3.4028235E37F) {
-            this.m_36222_(Stats.f_12933_, Math.round(f * 10.0F));
-         }
-
-         if (f2 != 0.0F) {
-            this.m_36399_(p_36312_.m_19377_());
-            float f1 = this.m_21223_();
-            this.m_21153_(this.m_21223_() - f2);
-            this.m_21231_().m_19289_(p_36312_, f1, f2);
-            if (f2 < 3.4028235E37F) {
-               this.m_36222_(Stats.f_12931_, Math.round(f2 * 10.0F));
-            }
-
-         }
-      }
+         p_36313_ = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, p_36312_, p_36313_);
+         if (p_36313_ <= 0) return;
+         p_36313_ = this.m_21161_(p_36312_, p_36313_);
+         p_36313_ = this.m_6515_(p_36312_, p_36313_);
+         float f2 = Math.max(p_36313_ - this.m_6103_(), 0.0F);
+         this.m_7911_(this.m_6103_() - (p_36313_ - f2));
+         f2 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, p_36312_, f2);
+         float f = p_36313_ - f2;
+         if (f > 0.0F && f < 3.4028235E37F) {
+            this.m_36222_(Stats.f_12933_, Math.round(f * 10.0F));
+         }
+
+         if (f2 != 0.0F) {
+            this.causeFoodExhaustion(p_36312_.m_19377_(), EntityExhaustionEvent.ExhaustionReason.DAMAGED); // CraftBukkit - EntityExhaustionEvent
+            float f1 = this.m_21223_();
+            this.m_21231_().m_19289_(p_36312_, f1, f2);
+            this.m_21153_(f1 - f2); // Forge: moved to fix MC-121048
+            if (f2 < 3.4028235E37F) {
+               this.m_36222_(Stats.f_12931_, Math.round(f2 * 10.0F));
+            }
+         }
+      }
+   }
+
+   protected boolean damageEntity0(DamageSource p_36312_, float p_36313_) {
+      if (true) {
+         return super.damageEntity0(p_36312_, p_36313_);
+      }
+      if (!this.m_6673_(p_36312_)) {
+         p_36313_ = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, p_36312_, p_36313_);
+         if (p_36313_ <= 0) return false;
+         p_36313_ = this.m_21161_(p_36312_, p_36313_);
+         p_36313_ = this.m_6515_(p_36312_, p_36313_);
+         float f2 = Math.max(p_36313_ - this.m_6103_(), 0.0F);
+         this.m_7911_(this.m_6103_() - (p_36313_ - f2));
+         f2 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, p_36312_, f2);
+         float f = p_36313_ - f2;
+         if (f > 0.0F && f < 3.4028235E37F) {
+            this.m_36222_(Stats.f_12933_, Math.round(f * 10.0F));
+         }
+
+         if (f2 != 0.0F) {
+            this.causeFoodExhaustion(p_36312_.m_19377_(), EntityExhaustionEvent.ExhaustionReason.DAMAGED); // CraftBukkit - EntityExhaustionEvent
+            float f1 = this.m_21223_();
+            this.m_21231_().m_19289_(p_36312_, f1, f2);
+            this.m_21153_(f1 - f2); // Forge: moved to fix MC-121048
+            if (f2 < 3.4028235E37F) {
+               this.m_36222_(Stats.f_12931_, Math.round(f2 * 10.0F));
+            }
+         }
+      }
+      return false;
    }
 
    protected boolean m_6046_() {
@@ -939,6 +_,8 @@
 
          return InteractionResult.PASS;
       } else {
+         InteractionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntity(this, p_36158_, p_36159_);
+         if (cancelResult != null) return cancelResult;
          ItemStack itemstack = this.m_21120_(p_36159_);
          ItemStack itemstack1 = itemstack.m_41777_();
          InteractionResult interactionresult = p_36158_.m_6096_(this, p_36159_);
@@ -947,6 +_,9 @@
                itemstack.m_41764_(itemstack1.m_41613_());
             }
 
+            if (!this.f_36077_.f_35937_ && itemstack.m_41619_()) {
+               net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, p_36159_);
+            }
             return interactionresult;
          } else {
             if (!itemstack.m_41619_() && p_36158_ instanceof LivingEntity) {
@@ -957,6 +_,7 @@
                InteractionResult interactionresult1 = itemstack.m_41647_(this, (LivingEntity)p_36158_, p_36159_);
                if (interactionresult1.m_19077_()) {
                   if (itemstack.m_41619_() && !this.f_36077_.f_35937_) {
+                     net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, p_36159_);
                      this.m_21008_(p_36159_, ItemStack.f_41583_);
                   }
 
@@ -986,6 +_,7 @@
       return !this.f_36077_.f_35935_;
    }
 
+   // Forge: Don't update this method to use IForgeEntity#getStepHeight() - https://github.com/MinecraftForge/MinecraftForge/issues/8922
    protected Vec3 m_5763_(Vec3 p_36201_, MoverType p_36202_) {
       if (!this.f_36077_.f_35935_ && (p_36202_ == MoverType.SELF || p_36202_ == MoverType.PLAYER) && this.m_36343_() && this.m_36386_()) {
          double d0 = p_36201_.f_82479_;
@@ -1037,10 +_,11 @@
    }
 
    private boolean m_36386_() {
-      return this.f_19861_ || this.f_19789_ < this.f_19793_ && !this.f_19853_.m_45756_(this, this.m_142469_().m_82386_(0.0D, (double)(this.f_19789_ - this.f_19793_), 0.0D));
+      return this.f_19861_ || this.f_19789_ < this.getStepHeight() && !this.f_19853_.m_45756_(this, this.m_142469_().m_82386_(0.0D, (double)(this.f_19789_ - this.getStepHeight()), 0.0D));
    }
 
    public void m_5706_(Entity p_36347_) {
+      if (!net.minecraftforge.common.ForgeHooks.onPlayerAttackTarget(this, p_36347_)) return;
       if (p_36347_.m_6097_()) {
          if (!p_36347_.m_7313_(this)) {
             float f = (float)this.m_21133_(Attributes.f_22281_);
@@ -1054,11 +_,11 @@
             float f2 = this.m_36403_(0.5F);
             f *= 0.2F + f2 * f2 * 0.8F;
             f1 *= f2;
-            this.m_36334_();
+            // this.resetAttackStrengthTicker(); // CraftBukkit - Moved to EntityLiving to reset the cooldown after the damage is dealt
             if (f > 0.0F || f1 > 0.0F) {
                boolean flag = f2 > 0.9F;
                boolean flag1 = false;
-               int i = 0;
+               float i = (float)this.m_21133_(Attributes.f_22282_); // Forge: Initialize this value to the attack knockback attribute of the player, which is by default 0
                i += EnchantmentHelper.m_44894_(this);
                if (this.m_20142_() && flag) {
                   this.f_19853_.m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12314_, this.m_5720_(), 1.0F, 1.0F);
@@ -1068,8 +_,10 @@
 
                boolean flag2 = flag && this.f_19789_ > 0.0F && !this.f_19861_ && !this.m_6147_() && !this.m_20069_() && !this.m_21023_(MobEffects.f_19610_) && !this.m_20159_() && p_36347_ instanceof LivingEntity;
                flag2 = flag2 && !this.m_20142_();
+               net.minecraftforge.event.entity.player.CriticalHitEvent hitResult = net.minecraftforge.common.ForgeHooks.getCriticalHit(this, p_36347_, flag2, flag2 ? 1.5F : 1.0F);
+               flag2 = hitResult != null;
                if (flag2) {
-                  f *= 1.5F;
+                  f *= hitResult.getDamageModifier();
                }
 
                f += f1;
@@ -1077,9 +_,7 @@
                double d0 = (double)(this.f_19787_ - this.f_19867_);
                if (flag && !flag2 && !flag1 && this.f_19861_ && d0 < (double)this.m_6113_()) {
                   ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-                  if (itemstack.m_41720_() instanceof SwordItem) {
-                     flag3 = true;
-                  }
+                  flag3 = itemstack.canPerformAction(net.minecraftforge.common.ToolActions.SWORD_SWEEP);
                }
 
                float f4 = 0.0F;
@@ -1088,8 +_,14 @@
                if (p_36347_ instanceof LivingEntity) {
                   f4 = ((LivingEntity)p_36347_).m_21223_();
                   if (j > 0 && !p_36347_.m_6060_()) {
-                     flag4 = true;
-                     p_36347_.m_20254_(1);
+                     // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+                     EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), p_36347_.getBukkitEntity(), 1);
+                     org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+                     if (!combustEvent.isCancelled()) {
+                        flag4 = true;
+                        p_36347_.setSecondsOnFire(combustEvent.getDuration(), false);
+                     }
+                     // CraftBukkit end
                   }
                }
 
@@ -1110,10 +_,12 @@
                   if (flag3) {
                      float f3 = 1.0F + EnchantmentHelper.m_44821_(this) * f;
 
-                     for(LivingEntity livingentity : this.f_19853_.m_45976_(LivingEntity.class, p_36347_.m_142469_().m_82377_(1.0D, 0.25D, 1.0D))) {
-                        if (livingentity != this && livingentity != p_36347_ && !this.m_7307_(livingentity) && (!(livingentity instanceof ArmorStand) || !((ArmorStand)livingentity).m_31677_()) && this.m_20280_(livingentity) < 9.0D) {
-                           livingentity.m_147240_((double)0.4F, (double)Mth.m_14031_(this.m_146908_() * ((float)Math.PI / 180F)), (double)(-Mth.m_14089_(this.m_146908_() * ((float)Math.PI / 180F))));
-                           livingentity.m_6469_(DamageSource.m_19344_(this), f3);
+                     for(LivingEntity livingentity : this.f_19853_.m_45976_(LivingEntity.class, this.m_21120_(InteractionHand.MAIN_HAND).getSweepHitBox(this, p_36347_))) {
+                        if (livingentity != this && livingentity != p_36347_ && !this.m_7307_(livingentity) && (!(livingentity instanceof ArmorStand) || !((ArmorStand)livingentity).m_31677_()) && this.canHit(livingentity, 0)) { // Original check was dist < 3, range is 3, so vanilla used padding=0
+                           // CraftBukkit start - Only apply knockback if the damage hits
+                           if (livingentity.m_6469_(DamageSource.m_19344_(this).sweep(), f3)) {
+                              livingentity.m_147240_((double) 0.4F, (double) Mth.m_14031_(this.m_146908_() * ((float) Math.PI / 180F)), (double) (-Mth.m_14089_(this.m_146908_() * ((float) Math.PI / 180F))));
+                           }
                         }
                      }
 
@@ -1122,10 +_,24 @@
                   }
 
                   if (p_36347_ instanceof ServerPlayer && p_36347_.f_19864_) {
-                     ((ServerPlayer)p_36347_).f_8906_.m_141995_(new ClientboundSetEntityMotionPacket(p_36347_));
-                     p_36347_.f_19864_ = false;
-                     p_36347_.m_20256_(vec3);
+                     // CraftBukkit start - Add Velocity Event
+                     boolean cancelled = false;
+                     org.bukkit.entity.Player player = (org.bukkit.entity.Player) p_36347_.getBukkitEntity();
+                     org.bukkit.util.Vector velocity = CraftVector.toBukkit(vec3);
+                     PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+                     f_19853_.getCraftServer().getPluginManager().callEvent(event);
+                     if (event.isCancelled()) {
+                        cancelled = true;
+                     } else if (!velocity.equals(event.getVelocity())) {
+                        player.setVelocity(event.getVelocity());
+                     }
+                     if (!cancelled) {
+                        ((ServerPlayer) p_36347_).f_8906_.m_141995_(new ClientboundSetEntityMotionPacket(p_36347_));
+                        p_36347_.f_19864_ = false;
+                        p_36347_.m_20256_(vec3);
+                     }
                   }
+                  // CraftBukkit end
 
                   if (flag2) {
                      this.f_19853_.m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12313_, this.m_5720_(), 1.0F, 1.0F);
@@ -1152,13 +_,15 @@
                   EnchantmentHelper.m_44896_(this, p_36347_);
                   ItemStack itemstack1 = this.m_21205_();
                   Entity entity = p_36347_;
-                  if (p_36347_ instanceof EnderDragonPart) {
-                     entity = ((EnderDragonPart)p_36347_).f_31010_;
+                  if (p_36347_ instanceof net.minecraftforge.entity.PartEntity) {
+                     entity = ((net.minecraftforge.entity.PartEntity<?>) p_36347_).getParent();
                   }
 
                   if (!this.f_19853_.f_46443_ && !itemstack1.m_41619_() && entity instanceof LivingEntity) {
+                     ItemStack copy = itemstack1.m_41777_();
                      itemstack1.m_41640_((LivingEntity)entity, this);
                      if (itemstack1.m_41619_()) {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, copy, InteractionHand.MAIN_HAND);
                         this.m_21008_(InteractionHand.MAIN_HAND, ItemStack.f_41583_);
                      }
                   }
@@ -1167,7 +_,13 @@
                      float f5 = f4 - ((LivingEntity)p_36347_).m_21223_();
                      this.m_36222_(Stats.f_12928_, Math.round(f5 * 10.0F));
                      if (j > 0) {
-                        p_36347_.m_20254_(j * 4);
+                        // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+                        EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), j * 4);
+                        org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+                        if (!combustEvent.isCancelled()) {
+                           entity.setSecondsOnFire(combustEvent.getDuration(), false);
+                        }
+                        // CraftBukkit end
                      }
 
                      if (this.f_19853_ instanceof ServerLevel && f5 > 2.0F) {
@@ -1175,13 +_,17 @@
                         ((ServerLevel)this.f_19853_).m_8767_(ParticleTypes.f_123798_, p_36347_.m_20185_(), p_36347_.m_20227_(0.5D), p_36347_.m_20189_(), k, 0.1D, 0.0D, 0.1D, 0.2D);
                      }
                   }
-
-                  this.m_36399_(0.1F);
+                  this.causeFoodExhaustion(0.1F, EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent
                } else {
                   this.f_19853_.m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12315_, this.m_5720_(), 1.0F, 1.0F);
                   if (flag4) {
                      p_36347_.m_20095_();
                   }
+                  // CraftBukkit start - resync on cancelled event
+                  if (this instanceof ServerPlayer) {
+                     ((ServerPlayer) this).getBukkitEntity().updateInventory();
+                  }
+                  // CraftBukkit end
                }
             }
 
@@ -1200,7 +_,7 @@
       }
 
       if (this.f_19796_.nextFloat() < f) {
-         this.m_36335_().m_41524_(Items.f_42740_, 100);
+         this.m_36335_().m_41524_(this.m_21211_().m_41720_(), 100);
          this.m_5810_();
          this.f_19853_.m_7605_(this, (byte)30);
       }
@@ -1253,13 +_,21 @@
    public void m_141945_(ItemStack p_150098_, ItemStack p_150099_, ClickAction p_150100_) {
    }
 
+   // CraftBukkit start
+   public Player forceSleep(boolean isForceSleep) {
+      CatServerCaptures.getCatServerCaptures().captureIsForceSleep(isForceSleep);
+      return this;
+   }
+
    public Either<Player.BedSleepingProblem, Unit> m_7720_(BlockPos p_36203_) {
       this.m_5802_(p_36203_);
       this.f_36110_ = 0;
       return Either.right(Unit.INSTANCE);
    }
+   // CraftBukkit end
 
    public void m_6145_(boolean p_36226_, boolean p_36227_) {
+      net.minecraftforge.event.ForgeEventFactory.onPlayerWakeup(this, p_36226_, p_36227_);
       super.m_5796_();
       if (this.f_19853_ instanceof ServerLevel && p_36227_) {
          ((ServerLevel)this.f_19853_).m_8878_();
@@ -1285,7 +_,7 @@
       } else if (block instanceof BedBlock && BedBlock.m_49488_(p_36131_)) {
          return BedBlock.m_49458_(EntityType.f_20532_, p_36131_, p_36132_, p_36133_);
       } else if (!p_36134_) {
-         return Optional.empty();
+         return blockstate.getRespawnPosition(EntityType.f_20532_, p_36131_, p_36132_, p_36133_, null);
       } else {
          boolean flag = block.m_5568_();
          boolean flag1 = p_36131_.m_8055_(p_36132_.m_7494_()).m_60734_().m_5568_();
@@ -1337,9 +_,9 @@
       super.m_6135_();
       this.m_36220_(Stats.f_12926_);
       if (this.m_20142_()) {
-         this.m_36399_(0.2F);
+         this.causeFoodExhaustion(0.2F, EntityExhaustionEvent.ExhaustionReason.JUMP_SPRINT); // CraftBukkit - EntityExhaustionEvent
       } else {
-         this.m_36399_(0.05F);
+         this.causeFoodExhaustion(0.05F, EntityExhaustionEvent.ExhaustionReason.JUMP); // CraftBukkit - EntityExhaustionEvent
       }
 
    }
@@ -1366,7 +_,11 @@
          this.m_20334_(vec3.f_82479_, d5 * 0.6D, vec3.f_82481_);
          this.f_20887_ = f;
          this.m_183634_();
-         this.m_20115_(7, false);
+         // CraftBukkit start
+         if (m_20291_(7) && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) {
+            this.m_20115_(7, false);
+         }
+         // CraftBukkit end
       } else {
          super.m_7023_(p_36359_);
       }
@@ -1397,19 +_,19 @@
             int i = Math.round((float)Math.sqrt(p_36379_ * p_36379_ + p_36380_ * p_36380_ + p_36381_ * p_36381_) * 100.0F);
             if (i > 0) {
                this.m_36222_(Stats.f_12924_, i);
-               this.m_36399_(0.01F * (float)i * 0.01F);
+               this.causeFoodExhaustion(0.01F * (float)i * 0.01F, EntityExhaustionEvent.ExhaustionReason.SWIM); // CraftBukkit - EntityExhaustionEvent
             }
          } else if (this.m_204029_(FluidTags.f_13131_)) {
             int j = Math.round((float)Math.sqrt(p_36379_ * p_36379_ + p_36380_ * p_36380_ + p_36381_ * p_36381_) * 100.0F);
             if (j > 0) {
                this.m_36222_(Stats.f_13001_, j);
-               this.m_36399_(0.01F * (float)j * 0.01F);
+               this.causeFoodExhaustion(0.01F * (float)j * 0.01F, EntityExhaustionEvent.ExhaustionReason.WALK_UNDERWATER); // CraftBukkit - EntityExhaustionEvent
             }
          } else if (this.m_20069_()) {
             int k = Math.round((float)Math.sqrt(p_36379_ * p_36379_ + p_36381_ * p_36381_) * 100.0F);
             if (k > 0) {
                this.m_36222_(Stats.f_12997_, k);
-               this.m_36399_(0.01F * (float)k * 0.01F);
+               this.causeFoodExhaustion(0.01F * (float)k * 0.01F, EntityExhaustionEvent.ExhaustionReason.WALK_ON_WATER); // CraftBukkit - EntityExhaustionEvent
             }
          } else if (this.m_6147_()) {
             if (p_36380_ > 0.0D) {
@@ -1420,13 +_,13 @@
             if (l > 0) {
                if (this.m_20142_()) {
                   this.m_36222_(Stats.f_12996_, l);
-                  this.m_36399_(0.1F * (float)l * 0.01F);
+                  this.causeFoodExhaustion(0.1F * (float)l * 0.01F, EntityExhaustionEvent.ExhaustionReason.SPRINT); // CraftBukkit - EntityExhaustionEvent
                } else if (this.m_6047_()) {
                   this.m_36222_(Stats.f_12995_, l);
-                  this.m_36399_(0.0F * (float)l * 0.01F);
+                  this.causeFoodExhaustion(0.0F * (float)l * 0.01F, EntityExhaustionEvent.ExhaustionReason.CROUCH); // CraftBukkit - EntityExhaustionEvent
                } else {
                   this.m_36222_(Stats.f_12994_, l);
-                  this.m_36399_(0.0F * (float)l * 0.01F);
+                  this.causeFoodExhaustion(0.0F * (float)l * 0.01F, EntityExhaustionEvent.ExhaustionReason.WALK); // CraftBukkit - EntityExhaustionEvent
                }
             }
          } else if (this.m_21255_()) {
@@ -1465,6 +_,7 @@
 
    public boolean m_142535_(float p_150093_, float p_150094_, DamageSource p_150095_) {
       if (this.f_36077_.f_35936_) {
+         net.minecraftforge.event.ForgeEventFactory.onPlayerFall(this, p_150093_, p_150093_);
          return false;
       } else {
          if (p_150093_ >= 2.0F) {
@@ -1478,7 +_,7 @@
    public boolean m_36319_() {
       if (!this.f_19861_ && !this.m_21255_() && !this.m_20069_() && !this.m_21023_(MobEffects.f_19620_)) {
          ItemStack itemstack = this.m_6844_(EquipmentSlot.CHEST);
-         if (itemstack.m_150930_(Items.f_42741_) && ElytraItem.m_41140_(itemstack)) {
+         if (itemstack.canElytraFly(this)) {
             this.m_36320_();
             return true;
          }
@@ -1488,12 +_,24 @@
    }
 
    public void m_36320_() {
-      this.m_20115_(7, true);
+      // CraftBukkit start
+      if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callToggleGlideEvent(this, true).isCancelled()) {
+         this.m_20115_(7, true);
+      } else {
+         // SPIGOT-5542: must toggle like below
+         this.m_20115_(7, true);
+         this.m_20115_(7, false);
+      }
+      // CraftBukkit end
    }
 
    public void m_36321_() {
-      this.m_20115_(7, true);
-      this.m_20115_(7, false);
+      // CraftBukkit start
+      if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) {
+         this.m_20115_(7, true);
+         this.m_20115_(7, false);
+      }
+      // CraftBukkit end
    }
 
    protected void m_5841_() {
@@ -1519,6 +_,10 @@
    }
 
    public void m_6756_(int p_36291_) {
+      net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange(this, p_36291_);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
+      p_36291_ = event.getAmount();
+
       this.m_36401_(p_36291_);
       this.f_36080_ += (float)p_36291_ / (float)this.m_36323_();
       this.f_36079_ = Mth.m_14045_(this.f_36079_ + p_36291_, 0, Integer.MAX_VALUE);
@@ -1547,7 +_,7 @@
    }
 
    public void m_7408_(ItemStack p_36172_, int p_36173_) {
-      this.f_36078_ -= p_36173_;
+      m_6749_(-p_36173_);
       if (this.f_36078_ < 0) {
          this.f_36078_ = 0;
          this.f_36080_ = 0.0F;
@@ -1558,6 +_,10 @@
    }
 
    public void m_6749_(int p_36276_) {
+      net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange(this, p_36276_);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
+      p_36276_ = event.getLevels();
+
       this.f_36078_ += p_36276_;
       if (this.f_36078_ < 0) {
          this.f_36078_ = 0;
@@ -1581,14 +_,27 @@
       }
    }
 
+   // CatServer start
+   private AtomicReference<EntityExhaustionEvent.ExhaustionReason> catserver$causeFoodReason = new AtomicReference<>(EntityExhaustionEvent.ExhaustionReason.UNKNOWN);
+   public void causeFoodExhaustion(float p_36400_, EntityExhaustionEvent.ExhaustionReason reason) {
+      this.catserver$causeFoodReason.set(reason);
+      this.m_36399_(p_36400_);
+   }
+
    public void m_36399_(float p_36400_) {
       if (!this.f_36077_.f_35934_) {
          if (!this.f_19853_.f_46443_) {
-            this.f_36097_.m_38703_(p_36400_);
+            var reason = this.catserver$causeFoodReason.getAndSet(EntityExhaustionEvent.ExhaustionReason.UNKNOWN);
+            // CraftBukkit start
+            EntityExhaustionEvent event = CraftEventFactory.callPlayerExhaustionEvent(this, reason, p_36400_);
+            if (!event.isCancelled()) {
+               this.f_36097_.m_38703_(event.getExhaustion());
+            }
+            // CraftBukkit end
          }
-
       }
    }
+   // CatServer end
 
    public FoodData m_36324_() {
       return this.f_36097_;
@@ -1658,16 +_,22 @@
       }
    }
 
+   public void setItemSlot(EquipmentSlot p_36161_, ItemStack p_36162_, boolean silent) {
+      CatServerCaptures.getCatServerCaptures().captureIsSlient(silent);
+      this.m_8061_(p_36161_, p_36162_);
+   }
+
    public void m_8061_(EquipmentSlot p_36161_, ItemStack p_36162_) {
       this.m_181122_(p_36162_);
+      var silent = CatServerCaptures.getCatServerCaptures().getCaptureIsSlient();
       if (p_36161_ == EquipmentSlot.MAINHAND) {
-         this.m_147218_(p_36162_);
+         this.equipEventAndSound(p_36162_, silent); // CraftBukkit
          this.f_36093_.f_35974_.set(this.f_36093_.f_35977_, p_36162_);
       } else if (p_36161_ == EquipmentSlot.OFFHAND) {
-         this.m_147218_(p_36162_);
+         this.equipEventAndSound(p_36162_, silent); // CraftBukkit
          this.f_36093_.f_35976_.set(0, p_36162_);
       } else if (p_36161_.m_20743_() == EquipmentSlot.Type.ARMOR) {
-         this.m_147218_(p_36162_);
+         this.equipEventAndSound(p_36162_, silent); // CraftBukkit
          this.f_36093_.f_35975_.set(p_36161_.m_20749_(), p_36162_);
       }
 
@@ -1722,7 +_,7 @@
             }
 
             p_36293_.m_6034_(this.m_20185_(), this.m_20186_() + (double)0.7F, this.m_20189_());
-            ((ServerLevel)this.f_19853_).m_8847_(p_36293_);
+            ((ServerLevel)this.f_19853_).addWithUUID(p_36293_, CreatureSpawnEvent.SpawnReason.SHOULDER_ENTITY); // CraftBukkit
          });
       }
 
@@ -1745,7 +_,11 @@
    }
 
    public Component m_5446_() {
-      MutableComponent mutablecomponent = PlayerTeam.m_83348_(this.m_5647_(), this.m_7755_());
+      if (this.displayname == null) this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.m_7755_());
+      MutableComponent mutablecomponent = new TextComponent("");
+      mutablecomponent = prefixes.stream().reduce(mutablecomponent, MutableComponent::m_7220_);
+      mutablecomponent = mutablecomponent.m_7220_(PlayerTeam.m_83348_(this.m_5647_(), this.displayname));
+      mutablecomponent = suffixes.stream().reduce(mutablecomponent, MutableComponent::m_7220_);
       return this.m_36218_(mutablecomponent);
    }
 
@@ -1835,7 +_,7 @@
       return this.f_19804_.m_135370_(f_36091_);
    }
 
-   protected void m_36362_(CompoundTag p_36363_) {
+   public void m_36362_(CompoundTag p_36363_) {
       this.f_19804_.m_135381_(f_36091_, p_36363_);
    }
 
@@ -1843,7 +_,7 @@
       return this.f_19804_.m_135370_(f_36092_);
    }
 
-   protected void m_36364_(CompoundTag p_36365_) {
+   public void m_36364_(CompoundTag p_36365_) {
       this.f_19804_.m_135381_(f_36092_, p_36365_);
    }
 
@@ -1895,24 +_,24 @@
          Predicate<ItemStack> predicate = ((ProjectileWeaponItem)p_36349_.m_41720_()).m_6442_();
          ItemStack itemstack = ProjectileWeaponItem.m_43010_(this, predicate);
          if (!itemstack.m_41619_()) {
-            return itemstack;
+            return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, itemstack);
          } else {
             predicate = ((ProjectileWeaponItem)p_36349_.m_41720_()).m_6437_();
 
             for(int i = 0; i < this.f_36093_.m_6643_(); ++i) {
                ItemStack itemstack1 = this.f_36093_.m_8020_(i);
                if (predicate.test(itemstack1)) {
-                  return itemstack1;
+                  return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, itemstack1);
                }
             }
 
-            return this.f_36077_.f_35937_ ? new ItemStack(Items.f_42412_) : ItemStack.f_41583_;
+            return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, this.f_36077_.f_35937_ ? new ItemStack(Items.f_42412_) : ItemStack.f_41583_);
          }
       }
    }
 
    public ItemStack m_5584_(Level p_36185_, ItemStack p_36186_) {
-      this.m_36324_().m_38712_(p_36186_.m_41720_(), p_36186_);
+      this.m_36324_().eat(p_36186_.m_41720_(), p_36186_, this);
       this.m_36246_(Stats.f_12982_.m_12902_(p_36186_.m_41720_()));
       p_36185_.m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12321_, SoundSource.PLAYERS, 0.5F, p_36185_.f_46441_.nextFloat() * 0.1F + 0.9F);
       if (this instanceof ServerPlayer) {
@@ -1968,7 +_,14 @@
       return false;
    }
 
-   public static enum BedSleepingProblem {
+   public boolean fauxSleeping;
+   public int oldLevel = -1;
+
+    public CraftHumanEntity getBukkitEntity() {
+       return (CraftHumanEntity) super.getBukkitEntity();
+    }
+
+    public static enum BedSleepingProblem {
       NOT_POSSIBLE_HERE,
       NOT_POSSIBLE_NOW(new TranslatableComponent("block.minecraft.bed.no_sleep")),
       TOO_FAR_AWAY(new TranslatableComponent("block.minecraft.bed.too_far_away")),
@@ -1991,5 +_,63 @@
       public Component m_36423_() {
          return this.f_36413_;
       }
+   }
+
+   // =========== FORGE START ==============//
+   public Collection<MutableComponent> getPrefixes() {
+       return this.prefixes;
+   }
+
+   public Collection<MutableComponent> getSuffixes() {
+       return this.suffixes;
+   }
+
+   private Component displayname = null;
+   /**
+    * Force the displayed name to refresh, by firing {@link net.minecraftforge.event.entity.player.PlayerEvent.NameFormat}, using the real player name as event parameter.
+    */
+   public void refreshDisplayName() {
+      this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.m_7755_());
+   }
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerMainHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.PlayerMainInvWrapper(f_36093_));
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerEquipmentHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.CombinedInvWrapper(
+                     new net.minecraftforge.items.wrapper.PlayerArmorInvWrapper(f_36093_),
+                     new net.minecraftforge.items.wrapper.PlayerOffhandInvWrapper(f_36093_)));
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerJoinedHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.PlayerInvWrapper(f_36093_));
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (this.m_6084_() && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+         if (facing == null) return playerJoinedHandler.cast();
+         else if (facing.m_122434_().m_122478_()) return playerMainHandler.cast();
+         else if (facing.m_122434_().m_122479_()) return playerEquipmentHandler.cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   /**
+    * Force a pose for the player. If set, the vanilla pose determination and clearance check is skipped. Make sure the pose is clear yourself (e.g. in PlayerTick).
+    * This has to be set just once, do not set it every tick.
+    * Make sure to clear (null) the pose if not required anymore and only use if necessary.
+    */
+   public void setForcedPose(@Nullable Pose pose) {
+      this.forcedPose = pose;
+   }
+
+   /**
+    * @return The forced pose if set, null otherwise
+    */
+   @Nullable
+   public Pose getForcedPose() {
+      return this.forcedPose;
    }
 }
