--- ../src-base/minecraft/net/minecraft/util/datafix/DataFixesManager.java
+++ ../src-work/minecraft/net/minecraft/util/datafix/DataFixesManager.java
@@ -165,6 +165,7 @@
     public static DataFixer createFixer()
     {
         DataFixer datafixer = new DataFixer(1343);
+        datafixer = new net.minecraftforge.common.util.CompoundDataFixer(datafixer);
         WorldInfo.registerFixes(datafixer);
         EntityPlayerMP.registerFixesPlayerMP(datafixer);
         EntityPlayer.registerFixesPlayer(datafixer);
