--- ../src-base/minecraft/net/minecraft/client/gui/spectator/SpectatorMenu.java
+++ ../src-work/minecraft/net/minecraft/client/gui/spectator/SpectatorMenu.java
@@ -16,10 +16,10 @@
 @SideOnly(Side.CLIENT)
 public class SpectatorMenu
 {
-    private static final ISpectatorMenuObject CLOSE_ITEM = new SpectatorMenu.EndSpectatorObject();
-    private static final ISpectatorMenuObject SCROLL_LEFT = new SpectatorMenu.MoveMenuObject(-1, true);
-    private static final ISpectatorMenuObject SCROLL_RIGHT_ENABLED = new SpectatorMenu.MoveMenuObject(1, true);
-    private static final ISpectatorMenuObject SCROLL_RIGHT_DISABLED = new SpectatorMenu.MoveMenuObject(1, false);
+    private static final ISpectatorMenuObject CLOSE_ITEM = new EndSpectatorObject();
+    private static final ISpectatorMenuObject SCROLL_LEFT = new MoveMenuObject(-1, true);
+    private static final ISpectatorMenuObject SCROLL_RIGHT_ENABLED = new MoveMenuObject(1, true);
+    private static final ISpectatorMenuObject SCROLL_RIGHT_DISABLED = new MoveMenuObject(1, false);
     public static final ISpectatorMenuObject EMPTY_SLOT = new ISpectatorMenuObject()
     {
         public void selectItem(SpectatorMenu menu)
