--- a/net/minecraft/world/level/storage/loot/parameters/LootContextParams.java
+++ b/net/minecraft/world/level/storage/loot/parameters/LootContextParams.java
@@ -20,6 +_,7 @@
    public static final LootContextParam<BlockEntity> f_81462_ = m_81466_("block_entity");
    public static final LootContextParam<ItemStack> f_81463_ = m_81466_("tool");
    public static final LootContextParam<Float> f_81464_ = m_81466_("explosion_radius");
+   public static final LootContextParam<Integer> LOOTING_MOD = new LootContextParam<>(new ResourceLocation("bukkit:looting_mod")); // CraftBukkit
 
    private static <T> LootContextParam<T> m_81466_(String p_81467_) {
       return new LootContextParam<>(new ResourceLocation(p_81467_));
