--- ../src-base/minecraft/net/minecraftforge/fml/common/event/FMLServerStartingEvent.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/event/FMLServerStartingEvent.java
@@ -23,6 +23,7 @@
 import net.minecraft.command.ICommand;
 import net.minecraft.server.MinecraftServer;
 import net.minecraftforge.fml.common.LoaderState.ModState;
+import org.bukkit.command.Command;
 
 /**
  * Called after {@link FMLServerAboutToStartEvent} and before {@link FMLServerStartedEvent}.
