--- a/net/minecraft/server/players/BanListEntry.java
+++ b/net/minecraft/server/players/BanListEntry.java
@@ -25,7 +_,7 @@
    }
 
    protected BanListEntry(T p_10950_, JsonObject p_10951_) {
-      super(p_10950_);
+      super(checkExpiry(p_10950_, p_10951_)); // CraftBukkit
 
       Date date;
       try {
@@ -77,4 +_,21 @@
       p_10959_.addProperty("expires", this.f_10946_ == null ? "forever" : f_10943_.format(this.f_10946_));
       p_10959_.addProperty("reason", this.f_10947_);
    }
+
+   // CraftBukkit start
+   private static <T> T checkExpiry(T object, JsonObject jsonobject) {
+      Date expires = null;
+      try {
+         expires = jsonobject.has("expires") ? f_10943_.parse(jsonobject.get("expires").getAsString()) : null;
+      } catch (ParseException ex) {
+         // Guess we don't have a date
+      }
+
+      if (expires == null || expires.after(new Date())) {
+         return object;
+      } else {
+         return null;
+      }
+   }
+   // CraftBukkit end
 }
