--- a/net/minecraft/server/players/GameProfileCache.java
+++ b/net/minecraft/server/players/GameProfileCache.java
@@ -42,6 +_,7 @@
 import net.minecraft.Util;
 import net.minecraft.world.entity.player.Player;
 import org.slf4j.Logger;
+import org.spigotmc.SpigotConfig;
 
 public class GameProfileCache {
    private static final Logger f_10964_ = LogUtils.getLogger();
@@ -115,7 +_,7 @@
       Date date = calendar.getTime();
       GameProfileCache.GameProfileInfo gameprofilecache$gameprofileinfo = new GameProfileCache.GameProfileInfo(p_10992_, date);
       this.m_10979_(gameprofilecache$gameprofileinfo);
-      this.m_11006_();
+      if( !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly ) this.m_11006_(); // Spigot - skip saving if disabled
    }
 
    private long m_11008_() {
@@ -145,7 +_,7 @@
          }
       }
 
-      if (flag) {
+      if (flag && !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) { // Spigot - skip saving if disabled
          this.m_11006_();
       }
 
@@ -250,7 +_,7 @@
    public void m_11006_() {
       JsonArray jsonarray = new JsonArray();
       DateFormat dateformat = m_11009_();
-      this.m_10977_(1000).forEach((p_143962_) -> {
+      this.m_10977_(SpigotConfig.userCacheCap).forEach((p_143962_) -> { // Spigot
          jsonarray.add(m_10981_(p_143962_, dateformat));
       });
       String s = this.f_10969_.toJson((JsonElement)jsonarray);
