--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -20,6 +_,8 @@
 import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtIo;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.util.StringUtil;
@@ -54,7 +_,7 @@
       }).toArray((p_11070_) -> {
          return new String[p_11070_];
       });
-      if (p_11087_.m_129797_()) {
+      if (p_11087_.m_129797_() || org.spigotmc.SpigotConfig.bungee) { // Spigot: bungee = online mode, for now.
          p_11087_.m_129926_().findProfilesByNames(astring, Agent.MINECRAFT, p_11089_);
       } else {
          for(String s : astring) {
@@ -73,7 +_,7 @@
             try {
                userbanlist.m_11399_();
             } catch (IOException ioexception1) {
-               f_11066_.warn("Could not load existing file {}", userbanlist.m_11385_().getName(), ioexception1);
+               f_11066_.warn("Could not load existing file {}", userbanlist.m_11385_().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -126,7 +_,7 @@
             try {
                ipbanlist.m_11399_();
             } catch (IOException ioexception1) {
-               f_11066_.warn("Could not load existing file {}", ipbanlist.m_11385_().getName(), ioexception1);
+               f_11066_.warn("Could not load existing file {}", ipbanlist.m_11385_().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -162,7 +_,7 @@
             try {
                serveroplist.m_11399_();
             } catch (IOException ioexception1) {
-               f_11066_.warn("Could not load existing file {}", serveroplist.m_11385_().getName(), ioexception1);
+               f_11066_.warn("Could not load existing file {}", serveroplist.m_11385_().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -204,7 +_,7 @@
             try {
                userwhitelist.m_11399_();
             } catch (IOException ioexception1) {
-               f_11066_.warn("Could not load existing file {}", userwhitelist.m_11385_().getName(), ioexception1);
+               f_11066_.warn("Could not load existing file {}", userwhitelist.m_11385_().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -315,6 +_,28 @@
                private void m_11167_(File p_11168_, String p_11169_, String p_11170_) {
                   File file5 = new File(file1, p_11169_ + ".dat");
                   File file6 = new File(p_11168_, p_11170_ + ".dat");
+                  // CraftBukkit start - Use old file name to seed lastKnownName
+                  CompoundTag root = null;
+
+                  try {
+                     root = NbtIo.m_128939_(new java.io.FileInputStream(file5));
+                  } catch (Exception exception) {
+                     exception.printStackTrace();
+                  }
+
+                  if (root != null) {
+                     if (!root.m_128441_("bukkit")) {
+                        root.m_128365_("bukkit", new CompoundTag());
+                     }
+                     CompoundTag data = root.m_128469_("bukkit");
+                     data.m_128359_("lastKnownName", p_11169_);
+                     try {
+                        NbtIo.m_128947_(root, new java.io.FileOutputStream(file2));
+                     } catch (Exception exception) {
+                        exception.printStackTrace();
+                     }
+                  }
+                  // CraftBukkit end
                   OldUsersConverter.m_11093_(p_11168_);
                   if (!file5.renameTo(file6)) {
                      throw new OldUsersConverter.ConversionError("Could not convert file for " + p_11169_);
