--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.players;
 
+import catserver.server.BukkitInjector;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
@@ -23,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.ChatFormatting;
 import net.minecraft.FileUtil;
+import net.minecraft.Util;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.LayeredRegistryAccess;
@@ -102,8 +_,32 @@
 import net.minecraft.world.scores.Objective;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Team;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R1.CraftServer;
+import org.bukkit.craftbukkit.v1_20_R1.CraftWorld;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftChatMessage;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.stream.Collectors;
+import net.minecraft.server.dedicated.DedicatedServer;
+import net.minecraft.server.network.ServerLoginPacketListenerImpl;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.CraftServer;
+import org.bukkit.craftbukkit.v1_20_R1.CraftWorld;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftChatMessage;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation;
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.event.player.PlayerRespawnEvent;
+import org.bukkit.event.player.PlayerRespawnEvent.RespawnReason;
+import org.bukkit.event.player.PlayerSpawnChangeEvent;
+// CraftBukkit end
+
 public abstract class PlayerList {
    public static final File f_11189_ = new File("banned-players.json");
    public static final File f_11190_ = new File("banned-ips.json");
@@ -114,26 +_,36 @@
    private static final int f_143987_ = 600;
    private static final SimpleDateFormat f_11194_ = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
    private final MinecraftServer f_11195_;
-   private final List<ServerPlayer> f_11196_ = Lists.newArrayList();
+   public final List<ServerPlayer> f_11196_ = new java.util.concurrent.CopyOnWriteArrayList<>(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety // CatServer - private -> public
    private final Map<UUID, ServerPlayer> f_11197_ = Maps.newHashMap();
    private final UserBanList f_11198_ = new UserBanList(f_11189_);
    private final IpBanList f_11199_ = new IpBanList(f_11190_);
    private final ServerOpList f_11200_ = new ServerOpList(f_11191_);
    private final UserWhiteList f_11201_ = new UserWhiteList(f_11192_);
+   // CraftBukkit start // CatServer - recover
    private final Map<UUID, ServerStatsCounter> f_11202_ = Maps.newHashMap();
    private final Map<UUID, PlayerAdvancements> f_11203_ = Maps.newHashMap();
-   private final PlayerDataStorage f_11204_;
+   // CraftBukkit end
+   public final PlayerDataStorage f_11204_; // CatServer - private -> public
    private boolean f_11205_;
    private final LayeredRegistryAccess<RegistryLayer> f_243858_;
    private final RegistryAccess.Frozen f_256838_;
-   protected final int f_11193_;
+   public int f_11193_; // CatServer - protected -> public  remove final
    private int f_11207_;
    private int f_184208_;
    private boolean f_11209_;
    private static final boolean f_143988_ = false;
    private int f_11210_;
+   private final List<ServerPlayer> playersView = java.util.Collections.unmodifiableList(f_11196_);
+   // CraftBukkit start
+   private CraftServer cserver;
 
    public PlayerList(MinecraftServer p_203842_, LayeredRegistryAccess<RegistryLayer> p_251844_, PlayerDataStorage p_203844_, int p_203845_) {
+      this.cserver = p_203842_.server = new CraftServer((DedicatedServer) p_203842_, this);
+      p_203842_.console = org.bukkit.craftbukkit.v1_20_R1.command.ColouredConsoleSender.getInstance();
+      // p_203842_.reader.addCompleter(new org.bukkit.craftbukkit.v1_20_R1.command.ConsoleCommandCompleter(p_203842_.server)); // CatServer - remove
+      BukkitInjector.registerAll(); // CatServer
+      // CraftBukkit end
       this.f_11195_ = p_203842_;
       this.f_243858_ = p_251844_;
       this.f_256838_ = (new RegistryAccess.ImmutableRegistryAccess(RegistrySynchronization.m_257599_(p_251844_))).m_203557_();
@@ -154,6 +_,12 @@
       }
 
       CompoundTag compoundtag = this.m_11224_(p_11263_);
+      // CraftBukkit start - Better rename detection
+      if (compoundtag != null && compoundtag.m_128441_("bukkit")) {
+         CompoundTag bukkit = compoundtag.m_128469_("bukkit");
+         s = bukkit.m_128425_("lastKnownName", 8) ? bukkit.m_128461_("lastKnownName") : s;
+      }
+      // CraftBukkit end
       ResourceKey<Level> resourcekey = compoundtag != null ? DimensionType.m_63911_(new Dynamic<>(NbtOps.f_128958_, compoundtag.m_128423_("Dimension"))).resultOrPartial(f_11188_::error).orElse(Level.f_46428_) : Level.f_46428_;
       ServerLevel serverlevel = this.f_11195_.m_129880_(resourcekey);
       ServerLevel serverlevel1;
@@ -167,22 +_,39 @@
       p_11263_.m_284127_(serverlevel1);
       String s1 = "local";
       if (p_11262_.m_129523_() != null) {
-         s1 = p_11262_.m_129523_().toString();
+         s1 = net.minecraftforge.network.DualStackUtils.getAddressString(p_11262_.m_129523_());
       }
 
-      f_11188_.info("{}[{}] logged in with entity id {} at ({}, {}, {})", p_11263_.m_7755_().getString(), s1, p_11263_.m_19879_(), p_11263_.m_20185_(), p_11263_.m_20186_(), p_11263_.m_20189_());
+      // Spigot start - spawn location event
+      org.bukkit.entity.Player spawnPlayer = p_11263_.getBukkitEntity();
+      org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new org.spigotmc.event.player.PlayerSpawnLocationEvent(spawnPlayer, spawnPlayer.getLocation());
+      cserver.getPluginManager().callEvent(ev);
+      Location loc = ev.getSpawnLocation();
+      serverlevel1 = ((CraftWorld) loc.getWorld()).getHandle();
+
+      p_11263_.spawnIn(serverlevel1);
+      p_11263_.f_8941_.m_9260_((ServerLevel) p_11263_.m_9236_());
+      p_11263_.m_19890_(loc.getX(), loc.getY(), loc.getZ(), loc.getYaw(), loc.getPitch());
+      // Spigot end
+
+      // CraftBukkit - Moved message to after join
+      // LOGGER.info("{}[{}] logged in with entity id {} at ({}, {}, {})", p_11263_.getName().getString(), s1, p_11263_.getId(), p_11263_.getX(), p_11263_.getY(), p_11263_.getZ());
       LevelData leveldata = serverlevel1.m_6106_();
       p_11263_.m_143427_(compoundtag);
       ServerGamePacketListenerImpl servergamepacketlistenerimpl = new ServerGamePacketListenerImpl(this.f_11195_, p_11262_, p_11263_);
+      net.minecraftforge.network.NetworkHooks.sendMCRegistryPackets(p_11262_, "PLAY_TO_CLIENT");
       GameRules gamerules = serverlevel1.m_46469_();
       boolean flag = gamerules.m_46207_(GameRules.f_46156_);
       boolean flag1 = gamerules.m_46207_(GameRules.f_46145_);
-      servergamepacketlistenerimpl.m_9829_(new ClientboundLoginPacket(p_11263_.m_19879_(), leveldata.m_5466_(), p_11263_.f_8941_.m_9290_(), p_11263_.f_8941_.m_9293_(), this.f_11195_.m_129784_(), this.f_256838_, serverlevel1.m_220362_(), serverlevel1.m_46472_(), BiomeManager.m_47877_(serverlevel1.m_7328_()), this.m_11310_(), this.f_11207_, this.f_184208_, flag1, !flag, serverlevel1.m_46659_(), serverlevel1.m_8584_(), p_11263_.m_219759_(), p_11263_.m_287157_()));
+      // Spigot - view distance
+      servergamepacketlistenerimpl.m_9829_(new ClientboundLoginPacket(p_11263_.m_19879_(), leveldata.m_5466_(), p_11263_.f_8941_.m_9290_(), p_11263_.f_8941_.m_9293_(), this.f_11195_.m_129784_(), this.f_256838_, serverlevel1.m_220362_(), serverlevel1.m_46472_(), BiomeManager.m_47877_(serverlevel1.m_7328_()), this.m_11310_(), serverlevel1.spigotConfig.viewDistance, serverlevel1.spigotConfig.simulationDistance, flag1, !flag, serverlevel1.m_46659_(), serverlevel1.m_8584_(), p_11263_.m_219759_(), p_11263_.m_287157_()));
+      p_11263_.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
       servergamepacketlistenerimpl.m_9829_(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.f_244280_.m_245829_(serverlevel1.m_246046_())));
       servergamepacketlistenerimpl.m_9829_(new ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.f_132012_, (new FriendlyByteBuf(Unpooled.buffer())).m_130070_(this.m_7873_().getServerModName())));
       servergamepacketlistenerimpl.m_9829_(new ClientboundChangeDifficultyPacket(leveldata.m_5472_(), leveldata.m_5474_()));
       servergamepacketlistenerimpl.m_9829_(new ClientboundPlayerAbilitiesPacket(p_11263_.m_150110_()));
       servergamepacketlistenerimpl.m_9829_(new ClientboundSetCarriedItemPacket(p_11263_.m_150109_().f_35977_));
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.OnDatapackSyncEvent(this, p_11263_));
       servergamepacketlistenerimpl.m_9829_(new ClientboundUpdateRecipesPacket(this.f_11195_.m_129894_().m_44051_()));
       servergamepacketlistenerimpl.m_9829_(new ClientboundUpdateTagsPacket(TagNetworkSerialization.m_245799_(this.f_243858_)));
       this.m_11289_(p_11263_);
@@ -196,21 +_,70 @@
       } else {
          mutablecomponent = Component.m_237110_("multiplayer.player.joined.renamed", p_11263_.m_5446_(), s);
       }
+      // CraftBukkit start
+      mutablecomponent.m_130940_(ChatFormatting.YELLOW);
+      String joinMessage = CraftChatMessage.fromComponent(mutablecomponent);
 
-      this.m_240416_(mutablecomponent.m_130940_(ChatFormatting.YELLOW), false);
       servergamepacketlistenerimpl.m_9774_(p_11263_.m_20185_(), p_11263_.m_20186_(), p_11263_.m_20189_(), p_11263_.m_146908_(), p_11263_.m_146909_());
       ServerStatus serverstatus = this.f_11195_.m_129928_();
       if (serverstatus != null) {
          p_11263_.m_215109_(serverstatus);
       }
 
-      p_11263_.f_8906_.m_9829_(ClientboundPlayerInfoUpdatePacket.m_247122_(this.f_11196_));
+      // p_11263_.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(this.players)); // CraftBukkit - 1.20.1 replaced with loop below
       this.f_11196_.add(p_11263_);
       this.f_11197_.put(p_11263_.m_20148_(), p_11263_);
-      this.m_11268_(ClientboundPlayerInfoUpdatePacket.m_247122_(List.of(p_11263_)));
+      // this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(p_11263_))); // CraftBukkit - replaced with loop below
+
+      // CraftBukkit start
+      CraftPlayer bukkitPlayer = p_11263_.getBukkitEntity();
+
+      // Ensure that player inventory is populated with its viewer
+      p_11263_.f_36096_.transferTo(p_11263_.f_36096_, bukkitPlayer);
+
+      PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, joinMessage);
+      cserver.getPluginManager().callEvent(playerJoinEvent);
+
+      if (!p_11263_.f_8906_.m_6198_()) {
+         return;
+      }
+
+      joinMessage = playerJoinEvent.getJoinMessage();
+
+      if (joinMessage != null && joinMessage.length() > 0) {
+         for (Component line : org.bukkit.craftbukkit.v1_20_R1.util.CraftChatMessage.fromString(joinMessage)) {
+            f_11195_.m_6846_().m_240416_(line, false);
+         }
+      }
+      // CraftBukkit end
+
+      // CraftBukkit start - sendAll above replaced with this loop
+      ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.m_247122_(List.of(p_11263_));
+
+      for (int i = 0; i < this.f_11196_.size(); ++i) {
+         ServerPlayer entityplayer1 = this.f_11196_.get(i);
+         if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer)) {
+            entityplayer1.f_8906_.m_9829_(packet);
+         }
+         if (!bukkitPlayer.canSee(entityplayer1.getBukkitEntity())) {
+            continue;
+         }
+         p_11263_.f_8906_.m_9829_(ClientboundPlayerInfoUpdatePacket.m_247122_(List.of(entityplayer1)));
+      }
+      p_11263_.sentListPacket = true;
+      // CraftBukkit end
+
+      p_11263_.m_20088_().refresh(p_11263_); // CraftBukkit - 1.20.1 BungeeCord#2321, send complete data to self on spawn
+
       this.m_11229_(p_11263_, serverlevel1);
-      serverlevel1.m_8834_(p_11263_);
-      this.f_11195_.m_129901_().m_136293_(p_11263_);
+
+      // CraftBukkit start - Only add if the player wasn't moved in the event
+      if (p_11263_.m_9236_() == serverlevel1 && !serverlevel1.m_6907_().contains(p_11263_)) {
+         serverlevel1.m_8834_(p_11263_);
+         this.f_11195_.m_129901_().m_136293_(p_11263_);
+      }
+      serverlevel1 = p_11263_.m_284548_(); // CraftBukkit - Update in case join event changed it
+      // CraftBukkit end
       this.f_11195_.m_214042_().ifPresent((p_215606_) -> {
          p_11263_.m_143408_(p_215606_.f_236743_(), p_215606_.f_236744_(), p_215606_.f_236745_(), p_215606_.f_236746_());
       });
@@ -221,8 +_,11 @@
 
       if (compoundtag != null && compoundtag.m_128425_("RootVehicle", 10)) {
          CompoundTag compoundtag1 = compoundtag.m_128469_("RootVehicle");
-         Entity entity1 = EntityType.m_20645_(compoundtag1.m_128469_("Entity"), serverlevel1, (p_215603_) -> {
-            return !serverlevel1.m_8847_(p_215603_) ? null : p_215603_;
+         // CraftBukkit start
+         ServerLevel finalServerLevel1 = serverlevel1;
+         Entity entity1 = EntityType.m_20645_(compoundtag1.m_128469_("Entity"), finalServerLevel1, (p_215603_) -> {
+            return !finalServerLevel1.m_8847_(p_215603_) ? null : p_215603_;
+            // CraftBukkit end
          });
          if (entity1 != null) {
             UUID uuid;
@@ -255,9 +_,12 @@
       }
 
       p_11263_.m_143429_();
+      net.minecraftforge.event.ForgeEventFactory.firePlayerLoggedIn( p_11263_ );
+      // CraftBukkit - Moved from above, added world
+      PlayerList.f_11188_.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", p_11263_.m_7755_().getString(), s1, p_11263_.m_19879_(), serverlevel1.f_8549_.m_5462_(), p_11263_.m_20185_(), p_11263_.m_20186_(), p_11263_.m_20189_());
    }
 
-   protected void m_11273_(ServerScoreboard p_11274_, ServerPlayer p_11275_) {
+   public void m_11273_(ServerScoreboard p_11274_, ServerPlayer p_11275_) { // CatServer - protected -> public
       Set<Objective> set = Sets.newHashSet();
 
       for(PlayerTeam playerteam : p_11274_.m_83491_()) {
@@ -278,25 +_,26 @@
    }
 
    public void m_184209_(ServerLevel p_184210_) {
+      if (f_11204_ != null) return; // CraftBukkit
       p_184210_.m_6857_().m_61929_(new BorderChangeListener() {
          public void m_6312_(WorldBorder p_11321_, double p_11322_) {
-            PlayerList.this.m_11268_(new ClientboundSetBorderSizePacket(p_11321_));
+            PlayerList.this.broadcastAll(new ClientboundSetBorderSizePacket(p_11321_), p_11321_.world); // CraftBukkit
          }
 
          public void m_6689_(WorldBorder p_11328_, double p_11329_, double p_11330_, long p_11331_) {
-            PlayerList.this.m_11268_(new ClientboundSetBorderLerpSizePacket(p_11328_));
+            PlayerList.this.broadcastAll(new ClientboundSetBorderLerpSizePacket(p_11328_), p_11328_.world); // CraftBukkit
          }
 
          public void m_7721_(WorldBorder p_11324_, double p_11325_, double p_11326_) {
-            PlayerList.this.m_11268_(new ClientboundSetBorderCenterPacket(p_11324_));
+            PlayerList.this.broadcastAll(new ClientboundSetBorderCenterPacket(p_11324_), p_11324_.world); // CraftBukkit
          }
 
          public void m_5904_(WorldBorder p_11333_, int p_11334_) {
-            PlayerList.this.m_11268_(new ClientboundSetBorderWarningDelayPacket(p_11333_));
+            PlayerList.this.broadcastAll(new ClientboundSetBorderWarningDelayPacket(p_11333_), p_11333_.world); // CraftBukkit
          }
 
          public void m_5903_(WorldBorder p_11339_, int p_11340_) {
-            PlayerList.this.m_11268_(new ClientboundSetBorderWarningDistancePacket(p_11339_));
+            PlayerList.this.broadcastAll(new ClientboundSetBorderWarningDistancePacket(p_11339_), p_11339_.world); // CraftBukkit
          }
 
          public void m_6315_(WorldBorder p_11336_, double p_11337_) {
@@ -315,6 +_,7 @@
          compoundtag1 = compoundtag;
          p_11225_.m_20258_(compoundtag);
          f_11188_.debug("loading single player");
+         net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(p_11225_, this.f_11204_, p_11225_.m_20148_().toString());
       } else {
          compoundtag1 = this.f_11204_.m_78435_(p_11225_);
       }
@@ -323,6 +_,8 @@
    }
 
    protected void m_6765_(ServerPlayer p_11277_) {
+      if (p_11277_.f_8906_ == null) return;
+      if (!p_11277_.getBukkitEntity().isPersistent()) return; // CraftBukkit
       this.f_11204_.m_78433_(p_11277_);
       ServerStatsCounter serverstatscounter = this.f_11202_.get(p_11277_.m_20148_());
       if (serverstatscounter != null) {
@@ -336,9 +_,30 @@
 
    }
 
+   // CatServer start
+   private @Nullable String quitMessage;
+   public String removeCb(ServerPlayer p_11287_) { // CraftBukkit - return string
+      this.m_11286_(p_11287_);
+      var quitText = this.quitMessage;
+      this.quitMessage = null;
+      return quitText;
+   }
+
    public void m_11286_(ServerPlayer p_11287_) {
+      this.quitMessage = null; // CatServer - reset
+      net.minecraftforge.event.ForgeEventFactory.firePlayerLoggedOut(p_11287_);
       ServerLevel serverlevel = p_11287_.m_284548_();
       p_11287_.m_36220_(Stats.f_12989_);
+      // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
+      // See SPIGOT-5799, SPIGOT-6145
+      if (p_11287_.f_36096_ != p_11287_.f_36095_) {
+         p_11287_.m_6915_();
+      }
+      PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(p_11287_.getBukkitEntity(), p_11287_.kickLeaveMessage != null ? p_11287_.kickLeaveMessage : "\u00A7e" + p_11287_.m_6302_() + " left the game");
+      cserver.getPluginManager().callEvent(playerQuitEvent);
+      p_11287_.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
+      // p_11287_.doTick(); // SPIGOT-924 // CatServer - Dont doTick here.
+      // CraftBukkit end
       this.m_6765_(p_11287_);
       if (p_11287_.m_20159_()) {
          Entity entity = p_11287_.m_20201_();
@@ -360,79 +_,229 @@
       ServerPlayer serverplayer = this.f_11197_.get(uuid);
       if (serverplayer == p_11287_) {
          this.f_11197_.remove(uuid);
+         // CraftBukkit start // CatServer - recover
          this.f_11202_.remove(uuid);
          this.f_11203_.remove(uuid);
+         // CraftBukkit end
       }
 
-      this.m_11268_(new ClientboundPlayerInfoRemovePacket(List.of(p_11287_.m_20148_())));
+      // CraftBukkit start
+      // this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(p_11287_.getUUID())));
+      ClientboundPlayerInfoRemovePacket packet = new ClientboundPlayerInfoRemovePacket(List.of(p_11287_.m_20148_()));
+      for (int i = 0; i < f_11196_.size(); i++) {
+         ServerPlayer entityplayer2 = (ServerPlayer) this.f_11196_.get(i);
+         if (entityplayer2.getBukkitEntity().canSee(p_11287_.getBukkitEntity())) {
+            entityplayer2.f_8906_.m_9829_(packet);
+         } else {
+            entityplayer2.getBukkitEntity().onEntityRemove(p_11287_);
+         }
+      }
+      // This removes the scoreboard (and player reference) for the specific player in the manager
+      cserver.getScoreboardManager().removePlayer(p_11287_.getBukkitEntity());
+      // CraftBukkit end
+      this.quitMessage = playerQuitEvent.getQuitMessage(); // CraftBukkit
    }
+   // CatServer end
 
    @Nullable
-   public Component m_6418_(SocketAddress p_11257_, GameProfile p_11258_) {
-      if (this.f_11198_.m_11406_(p_11258_)) {
+   // CraftBukkit start - Whole method, SocketAddress to LoginListener, added hostname to signature, return ServerPlayer
+   public ServerPlayer canPlayerLogin(ServerLoginPacketListenerImpl serverLoginPacketListener, GameProfile p_11258_) {
+      MutableComponent ichatmutablecomponent;
+      // Moved from processLogin
+      UUID uuid = UUIDUtil.m_235875_(p_11258_);
+      List<ServerPlayer> list = Lists.newArrayList();
+      for (ServerPlayer value : this.f_11196_) {
+         if (value.m_20148_().equals(uuid)) {
+            list.add(value);
+         }
+      }
+      for (ServerPlayer serverPlayer : list) {
+         m_6765_(serverPlayer); // CraftBukkit - Force the player's inventory to be saved
+         serverPlayer.f_8906_.m_9942_(Component.m_237115_("multiplayer.disconnect.duplicate_login"));
+      }
+      // Instead of kicking then returning, we need to store the kick reason
+      // in the event, check with plugins to see if it's ok, and THEN kick
+      // depending on the outcome.
+      SocketAddress socketaddress = serverLoginPacketListener.f_10013_.m_129523_();
+      ServerPlayer entity = new ServerPlayer(this.f_11195_, this.f_11195_.m_129880_(Level.f_46428_), p_11258_);
+      org.bukkit.entity.Player player = entity.getBukkitEntity();
+      PlayerLoginEvent event = new PlayerLoginEvent(player, serverLoginPacketListener.f_10013_.hostname, ((java.net.InetSocketAddress) socketaddress).getAddress(), ((java.net.InetSocketAddress) serverLoginPacketListener.f_10013_.f_129468_.remoteAddress()).getAddress());
+
+      if (m_11295_().m_11406_(p_11258_) && !m_11295_().m_11388_(p_11258_).m_7524_()) {
          UserBanListEntry userbanlistentry = this.f_11198_.m_11388_(p_11258_);
-         MutableComponent mutablecomponent1 = Component.m_237110_("multiplayer.disconnect.banned.reason", userbanlistentry.m_10962_());
+         ichatmutablecomponent = Component.m_237110_("multiplayer.disconnect.banned.reason", userbanlistentry.m_10962_());
          if (userbanlistentry.m_10961_() != null) {
-            mutablecomponent1.m_7220_(Component.m_237110_("multiplayer.disconnect.banned.expiration", f_11194_.format(userbanlistentry.m_10961_())));
+            ichatmutablecomponent.m_7220_(Component.m_237110_("multiplayer.disconnect.banned.expiration", f_11194_.format(userbanlistentry.m_10961_())));
          }
 
-         return mutablecomponent1;
+         // return ichatmutablecomponent;
+         event.disallow(PlayerLoginEvent.Result.KICK_BANNED, CraftChatMessage.fromComponent(ichatmutablecomponent));
       } else if (!this.m_5764_(p_11258_)) {
-         return Component.m_237115_("multiplayer.disconnect.not_whitelisted");
-      } else if (this.f_11199_.m_11041_(p_11257_)) {
-         IpBanListEntry ipbanlistentry = this.f_11199_.m_11043_(p_11257_);
-         MutableComponent mutablecomponent = Component.m_237110_("multiplayer.disconnect.banned_ip.reason", ipbanlistentry.m_10962_());
+         ichatmutablecomponent = Component.m_237115_("multiplayer.disconnect.not_whitelisted");
+         event.disallow(PlayerLoginEvent.Result.KICK_WHITELIST, org.spigotmc.SpigotConfig.whitelistMessage); // Spigot
+      } else if (m_11299_().m_11041_(socketaddress) && !m_11299_().m_11043_(socketaddress).m_7524_()) {
+         IpBanListEntry ipbanlistentry = this.f_11199_.m_11043_(socketaddress);
+         ichatmutablecomponent = Component.m_237110_("multiplayer.disconnect.banned_ip.reason", ipbanlistentry.m_10962_());
          if (ipbanlistentry.m_10961_() != null) {
-            mutablecomponent.m_7220_(Component.m_237110_("multiplayer.disconnect.banned_ip.expiration", f_11194_.format(ipbanlistentry.m_10961_())));
+            ichatmutablecomponent.m_7220_(Component.m_237110_("multiplayer.disconnect.banned_ip.expiration", f_11194_.format(ipbanlistentry.m_10961_())));
          }
 
-         return mutablecomponent;
+         // return mutablecomponent;
+         event.disallow(PlayerLoginEvent.Result.KICK_BANNED, CraftChatMessage.fromComponent(ichatmutablecomponent));
       } else {
-         return this.f_11196_.size() >= this.f_11193_ && !this.m_5765_(p_11258_) ? Component.m_237115_("multiplayer.disconnect.server_full") : null;
-      }
+         // return this.players.size() >= this.maxPlayers && !this.canBypassPlayerLimit(p_11258_) ? Component.translatable("multiplayer.disconnect.server_full") : null;
+         if (this.f_11196_.size() >= this.f_11193_ && !this.m_5765_(p_11258_)) {
+            event.disallow(PlayerLoginEvent.Result.KICK_FULL, org.spigotmc.SpigotConfig.serverFullMessage); // Spigot
+         }
+      }
+
+      cserver.getPluginManager().callEvent(event);
+      if (event.getResult() != PlayerLoginEvent.Result.ALLOWED) {
+         serverLoginPacketListener.disconnect(event.getKickMessage());
+         return null;
+      }
+      return entity;
    }
 
-   public ServerPlayer m_215624_(GameProfile p_215625_) {
-      UUID uuid = UUIDUtil.m_235875_(p_215625_);
+   public ServerPlayer getPlayerForLogin(GameProfile p_215625_, ServerPlayer player) { // CraftBukkit - added EntityPlayer
+      /* CraftBukkit start - Moved up
+      UUID uuid = UUIDUtil.getOrCreatePlayerUUID(p_215625_);
       List<ServerPlayer> list = Lists.newArrayList();
 
-      for(int i = 0; i < this.f_11196_.size(); ++i) {
-         ServerPlayer serverplayer = this.f_11196_.get(i);
-         if (serverplayer.m_20148_().equals(uuid)) {
+      for(int i = 0; i < this.players.size(); ++i) {
+         ServerPlayer serverplayer = this.players.get(i);
+         if (serverplayer.getUUID().equals(uuid)) {
             list.add(serverplayer);
          }
       }
 
-      ServerPlayer serverplayer2 = this.f_11197_.get(p_215625_.getId());
+      ServerPlayer serverplayer2 = this.playersByUUID.get(p_215625_.getId());
       if (serverplayer2 != null && !list.contains(serverplayer2)) {
          list.add(serverplayer2);
       }
 
       for(ServerPlayer serverplayer1 : list) {
-         serverplayer1.f_8906_.m_9942_(Component.m_237115_("multiplayer.disconnect.duplicate_login"));
-      }
-
-      return new ServerPlayer(this.f_11195_, this.f_11195_.m_129783_(), p_215625_);
+         serverplayer1.connection.disconnect(Component.translatable("multiplayer.disconnect.duplicate_login"));
+      }
+
+      return new ServerPlayer(this.server, this.server.overworld(), p_215625_);
+      */
+      return player;
+      // CraftBukkit end
+   }
+
+   // CatServer start - respawn logic
+   public ServerPlayer respawn(ServerPlayer p_11237_, ServerLevel worldServer, boolean p_11238_, Location location, boolean avoidSuffocation, RespawnReason reason) {
+      try {
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$worldServer.push(this, worldServer);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$location.push(this, location);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$avoidSuffocation.push(this, avoidSuffocation);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$reason.push(this, reason);
+
+         return m_11236_(p_11237_, p_11238_);
+      } finally {
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$worldServer.pop(this);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$location.pop(this);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$avoidSuffocation.pop(this);
+         catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$reason.pop(this);
+      }
+   }
+
+   public ServerPlayer respawn(ServerPlayer entityplayer, boolean p_11238_, RespawnReason reason) { // CatServer - 1.20.1
+      return this.respawn(entityplayer, this.f_11195_.m_129880_(entityplayer.m_8963_()), p_11238_, null, true, reason);
    }
 
    public ServerPlayer m_11236_(ServerPlayer p_11237_, boolean p_11238_) {
+      ServerLevel respawnLevel = catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$worldServer.get(this);
+      boolean avoidSuffocation = catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$avoidSuffocation.get(this);
+      Location location = catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$location.get(this);
+      RespawnReason reason = catserver.server.CatServerCaptures.getCatServerCaptures().PlayerList$respawn$reason.get(this);
+
+      // CraftBukkit start // CatServer
+      if (respawnLevel == null) {
+         respawnLevel = this.f_11195_.m_129880_(p_11237_.m_8963_());
+      }
+
+      p_11237_.m_8127_(); // CraftBukkit
+
       this.f_11196_.remove(p_11237_);
       p_11237_.m_284548_().m_143261_(p_11237_, Entity.RemovalReason.DISCARDED);
       BlockPos blockpos = p_11237_.m_8961_();
       float f = p_11237_.m_8962_();
       boolean flag = p_11237_.m_8964_();
+
+      org.bukkit.World fromWorld = p_11237_.getBukkitEntity().getWorld();
+      p_11237_.f_8944_ = false;
+      // CraftBukkit end
+
+      boolean flag3 = false;
       ServerLevel serverlevel = this.f_11195_.m_129880_(p_11237_.m_8963_());
-      Optional<Vec3> optional;
-      if (serverlevel != null && blockpos != null) {
-         optional = Player.m_36130_(serverlevel, blockpos, f, flag, p_11238_);
+
+      // CraftBukkit start - fire PlayerRespawnEvent
+      if (location == null) {
+         boolean isBedSpawn = false;
+         if (serverlevel != null) {
+            Optional<Vec3> optional;
+            if (blockpos != null) {
+               optional = net.minecraft.world.entity.player.Player.m_36130_(serverlevel, blockpos, f, flag, p_11238_);
+            } else {
+               optional = Optional.empty();
+            }
+            if (optional.isPresent()) {
+               BlockState blockState = serverlevel.m_8055_(blockpos);
+               boolean flag4 = blockState.m_60713_(Blocks.f_50724_);
+
+               Vec3 vec3 = optional.get();
+               float f2;
+
+               if (!blockState.m_204336_(BlockTags.f_13038_) && !flag4) {
+                  f2 = f;
+               } else {
+                  Vec3 vec31 = Vec3.m_82539_(blockpos).m_82546_(vec3).m_82541_();
+                  f2 = (float) Mth.m_14175_(Mth.m_14136_(vec31.f_82481_, vec31.f_82479_) * 57.2957763671875 - 90.0);
+               }
+               // p_11237_.setRespawnPosition(serverlevel.dimension(), blockpos, f2, flag, false); // CraftBukkit - 1.20.1 not required, just copies old location into reused entity
+               flag3 = (!flag && flag4);
+               isBedSpawn = true;
+               location = CraftLocation.toBukkit(vec3, serverlevel.getWorld(), f2, 0.0F);
+            } else if (blockpos != null) {
+               p_11237_.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132153_, 0.0f));
+               p_11237_.setRespawnPosition(Level.f_46428_, null, 0f, false, false, PlayerSpawnChangeEvent.Cause.RESET); // CraftBukkit - SPIGOT-5988: Clear respawn location when obstructed
+            }
+         }
+         if (location == null) {
+            serverlevel = this.f_11195_.m_129880_(Level.f_46428_);
+            assert serverlevel != null; // Why might be null ?
+            blockpos = p_11237_.getSpawnPoint(serverlevel);
+            location = CraftLocation.toBukkit(blockpos, serverlevel.getWorld()).add(0.5f, 0.1f, 0.5f);
+         }
+         org.bukkit.entity.Player respawnPlayer = p_11237_.getBukkitEntity();
+         PlayerRespawnEvent respawnEvent = new PlayerRespawnEvent(respawnPlayer, location, isBedSpawn && !flag3, flag3, reason);
+         this.cserver.getPluginManager().callEvent(respawnEvent);
+         if (p_11237_.f_8906_.isDisconnected()) return p_11237_; // Spigot
+         location = respawnEvent.getRespawnLocation();
+         if (!p_11238_) p_11237_.reset();
       } else {
-         optional = Optional.empty();
+         location.setWorld(respawnLevel.getWorld());
       }
 
-      ServerLevel serverlevel1 = serverlevel != null && optional.isPresent() ? serverlevel : this.f_11195_.m_129783_();
-      ServerPlayer serverplayer = new ServerPlayer(this.f_11195_, serverlevel1, p_11237_.m_36316_());
+      ServerLevel serverlevel1 = ((CraftWorld) location.getWorld()).getHandle();
+      ServerPlayer serverplayer = new ServerPlayer(this.f_11195_, serverlevel1, p_11237_.m_36316_()); // CatServer - use forge logic: create new ServerPlayer
+      // CraftBukkit end
+
+      // Forward
+      p_11237_.getBukkitEntity().setHandle(serverplayer);
+      serverplayer.setBukkitEntity(p_11237_.getBukkitEntity());
+      p_11237_.f_8906_.f_9743_ = serverplayer;
+
       serverplayer.f_8906_ = p_11237_.f_8906_;
       serverplayer.m_9015_(p_11237_, p_11238_);
+      serverplayer.m_9158_(p_11237_.m_8963_(), p_11237_.m_8961_(), p_11237_.m_8962_(), p_11237_.m_8964_(), false);
+
+//      if (!p_11238_) {
+//         serverplayer.getInventory().replaceWith(p_11237_.getInventory()); // Keep inventory // Called in each if case in ServerPlayerEntity#restoreFrom
+//      }
+
       serverplayer.m_20234_(p_11237_.m_19879_());
       serverplayer.m_36163_(p_11237_.m_5737_());
 
@@ -440,50 +_,71 @@
          serverplayer.m_20049_(s);
       }
 
-      boolean flag2 = false;
-      if (optional.isPresent()) {
-         BlockState blockstate = serverlevel1.m_8055_(blockpos);
-         boolean flag1 = blockstate.m_60713_(Blocks.f_50724_);
-         Vec3 vec3 = optional.get();
-         float f1;
-         if (!blockstate.m_204336_(BlockTags.f_13038_) && !flag1) {
-            f1 = f;
-         } else {
-            Vec3 vec31 = Vec3.m_82539_(blockpos).m_82546_(vec3).m_82541_();
-            f1 = (float)Mth.m_14175_(Mth.m_14136_(vec31.f_82481_, vec31.f_82479_) * (double)(180F / (float)Math.PI) - 90.0D);
-         }
-
-         serverplayer.m_7678_(vec3.f_82479_, vec3.f_82480_, vec3.f_82481_, f1, 0.0F);
-         serverplayer.m_9158_(serverlevel1.m_46472_(), blockpos, f, flag, false);
-         flag2 = !p_11238_ && flag1;
-      } else if (blockpos != null) {
-         serverplayer.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132153_, 0.0F));
-      }
-
-      while(!serverlevel1.m_45786_(serverplayer) && serverplayer.m_20186_() < (double)serverlevel1.m_151558_()) {
+      serverplayer.m_7678_(location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch());
+      serverplayer.f_8906_.m_9953_();
+
+      while (avoidSuffocation && !serverlevel1.m_45786_(serverplayer) && serverplayer.m_20186_() < serverlevel1.m_151558_()) {
          serverplayer.m_6034_(serverplayer.m_20185_(), serverplayer.m_20186_() + 1.0D, serverplayer.m_20189_());
       }
 
-      byte b0 = (byte)(p_11238_ ? 1 : 0);
-      LevelData leveldata = serverplayer.m_9236_().m_6106_();
-      serverplayer.f_8906_.m_9829_(new ClientboundRespawnPacket(serverplayer.m_9236_().m_220362_(), serverplayer.m_9236_().m_46472_(), BiomeManager.m_47877_(serverplayer.m_284548_().m_7328_()), serverplayer.f_8941_.m_9290_(), serverplayer.f_8941_.m_9293_(), serverplayer.m_9236_().m_46659_(), serverplayer.m_284548_().m_8584_(), b0, serverplayer.m_219759_(), serverplayer.m_287157_()));
-      serverplayer.f_8906_.m_9774_(serverplayer.m_20185_(), serverplayer.m_20186_(), serverplayer.m_20189_(), serverplayer.m_146908_(), serverplayer.m_146909_());
+      byte b0 = (byte)(flag ? 1 : 0);
+      // CraftBukkit start
+      LevelData leveldata = serverplayer.m_9236_().m_6106_(); // CatServer - recover
+      serverplayer.f_8906_.m_9829_(new ClientboundRespawnPacket(serverplayer.m_9236_().m_220362_(), serverplayer.m_9236_().m_46472_(), BiomeManager.m_47877_(serverplayer.m_284548_().m_7328_()), serverplayer.f_8941_.m_9290_(), serverplayer.f_8941_.m_9293_(), serverplayer.m_284548_().m_46659_(), serverplayer.m_284548_().m_8584_(), b0, serverplayer.m_219759_(), serverplayer.m_287157_()));
+      serverplayer.f_8906_.m_9829_(new ClientboundSetChunkCacheRadiusPacket(serverplayer.m_9236_().spigotConfig.viewDistance));
+      serverplayer.f_8906_.m_9829_(new ClientboundSetSimulationDistancePacket(serverplayer.m_9236_().spigotConfig.simulationDistance));
+
+      serverplayer.m_284535_(serverlevel1);
+      serverplayer.f_8906_.teleport(CraftLocation.toBukkit(serverplayer.m_20182_(), serverlevel.getWorld(), serverplayer.m_146908_(), serverplayer.m_146909_()));
+      serverplayer.m_20260_(false);
+
+      // serverplayer.connection.teleport(serverplayer.getX(), serverplayer.getY(), serverplayer.getZ(), serverplayer.getYRot(), serverplayer.getXRot());
       serverplayer.f_8906_.m_9829_(new ClientboundSetDefaultSpawnPositionPacket(serverlevel1.m_220360_(), serverlevel1.m_220361_()));
       serverplayer.f_8906_.m_9829_(new ClientboundChangeDifficultyPacket(leveldata.m_5472_(), leveldata.m_5474_()));
       serverplayer.f_8906_.m_9829_(new ClientboundSetExperiencePacket(serverplayer.f_36080_, serverplayer.f_36079_, serverplayer.f_36078_));
+
       this.m_11229_(serverplayer, serverlevel1);
       this.m_11289_(serverplayer);
-      serverlevel1.m_8845_(serverplayer);
-      this.f_11196_.add(serverplayer);
-      this.f_11197_.put(serverplayer.m_20148_(), serverplayer);
-      serverplayer.m_143429_();
+
+      if (!serverplayer.f_8906_.isDisconnected()) {
+         serverlevel1.m_8845_(serverplayer);
+         this.f_11196_.add(serverplayer);
+         this.f_11197_.put(serverplayer.m_20148_(), serverplayer);
+      }
+      serverplayer.m_143429_(); // CatServer - recover
       serverplayer.m_21153_(serverplayer.m_21223_());
-      if (flag2) {
+
+      net.minecraftforge.event.ForgeEventFactory.firePlayerRespawnEvent(serverplayer, p_11238_);
+
+      if (flag) {
          serverplayer.f_8906_.m_9829_(new ClientboundSoundPacket(SoundEvents.f_12377_, SoundSource.BLOCKS, (double)blockpos.m_123341_(), (double)blockpos.m_123342_(), (double)blockpos.m_123343_(), 1.0F, 1.0F, serverlevel1.m_213780_().m_188505_()));
       }
 
+      // Added from changeDimension
+      this.m_11292_(serverplayer); // Update health, etc...
+      serverplayer.m_6885_();
+      for (MobEffectInstance mobEffect : serverplayer.m_21220_()) {
+         serverplayer.f_8906_.m_9829_(new ClientboundUpdateMobEffectPacket(serverplayer.m_19879_(), mobEffect));
+      }
+
+      // Fire advancement trigger
+      serverplayer.m_9209_(((CraftWorld) fromWorld).getHandle());
+
+      // Don't fire on respawn
+      if (fromWorld != location.getWorld()) {
+         PlayerChangedWorldEvent event = new PlayerChangedWorldEvent(serverplayer.getBukkitEntity(), fromWorld);
+         Bukkit.getPluginManager().callEvent(event);
+      }
+
+      // Save player file again if they were disconnected
+      if (serverplayer.f_8906_.isDisconnected()) {
+         this.m_6765_(serverplayer);
+      }
+
+      // CraftBukkit end
       return serverplayer;
    }
+   // CatServer end
 
    public void m_11289_(ServerPlayer p_11290_) {
       GameProfile gameprofile = p_11290_.m_36316_();
@@ -493,12 +_,40 @@
 
    public void m_11288_() {
       if (++this.f_11210_ > 600) {
-         this.m_11268_(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), this.f_11196_));
+         // CraftBukkit start
+         for (int i = 0; i < this.f_11196_.size(); ++i) {
+            final ServerPlayer target = (ServerPlayer) this.f_11196_.get(i);
+            target.f_8906_.m_9829_(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), this.f_11196_.stream().filter(new Predicate<ServerPlayer>() {
+               @Override
+               public boolean test(ServerPlayer input) {
+                  return target.getBukkitEntity().canSee(input.getBukkitEntity());
+               }
+            }).collect(Collectors.toList())));
+         }
+         // CraftBukkit end
          this.f_11210_ = 0;
       }
 
    }
 
+   // CraftBukkit start - add a world/entity limited version
+   public void broadcastAll(Packet packet, Player entityhuman) {
+      for (int i = 0; i < this.f_11196_.size(); ++i) {
+         ServerPlayer entityplayer =  this.f_11196_.get(i);
+         if (entityhuman != null && !entityplayer.getBukkitEntity().canSee(entityhuman.getBukkitEntity())) {
+            continue;
+         }
+         ((ServerPlayer) this.f_11196_.get(i)).f_8906_.m_9829_(packet);
+      }
+   }
+
+   public void broadcastAll(Packet packet, Level world) {
+      for (int i = 0; i < world.m_6907_().size(); ++i) {
+         ((ServerPlayer) world.m_6907_().get(i)).f_8906_.m_9829_(packet);
+      }
+   }
+   // CraftBukkit end
+
    public void m_11268_(Packet<?> p_11269_) {
       for(ServerPlayer serverplayer : this.f_11196_) {
          serverplayer.f_8906_.m_9829_(p_11269_);
@@ -562,6 +_,7 @@
    }
 
    public void m_5749_(GameProfile p_11254_) {
+      if (net.minecraftforge.event.ForgeEventFactory.onPermissionChanged(p_11254_, this.f_11195_.m_7022_(), this)) return;
       this.f_11200_.m_11381_(new ServerOpListEntry(p_11254_, this.f_11195_.m_7022_(), this.f_11200_.m_11351_(p_11254_)));
       ServerPlayer serverplayer = this.m_11259_(p_11254_.getId());
       if (serverplayer != null) {
@@ -571,6 +_,7 @@
    }
 
    public void m_5750_(GameProfile p_11281_) {
+      if (net.minecraftforge.event.ForgeEventFactory.onPermissionChanged(p_11281_, 0, this)) return;
       this.f_11200_.m_11393_(p_11281_);
       ServerPlayer serverplayer = this.m_11259_(p_11281_.getId());
       if (serverplayer != null) {
@@ -593,6 +_,7 @@
          p_11227_.f_8906_.m_9829_(new ClientboundEntityEventPacket(p_11227_, b0));
       }
 
+      p_11227_.getBukkitEntity().recalculatePermissions(); // CraftBukkit
       this.f_11195_.m_129892_().m_82095_(p_11227_);
    }
 
@@ -618,6 +_,11 @@
    public void m_11241_(@Nullable Player p_11242_, double p_11243_, double p_11244_, double p_11245_, double p_11246_, ResourceKey<Level> p_11247_, Packet<?> p_11248_) {
       for(int i = 0; i < this.f_11196_.size(); ++i) {
          ServerPlayer serverplayer = this.f_11196_.get(i);
+         // CraftBukkit start - Test if player receiving packet can see the source of the packet
+         if (p_11242_ != null && !serverplayer.getBukkitEntity().canSee(p_11242_.getBukkitEntity())) {
+            continue;
+         }
+         // CraftBukkit end
          if (serverplayer != p_11242_ && serverplayer.m_9236_().m_46472_() == p_11247_) {
             double d0 = p_11243_ - serverplayer.m_20185_();
             double d1 = p_11244_ - serverplayer.m_20186_();
@@ -657,22 +_,34 @@
    }
 
    public void m_11229_(ServerPlayer p_11230_, ServerLevel p_11231_) {
-      WorldBorder worldborder = this.f_11195_.m_129783_().m_6857_();
+      WorldBorder worldborder = p_11230_.m_9236_().m_6857_(); // CraftBukkit
       p_11230_.f_8906_.m_9829_(new ClientboundInitializeBorderPacket(worldborder));
       p_11230_.f_8906_.m_9829_(new ClientboundSetTimePacket(p_11231_.m_46467_(), p_11231_.m_46468_(), p_11231_.m_46469_().m_46207_(GameRules.f_46140_)));
       p_11230_.f_8906_.m_9829_(new ClientboundSetDefaultSpawnPositionPacket(p_11231_.m_220360_(), p_11231_.m_220361_()));
       if (p_11231_.m_46471_()) {
-         p_11230_.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132154_, 0.0F));
-         p_11230_.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132160_, p_11231_.m_46722_(1.0F)));
-         p_11230_.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132161_, p_11231_.m_46661_(1.0F)));
+         // CraftBukkit start - handle player weather
+         // p_11230_.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
+         // p_11230_.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, p_11231_.getRainLevel(1.0F)));
+         // p_11230_.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, p_11231_.getThunderLevel(1.0F)));
+         p_11230_.setPlayerWeather(org.bukkit.WeatherType.DOWNFALL, false);
+         p_11230_.updateWeather(-p_11231_.f_46438_, p_11231_.f_46438_, -p_11231_.f_46440_, p_11231_.f_46440_);
+         // CraftBukkit end
       }
 
    }
 
    public void m_11292_(ServerPlayer p_11293_) {
       p_11293_.f_36095_.m_150429_();
-      p_11293_.m_9233_();
+      // p_11293_.resetSentInfo();
+      p_11293_.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
+      p_11293_.m_20088_().refresh(p_11293_); // CraftBukkkit - 1.20.1 SPIGOT-7218: sync metadata
       p_11293_.f_8906_.m_9829_(new ClientboundSetCarriedItemPacket(p_11293_.m_150109_().f_35977_));
+      // CraftBukkit start - from GameRules
+      int i = p_11293_.m_9236_().m_46469_().m_46207_(GameRules.f_46145_) ? 22 : 23;
+      p_11293_.f_8906_.m_9829_(new ClientboundEntityEventPacket(p_11293_, (byte) i));
+      float immediateRespawn = p_11293_.m_9236_().m_46469_().m_46207_(GameRules.f_46156_) ? 1.0F: 0.0F;
+      p_11293_.f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132164_, immediateRespawn));
+      // CraftBukkit end
    }
 
    public int m_11309_() {
@@ -725,11 +_,20 @@
    }
 
    public void m_11313_() {
-      for(int i = 0; i < this.f_11196_.size(); ++i) {
-         (this.f_11196_.get(i)).f_8906_.m_9942_(Component.m_237115_("multiplayer.disconnect.server_shutdown"));
+      // CraftBukkit start - disconnect safely
+      for (ServerPlayer player : this.f_11196_) {
+         player.f_8906_.disconnect(this.f_11195_.server.getShutdownMessage()); // CraftBukkit - add custom shutdown message
       }
+      // CraftBukkit end
+   }
 
+   // CraftBukkit start
+   public void broadcastMessage(Component[] iChatBaseComponents) {
+      for (Component component : iChatBaseComponents) {
+         m_240416_(component, false);
+      }
    }
+   // CraftBukkit end
 
    public void m_240416_(Component p_240618_, boolean p_240644_) {
       this.m_240502_(p_240618_, (p_215639_) -> {
@@ -779,22 +_,31 @@
       return p_251384_.m_245272_() && !p_251384_.m_240431_(Instant.now());
    }
 
-   public ServerStatsCounter m_11239_(Player p_11240_) {
-      UUID uuid = p_11240_.m_20148_();
-      ServerStatsCounter serverstatscounter = this.f_11202_.get(uuid);
+   // CraftBukkit start
+   public ServerStatsCounter getPlayerStats(ServerPlayer p_11240_) {
+      ServerStatsCounter serverstatisticmanager = p_11240_.m_8951_();
+      return serverstatisticmanager == null ? getPlayerStats(p_11240_.m_20148_(), p_11240_.m_5446_().getString()) : serverstatisticmanager;
+   }
+
+   public ServerStatsCounter getPlayerStats(UUID uuid, String displayName) {
+      ServerPlayer player = this.m_11259_(uuid);
+      ServerStatsCounter serverstatscounter = player == null ? null : (ServerStatsCounter) player.m_8951_();
+      // CraftBukkit end
+
       if (serverstatscounter == null) {
          File file1 = this.f_11195_.m_129843_(LevelResource.f_78175_).toFile();
          File file2 = new File(file1, uuid + ".json");
+         // CatServer start
          if (!file2.exists()) {
-            File file3 = new File(file1, p_11240_.m_7755_().getString() + ".json");
-            Path path = file3.toPath();
-            if (FileUtil.m_133728_(path) && FileUtil.m_133734_(path) && path.startsWith(file1.getPath()) && file3.isFile()) {
+            File file3 = new File(file1, displayName + ".json"); // CraftBukkit
+            if(file3.exists() && file3.isFile()) {
                file3.renameTo(file2);
             }
          }
+         // CatServer end
 
          serverstatscounter = new ServerStatsCounter(this.f_11195_, file2);
-         this.f_11202_.put(uuid, serverstatscounter);
+         // this.stats.put(uuid, serverstatscounter); // CraftBukkit
       }
 
       return serverstatscounter;
@@ -802,13 +_,15 @@
 
    public PlayerAdvancements m_11296_(ServerPlayer p_11297_) {
       UUID uuid = p_11297_.m_20148_();
-      PlayerAdvancements playeradvancements = this.f_11203_.get(uuid);
+      PlayerAdvancements playeradvancements = this.f_11203_.get(uuid); // CraftBukkit // CatServer - recover
       if (playeradvancements == null) {
          Path path = this.f_11195_.m_129843_(LevelResource.f_78174_).resolve(uuid + ".json");
          playeradvancements = new PlayerAdvancements(this.f_11195_.m_129933_(), this, this.f_11195_.m_129889_(), path, p_11297_);
-         this.f_11203_.put(uuid, playeradvancements);
+         this.f_11203_.put(uuid, playeradvancements); // CraftBukkit // CatServer - recover
       }
 
+      // Forge: don't overwrite active player with a fake one.
+      if (!(p_11297_ instanceof net.minecraftforge.common.util.FakePlayer))
       playeradvancements.m_135979_(p_11297_);
       return playeradvancements;
    }
@@ -838,7 +_,7 @@
    }
 
    public List<ServerPlayer> m_11314_() {
-      return this.f_11196_;
+      return this.playersView; //Unmodifiable view, we don't want people removing things without us knowing.
    }
 
    @Nullable
@@ -851,10 +_,13 @@
    }
 
    public void m_11315_() {
+      // CraftBukkit start // CatServer - recover
       for(PlayerAdvancements playeradvancements : this.f_11203_.values()) {
          playeradvancements.m_135981_(this.f_11195_.m_129889_());
       }
+      // CraftBukkit end
 
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.OnDatapackSyncEvent(this, null));
       this.m_11268_(new ClientboundUpdateTagsPacket(TagNetworkSerialization.m_245799_(this.f_243858_)));
       ClientboundUpdateRecipesPacket clientboundupdaterecipespacket = new ClientboundUpdateRecipesPacket(this.f_11195_.m_129894_().m_44051_());
 
