--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -14,8 +_,13 @@
    }
 
    public boolean m_144004_(int p_144005_, List<ServerPlayer> p_144006_) {
-      int i = (int)p_144006_.stream().filter(Player::m_36317_).count();
-      return i >= this.m_144010_(p_144005_);
+      //int i = (int)p_144006_.stream().filter(Player::isSleepingLongEnough).count();
+      // CraftBukkit start
+      int j = (int) p_144006_.stream().filter((eh) -> { return eh.m_36317_() || eh.fauxSleeping; }).count();
+      boolean anyDeepSleep = p_144006_.stream().anyMatch(Player::m_36317_);
+
+      return anyDeepSleep && j >= this.m_144010_(p_144005_);
+      // CraftBukkit end
    }
 
    public int m_144010_(int p_144011_) {
@@ -35,16 +_,22 @@
       int j = this.f_143999_;
       this.f_143998_ = 0;
       this.f_143999_ = 0;
+      boolean anySleep = false; // CraftBukkit
 
       for(ServerPlayer serverplayer : p_144008_) {
          if (!serverplayer.m_5833_()) {
             ++this.f_143998_;
-            if (serverplayer.m_5803_()) {
+            if (serverplayer.m_5803_() || serverplayer.fauxSleeping) { // CraftBukkit
                ++this.f_143999_;
             }
+            // CraftBukkit start
+            if (serverplayer.m_5803_()) {
+               anySleep = true;
+            }
+            // CraftBukkit end
          }
       }
 
-      return (j > 0 || this.f_143999_ > 0) && (i != this.f_143998_ || j != this.f_143999_);
+      return anySleep && (j > 0 || this.f_143999_ > 0) && (i != this.f_143998_ || j != this.f_143999_);
    }
 }
