--- a/net/minecraft/server/players/StoredUserList.java
+++ b/net/minecraft/server/players/StoredUserList.java
@@ -53,6 +_,12 @@
       return this.f_11377_.get(this.m_5981_(p_11389_));
    }
 
+   // CraftBukkit start
+   public Collection<V> getValues() {
+      return this.f_11377_.values();
+   }
+   // CraftBukkit end
+
    public void m_11393_(K p_11394_) {
       this.f_11377_.remove(this.m_5981_(p_11394_));
 
