--- a/net/minecraft/world/level/storage/loot/LootTables.java
+++ b/net/minecraft/world/level/storage/loot/LootTables.java
@@ -18,6 +_,7 @@
    private static final Logger f_79188_ = LogUtils.getLogger();
    private static final Gson f_79189_ = Deserializers.m_78800_().create();
    private Map<ResourceLocation, LootTable> f_79190_ = ImmutableMap.of();
+   public Map<LootTable, ResourceLocation> lootTableToKey = ImmutableMap.of(); // CraftBukkit
    private final PredicateManager f_79191_;
 
    public LootTables(PredicateManager p_79194_) {
@@ -25,6 +_,18 @@
       this.f_79191_ = p_79194_;
    }
 
+   // CatServer start
+   public ResourceLocation getKey(LootTable loot) {
+      ResourceLocation resourceLocation = ResourceLocation.m_135820_("");
+      for (Map.Entry<ResourceLocation, LootTable> tableEntry : f_79190_.entrySet()) {
+         if (loot.equals(tableEntry.getValue())) {
+            resourceLocation = tableEntry.getKey();
+         }
+      }
+      return resourceLocation;
+   }
+   // CatServer end
+
    public LootTable m_79217_(ResourceLocation p_79218_) {
       return this.f_79190_.getOrDefault(p_79218_, LootTable.f_79105_);
    }
@@ -37,8 +_,8 @@
       }
 
       p_79214_.forEach((p_79198_, p_79199_) -> {
-         try {
-            LootTable loottable = f_79189_.fromJson(p_79199_, LootTable.class);
+         try (net.minecraft.server.packs.resources.Resource res = p_79215_.m_142591_(getPreparedPath(p_79198_));){
+            LootTable loottable = net.minecraftforge.common.ForgeHooks.loadLootTable(f_79189_, p_79198_, p_79199_, res == null || !res.m_7816_().equals("Default"), this);
             builder.put(p_79198_, loottable);
          } catch (Exception exception) {
             f_79188_.error("Couldn't parse loot table {}", p_79198_, exception);
@@ -55,6 +_,11 @@
          f_79188_.warn("Found validation problem in {}: {}", p_79211_, p_79212_);
       });
       this.f_79190_ = immutablemap;
+      // CraftBukkit start - build a reversed registry map
+      ImmutableMap.Builder<LootTable, ResourceLocation> lootTableToKeyBuilder = ImmutableMap.builder();
+      this.f_79190_.forEach((lootTable, key) -> lootTableToKeyBuilder.put(key, lootTable));
+      this.lootTableToKey = lootTableToKeyBuilder.build();
+      // CraftBukkit end
    }
 
    public static void m_79202_(ValidationContext p_79203_, ResourceLocation p_79204_, LootTable p_79205_) {
