--- ../src-base/minecraft/net/minecraft/world/gen/layer/GenLayerEdge.java
+++ ../src-work/minecraft/net/minecraft/world/gen/layer/GenLayerEdge.java
@@ -2,9 +2,9 @@
 
 public class GenLayerEdge extends GenLayer
 {
-    private final GenLayerEdge.Mode mode;
+    private final Mode mode;
 
-    public GenLayerEdge(long p_i45474_1_, GenLayer p_i45474_3_, GenLayerEdge.Mode p_i45474_4_)
+    public GenLayerEdge(long p_i45474_1_, GenLayer p_i45474_3_, Mode p_i45474_4_)
     {
         super(p_i45474_1_);
         this.parent = p_i45474_3_;
