--- ../src-base/minecraft/net/minecraft/world/gen/layer/GenLayerVoronoiZoom.java
+++ ../src-work/minecraft/net/minecraft/world/gen/layer/GenLayerVoronoiZoom.java
@@ -28,7 +28,7 @@
 
             for (int j2 = aint[l1 + 0 + (k1 + 1) * k]; l1 < k - 1; ++l1)
             {
-                double d0 = 3.6D;
+                //double d0 = 3.6D;
                 this.initChunkSeed((long)(l1 + i << 2), (long)(k1 + j << 2));
                 double d1 = ((double)this.nextInt(1024) / 1024.0D - 0.5D) * 3.6D;
                 double d2 = ((double)this.nextInt(1024) / 1024.0D - 0.5D) * 3.6D;
