--- a/net/minecraft/entity/ai/goal/BreakBlockGoal.java
+++ b/net/minecraft/entity/ai/goal/BreakBlockGoal.java
@@ -11,7 +_,6 @@
 import net.minecraft.particles.ParticleTypes;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.vector.Vector3d;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
@@ -19,6 +_,8 @@
 import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.chunk.IChunk;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class BreakBlockGoal extends MoveToBlockGoal {
    private final Block field_203117_f;
@@ -32,7 +_,7 @@
    }
 
    public boolean func_75250_a() {
-      if (!this.field_203118_g.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+      if (!net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.field_203118_g.field_70170_p, this.field_179494_b, this.field_203118_g)) {
          return false;
       } else if (this.field_179496_a > 0) {
          --this.field_179496_a;
@@ -91,6 +_,13 @@
          }
 
          if (this.field_203119_h > 60) {
+            // CraftBukkit start - Step on eggs
+            EntityInteractEvent event = new EntityInteractEvent(this.field_203118_g.getBukkitEntity(), CraftBlock.at(world, blockpos1));
+            world.getCBServer().getPluginManager().callEvent((EntityInteractEvent) event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             world.func_217377_a(blockpos1, false);
             if (!world.field_72995_K) {
                for(int i = 0; i < 20; ++i) {
