--- a/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
@@ -8,6 +_,7 @@
 import net.minecraft.entity.passive.IronGolemEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.util.math.AxisAlignedBB;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class DefendVillageTargetGoal extends TargetGoal {
    private final IronGolemEntity field_75305_a;
@@ -44,7 +_,7 @@
    }
 
    public void func_75249_e() {
-      this.field_75305_a.func_70624_b(this.field_75304_b);
+      this.field_75305_a.setGoalTarget(this.field_75304_b, TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
       super.func_75249_e();
    }
 }
