--- a/net/minecraft/entity/ai/goal/EatGrassGoal.java
+++ b/net/minecraft/entity/ai/goal/EatGrassGoal.java
@@ -10,6 +_,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class EatGrassGoal extends Goal {
    private static final Predicate<BlockState> field_179505_b = BlockStateMatcher.func_177638_a(Blocks.field_150349_c);
@@ -59,7 +_,8 @@
       if (this.field_151502_a == 4) {
          BlockPos blockpos = this.field_151500_b.func_233580_cy_();
          if (field_179505_b.test(this.field_151501_c.func_180495_p(blockpos))) {
-            if (this.field_151501_c.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+            // CraftBukkit
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_151501_c, this.field_151500_b) || !CraftEventFactory.callEntityChangeBlockEvent(this.field_151500_b, blockpos, Blocks.field_150350_a.func_176223_P(), !this.field_151501_c.func_82736_K().func_223586_b(GameRules.field_223599_b)).isCancelled()) {
                this.field_151501_c.func_175655_b(blockpos, false);
             }
 
@@ -67,7 +_,8 @@
          } else {
             BlockPos blockpos1 = blockpos.func_177977_b();
             if (this.field_151501_c.func_180495_p(blockpos1).func_203425_a(Blocks.field_196658_i)) {
-               if (this.field_151501_c.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+               // CraftBukkit
+               if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_151501_c, this.field_151500_b) || !CraftEventFactory.callEntityChangeBlockEvent(this.field_151500_b, blockpos, Blocks.field_150350_a.func_176223_P(), !this.field_151501_c.func_82736_K().func_223586_b(GameRules.field_223599_b)).isCancelled()) {
                   this.field_151501_c.func_217379_c(2001, blockpos1, Block.func_196246_j(Blocks.field_196658_i.func_176223_P()));
                   this.field_151501_c.func_180501_a(blockpos1, Blocks.field_150346_d.func_176223_P(), 2);
                }
