--- a/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
@@ -12,6 +_,9 @@
 import net.minecraft.pathfinding.WalkNodeProcessor;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IWorldReader;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftEntity;
+import org.bukkit.event.entity.EntityTeleportEvent;
 
 public class FollowOwnerGoal extends Goal {
    private final TameableEntity field_75338_d;
@@ -113,7 +_,17 @@
       } else if (!this.func_226329_a_(new BlockPos(p_226328_1_, p_226328_2_, p_226328_3_))) {
          return false;
       } else {
-         this.field_75338_d.func_70012_b((double)p_226328_1_ + 0.5D, (double)p_226328_2_, (double)p_226328_3_ + 0.5D, this.field_75338_d.field_70177_z, this.field_75338_d.field_70125_A);
+         // CraftBukkit start
+         CraftEntity entity = this.field_75338_d.getBukkitEntity();
+         Location to = new Location(entity.getWorld(), (double) ((float) p_226328_1_ + 0.5f), (double) p_226328_2_, (double) ((float) p_226328_3_ + 0.5f), this.field_75338_d.field_70177_z, this.field_75338_d.field_70125_A);
+         EntityTeleportEvent event = new EntityTeleportEvent(entity, entity.getLocation(), to);
+         this.field_75338_d.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return false;
+         }
+         to = event.getTo();
+         this.field_75338_d.func_70012_b(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         // CraftBukkit end
          this.field_75337_g.func_75499_g();
          return true;
       }
