--- a/net/minecraft/entity/ai/goal/HurtByTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/HurtByTargetGoal.java
@@ -52,7 +_,7 @@
    }
 
    public void func_75249_e() {
-      this.field_75299_d.func_70624_b(this.field_75299_d.func_70643_av());
+      this.field_75299_d.setGoalTarget(this.field_75299_d.func_70643_av(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
       this.field_188509_g = this.field_75299_d.func_70638_az();
       this.field_142052_b = this.field_75299_d.func_142015_aE();
       this.field_188510_h = 300;
@@ -102,6 +_,6 @@
    }
 
    protected void func_220793_a(MobEntity p_220793_1_, LivingEntity p_220793_2_) {
-      p_220793_1_.func_70624_b(p_220793_2_);
+      p_220793_1_.setGoalTarget(p_220793_2_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit - reason
    }
 }
