--- a/net/minecraft/entity/ai/goal/NearestAttackableTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/NearestAttackableTargetGoal.java
@@ -55,7 +_,7 @@
    }
 
    public void func_75249_e() {
-      this.field_75299_d.func_70624_b(this.field_75309_a);
+     this.field_75299_d.setGoalTarget(this.field_75309_a, field_75309_a instanceof ServerPlayerEntity ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit - reason
       super.func_75249_e();
    }
 
