--- a/net/minecraft/entity/ai/goal/OwnerHurtByTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/OwnerHurtByTargetGoal.java
@@ -32,7 +_,7 @@
    }
 
    public void func_75249_e() {
-      this.field_75299_d.func_70624_b(this.field_75315_b);
+      this.field_75299_d.setGoalTarget(this.field_75315_b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit - reason
       LivingEntity livingentity = this.field_75316_a.func_70902_q();
       if (livingentity != null) {
          this.field_142051_e = livingentity.func_142015_aE();
