--- a/net/minecraft/entity/ai/goal/PanicGoal.java
+++ b/net/minecraft/entity/ai/goal/PanicGoal.java
@@ -68,6 +_,12 @@
    }
 
    public boolean func_75253_b() {
+      // CraftBukkit start - introduce a temporary timeout hack until this is fixed properly
+      if ((this.field_75267_a.field_70173_aa - this.field_75267_a.field_70756_c) > 100) {
+         this.field_75267_a.func_70604_c(null);
+         return false;
+      }
+      // CraftBukkit end
       return !this.field_75267_a.func_70661_as().func_75500_f();
    }
 
