--- a/net/minecraft/entity/ai/goal/SitGoal.java
+++ b/net/minecraft/entity/ai/goal/SitGoal.java
@@ -18,7 +_,7 @@
 
    public boolean func_75250_a() {
       if (!this.field_75272_a.func_70909_n()) {
-         return false;
+         return this.field_75272_a.func_233685_eM_() && this.field_75272_a.func_70638_az() == null; // CraftBukkit - Allow sitting for wild animals
       } else if (this.field_75272_a.func_203005_aq()) {
          return false;
       } else if (!this.field_75272_a.func_233570_aj_()) {
