--- a/net/minecraft/entity/ai/goal/TargetGoal.java
+++ b/net/minecraft/entity/ai/goal/TargetGoal.java
@@ -10,6 +_,7 @@
 import net.minecraft.pathfinding.PathPoint;
 import net.minecraft.scoreboard.Team;
 import net.minecraft.util.math.MathHelper;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public abstract class TargetGoal extends Goal {
    protected final MobEntity field_75299_d;
@@ -62,7 +_,7 @@
                if (livingentity instanceof PlayerEntity && ((PlayerEntity)livingentity).field_71075_bZ.field_75102_a) {
                   return false;
                } else {
-                  this.field_75299_d.func_70624_b(livingentity);
+                  this.field_75299_d.setGoalTarget(livingentity, EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit
                   return true;
                }
             }
@@ -81,7 +_,7 @@
    }
 
    public void func_75251_c() {
-      this.field_75299_d.func_70624_b((LivingEntity)null);
+      this.field_75299_d.setGoalTarget((LivingEntity)null, EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.field_188509_g = null;
    }
 
