--- a/net/minecraft/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/entity/ai/goal/TemptGoal.java
@@ -1,13 +_,18 @@
 package net.minecraft.entity.ai.goal;
 
 import java.util.EnumSet;
+
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+
 import net.minecraft.entity.CreatureEntity;
 import net.minecraft.entity.EntityPredicate;
+import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.Ingredient;
-import net.minecraft.pathfinding.FlyingPathNavigator;
-import net.minecraft.pathfinding.GroundPathNavigator;
 
 public class TemptGoal extends Goal {
    private static final EntityPredicate field_220762_c = (new EntityPredicate()).func_221013_a(10.0D).func_221008_a().func_221011_b().func_221009_d().func_221014_c();
@@ -34,9 +_,9 @@
       this.field_151484_k = p_i47823_5_;
       this.field_75285_l = p_i47823_4_;
       this.func_220684_a(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.LOOK));
-      if (!(p_i47823_1_.func_70661_as() instanceof GroundPathNavigator) && !(p_i47823_1_.func_70661_as() instanceof FlyingPathNavigator)) {
+      /* if (!(p_i47823_1_.getNavigation() instanceof GroundPathNavigator) && !(p_i47823_1_.getNavigation() instanceof FlyingPathNavigator)) {
          throw new IllegalArgumentException("Unsupported mob type for TemptGoal");
-      }
+      } */
    }
 
    public boolean func_75250_a() {
@@ -45,12 +_,24 @@
          return false;
       } else {
          this.field_75289_h = this.field_75284_a.field_70170_p.func_217370_a(field_220762_c, this.field_75284_a);
-         if (this.field_75289_h == null) {
-            return false;
-         } else {
-            return this.func_188508_a(this.field_75289_h.func_184614_ca()) || this.func_188508_a(this.field_75289_h.func_184592_cb());
+         // CraftBukkit start
+         boolean tempt = this.field_75289_h != null && (this.func_188508_a(this.field_75289_h.func_184614_ca()) || this.func_188508_a(this.field_75289_h.func_184592_cb()));
+         if (field_75289_h instanceof LivingEntity) {
+            if (tempt) {
+               EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(this.field_75284_a, this.field_75289_h, EntityTargetEvent.TargetReason.TEMPT);
+               if (event.isCancelled()) {
+                  return false;
+               }
+               quarkFix(((event.getTarget() == null) ? null : ((CraftPlayer) event.getTarget()).getHandle()));
+            }
          }
+         return tempt;
+         // CraftBukkit end
       }
+   }
+
+   public void quarkFix(PlayerEntity player) {
+      this.field_75289_h = player;
    }
 
    protected boolean func_188508_a(ItemStack p_188508_1_) {
