--- a/net/minecraft/entity/ai/goal/TriggerSkeletonTrapGoal.java
+++ b/net/minecraft/entity/ai/goal/TriggerSkeletonTrapGoal.java
@@ -28,6 +_,10 @@
 
    public void func_75246_d() {
       ServerWorld serverworld = (ServerWorld)this.field_188516_a.field_70170_p;
+      // Forge: Trigger the trap in a tick task to avoid crashes when mods add goals to skeleton horses
+      // (MC-206338/Forge PR #7509)
+      serverworld.func_73046_m().func_212871_a_(new net.minecraft.util.concurrent.TickDelayedTask(serverworld.func_73046_m().func_71259_af(), () -> {
+         if (!this.field_188516_a.func_70089_S()) return;
       DifficultyInstance difficultyinstance = serverworld.func_175649_E(this.field_188516_a.func_233580_cy_());
       this.field_188516_a.func_190691_p(false);
       this.field_188516_a.func_110234_j(true);
@@ -35,19 +_,20 @@
       LightningBoltEntity lightningboltentity = EntityType.field_200728_aG.func_200721_a(serverworld);
       lightningboltentity.func_225653_b_(this.field_188516_a.func_226277_ct_(), this.field_188516_a.func_226278_cu_(), this.field_188516_a.func_226281_cx_());
       lightningboltentity.func_233623_a_(true);
-      serverworld.func_217376_c(lightningboltentity);
+      ((ServerWorld)this.field_188516_a.field_70170_p).strikeLightning(lightningboltentity, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRAP); // CraftBukkit
       SkeletonEntity skeletonentity = this.func_188514_a(difficultyinstance, this.field_188516_a);
-      skeletonentity.func_184220_m(this.field_188516_a);
-      serverworld.func_242417_l(skeletonentity);
+      if (skeletonentity != null) skeletonentity.func_184220_m(this.field_188516_a); // CraftBukkit
+      serverworld.addAllEntities(skeletonentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRAP); // CraftBukkit
 
       for(int i = 0; i < 3; ++i) {
          AbstractHorseEntity abstracthorseentity = this.func_188515_a(difficultyinstance);
+         if (abstracthorseentity == null) continue; // CraftBukkit
          SkeletonEntity skeletonentity1 = this.func_188514_a(difficultyinstance, abstracthorseentity);
-         skeletonentity1.func_184220_m(abstracthorseentity);
+         if (skeletonentity1 != null) skeletonentity1.func_184220_m(abstracthorseentity); // CraftBukkit
          abstracthorseentity.func_70024_g(this.field_188516_a.func_70681_au().nextGaussian() * 0.5D, 0.0D, this.field_188516_a.func_70681_au().nextGaussian() * 0.5D);
-         serverworld.func_242417_l(abstracthorseentity);
+         serverworld.addAllEntities(abstracthorseentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit
       }
-
+      }));
    }
 
    private AbstractHorseEntity func_188515_a(DifficultyInstance p_188515_1_) {
