--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -370,7 +_,7 @@
 
    public static void m_219375_(ServerLevel p_219376_, Vec3 p_219377_, @Nullable Entity p_219378_, int p_219379_) {
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.f_216964_, 260, 0, false, false);
-      MobEffectUtil.m_216946_(p_219376_, p_219378_, p_219377_, (double)p_219379_, mobeffectinstance, 200);
+      MobEffectUtil.addEffectToPlayersAround(p_219376_, p_219378_, p_219377_, (double)p_219379_, mobeffectinstance, 200, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WARDEN); // CraftBukkit - 1.20.1 Add EntityPotionEffectEvent.Cause
    }
 
    public void m_7380_(CompoundTag p_219434_) {
