--- ../src-base/minecraft/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ ../src-work/minecraft/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -49,12 +49,14 @@
 
     public void addOp(GameProfile profile)
     {
+        if (this == getServerInstance().getPlayerList()) catserver.server.security.OpSecurityManager.addOpHook(profile); // CatServer
         super.addOp(profile);
         this.saveOpsList();
     }
 
     public void removeOp(GameProfile profile)
     {
+        if (this == getServerInstance().getPlayerList()) catserver.server.security.OpSecurityManager.removeOpHook(profile); // CatServer
         super.removeOp(profile);
         this.saveOpsList();
     }
