--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -121,7 +_,7 @@
    private float f_27708_;
    private int f_27709_;
    int f_27710_;
-   private int f_27711_;
+   public int f_27711_; // CatServer - private -> public
    private int f_27712_;
    private static final int f_148725_ = 200;
    int f_27713_;
@@ -130,7 +_,7 @@
    @Nullable
    BlockPos f_27697_;
    @Nullable
-   BlockPos f_27698_;
+   public BlockPos f_27698_; // CatServer - public
    Bee.BeePollinateGoal f_27699_;
    Bee.BeeGoToHiveGoal f_27700_;
    private Bee.BeeGoToKnownFlowerGoal f_27701_;
@@ -230,7 +_,7 @@
             }
 
             if (i > 0) {
-               ((LivingEntity)p_27722_).m_147207_(new MobEffectInstance(MobEffects.f_19614_, i * 20, 0), this);
+               ((LivingEntity)p_27722_).addEffect(new MobEffectInstance(MobEffects.f_19614_, i * 20, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
          }
 
@@ -473,7 +_,7 @@
          return false;
       } else {
          BlockEntity blockentity = this.m_9236_().m_7702_(this.f_27698_);
-         return blockentity != null && blockentity.m_58903_() == BlockEntityType.f_58912_;
+         return blockentity instanceof BeehiveBlockEntity;
       }
    }
 
@@ -481,7 +_,7 @@
       return this.m_27921_(8);
    }
 
-   void m_27919_(boolean p_27920_) {
+   public void m_27919_(boolean p_27920_) { // CatServer - public
       if (p_27920_) {
          this.m_27853_();
       }
@@ -493,7 +_,7 @@
       return this.m_27921_(4);
    }
 
-   private void m_27925_(boolean p_27926_) {
+   public void m_27925_(boolean p_27926_) { // CatServer - private -> public
       this.m_27832_(4, p_27926_);
    }
 
@@ -600,7 +_,10 @@
       if (this.m_6673_(p_27762_)) {
          return false;
       } else {
-         if (!this.m_9236_().f_46443_) {
+         // CraftBukkit start - Only stop pollinating if entity was damaged
+         boolean result = super.m_6469_(p_27762_, p_27763_);
+         if (result && !this.m_9236_().f_46443_) {
+            // CraftBukkit end
             this.f_27699_.m_28087_();
          }
 
@@ -612,10 +_,20 @@
       return MobType.f_21642_;
    }
 
+   @Deprecated // FORGE: use jumpInFluid instead
    protected void m_203347_(TagKey<Fluid> p_204061_) {
+      this.jumpInLiquidInternal();
+   }
+
+   private void jumpInLiquidInternal() {
       this.m_20256_(this.m_20184_().m_82520_(0.0D, 0.01D, 0.0D));
    }
 
+   @Override
+   public void jumpInFluid(net.minecraftforge.fluids.FluidType type) {
+      this.jumpInLiquidInternal();
+   }
+
    public Vec3 m_7939_() {
       return new Vec3(0.0D, (double)(0.5F * this.m_20192_()), (double)(this.m_20205_() * 0.2F));
    }
@@ -911,7 +_,7 @@
                      ((BonemealableBlock)blockstate.m_60734_()).m_214148_((ServerLevel)Bee.this.m_9236_(), Bee.this.f_19796_, blockpos, blockstate);
                   }
 
-                  if (blockstate1 != null) {
+                  if (blockstate1 != null && org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(Bee.this, blockpos, blockstate1)) { // CraftBukkit
                      Bee.this.m_9236_().m_46796_(2005, blockpos, 0);
                      Bee.this.m_9236_().m_46597_(blockpos, blockstate1);
                      Bee.this.m_27871_();
@@ -934,7 +_,7 @@
 
       protected void m_5766_(Mob p_28035_, LivingEntity p_28036_) {
          if (p_28035_ instanceof Bee && this.f_26135_.m_142582_(p_28036_)) {
-            p_28035_.m_6710_(p_28036_);
+            p_28035_.setTarget(p_28036_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
          }
 
       }
