--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -130,7 +_,8 @@
    @Nullable
    BlockPos f_27697_;
    @Nullable
-   public BlockPos f_27698_;
+   public
+   BlockPos f_27698_;
    Bee.BeePollinateGoal f_27699_;
    Bee.BeeGoToHiveGoal f_27700_;
    private Bee.BeeGoToKnownFlowerGoal f_27701_;
@@ -471,7 +_,7 @@
          return false;
       } else {
          BlockEntity blockentity = this.f_19853_.m_7702_(this.f_27698_);
-         return blockentity != null && blockentity.m_58903_() == BlockEntityType.f_58912_;
+         return blockentity instanceof BeehiveBlockEntity;
       }
    }
 
@@ -479,7 +_,7 @@
       return this.m_27921_(8);
    }
 
-   void m_27919_(boolean p_27920_) {
+   public void m_27919_(boolean p_27920_) {
       if (p_27920_) {
          this.m_27853_();
       }
@@ -491,7 +_,7 @@
       return this.m_27921_(4);
    }
 
-   private void m_27925_(boolean p_27926_) {
+   public void m_27925_(boolean p_27926_) {
       this.m_27832_(4, p_27926_);
    }
 
@@ -504,7 +_,7 @@
    }
 
    boolean m_27889_(BlockPos p_27890_) {
-      return !this.m_27816_(p_27890_, 32);
+      return !this.closerToCenterThan(p_27890_, 32);
    }
 
    private void m_27832_(int p_27833_, boolean p_27834_) {
@@ -621,7 +_,7 @@
       return new Vec3(0.0D, (double)(0.5F * this.m_20192_()), (double)(this.m_20205_() * 0.2F));
    }
 
-   boolean m_27816_(BlockPos p_27817_, int p_27818_) {
+   boolean closerToCenterThan(BlockPos p_27817_, int p_27818_) {
       return p_27817_.m_123314_(this.m_142538_(), (double)p_27818_);
    }
 
@@ -751,7 +_,7 @@
             if (this.f_27980_ > this.m_183277_(600)) {
                this.m_28007_();
             } else if (!Bee.this.f_21344_.m_26572_()) {
-               if (!Bee.this.m_27816_(Bee.this.f_27698_, 16)) {
+               if (!Bee.this.closerToCenterThan(Bee.this.f_27698_, 16)) {
                   if (Bee.this.m_27889_(Bee.this.f_27698_)) {
                      this.m_28008_();
                   } else {
@@ -813,7 +_,7 @@
       }
 
       private boolean m_28001_(BlockPos p_28002_) {
-         if (Bee.this.m_27816_(p_28002_, 2)) {
+         if (Bee.this.closerToCenterThan(p_28002_, 2)) {
             return true;
          } else {
             Path path = Bee.this.f_21344_.m_26570_();
@@ -831,7 +_,7 @@
       }
 
       public boolean m_7989_() {
-         return Bee.this.f_27697_ != null && !Bee.this.m_21536_() && this.m_28020_() && Bee.this.m_27896_(Bee.this.f_27697_) && !Bee.this.m_27816_(Bee.this.f_27697_, 2);
+         return Bee.this.f_27697_ != null && !Bee.this.m_21536_() && this.m_28020_() && Bee.this.m_27896_(Bee.this.f_27697_) && !Bee.this.closerToCenterThan(Bee.this.f_27697_, 2);
       }
 
       public boolean m_8011_() {
@@ -1207,7 +_,7 @@
       @Nullable
       private Vec3 m_28097_() {
          Vec3 vec3;
-         if (Bee.this.m_27872_() && !Bee.this.m_27816_(Bee.this.f_27698_, 22)) {
+         if (Bee.this.m_27872_() && !Bee.this.closerToCenterThan(Bee.this.f_27698_, 22)) {
             Vec3 vec31 = Vec3.m_82512_(Bee.this.f_27698_);
             vec3 = vec31.m_82546_(Bee.this.m_20182_()).m_82541_();
          } else {
