--- a/net/minecraft/world/entity/animal/Bucketable.java
+++ b/net/minecraft/world/entity/animal/Bucketable.java
@@ -15,6 +_,13 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketEntityEvent;
+// CraftBukkit end
+
 public interface Bucketable {
    boolean m_27487_();
 
@@ -91,9 +_,22 @@
    static <T extends LivingEntity & Bucketable> Optional<InteractionResult> m_148828_(Player p_148829_, InteractionHand p_148830_, T p_148831_) {
       ItemStack itemstack = p_148829_.m_21120_(p_148830_);
       if (itemstack.m_41720_() == Items.f_42447_ && p_148831_.m_6084_()) {
-         p_148831_.m_5496_(p_148831_.m_142623_(), 1.0F, 1.0F);
+         // CraftBukkit start
+         // p_148831_.playSound(p_148831_.getPickupSound(), 1.0F, 1.0F);
          ItemStack itemstack1 = p_148831_.m_28282_();
          p_148831_.m_6872_(itemstack1);
+
+         PlayerBucketEntityEvent playerBucketFishEvent = CraftEventFactory.callPlayerFishBucketEvent(p_148831_, p_148829_, itemstack, itemstack1, p_148830_);
+         itemstack1 = CraftItemStack.asNMSCopy(playerBucketFishEvent.getEntityBucket());
+         if (playerBucketFishEvent.isCancelled()) {
+            ((ServerPlayer) p_148829_).f_36096_.m_150429_(); // We need to update inventory to resync client's bucket
+            ((ServerPlayer) p_148829_).f_8906_.m_9829_(new ClientboundAddEntityPacket(p_148831_)); // We need to play out these packets as the client assumes the fish is gone
+            p_148831_.m_20088_().refresh((ServerPlayer) p_148829_); // Need to send data such as the display name to client
+            return Optional.of(InteractionResult.FAIL);
+         }
+         p_148831_.m_5496_(p_148831_.m_142623_(), 1.0F, 1.0F);
+         // CraftBukkit end
+
          ItemStack itemstack2 = ItemUtils.m_41817_(itemstack, p_148829_, itemstack1, false);
          p_148829_.m_21008_(p_148830_, itemstack2);
          Level level = p_148831_.m_9236_();
