--- a/net/minecraft/world/entity/animal/Bucketable.java
+++ b/net/minecraft/world/entity/animal/Bucketable.java
@@ -3,6 +_,8 @@
 import java.util.Optional;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.protocol.game.ClientboundAddMobPacket;
+import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.InteractionHand;
@@ -14,6 +_,9 @@
 import net.minecraft.world.item.ItemUtils;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketEntityEvent;
 
 public interface Bucketable {
    boolean m_142392_();
@@ -91,9 +_,20 @@
    static <T extends LivingEntity & Bucketable> Optional<InteractionResult> m_148828_(Player p_148829_, InteractionHand p_148830_, T p_148831_) {
       ItemStack itemstack = p_148829_.m_21120_(p_148830_);
       if (itemstack.m_41720_() == Items.f_42447_ && p_148831_.m_6084_()) {
-         p_148831_.m_5496_(p_148831_.m_142623_(), 1.0F, 1.0F);
          ItemStack itemstack1 = p_148831_.m_142563_();
          p_148831_.m_142146_(itemstack1);
+
+         PlayerBucketEntityEvent playerBucketFishEvent = CraftEventFactory.callPlayerFishBucketEvent(p_148831_, p_148829_, itemstack, itemstack1);
+         itemstack1 = CraftItemStack.asNMSCopy(playerBucketFishEvent.getEntityBucket());
+         if (playerBucketFishEvent.isCancelled()) {
+            ((ServerPlayer) p_148829_).f_36096_.m_150429_(); // We need to update inventory to resync client's bucket
+            ((ServerPlayer) p_148829_).f_8906_.m_141995_(new ClientboundAddMobPacket(p_148831_)); // We need to play out these packets as the client assumes the fish is gone
+            ((ServerPlayer) p_148829_).f_8906_.m_141995_(new ClientboundSetEntityDataPacket(p_148831_.m_142049_(), p_148831_.m_20088_(), true)); // Need to send data such as the display name to client
+            return Optional.of(InteractionResult.FAIL);
+         }
+         p_148831_.m_5496_(p_148831_.m_142623_(), 1.0F, 1.0F);
+         // CraftBukkit end
+
          ItemStack itemstack2 = ItemUtils.m_41817_(itemstack, p_148829_, itemstack1, false);
          p_148829_.m_21008_(p_148830_, itemstack2);
          Level level = p_148831_.f_19853_;
