--- a/net/minecraft/world/entity/animal/Chicken.java
+++ b/net/minecraft/world/entity/animal/Chicken.java
@@ -89,7 +_,9 @@
       this.f_28226_ += this.f_28230_ * 2.0F;
       if (!this.m_9236_().f_46443_ && this.m_6084_() && !this.m_6162_() && !this.m_28264_() && --this.f_28231_ <= 0) {
          this.m_5496_(SoundEvents.f_11752_, 1.0F, (this.f_19796_.m_188501_() - this.f_19796_.m_188501_()) * 0.2F + 1.0F);
+         this.forceDrops = true; // CraftBukkit
          this.m_19998_(Items.f_42521_);
+         this.forceDrops = false; // CraftBukkit
          this.m_146850_(GameEvent.f_157810_);
          this.f_28231_ = this.f_19796_.m_188503_(6000) + 6000;
       }
