--- a/net/minecraft/world/entity/animal/Chicken.java
+++ b/net/minecraft/world/entity/animal/Chicken.java
@@ -69,16 +_,21 @@
    }
 
    public void m_8107_() {
+      // CraftBukkit start
+      if (this.m_28264_()) {
+         this.setPersistenceRequired(!this.m_6785_(0));
+      }
+      // CraftBukkit end
       super.m_8107_();
       this.f_28229_ = this.f_28226_;
       this.f_28228_ = this.f_28227_;
-      this.f_28227_ += (this.f_19861_ ? -1.0F : 4.0F) * 0.3F;
+      this.f_28227_ = (float)((double)this.f_28227_ + (double)(this.f_19861_ ? -1 : 4) * 0.3D);
       this.f_28227_ = Mth.m_14036_(this.f_28227_, 0.0F, 1.0F);
       if (!this.f_19861_ && this.f_28230_ < 1.0F) {
          this.f_28230_ = 1.0F;
       }
 
-      this.f_28230_ *= 0.9F;
+      this.f_28230_ = (float)((double)this.f_28230_ * 0.9D);
       Vec3 vec3 = this.m_20184_();
       if (!this.f_19861_ && vec3.f_82480_ < 0.0D) {
          this.m_20256_(vec3.m_82542_(1.0D, 0.6D, 1.0D));
@@ -87,7 +_,9 @@
       this.f_28226_ += this.f_28230_ * 2.0F;
       if (!this.f_19853_.f_46443_ && this.m_6084_() && !this.m_6162_() && !this.m_28264_() && --this.f_28231_ <= 0) {
          this.m_5496_(SoundEvents.f_11752_, 1.0F, (this.f_19796_.nextFloat() - this.f_19796_.nextFloat()) * 0.2F + 1.0F);
+         this.forceDrops = true; // CraftBukkit
          this.m_19998_(Items.f_42521_);
+         this.forceDrops = false; // CraftBukkit
          this.f_28231_ = this.f_19796_.nextInt(6000) + 6000;
       }
 
