--- a/net/minecraft/world/entity/animal/Cow.java
+++ b/net/minecraft/world/entity/animal/Cow.java
@@ -29,6 +_,8 @@
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
 
 public class Cow extends Animal {
    public Cow(EntityType<? extends Cow> p_28285_, Level p_28286_) {
@@ -73,8 +_,14 @@
    public InteractionResult m_6071_(Player p_28298_, InteractionHand p_28299_) {
       ItemStack itemstack = p_28298_.m_21120_(p_28299_);
       if (itemstack.m_150930_(Items.f_42446_) && !this.m_6162_()) {
+         // CraftBukkit start - Got milk?
+         org.bukkit.event.player.PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_28298_.f_19853_, p_28298_, this.m_142538_(), this.m_142538_(), null, itemstack, Items.f_42455_);
+         if (event.isCancelled()) {
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          p_28298_.m_5496_(SoundEvents.f_11833_, 1.0F, 1.0F);
-         ItemStack itemstack1 = ItemUtils.m_41813_(itemstack, p_28298_, Items.f_42455_.m_7968_());
+         ItemStack itemstack1 = ItemUtils.m_41813_(itemstack, p_28298_, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
          p_28298_.m_21008_(p_28299_, itemstack1);
          return InteractionResult.m_19078_(this.f_19853_.f_46443_);
       } else {
