--- a/net/minecraft/world/entity/animal/Dolphin.java
+++ b/net/minecraft/world/entity/animal/Dolphin.java
@@ -72,6 +_,13 @@
       return !p_28369_.m_32063_() && p_28369_.m_6084_() && p_28369_.m_20069_();
    };
 
+   // CraftBukkit start - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
+   @Override
+   public int getDefaultMaxAirSupply() {
+      return f_148892_;
+   }
+   // CraftBukkit end
+
    public Dolphin(EntityType<? extends Dolphin> p_28316_, Level p_28317_) {
       super(p_28316_, p_28317_);
       this.f_21342_ = new SmoothSwimmingMoveControl(this, 85, 10, 0.02F, 0.1F, true);
@@ -178,7 +_,7 @@
    }
 
    public int m_6062_() {
-      return 4800;
+      return maxAirTicks;
    }
 
    protected int m_7305_(int p_28389_) {
@@ -214,6 +_,12 @@
       if (this.m_6844_(EquipmentSlot.MAINHAND).m_41619_()) {
          ItemStack itemstack = p_28357_.m_32055_();
          if (this.m_7252_(itemstack)) {
+            // CraftBukkit start - call EntityPickupItemEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPickupItemEvent(this, p_28357_, 0, false).isCancelled()) {
+               return;
+            }
+            itemstack = p_28357_.m_32055_(); // update ItemStack from event
+            // CraftBukkit end
             this.m_21053_(p_28357_);
             this.m_8061_(EquipmentSlot.MAINHAND, itemstack);
             this.f_21347_[EquipmentSlot.MAINHAND.m_20749_()] = 2.0F;
@@ -356,7 +_,7 @@
       }
 
       public boolean m_8036_() {
-         return this.f_28399_.m_28377_() && this.f_28399_.m_20146_() >= 100;
+         return this.f_28399_.m_28377_() && this.f_28399_.m_20146_() >= 100 && this.f_28399_.f_19853_.getWorld().canGenerateStructures(); // MC-151364, SPIGOT-5494: hangs if generate-structures=false
       }
 
       public boolean m_8045_() {
@@ -445,7 +_,7 @@
       }
 
       public void m_8056_() {
-         this.f_28411_.m_147207_(new MobEffectInstance(MobEffects.f_19593_, 100), this.f_28409_);
+         this.f_28411_.addEffect(new MobEffectInstance(MobEffects.f_19593_, 100), this.f_28409_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
       }
 
       public void m_8041_() {
@@ -462,7 +_,7 @@
          }
 
          if (this.f_28411_.m_6069_() && this.f_28411_.f_19853_.f_46441_.nextInt(6) == 0) {
-            this.f_28411_.m_147207_(new MobEffectInstance(MobEffects.f_19593_, 100), this.f_28409_);
+            this.f_28411_.addEffect(new MobEffectInstance(MobEffects.f_19593_, 100), this.f_28409_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
          }
 
       }
