--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -60,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class Panda extends Animal {
    private static final EntityDataAccessor<Integer> f_29073_ = SynchedEntityData.m_135353_(Panda.class, EntityDataSerializers.f_135028_);
@@ -501,7 +_,8 @@
    }
 
    protected void m_7581_(ItemEntity p_29121_) {
-      if (this.m_6844_(EquipmentSlot.MAINHAND).m_41619_() && f_29071_.test(p_29121_)) {
+     // if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_29121_)) {
+      if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPickupItemEvent(this, p_29121_, 0, !(this.m_6844_(EquipmentSlot.MAINHAND).m_41619_() && f_29071_.test(p_29121_))).isCancelled()) { // CraftBukkit
          this.m_21053_(p_29121_);
          ItemStack itemstack = p_29121_.m_32055_();
          this.m_8061_(EquipmentSlot.MAINHAND, itemstack);
@@ -513,10 +_,7 @@
    }
 
    public boolean m_6469_(DamageSource p_29097_, float p_29098_) {
-      if (!this.f_19853_.f_46443_) {
-         this.m_29208_(false);
-      }
-
+      this.m_29208_(false);
       return super.m_6469_(p_29097_, p_29098_);
    }
 
@@ -835,7 +_,7 @@
 
       protected void m_5766_(Mob p_29295_, LivingEntity p_29296_) {
          if (p_29295_ instanceof Panda && ((Panda)p_29295_).m_5912_()) {
-            p_29295_.m_6710_(p_29296_);
+            p_29295_.setTarget(p_29296_, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit
          }
 
       }
@@ -937,8 +_,20 @@
          this.f_29320_ = p_29322_;
       }
 
-      protected boolean m_202729_() {
-         return this.f_25684_.m_203117_() || this.f_25684_.m_6060_();
+      public boolean m_8036_() {
+         if (!this.f_29320_.m_6060_()) {
+            return false;
+         } else {
+            BlockPos blockpos = this.m_198172_(this.f_25684_.f_19853_, this.f_25684_, 5);
+            if (blockpos != null) {
+               this.f_25686_ = (double)blockpos.m_123341_();
+               this.f_25687_ = (double)blockpos.m_123342_();
+               this.f_25688_ = (double)blockpos.m_123343_();
+               return true;
+            } else {
+               return this.m_25702_();
+            }
+         }
       }
 
       public boolean m_8045_() {
@@ -965,10 +_,18 @@
                return false;
             } else {
                float f = this.f_29326_.m_146908_() * ((float)Math.PI / 180F);
+               int i = 0;
+               int j = 0;
                float f1 = -Mth.m_14031_(f);
                float f2 = Mth.m_14089_(f);
-               int i = (double)Math.abs(f1) > 0.5D ? Mth.m_14205_((double)f1) : 0;
-               int j = (double)Math.abs(f2) > 0.5D ? Mth.m_14205_((double)f2) : 0;
+               if ((double)Math.abs(f1) > 0.5D) {
+                  i = (int)((float)i + f1 / Math.abs(f1));
+               }
+
+               if ((double)Math.abs(f2) > 0.5D) {
+                  j = (int)((float)j + f2 / Math.abs(f2));
+               }
+
                if (this.f_29326_.f_19853_.m_8055_(this.f_29326_.m_142538_().m_142082_(i, -1, j)).m_60795_()) {
                   return true;
                } else if (this.f_29326_.m_29163_() && this.f_29326_.f_19796_.nextInt(m_186073_(60)) == 1) {
