--- a/net/minecraft/world/entity/animal/Parrot.java
+++ b/net/minecraft/world/entity/animal/Parrot.java
@@ -195,13 +_,13 @@
    private void m_29442_() {
       this.f_29353_ = this.f_29350_;
       this.f_29352_ = this.f_29351_;
-      this.f_29351_ += (float)(!this.f_19861_ && !this.m_20159_() ? 4 : -1) * 0.3F;
+      this.f_29351_ = (float)((double)this.f_29351_ + (double)(!this.f_19861_ && !this.m_20159_() ? 4 : -1) * 0.3D);
       this.f_29351_ = Mth.m_14036_(this.f_29351_, 0.0F, 1.0F);
       if (!this.f_19861_ && this.f_29359_ < 1.0F) {
          this.f_29359_ = 1.0F;
       }
 
-      this.f_29359_ *= 0.9F;
+      this.f_29359_ = (float)((double)this.f_29359_ * 0.9D);
       Vec3 vec3 = this.m_20184_();
       if (!this.f_19861_ && vec3.f_82480_ < 0.0D) {
          this.m_20256_(vec3.m_82542_(1.0D, 0.6D, 1.0D));
@@ -240,7 +_,7 @@
          }
 
          if (!this.f_19853_.f_46443_) {
-            if (this.f_19796_.nextInt(10) == 0) {
+            if (this.f_19796_.nextInt(10) == 0 && (!net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_29414_) || !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityTameEvent(this, p_29414_).isCancelled())) {
                this.m_21828_(p_29414_);
                this.f_19853_.m_7605_(this, (byte)7);
             } else {
@@ -254,7 +_,7 @@
             itemstack.m_41774_(1);
          }
 
-         this.m_7292_(new MobEffectInstance(MobEffects.f_19614_, 900));
+         this.addEffect(new MobEffectInstance(MobEffects.f_19614_, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
          if (p_29414_.m_7500_() || !this.m_20147_()) {
             this.m_6469_(DamageSource.m_19344_(p_29414_), Float.MAX_VALUE);
          }
@@ -351,7 +_,7 @@
    }
 
    public boolean m_6094_() {
-      return true;
+      return super.m_6094_(); // CraftBukkit - collidable API
    }
 
    protected void m_7324_(Entity p_29367_) {
@@ -364,10 +_,7 @@
       if (this.m_6673_(p_29378_)) {
          return false;
       } else {
-         if (!this.f_19853_.f_46443_) {
-            this.m_21839_(false);
-         }
-
+         this.m_21839_(false); // FoxServer TODO
          return super.m_6469_(p_29378_, p_29379_);
       }
    }
