--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -132,7 +_,7 @@
    private void m_29605_(Mob p_29606_) {
       int i = this.m_29631_();
       if (p_29606_.m_6469_(this.m_269291_().m_269333_(this), (float)(1 + i))) {
-         p_29606_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this);
+         p_29606_.addEffect(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
          this.m_5496_(SoundEvents.f_12295_, 1.0F, 1.0F);
       }
 
@@ -145,7 +_,7 @@
             ((ServerPlayer)p_29617_).f_8906_.m_9829_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132162_, 0.0F));
          }
 
-         p_29617_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this);
+         p_29617_.addEffect(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
       }
 
    }
