--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -47,7 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.crafting.Ingredient;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
@@ -91,7 +_,13 @@
       this.m_29725_(0.0D);
    }
 
-   protected void m_8099_() {
+   // CraftBukkit start - code from constructor    \
+   public void initializePathFinderGoals(){
+      this.m_29725_(0.0D);
+   }
+   // CraftBukkit end
+
+   public void m_8099_() {
       this.f_21345_.m_25352_(1, new FloatGoal(this));
       this.f_21345_.m_25352_(1, new ClimbOnTopOfPowderSnowGoal(this, this.f_19853_));
       this.f_21345_.m_25352_(1, new Rabbit.RabbitPanicGoal(this, 2.2D));
@@ -510,7 +_,7 @@
 
       public boolean m_8036_() {
          if (this.f_25600_ <= 0) {
-            if (!this.f_29778_.f_19853_.m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_29778_.f_19853_, this.f_29778_)) {
                return false;
             }
 
@@ -537,9 +_,23 @@
             if (this.f_29780_ && block instanceof CarrotBlock) {
                int i = blockstate.m_61143_(CarrotBlock.f_52244_);
                if (i == 0) {
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(this.f_29778_, blockpos, Blocks.f_50016_.m_49966_()).isCancelled()) {
+                     return;
+                  }
+                  // CraftBukkit end
                   level.m_7731_(blockpos, Blocks.f_50016_.m_49966_(), 2);
                   level.m_46953_(blockpos, true, this.f_29778_);
                } else {
+                  // CraftBukkit starts
+                  if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(
+                          this.f_29778_,
+                          blockpos,
+                          blockstate.m_61124_(CarrotBlock.f_52244_, i - 1)
+                  ).isCancelled()) {
+                     return;
+                  }
+                  // CraftBukkit end
                   level.m_7731_(blockpos, blockstate.m_61124_(CarrotBlock.f_52244_, Integer.valueOf(i - 1)), 2);
                   level.m_46796_(2001, blockpos, Block.m_49956_(blockstate));
                }
