--- a/net/minecraft/world/entity/animal/Squid.java
+++ b/net/minecraft/world/entity/animal/Squid.java
@@ -100,7 +_,7 @@
          if (this.f_19853_.f_46443_) {
             this.f_29940_ = ((float)Math.PI * 2F);
          } else {
-            this.f_29940_ -= ((float)Math.PI * 2F);
+            this.f_29940_ = (float)((double)this.f_29940_ - (Math.PI * 2D));
             if (this.f_19796_.nextInt(10) == 0) {
                this.f_29945_ = 1.0F / (this.f_19796_.nextFloat() + 1.0F) * 0.2F;
             }
@@ -133,7 +_,7 @@
          double d0 = vec3.m_165924_();
          this.f_20883_ += (-((float)Mth.m_14136_(vec3.f_82479_, vec3.f_82481_)) * (180F / (float)Math.PI) - this.f_20883_) * 0.1F;
          this.m_146922_(this.f_20883_);
-         this.f_29951_ += (float)Math.PI * this.f_29946_ * 1.5F;
+         this.f_29951_ = (float)((double)this.f_29951_ + Math.PI * (double)this.f_29946_ * 1.5D);
          this.f_29938_ += (-((float)Mth.m_14136_(d0, vec3.f_82480_)) * (180F / (float)Math.PI) - this.f_29938_) * 0.1F;
       } else {
          this.f_29942_ = Mth.m_14154_(Mth.m_14031_(this.f_29940_)) * (float)Math.PI * 0.25F;
@@ -148,17 +_,14 @@
             this.m_20334_(0.0D, d1 * (double)0.98F, 0.0D);
          }
 
-         this.f_29938_ += (-90.0F - this.f_29938_) * 0.02F;
+         this.f_29938_ = (float)((double)this.f_29938_ + (double)(-90.0F - this.f_29938_) * 0.02D);
       }
 
    }
 
    public boolean m_6469_(DamageSource p_29963_, float p_29964_) {
       if (super.m_6469_(p_29963_, p_29964_) && this.m_142581_() != null) {
-         if (!this.f_19853_.f_46443_) {
-            this.m_29982_();
-         }
-
+         this.m_29982_();
          return true;
       } else {
          return false;
@@ -243,13 +_,13 @@
                double d0 = vec3.m_82553_();
                if (d0 > 0.0D) {
                   vec3.m_82541_();
-                  double d1 = 3.0D;
+                  float f = 3.0F;
                   if (d0 > 5.0D) {
-                     d1 -= (d0 - 5.0D) / 5.0D;
+                     f = (float)((double)f - (d0 - 5.0D) / 5.0D);
                   }
 
-                  if (d1 > 0.0D) {
-                     vec3 = vec3.m_82490_(d1);
+                  if (f > 0.0F) {
+                     vec3 = vec3.m_82490_((double)f);
                   }
                }
 
