--- a/net/minecraft/world/entity/animal/TropicalFish.java
+++ b/net/minecraft/world/entity/animal/TropicalFish.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.entity.animal;
 
 import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -25,7 +_,6 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.biome.Biomes;
-import net.minecraft.world.level.block.Blocks;
 
 public class TropicalFish extends AbstractSchoolingFish {
    public static final String f_149057_ = "BucketVariantTag";
@@ -194,7 +_,7 @@
    }
 
    public static boolean m_186231_(EntityType<TropicalFish> p_186232_, LevelAccessor p_186233_, MobSpawnType p_186234_, BlockPos p_186235_, Random p_186236_) {
-      return p_186233_.m_6425_(p_186235_.m_7495_()).m_205070_(FluidTags.f_13131_) && p_186233_.m_8055_(p_186235_.m_7494_()).m_60713_(Blocks.f_49990_) && (p_186233_.m_204166_(p_186235_).m_203565_(Biomes.f_151785_) || WaterAnimal.m_186237_(p_186232_, p_186233_, p_186234_, p_186235_, p_186236_));
+      return p_186233_.m_6425_(p_186235_.m_7495_()).m_205070_(FluidTags.f_13131_) && (Objects.equals(p_186233_.m_204166_(p_186235_), Optional.of(Biomes.f_151785_)) || WaterAnimal.m_186237_(p_186232_, p_186233_, p_186234_, p_186235_, p_186236_));
    }
 
    static enum Pattern {
