--- a/net/minecraft/world/entity/animal/TropicalFish.java
+++ b/net/minecraft/world/entity/animal/TropicalFish.java
@@ -77,7 +_,7 @@
       this.m_30056_(p_30029_.m_128451_("Variant"));
    }
 
-   private void m_30056_(int p_30057_) {
+   public void m_30056_(int p_30057_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_30011_, p_30057_);
    }
 
@@ -85,7 +_,7 @@
       return !this.f_30010_;
    }
 
-   private int m_30042_() {
+   public int m_30042_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_30011_);
    }
 
