--- a/net/minecraft/world/entity/animal/Turtle.java
+++ b/net/minecraft/world/entity/animal/Turtle.java
@@ -284,7 +_,9 @@
    protected void m_142669_() {
       super.m_142669_();
       if (!this.m_6162_() && this.f_19853_.m_46469_().m_46207_(GameRules.f_46135_)) {
+         this.forceDrops = true; // CraftBukkit
          this.m_20000_(Items.f_42355_, 1);
+         this.forceDrops = false; // CraftBukkit
       }
 
    }
@@ -308,7 +_,9 @@
    }
 
    public void m_8038_(ServerLevel p_30140_, LightningBolt p_30141_) {
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = p_30141_; // CraftBukkit
       this.m_6469_(DamageSource.f_19306_, Float.MAX_VALUE);
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = null; // CraftBukkit
    }
 
    static class TurtleBreedGoal extends BreedGoal {
@@ -467,8 +_,12 @@
                this.f_30274_.m_30236_(true);
             } else if (this.f_30274_.f_30129_ > this.m_183277_(200)) {
                Level level = this.f_30274_.f_19853_;
-               level.m_5594_((Player)null, blockpos, SoundEvents.f_12486_, SoundSource.BLOCKS, 0.3F, 0.9F + level.f_46441_.nextFloat() * 0.2F);
-               level.m_7731_(this.f_25602_.m_7494_(), Blocks.f_50578_.m_49966_().m_61124_(TurtleEggBlock.f_57754_, Integer.valueOf(this.f_30274_.f_19796_.nextInt(4) + 1)), 3);
+
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(this.f_30274_, this.f_25602_.m_7494_(), Blocks.f_50578_.m_49966_().m_61124_(TurtleEggBlock.f_57754_, this.f_30274_.f_19796_.nextInt(4) + 1)).isCancelled()) {
+                  level.m_5594_((Player)null, blockpos, SoundEvents.f_12486_, SoundSource.BLOCKS, 0.3F, 0.9F + level.f_46441_.nextFloat() * 0.2F);
+                  level.m_7731_(this.f_25602_.m_7494_(), Blocks.f_50578_.m_49966_().m_61124_(TurtleEggBlock.f_57754_, Integer.valueOf(this.f_30274_.f_19796_.nextInt(4) + 1)), 3);
+               }
                this.f_30274_.m_30234_(false);
                this.f_30274_.m_30236_(false);
                this.f_30274_.m_27601_(600);
@@ -536,7 +_,7 @@
       }
 
       public boolean m_8036_() {
-         if (!this.m_202729_()) {
+         if (this.f_25684_.m_142581_() == null && !this.f_25684_.m_6060_()) {
             return false;
          } else {
             BlockPos blockpos = this.m_198172_(this.f_25684_.f_19853_, this.f_25684_, 7);
