--- a/net/minecraft/world/entity/animal/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/WaterAnimal.java
@@ -71,6 +_,6 @@
    public static boolean m_186237_(EntityType<? extends WaterAnimal> p_186238_, LevelAccessor p_186239_, MobSpawnType p_186240_, BlockPos p_186241_, Random p_186242_) {
       int i = p_186239_.m_5736_();
       int j = i - 13;
-      return p_186241_.m_123342_() >= j && p_186241_.m_123342_() <= i && p_186239_.m_6425_(p_186241_.m_7495_()).m_205070_(FluidTags.f_13131_) && p_186239_.m_8055_(p_186241_.m_7494_()).m_60713_(Blocks.f_49990_);
+      return p_186239_.m_6425_(p_186241_.m_7495_()).m_205070_(FluidTags.f_13131_) && p_186239_.m_8055_(p_186241_.m_7494_()).m_60713_(Blocks.f_49990_) && p_186241_.m_123342_() >= j && p_186241_.m_123342_() <= i;
    }
 }
