--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -68,6 +_,8 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class Wolf extends TamableAnimal implements NeutralMob {
    private static final EntityDataAccessor<Boolean> f_30358_ = SynchedEntityData.m_135353_(Wolf.class, EntityDataSerializers.f_135035_);
@@ -308,7 +_,7 @@
       super.m_7105_(p_30443_);
       if (p_30443_) {
          this.m_21051_(Attributes.f_22276_).m_22100_(20.0D);
-         this.m_21153_(20.0F);
+         this.m_21153_(this.m_21233_()); // CraftBukkit - 20.0 -> getMaxHealth()
       } else {
          this.m_21051_(Attributes.f_22276_).m_22100_(8.0D);
       }
@@ -329,7 +_,7 @@
                   itemstack.m_41774_(1);
                }
 
-               this.m_5634_((float)item.m_41473_().m_38744_());
+               this.heal((float)itemstack.getFoodProperties(this).m_38744_(), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
                this.m_146859_(GameEvent.f_157771_, this.m_146901_());
                return InteractionResult.SUCCESS;
             }
@@ -340,7 +_,7 @@
                   this.m_21839_(!this.m_21827_());
                   this.f_20899_ = false;
                   this.f_21344_.m_26573_();
-                  this.m_6710_((LivingEntity)null);
+                  this.setTarget((LivingEntity)null, EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
                   return InteractionResult.SUCCESS;
                }
 
@@ -361,7 +_,7 @@
                itemstack.m_41774_(1);
             }
 
-            if (this.f_19796_.nextInt(3) == 0) {
+            if (this.f_19796_.nextInt(3) == 0 && (!net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_30412_) || !CraftEventFactory.callEntityTameEvent(this, p_30412_).isCancelled())) {
                this.m_21828_(p_30412_);
                this.f_21344_.m_26573_();
                this.m_6710_((LivingEntity)null);
@@ -401,7 +_,7 @@
 
    public boolean m_6898_(ItemStack p_30440_) {
       Item item = p_30440_.m_41720_();
-      return item.m_41472_() && item.m_41473_().m_38746_();
+      return item.m_41472_() && p_30440_.getFoodProperties(this).m_38746_();
    }
 
    public int m_5792_() {
@@ -534,7 +_,7 @@
          super.m_8037_();
       }
    }
-
+   
    class WolfPanicGoal extends PanicGoal {
       public WolfPanicGoal(double p_203124_) {
          super(Wolf.this, p_203124_);
@@ -544,4 +_,16 @@
          return this.f_25684_.m_203117_() || this.f_25684_.m_6060_();
       }
    }
+
+   // CraftBukkit - add overriden version
+   @Override
+   public boolean setTarget(LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fire) {
+      if (!super.setTarget(entityliving, reason, fire)) {
+         return false;
+      }
+      entityliving = m_5448_();
+
+      return true;
+   }
+   // CraftBukkit end
 }
