--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -18,6 +_,13 @@
       public boolean m_6102_() {
          return false;
       }
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         throw new UnsupportedOperationException("Not supported yet.");
+      }
+      // CraftBukkit end
    };
 
    void m_213846_(Component p_230797_);
@@ -31,4 +_,6 @@
    default boolean m_142559_() {
       return false;
    }
+
+   org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper); // CraftBukkit
 }
