--- a/net/minecraft/commands/SharedSuggestionProvider.java
+++ b/net/minecraft/commands/SharedSuggestionProvider.java
@@ -72,6 +_,8 @@
 
    boolean m_6761_(int p_82986_);
 
+   boolean hasPermission(int p_82986_, String bukkitPermission); // CatServer - 1.20.1
+
    static <T> void m_82944_(Iterable<T> p_82945_, String p_82946_, Function<T, ResourceLocation> p_82947_, Consumer<T> p_82948_) {
       boolean flag = p_82946_.indexOf(58) > -1;
 
@@ -82,7 +_,7 @@
             if (m_82949_(p_82946_, s)) {
                p_82948_.accept(t);
             }
-         } else if (m_82949_(p_82946_, resourcelocation.m_135827_()) || resourcelocation.m_135827_().equals("minecraft") && m_82949_(p_82946_, resourcelocation.m_135815_())) {
+         } else if (m_82949_(p_82946_, resourcelocation.m_135827_()) || m_82949_(p_82946_, resourcelocation.m_135815_())) { // Forge: Removed "minecraft" namespace requirement for path searching to allow modded path suggestions to show up
             p_82948_.accept(t);
          }
       }
