--- ../src-base/minecraft/net/minecraft/scoreboard/ScoreCriteriaColored.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/ScoreCriteriaColored.java
@@ -22,8 +22,8 @@
         return false;
     }
 
-    public IScoreCriteria.EnumRenderType getRenderType()
+    public EnumRenderType getRenderType()
     {
-        return IScoreCriteria.EnumRenderType.INTEGER;
+        return EnumRenderType.INTEGER;
     }
 }
