--- ../src-base/minecraft/net/minecraft/scoreboard/ScoreCriteriaHealth.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/ScoreCriteriaHealth.java
@@ -12,8 +12,8 @@
         return true;
     }
 
-    public IScoreCriteria.EnumRenderType getRenderType()
+    public EnumRenderType getRenderType()
     {
-        return IScoreCriteria.EnumRenderType.HEARTS;
+        return EnumRenderType.HEARTS;
     }
 }
