--- ../src-base/minecraft/net/minecraft/scoreboard/ScorePlayerTeam.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/ScorePlayerTeam.java
@@ -18,10 +18,10 @@
     private String suffix = "";
     private boolean allowFriendlyFire = true;
     private boolean canSeeFriendlyInvisibles = true;
-    private Team.EnumVisible nameTagVisibility = Team.EnumVisible.ALWAYS;
-    private Team.EnumVisible deathMessageVisibility = Team.EnumVisible.ALWAYS;
+    private EnumVisible nameTagVisibility = EnumVisible.ALWAYS;
+    private EnumVisible deathMessageVisibility = EnumVisible.ALWAYS;
     private TextFormatting color = TextFormatting.RESET;
-    private Team.CollisionRule collisionRule = Team.CollisionRule.ALWAYS;
+    private CollisionRule collisionRule = CollisionRule.ALWAYS;
 
     public ScorePlayerTeam(Scoreboard scoreboardIn, String name)
     {
@@ -119,34 +119,34 @@
         this.scoreboard.broadcastTeamInfoUpdate(this);
     }
 
-    public Team.EnumVisible getNameTagVisibility()
+    public EnumVisible getNameTagVisibility()
     {
         return this.nameTagVisibility;
     }
 
-    public Team.EnumVisible getDeathMessageVisibility()
+    public EnumVisible getDeathMessageVisibility()
     {
         return this.deathMessageVisibility;
     }
 
-    public void setNameTagVisibility(Team.EnumVisible visibility)
+    public void setNameTagVisibility(EnumVisible visibility)
     {
         this.nameTagVisibility = visibility;
         this.scoreboard.broadcastTeamInfoUpdate(this);
     }
 
-    public void setDeathMessageVisibility(Team.EnumVisible visibility)
+    public void setDeathMessageVisibility(EnumVisible visibility)
     {
         this.deathMessageVisibility = visibility;
         this.scoreboard.broadcastTeamInfoUpdate(this);
     }
 
-    public Team.CollisionRule getCollisionRule()
+    public CollisionRule getCollisionRule()
     {
         return this.collisionRule;
     }
 
-    public void setCollisionRule(Team.CollisionRule rule)
+    public void setCollisionRule(CollisionRule rule)
     {
         this.collisionRule = rule;
         this.scoreboard.broadcastTeamInfoUpdate(this);
