--- ../src-base/minecraft/net/minecraft/scoreboard/Team.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/Team.java
@@ -38,9 +38,9 @@
 
     public abstract Collection<String> getMembershipCollection();
 
-    public abstract Team.EnumVisible getDeathMessageVisibility();
+    public abstract EnumVisible getDeathMessageVisibility();
 
-    public abstract Team.CollisionRule getCollisionRule();
+    public abstract CollisionRule getCollisionRule();
 
     public static enum CollisionRule
     {
@@ -49,7 +49,7 @@
         HIDE_FOR_OTHER_TEAMS("pushOtherTeams", 2),
         HIDE_FOR_OWN_TEAM("pushOwnTeam", 3);
 
-        private static final Map<String, Team.CollisionRule> nameMap = Maps.<String, Team.CollisionRule>newHashMap();
+        private static final Map<String, CollisionRule> nameMap = Maps.<String, CollisionRule>newHashMap();
         public final String name;
         public final int id;
 
@@ -72,7 +72,7 @@
 
         static
         {
-            for (Team.CollisionRule team$collisionrule : values())
+            for (CollisionRule team$collisionrule : values())
             {
                 nameMap.put(team$collisionrule.name, team$collisionrule);
             }
@@ -86,7 +86,7 @@
         HIDE_FOR_OTHER_TEAMS("hideForOtherTeams", 2),
         HIDE_FOR_OWN_TEAM("hideForOwnTeam", 3);
 
-        private static final Map<String, Team.EnumVisible> nameMap = Maps.<String, Team.EnumVisible>newHashMap();
+        private static final Map<String, EnumVisible> nameMap = Maps.<String, EnumVisible>newHashMap();
         public final String internalName;
         public final int id;
 
@@ -109,7 +109,7 @@
 
         static
         {
-            for (Team.EnumVisible team$enumvisible : values())
+            for (EnumVisible team$enumvisible : values())
             {
                 nameMap.put(team$enumvisible.internalName, team$enumvisible);
             }
