--- ../src-base/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
+++ ../src-work/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
@@ -36,10 +36,10 @@
 {
     protected GuiScreen parentScreen;
     protected String screenTitle = "Select world";
-    private GuiStats.StatsGeneral generalStats;
-    private GuiStats.StatsItem itemStats;
-    private GuiStats.StatsBlock blockStats;
-    private GuiStats.StatsMobsList mobStats;
+    private StatsGeneral generalStats;
+    private StatsItem itemStats;
+    private StatsBlock blockStats;
+    private StatsMobsList mobStats;
     private final StatisticsManager stats;
     private GuiSlot displaySlot;
     private boolean doesGuiPauseGame = true;
@@ -69,13 +69,13 @@
 
     public void initLists()
     {
-        this.generalStats = new GuiStats.StatsGeneral(this.mc);
+        this.generalStats = new StatsGeneral(this.mc);
         this.generalStats.registerScrollButtons(1, 1);
-        this.itemStats = new GuiStats.StatsItem(this.mc);
+        this.itemStats = new StatsItem(this.mc);
         this.itemStats.registerScrollButtons(1, 1);
-        this.blockStats = new GuiStats.StatsBlock(this.mc);
+        this.blockStats = new StatsBlock(this.mc);
         this.blockStats.registerScrollButtons(1, 1);
-        this.mobStats = new GuiStats.StatsMobsList(this.mc);
+        this.mobStats = new StatsMobsList(this.mc);
         this.mobStats.registerScrollButtons(1, 1);
     }
 
@@ -483,7 +483,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    class StatsBlock extends GuiStats.Stats
+    class StatsBlock extends Stats
     {
         public StatsBlock(Minecraft p_i47554_2_)
         {
@@ -709,7 +709,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    class StatsItem extends GuiStats.Stats
+    class StatsItem extends Stats
     {
         public StatsItem(Minecraft mcIn)
         {
