--- a/net/minecraft/stats/RecipeBook.java
+++ b/net/minecraft/stats/RecipeBook.java
@@ -9,7 +_,7 @@
 import net.minecraft.world.item.crafting.Recipe;
 
 public class RecipeBook {
-   protected final Set<ResourceLocation> f_12680_ = Sets.newHashSet();
+   public final Set<ResourceLocation> f_12680_ = Sets.newHashSet();
    protected final Set<ResourceLocation> f_12681_ = Sets.newHashSet();
    private final RecipeBookSettings f_12682_ = new RecipeBookSettings();
 
