--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -17,6 +_,7 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.crafting.RecipeManager;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 import org.slf4j.Logger;
 
 public class ServerRecipeBook extends RecipeBook {
@@ -29,7 +_,7 @@
 
       for(Recipe<?> recipe : p_12792_) {
          ResourceLocation resourcelocation = recipe.m_6423_();
-         if (!this.f_12680_.contains(resourcelocation) && !recipe.m_5598_()) {
+         if (!this.f_12680_.contains(resourcelocation) && !recipe.m_5598_() && CraftEventFactory.handlePlayerRecipeListUpdateEvent(p_12793_, resourcelocation)) { // CraftBukkit
             this.m_12702_(resourcelocation);
             this.m_12719_(resourcelocation);
             list.add(resourcelocation);
@@ -60,6 +_,7 @@
    }
 
    private void m_12801_(ClientboundRecipePacket.State p_12802_, ServerPlayer p_12803_, List<ResourceLocation> p_12804_) {
+      if (p_12803_.f_8906_ == null) return; // SPIGOT-4478 during PlayerLoginEvent
       p_12803_.f_8906_.m_141995_(new ClientboundRecipePacket(p_12802_, p_12804_, Collections.emptyList(), this.m_12684_()));
    }
 
