--- a/net/minecraft/stats/ServerStatisticsManager.java
+++ b/net/minecraft/stats/ServerStatisticsManager.java
@@ -43,6 +_,13 @@
    public ServerStatisticsManager(MinecraftServer p_i45306_1_, File p_i45306_2_) {
       this.field_150890_c = p_i45306_1_;
       this.field_150887_d = p_i45306_2_;
+      // Spigot start
+      for ( Map.Entry<ResourceLocation, Integer> entry : org.spigotmc.SpigotConfig.forcedStats.entrySet() )
+      {
+         Stat<ResourceLocation> wrapper = Stats.field_199092_j.func_199076_b( entry.getKey() );
+         this.field_150875_a.put( wrapper, entry.getValue().intValue() );
+      }
+      // Spigot end
       if (p_i45306_2_.isFile()) {
          try {
             this.func_199062_a(p_i45306_1_.func_195563_aC(), FileUtils.readFileToString(p_i45306_2_));
@@ -56,6 +_,7 @@
    }
 
    public void func_150883_b() {
+      if ( org.spigotmc.SpigotConfig.disableStatSaving ) return; // Spigot
       try {
          FileUtils.writeStringToFile(this.field_150887_d, this.func_199061_b());
       } catch (IOException ioexception) {
@@ -65,6 +_,7 @@
    }
 
    public void func_150873_a(PlayerEntity p_150873_1_, Stat<?> p_150873_2_, int p_150873_3_) {
+      if ( org.spigotmc.SpigotConfig.disableStatSaving ) return; // Spigot
       super.func_150873_a(p_150873_1_, p_150873_2_, p_150873_3_);
       this.field_150888_e.add(p_150873_2_);
    }
