--- a/net/minecraft/stats/StatisticsManager.java
+++ b/net/minecraft/stats/StatisticsManager.java
@@ -16,6 +_,14 @@
 
    public void func_150871_b(PlayerEntity p_150871_1_, Stat<?> p_150871_2_, int p_150871_3_) {
       int i = (int)Math.min((long)this.func_77444_a(p_150871_2_) + (long)p_150871_3_, 2147483647L);
+
+      // CraftBukkit start - fire Statistic events
+      org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.handleStatisticsIncrease(p_150871_1_, p_150871_2_, this.func_77444_a(p_150871_2_), i);
+      if (cancellable != null && cancellable.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
+
       this.func_150873_a(p_150871_1_, p_150871_2_, i);
    }
 
