--- a/net/minecraft/entity/boss/dragon/phase/FlamingSittingPhase.java
+++ b/net/minecraft/entity/boss/dragon/phase/FlamingSittingPhase.java
@@ -59,7 +_,7 @@
          double d3 = d2;
          BlockPos.Mutable blockpos$mutable = new BlockPos.Mutable(d0, d2, d1);
 
-         while(this.field_188661_a.field_70170_p.func_175623_d(blockpos$mutable)) {
+         while(this.field_188661_a.field_70170_p.func_175623_d(blockpos$mutable) && d2 > 0) { //Paper
             --d3;
             if (d3 < 0.0D) {
                d3 = d2;
