--- a/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
+++ b/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
@@ -3,6 +_,8 @@
 import net.minecraft.entity.boss.dragon.EnderDragonEntity;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftEnderDragon;
+import org.bukkit.event.entity.EnderDragonChangePhaseEvent;
 
 public class PhaseManager {
    private static final Logger field_188759_a = LogManager.getLogger();
@@ -20,6 +_,16 @@
          if (this.field_188762_d != null) {
             this.field_188762_d.func_188658_e();
          }
+
+         // CraftBukkit start - Call EnderDragonChangePhaseEvent
+         EnderDragonChangePhaseEvent event = new EnderDragonChangePhaseEvent((CraftEnderDragon) this.field_188760_b.getBukkitEntity(),
+             (this.field_188762_d == null) ? null : CraftEnderDragon.getBukkitPhase(this.field_188762_d.func_188652_i()), CraftEnderDragon.getBukkitPhase(p_188758_1_));
+         this.field_188760_b.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return;
+         }
+         if (event.getNewPhase() != org.bukkit.entity.EnderDragon.Phase.CUSTOM) p_188758_1_ = CraftEnderDragon.getMinecraftPhase(event.getNewPhase()); // LoliServer
+         // CraftBukkit end
 
          this.field_188762_d = this.func_188757_b(p_188758_1_);
          if (!this.field_188760_b.field_70170_p.field_72995_K) {
