--- ../src-base/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
@@ -22,7 +22,11 @@
     {
         File file1 = this.getWorldDirectory();
 
-        if (provider.getSaveFolder() != null)
+        // Svarka
+        // To workaround the issue of Bukkit relying on every world having a seperate container
+        // we won't be generating a DIMXX folder for chunk loaders since this name is already generated
+        // for the world container with provider.getSaveFolder().
+        /*if (provider.getSaveFolder() != null)
         {
             File file3 = new File(file1, provider.getSaveFolder());
             file3.mkdirs();
@@ -31,7 +35,8 @@
         else
         {
             return new AnvilChunkLoader(file1, this.dataFixer);
-        }
+        }*/
+        return new AnvilChunkLoader(file1,this.dataFixer);
     }
 
     public void saveWorldInfoWithPlayer(WorldInfo worldInformation, @Nullable NBTTagCompound tagCompound)
