--- ../src-base/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
@@ -21,23 +21,21 @@
     public IChunkLoader getChunkLoader(WorldProvider provider)
     {
         File file1 = this.getWorldDirectory();
+        // To workaround the issue of Bukkit relying on every world having a seperate container
+        // we won't be generating a DIMXX folder for chunk loaders since this name is already generated
+        // for the world container with provider.getSaveFolder().
 
-        if (provider instanceof WorldProviderHell)
+        /*if (provider.getSaveFolder() != null)
         {
-            File file3 = new File(file1, "DIM-1");
+            File file3 = new File(file1, provider.getSaveFolder());
             file3.mkdirs();
             return new AnvilChunkLoader(file3, this.dataFixer);
         }
-        else if (provider instanceof WorldProviderEnd)
-        {
-            File file2 = new File(file1, "DIM1");
-            file2.mkdirs();
-            return new AnvilChunkLoader(file2, this.dataFixer);
-        }
         else
         {
             return new AnvilChunkLoader(file1, this.dataFixer);
-        }
+        }*/
+        return new AnvilChunkLoader(file1,this.dataFixer);
     }
 
     public void saveWorldInfoWithPlayer(WorldInfo worldInformation, @Nullable NBTTagCompound tagCompound)
