--- ../src-base/minecraft/net/minecraft/world/chunk/storage/ChunkLoader.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/ChunkLoader.java
@@ -10,11 +10,11 @@
 
 public class ChunkLoader
 {
-    public static ChunkLoader.AnvilConverterData load(NBTTagCompound nbt)
+    public static AnvilConverterData load(NBTTagCompound nbt)
     {
         int i = nbt.getInteger("xPos");
         int j = nbt.getInteger("zPos");
-        ChunkLoader.AnvilConverterData chunkloader$anvilconverterdata = new ChunkLoader.AnvilConverterData(i, j);
+        AnvilConverterData chunkloader$anvilconverterdata = new AnvilConverterData(i, j);
         chunkloader$anvilconverterdata.blocks = nbt.getByteArray("Blocks");
         chunkloader$anvilconverterdata.data = new NibbleArrayReader(nbt.getByteArray("Data"), 7);
         chunkloader$anvilconverterdata.skyLight = new NibbleArrayReader(nbt.getByteArray("SkyLight"), 7);
@@ -37,7 +37,7 @@
         return chunkloader$anvilconverterdata;
     }
 
-    public static void convertToAnvilFormat(ChunkLoader.AnvilConverterData converterData, NBTTagCompound compound, BiomeProvider provider)
+    public static void convertToAnvilFormat(AnvilConverterData converterData, NBTTagCompound compound, BiomeProvider provider)
     {
         compound.setInteger("xPos", converterData.x);
         compound.setInteger("zPos", converterData.z);
