--- ../src-base/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
@@ -27,6 +27,22 @@
         }
     }
 
+    public ExtendedBlockStorage(int y, boolean flag, char[] blockIds) {
+        this.yBase = y;
+        this.data = new BlockStateContainer();
+        for (int i = 0; i < blockIds.length; i++) {
+            int xx = i & 15;
+            int yy = (i >> 8) & 15;
+            int zz = (i >> 4) & 15;
+            this.data.set(xx, yy, zz, Block.BLOCK_STATE_IDS.getByValue(blockIds[i]));
+        }
+        this.blockLight = new NibbleArray();
+        if (flag) {
+            this.skyLight = new NibbleArray();
+        }
+        recalculateRefCounts();
+    }
+
     public IBlockState get(int x, int y, int z)
     {
         return this.data.get(x, y, z);
@@ -34,6 +50,8 @@
 
     public void set(int x, int y, int z, IBlockState state)
     {
+        if (state instanceof net.minecraftforge.common.property.IExtendedBlockState)
+            state = ((net.minecraftforge.common.property.IExtendedBlockState) state).getClean();
         IBlockState iblockstate = this.get(x, y, z);
         Block block = iblockstate.getBlock();
         Block block1 = state.getBlock();
@@ -63,7 +81,8 @@
 
     public boolean isEmpty()
     {
-        return this.blockRefCount == 0;
+        // return this.blockRefCount == 0;
+        return false; // CraftBukkit - MC-80966
     }
 
     public boolean needsRandomTick()
