--- ../src-base/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
@@ -14,7 +14,23 @@
     private final BlockStateContainer data;
     private NibbleArray blocklightArray;
     private NibbleArray skylightArray;
-
+    // CraftBukkit start
+    public ExtendedBlockStorage(int y, boolean flag, char[] blockIds) {
+    	this.yBase = y;
+    	this.data = new BlockStateContainer();
+    	for (int i = 0; i < blockIds.length; i++) {
+    		int xx = i & 15;
+    		int yy = (i >> 8) & 15;
+    		int zz = (i >> 4) & 15;
+    		this.data.set(xx, yy, zz, Block.BLOCK_STATE_IDS.getByValue(blockIds[i]));
+    	}
+    	this.blocklightArray = new NibbleArray();
+    	if (flag) {
+    		this.skylightArray = new NibbleArray();
+    	}
+    	removeInvalidBlocks();
+    }
+    // CraftBukkit end
     public ExtendedBlockStorage(int y, boolean storeSkylight)
     {
         this.yBase = y;
