--- ../src-base/minecraft/net/minecraft/world/chunk/storage/IChunkLoader.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/IChunkLoader.java
@@ -20,4 +20,10 @@
     void flush();
 
     boolean isChunkGeneratedAt(int x, int z);
+
+    // CatServer start - Add spigot method
+    default void saveChunk(World world, Chunk chunk, boolean unloaded) throws MinecraftException, IOException {
+        saveChunk(world, chunk);
+    }
+    // CatServer end
 }
