--- a/net/minecraft/entity/passive/horse/TraderLlamaEntity.java
+++ b/net/minecraft/entity/passive/horse/TraderLlamaEntity.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class TraderLlamaEntity extends LlamaEntity {
    private int field_213802_bJ = 47999;
@@ -135,7 +_,7 @@
       }
 
       public void func_75249_e() {
-         this.field_75299_d.func_70624_b(this.field_220801_c);
+         this.field_75299_d.setGoalTarget(this.field_220801_c, TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit
          Entity entity = this.field_220800_b.func_110166_bE();
          if (entity instanceof WanderingTraderEntity) {
             this.field_220802_d = ((WanderingTraderEntity)entity).func_142015_aE();
