--- a/net/minecraft/enchantment/DamageEnchantment.java
+++ b/net/minecraft/enchantment/DamageEnchantment.java
@@ -8,6 +_,7 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.potion.Effects;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class DamageEnchantment extends Enchantment {
    private static final String[] field_77359_A = new String[]{"all", "undead", "arthropods"};
@@ -56,7 +_,7 @@
          LivingEntity livingentity = (LivingEntity)p_151368_2_;
          if (this.field_77361_a == 2 && livingentity.func_70668_bt() == CreatureAttribute.field_223224_c_) {
             int i = 20 + p_151368_1_.func_70681_au().nextInt(10 * p_151368_3_);
-            livingentity.func_195064_c(new EffectInstance(Effects.field_76421_d, i, 3));
+            livingentity.addEffect(new EffectInstance(Effects.field_76421_d, i, 3), Cause.ATTACK);  // CraftBukkit
          }
       }
 
