--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -28,6 +_,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class Bat extends AmbientCreature {
    public static final float f_148698_ = 74.48451F;
    public static final int f_148699_ = Mth.m_14167_(2.4166098F);
@@ -125,13 +_,13 @@
                this.f_20885_ = (float)this.f_19796_.m_188503_(360);
             }
 
-            if (this.m_9236_().m_45946_(f_27408_, this) != null) {
+            if (this.m_9236_().m_45946_(f_27408_, this) != null && CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
                this.m_27456_(false);
                if (!flag) {
                   this.m_9236_().m_5898_((Player)null, 1025, blockpos, 0);
                }
             }
-         } else {
+         } else if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
             this.m_27456_(false);
             if (!flag) {
                this.m_9236_().m_5898_((Player)null, 1025, blockpos, 0);
@@ -157,7 +_,11 @@
          this.f_20902_ = 0.5F;
          this.m_146922_(this.m_146908_() + f1);
          if (this.f_19796_.m_188503_(100) == 0 && this.m_9236_().m_8055_(blockpos1).m_60796_(this.m_9236_(), blockpos1)) {
-            this.m_27456_(true);
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (CraftEventFactory.handleBatToggleSleepEvent(this, false)) {
+               this.m_27456_(true);
+            }
+            // CraftBukkit End
          }
       }
 
@@ -178,7 +_,7 @@
       if (this.m_6673_(p_27424_)) {
          return false;
       } else {
-         if (!this.m_9236_().f_46443_ && this.m_27452_()) {
+         if (!this.m_9236_().f_46443_ && this.m_27452_() && CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
             this.m_27456_(false);
          }
 
