--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -27,6 +_,7 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class Bat extends AmbientCreature {
    public static final float f_148698_ = 74.48451F;
@@ -123,16 +_,22 @@
             }
 
             if (this.f_19853_.m_45946_(f_27408_, this) != null) {
+               // CraftBukkit Start - Call BatToggleSleepEvent
+               if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+                  this.m_27456_(false);
+                  if (!flag) {
+                     this.f_19853_.m_5898_((Player) null, 1025, blockpos, 0);
+                  }
+               }
+               // CraftBukkit End
+            }
+         } else {
+            if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
                this.m_27456_(false);
                if (!flag) {
                   this.f_19853_.m_5898_((Player)null, 1025, blockpos, 0);
                }
             }
-         } else {
-            this.m_27456_(false);
-            if (!flag) {
-               this.f_19853_.m_5898_((Player)null, 1025, blockpos, 0);
-            }
          }
       } else {
          if (this.f_27409_ != null && (!this.f_19853_.m_46859_(this.f_27409_) || this.f_27409_.m_123342_() <= this.f_19853_.m_141937_())) {
@@ -154,7 +_,11 @@
          this.f_20902_ = 0.5F;
          this.m_146922_(this.m_146908_() + f1);
          if (this.f_19796_.nextInt(100) == 0 && this.f_19853_.m_8055_(blockpos1).m_60796_(this.f_19853_, blockpos1)) {
-            this.m_27456_(true);
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (CraftEventFactory.handleBatToggleSleepEvent(this, false)) {
+               this.m_27456_(true);
+            }
+            // CraftBukkit End
          }
       }
 
@@ -180,7 +_,11 @@
          return false;
       } else {
          if (!this.f_19853_.f_46443_ && this.m_27452_()) {
-            this.m_27456_(false);
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+               this.m_27456_(false);
+            }
+            // CraftBukkit End - Call BatToggleSleepEvent
          }
 
          return super.m_6469_(p_27424_, p_27425_);
