--- a/net/minecraft/world/entity/monster/piglin/PiglinBruteAi.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinBruteAi.java
@@ -108,7 +_,7 @@
 
    private static Optional<? extends LivingEntity> m_35091_(AbstractPiglin p_35092_, MemoryModuleType<? extends LivingEntity> p_35093_) {
       return p_35092_.m_6274_().m_21952_(p_35093_).filter((p_35108_) -> {
-         return p_35108_.m_19950_(p_35092_, 12.0D);
+         return p_35108_.closerToCenterThan(p_35092_, 12.0D);
       });
    }
 
