--- a/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -7,6 +_,8 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.Random;
+import java.util.stream.Collectors;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
@@ -56,8 +_,6 @@
 import net.minecraft.world.entity.monster.hoglin.Hoglin;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.schedule.Activity;
-import net.minecraft.world.item.ArmorItem;
-import net.minecraft.world.item.ArmorMaterials;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
@@ -68,6 +_,9 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.entity.PiglinBarterEvent;
 
 public class PiglinAi {
    public static final int f_149940_ = 8;
@@ -211,23 +_,27 @@
    protected static void m_34843_(Piglin p_34844_, ItemEntity p_34845_) {
       m_35006_(p_34844_);
       ItemStack itemstack;
-      if (p_34845_.m_32055_().m_150930_(Items.f_42587_)) {
+      // CraftBukkit start
+      if (p_34845_.m_32055_().m_150930_(Items.f_42587_) && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPickupItemEvent(p_34844_, p_34845_, 0, false).isCancelled()) {
          p_34844_.m_7938_(p_34845_, p_34845_.m_32055_().m_41613_());
          itemstack = p_34845_.m_32055_();
          p_34845_.m_146870_();
-      } else {
+      } else if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPickupItemEvent(p_34844_, p_34845_, p_34845_.m_32055_().m_41613_() - 1, false).isCancelled()) {
          p_34844_.m_7938_(p_34845_, 1);
          itemstack = m_34822_(p_34845_);
+      } else {
+         return;
       }
+      // CraftBukkit end
 
-      if (m_149965_(itemstack)) {
+      if (isLovedItem(itemstack, p_34844_)) { // CraftBukkit - Changes to allow for custom payment in bartering
          p_34844_.m_6274_().m_21936_(MemoryModuleType.f_26337_);
          m_34932_(p_34844_, itemstack);
          m_34938_(p_34844_);
       } else if (m_149969_(itemstack) && !m_35018_(p_34844_)) {
          m_35014_(p_34844_);
       } else {
-         boolean flag = p_34844_.m_21540_(itemstack);
+         boolean flag = p_34844_.equipItemIfPossible(itemstack, p_34845_); // CraftBukkit
          if (!flag) {
             m_34952_(p_34844_, itemstack);
          }
@@ -258,9 +_,14 @@
       ItemStack itemstack = p_34868_.m_21120_(InteractionHand.OFF_HAND);
       p_34868_.m_21008_(InteractionHand.OFF_HAND, ItemStack.f_41583_);
       if (p_34868_.m_34667_()) {
-         boolean flag = m_149967_(itemstack);
+         boolean flag = isBarterCurrency( itemstack, p_34868_ );
          if (p_34869_ && flag) {
-            m_34860_(p_34868_, m_34996_(p_34868_));
+            // CraftBukkit start
+            PiglinBarterEvent event = CraftEventFactory.callPiglinBarterEvent(p_34868_, m_34996_(p_34868_), itemstack);
+            if (!event.isCancelled()) {
+               m_34860_(p_34868_, event.getOutcome().stream().map( CraftItemStack::asNMSCopy).collect( Collectors.toList()));
+            }
+            // CraftBukkit end
          } else if (!flag) {
             boolean flag1 = p_34868_.m_21540_(itemstack);
             if (!flag1) {
@@ -271,7 +_,7 @@
          boolean flag2 = p_34868_.m_21540_(itemstack);
          if (!flag2) {
             ItemStack itemstack1 = p_34868_.m_21205_();
-            if (m_149965_(itemstack1)) {
+            if (isLovedItem(itemstack1, p_34868_)) { // CraftBukkit - Changes to allow for custom payment in bartering
                m_34952_(p_34868_, itemstack1);
             } else {
                m_34860_(p_34868_, Collections.singletonList(itemstack1));
@@ -345,7 +_,7 @@
          return false;
       } else if (m_35024_(p_34858_) && p_34858_.m_6274_().m_21874_(MemoryModuleType.f_26372_)) {
          return false;
-      } else if (m_149967_(p_34859_)) {
+      } else if (isBarterCurrency(p_34859_, p_34858_)) { // CraftBukkit
          return m_35028_(p_34858_);
       } else {
          boolean flag = p_34858_.m_34780_(p_34859_);
@@ -361,6 +_,12 @@
       }
    }
 
+   // CraftBukkit start - Added method to allow checking for custom payment items
+   protected static boolean isLovedItem(ItemStack itemstack, Piglin piglin) {
+      return m_149965_(itemstack) || (piglin.interestItems.contains(itemstack.m_41720_()) || piglin.allowedBarterItems.contains(itemstack.m_41720_()));
+   }
+   // CraftBukkit end
+
    protected static boolean m_149965_(ItemStack p_149966_) {
       return p_149966_.m_204117_(ItemTags.f_13151_);
    }
@@ -384,7 +_,7 @@
       Brain<Piglin> brain = p_34999_.m_6274_();
       if (brain.m_21874_(MemoryModuleType.f_26351_)) {
          LivingEntity livingentity = brain.m_21952_(MemoryModuleType.f_26351_).get();
-         return p_34999_.m_19950_(livingentity, 6.0D);
+         return p_34999_.closerToCenterThan(livingentity, 6.0D);
       } else {
          return false;
       }
@@ -445,7 +_,7 @@
    }
 
    protected static boolean m_34909_(Piglin p_34910_, ItemStack p_34911_) {
-      return !m_35024_(p_34910_) && !m_35020_(p_34910_) && p_34910_.m_34667_() && m_149967_(p_34911_);
+      return !m_35024_(p_34910_) && !m_35020_(p_34910_) && p_34910_.m_34667_() && isBarterCurrency(p_34911_, p_34910_); // CraftBukkit
    }
 
    protected static void m_34837_(Piglin p_34838_, LivingEntity p_34839_) {
@@ -483,6 +_,13 @@
       }
    }
 
+   // CraftBukkit start - Changes to allow custom payment for bartering
+   private static boolean isBarterCurrency(ItemStack itemstack, Piglin piglin) {
+      return m_149967_(itemstack) || piglin.allowedBarterItems.contains(itemstack.m_41720_());
+   }
+   // CraftBukkit end
+
+
    protected static void m_34826_(AbstractPiglin p_34827_, LivingEntity p_34828_) {
       if (!p_34827_.m_6274_().m_21954_(Activity.f_37991_)) {
          if (Sensor.m_182377_(p_34827_, p_34828_)) {
@@ -526,7 +_,7 @@
 
    private static boolean m_35002_(Piglin p_35003_) {
       Brain<Piglin> brain = p_35003_.m_6274_();
-      return !brain.m_21874_(MemoryModuleType.f_26383_) ? false : brain.m_21952_(MemoryModuleType.f_26383_).get().m_19950_(p_35003_, 12.0D);
+      return !brain.m_21874_(MemoryModuleType.f_26383_) ? false : brain.m_21952_(MemoryModuleType.f_26383_).get().closerToCenterThan(p_35003_, 12.0D);
    }
 
    protected static boolean m_34965_(Piglin p_34966_) {
@@ -546,7 +_,7 @@
    public static boolean m_34808_(LivingEntity p_34809_) {
       for(ItemStack itemstack : p_34809_.m_6168_()) {
          Item item = itemstack.m_41720_();
-         if (item instanceof ArmorItem && ((ArmorItem)item).m_40401_() == ArmorMaterials.GOLD) {
+         if (itemstack.makesPiglinsNeutral(p_34809_)) {
             return true;
          }
       }
@@ -704,7 +_,7 @@
    }
 
    private static boolean m_34918_(LivingEntity p_34919_) {
-      return p_34919_.m_21055_(Items.f_42717_);
+      return p_34919_.m_21093_(is -> is.m_41720_() instanceof net.minecraft.world.item.CrossbowItem);
    }
 
    private static void m_34938_(LivingEntity p_34939_) {
@@ -752,7 +_,7 @@
    }
 
    private static boolean m_35028_(Piglin p_35029_) {
-      return p_35029_.m_21206_().m_41619_() || !m_149965_(p_35029_.m_21206_());
+      return p_35029_.m_21206_().m_41619_() || !isLovedItem(p_35029_.m_21206_(), p_35029_); // CraftBukkit - Changes to allow custom payment for bartering
    }
 
    public static boolean m_34806_(EntityType<?> p_34807_) {
