--- a/net/minecraft/world/entity/monster/piglin/StopAdmiringIfItemTooFarAway.java
+++ b/net/minecraft/world/entity/monster/piglin/StopAdmiringIfItemTooFarAway.java
@@ -24,7 +_,7 @@
          if (!optional.isPresent()) {
             return true;
          } else {
-            return !optional.get().m_19950_(p_35222_, (double)this.f_35210_);
+            return !optional.get().closerToCenterThan(p_35222_, (double)this.f_35210_);
          }
       }
    }
