--- ../src-base/minecraft/net/minecraft/nbt/NBTTagList.java
+++ ../src-work/minecraft/net/minecraft/nbt/NBTTagList.java
@@ -9,7 +9,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class NBTTagList extends NBTBase
+public class NBTTagList extends NBTBase implements Iterable<NBTBase>
 {
     private static final Logger LOGGER = LogManager.getLogger();
     private List<NBTBase> tagList = Lists.<NBTBase>newArrayList();
@@ -281,4 +281,5 @@
     {
         return this.tagType;
     }
+    @Override public java.util.Iterator<NBTBase> iterator() {return tagList.iterator();}
 }
