--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class EatBlockGoal extends Goal {
    private static final int f_148085_ = 40;
@@ -61,7 +_,8 @@
       if (this.f_25204_ == this.m_183277_(4)) {
          BlockPos blockpos = this.f_25202_.m_142538_();
          if (f_25201_.test(this.f_25203_.m_8055_(blockpos))) {
-            if (this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)) {
+            // CatServer start
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_25203_, this.f_25202_) || (!CraftEventFactory.callEntityChangeBlockEvent(this.f_25202_, blockpos, Blocks.f_50016_.m_49966_(), !this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)).isCancelled())) {
                this.f_25203_.m_46961_(blockpos, false);
             }
 
@@ -70,7 +_,8 @@
          } else {
             BlockPos blockpos1 = blockpos.m_7495_();
             if (this.f_25203_.m_8055_(blockpos1).m_60713_(Blocks.f_50440_)) {
-               if (this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)) {
+               if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_25203_, this.f_25202_) || !CraftEventFactory.callEntityChangeBlockEvent(this.f_25202_, blockpos1, Blocks.f_50016_.m_49966_(), !this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)).isCancelled()) {
+                  // CatServer end
                   this.f_25203_.m_46796_(2001, blockpos1, Block.m_49956_(Blocks.f_50440_.m_49966_()));
                   this.f_25203_.m_7731_(blockpos1, Blocks.f_50493_.m_49966_(), 2);
                }
