--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -11,6 +_,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class EatBlockGoal extends Goal {
    private static final int f_148085_ = 40;
    private static final Predicate<BlockState> f_25201_ = BlockStatePredicate.m_61287_(Blocks.f_50034_);
@@ -60,7 +_,7 @@
       if (this.f_25204_ == this.m_183277_(4)) {
          BlockPos blockpos = this.f_25202_.m_20183_();
          if (f_25201_.test(this.f_25203_.m_8055_(blockpos))) {
-            if (this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (CraftEventFactory.callEntityChangeBlockEvent(this.f_25202_, blockpos, Blocks.f_50016_.m_49966_(), !net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_25203_, this.f_25202_))) { // CatServer
                this.f_25203_.m_46961_(blockpos, false);
             }
 
@@ -68,7 +_,7 @@
          } else {
             BlockPos blockpos1 = blockpos.m_7495_();
             if (this.f_25203_.m_8055_(blockpos1).m_60713_(Blocks.f_50440_)) {
-               if (this.f_25203_.m_46469_().m_46207_(GameRules.f_46132_)) {
+               if (CraftEventFactory.callEntityChangeBlockEvent(this.f_25202_, blockpos1, Blocks.f_50016_.m_49966_(), !net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_25203_, this.f_25202_))) { // CatServer
                   this.f_25203_.m_46796_(2001, blockpos1, Block.m_49956_(Blocks.f_50440_.m_49966_()));
                   this.f_25203_.m_7731_(blockpos1, Blocks.f_50493_.m_49966_(), 2);
                }
