--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -12,6 +_,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftEntity;
+import org.bukkit.event.entity.EntityTeleportEvent;
 
 public class FollowOwnerGoal extends Goal {
    public static final int f_148087_ = 12;
@@ -117,7 +_,18 @@
       } else if (!this.m_25307_(new BlockPos(p_25304_, p_25305_, p_25306_))) {
          return false;
       } else {
-         this.f_25283_.m_7678_((double)p_25304_ + 0.5D, (double)p_25305_, (double)p_25306_ + 0.5D, this.f_25283_.m_146908_(), this.f_25283_.m_146909_());
+         // CraftBukkit start
+         CraftEntity entity = this.f_25283_.getBukkitEntity();
+         Location to = new Location(entity.getWorld(), (double)p_25304_ + 0.5D, (double)p_25305_, (double)p_25306_ + 0.5D, this.f_25283_.m_146908_(), this.f_25283_.m_146909_());
+         EntityTeleportEvent event = new EntityTeleportEvent(entity, entity.getLocation(), to);
+         this.f_25283_.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return false;
+         }
+         to = event.getTo();
+
+         this.f_25283_.m_7678_(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         // CraftBukkit end
          this.f_25287_.m_26573_();
          return true;
       }
