--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -13,6 +_,12 @@
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTeleportEvent;
+// CraftBukkit end
+
 public class FollowOwnerGoal extends Goal {
    public static final int f_148087_ = 12;
    private static final int f_148088_ = 2;
@@ -119,7 +_,14 @@
       } else if (!this.m_25307_(new BlockPos(p_25304_, p_25305_, p_25306_))) {
          return false;
       } else {
-         this.f_25283_.m_7678_((double)p_25304_ + 0.5D, (double)p_25305_, (double)p_25306_ + 0.5D, this.f_25283_.m_146908_(), this.f_25283_.m_146909_());
+         // CraftBukkit start
+         EntityTeleportEvent event = CraftEventFactory.callEntityTeleportEvent(this.f_25283_, (double) p_25304_ + 0.5D, (double) p_25305_, (double) p_25306_ + 0.5D);
+         if (event.isCancelled()) {
+            return false;
+         }
+         Location to = event.getTo();
+         this.f_25283_.m_7678_(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         // CraftBukkit end
          this.f_25287_.m_26573_();
          return true;
       }
