--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -12,7 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
@@ -20,6 +_,8 @@
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class RemoveBlockGoal extends MoveToBlockGoal {
    private final Block f_25836_;
@@ -34,7 +_,7 @@
    }
 
    public boolean m_8036_() {
-      if (!this.f_25837_.f_19853_.m_46469_().m_46207_(GameRules.f_46132_)) {
+      if (!net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.f_25837_.f_19853_, this.f_25602_, this.f_25837_)) {
          return false;
       } else if (this.f_25600_ > 0) {
          --this.f_25600_;
@@ -93,6 +_,13 @@
          }
 
          if (this.f_25838_ > 60) {
+            // CraftBukkit start - Step on eggs
+            EntityInteractEvent event = new EntityInteractEvent(this.f_25837_.getBukkitEntity(), CraftBlock.at(level, blockpos1));
+            level.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             level.m_7471_(blockpos1, false);
             if (!level.f_46443_) {
                for(int i = 0; i < 20; ++i) {
