--- a/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java
+++ b/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java
@@ -18,7 +_,7 @@
 
    public boolean m_8036_() {
       if (!this.f_25896_.m_21824_()) {
-         return false;
+         return this.f_25896_.m_21827_() && this.f_25896_.m_5448_() == null; // CraftBukkit - Allow sitting for wild animals
       } else if (this.f_25896_.m_20072_()) {
          return false;
       } else if (!this.f_25896_.m_20096_()) {
