--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -7,6 +_,10 @@
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.crafting.Ingredient;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
 
 public class TemptGoal extends Goal {
    private static final TargetingConditions f_25926_ = TargetingConditions.m_148353_().m_26883_(10.0D).m_148355_();
@@ -40,6 +_,15 @@
          return false;
       } else {
          this.f_25925_ = this.f_25924_.f_19853_.m_45946_(this.f_148137_, this.f_25924_);
+         // CraftBukkit start
+         if (this.f_25925_ != null) {
+            EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(this.f_25924_, this.f_25925_, EntityTargetEvent.TargetReason.TEMPT);
+            if (event.isCancelled()) {
+               return false;
+            }
+            this.f_25925_ = (event.getTarget() == null) ? null : (Player) ((CraftLivingEntity) event.getTarget()).getHandle();
+         }
+         // CraftBukkit end
          return this.f_25925_ != null;
       }
    }
