--- ../src-base/minecraft/net/minecraftforge/common/model/Models.java
+++ ../src-work/minecraft/net/minecraftforge/common/model/Models.java
@@ -23,6 +23,8 @@
 import com.google.common.collect.Iterators;
 import com.google.common.collect.UnmodifiableIterator;
 
+import java.util.Collections;
+
 public enum Models
 {
     ;
@@ -32,12 +34,12 @@
         return new HiddenModelPart(path);
     }
 
-    public static UnmodifiableIterator<String> getParts(IModelPart part)
+/*    public static UnmodifiableIterator<String> getParts(IModelPart part)
     {
         if(part instanceof HiddenModelPart)
         {
             return ((HiddenModelPart) part).getPath().iterator();
         }
         return Iterators.emptyIterator();
-    }
+    }*/
 }
