--- a/net/minecraft/world/BossInfo.java
+++ b/net/minecraft/world/BossInfo.java
@@ -6,10 +_,10 @@
 
 public abstract class BossInfo {
    private final UUID field_186756_h;
-   protected ITextComponent field_186749_a;
+   public ITextComponent field_186749_a;
    protected float field_186750_b;
-   protected BossInfo.Color field_186751_c;
-   protected BossInfo.Overlay field_186752_d;
+   public BossInfo.Color field_186751_c;
+   public BossInfo.Overlay field_186752_d;
    protected boolean field_186753_e;
    protected boolean field_186754_f;
    protected boolean field_186755_g;
