--- a/net/minecraft/world/CompoundContainer.java
+++ b/net/minecraft/world/CompoundContainer.java
@@ -2,11 +_,59 @@
 
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public class CompoundContainer implements Container {
    public final Container f_18910_;
    public final Container f_18911_;
 
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+
+   public List<ItemStack> getContents() {
+      List<ItemStack> result = new ArrayList<ItemStack>(this.m_6643_());
+      for (int i = 0; i < this.m_6643_(); i++) {
+         result.add(this.m_8020_(i));
+      }
+      return result;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      this.f_18910_.onOpen(who);
+      this.f_18911_.onOpen(who);
+      transaction.add(who);
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      this.f_18910_.onClose(who);
+      this.f_18911_.onClose(who);
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return null; // This method won't be called since CraftInventoryDoubleChest doesn't defer to here
+   }
+
+   public void setMaxStackSize(int size) {
+      this.f_18910_.setMaxStackSize(size);
+      this.f_18911_.setMaxStackSize(size);
+   }
+
+   @Override
+   public Location getLocation() {
+      return f_18910_.getLocation(); // TODO: right?
+   }
+   // CraftBukkit end
+
    public CompoundContainer(Container p_18913_, Container p_18914_) {
       this.f_18910_ = p_18913_;
       this.f_18911_ = p_18914_;
@@ -46,7 +_,7 @@
    }
 
    public int m_6893_() {
-      return this.f_18910_.m_6893_();
+      return Math.min(this.f_18910_.m_6893_(), this.f_18911_.m_6893_()); // CraftBukkit - check both sides
    }
 
    public void m_6596_() {
