--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -4,6 +_,9 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Recipe;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public interface Container extends Clearable {
    int f_146642_ = 64;
@@ -61,4 +_,33 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   java.util.List<ItemStack> getContents();
+
+   void onOpen(CraftHumanEntity who);
+
+   void onClose(CraftHumanEntity who);
+
+   // CauldronX start
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+      return new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+   }
+   // CauldronX end
+
+   org.bukkit.inventory.InventoryHolder getOwner();
+
+   void setMaxStackSize(int size);
+
+   org.bukkit.Location getLocation();
+
+   default Recipe getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(Recipe recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
