--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -9,6 +_,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 
+// CraftBukkit start
+import net.minecraft.world.item.crafting.Recipe;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+// CraftBukkit end
+
 public interface Container extends Clearable {
    int f_146642_ = 64;
    int f_271421_ = 8;
@@ -27,7 +_,7 @@
 
    default int m_6893_() {
       return 64;
-   }
+   } // CraftBukkit // CatServer - recover
 
    void m_6596_();
 
@@ -78,10 +_,14 @@
    }
 
    static boolean m_272074_(BlockEntity p_273154_, Player p_273222_) {
-      return m_271806_(p_273154_, p_273222_, 8);
+      return stillValidBlockEntity(p_273154_, p_273222_, p_273222_.m_21133_(net.minecraftforge.common.ForgeMod.BLOCK_REACH.get()) + 3.5); // Vanilla uses 8, default reach is 4.5 so add 3.5 padding
    }
 
    static boolean m_271806_(BlockEntity p_272877_, Player p_272670_, int p_273411_) {
+       return stillValidBlockEntity(p_272877_, p_272670_, (double)p_273411_);
+   }
+
+   static boolean stillValidBlockEntity(BlockEntity p_272877_, Player p_272670_, double p_273411_) {
       Level level = p_272877_.m_58904_();
       BlockPos blockpos = p_272877_.m_58899_();
       if (level == null) {
@@ -92,4 +_,46 @@
          return p_272670_.m_20275_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.5D, (double)blockpos.m_123343_() + 0.5D) <= (double)(p_273411_ * p_273411_);
       }
    }
+
+   // CraftBukkit start
+   default java.util.List<ItemStack> getContents() {
+      return java.util.Collections.emptyList(); // CatServer
+   }
+
+   default void onOpen(CraftHumanEntity who) {} // CatServer
+
+   default void onClose(CraftHumanEntity who) {} // CatServer
+
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+      return java.util.Collections.emptyList(); // CatServer
+   }
+
+   // CatServer start
+   default @org.jetbrains.annotations.Nullable org.bukkit.inventory.InventoryHolder getOwner() { // Paper - annotation
+      return this instanceof BlockEntity ? catserver.server.inventory.CatInventoryUtils.getOwner((net.minecraft.world.level.block.entity.BlockEntity) this) : null;
+   }
+   // CatServer end
+
+   default void setMaxStackSize(int size) {} // CatServer
+
+   // CatServer start
+   default org.bukkit.Location getLocation() {
+      if (this instanceof BlockEntity) {
+         BlockPos blockPos = ((BlockEntity) this).m_58899_();
+         return new org.bukkit.Location(((BlockEntity) this).m_58904_().getWorld(), blockPos.m_123341_(), blockPos.m_123342_(), blockPos.m_123343_());
+      } else {
+         return null;
+      }
+   }
+   // CatServer end
+
+   default Recipe getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(Recipe recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
