--- a/net/minecraft/world/GameRules.java
+++ b/net/minecraft/world/GameRules.java
@@ -15,6 +_,8 @@
 import java.util.function.Function;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.event.world.WorldGameRuleChangeEvent;
 import net.minecraft.command.CommandSource;
 import net.minecraft.command.Commands;
 import net.minecraft.entity.player.ServerPlayerEntity;
@@ -26,6 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
 
 public class GameRules {
    private static final Logger field_223622_y = LogManager.getLogger();
@@ -78,6 +_,7 @@
    public static final GameRules.RuleKey<GameRules.BooleanValue> field_234895_F_ = func_234903_a_("forgiveDeadPlayers", GameRules.Category.MOBS, GameRules.BooleanValue.func_223568_b(true));
    public static final GameRules.RuleKey<GameRules.BooleanValue> field_234896_G_ = func_234903_a_("universalAnger", GameRules.Category.MOBS, GameRules.BooleanValue.func_223568_b(false));
    private final Map<GameRules.RuleKey<?>, GameRules.RuleValue<?>> field_82771_a;
+   public static GameRules.RuleKey<?> ruleKey;
 
    public static <T extends GameRules.RuleValue<T>> GameRules.RuleKey<T> func_234903_a_(String p_234903_0_, GameRules.Category p_234903_1_, GameRules.RuleType<T> p_234903_2_) {
       GameRules.RuleKey<T> rulekey = new GameRules.RuleKey<>(p_234903_0_, p_234903_1_);
@@ -179,9 +_,20 @@
          this.field_223575_a = p_i51535_2_;
       }
 
+      // Paper start
       protected void func_223555_a(CommandContext<CommandSource> p_223555_1_, String p_223555_2_) {
-         this.field_223575_a = BoolArgumentType.getBool(p_223555_1_, p_223555_2_);
+         if (ruleKey != null) {
+            WorldGameRuleChangeEvent event = new WorldGameRuleChangeEvent(p_223555_1_.getSource().func_197023_e().getWorld(), p_223555_1_.getSource().getBukkitSender(), org.bukkit.GameRule.getByName(ruleKey.toString()), String.valueOf(BoolArgumentType.getBool(p_223555_1_, p_223555_2_)));
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            this.field_223575_a = Boolean.parseBoolean(event.getValue());
+         } else {
+            this.field_223575_a = BoolArgumentType.getBool(p_223555_1_, p_223555_2_);
+         }
       }
+      // Paper end
 
       public boolean func_223572_a() {
          return this.field_223575_a;
@@ -196,7 +_,7 @@
          return Boolean.toString(this.field_223575_a);
       }
 
-      protected void func_223553_a(String p_223553_1_) {
+      public void func_223553_a(String p_223553_1_) {
          this.field_223575_a = Boolean.parseBoolean(p_223553_1_);
       }
 
@@ -274,9 +_,20 @@
          this.field_223566_a = p_i51534_2_;
       }
 
+      // Paper start
       protected void func_223555_a(CommandContext<CommandSource> p_223555_1_, String p_223555_2_) {
-         this.field_223566_a = IntegerArgumentType.getInteger(p_223555_1_, p_223555_2_);
+         if (ruleKey != null) {
+            WorldGameRuleChangeEvent event = new WorldGameRuleChangeEvent(p_223555_1_.getSource().func_197023_e().getWorld(), p_223555_1_.getSource().getBukkitSender(), org.bukkit.GameRule.getByName(ruleKey.toString()), String.valueOf(IntegerArgumentType.getInteger(p_223555_1_, p_223555_2_)));
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            this.field_223566_a = Integer.parseInt(event.getValue());
+         } else {
+            this.field_223566_a = IntegerArgumentType.getInteger(p_223555_1_, p_223555_2_);
+         }
       }
+      // Paper end
 
       public int func_223560_a() {
          return this.field_223566_a;
@@ -286,7 +_,7 @@
          return Integer.toString(this.field_223566_a);
       }
 
-      protected void func_223553_a(String p_223553_1_) {
+      public void func_223553_a(String p_223553_1_) {
          this.field_223566_a = func_223563_b(p_223553_1_);
       }
 
@@ -405,19 +_,25 @@
 
       protected abstract void func_223555_a(CommandContext<CommandSource> p_223555_1_, String p_223555_2_);
 
+      // Paper start
       public void func_223554_b(CommandContext<CommandSource> p_223554_1_, String p_223554_2_) {
          this.func_223555_a(p_223554_1_, p_223554_2_);
          this.func_223556_a(p_223554_1_.getSource().func_197028_i());
       }
-
-      protected void func_223556_a(@Nullable MinecraftServer p_223556_1_) {
+      // Paper end
+
+      public void setPaperRuleKey(GameRules.RuleKey<T> paperRuleKey) {
+         ruleKey = paperRuleKey;
+      }
+
+      public void func_223556_a(@Nullable MinecraftServer p_223556_1_) {
          if (p_223556_1_ != null) {
             this.field_223558_a.field_223584_c.accept(p_223556_1_, this.func_223213_e_());
          }
 
       }
 
-      protected abstract void func_223553_a(String p_223553_1_);
+      public abstract void func_223553_a(String p_223553_1_);
 
       public abstract String func_223552_b();
 
