--- a/net/minecraft/world/IBlockReader.java
+++ b/net/minecraft/world/IBlockReader.java
@@ -25,8 +_,12 @@
 
    FluidState func_204610_c(BlockPos p_204610_1_);
 
+   default boolean isPosLoaded(BlockPos pos) {
+      return false;
+   }
+
    default int func_217298_h(BlockPos p_217298_1_) {
-      return this.func_180495_p(p_217298_1_).func_185906_d();
+      return this.func_180495_p(p_217298_1_).getLightValue(this, p_217298_1_);
    }
 
    default int func_201572_C() {
@@ -41,19 +_,25 @@
       return BlockPos.func_239581_a_(p_234853_1_).map(this::func_180495_p);
    }
 
+   // CraftBukkit start
+   default BlockRayTraceResult rayTraceBlock(RayTraceContext p_217297_1_, BlockPos p_217297_2_) {
+      BlockState blockstate = this.func_180495_p(p_217297_2_);
+      FluidState fluidstate = this.func_204610_c(p_217297_2_);
+      Vector3d vector3d = p_217297_1_.func_222253_b();
+      Vector3d vector3d1 = p_217297_1_.func_222250_a();
+      VoxelShape voxelshape = p_217297_1_.func_222251_a(blockstate, this, p_217297_2_);
+      BlockRayTraceResult blockraytraceresult = this.func_217296_a(vector3d, vector3d1, p_217297_2_, voxelshape, blockstate);
+      VoxelShape voxelshape1 = p_217297_1_.func_222252_a(fluidstate, this, p_217297_2_);
+      BlockRayTraceResult blockraytraceresult1 = voxelshape1.func_212433_a(vector3d, vector3d1, p_217297_2_);
+      double d0 = blockraytraceresult == null ? Double.MAX_VALUE : p_217297_1_.func_222253_b().func_72436_e(blockraytraceresult.func_216347_e());
+      double d1 = blockraytraceresult1 == null ? Double.MAX_VALUE : p_217297_1_.func_222253_b().func_72436_e(blockraytraceresult1.func_216347_e());
+      return d0 <= d1 ? blockraytraceresult : blockraytraceresult1;
+   }
+   // CraftBukkit end
+
    default BlockRayTraceResult func_217299_a(RayTraceContext p_217299_1_) {
       return func_217300_a(p_217299_1_, (p_217297_1_, p_217297_2_) -> {
-         BlockState blockstate = this.func_180495_p(p_217297_2_);
-         FluidState fluidstate = this.func_204610_c(p_217297_2_);
-         Vector3d vector3d = p_217297_1_.func_222253_b();
-         Vector3d vector3d1 = p_217297_1_.func_222250_a();
-         VoxelShape voxelshape = p_217297_1_.func_222251_a(blockstate, this, p_217297_2_);
-         BlockRayTraceResult blockraytraceresult = this.func_217296_a(vector3d, vector3d1, p_217297_2_, voxelshape, blockstate);
-         VoxelShape voxelshape1 = p_217297_1_.func_222252_a(fluidstate, this, p_217297_2_);
-         BlockRayTraceResult blockraytraceresult1 = voxelshape1.func_212433_a(vector3d, vector3d1, p_217297_2_);
-         double d0 = blockraytraceresult == null ? Double.MAX_VALUE : p_217297_1_.func_222253_b().func_72436_e(blockraytraceresult.func_216347_e());
-         double d1 = blockraytraceresult1 == null ? Double.MAX_VALUE : p_217297_1_.func_222253_b().func_72436_e(blockraytraceresult1.func_216347_e());
-         return d0 <= d1 ? blockraytraceresult : blockraytraceresult1;
+         return this.rayTraceBlock(p_217299_1_, p_217297_2_); // CraftBukkit - moved into separate method
       }, (p_217302_0_) -> {
          Vector3d vector3d = p_217302_0_.func_222253_b().func_178788_d(p_217302_0_.func_222250_a());
          return BlockRayTraceResult.func_216352_a(p_217302_0_.func_222250_a(), Direction.func_210769_a(vector3d.field_72450_a, vector3d.field_72448_b, vector3d.field_72449_c), new BlockPos(p_217302_0_.func_222250_a()));
