--- a/net/minecraft/world/IServerWorld.java
+++ b/net/minecraft/world/IServerWorld.java
@@ -6,7 +_,18 @@
 public interface IServerWorld extends IWorld {
    ServerWorld func_201672_e();
 
+   default ServerWorld getMinecraftWorld() {
+      return this.func_201672_e(); // CatServer - remap to real method
+   }
+
    default void func_242417_l(Entity p_242417_1_) {
-      p_242417_1_.func_226276_cg_().forEach(this::func_217376_c);
-   }
+      if (this.func_201672_e() != null && this.func_201672_e().dummyFreshEntitySpawn) return; // Fix Pehkui Startup
+      this.addAllEntities(p_242417_1_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+      // CraftBukkit start
+   }
+
+   default void addAllEntities(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      entity.func_226276_cg_().forEach((e) -> this.addEntity(e, reason));
+   }
+   // CraftBukkit end
 }
