--- a/net/minecraft/world/IWorld.java
+++ b/net/minecraft/world/IWorld.java
@@ -10,6 +_,7 @@
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.chunk.AbstractChunkProvider;
+import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.storage.IWorldInfo;
 
 public interface IWorld extends IBiomeReader, IDayTimeReader {
@@ -52,5 +_,9 @@
 
    default void func_217379_c(int p_217379_1_, BlockPos p_217379_2_, int p_217379_3_) {
       this.func_217378_a((PlayerEntity)null, p_217379_1_, p_217379_2_, p_217379_3_);
+   }
+
+   default ServerWorld getMinecraftWorld() {
+      return this.getMinecraftWorld();
    }
 }
