--- a/net/minecraft/world/LockCode.java
+++ b/net/minecraft/world/LockCode.java
@@ -3,6 +_,8 @@
 import javax.annotation.concurrent.Immutable;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.ChatColor;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftChatMessage;
 
 @Immutable
 public class LockCode {
@@ -15,7 +_,19 @@
    }
 
    public boolean m_19107_(ItemStack p_19108_) {
-      return this.f_19103_.isEmpty() || !p_19108_.m_41619_() && p_19108_.m_41788_() && this.f_19103_.equals(p_19108_.m_41786_().getString());
+      // CraftBukkit start - SPIGOT-6307: Check for color codes if the lock contains color codes
+      if (this.f_19103_.isEmpty()) return true;
+      if (!p_19108_.m_41619_() && p_19108_.m_41788_()) {
+         if (this.f_19103_.indexOf(ChatColor.COLOR_CHAR) == -1) {
+            // The lock key contains no color codes, so let's ignore colors in the item display name (vanilla Minecraft behavior):
+            return this.f_19103_.equals(p_19108_.m_41786_().getString());
+         } else {
+            // The lock key contains color codes, so let's take them into account:
+            return this.f_19103_.equals(CraftChatMessage.fromComponent(p_19108_.m_41786_()));
+         }
+      }
+      return false;
+      // CraftBukkit end
    }
 
    public void m_19109_(CompoundTag p_19110_) {
