--- a/net/minecraft/world/MenuProvider.java
+++ b/net/minecraft/world/MenuProvider.java
@@ -5,4 +_,12 @@
 
 public interface MenuProvider extends MenuConstructor {
    Component m_5446_();
+
+   // CatServer start
+   net.minecraft.core.BlockPos[] blockPosForEvent = new net.minecraft.core.BlockPos[1];
+   default MenuProvider setBlockPosForEvent(net.minecraft.core.BlockPos pos) {
+      this.blockPosForEvent[0] = pos;
+      return this;
+   }
+   // CatServer end
 }
