--- a/net/minecraft/world/SimpleContainer.java
+++ b/net/minecraft/world/SimpleContainer.java
@@ -13,13 +_,91 @@
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
 public class SimpleContainer implements Container, StackedContentsCompatible {
    private final int f_19146_;
    private final NonNullList<ItemStack> f_19147_;
    @Nullable
    private List<ContainerListener> f_19148_;
 
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private int maxStack = MAX_STACK;
+   protected @Nullable org.bukkit.inventory.InventoryHolder bukkitOwner; // Paper - annotation
+
+   public List<ItemStack> getContents() {
+      return this.f_19147_;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int i) {
+      maxStack = i;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      // Paper start
+      if (this.bukkitOwner == null && this.bukkitOwnerCreator != null) {
+         this.bukkitOwner = this.bukkitOwnerCreator.get();
+      }
+      // Paper end
+      return bukkitOwner;
+   }
+
+   @Override
+   public Location getLocation() {
+      return null;
+   }
+
+   public SimpleContainer(SimpleContainer original) {
+      this(original.f_19146_);
+      for (int slot = 0; slot < original.f_19146_; slot++) {
+         this.f_19147_.set(slot, original.f_19147_.get(slot).m_41777_());
+      }
+   }
+
    public SimpleContainer(int p_19150_) {
+      this(p_19150_, null);
+   }
+
+   // Paper start
+   private @Nullable java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> bukkitOwnerCreator;
+   public SimpleContainer(java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> bukkitOwnerCreator, int size) {
+      this(size);
+      this.bukkitOwnerCreator = bukkitOwnerCreator;
+   }
+
+   // CatServer start
+   public SimpleContainer setBukkitOwnerCreator(java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> bukkitOwnerCreator) {
+      this.bukkitOwnerCreator = bukkitOwnerCreator;
+      return this;
+   }
+   // CatServer end
+   // Paper end
+
+   public SimpleContainer(int p_19150_, org.bukkit.inventory.InventoryHolder bukkitOwner) {
+      this.bukkitOwner = bukkitOwner;
+      // CraftBukkit end
       this.f_19146_ = p_19150_;
       this.f_19147_ = NonNullList.m_122780_(p_19150_, ItemStack.f_41583_);
    }
