--- a/net/minecraft/world/Teleporter.java
+++ b/net/minecraft/world/Teleporter.java
@@ -5,6 +_,7 @@
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.NetherPortalBlock;
+import net.minecraft.entity.Entity;
 import net.minecraft.state.properties.BlockStateProperties;
 import net.minecraft.util.Direction;
 import net.minecraft.util.TeleportationRepositioner;
@@ -19,7 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.server.TicketType;
 
-public class Teleporter {
+public class Teleporter implements net.minecraftforge.common.util.ITeleporter {
    protected final ServerWorld field_85192_a;
 
    public Teleporter(ServerWorld p_i1963_1_) {
@@ -27,13 +_,19 @@
    }
 
    public Optional<TeleportationRepositioner.Result> func_242957_a(BlockPos p_242957_1_, boolean p_242957_2_) {
+      // CraftBukkit start
+      return findPortal(p_242957_1_, p_242957_2_ ? 16 : 128); // Search Radius
+   }
+
+   public Optional<TeleportationRepositioner.Result> findPortal(BlockPos pos, int i) {
       PointOfInterestManager pointofinterestmanager = this.field_85192_a.func_217443_B();
-      int i = p_242957_2_ ? 16 : 128;
-      pointofinterestmanager.func_226347_a_(this.field_85192_a, p_242957_1_, i);
+      // int i = isNether ? 16 : 128;
+      // CraftBukkit end
+      pointofinterestmanager.func_226347_a_(this.field_85192_a, pos, i);
       Optional<PointOfInterest> optional = pointofinterestmanager.func_226353_b_((p_242952_0_) -> {
          return p_242952_0_ == PointOfInterestType.field_226358_u_;
-      }, p_242957_1_, i, PointOfInterestManager.Status.ANY).sorted(Comparator.<PointOfInterest>comparingDouble((p_242954_1_) -> {
-         return p_242954_1_.func_218261_f().func_177951_i(p_242957_1_);
+      }, pos, i, PointOfInterestManager.Status.ANY).sorted(Comparator.<PointOfInterest>comparingDouble((p_242954_1_) -> {
+         return p_242954_1_.func_218261_f().func_177951_i(pos);
       }).thenComparingInt((p_242959_0_) -> {
          return p_242959_0_.func_218261_f().func_177956_o();
       })).filter((p_242958_1_) -> {
@@ -50,16 +_,21 @@
    }
 
    public Optional<TeleportationRepositioner.Result> func_242956_a(BlockPos p_242956_1_, Direction.Axis p_242956_2_) {
-      Direction direction = Direction.func_181076_a(Direction.AxisDirection.POSITIVE, p_242956_2_);
+      // CraftBukkit start
+      return this.createPortal(p_242956_1_, p_242956_2_, null, 16);
+   }
+   public Optional<TeleportationRepositioner.Result> createPortal(BlockPos pos, Direction.Axis axis, Entity entity, int createRadius) {
+      // CraftBukkit end
+      Direction direction = Direction.func_181076_a(Direction.AxisDirection.POSITIVE, axis);
       double d0 = -1.0D;
       BlockPos blockpos = null;
       double d1 = -1.0D;
       BlockPos blockpos1 = null;
       WorldBorder worldborder = this.field_85192_a.func_175723_af();
       int i = this.field_85192_a.func_234938_ad_() - 1;
-      BlockPos.Mutable blockpos$mutable = p_242956_1_.func_239590_i_();
+      BlockPos.Mutable blockpos$mutable = pos.func_239590_i_();
 
-      for(BlockPos.Mutable blockpos$mutable1 : BlockPos.func_243514_a(p_242956_1_, 16, Direction.EAST, Direction.SOUTH)) {
+      for(BlockPos.Mutable blockpos$mutable1 : BlockPos.func_243514_a(pos, createRadius, Direction.EAST, Direction.SOUTH)) { // CraftBukkit
          int j = Math.min(i, this.field_85192_a.func_201676_a(Heightmap.Type.MOTION_BLOCKING, blockpos$mutable1.func_177958_n(), blockpos$mutable1.func_177952_p()));
          int k = 1;
          if (worldborder.func_177746_a(blockpos$mutable1) && worldborder.func_177746_a(blockpos$mutable1.func_189534_c(direction, 1))) {
@@ -77,7 +_,7 @@
                      if (j1 <= 0 || j1 >= 3) {
                         blockpos$mutable1.func_185336_p(l);
                         if (this.func_242955_a(blockpos$mutable1, blockpos$mutable, direction, 0)) {
-                           double d2 = p_242956_1_.func_177951_i(blockpos$mutable1);
+                           double d2 = pos.func_177951_i(blockpos$mutable1);
                            if (this.func_242955_a(blockpos$mutable1, blockpos$mutable, direction, -1) && this.func_242955_a(blockpos$mutable1, blockpos$mutable, direction, 1) && (d0 == -1.0D || d0 > d2)) {
                               d0 = d2;
                               blockpos = blockpos$mutable1.func_185334_h();
@@ -100,8 +_,9 @@
          d0 = d1;
       }
 
+      org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator(this.field_85192_a); // CraftBukkit - Use BlockStateListPopulator
       if (d0 == -1.0D) {
-         blockpos = (new BlockPos(p_242956_1_.func_177958_n(), MathHelper.func_76125_a(p_242956_1_.func_177956_o(), 70, this.field_85192_a.func_234938_ad_() - 10), p_242956_1_.func_177952_p())).func_185334_h();
+         blockpos = (new BlockPos(pos.func_177958_n(), MathHelper.func_76125_a(pos.func_177956_o(), 70, this.field_85192_a.func_234938_ad_() - 10), pos.func_177952_p())).func_185334_h();
          Direction direction1 = direction.func_176746_e();
          if (!worldborder.func_177746_a(blockpos)) {
             return Optional.empty();
@@ -112,7 +_,7 @@
                for(int i3 = -1; i3 < 3; ++i3) {
                   BlockState blockstate1 = i3 < 0 ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P();
                   blockpos$mutable.func_239621_a_(blockpos, k2 * direction.func_82601_c() + l1 * direction1.func_82601_c(), i3, k2 * direction.func_82599_e() + l1 * direction1.func_82599_e());
-                  this.field_85192_a.func_175656_a(blockpos$mutable, blockstate1);
+                  blockList.func_180501_a(blockpos$mutable, blockstate1, 3); // CraftBukkit
                }
             }
          }
@@ -122,19 +_,29 @@
          for(int i2 = -1; i2 < 4; ++i2) {
             if (k1 == -1 || k1 == 2 || i2 == -1 || i2 == 3) {
                blockpos$mutable.func_239621_a_(blockpos, k1 * direction.func_82601_c(), i2, k1 * direction.func_82599_e());
-               this.field_85192_a.func_180501_a(blockpos$mutable, Blocks.field_150343_Z.func_176223_P(), 3);
+               blockList.func_180501_a(blockpos$mutable, Blocks.field_150343_Z.func_176223_P(), 3); // CraftBukkit
             }
          }
       }
 
-      BlockState blockstate = Blocks.field_150427_aO.func_176223_P().func_206870_a(NetherPortalBlock.field_176550_a, p_242956_2_);
+      BlockState blockstate = Blocks.field_150427_aO.func_176223_P().func_206870_a(NetherPortalBlock.field_176550_a, axis);
 
       for(int j2 = 0; j2 < 2; ++j2) {
          for(int l2 = 0; l2 < 3; ++l2) {
             blockpos$mutable.func_239621_a_(blockpos, j2 * direction.func_82601_c(), l2, j2 * direction.func_82599_e());
-            this.field_85192_a.func_180501_a(blockpos$mutable, blockstate, 18);
+            blockList.func_180501_a(blockpos$mutable, blockstate, 18); // CraftBukkit
          }
       }
+
+      // CraftBukkit start
+      org.bukkit.World bworld = this.field_85192_a.getWorld();
+      org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blockList.getList(), bworld, (entity == null) ? null : entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.NETHER_PAIR);
+      this.field_85192_a.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return Optional.empty();
+      }
+      blockList.updateList();
+      // CraftBukkit end
 
       return Optional.of(new TeleportationRepositioner.Result(blockpos.func_185334_h(), 2, 3));
    }
