--- a/net/minecraft/world/chunk/listener/LoggingChunkStatusListener.java
+++ b/net/minecraft/world/chunk/listener/LoggingChunkStatusListener.java
@@ -26,15 +_,21 @@
       this.field_219515_d = this.field_219516_e;
    }
 
+   private static int lastPercentage = -1;
+
    public void func_219508_a(ChunkPos p_219508_1_, @Nullable ChunkStatus p_219508_2_) {
       if (p_219508_2_ == ChunkStatus.field_222617_m) {
          ++this.field_219514_c;
       }
 
       int i = this.func_219511_c();
+      int calc = MathHelper.func_76125_a(i, 0, 100);
       if (Util.func_211177_b() > this.field_219516_e) {
          this.field_219516_e += 500L;
-         field_219512_a.info((new TranslationTextComponent("menu.preparingSpawn", MathHelper.func_76125_a(i, 0, 100))).getString());
+         if(lastPercentage != calc){
+            field_219512_a.info((new TranslationTextComponent("menu.preparingSpawn", MathHelper.func_76125_a(i, 0, 100))).getString());
+            lastPercentage = calc;
+         }
       }
 
    }
