--- ../src-base/minecraft/net/minecraft/world/biome/BiomeEndDecorator.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeEndDecorator.java
@@ -17,7 +17,7 @@
 
 public class BiomeEndDecorator extends BiomeDecorator
 {
-    private static final LoadingCache<Long, WorldGenSpikes.EndSpike[]> SPIKE_CACHE = CacheBuilder.newBuilder().expireAfterWrite(5L, TimeUnit.MINUTES).<Long, WorldGenSpikes.EndSpike[]>build(new BiomeEndDecorator.SpikeCacheLoader());
+    private static final LoadingCache<Long, WorldGenSpikes.EndSpike[]> SPIKE_CACHE = CacheBuilder.newBuilder().expireAfterWrite(5L, TimeUnit.MINUTES).<Long, WorldGenSpikes.EndSpike[]>build(new SpikeCacheLoader());
     private final WorldGenSpikes spikeGen = new WorldGenSpikes();
 
     protected void genDecorations(Biome biomeIn, World worldIn, Random random)
