--- ../src-base/minecraft/net/minecraft/world/biome/BiomeSavannaMutated.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeSavannaMutated.java
@@ -9,7 +9,7 @@
 
 public class BiomeSavannaMutated extends BiomeSavanna
 {
-    public BiomeSavannaMutated(Biome.BiomeProperties properties)
+    public BiomeSavannaMutated(BiomeProperties properties)
     {
         super(properties);
         this.decorator.treesPerChunk = 2;
