--- ../src-base/minecraft/net/minecraft/world/biome/BiomeEnd.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeEnd.java
@@ -7,14 +7,14 @@
 
 public class BiomeEnd extends Biome
 {
-    public BiomeEnd(Biome.BiomeProperties properties)
+    public BiomeEnd(BiomeProperties properties)
     {
         super(properties);
         this.spawnableMonsterList.clear();
         this.spawnableCreatureList.clear();
         this.spawnableWaterCreatureList.clear();
         this.spawnableCaveCreatureList.clear();
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityEnderman.class, 10, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityEnderman.class, 10, 4, 4));
         this.topBlock = Blocks.DIRT.getDefaultState();
         this.fillerBlock = Blocks.DIRT.getDefaultState();
         this.decorator = new BiomeEndDecorator();
