--- ../src-base/minecraft/net/minecraft/world/biome/BiomeSnow.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeSnow.java
@@ -56,7 +56,7 @@
 
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.superIcy)
+        if (this.superIcy && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ICE))
         {
             for (int i = 0; i < 3; ++i)
             {
