--- ../src-base/minecraft/net/minecraft/world/biome/BiomeHell.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeHell.java
@@ -7,17 +7,17 @@
 
 public class BiomeHell extends Biome
 {
-    public BiomeHell(Biome.BiomeProperties properties)
+    public BiomeHell(BiomeProperties properties)
     {
         super(properties);
         this.spawnableMonsterList.clear();
         this.spawnableCreatureList.clear();
         this.spawnableWaterCreatureList.clear();
         this.spawnableCaveCreatureList.clear();
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityGhast.class, 50, 4, 4));
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityPigZombie.class, 100, 4, 4));
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityMagmaCube.class, 2, 4, 4));
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityEnderman.class, 1, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityGhast.class, 50, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityPigZombie.class, 100, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityMagmaCube.class, 2, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityEnderman.class, 1, 4, 4));
         this.decorator = new BiomeHellDecorator();
     }
 }
