--- ../src-base/minecraft/net/minecraft/world/biome/BiomeOcean.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeOcean.java
@@ -2,14 +2,14 @@
 
 public class BiomeOcean extends Biome
 {
-    public BiomeOcean(Biome.BiomeProperties properties)
+    public BiomeOcean(BiomeProperties properties)
     {
         super(properties);
         this.spawnableCreatureList.clear();
     }
 
-    public Biome.TempCategory getTempCategory()
+    public TempCategory getTempCategory()
     {
-        return Biome.TempCategory.OCEAN;
+        return TempCategory.OCEAN;
     }
 }
