--- ../src-base/minecraft/net/minecraft/world/biome/BiomeMesa.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeMesa.java
@@ -33,7 +33,7 @@
     private final boolean brycePillars;
     private final boolean hasForest;
 
-    public BiomeMesa(boolean p_i46704_1_, boolean p_i46704_2_, Biome.BiomeProperties properties)
+    public BiomeMesa(boolean p_i46704_1_, boolean p_i46704_2_, BiomeProperties properties)
     {
         super(properties);
         this.brycePillars = p_i46704_1_;
@@ -56,7 +56,7 @@
 
     public BiomeDecorator createBiomeDecorator()
     {
-        return new BiomeMesa.Decorator();
+        return new Decorator();
     }
 
     public WorldGenAbstractTree getRandomTreeFeature(Random rand)
@@ -336,6 +336,7 @@
         protected void generateOres(World worldIn, Random random)
         {
             super.generateOres(worldIn, random);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(worldIn, random, goldGen, chunkPos, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.genStandardOre1(worldIn, random, 20, this.goldGen, 32, 80);
         }
     }
