--- ../src-base/minecraft/net/minecraft/world/biome/BiomeMushroomIsland.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeMushroomIsland.java
@@ -5,7 +5,7 @@
 
 public class BiomeMushroomIsland extends Biome
 {
-    public BiomeMushroomIsland(Biome.BiomeProperties properties)
+    public BiomeMushroomIsland(BiomeProperties properties)
     {
         super(properties);
         this.decorator.treesPerChunk = -100;
@@ -17,6 +17,6 @@
         this.spawnableMonsterList.clear();
         this.spawnableCreatureList.clear();
         this.spawnableWaterCreatureList.clear();
-        this.spawnableCreatureList.add(new Biome.SpawnListEntry(EntityMooshroom.class, 8, 4, 8));
+        this.spawnableCreatureList.add(new SpawnListEntry(EntityMooshroom.class, 8, 4, 8));
     }
 }
