--- ../src-base/minecraft/net/minecraft/world/biome/BiomeStoneBeach.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeStoneBeach.java
@@ -4,7 +4,7 @@
 
 public class BiomeStoneBeach extends Biome
 {
-    public BiomeStoneBeach(Biome.BiomeProperties properties)
+    public BiomeStoneBeach(BiomeProperties properties)
     {
         super(properties);
         this.spawnableCreatureList.clear();
