--- ../src-base/minecraft/net/minecraft/world/biome/BiomeTaiga.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeTaiga.java
@@ -25,17 +25,17 @@
     private static final WorldGenMegaPineTree MEGA_PINE_GENERATOR = new WorldGenMegaPineTree(false, false);
     private static final WorldGenMegaPineTree MEGA_SPRUCE_GENERATOR = new WorldGenMegaPineTree(false, true);
     private static final WorldGenBlockBlob FOREST_ROCK_GENERATOR = new WorldGenBlockBlob(Blocks.MOSSY_COBBLESTONE, 0);
-    private final BiomeTaiga.Type type;
+    private final Type type;
 
-    public BiomeTaiga(BiomeTaiga.Type typeIn, Biome.BiomeProperties properties)
+    public BiomeTaiga(Type typeIn, BiomeProperties properties)
     {
         super(properties);
         this.type = typeIn;
-        this.spawnableCreatureList.add(new Biome.SpawnListEntry(EntityWolf.class, 8, 4, 4));
-        this.spawnableCreatureList.add(new Biome.SpawnListEntry(EntityRabbit.class, 4, 2, 3));
+        this.spawnableCreatureList.add(new SpawnListEntry(EntityWolf.class, 8, 4, 4));
+        this.spawnableCreatureList.add(new SpawnListEntry(EntityRabbit.class, 4, 2, 3));
         this.decorator.treesPerChunk = 10;
 
-        if (typeIn != BiomeTaiga.Type.MEGA && typeIn != BiomeTaiga.Type.MEGA_SPRUCE)
+        if (typeIn != Type.MEGA && typeIn != Type.MEGA_SPRUCE)
         {
             this.decorator.grassPerChunk = 1;
             this.decorator.mushroomsPerChunk = 1;
@@ -50,9 +50,9 @@
 
     public WorldGenAbstractTree getRandomTreeFeature(Random rand)
     {
-        if ((this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE) && rand.nextInt(3) == 0)
+        if ((this.type == Type.MEGA || this.type == Type.MEGA_SPRUCE) && rand.nextInt(3) == 0)
         {
-            return this.type != BiomeTaiga.Type.MEGA_SPRUCE && rand.nextInt(13) != 0 ? MEGA_PINE_GENERATOR : MEGA_SPRUCE_GENERATOR;
+            return this.type != Type.MEGA_SPRUCE && rand.nextInt(13) != 0 ? MEGA_PINE_GENERATOR : MEGA_SPRUCE_GENERATOR;
         }
         else
         {
@@ -67,7 +67,7 @@
 
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE)
+        if ((this.type == Type.MEGA || this.type == Type.MEGA_SPRUCE) && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ROCK))
         {
             int i = rand.nextInt(3);
 
@@ -82,6 +82,7 @@
 
         DOUBLE_PLANT_GENERATOR.setPlantType(BlockDoublePlant.EnumPlantType.FERN);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FLOWERS))
         for (int i1 = 0; i1 < 7; ++i1)
         {
             int j1 = rand.nextInt(16) + 8;
@@ -95,7 +96,7 @@
 
     public void genTerrainBlocks(World worldIn, Random rand, ChunkPrimer chunkPrimerIn, int x, int z, double noiseVal)
     {
-        if (this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE)
+        if (this.type == Type.MEGA || this.type == Type.MEGA_SPRUCE)
         {
             this.topBlock = Blocks.GRASS.getDefaultState();
             this.fillerBlock = Blocks.DIRT.getDefaultState();
