--- ../src-base/minecraft/net/minecraft/world/biome/BiomeCache.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeCache.java
@@ -10,24 +10,24 @@
 {
     private final BiomeProvider provider;
     private long lastCleanupTime;
-    private final Long2ObjectMap<BiomeCache.Block> cacheMap = new Long2ObjectOpenHashMap<BiomeCache.Block>(4096);
-    private final List<BiomeCache.Block> cache = Lists.<BiomeCache.Block>newArrayList();
+    private final Long2ObjectMap<Block> cacheMap = new Long2ObjectOpenHashMap<Block>(4096);
+    private final List<Block> cache = Lists.<Block>newArrayList();
 
     public BiomeCache(BiomeProvider provider)
     {
         this.provider = provider;
     }
 
-    public BiomeCache.Block getEntry(int x, int z)
+    public Block getEntry(int x, int z)
     {
         x = x >> 4;
         z = z >> 4;
         long i = (long)x & 4294967295L | ((long)z & 4294967295L) << 32;
-        BiomeCache.Block biomecache$block = (BiomeCache.Block)this.cacheMap.get(i);
+        Block biomecache$block = (Block)this.cacheMap.get(i);
 
         if (biomecache$block == null)
         {
-            biomecache$block = new BiomeCache.Block(x, z);
+            biomecache$block = new Block(x, z);
             this.cacheMap.put(i, biomecache$block);
             this.cache.add(biomecache$block);
         }
@@ -53,7 +53,7 @@
 
             for (int k = 0; k < this.cache.size(); ++k)
             {
-                BiomeCache.Block biomecache$block = this.cache.get(k);
+                Block biomecache$block = this.cache.get(k);
                 long l = i - biomecache$block.lastAccessTime;
 
                 if (l > 30000L || l < 0L)
