--- ../src-base/minecraft/net/minecraft/world/biome/BiomeSnow.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeSnow.java
@@ -20,7 +20,7 @@
     private final WorldGenIceSpike iceSpike = new WorldGenIceSpike();
     private final WorldGenIcePath icePatch = new WorldGenIcePath(4);
 
-    public BiomeSnow(boolean superIcyIn, Biome.BiomeProperties properties)
+    public BiomeSnow(boolean superIcyIn, BiomeProperties properties)
     {
         super(properties);
         this.superIcy = superIcyIn;
@@ -31,13 +31,13 @@
         }
 
         this.spawnableCreatureList.clear();
-        this.spawnableCreatureList.add(new Biome.SpawnListEntry(EntityRabbit.class, 10, 2, 3));
-        this.spawnableCreatureList.add(new Biome.SpawnListEntry(EntityPolarBear.class, 1, 1, 2));
-        Iterator<Biome.SpawnListEntry> iterator = this.spawnableMonsterList.iterator();
+        this.spawnableCreatureList.add(new SpawnListEntry(EntityRabbit.class, 10, 2, 3));
+        this.spawnableCreatureList.add(new SpawnListEntry(EntityPolarBear.class, 1, 1, 2));
+        Iterator<SpawnListEntry> iterator = this.spawnableMonsterList.iterator();
 
         while (iterator.hasNext())
         {
-            Biome.SpawnListEntry biome$spawnlistentry = iterator.next();
+            SpawnListEntry biome$spawnlistentry = iterator.next();
 
             if (biome$spawnlistentry.entityClass == EntitySkeleton.class)
             {
@@ -45,8 +45,8 @@
             }
         }
 
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntitySkeleton.class, 20, 4, 4));
-        this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityStray.class, 80, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntitySkeleton.class, 20, 4, 4));
+        this.spawnableMonsterList.add(new SpawnListEntry(EntityStray.class, 80, 4, 4));
     }
 
     public float getSpawningChance()
@@ -56,7 +56,7 @@
 
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.superIcy)
+        if (this.superIcy && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ICE))
         {
             for (int i = 0; i < 3; ++i)
             {
