--- ../src-base/minecraft/net/minecraft/world/biome/BiomeColorHelper.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeColorHelper.java
@@ -8,21 +8,21 @@
 @SideOnly(Side.CLIENT)
 public class BiomeColorHelper
 {
-    private static final BiomeColorHelper.ColorResolver GRASS_COLOR = new BiomeColorHelper.ColorResolver()
+    private static final ColorResolver GRASS_COLOR = new ColorResolver()
     {
         public int getColorAtPos(Biome biome, BlockPos blockPosition)
         {
             return biome.getGrassColorAtPos(blockPosition);
         }
     };
-    private static final BiomeColorHelper.ColorResolver FOLIAGE_COLOR = new BiomeColorHelper.ColorResolver()
+    private static final ColorResolver FOLIAGE_COLOR = new ColorResolver()
     {
         public int getColorAtPos(Biome biome, BlockPos blockPosition)
         {
             return biome.getFoliageColorAtPos(blockPosition);
         }
     };
-    private static final BiomeColorHelper.ColorResolver WATER_COLOR = new BiomeColorHelper.ColorResolver()
+    private static final ColorResolver WATER_COLOR = new ColorResolver()
     {
         public int getColorAtPos(Biome biome, BlockPos blockPosition)
         {
@@ -30,7 +30,7 @@
         }
     };
 
-    private static int getColorAtPos(IBlockAccess blockAccess, BlockPos pos, BiomeColorHelper.ColorResolver colorResolver)
+    private static int getColorAtPos(IBlockAccess blockAccess, BlockPos pos, ColorResolver colorResolver)
     {
         int i = 0;
         int j = 0;
