--- ../src-base/minecraft/net/minecraft/world/biome/BiomeForestMutated.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeForestMutated.java
@@ -5,9 +5,9 @@
 
 public class BiomeForestMutated extends BiomeForest
 {
-    public BiomeForestMutated(Biome.BiomeProperties properties)
+    public BiomeForestMutated(BiomeProperties properties)
     {
-        super(BiomeForest.Type.BIRCH, properties);
+        super(Type.BIRCH, properties);
     }
 
     public WorldGenAbstractTree getRandomTreeFeature(Random rand)
