--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -43,6 +_,8 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
 
 public class Goat extends Animal {
    public static final EntityDimensions f_149342_ = EntityDimensions.m_20395_(0.9F, 1.3F).m_20388_(0.7F);
@@ -151,8 +_,14 @@
    public InteractionResult m_6071_(Player p_149379_, InteractionHand p_149380_) {
       ItemStack itemstack = p_149379_.m_21120_(p_149380_);
       if (itemstack.m_150930_(Items.f_42446_) && !this.m_6162_()) {
+         // CraftBukkit start - Got milk?
+         org.bukkit.event.player.PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_149379_.f_19853_, p_149379_, this.m_142538_(), this.m_142538_(), null, itemstack, Items.f_42455_);
+         if (event.isCancelled()) {
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          p_149379_.m_5496_(this.m_149403_(), 1.0F, 1.0F);
-         ItemStack itemstack1 = ItemUtils.m_41813_(itemstack, p_149379_, Items.f_42455_.m_7968_());
+         ItemStack itemstack1 = ItemUtils.m_41813_(itemstack, p_149379_, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
          p_149379_.m_21008_(p_149380_, itemstack1);
          return InteractionResult.m_19078_(this.f_19853_.f_46443_);
       } else {
