--- ../src-base/minecraft/net/minecraft/entity/passive/AbstractChestHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/AbstractChestHorse.java
@@ -72,7 +72,7 @@
 
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+        // super.onDeath(cause); // CraftBukkit - moved down
 
         if (this.hasChest())
         {
@@ -81,8 +81,10 @@
                 this.dropItem(Item.getItemFromBlock(Blocks.CHEST), 1);
             }
 
-            this.setChested(false);
+            // this.setChested(false); // CraftBukkit - moved down
         }
+        super.onDeath(cause);
+        this.setChested(false);
     }
 
     public static void registerFixesAbstractChestHorse(DataFixer fixer, Class<?> entityClass)
