--- ../src-base/minecraft/net/minecraft/entity/passive/EntityRabbit.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityRabbit.java
@@ -68,8 +68,13 @@
         this.setSize(0.4F, 0.5F);
         this.jumpHelper = new EntityRabbit.RabbitJumpHelper(this);
         this.moveHelper = new EntityRabbit.RabbitMoveHelper(this);
+        this.initializePathFinderGoals(); // CraftBukkit - moved code
+    }
+    // CraftBukkit start - code from constructor
+    public void initializePathFinderGoals(){
         this.setMovementSpeed(0.0D);
     }
+    // CraftBukkit end
 
     protected void initEntityAI()
     {
