--- ../src-base/minecraft/net/minecraft/entity/passive/EntityAnimal.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityAnimal.java
@@ -23,6 +23,8 @@
     private int inLove;
     private EntityPlayer playerInLove;
 
+    public ItemStack breedItem; // CraftBukkit - Add breedItem variable
+
     public EntityAnimal(World worldIn)
     {
         super(worldIn);
@@ -110,7 +112,7 @@
         return 120;
     }
 
-    protected boolean canDespawn()
+    public boolean canDespawn()
     {
         return false;
     }
@@ -159,6 +161,7 @@
     {
         this.inLove = 600;
         this.playerInLove = player;
+        this.breedItem = player.inventory.getCurrentItem(); // CraftBukkit
         this.worldObj.setEntityState(this, (byte)18);
     }
 
