--- ../src-base/minecraft/net/minecraft/entity/passive/EntityLlama.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityLlama.java
@@ -66,7 +66,7 @@
         this.setSize(0.9F, 1.87F);
     }
 
-    private void setStrength(int strengthIn)
+    public void setStrength(int strengthIn)
     {
         this.dataManager.set(DATA_STRENGTH_ID, Integer.valueOf(Math.max(1, Math.min(5, strengthIn))));
     }
@@ -120,8 +120,8 @@
         this.tasks.addTask(6, new EntityAIWanderAvoidWater(this, 0.7D));
         this.tasks.addTask(7, new EntityAIWatchClosest(this, EntityPlayer.class, 6.0F));
         this.tasks.addTask(8, new EntityAILookIdle(this));
-        this.targetTasks.addTask(1, new EntityLlama.AIHurtByTarget(this));
-        this.targetTasks.addTask(2, new EntityLlama.AIDefendTarget(this));
+        this.targetTasks.addTask(1, new AIHurtByTarget(this));
+        this.targetTasks.addTask(2, new AIDefendTarget(this));
     }
 
     protected void applyEntityAttributes()
@@ -249,14 +249,14 @@
         this.setRandomStrength();
         int i;
 
-        if (livingdata instanceof EntityLlama.GroupData)
+        if (livingdata instanceof GroupData)
         {
-            i = ((EntityLlama.GroupData)livingdata).variant;
+            i = ((GroupData)livingdata).variant;
         }
         else
         {
             i = this.rand.nextInt(4);
-            livingdata = new EntityLlama.GroupData(i);
+            livingdata = new GroupData(i);
         }
 
         this.setVariant(i);
