--- ../src-base/minecraft/net/minecraft/entity/passive/EntityCow.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityCow.java
@@ -25,6 +25,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 
 public class EntityCow extends EntityAnimal
 {
@@ -95,16 +97,24 @@
 
         if (itemstack.getItem() == Items.BUCKET && !player.capabilities.isCreativeMode && !this.isChild())
         {
+            org.bukkit.Location loc = this.getBukkitEntity().getLocation();
+            org.bukkit.event.player.PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(player, loc.getBlockX(), loc.getBlockY(), loc.getBlockZ(), null, itemstack, Items.MILK_BUCKET);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+            ItemStack result = CraftItemStack.asNMSCopy(event.getItemStack());
+
             player.playSound(SoundEvents.ENTITY_COW_MILK, 1.0F, 1.0F);
             itemstack.shrink(1);
 
             if (itemstack.isEmpty())
             {
-                player.setHeldItem(hand, new ItemStack(Items.MILK_BUCKET));
+                player.setHeldItem(hand, result);
             }
-            else if (!player.inventory.addItemStackToInventory(new ItemStack(Items.MILK_BUCKET)))
+            else if (!player.inventory.addItemStackToInventory(result))
             {
-                player.dropItem(new ItemStack(Items.MILK_BUCKET), false);
+                player.dropItem(result, false);
             }
 
             return true;
