--- ../src-base/minecraft/net/minecraft/entity/passive/EntitySheep.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntitySheep.java
@@ -4,6 +4,9 @@
 import java.util.Map;
 import java.util.Random;
 import javax.annotation.Nullable;
+
+import org.bukkit.inventory.InventoryView;
+
 import net.minecraft.block.Block;
 import net.minecraft.entity.EntityAgeable;
 import net.minecraft.entity.EntityLiving;
