--- ../src-base/minecraft/net/minecraft/entity/passive/EntityTameable.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityTameable.java
@@ -223,6 +223,11 @@
         return this.aiSit;
     }
 
+    public void setAISit(EntityAISit aiSit)
+    {
+        this.aiSit = aiSit;
+    }
+
     public boolean shouldAttackEntity(EntityLivingBase target, EntityLivingBase owner)
     {
         return true;
