--- a/net/minecraft/block/AbstractFireBlock.java
+++ b/net/minecraft/block/AbstractFireBlock.java
@@ -6,6 +_,7 @@
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.BlockItemUseContext;
 import net.minecraft.particles.ParticleTypes;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.Direction;
 import net.minecraft.util.SoundCategory;
@@ -13,7 +_,9 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
@@ -111,7 +_,15 @@
       if (!p_196262_4_.func_230279_az_()) {
          p_196262_4_.func_241209_g_(p_196262_4_.func_223314_ad() + 1);
          if (p_196262_4_.func_223314_ad() == 0) {
-            p_196262_4_.func_70015_d(8);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock.at(p_196262_2_, p_196262_3_), p_196262_4_.getBukkitEntity(), 8);
+            p_196262_2_.getCBServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               p_196262_4_.putCallCombustEvent(false);
+               p_196262_4_.func_70015_d(event.getDuration());
+               p_196262_4_.putCallCombustEvent(true);
+            }
+            // CraftBukkit end
          }
 
          p_196262_4_.func_70097_a(DamageSource.field_76372_a, this.field_235325_g_);
@@ -124,21 +_,22 @@
       if (!p_220082_4_.func_203425_a(p_220082_1_.func_177230_c())) {
          if (func_242649_a(p_220082_2_)) {
             Optional<PortalSize> optional = PortalSize.func_242964_a(p_220082_2_, p_220082_3_, Direction.Axis.X);
+            optional =  net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_220082_2_, p_220082_3_, optional);
             if (optional.isPresent()) {
-               optional.get().func_150859_c();
+               optional.get().placePortalBlocks();
                return;
             }
          }
 
          if (!p_220082_1_.func_196955_c(p_220082_2_, p_220082_3_)) {
-            p_220082_2_.func_217377_a(p_220082_3_, false);
+            fireExtinguished(p_220082_2_, p_220082_3_); // CraftBukkit - fuel block broke
          }
 
       }
    }
 
    private static boolean func_242649_a(World p_242649_0_) {
-      return p_242649_0_.func_234923_W_() == World.field_234918_g_ || p_242649_0_.func_234923_W_() == World.field_234919_h_;
+      return p_242649_0_.func_230315_m_() != MinecraftServer.getServer().field_240767_f_.func_230520_a_().func_230516_a_(DimensionType.field_236001_e_); // LoliServer - Custom world to teleport
    }
 
    public void func_176208_a(World p_176208_1_, BlockPos p_176208_2_, BlockState p_176208_3_, PlayerEntity p_176208_4_) {
@@ -179,4 +_,12 @@
          }
       }
    }
+
+   // CraftBukkit start
+   protected void fireExtinguished(IWorld world, BlockPos position) {
+      if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockFadeEvent(world, position, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+         world.func_217377_a(position, false);
+      }
+   }
+   // CraftBukkit end
 }
