--- a/net/minecraft/block/AbstractPressurePlateBlock.java
+++ b/net/minecraft/block/AbstractPressurePlateBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public abstract class AbstractPressurePlateBlock extends Block {
    protected static final VoxelShape field_185509_a = Block.func_208617_a(1.0D, 0.0D, 1.0D, 15.0D, 0.5D, 15.0D);
@@ -66,6 +_,18 @@
       int i = this.func_180669_e(p_180666_1_, p_180666_2_);
       boolean flag = p_180666_4_ > 0;
       boolean flag1 = i > 0;
+
+      // CraftBukkit start - Interact Pressure Plate
+      org.bukkit.World bworld = p_180666_1_.getWorld();
+      org.bukkit.plugin.PluginManager manager = p_180666_1_.getCBServer().getPluginManager();
+      if (flag != flag1) {
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(p_180666_2_.func_177958_n(), p_180666_2_.func_177956_o(), p_180666_2_.func_177952_p()), p_180666_4_, i);
+         manager.callEvent(eventRedstone);
+         flag1 = eventRedstone.getNewCurrent() > 0;
+         i = eventRedstone.getNewCurrent();
+      }
+      // CraftBukkit end
+
       if (p_180666_4_ != i) {
          BlockState blockstate = this.func_176575_a(p_180666_3_, i);
          p_180666_1_.func_180501_a(p_180666_2_, blockstate, 2);
