--- a/net/minecraft/block/AbstractButtonBlock.java
+++ b/net/minecraft/block/AbstractButtonBlock.java
@@ -23,6 +_,10 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+// CraftBukkit start
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
 
 public abstract class AbstractButtonBlock extends HorizontalFaceBlock {
    public static final BooleanProperty field_176584_b = BlockStateProperties.field_208194_u;
@@ -90,6 +_,19 @@
       if (p_225533_1_.func_177229_b(field_176584_b)) {
          return ActionResultType.CONSUME;
       } else {
+         // CraftBukkit start
+         boolean powered = ((Boolean) p_225533_1_.func_177229_b(field_176584_b));
+         org.bukkit.block.Block block = p_225533_2_.getWorld().getBlockAt(p_225533_3_.func_177958_n(), p_225533_3_.func_177956_o(), p_225533_3_.func_177952_p());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_225533_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return ActionResultType.SUCCESS;
+         }
+         // CraftBukkit end
          this.func_226910_d_(p_225533_1_, p_225533_2_, p_225533_3_);
          this.func_196367_a(p_225533_4_, p_225533_2_, p_225533_3_, true);
          return ActionResultType.func_233537_a_(p_225533_2_.field_72995_K);
@@ -135,6 +_,16 @@
          if (this.field_150047_a) {
             this.func_185616_e(p_225534_1_, p_225534_2_, p_225534_3_);
          } else {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_225534_2_.getWorld().getBlockAt(p_225534_3_.func_177958_n(), p_225534_3_.func_177956_o(), p_225534_3_.func_177952_p());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+            p_225534_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+               return;
+            }
+            // CraftBukkit end
             p_225534_2_.func_180501_a(p_225534_3_, p_225534_1_.func_206870_a(field_176584_b, Boolean.valueOf(false)), 3);
             this.func_196368_e(p_225534_1_, p_225534_2_, p_225534_3_);
             this.func_196367_a((PlayerEntity)null, p_225534_2_, p_225534_3_, false);
@@ -153,7 +_,43 @@
       List<? extends Entity> list = p_185616_2_.func_217357_a(AbstractArrowEntity.class, p_185616_1_.func_196954_c(p_185616_2_, p_185616_3_).func_197752_a().func_186670_a(p_185616_3_));
       boolean flag = !list.isEmpty();
       boolean flag1 = p_185616_1_.func_177229_b(field_176584_b);
+      // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+      if (flag1 != flag && flag) {
+         org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+         boolean allowed = false;
+
+         // If all of the events are cancelled block the button press, else allow
+         for (Object object : list) {
+            if (object != null) {
+               EntityInteractEvent event = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+               p_185616_2_.getCBServer().getPluginManager().callEvent(event);
+
+               if (!event.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+
+         if (!allowed) {
+            return;
+         }
+      }
+         // CraftBukkit end
       if (flag != flag1) {
+         // CraftBukkit start
+         boolean powered = flag1;
+         org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_185616_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((flag && eventRedstone.getNewCurrent() <= 0) || (!flag && eventRedstone.getNewCurrent() > 0)) {
+            return;
+         }
+         // CraftBukkit end
          p_185616_2_.func_180501_a(p_185616_3_, p_185616_1_.func_206870_a(field_176584_b, Boolean.valueOf(flag)), 3);
          this.func_196368_e(p_185616_1_, p_185616_2_, p_185616_3_);
          this.func_196367_a((PlayerEntity)null, p_185616_2_, p_185616_3_, flag);
