--- a/net/minecraft/block/BambooSaplingBlock.java
+++ b/net/minecraft/block/BambooSaplingBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class BambooSaplingBlock extends Block implements IGrowable {
    protected static final VoxelShape field_220088_a = Block.func_208617_a(4.0D, 0.0D, 4.0D, 12.0D, 12.0D, 12.0D);
@@ -78,6 +_,6 @@
    }
 
    protected void func_220087_a(World p_220087_1_, BlockPos p_220087_2_) {
-      p_220087_1_.func_180501_a(p_220087_2_.func_177984_a(), Blocks.field_222405_kQ.func_176223_P().func_206870_a(BambooBlock.field_220265_e, BambooLeaves.SMALL), 3);
+      CraftEventFactory.handleBlockSpreadEvent(p_220087_1_, p_220087_2_, p_220087_2_.func_177984_a(), Blocks.field_222405_kQ.func_176223_P().func_206870_a(BambooBlock.field_220265_e, BambooLeaves.SMALL), 3); // CraftBukkit - BlockSpreadEvent
    }
 }
