--- a/net/minecraft/block/BedBlock.java
+++ b/net/minecraft/block/BedBlock.java
@@ -100,10 +_,20 @@
 
             return ActionResultType.SUCCESS;
          } else {
+            // CraftBukkit start
+            BlockState finalblockstate = p_225533_1_;
+            BlockPos finalblockpos = p_225533_3_;
+
+            // CraftBukkit end
             p_225533_4_.func_213819_a(p_225533_3_).ifLeft((p_220173_1_) -> {
-               if (p_220173_1_ != null) {
-                  p_225533_4_.func_146105_b(p_220173_1_.func_221259_a(), true);
-               }
+               // CraftBukkit start - handling bed explosion from below here
+               if(p_220173_1_ == PlayerEntity.SleepResult.NOT_POSSIBLE_HERE) {
+                  this.explodeBed(finalblockstate, p_225533_2_, finalblockpos);
+               } else
+                  // CraftBukkit end
+                  if (p_220173_1_ != null) {
+                     p_225533_4_.func_146105_b(p_220173_1_.func_221259_a(), true);
+                  }
 
             });
             return ActionResultType.SUCCESS;
@@ -111,8 +_,24 @@
       }
    }
 
+   // CraftBukkit start
+   private ActionResultType explodeBed(BlockState state, World world, BlockPos pos) {
+      world.func_217377_a(pos, false);
+      BlockPos pos1 = pos.func_177972_a((state.func_177229_b(BedBlock.field_185512_D)).func_176734_d());
+
+      if(world.func_180495_p(pos1).func_177230_c() == this) {
+         world.func_217377_a(pos1, false);
+      }
+
+      world.func_230546_a_((Entity)null, DamageSource.func_233546_a_(), (ExplosionContext) null, (double)pos.func_177958_n() + 0.5D, (double)pos.func_177956_o() + 0.5D, (double)pos.func_177952_p() + 0.5D, 5.0F, true, Explosion.Mode.DESTROY);
+
+      return ActionResultType.SUCCESS;
+   }
+   // CraftBukkit end
+
    public static boolean func_235330_a_(World p_235330_0_) {
-      return p_235330_0_.func_230315_m_().func_241510_j_();
+      // CraftBukkit - moved world and biome check into EntityHuman
+      return true || p_235330_0_.func_230315_m_().func_241510_j_();
    }
 
    private boolean func_226861_a_(World p_226861_1_, BlockPos p_226861_2_) {
