--- a/net/minecraft/block/CakeBlock.java
+++ b/net/minecraft/block/CakeBlock.java
@@ -1,6 +_,7 @@
 package net.minecraft.block;
 
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.pathfinding.PathType;
 import net.minecraft.state.IntegerProperty;
@@ -52,7 +_,17 @@
          return ActionResultType.PASS;
       } else {
          p_226911_4_.func_195066_a(Stats.field_188076_J);
-         p_226911_4_.func_71024_bL().func_75122_a(2, 0.1F);
+         // CraftBukkit start
+         int oldFoodLevel = p_226911_4_.func_71024_bL().field_75127_a;
+
+         org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callFoodLevelChangeEvent(p_226911_4_, 2 + oldFoodLevel);
+
+         if (!event.isCancelled()) {
+            p_226911_4_.func_71024_bL().func_75122_a(event.getFoodLevel() - oldFoodLevel, 0.1F);
+         }
+
+         ((ServerPlayerEntity) p_226911_4_).getBukkitEntity().sendHealthUpdate();
+         // CraftBukkit end
          int i = p_226911_3_.func_177229_b(field_176589_a);
          if (i < 6) {
             p_226911_1_.func_180501_a(p_226911_2_, p_226911_3_.func_206870_a(field_176589_a, Integer.valueOf(i + 1)), 3);
