--- a/net/minecraft/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/block/CarvedPumpkinBlock.java
@@ -21,6 +_,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class CarvedPumpkinBlock extends HorizontalBlock implements IArmorVanishable {
    public static final DirectionProperty field_196359_a = HorizontalBlock.field_185512_D;
@@ -53,17 +_,27 @@
 
    private void func_196358_b(World p_196358_1_, BlockPos p_196358_2_) {
       BlockPattern.PatternHelper blockpattern$patternhelper = this.func_196355_e().func_177681_a(p_196358_1_, p_196358_2_);
+      BlockStateListPopulator blockList = new BlockStateListPopulator(p_196358_1_); // CraftBukkit - Use BlockStateListPopulator
       if (blockpattern$patternhelper != null) {
          for(int i = 0; i < this.func_196355_e().func_177685_b(); ++i) {
             CachedBlockInfo cachedblockinfo = blockpattern$patternhelper.func_177670_a(0, i, 0);
-            p_196358_1_.func_180501_a(cachedblockinfo.func_177508_d(), Blocks.field_150350_a.func_176223_P(), 2);
-            p_196358_1_.func_217379_c(2001, cachedblockinfo.func_177508_d(), Block.func_196246_j(cachedblockinfo.func_177509_a()));
+            blockList.func_180501_a(cachedblockinfo.func_177508_d(), Blocks.field_150350_a.func_176223_P(), 2); // CraftBukkit
+            //world.playEvent(2001, cachedblockinfo.getPos(), Block.getStateId(cachedblockinfo.getBlockState())); // CraftBukkit
          }
 
          SnowGolemEntity snowgolementity = EntityType.field_200745_ak.func_200721_a(p_196358_1_);
          BlockPos blockpos1 = blockpattern$patternhelper.func_177670_a(0, 2, 0).func_177508_d();
          snowgolementity.func_70012_b((double)blockpos1.func_177958_n() + 0.5D, (double)blockpos1.func_177956_o() + 0.05D, (double)blockpos1.func_177952_p() + 0.5D, 0.0F, 0.0F);
-         p_196358_1_.func_217376_c(snowgolementity);
+
+         // CraftBukkit start
+         if (!p_196358_1_.addEntity(snowgolementity, SpawnReason.BUILD_SNOWMAN)) {
+            return;
+         }
+         for (BlockPos blockPos : blockList.getBlocks()) {
+            p_196358_1_.func_217379_c(2001, blockPos, Block.func_196246_j(p_196358_1_.func_180495_p(blockPos)));
+         }
+         blockList.updateList();
+         // CraftBukkit end
 
          for(ServerPlayerEntity serverplayerentity : p_196358_1_.func_217357_a(ServerPlayerEntity.class, snowgolementity.func_174813_aQ().func_186662_g(5.0D))) {
             CriteriaTriggers.field_192133_m.func_192229_a(serverplayerentity, snowgolementity);
@@ -79,8 +_,8 @@
             for(int j = 0; j < this.func_196357_g().func_177684_c(); ++j) {
                for(int k = 0; k < this.func_196357_g().func_177685_b(); ++k) {
                   CachedBlockInfo cachedblockinfo2 = blockpattern$patternhelper.func_177670_a(j, k, 0);
-                  p_196358_1_.func_180501_a(cachedblockinfo2.func_177508_d(), Blocks.field_150350_a.func_176223_P(), 2);
-                  p_196358_1_.func_217379_c(2001, cachedblockinfo2.func_177508_d(), Block.func_196246_j(cachedblockinfo2.func_177509_a()));
+                  blockList.func_180501_a(cachedblockinfo2.func_177508_d(), Blocks.field_150350_a.func_176223_P(), 2); // CraftBukkit
+                  //world.playEvent(2001, cachedblockinfo2.getPos(), Block.getStateId(cachedblockinfo2.getBlockState())); // CraftBukkit
                }
             }
 
@@ -88,7 +_,16 @@
             IronGolemEntity irongolementity = EntityType.field_200757_aw.func_200721_a(p_196358_1_);
             irongolementity.func_70849_f(true);
             irongolementity.func_70012_b((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.05D, (double)blockpos.func_177952_p() + 0.5D, 0.0F, 0.0F);
-            p_196358_1_.func_217376_c(irongolementity);
+
+            // CraftBukkit start
+            if (!p_196358_1_.addEntity(irongolementity, SpawnReason.BUILD_IRONGOLEM)) {
+               return;
+            }
+            for (BlockPos blockPos : blockList.getBlocks()) {
+               p_196358_1_.func_217379_c(2001, blockPos, Block.func_196246_j(p_196358_1_.func_180495_p(blockPos)));
+            }
+            blockList.updateList();
+            // CraftBukkit end
 
             for(ServerPlayerEntity serverplayerentity1 : p_196358_1_.func_217357_a(ServerPlayerEntity.class, irongolementity.func_174813_aQ().func_186662_g(5.0D))) {
                CriteriaTriggers.field_192133_m.func_192229_a(serverplayerentity1, irongolementity);
