--- a/net/minecraft/block/CauldronBlock.java
+++ b/net/minecraft/block/CauldronBlock.java
@@ -30,7 +_,7 @@
 import net.minecraft.util.math.shapes.VoxelShapes;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
-
+import org.bukkit.event.block.CauldronLevelChangeEvent; // CraftBukkit
 public class CauldronBlock extends Block {
    public static final IntegerProperty field_176591_a = BlockStateProperties.field_208130_ae;
    private static final VoxelShape field_196403_b = func_208617_a(2.0D, 4.0D, 2.0D, 14.0D, 16.0D, 14.0D);
@@ -53,8 +_,13 @@
       int i = p_196262_1_.func_177229_b(field_176591_a);
       float f = (float)p_196262_3_.func_177956_o() + (6.0F + (float)(3 * i)) / 16.0F;
       if (!p_196262_2_.field_72995_K && p_196262_4_.func_70027_ad() && i > 0 && p_196262_4_.func_226278_cu_() <= (double)f) {
+         // CraftBukkit start
+         if (!this.changeLevel(p_196262_2_, p_196262_3_, p_196262_1_, i - 1, p_196262_4_, CauldronLevelChangeEvent.ChangeReason.EXTINGUISH)) {
+            return;
+         }
          p_196262_4_.func_70066_B();
-         this.func_176590_a(p_196262_2_, p_196262_3_, p_196262_1_, i - 1);
+          // this.setWaterLevel(worldIn, pos, state, i - 1);
+          // CraftBukkit end
       }
 
    }
@@ -68,18 +_,27 @@
          Item item = itemstack.func_77973_b();
          if (item == Items.field_151131_as) {
             if (i < 3 && !p_225533_2_.field_72995_K) {
+               // CraftBukkit start
+               if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, 3, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+                  return ActionResultType.SUCCESS;
+               }
                if (!p_225533_4_.field_71075_bZ.field_75098_d) {
                   p_225533_4_.func_184611_a(p_225533_5_, new ItemStack(Items.field_151133_ar));
                }
 
                p_225533_4_.func_195066_a(Stats.field_188077_K);
-               this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, 3);
+               //this.setWaterLevel(worldIn, pos, state, 3);
+               // CraftBukkit end
                p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_187624_K, SoundCategory.BLOCKS, 1.0F, 1.0F);
             }
 
             return ActionResultType.func_233537_a_(p_225533_2_.field_72995_K);
          } else if (item == Items.field_151133_ar) {
             if (i == 3 && !p_225533_2_.field_72995_K) {
+               // CraftBukkit start
+               if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, 0, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+                  return ActionResultType.SUCCESS;
+               }
                if (!p_225533_4_.field_71075_bZ.field_75098_d) {
                   itemstack.func_190918_g(1);
                   if (itemstack.func_190926_b()) {
@@ -90,13 +_,18 @@
                }
 
                p_225533_4_.func_195066_a(Stats.field_188078_L);
-               this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, 0);
+               //this.setWaterLevel(worldIn, pos, state, 0);
+               // CraftBukkit end
                p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_187630_M, SoundCategory.BLOCKS, 1.0F, 1.0F);
             }
 
             return ActionResultType.func_233537_a_(p_225533_2_.field_72995_K);
          } else if (item == Items.field_151069_bo) {
             if (i > 0 && !p_225533_2_.field_72995_K) {
+               // CraftBukkit start
+               if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, i - 1, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                  return ActionResultType.SUCCESS;
+               }
                if (!p_225533_4_.field_71075_bZ.field_75098_d) {
                   ItemStack itemstack4 = PotionUtils.func_185188_a(new ItemStack(Items.field_151068_bn), Potions.field_185230_b);
                   p_225533_4_.func_195066_a(Stats.field_188078_L);
@@ -111,12 +_,17 @@
                }
 
                p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_187615_H, SoundCategory.BLOCKS, 1.0F, 1.0F);
-               this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, i - 1);
+               //this.setWaterLevel(worldIn, pos, state, i - 1);
+               // CraftBukkit end
             }
 
             return ActionResultType.func_233537_a_(p_225533_2_.field_72995_K);
          } else if (item == Items.field_151068_bn && PotionUtils.func_185191_c(itemstack) == Potions.field_185230_b) {
             if (i < 3 && !p_225533_2_.field_72995_K) {
+               // CraftBukkit start
+               if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, i - 1, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                  return ActionResultType.SUCCESS;
+               }
                if (!p_225533_4_.field_71075_bZ.field_75098_d) {
                   ItemStack itemstack3 = new ItemStack(Items.field_151069_bo);
                   p_225533_4_.func_195066_a(Stats.field_188078_L);
@@ -127,7 +_,8 @@
                }
 
                p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_191241_J, SoundCategory.BLOCKS, 1.0F, 1.0F);
-               this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, i + 1);
+               //this.setWaterLevel(worldIn, pos, state, i + 1);
+               // CraftBukkit end
             }
 
             return ActionResultType.func_233537_a_(p_225533_2_.field_72995_K);
@@ -135,8 +_,13 @@
             if (i > 0 && item instanceof IDyeableArmorItem) {
                IDyeableArmorItem idyeablearmoritem = (IDyeableArmorItem)item;
                if (idyeablearmoritem.func_200883_f_(itemstack) && !p_225533_2_.field_72995_K) {
+                  // CraftBukkit start
+                  if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, i - 1, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                     return ActionResultType.SUCCESS;
+                  }
                   idyeablearmoritem.func_200884_g(itemstack);
-                  this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, i - 1);
+                  //this.setWaterLevel(worldIn, pos, state, i - 1);
+                  // CraftBukkit end
                   p_225533_4_.func_195066_a(Stats.field_188079_M);
                   return ActionResultType.SUCCESS;
                }
@@ -144,13 +_,18 @@
 
             if (i > 0 && item instanceof BannerItem) {
                if (BannerTileEntity.func_175113_c(itemstack) > 0 && !p_225533_2_.field_72995_K) {
+                  // CraftBukkit start
+                  if (!this.changeLevel(p_225533_2_, p_225533_3_, p_225533_1_, i - 1, p_225533_4_, CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                     return ActionResultType.SUCCESS;
+                  }
                   ItemStack itemstack2 = itemstack.func_77946_l();
                   itemstack2.func_190920_e(1);
                   BannerTileEntity.func_175117_e(itemstack2);
                   p_225533_4_.func_195066_a(Stats.field_188080_N);
                   if (!p_225533_4_.field_71075_bZ.field_75098_d) {
                      itemstack.func_190918_g(1);
-                     this.func_176590_a(p_225533_2_, p_225533_3_, p_225533_1_, i - 1);
+                     //this.setWaterLevel(worldIn, pos, state, i - 1);
+                     // CraftBukkit end
                   }
 
                   if (itemstack.func_190926_b()) {
@@ -185,18 +_,36 @@
       }
    }
 
+   // CraftBukkit start
    public void func_176590_a(World p_176590_1_, BlockPos p_176590_2_, BlockState p_176590_3_, int p_176590_4_) {
-      p_176590_1_.func_180501_a(p_176590_2_, p_176590_3_.func_206870_a(field_176591_a, Integer.valueOf(MathHelper.func_76125_a(p_176590_4_, 0, 3))), 2);
+      this.changeLevel(p_176590_1_, p_176590_2_, p_176590_3_, p_176590_4_, null, CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
       p_176590_1_.func_175666_e(p_176590_2_, this);
    }
 
+   private boolean changeLevel(World world, BlockPos blockposition, BlockState iblockdata, int i, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+      int newLevel = Integer.valueOf(MathHelper.func_76125_a(i, 0, 3));
+      CauldronLevelChangeEvent event = new CauldronLevelChangeEvent(
+              world.getWorld().getBlockAt(blockposition.func_177958_n(), blockposition.func_177956_o(), blockposition.func_177952_p()),
+              (entity == null) ? null : entity.getBukkitEntity(), reason, iblockdata.func_177229_b(CauldronBlock.field_176591_a), newLevel
+      );
+      world.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return false;
+      }
+      world.func_180501_a(blockposition, iblockdata.func_206870_a(field_176591_a, event.getNewLevel()), 2);
+
+      world.func_175666_e(blockposition, this);
+      return true;
+      // CraftBukkit end
+   }
+
    public void func_176224_k(World p_176224_1_, BlockPos p_176224_2_) {
       if (p_176224_1_.field_73012_v.nextInt(20) == 1) {
          float f = p_176224_1_.func_226691_t_(p_176224_2_).func_225486_c(p_176224_2_);
          if (!(f < 0.15F)) {
             BlockState blockstate = p_176224_1_.func_180495_p(p_176224_2_);
             if (blockstate.func_177229_b(field_176591_a) < 3) {
-               p_176224_1_.func_180501_a(p_176224_2_, blockstate.func_235896_a_(field_176591_a), 2);
+               this.func_176590_a(p_176224_1_, p_176224_2_, blockstate.func_235896_a_(field_176591_a), 2); // CraftBukkit
             }
 
          }
