--- a/net/minecraft/block/CocoaBlock.java
+++ b/net/minecraft/block/CocoaBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class CocoaBlock extends HorizontalBlock implements IGrowable {
    public static final IntegerProperty field_176501_a = BlockStateProperties.field_208167_T;
@@ -35,10 +_,11 @@
    }
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_2_.field_73012_v.nextInt(5) == 0) {
+      if (true) {
          int i = p_225542_1_.func_177229_b(field_176501_a);
-         if (i < 2) {
-            p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_.func_206870_a(field_176501_a, Integer.valueOf(i + 1)), 2);
+         if (i < 2 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_225542_2_, p_225542_3_, p_225542_1_, p_225542_2_.field_73012_v.nextInt(Math.max(1, (int) (100.0F / p_225542_2_.spigotConfig.cocoaModifier) * 5)) == 0)) { // Spigot
+            CraftEventFactory.handleBlockGrowEvent(p_225542_2_, p_225542_3_, p_225542_1_.func_206870_a(field_176501_a, Integer.valueOf(i + 1)), 2); // CraftBukkkit
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_225542_2_, p_225542_3_, p_225542_1_);
          }
       }
 
@@ -95,7 +_,7 @@
    }
 
    public void func_225535_a_(ServerWorld p_225535_1_, Random p_225535_2_, BlockPos p_225535_3_, BlockState p_225535_4_) {
-      p_225535_1_.func_180501_a(p_225535_3_, p_225535_4_.func_206870_a(field_176501_a, Integer.valueOf(p_225535_4_.func_177229_b(field_176501_a) + 1)), 2);
+      CraftEventFactory.handleBlockGrowEvent(p_225535_1_, p_225535_3_, p_225535_4_.func_206870_a(field_176501_a, Integer.valueOf(p_225535_4_.func_177229_b(field_176501_a) + 1)), 2); // CraftBukkit
    }
 
    protected void func_206840_a(StateContainer.Builder<Block, BlockState> p_206840_1_) {
