--- a/net/minecraft/block/CommandBlockBlock.java
+++ b/net/minecraft/block/CommandBlockBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class CommandBlockBlock extends ContainerBlock {
    private static final Logger field_193388_c = LogManager.getLogger();
@@ -50,6 +_,17 @@
             CommandBlockTileEntity commandblocktileentity = (CommandBlockTileEntity)tileentity;
             boolean flag = p_220069_2_.func_175640_z(p_220069_3_);
             boolean flag1 = commandblocktileentity.func_184255_d();
+
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = p_220069_2_.getWorld().getBlockAt(p_220069_3_.func_177958_n(), p_220069_3_.func_177956_o(), p_220069_3_.func_177952_p());
+            int old = flag1 ? 15 : 0;
+            int current = flag ? 15 : 0;
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, old, current);
+            p_220069_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+            flag = eventRedstone.getNewCurrent() > 0;
+            // CraftBukkit end
+
             commandblocktileentity.func_184250_a(flag);
             if (!flag1 && !commandblocktileentity.func_184254_e() && commandblocktileentity.func_184251_i() != CommandBlockTileEntity.Mode.SEQUENCE) {
                if (flag) {
