--- a/net/minecraft/block/ConcretePowderBlock.java
+++ b/net/minecraft/block/ConcretePowderBlock.java
@@ -10,6 +_,8 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlockState;
+import org.bukkit.event.block.BlockFormEvent;
 
 public class ConcretePowderBlock extends FallingBlock {
    private final BlockState field_200294_a;
@@ -21,7 +_,7 @@
 
    public void func_176502_a_(World p_176502_1_, BlockPos p_176502_2_, BlockState p_176502_3_, BlockState p_176502_4_, FallingBlockEntity p_176502_5_) {
       if (func_230137_b_(p_176502_1_, p_176502_2_, p_176502_4_)) {
-         p_176502_1_.func_180501_a(p_176502_2_, this.field_200294_a, 3);
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.handleBlockFormEvent(p_176502_1_, p_176502_2_, this.field_200294_a, 3); // CraftBukkit
       }
 
    }
@@ -30,7 +_,19 @@
       IBlockReader iblockreader = p_196258_1_.func_195991_k();
       BlockPos blockpos = p_196258_1_.func_195995_a();
       BlockState blockstate = iblockreader.func_180495_p(blockpos);
-      return func_230137_b_(iblockreader, blockpos, blockstate) ? this.field_200294_a : super.func_196258_a(p_196258_1_);
+      // CraftBukkit start
+      if (!func_230137_b_(iblockreader, blockpos, blockstate)) {
+         return super.func_196258_a(p_196258_1_);
+      }
+      CraftBlockState blockState = CraftBlockState.getBlockState(p_196258_1_.func_195991_k(), blockpos);
+      blockState.setData(this.field_200294_a);
+      BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+      p_196258_1_.func_195991_k().getCBServer().getPluginManager().callEvent(event);
+      if (!event.isCancelled()) {
+         return blockState.getHandle();
+      }
+      return super.func_196258_a(p_196258_1_);
+      // CraftBukkit end
    }
 
    private static boolean func_230137_b_(IBlockReader p_230137_0_, BlockPos p_230137_1_, BlockState p_230137_2_) {
@@ -61,7 +_,18 @@
    }
 
    public BlockState func_196271_a(BlockState p_196271_1_, Direction p_196271_2_, BlockState p_196271_3_, IWorld p_196271_4_, BlockPos p_196271_5_, BlockPos p_196271_6_) {
-      return func_196441_b(p_196271_4_, p_196271_5_) ? this.field_200294_a : super.func_196271_a(p_196271_1_, p_196271_2_, p_196271_3_, p_196271_4_, p_196271_5_, p_196271_6_);
+      // CraftBukkit start
+      if (func_196441_b(p_196271_4_, p_196271_5_)) {
+         CraftBlockState blockState = CraftBlockState.getBlockState(p_196271_4_, p_196271_5_);
+         blockState.setData(this.field_200294_a);
+         BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+         p_196271_4_.getMinecraftWorld().getCBServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            return blockState.getHandle();
+         }
+      }
+      return super.func_196271_a(p_196271_1_, p_196271_2_, p_196271_3_, p_196271_4_, p_196271_5_, p_196271_6_);
+      // CraftBukkit end
    }
 
    @OnlyIn(Dist.CLIENT)
