--- a/net/minecraft/block/CoralBlock.java
+++ b/net/minecraft/block/CoralBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class CoralBlock extends Block {
    private final Block field_204403_a;
@@ -21,6 +_,11 @@
 
    public void func_225534_a_(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
       if (!this.func_203943_a(p_225534_2_, p_225534_3_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_225534_2_, p_225534_3_, this.field_204403_a.func_176223_P()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_225534_2_.func_180501_a(p_225534_3_, this.field_204403_a.func_176223_P(), 2);
       }
 
