--- a/net/minecraft/block/DaylightDetectorBlock.java
+++ b/net/minecraft/block/DaylightDetectorBlock.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.LightType;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class DaylightDetectorBlock extends ContainerBlock {
    public static final IntegerProperty field_176436_a = BlockStateProperties.field_208136_ak;
@@ -56,6 +_,7 @@
 
          i = MathHelper.func_76125_a(i, 0, 15);
          if (p_196319_0_.func_177229_b(field_176436_a) != i) {
+            i = CraftEventFactory.callRedstoneChange(p_196319_1_, p_196319_2_, p_196319_0_.func_177229_b(field_176436_a), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
             p_196319_1_.func_180501_a(p_196319_2_, p_196319_0_.func_206870_a(field_176436_a, Integer.valueOf(i)), 3);
          }
 
