--- a/net/minecraft/block/DispenserBlock.java
+++ b/net/minecraft/block/DispenserBlock.java
@@ -40,9 +_,10 @@
 public class DispenserBlock extends ContainerBlock {
    public static final DirectionProperty field_176441_a = DirectionalBlock.field_176387_N;
    public static final BooleanProperty field_176440_b = BlockStateProperties.field_208197_x;
-   private static final Map<Item, IDispenseItemBehavior> field_149943_a = Util.func_200696_a(new Object2ObjectOpenHashMap<>(), (p_212564_0_) -> {
+   public static final Map<Item, IDispenseItemBehavior> field_149943_a = Util.func_200696_a(new Object2ObjectOpenHashMap<>(), (p_212564_0_) -> {
       p_212564_0_.defaultReturnValue(new DefaultDispenseItemBehavior());
    });
+   public static boolean eventFired = false; // CraftBukkit
 
    public static void func_199774_a(IItemProvider p_199774_0_, IDispenseItemBehavior p_199774_1_) {
       field_149943_a.put(p_199774_0_.func_199767_j(), p_199774_1_);
@@ -71,7 +_,7 @@
       }
    }
 
-   protected void func_176439_d(ServerWorld p_176439_1_, BlockPos p_176439_2_) {
+   public void func_176439_d(ServerWorld p_176439_1_, BlockPos p_176439_2_) {
       ProxyBlockSource proxyblocksource = new ProxyBlockSource(p_176439_1_, p_176439_2_);
       DispenserTileEntity dispensertileentity = proxyblocksource.func_150835_j();
       int i = dispensertileentity.func_146017_i();
@@ -81,6 +_,7 @@
          ItemStack itemstack = dispensertileentity.func_70301_a(i);
          IDispenseItemBehavior idispenseitembehavior = this.func_149940_a(itemstack);
          if (idispenseitembehavior != IDispenseItemBehavior.field_223216_a_) {
+            eventFired = false; // CraftBukkit - reset event status
             dispensertileentity.func_70299_a(i, idispenseitembehavior.dispense(proxyblocksource, itemstack));
          }
 
