--- a/net/minecraft/block/DoorBlock.java
+++ b/net/minecraft/block/DoorBlock.java
@@ -32,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DoorBlock extends Block {
    public static final DirectionProperty field_176520_a = HorizontalBlock.field_185512_D;
@@ -177,8 +_,21 @@
    }
 
    public void func_220069_a(BlockState p_220069_1_, World p_220069_2_, BlockPos p_220069_3_, Block p_220069_4_, BlockPos p_220069_5_, boolean p_220069_6_) {
-      boolean flag = p_220069_2_.func_175640_z(p_220069_3_) || p_220069_2_.func_175640_z(p_220069_3_.func_177972_a(p_220069_1_.func_177229_b(field_176523_O) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
-      if (p_220069_4_ != this && flag != p_220069_1_.func_177229_b(field_176522_N)) {
+      // CraftBukkit start
+      BlockPos otherHalf = p_220069_3_.func_177972_a(p_220069_1_.func_177229_b(field_176523_O) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN);
+      org.bukkit.World bworld = p_220069_2_.getWorld();
+      org.bukkit.block.Block bukkitBlock = bworld.getBlockAt(p_220069_3_.func_177958_n(), p_220069_3_.func_177956_o(), p_220069_3_.func_177952_p());
+      org.bukkit.block.Block blockTop = bworld.getBlockAt(otherHalf.func_177958_n(), otherHalf.func_177956_o(), otherHalf.func_177952_p());
+      int power = bukkitBlock.getBlockPower();
+      int powerTop = blockTop.getBlockPower();
+      if (powerTop > power) power = powerTop;
+      int oldPower = (Boolean) p_220069_1_.func_177229_b(DoorBlock.field_176522_N) ? 15 : 0;
+      if(oldPower == 0 ^ power == 0){
+         BlockRedstoneEvent blockRedstoneEvent = new BlockRedstoneEvent(bukkitBlock, oldPower, power);
+         p_220069_2_.getCBServer().getPluginManager().callEvent(blockRedstoneEvent);
+
+         boolean flag = blockRedstoneEvent.getNewCurrent() > 0;
+         // CraftBukkit end
          if (flag != p_220069_1_.func_177229_b(field_176519_b)) {
             this.func_196426_b(p_220069_2_, p_220069_3_, flag);
          }
