--- a/net/minecraft/block/DragonEggBlock.java
+++ b/net/minecraft/block/DragonEggBlock.java
@@ -12,6 +_,7 @@
 import net.minecraft.util.math.shapes.VoxelShape;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
+import org.bukkit.event.block.BlockFromToEvent;
 
 public class DragonEggBlock extends FallingBlock {
    protected static final VoxelShape field_196444_a = Block.func_208617_a(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
@@ -37,6 +_,19 @@
       for(int i = 0; i < 1000; ++i) {
          BlockPos blockpos = p_196443_3_.func_177982_a(p_196443_2_.field_73012_v.nextInt(16) - p_196443_2_.field_73012_v.nextInt(16), p_196443_2_.field_73012_v.nextInt(8) - p_196443_2_.field_73012_v.nextInt(8), p_196443_2_.field_73012_v.nextInt(16) - p_196443_2_.field_73012_v.nextInt(16));
          if (p_196443_2_.func_180495_p(blockpos).func_196958_f()) {
+            // CraftBukkit start
+            org.bukkit.block.Block from = p_196443_2_.getWorld().getBlockAt(p_196443_3_.func_177958_n(),p_196443_3_.func_177956_o(),p_196443_3_.func_177952_p());
+            org.bukkit.block.Block to = p_196443_2_.getWorld().getBlockAt(blockpos.func_177958_n(),blockpos.func_177956_o(),blockpos.func_177952_p());
+            BlockFromToEvent event = new BlockFromToEvent(from, to);
+            p_196443_2_.getCBServer().getPluginManager().callEvent(event);
+
+            if(event.isCancelled()){
+               return;
+            }
+
+            blockpos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+            // CraftBukkit end
+
             if (p_196443_2_.field_72995_K) {
                for(int j = 0; j < 128; ++j) {
                   double d0 = p_196443_2_.field_73012_v.nextDouble();
