--- a/net/minecraft/block/EndPortalBlock.java
+++ b/net/minecraft/block/EndPortalBlock.java
@@ -2,9 +_,11 @@
 
 import java.util.Random;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.fluid.Fluid;
 import net.minecraft.item.ItemStack;
 import net.minecraft.particles.ParticleTypes;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.EndPortalTileEntity;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.RegistryKey;
@@ -13,11 +_,14 @@
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
 import net.minecraft.util.math.shapes.VoxelShapes;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndPortalBlock extends ContainerBlock {
    protected static final VoxelShape field_196323_a = Block.func_208617_a(0.0D, 0.0D, 0.0D, 16.0D, 12.0D, 16.0D);
@@ -36,11 +_,36 @@
 
    public void func_196262_a(BlockState p_196262_1_, World p_196262_2_, BlockPos p_196262_3_, Entity p_196262_4_) {
       if (p_196262_2_ instanceof ServerWorld && !p_196262_4_.func_184218_aH() && !p_196262_4_.func_184207_aI() && p_196262_4_.func_184222_aU() && VoxelShapes.func_197879_c(VoxelShapes.func_197881_a(p_196262_4_.func_174813_aQ().func_72317_d((double)(-p_196262_3_.func_177958_n()), (double)(-p_196262_3_.func_177956_o()), (double)(-p_196262_3_.func_177952_p()))), p_196262_1_.func_196954_c(p_196262_2_, p_196262_3_), IBooleanFunction.field_223238_i_)) {
-         RegistryKey<World> registrykey = p_196262_2_.func_234923_W_() == World.field_234920_i_ ? World.field_234918_g_ : World.field_234920_i_;
+         //LoliServer start
+         RegistryKey<World> originRegistrykey = p_196262_2_.func_234923_W_() == World.field_234920_i_ ? World.field_234918_g_ : World.field_234920_i_;
+         RegistryKey<World> originRegistrykey1 = p_196262_2_.dimension1() == World.field_234920_i_ ? World.field_234918_g_ : World.field_234920_i_;
+         //LoliServer end
+         //CraftBukkit start
+         RegistryKey<World> registrykey = p_196262_2_.func_230315_m_() == MinecraftServer.getServer().field_240767_f_.func_230520_a_().func_230516_a_(DimensionType.field_236001_e_) ? World.field_234918_g_ : World.field_234920_i_; // CraftBukkit - SPIGOT-6152: send back to main overworld in custom ends
+         //CraftBukkit end
+         //LoliServer start
+         if (originRegistrykey != originRegistrykey1) {
+            registrykey = originRegistrykey;
+         }
+         //LoliServer end
          ServerWorld serverworld = ((ServerWorld)p_196262_2_).func_73046_m().func_71218_a(registrykey);
          if (serverworld == null) {
-            return;
-         }
+            //return; // CraftBukkit - always fire event in case plugins wish to change it
+         }
+
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_196262_4_.getBukkitEntity(), new org.bukkit.Location(p_196262_2_.getWorld(), p_196262_3_.func_177958_n(), p_196262_3_.func_177956_o(), p_196262_3_.func_177952_p()));
+         p_196262_2_.getCBServer().getPluginManager().callEvent(event);
+         // LoliServer start - don't teleport when serverworld == null
+         if (serverworld == null) {
+            return;
+         }
+         // LoliServer end
+         if (p_196262_4_ instanceof ServerPlayerEntity) {
+            ((ServerPlayerEntity) p_196262_4_).changeDimension(serverworld, serverworld.func_85176_s(), PlayerTeleportEvent.TeleportCause.END_PORTAL);
+            return;
+         }
+         // CraftBukkit end
 
          p_196262_4_.func_241206_a_(serverworld);
       }
