--- a/net/minecraft/block/FarmlandBlock.java
+++ b/net/minecraft/block/FarmlandBlock.java
@@ -2,8 +_,6 @@
 
 import java.util.Random;
 import net.minecraft.entity.Entity;
-import net.minecraft.entity.LivingEntity;
-import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.BlockItemUseContext;
 import net.minecraft.pathfinding.PathType;
 import net.minecraft.state.IntegerProperty;
@@ -14,12 +_,12 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class FarmlandBlock extends Block {
    public static final IntegerProperty field_176531_a = BlockStateProperties.field_208133_ah;
@@ -66,31 +_,35 @@
       int i = p_225542_1_.func_177229_b(field_176531_a);
       if (!func_176530_e(p_225542_2_, p_225542_3_) && !p_225542_2_.func_175727_C(p_225542_3_.func_177984_a())) {
          if (i > 0) {
-            p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_.func_206870_a(field_176531_a, Integer.valueOf(i - 1)), 2);
+            CraftEventFactory.handleMoistureChangeEvent(p_225542_2_, p_225542_3_, (BlockState) p_225542_1_.func_206870_a(FarmlandBlock.field_176531_a, i - 1), 2); // CraftBukkit
          } else if (!func_176529_d(p_225542_2_, p_225542_3_)) {
             func_199610_d(p_225542_1_, p_225542_2_, p_225542_3_);
          }
       } else if (i < 7) {
-         p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_.func_206870_a(field_176531_a, Integer.valueOf(7)), 2);
+         CraftEventFactory.handleMoistureChangeEvent(p_225542_2_, p_225542_3_, (BlockState) p_225542_1_.func_206870_a(FarmlandBlock.field_176531_a, 7), 2); // CraftBukkit
       }
 
    }
 
    public void func_180658_a(World p_180658_1_, BlockPos p_180658_2_, Entity p_180658_3_, float p_180658_4_) {
-      if (!p_180658_1_.field_72995_K && p_180658_1_.field_73012_v.nextFloat() < p_180658_4_ - 0.5F && p_180658_3_ instanceof LivingEntity && (p_180658_3_ instanceof PlayerEntity || p_180658_1_.func_82736_K().func_223586_b(GameRules.field_223599_b)) && p_180658_3_.func_213311_cf() * p_180658_3_.func_213311_cf() * p_180658_3_.func_213302_cg() > 0.512F) {
+      if (!p_180658_1_.field_72995_K && net.minecraftforge.common.ForgeHooks.onFarmlandTrample(p_180658_1_, p_180658_2_, Blocks.field_150346_d.func_176223_P(), p_180658_4_, p_180658_3_)) { // Forge: Move logic to Entity#canTrample
          func_199610_d(p_180658_1_.func_180495_p(p_180658_2_), p_180658_1_, p_180658_2_);
       }
-
-      super.func_180658_a(p_180658_1_, p_180658_2_, p_180658_3_, p_180658_4_);
    }
 
    public static void func_199610_d(BlockState p_199610_0_, World p_199610_1_, BlockPos p_199610_2_) {
+      // CraftBukkit start
+      if (CraftEventFactory.callBlockFadeEvent(p_199610_1_, p_199610_2_, Blocks.field_150346_d.func_176223_P()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_199610_1_.func_175656_a(p_199610_2_, func_199601_a(p_199610_0_, Blocks.field_150346_d.func_176223_P(), p_199610_1_, p_199610_2_));
    }
 
-   private static boolean func_176529_d(IBlockReader p_176529_0_, BlockPos p_176529_1_) {
-      Block block = p_176529_0_.func_180495_p(p_176529_1_.func_177984_a()).func_177230_c();
-      return block instanceof CropsBlock || block instanceof StemBlock || block instanceof AttachedStemBlock;
+   private boolean func_176529_d(IBlockReader p_176529_0_, BlockPos p_176529_1_) {
+      BlockState plant = p_176529_0_.func_180495_p(p_176529_1_.func_177984_a());
+      BlockState state = p_176529_0_.func_180495_p(p_176529_1_);
+      return plant.func_177230_c() instanceof net.minecraftforge.common.IPlantable && state.canSustainPlant(p_176529_0_, p_176529_1_, Direction.UP, (net.minecraftforge.common.IPlantable)plant.func_177230_c());
    }
 
    private static boolean func_176530_e(IWorldReader p_176530_0_, BlockPos p_176530_1_) {
@@ -100,7 +_,7 @@
          }
       }
 
-      return false;
+      return net.minecraftforge.common.FarmlandWaterManager.hasBlockWaterTicket(p_176530_0_, p_176530_1_);
    }
 
    protected void func_206840_a(StateContainer.Builder<Block, BlockState> p_206840_1_) {
