--- a/net/minecraft/block/FenceGateBlock.java
+++ b/net/minecraft/block/FenceGateBlock.java
@@ -18,6 +_,8 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class FenceGateBlock extends HorizontalBlock {
    public static final BooleanProperty field_176466_a = BlockStateProperties.field_208193_t;
@@ -121,6 +_,19 @@
    public void func_220069_a(BlockState p_220069_1_, World p_220069_2_, BlockPos p_220069_3_, Block p_220069_4_, BlockPos p_220069_5_, boolean p_220069_6_) {
       if (!p_220069_2_.field_72995_K) {
          boolean flag = p_220069_2_.func_175640_z(p_220069_3_);
+
+         // CraftBukkit start
+         boolean oldPowered = p_220069_1_.func_177229_b(FenceGateBlock.field_176465_b);
+         if (oldPowered != flag) {
+            int newPower = flag ? 15 : 0;
+            int oldPower = oldPowered ? 15 : 0;
+            org.bukkit.block.Block bukkitBlock = CraftBlock.at(p_220069_2_, p_220069_3_);
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, oldPower, newPower);
+            p_220069_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+            flag = eventRedstone.getNewCurrent() > 0;
+         }
+         // CraftBukkit end
+
          if (p_220069_1_.func_177229_b(field_176465_b) != flag) {
             p_220069_2_.func_180501_a(p_220069_3_, p_220069_1_.func_206870_a(field_176465_b, Boolean.valueOf(flag)).func_206870_a(field_176466_a, Boolean.valueOf(flag)), 2);
             if (p_220069_1_.func_177229_b(field_176466_a) != flag) {
