--- a/net/minecraft/block/FungusBlock.java
+++ b/net/minecraft/block/FungusBlock.java
@@ -40,6 +_,13 @@
    }
 
    public void func_225535_a_(ServerWorld p_225535_1_, Random p_225535_2_, BlockPos p_225535_3_, BlockState p_225535_4_) {
+      // CraftBukkit start
+      if (this == Blocks.field_235373_mm_) { // WARPED_FUNGUS
+         SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+      } else if (this == Blocks.field_235382_mv_) { // CRIMSON_FUNGUS
+         SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+      }
+      // CraftBukkit end
       this.field_235497_b_.get().func_242765_a(p_225535_1_, p_225535_1_.func_72863_F().func_201711_g(), p_225535_2_, p_225535_3_);
    }
 }
