--- a/net/minecraft/block/GrassBlock.java
+++ b/net/minecraft/block/GrassBlock.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.gen.feature.ConfiguredFeature;
 import net.minecraft.world.gen.feature.FlowersFeature;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class GrassBlock extends SpreadableSnowyDirtBlock implements IGrowable {
    public GrassBlock(AbstractBlock.Properties p_i48388_1_) {
@@ -58,7 +_,7 @@
             }
 
             if (blockstate1.func_196955_c(p_225535_1_, blockpos1)) {
-               p_225535_1_.func_180501_a(blockpos1, blockstate1, 3);
+               CraftEventFactory.handleBlockGrowEvent(p_225535_1_, blockpos1, blockstate1, 3); // CraftBukkit
             }
          }
       }
