--- a/net/minecraft/block/GrassPathBlock.java
+++ b/net/minecraft/block/GrassPathBlock.java
@@ -36,6 +_,11 @@
    }
 
    public void func_225534_a_(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
+      // CraftBukkit start - do not fade if the block is valid here
+      if (p_225534_1_.func_196955_c(p_225534_2_, p_225534_3_)) {
+         return;
+      }
+      // CraftBukkit end
       FarmlandBlock.func_199610_d(p_225534_1_, p_225534_2_, p_225534_3_);
    }
 
