--- a/net/minecraft/block/IceBlock.java
+++ b/net/minecraft/block/IceBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.LightType;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class IceBlock extends BreakableBlock {
    public IceBlock(AbstractBlock.Properties p_i48375_1_) {
@@ -43,6 +_,11 @@
    }
 
    protected void func_196454_d(BlockState p_196454_1_, World p_196454_2_, BlockPos p_196454_3_) {
+      // CraftBukkit start
+      if (CraftEventFactory.callBlockFadeEvent(p_196454_2_, p_196454_3_, p_196454_2_.func_230315_m_().func_236040_e_() ? Blocks.field_150350_a.func_176223_P() : Blocks.field_150355_j.func_176223_P()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       if (p_196454_2_.func_230315_m_().func_236040_e_()) {
          p_196454_2_.func_217377_a(p_196454_3_, false);
       } else {
