--- a/net/minecraft/block/JukeboxBlock.java
+++ b/net/minecraft/block/JukeboxBlock.java
@@ -55,12 +_,18 @@
    public void func_176431_a(IWorld p_176431_1_, BlockPos p_176431_2_, BlockState p_176431_3_, ItemStack p_176431_4_) {
       TileEntity tileentity = p_176431_1_.func_175625_s(p_176431_2_);
       if (tileentity instanceof JukeboxTileEntity) {
-         ((JukeboxTileEntity)tileentity).func_195535_a(p_176431_4_.func_77946_l());
+         // CraftBukkit start - There can only be one
+         p_176431_4_ = p_176431_4_.func_77946_l();
+         if (!p_176431_4_.func_190926_b()) {
+            p_176431_4_.func_190920_e(1);
+         }
+         ((JukeboxTileEntity) tileentity).func_195535_a(p_176431_4_);
+         // CraftBukkit end
          p_176431_1_.func_180501_a(p_176431_2_, p_176431_3_.func_206870_a(field_176432_a, Boolean.valueOf(true)), 2);
       }
    }
 
-   private void func_203419_a(World p_203419_1_, BlockPos p_203419_2_) {
+   public void func_203419_a(World p_203419_1_, BlockPos p_203419_2_) { // private->public CraftBukkit
       if (!p_203419_1_.field_72995_K) {
          TileEntity tileentity = p_203419_1_.func_175625_s(p_203419_2_);
          if (tileentity instanceof JukeboxTileEntity) {
