--- a/net/minecraft/block/LecternBlock.java
+++ b/net/minecraft/block/LecternBlock.java
@@ -178,11 +_,12 @@
    }
 
    private void func_220150_d(BlockState p_220150_1_, World p_220150_2_, BlockPos p_220150_3_) {
-      TileEntity tileentity = p_220150_2_.func_175625_s(p_220150_3_);
+      TileEntity tileentity = p_220150_2_.getTileEntity(p_220150_3_, false); // CraftBukkit - don't validate, type may be changed already
       if (tileentity instanceof LecternTileEntity) {
          LecternTileEntity lecterntileentity = (LecternTileEntity)tileentity;
          Direction direction = p_220150_1_.func_177229_b(field_220156_a);
          ItemStack itemstack = lecterntileentity.func_214033_c().func_77946_l();
+         if(itemstack.func_190926_b()) return;  // CraftBukkit - SPIGOT-5500
          float f = 0.25F * (float)direction.func_82601_c();
          float f1 = 0.25F * (float)direction.func_82599_e();
          ItemEntity itementity = new ItemEntity(p_220150_2_, (double)p_220150_3_.func_177958_n() + 0.5D + (double)f, (double)(p_220150_3_.func_177956_o() + 1), (double)p_220150_3_.func_177952_p() + 0.5D + (double)f1, itemstack);
