--- a/net/minecraft/block/LeverBlock.java
+++ b/net/minecraft/block/LeverBlock.java
@@ -21,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class LeverBlock extends HorizontalFaceBlock {
    public static final BooleanProperty field_176359_b = BlockStateProperties.field_208194_u;
@@ -81,6 +_,18 @@
 
          return ActionResultType.SUCCESS;
       } else {
+         // CraftBukkit start - Interact Lever
+         boolean powered = p_225533_1_.func_177229_b(LeverBlock.field_176359_b); // Old powered state
+         org.bukkit.block.Block block = p_225533_2_.getWorld().getBlockAt(p_225533_3_.func_177958_n(), p_225533_3_.func_177956_o(), p_225533_3_.func_177952_p());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_225533_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return ActionResultType.SUCCESS;
+         }
+         // CraftBukkit end
+
          BlockState blockstate = this.func_226939_d_(p_225533_1_, p_225533_2_, p_225533_3_);
          float f = blockstate.func_177229_b(field_176359_b) ? 0.6F : 0.5F;
          p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_187750_dc, SoundCategory.BLOCKS, 0.3F, f);
