--- a/net/minecraft/block/LilyPadBlock.java
+++ b/net/minecraft/block/LilyPadBlock.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class LilyPadBlock extends BushBlock {
    protected static final VoxelShape field_185523_a = Block.func_208617_a(1.0D, 0.0D, 1.0D, 15.0D, 1.5D, 15.0D);
@@ -21,7 +_,7 @@
 
    public void func_196262_a(BlockState p_196262_1_, World p_196262_2_, BlockPos p_196262_3_, Entity p_196262_4_) {
       super.func_196262_a(p_196262_1_, p_196262_2_, p_196262_3_, p_196262_4_);
-      if (p_196262_2_ instanceof ServerWorld && p_196262_4_ instanceof BoatEntity) {
+      if (p_196262_2_ instanceof ServerWorld && p_196262_4_ instanceof BoatEntity && !CraftEventFactory.callEntityChangeBlockEvent(p_196262_4_, p_196262_3_, Blocks.field_150350_a.func_176223_P()).isCancelled()) { // CraftBukkit
          p_196262_2_.func_225521_a_(new BlockPos(p_196262_3_), true, p_196262_4_);
       }
 
