--- a/net/minecraft/block/MagmaBlock.java
+++ b/net/minecraft/block/MagmaBlock.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class MagmaBlock extends Block {
    public MagmaBlock(AbstractBlock.Properties p_i48366_1_) {
@@ -23,7 +_,9 @@
 
    public void func_176199_a(World p_176199_1_, BlockPos p_176199_2_, Entity p_176199_3_) {
       if (!p_176199_3_.func_230279_az_() && p_176199_3_ instanceof LivingEntity && !EnchantmentHelper.func_189869_j((LivingEntity)p_176199_3_)) {
+         CraftEventFactory.blockDamage = p_176199_1_.getWorld().getBlockAt(p_176199_2_.func_177958_n(), p_176199_2_.func_177956_o(), p_176199_2_.func_177952_p()); // CraftBukkit
          p_176199_3_.func_70097_a(DamageSource.field_190095_e, 1.0F);
+         CraftEventFactory.blockDamage = null; // CraftBukkit
       }
 
       super.func_176199_a(p_176199_1_, p_176199_2_, p_176199_3_);
