--- a/net/minecraft/block/MushroomBlock.java
+++ b/net/minecraft/block/MushroomBlock.java
@@ -11,6 +_,8 @@
 import net.minecraft.world.gen.feature.ConfiguredFeature;
 import net.minecraft.world.gen.feature.Features;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.TreeType;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class MushroomBlock extends BushBlock implements IGrowable {
    protected static final VoxelShape field_196385_a = Block.func_208617_a(5.0D, 0.0D, 5.0D, 11.0D, 6.0D, 11.0D);
@@ -24,7 +_,7 @@
    }
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_4_.nextInt(25) == 0) {
+      if (p_225542_4_.nextInt(Math.max(1, (int) (100.0F / p_225542_2_.spigotConfig.mushroomModifier) * 25)) == 0) { // Spigot
          int i = 5;
          int j = 4;
 
@@ -48,7 +_,7 @@
          }
 
          if (p_225542_2_.func_175623_d(blockpos1) && p_225542_1_.func_196955_c(p_225542_2_, blockpos1)) {
-            p_225542_2_.func_180501_a(blockpos1, p_225542_1_, 2);
+            CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, p_225542_3_, blockpos1, p_225542_1_, 2); // CraftBukkit
          }
       }
 
@@ -64,7 +_,7 @@
       if (blockstate.func_235714_a_(BlockTags.field_242171_aD)) {
          return true;
       } else {
-         return p_196260_2_.func_226659_b_(p_196260_3_, 0) < 13 && this.func_200014_a_(blockstate, p_196260_2_, blockpos);
+         return p_196260_2_.func_226659_b_(p_196260_3_, 0) < 13 && blockstate.canSustainPlant(p_196260_2_, blockpos, net.minecraft.util.Direction.UP, this);
       }
    }
 
@@ -72,6 +_,7 @@
       p_226940_1_.func_217377_a(p_226940_2_, false);
       ConfiguredFeature<?, ?> configuredfeature;
       if (this == Blocks.field_150338_P) {
+         SaplingBlock.treeType = TreeType.BROWN_MUSHROOM; // CraftBukkit
          configuredfeature = Features.field_243860_bF;
       } else {
          if (this != Blocks.field_150337_Q) {
@@ -79,6 +_,7 @@
             return false;
          }
 
+         SaplingBlock.treeType = TreeType.RED_MUSHROOM; // CraftBukkit
          configuredfeature = Features.field_243861_bG;
       }
 
