--- a/net/minecraft/block/NetherPortalBlock.java
+++ b/net/minecraft/block/NetherPortalBlock.java
@@ -26,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
 
 public class NetherPortalBlock extends Block {
    public static final EnumProperty<Direction.Axis> field_176550_a = BlockStateProperties.field_208199_z;
@@ -48,13 +_,14 @@
    }
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_2_.func_230315_m_().func_236043_f_() && p_225542_2_.func_82736_K().func_223586_b(GameRules.field_223601_d) && p_225542_4_.nextInt(2000) < p_225542_2_.func_175659_aa().func_151525_a()) {
+      if (p_225542_2_.spigotConfig.enableZombiePigmenPortalSpawns && p_225542_2_.func_230315_m_().func_236043_f_() && p_225542_2_.func_82736_K().func_223586_b(GameRules.field_223601_d) && p_225542_4_.nextInt(2000) < p_225542_2_.func_175659_aa().func_151525_a()) { // Spigot
          while(p_225542_2_.func_180495_p(p_225542_3_).func_203425_a(this)) {
             p_225542_3_ = p_225542_3_.func_177977_b();
          }
 
          if (p_225542_2_.func_180495_p(p_225542_3_).func_215688_a(p_225542_2_, p_225542_3_, EntityType.field_233592_ba_)) {
-            Entity entity = EntityType.field_233592_ba_.func_220342_a(p_225542_2_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, p_225542_3_.func_177984_a(), SpawnReason.STRUCTURE, false, false);
+            // CraftBukkit - set spawn reason to NETHER_PORTAL
+            Entity entity = EntityType.field_233592_ba_.spawn(p_225542_2_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, p_225542_3_.func_177984_a(), SpawnReason.STRUCTURE, false, false, CreatureSpawnEvent.SpawnReason.NETHER_PORTAL);
             if (entity != null) {
                entity.func_242279_ag();
             }
@@ -72,6 +_,10 @@
 
    public void func_196262_a(BlockState p_196262_1_, World p_196262_2_, BlockPos p_196262_3_, Entity p_196262_4_) {
       if (!p_196262_4_.func_184218_aH() && !p_196262_4_.func_184207_aI() && p_196262_4_.func_184222_aU()) {
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_196262_4_.getBukkitEntity(), new org.bukkit.Location(p_196262_2_.getWorld(), p_196262_3_.func_177958_n(), p_196262_3_.func_177956_o(), p_196262_3_.func_177952_p()));
+         p_196262_2_.getCBServer().getPluginManager().callEvent(event);
+         // CraftBukkit end
          p_196262_4_.func_181015_d(p_196262_3_);
       }
 
