--- a/net/minecraft/block/ObserverBlock.java
+++ b/net/minecraft/block/ObserverBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ObserverBlock extends DirectionalBlock {
    public static final BooleanProperty field_190963_a = BlockStateProperties.field_208194_u;
@@ -36,8 +_,18 @@
 
    public void func_225534_a_(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
       if (p_225534_1_.func_177229_b(field_190963_a)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_225534_2_.func_180501_a(p_225534_3_, p_225534_1_.func_206870_a(field_190963_a, Boolean.valueOf(false)), 2);
       } else {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 0, 15).getNewCurrent() != 15) {
+            return;
+         }
+         // CraftBukkit end
          p_225534_2_.func_180501_a(p_225534_3_, p_225534_1_.func_206870_a(field_190963_a, Boolean.valueOf(true)), 2);
          p_225534_2_.func_205220_G_().func_205360_a(p_225534_3_, this, 2);
       }
