--- a/net/minecraft/block/PortalInfo.java
+++ b/net/minecraft/block/PortalInfo.java
@@ -1,12 +_,26 @@
 package net.minecraft.block;
 
 import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftPortalEvent;
 
 public class PortalInfo {
    public final Vector3d field_222505_a;
    public final Vector3d field_222506_b;
    public final float field_242960_c;
    public final float field_242961_d;
+   // CraftBukkit start
+   public ServerWorld world;
+   public CraftPortalEvent portalEventInfo;
+   // CraftBukkit end
+
+   // CatServer start
+   public PortalInfo putCB(ServerWorld world, CraftPortalEvent portalEventInfo) {
+      this.world = world;
+      this.portalEventInfo = portalEventInfo;
+      return this;
+   }
+   // CatServer end
 
    public PortalInfo(Vector3d p_i242042_1_, Vector3d p_i242042_2_, float p_i242042_3_, float p_i242042_4_) {
       this.field_222505_a = p_i242042_1_;
