--- a/net/minecraft/block/PressurePlateBlock.java
+++ b/net/minecraft/block/PressurePlateBlock.java
@@ -14,6 +_,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class PressurePlateBlock extends AbstractPressurePlateBlock {
    public static final BooleanProperty field_176580_a = BlockStateProperties.field_208194_u;
@@ -67,6 +_,25 @@
 
       if (!list.isEmpty()) {
          for(Entity entity : list) {
+
+            // CraftBukkit start - Call interact event when turning on a pressure plate
+            if (this.func_176576_e(p_180669_1_.func_180495_p(p_180669_2_)) == 0) {
+               org.bukkit.World bworld = p_180669_1_.getWorld();
+               org.bukkit.plugin.PluginManager manager = p_180669_1_.getCBServer().getPluginManager();
+               org.bukkit.event.Cancellable cancellable;
+               if (entity instanceof PlayerEntity) {
+                  cancellable = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) entity, org.bukkit.event.block.Action.PHYSICAL, p_180669_2_, null, null, null);
+               } else {
+                  cancellable = new EntityInteractEvent(entity.getBukkitEntity(), bworld.getBlockAt(p_180669_2_.func_177958_n(), p_180669_2_.func_177956_o(), p_180669_2_.func_177952_p()));
+                  manager.callEvent((EntityInteractEvent) cancellable);
+               }
+               // We only want to block turning the plate on if all events are cancelled
+               if (cancellable.isCancelled()) {
+                  continue;
+               }
+            }
+            // CraftBukkit end
+
             if (!entity.func_145773_az()) {
                return 15;
             }
