--- a/net/minecraft/block/RedstoneOreBlock.java
+++ b/net/minecraft/block/RedstoneOreBlock.java
@@ -1,8 +_,7 @@
 package net.minecraft.block;
 
 import java.util.Random;
-import net.minecraft.enchantment.EnchantmentHelper;
-import net.minecraft.enchantment.Enchantments;
+
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.BlockItem;
@@ -20,6 +_,9 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
 
 public class RedstoneOreBlock extends Block {
    public static final BooleanProperty field_196501_a = RedstoneTorchBlock.field_196528_a;
@@ -30,30 +_,50 @@
    }
 
    public void func_196270_a(BlockState p_196270_1_, World p_196270_2_, BlockPos p_196270_3_, PlayerEntity p_196270_4_) {
-      func_196500_d(p_196270_1_, p_196270_2_, p_196270_3_);
+      interact(p_196270_1_, p_196270_2_, p_196270_3_, p_196270_4_);  // CraftBukkit - add entityhuman
       super.func_196270_a(p_196270_1_, p_196270_2_, p_196270_3_, p_196270_4_);
    }
 
    public void func_176199_a(World p_176199_1_, BlockPos p_176199_2_, Entity p_176199_3_) {
-      func_196500_d(p_176199_1_.func_180495_p(p_176199_2_), p_176199_1_, p_176199_2_);
-      super.func_176199_a(p_176199_1_, p_176199_2_, p_176199_3_);
+
+      // CraftBukkit start
+      if (p_176199_3_ instanceof PlayerEntity) {
+         PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) p_176199_3_, org.bukkit.event.block.Action.PHYSICAL, p_176199_2_, null, null, null);
+         if (!event.isCancelled()) {
+            interact(p_176199_1_.func_180495_p(p_176199_2_), p_176199_1_, p_176199_2_, p_176199_3_); // add entity
+            super.func_176199_a(p_176199_1_, p_176199_2_, p_176199_3_);
+         }
+      } else {
+         EntityInteractEvent event = new EntityInteractEvent(p_176199_3_.getBukkitEntity(), p_176199_1_.getWorld().getBlockAt(p_176199_2_.func_177958_n(), p_176199_2_.func_177956_o(), p_176199_2_.func_177952_p()));
+         p_176199_1_.getCBServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            interact(p_176199_1_.func_180495_p(p_176199_2_), p_176199_1_, p_176199_2_, p_176199_3_); // add entity
+            super.func_176199_a(p_176199_1_, p_176199_2_, p_176199_3_);
+         }
+      }
+      // CraftBukkit end
    }
 
    public ActionResultType func_225533_a_(BlockState p_225533_1_, World p_225533_2_, BlockPos p_225533_3_, PlayerEntity p_225533_4_, Hand p_225533_5_, BlockRayTraceResult p_225533_6_) {
       if (p_225533_2_.field_72995_K) {
          func_180691_e(p_225533_2_, p_225533_3_);
       } else {
-         func_196500_d(p_225533_1_, p_225533_2_, p_225533_3_);
+         interact(p_225533_1_, p_225533_2_, p_225533_3_, p_225533_4_); // CraftBukkit - add entityhuman
       }
 
       ItemStack itemstack = p_225533_4_.func_184586_b(p_225533_5_);
       return itemstack.func_77973_b() instanceof BlockItem && (new BlockItemUseContext(p_225533_4_, p_225533_5_, itemstack, p_225533_6_)).func_196011_b() ? ActionResultType.PASS : ActionResultType.SUCCESS;
    }
 
-   private static void func_196500_d(BlockState p_196500_0_, World p_196500_1_, BlockPos p_196500_2_) {
-      func_180691_e(p_196500_1_, p_196500_2_);
-      if (!p_196500_0_.func_177229_b(field_196501_a)) {
-         p_196500_1_.func_180501_a(p_196500_2_, p_196500_0_.func_206870_a(field_196501_a, Boolean.valueOf(true)), 3);
+   private static void interact(BlockState state, World world, BlockPos pos, Entity entity) { // CraftBukkit - add Entity
+      func_180691_e(world, pos);
+      if (!state.func_177229_b(field_196501_a)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.func_206870_a(field_196501_a, true)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
+         world.func_180501_a(pos, state.func_206870_a(field_196501_a, Boolean.valueOf(true)), 3);
       }
 
    }
@@ -64,6 +_,11 @@
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (p_225542_1_.func_177229_b(field_196501_a)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_225542_2_, p_225542_3_, p_225542_1_.func_206870_a(field_196501_a, true)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_.func_206870_a(field_196501_a, Boolean.valueOf(false)), 3);
       }
 
@@ -71,11 +_,11 @@
 
    public void func_220062_a(BlockState p_220062_1_, ServerWorld p_220062_2_, BlockPos p_220062_3_, ItemStack p_220062_4_) {
       super.func_220062_a(p_220062_1_, p_220062_2_, p_220062_3_, p_220062_4_);
-      if (EnchantmentHelper.func_77506_a(Enchantments.field_185306_r, p_220062_4_) == 0) {
-         int i = 1 + p_220062_2_.field_73012_v.nextInt(5);
-         this.func_180637_b(p_220062_2_, p_220062_3_, i);
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.IWorldReader world, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? 1 + RANDOM.nextInt(5) : 0;
    }
 
    @OnlyIn(Dist.CLIENT)
