--- a/net/minecraft/block/RedstoneDiodeBlock.java
+++ b/net/minecraft/block/RedstoneDiodeBlock.java
@@ -16,6 +_,7 @@
 import net.minecraft.world.TickPriority;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public abstract class RedstoneDiodeBlock extends HorizontalBlock {
    protected static final VoxelShape field_196347_b = Block.func_208617_a(0.0D, 0.0D, 0.0D, 16.0D, 2.0D, 16.0D);
@@ -38,8 +_,18 @@
          boolean flag = p_225534_1_.func_177229_b(field_196348_c);
          boolean flag1 = this.func_176404_e(p_225534_2_, p_225534_3_, p_225534_1_);
          if (flag && !flag1) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 15, 0).getNewCurrent() != 0) {
+               return;
+            }
+            // CraftBukkit end
             p_225534_2_.func_180501_a(p_225534_3_, p_225534_1_.func_206870_a(field_196348_c, Boolean.valueOf(false)), 2);
          } else if (!flag) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 0, 15).getNewCurrent() != 15) {
+               return;
+            }
+            // CraftBukkit end
             p_225534_2_.func_180501_a(p_225534_3_, p_225534_1_.func_206870_a(field_196348_c, Boolean.valueOf(true)), 2);
             if (!flag1) {
                p_225534_2_.func_205220_G_().func_205362_a(p_225534_3_, this, this.func_196346_i(p_225534_1_), TickPriority.VERY_HIGH);
@@ -65,7 +_,7 @@
       if (p_220069_1_.func_196955_c(p_220069_2_, p_220069_3_)) {
          this.func_176398_g(p_220069_2_, p_220069_3_, p_220069_1_);
       } else {
-         TileEntity tileentity = this.func_235695_q_() ? p_220069_2_.func_175625_s(p_220069_3_) : null;
+         TileEntity tileentity = p_220069_1_.hasTileEntity() ? p_220069_2_.func_175625_s(p_220069_3_) : null;
          func_220059_a(p_220069_1_, p_220069_2_, p_220069_3_, tileentity);
          p_220069_2_.func_217377_a(p_220069_3_, false);
 
@@ -163,6 +_,8 @@
    protected void func_176400_h(World p_176400_1_, BlockPos p_176400_2_, BlockState p_176400_3_) {
       Direction direction = p_176400_3_.func_177229_b(field_185512_D);
       BlockPos blockpos = p_176400_2_.func_177972_a(direction.func_176734_d());
+      if (net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(p_176400_1_, p_176400_2_, p_176400_1_.func_180495_p(p_176400_2_), java.util.EnumSet.of(direction.func_176734_d()), false).isCanceled())
+         return;
       p_176400_1_.func_190524_a(blockpos, this, p_176400_2_);
       p_176400_1_.func_175695_a(blockpos, this, direction);
    }
