--- a/net/minecraft/block/ScaffoldingBlock.java
+++ b/net/minecraft/block/ScaffoldingBlock.java
@@ -19,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ScaffoldingBlock extends Block implements IWaterLoggable {
    private static final VoxelShape field_220121_d;
@@ -83,7 +_,7 @@
    public void func_225534_a_(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
       int i = func_220117_a(p_225534_2_, p_225534_3_);
       BlockState blockstate = p_225534_1_.func_206870_a(field_220118_a, Integer.valueOf(i)).func_206870_a(field_220120_c, Boolean.valueOf(this.func_220116_a(p_225534_2_, p_225534_3_, i)));
-      if (blockstate.func_177229_b(field_220118_a) == 7) {
+      if (blockstate.func_177229_b(field_220118_a) == 7 && !CraftEventFactory.callBlockFadeEvent(p_225534_2_, p_225534_3_, Blocks.field_150350_a.func_176223_P()).isCancelled()) { // CraftBukkit - BlockFadeEvent
          if (p_225534_1_.func_177229_b(field_220118_a) == 7) {
             p_225534_2_.func_217376_c(new FallingBlockEntity(p_225534_2_, (double)p_225534_3_.func_177958_n() + 0.5D, (double)p_225534_3_.func_177956_o(), (double)p_225534_3_.func_177952_p() + 0.5D, blockstate.func_206870_a(field_220119_b, Boolean.valueOf(false))));
          } else {
