--- a/net/minecraft/block/ShulkerBoxBlock.java
+++ b/net/minecraft/block/ShulkerBoxBlock.java
@@ -50,7 +_,7 @@
    public static final EnumProperty<Direction> field_190957_a = DirectionalBlock.field_176387_N;
    public static final ResourceLocation field_220169_b = new ResourceLocation("contents");
    @Nullable
-   private final DyeColor field_190958_b;
+   public final DyeColor field_190958_b;
 
    public ShulkerBoxBlock(@Nullable DyeColor p_i48334_1_, AbstractBlock.Properties p_i48334_2_) {
       super(p_i48334_2_);
