--- a/net/minecraft/block/SilverfishBlock.java
+++ b/net/minecraft/block/SilverfishBlock.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class SilverfishBlock extends Block {
    private final Block field_196469_a;
@@ -34,7 +_,7 @@
    private void func_235505_a_(ServerWorld p_235505_1_, BlockPos p_235505_2_) {
       SilverfishEntity silverfishentity = EntityType.field_200740_af.func_200721_a(p_235505_1_);
       silverfishentity.func_70012_b((double)p_235505_2_.func_177958_n() + 0.5D, (double)p_235505_2_.func_177956_o(), (double)p_235505_2_.func_177952_p() + 0.5D, 0.0F, 0.0F);
-      p_235505_1_.func_217376_c(silverfishentity);
+      p_235505_1_.addEntity(silverfishentity, SpawnReason.SILVERFISH_BLOCK); // CraftBukkit - add SpawnReason
       silverfishentity.func_70656_aK();
    }
 
