--- a/net/minecraft/block/SnowBlock.java
+++ b/net/minecraft/block/SnowBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.LightType;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class SnowBlock extends Block {
    public static final IntegerProperty field_176315_a = BlockStateProperties.field_208129_ad;
@@ -79,6 +_,11 @@
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (p_225542_2_.func_226658_a_(LightType.BLOCK, p_225542_3_) > 11) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_225542_2_, p_225542_3_, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          func_220075_c(p_225542_1_, p_225542_2_, p_225542_3_);
          p_225542_2_.func_217377_a(p_225542_3_, false);
       }
