--- a/net/minecraft/block/TrapDoorBlock.java
+++ b/net/minecraft/block/TrapDoorBlock.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TrapDoorBlock extends HorizontalBlock implements IWaterLoggable {
    public static final BooleanProperty field_176283_b = BlockStateProperties.field_208193_t;
@@ -101,6 +_,17 @@
       if (!p_220069_2_.field_72995_K) {
          boolean flag = p_220069_2_.func_175640_z(p_220069_3_);
          if (flag != p_220069_1_.func_177229_b(field_196381_c)) {
+            // CraftBukkit start
+            org.bukkit.World bworld = p_220069_2_.getWorld();
+            org.bukkit.block.Block bblock = bworld.getBlockAt(p_220069_3_.func_177958_n(), p_220069_3_.func_177956_o(), p_220069_3_.func_177952_p());
+            int power = bblock.getBlockPower();
+            int oldPower = (Boolean) p_220069_1_.func_177229_b(field_176283_b) ? 15 : 0;
+            if (oldPower == 0 ^ power == 0 || p_220069_4_.func_176223_P().func_185897_m()) {
+               BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bblock, oldPower, power);
+               p_220069_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+               flag = eventRedstone.getNewCurrent() > 0;
+            }
+            // CraftBukkit end
             if (p_220069_1_.func_177229_b(field_176283_b) != flag) {
                p_220069_1_ = p_220069_1_.func_206870_a(field_176283_b, Boolean.valueOf(flag));
                this.func_185731_a((PlayerEntity)null, p_220069_2_, p_220069_3_, flag);
@@ -147,4 +_,17 @@
 
       return super.func_196271_a(p_196271_1_, p_196271_2_, p_196271_3_, p_196271_4_, p_196271_5_, p_196271_6_);
    }
+
+   //Forge Start
+   @Override
+   public boolean isLadder(BlockState state, net.minecraft.world.IWorldReader world, BlockPos pos, net.minecraft.entity.LivingEntity entity) {
+      if (state.func_177229_b(field_176283_b)) {
+         BlockPos downPos = pos.func_177977_b();
+         BlockState down = world.func_180495_p(downPos);
+         return down.func_177230_c().makesOpenTrapdoorAboveClimbable(down, world, downPos, state);
+      }
+      return false;
+   }
+   //Forge End
+
 }
