--- a/net/minecraft/block/TripWireBlock.java
+++ b/net/minecraft/block/TripWireBlock.java
@@ -20,6 +_,8 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class TripWireBlock extends Block {
    public static final BooleanProperty field_176293_a = BlockStateProperties.field_208194_u;
@@ -121,6 +_,36 @@
             }
          }
       }
+
+      // CraftBukkit start - Call interact even when triggering connected tripwire
+      if (flag != flag1 && flag1 && (Boolean) blockstate.func_177229_b(field_176294_M)) {
+         org.bukkit.World bworld = p_176288_1_.getWorld();
+         org.bukkit.plugin.PluginManager manager = p_176288_1_.getCBServer().getPluginManager();
+         org.bukkit.block.Block block = bworld.getBlockAt(p_176288_2_.func_177958_n(), p_176288_2_.func_177956_o(), p_176288_2_.func_177952_p());
+         boolean allowed = false;
+         // If all of the events are cancelled block the tripwire trigger, else allow
+         for (Object object : list) {
+            if (object != null) {
+               org.bukkit.event.Cancellable cancellable;
+               if (object instanceof PlayerEntity) {
+                  cancellable = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) object, org.bukkit.event.block.Action.PHYSICAL, p_176288_2_, null, null, null);
+               } else if (object instanceof Entity) {
+                  cancellable = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+                  manager.callEvent((EntityInteractEvent) cancellable);
+               } else {
+                  continue;
+               }
+               if (!cancellable.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+         if (!allowed) {
+            return;
+         }
+      }
+      // CraftBukkit end
 
       if (flag1 != flag) {
          blockstate = blockstate.func_206870_a(field_176293_a, Boolean.valueOf(flag1));
