--- a/net/minecraft/block/TripWireHookBlock.java
+++ b/net/minecraft/block/TripWireHookBlock.java
@@ -24,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TripWireHookBlock extends Block {
    public static final DirectionProperty field_176264_a = HorizontalBlock.field_185512_D;
@@ -137,6 +_,15 @@
          this.func_180694_a(p_176260_1_, blockpos1, flag2, flag3, flag, flag1);
       }
 
+      // CraftBukkit start
+      org.bukkit.block.Block block = p_176260_1_.getWorld().getBlockAt(p_176260_2_.func_177958_n(), p_176260_2_.func_177956_o(), p_176260_2_.func_177952_p());
+      BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+      p_176260_1_.getCBServer().getPluginManager().callEvent(eventRedstone);
+      if (eventRedstone.getNewCurrent() > 0) {
+         return;
+      }
+      // CraftBukkit end
+
       this.func_180694_a(p_176260_1_, p_176260_2_, flag2, flag3, flag, flag1);
       if (!p_176260_4_) {
          p_176260_1_.func_180501_a(p_176260_2_, blockstate1.func_206870_a(field_176264_a, direction), 3);
@@ -150,8 +_,8 @@
             BlockPos blockpos2 = p_176260_2_.func_177967_a(direction, k);
             BlockState blockstate2 = ablockstate[k];
             if (blockstate2 != null) {
-               p_176260_1_.func_180501_a(blockpos2, blockstate2.func_206870_a(field_176265_M, Boolean.valueOf(flag2)), 3);
-               if (!p_176260_1_.func_180495_p(blockpos2).func_196958_f()) {
+               if (!p_176260_1_.func_180495_p(blockpos2).func_196958_f()) { // FORGE: fix MC-129055
+                  p_176260_1_.func_180501_a(blockpos2, blockstate2.func_206870_a(field_176265_M, Boolean.valueOf(flag2)), 3);
                }
             }
          }
