--- a/net/minecraft/block/TurtleEggBlock.java
+++ b/net/minecraft/block/TurtleEggBlock.java
@@ -21,10 +_,12 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class TurtleEggBlock extends Block {
    private static final VoxelShape field_203172_c = Block.func_208617_a(3.0D, 0.0D, 3.0D, 12.0D, 7.0D, 12.0D);
@@ -55,6 +_,18 @@
          if (!p_203167_1_.field_72995_K && p_203167_1_.field_73012_v.nextInt(p_203167_4_) == 0) {
             BlockState blockstate = p_203167_1_.func_180495_p(p_203167_2_);
             if (blockstate.func_203425_a(Blocks.field_203213_jA)) {
+               // CraftBukkit start - Step on eggs
+               org.bukkit.event.Cancellable cancellable;
+               if (p_203167_3_ instanceof PlayerEntity) {
+                  cancellable = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) p_203167_3_, org.bukkit.event.block.Action.PHYSICAL, p_203167_2_, null, null, null);
+               } else {
+                  cancellable = new EntityInteractEvent(p_203167_3_.getBukkitEntity(), CraftBlock.at(p_203167_1_, p_203167_2_));
+                  p_203167_1_.getCBServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+               }
+               if (cancellable.isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                this.func_203166_c(p_203167_1_, p_203167_2_, blockstate);
             }
          }
@@ -78,9 +_,19 @@
       if (this.func_203169_a(p_225542_2_) && func_203168_a(p_225542_2_, p_225542_3_)) {
          int i = p_225542_1_.func_177229_b(field_203170_a);
          if (i < 2) {
+            // CraftBukkit start - Call BlockGrowEvent
+            if (!CraftEventFactory.handleBlockGrowEvent(p_225542_2_, p_225542_3_, p_225542_1_.func_206870_a(field_203170_a, i + 1), 2)) {
+               return;
+            }
+            // CraftBukkit end
             p_225542_2_.func_184133_a((PlayerEntity)null, p_225542_3_, SoundEvents.field_203280_iy, SoundCategory.BLOCKS, 0.7F, 0.9F + p_225542_4_.nextFloat() * 0.2F);
-            p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_.func_206870_a(field_203170_a, Integer.valueOf(i + 1)), 2);
+			// worldIn.setBlockState(pos, state.with(HATCH, Integer.valueOf(i + 1)), 2); // CraftBukkit - handled above
          } else {
+            // CraftBukkit start - Call BlockFadeEvent
+            if (CraftEventFactory.callBlockFadeEvent(p_225542_2_, p_225542_3_, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             p_225542_2_.func_184133_a((PlayerEntity)null, p_225542_3_, SoundEvents.field_203279_ix, SoundCategory.BLOCKS, 0.7F, 0.9F + p_225542_4_.nextFloat() * 0.2F);
             p_225542_2_.func_217377_a(p_225542_3_, false);
 
@@ -90,7 +_,7 @@
                turtleentity.func_70873_a(-24000);
                turtleentity.func_203011_g(p_225542_3_);
                turtleentity.func_70012_b((double)p_225542_3_.func_177958_n() + 0.3D + (double)j * 0.2D, (double)p_225542_3_.func_177956_o(), (double)p_225542_3_.func_177952_p() + 0.3D, 0.0F, 0.0F);
-               p_225542_2_.func_217376_c(turtleentity);
+               p_225542_2_.addEntity(turtleentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // CraftBukkit
             }
          }
       }
@@ -149,7 +_,7 @@
          if (!(p_212570_2_ instanceof LivingEntity)) {
             return false;
          } else {
-            return p_212570_2_ instanceof PlayerEntity || p_212570_1_.func_82736_K().func_223586_b(GameRules.field_223599_b);
+            return p_212570_2_ instanceof PlayerEntity || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_212570_1_, p_212570_2_);
          }
       } else {
          return false;
