--- a/net/minecraft/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/block/WeightedPressurePlateBlock.java
@@ -11,6 +_,8 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class WeightedPressurePlateBlock extends AbstractPressurePlateBlock {
    public static final IntegerProperty field_176579_a = BlockStateProperties.field_208136_ak;
@@ -23,7 +_,25 @@
    }
 
    protected int func_180669_e(World p_180669_1_, BlockPos p_180669_2_) {
-      int i = Math.min(p_180669_1_.func_217357_a(Entity.class, field_185511_c.func_186670_a(p_180669_2_)).size(), this.field_150068_a);
+      // CraftBukkit start
+      int i = 0;
+      java.util.Iterator iterator = p_180669_1_.func_217357_a(Entity.class, WeightedPressurePlateBlock.field_185511_c.func_186670_a(p_180669_2_)).iterator();
+      while (iterator.hasNext()) {
+         Entity entity = (Entity) iterator.next();
+         org.bukkit.event.Cancellable cancellable;
+         if (entity instanceof PlayerEntity) {
+            cancellable = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) entity, org.bukkit.event.block.Action.PHYSICAL, p_180669_2_, null, null, null);
+         } else {
+            cancellable = new EntityInteractEvent(entity.getBukkitEntity(), p_180669_1_.getWorld().getBlockAt(p_180669_2_.func_177958_n(), p_180669_2_.func_177956_o(), p_180669_2_.func_177952_p()));
+            p_180669_1_.getCBServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+         }
+         // We only want to block turning the plate on if all events are cancelled
+         if (!cancellable.isCancelled()) {
+            i++;
+         }
+      }
+      i = Math.min(i, this.field_150068_a);
+      // CraftBukkit end
       if (i > 0) {
          float f = (float)Math.min(this.field_150068_a, i) / (float)this.field_150068_a;
          return MathHelper.func_76123_f(f * 15.0F);
