--- a/net/minecraft/block/WitherRoseBlock.java
+++ b/net/minecraft/block/WitherRoseBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class WitherRoseBlock extends FlowerBlock {
    public WitherRoseBlock(Effect p_i49968_1_, AbstractBlock.Properties p_i49968_2_) {
@@ -47,7 +_,7 @@
          if (p_196262_4_ instanceof LivingEntity) {
             LivingEntity livingentity = (LivingEntity)p_196262_4_;
             if (!livingentity.func_180431_b(DamageSource.field_82727_n)) {
-               livingentity.func_195064_c(new EffectInstance(Effects.field_82731_v, 40));
+               livingentity.addEffect(new EffectInstance(Effects.field_82731_v, 40), Cause.WITHER_ROSE); // CraftBukkit
             }
          }
 
