--- a/net/minecraft/block/WitherSkeletonSkullBlock.java
+++ b/net/minecraft/block/WitherSkeletonSkullBlock.java
@@ -21,6 +_,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class WitherSkeletonSkullBlock extends SkullBlock {
    @Nullable
@@ -42,6 +_,7 @@
    }
 
    public static void func_196298_a(World p_196298_0_, BlockPos p_196298_1_, SkullTileEntity p_196298_2_) {
+      if (p_196298_0_.captureBlockStates) return; // CraftBukkit
       if (!p_196298_0_.field_72995_K) {
          BlockState blockstate = p_196298_2_.func_195044_w();
          boolean flag = blockstate.func_203425_a(Blocks.field_196705_eO) || blockstate.func_203425_a(Blocks.field_196704_eN);
@@ -67,7 +_,7 @@
                   CriteriaTriggers.field_192133_m.func_192229_a(serverplayerentity, witherentity);
                }
 
-               p_196298_0_.func_217376_c(witherentity);
+               p_196298_0_.addEntity(witherentity, SpawnReason.BUILD_WITHER);
 
                for(int k = 0; k < blockpattern.func_177684_c(); ++k) {
                   for(int l = 0; l < blockpattern.func_177685_b(); ++l) {
