--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -22,6 +_,11 @@
 import org.apache.commons.lang3.ObjectUtils;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
+import net.minecraft.server.level.ServerPlayer;
+// CraftBukkit end
+
 public class SynchedEntityData {
    private static final Logger f_135342_ = LogUtils.getLogger();
    private static final Object2IntMap<Class<? extends Entity>> f_135343_ = new Object2IntOpenHashMap<>();
@@ -35,12 +_,30 @@
       this.f_135344_ = p_135351_;
    }
 
+   // CatServer start - [compat:plugin:ProtocolLib:5.4.0:com.comphenix.protocol.wrappers.LegacyDataWatcher#newHandle] a constructor from higher mc versions
+   SynchedEntityData(Entity p_135351_, SynchedEntityData.DataItem<?>[] adatawatcher_item) {
+      this.f_135344_ = p_135351_;
+      for (DataItem<?> dataItem : adatawatcher_item) {
+         f_135345_.put(dataItem.f_135390_.m_135015_(), dataItem);
+      }
+   }
+   // CatServer end
+
+   // CraftBukkit start - add method from above
+   public <T> void markDirty(EntityDataAccessor<T> datawatcherobject) {
+      this.m_135379_(datawatcherobject).m_135401_(true);
+      this.f_135348_ = true;
+   }
+   // CraftBukkit end
+
    public static <T> EntityDataAccessor<T> m_135353_(Class<? extends Entity> p_135354_, EntityDataSerializer<T> p_135355_) {
-      if (f_135342_.isDebugEnabled()) {
+      if (true || f_135342_.isDebugEnabled()) { // Forge: This is very useful for mods that register keys on classes that are not their own
          try {
             Class<?> oclass = Class.forName(Thread.currentThread().getStackTrace()[2].getClassName());
             if (!oclass.equals(p_135354_)) {
-               f_135342_.debug("defineId called for: {} from {}", p_135354_, oclass, new RuntimeException());
+               // Forge: log at warn, mods should not add to classes that they don't own, and only add stacktrace when in debug is enabled as it is mostly not needed and consumes time
+               if (f_135342_.isDebugEnabled()) f_135342_.warn("defineId called for: {} from {}", p_135354_, oclass, new RuntimeException());
+               else f_135342_.warn("defineId called for: {} from {}", p_135354_, oclass);
             }
          } catch (ClassNotFoundException classnotfoundexception) {
          }
@@ -209,6 +_,18 @@
    public boolean m_135388_() {
       return this.f_135345_.isEmpty();
    }
+
+   // CraftBukkit start - 1.20.1
+   public void refresh(ServerPlayer to) {
+      if (!this.m_135388_()) {
+         List<SynchedEntityData.DataValue<?>> list = this.m_252804_();
+
+         if (list != null) {
+            to.f_8906_.m_9829_(new ClientboundSetEntityDataPacket(this.f_135344_.m_19879_(), list));
+         }
+      }
+   }
+   // CraftBukkit end
 
    public static class DataItem<T> {
       final EntityDataAccessor<T> f_135390_;
