--- ../src-base/minecraft/net/minecraft/client/main/Main.java
+++ ../src-work/minecraft/net/minecraft/client/main/Main.java
@@ -17,6 +17,7 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.util.JsonUtils;
 import net.minecraft.util.Session;
+import net.minecraft.world.MinecraftException;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
@@ -111,7 +112,11 @@
         {
             public void run()
             {
-                Minecraft.stopIntegratedServer();
+                try {
+                    Minecraft.stopIntegratedServer();
+                } catch (MinecraftException exceptionWorldConflict) {
+                    exceptionWorldConflict.printStackTrace();
+                }
             }
         });
         Thread.currentThread().setName("Client thread");
