--- ../src-base/minecraft/net/minecraft/client/main/Main.java
+++ ../src-work/minecraft/net/minecraft/client/main/Main.java
@@ -15,6 +15,7 @@
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
 import net.minecraft.client.Minecraft;
+import net.minecraft.server.ExceptionWorldConflict;
 import net.minecraft.util.JsonUtils;
 import net.minecraft.util.Session;
 import net.minecraftforge.fml.relauncher.Side;
@@ -111,7 +112,11 @@
         {
             public void run()
             {
-                Minecraft.stopIntegratedServer();
+                try {
+                    Minecraft.stopIntegratedServer();
+                } catch (ExceptionWorldConflict exceptionWorldConflict) {
+                    exceptionWorldConflict.printStackTrace();
+                }
             }
         });
         Thread.currentThread().setName("Client thread");
