--- a/net/minecraft/nbt/ByteArrayTag.java
+++ b/net/minecraft/nbt/ByteArrayTag.java
@@ -13,6 +_,7 @@
       public ByteArrayTag m_7300_(DataInput p_128247_, int p_128248_, NbtAccounter p_128249_) throws IOException {
          p_128249_.m_6800_(192L);
          int i = p_128247_.readInt();
+         com.google.common.base.Preconditions.checkArgument( i < 1 << 24); // Spigot
          p_128249_.m_6800_(8L * (long)i);
          byte[] abyte = new byte[i];
          p_128247_.readFully(abyte);
