--- a/net/minecraft/nbt/IntArrayTag.java
+++ b/net/minecraft/nbt/IntArrayTag.java
@@ -13,6 +_,7 @@
       public IntArrayTag m_7300_(DataInput p_128662_, int p_128663_, NbtAccounter p_128664_) throws IOException {
          p_128664_.m_6800_(192L);
          int i = p_128662_.readInt();
+         com.google.common.base.Preconditions.checkArgument( i < 1 << 24); // Spigot
          p_128664_.m_6800_(32L * (long)i);
          int[] aint = new int[i];
 
