--- a/net/minecraft/nbt/JsonToNBT.java
+++ b/net/minecraft/nbt/JsonToNBT.java
@@ -77,7 +_,7 @@
       }
    }
 
-   private INBT func_193596_c(String p_193596_1_) {
+   public INBT func_193596_c(String p_193596_1_) {
       try {
          if (field_193617_c.matcher(p_193596_1_).matches()) {
             return FloatNBT.func_229689_a_(Float.parseFloat(p_193596_1_.substring(0, p_193596_1_.length() - 1)));
@@ -201,7 +_,7 @@
       }
    }
 
-   private INBT func_193606_k() throws CommandSyntaxException {
+   public INBT func_193606_k() throws CommandSyntaxException {
       this.func_193604_b('[');
       int i = this.field_197663_n.getCursor();
       char c0 = this.field_197663_n.read();
