--- a/net/minecraft/nbt/TagParser.java
+++ b/net/minecraft/nbt/TagParser.java
@@ -83,7 +_,7 @@
       }
    }
 
-   private Tag m_129368_(String p_129369_) {
+   public Tag m_129368_(String p_129369_) {
       try {
          if (f_129342_.matcher(p_129369_).matches()) {
             return FloatTag.m_128566_(Float.parseFloat(p_129369_.substring(0, p_129369_.length() - 1)));
@@ -207,7 +_,7 @@
       }
    }
 
-   private Tag m_129375_() throws CommandSyntaxException {
+   public Tag m_129375_() throws CommandSyntaxException {
       this.m_129352_('[');
       int i = this.f_129347_.getCursor();
       char c0 = this.f_129347_.read();
