--- a/net/minecraft/world/chunk/NibbleArray.java
+++ b/net/minecraft/world/chunk/NibbleArray.java
@@ -33,7 +_,7 @@
       return p_177483_2_ << 8 | p_177483_3_ << 4 | p_177483_1_;
    }
 
-   private int func_177480_a(int p_177480_1_) {
+   public int func_177480_a(int p_177480_1_) {
       if (this.field_76585_a == null) {
          return 0;
       } else {
@@ -42,7 +_,7 @@
       }
    }
 
-   private void func_177482_a(int p_177482_1_, int p_177482_2_) {
+   public void func_177482_a(int p_177482_1_, int p_177482_2_) {
       if (this.field_76585_a == null) {
          this.field_76585_a = new byte[2048];
       }
