--- a/net/minecraft/loot/functions/LootingEnchantBonus.java
+++ b/net/minecraft/loot/functions/LootingEnchantBonus.java
@@ -43,8 +_,14 @@
    public ItemStack func_215859_a(ItemStack p_215859_1_, LootContext p_215859_2_) {
       Entity entity = p_215859_2_.func_216031_c(LootParameters.field_216284_d);
       if (entity instanceof LivingEntity) {
-         int i = EnchantmentHelper.func_185283_h((LivingEntity)entity);
-         if (i == 0) {
+         int i = p_215859_2_.getLootingModifier();
+         // CraftBukkit start - use lootingModifier if set by plugin
+         if (p_215859_2_.func_216033_a(LootParameters.LOOTING_MOD)) {
+            i = p_215859_2_.func_216031_c(LootParameters.LOOTING_MOD);
+         }
+         // CraftBukkit end
+
+         if (i <= 0) { // CraftBukkit - account for possible negative looting values from Bukkit
             return p_215859_1_;
          }
 
