--- a/net/minecraft/world/storage/DerivedWorldInfo.java
+++ b/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -13,7 +_,15 @@
 public class DerivedWorldInfo implements IServerWorldInfo {
    private final IServerConfiguration field_237244_a_;
    private final IServerWorldInfo field_76115_a;
-
+   private String properName;
+
+   public DerivedWorldInfo(IServerConfiguration configuration, IServerWorldInfo delegate, String properName) {
+       this.field_237244_a_ = configuration;
+       this.field_76115_a = delegate;
+       this.properName = properName;
+   }
+
+   @Deprecated
    public DerivedWorldInfo(IServerConfiguration p_i232150_1_, IServerWorldInfo p_i232150_2_) {
       this.field_237244_a_ = p_i232150_1_;
       this.field_76115_a = p_i232150_2_;
@@ -44,7 +_,7 @@
    }
 
    public String func_76065_j() {
-      return this.field_237244_a_.func_76065_j();
+      return properName != null ? properName : this.field_237244_a_.func_76065_j();
    }
 
    public int func_230395_g_() {
