--- a/net/minecraft/world/storage/ServerWorldInfo.java
+++ b/net/minecraft/world/storage/ServerWorldInfo.java
@@ -13,12 +_,14 @@
 import net.minecraft.command.TimerCallbackManager;
 import net.minecraft.command.TimerCallbackSerializers;
 import net.minecraft.crash.CrashReportCategory;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.INBT;
 import net.minecraft.nbt.ListNBT;
 import net.minecraft.nbt.NBTDynamicOps;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.nbt.StringNBT;
+import net.minecraft.network.play.server.SServerDifficultyPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.SharedConstants;
 import net.minecraft.util.UUIDCodec;
@@ -34,14 +_,18 @@
 import net.minecraft.world.WorldSettings;
 import net.minecraft.world.border.WorldBorder;
 import net.minecraft.world.gen.settings.DimensionGeneratorSettings;
+import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
 
 public class ServerWorldInfo implements IServerWorldInfo, IServerConfiguration {
    private static final Logger field_237341_a_ = LogManager.getLogger();
-   private WorldSettings field_237342_b_;
+   public WorldSettings field_237342_b_;
    private final DimensionGeneratorSettings field_237343_c_;
    private final Lifecycle field_237344_d_;
    private int field_237345_e_;
@@ -75,6 +_,7 @@
    private final Set<String> field_237338_B_;
    private boolean field_237339_C_;
    private final TimerCallbackManager<MinecraftServer> field_237340_D_;
+   public ServerWorld world; // CraftBukkit
 
    private ServerWorldInfo(@Nullable DataFixer p_i242043_1_, int p_i242043_2_, @Nullable CompoundNBT p_i242043_3_, boolean p_i242043_4_, int p_i242043_5_, int p_i242043_6_, int p_i242043_7_, float p_i242043_8_, long p_i242043_9_, long p_i242043_11_, int p_i242043_13_, int p_i242043_14_, int p_i242043_15_, boolean p_i242043_16_, int p_i242043_17_, boolean p_i242043_18_, boolean p_i242043_19_, boolean p_i242043_20_, WorldBorder.Serializer p_i242043_21_, int p_i242043_22_, int p_i242043_23_, @Nullable UUID p_i242043_24_, LinkedHashSet<String> p_i242043_25_, TimerCallbackManager<MinecraftServer> p_i242043_26_, @Nullable CompoundNBT p_i242043_27_, CompoundNBT p_i242043_28_, WorldSettings p_i242043_29_, DimensionGeneratorSettings p_i242043_30_, Lifecycle p_i242043_31_) {
       this.field_237350_j_ = p_i242043_1_;
@@ -189,6 +_,7 @@
          p_237370_2_.func_186854_a("WanderingTraderId", this.field_237337_A_);
       }
 
+      p_237370_2_.func_74778_a("Bukkit.Version", "CatServer/1.16.5/" + org.bukkit.craftbukkit.v1_16_R3.util.Versioning.getBukkitVersion()); // CraftBukkit
    }
 
    public int func_76079_c() {
@@ -286,6 +_,19 @@
    }
 
    public void func_76069_a(boolean p_76069_1_) {
+      // CraftBukkit start
+      if (this.field_237358_r_ == p_76069_1_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(func_76065_j());
+      if (world != null) {
+         ThunderChangeEvent thunder = new ThunderChangeEvent(world, p_76069_1_);
+         Bukkit.getServer().getPluginManager().callEvent(thunder);
+         if (thunder.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.field_237358_r_ = p_76069_1_;
    }
 
@@ -302,6 +_,19 @@
    }
 
    public void func_76084_b(boolean p_76084_1_) {
+      // CraftBukkit start
+      if (this.field_237356_p_ == p_76084_1_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(func_76065_j());
+      if (world != null) {
+         WeatherChangeEvent weather = new WeatherChangeEvent(world, p_76084_1_);
+         Bukkit.getServer().getPluginManager().callEvent(weather);
+         if (weather.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.field_237356_p_ = p_76084_1_;
    }
 
@@ -355,6 +_,12 @@
 
    public void func_230409_a_(Difficulty p_230409_1_) {
       this.field_237342_b_ = this.field_237342_b_.func_234948_a_(p_230409_1_);
+      /*/ CraftBukkit start
+      SServerDifficultyPacket packet = new SServerDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+      for (ServerPlayerEntity player : world.getPlayers()) {
+         player.connection.sendPacket(packet);
+      }
+      // CraftBukkit end*/
    }
 
    public boolean func_176123_z() {
@@ -449,4 +_,12 @@
    public WorldSettings func_230408_H_() {
       return this.field_237342_b_.func_234959_h_();
    }
+
+   // CraftBukkit start - Check if the name stored in NBT is the correct one
+   public void checkName(String name) {
+      if (!this.field_237342_b_.field_234943_a_.equals(name)) {
+         this.field_237342_b_.field_234943_a_ = name;
+      }
+   }
+   // CraftBukkit end
 }
