--- a/net/minecraft/block/trees/BigTree.java
+++ b/net/minecraft/block/trees/BigTree.java
@@ -34,6 +_,7 @@
          return false;
       } else {
          configuredfeature.field_222738_b.func_227373_a_();
+         setTreeType(configuredfeature); // CraftBukkit
          BlockState blockstate = Blocks.field_150350_a.func_176223_P();
          p_235678_1_.func_180501_a(p_235678_3_.func_177982_a(p_235678_6_, 0, p_235678_7_), blockstate, 4);
          p_235678_1_.func_180501_a(p_235678_3_.func_177982_a(p_235678_6_ + 1, 0, p_235678_7_), blockstate, 4);
