--- a/net/minecraft/block/trees/Tree.java
+++ b/net/minecraft/block/trees/Tree.java
@@ -4,13 +_,16 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
+import net.minecraft.block.SaplingBlock;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.gen.ChunkGenerator;
 import net.minecraft.world.gen.feature.BaseTreeFeatureConfig;
 import net.minecraft.world.gen.feature.ConfiguredFeature;
+import net.minecraft.world.gen.feature.Features;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.TreeType;
 
 public abstract class Tree {
    @Nullable
@@ -21,6 +_,7 @@
       if (configuredfeature == null) {
          return false;
       } else {
+         setTreeType(configuredfeature); // CraftBukkit
          p_230339_1_.func_180501_a(p_230339_3_, Blocks.field_150350_a.func_176223_P(), 4);
          configuredfeature.field_222738_b.func_227373_a_();
          if (configuredfeature.func_242765_a(p_230339_1_, p_230339_2_, p_230339_5_, p_230339_3_)) {
@@ -41,4 +_,47 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(final ConfiguredFeature<?, ?> worldgentreeabstract) {
+      if (worldgentreeabstract == Features.field_243862_bH || worldgentreeabstract == Features.field_243879_bY) {
+         SaplingBlock.treeType = TreeType.TREE;
+      } else if (worldgentreeabstract == Features.field_243861_bG) {
+         SaplingBlock.treeType = TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == Features.field_243860_bF) {
+         SaplingBlock.treeType = TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == Features.field_243868_bN) {
+         SaplingBlock.treeType = TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == Features.field_243870_bP) {
+         SaplingBlock.treeType = TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == Features.field_243867_bM) {
+         SaplingBlock.treeType = TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == Features.field_243866_bL) {
+         SaplingBlock.treeType = TreeType.REDWOOD;
+      } else if (worldgentreeabstract == Features.field_243865_bK) {
+         SaplingBlock.treeType = TreeType.ACACIA;
+      } else if (worldgentreeabstract == Features.field_243864_bJ || worldgentreeabstract == Features.field_243919_cb) {
+         SaplingBlock.treeType = TreeType.BIRCH;
+      } else if (worldgentreeabstract == Features.field_243877_bW) {
+         SaplingBlock.treeType = TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == Features.field_243875_bU) {
+         SaplingBlock.treeType = TreeType.SWAMP;
+      } else if (worldgentreeabstract == Features.field_243869_bO || worldgentreeabstract == Features.field_243922_ce) {
+         SaplingBlock.treeType = TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == Features.field_243876_bV) {
+         SaplingBlock.treeType = TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == Features.field_243863_bI) {
+         SaplingBlock.treeType = TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == Features.field_243872_bR) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == Features.field_243873_bS) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract != Features.field_243871_bQ) {
+         SaplingBlock.treeType = TreeType.JUNGLE;
+      } else {
+         SaplingBlock.treeType = TreeType.FORGE_MODS;
+      }
+
+   }
+   // CraftBukkit start
 }
