--- a/net/minecraft/inventory/container/AbstractFurnaceContainer.java
+++ b/net/minecraft/inventory/container/AbstractFurnaceContainer.java
@@ -19,6 +_,8 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryFurnace;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public abstract class AbstractFurnaceContainer extends RecipeBookContainer<IInventory> {
    private final IInventory field_217063_d;
@@ -27,6 +_,23 @@
    private final IRecipeType<? extends AbstractCookingRecipe> field_217065_f;
    private final RecipeBookCategory field_242384_g;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+   // CraftBukkit stop
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryFurnace inventory = new CraftInventoryFurnace((AbstractFurnaceTileEntity) this.field_217063_d);
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    protected AbstractFurnaceContainer(ContainerType<?> p_i241921_1_, IRecipeType<? extends AbstractCookingRecipe> p_i241921_2_, RecipeBookCategory p_i241921_3_, int p_i241921_4_, PlayerInventory p_i241921_5_) {
       this(p_i241921_1_, p_i241921_2_, p_i241921_3_, p_i241921_4_, p_i241921_5_, new Inventory(3), new IntArray(4));
    }
@@ -43,7 +_,7 @@
       this.func_75146_a(new Slot(p_i241922_6_, 0, 56, 17));
       this.func_75146_a(new FurnaceFuelSlot(this, p_i241922_6_, 1, 56, 53));
       this.func_75146_a(new FurnaceResultSlot(p_i241922_5_.field_70458_d, p_i241922_6_, 2, 116, 35));
-
+      this.player = p_i241922_5_; // CraftBukkit - save player
       for(int i = 0; i < 3; ++i) {
          for(int j = 0; j < 9; ++j) {
             this.func_75146_a(new Slot(p_i241922_5_, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
@@ -94,6 +_,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_217063_d.func_70300_a(p_75145_1_);
    }
 
@@ -150,7 +_,7 @@
    }
 
    protected boolean func_217058_b(ItemStack p_217058_1_) {
-      return AbstractFurnaceTileEntity.func_213991_b(p_217058_1_);
+      return net.minecraftforge.common.ForgeHooks.getBurnTime(p_217058_1_, this.field_217065_f) > 0;
    }
 
    @OnlyIn(Dist.CLIENT)
