--- a/net/minecraft/inventory/container/BrewingStandContainer.java
+++ b/net/minecraft/inventory/container/BrewingStandContainer.java
@@ -6,22 +_,27 @@
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
-import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
 import net.minecraft.potion.Potion;
-import net.minecraft.potion.PotionBrewing;
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.IIntArray;
 import net.minecraft.util.IntArray;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryBrewer;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class BrewingStandContainer extends Container {
    private final IInventory field_75188_e;
    private final IIntArray field_216983_d;
    private final Slot field_75186_f;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+   // CraftBukkit end
+
    public BrewingStandContainer(int p_i50095_1_, PlayerInventory p_i50095_2_) {
       this(p_i50095_1_, p_i50095_2_, new Inventory(5), new IntArray(2));
    }
@@ -30,6 +_,7 @@
       super(ContainerType.field_221517_k, p_i50096_1_);
       func_216962_a(p_i50096_3_, 5);
       func_216959_a(p_i50096_4_, 2);
+      player = p_i50096_2_;  // CraftBukkit
       this.field_75188_e = p_i50096_3_;
       this.field_216983_d = p_i50096_4_;
       this.func_75146_a(new BrewingStandContainer.PotionSlot(p_i50096_3_, 0, 56, 51));
@@ -48,10 +_,22 @@
       for(int k = 0; k < 9; ++k) {
          this.func_75146_a(new Slot(p_i50096_2_, k, 8 + k * 18, 142));
       }
+   }
 
-   }
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+   }
+      CraftInventoryBrewer inventory = new CraftInventoryBrewer(this.field_75188_e);
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_75188_e.func_70300_a(p_75145_1_);
    }
 
@@ -143,7 +_,7 @@
       }
 
       public boolean func_75214_a(ItemStack p_75214_1_) {
-         return PotionBrewing.func_185205_a(p_75214_1_);
+         return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(p_75214_1_);
       }
 
       public int func_75219_a() {
@@ -167,6 +_,7 @@
       public ItemStack func_190901_a(PlayerEntity p_190901_1_, ItemStack p_190901_2_) {
          Potion potion = PotionUtils.func_185191_c(p_190901_2_);
          if (p_190901_1_ instanceof ServerPlayerEntity) {
+            net.minecraftforge.event.ForgeEventFactory.onPlayerBrewedPotion(p_190901_1_, p_190901_2_);
             CriteriaTriggers.field_192130_j.func_192173_a((ServerPlayerEntity)p_190901_1_, potion);
          }
 
@@ -175,8 +_,7 @@
       }
 
       public static boolean func_75243_a_(ItemStack p_75243_0_) {
-         Item item = p_75243_0_.func_77973_b();
-         return item == Items.field_151068_bn || item == Items.field_185155_bH || item == Items.field_185156_bI || item == Items.field_151069_bo;
+         return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(p_75243_0_);
       }
    }
 }
