--- a/net/minecraft/inventory/container/CartographyContainer.java
+++ b/net/minecraft/inventory/container/CartographyContainer.java
@@ -14,8 +_,28 @@
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.storage.MapData;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryCartography;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
+import org.bukkit.entity.Player;
 
 public class CartographyContainer extends Container {
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryCartography inventory = new CraftInventoryCartography(this.field_216998_c, this.field_217001_f);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    private final IWorldPosCallable field_216999_d;
    private long field_226605_f_;
    public final IInventory field_216998_c = new Inventory(2) {
@@ -23,12 +_,26 @@
          CartographyContainer.this.func_75130_a(this);
          super.func_70296_d();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return field_216999_d.getLocation();
+      }
+      // CraftBukkit end
    };
    private final CraftResultInventory field_217001_f = new CraftResultInventory() {
       public void func_70296_d() {
          CartographyContainer.this.func_75130_a(this);
          super.func_70296_d();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return field_216999_d.getLocation();
+      }
+      // CraftBukkit end
    };
 
    public CartographyContainer(int p_i50093_1_, PlayerInventory p_i50093_2_) {
@@ -80,9 +_,12 @@
          this.func_75146_a(new Slot(p_i50094_2_, k, 8 + k * 18, 142));
       }
 
+      player = (Player) p_i50094_2_.field_70458_d.getBukkitEntity(); // CraftBukkit
+
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return func_216963_a(this.field_216999_d, p_75145_1_, Blocks.field_222425_lN);
    }
 
