--- a/net/minecraft/inventory/container/DispenserContainer.java
+++ b/net/minecraft/inventory/container/DispenserContainer.java
@@ -5,16 +_,26 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class DispenserContainer extends Container {
    private final IInventory field_178146_a;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+   // CraftBukkit end
+
    public DispenserContainer(int p_i50087_1_, PlayerInventory p_i50087_2_) {
       this(p_i50087_1_, p_i50087_2_, new Inventory(9));
    }
 
    public DispenserContainer(int p_i50088_1_, PlayerInventory p_i50088_2_, IInventory p_i50088_3_) {
       super(ContainerType.field_221513_g, p_i50088_1_);
+      // CraftBukkit start - Save player
+      this.player = p_i50088_2_;
+      // CraftBukkit end
       func_216962_a(p_i50088_3_, 9);
       this.field_178146_a = p_i50088_3_;
       p_i50088_3_.func_174889_b(p_i50088_2_.field_70458_d);
@@ -38,6 +_,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_178146_a.func_70300_a(p_75145_1_);
    }
 
@@ -75,4 +_,16 @@
       super.func_75134_a(p_75134_1_);
       this.field_178146_a.func_174886_c(p_75134_1_);
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventory inventory = new CraftInventory(this.field_178146_a);
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
