--- a/net/minecraft/inventory/container/FurnaceResultSlot.java
+++ b/net/minecraft/inventory/container/FurnaceResultSlot.java
@@ -40,9 +_,10 @@
    protected void func_75208_c(ItemStack p_75208_1_) {
       p_75208_1_.func_77980_a(this.field_75229_a.field_70170_p, this.field_75229_a, this.field_75228_b);
       if (!this.field_75229_a.field_70170_p.field_72995_K && this.field_75224_c instanceof AbstractFurnaceTileEntity) {
-         ((AbstractFurnaceTileEntity)this.field_75224_c).func_235645_d_(this.field_75229_a);
+         ((AbstractFurnaceTileEntity)this.field_75224_c).unlockRecipes(this.field_75229_a, p_75208_1_, this.field_75228_b); // CraftBukkit
       }
 
       this.field_75228_b = 0;
+      net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerSmeltedEvent(this.field_75229_a, p_75208_1_);
    }
 }
