--- a/net/minecraft/inventory/container/GrindstoneContainer.java
+++ b/net/minecraft/inventory/container/GrindstoneContainer.java
@@ -17,6 +_,10 @@
 import net.minecraft.item.Items;
 import net.minecraft.util.IWorldPosCallable;
 import net.minecraft.world.World;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryGrindstone;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
+import org.bukkit.entity.Player;
 
 public class GrindstoneContainer extends Container {
    private final IInventory field_217013_c = new CraftResultInventory();
@@ -25,9 +_,31 @@
          super.func_70296_d();
          GrindstoneContainer.this.func_75130_a(this);
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return field_217014_d.getLocation();
+      }
+      // CraftBukkit end
    };
    private final IWorldPosCallable field_217015_e;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Player cbPlayer;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryGrindstone inventory = new CraftInventoryGrindstone(this.field_217014_d, this.field_217013_c);
+      bukkitEntity = new CraftInventoryView(this.cbPlayer, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public GrindstoneContainer(int p_i50080_1_, PlayerInventory p_i50080_2_) {
       this(p_i50080_1_, p_i50080_2_, IWorldPosCallable.field_221489_a);
    }
@@ -105,6 +_,7 @@
          this.func_75146_a(new Slot(p_i50081_2_, k, 8 + k * 18, 142));
       }
 
+      cbPlayer = (Player) p_i50081_2_.field_70458_d.getBukkitEntity(); // CraftBukkit
    }
 
    public void func_75130_a(IInventory p_75130_1_) {
@@ -141,12 +_,13 @@
             }
 
             Item item = itemstack.func_77973_b();
-            int k = item.func_77612_l() - itemstack.func_77952_i();
-            int l = item.func_77612_l() - itemstack1.func_77952_i();
-            int i1 = k + l + item.func_77612_l() * 5 / 100;
-            i = Math.max(item.func_77612_l() - i1, 0);
+            int k = itemstack.func_77958_k() - itemstack.func_77952_i();
+            int l = itemstack.func_77958_k() - itemstack1.func_77952_i();
+            int i1 = k + l + itemstack.func_77958_k() * 5 / 100;
+            i = Math.max(itemstack.func_77958_k() - i1, 0);
             itemstack2 = this.func_217011_b(itemstack, itemstack1);
-            if (!itemstack2.func_77984_f()) {
+            if (!itemstack2.isRepairable()) i = itemstack.func_77952_i();
+            if (!itemstack2.func_77984_f() || !itemstack2.isRepairable()) {
                if (!ItemStack.func_77989_b(itemstack, itemstack1)) {
                   this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
                   this.func_75142_b();
@@ -219,6 +_,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return func_216963_a(this.field_217015_e, p_75145_1_, Blocks.field_222427_lP);
    }
 
