--- a/net/minecraft/inventory/container/HopperContainer.java
+++ b/net/minecraft/inventory/container/HopperContainer.java
@@ -5,10 +_,28 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class HopperContainer extends Container {
    private final IInventory field_94538_a;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventory inventory = new CraftInventory(this.field_94538_a);
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+
+   // CraftBukkit stop
+
    public HopperContainer(int p_i50078_1_, PlayerInventory p_i50078_2_) {
       this(p_i50078_1_, p_i50078_2_, new Inventory(5));
    }
@@ -16,6 +_,7 @@
    public HopperContainer(int p_i50079_1_, PlayerInventory p_i50079_2_, IInventory p_i50079_3_) {
       super(ContainerType.field_221522_p, p_i50079_1_);
       this.field_94538_a = p_i50079_3_;
+      this.player = p_i50079_2_; // CraftBukkit - save player
       func_216962_a(p_i50079_3_, 5);
       p_i50079_3_.func_174889_b(p_i50079_2_.field_70458_d);
       int i = 51;
@@ -37,6 +_,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_94538_a.func_70300_a(p_75145_1_);
    }
 
