--- a/net/minecraft/inventory/container/MerchantContainer.java
+++ b/net/minecraft/inventory/container/MerchantContainer.java
@@ -13,6 +_,7 @@
 import net.minecraft.util.SoundCategory;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class MerchantContainer extends Container {
    private final IMerchant field_75178_e;
@@ -24,6 +_,19 @@
    @OnlyIn(Dist.CLIENT)
    private boolean field_223433_g;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity == null) {
+         bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), new org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryMerchant(field_75178_e, field_75176_f), this);
+      }
+      return bukkitEntity;
+   }
+   // CraftBukkit stop
+
    public MerchantContainer(int p_i50068_1_, PlayerInventory p_i50068_2_) {
       this(p_i50068_1_, p_i50068_2_, new NPCMerchant(p_i50068_2_.field_70458_d));
    }
@@ -31,6 +_,7 @@
    public MerchantContainer(int p_i50069_1_, PlayerInventory p_i50069_2_, IMerchant p_i50069_3_) {
       super(ContainerType.field_221525_s, p_i50069_1_);
       this.field_75178_e = p_i50069_3_;
+      this.player = p_i50069_2_; // CraftBukkit - save player
       this.field_75176_f = new MerchantInventory(p_i50069_3_);
       this.func_75146_a(new Slot(this.field_75176_f, 0, 136, 37));
       this.func_75146_a(new Slot(this.field_75176_f, 1, 162, 37));
@@ -147,7 +_,7 @@
    }
 
    private void func_223132_j() {
-      if (!this.field_75178_e.func_190670_t_().field_72995_K) {
+      if (!this.field_75178_e.func_190670_t_().field_72995_K && this.field_75178_e instanceof Entity) { // CraftBukkit - SPIGOT-5035
          Entity entity = (Entity)this.field_75178_e;
          this.field_75178_e.func_190670_t_().func_184134_a(entity.func_226277_ct_(), entity.func_226278_cu_(), entity.func_226281_cx_(), this.field_75178_e.func_213714_ea(), SoundCategory.NEUTRAL, 1.0F, 1.0F, false);
       }
