--- a/net/minecraft/inventory/container/RecipeBookContainer.java
+++ b/net/minecraft/inventory/container/RecipeBookContainer.java
@@ -8,8 +_,10 @@
 import net.minecraft.item.crafting.ServerRecipePlacer;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.inventory.InventoryView;
 
 public abstract class RecipeBookContainer<C extends IInventory> extends Container {
+
    public RecipeBookContainer(ContainerType<?> p_i50067_1_, int p_i50067_2_) {
       super(p_i50067_1_, p_i50067_2_);
    }
@@ -32,6 +_,10 @@
 
    @OnlyIn(Dist.CLIENT)
    public abstract int func_203721_h();
+
+   public java.util.List<net.minecraft.client.util.RecipeBookCategories> getRecipeBookCategories() {
+      return net.minecraft.client.util.RecipeBookCategories.func_243236_a(this.func_241850_m());
+   }
 
    @OnlyIn(Dist.CLIENT)
    public abstract RecipeBookCategory func_241850_m();
