--- a/net/minecraft/inventory/container/ShulkerBoxContainer.java
+++ b/net/minecraft/inventory/container/ShulkerBoxContainer.java
@@ -5,10 +_,26 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class ShulkerBoxContainer extends Container {
    private final IInventory field_190899_a;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   private PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), new CraftInventory(this.field_190899_a), this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public ShulkerBoxContainer(int p_i50065_1_, PlayerInventory p_i50065_2_) {
       this(p_i50065_1_, p_i50065_2_, new Inventory(27));
    }
@@ -16,6 +_,7 @@
    public ShulkerBoxContainer(int p_i50066_1_, PlayerInventory p_i50066_2_, IInventory p_i50066_3_) {
       super(ContainerType.field_221526_t, p_i50066_1_);
       func_216962_a(p_i50066_3_, 27);
+      this.player = p_i50066_2_; // CraftBukkit - save player
       this.field_190899_a = p_i50066_3_;
       p_i50066_3_.func_174889_b(p_i50066_2_.field_70458_d);
       int i = 3;
