--- a/net/minecraft/inventory/container/SmithingTableContainer.java
+++ b/net/minecraft/inventory/container/SmithingTableContainer.java
@@ -11,6 +_,7 @@
 import net.minecraft.item.crafting.SmithingRecipe;
 import net.minecraft.util.IWorldPosCallable;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class SmithingTableContainer extends AbstractRepairContainer {
    private final World field_234651_g_;
@@ -18,6 +_,10 @@
    private SmithingRecipe field_234652_h_;
    private final List<SmithingRecipe> field_241443_i_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   // CraftBukkit end
+
    public SmithingTableContainer(int p_i231590_1_, PlayerInventory p_i231590_2_) {
       this(p_i231590_1_, p_i231590_2_, IWorldPosCallable.field_221489_a);
    }
@@ -56,12 +_,15 @@
    public void func_82848_d() {
       List<SmithingRecipe> list = this.field_234651_g_.func_199532_z().func_215370_b(IRecipeType.field_234827_g_, this.field_234643_d_, this.field_234651_g_);
       if (list.isEmpty()) {
-         this.field_234642_c_.func_70299_a(0, ItemStack.field_190927_a);
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), ItemStack.field_190927_a); // CraftBukkit
       } else {
          this.field_234652_h_ = list.get(0);
          ItemStack itemstack = this.field_234652_h_.func_77572_b(this.field_234643_d_);
          this.field_234642_c_.func_193056_a(this.field_234652_h_);
-         this.field_234642_c_.func_70299_a(0, itemstack);
+         // CraftBukkit start
+         this.field_234643_d_.setCurrentRecipe(this.field_234652_h_);
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), itemstack);
+         // CraftBukkit end
       }
 
    }
@@ -75,4 +_,16 @@
    public boolean func_94530_a(ItemStack p_94530_1_, Slot p_94530_2_) {
       return p_94530_2_.field_75224_c != this.field_234642_c_ && super.func_94530_a(p_94530_1_, p_94530_2_);
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventorySmithing(this.field_234644_e_.getLocation(), this.field_234643_d_, this.field_234642_c_);
+      bukkitEntity = new CraftInventoryView(this.field_234645_f_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
