--- a/net/minecraft/world/item/ArmorStandItem.java
+++ b/net/minecraft/world/item/ArmorStandItem.java
@@ -47,6 +_,11 @@
 
                float f = (float)Mth.m_14143_((Mth.m_14177_(p_40510_.m_7074_() - 180.0F) + 22.5F) / 45.0F) * 45.0F;
                armorstand.m_7678_(armorstand.m_20185_(), armorstand.m_20186_(), armorstand.m_20189_(), f, 0.0F);
+               // CraftBukkit start
+               if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityPlaceEvent(p_40510_, armorstand).isCancelled()) {
+                  return InteractionResult.FAIL;
+               }
+               // CraftBukkit end
                serverlevel.m_47205_(armorstand);
                level.m_6263_((Player)null, armorstand.m_20185_(), armorstand.m_20186_(), armorstand.m_20189_(), SoundEvents.f_11684_, SoundSource.BLOCKS, 0.75F, 0.8F);
                armorstand.m_146852_(GameEvent.f_157810_, p_40510_.m_43723_());
