--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 
@@ -47,6 +_,13 @@
          }
 
          if (hitresult.m_6662_() == HitResult.Type.BLOCK) {
+            // CraftBukkit start - Boat placement
+            BlockHitResult blockHitResult = (BlockHitResult) hitresult;
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent(p_40623_, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, blockHitResult.m_82425_(), blockHitResult.m_82434_(), itemstack, p_40624_);
+            if (event.isCancelled()) {
+               return InteractionResultHolder.m_19098_(itemstack);
+            }
+            // CraftBukkit end
             Boat boat = new Boat(p_40622_, hitresult.m_82450_().f_82479_, hitresult.m_82450_().f_82480_, hitresult.m_82450_().f_82481_);
             boat.m_38332_(this.f_40616_);
             boat.m_146922_(p_40623_.m_146908_());
@@ -54,7 +_,14 @@
                return InteractionResultHolder.m_19100_(itemstack);
             } else {
                if (!p_40622_.f_46443_) {
-                  p_40622_.m_7967_(boat);
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPlaceEvent(p_40622_, blockHitResult.m_82425_(), blockHitResult.m_82434_(), p_40623_, boat).isCancelled()) {
+                     return InteractionResultHolder.m_19100_(itemstack);
+                  }
+                  if (!p_40622_.m_7967_(boat)) {
+                     return InteractionResultHolder.m_19098_(itemstack);
+                  }
+                  // CraftBukkit end
                   p_40622_.m_142346_(p_40623_, GameEvent.f_157810_, new BlockPos(hitresult.m_82450_()));
                   if (!p_40623_.m_150110_().f_35937_) {
                      itemstack.m_41774_(1);
