--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -47,6 +_,14 @@
          }
 
          if (hitresult.m_6662_() == HitResult.Type.BLOCK) {
+            // TODO
+            /*/ CraftBukkit start - Boat placement
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent(p_40623_, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, hitresult.getBlockPos(), hitresult.getDirection(), itemstack, p_40624_);
+
+            if (event.isCancelled()) {
+               return InteractionResultHolder.pass(itemstack);
+            }
+            // CraftBukkit end*/
             Boat boat = new Boat(p_40622_, hitresult.m_82450_().f_82479_, hitresult.m_82450_().f_82480_, hitresult.m_82450_().f_82481_);
             boat.m_38332_(this.f_40616_);
             boat.m_146922_(p_40623_.m_146908_());
