--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -49,6 +_,13 @@
          }
 
          if (hitresult.m_6662_() == HitResult.Type.BLOCK) {
+            // CraftBukkit start - Boat placement
+            net.minecraft.world.phys.BlockHitResult blockHitResult = (net.minecraft.world.phys.BlockHitResult) hitresult;
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerInteractEvent(p_40623_, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, blockHitResult.m_82425_(), blockHitResult.m_82434_(), itemstack, false, p_40624_, blockHitResult.m_82450_());
+            if (event.isCancelled()) {
+               return InteractionResultHolder.m_19098_(itemstack);
+            }
+            // CraftBukkit end
             Boat boat = this.m_220016_(p_40622_, hitresult);
             boat.m_28464_(this.f_40616_);
             boat.m_146922_(p_40623_.m_146908_());
@@ -56,7 +_,14 @@
                return InteractionResultHolder.m_19100_(itemstack);
             } else {
                if (!p_40622_.f_46443_) {
-                  p_40622_.m_7967_(boat);
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityPlaceEvent(p_40622_, blockHitResult.m_82425_(), blockHitResult.m_82434_(), p_40623_, boat, p_40624_).isCancelled()) {
+                     return InteractionResultHolder.m_19100_(itemstack);
+                  }
+                  if (!p_40622_.m_7967_(boat)) {
+                     return InteractionResultHolder.m_19098_(itemstack);
+                  }
+                  // CraftBukkit end
                   p_40622_.m_220400_(p_40623_, GameEvent.f_157810_, hitresult.m_82450_());
                   if (!p_40623_.m_150110_().f_35937_) {
                      itemstack.m_41774_(1);
