--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.item;
 
+import java.util.Objects;
+import java.util.Optional;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -7,6 +_,7 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
 import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.world.InteractionResult;
@@ -33,6 +_,32 @@
       Level level = p_40637_.m_43725_();
       BlockPos blockpos = p_40637_.m_8083_();
       BlockPos blockpos1 = blockpos.m_142300_(p_40637_.m_43719_());
+      if (applyBonemeal(p_40637_.m_43722_(), level, blockpos, p_40637_.m_43723_())) {
+         if (!level.f_46443_) {
+            level.m_46796_(1505, blockpos, 0);
+         }
+
+         return InteractionResult.m_19078_(level.f_46443_);
+      } else {
+         BlockState blockstate = level.m_8055_(blockpos);
+         boolean flag = blockstate.m_60783_(level, blockpos, p_40637_.m_43719_());
+         if (flag && m_40631_(p_40637_.m_43722_(), level, blockpos1, p_40637_.m_43719_())) {
+            if (!level.f_46443_) {
+               level.m_46796_(1505, blockpos1, 0);
+            }
+
+            return InteractionResult.m_19078_(level.f_46443_);
+         } else {
+            return InteractionResult.PASS;
+         }
+      }
+   }
+
+   // CraftBukkit FoxServer start
+   public static InteractionResult applyBonemeal(UseOnContext p_40637_) {
+      Level level = p_40637_.m_43725_();
+      BlockPos blockpos = p_40637_.m_8083_();
+      BlockPos blockpos1 = blockpos.m_142300_(p_40637_.m_43719_());
       if (m_40627_(p_40637_.m_43722_(), level, blockpos)) {
          if (!level.f_46443_) {
             level.m_46796_(1505, blockpos, 0);
@@ -53,9 +_,19 @@
          }
       }
    }
+   // CraftBukkit FoxServer end
 
+   @Deprecated //Forge: Use Player/Hand version
    public static boolean m_40627_(ItemStack p_40628_, Level p_40629_, BlockPos p_40630_) {
+      if (p_40629_ instanceof net.minecraft.server.level.ServerLevel)
+         return applyBonemeal(p_40628_, p_40629_, p_40630_, net.minecraftforge.common.util.FakePlayerFactory.getMinecraft((net.minecraft.server.level.ServerLevel)p_40629_));
+      return false;
+   }
+
+   public static boolean applyBonemeal(ItemStack p_40628_, Level p_40629_, BlockPos p_40630_, net.minecraft.world.entity.player.Player player) {
       BlockState blockstate = p_40629_.m_8055_(p_40630_);
+      int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, p_40629_, p_40630_, blockstate, p_40628_);
+      if (hook != 0) return hook > 0;
       if (blockstate.m_60734_() instanceof BonemealableBlock) {
          BonemealableBlock bonemealableblock = (BonemealableBlock)blockstate.m_60734_();
          if (bonemealableblock.m_7370_(p_40629_, p_40630_, blockstate, p_40629_.f_46443_)) {
