--- a/net/minecraft/world/item/CrossbowItem.java
+++ b/net/minecraft/world/item/CrossbowItem.java
@@ -21,6 +_,7 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResultHolder;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.monster.CrossbowAttackMob;
 import net.minecraft.world.entity.player.Player;
@@ -105,6 +_,10 @@
          if (itemstack.m_41619_() && flag) {
             itemstack = new ItemStack(Items.f_42412_);
             itemstack1 = itemstack.m_41777_();
+            // CraftBukkit start - SPIGOT-4870, MC-150847
+         } else if (itemstack1.m_41619_()) {
+            return false;
+            // CraftBukkit end
          }
 
          if (!m_40862_(p_40860_, p_40861_, itemstack, k > 0, flag)) {
@@ -217,10 +_,28 @@
             projectile.m_6686_((double)vector3f.m_122239_(), (double)vector3f.m_122260_(), (double)vector3f.m_122269_(), p_40902_, p_40903_);
          }
 
+         // CraftBukkit start
+         org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityShootBowEvent(p_40896_, p_40898_, p_40899_, projectile, p_40896_.m_7655_(), p_40900_, true);
+         if (event.isCancelled()) {
+            event.getProjectile().remove();
+            return;
+         }
+         // CraftBukkit end
+
          p_40898_.m_41622_(flag ? 3 : 1, p_40896_, (p_40858_) -> {
             p_40858_.m_21190_(p_40897_);
          });
          p_40895_.m_7967_(projectile);
+         // CraftBukkit start
+         if (event.getProjectile() == ((Entity) projectile).getBukkitEntity()) {
+            if (!p_40895_.m_7967_((Entity) projectile)) {
+               if (p_40896_ instanceof ServerPlayer) {
+                  ((ServerPlayer) p_40896_).getBukkitEntity().updateInventory();
+               }
+               return;
+            }
+         }
+         // CraftBukkit end
          p_40895_.m_6263_((Player)null, p_40896_.m_20185_(), p_40896_.m_20186_(), p_40896_.m_20189_(), SoundEvents.f_11847_, SoundSource.PLAYERS, 1.0F, p_40900_);
       }
    }
