--- a/net/minecraft/world/item/DebugStickItem.java
+++ b/net/minecraft/world/item/DebugStickItem.java
@@ -49,7 +_,7 @@
       return InteractionResult.m_19078_(level.f_46443_);
    }
 
-   private boolean m_150802_(Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_) {
+   public boolean m_150802_(Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_) { // CatServer - private -> public
       if (!p_150803_.m_36337_()) {
          return false;
       } else {
