--- a/net/minecraft/world/item/DebugStickItem.java
+++ b/net/minecraft/world/item/DebugStickItem.java
@@ -52,7 +_,7 @@
    }
 
    private boolean m_150802_(Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_) {
-      if (!p_150803_.m_36337_()) {
+      if (!p_150803_.m_36337_() && !(p_150803_.m_150110_().f_35937_ && p_150803_.getBukkitEntity().hasPermission("minecraft.debugstick")) && !p_150803_.getBukkitEntity().hasPermission("minecraft.debugstick.always")) { // Spigot
          return false;
       } else {
          Block block = p_150804_.m_60734_();
