--- a/net/minecraft/world/item/DyeItem.java
+++ b/net/minecraft/world/item/DyeItem.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.animal.Sheep;
 import net.minecraft.world.entity.player.Player;
+import org.bukkit.event.entity.SheepDyeWoolEvent;
 
 public class DyeItem extends Item {
    private static final Map<DyeColor, DyeItem> f_41076_ = Maps.newEnumMap(DyeColor.class);
@@ -26,7 +_,16 @@
          if (sheep.m_6084_() && !sheep.m_29875_() && sheep.m_29874_() != this.f_41077_) {
             sheep.f_19853_.m_6269_(p_41086_, sheep, SoundEvents.f_144133_, SoundSource.PLAYERS, 1.0F, 1.0F);
             if (!p_41086_.f_19853_.f_46443_) {
-               sheep.m_29855_(this.f_41077_);
+               // CraftBukkit start
+               byte bColor = (byte) this.f_41077_.m_41060_();
+               SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) sheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor), (org.bukkit.entity.Player) p_41086_.getBukkitEntity());
+               sheep.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+               if (event.isCancelled()) {
+                  return InteractionResult.PASS;
+               }
+               sheep.m_29855_(DyeColor.m_41053_((byte) event.getColor().getWoolData()));
+               // CraftBukkit end
                p_41085_.m_41774_(1);
             }
 
