--- a/net/minecraft/world/item/EggItem.java
+++ b/net/minecraft/world/item/EggItem.java
@@ -16,13 +_,21 @@
 
    public InteractionResultHolder<ItemStack> m_7203_(Level p_41128_, Player p_41129_, InteractionHand p_41130_) {
       ItemStack itemstack = p_41129_.m_21120_(p_41130_);
-      p_41128_.m_6263_((Player)null, p_41129_.m_20185_(), p_41129_.m_20186_(), p_41129_.m_20189_(), SoundEvents.f_11877_, SoundSource.PLAYERS, 0.5F, 0.4F / (p_41128_.m_5822_().nextFloat() * 0.4F + 0.8F));
       if (!p_41128_.f_46443_) {
          ThrownEgg thrownegg = new ThrownEgg(p_41128_, p_41129_);
          thrownegg.m_37446_(itemstack);
          thrownegg.m_37251_(p_41129_, p_41129_.m_146909_(), p_41129_.m_146908_(), 0.0F, 1.5F, 1.0F);
-         p_41128_.m_7967_(thrownegg);
+         // CraftBukkit start
+         if (!p_41128_.m_7967_(thrownegg)) {
+            if (p_41129_ instanceof net.minecraft.server.level.ServerPlayer) {
+               ((net.minecraft.server.level.ServerPlayer) p_41129_).getBukkitEntity().updateInventory();
+            }
+            return InteractionResultHolder.m_19100_(itemstack);
+         }
+         // CraftBukkit end
       }
+      p_41128_.m_6263_((Player)null, p_41129_.m_20185_(), p_41129_.m_20186_(), p_41129_.m_20189_(), SoundEvents.f_11877_, SoundSource.PLAYERS, 0.5F, 0.4F / (p_41128_.m_5822_().nextFloat() * 0.4F + 0.8F));
+
 
       p_41129_.m_36246_(Stats.f_12982_.m_12902_(this));
       if (!p_41129_.m_150110_().f_35937_) {
