--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -53,7 +_,26 @@
                   }
                }
 
-               level.m_6798_(1038, blockpos1.m_142082_(1, 0, 1), 0);
+               //level.globalLevelEvent(1038, blockpos1.offset(1, 0, 1), 0);
+               // CraftBukkit start - Use relative location for far away sounds
+               // world.b(1038, blockposition1.c(1, 0, 1), 0);
+               int viewDistance = level.getCraftServer().getViewDistance() * 16;
+               BlockPos soundPos = blockpos1.m_142082_(1, 0, 1);
+               for (ServerPlayer player : level.m_142572_().m_6846_().f_11196_) {
+                  double deltaX = soundPos.m_123341_() - player.m_20185_();
+                  double deltaZ = soundPos.m_123343_() - player.m_20189_();
+                  double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                  if (level.spigotConfig.endPortalSoundRadius > 0 && distanceSquared > level.spigotConfig.endPortalSoundRadius * level.spigotConfig.endPortalSoundRadius) continue; // Spigot
+                  if (distanceSquared > viewDistance * viewDistance) {
+                     double deltaLength = Math.sqrt(distanceSquared);
+                     double relativeX = player.m_20185_() + (deltaX / deltaLength) * viewDistance;
+                     double relativeZ = player.m_20189_() + (deltaZ / deltaLength) * viewDistance;
+                     player.f_8906_.m_141995_(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1038, new BlockPos((int) relativeX, (int) soundPos.m_123342_(), (int) relativeZ), 0, true));
+                  } else {
+                     player.f_8906_.m_141995_(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1038, soundPos, 0, true));
+                  }
+               }
+               // CraftBukkit end
             }
 
             return InteractionResult.CONSUME;
@@ -77,7 +_,11 @@
                EyeOfEnder eyeofender = new EyeOfEnder(p_41184_, p_41185_.m_20185_(), p_41185_.m_20227_(0.5D), p_41185_.m_20189_());
                eyeofender.m_36972_(itemstack);
                eyeofender.m_36967_(blockpos);
-               p_41184_.m_7967_(eyeofender);
+               // CraftBukkit start
+               if (!p_41184_.m_7967_(eyeofender)) {
+                  return new InteractionResultHolder(InteractionResult.FAIL, itemstack);
+               }
+               // CraftBukkit end
                if (p_41185_ instanceof ServerPlayer) {
                   CriteriaTriggers.f_10579_.m_73935_((ServerPlayer)p_41185_, blockpos);
                }
