--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -79,7 +_,11 @@
                eyeofender.m_36972_(itemstack);
                eyeofender.m_36967_(blockpos);
                p_41184_.m_214171_(GameEvent.f_157778_, eyeofender.m_20182_(), GameEvent.Context.m_223717_(p_41185_));
-               p_41184_.m_7967_(eyeofender);
+               // CraftBukkit start
+               if (!p_41184_.m_7967_(eyeofender)) {
+                  return new InteractionResultHolder<>(InteractionResult.FAIL, itemstack);
+               }
+               // CraftBukkit end
                if (p_41185_ instanceof ServerPlayer) {
                   CriteriaTriggers.f_10579_.m_73935_((ServerPlayer)p_41185_, blockpos);
                }
