--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -77,7 +_,11 @@
                EyeOfEnder eyeofender = new EyeOfEnder(p_41184_, p_41185_.m_20185_(), p_41185_.m_20227_(0.5D), p_41185_.m_20189_());
                eyeofender.m_36972_(itemstack);
                eyeofender.m_36967_(blockpos);
-               p_41184_.m_7967_(eyeofender);
+               // CraftBukkit start
+               if (!p_41184_.m_7967_(eyeofender)) {
+                  return new InteractionResultHolder<>(InteractionResult.FAIL, itemstack);
+               }
+               // CraftBukkit end
                if (p_41185_ instanceof ServerPlayer) {
                   CriteriaTriggers.f_10579_.m_73935_((ServerPlayer)p_41185_, blockpos);
                }
