--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -29,12 +_,28 @@
       if (!CampfireBlock.m_51321_(blockstate) && !CandleBlock.m_152845_(blockstate) && !CandleCakeBlock.m_152910_(blockstate)) {
          blockpos = blockpos.m_142300_(p_41204_.m_43719_());
          if (BaseFireBlock.m_49255_(level, blockpos, p_41204_.m_8125_())) {
+            // CraftBukkit start - fire BlockIgniteEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, p_41204_.m_43723_()).isCancelled()) {
+               if (!p_41204_.m_43723_().m_150110_().f_35937_) {
+                  p_41204_.m_43722_().m_41774_(1);
+               }
+               return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             this.m_41205_(level, blockpos);
             level.m_46597_(blockpos, BaseFireBlock.m_49245_(level, blockpos));
             level.m_142346_(p_41204_.m_43723_(), GameEvent.f_157797_, blockpos);
             flag = true;
          }
       } else {
+         // CraftBukkit start - fire BlockIgniteEvent
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, p_41204_.m_43723_()).isCancelled()) {
+            if (!p_41204_.m_43723_().m_150110_().f_35937_) {
+               p_41204_.m_43722_().m_41774_(1);
+            }
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          this.m_41205_(level, blockpos);
          level.m_46597_(blockpos, blockstate.m_61124_(BlockStateProperties.f_61443_, Boolean.valueOf(true)));
          level.m_142346_(p_41204_.m_43723_(), GameEvent.f_157797_, blockpos);
