--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -11,6 +_,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.event.player.PlayerFishEvent;
+import org.bukkit.craftbukkit.v1_20_R1.CraftEquipmentSlot;
+// CraftBukkit end
+
 public class FishingRodItem extends Item implements Vanishable {
    public FishingRodItem(Item.Properties p_41285_) {
       super(p_41285_);
@@ -29,11 +_,21 @@
          p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11939_, SoundSource.NEUTRAL, 1.0F, 0.4F / (p_41290_.m_213780_().m_188501_() * 0.4F + 0.8F));
          p_41291_.m_146850_(GameEvent.f_223697_);
       } else {
-         p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11941_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_41290_.m_213780_().m_188501_() * 0.4F + 0.8F));
+         // p_41290_.playSound((Player)null, p_41291_.getX(), p_41291_.getY(), p_41291_.getZ(), SoundEvents.FISHING_BOBBER_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_41290_.getRandom().nextFloat() * 0.4F + 0.8F));
          if (!p_41290_.f_46443_) {
             int k = EnchantmentHelper.m_44916_(itemstack);
             int j = EnchantmentHelper.m_44904_(itemstack);
-            p_41290_.m_7967_(new FishingHook(p_41291_, p_41290_, j, k));
+            // CraftBukkit start
+            FishingHook entityfishinghook = new FishingHook(p_41291_, p_41290_, j, k);
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) p_41291_.getBukkitEntity(), null, (org.bukkit.entity.FishHook) entityfishinghook.getBukkitEntity(), CraftEquipmentSlot.getHand(p_41292_), PlayerFishEvent.State.FISHING);
+            p_41290_.getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            if (playerFishEvent.isCancelled()) {
+               p_41291_.f_36083_ = null;
+               return InteractionResultHolder.m_19098_(itemstack);
+            }
+            p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11941_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_41290_.m_213780_().m_188501_() * 0.4F + 0.8F));
+            p_41290_.m_7967_(entityfishinghook);
+            // CraftBukkit end
          }
 
          p_41291_.m_36246_(Stats.f_12982_.m_12902_(this));
@@ -45,5 +_,12 @@
 
    public int m_6473_() {
       return 1;
+   }
+
+    /* ******************** FORGE START ******************** */
+
+    @Override
+   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+      return net.minecraftforge.common.ToolActions.DEFAULT_FISHING_ROD_ACTIONS.contains(toolAction);
    }
 }
