--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -10,6 +_,8 @@
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.SoundCategory;
+import org.bukkit.event.player.PlayerFishEvent;
 
 public class FishingRodItem extends Item implements Vanishable {
    public FishingRodItem(Item.Properties p_41285_) {
@@ -29,11 +_,22 @@
          p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11939_, SoundSource.NEUTRAL, 1.0F, 0.4F / (p_41290_.m_5822_().nextFloat() * 0.4F + 0.8F));
          p_41290_.m_151545_(p_41291_, GameEvent.f_157814_, p_41291_);
       } else {
-         p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11941_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_41290_.m_5822_().nextFloat() * 0.4F + 0.8F));
          if (!p_41290_.f_46443_) {
             int k = EnchantmentHelper.m_44916_(itemstack);
             int j = EnchantmentHelper.m_44904_(itemstack);
             p_41290_.m_7967_(new FishingHook(p_41291_, p_41290_, j, k));
+            // CraftBukkit start
+            FishingHook entityfishinghook = new FishingHook(p_41291_, p_41290_, j, k);
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) p_41291_.getBukkitEntity(), null, (org.bukkit.entity.FishHook) entityfishinghook.getBukkitEntity(), PlayerFishEvent.State.FISHING);
+            p_41290_.getCraftServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled()) {
+               p_41291_.f_36083_ = null;
+               return InteractionResultHolder.m_19098_(itemstack);
+            }
+            p_41290_.m_6263_((Player)null, p_41291_.m_20185_(), p_41291_.m_20186_(), p_41291_.m_20189_(), SoundEvents.f_11941_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_41290_.m_5822_().nextFloat() * 0.4F + 0.8F));
+            p_41290_.m_7967_(entityfishinghook);
+            // CraftBukkit end
          }
 
          p_41291_.m_36246_(Stats.f_12982_.m_12902_(this));
