--- a/net/minecraft/world/item/FlintAndSteelItem.java
+++ b/net/minecraft/world/item/FlintAndSteelItem.java
@@ -30,6 +_,14 @@
       if (!CampfireBlock.m_51321_(blockstate) && !CandleBlock.m_152845_(blockstate) && !CandleCakeBlock.m_152910_(blockstate)) {
          BlockPos blockpos1 = blockpos.m_142300_(p_41297_.m_43719_());
          if (BaseFireBlock.m_49255_(level, blockpos1, p_41297_.m_8125_())) {
+            // CraftBukkit start - Store the clicked block
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+               p_41297_.m_43722_().m_41622_(1, player, (entityhuman1) -> {
+                  entityhuman1.m_21190_(p_41297_.m_43724_());
+               });
+               return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             level.m_5594_(player, blockpos1, SoundEvents.f_11942_, SoundSource.BLOCKS, 1.0F, level.m_5822_().nextFloat() * 0.4F + 0.8F);
             BlockState blockstate1 = BaseFireBlock.m_49245_(level, blockpos1);
             level.m_7731_(blockpos1, blockstate1, 11);
@@ -47,6 +_,14 @@
             return InteractionResult.FAIL;
          }
       } else {
+         // CraftBukkit start - Store the clicked block
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+            p_41297_.m_43722_().m_41622_(1, player, (entityhuman1) -> {
+               entityhuman1.m_21190_(p_41297_.m_43724_());
+            });
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          level.m_5594_(player, blockpos, SoundEvents.f_11942_, SoundSource.BLOCKS, 1.0F, level.m_5822_().nextFloat() * 0.4F + 0.8F);
          level.m_7731_(blockpos, blockstate.m_61124_(BlockStateProperties.f_61443_, Boolean.valueOf(true)), 11);
          level.m_142346_(player, GameEvent.f_157797_, blockpos);
