--- a/net/minecraft/world/item/HangingEntityItem.java
+++ b/net/minecraft/world/item/HangingEntityItem.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.event.hanging.HangingPlaceEvent;
 
 public class HangingEntityItem extends Item {
    private final EntityType<? extends HangingEntity> f_41322_;
@@ -52,6 +_,18 @@
 
          if (hangingentity.m_7088_()) {
             if (!level.f_46443_) {
+               // CraftBukkit start - fire HangingPlaceEvent
+               org.bukkit.entity.Player who = (p_41331_.m_43723_() == null) ? null : (org.bukkit.entity.Player) p_41331_.m_43723_().getBukkitEntity();
+               org.bukkit.block.Block blockClicked = level.getWorld().getBlockAt(blockpos.m_123341_(), blockpos.m_123342_(), blockpos.m_123343_());
+               org.bukkit.block.BlockFace blockFace = org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.notchToBlockFace(direction);
+
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) hangingentity.getBukkitEntity(), who, blockClicked, blockFace, org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asBukkitCopy(itemstack));
+               level.getCraftServer().getPluginManager().callEvent(event);
+
+               if (event.isCancelled()) {
+                  return InteractionResult.FAIL;
+               }
+               // CraftBukkit end
                hangingentity.m_7084_();
                level.m_142346_(player, GameEvent.f_157810_, blockpos);
                level.m_7967_(hangingentity);
