--- a/net/minecraft/world/item/HangingEntityItem.java
+++ b/net/minecraft/world/item/HangingEntityItem.java
@@ -20,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.event.hanging.HangingPlaceEvent;
+// CraftBukkit end
+
 public class HangingEntityItem extends Item {
    private static final Component f_268736_ = Component.m_237115_("painting.random").m_130940_(ChatFormatting.GRAY);
    private final EntityType<? extends HangingEntity> f_41322_;
@@ -64,6 +_,17 @@
 
          if (hangingentity.m_7088_()) {
             if (!level.f_46443_) {
+               // CraftBukkit start - fire HangingPlaceEvent
+               org.bukkit.entity.Player who = (p_41331_.m_43723_() == null) ? null : (org.bukkit.entity.Player) p_41331_.m_43723_().getBukkitEntity();
+               org.bukkit.block.Block blockClicked = level.getWorld().getBlockAt(blockpos.m_123341_(), blockpos.m_123342_(), blockpos.m_123343_());
+               org.bukkit.block.BlockFace blockFace = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.notchToBlockFace(direction);
+               org.bukkit.inventory.EquipmentSlot hand = org.bukkit.craftbukkit.v1_20_R1.CraftEquipmentSlot.getHand(p_41331_.m_43724_());
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) ((HangingEntity) hangingentity).getBukkitEntity(), who, blockClicked, blockFace, hand, org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asBukkitCopy(itemstack));
+               level.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  return InteractionResult.FAIL;
+               }
+               // CraftBukkit end
                hangingentity.m_7084_();
                level.m_220400_(player, GameEvent.f_157810_, hangingentity.m_20182_());
                level.m_7967_(hangingentity);
