--- a/net/minecraft/world/item/ItemCooldowns.java
+++ b/net/minecraft/world/item/ItemCooldowns.java
@@ -7,8 +_,8 @@
 import net.minecraft.util.Mth;
 
 public class ItemCooldowns {
-   private final Map<Item, ItemCooldowns.CooldownInstance> f_41515_ = Maps.newHashMap();
-   private int f_41516_;
+   public final Map<Item, ItemCooldowns.CooldownInstance> f_41515_ = Maps.newHashMap();
+   public int f_41516_;
 
    public boolean m_41519_(Item p_41520_) {
       return this.m_41521_(p_41520_, 0.0F) > 0.0F;
@@ -57,9 +_,9 @@
    protected void m_7432_(Item p_41531_) {
    }
 
-   static class CooldownInstance {
+   public static class CooldownInstance {
       final int f_41533_;
-      final int f_41534_;
+      public final int f_41534_;
 
       CooldownInstance(int p_186358_, int p_186359_) {
          this.f_41533_ = p_186358_;
