--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.event.hanging.HangingPlaceEvent;
 
 public class LeadItem extends Item {
    public LeadItem(Item.Properties p_42828_) {
@@ -44,8 +_,22 @@
          if (mob.m_21524_() == p_42830_) {
             if (leashfenceknotentity == null) {
                leashfenceknotentity = LeashFenceKnotEntity.m_31844_(p_42831_, p_42832_);
+               // CraftBukkit start - fire HangingPlaceEvent
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) leashfenceknotentity.getBukkitEntity(), p_42830_ != null ? (org.bukkit.entity.Player) p_42830_.getBukkitEntity() : null, p_42831_.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF);
+               p_42831_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  leashfenceknotentity.m_146870_();
+                  return InteractionResult.PASS;
+               }
+               // CraftBukkit end
                leashfenceknotentity.m_7084_();
             }
+
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerLeashEntityEvent(mob, leashfenceknotentity, p_42830_).isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
 
             mob.m_21463_(leashfenceknotentity, true);
             flag = true;
