--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -12,6 +_,11 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.CraftEquipmentSlot;
+import org.bukkit.event.hanging.HangingPlaceEvent;
+// CraftBukkit end
+
 public class LeadItem extends Item {
    public LeadItem(Item.Properties p_42828_) {
       super(p_42828_);
@@ -24,7 +_,7 @@
       if (blockstate.m_204336_(BlockTags.f_13039_)) {
          Player player = p_42834_.m_43723_();
          if (!level.f_46443_ && player != null) {
-            m_42829_(player, level, blockpos);
+            bindPlayerMobs(player, level, blockpos, p_42834_.m_43724_()); // CraftBukkit - 1.20.1 Pass hand
          }
 
          return InteractionResult.m_19078_(level.f_46443_);
@@ -33,6 +_,17 @@
       }
    }
 
+   // CatServer start - 1.20.1
+   private static net.minecraft.world.InteractionHand catserver$bindPlayerMobsEnumhand = net.minecraft.world.InteractionHand.MAIN_HAND;
+   public static InteractionResult bindPlayerMobs(Player p_42830_, Level p_42831_, BlockPos p_42832_, net.minecraft.world.InteractionHand enumhand) { // CraftBukkit - 1.20.1 Add EnumHand
+      try {
+         LeadItem.catserver$bindPlayerMobsEnumhand = enumhand;
+         return LeadItem.m_42829_(p_42830_, p_42831_, p_42832_);
+      } finally {
+         LeadItem.catserver$bindPlayerMobsEnumhand = net.minecraft.world.InteractionHand.MAIN_HAND;
+      }
+   }
+
    public static InteractionResult m_42829_(Player p_42830_, Level p_42831_, BlockPos p_42832_) {
       LeashFenceKnotEntity leashfenceknotentity = null;
       boolean flag = false;
@@ -45,8 +_,22 @@
          if (mob.m_21524_() == p_42830_) {
             if (leashfenceknotentity == null) {
                leashfenceknotentity = LeashFenceKnotEntity.m_31844_(p_42831_, p_42832_);
+               // CraftBukkit start - 1.20.1 fire HangingPlaceEvent
+               org.bukkit.inventory.EquipmentSlot hand = CraftEquipmentSlot.getHand(LeadItem.catserver$bindPlayerMobsEnumhand);
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) leashfenceknotentity.getBukkitEntity(), p_42830_ != null ? (org.bukkit.entity.Player) p_42830_.getBukkitEntity() : null, p_42831_.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF, hand);
+               p_42831_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  leashfenceknotentity.m_146870_();
+                  return InteractionResult.PASS;
+               }
+               // CraftBukkit end
                leashfenceknotentity.m_7084_();
             }
+            // CraftBukkit start - 1.20.1
+            if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerLeashEntityEvent(mob, leashfenceknotentity, p_42830_, LeadItem.catserver$bindPlayerMobsEnumhand).isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
 
             mob.m_21463_(leashfenceknotentity, true);
             flag = true;
@@ -59,4 +_,5 @@
 
       return flag ? InteractionResult.SUCCESS : InteractionResult.PASS;
    }
+   // CatServer end
 }
