--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -31,6 +_,8 @@
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.material.MaterialColor;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import org.bukkit.Bukkit;
+import org.bukkit.event.server.MapInitializeEvent;
 
 public class MapItem extends ComplexItem {
    public static final int f_151104_ = 128;
@@ -55,20 +_,34 @@
 
    @Nullable
    public static MapItemSavedData m_42853_(ItemStack p_42854_, Level p_42855_) {
-      Integer integer = m_151131_(p_42854_);
-      return m_151128_(integer, p_42855_);
+      // Forge: Add instance method so that mods can override
+      Item map = p_42854_.m_41720_();
+      if(map instanceof MapItem) {
+         return ((MapItem)map).getCustomMapData(p_42854_, p_42855_);
+      }
+      return null;
+   }
+
+   @Nullable
+   protected MapItemSavedData getCustomMapData(ItemStack p_42910_, Level p_42911_) {
+      Integer integer = m_151131_(p_42910_);
+      return m_151128_(integer, p_42911_);
    }
 
    @Nullable
    public static Integer m_151131_(ItemStack p_151132_) {
       CompoundTag compoundtag = p_151132_.m_41783_();
-      return compoundtag != null && compoundtag.m_128425_("map", 99) ? compoundtag.m_128451_("map") : null;
+      return compoundtag != null && compoundtag.m_128425_("map", 99) ? compoundtag.m_128451_("map") : -1; // CraftBukkit - make new maps for no tag
    }
 
-   private static int m_151120_(Level p_151121_, int p_151122_, int p_151123_, int p_151124_, boolean p_151125_, boolean p_151126_, ResourceKey<Level> p_151127_) {
+   public static int m_151120_(Level p_151121_, int p_151122_, int p_151123_, int p_151124_, boolean p_151125_, boolean p_151126_, ResourceKey<Level> p_151127_) {
       MapItemSavedData mapitemsaveddata = MapItemSavedData.m_164780_((double)p_151122_, (double)p_151123_, (byte)p_151124_, p_151125_, p_151126_, p_151127_);
       int i = p_151121_.m_7354_();
       p_151121_.m_142325_(m_42848_(i), mapitemsaveddata);
+      // CraftBukkit start
+      MapInitializeEvent event = new MapInitializeEvent(mapitemsaveddata.mapView);
+      Bukkit.getServer().getPluginManager().callEvent(event);
+      // CraftBukkit end
       return i;
    }
 
