--- a/net/minecraft/world/item/MinecartItem.java
+++ b/net/minecraft/world/item/MinecartItem.java
@@ -15,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class MinecartItem extends Item {
    private static final DispenseItemBehavior f_42934_ = new DefaultDispenseItemBehavior() {
@@ -28,7 +_,7 @@
          double d2 = p_42949_.m_7094_() + (double)direction.m_122431_() * 1.125D;
          BlockPos blockpos = p_42949_.m_7961_().m_142300_(direction);
          BlockState blockstate = level.m_8055_(blockpos);
-         RailShape railshape = blockstate.m_60734_() instanceof BaseRailBlock ? blockstate.m_61143_(((BaseRailBlock)blockstate.m_60734_()).m_7978_()) : RailShape.NORTH_SOUTH;
+         RailShape railshape = blockstate.m_60734_() instanceof BaseRailBlock ? ((BaseRailBlock)blockstate.m_60734_()).getRailDirection(blockstate, level, blockpos, null) : RailShape.NORTH_SOUTH;
          double d3;
          if (blockstate.m_204336_(BlockTags.f_13034_)) {
             if (railshape.m_61745_()) {
@@ -50,13 +_,38 @@
             }
          }
 
-         AbstractMinecart abstractminecart = AbstractMinecart.m_38119_(level, d0, d1 + d3, d2, ((MinecartItem)p_42950_.m_41720_()).f_42935_);
+         //AbstractMinecart abstractminecart = AbstractMinecart.createMinecart(level, d0, d1 + d3, d2, ((MinecartItem)p_42950_.getItem()).type);
+         // CraftBukkit start
+         ItemStack itemstack1 = p_42950_.m_41620_(1);
+         org.bukkit.block.Block block2 = level.getWorld().getBlockAt(p_42949_.m_7961_().m_123341_(), p_42949_.m_7961_().m_123342_(), p_42949_.m_7961_().m_123343_());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+         BlockDispenseEvent event = new BlockDispenseEvent(block2, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2));
+         if (!DispenserBlock.eventFired) {
+            level.getCraftServer().getPluginManager().callEvent(event);
+         }
+
+         if (event.isCancelled()) {
+            p_42950_.m_41769_(1);
+            return p_42950_;
+         }
+
+         if (!event.getItem().equals(craftItem)) {
+            p_42950_.m_41769_(1);
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.f_52661_.get(eventStack.m_41720_());
+            if (idispensebehavior != DispenseItemBehavior.f_123393_ && idispensebehavior != this) {
+               idispensebehavior.m_6115_(p_42949_, eventStack);
+               return p_42950_;
+            }
+         }
+         itemstack1 = CraftItemStack.asNMSCopy(event.getItem());
+         AbstractMinecart abstractminecart = AbstractMinecart.m_38119_(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), ((MinecartItem) itemstack1.m_41720_()).f_42935_);
+
          if (p_42950_.m_41788_()) {
             abstractminecart.m_6593_(p_42950_.m_41786_());
          }
-
-         level.m_7967_(abstractminecart);
-         p_42950_.m_41774_(1);
+         if (!level.m_7967_(abstractminecart)) p_42950_.m_41769_(1);
          return p_42950_;
       }
 
@@ -81,7 +_,7 @@
       } else {
          ItemStack itemstack = p_42943_.m_43722_();
          if (!level.f_46443_) {
-            RailShape railshape = blockstate.m_60734_() instanceof BaseRailBlock ? blockstate.m_61143_(((BaseRailBlock)blockstate.m_60734_()).m_7978_()) : RailShape.NORTH_SOUTH;
+            RailShape railshape = blockstate.m_60734_() instanceof BaseRailBlock ? ((BaseRailBlock)blockstate.m_60734_()).getRailDirection(blockstate, level, blockpos, null) : RailShape.NORTH_SOUTH;
             double d0 = 0.0D;
             if (railshape.m_61745_()) {
                d0 = 0.5D;
@@ -92,7 +_,13 @@
                abstractminecart.m_6593_(itemstack.m_41786_());
             }
 
-            level.m_7967_(abstractminecart);
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityPlaceEvent(p_42943_, abstractminecart).isCancelled()) {
+               return InteractionResult.FAIL;
+            }
+            // CraftBukkit end
+            if (!level.m_7967_(abstractminecart)) return InteractionResult.PASS; // CraftBukkit
+
             level.m_142346_(p_42943_.m_43723_(), GameEvent.f_157810_, blockpos);
          }
 
