--- a/net/minecraft/world/item/PlayerHeadItem.java
+++ b/net/minecraft/world/item/PlayerHeadItem.java
@@ -45,6 +_,16 @@
          SkullBlockEntity.m_155738_(gameprofile, (p_151177_) -> {
             p_151179_.m_128365_("SkullOwner", NbtUtils.m_129230_(new CompoundTag(), p_151177_));
          });
+         // CraftBukkit start
+      } else {
+         net.minecraft.nbt.ListTag textures = p_151179_.m_128469_("SkullOwner").m_128469_("Properties").m_128437_("textures", 10); // Safe due to method contracts
+         for (int i = 0; i < textures.size(); i++) {
+            if (textures.get(i) instanceof CompoundTag && !((CompoundTag) textures.get(i)).m_128425_("Signature", 8) && ((CompoundTag) textures.get(i)).m_128461_("Value").trim().isEmpty()) {
+               p_151179_.m_128473_("SkullOwner");
+               break;
+            }
+         }
+         // CraftBukkit end
       }
 
    }
