--- a/net/minecraft/world/item/PotionItem.java
+++ b/net/minecraft/world/item/PotionItem.java
@@ -50,7 +_,7 @@
             if (mobeffectinstance.m_19544_().m_8093_()) {
                mobeffectinstance.m_19544_().m_19461_(player, player, p_42986_, mobeffectinstance.m_19564_(), 1.0D);
             } else {
-               p_42986_.m_7292_(new MobEffectInstance(mobeffectinstance));
+               p_42986_.addEffect(new MobEffectInstance(mobeffectinstance), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
             }
          }
       }
