--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -9,6 +_,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SignItem extends StandingAndWallBlockItem {
+
+   public static BlockPos openSign; // CraftBukkit
+
    public SignItem(Item.Properties p_43126_, Block p_43127_, Block p_43128_) {
       super(p_43127_, p_43128_, p_43126_);
    }
@@ -17,6 +_,11 @@
       boolean flag = super.m_7274_(p_43130_, p_43131_, p_43132_, p_43133_, p_43134_);
       if (!p_43131_.f_46443_ && !flag && p_43132_ != null) {
          p_43132_.m_7739_((SignBlockEntity)p_43131_.m_7702_(p_43130_));
+         // FoxServer TODO
+         // CraftBukkit start - SPIGOT-4678
+         // entityhuman.openTextEdit((TileEntitySign) world.getBlockEntity(blockposition));
+         SignItem.openSign = p_43130_;
+         // CraftBukkit end
       }
 
       return flag;
