--- a/net/minecraft/world/item/SnowballItem.java
+++ b/net/minecraft/world/item/SnowballItem.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.Snowball;
 import net.minecraft.world.level.Level;
+import org.bukkit.SoundCategory;
 
 public class SnowballItem extends Item {
    public SnowballItem(Item.Properties p_43140_) {
@@ -16,18 +_,23 @@
 
    public InteractionResultHolder<ItemStack> m_7203_(Level p_43142_, Player p_43143_, InteractionHand p_43144_) {
       ItemStack itemstack = p_43143_.m_21120_(p_43144_);
-      p_43142_.m_6263_((Player)null, p_43143_.m_20185_(), p_43143_.m_20186_(), p_43143_.m_20189_(), SoundEvents.f_12473_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_43142_.m_5822_().nextFloat() * 0.4F + 0.8F));
       if (!p_43142_.f_46443_) {
          Snowball snowball = new Snowball(p_43142_, p_43143_);
          snowball.m_37446_(itemstack);
          snowball.m_37251_(p_43143_, p_43143_.m_146909_(), p_43143_.m_146908_(), 0.0F, 1.5F, 1.0F);
          p_43142_.m_7967_(snowball);
+         if (p_43142_.m_7967_(snowball)) {
+            if (!p_43143_.m_150110_().f_35937_) {
+               itemstack.m_41774_(1);
+            }
+
+            p_43142_.m_6263_((Player) null, p_43143_.m_20185_(), p_43143_.m_20186_(), p_43143_.m_20189_(), SoundEvents.f_12473_, SoundSource.NEUTRAL, 0.5F, 0.4F / (p_43142_.m_5822_().nextFloat() * 0.4F + 0.8F));
+         } else if (p_43143_ instanceof net.minecraft.server.level.ServerPlayer) {
+            ((net.minecraft.server.level.ServerPlayer) p_43143_).getBukkitEntity().updateInventory();
+         }
       }
 
       p_43143_.m_36246_(Stats.f_12982_.m_12902_(this));
-      if (!p_43143_.m_150110_().f_35937_) {
-         itemstack.m_41774_(1);
-      }
 
       return InteractionResultHolder.m_19092_(itemstack, p_43142_.m_5776_());
    }
