--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -10,6 +_,13 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.shapes.CollisionContext;
 
+// CraftBukkit start
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.block.data.CraftBlockData;
+import org.bukkit.event.block.BlockCanBuildEvent;
+// CraftBukkit end
+
 public class StandingAndWallBlockItem extends BlockItem {
    protected final Block f_43246_;
    private final Direction f_244386_;
@@ -41,11 +_,28 @@
          }
       }
 
-      return blockstate1 != null && levelreader.m_45752_(blockstate1, blockpos, CollisionContext.m_82749_()) ? blockstate1 : null;
+      // CraftBukkit start - 1.20.1
+      if (blockstate1 != null) {
+         boolean defaultReturn = levelreader.m_45752_(blockstate1, blockpos, CollisionContext.m_82749_());
+         org.bukkit.entity.Player player = (p_43255_.m_43723_() instanceof ServerPlayer) ? (org.bukkit.entity.Player) p_43255_.m_43723_().getBukkitEntity() : null;
+
+         BlockCanBuildEvent event = new BlockCanBuildEvent(CraftBlock.at((net.minecraft.world.level.Level)levelreader, blockpos), player, CraftBlockData.fromData(blockstate1), defaultReturn);
+         p_43255_.m_43725_().getCraftServer().getPluginManager().callEvent(event);
+
+         return (event.isBuildable()) ? blockstate1 : null;
+      } else {
+         return null;
+      }
+      // CraftBukkit end
    }
 
    public void m_6192_(Map<Block, Item> p_43252_, Item p_43253_) {
       super.m_6192_(p_43252_, p_43253_);
       p_43252_.put(this.f_43246_, p_43253_);
+   }
+
+   public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+      super.removeFromBlockToItemMap(blockToItemMap, itemIn);
+      blockToItemMap.remove(this.f_43246_);
    }
 }
