--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -59,9 +_,6 @@
             int j = EnchantmentHelper.m_44932_(p_43394_);
             if (j <= 0 || player.m_20070_()) {
                if (!p_43395_.f_46443_) {
-                  p_43394_.m_41622_(1, player, (p_43388_) -> {
-                     p_43388_.m_21190_(p_43396_.m_7655_());
-                  });
                   if (j == 0) {
                      ThrownTrident throwntrident = new ThrownTrident(p_43395_, player, p_43394_);
                      throwntrident.m_37251_(player, player.m_146909_(), player.m_146908_(), 0.0F, 2.5F + (float)j * 0.5F, 1.0F);
@@ -69,16 +_,37 @@
                         throwntrident.f_36705_ = AbstractArrow.Pickup.CREATIVE_ONLY;
                      }
 
-                     p_43395_.m_7967_(throwntrident);
+                     // p_43395_.addFreshEntity(throwntrident);
+                     // CraftBukkit start
+                     if (!p_43395_.m_7967_(throwntrident)) {
+                        if (player instanceof net.minecraft.server.level.ServerPlayer) {
+                           ((net.minecraft.server.level.ServerPlayer) player).getBukkitEntity().updateInventory();
+                        }
+                        return;
+                     }
+
+                     p_43394_.m_41622_(1, player, (entityhuman1) -> {
+                        entityhuman1.m_21190_(p_43396_.m_7655_());
+                     });
+                     throwntrident.f_37555_ = p_43394_.m_41777_(); // SPIGOT-4511 update since damage call moved
+                     // CraftBukkit end
                      p_43395_.m_6269_((Player)null, throwntrident, SoundEvents.f_12520_, SoundSource.PLAYERS, 1.0F, 1.0F);
                      if (!player.m_150110_().f_35937_) {
                         player.m_150109_().m_36057_(p_43394_);
                      }
+                  } else {
+                     p_43394_.m_41622_(1, player, (p_43388_) -> {
+                        p_43388_.m_21190_(p_43396_.m_7655_());
+                     });
                   }
                }
 
                player.m_36246_(Stats.f_12982_.m_12902_(this));
                if (j > 0) {
+                  // CraftBukkit start
+                  org.bukkit.event.player.PlayerRiptideEvent event = new org.bukkit.event.player.PlayerRiptideEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(p_43394_));
+                  event.getPlayer().getServer().getPluginManager().callEvent(event);
+                  // CraftBukkit end
                   float f7 = player.m_146908_();
                   float f = player.m_146909_();
                   float f1 = -Mth.m_14031_(f7 * ((float)Math.PI / 180F)) * Mth.m_14089_(f * ((float)Math.PI / 180F));
