--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -54,6 +_,17 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
+
 public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob {
    private static final EntityDataAccessor<Integer> f_31420_ = SynchedEntityData.m_135353_(WitherBoss.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Integer> f_31433_ = SynchedEntityData.m_135353_(WitherBoss.class, EntityDataSerializers.f_135028_);
@@ -68,7 +_,7 @@
    private final int[] f_31427_ = new int[2];
    private final int[] f_31428_ = new int[2];
    private int f_31429_;
-   private final ServerBossEvent f_31430_ = (ServerBossEvent)(new ServerBossEvent(this.m_5446_(), BossEvent.BossBarColor.PURPLE, BossEvent.BossBarOverlay.PROGRESS)).m_7003_(true);
+   public final ServerBossEvent f_31430_ = (ServerBossEvent)(new ServerBossEvent(this.m_5446_(), BossEvent.BossBarColor.PURPLE, BossEvent.BossBarOverlay.PROGRESS)).m_7003_(true); // CatServer - private -> public
    private static final Predicate<LivingEntity> f_31431_ = (p_31504_) -> {
       return p_31504_.m_6336_() != MobType.f_21641_ && p_31504_.m_5789_();
    };
@@ -219,15 +_,39 @@
          int k1 = this.m_31502_() - 1;
          this.f_31430_.m_142711_(1.0F - (float)k1 / 220.0F);
          if (k1 <= 0) {
-            this.m_9236_().m_255391_(this, this.m_20185_(), this.m_20188_(), this.m_20189_(), 7.0F, false, Level.ExplosionInteraction.MOB);
+            // CraftBukkit start
+            // this.level().explode(this, this.getX(), this.getEyeY(), this.getZ(), 7.0F, false, Level.ExplosionInteraction.MOB);
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 7.0F, false);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               this.m_9236_().m_255391_(this, this.m_20185_(), this.m_20188_(), this.m_20189_(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB);
+            }
+            // CraftBukkit end
             if (!this.m_20067_()) {
-               this.m_9236_().m_6798_(1023, this.m_20183_(), 0);
+               // CraftBukkit start - Use relative location for far away sounds
+               // this.level().globalLevelEvent(1023, this.blockPosition(), 0);
+               int viewDistance = ((ServerLevel) this.m_9236_()).getCraftServer().getViewDistance() * 16;
+               for (ServerPlayer player : (List<ServerPlayer>) MinecraftServer.getServer().m_6846_().f_11196_) {
+                  double deltaX = this.m_20185_() - player.m_20185_();
+                  double deltaZ = this.m_20189_() - player.m_20189_();
+                  double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                  if ( this.m_9236_().spigotConfig.witherSpawnSoundRadius > 0 && distanceSquared > this.m_9236_().spigotConfig.witherSpawnSoundRadius * this.m_9236_().spigotConfig.witherSpawnSoundRadius ) continue; // Spigot
+                  if (distanceSquared > viewDistance * viewDistance) {
+                     double deltaLength = Math.sqrt(distanceSquared);
+                     double relativeX = player.m_20185_() + (deltaX / deltaLength) * viewDistance;
+                     double relativeZ = player.m_20189_() + (deltaZ / deltaLength) * viewDistance;
+                     player.f_8906_.m_9829_(new ClientboundLevelEventPacket(1023, new BlockPos((int) relativeX, (int) this.m_20186_(), (int) relativeZ), 0, true));
+                  } else {
+                     player.f_8906_.m_9829_(new ClientboundLevelEventPacket(1023, this.m_20183_(), 0, true));
+                  }
+               }
+               // CraftBukkit end
             }
          }
 
          this.m_31510_(k1);
          if (this.f_19797_ % 10 == 0) {
-            this.m_5634_(10.0F);
+            this.heal(10.0F, EntityRegainHealthEvent.RegainReason.WITHER_SPAWN); // CraftBukkit
          }
 
       } else {
@@ -265,6 +_,7 @@
                   List<LivingEntity> list = this.m_9236_().m_45971_(LivingEntity.class, f_31432_, this, this.m_20191_().m_82377_(20.0D, 8.0D, 20.0D));
                   if (!list.isEmpty()) {
                      LivingEntity livingentity1 = list.get(this.f_19796_.m_188503_(list.size()));
+                     if (CraftEventFactory.callEntityTargetLivingEvent(this, livingentity1, EntityTargetEvent.TargetReason.CLOSEST_ENTITY).isCancelled()) continue; // CraftBukkit
                      this.m_31454_(i, livingentity1.m_19879_());
                   }
                }
@@ -279,7 +_,7 @@
 
          if (this.f_31429_ > 0) {
             --this.f_31429_;
-            if (this.f_31429_ == 0 && this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (this.f_31429_ == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this)) {
                int j1 = Mth.m_14107_(this.m_20186_());
                int i2 = Mth.m_14107_(this.m_20185_());
                int j2 = Mth.m_14107_(this.m_20189_());
@@ -293,7 +_,12 @@
                         int i1 = j2 + k2;
                         BlockPos blockpos = new BlockPos(l2, l, i1);
                         BlockState blockstate = this.m_9236_().m_8055_(blockpos);
-                        if (m_31491_(blockstate)) {
+                        if (blockstate.canEntityDestroy(this.m_9236_(), blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, blockstate)) {
+                           // CraftBukkit start
+                           if (!CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.f_50016_.m_49966_())) {
+                              continue;
+                           }
+                           // CraftBukkit end
                            flag = this.m_9236_().m_46953_(blockpos, true, this) || flag;
                         }
                      }
@@ -307,13 +_,17 @@
          }
 
          if (this.f_19797_ % 20 == 0) {
-            this.m_5634_(1.0F);
+            this.heal(1.0F, EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
          }
 
          this.f_31430_.m_142711_(this.m_21223_() / this.m_21233_());
       }
    }
 
+   /**
+    * @deprecated Forge: Use {@link BlockState#canEntityDestroy(net.minecraft.world.level.BlockGetter, BlockPos, Entity)} instead.
+    */
+   @Deprecated
    public static boolean m_31491_(BlockState p_31492_) {
       return !p_31492_.m_60795_() && !p_31492_.m_204336_(BlockTags.f_13070_);
    }
