--- a/net/minecraft/server/rcon/thread/RconClient.java
+++ b/net/minecraft/server/rcon/thread/RconClient.java
@@ -9,6 +_,7 @@
 import java.nio.charset.StandardCharsets;
 import java.util.Locale;
 import net.minecraft.server.ServerInterface;
+import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.server.rcon.PktUtils;
 import org.slf4j.Logger;
 
@@ -23,11 +_,14 @@
    private final Socket f_11581_;
    private final byte[] f_11582_ = new byte[1460];
    private final String f_11583_;
-   private final ServerInterface f_11584_;
+   // CraftBukkit start - 1.20.1
+   private final ServerInterface f_11584_; // CatServer - recover
+   // private final net.minecraft.server.rcon.RconConsoleSource rconConsoleSource; // CatServer - recover
+   // CraftBukkit end
 
    RconClient(ServerInterface p_11587_, String p_11588_, Socket p_11589_) {
       super("RCON Client " + p_11589_.getInetAddress());
-      this.f_11584_ = p_11587_;
+      this.f_11584_ = p_11587_; // CraftBukkit // CatServer - recover
       this.f_11581_ = p_11589_;
 
       try {
@@ -37,6 +_,9 @@
       }
 
       this.f_11583_ = p_11588_;
+      // this.rconConsoleSource = new net.minecraft.server.rcon.RconConsoleSource(this.serverInterface, socket.getRemoteSocketAddress()); // CraftBukkit - 1.20.1 // CatServer - remove
+      ((DedicatedServer) this.f_11584_).f_139602_ = new net.minecraft.server.rcon.RconConsoleSource((DedicatedServer) this.f_11584_); // CatServer - 1.20.1
+      ((DedicatedServer) this.f_11584_).f_139602_.socketAddress = p_11589_.getRemoteSocketAddress(); // CatServer - 1.20.1
    }
 
    public void run() {
@@ -66,7 +_,7 @@
                         String s1 = PktUtils.m_11488_(this.f_11582_, j, i);
 
                         try {
-                           this.m_11594_(l, this.f_11584_.m_7261_(s1));
+                           this.m_11594_(l, this.f_11584_.m_7261_(s1)); // CraftBukkit - 1.20.1 // CatServer - recover
                         } catch (Exception exception) {
                            this.m_11594_(l, "Error executing: " + s1 + " (" + exception.getMessage() + ")");
                         }
@@ -123,13 +_,14 @@
    }
 
    private void m_11594_(int p_11595_, String p_11596_) throws IOException {
-      int i = p_11596_.length();
-
+      byte[] whole = p_11596_.getBytes(StandardCharsets.UTF_8);
+      int i = whole.length;
+      int start = 0;
       do {
          int j = 4096 <= i ? 4096 : i;
-         this.m_11590_(p_11595_, 0, p_11596_.substring(0, j));
-         p_11596_ = p_11596_.substring(j);
-         i = p_11596_.length();
+         this.m_11590_(p_11595_, 0, new String(java.util.Arrays.copyOfRange(whole, start, j+start), StandardCharsets.UTF_8));
+         i -= j;
+         start += j;
       } while(0 != i);
 
    }
