--- ../src-base/minecraft/net/minecraft/client/renderer/tileentity/TileEntitySpecialRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/tileentity/TileEntitySpecialRenderer.java
@@ -77,6 +77,8 @@
         return false;
     }
 
+    public void renderTileEntityFast(T te, double x, double y, double z, float partialTicks, int destroyStage, float partial, net.minecraft.client.renderer.BufferBuilder buffer) {}
+
     protected void drawNameplate(T te, String str, double x, double y, double z, int maxDistance)
     {
         Entity entity = this.rendererDispatcher.entity;
