--- ../src-base/minecraft/ibxm/LogTable.java
+++ ../src-work/minecraft/ibxm/LogTable.java
@@ -77,7 +77,6 @@
 		y = eval_table( exp_2_table, x & IBXM.FP_MASK ) << IBXM.FP_SHIFT;
 		return y >> IBXM.FP_SHIFT - ( x >> IBXM.FP_SHIFT );
 	}
-
 	private static int eval_table( int[] table, int x ) {
 		int table_idx, table_frac, c, m, y;
 		table_idx = x >> INTERP_SHIFT;
@@ -88,4 +87,3 @@
 		return y >> 15 - IBXM.FP_SHIFT;
 	}
 }
-
