--- a/net/minecraft/commands/arguments/blocks/BlockStateParser.java
+++ b/net/minecraft/commands/arguments/blocks/BlockStateParser.java
@@ -64,7 +_,7 @@
    private final StringReader f_116749_;
    private final boolean f_116750_;
    private final boolean f_234671_;
-   private final Map<Property<?>, Comparable<?>> f_116751_ = Maps.newHashMap();
+   private final Map<Property<?>, Comparable<?>> f_116751_ = Maps.newLinkedHashMap(); // CraftBukkit - stable
    private final Map<String, String> f_116752_ = Maps.newHashMap();
    private ResourceLocation f_116753_ = new ResourceLocation("");
    @Nullable
