--- ../src-base/minecraft/net/minecraft/tileentity/MobSpawnerBaseLogic.java
+++ ../src-work/minecraft/net/minecraft/tileentity/MobSpawnerBaseLogic.java
@@ -20,7 +20,7 @@
 
 public abstract class MobSpawnerBaseLogic
 {
-    private int spawnDelay = 20;
+    public int spawnDelay = 20;
     private final List<WeightedSpawnerEntity> minecartToSpawn = Lists.<WeightedSpawnerEntity>newArrayList();
     private WeightedSpawnerEntity randomEntity = new WeightedSpawnerEntity();
     private double mobRotation;
@@ -33,7 +33,7 @@
     private int activatingRangeFromPlayer = 16;
     private int spawnRange = 4;
 
-    private String getEntityNameToSpawn()
+    public String getEntityNameToSpawn()
     {
         return this.randomEntity.getNbt().getString("id");
     }
