--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityBanner.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityBanner.java
@@ -16,8 +16,8 @@
 
 public class TileEntityBanner extends TileEntity
 {
-    private int baseColor;
-    private NBTTagList patterns;
+    public int baseColor;
+    public NBTTagList patterns;
     private boolean patternDataSet;
     private List<TileEntityBanner.EnumBannerPattern> patternList;
     private List<EnumDyeColor> colorList;
