--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityCommandBlock.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityCommandBlock.java
@@ -25,6 +25,9 @@
     private boolean sendToClient;
     private final CommandBlockBaseLogic commandBlockLogic = new CommandBlockBaseLogic()
     {
+        {
+            sender = new org.bukkit.craftbukkit.command.CraftBlockCommandSender(this); // CraftBukkit - add sender
+        }
         public BlockPos getPosition()
         {
             return TileEntityCommandBlock.this.pos;
