--- ../src-base/minecraft/net/minecraft/tileentity/TileEntitySkull.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntitySkull.java
@@ -22,7 +22,7 @@
 public class TileEntitySkull extends TileEntity implements ITickable
 {
     private int skullType;
-    private int skullRotation;
+    public int skullRotation;
     private GameProfile playerProfile;
     private int dragonAnimatedTicks;
     private boolean dragonAnimated;
