--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityHopper.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityHopper.java
@@ -27,12 +27,49 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import luohuayu.CatServer.tileentity.CBTileEntityLockableLoot;
 
-public class TileEntityHopper extends TileEntityLockableLoot implements IHopper, ITickable
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+
+//CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.inventory.InventoryMoveItemEvent;
+import org.bukkit.event.inventory.InventoryPickupItemEvent;
+import org.bukkit.inventory.Inventory;
+// CraftBukkit end
+
+public class TileEntityHopper extends CBTileEntityLockableLoot implements IHopper, ITickable
 {
     private ItemStack[] inventory = new ItemStack[5];
     private String customName;
     private int transferCooldown = -1;
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+    
+    public ItemStack[] getContents() {
+    	return this.inventory;
+    }
+    
+    public void onOpen(CraftHumanEntity who) {
+    	transaction.add(who);
+    }
+    	
+    public void onClose(CraftHumanEntity who) {
+    	transaction.remove(who);
+    }
+    
+    public List<HumanEntity> getViewers() {
+    	return transaction;
+    }
+    
+    public void setMaxStackSize(int size) {
+    	maxStack = size;
+    }
+    // CraftBukkit end
 
     public static void registerFixesHopper(DataFixer fixer)
     {
