--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityEndGateway.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityEndGateway.java
@@ -30,8 +30,8 @@
     private static final Logger LOG = LogManager.getLogger();
     private long age;
     private int teleportCooldown;
-    private BlockPos exitPortal;
-    private boolean exactTeleport;
+    public BlockPos exitPortal;
+    public boolean exactTeleport;
 
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
