--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
@@ -23,6 +23,7 @@
 import net.minecraft.item.crafting.FurnaceRecipes;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ITickable;
 import net.minecraft.util.datafix.DataFixer;
@@ -31,9 +32,48 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import luohuayu.CatServer.tileentity.CBTileEntityLockable;
+import java.util.List;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
-public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory
+//CraftBukkit start
+import java.util.List;
+
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.inventory.FurnaceBurnEvent;
+import org.bukkit.event.inventory.FurnaceSmeltEvent;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+// CraftBukkit end
+
+public class TileEntityFurnace extends CBTileEntityLockable implements ITickable, ISidedInventory
 {
+	// CraftBukkit start - add fields and methods
+	private int lastTick = MinecraftServer.currentTick;
+	private int maxStack = MAX_STACK;
+	public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+	
+	public ItemStack[] getContents() {
+		return this.furnaceItemStacks;
+	}
+	
+	public void onOpen(CraftHumanEntity who) {
+		transaction.add(who);
+	}
+		
+	public void onClose(CraftHumanEntity who) {
+		transaction.remove(who);
+	}
+	
+	public List<HumanEntity> getViewers() {
+		return transaction;
+	}
+	
+	public void setMaxStackSize(int size) {
+		maxStack = size;
+	}
+	// CraftBukkit end
     private static final int[] SLOTS_TOP = new int[] {0};
     private static final int[] SLOTS_BOTTOM = new int[] {2, 1};
     private static final int[] SLOTS_SIDES = new int[] {1};
