--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityCommandBlock.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityCommandBlock.java
@@ -134,7 +134,7 @@
         boolean flag = this.auto;
         this.auto = autoIn;
 
-        if (!flag && autoIn && !this.powered && this.world != null && this.getMode() != TileEntityCommandBlock.Mode.SEQUENCE)
+        if (!flag && autoIn && !this.powered && this.world != null && this.getMode() != Mode.SEQUENCE)
         {
             Block block = this.getBlockType();
 
@@ -183,21 +183,21 @@
         this.sendToClient = p_184252_1_;
     }
 
-    public TileEntityCommandBlock.Mode getMode()
+    public Mode getMode()
     {
         Block block = this.getBlockType();
 
         if (block == Blocks.COMMAND_BLOCK)
         {
-            return TileEntityCommandBlock.Mode.REDSTONE;
+            return Mode.REDSTONE;
         }
         else if (block == Blocks.REPEATING_COMMAND_BLOCK)
         {
-            return TileEntityCommandBlock.Mode.AUTO;
+            return Mode.AUTO;
         }
         else
         {
-            return block == Blocks.CHAIN_COMMAND_BLOCK ? TileEntityCommandBlock.Mode.SEQUENCE : TileEntityCommandBlock.Mode.REDSTONE;
+            return block == Blocks.CHAIN_COMMAND_BLOCK ? Mode.SEQUENCE : Mode.REDSTONE;
         }
     }
 
