--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityChest.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityChest.java
@@ -1,6 +1,12 @@
 package net.minecraft.tileentity;
 
+import java.util.List;
+
 import javax.annotation.Nullable;
+
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.entity.player.EntityPlayer;
@@ -22,8 +28,9 @@
 import net.minecraft.util.datafix.walkers.ItemStackDataLists;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
+import luohuayu.CatServer.tileentity.CBTileEntityLockableLoot;
 
-public class TileEntityChest extends TileEntityLockableLoot implements ITickable, IInventory
+public class TileEntityChest extends CBTileEntityLockableLoot implements ITickable, IInventory
 {
     private ItemStack[] chestContents = new ItemStack[27];
     public boolean adjacentChestChecked;
@@ -37,7 +44,30 @@
     private int ticksSinceSync;
     private BlockChest.Type cachedChestType;
     private String customName;
-
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+    
+    public ItemStack[] getContents() {
+    	return this.chestContents;
+    }
+    
+    public void onOpen(CraftHumanEntity who) {
+    	transaction.add(who);
+    }
+    
+    public void onClose(CraftHumanEntity who) {
+    	transaction.remove(who);
+    }
+    
+    public List<HumanEntity> getViewers() {
+    	return transaction;
+    }
+    
+    public void setMaxStackSize(int size) {
+    	maxStack = size;
+    }
+    // CraftBukkit end
     public TileEntityChest()
     {
     }
