--- a/net/minecraft/world/level/block/AbstractCandleBlock.java
+++ b/net/minecraft/world/level/block/AbstractCandleBlock.java
@@ -35,6 +_,11 @@
 
    public void m_5581_(Level p_151905_, BlockState p_151906_, BlockHitResult p_151907_, Projectile p_151908_) {
       if (!p_151905_.f_46443_ && p_151908_.m_6060_() && this.m_142595_(p_151906_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(p_151905_, p_151907_.m_82425_(), p_151908_).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          m_151918_(p_151905_, p_151906_, p_151907_.m_82425_(), true);
       }
 
