--- a/net/minecraft/world/level/block/BambooBlock.java
+++ b/net/minecraft/world/level/block/BambooBlock.java
@@ -7,7 +_,6 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.SwordItem;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
@@ -26,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class BambooBlock extends Block implements BonemealableBlock {
+public class BambooBlock extends Block implements BonemealableBlock, net.minecraftforge.common.IPlantable {
    protected static final float f_152092_ = 3.0F;
    protected static final float f_152093_ = 5.0F;
    protected static final float f_152094_ = 1.5F;
@@ -114,10 +_,12 @@
 
    public void m_7455_(BlockState p_48936_, ServerLevel p_48937_, BlockPos p_48938_, Random p_48939_) {
       if (p_48936_.m_61143_(f_48871_) == 0) {
-         if (p_48939_.nextInt(3) == 0 && p_48937_.m_46859_(p_48938_.m_7494_()) && p_48937_.m_45524_(p_48938_.m_7494_(), 0) >= 9) {
+         //if (p_48937_.isEmptyBlock(p_48938_.above()) && p_48937_.getRawBrightness(p_48938_.above(), 0) >= 9) {
+         if (p_48939_.nextInt(Math.max(1, (int) (100.0F / p_48937_.spigotConfig.bambooModifier) * 3)) == 0 && p_48937_.m_46859_(p_48938_.m_7494_()) && p_48937_.m_45524_(p_48938_.m_7494_(), 0) >= 9) { // Spigot
             int i = this.m_48932_(p_48937_, p_48938_) + 1;
-            if (i < 16) {
+            if (i < 16 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_48937_, p_48938_, p_48936_, p_48939_.nextInt(3) == 0)) {
                this.m_48910_(p_48936_, p_48937_, p_48938_, p_48939_, i);
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_48937_, p_48938_, p_48936_);
             }
          }
 
@@ -171,7 +_,7 @@
    }
 
    public float m_5880_(BlockState p_48901_, Player p_48902_, BlockGetter p_48903_, BlockPos p_48904_) {
-      return p_48902_.m_21205_().m_41720_() instanceof SwordItem ? 1.0F : super.m_5880_(p_48901_, p_48902_, p_48903_, p_48904_);
+      return p_48902_.m_21205_().canPerformAction(net.minecraftforge.common.ToolActions.SWORD_DIG) ? 1.0F : super.m_5880_(p_48901_, p_48902_, p_48903_, p_48904_);
    }
 
    protected void m_48910_(BlockState p_48911_, Level p_48912_, BlockPos p_48913_, Random p_48914_, int p_48915_) {
@@ -179,13 +_,17 @@
       BlockPos blockpos = p_48913_.m_6625_(2);
       BlockState blockstate1 = p_48912_.m_8055_(blockpos);
       BambooLeaves bambooleaves = BambooLeaves.NONE;
+      boolean shouldUpdateOthers = false; // CraftBukkit
       if (p_48915_ >= 1) {
          if (blockstate.m_60713_(Blocks.f_50571_) && blockstate.m_61143_(f_48870_) != BambooLeaves.NONE) {
             if (blockstate.m_60713_(Blocks.f_50571_) && blockstate.m_61143_(f_48870_) != BambooLeaves.NONE) {
                bambooleaves = BambooLeaves.LARGE;
                if (blockstate1.m_60713_(Blocks.f_50571_)) {
-                  p_48912_.m_7731_(p_48913_.m_7495_(), blockstate.m_61124_(f_48870_, BambooLeaves.SMALL), 3);
-                  p_48912_.m_7731_(blockpos, blockstate1.m_61124_(f_48870_, BambooLeaves.NONE), 3);
+                  // CraftBukkit start - moved down
+                  // world.setBlock(blockposition.below(), (IBlockData) iblockdata1.setValue(BlockBamboo.LEAVES, BlockPropertyBambooSize.SMALL), 3);
+                  // world.setBlock(blockposition1, (IBlockData) iblockdata2.setValue(BlockBamboo.LEAVES, BlockPropertyBambooSize.NONE), 3);
+                  shouldUpdateOthers = true;
+                  // CraftBukkit end
                }
             }
          } else {
@@ -195,7 +_,14 @@
 
       int i = p_48911_.m_61143_(f_48869_) != 1 && !blockstate1.m_60713_(Blocks.f_50571_) ? 0 : 1;
       int j = (p_48915_ < 11 || !(p_48914_.nextFloat() < 0.25F)) && p_48915_ != 15 ? 0 : 1;
-      p_48912_.m_7731_(p_48913_.m_7494_(), this.m_49966_().m_61124_(f_48869_, Integer.valueOf(i)).m_61124_(f_48870_, bambooleaves).m_61124_(f_48871_, Integer.valueOf(j)), 3);
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_48912_, p_48913_, p_48913_.m_7494_(), ((this.m_49966_().m_61124_(BambooBlock.f_48869_, j)).m_61124_(BambooBlock.f_48870_, bambooleaves)).m_61124_(BambooBlock.f_48871_, Integer.valueOf(j)), 3)) {
+         if (shouldUpdateOthers) {
+            p_48912_.m_7731_(p_48913_.m_7495_(), blockstate.m_61124_(f_48870_, BambooLeaves.SMALL), 3);
+            p_48912_.m_7731_(blockpos, blockstate1.m_61124_(f_48870_, BambooLeaves.NONE), 3);
+         }
+      }
+      // CraftBukkit end
    }
 
    protected int m_48882_(BlockGetter p_48883_, BlockPos p_48884_) {
@@ -212,5 +_,12 @@
       }
 
       return i;
+   }
+
+   @Override
+   public BlockState getPlant(BlockGetter world, BlockPos pos) {
+      BlockState state = world.m_8055_(pos);
+      if (state.m_60734_() != this) return m_49966_();
+      return state;
    }
 }
