--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -113,7 +_,13 @@
       if (!p_49263_.m_5825_()) {
          p_49263_.m_7311_(p_49263_.m_20094_() + 1);
          if (p_49263_.m_20094_() == 0) {
-            p_49263_.m_20254_(8);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_49261_, p_49262_), p_49263_.getBukkitEntity(), 8);
+            p_49261_.getCraftServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               p_49263_.setSecondsOnFire(event.getDuration(), false);
+            }
+            // CraftBukkit end
          }
 
          p_49263_.m_6469_(DamageSource.f_19305_, this.f_49238_);
@@ -126,14 +_,15 @@
       if (!p_49282_.m_60713_(p_49279_.m_60734_())) {
          if (m_49248_(p_49280_)) {
             Optional<PortalShape> optional = PortalShape.m_77708_(p_49280_, p_49281_, Direction.Axis.X);
+            optional = net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_49280_, p_49281_, optional);
             if (optional.isPresent()) {
-               optional.get().m_77743_();
+               optional.get().createPortalBlocks();
                return;
             }
          }
 
          if (!p_49279_.m_60710_(p_49280_, p_49281_)) {
-            p_49280_.m_7471_(p_49281_, false);
+            fireExtinguished(p_49280_, p_49281_); // CraftBukkit - fuel block broke
          }
 
       }
@@ -185,4 +_,12 @@
          }
       }
    }
+
+   // CraftBukkit start
+   protected void fireExtinguished(net.minecraft.world.level.LevelAccessor world, BlockPos position) {
+      if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(world, position, Blocks.f_50016_.m_49966_()).isCancelled()) {
+         world.m_7471_(position, false);
+      }
+   }
+   // CraftBukkit end
 }
