--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public abstract class BasePressurePlateBlock extends Block {
    protected static final VoxelShape f_49285_ = Block.m_49796_(1.0D, 0.0D, 1.0D, 15.0D, 0.5D, 15.0D);
@@ -70,6 +_,18 @@
       int i = this.m_6693_(p_152145_, p_152146_);
       boolean flag = p_152148_ > 0;
       boolean flag1 = i > 0;
+      // CraftBukkit start - Interact Pressure Plate
+      org.bukkit.World bworld = p_152145_.getWorld();
+      org.bukkit.plugin.PluginManager manager = p_152145_.getCraftServer().getPluginManager();
+
+      if (flag != flag1) {
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(p_152146_.m_123341_(), p_152146_.m_123342_(), p_152146_.m_123343_()), p_152148_, i);
+         manager.callEvent(eventRedstone);
+
+         flag1 = eventRedstone.getNewCurrent() > 0;
+         i = eventRedstone.getNewCurrent();
+      }
+      // CraftBukkit end
       if (p_152148_ != i) {
          BlockState blockstate = this.m_7422_(p_152147_, i);
          p_152145_.m_7731_(p_152146_, blockstate, 2);
