--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -43,6 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.apache.commons.lang3.ArrayUtils;
+import org.bukkit.World;
 
 public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock {
    public static final EnumProperty<BedPart> f_49440_ = BlockStateProperties.f_61391_;
@@ -84,7 +_,8 @@
             }
          }
 
-         if (!m_49488_(p_49516_)) {
+         // CraftBukkit - moved world and biome check into EntityHuman
+         if (false && !m_49488_(p_49516_)) {
             p_49516_.m_7471_(p_49517_, false);
             BlockPos blockpos = p_49517_.m_142300_(p_49515_.m_61143_(f_54117_).m_122424_());
             if (p_49516_.m_8055_(blockpos).m_60713_(this)) {
@@ -100,7 +_,16 @@
 
             return InteractionResult.SUCCESS;
          } else {
+            // CraftBukkit start
+            BlockState finaliblockdata = p_49515_;
+            BlockPos finalblockposition = p_49517_;
+            // CraftBukkit end
             p_49518_.m_7720_(p_49517_).ifLeft((p_49477_) -> {
+               // CraftBukkit start - handling bed explosion from below here
+               if (!p_49516_.m_6042_().m_63961_()) {
+                  this.explodeBed(finaliblockdata, p_49516_, finalblockposition);
+               } else
+                  // CraftBukkit end
                if (p_49477_ != null) {
                   p_49518_.m_5661_(p_49477_.m_36423_(), true);
                }
@@ -111,8 +_,25 @@
       }
    }
 
+   // CraftBukkit start
+   private InteractionResult explodeBed(BlockState iblockdata, Level world, BlockPos blockposition) {
+      {
+         {
+            world.m_7471_(blockposition, false);
+            BlockPos blockposition1 = blockposition.m_142300_(iblockdata.m_61143_(BedBlock.f_54117_).m_122424_());
+            if (world.m_8055_(blockposition1).m_60734_() == this) {
+               world.m_7471_(blockposition1, false);
+            }
+            world.m_7703_((Entity) null, DamageSource.m_19334_(), (ExplosionDamageCalculator) null, (double) blockposition.m_123341_() + 0.5D, (double) blockposition.m_123342_() + 0.5D, (double) blockposition.m_123343_() + 0.5D, 5.0F, true, Explosion.BlockInteraction.DESTROY);
+            return InteractionResult.SUCCESS;
+         }
+      }
+   }
+   // CraftBukkit end
+
    public static boolean m_49488_(Level p_49489_) {
-      return p_49489_.m_6042_().m_63961_();
+      // CraftBukkit - moved world and biome check into EntityHuman
+      return true || p_49489_.m_6042_().m_63961_();
    }
 
    private boolean m_49490_(Level p_49491_, BlockPos p_49492_) {
