--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -126,6 +_,11 @@
          if (p_152192_ == null) {
             p_152192_ = p_152190_.m_8055_(p_152191_).m_61143_(f_49679_);
          }
+         // CraftBukkit start - 1.20.1
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBellRingEvent(p_152190_, p_152191_, p_152192_, p_152189_)) {
+            return false;
+         }
+         // CraftBukkit end
 
          ((BellBlockEntity)blockentity).m_58834_(p_152192_);
          p_152190_.m_5594_((Player)null, p_152191_, SoundEvents.f_11699_, SoundSource.BLOCKS, 2.0F, 1.0F);
@@ -247,4 +_,16 @@
    public boolean m_7357_(BlockState p_49717_, BlockGetter p_49718_, BlockPos p_49719_, PathComputationType p_49720_) {
       return false;
    }
+
+   // CraftBukkit start - 1.20.1 fix MC-253819
+   @Override
+   public BlockState m_6843_(BlockState iblockdata, Rotation enumblockrotation) {
+      return iblockdata.m_61124_(f_49679_, enumblockrotation.m_55954_(iblockdata.m_61143_(f_49679_)));
+   }
+
+   @Override
+   public BlockState m_6943_(BlockState iblockdata, Mirror enumblockmirror) {
+      return iblockdata.m_60717_(enumblockmirror.m_54846_(iblockdata.m_61143_(f_49679_)));
+   }
+   // CraftBukkit end
 }
