--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -40,6 +_,11 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class BigDripleafBlock extends HorizontalDirectionalBlock implements BonemealableBlock, SimpleWaterloggedBlock {
    private static final BooleanProperty f_152200_ = BlockStateProperties.f_61362_;
    private static final EnumProperty<Tilt> f_152201_ = BlockStateProperties.f_155996_;
@@ -104,7 +_,7 @@
    }
 
    public void m_5581_(Level p_152228_, BlockState p_152229_, BlockHitResult p_152230_, Projectile p_152231_) {
-      this.m_152282_(p_152229_, p_152228_, p_152230_.m_82425_(), Tilt.FULL, SoundEvents.f_144131_);
+      this.setTiltAndScheduleTick(p_152229_, p_152228_, p_152230_.m_82425_(), Tilt.FULL, SoundEvents.f_144131_, p_152231_); // CraftBukkit
    }
 
    public FluidState m_5888_(BlockState p_152312_) {
@@ -152,7 +_,19 @@
    public void m_7892_(BlockState p_152266_, Level p_152267_, BlockPos p_152268_, Entity p_152269_) {
       if (!p_152267_.f_46443_) {
          if (p_152266_.m_61143_(f_152201_) == Tilt.NONE && m_152301_(p_152268_, p_152269_) && !p_152267_.m_276867_(p_152268_)) {
-            this.m_152282_(p_152266_, p_152267_, p_152268_, Tilt.UNSTABLE, (SoundEvent)null);
+            // CraftBukkit start - tilt dripleaf
+            org.bukkit.event.Cancellable cancellable;
+            if (p_152269_ instanceof Player) {
+               cancellable = CraftEventFactory.callPlayerInteractEvent((Player) p_152269_, org.bukkit.event.block.Action.PHYSICAL, p_152268_, null, null, null);
+            } else {
+               cancellable = new EntityInteractEvent(p_152269_.getBukkitEntity(), p_152267_.getWorld().getBlockAt(p_152268_.m_123341_(), p_152268_.m_123342_(), p_152268_.m_123343_()));
+               p_152267_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+            if (cancellable.isCancelled()) {
+               return;
+            }
+            this.setTiltAndScheduleTick(p_152266_, p_152267_, p_152268_, Tilt.UNSTABLE, (SoundEvent)null, p_152269_);
+            // CraftBukkit end
          }
 
       }
@@ -164,9 +_,9 @@
       } else {
          Tilt tilt = p_220798_.m_61143_(f_152201_);
          if (tilt == Tilt.UNSTABLE) {
-            this.m_152282_(p_220798_, p_220799_, p_220800_, Tilt.PARTIAL, SoundEvents.f_144131_);
+            this.setTiltAndScheduleTick(p_220798_, p_220799_, p_220800_, Tilt.PARTIAL, SoundEvents.f_144131_, null); // CraftBukkit
          } else if (tilt == Tilt.PARTIAL) {
-            this.m_152282_(p_220798_, p_220799_, p_220800_, Tilt.FULL, SoundEvents.f_144131_);
+            this.setTiltAndScheduleTick(p_220798_, p_220799_, p_220800_, Tilt.FULL, SoundEvents.f_144131_, null); // CraftBukkit
          } else if (tilt == Tilt.FULL) {
             m_152313_(p_220798_, p_220799_, p_220800_);
          }
@@ -191,7 +_,13 @@
    }
 
    private void m_152282_(BlockState p_152283_, Level p_152284_, BlockPos p_152285_, Tilt p_152286_, @Nullable SoundEvent p_152287_) {
-      m_152277_(p_152283_, p_152284_, p_152285_, p_152286_);
+      // CatServer start
+      this.setTiltAndScheduleTick(p_152283_, p_152284_, p_152285_, p_152286_, p_152287_, null);
+   }
+
+   private void setTiltAndScheduleTick(BlockState p_152283_, Level p_152284_, BlockPos p_152285_, Tilt p_152286_, @Nullable SoundEvent p_152287_, @Nullable Entity entity) {
+      if (!setTilt(p_152283_, p_152284_, p_152285_, p_152286_, entity)) return;
+      // CatServer end
       if (p_152287_ != null) {
          m_152232_(p_152284_, p_152285_, p_152287_);
       }
@@ -204,7 +_,7 @@
    }
 
    private static void m_152313_(BlockState p_152314_, Level p_152315_, BlockPos p_152316_) {
-      m_152277_(p_152314_, p_152315_, p_152316_, Tilt.NONE);
+      setTilt(p_152314_, p_152315_, p_152316_, Tilt.NONE, (Entity)null); // CraftBukkit
       if (p_152314_.m_61143_(f_152201_) != Tilt.NONE) {
          m_152232_(p_152315_, p_152316_, SoundEvents.f_144132_);
       }
@@ -212,12 +_,23 @@
    }
 
    private static void m_152277_(BlockState p_152278_, Level p_152279_, BlockPos p_152280_, Tilt p_152281_) {
+      // CatServer start
+      setTilt(p_152278_, p_152279_, p_152280_, p_152281_, (Entity)null);
+   }
+
+   private static boolean setTilt(BlockState p_152278_, Level p_152279_, BlockPos p_152280_, Tilt p_152281_, @Nullable Entity entity) {
+      if (entity != null) {
+         if (!CraftEventFactory.callEntityChangeBlockEvent(entity, p_152280_, p_152278_.m_61124_(f_152201_, p_152281_))) {
+            return false;
+         }
+      }
+      // CatServer end
       Tilt tilt = p_152278_.m_61143_(f_152201_);
       p_152279_.m_7731_(p_152280_, p_152278_.m_61124_(f_152201_, p_152281_), 2);
       if (p_152281_.m_156084_() && p_152281_ != tilt) {
          p_152279_.m_142346_((Entity)null, GameEvent.f_157792_, p_152280_);
       }
-
+      return true; // CraftBukkit
    }
 
    public VoxelShape m_5939_(BlockState p_152307_, BlockGetter p_152308_, BlockPos p_152309_, CollisionContext p_152310_) {
