--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -39,7 +_,7 @@
 
          if (block != null) {
             BlockState blockstate1 = block.m_49966_().m_61124_(AmethystClusterBlock.f_152006_, direction).m_61124_(AmethystClusterBlock.f_152005_, Boolean.valueOf(blockstate.m_60819_().m_76152_() == Fluids.f_76193_));
-            p_152729_.m_46597_(blockpos, blockstate1);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_152729_, p_152730_, blockpos, blockstate1); // CraftBukkit
          }
 
       }
