--- a/net/minecraft/world/level/block/ButtonBlock.java
+++ b/net/minecraft/world/level/block/ButtonBlock.java
@@ -27,6 +_,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public abstract class ButtonBlock extends FaceAttachedHorizontalDirectionalBlock {
    public static final BooleanProperty f_51045_ = BlockStateProperties.f_61448_;
    private static final int f_152736_ = 1;
@@ -97,6 +_,19 @@
       if (p_51088_.m_61143_(f_51045_)) {
          return InteractionResult.CONSUME;
       } else {
+         // CraftBukkit start
+         boolean powered = ((Boolean) p_51088_.m_61143_(f_51045_));
+         org.bukkit.block.Block block = p_51089_.getWorld().getBlockAt(p_51090_.m_123341_(), p_51090_.m_123342_(), p_51090_.m_123343_());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_51089_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          this.m_51116_(p_51088_, p_51089_, p_51090_);
          this.m_51067_(p_51091_, p_51089_, p_51090_, true);
          p_51089_.m_142346_(p_51091_, GameEvent.f_157798_, p_51090_);
@@ -143,6 +_,16 @@
          if (this.f_51062_) {
             this.m_51120_(p_51073_, p_51074_, p_51075_);
          } else {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_51074_.getWorld().getBlockAt(p_51075_.m_123341_(), p_51075_.m_123342_(), p_51075_.m_123343_());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+            p_51074_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+               return;
+            }
+            // CraftBukkit end
             p_51074_.m_7731_(p_51075_, p_51073_.m_61124_(f_51045_, Boolean.valueOf(false)), 3);
             this.m_51124_(p_51073_, p_51074_, p_51075_);
             this.m_51067_((Player)null, p_51074_, p_51075_, false);
@@ -162,11 +_,48 @@
       List<? extends Entity> list = p_51122_.m_45976_(AbstractArrow.class, p_51121_.m_60808_(p_51122_, p_51123_).m_83215_().m_82338_(p_51123_));
       boolean flag = !list.isEmpty();
       boolean flag1 = p_51121_.m_61143_(f_51045_);
+
+      // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+      if (flag1 != flag && flag) {
+         org.bukkit.block.Block block = p_51122_.getWorld().getBlockAt(p_51123_.m_123341_(), p_51123_.m_123342_(), p_51123_.m_123343_());
+         boolean allowed = false;
+
+         // If all of the events are cancelled block the button press, else allow
+         for (Object object : list) {
+            if (object != null) {
+               EntityInteractEvent event = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+               p_51122_.getCraftServer().getPluginManager().callEvent(event);
+
+               if (!event.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+
+         if (!allowed) {
+            return;
+         }
+      }
+      // CraftBukkit end
       if (flag != flag1) {
+         // CraftBukkit start
+         boolean powered = flag1;
+         org.bukkit.block.Block block = p_51122_.getWorld().getBlockAt(p_51123_.m_123341_(), p_51123_.m_123342_(), p_51123_.m_123343_());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_51122_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((flag && eventRedstone.getNewCurrent() <= 0) || (!flag && eventRedstone.getNewCurrent() > 0)) {
+            return;
+         }
+         // CraftBukkit end
          p_51122_.m_7731_(p_51123_, p_51121_.m_61124_(f_51045_, Boolean.valueOf(flag)), 3);
          this.m_51124_(p_51121_, p_51122_, p_51123_);
          this.m_51067_((Player)null, p_51122_, p_51123_, flag);
-         p_51122_.m_142346_(list.stream().findFirst().orElse(null), flag ? GameEvent.f_157798_ : GameEvent.f_157800_, p_51123_);
+         p_51122_.m_142346_(list.stream().findFirst().orElse(null), flag ? GameEvent.f_157798_ : GameEvent.f_157800_, p_51123_); // CraftBukkit - decompile error
       }
 
       if (flag) {
