--- a/net/minecraft/world/level/block/CakeBlock.java
+++ b/net/minecraft/world/level/block/CakeBlock.java
@@ -79,7 +_,19 @@
          return InteractionResult.PASS;
       } else {
          p_51189_.m_36220_(Stats.f_12942_);
-         p_51189_.m_36324_().m_38707_(2, 0.1F);
+         //p_51189_.getFoodData().eat(2, 0.1F);
+         // CraftBukkit start
+         // entityhuman.getFoodData().eat(2, 0.1F);
+         int oldFoodLevel = p_51189_.m_36324_().f_38696_;
+
+         org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callFoodLevelChangeEvent(p_51189_, 2 + oldFoodLevel);
+
+         if (!event.isCancelled()) {
+            p_51189_.m_36324_().m_38707_(event.getFoodLevel() - oldFoodLevel, 0.1F);
+         }
+
+         ((net.minecraft.server.level.ServerPlayer) p_51189_).getBukkitEntity().sendHealthUpdate();
+         // CraftBukkit end
          int i = p_51188_.m_61143_(f_51180_);
          p_51186_.m_142346_(p_51189_, GameEvent.f_157806_, p_51187_);
          if (i < 6) {
