--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -45,6 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
 
 public class CampfireBlock extends BaseEntityBlock implements SimpleWaterloggedBlock {
    protected static final VoxelShape f_51226_ = Block.m_49796_(0.0D, 0.0D, 0.0D, 16.0D, 7.0D, 16.0D);
@@ -85,7 +_,9 @@
 
    public void m_7892_(BlockState p_51269_, Level p_51270_, BlockPos p_51271_, Entity p_51272_) {
       if (!p_51272_.m_5825_() && p_51269_.m_61143_(f_51227_) && p_51272_ instanceof LivingEntity && !EnchantmentHelper.m_44938_((LivingEntity)p_51272_)) {
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.blockDamage = CraftBlock.at(p_51270_, p_51271_); // CraftBukkit
          p_51272_.m_6469_(DamageSource.f_19305_, (float)this.f_51233_);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.blockDamage = null; // CraftBukkit
       }
 
       super.m_7892_(p_51269_, p_51270_, p_51271_, p_51272_);
@@ -182,6 +_,11 @@
    public void m_5581_(Level p_51244_, BlockState p_51245_, BlockHitResult p_51246_, Projectile p_51247_) {
       BlockPos blockpos = p_51246_.m_82425_();
       if (!p_51244_.f_46443_ && p_51247_.m_6060_() && p_51247_.m_142265_(p_51244_, blockpos) && !p_51245_.m_61143_(f_51227_) && !p_51245_.m_61143_(f_51229_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(p_51244_, blockpos, p_51247_).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_51244_.m_7731_(blockpos, p_51245_.m_61124_(BlockStateProperties.f_61443_, Boolean.valueOf(true)), 11);
       }
 
@@ -205,7 +_,7 @@
             return true;
          }
 
-         boolean flag = Shapes.m_83157_(f_51231_, blockstate.m_60742_(p_51249_, p_51250_, CollisionContext.m_82749_()), BooleanOp.f_82689_);
+         boolean flag = Shapes.m_83157_(f_51231_, blockstate.m_60742_(p_51249_, blockpos, CollisionContext.m_82749_()), BooleanOp.f_82689_);//Forge fix: MC-201374
          if (flag) {
             BlockState blockstate1 = p_51249_.m_8055_(blockpos.m_7495_());
             return m_51319_(blockstate1);
