--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -23,6 +_,8 @@
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
 import net.minecraft.world.level.block.state.properties.DirectionProperty;
 import net.minecraft.world.level.material.Material;
+import org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class CarvedPumpkinBlock extends HorizontalDirectionalBlock implements Wearable {
    public static final DirectionProperty f_51367_ = HorizontalDirectionalBlock.f_54117_;
@@ -55,17 +_,26 @@
 
    private void m_51378_(Level p_51379_, BlockPos p_51380_) {
       BlockPattern.BlockPatternMatch blockpattern$blockpatternmatch = this.m_51393_().m_61184_(p_51379_, p_51380_);
+      BlockStateListPopulator blockList = new BlockStateListPopulator(p_51379_); // CraftBukkit - Use BlockStateListPopulator
       if (blockpattern$blockpatternmatch != null) {
          for(int i = 0; i < this.m_51393_().m_61202_(); ++i) {
             BlockInWorld blockinworld = blockpattern$blockpatternmatch.m_61229_(0, i, 0);
-            p_51379_.m_7731_(blockinworld.m_61176_(), Blocks.f_50016_.m_49966_(), 2);
-            p_51379_.m_46796_(2001, blockinworld.m_61176_(), Block.m_49956_(blockinworld.m_61168_()));
+            blockList.m_7731_(blockinworld.m_61176_(), Blocks.f_50016_.m_49966_(), 2); // CraftBukkit
+            // p_51379_.levelEvent(2001, blockinworld.getPos(), Block.getId(blockinworld.getState()));
          }
 
          SnowGolem snowgolem = EntityType.f_20528_.m_20615_(p_51379_);
          BlockPos blockpos1 = blockpattern$blockpatternmatch.m_61229_(0, 2, 0).m_61176_();
          snowgolem.m_7678_((double)blockpos1.m_123341_() + 0.5D, (double)blockpos1.m_123342_() + 0.05D, (double)blockpos1.m_123343_() + 0.5D, 0.0F, 0.0F);
-         p_51379_.m_7967_(snowgolem);
+         // CraftBukkit start
+         if (!p_51379_.addFreshEntity(snowgolem, CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN)) {
+            return;
+         }
+         for (BlockPos pos : blockList.getBlocks()) {
+            p_51379_.m_46796_(2001, pos, Block.m_49956_(p_51379_.m_8055_(pos)));
+         }
+         blockList.updateList();
+         // CraftBukkit end
 
          for(ServerPlayer serverplayer : p_51379_.m_45976_(ServerPlayer.class, snowgolem.m_142469_().m_82400_(5.0D))) {
             CriteriaTriggers.f_10580_.m_68256_(serverplayer, snowgolem);
@@ -81,8 +_,8 @@
             for(int j = 0; j < this.m_51397_().m_61203_(); ++j) {
                for(int k = 0; k < this.m_51397_().m_61202_(); ++k) {
                   BlockInWorld blockinworld2 = blockpattern$blockpatternmatch.m_61229_(j, k, 0);
-                  p_51379_.m_7731_(blockinworld2.m_61176_(), Blocks.f_50016_.m_49966_(), 2);
-                  p_51379_.m_46796_(2001, blockinworld2.m_61176_(), Block.m_49956_(blockinworld2.m_61168_()));
+                  blockList.m_7731_(blockinworld2.m_61176_(), Blocks.f_50016_.m_49966_(), 2); // CraftBukkit
+                  // p_51379_.levelEvent(2001, blockinworld2.getPos(), Block.getId(blockinworld2.getState())); // CraftBukkit
                }
             }
 
@@ -90,7 +_,15 @@
             IronGolem irongolem = EntityType.f_20460_.m_20615_(p_51379_);
             irongolem.m_28887_(true);
             irongolem.m_7678_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.05D, (double)blockpos.m_123343_() + 0.5D, 0.0F, 0.0F);
-            p_51379_.m_7967_(irongolem);
+            // CraftBukkit start
+            if (!p_51379_.addFreshEntity(irongolem, CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM)) {
+               return;
+            }
+            for (BlockPos pos : blockList.getBlocks()) {
+               p_51379_.m_46796_(2001, pos, Block.m_49956_(p_51379_.m_8055_(pos)));
+            }
+            blockList.updateList();
+            // CraftBukkit end
 
             for(ServerPlayer serverplayer1 : p_51379_.m_45976_(ServerPlayer.class, irongolem.m_142469_().m_82400_(5.0D))) {
                CriteriaTriggers.f_10580_.m_68256_(serverplayer1, irongolem);
