--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class CauldronBlock extends AbstractCauldronBlock {
    private static final float f_182448_ = 0.05F;
@@ -50,12 +_,15 @@
       return true;
    }
 
+   // FoxServer TODO
    protected void m_142310_(BlockState p_152940_, Level p_152941_, BlockPos p_152942_, Fluid p_152943_) {
       if (p_152943_ == Fluids.f_76193_) {
+         //LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, Blocks.WATER_CAULDRON.defaultBlockState(), null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
          p_152941_.m_46597_(p_152942_, Blocks.f_152476_.m_49966_());
          p_152941_.m_46796_(1047, p_152942_, 0);
          p_152941_.m_142346_((Entity)null, GameEvent.f_157769_, p_152942_);
       } else if (p_152943_ == Fluids.f_76195_) {
+         //LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, Blocks.WATER_CAULDRON.defaultBlockState(), null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
          p_152941_.m_46597_(p_152942_, Blocks.f_152477_.m_49966_());
          p_152941_.m_46796_(1046, p_152942_, 0);
          p_152941_.m_142346_((Entity)null, GameEvent.f_157769_, p_152942_);
