--- a/net/minecraft/world/level/block/CaveVinesPlantBlock.java
+++ b/net/minecraft/world/level/block/CaveVinesPlantBlock.java
@@ -35,7 +_,7 @@
    }
 
    public InteractionResult m_6227_(BlockState p_153021_, Level p_153022_, BlockPos p_153023_, Player p_153024_, InteractionHand p_153025_, BlockHitResult p_153026_) {
-      return CaveVines.m_152953_(p_153021_, p_153022_, p_153023_);
+      return CaveVines.use(p_153021_, p_153022_, p_153023_, p_153024_); // CraftBukkit
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_153031_) {
