--- a/net/minecraft/world/level/block/CaveVines.java
+++ b/net/minecraft/world/level/block/CaveVines.java
@@ -7,8 +_,6 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
@@ -21,7 +_,24 @@
 
    static InteractionResult m_152953_(BlockState p_152954_, Level p_152955_, BlockPos p_152956_) {
       if (p_152954_.m_61143_(f_152949_)) {
-         Block.m_49840_(p_152955_, p_152956_, new ItemStack(Items.f_151079_, 1));
+         /*
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(entity, p_152956_, p_152954_.setValue(CaveVines.BERRIES, false)).isCancelled()) {
+            return InteractionResult.SUCCESS;
+         }
+
+         if (entity instanceof Player) {
+            PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(p_152955_, p_152956_, (Player) entity, Collections.singletonList(new ItemStack(Items.GLOW_BERRIES, 1)));
+            if (event.isCancelled()) {
+               return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+            }
+            for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+               Block.popResource(p_152955_, p_152956_, CraftItemStack.asNMSCopy(itemStack));
+            }
+         } else {
+            Block.popResource(p_152955_, p_152956_, new ItemStack(Items.GLOW_BERRIES, 1));
+         }
+         // CraftBukkit end*/
          float f = Mth.m_144924_(p_152955_.f_46441_, 0.8F, 1.2F);
          p_152955_.m_5594_((Player)null, p_152956_, SoundEvents.f_144088_, SoundSource.BLOCKS, 1.0F, f);
          p_152955_.m_7731_(p_152956_, p_152954_.m_61124_(f_152949_, Boolean.valueOf(false)), 2);
