--- a/net/minecraft/world/level/block/CaveVinesBlock.java
+++ b/net/minecraft/world/level/block/CaveVinesBlock.java
@@ -50,6 +_,7 @@
 
    public InteractionResult m_6227_(BlockState p_152980_, Level p_152981_, BlockPos p_152982_, Player p_152983_, InteractionHand p_152984_, BlockHitResult p_152985_) {
       return CaveVines.m_152953_(p_152980_, p_152981_, p_152982_);
+      //return CaveVines.use(iblockdata, world, blockposition, entityhuman); // CraftBukkit FoxServer TODO
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_152993_) {
