--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -60,6 +_,7 @@
       if (p_153050_.nextFloat() < f1) {
          this.m_142123_(p_153047_).ifPresent((p_153039_) -> {
             p_153048_.m_46597_(p_153049_, p_153039_);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockFormEvent(p_153048_, p_153049_, p_153039_); // CraftBukkit
          });
       }
 
