--- a/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
+++ b/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
@@ -112,7 +_,7 @@
       }
    }
 
-   private static int m_261279_(Vec2 p_261991_) {
+   public static int m_261279_(Vec2 p_261991_) { // CraftBukkit - 1.20.1 private -> public
       int i = p_261991_.f_82471_ >= 0.5F ? 0 : 1;
       int j = m_261181_(p_261991_.f_82470_);
       return j + i * 3;
