--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -17,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.BlockHitResult;
 
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory; // CraftBukkit
+
 public class ChorusFlowerBlock extends Block {
    public static final int f_153067_ = 5;
    public static final IntegerProperty f_51647_ = BlockStateProperties.f_61408_;
@@ -43,11 +_,11 @@
       BlockPos blockpos = p_220982_.m_7494_();
       if (p_220981_.m_46859_(blockpos) && blockpos.m_123342_() < p_220981_.m_151558_()) {
          int i = p_220980_.m_61143_(f_51647_);
-         if (i < 5) {
+         if (i < 5 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_220981_, blockpos, p_220980_, true)) {
             boolean flag = false;
             boolean flag1 = false;
             BlockState blockstate = p_220981_.m_8055_(p_220982_.m_7495_());
-            if (blockstate.m_60713_(Blocks.f_50259_)) {
+            if (blockstate.m_60713_(Blocks.f_50259_) || blockstate.m_204336_(net.minecraftforge.common.Tags.Blocks.CHORUS_ADDITIONALLY_GROWS_ON)) {
                flag = true;
             } else if (blockstate.m_60713_(this.f_51648_)) {
                int j = 1;
@@ -55,7 +_,7 @@
                for(int k = 0; k < 4; ++k) {
                   BlockState blockstate1 = p_220981_.m_8055_(p_220982_.m_6625_(j + 1));
                   if (!blockstate1.m_60713_(this.f_51648_)) {
-                     if (blockstate1.m_60713_(Blocks.f_50259_)) {
+                     if (blockstate1.m_60713_(Blocks.f_50259_) || blockstate1.m_204336_(net.minecraftforge.common.Tags.Blocks.CHORUS_ADDITIONALLY_GROWS_ON)) {
                         flag1 = true;
                      }
                      break;
@@ -72,8 +_,12 @@
             }
 
             if (flag && m_51697_(p_220981_, blockpos, (Direction)null) && p_220981_.m_46859_(p_220982_.m_6630_(2))) {
-               p_220981_.m_7731_(p_220982_, this.f_51648_.m_51710_(p_220981_, p_220982_), 2);
-               this.m_51661_(p_220981_, blockpos, i);
+               // CraftBukkit start - add event
+               if (CraftEventFactory.handleBlockSpreadEvent(p_220981_, p_220982_, blockpos, this.m_49966_().m_61124_(f_51647_, i), 2)) {
+                  p_220981_.m_7731_(p_220982_, this.f_51648_.m_51710_(p_220981_, p_220982_), 2);
+                  this.m_51661_(p_220981_, blockpos, i);
+               }
+               // CraftBukkit end
             } else if (i < 4) {
                int l = p_220983_.m_188503_(4);
                if (flag1) {
@@ -86,20 +_,32 @@
                   Direction direction = Direction.Plane.HORIZONTAL.m_235690_(p_220983_);
                   BlockPos blockpos1 = p_220982_.m_121945_(direction);
                   if (p_220981_.m_46859_(blockpos1) && p_220981_.m_46859_(blockpos1.m_7495_()) && m_51697_(p_220981_, blockpos1, direction.m_122424_())) {
-                     this.m_51661_(p_220981_, blockpos1, i + 1);
-                     flag2 = true;
+                     // CraftBukkit start - add event
+                     if (CraftEventFactory.handleBlockSpreadEvent(p_220981_, p_220982_, blockpos1, this.m_49966_().m_61124_(f_51647_, i + 1), 2)) {
+                        this.m_51661_(p_220981_, blockpos1, i + 1);
+                        flag2 = true;
+                     }
+                     // CraftBukkit end
                   }
                }
 
                if (flag2) {
                   p_220981_.m_7731_(p_220982_, this.f_51648_.m_51710_(p_220981_, p_220982_), 2);
                } else {
-                  this.m_51658_(p_220981_, p_220982_);
+                  // CraftBukkit start - add event
+                  if (CraftEventFactory.handleBlockGrowEvent(p_220981_, p_220982_, this.m_49966_().m_61124_(f_51647_, 5), 2)) {
+                     this.m_51658_(p_220981_, p_220982_);
+                  }
+                  // CraftBukkit end
                }
             } else {
-               this.m_51658_(p_220981_, p_220982_);
+               // CraftBukkit start - add event
+               if (CraftEventFactory.handleBlockGrowEvent(p_220981_, p_220982_, this.m_49966_().m_61124_(f_51647_, 5), 2)) {
+                  this.m_51658_(p_220981_, p_220982_);
+               }
+               // CraftBukkit end
             }
-
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_220981_, p_220982_, p_220980_);
          }
       }
    }
@@ -134,7 +_,7 @@
 
    public boolean m_7898_(BlockState p_51683_, LevelReader p_51684_, BlockPos p_51685_) {
       BlockState blockstate = p_51684_.m_8055_(p_51685_.m_7495_());
-      if (!blockstate.m_60713_(this.f_51648_) && !blockstate.m_60713_(Blocks.f_50259_)) {
+      if (!blockstate.m_60713_(this.f_51648_) && !blockstate.m_60713_(Blocks.f_50259_) && !blockstate.m_204336_(net.minecraftforge.common.Tags.Blocks.CHORUS_ADDITIONALLY_GROWS_ON)) {
          if (!blockstate.m_60795_()) {
             return false;
          } else {
@@ -214,6 +_,11 @@
    public void m_5581_(Level p_51654_, BlockState p_51655_, BlockHitResult p_51656_, Projectile p_51657_) {
       BlockPos blockpos = p_51656_.m_82425_();
       if (!p_51654_.f_46443_ && p_51657_.m_142265_(p_51654_, blockpos) && p_51657_.m_6095_().m_204039_(EntityTypeTags.f_13124_)) {
+         // CraftBukkit start
+         if (!CraftEventFactory.callEntityChangeBlockEvent(p_51657_, blockpos, Blocks.f_50016_.m_49966_())) {
+            return;
+         }
+         // CraftBukkit end
          p_51654_.m_46953_(blockpos, true, p_51657_);
       }
 
