--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -20,6 +_,8 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory; // CraftBukkit
+
 public class CocoaBlock extends HorizontalDirectionalBlock implements BonemealableBlock {
    public static final int f_153068_ = 2;
    public static final IntegerProperty f_51736_ = BlockStateProperties.f_61406_;
@@ -47,10 +_,12 @@
    }
 
    public void m_213898_(BlockState p_221000_, ServerLevel p_221001_, BlockPos p_221002_, RandomSource p_221003_) {
-      if (p_221001_.f_46441_.m_188503_(5) == 0) {
+      if (true) {
          int i = p_221000_.m_61143_(f_51736_);
-         if (i < 2) {
-            p_221001_.m_7731_(p_221002_, p_221000_.m_61124_(f_51736_, Integer.valueOf(i + 1)), 2);
+         if (i < 2 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_221001_, p_221002_, p_221000_, p_221001_.f_46441_.m_188501_() < (p_221001_.spigotConfig.cocoaModifier / (100.0f * 5)))) { // Spigot - SPIGOT-7159: Better modifier resolution
+            // p_221001_.setBlock(p_221002_, p_221000_.setValue(AGE, i + 1), 2);
+            CraftEventFactory.handleBlockGrowEvent(p_221001_, p_221002_, p_221000_.m_61124_(f_51736_, i + 1), 2); // CraftBukkit
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_221001_, p_221002_, p_221000_);
          }
       }
 
@@ -107,7 +_,7 @@
    }
 
    public void m_214148_(ServerLevel p_220990_, RandomSource p_220991_, BlockPos p_220992_, BlockState p_220993_) {
-      p_220990_.m_7731_(p_220992_, p_220993_.m_61124_(f_51736_, Integer.valueOf(p_220993_.m_61143_(f_51736_) + 1)), 2);
+      CraftEventFactory.handleBlockGrowEvent(p_220990_, p_220992_, p_220993_.m_61124_(f_51736_, p_220993_.m_61143_(f_51736_) + 1), 2); // CraftBukkit
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_51778_) {
