--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -19,6 +_,7 @@
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class CocoaBlock extends HorizontalDirectionalBlock implements BonemealableBlock {
    public static final int f_153068_ = 2;
@@ -47,10 +_,12 @@
    }
 
    public void m_7455_(BlockState p_51782_, ServerLevel p_51783_, BlockPos p_51784_, Random p_51785_) {
-      if (p_51783_.f_46441_.nextInt(5) == 0) {
+      //if (true) {
+      if (p_51783_.f_46441_.nextInt(Math.max(1, (int) (100.0F / p_51783_.spigotConfig.cocoaModifier) * 5)) == 0) { // Spigot
          int i = p_51782_.m_61143_(f_51736_);
-         if (i < 2) {
-            p_51783_.m_7731_(p_51784_, p_51782_.m_61124_(f_51736_, Integer.valueOf(i + 1)), 2);
+         if (i < 2 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_51783_, p_51784_, p_51782_, p_51783_.f_46441_.nextInt(5) == 0)) {
+            CraftEventFactory.handleBlockGrowEvent(p_51783_, p_51784_,  p_51782_.m_61124_(CocoaBlock.f_51736_, i + 1), 2); // CraftBukkkit
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_51783_, p_51784_, p_51782_);
          }
       }
 
@@ -107,7 +_,7 @@
    }
 
    public void m_7719_(ServerLevel p_51745_, Random p_51746_, BlockPos p_51747_, BlockState p_51748_) {
-      p_51745_.m_7731_(p_51747_, p_51748_.m_61124_(f_51736_, Integer.valueOf(p_51748_.m_61143_(f_51736_) + 1)), 2);
+      CraftEventFactory.handleBlockGrowEvent(p_51745_, p_51747_, p_51748_.m_61124_(CocoaBlock.f_51736_, p_51748_.m_61143_(CocoaBlock.f_51736_) + 1), 2); // CraftBukkit
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_51778_) {
