--- a/net/minecraft/world/level/block/CommandBlock.java
+++ b/net/minecraft/world/level/block/CommandBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.DirectionProperty;
 import net.minecraft.world.phys.BlockHitResult;
 import org.slf4j.Logger;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
    private static final Logger f_51795_ = LogUtils.getLogger();
@@ -52,6 +_,15 @@
             CommandBlockEntity commandblockentity = (CommandBlockEntity)blockentity;
             boolean flag = p_51839_.m_46753_(p_51840_);
             boolean flag1 = commandblockentity.m_59142_();
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = p_51839_.getWorld().getBlockAt(p_51840_.m_123341_(), p_51840_.m_123342_(), p_51840_.m_123343_());
+            int old = flag1 ? 15 : 0;
+            int current = flag ? 15 : 0;
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, old, current);
+            p_51839_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+            flag = eventRedstone.getNewCurrent() > 0;
+            // CraftBukkit end
             commandblockentity.m_59135_(flag);
             if (!flag1 && !commandblockentity.m_59143_() && commandblockentity.m_59148_() != CommandBlockEntity.Mode.SEQUENCE) {
                if (flag) {
