--- a/net/minecraft/world/level/block/CommandBlock.java
+++ b/net/minecraft/world/level/block/CommandBlock.java
@@ -27,6 +_,8 @@
 import net.minecraft.world.phys.BlockHitResult;
 import org.slf4j.Logger;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
    private static final Logger f_51795_ = LogUtils.getLogger();
    public static final DirectionProperty f_51793_ = DirectionalBlock.f_52588_;
@@ -52,6 +_,14 @@
             CommandBlockEntity commandblockentity = (CommandBlockEntity)blockentity;
             boolean flag = p_51839_.m_276867_(p_51840_);
             boolean flag1 = commandblockentity.m_59142_();
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = p_51839_.getWorld().getBlockAt(p_51840_.m_123341_(), p_51840_.m_123342_(), p_51840_.m_123343_());
+            int old = flag1 ? 15 : 0;
+            int current = flag ? 15 : 0;
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, old, current);
+            p_51839_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+            flag = eventRedstone.getNewCurrent() > 0;
+            // CraftBukkit end
             commandblockentity.m_59135_(flag);
             if (!flag1 && !commandblockentity.m_59143_() && commandblockentity.m_59148_() != CommandBlockEntity.Mode.SEQUENCE) {
                if (flag) {
