--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -37,6 +_,13 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import net.minecraft.world.entity.Entity;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftBlockInventoryHolder;
+import org.bukkit.craftbukkit.v1_18_R2.util.DummyGeneratorAccess;
+// CraftBukkit end
+
+
 public class ComposterBlock extends Block implements WorldlyContainerHolder {
    public static final int f_153088_ = 8;
    public static final int f_153089_ = 0;
@@ -213,25 +_,39 @@
 
          return InteractionResult.m_19078_(p_51950_.f_46443_);
       } else if (i == 8) {
-         m_51998_(p_51949_, p_51950_, p_51951_);
+         extractProduce(p_51949_, p_51950_, p_51951_, (Entity) null);
          return InteractionResult.m_19078_(p_51950_.f_46443_);
       } else {
          return InteractionResult.PASS;
       }
    }
 
-   public static BlockState m_51929_(BlockState p_51930_, ServerLevel p_51931_, ItemStack p_51932_, BlockPos p_51933_) {
+   public static BlockState insertItem(BlockState p_51930_, ServerLevel p_51931_, ItemStack p_51932_, BlockPos p_51933_, Entity p_51934_) {
       int i = p_51930_.m_61143_(f_51913_);
       if (i < 7 && f_51914_.containsKey(p_51932_.m_41720_())) {
-         BlockState blockstate = m_51983_(p_51930_, p_51931_, p_51933_, p_51932_);
+         // CraftBukkit start
+         double rand = p_51931_.m_5822_().nextDouble();
+         BlockState iblockdata1 = addItem(p_51930_, DummyGeneratorAccess.INSTANCE, p_51933_, p_51932_, rand);
+         if (p_51930_ == iblockdata1 || org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_51934_, p_51933_, iblockdata1).isCancelled()) {
+            return p_51930_;
+         }
+         iblockdata1 = addItem(p_51930_, (Level) p_51931_, p_51933_, p_51932_, rand);
+         // CraftBukkit end
          p_51932_.m_41774_(1);
-         return blockstate;
+         return p_51930_;
       } else {
          return p_51930_;
       }
    }
 
-   public static BlockState m_51998_(BlockState p_51999_, Level p_52000_, BlockPos p_52001_) {
+   public static BlockState extractProduce(BlockState p_51999_, Level p_52000_, BlockPos p_52001_, Entity p_52002_) {
+      if (p_52002_ != null) {
+         BlockState iblockdata1 = m_52002_(p_51999_, DummyGeneratorAccess.INSTANCE, p_52001_);
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_52002_, p_52001_, iblockdata1).isCancelled()) {
+            return p_51999_;
+         }
+      }
+      // CraftBukkit end
       if (!p_52000_.f_46443_) {
          float f = 0.7F;
          double d0 = (double)(p_52000_.f_46441_.nextFloat() * 0.7F) + (double)0.15F;
@@ -253,10 +_,17 @@
       return blockstate;
    }
 
+
    static BlockState m_51983_(BlockState p_51984_, LevelAccessor p_51985_, BlockPos p_51986_, ItemStack p_51987_) {
+      // CraftBukkit start
+      return addItem(p_51984_, p_51985_, p_51986_, p_51987_, p_51985_.m_5822_().nextDouble());
+   }
+
+   static BlockState addItem(BlockState p_51984_, LevelAccessor p_51985_, BlockPos p_51986_, ItemStack p_51987_, double rand) {
+      // CraftBukkit end
       int i = p_51984_.m_61143_(f_51913_);
       float f = f_51914_.getFloat(p_51987_.m_41720_());
-      if ((i != 0 || !(f > 0.0F)) && !(p_51985_.m_5822_().nextDouble() < (double)f)) {
+      if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
          return p_51984_;
       } else {
          int j = i + 1;
@@ -297,15 +_,17 @@
    public WorldlyContainer m_5840_(BlockState p_51956_, LevelAccessor p_51957_, BlockPos p_51958_) {
       int i = p_51956_.m_61143_(f_51913_);
       if (i == 8) {
-         return new ComposterBlock.OutputContainer(p_51956_, p_51957_, p_51958_, new ItemStack(Items.f_42499_));
+         return (WorldlyContainer) (new ComposterBlock.OutputContainer(p_51956_, p_51957_, p_51958_, new ItemStack(Items.f_42499_)));
       } else {
-         return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer());
+         return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer(p_51957_, p_51958_));
       }
    }
 
-   static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
-      public EmptyContainer() {
+   public static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
+
+      public EmptyContainer(LevelAccessor p_52023_, BlockPos p_52024_) { // CraftBukkit
          super(0);
+         this.bukkitOwner = new CraftBlockInventoryHolder(p_52023_, p_52024_, this); // CraftBukkit
       }
 
       public int[] m_7071_(Direction p_52012_) {
@@ -321,7 +_,7 @@
       }
    }
 
-   static class InputContainer extends SimpleContainer implements WorldlyContainer {
+   public static class InputContainer extends SimpleContainer implements WorldlyContainer {
       private final BlockState f_52017_;
       private final LevelAccessor f_52018_;
       private final BlockPos f_52019_;
@@ -332,6 +_,7 @@
          this.f_52017_ = p_52022_;
          this.f_52018_ = p_52023_;
          this.f_52019_ = p_52024_;
+         this.bukkitOwner = new CraftBlockInventoryHolder(p_52023_, p_52024_, this); // CraftBukkit
       }
 
       public int m_6893_() {
@@ -362,7 +_,7 @@
       }
    }
 
-   static class OutputContainer extends SimpleContainer implements WorldlyContainer {
+   public static class OutputContainer extends SimpleContainer implements WorldlyContainer {
       private final BlockState f_52037_;
       private final LevelAccessor f_52038_;
       private final BlockPos f_52039_;
@@ -373,6 +_,7 @@
          this.f_52037_ = p_52042_;
          this.f_52038_ = p_52043_;
          this.f_52039_ = p_52044_;
+         this.bukkitOwner = new CraftBlockInventoryHolder(p_52043_, p_52044_, this); // CraftBukkit
       }
 
       public int m_6893_() {
@@ -392,8 +_,15 @@
       }
 
       public void m_6596_() {
-         ComposterBlock.m_52002_(this.f_52037_, this.f_52038_, this.f_52039_);
-         this.f_52040_ = true;
+         // CraftBukkit start - allow putting items back (eg cancelled InventoryMoveItemEvent)
+         if (this.m_7983_()) {
+            ComposterBlock.m_52002_(this.f_52037_, this.f_52038_, this.f_52039_);
+            this.f_52040_ = true;
+         } else {
+            this.f_52038_.m_7731_(this.f_52039_, this.f_52037_, 3);
+            this.f_52040_ = false;
+         }
+         // CraftBukkit end
       }
    }
 }
