--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -10,6 +_,11 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockState;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockStates;
+import org.bukkit.event.block.BlockFormEvent;
 
 public class ConcretePowderBlock extends FallingBlock {
    private final BlockState f_52058_;
@@ -21,7 +_,7 @@
 
    public void m_142216_(Level p_52068_, BlockPos p_52069_, BlockState p_52070_, BlockState p_52071_, FallingBlockEntity p_52072_) {
       if (m_52080_(p_52068_, p_52069_, p_52071_)) {
-         p_52068_.m_7731_(p_52069_, this.f_52058_, 3);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockFormEvent(p_52068_, p_52069_, this.f_52058_, 3); // CraftBukkit
       }
 
    }
@@ -31,6 +_,26 @@
       BlockPos blockpos = p_52063_.m_8083_();
       BlockState blockstate = blockgetter.m_8055_(blockpos);
       return m_52080_(blockgetter, blockpos, blockstate) ? this.f_52058_ : super.m_5573_(p_52063_);
+      /* FoxServer TODO
+      // CraftBukkit start
+      if (!shouldSolidify(blockgetter, blockpos, blockstate)) {
+         return super.getStateForPlacement(p_52063_);
+      }
+
+      // TODO: An event factory call for methods like this
+      CraftBlockState blockState = CraftBlockStates.getBlockState(blockgetter, blockpos);
+      blockState.setData(this.concrete);
+
+      BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+      blockgetter.getServer().server.getPluginManager().callEvent(event);
+
+      if (!event.isCancelled()) {
+         return blockState.getHandle();
+      }
+
+      return super.getStateForPlacement(p_52063_);
+      // CraftBukkit end
+       */
    }
 
    private static boolean m_52080_(BlockGetter p_52081_, BlockPos p_52082_, BlockState p_52083_) {
@@ -61,7 +_,26 @@
    }
 
    public BlockState m_7417_(BlockState p_52074_, Direction p_52075_, BlockState p_52076_, LevelAccessor p_52077_, BlockPos p_52078_, BlockPos p_52079_) {
-      return m_52064_(p_52077_, p_52078_) ? this.f_52058_ : super.m_7417_(p_52074_, p_52075_, p_52076_, p_52077_, p_52078_, p_52079_);
+      //return touchesLiquid(p_52077_, p_52078_) ? this.concrete : super.updateShape(p_52074_, p_52075_, p_52076_, p_52077_, p_52078_, p_52079_);
+      // CraftBukkit start
+      if (m_52064_(p_52077_, p_52078_)) {
+         // Suppress during worldgen
+         if (!(p_52077_ instanceof World)) {
+            return this.f_52058_;
+         }
+         CraftBlockState blockState = CraftBlockStates.getBlockState(p_52077_, p_52078_);
+         blockState.setData(this.f_52058_);
+
+         BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+         Bukkit.getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            return blockState.getHandle();
+         }
+      }
+
+      return super.m_7417_(p_52074_, p_52075_, p_52076_, p_52077_, p_52078_, p_52079_);
+      // CraftBukkit end
    }
 
    public int m_6248_(BlockState p_52085_, BlockGetter p_52086_, BlockPos p_52087_) {
