--- a/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -29,6 +_,11 @@
 
    public void m_213897_(BlockState p_221030_, ServerLevel p_221031_, BlockPos p_221032_, RandomSource p_221033_) {
       if (!m_49186_(p_221030_, p_221031_, p_221032_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callBlockFadeEvent(p_221031_, p_221032_, this.f_52172_.m_49966_().m_61124_(f_49158_, false)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_221031_.m_7731_(p_221032_, this.f_52172_.m_49966_().m_61124_(f_49158_, Boolean.valueOf(false)), 2);
       }
 
