--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -23,6 +_,11 @@
 
    public void m_7458_(BlockState p_52138_, ServerLevel p_52139_, BlockPos p_52140_, Random p_52141_) {
       if (!this.m_52134_(p_52139_, p_52140_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52139_, p_52140_, this.f_52128_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52139_.m_7731_(p_52140_, this.f_52128_.m_49966_(), 2);
       }
 
