--- a/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/net/minecraft/world/level/block/CoralFanBlock.java
@@ -24,6 +_,11 @@
 
    public void m_7458_(BlockState p_52154_, ServerLevel p_52155_, BlockPos p_52156_, Random p_52157_) {
       if (!m_49186_(p_52154_, p_52155_, p_52156_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52155_, p_52156_, this.f_52149_.m_49966_().m_61124_(CoralFanBlock.f_49158_, false)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52155_.m_7731_(p_52156_, this.f_52149_.m_49966_().m_61124_(f_49158_, Boolean.valueOf(false)), 2);
       }
 
