--- a/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -24,6 +_,11 @@
 
    public void m_7458_(BlockState p_52205_, ServerLevel p_52206_, BlockPos p_52207_, Random p_52208_) {
       if (!m_49186_(p_52205_, p_52206_, p_52207_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52206_, p_52207_, this.f_52200_.m_49966_().m_61124_(CoralWallFanBlock.f_49158_, false).m_61124_(CoralWallFanBlock.f_49192_, p_52205_.m_61143_(CoralWallFanBlock.f_49192_))).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52206_.m_7731_(p_52207_, this.f_52200_.m_49966_().m_61124_(f_49158_, Boolean.valueOf(false)).m_61124_(f_49192_, p_52205_.m_61143_(f_49192_)), 2);
       }
 
