--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -60,6 +_,7 @@
 
       i = Mth.m_14045_(i, 0, 15);
       if (p_52411_.m_61143_(f_52377_) != i) {
+         i = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRedstoneChange(p_52412_, p_52413_, ((Integer) p_52411_.m_61143_(f_52377_)), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
          p_52412_.m_7731_(p_52413_, p_52411_.m_61124_(f_52377_, Integer.valueOf(i)), 3);
       }
 
