--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.Property;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DetectorRailBlock extends BaseRailBlock {
    public static final EnumProperty<RailShape> f_52427_ = BlockStateProperties.f_61404_;
@@ -75,6 +_,17 @@
             flag1 = true;
          }
 
+         // CraftBukkit start
+         if (flag != flag1) {
+            org.bukkit.block.Block block = p_52433_.getWorld().getBlockAt(p_52434_.m_123341_(), p_52434_.m_123342_(), p_52434_.m_123343_());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+            p_52433_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+         }
+         // CraftBukkit end
+
          if (flag1 && !flag) {
             BlockState blockstate = p_52435_.m_61124_(f_52428_, Boolean.valueOf(true));
             p_52433_.m_7731_(p_52434_, blockstate, 3);
@@ -135,7 +_,9 @@
             return list.get(0).m_38534_().m_45436_();
          }
 
-         List<AbstractMinecart> list1 = this.m_52436_(p_52455_, p_52456_, AbstractMinecart.class, EntitySelector.f_20405_);
+         List<AbstractMinecart> carts = this.m_52436_(p_52455_, p_52456_, AbstractMinecart.class, e -> e.m_6084_());
+         if (!carts.isEmpty() && carts.get(0).getComparatorLevel() > -1) return carts.get(0).getComparatorLevel();
+         List<AbstractMinecart> list1 = carts.stream().filter(EntitySelector.f_20405_).collect(java.util.stream.Collectors.toList());
          if (!list1.isEmpty()) {
             return AbstractContainerMenu.m_38938_((Container)list1.get(0));
          }
