--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.Property;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DetectorRailBlock extends BaseRailBlock {
    public static final EnumProperty<RailShape> f_52427_ = BlockStateProperties.f_61404_;
@@ -31,6 +_,10 @@
 
    public DetectorRailBlock(BlockBehaviour.Properties p_52431_) {
       super(true, p_52431_);
+      this.registerDefaultState();
+   }
+
+   protected void registerDefaultState() {
       this.m_49959_(this.f_49792_.m_61090_().m_61124_(f_52428_, Boolean.valueOf(false)).m_61124_(f_52427_, RailShape.NORTH_SOUTH).m_61124_(f_152149_, Boolean.valueOf(false)));
    }
 
@@ -75,6 +_,15 @@
             flag1 = true;
          }
 
+         // CraftBukkit start
+         if (flag != flag1) {
+            org.bukkit.block.Block block = p_52433_.getWorld().getBlockAt(p_52434_.m_123341_(), p_52434_.m_123342_(), p_52434_.m_123343_());
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+            p_52433_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+            flag1 = eventRedstone.getNewCurrent() > 0;
+         }
+         // CraftBukkit end
+
          if (flag1 && !flag) {
             BlockState blockstate = p_52435_.m_61124_(f_52428_, Boolean.valueOf(true));
             p_52433_.m_7731_(p_52434_, blockstate, 3);
@@ -135,7 +_,9 @@
             return list.get(0).m_38534_().m_45436_();
          }
 
-         List<AbstractMinecart> list1 = this.m_52436_(p_52455_, p_52456_, AbstractMinecart.class, EntitySelector.f_20405_);
+         List<AbstractMinecart> carts = this.m_52436_(p_52455_, p_52456_, AbstractMinecart.class, e -> e.m_6084_());
+         if (!carts.isEmpty() && carts.get(0).getComparatorLevel() > -1) return carts.get(0).getComparatorLevel();
+         List<AbstractMinecart> list1 = carts.stream().filter(EntitySelector.f_20405_).collect(java.util.stream.Collectors.toList());
          if (!list1.isEmpty()) {
             return AbstractContainerMenu.m_38938_((Container)list1.get(0));
          }
@@ -270,6 +_,6 @@
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_52469_) {
-      p_52469_.m_61104_(f_52427_, f_52428_, f_152149_);
+      p_52469_.m_61104_(m_7978_(), f_52428_, f_152149_);
    }
 }
