--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.ticks.TickPriority;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public abstract class DiodeBlock extends HorizontalDirectionalBlock {
    protected static final VoxelShape f_52495_ = Block.m_49796_(0.0D, 0.0D, 0.0D, 16.0D, 2.0D, 16.0D);
@@ -40,8 +_,18 @@
          boolean flag = p_52515_.m_61143_(f_52496_);
          boolean flag1 = this.m_7320_(p_52516_, p_52517_, p_52515_);
          if (flag && !flag1) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(p_52516_, p_52517_, 15, 0).getNewCurrent() != 0) {
+               return;
+            }
+            // CraftBukkit end
             p_52516_.m_7731_(p_52517_, p_52515_.m_61124_(f_52496_, Boolean.valueOf(false)), 2);
          } else if (!flag) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(p_52516_, p_52517_, 0, 15).getNewCurrent() != 15) {
+               return;
+            }
+            // CraftBukkit end
             p_52516_.m_7731_(p_52517_, p_52515_.m_61124_(f_52496_, Boolean.valueOf(true)), 2);
             if (!flag1) {
                p_52516_.m_186464_(p_52517_, this, this.m_6112_(p_52515_), TickPriority.VERY_HIGH);
@@ -165,6 +_,8 @@
    protected void m_52580_(Level p_52581_, BlockPos p_52582_, BlockState p_52583_) {
       Direction direction = p_52583_.m_61143_(f_54117_);
       BlockPos blockpos = p_52582_.m_142300_(direction.m_122424_());
+      if (net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(p_52581_, p_52582_, p_52581_.m_8055_(p_52582_), java.util.EnumSet.of(direction.m_122424_()), false).isCanceled())
+         return;
       p_52581_.m_46586_(blockpos, this, p_52582_);
       p_52581_.m_46590_(blockpos, this, direction);
    }
