--- a/net/minecraft/world/level/block/DirtPathBlock.java
+++ b/net/minecraft/world/level/block/DirtPathBlock.java
@@ -38,6 +_,11 @@
    }
 
    public void m_7458_(BlockState p_153133_, ServerLevel p_153134_, BlockPos p_153135_, Random p_153136_) {
+      // CraftBukkit start - do not fade if the block is valid here
+      if (p_153133_.m_60710_(p_153134_, p_153135_)) {
+         return;
+      }
+      // CraftBukkit end
       FarmBlock.m_53296_(p_153133_, p_153134_, p_153135_);
    }
 
