--- a/net/minecraft/world/level/block/DispenserBlock.java
+++ b/net/minecraft/world/level/block/DispenserBlock.java
@@ -40,10 +_,11 @@
 public class DispenserBlock extends BaseEntityBlock {
    public static final DirectionProperty f_52659_ = DirectionalBlock.f_52588_;
    public static final BooleanProperty f_52660_ = BlockStateProperties.f_61360_;
-   private static final Map<Item, DispenseItemBehavior> f_52661_ = Util.m_137469_(new Object2ObjectOpenHashMap<>(), (p_52723_) -> {
+   public static final Map<Item, DispenseItemBehavior> f_52661_ = Util.m_137469_(new Object2ObjectOpenHashMap<>(), (p_52723_) -> {
       p_52723_.defaultReturnValue(new DefaultDispenseItemBehavior());
    });
    private static final int f_153160_ = 4;
+   public static boolean eventFired = false; // CraftBukkit
 
    public static void m_52672_(ItemLike p_52673_, DispenseItemBehavior p_52674_) {
       f_52661_.put(p_52673_.m_5456_(), p_52674_);
@@ -72,7 +_,7 @@
       }
    }
 
-   protected void m_5824_(ServerLevel p_52665_, BlockPos p_52666_) {
+   public void m_5824_(ServerLevel p_52665_, BlockPos p_52666_) {
       BlockSourceImpl blocksourceimpl = new BlockSourceImpl(p_52665_, p_52666_);
       DispenserBlockEntity dispenserblockentity = blocksourceimpl.m_8118_();
       int i = dispenserblockentity.m_59248_();
@@ -83,6 +_,7 @@
          ItemStack itemstack = dispenserblockentity.m_8020_(i);
          DispenseItemBehavior dispenseitembehavior = this.m_7216_(itemstack);
          if (dispenseitembehavior != DispenseItemBehavior.f_123393_) {
+            eventFired = false; // CraftBukkit - reset event status
             dispenserblockentity.m_6836_(i, dispenseitembehavior.m_6115_(blocksourceimpl, itemstack));
          }
 
