--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -32,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DoorBlock extends Block {
    public static final DirectionProperty f_52726_ = HorizontalDirectionalBlock.f_54117_;
@@ -180,8 +_,22 @@
    }
 
    public void m_6861_(BlockState p_52776_, Level p_52777_, BlockPos p_52778_, Block p_52779_, BlockPos p_52780_, boolean p_52781_) {
-      boolean flag = p_52777_.m_46753_(p_52778_) || p_52777_.m_46753_(p_52778_.m_142300_(p_52776_.m_61143_(f_52730_) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
-      if (!this.m_49966_().m_60713_(p_52779_) && flag != p_52776_.m_61143_(f_52729_)) {
+      // CraftBukkit start
+      BlockPos otherHalf = p_52778_.m_142300_(p_52776_.m_61143_(f_52730_) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN);
+
+      org.bukkit.World bworld = p_52777_.getWorld();
+      org.bukkit.block.Block bukkitBlock = bworld.getBlockAt(p_52778_.m_123341_(), p_52778_.m_123342_(), p_52778_.m_123343_());
+      org.bukkit.block.Block blockTop = bworld.getBlockAt(otherHalf.m_123341_(), otherHalf.m_123342_(), otherHalf.m_123343_());
+
+      int power = bukkitBlock.getBlockPower();
+      int powerTop = blockTop.getBlockPower();
+      if (powerTop > power) power = powerTop;
+      int oldPower = (Boolean) p_52776_.m_61143_(f_52729_) ? 15 : 0;
+
+      if (oldPower == 0 ^ power == 0) {
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, oldPower, power);
+         p_52777_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+         boolean flag = eventRedstone.getNewCurrent() > 0;
          if (flag != p_52776_.m_61143_(f_52727_)) {
             this.m_52759_(p_52777_, p_52778_, flag);
             p_52777_.m_151555_(flag ? GameEvent.f_157796_ : GameEvent.f_157793_, p_52778_);
