--- a/net/minecraft/world/level/block/DragonEggBlock.java
+++ b/net/minecraft/world/level/block/DragonEggBlock.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockFromToEvent;
 
 public class DragonEggBlock extends FallingBlock {
    protected static final VoxelShape f_52908_ = Block.m_49796_(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
@@ -42,6 +_,18 @@
       for(int i = 0; i < 1000; ++i) {
          BlockPos blockpos = p_52938_.m_142082_(p_52937_.f_46441_.nextInt(16) - p_52937_.f_46441_.nextInt(16), p_52937_.f_46441_.nextInt(8) - p_52937_.f_46441_.nextInt(8), p_52937_.f_46441_.nextInt(16) - p_52937_.f_46441_.nextInt(16));
          if (p_52937_.m_8055_(blockpos).m_60795_() && worldborder.m_61937_(blockpos)) {
+            // CraftBukkit start
+            org.bukkit.block.Block from = p_52937_.getWorld().getBlockAt(p_52938_.m_123341_(), p_52938_.m_123342_(), p_52938_.m_123343_());
+            org.bukkit.block.Block to = p_52937_.getWorld().getBlockAt(blockpos.m_123341_(), blockpos.m_123342_(), blockpos.m_123343_());
+            BlockFromToEvent event = new BlockFromToEvent(from, to);
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+
+            blockpos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+            // CraftBukkit end
             if (p_52937_.f_46443_) {
                for(int j = 0; j < 128; ++j) {
                   double d0 = p_52937_.f_46441_.nextDouble();
