--- a/net/minecraft/world/level/block/EnchantmentTableBlock.java
+++ b/net/minecraft/world/level/block/EnchantmentTableBlock.java
@@ -1,6 +_,5 @@
 package net.minecraft.world.level.block;
 
-import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -31,18 +_,11 @@
 
 public class EnchantmentTableBlock extends BaseEntityBlock {
    protected static final VoxelShape f_52950_ = Block.m_49796_(0.0D, 0.0D, 0.0D, 16.0D, 12.0D, 16.0D);
-   public static final List<BlockPos> f_207902_ = BlockPos.m_121886_(-2, 0, -2, 2, 1, 2).filter((p_207914_) -> {
-      return Math.abs(p_207914_.m_123341_()) == 2 || Math.abs(p_207914_.m_123343_()) == 2;
-   }).map(BlockPos::m_7949_).toList();
 
    public EnchantmentTableBlock(BlockBehaviour.Properties p_52953_) {
       super(p_52953_);
    }
 
-   public static boolean m_207909_(Level p_207910_, BlockPos p_207911_, BlockPos p_207912_) {
-      return p_207910_.m_8055_(p_207911_.m_141952_(p_207912_)).m_60713_(Blocks.f_50078_) && p_207910_.m_46859_(p_207911_.m_142082_(p_207912_.m_123341_() / 2, p_207912_.m_123342_(), p_207912_.m_123343_() / 2));
-   }
-
    public boolean m_7923_(BlockState p_52997_) {
       return true;
    }
@@ -54,9 +_,24 @@
    public void m_7100_(BlockState p_52981_, Level p_52982_, BlockPos p_52983_, Random p_52984_) {
       super.m_7100_(p_52981_, p_52982_, p_52983_, p_52984_);
 
-      for(BlockPos blockpos : f_207902_) {
-         if (p_52984_.nextInt(16) == 0 && m_207909_(p_52982_, p_52983_, blockpos)) {
-            p_52982_.m_7106_(ParticleTypes.f_123809_, (double)p_52983_.m_123341_() + 0.5D, (double)p_52983_.m_123342_() + 2.0D, (double)p_52983_.m_123343_() + 0.5D, (double)((float)blockpos.m_123341_() + p_52984_.nextFloat()) - 0.5D, (double)((float)blockpos.m_123342_() - p_52984_.nextFloat() - 1.0F), (double)((float)blockpos.m_123343_() + p_52984_.nextFloat()) - 0.5D);
+      for(int i = -2; i <= 2; ++i) {
+         for(int j = -2; j <= 2; ++j) {
+            if (i > -2 && i < 2 && j == -1) {
+               j = 2;
+            }
+
+            if (p_52984_.nextInt(16) == 0) {
+               for(int k = 0; k <= 1; ++k) {
+                  BlockPos blockpos = p_52983_.m_142082_(i, k, j);
+                  if (p_52982_.m_8055_(blockpos).getEnchantPowerBonus(p_52982_, blockpos) > 0) {
+                     if (!p_52982_.m_46859_(p_52983_.m_142082_(i / 2, 0, j / 2))) {
+                        break;
+                     }
+
+                     p_52982_.m_7106_(ParticleTypes.f_123809_, (double)p_52983_.m_123341_() + 0.5D, (double)p_52983_.m_123342_() + 2.0D, (double)p_52983_.m_123343_() + 0.5D, (double)((float)i + p_52984_.nextFloat()) - 0.5D, (double)((float)k - p_52984_.nextFloat() - 1.0F), (double)((float)j + p_52984_.nextFloat()) - 0.5D);
+                  }
+               }
+            }
          }
       }
 
@@ -89,8 +_,8 @@
       BlockEntity blockentity = p_52994_.m_7702_(p_52995_);
       if (blockentity instanceof EnchantmentTableBlockEntity) {
          Component component = ((Nameable)blockentity).m_5446_();
-         return new SimpleMenuProvider((p_207906_, p_207907_, p_207908_) -> {
-            return new EnchantmentMenu(p_207906_, p_207907_, ContainerLevelAccess.m_39289_(p_52994_, p_52995_));
+         return new SimpleMenuProvider((p_52959_, p_52960_, p_52961_) -> {
+            return new EnchantmentMenu(p_52959_, p_52960_, ContainerLevelAccess.m_39289_(p_52994_, p_52995_));
          }, component);
       } else {
          return null;
