--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -5,6 +_,7 @@
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.BlockGetter;
@@ -18,6 +_,9 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.Server;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndPortalBlock extends BaseEntityBlock {
    protected static final VoxelShape f_53014_ = Block.m_49796_(0.0D, 6.0D, 0.0D, 16.0D, 12.0D, 16.0D);
@@ -41,7 +_,15 @@
          if (serverlevel == null) {
             return;
          }
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_53028_.getBukkitEntity(), new org.bukkit.Location(p_53026_.getWorld(), p_53027_.m_123341_(), p_53027_.m_123342_(), p_53027_.m_123343_()));
+         p_53026_.getCraftServer().getPluginManager().callEvent(event);
 
+         if (p_53028_ instanceof ServerPlayer) {
+            //((ServerPlayer) p_53028_).changeDimension(serverlevel, PlayerTeleportEvent.TeleportCause.END_PORTAL); FoxServer TODO
+            //return;
+         }
+         // CraftBukkit end
          p_53028_.m_5489_(serverlevel);
       }
 
