--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -19,6 +_,12 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+// CraftBukkit end
+
 public class EndPortalBlock extends BaseEntityBlock {
    protected static final VoxelShape f_53014_ = Block.m_49796_(0.0D, 6.0D, 0.0D, 16.0D, 12.0D, 16.0D);
 
@@ -36,11 +_,20 @@
 
    public void m_7892_(BlockState p_53025_, Level p_53026_, BlockPos p_53027_, Entity p_53028_) {
       if (p_53026_ instanceof ServerLevel && p_53028_.m_6072_() && Shapes.m_83157_(Shapes.m_83064_(p_53028_.m_20191_().m_82386_((double)(-p_53027_.m_123341_()), (double)(-p_53027_.m_123342_()), (double)(-p_53027_.m_123343_()))), p_53025_.m_60808_(p_53026_, p_53027_), BooleanOp.f_82689_)) {
-         ResourceKey<Level> resourcekey = p_53026_.m_46472_() == Level.f_46430_ ? Level.f_46428_ : Level.f_46430_;
+         ResourceKey<Level> resourcekey = p_53026_.m_46472_() == Level.f_46430_ ? Level.f_46428_ : Level.f_46430_; // CraftBukkit - SPIGOT-6152: send back to main overworld in custom ends // CatServer - recover
          ServerLevel serverlevel = ((ServerLevel)p_53026_).m_7654_().m_129880_(resourcekey);
          if (serverlevel == null) {
+            // return; // CraftBukkit - always fire event in case plugins wish to change it
+         }
+
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_53028_.getBukkitEntity(), new org.bukkit.Location(p_53026_.getWorld(), p_53027_.m_123341_(), p_53027_.m_123342_(), p_53027_.m_123343_()));
+         p_53026_.getCraftServer().getPluginManager().callEvent(event);
+         if (p_53028_ instanceof ServerPlayer) {
+            ((ServerPlayer) p_53028_).changeDimension(serverlevel, PlayerTeleportEvent.TeleportCause.END_PORTAL);
             return;
          }
+         // CraftBukkit end
 
          p_53028_.m_5489_(serverlevel);
       }
