--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -6,11 +_,9 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
@@ -23,6 +_,8 @@
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class FarmBlock extends Block {
    public static final IntegerProperty f_53243_ = BlockStateProperties.f_61423_;
@@ -70,31 +_,54 @@
       int i = p_53285_.m_61143_(f_53243_);
       if (!m_53258_(p_53286_, p_53287_) && !p_53286_.m_46758_(p_53287_.m_7494_())) {
          if (i > 0) {
-            p_53286_.m_7731_(p_53287_, p_53285_.m_61124_(f_53243_, Integer.valueOf(i - 1)), 2);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleMoistureChangeEvent(p_53286_, p_53287_, p_53285_.m_61124_(f_53243_, i - 1), 2); // CraftBukkit
          } else if (!m_53250_(p_53286_, p_53287_)) {
             m_53296_(p_53285_, p_53286_, p_53287_);
          }
       } else if (i < 7) {
-         p_53286_.m_7731_(p_53287_, p_53285_.m_61124_(f_53243_, Integer.valueOf(7)), 2);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleMoistureChangeEvent(p_53286_, p_53287_, p_53285_.m_61124_(f_53243_, 7), 2); // CraftBukkit
       }
 
    }
 
    public void m_142072_(Level p_153227_, BlockState p_153228_, BlockPos p_153229_, Entity p_153230_, float p_153231_) {
-      if (!p_153227_.f_46443_ && p_153227_.f_46441_.nextFloat() < p_153231_ - 0.5F && p_153230_ instanceof LivingEntity && (p_153230_ instanceof Player || p_153227_.m_46469_().m_46207_(GameRules.f_46132_)) && p_153230_.m_20205_() * p_153230_.m_20205_() * p_153230_.m_20206_() > 0.512F) {
+      super.m_142072_(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
+      if (!p_153227_.f_46443_ && net.minecraftforge.common.ForgeHooks.onFarmlandTrample(p_153227_, p_153229_, Blocks.f_50493_.m_49966_(), p_153231_, p_153230_)) { // Forge: Move logic to Entity#canTrample
+         // CraftBukkit start - Interact soil
+         org.bukkit.event.Cancellable cancellable;
+         if (p_153230_ instanceof Player) {
+            cancellable = CraftEventFactory.callPlayerInteractEvent((Player) p_153230_, org.bukkit.event.block.Action.PHYSICAL, p_153229_, null, null, null);
+         } else {
+            cancellable = new EntityInteractEvent(p_153230_.getBukkitEntity(), p_153227_.getWorld().getBlockAt(p_153229_.m_123341_(), p_153229_.m_123342_(), p_153229_.m_123343_()));
+            p_153227_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+         }
+
+         if (cancellable.isCancelled()) {
+            return;
+         }
+         if (CraftEventFactory.callEntityChangeBlockEvent(p_153230_, p_153229_, Blocks.f_50493_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          m_53296_(p_153228_, p_153227_, p_153229_);
       }
 
-      super.m_142072_(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_);
+      //super.fallOn(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_);
    }
 
    public static void m_53296_(BlockState p_53297_, Level p_53298_, BlockPos p_53299_) {
+      // CraftBukkit start
+      if (CraftEventFactory.callBlockFadeEvent(p_53298_, p_53299_, Blocks.f_50493_.m_49966_()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_53298_.m_46597_(p_53299_, m_49897_(p_53297_, Blocks.f_50493_.m_49966_(), p_53298_, p_53299_));
    }
 
    private static boolean m_53250_(BlockGetter p_53251_, BlockPos p_53252_) {
-      Block block = p_53251_.m_8055_(p_53252_.m_7494_()).m_60734_();
-      return block instanceof CropBlock || block instanceof StemBlock || block instanceof AttachedStemBlock;
+      BlockState plant = p_53251_.m_8055_(p_53252_.m_7494_());
+      BlockState state = p_53251_.m_8055_(p_53252_);
+      return plant.m_60734_() instanceof net.minecraftforge.common.IPlantable && state.canSustainPlant(p_53251_, p_53252_, Direction.UP, (net.minecraftforge.common.IPlantable)plant.m_60734_());
    }
 
    private static boolean m_53258_(LevelReader p_53259_, BlockPos p_53260_) {
@@ -104,7 +_,7 @@
          }
       }
 
-      return false;
+      return net.minecraftforge.common.FarmlandWaterManager.hasBlockWaterTicket(p_53259_, p_53260_);
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_53283_) {
