--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -27,6 +_,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityInteractEvent;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class FarmBlock extends Block {
    public static final IntegerProperty f_53243_ = BlockStateProperties.f_61423_;
    protected static final VoxelShape f_53244_ = Block.m_49796_(0.0D, 0.0D, 0.0D, 16.0D, 15.0D, 16.0D);
@@ -73,42 +_,66 @@
       int i = p_221139_.m_61143_(f_53243_);
       if (!m_53258_(p_221140_, p_221141_) && !p_221140_.m_46758_(p_221141_.m_7494_())) {
          if (i > 0) {
-            p_221140_.m_7731_(p_221141_, p_221139_.m_61124_(f_53243_, Integer.valueOf(i - 1)), 2);
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleMoistureChangeEvent(p_221140_, p_221141_, p_221139_.m_61124_(f_53243_, i - 1), 2); // CraftBukkit
          } else if (!m_53250_(p_221140_, p_221141_)) {
             m_269406_((Entity)null, p_221139_, p_221140_, p_221141_);
          }
       } else if (i < 7) {
-         p_221140_.m_7731_(p_221141_, p_221139_.m_61124_(f_53243_, Integer.valueOf(7)), 2);
+         org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleMoistureChangeEvent(p_221140_, p_221141_, p_221139_.m_61124_(f_53243_, 7), 2); // CraftBukkit
       }
 
    }
 
    public void m_142072_(Level p_153227_, BlockState p_153228_, BlockPos p_153229_, Entity p_153230_, float p_153231_) {
-      if (!p_153227_.f_46443_ && p_153227_.f_46441_.m_188501_() < p_153231_ - 0.5F && p_153230_ instanceof LivingEntity && (p_153230_ instanceof Player || p_153227_.m_46469_().m_46207_(GameRules.f_46132_)) && p_153230_.m_20205_() * p_153230_.m_20205_() * p_153230_.m_20206_() > 0.512F) {
+      super.m_142072_(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
+      if (!p_153227_.f_46443_ && net.minecraftforge.common.ForgeHooks.onFarmlandTrample(p_153227_, p_153229_, Blocks.f_50493_.m_49966_(), p_153231_, p_153230_)) { // Forge: Move logic to Entity#canTrample
+         // CraftBukkit start - Interact soil
+         org.bukkit.event.Cancellable cancellable;
+         if (p_153230_ instanceof Player) {
+            cancellable = CraftEventFactory.callPlayerInteractEvent((Player) p_153230_, org.bukkit.event.block.Action.PHYSICAL, p_153229_, null, null, null);
+         } else {
+            cancellable = new EntityInteractEvent(p_153230_.getBukkitEntity(), p_153227_.getWorld().getBlockAt(p_153229_.m_123341_(), p_153229_.m_123342_(), p_153229_.m_123343_()));
+            p_153227_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+         }
+         if (cancellable.isCancelled()) {
+            return;
+         }
+         if (!CraftEventFactory.callEntityChangeBlockEvent(p_153230_, p_153229_, Blocks.f_50493_.m_49966_())) {
+            return;
+         }
+         // CraftBukkit end
          m_269406_(p_153230_, p_153228_, p_153227_, p_153229_);
       }
 
-      super.m_142072_(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_);
+      // super.fallOn(p_153227_, p_153228_, p_153229_, p_153230_, p_153231_); // CraftBukkit - moved up
    }
 
    public static void m_269406_(@Nullable Entity p_270981_, BlockState p_270402_, Level p_270568_, BlockPos p_270551_) {
+      // CraftBukkit start
+      if (CraftEventFactory.callBlockFadeEvent(p_270568_, p_270551_, Blocks.f_50493_.m_49966_()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       BlockState blockstate = m_49897_(p_270402_, Blocks.f_50493_.m_49966_(), p_270568_, p_270551_);
       p_270568_.m_46597_(p_270551_, blockstate);
       p_270568_.m_220407_(GameEvent.f_157792_, p_270551_, GameEvent.Context.m_223719_(p_270981_, blockstate));
    }
 
    private static boolean m_53250_(BlockGetter p_279219_, BlockPos p_279209_) {
-      return p_279219_.m_8055_(p_279209_.m_7494_()).m_204336_(BlockTags.f_279534_);
+      BlockState plant = p_279219_.m_8055_(p_279209_.m_7494_());
+      BlockState state = p_279219_.m_8055_(p_279209_);
+      return plant.m_60734_() instanceof net.minecraftforge.common.IPlantable && state.canSustainPlant(p_279219_, p_279209_, Direction.UP, (net.minecraftforge.common.IPlantable)plant.m_60734_());
    }
 
    private static boolean m_53258_(LevelReader p_53259_, BlockPos p_53260_) {
+      BlockState state = p_53259_.m_8055_(p_53260_);
       for(BlockPos blockpos : BlockPos.m_121940_(p_53260_.m_7918_(-4, 0, -4), p_53260_.m_7918_(4, 1, 4))) {
-         if (p_53259_.m_6425_(blockpos).m_205070_(FluidTags.f_13131_)) {
+         if (state.canBeHydrated(p_53259_, p_53260_, p_53259_.m_6425_(blockpos), blockpos)) {
             return true;
          }
       }
 
-      return false;
+      return net.minecraftforge.common.FarmlandWaterManager.hasBlockWaterTicket(p_53259_, p_53260_);
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_53283_) {
