--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -126,6 +_,17 @@
    public void m_6861_(BlockState p_53372_, Level p_53373_, BlockPos p_53374_, Block p_53375_, BlockPos p_53376_, boolean p_53377_) {
       if (!p_53373_.f_46443_) {
          boolean flag = p_53373_.m_46753_(p_53374_);
+         // CraftBukkit start
+         boolean oldPowered = p_53372_.m_61143_(FenceGateBlock.f_53342_);
+         if (oldPowered != flag) {
+            int newPower = flag ? 15 : 0;
+            int oldPower = oldPowered ? 15 : 0;
+            org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_53373_, p_53374_);
+            org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(bukkitBlock, oldPower, newPower);
+            p_53373_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+            flag = eventRedstone.getNewCurrent() > 0;
+         }
+         // CraftBukkit end
          if (p_53372_.m_61143_(f_53342_) != flag) {
             p_53373_.m_7731_(p_53374_, p_53372_.m_61124_(f_53342_, Boolean.valueOf(flag)).m_61124_(f_53341_, Boolean.valueOf(flag)), 2);
             if (p_53372_.m_61143_(f_53341_) != flag) {
