--- a/net/minecraft/world/level/block/FungusBlock.java
+++ b/net/minecraft/world/level/block/FungusBlock.java
@@ -44,6 +_,13 @@
    }
 
    public void m_7719_(ServerLevel p_53603_, Random p_53604_, BlockPos p_53605_, BlockState p_53606_) {
+      // CraftBukkit start
+      if (this == Blocks.f_50691_) {
+         SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+      } else if (this == Blocks.f_50700_) {
+         SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+      }
+      // CraftBukkit end
       this.f_53597_.get().m_203334_().m_65385_(p_53603_, p_53603_.m_7726_().m_8481_(), p_53604_, p_53605_);
    }
 }
