--- a/net/minecraft/world/level/block/IceBlock.java
+++ b/net/minecraft/world/level/block/IceBlock.java
@@ -45,6 +_,11 @@
    }
 
    protected void m_54168_(BlockState p_54169_, Level p_54170_, BlockPos p_54171_) {
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_54170_, p_54171_, p_54170_.m_6042_().m_63951_() ? Blocks.f_50016_.m_49966_() : Blocks.f_49990_.m_49966_()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       if (p_54170_.m_6042_().m_63951_()) {
          p_54170_.m_7471_(p_54171_, false);
       } else {
