--- a/net/minecraft/world/level/block/InfestedBlock.java
+++ b/net/minecraft/world/level/block/InfestedBlock.java
@@ -39,7 +_,7 @@
       Silverfish silverfish = EntityType.f_20523_.m_20615_(p_54181_);
       if (silverfish != null) {
          silverfish.m_7678_((double)p_54182_.m_123341_() + 0.5D, (double)p_54182_.m_123342_(), (double)p_54182_.m_123343_() + 0.5D, 0.0F, 0.0F);
-         p_54181_.m_7967_(silverfish);
+         p_54181_.addFreshEntity(silverfish, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SILVERFISH_BLOCK); // CraftBukkit - add SpawnReason
          silverfish.m_21373_();
       }
 
