--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -54,12 +_,18 @@
    public void m_54269_(LevelAccessor p_54270_, BlockPos p_54271_, BlockState p_54272_, ItemStack p_54273_) {
       BlockEntity blockentity = p_54270_.m_7702_(p_54271_);
       if (blockentity instanceof JukeboxBlockEntity) {
-         ((JukeboxBlockEntity)blockentity).m_59517_(p_54273_.m_41777_());
+         // CraftBukkit start - There can only be one
+         p_54273_ = p_54273_.m_41777_();
+         if (!p_54273_.m_41619_()) {
+            p_54273_.m_41764_(1);
+         }
+         ((JukeboxBlockEntity) blockentity).m_59517_(p_54273_);
+         // CraftBukkit end
          p_54270_.m_7731_(p_54271_, p_54272_.m_61124_(f_54254_, Boolean.valueOf(true)), 2);
       }
    }
 
-   private void m_54260_(Level p_54261_, BlockPos p_54262_) {
+   public void m_54260_(Level p_54261_, BlockPos p_54262_) {
       if (!p_54261_.f_46443_) {
          BlockEntity blockentity = p_54261_.m_7702_(p_54262_);
          if (blockentity instanceof JukeboxBlockEntity) {
