--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -59,7 +_,7 @@
       }
    }
 
-   private void m_54260_(Level p_54261_, BlockPos p_54262_) {
+   public void m_54260_(Level p_54261_, BlockPos p_54262_) {
       if (!p_54261_.f_46443_) {
          BlockEntity blockentity = p_54261_.m_7702_(p_54262_);
          if (blockentity instanceof JukeboxBlockEntity) {
