--- a/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -2,6 +_,7 @@
 
 import java.util.Map;
 import java.util.function.Predicate;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.cauldron.CauldronInteraction;
 import net.minecraft.world.entity.Entity;
@@ -15,6 +_,9 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockState;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockStates;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class LayeredCauldronBlock extends AbstractCauldronBlock {
    public static final int f_153512_ = 1;
@@ -45,15 +_,19 @@
    }
 
    protected double m_142446_(BlockState p_153528_) {
-      return (6.0D + (double)p_153528_.m_61143_(f_153514_).intValue() * 3.0D) / 16.0D;
+      return (6.0D + (double) p_153528_.m_61143_(f_153514_).intValue() * 3.0D) / 16.0D;
    }
 
    public void m_7892_(BlockState p_153534_, Level p_153535_, BlockPos p_153536_, Entity p_153537_) {
       if (!p_153535_.f_46443_ && p_153537_.m_6060_() && this.m_151979_(p_153534_, p_153536_, p_153537_)) {
-         p_153537_.m_20095_();
+         // CraftBukkit start
          if (p_153537_.m_142265_(p_153535_, p_153536_)) {
-            this.m_142266_(p_153534_, p_153535_, p_153536_);
+            if (!lowerFillLevel(p_153534_, p_153535_, p_153536_, p_153537_, CauldronLevelChangeEvent.ChangeReason.EXTINGUISH)) {
+               return;
+            }
          }
+         p_153537_.m_20095_();
+         // CraftBukkit end
       }
 
    }
@@ -63,13 +_,37 @@
    }
 
    public static void m_153559_(BlockState p_153560_, Level p_153561_, BlockPos p_153562_) {
+      // CraftBukkit start
+      lowerFillLevel(p_153560_, p_153561_, p_153562_, null, CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
+   }
+
+   public static boolean lowerFillLevel(BlockState p_153560_, Level p_153561_, BlockPos p_153562_, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
       int i = p_153560_.m_61143_(f_153514_) - 1;
-      p_153561_.m_46597_(p_153562_, i == 0 ? Blocks.f_50256_.m_49966_() : p_153560_.m_61124_(f_153514_, Integer.valueOf(i)));
-   }
+      return changeLevel(p_153560_, p_153561_, p_153562_, i == 0 ? Blocks.f_50256_.m_49966_() : p_153560_.m_61124_(LayeredCauldronBlock.f_153514_, i), entity, reason);
+   }
+
+   // CraftBukkit start
+   public static boolean changeLevel(BlockState iblockdata, Level world, BlockPos blockposition, BlockState newBlock, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+      CraftBlockState newState = CraftBlockStates.getBlockState(world, blockposition);
+      newState.setData(newBlock);
+
+      CauldronLevelChangeEvent event = new CauldronLevelChangeEvent(
+              world.getWorld().getBlockAt(blockposition.m_123341_(), blockposition.m_123342_(), blockposition.m_123343_()),
+              (entity == null) ? null : entity.getBukkitEntity(), reason, newState
+      );
+      world.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return false;
+      }
+      newState.update(true);
+      return true;
+   }
+   // CraftBukkit end
+
 
    public void m_141997_(BlockState p_153539_, Level p_153540_, BlockPos p_153541_, Biome.Precipitation p_153542_) {
       if (CauldronBlock.m_182450_(p_153540_, p_153542_) && p_153539_.m_61143_(f_153514_) != 3 && this.f_153519_.test(p_153542_)) {
-         p_153540_.m_46597_(p_153541_, p_153539_.m_61122_(f_153514_));
+         changeLevel(p_153539_, p_153540_, p_153541_, (BlockState) p_153539_.m_61122_(LayeredCauldronBlock.f_153514_), null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
       }
    }
 
@@ -83,7 +_,11 @@
 
    protected void m_142310_(BlockState p_153544_, Level p_153545_, BlockPos p_153546_, Fluid p_153547_) {
       if (!this.m_142596_(p_153544_)) {
-         p_153545_.m_46597_(p_153546_, p_153544_.m_61124_(f_153514_, Integer.valueOf(p_153544_.m_61143_(f_153514_) + 1)));
+         // CraftBukkit start
+         if (!changeLevel(p_153544_, p_153545_, p_153546_, (BlockState) p_153544_.m_61124_(LayeredCauldronBlock.f_153514_, (Integer) p_153544_.m_61143_(LayeredCauldronBlock.f_153514_) + 1), null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) {
+            return;
+         }
+         // CraftBukkit end
          p_153545_.m_46796_(1047, p_153546_, 0);
       }
    }
