--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -18,8 +_,9 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.LeavesDecayEvent;
 
-public class LeavesBlock extends Block {
+public class LeavesBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final int f_153563_ = 7;
    public static final IntegerProperty f_54418_ = BlockStateProperties.f_61414_;
    public static final BooleanProperty f_54419_ = BlockStateProperties.f_61447_;
@@ -40,6 +_,14 @@
 
    public void m_7455_(BlockState p_54451_, ServerLevel p_54452_, BlockPos p_54453_, Random p_54454_) {
       if (!p_54451_.m_61143_(f_54419_) && p_54451_.m_61143_(f_54418_) == 7) {
+         // CraftBukkit start
+         LeavesDecayEvent event = new LeavesDecayEvent(p_54452_.getWorld().getBlockAt(p_54453_.m_123341_(), p_54453_.m_123342_(), p_54453_.m_123343_()));
+         p_54452_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (event.isCancelled() || p_54452_.m_8055_(p_54453_).m_60734_() != this) {
+            return;
+         }
+         // CraftBukkit end
          m_49950_(p_54451_, p_54452_, p_54453_);
          p_54452_.m_7471_(p_54453_, false);
       }
