--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -23,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
+public class LeavesBlock extends Block implements SimpleWaterloggedBlock, net.minecraftforge.common.IForgeShearable {
    public static final int f_153563_ = 7;
    public static final IntegerProperty f_54418_ = BlockStateProperties.f_61414_;
    public static final BooleanProperty f_54419_ = BlockStateProperties.f_61447_;
@@ -45,6 +_,13 @@
 
    public void m_213898_(BlockState p_221379_, ServerLevel p_221380_, BlockPos p_221381_, RandomSource p_221382_) {
       if (this.m_221385_(p_221379_)) {
+         // CraftBukkit start
+         org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(p_221380_.getWorld().getBlockAt(p_221381_.m_123341_(), p_221381_.m_123342_(), p_221381_.m_123343_()));
+         p_221380_.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled() || p_221380_.m_8055_(p_221381_).m_60734_() != this) {
+            return;
+         }
+         // CraftBukkit end
          m_49950_(p_221379_, p_221380_, p_221381_);
          p_221380_.m_7471_(p_221381_, false);
       }
