--- a/net/minecraft/world/level/block/LecternBlock.java
+++ b/net/minecraft/world/level/block/LecternBlock.java
@@ -181,10 +_,11 @@
    }
 
    private void m_54587_(BlockState p_54588_, Level p_54589_, BlockPos p_54590_) {
-      BlockEntity blockentity = p_54589_.m_7702_(p_54590_);
+      BlockEntity blockentity = p_54589_.getBlockEntity(p_54590_, false); // CraftBukkit - don't validate, type may be changed already
       if (blockentity instanceof LecternBlockEntity lecternblockentity) {
          Direction direction = p_54588_.m_61143_(f_54465_);
          ItemStack itemstack = lecternblockentity.m_59566_().m_41777_();
+         if (itemstack.m_41619_()) return; // CraftBukkit - SPIGOT-5500
          float f = 0.25F * (float)direction.m_122429_();
          float f1 = 0.25F * (float)direction.m_122431_();
          ItemEntity itementity = new ItemEntity(p_54589_, (double)p_54590_.m_123341_() + 0.5D + (double)f, (double)(p_54590_.m_123342_() + 1), (double)p_54590_.m_123343_() + 0.5D + (double)f1, itemstack);
