--- a/net/minecraft/world/level/block/LeverBlock.java
+++ b/net/minecraft/world/level/block/LeverBlock.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class LeverBlock extends FaceAttachedHorizontalDirectionalBlock {
    public static final BooleanProperty f_54622_ = BlockStateProperties.f_61448_;
@@ -85,6 +_,17 @@
 
          return InteractionResult.SUCCESS;
       } else {
+         // CraftBukkit start - Interact Lever
+         boolean powered = p_54640_.m_61143_(f_54622_); // Old powered state
+         org.bukkit.block.Block block = p_54641_.getWorld().getBlockAt(p_54642_.m_123341_(), p_54642_.m_123342_(), p_54642_.m_123343_());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_54641_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          BlockState blockstate = this.m_54676_(p_54640_, p_54641_, p_54642_);
          float f = blockstate.m_61143_(f_54622_) ? 0.6F : 0.5F;
          p_54641_.m_5594_((Player)null, p_54642_, SoundEvents.f_12088_, SoundSource.BLOCKS, 0.3F, f);
