--- a/net/minecraft/world/level/block/LightningRodBlock.java
+++ b/net/minecraft/world/level/block/LightningRodBlock.java
@@ -30,6 +_,8 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class LightningRodBlock extends RodBlock implements SimpleWaterloggedBlock {
    public static final BooleanProperty f_153702_ = BlockStateProperties.f_61362_;
@@ -70,6 +_,18 @@
    }
 
    public void m_153760_(BlockState p_153761_, Level p_153762_, BlockPos p_153763_) {
+      // CraftBukkit start
+      boolean powered = p_153761_.m_61143_(LightningRodBlock.f_153703_);
+      int old = (powered) ? 15 : 0;
+      int current = (!powered) ? 15 : 0;
+
+      BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(CraftBlock.at(p_153762_, p_153763_), old, current);
+      p_153762_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+      if (eventRedstone.getNewCurrent() <= 0) {
+         return;
+      }
+      // CraftBukkit end
       p_153762_.m_7731_(p_153763_, p_153761_.m_61124_(f_153703_, Boolean.valueOf(true)), 3);
       this.m_153764_(p_153761_, p_153762_, p_153763_);
       p_153762_.m_186460_(p_153763_, this, 8);
@@ -118,7 +_,7 @@
             lightningbolt.m_20219_(Vec3.m_82539_(blockpos.m_7494_()));
             Entity entity = p_153716_.m_37282_();
             lightningbolt.m_20879_(entity instanceof ServerPlayer ? (ServerPlayer)entity : null);
-            p_153713_.m_7967_(lightningbolt);
+            ((ServerLevel) p_153713_).strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRIDENT); // CraftBukkit
             p_153713_.m_5594_((Player)null, blockpos, SoundEvents.f_12521_, SoundSource.WEATHER, 5.0F, 1.0F);
          }
       }
