--- a/net/minecraft/world/level/block/MagmaBlock.java
+++ b/net/minecraft/world/level/block/MagmaBlock.java
@@ -27,7 +_,9 @@
 
    public void m_141947_(Level p_153777_, BlockPos p_153778_, BlockState p_153779_, Entity p_153780_) {
       if (!p_153780_.m_5825_() && p_153780_ instanceof LivingEntity && !EnchantmentHelper.m_44938_((LivingEntity)p_153780_)) {
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.blockDamage = p_153777_.getWorld().getBlockAt(p_153778_.m_123341_(), p_153778_.m_123342_(), p_153778_.m_123343_()); // CraftBukkit
          p_153780_.m_6469_(DamageSource.f_19309_, 1.0F);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.blockDamage = null; // CraftBukkit
       }
 
       super.m_141947_(p_153777_, p_153778_, p_153779_, p_153780_);
