--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -136,14 +_,19 @@
                p_221702_.m_46865_(p_221703_.f_221717_()).m_8113_(p_221703_.f_221717_());
             }
 
-            return p_221702_.m_7731_(p_221703_.f_221717_(), blockstate, 2);
+            return org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockSpreadEvent(p_221702_, p_221703_.source(), p_221703_.f_221717_(), blockstate, 2); // CraftBukkit - 1.20.1
          } else {
             return false;
          }
       }
    }
 
-   public static record SpreadPos(BlockPos f_221717_, Direction f_221718_) {
+   public static record SpreadPos(BlockPos f_221717_, Direction f_221718_, BlockPos source) { // CraftBukkit - 1.20.1
+      // CatServer start - 1.20.1
+      public SpreadPos(BlockPos f_221717_, Direction f_221718_) {
+         this(f_221717_, f_221718_, f_221717_);
+      }
+      // CatServer end
    }
 
    @FunctionalInterface
@@ -154,17 +_,17 @@
    public static enum SpreadType {
       SAME_POSITION {
          public MultifaceSpreader.SpreadPos m_213941_(BlockPos p_221751_, Direction p_221752_, Direction p_221753_) {
-            return new MultifaceSpreader.SpreadPos(p_221751_, p_221752_);
+            return new MultifaceSpreader.SpreadPos(p_221751_, p_221752_, p_221751_); // CraftBukkit - 1.20.1
          }
       },
       SAME_PLANE {
          public MultifaceSpreader.SpreadPos m_213941_(BlockPos p_221758_, Direction p_221759_, Direction p_221760_) {
-            return new MultifaceSpreader.SpreadPos(p_221758_.m_121945_(p_221759_), p_221760_);
+            return new MultifaceSpreader.SpreadPos(p_221758_.m_121945_(p_221759_), p_221760_, p_221758_); // CraftBukkit - 1.20.1
          }
       },
       WRAP_AROUND {
          public MultifaceSpreader.SpreadPos m_213941_(BlockPos p_221765_, Direction p_221766_, Direction p_221767_) {
-            return new MultifaceSpreader.SpreadPos(p_221765_.m_121945_(p_221766_).m_121945_(p_221767_), p_221766_.m_122424_());
+            return new MultifaceSpreader.SpreadPos(p_221765_.m_121945_(p_221766_).m_121945_(p_221767_), p_221766_.m_122424_(), p_221765_); // CraftBukkit - 1.20.1
          }
       };
 
