--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -231,6 +_,7 @@
       }
    }
 
+   // FoxServer TODO
    private boolean m_153834_(LevelAccessor p_153835_, BlockPos p_153836_, Direction p_153837_, boolean p_153838_) {
       BlockState blockstate = p_153835_.m_8055_(p_153836_);
       BlockState blockstate1 = this.m_153940_(blockstate, p_153835_, p_153836_, p_153837_);
