--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -190,7 +_,7 @@
       Optional<Pair<BlockPos, Direction>> optional = this.m_153855_(p_153867_, p_153868_, p_153869_, p_153870_, p_153871_);
       if (optional.isPresent()) {
          Pair<BlockPos, Direction> pair = optional.get();
-         return this.m_153834_(p_153868_, pair.getFirst(), pair.getSecond(), p_153872_);
+         return this.spreadToFace(p_153868_, pair.getFirst(), pair.getSecond(), p_153872_, p_153869_); // CraftBukkit
       } else {
          return false;
       }
@@ -231,6 +_,18 @@
       }
    }
 
+   private boolean spreadToFace(LevelAccessor p_153835_, BlockPos p_153836_, Direction p_153837_, boolean p_153838_, BlockPos source) {
+      // CatServer start
+      catserver.server.CatServerCaptures.getCatServerCaptures().captureBlockPos(source);
+      try {
+         return this.m_153834_(p_153835_, p_153836_, p_153837_, p_153838_);
+      } finally {
+         // CatServer - Make sure reset
+         catserver.server.CatServerCaptures.getCatServerCaptures().getCaptureBlockPos(); // CatServer
+      }
+      // CatServer end
+   }
+
    private boolean m_153834_(LevelAccessor p_153835_, BlockPos p_153836_, Direction p_153837_, boolean p_153838_) {
       BlockState blockstate = p_153835_.m_8055_(p_153836_);
       BlockState blockstate1 = this.m_153940_(blockstate, p_153835_, p_153836_, p_153837_);
@@ -239,7 +_,12 @@
             p_153835_.m_46865_(p_153836_).m_8113_(p_153836_);
          }
 
-         return p_153835_.m_7731_(p_153836_, blockstate1, 2);
+         var source = catserver.server.CatServerCaptures.getCatServerCaptures().getCaptureBlockPos(); // CatServer
+         if (source != null) {
+            return org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_153835_, source, p_153836_, blockstate1, 2); // CraftBukkit
+         } else {
+            return p_153835_.m_7731_(p_153836_, blockstate1, 2);
+         }
       } else {
          return false;
       }
