--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -32,7 +_,7 @@
    }
 
    public void m_213898_(BlockState p_221784_, ServerLevel p_221785_, BlockPos p_221786_, RandomSource p_221787_) {
-      if (p_221787_.m_188503_(25) == 0) {
+      if (p_221787_.m_188501_() < (p_221785_.spigotConfig.mushroomModifier / (100.0f * 25))) { // Spigot - SPIGOT-7159: Better modifier resolution
          int i = 5;
          int j = 4;
 
@@ -56,7 +_,7 @@
          }
 
          if (p_221785_.m_46859_(blockpos1) && p_221784_.m_60710_(p_221785_, blockpos1)) {
-            p_221785_.m_7731_(blockpos1, p_221784_, 2);
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockSpreadEvent(p_221785_, p_221786_, blockpos1, p_221784_, 2); // CraftBukkit
          }
       }
 
@@ -72,7 +_,7 @@
       if (blockstate.m_204336_(BlockTags.f_13057_)) {
          return true;
       } else {
-         return p_54881_.m_45524_(p_54882_, 0) < 13 && this.m_6266_(blockstate, p_54881_, blockpos);
+         return p_54881_.m_45524_(p_54882_, 0) < 13 && blockstate.canSustainPlant(p_54881_, blockpos, net.minecraft.core.Direction.UP, this);
       }
    }
 
@@ -81,8 +_,11 @@
       if (optional.isEmpty()) {
          return false;
       } else {
+         var event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_221774_, p_221777_, p_221775_, optional.get());
+         if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY)) return false;
          p_221774_.m_7471_(p_221775_, false);
-         if (optional.get().m_203334_().m_224953_(p_221774_, p_221774_.m_7726_().m_8481_(), p_221777_, p_221775_)) {
+         SaplingBlock.treeType = (this == Blocks.f_50072_) ? org.bukkit.TreeType.BROWN_MUSHROOM : org.bukkit.TreeType.RED_MUSHROOM; // CraftBukkit // CatServer - BROWN_MUSHROOM -> RED_MUSHROOM craftbukkit fixed in higher versions
+         if (event.getFeature().m_203334_().m_224953_(p_221774_, p_221774_.m_7726_().m_8481_(), p_221777_, p_221775_)) {
             return true;
          } else {
             p_221774_.m_7731_(p_221775_, p_221776_, 3);
