--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.TreeType;
 
 public class MushroomBlock extends BushBlock implements BonemealableBlock {
    protected static final float f_153980_ = 3.0F;
@@ -30,7 +_,7 @@
    }
 
    public void m_7455_(BlockState p_54884_, ServerLevel p_54885_, BlockPos p_54886_, Random p_54887_) {
-      if (p_54887_.nextInt(25) == 0) {
+      if (p_54887_.nextInt(Math.max(1, (int) (100.0F / p_54885_.spigotConfig.mushroomModifier) * 25)) == 0) { // Spigot
          int i = 5;
          int j = 4;
 
@@ -54,7 +_,7 @@
          }
 
          if (p_54885_.m_46859_(blockpos1) && p_54884_.m_60710_(p_54885_, blockpos1)) {
-            p_54885_.m_7731_(blockpos1, p_54884_, 2);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_54885_, p_54886_, blockpos1, p_54884_, 2); // CraftBukkit
          }
       }
 
@@ -70,12 +_,13 @@
       if (blockstate.m_204336_(BlockTags.f_13057_)) {
          return true;
       } else {
-         return p_54881_.m_45524_(p_54882_, 0) < 13 && this.m_6266_(blockstate, p_54881_, blockpos);
+         return p_54881_.m_45524_(p_54882_, 0) < 13 && blockstate.canSustainPlant(p_54881_, blockpos, net.minecraft.core.Direction.UP, this);
       }
    }
 
    public boolean m_54859_(ServerLevel p_54860_, BlockPos p_54861_, BlockState p_54862_, Random p_54863_) {
       p_54860_.m_7471_(p_54861_, false);
+      SaplingBlock.treeType = (this == Blocks.f_50072_) ? TreeType.BROWN_MUSHROOM : TreeType.BROWN_MUSHROOM; // CraftBukkit
       if (this.f_153981_.get().m_203334_().m_65385_(p_54860_, p_54860_.m_7726_().m_8481_(), p_54863_, p_54861_)) {
          return true;
       } else {
