--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -46,13 +_,13 @@
    }
 
    public void m_213898_(BlockState p_221799_, ServerLevel p_221800_, BlockPos p_221801_, RandomSource p_221802_) {
-      if (p_221800_.m_6042_().f_63858_() && p_221800_.m_46469_().m_46207_(GameRules.f_46134_) && p_221802_.m_188503_(2000) < p_221800_.m_46791_().m_19028_()) {
+      if (p_221800_.spigotConfig.enableZombiePigmenPortalSpawns && p_221800_.m_6042_().f_63858_() && p_221800_.m_46469_().m_46207_(GameRules.f_46134_) && p_221802_.m_188503_(2000) < p_221800_.m_46791_().m_19028_()) { // Spigot
          while(p_221800_.m_8055_(p_221801_).m_60713_(this)) {
             p_221801_ = p_221801_.m_7495_();
          }
 
          if (p_221800_.m_8055_(p_221801_).m_60643_(p_221800_, p_221801_, EntityType.f_20531_)) {
-            Entity entity = EntityType.f_20531_.m_262496_(p_221800_, p_221801_.m_7494_(), MobSpawnType.STRUCTURE);
+            Entity entity = EntityType.f_20531_.spawn(p_221800_, p_221801_.m_7494_(), MobSpawnType.STRUCTURE, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NETHER_PORTAL); // CraftBukkit - set spawn reason to NETHER_PORTAL
             if (entity != null) {
                entity.m_20091_();
             }
@@ -70,6 +_,10 @@
 
    public void m_7892_(BlockState p_54915_, Level p_54916_, BlockPos p_54917_, Entity p_54918_) {
       if (p_54918_.m_6072_()) {
+         // CraftBukkit start - Entity in portal
+         org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(p_54918_.getBukkitEntity(), new org.bukkit.Location(p_54916_.getWorld(), p_54917_.m_123341_(), p_54917_.m_123342_(), p_54917_.m_123343_()));
+         p_54916_.getCraftServer().getPluginManager().callEvent(event);
+         // CraftBukkit end
          p_54918_.m_20221_(p_54917_);
       }
 
