--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.level.portal.PortalShape;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
 
 public class NetherPortalBlock extends Block {
    public static final EnumProperty<Direction.Axis> f_54904_ = BlockStateProperties.f_61364_;
@@ -49,13 +_,13 @@
    }
 
    public void m_7455_(BlockState p_54937_, ServerLevel p_54938_, BlockPos p_54939_, Random p_54940_) {
-      if (p_54938_.m_6042_().m_63956_() && p_54938_.m_46469_().m_46207_(GameRules.f_46134_) && p_54940_.nextInt(2000) < p_54938_.m_46791_().m_19028_()) {
+      if (p_54938_.spigotConfig.enableZombiePigmenPortalSpawns && p_54938_.m_6042_().m_63956_() && p_54938_.m_46469_().m_46207_(GameRules.f_46134_) && p_54940_.nextInt(2000) < p_54938_.m_46791_().m_19028_()) { // Spigot
          while(p_54938_.m_8055_(p_54939_).m_60713_(this)) {
             p_54939_ = p_54939_.m_7495_();
          }
 
          if (p_54938_.m_8055_(p_54939_).m_60643_(p_54938_, p_54939_, EntityType.f_20531_)) {
-            Entity entity = EntityType.f_20531_.m_20600_(p_54938_, (CompoundTag)null, (Component)null, (Player)null, p_54939_.m_7494_(), MobSpawnType.STRUCTURE, false, false);
+            Entity entity = EntityType.f_20531_.m_20600_(p_54938_, (CompoundTag)null, (Component)null, (Player)null, p_54939_.m_7494_(), MobSpawnType.STRUCTURE, false, false); // FoxServer TODO
             if (entity != null) {
                entity.m_20091_();
             }
@@ -73,6 +_,10 @@
 
    public void m_7892_(BlockState p_54915_, Level p_54916_, BlockPos p_54917_, Entity p_54918_) {
       if (!p_54918_.m_20159_() && !p_54918_.m_20160_() && p_54918_.m_6072_()) {
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_54918_.getBukkitEntity(), new org.bukkit.Location(p_54916_.getWorld(), p_54917_.m_123341_(), p_54917_.m_123342_(), p_54917_.m_123343_()));
+         p_54916_.getCraftServer().getPluginManager().callEvent(event);
+         // CraftBukkit end
          p_54918_.m_20221_(p_54917_);
       }
 
