--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -67,6 +_,7 @@
       if (flag != p_55041_.m_61143_(f_55012_)) {
          if (flag) {
             this.m_260916_((Entity)null, p_55041_, p_55042_, p_55043_);
+            p_55041_ = p_55042_.m_8055_(p_55043_); // CraftBukkit - SPIGOT-5617: update in case changed in event
          }
 
          p_55042_.m_7731_(p_55043_, p_55041_.m_61124_(f_55012_, Boolean.valueOf(flag)), 3);
@@ -76,6 +_,12 @@
 
    private void m_260916_(@Nullable Entity p_261664_, BlockState p_261606_, Level p_261819_, BlockPos p_262042_) {
       if (p_261606_.m_61143_(f_55011_).m_280504_() || p_261819_.m_8055_(p_262042_.m_7494_()).m_60795_()) {
+         // CraftBukkit start
+         org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callNotePlayEvent(p_261819_, p_262042_, p_261606_.m_61143_(f_55011_), p_261606_.m_61143_(f_55013_));
+         if (event.isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_261819_.m_7696_(p_262042_, this, 0, 0);
          p_261819_.m_142346_(p_261664_, GameEvent.f_223699_, p_262042_);
       }
@@ -89,7 +_,9 @@
       } else if (p_55035_.f_46443_) {
          return InteractionResult.SUCCESS;
       } else {
-         p_55034_ = p_55034_.m_61122_(f_55013_);
+         int _new = net.minecraftforge.common.ForgeHooks.onNoteChange(p_55035_, p_55036_, p_55034_, p_55034_.m_61143_(f_55013_), p_55034_.m_61122_(f_55013_).m_61143_(f_55013_));
+         if (_new == -1) return InteractionResult.FAIL;
+         p_55034_ = p_55034_.m_61124_(f_55013_, _new);
          p_55035_.m_7731_(p_55036_, p_55034_, 3);
          this.m_260916_(p_55037_, p_55034_, p_55035_, p_55036_);
          p_55037_.m_36220_(Stats.f_12960_);
@@ -109,6 +_,9 @@
    }
 
    public boolean m_8133_(BlockState p_55023_, Level p_55024_, BlockPos p_55025_, int p_55026_, int p_55027_) {
+      net.minecraftforge.event.level.NoteBlockEvent.Play e = new net.minecraftforge.event.level.NoteBlockEvent.Play(p_55024_, p_55025_, p_55023_, p_55023_.m_61143_(f_55013_), p_55023_.m_61143_(f_55011_));
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e)) return false;
+      p_55023_ = p_55023_.m_61124_(f_55013_, e.getVanillaNoteId()).m_61124_(f_55011_, e.getInstrument());
       NoteBlockInstrument noteblockinstrument = p_55023_.m_61143_(f_55011_);
       float f;
       if (noteblockinstrument.m_262503_()) {
