--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -43,7 +_,9 @@
       boolean flag = p_55042_.m_46753_(p_55043_);
       if (flag != p_55041_.m_61143_(f_55012_)) {
          if (flag) {
-            this.m_55019_(p_55042_, p_55043_);
+            this.playNote(p_55042_, p_55043_, p_55041_); // CraftBukkit
+            p_55041_ = p_55042_.m_8055_(p_55043_); // CraftBukkit - SPIGOT-5617: update in case changed in event
+
          }
 
          p_55042_.m_7731_(p_55043_, p_55041_.m_61124_(f_55012_, Boolean.valueOf(flag)), 3);
@@ -51,9 +_,14 @@
 
    }
 
-   private void m_55019_(Level p_55020_, BlockPos p_55021_) {
+   private void playNote(Level p_55020_, BlockPos p_55021_, BlockState data) {
       if (p_55020_.m_8055_(p_55021_.m_7494_()).m_60795_()) {
-         p_55020_.m_7696_(p_55021_, this, 0, 0);
+         // CraftBukkit start
+         org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callNotePlayEvent(p_55020_, p_55021_, data.m_61143_(f_55011_), data.m_61143_(f_55013_));
+         if (!event.isCancelled()) {
+            p_55020_.m_7696_(p_55021_, this, 0, 0);
+         }
+         // CraftBukkit end
       }
 
    }
@@ -62,9 +_,11 @@
       if (p_55035_.f_46443_) {
          return InteractionResult.SUCCESS;
       } else {
-         p_55034_ = p_55034_.m_61122_(f_55013_);
+         int _new = net.minecraftforge.common.ForgeHooks.onNoteChange(p_55035_, p_55036_, p_55034_, p_55034_.m_61143_(f_55013_), p_55034_.m_61122_(f_55013_).m_61143_(f_55013_));
+         if (_new == -1) return InteractionResult.FAIL;
+         p_55034_ = p_55034_.m_61124_(f_55013_, _new);
          p_55035_.m_7731_(p_55036_, p_55034_, 3);
-         this.m_55019_(p_55035_, p_55036_);
+         this.playNote(p_55035_, p_55036_, p_55034_); // CraftBukkit
          p_55037_.m_36220_(Stats.f_12960_);
          return InteractionResult.CONSUME;
       }
@@ -72,12 +_,15 @@
 
    public void m_6256_(BlockState p_55029_, Level p_55030_, BlockPos p_55031_, Player p_55032_) {
       if (!p_55030_.f_46443_) {
-         this.m_55019_(p_55030_, p_55031_);
+         this.playNote(p_55030_, p_55031_, p_55029_); // CraftBukkit
          p_55032_.m_36220_(Stats.f_12959_);
       }
    }
 
    public boolean m_8133_(BlockState p_55023_, Level p_55024_, BlockPos p_55025_, int p_55026_, int p_55027_) {
+      net.minecraftforge.event.world.NoteBlockEvent.Play e = new net.minecraftforge.event.world.NoteBlockEvent.Play(p_55024_, p_55025_, p_55023_, p_55023_.m_61143_(f_55013_), p_55023_.m_61143_(f_55011_));
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e)) return false;
+      p_55023_ = p_55023_.m_61124_(f_55013_, e.getVanillaNoteId()).m_61124_(f_55011_, e.getInstrument());
       int i = p_55023_.m_61143_(f_55013_);
       float f = (float)Math.pow(2.0D, (double)(i - 12) / 12.0D);
       p_55024_.m_5594_((Player)null, p_55025_, p_55023_.m_61143_(f_55011_).m_61668_(), SoundSource.RECORDS, 3.0F, f);
