--- a/net/minecraft/world/level/block/NyliumBlock.java
+++ b/net/minecraft/world/level/block/NyliumBlock.java
@@ -27,6 +_,11 @@
 
    public void m_7455_(BlockState p_55074_, ServerLevel p_55075_, BlockPos p_55076_, Random p_55077_) {
       if (!m_55078_(p_55074_, p_55075_, p_55076_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_55075_, p_55076_, Blocks.f_50134_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_55075_.m_46597_(p_55076_, Blocks.f_50134_.m_49966_());
       }
 
