--- a/net/minecraft/world/level/block/NyliumBlock.java
+++ b/net/minecraft/world/level/block/NyliumBlock.java
@@ -30,6 +_,11 @@
 
    public void m_213898_(BlockState p_221835_, ServerLevel p_221836_, BlockPos p_221837_, RandomSource p_221838_) {
       if (!m_55078_(p_221835_, p_221836_, p_221837_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callBlockFadeEvent(p_221836_, p_221837_, Blocks.f_50134_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_221836_.m_46597_(p_221837_, Blocks.f_50134_.m_49966_());
       }
 
