--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class ObserverBlock extends DirectionalBlock {
    public static final BooleanProperty f_55082_ = BlockStateProperties.f_61448_;
@@ -36,8 +_,18 @@
 
    public void m_7458_(BlockState p_55096_, ServerLevel p_55097_, BlockPos p_55098_, Random p_55099_) {
       if (p_55096_.m_61143_(f_55082_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_55097_, p_55098_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_55097_.m_7731_(p_55098_, p_55096_.m_61124_(f_55082_, Boolean.valueOf(false)), 2);
       } else {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_55097_, p_55098_, 0, 15).getNewCurrent() != 15) {
+            return;
+         }
+         // CraftBukkit end
          p_55097_.m_7731_(p_55098_, p_55096_.m_61124_(f_55082_, Boolean.valueOf(true)), 2);
          p_55097_.m_186460_(p_55098_, this, 2);
       }
