--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -14,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory; // CraftBukkit
+
 public class ObserverBlock extends DirectionalBlock {
    public static final BooleanProperty f_55082_ = BlockStateProperties.f_61448_;
 
@@ -36,8 +_,18 @@
 
    public void m_213897_(BlockState p_221840_, ServerLevel p_221841_, BlockPos p_221842_, RandomSource p_221843_) {
       if (p_221840_.m_61143_(f_55082_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_221841_, p_221842_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_221841_.m_7731_(p_221842_, p_221840_.m_61124_(f_55082_, Boolean.valueOf(false)), 2);
       } else {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_221841_, p_221842_, 0, 15).getNewCurrent() != 15) {
+            return;
+         }
+         // CraftBukkit end
          p_221841_.m_7731_(p_221842_, p_221840_.m_61124_(f_55082_, Boolean.valueOf(true)), 2);
          p_221841_.m_186460_(p_221842_, this, 2);
       }
