--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -42,6 +_,8 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class PointedDripstoneBlock extends Block implements Fallable, SimpleWaterloggedBlock {
    public static final DirectionProperty f_154009_ = BlockStateProperties.f_155997_;
@@ -117,6 +_,11 @@
    public void m_5581_(Level p_154042_, BlockState p_154043_, BlockHitResult p_154044_, Projectile p_154045_) {
       BlockPos blockpos = p_154044_.m_82425_();
       if (!p_154042_.f_46443_ && p_154045_.m_142265_(p_154042_, blockpos) && p_154045_ instanceof ThrownTrident && p_154045_.m_20184_().m_82553_() > 0.6D) {
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityChangeBlockEvent(p_154045_, blockpos, Blocks.f_50016_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_154042_.m_46961_(blockpos, true);
       }
 
@@ -124,7 +_,9 @@
 
    public void m_142072_(Level p_154047_, BlockState p_154048_, BlockPos p_154049_, Entity p_154050_, float p_154051_) {
       if (p_154048_.m_61143_(f_154009_) == Direction.UP && p_154048_.m_61143_(f_154010_) == DripstoneThickness.TIP) {
+         CraftEventFactory.blockDamage = CraftBlock.at(p_154047_, p_154049_); // CraftBukkit
          p_154050_.m_142535_(p_154051_ + 2.0F, 2.0F, DamageSource.f_146703_);
+         CraftEventFactory.blockDamage = null; // CraftBukkit
       } else {
          super.m_142072_(p_154047_, p_154048_, p_154049_, p_154050_, p_154051_);
       }
@@ -341,7 +_,7 @@
       if (m_154143_(blockstate, p_154038_.m_122424_())) {
          m_154230_(blockstate, p_154036_, blockpos);
       } else if (blockstate.m_60795_() || blockstate.m_60713_(Blocks.f_49990_)) {
-         m_154087_(p_154036_, blockpos, p_154038_, DripstoneThickness.TIP);
+         createDripstone(p_154036_, blockpos, p_154038_, DripstoneThickness.TIP, p_154037_); // CraftBukkit
       }
 
    }
@@ -351,6 +_,11 @@
       p_154088_.m_7731_(p_154089_, blockstate, 3);
    }
 
+   private static void createDripstone(LevelAccessor p_154088_, BlockPos p_154089_, Direction p_154090_, DripstoneThickness p_154091_, BlockPos source) {  // CraftBukkit
+      BlockState blockstate = Blocks.f_152588_.m_49966_().m_61124_(f_154009_, p_154090_).m_61124_(f_154010_, p_154091_).m_61124_(f_154011_, Boolean.valueOf(p_154088_.m_6425_(p_154089_).m_76152_() == Fluids.f_76193_));
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_154088_, source, p_154089_, blockstate, 3); // CraftBukkit
+   }
+
    private static void m_154230_(BlockState p_154231_, LevelAccessor p_154232_, BlockPos p_154233_) {
       BlockPos blockpos;
       BlockPos blockpos1;
@@ -362,8 +_,8 @@
          blockpos1 = p_154233_.m_7495_();
       }
 
-      m_154087_(p_154232_, blockpos, Direction.DOWN, DripstoneThickness.TIP_MERGE);
-      m_154087_(p_154232_, blockpos1, Direction.UP, DripstoneThickness.TIP_MERGE);
+      createDripstone(p_154232_, blockpos, Direction.DOWN, DripstoneThickness.TIP_MERGE, p_154233_); // CraftBukkit
+      createDripstone(p_154232_, blockpos1, Direction.UP, DripstoneThickness.TIP_MERGE, p_154233_); // CraftBukkit
    }
 
    public static void m_154062_(Level p_154063_, BlockPos p_154064_, BlockState p_154065_) {
