--- a/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -64,7 +_,13 @@
 
       p_154266_.m_146924_(true);
       if (!p_154264_.f_46443_) {
-         if (p_154266_.m_6060_() && (p_154264_.m_46469_().m_46207_(GameRules.f_46132_) || p_154266_ instanceof Player) && p_154266_.m_142265_(p_154264_, p_154265_)) {
+         //if (p_154266_.isOnFire() && (p_154264_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || p_154266_ instanceof Player) && p_154266_.mayInteract(p_154264_, p_154265_)) {
+         // CraftBukkit start
+         if (p_154266_.m_6060_() && p_154266_.m_142265_(p_154264_, p_154265_)) {
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_154266_, p_154265_, Blocks.f_50016_.m_49966_(), !(p_154264_.m_46469_().m_46207_(GameRules.f_46132_) || p_154266_ instanceof Player)).isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             p_154264_.m_46961_(p_154265_, false);
          }
 
@@ -109,7 +_,7 @@
       if (p_154256_.m_6095_().m_204039_(EntityTypeTags.f_144291_)) {
          return true;
       } else {
-         return p_154256_ instanceof LivingEntity ? ((LivingEntity)p_154256_).m_6844_(EquipmentSlot.FEET).m_150930_(Items.f_42463_) : false;
+         return p_154256_ instanceof LivingEntity ? ((LivingEntity)p_154256_).m_6844_(EquipmentSlot.FEET).canWalkOnPowderedSnow((LivingEntity)p_154256_) : false;
       }
    }
 
