--- a/net/minecraft/world/level/block/PressurePlateBlock.java
+++ b/net/minecraft/world/level/block/PressurePlateBlock.java
@@ -16,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 import net.minecraft.world.level.material.Material;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class PressurePlateBlock extends BasePressurePlateBlock {
    public static final BooleanProperty f_55249_ = BlockStateProperties.f_61448_;
@@ -69,6 +_,25 @@
 
       if (!list.isEmpty()) {
          for(Entity entity : list) {
+            // CraftBukkit start - Call interact event when turning on a pressure plate
+            if (this.m_6016_(p_55264_.m_8055_(p_55265_)) == 0) {
+               org.bukkit.World bworld = p_55264_.getWorld();
+               org.bukkit.plugin.PluginManager manager = p_55264_.getCraftServer().getPluginManager();
+               org.bukkit.event.Cancellable cancellable;
+
+               if (entity instanceof Player) {
+                  cancellable = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, p_55265_, null, null, null);
+               } else {
+                  cancellable = new EntityInteractEvent(entity.getBukkitEntity(), bworld.getBlockAt(p_55265_.m_123341_(), p_55265_.m_123342_(), p_55265_.m_123343_()));
+                  manager.callEvent((EntityInteractEvent) cancellable);
+               }
+
+               // We only want to block turning the plate on if all events are cancelled
+               if (cancellable.isCancelled()) {
+                  continue;
+               }
+            }
+            // CraftBukkit end
             if (!entity.m_6090_()) {
                return 15;
             }
