--- a/net/minecraft/world/level/block/RedstoneLampBlock.java
+++ b/net/minecraft/world/level/block/RedstoneLampBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class RedstoneLampBlock extends Block {
    public static final BooleanProperty f_55654_ = RedstoneTorchBlock.f_55674_;
@@ -31,6 +_,11 @@
             if (flag) {
                p_55667_.m_186460_(p_55668_, this, 4);
             } else {
+               // CraftBukkit start
+               if (CraftEventFactory.callRedstoneChange(p_55667_, p_55668_, 0, 15).getNewCurrent() != 15) {
+                  return;
+               }
+               // CraftBukkit end
                p_55667_.m_7731_(p_55668_, p_55666_.m_61122_(f_55654_), 2);
             }
          }
@@ -40,6 +_,11 @@
 
    public void m_7458_(BlockState p_55661_, ServerLevel p_55662_, BlockPos p_55663_, Random p_55664_) {
       if (p_55661_.m_61143_(f_55654_) && !p_55662_.m_46753_(p_55663_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_55662_, p_55663_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_55662_.m_7731_(p_55663_, p_55661_.m_61122_(f_55654_), 2);
       }
 
