--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -16,6 +_,7 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class RedstoneTorchBlock extends TorchBlock {
    public static final BooleanProperty f_55674_ = BlockStateProperties.f_61443_;
@@ -62,8 +_,24 @@
          list.remove(0);
       }
 
+      // CraftBukkit start
+      org.bukkit.plugin.PluginManager manager = p_55690_.getCraftServer().getPluginManager();
+      org.bukkit.block.Block block = p_55690_.getWorld().getBlockAt(p_55691_.m_123341_(), p_55691_.m_123342_(), p_55691_.m_123343_());
+      int oldCurrent = ((Boolean) p_55689_.m_61143_(f_55674_)).booleanValue() ? 15 : 0;
+
+      BlockRedstoneEvent event = new BlockRedstoneEvent(block, oldCurrent, oldCurrent);
+      // CraftBukkit end
       if (p_55689_.m_61143_(f_55674_)) {
          if (flag) {
+            // CraftBukkit start
+            if (oldCurrent != 0) {
+               event.setNewCurrent(0);
+               manager.callEvent(event);
+               if (event.getNewCurrent() != 0) {
+                  return;
+               }
+            }
+            // CraftBukkit end
             p_55690_.m_7731_(p_55691_, p_55689_.m_61124_(f_55674_, Boolean.valueOf(false)), 3);
             if (m_55684_(p_55690_, p_55691_, true)) {
                p_55690_.m_46796_(1502, p_55691_, 0);
@@ -71,6 +_,15 @@
             }
          }
       } else if (!flag && !m_55684_(p_55690_, p_55691_, false)) {
+         // CraftBukkit start
+         if (oldCurrent != 15) {
+            event.setNewCurrent(15);
+            manager.callEvent(event);
+            if (event.getNewCurrent() != 15) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_55690_.m_7731_(p_55691_, p_55689_.m_61124_(f_55674_, Boolean.valueOf(true)), 3);
       }
 
