--- a/net/minecraft/world/level/block/RedstoneLampBlock.java
+++ b/net/minecraft/world/level/block/RedstoneLampBlock.java
@@ -11,6 +_,8 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory; // CraftBukkit
+
 public class RedstoneLampBlock extends Block {
    public static final BooleanProperty f_55654_ = RedstoneTorchBlock.f_55674_;
 
@@ -31,6 +_,11 @@
             if (flag) {
                p_55667_.m_186460_(p_55668_, this, 4);
             } else {
+               // CraftBukkit start
+               if (CraftEventFactory.callRedstoneChange(p_55667_, p_55668_, 0, 15).getNewCurrent() != 15) {
+                  return;
+               }
+               // CraftBukkit end
                p_55667_.m_7731_(p_55668_, p_55666_.m_61122_(f_55654_), 2);
             }
          }
@@ -40,6 +_,11 @@
 
    public void m_213897_(BlockState p_221937_, ServerLevel p_221938_, BlockPos p_221939_, RandomSource p_221940_) {
       if (p_221937_.m_61143_(f_55654_) && !p_221938_.m_276867_(p_221939_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_221938_, p_221939_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_221938_.m_7731_(p_221939_, p_221937_.m_61122_(f_55654_), 2);
       }
 
