--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -12,14 +_,14 @@
 import net.minecraft.world.item.BlockItem;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.context.BlockPlaceContext;
-import net.minecraft.world.item.enchantment.EnchantmentHelper;
-import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 import net.minecraft.world.phys.BlockHitResult;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class RedStoneOreBlock extends Block {
    public static final BooleanProperty f_55450_ = RedstoneTorchBlock.f_55674_;
@@ -30,29 +_,50 @@
    }
 
    public void m_6256_(BlockState p_55467_, Level p_55468_, BlockPos p_55469_, Player p_55470_) {
-      m_55492_(p_55467_, p_55468_, p_55469_);
+      interact(p_55467_, p_55468_, p_55469_, p_55470_);
       super.m_6256_(p_55467_, p_55468_, p_55469_, p_55470_);
    }
 
    public void m_141947_(Level p_154299_, BlockPos p_154300_, BlockState p_154301_, Entity p_154302_) {
-      m_55492_(p_154301_, p_154299_, p_154300_);
-      super.m_141947_(p_154299_, p_154300_, p_154301_, p_154302_);
+      // CraftBukkit start
+      // interact(iblomckdata, world, blockgposition);
+      // super.stepOn(world, blockposition, iblockdata, entity);
+      if (p_154302_ instanceof Player) {
+         org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) p_154302_, org.bukkit.event.block.Action.PHYSICAL, p_154300_, null, null, null);
+         if (!event.isCancelled()) {
+            interact(p_154299_.m_8055_(p_154300_), p_154299_, p_154300_, p_154302_); // add entity
+            super.m_141947_(p_154299_, p_154300_, p_154301_, p_154302_);
+         }
+      } else {
+         EntityInteractEvent event = new EntityInteractEvent(p_154302_.getBukkitEntity(), p_154299_.getWorld().getBlockAt(p_154300_.m_123341_(), p_154300_.m_123342_(), p_154300_.m_123343_()));
+         p_154299_.getCraftServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            interact(p_154299_.m_8055_(p_154300_), p_154299_, p_154300_, p_154302_); // add entity
+            super.m_141947_(p_154299_, p_154300_, p_154301_, p_154302_);
+         }
+      }
+      // CraftBukkit end
    }
 
    public InteractionResult m_6227_(BlockState p_55472_, Level p_55473_, BlockPos p_55474_, Player p_55475_, InteractionHand p_55476_, BlockHitResult p_55477_) {
       if (p_55473_.f_46443_) {
          m_55454_(p_55473_, p_55474_);
       } else {
-         m_55492_(p_55472_, p_55473_, p_55474_);
+         interact(p_55472_, p_55473_, p_55474_, p_55475_);
       }
 
       ItemStack itemstack = p_55475_.m_21120_(p_55476_);
       return itemstack.m_41720_() instanceof BlockItem && (new BlockPlaceContext(p_55475_, p_55476_, itemstack, p_55477_)).m_7059_() ? InteractionResult.PASS : InteractionResult.SUCCESS;
    }
 
-   private static void m_55492_(BlockState p_55493_, Level p_55494_, BlockPos p_55495_) {
+   private static void interact(BlockState p_55493_, Level p_55494_, BlockPos p_55495_, Entity entity) { // CraftBukkit - add Entity
       m_55454_(p_55494_, p_55495_);
       if (!p_55493_.m_61143_(f_55450_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityChangeBlockEvent(entity, p_55495_, p_55493_.m_61124_(f_55450_, true)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_55494_.m_7731_(p_55495_, p_55493_.m_61124_(f_55450_, Boolean.valueOf(true)), 3);
       }
 
@@ -64,6 +_,11 @@
 
    public void m_7455_(BlockState p_55488_, ServerLevel p_55489_, BlockPos p_55490_, Random p_55491_) {
       if (p_55488_.m_61143_(f_55450_)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_55489_, p_55490_, p_55488_.m_61124_(f_55450_, false)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_55489_.m_7731_(p_55490_, p_55488_.m_61124_(f_55450_, Boolean.valueOf(false)), 3);
       }
 
@@ -71,11 +_,11 @@
 
    public void m_8101_(BlockState p_55462_, ServerLevel p_55463_, BlockPos p_55464_, ItemStack p_55465_) {
       super.m_8101_(p_55462_, p_55463_, p_55464_, p_55465_);
-      if (EnchantmentHelper.m_44843_(Enchantments.f_44985_, p_55465_) == 0) {
-         int i = 1 + p_55463_.f_46441_.nextInt(5);
-         this.m_49805_(p_55463_, p_55464_, i);
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? 1 + RANDOM.nextInt(5) : 0;
    }
 
    public void m_7100_(BlockState p_55479_, Level p_55480_, BlockPos p_55481_, Random p_55482_) {
