--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -17,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public class RedstoneTorchBlock extends TorchBlock {
    public static final BooleanProperty f_55674_ = BlockStateProperties.f_61443_;
    private static final Map<BlockGetter, List<RedstoneTorchBlock.Toggle>> f_55675_ = new WeakHashMap<>();
@@ -62,8 +_,24 @@
          list.remove(0);
       }
 
+      // CraftBukkit start
+      org.bukkit.plugin.PluginManager manager = p_221950_.getCraftServer().getPluginManager();
+      org.bukkit.block.Block block = p_221950_.getWorld().getBlockAt(p_221951_.m_123341_(), p_221951_.m_123342_(), p_221951_.m_123343_());
+      int oldCurrent = (Boolean) p_221949_.m_61143_(f_55674_) ? 15 : 0;
+      BlockRedstoneEvent event = new BlockRedstoneEvent(block, oldCurrent, oldCurrent);
+      // CraftBukkit end
+
       if (p_221949_.m_61143_(f_55674_)) {
          if (flag) {
+            // CraftBukkit start
+            if (oldCurrent != 0) {
+               event.setNewCurrent(0);
+               manager.callEvent(event);
+               if (event.getNewCurrent() != 0) {
+                  return;
+               }
+            }
+            // CraftBukkit end
             p_221950_.m_7731_(p_221951_, p_221949_.m_61124_(f_55674_, Boolean.valueOf(false)), 3);
             if (m_55684_(p_221950_, p_221951_, true)) {
                p_221950_.m_46796_(1502, p_221951_, 0);
@@ -71,6 +_,15 @@
             }
          }
       } else if (!flag && !m_55684_(p_221950_, p_221951_, false)) {
+         // CraftBukkit start
+         if (oldCurrent != 15) {
+            event.setNewCurrent(15);
+            manager.callEvent(event);
+            if (event.getNewCurrent() != 15) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_221950_.m_7731_(p_221951_, p_221949_.m_61124_(f_55674_, Boolean.valueOf(true)), 3);
       }
 
