--- a/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -72,7 +_,7 @@
          if (!p_55875_.f_46443_) {
             ServerPlayer serverplayer = (ServerPlayer)p_55877_;
             if (serverplayer.m_8963_() != p_55875_.m_46472_() || !p_55876_.equals(serverplayer.m_8961_())) {
-               serverplayer.m_9158_(p_55875_.m_46472_(), p_55876_, 0.0F, false, true);
+               serverplayer.setRespawnPosition(p_55875_.m_46472_(), p_55876_, 0.0F, false, true, org.bukkit.event.player.PlayerSpawnChangeEvent.Cause.RESPAWN_ANCHOR); // CraftBukkit
                p_55875_.m_6263_((Player)null, (double)p_55876_.m_123341_() + 0.5D, (double)p_55876_.m_123342_() + 0.5D, (double)p_55876_.m_123343_() + 0.5D, SoundEvents.f_12326_, SoundSource.BLOCKS, 1.0F, 1.0F);
                return InteractionResult.SUCCESS;
             }
