--- a/net/minecraft/world/level/block/RootedDirtBlock.java
+++ b/net/minecraft/world/level/block/RootedDirtBlock.java
@@ -22,6 +_,7 @@
    }
 
    public void m_7719_(ServerLevel p_154361_, Random p_154362_, BlockPos p_154363_, BlockState p_154364_) {
-      p_154361_.m_46597_(p_154363_.m_7495_(), Blocks.f_152548_.m_49966_());
+      //p_154361_.setBlockAndUpdate(p_154363_.below(), Blocks.HANGING_ROOTS.defaultBlockState());
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_154361_, p_154363_, p_154363_.m_7495_(), Blocks.f_152548_.m_49966_()); // CraftBukkit
    }
 }
