--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -15,11 +_,19 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.TreeType;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation;
+import org.bukkit.event.world.StructureGrowEvent;
+// CraftBukkit end
+
 public class SaplingBlock extends BushBlock implements BonemealableBlock {
    public static final IntegerProperty f_55973_ = BlockStateProperties.f_61387_;
    protected static final float f_154380_ = 6.0F;
    protected static final VoxelShape f_55974_ = Block.m_49796_(2.0D, 0.0D, 2.0D, 14.0D, 12.0D, 14.0D);
    private final AbstractTreeGrower f_55975_;
+   public static TreeType treeType; // CraftBukkit
 
    public SaplingBlock(AbstractTreeGrower p_55978_, BlockBehaviour.Properties p_55979_) {
       super(p_55979_);
@@ -32,7 +_,8 @@
    }
 
    public void m_213898_(BlockState p_222011_, ServerLevel p_222012_, BlockPos p_222013_, RandomSource p_222014_) {
-      if (p_222012_.m_46803_(p_222013_.m_7494_()) >= 9 && p_222014_.m_188503_(7) == 0) {
+      if (!p_222012_.isAreaLoaded(p_222013_, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
+      if (p_222012_.m_46803_(p_222013_.m_7494_()) >= 9 && p_222014_.m_188501_() < (p_222012_.spigotConfig.saplingModifier / (100.0f * 7))) { // Spigot - SPIGOT-7159: Better modifier resolution
          this.m_222000_(p_222012_, p_222013_, p_222011_, p_222014_);
       }
 
@@ -42,7 +_,32 @@
       if (p_222003_.m_61143_(f_55973_) == 0) {
          p_222001_.m_7731_(p_222002_, p_222003_.m_61122_(f_55973_), 4);
       } else {
-         this.f_55975_.m_213817_(p_222001_, p_222001_.m_7726_().m_8481_(), p_222002_, p_222003_, p_222004_);
+         // CraftBukkit start
+         if (p_222001_.captureTreeGeneration) {
+            this.f_55975_.m_213817_(p_222001_, p_222001_.m_7726_().m_8481_(), p_222002_, p_222003_, p_222004_);
+         } else {
+            p_222001_.captureTreeGeneration = true;
+            this.f_55975_.m_213817_(p_222001_, p_222001_.m_7726_().m_8481_(), p_222002_, p_222003_, p_222004_);
+            p_222001_.captureTreeGeneration = false;
+            if (p_222001_.capturedBlockStates.size() > 0) {
+               TreeType treeType = SaplingBlock.treeType;
+               SaplingBlock.treeType = null;
+               Location location = CraftLocation.toBukkit(p_222002_, p_222001_.getWorld());
+               java.util.List<org.bukkit.block.BlockState> blocks = new java.util.ArrayList<>(p_222001_.capturedBlockStates.values());
+               p_222001_.capturedBlockStates.clear();
+               StructureGrowEvent event = null;
+               if (treeType != null) {
+                  event = new StructureGrowEvent(location, treeType, false, null, blocks);
+                  org.bukkit.Bukkit.getPluginManager().callEvent(event);
+               }
+               if (event == null || !event.isCancelled()) {
+                  for (org.bukkit.block.BlockState blockstate : blocks) {
+                     blockstate.update(true);
+                  }
+               }
+            }
+         }
+         // CraftBukkit end
       }
 
    }
