--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -87,9 +_,9 @@
    public void m_7458_(BlockState p_56032_, ServerLevel p_56033_, BlockPos p_56034_, Random p_56035_) {
       int i = m_56024_(p_56033_, p_56034_);
       BlockState blockstate = p_56032_.m_61124_(f_56012_, Integer.valueOf(i)).m_61124_(f_56014_, Boolean.valueOf(this.m_56027_(p_56033_, p_56034_, i)));
-      if (blockstate.m_61143_(f_56012_) == 7) {
+      if (blockstate.m_61143_(f_56012_) == 7 && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_56033_, p_56034_, Blocks.f_50016_.m_49966_()).isCancelled()) { // CraftBukkit - BlockFadeEvent
          if (p_56032_.m_61143_(f_56012_) == 7) {
-            FallingBlockEntity.m_201971_(p_56033_, p_56034_, blockstate);
+            p_56033_.m_7967_(new FallingBlockEntity(p_56033_, (double)p_56034_.m_123341_() + 0.5D, (double)p_56034_.m_123342_(), (double)p_56034_.m_123343_() + 0.5D, blockstate.m_61124_(f_56013_, Boolean.valueOf(false))));
          } else {
             p_56033_.m_46961_(p_56034_, true);
          }
@@ -153,6 +_,6 @@
       VoxelShape voxelshape6 = Block.m_49796_(14.0D, 0.0D, 0.0D, 16.0D, 2.0D, 16.0D);
       VoxelShape voxelshape7 = Block.m_49796_(0.0D, 0.0D, 14.0D, 16.0D, 2.0D, 16.0D);
       VoxelShape voxelshape8 = Block.m_49796_(0.0D, 0.0D, 0.0D, 16.0D, 2.0D, 2.0D);
-      f_56016_ = Shapes.m_83124_(f_56017_, f_56015_, voxelshape6, voxelshape5, voxelshape8, voxelshape7);
+      f_56016_ = Shapes.m_83124_(ScaffoldingBlock.f_56017_, f_56015_, voxelshape6, voxelshape5, voxelshape8, voxelshape7);
    }
 }
