--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -40,6 +_,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
+// CraftBukkit end
+
 public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterloggedBlock {
    public static final int f_154383_ = 30;
    public static final int f_154384_ = 10;
@@ -88,6 +_,18 @@
 
    public void m_141947_(Level p_222132_, BlockPos p_222133_, BlockState p_222134_, Entity p_222135_) {
       if (!p_222132_.m_5776_() && m_154489_(p_222134_) && p_222135_.m_6095_() != EntityType.f_217015_) {
+         // CraftBukkit start - 1.20.1
+         org.bukkit.event.Cancellable cancellable;
+         if (p_222135_ instanceof Player) {
+            cancellable = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerInteractEvent((Player) p_222135_, org.bukkit.event.block.Action.PHYSICAL, p_222133_, null, null, null);
+         } else {
+            cancellable = new org.bukkit.event.entity.EntityInteractEvent(p_222135_.getBukkitEntity(), p_222132_.getWorld().getBlockAt(p_222133_.m_123341_(), p_222133_.m_123342_(), p_222133_.m_123343_()));
+            p_222132_.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+         }
+         if (cancellable.isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          BlockEntity blockentity = p_222132_.m_7702_(p_222133_);
          if (blockentity instanceof SculkSensorBlockEntity) {
             SculkSensorBlockEntity sculksensorblockentity = (SculkSensorBlockEntity)blockentity;
@@ -177,6 +_,14 @@
    }
 
    public static void m_154407_(Level p_154408_, BlockPos p_154409_, BlockState p_154410_) {
+      // CraftBukkit start
+      BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(CraftBlock.at(p_154408_, p_154409_), p_154410_.m_61143_(SculkSensorBlock.f_154387_), 0);
+      p_154408_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+      if (eventRedstone.getNewCurrent() > 0) {
+         p_154408_.m_7731_(p_154409_, p_154410_.m_61124_(SculkSensorBlock.f_154387_, eventRedstone.getNewCurrent()), 3);
+         return;
+      }
+      // CraftBukkit end
       p_154408_.m_7731_(p_154409_, p_154410_.m_61124_(f_154386_, SculkSensorPhase.COOLDOWN).m_61124_(f_154387_, Integer.valueOf(0)), 3);
       p_154408_.m_186460_(p_154409_, p_154410_.m_60734_(), 10);
       m_276869_(p_154408_, p_154409_, p_154410_);
@@ -188,6 +_,14 @@
    }
 
    public void m_277033_(@Nullable Entity p_277529_, Level p_277340_, BlockPos p_277386_, BlockState p_277799_, int p_277993_, int p_278003_) {
+      // CraftBukkit start
+      BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(CraftBlock.at(p_277340_, p_277386_), p_277799_.m_61143_(SculkSensorBlock.f_154387_), p_278003_);
+      p_277340_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+      if (eventRedstone.getNewCurrent() <= 0) {
+         return;
+      }
+      p_278003_ = eventRedstone.getNewCurrent();
+      // CraftBukkit end
       p_277340_.m_7731_(p_277386_, p_277799_.m_61124_(f_154386_, SculkSensorPhase.ACTIVE).m_61124_(f_154387_, Integer.valueOf(p_277993_)), 3);
       p_277340_.m_186460_(p_277386_, p_277799_.m_60734_(), this.m_278716_());
       m_276869_(p_277340_, p_277386_, p_277799_);
@@ -252,9 +_,11 @@
 
    public void m_213646_(BlockState p_222142_, ServerLevel p_222143_, BlockPos p_222144_, ItemStack p_222145_, boolean p_222146_) {
       super.m_213646_(p_222142_, p_222143_, p_222144_, p_222145_, p_222146_);
-      if (p_222146_) {
-         this.m_220822_(p_222143_, p_222144_, p_222145_, ConstantInt.m_146483_(5));
-      }
-
+
+   }
+
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader level, RandomSource randomSource, BlockPos pos, int fortuneLevel, int silkTouchLevel) {
+      return silkTouchLevel == 0 ? 5 : 0;
    }
 }
