--- a/net/minecraft/world/level/block/SculkBlock.java
+++ b/net/minecraft/world/level/block/SculkBlock.java
@@ -27,8 +_,11 @@
             if (p_222076_.m_188503_(j) < i) {
                BlockPos blockpos1 = blockpos.m_7494_();
                BlockState blockstate = this.m_222067_(p_222074_, blockpos1, p_222076_, p_222077_.m_222282_());
-               p_222074_.m_7731_(blockpos1, blockstate, 3);
-               p_222074_.m_5594_((Player)null, blockpos, blockstate.m_60827_().m_56777_(), SoundSource.BLOCKS, 1.0F, 1.0F);
+               // CraftBukkit start - 1.20.1 Call BlockSpreadEvent
+               if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockSpreadEvent(p_222074_, p_222075_, blockpos1, blockstate, 3)) {
+                  p_222074_.m_5594_((Player) null, blockpos, blockstate.m_60827_().m_56777_(), SoundSource.BLOCKS, 1.0F, 1.0F);
+               }
+               // CraftBukkit end
             }
 
             return Math.max(0, i - j);
