--- a/net/minecraft/world/level/block/SculkSpreader.java
+++ b/net/minecraft/world/level/block/SculkSpreader.java
@@ -35,6 +_,13 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.world.level.Level;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.event.block.SculkBloomEvent;
+// CraftBukkit end
+
 public class SculkSpreader {
    public static final int f_222233_ = 24;
    public static final int f_222234_ = 1000;
@@ -49,6 +_,7 @@
    private final int f_222243_;
    private List<SculkSpreader.ChargeCursor> f_222244_ = new ArrayList<>();
    private static final Logger f_222245_ = LogUtils.getLogger();
+   Level level; // CraftBukkit - 1.20.1
 
    public SculkSpreader(boolean p_222248_, TagKey<Block> p_222249_, int p_222250_, int p_222251_, int p_222252_, int p_222253_) {
       this.f_222238_ = p_222248_;
@@ -130,6 +_,19 @@
 
    private void m_222260_(SculkSpreader.ChargeCursor p_222261_) {
       if (this.f_222244_.size() < 32) {
+         // CraftBukkit start - 1.20.1
+         if (!m_222282_()) { // CraftBukkit - SPIGOT-7475: Don't call event during world generation
+            CraftBlock bukkitBlock = CraftBlock.at(level, p_222261_.f_222288_);
+            SculkBloomEvent event = new SculkBloomEvent(bukkitBlock, p_222261_.m_222341_());
+            Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+
+            p_222261_.f_222289_ = event.getCharge();
+         }
+         // CraftBukkit end
+
          this.f_222244_.add(p_222261_);
       }
    }
