--- a/net/minecraft/world/level/block/SculkVeinBlock.java
+++ b/net/minecraft/world/level/block/SculkVeinBlock.java
@@ -84,45 +_,69 @@
    }
 
    public int m_213628_(SculkSpreader.ChargeCursor p_222369_, LevelAccessor p_222370_, BlockPos p_222371_, RandomSource p_222372_, SculkSpreader p_222373_, boolean p_222374_) {
-      if (p_222374_ && this.m_222375_(p_222373_, p_222370_, p_222369_.m_222304_(), p_222372_)) {
+      if (p_222374_ && this.attemptPlaceSculk(p_222373_, p_222370_, p_222369_.m_222304_(), p_222372_, p_222371_)) { // CraftBukkit - 1.20.1 add source block
          return p_222369_.m_222341_() - 1;
       } else {
          return p_222372_.m_188503_(p_222373_.m_222280_()) == 0 ? Mth.m_14143_((float)p_222369_.m_222341_() * 0.5F) : p_222369_.m_222341_();
       }
    }
 
+   // CatServer start - 1.20.1
+   private final ThreadLocal<BlockPos> catserver$attemptPlaceSculkSourceBlock = ThreadLocal.withInitial(() -> BlockPos.f_121853_);
+
+   private boolean attemptPlaceSculk(SculkSpreader p_222376_, LevelAccessor p_222377_, BlockPos p_222378_, RandomSource p_222379_, BlockPos sourceBlock) { // CraftBukkit
+      try {
+         this.catserver$attemptPlaceSculkSourceBlock.set(sourceBlock);
+         return this.m_222375_(p_222376_, p_222377_, p_222378_, p_222379_);
+      } finally {
+         this.catserver$attemptPlaceSculkSourceBlock.remove();
+      }
+   }
+   // CatServer end
+
    private boolean m_222375_(SculkSpreader p_222376_, LevelAccessor p_222377_, BlockPos p_222378_, RandomSource p_222379_) {
-      BlockState blockstate = p_222377_.m_8055_(p_222378_);
-      TagKey<Block> tagkey = p_222376_.m_222277_();
-
-      for(Direction direction : Direction.m_235667_(p_222379_)) {
-         if (m_153900_(blockstate, direction)) {
-            BlockPos blockpos = p_222378_.m_121945_(direction);
-            BlockState blockstate1 = p_222377_.m_8055_(blockpos);
-            if (blockstate1.m_204336_(tagkey)) {
-               BlockState blockstate2 = Blocks.f_220855_.m_49966_();
-               p_222377_.m_7731_(blockpos, blockstate2, 3);
-               Block.m_49897_(blockstate1, blockstate2, p_222377_, blockpos);
-               p_222377_.m_5594_((Player)null, blockpos, SoundEvents.f_215753_, SoundSource.BLOCKS, 1.0F, 1.0F);
-               this.f_222349_.m_221657_(blockstate2, p_222377_, blockpos, p_222376_.m_222282_());
-               Direction direction1 = direction.m_122424_();
-
-               for(Direction direction2 : f_153806_) {
-                  if (direction2 != direction1) {
-                     BlockPos blockpos1 = blockpos.m_121945_(direction2);
-                     BlockState blockstate3 = p_222377_.m_8055_(blockpos1);
-                     if (blockstate3.m_60713_(this)) {
-                        this.m_213805_(p_222377_, blockstate3, blockpos1, p_222379_);
-                     }
-                  }
+      try {
+         BlockState blockstate = p_222377_.m_8055_(p_222378_);
+         TagKey<Block> tagkey = p_222376_.m_222277_();
+
+         for(Direction direction : Direction.m_235667_(p_222379_)) {
+            if (m_153900_(blockstate, direction)) {
+               BlockPos blockpos = p_222378_.m_121945_(direction);
+               BlockState blockstate1 = p_222377_.m_8055_(blockpos);
+               if (blockstate1.m_204336_(tagkey)) {
+                  BlockState blockstate2 = Blocks.f_220855_.m_49966_();
+                  {
+                     BlockPos sourceBlock = this.catserver$attemptPlaceSculkSourceBlock.get(); // CatServer
+                     // CraftBukkit start - 1.20.1 Call BlockSpreadEvent
+                     if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockSpreadEvent(p_222377_, sourceBlock, blockpos, blockstate2, 3)) {
+                        return false;
+                     }
+                     // CraftBukkit end
+                  }
+                  Block.m_49897_(blockstate1, blockstate2, p_222377_, blockpos);
+                  p_222377_.m_5594_((Player)null, blockpos, SoundEvents.f_215753_, SoundSource.BLOCKS, 1.0F, 1.0F);
+                  this.f_222349_.m_221657_(blockstate2, p_222377_, blockpos, p_222376_.m_222282_());
+                  Direction direction1 = direction.m_122424_();
+
+                  for(Direction direction2 : f_153806_) {
+                     if (direction2 != direction1) {
+                        BlockPos blockpos1 = blockpos.m_121945_(direction2);
+                        BlockState blockstate3 = p_222377_.m_8055_(blockpos1);
+                        if (blockstate3.m_60713_(this)) {
+                           this.m_213805_(p_222377_, blockstate3, blockpos1, p_222379_);
+                        }
+                     }
+                  }
+
+                  return true;
                }
-
-               return true;
             }
          }
-      }
 
-      return false;
+         return false;
+      } finally {
+         this.catserver$attemptPlaceSculkSourceBlock.remove(); // CatServer
+      }
    }
 
    public static boolean m_222354_(LevelAccessor p_222355_, BlockState p_222356_, BlockPos p_222357_) {
