--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -52,7 +_,7 @@
    public static final EnumProperty<Direction> f_56183_ = DirectionalBlock.f_52588_;
    public static final ResourceLocation f_56184_ = new ResourceLocation("contents");
    @Nullable
-   private final DyeColor f_56185_;
+   public final DyeColor f_56185_;
 
    public ShulkerBoxBlock(@Nullable DyeColor p_56188_, BlockBehaviour.Properties p_56189_) {
       super(p_56189_);
