--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -101,7 +_,7 @@
             } else if (flag) {
                return InteractionResult.SUCCESS;
             } else if (!this.m_277189_(p_56281_, signblockentity) && p_56281_.m_36326_() && this.m_278656_(p_56281_, signblockentity, flag2)) {
-               this.m_276926_(p_56281_, signblockentity, flag2);
+               this.openTextEdit(p_56281_, signblockentity, flag2, org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT); // CraftBukkit - 1.20.1
                return InteractionResult.SUCCESS;
             } else {
                return InteractionResult.PASS;
@@ -146,7 +_,28 @@
       return woodtype;
    }
 
+   // CatServer start
+   private final ThreadLocal<org.bukkit.event.player.PlayerSignOpenEvent.Cause> catserver$openTextEditCause = ThreadLocal.withInitial(() -> org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN);
+
+   public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_, org.bukkit.event.player.PlayerSignOpenEvent.Cause cause) {
+      try {
+         this.catserver$openTextEditCause.set(cause);
+         // Craftbukkit start - 1.20.1
+         m_276926_(p_277738_, p_277467_, p_277771_);
+      } finally {
+         this.catserver$openTextEditCause.remove(); // CatServer - ensure reset
+      }
+   }
+
    public void m_276926_(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_) {
+      {
+         org.bukkit.event.player.PlayerSignOpenEvent.Cause cause = this.catserver$openTextEditCause.get();
+         this.catserver$openTextEditCause.remove();
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerSignOpenEvent(p_277738_, p_277467_, p_277771_, cause)) {
+            return;
+         }
+         // Craftbukkit end
+      }
       p_277467_.m_155713_(p_277738_.m_20148_());
       p_277738_.m_7739_(p_277467_, p_277771_);
    }
