--- a/net/minecraft/world/level/block/SnowLayerBlock.java
+++ b/net/minecraft/world/level/block/SnowLayerBlock.java
@@ -86,6 +_,11 @@
 
    public void m_213898_(BlockState p_222448_, ServerLevel p_222449_, BlockPos p_222450_, RandomSource p_222451_) {
       if (p_222449_.m_45517_(LightLayer.BLOCK, p_222450_) > 11) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callBlockFadeEvent(p_222449_, p_222450_, Blocks.f_50016_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          m_49950_(p_222448_, p_222449_, p_222450_);
          p_222449_.m_7471_(p_222450_, false);
       }
