--- a/net/minecraft/world/level/block/SnowLayerBlock.java
+++ b/net/minecraft/world/level/block/SnowLayerBlock.java
@@ -83,6 +_,11 @@
 
    public void m_7455_(BlockState p_56615_, ServerLevel p_56616_, BlockPos p_56617_, Random p_56618_) {
       if (p_56616_.m_45517_(LightLayer.BLOCK, p_56617_) > 11) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_56616_, p_56617_, Blocks.f_50016_.m_49966_()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          m_49950_(p_56615_, p_56616_, p_56617_);
          p_56616_.m_7471_(p_56617_, false);
       }
