--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -82,12 +_,13 @@
    public static final SoundType f_154680_ = new SoundType(1.0F, 1.0F, SoundEvents.f_144195_, SoundEvents.f_144199_, SoundEvents.f_144198_, SoundEvents.f_144197_, SoundEvents.f_144196_);
    public final float f_56731_;
    public final float f_56732_;
-   private final SoundEvent f_56733_;
+   public final SoundEvent f_56733_;
    private final SoundEvent f_56734_;
    private final SoundEvent f_56735_;
-   private final SoundEvent f_56737_;
+   public final SoundEvent f_56737_;
    private final SoundEvent f_56738_;
 
+   @Deprecated // Forge: Use {@link net.minecraftforge.common.util.ForgeSoundType} instead for suppliers
    public SoundType(float p_56766_, float p_56767_, SoundEvent p_56768_, SoundEvent p_56769_, SoundEvent p_56770_, SoundEvent p_56771_, SoundEvent p_56772_) {
       this.f_56731_ = p_56766_;
       this.f_56732_ = p_56767_;
