--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -5,7 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.BlockTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class StemBlock extends BushBlock implements BonemealableBlock {
    public static final int f_154724_ = 7;
@@ -43,22 +_,29 @@
    }
 
    public void m_7455_(BlockState p_57042_, ServerLevel p_57043_, BlockPos p_57044_, Random p_57045_) {
+      if (!p_57043_.isAreaLoaded(p_57044_, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
       if (p_57043_.m_45524_(p_57044_, 0) >= 9) {
          float f = CropBlock.m_52272_(this, p_57043_, p_57044_);
-         if (p_57045_.nextInt((int)(25.0F / f) + 1) == 0) {
+         // FoxServer TODO
+         if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_57043_, p_57044_, p_57042_, p_57045_.nextInt((int)(25.0F / f) + 1) == 0) && p_57045_.nextInt((int) ((100.0F / (this == Blocks.f_50189_ ? p_57043_.spigotConfig.pumpkinModifier : p_57043_.spigotConfig.melonModifier)) * (25.0F / f)) + 1) == 0) { // Spigot
             int i = p_57042_.m_61143_(f_57013_);
             if (i < 7) {
-               p_57042_ = p_57042_.m_61124_(f_57013_, Integer.valueOf(i + 1));
-               p_57043_.m_7731_(p_57044_, p_57042_, 2);
+               CraftEventFactory.handleBlockGrowEvent(p_57043_, p_57044_, p_57042_, 2); // CraftBukkit
             } else {
                Direction direction = Direction.Plane.HORIZONTAL.m_122560_(p_57045_);
                BlockPos blockpos = p_57044_.m_142300_(direction);
                BlockState blockstate = p_57043_.m_8055_(blockpos.m_7495_());
-               if (p_57043_.m_8055_(blockpos).m_60795_() && (blockstate.m_60713_(Blocks.f_50093_) || blockstate.m_204336_(BlockTags.f_144274_))) {
-                  p_57043_.m_46597_(blockpos, this.f_57015_.m_49966_());
+               Block block = blockstate.m_60734_();
+               if (p_57043_.m_46859_(blockpos) && (blockstate.canSustainPlant(p_57043_, blockpos.m_7495_(), Direction.UP, this) || block == Blocks.f_50093_ || block == Blocks.f_50493_ || block == Blocks.f_50546_ || block == Blocks.f_50599_ || block == Blocks.f_50440_)) {
+                  // CraftBukkit start
+                  if (!CraftEventFactory.handleBlockGrowEvent(p_57043_, blockpos, this.f_57015_.m_49966_())) {
+                     return;
+                  }
+                  // CraftBukkit end
                   p_57043_.m_46597_(p_57044_, this.f_57015_.m_7810_().m_49966_().m_61124_(HorizontalDirectionalBlock.f_54117_, direction));
                }
             }
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_57043_, p_57044_, p_57042_);
          }
 
       }
@@ -79,7 +_,7 @@
    public void m_7719_(ServerLevel p_57021_, Random p_57022_, BlockPos p_57023_, BlockState p_57024_) {
       int i = Math.min(7, p_57024_.m_61143_(f_57013_) + Mth.m_14072_(p_57021_.f_46441_, 2, 5));
       BlockState blockstate = p_57024_.m_61124_(f_57013_, Integer.valueOf(i));
-      p_57021_.m_7731_(p_57023_, blockstate, 2);
+      CraftEventFactory.handleBlockGrowEvent(p_57021_, p_57023_, blockstate, 2); // CraftBukkit
       if (i == 7) {
          blockstate.m_60735_(p_57021_, p_57023_, p_57021_.f_46441_);
       }
@@ -92,5 +_,11 @@
 
    public StemGrownBlock m_57056_() {
       return this.f_57015_;
+   }
+
+   //FORGE START
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.CROP;
    }
 }
