--- a/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.block;
 
+import java.util.Collections;
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
@@ -25,6 +_,10 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerHarvestBlockEvent;
 
 public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock {
    private static final float f_154738_ = 0.003F;
@@ -56,8 +_,9 @@
 
    public void m_7455_(BlockState p_57286_, ServerLevel p_57287_, BlockPos p_57288_, Random p_57289_) {
       int i = p_57286_.m_61143_(f_57244_);
-      if (i < 3 && p_57289_.nextInt(5) == 0 && p_57287_.m_45524_(p_57288_.m_7494_(), 0) >= 9) {
-         p_57287_.m_7731_(p_57288_, p_57286_.m_61124_(f_57244_, Integer.valueOf(i + 1)), 2);
+      if (i < 3 && p_57287_.m_45524_(p_57288_.m_7494_(), 0) >= 9 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_57287_, p_57288_, p_57286_,p_57289_.nextInt(5) == 0)) {
+         CraftEventFactory.handleBlockGrowEvent(p_57287_, p_57288_, p_57286_.m_61124_(f_57244_, i + 1), 2); // CraftBukkit
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_57287_, p_57288_, p_57286_);
       }
 
    }
@@ -69,7 +_,9 @@
             double d0 = Math.abs(p_57273_.m_20185_() - p_57273_.f_19790_);
             double d1 = Math.abs(p_57273_.m_20189_() - p_57273_.f_19792_);
             if (d0 >= (double)0.003F || d1 >= (double)0.003F) {
+               CraftEventFactory.blockDamage = CraftBlock.at(p_57271_, p_57272_); // CraftBukkit
                p_57273_.m_6469_(DamageSource.f_19325_, 1.0F);
+               CraftEventFactory.blockDamage = null; // CraftBukkit
             }
          }
 
@@ -83,7 +_,15 @@
          return InteractionResult.PASS;
       } else if (i > 1) {
          int j = 1 + p_57276_.f_46441_.nextInt(2);
-         m_49840_(p_57276_, p_57277_, new ItemStack(Items.f_42780_, j + (flag ? 1 : 0)));
+         // CraftBukkit start
+         PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(p_57276_, p_57277_, p_57278_, Collections.singletonList(new ItemStack(Items.f_42780_, j + (flag ? 1 : 0))));
+         if (event.isCancelled()) {
+            return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+         }
+         for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+            m_49840_(p_57276_, p_57277_, CraftItemStack.asNMSCopy(itemStack));
+         }
+         // CraftBukkit end
          p_57276_.m_5594_((Player)null, p_57277_, SoundEvents.f_12457_, SoundSource.BLOCKS, 1.0F, 0.8F + p_57276_.f_46441_.nextFloat() * 0.4F);
          p_57276_.m_7731_(p_57277_, p_57275_.m_61124_(f_57244_, Integer.valueOf(1)), 2);
          return InteractionResult.m_19078_(p_57276_.f_46443_);
