--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleWaterloggedBlock {
    public static final BooleanProperty f_57514_ = BlockStateProperties.f_61446_;
@@ -106,6 +_,19 @@
       if (!p_57548_.f_46443_) {
          boolean flag = p_57548_.m_46753_(p_57549_);
          if (flag != p_57547_.m_61143_(f_57516_)) {
+            // CraftBukkit start
+            org.bukkit.World bworld = p_57548_.getWorld();
+            org.bukkit.block.Block bblock = bworld.getBlockAt(p_57549_.m_123341_(), p_57549_.m_123342_(), p_57549_.m_123343_());
+
+            int power = bblock.getBlockPower();
+            int oldPower = (Boolean) p_57547_.m_61143_(f_57514_) ? 15 : 0;
+
+            if (oldPower == 0 ^ power == 0 || p_57550_.m_49966_().m_60803_()) {
+               BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bblock, oldPower, power);
+               p_57548_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+               flag = eventRedstone.getNewCurrent() > 0;
+            }
+            // CraftBukkit end
             if (p_57547_.m_61143_(f_57514_) != flag) {
                p_57547_ = p_57547_.m_61124_(f_57514_, Boolean.valueOf(flag));
                this.m_57527_((Player)null, p_57548_, p_57549_, flag);
@@ -152,4 +_,17 @@
 
       return super.m_7417_(p_57554_, p_57555_, p_57556_, p_57557_, p_57558_, p_57559_);
    }
+
+   //Forge Start
+   @Override
+   public boolean isLadder(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, net.minecraft.world.entity.LivingEntity entity) {
+      if (state.m_61143_(f_57514_)) {
+         BlockPos downPos = pos.m_7495_();
+         BlockState down = world.m_8055_(downPos);
+         return down.m_60734_().makesOpenTrapdoorAboveClimbable(down, world, downPos, state);
+      }
+      return false;
+   }
+   //Forge End
+
 }
