--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -8,7 +_,6 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.Items;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
@@ -21,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class TripWireBlock extends Block {
    public static final BooleanProperty f_57590_ = BlockStateProperties.f_61448_;
@@ -69,7 +_,7 @@
    }
 
    public void m_5707_(Level p_57615_, BlockPos p_57616_, BlockState p_57617_, Player p_57618_) {
-      if (!p_57615_.f_46443_ && !p_57618_.m_21205_().m_41619_() && p_57618_.m_21205_().m_150930_(Items.f_42574_)) {
+      if (!p_57615_.f_46443_ && !p_57618_.m_21205_().m_41619_() && p_57618_.m_21205_().canPerformAction(net.minecraftforge.common.ToolActions.SHEARS_DISARM)) {
          p_57615_.m_7731_(p_57616_, p_57617_.m_61124_(f_57592_, Boolean.valueOf(true)), 4);
          p_57615_.m_142346_(p_57618_, GameEvent.f_157781_, p_57616_);
       }
@@ -124,6 +_,39 @@
             }
          }
       }
+
+      // CraftBukkit start - Call interact even when triggering connected tripwire
+      if (flag != flag1 && flag1 && (Boolean)blockstate.m_61143_(f_57591_)) {
+         org.bukkit.World bworld = p_57608_.getWorld();
+         org.bukkit.plugin.PluginManager manager = p_57608_.getCraftServer().getPluginManager();
+         org.bukkit.block.Block block = bworld.getBlockAt(p_57609_.m_123341_(), p_57609_.m_123342_(), p_57609_.m_123343_());
+         boolean allowed = false;
+
+         // If all of the events are cancelled block the tripwire trigger, else allow
+         for (Object object : list) {
+            if (object != null) {
+               org.bukkit.event.Cancellable cancellable;
+               if (object instanceof Player) {
+                  cancellable = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) object, org.bukkit.event.block.Action.PHYSICAL, p_57609_, null, null, null);
+               } else if (object instanceof Entity) {
+                  cancellable = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+                  manager.callEvent((EntityInteractEvent) cancellable);
+               } else {
+                  continue;
+               }
+
+               if (!cancellable.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+
+         if (!allowed) {
+            return;
+         }
+      }
+      // CraftBukkit end
 
       if (flag1 != flag) {
          blockstate = blockstate.m_61124_(f_57590_, Boolean.valueOf(flag1));
