--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TripWireHookBlock extends Block {
    public static final DirectionProperty f_57667_ = HorizontalDirectionalBlock.f_54117_;
@@ -142,6 +_,17 @@
          this.m_57697_(p_57686_, blockpos1, flag2, flag3, flag, flag1);
       }
 
+      // CraftBukkit start
+      org.bukkit.block.Block block = p_57686_.getWorld().getBlockAt(p_57687_.m_123341_(), p_57687_.m_123342_(), p_57687_.m_123343_());
+
+      BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+      p_57686_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+      if (eventRedstone.getNewCurrent() > 0) {
+         return;
+      }
+      // CraftBukkit end
+
       this.m_57697_(p_57686_, p_57687_, flag2, flag3, flag, flag1);
       if (!p_57689_) {
          p_57686_.m_7731_(p_57687_, blockstate1.m_61124_(f_57667_, direction), 3);
@@ -155,8 +_,8 @@
             BlockPos blockpos2 = p_57687_.m_5484_(direction, k);
             BlockState blockstate2 = ablockstate[k];
             if (blockstate2 != null) {
+               if (!p_57686_.m_8055_(blockpos2).m_60795_()) { // FORGE: fix MC-129055
                p_57686_.m_7731_(blockpos2, blockstate2.m_61124_(f_57669_, Boolean.valueOf(flag2)), 3);
-               if (!p_57686_.m_8055_(blockpos2).m_60795_()) {
                }
             }
          }
