--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -29,6 +_,12 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityInteractEvent;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class TurtleEggBlock extends Block {
    public static final int f_154841_ = 2;
    public static final int f_154842_ = 1;
@@ -62,6 +_,18 @@
    private void m_154850_(Level p_154851_, BlockState p_154852_, BlockPos p_154853_, Entity p_154854_, int p_154855_) {
       if (this.m_57767_(p_154851_, p_154854_)) {
          if (!p_154851_.f_46443_ && p_154851_.f_46441_.m_188503_(p_154855_) == 0 && p_154852_.m_60713_(Blocks.f_50578_)) {
+            // CraftBukkit start - Step on eggs
+            org.bukkit.event.Cancellable cancellable;
+            if (p_154854_ instanceof Player) {
+               cancellable = CraftEventFactory.callPlayerInteractEvent((Player) p_154854_, org.bukkit.event.block.Action.PHYSICAL, p_154853_, null, null, null);
+            } else {
+               cancellable = new EntityInteractEvent(p_154854_.getBukkitEntity(), CraftBlock.at(p_154851_, p_154853_));
+               p_154851_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+            if (cancellable.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             this.m_57791_(p_154851_, p_154853_, p_154852_);
          }
 
@@ -85,9 +_,19 @@
       if (this.m_57765_(p_222645_) && m_57762_(p_222645_, p_222646_)) {
          int i = p_222644_.m_61143_(f_57753_);
          if (i < 2) {
+            // CraftBukkit start - Call BlockGrowEvent
+            if (!CraftEventFactory.handleBlockGrowEvent(p_222645_, p_222646_, p_222644_.m_61124_(f_57753_, i + 1), 2)) {
+               return;
+            }
+            // CraftBukkit end
             p_222645_.m_5594_((Player)null, p_222646_, SoundEvents.f_12534_, SoundSource.BLOCKS, 0.7F, 0.9F + p_222647_.m_188501_() * 0.2F);
-            p_222645_.m_7731_(p_222646_, p_222644_.m_61124_(f_57753_, Integer.valueOf(i + 1)), 2);
+            // p_222645_.setBlock(p_222646_, p_222644_.setValue(HATCH, Integer.valueOf(i + 1)), 2); // CraftBukkit - handled above
          } else {
+            // CraftBukkit start - Call BlockFadeEvent
+            if (CraftEventFactory.callBlockFadeEvent(p_222645_, p_222646_, Blocks.f_50016_.m_49966_()).isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             p_222645_.m_5594_((Player)null, p_222646_, SoundEvents.f_12535_, SoundSource.BLOCKS, 0.7F, 0.9F + p_222647_.m_188501_() * 0.2F);
             p_222645_.m_7471_(p_222646_, false);
 
@@ -98,7 +_,7 @@
                   turtle.m_146762_(-24000);
                   turtle.m_30219_(p_222646_);
                   turtle.m_7678_((double)p_222646_.m_123341_() + 0.3D + (double)j * 0.2D, (double)p_222646_.m_123342_(), (double)p_222646_.m_123343_() + 0.3D, 0.0F, 0.0F);
-                  p_222645_.m_7967_(turtle);
+                  p_222645_.addFreshEntity(turtle, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // CraftBukkit
                }
             }
          }
@@ -158,7 +_,7 @@
          if (!(p_57769_ instanceof LivingEntity)) {
             return false;
          } else {
-            return p_57769_ instanceof Player || p_57768_.m_46469_().m_46207_(GameRules.f_46132_);
+            return p_57769_ instanceof Player || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_57768_, p_57769_);
          }
       } else {
          return false;
