--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -21,8 +_,9 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty f_57833_ = PipeBlock.f_55152_;
    public static final BooleanProperty f_57834_ = PipeBlock.f_55148_;
    public static final BooleanProperty f_57835_ = PipeBlock.f_55149_;
@@ -157,7 +_,7 @@
    }
 
    public void m_7455_(BlockState p_57892_, ServerLevel p_57893_, BlockPos p_57894_, Random p_57895_) {
-      if (p_57895_.nextInt(4) == 0) {
+      if (p_57895_.nextInt(Math.max(1, (int) (100.0F / p_57893_.spigotConfig.vineModifier) * 4)) == 0 && p_57893_.isAreaLoaded(p_57894_, 4)) { // Forge: check area to prevent loading unloaded chunks
          Direction direction = Direction.m_122404_(p_57895_);
          BlockPos blockpos = p_57894_.m_7494_();
          if (direction.m_122434_().m_122479_() && !p_57892_.m_61143_(m_57883_(direction))) {
@@ -171,18 +_,22 @@
                   boolean flag1 = p_57892_.m_61143_(m_57883_(direction4));
                   BlockPos blockpos2 = blockpos4.m_142300_(direction3);
                   BlockPos blockpos3 = blockpos4.m_142300_(direction4);
+
+                  // CraftBukkit start - Call BlockSpreadEvent
+                  BlockPos source = p_57894_;
+
                   if (flag && m_57853_(p_57893_, blockpos2, direction3)) {
-                     p_57893_.m_7731_(blockpos4, this.m_49966_().m_61124_(m_57883_(direction3), Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_57893_, source, blockpos4, this.m_49966_().m_61124_(m_57883_(direction3), true), 2);
                   } else if (flag1 && m_57853_(p_57893_, blockpos3, direction4)) {
-                     p_57893_.m_7731_(blockpos4, this.m_49966_().m_61124_(m_57883_(direction4), Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_57893_, source, blockpos4, this.m_49966_().m_61124_(m_57883_(direction4), true), 2);
                   } else {
                      Direction direction1 = direction.m_122424_();
                      if (flag && p_57893_.m_46859_(blockpos2) && m_57853_(p_57893_, p_57894_.m_142300_(direction3), direction1)) {
-                        p_57893_.m_7731_(blockpos2, this.m_49966_().m_61124_(m_57883_(direction1), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_57893_, source, blockpos2, this.m_49966_().m_61124_(m_57883_(direction1), true), 2);
                      } else if (flag1 && p_57893_.m_46859_(blockpos3) && m_57853_(p_57893_, p_57894_.m_142300_(direction4), direction1)) {
-                        p_57893_.m_7731_(blockpos3, this.m_49966_().m_61124_(m_57883_(direction1), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_57893_, source, blockpos3, this.m_49966_().m_61124_(m_57883_(direction1), true), 2);
                      } else if ((double)p_57895_.nextFloat() < 0.05D && m_57853_(p_57893_, blockpos4.m_7494_(), Direction.UP)) {
-                        p_57893_.m_7731_(blockpos4, this.m_49966_().m_61124_(f_57833_, Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_57893_, source, blockpos4, this.m_49966_().m_61124_(f_57833_, true), 2);
                      }
                   }
                } else if (m_57853_(p_57893_, blockpos4, direction)) {
@@ -211,7 +_,7 @@
                   }
 
                   if (this.m_57911_(blockstate3)) {
-                     p_57893_.m_7731_(blockpos, blockstate3, 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_57893_, p_57894_, blockpos, blockstate3, 2); // CraftBukkit
                   }
 
                   return;
@@ -225,7 +_,7 @@
                   BlockState blockstate1 = blockstate.m_60795_() ? this.m_49966_() : blockstate;
                   BlockState blockstate2 = this.m_57870_(p_57892_, blockstate1, p_57895_);
                   if (blockstate1 != blockstate2 && this.m_57911_(blockstate2)) {
-                     p_57893_.m_7731_(blockpos1, blockstate2, 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_57893_, p_57894_, blockpos1, blockstate2, 2); // CraftBukkit
                   }
                }
             }
