--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -23,7 +_,9 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class VineBlock extends Block {
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory; // CraftBukkit
+
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty f_57833_ = PipeBlock.f_55152_;
    public static final BooleanProperty f_57834_ = PipeBlock.f_55148_;
    public static final BooleanProperty f_57835_ = PipeBlock.f_55149_;
@@ -158,7 +_,7 @@
 
    public void m_213898_(BlockState p_222655_, ServerLevel p_222656_, BlockPos p_222657_, RandomSource p_222658_) {
       if (p_222656_.m_46469_().m_46207_(GameRules.f_268705_)) {
-         if (p_222658_.m_188503_(4) == 0) {
+      if (p_222656_.f_46441_.m_188501_() < (p_222656_.spigotConfig.vineModifier / (100.0f * 4)) && p_222656_.isAreaLoaded(p_222657_, 4)) { // Forge: check area to prevent loading unloaded chunks // Spigot - SPIGOT-7159: Better modifier resolution
             Direction direction = Direction.m_235672_(p_222658_);
             BlockPos blockpos = p_222657_.m_7494_();
             if (direction.m_122434_().m_122479_() && !p_222655_.m_61143_(m_57883_(direction))) {
@@ -172,29 +_,35 @@
                      boolean flag1 = p_222655_.m_61143_(m_57883_(direction4));
                      BlockPos blockpos2 = blockpos4.m_121945_(direction3);
                      BlockPos blockpos3 = blockpos4.m_121945_(direction4);
+
+                     // CraftBukkit start - Call BlockSpreadEvent
+                     BlockPos source = p_222657_;
+
                      if (flag && m_57853_(p_222656_, blockpos2, direction3)) {
-                        p_222656_.m_7731_(blockpos4, this.m_49966_().m_61124_(m_57883_(direction3), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, (BlockState) this.m_49966_().m_61124_(m_57883_(direction3), true), 2);
                      } else if (flag1 && m_57853_(p_222656_, blockpos3, direction4)) {
-                        p_222656_.m_7731_(blockpos4, this.m_49966_().m_61124_(m_57883_(direction4), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, (BlockState) this.m_49966_().m_61124_(m_57883_(direction4), true), 2);
                      } else {
                         Direction direction1 = direction.m_122424_();
                         if (flag && p_222656_.m_46859_(blockpos2) && m_57853_(p_222656_, p_222657_.m_121945_(direction3), direction1)) {
-                           p_222656_.m_7731_(blockpos2, this.m_49966_().m_61124_(m_57883_(direction1), Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos2, (BlockState) this.m_49966_().m_61124_(m_57883_(direction1), true), 2);
                         } else if (flag1 && p_222656_.m_46859_(blockpos3) && m_57853_(p_222656_, p_222657_.m_121945_(direction4), direction1)) {
-                           p_222656_.m_7731_(blockpos3, this.m_49966_().m_61124_(m_57883_(direction1), Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos3, (BlockState) this.m_49966_().m_61124_(m_57883_(direction1), true), 2);
                         } else if ((double)p_222658_.m_188501_() < 0.05D && m_57853_(p_222656_, blockpos4.m_7494_(), Direction.UP)) {
-                           p_222656_.m_7731_(blockpos4, this.m_49966_().m_61124_(f_57833_, Boolean.valueOf(true)), 2);
+                           // p_222656_.setBlock(blockpos4, this.defaultBlockState().setValue(UP, Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, (BlockState) this.m_49966_().m_61124_(f_57833_, true), 2);
                         }
+                        // CraftBukkit end
                      }
                   } else if (m_57853_(p_222656_, blockpos4, direction)) {
-                     p_222656_.m_7731_(p_222657_, p_222655_.m_61124_(m_57883_(direction), Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockGrowEvent(p_222656_, p_222657_, (BlockState) p_222655_.m_61124_(m_57883_(direction), true), 2); // CraftBukkit - 1.20.1
                   }
 
                }
             } else {
                if (direction == Direction.UP && p_222657_.m_123342_() < p_222656_.m_151558_() - 1) {
                   if (this.m_57887_(p_222656_, p_222657_, direction)) {
-                     p_222656_.m_7731_(p_222657_, p_222655_.m_61124_(f_57833_, Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockGrowEvent(p_222656_, p_222657_, (BlockState) p_222655_.m_61124_(f_57833_, true), 2); // CraftBukkit - 1.20.1
                      return;
                   }
 
@@ -212,7 +_,7 @@
                      }
 
                      if (this.m_57911_(blockstate3)) {
-                        p_222656_.m_7731_(blockpos, blockstate3, 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, p_222657_, blockpos, blockstate3, 2); // CraftBukkit
                      }
 
                      return;
@@ -226,7 +_,7 @@
                      BlockState blockstate1 = blockstate.m_60795_() ? this.m_49966_() : blockstate;
                      BlockState blockstate2 = this.m_222650_(p_222655_, blockstate1, p_222658_);
                      if (blockstate1 != blockstate2 && this.m_57911_(blockstate2)) {
-                        p_222656_.m_7731_(blockpos1, blockstate2, 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, p_222657_, blockpos1, blockstate2, 2); // CraftBukkit
                      }
                   }
                }
