--- a/net/minecraft/world/level/block/WaterlilyBlock.java
+++ b/net/minecraft/world/level/block/WaterlilyBlock.java
@@ -23,7 +_,7 @@
 
    public void m_7892_(BlockState p_58164_, Level p_58165_, BlockPos p_58166_, Entity p_58167_) {
       super.m_7892_(p_58164_, p_58165_, p_58166_, p_58167_);
-      if (p_58165_ instanceof ServerLevel && p_58167_ instanceof Boat) {
+      if (p_58165_ instanceof ServerLevel && p_58167_ instanceof Boat && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_58167_, p_58166_, Blocks.f_50016_.m_49966_()).isCancelled()) { // CraftBukkit
          p_58165_.m_46953_(new BlockPos(p_58166_), true, p_58167_);
       }
 
