--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
    public static final IntegerProperty f_58198_ = BlockStateProperties.f_61426_;
@@ -25,7 +_,29 @@
    }
 
    protected int m_6693_(Level p_58213_, BlockPos p_58214_) {
-      int i = Math.min(p_58213_.m_45976_(Entity.class, f_49287_.m_82338_(p_58214_)).size(), this.f_58199_);
+      // CraftBukkit start
+      // int i = Math.min(p_58213_.getEntitiesOfClass(Entity.class, TOUCH_AABB.move(p_58214_)).size(), this.maxWeight);
+      int i = 0;
+      java.util.Iterator iterator = p_58213_.m_45976_(Entity.class, f_49287_.m_82338_(p_58214_)).iterator();
+      while (iterator.hasNext()) {
+         Entity entity = (Entity) iterator.next();
+         org.bukkit.event.Cancellable cancellable;
+
+         if (entity instanceof Player) {
+            cancellable = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, p_58214_, null, null, null);
+         } else {
+            cancellable = new EntityInteractEvent(entity.getBukkitEntity(), p_58213_.getWorld().getBlockAt(p_58214_.m_123341_(), p_58214_.m_123342_(), p_58214_.m_123343_()));
+            p_58213_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+         }
+
+         // We only want to block turning the plate on if all events are cancelled
+         if (!cancellable.isCancelled()) {
+            i++;
+         }
+      }
+
+      i = Math.min(i, this.f_58199_);
+      // CraftBukkit end
       if (i > 0) {
          float f = (float)Math.min(this.f_58199_, i) / (float)this.f_58199_;
          return Mth.m_14167_(f * 15.0F);
