--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -23,6 +_,8 @@
 import net.minecraft.world.level.block.state.predicate.BlockMaterialPredicate;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
 import net.minecraft.world.level.material.Material;
+import org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class WitherSkullBlock extends SkullBlock {
    @Nullable
@@ -44,6 +_,7 @@
    }
 
    public static void m_58255_(Level p_58256_, BlockPos p_58257_, SkullBlockEntity p_58258_) {
+      if (p_58256_.captureBlockStates) return; // CraftBukkit
       if (!p_58256_.f_46443_) {
          BlockState blockstate = p_58258_.m_58900_();
          boolean flag = blockstate.m_60713_(Blocks.f_50312_) || blockstate.m_60713_(Blocks.f_50313_);
@@ -51,11 +_,14 @@
             BlockPattern blockpattern = m_58273_();
             BlockPattern.BlockPatternMatch blockpattern$blockpatternmatch = blockpattern.m_61184_(p_58256_, p_58257_);
             if (blockpattern$blockpatternmatch != null) {
+               // CraftBukkit start - Use BlockStateListPopulator
+               BlockStateListPopulator blockList = new BlockStateListPopulator(p_58256_);
                for(int i = 0; i < blockpattern.m_61203_(); ++i) {
                   for(int j = 0; j < blockpattern.m_61202_(); ++j) {
                      BlockInWorld blockinworld = blockpattern$blockpatternmatch.m_61229_(i, j, 0);
-                     p_58256_.m_7731_(blockinworld.m_61176_(), Blocks.f_50016_.m_49966_(), 2);
-                     p_58256_.m_46796_(2001, blockinworld.m_61176_(), Block.m_49956_(blockinworld.m_61168_()));
+                     //p_58256_.setBlock(blockinworld.getPos(), Blocks.AIR.defaultBlockState(), 2);
+                     //p_58256_.levelEvent(2001, blockinworld.getPos(), Block.getId(blockinworld.getState()));
+                     blockList.m_7731_(blockinworld.m_61176_(), Blocks.f_50016_.m_49966_(), 2); // CraftBukkit
                   }
                }
 
@@ -64,12 +_,20 @@
                witherboss.m_7678_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.55D, (double)blockpos.m_123343_() + 0.5D, blockpattern$blockpatternmatch.m_61233_().m_122434_() == Direction.Axis.X ? 0.0F : 90.0F, 0.0F);
                witherboss.f_20883_ = blockpattern$blockpatternmatch.m_61233_().m_122434_() == Direction.Axis.X ? 0.0F : 90.0F;
                witherboss.m_31506_();
-
+               // CraftBukkit start
+               if (!p_58256_.addFreshEntity(witherboss, CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
+                  return;
+               }
+               for (BlockPos pos : blockList.getBlocks()) {
+                  p_58256_.m_46796_(2001, pos, Block.m_49956_(p_58256_.m_8055_(pos)));
+               }
+               blockList.updateList();
+               // CraftBukkit end
                for(ServerPlayer serverplayer : p_58256_.m_45976_(ServerPlayer.class, witherboss.m_142469_().m_82400_(50.0D))) {
                   CriteriaTriggers.f_10580_.m_68256_(serverplayer, witherboss);
                }
 
-               p_58256_.m_7967_(witherboss);
+               // p_58256_.addFreshEntity(witherboss); // CraftBukkit - moved up
 
                for(int k = 0; k < blockpattern.m_61203_(); ++k) {
                   for(int l = 0; l < blockpattern.m_61202_(); ++l) {
