--- a/net/minecraft/world/level/block/WitherRoseBlock.java
+++ b/net/minecraft/world/level/block/WitherRoseBlock.java
@@ -46,7 +_,7 @@
          if (p_58241_ instanceof LivingEntity) {
             LivingEntity livingentity = (LivingEntity)p_58241_;
             if (!livingentity.m_6673_(DamageSource.f_19320_)) {
-               livingentity.m_7292_(new MobEffectInstance(MobEffects.f_19615_, 40));
+               livingentity.addEffect(new MobEffectInstance(MobEffects.f_19615_, 40), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WITHER_ROSE); // CraftBukkit
             }
          }
 
