--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -42,6 +_,7 @@
    }
 
    public static void m_58255_(Level p_58256_, BlockPos p_58257_, SkullBlockEntity p_58258_) {
+      if (p_58256_.captureBlockStates) return; // CraftBukkit
       if (!p_58256_.f_46443_) {
          BlockState blockstate = p_58258_.m_58900_();
          boolean flag = blockstate.m_60713_(Blocks.f_50312_) || blockstate.m_60713_(Blocks.f_50313_);
@@ -50,17 +_,23 @@
             if (blockpattern$blockpatternmatch != null) {
                WitherBoss witherboss = EntityType.f_20496_.m_20615_(p_58256_);
                if (witherboss != null) {
-                  CarvedPumpkinBlock.m_245585_(p_58256_, blockpattern$blockpatternmatch);
+                  // CarvedPumpkinBlock.clearPatternBlocks(p_58256_, blockpattern$blockpatternmatch); // CraftBukkit - move down
                   BlockPos blockpos = blockpattern$blockpatternmatch.m_61229_(1, 2, 0).m_61176_();
                   witherboss.m_7678_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.55D, (double)blockpos.m_123343_() + 0.5D, blockpattern$blockpatternmatch.m_61233_().m_122434_() == Direction.Axis.X ? 0.0F : 90.0F, 0.0F);
                   witherboss.f_20883_ = blockpattern$blockpatternmatch.m_61233_().m_122434_() == Direction.Axis.X ? 0.0F : 90.0F;
                   witherboss.m_31506_();
+                  // CraftBukkit start
+                  if (!p_58256_.addFreshEntity(witherboss, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
+                     return;
+                  }
+                  CarvedPumpkinBlock.m_245585_(p_58256_, blockpattern$blockpatternmatch); // CraftBukkit - from above
+                  // CraftBukkit end
 
                   for(ServerPlayer serverplayer : p_58256_.m_45976_(ServerPlayer.class, witherboss.m_20191_().m_82400_(50.0D))) {
                      CriteriaTriggers.f_10580_.m_68256_(serverplayer, witherboss);
                   }
 
-                  p_58256_.m_7967_(witherboss);
+                  // p_58256_.addFreshEntity(witherboss); // CraftBukkit - moved up
                   CarvedPumpkinBlock.m_246758_(p_58256_, blockpattern$blockpatternmatch);
                }
 
