--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -27,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.Pose;
+import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
@@ -41,6 +_,8 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.player.PlayerPickupArrowEvent;
 
 public abstract class AbstractArrow extends Projectile {
    private static final double f_150120_ = 2.0D;
@@ -51,11 +_,11 @@
    private static final int f_150119_ = 4;
    @Nullable
    private BlockState f_36696_;
-   protected boolean f_36703_;
+   public boolean f_36703_;
    protected int f_36704_;
    public AbstractArrow.Pickup f_36705_ = AbstractArrow.Pickup.DISALLOWED;
    public int f_36706_;
-   private int f_36697_;
+   public int f_36697_;
    private double f_36698_ = 2.0D;
    private int f_36699_;
    private SoundEvent f_36700_ = this.m_7239_();
@@ -79,7 +_,6 @@
       if (p_36718_ instanceof Player) {
          this.f_36705_ = AbstractArrow.Pickup.ALLOWED;
       }
-
    }
 
    public void m_36740_(SoundEvent p_36741_) {
@@ -184,8 +_,8 @@
                }
             }
 
-            if (hitresult != null && !flag) {
-               this.m_6532_(hitresult);
+            if (hitresult != null && hitresult.m_6662_() != HitResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
+               this.preOnHit(hitresult); // CraftBukkit - projectile hit event
                this.f_19812_ = true;
             }
 
@@ -320,7 +_,13 @@
       boolean flag = entity.m_6095_() == EntityType.f_20566_;
       int k = entity.m_20094_();
       if (this.m_6060_() && !flag) {
-         entity.m_20254_(5);
+         // CraftBukkit start
+         EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+         org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+         if (!combustEvent.isCancelled()) {
+            entity.setSecondsOnFire(combustEvent.getDuration(), false);
+         }
+         // CraftBukkit end
       }
 
       if (entity.m_6469_(damagesource, (float)i)) {
@@ -472,7 +_,20 @@
 
    public void m_6123_(Player p_36766_) {
       if (!this.f_19853_.f_46443_ && (this.f_36703_ || this.m_36797_()) && this.f_36706_ <= 0) {
-         if (this.m_142470_(p_36766_)) {
+         // CraftBukkit start
+         ItemStack itemstack = this.m_7941_();
+         if (this.f_36705_ == Pickup.ALLOWED && !itemstack.m_41619_() && p_36766_.m_150109_().canHold(itemstack) > 0) {
+            ItemEntity item = new ItemEntity(this.f_19853_, this.m_20185_(), this.m_20186_(), this.m_20189_(), itemstack);
+            PlayerPickupArrowEvent event = new PlayerPickupArrowEvent((org.bukkit.entity.Player) p_36766_.getBukkitEntity(), new org.bukkit.craftbukkit.v1_18_R2.entity.CraftItem(this.f_19853_.getCraftServer(), this, item), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
+            // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+            this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            itemstack = item.m_32055_();
+         }
+         if ((this.f_36705_ == AbstractArrow.Pickup.ALLOWED && p_36766_.m_150109_().m_36054_(itemstack)) || (this.f_36705_ == AbstractArrow.Pickup.CREATIVE_ONLY && p_36766_.m_150110_().f_35937_)) {
+            // CraftBukkit end
             p_36766_.m_7938_(this, 1);
             this.m_146870_();
          }
