--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -13,11 +_,14 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public abstract class AbstractHurtingProjectile extends Projectile {
    public double f_36813_;
    public double f_36814_;
    public double f_36815_;
+   public float bukkitYield = 1; // CraftBukkit
+   public boolean isIncendiary = true; // CraftBukkit
 
    protected AbstractHurtingProjectile(EntityType<? extends AbstractHurtingProjectile> p_36833_, Level p_36834_) {
       super(p_36833_, p_36834_);
@@ -36,6 +_,15 @@
 
    }
 
+   public void setDirection(double p_36821_, double p_36822_, double p_36823_) {
+      double d0 = Math.sqrt(p_36821_ * p_36821_ + p_36822_ * p_36822_ + p_36823_ * p_36823_);
+      if (d0 != 0.0D) {
+         this.f_36813_ = p_36821_ / d0 * 0.1D;
+         this.f_36814_ = p_36822_ / d0 * 0.1D;
+         this.f_36815_ = p_36823_ / d0 * 0.1D;
+      }
+   }
+
    public AbstractHurtingProjectile(EntityType<? extends AbstractHurtingProjectile> p_36826_, LivingEntity p_36827_, double p_36828_, double p_36829_, double p_36830_, Level p_36831_) {
       this(p_36826_, p_36827_.m_20185_(), p_36827_.m_20186_(), p_36827_.m_20189_(), p_36828_, p_36829_, p_36830_, p_36831_);
       this.m_5602_(p_36827_);
@@ -64,8 +_,13 @@
          }
 
          HitResult hitresult = ProjectileUtil.m_37294_(this, this::m_5603_);
-         if (hitresult.m_6662_() != HitResult.Type.MISS) {
+         if (hitresult.m_6662_() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
             this.m_6532_(hitresult);
+            // CraftBukkit start - Fire ProjectileHitEvent
+            if (this.m_146910_()) {
+               CraftEventFactory.callProjectileHitEvent(this, hitresult);
+            }
+            // CraftBukkit end
          }
 
          this.m_20101_();
@@ -141,15 +_,17 @@
          this.m_5834_();
          Entity entity = p_36839_.m_7639_();
          if (entity != null) {
-            if (!this.f_19853_.f_46443_) {
-               Vec3 vec3 = entity.m_20154_();
-               this.m_20256_(vec3);
-               this.f_36813_ = vec3.f_82479_ * 0.1D;
-               this.f_36814_ = vec3.f_82480_ * 0.1D;
-               this.f_36815_ = vec3.f_82481_ * 0.1D;
-               this.m_5602_(entity);
+            // CraftBukkit start
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_36839_, p_36840_, false)) {
+               return false;
             }
-
+            // CraftBukkit end
+            Vec3 vec3 = entity.m_20154_();
+            this.m_20256_(vec3);
+            this.f_36813_ = vec3.f_82479_ * 0.1D;
+            this.f_36814_ = vec3.f_82480_ * 0.1D;
+            this.f_36815_ = vec3.f_82481_ * 0.1D;
+            this.m_5602_(entity);
             return true;
          } else {
             return false;
