--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -128,7 +_,7 @@
       return this.f_19804_.m_135370_(f_36854_);
    }
 
-   private void m_36882_(int p_36883_) {
+   public void m_36882_(int p_36883_) {
       this.f_36853_ = true;
       this.f_19804_.m_135381_(f_36854_, p_36883_);
    }
@@ -178,12 +_,12 @@
       Entity entity = this.m_150173_();
 
       for(MobEffectInstance mobeffectinstance : this.f_36855_.m_43488_()) {
-         p_36873_.m_147207_(new MobEffectInstance(mobeffectinstance.m_19544_(), Math.max(mobeffectinstance.m_19557_() / 8, 1), mobeffectinstance.m_19564_(), mobeffectinstance.m_19571_(), mobeffectinstance.m_19572_()), entity);
+         p_36873_.addEffect(new MobEffectInstance(mobeffectinstance.m_19544_(), Math.max(mobeffectinstance.m_19557_() / 8, 1), mobeffectinstance.m_19564_(), mobeffectinstance.m_19571_(), mobeffectinstance.m_19572_()), entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
       }
 
       if (!this.f_36852_.isEmpty()) {
          for(MobEffectInstance mobeffectinstance1 : this.f_36852_) {
-            p_36873_.m_147207_(mobeffectinstance1, entity);
+            p_36873_.addEffect(mobeffectinstance1, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
          }
       }
 
@@ -221,4 +_,23 @@
       }
 
    }
+
+   // CraftBukkit start accessor methods
+   public void refreshEffects() {
+      this.m_20088_().m_135381_(f_36854_, PotionUtils.m_43564_((Collection) PotionUtils.m_43561_(this.f_36855_, (Collection) this.f_36852_)));
+   }
+
+   public String getPotionType() {
+      return Registry.f_122828_.m_7981_(this.f_36855_).toString();
+   }
+
+   public void setPotionType(String string) {
+      this.f_36855_ = Registry.f_122828_.m_7745_(new net.minecraft.resources.ResourceLocation(string));
+      this.m_20088_().m_135381_(f_36854_, PotionUtils.m_43564_((Collection) PotionUtils.m_43561_(this.f_36855_, (Collection) this.f_36852_)));
+   }
+
+   public boolean isTipped() {
+      return !(this.f_36852_.isEmpty() && this.f_36855_ == Potions.f_43598_);
+   }
+   // CraftBukkit end
 }
