--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -24,10 +_,10 @@
 public class Arrow extends AbstractArrow {
    private static final int f_150131_ = 600;
    private static final int f_150129_ = -1;
-   private static final EntityDataAccessor<Integer> f_36854_ = SynchedEntityData.m_135353_(Arrow.class, EntityDataSerializers.f_135028_);
+   public static final EntityDataAccessor<Integer> f_36854_ = SynchedEntityData.m_135353_(Arrow.class, EntityDataSerializers.f_135028_);
    private static final byte f_150130_ = 0;
    private Potion f_36855_ = Potions.f_43598_;
-   private final Set<MobEffectInstance> f_36852_ = Sets.newHashSet();
+   public final Set<MobEffectInstance> f_36852_ = Sets.newHashSet();
    private boolean f_36853_;
 
    public Arrow(EntityType<? extends Arrow> p_36858_, Level p_36859_) {
@@ -42,6 +_,25 @@
       super(EntityType.f_20548_, p_36867_, p_36866_);
    }
 
+   // CraftBukkit start accessor methods
+   public void refreshEffects() {
+      this.m_20088_().m_135381_(Arrow.f_36854_, PotionUtils.m_43564_(PotionUtils.m_43561_(this.f_36855_, this.f_36852_)));
+   }
+
+   public String getPotionType() {
+      return Registry.f_122828_.m_7981_(this.f_36855_).toString();
+   }
+
+   public void setPotionType(String string) {
+      this.f_36855_ = Registry.f_122828_.m_7745_(new net.minecraft.resources.ResourceLocation(string));
+      this.m_20088_().m_135381_(Arrow.f_36854_, PotionUtils.m_43564_(PotionUtils.m_43561_(this.f_36855_, this.f_36852_)));
+   }
+
+   public boolean isTipped() {
+      return !(this.f_36852_.isEmpty() && this.f_36855_ == Potions.f_43598_);
+   }
+   // CraftBukkit end
+
    public void m_36878_(ItemStack p_36879_) {
       if (p_36879_.m_150930_(Items.f_42738_)) {
          this.f_36855_ = PotionUtils.m_43579_(p_36879_);
@@ -128,7 +_,7 @@
       return this.f_19804_.m_135370_(f_36854_);
    }
 
-   private void m_36882_(int p_36883_) {
+   public void m_36882_(int p_36883_) {
       this.f_36853_ = true;
       this.f_19804_.m_135381_(f_36854_, p_36883_);
    }
@@ -178,12 +_,12 @@
       Entity entity = this.m_150173_();
 
       for(MobEffectInstance mobeffectinstance : this.f_36855_.m_43488_()) {
-         p_36873_.m_147207_(new MobEffectInstance(mobeffectinstance.m_19544_(), Math.max(mobeffectinstance.m_19557_() / 8, 1), mobeffectinstance.m_19564_(), mobeffectinstance.m_19571_(), mobeffectinstance.m_19572_()), entity);
+         p_36873_.addEffect(new MobEffectInstance(mobeffectinstance.m_19544_(), Math.max(mobeffectinstance.m_19557_() / 8, 1), mobeffectinstance.m_19564_(), mobeffectinstance.m_19571_(), mobeffectinstance.m_19572_()), entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
       }
 
       if (!this.f_36852_.isEmpty()) {
          for(MobEffectInstance mobeffectinstance1 : this.f_36852_) {
-            p_36873_.m_147207_(mobeffectinstance1, entity);
+            p_36873_.addEffect(mobeffectinstance1, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
          }
       }
 
