--- a/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -113,7 +_,9 @@
       LivingEntity livingentity = this.m_19749_();
       if (p_36945_.m_6084_() && !p_36945_.m_20147_() && p_36945_ != livingentity) {
          if (livingentity == null) {
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = this; // CraftBukkit
             p_36945_.m_6469_(this.m_269291_().m_269425_(), 6.0F);
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = null; // CraftBukkit
          } else {
             if (livingentity.m_7307_(p_36945_)) {
                return;
