--- a/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -115,7 +_,9 @@
       LivingEntity livingentity = this.m_36947_();
       if (p_36945_.m_6084_() && !p_36945_.m_20147_() && p_36945_ != livingentity) {
          if (livingentity == null) {
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = this; // CraftBukkit
             p_36945_.m_6469_(DamageSource.f_19319_, 6.0F);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = this; // CraftBukkit
          } else {
             if (livingentity.m_7307_(p_36945_)) {
                return;
