--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -21,11 +_,11 @@
 
 public class EyeOfEnder extends Entity implements ItemSupplier {
    private static final EntityDataAccessor<ItemStack> f_36949_ = SynchedEntityData.m_135353_(EyeOfEnder.class, EntityDataSerializers.f_135033_);
-   private double f_36950_;
-   private double f_36951_;
-   private double f_36952_;
-   private int f_36953_;
-   private boolean f_36954_;
+   public double f_36950_;
+   public double f_36951_;
+   public double f_36952_;
+   public int f_36953_;
+   public boolean f_36954_;
 
    public EyeOfEnder(EntityType<? extends EyeOfEnder> p_36957_, Level p_36958_) {
       super(p_36957_, p_36958_);
@@ -37,7 +_,7 @@
    }
 
    public void m_36972_(ItemStack p_36973_) {
-      if (!p_36973_.m_150930_(Items.f_42545_) || p_36973_.m_41782_()) {
+      if (true || !p_36973_.m_150930_(Items.f_42545_) || p_36973_.m_41782_()) { // CraftBukkit - always allow item changing
          this.m_20088_().m_135381_(f_36949_, Util.m_137469_(p_36973_.m_41777_(), (p_36978_) -> {
             p_36978_.m_41764_(1);
          }));
@@ -164,7 +_,7 @@
 
    public void m_7378_(CompoundTag p_36970_) {
       ItemStack itemstack = ItemStack.m_41712_(p_36970_.m_128469_("Item"));
-      this.m_36972_(itemstack);
+      if (!itemstack.m_41619_()) this.m_36972_(itemstack); // CraftBukkit - SPIGOT-6103 summon, see also SPIGOT-5474
    }
 
    public float m_6073_() {
