--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -26,14 +_,16 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+
 public class FireworkRocketEntity extends Projectile implements ItemSupplier {
-   private static final EntityDataAccessor<ItemStack> f_37019_ = SynchedEntityData.m_135353_(FireworkRocketEntity.class, EntityDataSerializers.f_135033_);
+   public static final EntityDataAccessor<ItemStack> f_37019_ = SynchedEntityData.m_135353_(FireworkRocketEntity.class, EntityDataSerializers.f_135033_); // CatServer - private -> public
    private static final EntityDataAccessor<OptionalInt> f_37020_ = SynchedEntityData.m_135353_(FireworkRocketEntity.class, EntityDataSerializers.f_135044_);
-   private static final EntityDataAccessor<Boolean> f_37021_ = SynchedEntityData.m_135353_(FireworkRocketEntity.class, EntityDataSerializers.f_135035_);
-   private int f_37022_;
-   private int f_37023_;
+   public static final EntityDataAccessor<Boolean> f_37021_ = SynchedEntityData.m_135353_(FireworkRocketEntity.class, EntityDataSerializers.f_135035_); // CatServer - private -> public
+   public int f_37022_; // CatServer - private -> public
+   public int f_37023_; // CatServer - private -> public
    @Nullable
-   private LivingEntity f_37024_;
+   public LivingEntity f_37024_; // CatServer - private -> public
 
    public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> p_37027_, Level p_37028_) {
       super(p_37027_, p_37028_);
@@ -74,6 +_,21 @@
       this.m_5602_(p_37052_);
    }
 
+   // Spigot Start - copied from tick
+   @Override
+   public void inactiveTick() {
+      this.f_37022_ += 1;
+      if (!this.m_9236_().f_46443_ && this.f_37022_ > this.f_37023_) {
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.m_37080_();
+         }
+         // CraftBukkit end
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected void m_8097_() {
       this.f_19804_.m_135372_(f_37019_, ItemStack.f_41583_);
       this.f_19804_.m_135372_(f_37020_, OptionalInt.empty());
@@ -130,7 +_,7 @@
 
       HitResult hitresult = ProjectileUtil.m_278158_(this, this::m_5603_);
       if (!this.f_19794_) {
-         this.m_6532_(hitresult);
+         this.preOnHit(hitresult); // CraftBukkit - projectile hit event
          this.f_19812_ = true;
       }
 
@@ -145,9 +_,20 @@
       }
 
       if (!this.m_9236_().f_46443_ && this.f_37022_ > this.f_37023_) {
-         this.m_37080_();
-      }
-
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.m_37080_();
+         }
+         // CraftBukkit end
+      }
+
+   }
+
+   @Override
+   protected void m_6532_(HitResult result) {
+      if (result.m_6662_() == HitResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.m_6532_(result);
+      }
    }
 
    private void m_37080_() {
@@ -160,7 +_,11 @@
    protected void m_5790_(EntityHitResult p_37071_) {
       super.m_5790_(p_37071_);
       if (!this.m_9236_().f_46443_) {
-         this.m_37080_();
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.m_37080_();
+         }
+         // CraftBukkit end
       }
    }
 
@@ -168,7 +_,11 @@
       BlockPos blockpos = new BlockPos(p_37069_.m_82425_());
       this.m_9236_().m_8055_(blockpos).m_60682_(this.m_9236_(), blockpos, this);
       if (!this.m_9236_().m_5776_() && this.m_37086_()) {
-         this.m_37080_();
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.m_37080_();
+         }
+         // CraftBukkit end
       }
 
       super.m_8060_(p_37069_);
@@ -192,7 +_,9 @@
 
       if (f > 0.0F) {
          if (this.f_37024_ != null) {
+            CraftEventFactory.entityDamage = this; // CraftBukkit
             this.f_37024_.m_6469_(this.m_269291_().m_268994_(this, this.m_19749_()), 5.0F + (float)(listtag.size() * 2));
+            CraftEventFactory.entityDamage = null; // CraftBukkit
          }
 
          double d0 = 5.0D;
@@ -213,7 +_,9 @@
 
                if (flag) {
                   float f1 = f * (float)Math.sqrt((5.0D - (double)this.m_20270_(livingentity)) / 5.0D);
+                  CraftEventFactory.entityDamage = this; // CraftBukkit
                   livingentity.m_6469_(this.m_269291_().m_268994_(this, this.m_19749_()), f1);
+                  CraftEventFactory.entityDamage = null; // CraftBukkit
                }
             }
          }
