--- a/net/minecraft/world/entity/projectile/Fireball.java
+++ b/net/minecraft/world/entity/projectile/Fireball.java
@@ -60,6 +_,6 @@
    public void m_7378_(CompoundTag p_37009_) {
       super.m_7378_(p_37009_);
       ItemStack itemstack = ItemStack.m_41712_(p_37009_.m_128469_("Item"));
-      this.m_37010_(itemstack);
+      if (!itemstack.m_41619_()) this.m_37010_(itemstack); // CraftBukkit - SPIGOT-5474 probably came from bugged earlier versions
    }
 }
