--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -45,13 +_,18 @@
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import org.bukkit.entity.FishHook;
+import org.bukkit.event.player.PlayerFishEvent;
+// CraftBukkit end
+
 public class FishingHook extends Projectile {
    private static final Logger f_201977_ = LogUtils.getLogger();
    private final RandomSource f_37098_ = RandomSource.m_216327_();
    private boolean f_37099_;
    private int f_37100_;
    private static final int f_150136_ = 10;
-   private static final EntityDataAccessor<Integer> f_37101_ = SynchedEntityData.m_135353_(FishingHook.class, EntityDataSerializers.f_135028_);
+   public static final EntityDataAccessor<Integer> f_37101_ = SynchedEntityData.m_135353_(FishingHook.class, EntityDataSerializers.f_135028_); // CatServer - private -> public
    private static final EntityDataAccessor<Boolean> f_37102_ = SynchedEntityData.m_135353_(FishingHook.class, EntityDataSerializers.f_135035_);
    private int f_37103_;
    private int f_37089_;
@@ -60,11 +_,23 @@
    private float f_37092_;
    private boolean f_37093_ = true;
    @Nullable
-   private Entity f_37094_;
-   private FishingHook.FishHookState f_37095_ = FishingHook.FishHookState.FLYING;
+   public Entity f_37094_; // CatServer - private -> public
+   public FishingHook.FishHookState f_37095_ = FishingHook.FishHookState.FLYING; // CatServer - private -> public
    private final int f_37096_;
    private final int f_37097_;
 
+   // CraftBukkit start - Extra variables to enable modification of fishing wait time, values are minecraft defaults
+   public int minWaitTime = 100;
+   public int maxWaitTime = 600;
+   public int minLureTime = 20;
+   public int maxLureTime = 80;
+   public float minLureAngle = 0.0F;
+   public float maxLureAngle = 360.0F;
+   public boolean applyLure = true;
+   public boolean rainInfluenced = true;
+   public boolean skyInfluenced = true;
+   // CraftBukkit end
+
    private FishingHook(EntityType<? extends FishingHook> p_150141_, Level p_150142_, int p_150143_, int p_150144_) {
       super(p_150141_, p_150142_);
       this.f_19811_ = true;
@@ -229,8 +_,8 @@
    private boolean m_37136_(Player p_37137_) {
       ItemStack itemstack = p_37137_.m_21205_();
       ItemStack itemstack1 = p_37137_.m_21206_();
-      boolean flag = itemstack.m_150930_(Items.f_42523_);
-      boolean flag1 = itemstack1.m_150930_(Items.f_42523_);
+      boolean flag = itemstack.canPerformAction(net.minecraftforge.common.ToolActions.FISHING_ROD_CAST);
+      boolean flag1 = itemstack1.canPerformAction(net.minecraftforge.common.ToolActions.FISHING_ROD_CAST);
       if (!p_37137_.m_213877_() && p_37137_.m_6084_() && (flag || flag1) && !(this.m_20280_(p_37137_) > 1024.0D)) {
          return false;
       } else {
@@ -241,7 +_,7 @@
 
    private void m_37171_() {
       HitResult hitresult = ProjectileUtil.m_278158_(this, this::m_5603_);
-      this.m_6532_(hitresult);
+      if (hitresult.m_6662_() == HitResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) this.preOnHit(hitresult); // CraftBukkit - projectile hit event
    }
 
    protected boolean m_5603_(Entity p_37135_) {
@@ -270,11 +_,11 @@
       ServerLevel serverlevel = (ServerLevel)this.m_9236_();
       int i = 1;
       BlockPos blockpos = p_37146_.m_7494_();
-      if (this.f_19796_.m_188501_() < 0.25F && this.m_9236_().m_46758_(blockpos)) {
+      if (this.rainInfluenced && this.f_19796_.m_188501_() < 0.25F && this.m_9236_().m_46758_(blockpos)) { // CraftBukkit
          ++i;
       }
 
-      if (this.f_19796_.m_188501_() < 0.5F && !this.m_9236_().m_45527_(blockpos)) {
+      if (this.skyInfluenced && this.f_19796_.m_188501_() < 0.5F && !this.m_9236_().m_45527_(blockpos)) { // CraftBukkit
          --i;
       }
 
@@ -284,6 +_,10 @@
             this.f_37090_ = 0;
             this.f_37091_ = 0;
             this.m_20088_().m_135381_(f_37102_, false);
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.m_37168_().getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.FAILED_ATTEMPT);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            // CraftBukkit end
          }
       } else if (this.f_37091_ > 0) {
          this.f_37091_ -= i;
@@ -307,6 +_,13 @@
                serverlevel.m_8767_(ParticleTypes.f_123816_, d0, d1, d2, 0, (double)(-f4), 0.01D, (double)f3, 1.0D);
             }
          } else {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) this.m_37168_().getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.BITE);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            if (playerFishEvent.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             this.m_5496_(SoundEvents.f_11940_, 0.25F, 1.0F + (this.f_19796_.m_188501_() - this.f_19796_.m_188501_()) * 0.4F);
             double d3 = this.m_20186_() + 0.5D;
             serverlevel.m_8767_(ParticleTypes.f_123795_, this.m_20185_(), d3, this.m_20189_(), (int)(1.0F + this.m_20205_() * 20.0F), (double)this.m_20205_(), 0.0D, (double)this.m_20205_(), (double)0.2F);
@@ -338,12 +_,16 @@
          }
 
          if (this.f_37090_ <= 0) {
-            this.f_37092_ = Mth.m_216267_(this.f_19796_, 0.0F, 360.0F);
-            this.f_37091_ = Mth.m_216271_(this.f_19796_, 20, 80);
+            // CraftBukkit start - logic to modify fishing wait time, lure time, and lure angle
+            this.f_37092_ = Mth.m_216267_(this.f_19796_, this.minLureAngle, this.maxLureAngle);
+            this.f_37091_ = Mth.m_216271_(this.f_19796_, this.minLureTime, this.maxLureTime);
+            // CraftBukkit end
          }
       } else {
-         this.f_37090_ = Mth.m_216271_(this.f_19796_, 100, 600);
-         this.f_37090_ -= this.f_37097_ * 20 * 5;
+         // CraftBukkit start - logic to modify fishing wait time
+         this.f_37090_ = Mth.m_216271_(this.f_19796_, this.minWaitTime, this.maxWaitTime);
+         this.f_37090_ -= (this.applyLure) ? this.f_37097_ * 20 * 5 : 0;
+         // CraftBukkit end
       }
 
    }
@@ -403,26 +_,52 @@
       Player player = this.m_37168_();
       if (!this.m_9236_().f_46443_ && player != null && !this.m_37136_(player)) {
          int i = 0;
+         net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
          if (this.f_37094_ != null) {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) player.getBukkitEntity(), this.f_37094_.getBukkitEntity(), (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.CAUGHT_ENTITY);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            if (playerFishEvent.isCancelled()) {
+               return 0;
+            }
+            // CraftBukkit end
             this.m_150155_(this.f_37094_);
             CriteriaTriggers.f_10553_.m_40416_((ServerPlayer)player, p_37157_, this, Collections.emptyList());
             this.m_9236_().m_7605_(this, (byte)31);
             i = this.f_37094_ instanceof ItemEntity ? 3 : 5;
          } else if (this.f_37089_ > 0) {
-            LootParams lootparams = (new LootParams.Builder((ServerLevel)this.m_9236_())).m_287286_(LootContextParams.f_81460_, this.m_20182_()).m_287286_(LootContextParams.f_81463_, p_37157_).m_287286_(LootContextParams.f_81455_, this).m_287239_((float)this.f_37096_ + player.m_36336_()).m_287235_(LootContextParamSets.f_81414_);
+            LootParams lootparams = (new LootParams.Builder((ServerLevel)this.m_9236_())).m_287286_(LootContextParams.f_81460_, this.m_20182_()).m_287286_(LootContextParams.f_81463_, p_37157_).m_287286_(LootContextParams.f_81455_, this).m_287286_(LootContextParams.f_81458_, this.m_19749_()).m_287286_(LootContextParams.f_81455_, this).m_287239_((float)this.f_37096_ + player.m_36336_()).m_287235_(LootContextParamSets.f_81414_);
             LootTable loottable = this.m_9236_().m_7654_().m_278653_().m_278676_(BuiltInLootTables.f_78720_);
             List<ItemStack> list = loottable.m_287195_(lootparams);
+            event = new net.minecraftforge.event.entity.player.ItemFishedEvent(list, this.m_20096_() ? 2 : 1, this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            if (event.isCanceled()) {
+               this.m_146870_();
+               return event.getRodDamage();
+            }
             CriteriaTriggers.f_10553_.m_40416_((ServerPlayer)player, p_37157_, this, list);
 
             for(ItemStack itemstack : list) {
                ItemEntity itementity = new ItemEntity(this.m_9236_(), this.m_20185_(), this.m_20186_(), this.m_20189_(), itemstack);
+               // CraftBukkit start
+               PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) player.getBukkitEntity(), itementity.getBukkitEntity(), (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.CAUGHT_FISH);
+               playerFishEvent.setExpToDrop(this.f_19796_.m_188503_(6) + 1);
+               this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+               if (playerFishEvent.isCancelled()) {
+                  return 0;
+               }
+               // CraftBukkit end
                double d0 = player.m_20185_() - this.m_20185_();
                double d1 = player.m_20186_() - this.m_20186_();
                double d2 = player.m_20189_() - this.m_20189_();
                double d3 = 0.1D;
                itementity.m_20334_(d0 * 0.1D, d1 * 0.1D + Math.sqrt(Math.sqrt(d0 * d0 + d1 * d1 + d2 * d2)) * 0.08D, d2 * 0.1D);
                this.m_9236_().m_7967_(itementity);
-               player.m_9236_().m_7967_(new ExperienceOrb(player.m_9236_(), player.m_20185_(), player.m_20186_() + 0.5D, player.m_20189_() + 0.5D, this.f_19796_.m_188503_(6) + 1));
+               // CraftBukkit start - this.random.nextInt(6) + 1 -> playerFishEvent.getExpToDrop()
+               if (playerFishEvent.getExpToDrop() > 0) {
+                  player.m_9236_().m_7967_(new ExperienceOrb(player.m_9236_(), player.m_20185_(), player.m_20186_() + 0.5D, player.m_20189_() + 0.5D, playerFishEvent.getExpToDrop()));
+               }
+               // CraftBukkit end
                if (itemstack.m_204117_(ItemTags.f_13156_)) {
                   player.m_36222_(Stats.f_12939_, 1);
                }
@@ -431,12 +_,28 @@
             i = 1;
          }
 
-         if (this.m_20096_()) {
+         if (this.f_19861_) {
+            // CraftBukkit start
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) player.getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.IN_GROUND);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            if (playerFishEvent.isCancelled()) {
+               return 0;
+            }
+            // CraftBukkit end
             i = 2;
          }
+         // CraftBukkit start
+         if (i == 0) {
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) player.getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.REEL_IN);
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(playerFishEvent);
+            if (playerFishEvent.isCancelled()) {
+               return 0;
+            }
+         }
+         // CraftBukkit end
 
          this.m_146870_();
-         return i;
+         return event == null ? i : event.getRodDamage();
       } else {
          return 0;
       }
@@ -450,7 +_,7 @@
       super.m_7822_(p_37123_);
    }
 
-   protected void m_150155_(Entity p_150156_) {
+   public void m_150155_(Entity p_150156_) { // CatServer - protected -> public
       Entity entity = this.m_19749_();
       if (entity != null) {
          Vec3 vec3 = (new Vec3(entity.m_20185_() - this.m_20185_(), entity.m_20186_() - this.m_20186_(), entity.m_20189_() - this.m_20189_())).m_82490_(0.1D);
@@ -514,7 +_,7 @@
 
    }
 
-   static enum FishHookState {
+   public static enum FishHookState { // CatServer - make public
       FLYING,
       HOOKED_IN_ENTITY,
       BOBBING;
