--- a/net/minecraft/world/entity/projectile/LargeFireball.java
+++ b/net/minecraft/world/entity/projectile/LargeFireball.java
@@ -9,23 +_,35 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 
+import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
+
 public class LargeFireball extends Fireball {
-   private int f_37197_ = 1;
+   public int f_37197_ = 1; // CatServer - private -> public
 
    public LargeFireball(EntityType<? extends LargeFireball> p_37199_, Level p_37200_) {
       super(p_37199_, p_37200_);
+      isIncendiary = this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_); // CraftBukkit
    }
 
    public LargeFireball(Level p_181151_, LivingEntity p_181152_, double p_181153_, double p_181154_, double p_181155_, int p_181156_) {
       super(EntityType.f_20463_, p_181152_, p_181153_, p_181154_, p_181155_, p_181151_);
       this.f_37197_ = p_181156_;
+      isIncendiary = this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_); // CraftBukkit
    }
 
    protected void m_6532_(HitResult p_37218_) {
       super.m_6532_(p_37218_);
       if (!this.m_9236_().f_46443_) {
-         boolean flag = this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_);
-         this.m_9236_().m_255391_(this, this.m_20185_(), this.m_20186_(), this.m_20189_(), (float)this.f_37197_, flag, Level.ExplosionInteraction.MOB);
+         // TODO 1.19.3: The creation of Level.ExplosionInteraction means this code path will fire EntityMobGriefingEvent twice. Should we try and fix it? -SS
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this.m_19749_());
+         // CraftBukkit start - fire ExplosionPrimeEvent
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            // give 'this' instead of (Entity) null so we know what causes the damage
+            this.m_9236_().m_255391_(this, this.m_20185_(), this.m_20186_(), this.m_20189_(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB);
+         }
+         // CraftBukkit end
          this.m_146870_();
       }
 
@@ -52,7 +_,8 @@
    public void m_7378_(CompoundTag p_37220_) {
       super.m_7378_(p_37220_);
       if (p_37220_.m_128425_("ExplosionPower", 99)) {
-         this.f_37197_ = p_37220_.m_128445_("ExplosionPower");
+         // CraftBukkit - set bukkitYield when setting explosionpower
+         this.bukkitYield = this.f_37197_ = p_37220_.m_128445_("ExplosionPower");
       }
 
    }
