--- a/net/minecraft/world/entity/projectile/LargeFireball.java
+++ b/net/minecraft/world/entity/projectile/LargeFireball.java
@@ -10,24 +_,35 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class LargeFireball extends Fireball {
    public int f_37197_ = 1;
 
    public LargeFireball(EntityType<? extends LargeFireball> p_37199_, Level p_37200_) {
       super(p_37199_, p_37200_);
+      isIncendiary = this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_); // CraftBukkit
    }
 
    public LargeFireball(Level p_181151_, LivingEntity p_181152_, double p_181153_, double p_181154_, double p_181155_, int p_181156_) {
       super(EntityType.f_20463_, p_181152_, p_181153_, p_181154_, p_181155_, p_181151_);
       this.f_37197_ = p_181156_;
+      isIncendiary = this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_); // CraftBukkit
    }
 
    protected void m_6532_(HitResult p_37218_) {
       super.m_6532_(p_37218_);
       if (!this.f_19853_.f_46443_) {
-         boolean flag = this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_);
-         this.f_19853_.m_46518_((Entity)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), (float)this.f_37197_, flag, flag ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE);
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_19853_, this.m_37282_());
+         // CraftBukkit start - fire ExplosionPrimeEvent
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+         this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            // give 'this' instead of (Entity) null so we know what causes the damage
+            this.f_19853_.m_46518_(this, this.m_20185_(), this.m_20186_(), this.m_20189_(), event.getRadius(), event.getFire(), flag ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE);
+         }
+         // CraftBukkit end
          this.m_146870_();
       }
 
@@ -54,7 +_,7 @@
    public void m_7378_(CompoundTag p_37220_) {
       super.m_7378_(p_37220_);
       if (p_37220_.m_128425_("ExplosionPower", 99)) {
-         this.f_37197_ = p_37220_.m_128445_("ExplosionPower");
+         bukkitYield = this.f_37197_ = p_37220_.m_128445_("ExplosionPower");
       }
 
    }
