--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -57,7 +_,20 @@
       this.f_37312_ = p_37332_;
       this.f_37313_ = Direction.UP;
       this.m_37348_(p_37333_);
-   }
+      projectileSource = (org.bukkit.entity.LivingEntity) p_37331_.getBukkitEntity(); // CraftBukkit
+   }
+
+   // CraftBukkit start
+   public Entity getTarget() {
+      return this.f_37312_;
+   }
+
+   public void setTarget(Entity e) {
+      this.f_37312_ = e;
+      this.f_37313_ = Direction.UP;
+      this.m_37348_(Direction.Axis.X);
+   }
+   // CraftBukkit end
 
    public SoundSource m_5720_() {
       return SoundSource.HOSTILE;
@@ -211,7 +_,7 @@
          }
 
          HitResult hitresult = ProjectileUtil.m_37294_(this, this::m_5603_);
-         if (hitresult.m_6662_() != HitResult.Type.MISS) {
+         if (hitresult.m_6662_() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
             this.m_6532_(hitresult);
          }
       }
@@ -271,7 +_,7 @@
       if (flag) {
          this.m_19970_(livingentity, entity);
          if (entity instanceof LivingEntity) {
-            ((LivingEntity)entity).m_147207_(new MobEffectInstance(MobEffects.f_19620_, 200), MoreObjects.firstNonNull(entity1, this));
+            ((LivingEntity)entity).addEffect(new MobEffectInstance(MobEffects.f_19620_, 200), MoreObjects.firstNonNull(entity1, this), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
          }
       }
 
@@ -293,6 +_,11 @@
    }
 
    public boolean m_6469_(DamageSource p_37338_, float p_37339_) {
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_37338_, p_37339_, false)) {
+         return false;
+      }
+      // CraftBukkit end
       if (!this.f_19853_.f_46443_) {
          this.m_5496_(SoundEvents.f_12411_, 1.0F, 1.0F);
          ((ServerLevel)this.f_19853_).m_8767_(ParticleTypes.f_123797_, this.m_20185_(), this.m_20186_(), this.m_20189_(), 15, 0.2D, 0.2D, 0.2D, 0.0D);
