--- a/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -13,6 +_,8 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 
+import org.bukkit.event.entity.EntityCombustByEntityEvent; // CraftBukkit
+
 public class SmallFireball extends Fireball {
    public SmallFireball(EntityType<? extends SmallFireball> p_37364_, Level p_37365_) {
       super(p_37364_, p_37365_);
@@ -20,6 +_,11 @@
 
    public SmallFireball(Level p_37375_, LivingEntity p_37376_, double p_37377_, double p_37378_, double p_37379_) {
       super(EntityType.f_20527_, p_37376_, p_37377_, p_37378_, p_37379_, p_37375_);
+      // CraftBukkit start
+      if (this.m_19749_() != null && this.m_19749_() instanceof Mob) {
+         this.isIncendiary = this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_);
+      }
+      // CraftBukkit end
    }
 
    public SmallFireball(Level p_37367_, double p_37368_, double p_37369_, double p_37370_, double p_37371_, double p_37372_, double p_37373_) {
@@ -32,7 +_,13 @@
          Entity entity = p_37386_.m_82443_();
          Entity entity1 = this.m_19749_();
          int i = entity.m_20094_();
-         entity.m_20254_(5);
+         // CraftBukkit start - Entity damage by entity event + combust event
+         EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+         entity.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            entity.setSecondsOnFire(event.getDuration(), false);
+         }
+         // CraftBukkit end
          if (!entity.m_6469_(this.m_269291_().m_269453_(this, entity1), 5.0F)) {
             entity.m_7311_(i);
          } else if (entity1 instanceof LivingEntity) {
@@ -46,9 +_,9 @@
       super.m_8060_(p_37384_);
       if (!this.m_9236_().f_46443_) {
          Entity entity = this.m_19749_();
-         if (!(entity instanceof Mob) || this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+         if (!(entity instanceof Mob) || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), entity)) { // CraftBukkit // CatServer - recover
             BlockPos blockpos = p_37384_.m_82425_().m_121945_(p_37384_.m_82434_());
-            if (this.m_9236_().m_46859_(blockpos)) {
+            if (this.m_9236_().m_46859_(blockpos) && !org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callBlockIgniteEvent(m_9236_(), blockpos, this).isCancelled()) { // CraftBukkit
                this.m_9236_().m_46597_(blockpos, BaseFireBlock.m_49245_(this.m_9236_(), blockpos));
             }
          }
