--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -40,7 +_,7 @@
    protected void m_7761_(LivingEntity p_37422_) {
       super.m_7761_(p_37422_);
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.f_19619_, this.f_37409_, 0);
-      p_37422_.m_147207_(mobeffectinstance, this.m_150173_());
+      p_37422_.addEffect(mobeffectinstance, this.m_150173_(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
    }
 
    public void m_7378_(CompoundTag p_37424_) {
