--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -2,7 +_,9 @@
 
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.animal.Chicken;
@@ -11,6 +_,8 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.entity.Ageable;
+import org.bukkit.event.player.PlayerEggThrowEvent;
 
 public class ThrownEgg extends ThrowableItemProjectile {
    public ThrownEgg(EntityType<? extends ThrownEgg> p_37473_, Level p_37474_) {
@@ -44,18 +_,38 @@
    protected void m_6532_(HitResult p_37488_) {
       super.m_6532_(p_37488_);
       if (!this.f_19853_.f_46443_) {
-         if (this.f_19796_.nextInt(8) == 0) {
+         boolean hatching = this.f_19796_.nextInt(8) == 0; // CraftBukkit
+         if (true) {
             int i = 1;
             if (this.f_19796_.nextInt(32) == 0) {
                i = 4;
             }
 
-            for(int j = 0; j < i; ++j) {
-               Chicken chicken = EntityType.f_20555_.m_20615_(this.f_19853_);
-               chicken.m_146762_(-24000);
-               chicken.m_7678_(this.m_20185_(), this.m_20186_(), this.m_20189_(), this.m_146908_(), 0.0F);
-               this.f_19853_.m_7967_(chicken);
-            }
+            // CraftBukkit start
+            if (!hatching) {
+               i = 0;
+            }
+            org.bukkit.entity.EntityType hatchingType = org.bukkit.entity.EntityType.CHICKEN;
+            Entity shooter = this.m_37282_();
+            if (shooter instanceof ServerPlayer) {
+               PlayerEggThrowEvent event = new PlayerEggThrowEvent((org.bukkit.entity.Player) shooter.getBukkitEntity(), (org.bukkit.entity.Egg) this.getBukkitEntity(), hatching, (byte) i, hatchingType);
+               this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+               i = event.getNumHatches();
+               hatching = event.isHatching();
+               hatchingType = event.getHatchingType();
+            }
+
+            if (hatching) {
+               for (int j = 0; j < i; ++j) {
+                  Entity entity = f_19853_.getWorld().createEntity(new org.bukkit.Location(f_19853_.getWorld(), this.m_20185_(), this.m_20186_(), this.m_20189_(), this.m_146908_(), 0.0F), hatchingType.getEntityClass());
+                  if (entity.getBukkitEntity() instanceof Ageable) {
+                     ((Ageable) entity.getBukkitEntity()).setBaby();
+                  }
+                  f_19853_.getWorld().addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG);
+               }
+            }
+            // CraftBukkit end
          }
 
          this.f_19853_.m_7605_(this, (byte)3);
