--- a/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
@@ -30,12 +_,19 @@
       if (!p_37447_.m_150930_(this.m_7881_()) || p_37447_.m_41782_()) {
          this.m_20088_().m_135381_(f_37429_, Util.m_137469_(p_37447_.m_41777_(), (p_37451_) -> {
             p_37451_.m_41764_(1);
+            if (!p_37447_.m_41619_()) p_37447_.m_41764_(1); // CraftBukkit
          }));
       }
 
    }
 
    protected abstract Item m_7881_();
+
+   // CraftBukkit start
+   public Item getDefaultItemPublic() {
+      return m_7881_();
+   }
+   // CraftBukkit end
 
    protected ItemStack m_37454_() {
       return this.m_20088_().m_135370_(f_37429_);
