--- a/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
+++ b/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
@@ -35,8 +_,16 @@
    protected void m_6532_(HitResult p_37521_) {
       super.m_6532_(p_37521_);
       if (this.f_19853_ instanceof ServerLevel) {
-         this.f_19853_.m_46796_(2002, this.m_142538_(), PotionUtils.m_43559_(Potions.f_43599_));
+         // CraftBukkit - moved to after event
+         //this.level.levelEvent(2002, this.blockPosition(), PotionUtils.getColor(Potions.WATER));
          int i = 3 + this.f_19853_.f_46441_.nextInt(5) + this.f_19853_.f_46441_.nextInt(5);
+         // CraftBukkit start
+         org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callExpBottleEvent(this, i);
+         i = event.getExperience();
+         if (event.getShowEffect()) {
+            this.f_19853_.m_46796_(2002, this.m_142538_(), PotionUtils.m_43559_(Potions.f_43599_));
+         }
+         // CraftBukkit end
          ExperienceOrb.m_147082_((ServerLevel)this.f_19853_, this.m_20182_(), i);
          this.m_146870_();
       }
