--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -58,8 +_,8 @@
          }
       }
 
-      if (hitresult.m_6662_() != HitResult.Type.MISS && !flag) {
-         this.m_6532_(hitresult);
+      if (hitresult.m_6662_() != HitResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
+         this.preOnHit(hitresult); // CraftBukkit - projectile hit event
       }
 
       this.m_20101_();
