--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -140,7 +_,7 @@
             LightningBolt lightningbolt = EntityType.f_20465_.m_20615_(this.f_19853_);
             lightningbolt.m_20219_(Vec3.m_82539_(blockpos));
             lightningbolt.m_20879_(entity1 instanceof ServerPlayer ? (ServerPlayer)entity1 : null);
-            this.f_19853_.m_7967_(lightningbolt);
+            ((ServerLevel) this.f_19853_).strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRIDENT); // CraftBukkit
             soundevent = SoundEvents.f_12521_;
             f1 = 5.0F;
          }
