--- a/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
@@ -32,9 +_,15 @@
 
    }
 
+   // CraftBukkit start
+   public Item getDefaultItemPublic() {
+      return m_7881_();
+   }
+   // CraftBukkit end
+
    protected abstract Item m_7881_();
 
-   protected ItemStack m_37454_() {
+   public ItemStack m_37454_() { // CatServer - protected -> public
       return this.m_20088_().m_135370_(f_37429_);
    }
 
