--- ../src-base/minecraft/net/minecraftforge/fml/common/versioning/ComparableVersion.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/versioning/ComparableVersion.java
@@ -56,7 +56,7 @@
  *
  * @see <a href="https://cwiki.apache.org/confluence/display/MAVENOLD/Versioning">"Versioning" on Maven Wiki</a>
  * @author <a href="mailto:kenney@apache.org">Kenney Westerhof</a>
- * @author <a href="mailto:hboutemy@apache.org">Hervé Boutemy</a>
+ * @author <a href="mailto:hboutemy@apache.org">Herv? Boutemy</a>
  */
 public class ComparableVersion
     implements Comparable<ComparableVersion>
