--- ../src-base/minecraft/net/minecraft/client/resources/Language.java
+++ ../src-work/minecraft/net/minecraft/client/resources/Language.java
@@ -17,6 +17,7 @@
         this.region = regionIn;
         this.name = nameIn;
         this.bidirectional = bidirectionalIn;
+        this.javaLocale = new java.util.Locale(languageCode, region);
     }
 
     public String getLanguageCode()
@@ -55,4 +56,8 @@
     {
         return this.languageCode.compareTo(p_compareTo_1_.languageCode);
     }
+
+    // Forge: add access to Locale so modders can create correct string and number formatters
+    private final java.util.Locale javaLocale;
+    public java.util.Locale getJavaLocale() { return javaLocale; }
 }
