--- ../src-base/minecraft/net/minecraft/client/resources/ResourcePackRepository.java
+++ ../src-work/minecraft/net/minecraft/client/resources/ResourcePackRepository.java
@@ -66,8 +66,8 @@
     private IResourcePack serverResourcePack;
     private final ReentrantLock lock = new ReentrantLock();
     private ListenableFuture<Object> downloadingPacks;
-    private List<ResourcePackRepository.Entry> repositoryEntriesAll = Lists.<ResourcePackRepository.Entry>newArrayList();
-    private final List<ResourcePackRepository.Entry> repositoryEntries = Lists.<ResourcePackRepository.Entry>newArrayList();
+    private List<Entry> repositoryEntriesAll = Lists.<Entry>newArrayList();
+    private final List<Entry> repositoryEntries = Lists.<Entry>newArrayList();
 
     public ResourcePackRepository(File dirResourcepacksIn, File dirServerResourcepacksIn, IResourcePack rprDefaultResourcePackIn, MetadataSerializer rprMetadataSerializerIn, GameSettings settings)
     {
@@ -83,7 +83,7 @@
         {
             String s = iterator.next();
 
-            for (ResourcePackRepository.Entry resourcepackrepository$entry : this.repositoryEntriesAll)
+            for (Entry resourcepackrepository$entry : this.repositoryEntriesAll)
             {
                 if (resourcepackrepository$entry.getResourcePackName().equals(s))
                 {
@@ -161,11 +161,11 @@
 
     public void updateRepositoryEntriesAll()
     {
-        List<ResourcePackRepository.Entry> list = Lists.<ResourcePackRepository.Entry>newArrayList();
+        List<Entry> list = Lists.<Entry>newArrayList();
 
         for (File file1 : this.getResourcePackFiles())
         {
-            ResourcePackRepository.Entry resourcepackrepository$entry = new ResourcePackRepository.Entry(file1);
+            Entry resourcepackrepository$entry = new Entry(file1);
 
             if (this.repositoryEntriesAll.contains(resourcepackrepository$entry))
             {
@@ -192,7 +192,7 @@
 
         this.repositoryEntriesAll.removeAll(list);
 
-        for (ResourcePackRepository.Entry resourcepackrepository$entry1 : this.repositoryEntriesAll)
+        for (Entry resourcepackrepository$entry1 : this.repositoryEntriesAll)
         {
             resourcepackrepository$entry1.closeResourcePack();
         }
@@ -205,7 +205,7 @@
     {
         if (this.serverResourcePack != null)
         {
-            ResourcePackRepository.Entry resourcepackrepository$entry = new ResourcePackRepository.Entry(this.serverResourcePack);
+            Entry resourcepackrepository$entry = new Entry(this.serverResourcePack);
 
             try
             {
@@ -221,17 +221,17 @@
         return null;
     }
 
-    public List<ResourcePackRepository.Entry> getRepositoryEntriesAll()
+    public List<Entry> getRepositoryEntriesAll()
     {
         return ImmutableList.copyOf(this.repositoryEntriesAll);
     }
 
-    public List<ResourcePackRepository.Entry> getRepositoryEntries()
+    public List<Entry> getRepositoryEntries()
     {
         return ImmutableList.copyOf(this.repositoryEntries);
     }
 
-    public void setRepositories(List<ResourcePackRepository.Entry> repositories)
+    public void setRepositories(List<Entry> repositories)
     {
         this.repositoryEntries.clear();
         this.repositoryEntries.addAll(repositories);
@@ -338,7 +338,7 @@
 
     private boolean validatePack(File p_190112_1_)
     {
-        ResourcePackRepository.Entry resourcepackrepository$entry = new ResourcePackRepository.Entry(p_190112_1_);
+        Entry resourcepackrepository$entry = new Entry(p_190112_1_);
 
         try
         {
@@ -511,7 +511,7 @@
             }
             else
             {
-                return p_equals_1_ instanceof ResourcePackRepository.Entry ? this.toString().equals(p_equals_1_.toString()) : false;
+                return p_equals_1_ instanceof Entry ? this.toString().equals(p_equals_1_.toString()) : false;
             }
         }
 
