--- ../src-base/minecraft/net/minecraft/client/resources/SimpleReloadableResourceManager.java
+++ ../src-work/minecraft/net/minecraft/client/resources/SimpleReloadableResourceManager.java
@@ -92,6 +92,7 @@
 
     public void reloadResources(List<IResourcePack> resourcesPacksList)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar resReload = net.minecraftforge.fml.common.ProgressManager.push("Loading Resources", resourcesPacksList.size()+1, true);
         this.clearResources();
         LOGGER.info("Reloading ResourceManager: {}", (Object)JOINER_RESOURCE_PACKS.join(Iterables.transform(resourcesPacksList, new Function<IResourcePack, String>()
         {
@@ -103,23 +104,32 @@
 
         for (IResourcePack iresourcepack : resourcesPacksList)
         {
+            resReload.step(iresourcepack.getPackName());
             this.reloadResourcePack(iresourcepack);
         }
 
+        resReload.step("Reloading listeners");
         this.notifyReloadListeners();
+        net.minecraftforge.fml.common.ProgressManager.pop(resReload);
     }
 
     public void registerReloadListener(IResourceManagerReloadListener reloadListener)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar resReload = net.minecraftforge.fml.common.ProgressManager.push("Loading Resource", 1);
+        resReload.step(reloadListener.getClass());
         this.reloadListeners.add(reloadListener);
         reloadListener.onResourceManagerReload(this);
+        net.minecraftforge.fml.common.ProgressManager.pop(resReload);
     }
 
     private void notifyReloadListeners()
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar resReload = net.minecraftforge.fml.common.ProgressManager.push("Reloading", this.reloadListeners.size());
         for (IResourceManagerReloadListener iresourcemanagerreloadlistener : this.reloadListeners)
         {
+            resReload.step(iresourcemanagerreloadlistener.getClass());
             iresourcemanagerreloadlistener.onResourceManagerReload(this);
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(resReload);
     }
 }
