--- ../src-base/minecraft/net/minecraft/client/resources/SkinManager.java
+++ ../src-work/minecraft/net/minecraft/client/resources/SkinManager.java
@@ -59,7 +59,7 @@
 
     public ResourceLocation loadSkin(MinecraftProfileTexture profileTexture, Type textureType)
     {
-        return this.loadSkin(profileTexture, textureType, (SkinManager.SkinAvailableCallback)null);
+        return this.loadSkin(profileTexture, textureType, (SkinAvailableCallback)null);
     }
 
     public ResourceLocation loadSkin(final MinecraftProfileTexture profileTexture, final Type textureType, @Nullable final SkinManager.SkinAvailableCallback skinAvailableCallback)
@@ -109,7 +109,7 @@
         return resourcelocation;
     }
 
-    public void loadProfileTextures(final GameProfile profile, final SkinManager.SkinAvailableCallback skinAvailableCallback, final boolean requireSecure)
+    public void loadProfileTextures(final GameProfile profile, final SkinAvailableCallback skinAvailableCallback, final boolean requireSecure)
     {
         THREAD_POOL.submit(new Runnable()
         {
