--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -13,6 +_,14 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 
+// CraftBukkit start
+import net.minecraft.server.level.ServerLevel;
+import org.bukkit.craftbukkit.v1_20_R1.CraftWorld;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.block.data.CraftBlockData;
+import org.bukkit.event.block.BlockPhysicsEvent;
+// CraftBukkit end
+
 public interface NeighborUpdater {
    Direction[] f_230761_ = new Direction[]{Direction.WEST, Direction.EAST, Direction.DOWN, Direction.UP, Direction.NORTH, Direction.SOUTH};
 
@@ -39,7 +_,22 @@
 
    static void m_230763_(Level p_230764_, BlockState p_230765_, BlockPos p_230766_, Block p_230767_, BlockPos p_230768_, boolean p_230769_) {
       try {
+         // CraftBukkit start - 1.20.1
+         CraftWorld cworld = ((ServerLevel) p_230764_).getWorld();
+         if (cworld != null) {
+            BlockPhysicsEvent event = new BlockPhysicsEvent(CraftBlock.at(p_230764_, p_230766_), CraftBlockData.fromData(p_230765_), CraftBlock.at(p_230764_, p_230768_));
+            ((ServerLevel) p_230764_).getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_230765_.m_60690_(p_230764_, p_230766_, p_230767_, p_230768_, p_230769_);
+         // Spigot Start - 1.20.1
+      } catch (StackOverflowError ex) {
+         p_230764_.lastPhysicsProblem = new BlockPos(p_230766_);
+         // Spigot End
       } catch (Throwable throwable) {
          CrashReport crashreport = CrashReport.m_127521_(throwable, "Exception while updating neighbours");
          CrashReportCategory crashreportcategory = crashreport.m_127514_("Block being updated");
