--- ../src-base/minecraft/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -11,13 +11,13 @@
 @SideOnly(Side.CLIENT)
 public class DebugRenderer
 {
-    public final DebugRenderer.IDebugRenderer pathfinding;
-    public final DebugRenderer.IDebugRenderer water;
-    public final DebugRenderer.IDebugRenderer chunkBorder;
-    public final DebugRenderer.IDebugRenderer heightMap;
-    public final DebugRenderer.IDebugRenderer collisionBox;
-    public final DebugRenderer.IDebugRenderer neighborsUpdate;
-    public final DebugRenderer.IDebugRenderer solidFace;
+    public final IDebugRenderer pathfinding;
+    public final IDebugRenderer water;
+    public final IDebugRenderer chunkBorder;
+    public final IDebugRenderer heightMap;
+    public final IDebugRenderer collisionBox;
+    public final IDebugRenderer neighborsUpdate;
+    public final IDebugRenderer solidFace;
     private boolean chunkBorderEnabled;
     private boolean pathfindingEnabled;
     private boolean waterEnabled;
