--- ../src-base/minecraft/net/minecraft/entity/EntityTrackerEntry.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityTrackerEntry.java
@@ -64,6 +64,7 @@
 import net.minecraft.world.storage.MapData;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.spigotmc.AsyncCatcher;
 
 public class EntityTrackerEntry
 {
@@ -361,6 +362,7 @@
 
     public void updatePlayerEntity(EntityPlayerMP playerMP)
     {
+        AsyncCatcher.catchOp( "player tracker update"); // Spigot
         if (playerMP != this.trackedEntity)
         {
             if (this.isVisibleTo(playerMP))
@@ -635,6 +637,7 @@
 
     public void removeTrackedPlayerSymmetric(EntityPlayerMP playerMP)
     {
+        AsyncCatcher.catchOp( "player tracker clear"); // Spigot
         if (this.trackingPlayers.contains(playerMP))
         {
             this.trackingPlayers.remove(playerMP);
