--- ../src-base/minecraft/net/minecraft/entity/EntityHanging.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityHanging.java
@@ -18,6 +18,10 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import org.apache.commons.lang3.Validate;
+//CraftBukkit start
+import org.bukkit.entity.Hanging;
+import org.bukkit.event.hanging.HangingBreakEvent;
+// CraftBukkit end
 
 public abstract class EntityHanging extends Entity
 {
@@ -29,7 +33,7 @@
         }
     };
     private int tickCounter1;
-    protected BlockPos hangingPosition;
+    public BlockPos hangingPosition;
     @Nullable
     public EnumFacing facingDirection;
 
@@ -49,7 +53,7 @@
     {
     }
 
-    protected void updateFacingWithBoundingBox(EnumFacing facingDirectionIn)
+    public void updateFacingWithBoundingBox(EnumFacing facingDirectionIn)
     {
         Validate.notNull(facingDirectionIn);
         Validate.isTrue(facingDirectionIn.getAxis().isHorizontal());
@@ -61,7 +65,7 @@
 
     protected void updateBoundingBox()
     {
-        if (this.facingDirection != null)
+        /*if (this.facingDirection != null)
         {
             double d0 = (double)this.hangingPosition.getX() + 0.5D;
             double d1 = (double)this.hangingPosition.getY() + 0.5D;
@@ -95,10 +99,53 @@
             d7 = d7 / 32.0D;
             d8 = d8 / 32.0D;
             this.setEntityBoundingBox(new AxisAlignedBB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8));
+        }*/
+        if(this.facingDirection != null) {
+            this.setEntityBoundingBox(calculateBoundingBox(this, this.hangingPosition, this.facingDirection, getWidthPixels(), getHeightPixels()));
+        }
+    }
+    // CraftBukkit start - break out BB calc into own method
+    public static AxisAlignedBB calculateBoundingBox(Entity entity, BlockPos pos, EnumFacing direction, int width, int height) {
+        double d0 = (double)pos.getX() + 0.5D;
+        double d1 = (double)pos.getY() + 0.5D;
+        double d2 = (double)pos.getZ() + 0.5D;
+        double d3 = 0.46875D;
+        double d4 = offs(width);
+        double d5 = offs(height);
+        d0 = d0 - (double)direction.getFrontOffsetX() * 0.46875D;
+        d2 = d2 - (double)direction.getFrontOffsetZ() * 0.46875D;
+        d1 = d1 + d5;
+        EnumFacing enumfacing = direction.rotateYCCW();
+        d0 = d0 + d4 * (double)enumfacing.getFrontOffsetX();
+        d2 = d2 + d4 * (double)enumfacing.getFrontOffsetZ();
+        if(entity != null) {
+            entity.posX = d0;
+            entity.posY = d1;
+            entity.posZ = d2;
         }
+        //this.posX = d0;
+        //this.posY = d1;
+        //this.posZ = d2;
+        double d6 = (double)width;
+        double d7 = (double)height;
+        double d8 = (double)width;
+
+        if (direction.getAxis() == EnumFacing.Axis.Z)
+        {
+            d8 = 1.0D;
+        }
+        else
+        {
+            d6 = 1.0D;
+        }
+
+        d6 = d6 / 32.0D;
+        d7 = d7 / 32.0D;
+        d8 = d8 / 32.0D;
+        return new AxisAlignedBB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8);
     }
 
-    private double offs(int p_190202_1_)
+    private static double offs(int p_190202_1_)
     {
         return p_190202_1_ % 32 == 0 ? 0.5D : 0.0D;
     }
