--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -4,13 +_,25 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.protocol.Packet;
 
-public record ClientboundSystemChatPacket(Component f_237849_, boolean f_240374_) implements Packet<ClientGamePacketListener> {
+// Spigot start
+public record ClientboundSystemChatPacket(Component f_237849_, boolean f_240374_) implements Packet<ClientGamePacketListener> { // Spigot - Component -> String // CatServer - recover
+   // CatServer start - 1.20.1
+   public ClientboundSystemChatPacket(String serializedVanillaContent, boolean overlay) {
+      this(Component.Serializer.m_130701_(serializedVanillaContent), overlay);
+   }
+   // CatServer end
+
+   public ClientboundSystemChatPacket(net.md_5.bungee.api.chat.BaseComponent[] content, boolean overlay) {
+      this(net.md_5.bungee.chat.ComponentSerializer.toString(content), overlay); // CatServer - 1.20.1
+   }
+   // Spigot end
+
    public ClientboundSystemChatPacket(FriendlyByteBuf p_237852_) {
       this(p_237852_.m_130238_(), p_237852_.readBoolean());
    }
 
    public void m_5779_(FriendlyByteBuf p_237860_) {
-      p_237860_.m_130083_(this.f_237849_);
+      p_237860_.m_130083_(this.f_237849_); // Spigot // CatServer - recover
       p_237860_.writeBoolean(this.f_240374_);
    }
 
