--- a/net/minecraft/network/protocol/game/ClientboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundChatPacket.java
@@ -8,6 +_,7 @@
 
 public class ClientboundChatPacket implements Packet<ClientGamePacketListener> {
    private final Component f_131821_;
+   public net.md_5.bungee.api.chat.BaseComponent[] components; // Spigot
    private final ChatType f_131822_;
    private final UUID f_131823_;
 
@@ -24,7 +_,13 @@
    }
 
    public void m_5779_(FriendlyByteBuf p_131838_) {
-      p_131838_.m_130083_(this.f_131821_);
+      // Spigot start
+      if (components != null) {
+         p_131838_.m_130070_(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+      } else {
+         p_131838_.m_130083_(this.f_131821_);
+      }
+      // Spigot end
       p_131838_.writeByte(this.f_131822_.m_130610_());
       p_131838_.m_130077_(this.f_131823_);
    }
