--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -25,11 +_,19 @@
 
       for(short short1 : p_285027_) {
          this.f_132981_[j] = short1;
-         this.f_132982_[j] = p_285414_.m_62982_(SectionPos.m_123204_(short1), SectionPos.m_123220_(short1), SectionPos.m_123227_(short1));
+         this.f_132982_[j] = (p_285414_ != null) ? p_285414_.m_62982_(SectionPos.m_123204_(short1), SectionPos.m_123220_(short1), SectionPos.m_123227_(short1)) : net.minecraft.world.level.block.Blocks.f_50016_.m_49966_(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
          ++j;
       }
 
    }
+
+   // CraftBukkit start - 1.20.1 Add constructor
+   public ClientboundSectionBlocksUpdatePacket(SectionPos sectionposition, ShortSet shortset, BlockState[] states) {
+      this.f_132980_ = sectionposition;
+      this.f_132981_ = shortset.toShortArray();
+      this.f_132982_ = states;
+   }
+   // CraftBukkit end
 
    public ClientboundSectionBlocksUpdatePacket(FriendlyByteBuf p_179196_) {
       this.f_132980_ = SectionPos.m_123184_(p_179196_.readLong());
