--- a/net/minecraft/network/protocol/game/ClientboundSetDefaultSpawnPositionPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetDefaultSpawnPositionPacket.java
@@ -5,7 +_,7 @@
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundSetDefaultSpawnPositionPacket implements Packet<ClientGamePacketListener> {
-   private final BlockPos f_133111_;
+   public final BlockPos f_133111_;
    private final float f_133112_;
 
    public ClientboundSetDefaultSpawnPositionPacket(BlockPos p_133115_, float p_133116_) {
