--- a/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java
@@ -6,15 +_,18 @@
 public class ClientboundForgetLevelChunkPacket implements Packet<ClientGamePacketListener> {
    private final int f_132137_;
    private final int f_132138_;
+   private final net.minecraft.world.level.ChunkPos chunkPos; // CatServer - [1.20.1:compat:plugin:QuickShop-Hikari:6.2.0.10] ClientboundForgetLevelChunkPacket in higher version is a record with a ChunkPos field, and QuickShop-Hikari uses it.
 
    public ClientboundForgetLevelChunkPacket(int p_132141_, int p_132142_) {
       this.f_132137_ = p_132141_;
       this.f_132138_ = p_132142_;
+      this.chunkPos = new net.minecraft.world.level.ChunkPos(this.f_132137_, this.f_132138_); // CatServer - 1.20.1
    }
 
    public ClientboundForgetLevelChunkPacket(FriendlyByteBuf p_178858_) {
       this.f_132137_ = p_178858_.readInt();
       this.f_132138_ = p_178858_.readInt();
+      this.chunkPos = new net.minecraft.world.level.ChunkPos(this.f_132137_, this.f_132138_); // CatServer - 1.20.1
    }
 
    public void m_5779_(FriendlyByteBuf p_132151_) {
