--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -26,8 +_,10 @@
    }
 
    public ClientboundInitializeBorderPacket(WorldBorder p_178877_) {
-      this.f_178868_ = p_178877_.m_6347_();
-      this.f_178869_ = p_178877_.m_6345_();
+      // CraftBukkit start - multiply out nether border
+      this.f_178868_ = p_178877_.m_6347_() * p_178877_.world.m_6042_().f_63859_();
+      this.f_178869_ = p_178877_.m_6345_() * p_178877_.world.m_6042_().f_63859_();
+      // CraftBukkit end
       this.f_178870_ = p_178877_.m_61959_();
       this.f_178871_ = p_178877_.m_61961_();
       this.f_178872_ = p_178877_.m_61960_();
