--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -27,7 +_,7 @@
 
       for(short short1 : p_132987_) {
          this.f_132981_[j] = short1;
-         this.f_132982_[j] = p_132988_.m_62982_(SectionPos.m_123204_(short1), SectionPos.m_123220_(short1), SectionPos.m_123227_(short1));
+         this.f_132982_[j] = (p_132988_ != null) ? p_132988_.m_62982_(SectionPos.m_123204_(short1), SectionPos.m_123220_(short1), SectionPos.m_123227_(short1)) : net.minecraft.world.level.block.Blocks.f_50016_.m_49966_(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
          ++j;
       }
 
