--- a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -6,12 +_,14 @@
 
 public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener> {
    private final InteractionHand f_134707_;
+   public long timestamp; // Spigot
 
    public ServerboundUseItemPacket(InteractionHand p_134710_) {
       this.f_134707_ = p_134710_;
    }
 
    public ServerboundUseItemPacket(FriendlyByteBuf p_179798_) {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.f_134707_ = p_179798_.m_130066_(InteractionHand.class);
    }
 
