--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -16,14 +_,28 @@
    }
 
    public ServerboundChatPacket(FriendlyByteBuf p_179545_) {
-      this.f_133827_ = p_179545_.m_130136_(256);
+      this.f_133827_ = org.apache.commons.lang3.StringUtils.normalizeSpace(p_179545_.m_130136_(256)); // CraftBukkit - see PlayerConnection
    }
 
    public void m_5779_(FriendlyByteBuf p_133839_) {
       p_133839_.m_130070_(this.f_133827_);
    }
 
-   public void m_5797_(ServerGamePacketListener p_133836_) {
+   // Spigot Start
+   private static final java.util.concurrent.ExecutorService executors = java.util.concurrent.Executors.newCachedThreadPool(
+           new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon( true ).setNameFormat( "Async Chat Thread - #%d" ).build() );
+
+   public void m_5797_(final ServerGamePacketListener p_133836_) {
+      if (!f_133827_.startsWith("/")) {
+         executors.submit(new Runnable() {
+            @Override
+            public void run() {
+               p_133836_.m_7388_(ServerboundChatPacket.this);
+            }
+         });
+         return;
+      }
+      // Spigot end
       p_133836_.m_7388_(this);
    }
 
