--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -20,7 +_,15 @@
       this.f_241662_.m_242008_(p_133839_);
    }
 
+   // CatServer start
+   private static final java.util.concurrent.ExecutorService executors = java.util.concurrent.Executors.newCachedThreadPool(new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").build());
+
    public void m_5797_(ServerGamePacketListener p_133836_) {
+      if (!f_133827_.startsWith("/")) {
+         executors.submit(() -> p_133836_.m_7388_(ServerboundChatPacket.this));
+         return;
+      }
+      // CatServer end
       p_133836_.m_7388_(this);
    }
 }
