--- a/net/minecraft/network/protocol/game/ServerboundMovePlayerPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundMovePlayerPacket.java
@@ -4,14 +_,14 @@
 import net.minecraft.network.protocol.Packet;
 
 public abstract class ServerboundMovePlayerPacket implements Packet<ServerGamePacketListener> {
-   protected final double f_134118_;
-   protected final double f_134119_;
-   protected final double f_134120_;
-   protected final float f_134121_;
-   protected final float f_134122_;
+   public final double f_134118_;
+   public final double f_134119_;
+   public final double f_134120_;
+   public final float f_134121_;
+   public final float f_134122_;
    protected final boolean f_134123_;
-   protected final boolean f_134124_;
-   protected final boolean f_134125_;
+   public final boolean f_134124_;
+   public final boolean f_134125_;
 
    protected ServerboundMovePlayerPacket(double p_179675_, double p_179676_, double p_179677_, float p_179678_, float p_179679_, boolean p_179680_, boolean p_179681_, boolean p_179682_) {
       this.f_134118_ = p_179675_;
