--- ../src-base/minecraft/net/minecraft/server/gui/MinecraftServerGui.java
+++ ../src-work/minecraft/net/minecraft/server/gui/MinecraftServerGui.java
@@ -93,7 +93,7 @@
         }
         catch (Exception exception)
         {
-            LOGGER.error((String)"Couldn\'t build server GUI", (Throwable)exception);
+            LOGGER.error((String)"Couldn\'t build1 server GUI", (Throwable)exception);
         }
     }
 
