--- a/net/minecraft/tileentity/AbstractFurnaceTileEntity.java
+++ b/net/minecraft/tileentity/AbstractFurnaceTileEntity.java
@@ -6,6 +_,8 @@
 import it.unimi.dsi.fastutil.objects.Object2IntMap.Entry;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import net.minecraft.block.AbstractFurnaceBlock;
 import net.minecraft.block.BlockState;
@@ -36,16 +_,55 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.inventory.FurnaceBurnEvent;
+import org.bukkit.event.inventory.FurnaceExtractEvent;
+import org.bukkit.event.inventory.FurnaceSmeltEvent;
 
 public abstract class AbstractFurnaceTileEntity extends LockableTileEntity implements ISidedInventory, IRecipeHolder, IRecipeHelperPopulator, ITickableTileEntity {
    private static final int[] field_214015_g = new int[]{0};
    private static final int[] field_214016_h = new int[]{2, 1};
    private static final int[] field_214017_i = new int[]{1};
    protected NonNullList<ItemStack> field_214012_a = NonNullList.func_191197_a(3, ItemStack.field_190927_a);
-   private int field_214018_j;
+   public int field_214018_j;
    private int field_214019_k;
-   private int field_214020_l;
-   private int field_214021_m;
+   public int field_214020_l;
+   public int field_214021_m;
+
+   // CraftBukkit start - add fields and methods
+   private int maxStack = MAX_STACK;
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+
+   public List<ItemStack> getContents() {
+      return this.field_214012_a;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public int func_70297_j_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+   }
+   // CraftBukkit end
+
    protected final IIntArray field_214013_b = new IIntArray() {
       public int func_221476_a(int p_221476_1_) {
          switch(p_221476_1_) {
@@ -91,6 +_,7 @@
       this.field_214014_c = p_i49964_2_;
    }
 
+   @Deprecated //Forge - get burn times by calling ForgeHooks#getBurnTime(ItemStack)
    public static Map<Item, Integer> func_214001_f() {
       Map<Item, Integer> map = Maps.newLinkedHashMap();
       func_213996_a(map, Items.field_151129_at, 20000);
@@ -184,13 +_,13 @@
       return this.field_214018_j > 0;
    }
 
-   public void func_230337_a_(BlockState p_230337_1_, CompoundNBT p_230337_2_) {
+   public void func_230337_a_(BlockState p_230337_1_, CompoundNBT p_230337_2_) { //TODO: MARK
       super.func_230337_a_(p_230337_1_, p_230337_2_);
       this.field_214012_a = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
       ItemStackHelper.func_191283_b(p_230337_2_, this.field_214012_a);
-      this.field_214018_j = p_230337_2_.func_74765_d("BurnTime");
-      this.field_214020_l = p_230337_2_.func_74765_d("CookTime");
-      this.field_214021_m = p_230337_2_.func_74765_d("CookTimeTotal");
+      this.field_214018_j = p_230337_2_.func_74762_e("BurnTime");
+      this.field_214020_l = p_230337_2_.func_74762_e("CookTime");
+      this.field_214021_m = p_230337_2_.func_74762_e("CookTimeTotal");
       this.field_214019_k = this.func_213997_a(this.field_214012_a.get(1));
       CompoundNBT compoundnbt = p_230337_2_.func_74775_l("RecipesUsed");
 
@@ -202,9 +_,9 @@
 
    public CompoundNBT func_189515_b(CompoundNBT p_189515_1_) {
       super.func_189515_b(p_189515_1_);
-      p_189515_1_.func_74777_a("BurnTime", (short)this.field_214018_j);
-      p_189515_1_.func_74777_a("CookTime", (short)this.field_214020_l);
-      p_189515_1_.func_74777_a("CookTimeTotal", (short)this.field_214021_m);
+      p_189515_1_.func_74768_a("BurnTime", this.field_214018_j);
+      p_189515_1_.func_74768_a("CookTime", this.field_214020_l);
+      p_189515_1_.func_74768_a("CookTimeTotal", this.field_214021_m);
       ItemStackHelper.func_191282_a(p_189515_1_, this.field_214012_a);
       CompoundNBT compoundnbt = new CompoundNBT();
       this.field_214022_n.forEach((p_235643_1_, p_235643_2_) -> {
@@ -226,16 +_,26 @@
          if (this.func_214006_r() || !itemstack.func_190926_b() && !this.field_214012_a.get(0).func_190926_b()) {
             IRecipe<?> irecipe = this.field_145850_b.func_199532_z().func_215371_a((IRecipeType<AbstractCookingRecipe>)this.field_214014_c, this, this.field_145850_b).orElse(null);
             if (!this.func_214006_r() && this.func_214008_b(irecipe)) {
-               this.field_214018_j = this.func_213997_a(itemstack);
+               // CraftBukkit start
+               CraftItemStack fuel = CraftItemStack.asCraftMirror(itemstack);
+               FurnaceBurnEvent furnaceBurnEvent = new FurnaceBurnEvent(CraftBlock.at(this.field_145850_b, this.field_174879_c), fuel, this.func_213997_a(itemstack));
+               this.field_145850_b.getCBServer().getPluginManager().callEvent(furnaceBurnEvent);
+               if (furnaceBurnEvent.isCancelled()) {
+                  return;
+               }
+               this.field_214018_j = furnaceBurnEvent.getBurnTime();
                this.field_214019_k = this.field_214018_j;
-               if (this.func_214006_r()) {
+               if (this.func_214006_r() && furnaceBurnEvent.isBurning()) {
+                  // CraftBukkit end
                   flag1 = true;
+                  if (itemstack.hasContainerItem())
+                      this.field_214012_a.set(1, itemstack.getContainerItem());
+                  else
                   if (!itemstack.func_190926_b()) {
                      Item item = itemstack.func_77973_b();
                      itemstack.func_190918_g(1);
                      if (itemstack.func_190926_b()) {
-                        Item item1 = item.func_77668_q();
-                        this.field_214012_a.set(1, item1 == null ? ItemStack.field_190927_a : new ItemStack(item1));
+                        this.field_214012_a.set(1, itemstack.getContainerItem());
                      }
                   }
                }
@@ -270,7 +_,7 @@
 
    protected boolean func_214008_b(@Nullable IRecipe<?> p_214008_1_) {
       if (!this.field_214012_a.get(0).func_190926_b() && p_214008_1_ != null) {
-         ItemStack itemstack = p_214008_1_.func_77571_b();
+         ItemStack itemstack = ((IRecipe<ISidedInventory>) p_214008_1_).func_77572_b(this);
          if (itemstack.func_190926_b()) {
             return false;
          } else {
@@ -279,10 +_,10 @@
                return true;
             } else if (!itemstack1.func_77969_a(itemstack)) {
                return false;
-            } else if (itemstack1.func_190916_E() < this.func_70297_j_() && itemstack1.func_190916_E() < itemstack1.func_77976_d()) {
+            } else if (itemstack1.func_190916_E() + itemstack.func_190916_E() <= this.func_70297_j_() && itemstack1.func_190916_E() + itemstack.func_190916_E() <= itemstack1.func_77976_d()) { // Forge fix: make furnace respect stack sizes in furnace recipes
                return true;
             } else {
-               return itemstack1.func_190916_E() < itemstack.func_77976_d();
+               return itemstack1.func_190916_E() + itemstack.func_190916_E() <= itemstack.func_77976_d(); // Forge fix: make furnace respect stack sizes in furnace recipes
             }
          }
       } else {
@@ -293,12 +_,27 @@
    private void func_214007_c(@Nullable IRecipe<?> p_214007_1_) {
       if (p_214007_1_ != null && this.func_214008_b(p_214007_1_)) {
          ItemStack itemstack = this.field_214012_a.get(0);
-         ItemStack itemstack1 = p_214007_1_.func_77571_b();
+         ItemStack itemstack1 = ((IRecipe<ISidedInventory>) p_214007_1_).func_77572_b(this);
          ItemStack itemstack2 = this.field_214012_a.get(2);
-         if (itemstack2.func_190926_b()) {
-            this.field_214012_a.set(2, itemstack1.func_77946_l());
-         } else if (itemstack2.func_77973_b() == itemstack1.func_77973_b()) {
-            itemstack2.func_190917_f(1);
+
+         // CraftBukkit start - fire FurnaceSmeltEvent
+         CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+         org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+         FurnaceSmeltEvent furnaceSmeltEvent = new FurnaceSmeltEvent(this.field_145850_b.getWorld().getBlockAt(field_174879_c.func_177958_n(), field_174879_c.func_177956_o(), field_174879_c.func_177952_p()), source, result);
+         this.field_145850_b.getCBServer().getPluginManager().callEvent(furnaceSmeltEvent);
+         if (furnaceSmeltEvent.isCancelled()) {
+            return;
+         }
+         result = furnaceSmeltEvent.getResult();
+         itemstack1 = CraftItemStack.asNMSCopy(result);
+         if (!itemstack1.func_190926_b()) {
+            if (itemstack2.func_190926_b()) {
+               this.field_214012_a.set(2, itemstack1.func_77946_l());
+            } else if (CraftItemStack.asCraftMirror(itemstack2).isSimilar(result)) {
+               itemstack2.func_190917_f(itemstack1.func_190916_E());
+            } else {
+               return;
+            }
          }
 
          if (!this.field_145850_b.field_72995_K) {
@@ -318,7 +_,7 @@
          return 0;
       } else {
          Item item = p_213997_1_.func_77973_b();
-         return func_214001_f().getOrDefault(item, 0);
+         return net.minecraftforge.common.ForgeHooks.getBurnTime(p_213997_1_, this.field_214014_c);
       }
    }
 
@@ -327,7 +_,7 @@
    }
 
    public static boolean func_213991_b(ItemStack p_213991_0_) {
-      return func_214001_f().containsKey(p_213991_0_.func_77973_b());
+      return net.minecraftforge.common.ForgeHooks.getBurnTime(p_213991_0_, null) > 0;
    }
 
    public int[] func_180463_a(Direction p_180463_1_) {
@@ -410,7 +_,7 @@
          return true;
       } else {
          ItemStack itemstack = this.field_214012_a.get(1);
-         return func_213991_b(p_94041_2_) || p_94041_2_.func_77973_b() == Items.field_151133_ar && itemstack.func_77973_b() != Items.field_151133_ar;
+         return net.minecraftforge.common.ForgeHooks.getBurnTime(p_94041_2_, this.field_214014_c) > 0 || p_94041_2_.func_77973_b() == Items.field_151133_ar && itemstack.func_77973_b() != Items.field_151133_ar;
       }
    }
 
@@ -434,32 +_,52 @@
    public void func_201560_d(PlayerEntity p_201560_1_) {
    }
 
+   // CatServer start - Fix MCAR mod compatibility
+   public AtomicReference<PlayerEntity> catserver$awardPlayerEntity = new AtomicReference<>();
+   public AtomicReference<ItemStack> catserver$awardItemStack = new AtomicReference<>();
+   public AtomicInteger catserver$awardAmount = new AtomicInteger();
+
    public void func_235645_d_(PlayerEntity p_235645_1_) {
-      List<IRecipe<?>> list = this.func_235640_a_(p_235645_1_.field_70170_p, p_235645_1_.func_213303_ch());
+      List<IRecipe<?>> list = this.func_235640_a_(p_235645_1_.field_70170_p, p_235645_1_.func_213303_ch()); // TODO
       p_235645_1_.func_195065_a(list);
       this.field_214022_n.clear();
    }
 
+   public void unlockRecipes(PlayerEntity p_235645_1_, ItemStack itemstack, int amount) { // CraftBukkit
+      this.catserver$awardPlayerEntity.set(p_235645_1_);
+      this.catserver$awardItemStack.set(itemstack);
+      this.catserver$awardAmount.set(amount);
+      this.func_235645_d_(p_235645_1_);
+   }
+
    public List<IRecipe<?>> func_235640_a_(World p_235640_1_, Vector3d p_235640_2_) {
       List<IRecipe<?>> list = Lists.newArrayList();
 
       for(Entry<ResourceLocation> entry : this.field_214022_n.object2IntEntrySet()) {
          p_235640_1_.func_199532_z().func_215367_a(entry.getKey()).ifPresent((p_235642_4_) -> {
             list.add(p_235642_4_);
-            func_235641_a_(p_235640_1_, p_235640_2_, entry.getIntValue(), ((AbstractCookingRecipe)p_235642_4_).func_222138_b());
+            createExperience(p_235640_1_, p_235640_2_, entry.getIntValue(), ((AbstractCookingRecipe) p_235642_4_).func_222138_b(), catserver$awardPlayerEntity.getAndSet(null), catserver$awardItemStack.getAndSet(null), catserver$awardAmount.getAndSet(0)); // CatServer
          });
       }
 
       return list;
    }
 
-   private static void func_235641_a_(World p_235641_0_, Vector3d p_235641_1_, int p_235641_2_, float p_235641_3_) {
+   private void createExperience(World p_235641_0_, Vector3d p_235641_1_, int p_235641_2_, float p_235641_3_, PlayerEntity entityhuman, ItemStack itemstack, int amount) {
       int i = MathHelper.func_76141_d((float)p_235641_2_ * p_235641_3_);
       float f = MathHelper.func_226164_h_((float)p_235641_2_ * p_235641_3_);
       if (f != 0.0F && Math.random() < (double)f) {
          ++i;
       }
 
+      // CraftBukkit start - fire FurnaceExtractEvent
+      if (amount != 0) {
+         FurnaceExtractEvent event = new FurnaceExtractEvent((Player) entityhuman.getBukkitEntity(), CraftBlock.at(p_235641_0_, this.field_174879_c), org.bukkit.craftbukkit.v1_16_R3.util.CraftMagicNumbers.getMaterial(itemstack.func_77973_b()), amount, i);
+         this.field_145850_b.getCBServer().getPluginManager().callEvent(event);
+         i = event.getExpToDrop();
+      }
+      // CraftBukkit end
+
       while(i > 0) {
          int j = ExperienceOrbEntity.func_70527_a(i);
          i -= j;
@@ -467,11 +_,35 @@
       }
 
    }
+   // CatServer end
 
    public void func_194018_a(RecipeItemHelper p_194018_1_) {
       for(ItemStack itemstack : this.field_214012_a) {
          p_194018_1_.func_194112_a(itemstack);
       }
 
+   }
+
+   net.minecraftforge.common.util.LazyOptional<? extends net.minecraftforge.items.IItemHandler>[] handlers =
+           net.minecraftforge.items.wrapper.SidedInvWrapper.create(this, Direction.UP, Direction.DOWN, Direction.NORTH);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (!this.field_145846_f && facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+         if (facing == Direction.UP)
+            return handlers[0].cast();
+         else if (facing == Direction.DOWN)
+            return handlers[1].cast();
+         else
+            return handlers[2].cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   protected void invalidateCaps() {
+      super.invalidateCaps();
+      for (int x = 0; x < handlers.length; x++)
+        handlers[x].invalidate();
    }
 }
