--- a/net/minecraft/tileentity/BannerTileEntity.java
+++ b/net/minecraft/tileentity/BannerTileEntity.java
@@ -23,9 +_,9 @@
    @Nullable
    private ITextComponent field_190617_a;
    @Nullable
-   private DyeColor field_175120_a = DyeColor.WHITE;
+   public DyeColor field_175120_a = DyeColor.WHITE;
    @Nullable
-   private ListNBT field_175118_f;
+   public ListNBT field_175118_f;
    private boolean field_175119_g;
    @Nullable
    private List<Pair<BannerPattern, DyeColor>> field_175122_h;
@@ -99,6 +_,11 @@
       }
 
       this.field_175118_f = p_230337_2_.func_150295_c("Patterns", 10);
+      // CraftBukkit start
+      while (this.field_175118_f.size() > 20) {
+         this.field_175118_f.remove(20);
+      }
+      // CraftBukkit end
       this.field_175122_h = null;
       this.field_175119_g = true;
    }
