--- a/net/minecraft/tileentity/CampfireTileEntity.java
+++ b/net/minecraft/tileentity/CampfireTileEntity.java
@@ -5,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.CampfireBlock;
+import net.minecraft.entity.item.ItemEntity;
 import net.minecraft.inventory.IClearable;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
@@ -21,11 +_,14 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockCookEvent;
 
 public class CampfireTileEntity extends TileEntity implements IClearable, ITickableTileEntity {
    private final NonNullList<ItemStack> field_213987_a = NonNullList.func_191197_a(4, ItemStack.field_190927_a);
-   private final int[] field_213988_b = new int[4];
-   private final int[] field_213989_c = new int[4];
+   public final int[] field_213988_b = new int[4];
+   public final int[] field_213989_c = new int[4];
 
    public CampfireTileEntity() {
       super(TileEntityType.field_222488_F);
@@ -64,7 +_,23 @@
                   return p_213979_1_.func_77572_b(iinventory);
                }).orElse(itemstack);
                BlockPos blockpos = this.func_174877_v();
-               InventoryHelper.func_180173_a(this.field_145850_b, (double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p(), itemstack1);
+
+               // CraftBukkit start - fire BlockCookEvent
+               CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+               org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+               BlockCookEvent blockCookEvent = new BlockCookEvent(CraftBlock.at(this.field_145850_b, this.field_174879_c), source, result);
+               this.field_145850_b.getCBServer().getPluginManager().callEvent(blockCookEvent);
+               if (blockCookEvent.isCancelled()) {
+                  return;
+               }
+               result = blockCookEvent.getResult();
+               itemstack1 = CraftItemStack.asNMSCopy(result);
+               // CraftBukkit end
+               // Paper start
+               ItemEntity droppedItem = new ItemEntity(this.field_145850_b, blockpos.func_177958_n() + 0.5D, blockpos.func_177956_o() + 0.5D, blockpos.func_177952_p() + 0.5D, itemstack1.func_77979_a(this.field_145850_b.field_73012_v.nextInt(21) + 10));
+               droppedItem.func_213293_j(this.field_145850_b.field_73012_v.nextGaussian() * 0.05D, this.field_145850_b.field_73012_v.nextGaussian() * 0.05D + 0.2D, this.field_145850_b.field_73012_v.nextGaussian() * 0.05D);
+               this.field_145850_b.func_217376_c(droppedItem);
+               // Paper end
                this.field_213987_a.set(i, ItemStack.field_190927_a);
                this.func_213981_s();
             }
