--- a/net/minecraft/tileentity/CommandBlockLogic.java
+++ b/net/minecraft/tileentity/CommandBlockLogic.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
 
 public abstract class CommandBlockLogic implements ICommandSource {
    private static final SimpleDateFormat field_145766_a = new SimpleDateFormat("HH:mm:ss");
@@ -34,6 +_,11 @@
    private ITextComponent field_145762_d;
    private String field_145763_e = "";
    private ITextComponent field_145761_f = field_226655_c_;
+
+   // CraftBukkit start
+   @Override
+   public abstract CommandSender getBukkitSender(CommandSource wrapper);
+   // CraftBukkit end
 
    public int func_145760_g() {
       return this.field_145764_b;
