--- a/net/minecraft/tileentity/CommandBlockTileEntity.java
+++ b/net/minecraft/tileentity/CommandBlockTileEntity.java
@@ -15,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_16_R3.command.CraftBlockCommandSender;
 
 public class CommandBlockTileEntity extends TileEntity {
    private boolean field_184259_a;
@@ -22,7 +_,14 @@
    private boolean field_184261_g;
    private boolean field_184262_h;
    private final CommandBlockLogic field_145994_a = new CommandBlockLogic() {
-      public void func_145752_a(String p_145752_1_) {
+      // CraftBukkit start
+       @Override
+      public CommandSender getBukkitSender(CommandSource wrapper) {
+         return new CraftBlockCommandSender(wrapper, CommandBlockTileEntity.this);
+       }
+      // CraftBukkit end
+
+       public void func_145752_a(String p_145752_1_) {
          super.func_145752_a(p_145752_1_);
          CommandBlockTileEntity.this.func_70296_d();
       }
