--- a/net/minecraft/tileentity/EndGatewayTileEntity.java
+++ b/net/minecraft/tileentity/EndGatewayTileEntity.java
@@ -20,6 +_,7 @@
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
@@ -31,14 +_,18 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndGatewayTileEntity extends EndPortalTileEntity implements ITickableTileEntity {
    private static final Logger field_195503_a = LogManager.getLogger();
-   private long field_195504_f;
+   public long field_195504_f;
    private int field_195505_g;
    @Nullable
-   private BlockPos field_195506_h;
-   private boolean field_195507_i;
+   public BlockPos field_195506_h;
+   public boolean field_195507_i;
 
    public EndGatewayTileEntity() {
       super(TileEntityType.field_200991_v);
@@ -148,7 +_,7 @@
    public void func_195496_a(Entity p_195496_1_) {
       if (this.field_145850_b instanceof ServerWorld && !this.func_195500_d()) {
          this.field_195505_g = 100;
-         if (this.field_195506_h == null && this.field_145850_b.func_234923_W_() == World.field_234920_i_) {
+         if (this.field_195506_h == null && this.field_145850_b.getTypeKey() == DimensionType.field_236001_e_) { // CraftBukkit - work in alternate worlds
             this.func_227015_a_((ServerWorld)this.field_145850_b);
          }
 
@@ -170,6 +_,24 @@
             } else {
                entity = p_195496_1_.func_184208_bv();
             }
+
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (p_195496_1_ instanceof ServerPlayerEntity) {
+               org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer player = (CraftPlayer) p_195496_1_.getBukkitEntity();
+               org.bukkit.Location location = new Location(field_145850_b.getWorld(), (double) blockpos.func_177958_n() + 0.5D, (double) blockpos.func_177956_o() + 0.5D, (double) blockpos.func_177952_p() + 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+               PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+               ((ServerPlayerEntity) p_195496_1_).field_71135_a.teleport(teleEvent.getTo());
+               this.func_195500_d(); // CraftBukkit - call at end of method
+               return;
+            }
+            // CraftBukkit end
 
             entity.func_242279_ag();
             entity.func_223102_j((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p() + 0.5D);
