--- a/net/minecraft/tileentity/LecternTileEntity.java
+++ b/net/minecraft/tileentity/LecternTileEntity.java
@@ -1,5 +_,9 @@
 package net.minecraft.tileentity;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.LecternBlock;
@@ -25,9 +_,62 @@
 import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.server.ServerWorld;
-
-public class LecternTileEntity extends TileEntity implements IClearable, INamedContainerProvider {
-   private final IInventory field_214048_a = new IInventory() {
+import org.bukkit.Location;
+import org.bukkit.block.Lectern;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_16_R3.command.CraftBlockCommandSender;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+
+public class LecternTileEntity extends TileEntity implements IClearable, INamedContainerProvider, ICommandSource { // CraftBukkit - ICommandSource
+   // CraftBukkit start - add fields and methods
+   public final IInventory field_214048_a = new LecternInventory();
+   public class LecternInventory implements IInventory {
+
+       public List<HumanEntity> transaction = new ArrayList<>();
+       private int maxStack = 1;
+
+       @Override
+       public List<ItemStack> getContents() {
+          return Arrays.asList(field_214050_c);
+       }
+
+       @Override
+       public void onOpen(CraftHumanEntity who) {
+          transaction.add(who);
+       }
+
+       @Override
+       public void onClose(CraftHumanEntity who) {
+          transaction.remove(who);
+       }
+
+       @Override
+       public List<HumanEntity> getViewers() {
+          return transaction;
+       }
+
+       @Override
+       public void setMaxStackSize(int i) {
+          maxStack = i;
+       }
+
+       @Override
+       public Location getLocation() {
+          return new Location(field_145850_b.getWorld(), field_174879_c.func_177958_n(), field_174879_c.func_177956_o(), field_174879_c.func_177952_p());
+       }
+
+       @Override
+       public InventoryHolder getOwner() {
+          return (Lectern) LecternTileEntity.this.getOwner();
+       }
+
+       public LecternTileEntity getLectern(){
+          return LecternTileEntity.this;
+       }
+
+       // CraftBukkit end
       public int func_70302_i_() {
          return 1;
       }
@@ -36,13 +_,13 @@
          return LecternTileEntity.this.field_214050_c.func_190926_b();
       }
 
-      public ItemStack func_70301_a(int p_70301_1_) {
-         return p_70301_1_ == 0 ? LecternTileEntity.this.field_214050_c : ItemStack.field_190927_a;
+      public ItemStack func_70301_a(int index) {
+         return index == 0 ? LecternTileEntity.this.field_214050_c : ItemStack.field_190927_a;
       }
 
-      public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_) {
-         if (p_70298_1_ == 0) {
-            ItemStack itemstack = LecternTileEntity.this.field_214050_c.func_77979_a(p_70298_2_);
+      public ItemStack func_70298_a(int index, int count) {
+         if (index == 0) {
+            ItemStack itemstack = LecternTileEntity.this.field_214050_c.func_77979_a(count);
             if (LecternTileEntity.this.field_214050_c.func_190926_b()) {
                LecternTileEntity.this.func_214042_s();
             }
@@ -53,8 +_,8 @@
          }
       }
 
-      public ItemStack func_70304_b(int p_70304_1_) {
-         if (p_70304_1_ == 0) {
+      public ItemStack func_70304_b(int index) {
+         if (index == 0) {
             ItemStack itemstack = LecternTileEntity.this.field_214050_c;
             LecternTileEntity.this.field_214050_c = ItemStack.field_190927_a;
             LecternTileEntity.this.func_214042_s();
@@ -64,26 +_,34 @@
          }
       }
 
-      public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_) {
+      public void func_70299_a(int index, ItemStack stack) {
+         // CraftBukkit start
+          if (index == 0) {
+            LecternTileEntity.this.func_214045_a(stack);
+            if (LecternTileEntity.this.func_145831_w() != null) {
+               LecternBlock.func_220155_a(LecternTileEntity.this.func_145831_w(), LecternTileEntity.this.func_174877_v(), LecternTileEntity.this.func_195044_w(), LecternTileEntity.this.func_214046_f());
+            }
+         }
+         // CraftBukkit end
       }
 
       public int func_70297_j_() {
-         return 1;
+         return maxStack; // CraftBukkit
       }
 
       public void func_70296_d() {
          LecternTileEntity.this.func_70296_d();
       }
 
-      public boolean func_70300_a(PlayerEntity p_70300_1_) {
+      public boolean func_70300_a(PlayerEntity player) {
          if (LecternTileEntity.this.field_145850_b.func_175625_s(LecternTileEntity.this.field_174879_c) != LecternTileEntity.this) {
             return false;
          } else {
-            return p_70300_1_.func_70092_e((double)LecternTileEntity.this.field_174879_c.func_177958_n() + 0.5D, (double)LecternTileEntity.this.field_174879_c.func_177956_o() + 0.5D, (double)LecternTileEntity.this.field_174879_c.func_177952_p() + 0.5D) > 64.0D ? false : LecternTileEntity.this.func_214046_f();
+            return player.func_70092_e((double)LecternTileEntity.this.field_174879_c.func_177958_n() + 0.5D, (double)LecternTileEntity.this.field_174879_c.func_177956_o() + 0.5D, (double)LecternTileEntity.this.field_174879_c.func_177952_p() + 0.5D) > 64.0D ? false : LecternTileEntity.this.func_214046_f();
          }
       }
 
-      public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_) {
+      public boolean func_94041_b(int index, ItemStack stack) {
          return false;
       }
 
@@ -91,13 +_,13 @@
       }
    };
    private final IIntArray field_214049_b = new IIntArray() {
-      public int func_221476_a(int p_221476_1_) {
-         return p_221476_1_ == 0 ? LecternTileEntity.this.field_214051_g : 0;
+      public int func_221476_a(int index) {
+         return index == 0 ? LecternTileEntity.this.field_214051_g : 0;
       }
 
-      public void func_221477_a(int p_221477_1_, int p_221477_2_) {
-         if (p_221477_1_ == 0) {
-            LecternTileEntity.this.func_214035_a(p_221477_2_);
+      public void func_221477_a(int index, int value) {
+         if (index == 0) {
+            LecternTileEntity.this.func_214035_a(value);
          }
 
       }
@@ -140,7 +_,7 @@
       this.func_70296_d();
    }
 
-   private void func_214035_a(int p_214035_1_) {
+   public void func_214035_a(int p_214035_1_) {
       int i = MathHelper.func_76125_a(p_214035_1_, 0, this.field_214052_h - 1);
       if (i != this.field_214051_g) {
          this.field_214051_g = i;
@@ -167,6 +_,34 @@
       return p_214047_1_;
    }
 
+
+   // CraftBukkit start
+   @Override
+   public void func_145747_a(ITextComponent component, UUID senderUUID) {
+
+   }
+
+   @Override
+   public CommandSender getBukkitSender(final CommandSource wrapper) {
+      return (wrapper.func_197022_f() != null) ? wrapper.func_197022_f().getBukkitSender(wrapper) : new CraftBlockCommandSender(wrapper, this);
+   }
+
+   @Override
+   public boolean func_195039_a() {
+      return false;
+   }
+
+   @Override
+   public boolean func_195040_b() {
+      return false;
+   }
+
+   @Override
+   public boolean func_195041_r_() {
+      return false;
+   }
+   // CraftBukkit end
+
    private CommandSource func_214039_a(@Nullable PlayerEntity p_214039_1_) {
       String s;
       ITextComponent itextcomponent;
@@ -179,7 +_,8 @@
       }
 
       Vector3d vector3d = Vector3d.func_237489_a_(this.field_174879_c);
-      return new CommandSource(ICommandSource.field_213139_a_, vector3d, Vector2f.field_189974_a, (ServerWorld)this.field_145850_b, 2, s, itextcomponent, this.field_145850_b.func_73046_m(), p_214039_1_);
+      // CraftBukkit - this
+      return new CommandSource(this, vector3d, Vector2f.field_189974_a, (ServerWorld)this.field_145850_b, 2, s, itextcomponent, this.field_145850_b.func_73046_m(), p_214039_1_);
    }
 
    public boolean func_183000_F() {
@@ -213,7 +_,7 @@
    }
 
    public Container createMenu(int p_createMenu_1_, PlayerInventory p_createMenu_2_, PlayerEntity p_createMenu_3_) {
-      return new LecternContainer(p_createMenu_1_, this.field_214048_a, this.field_214049_b);
+      return new LecternContainer(p_createMenu_1_, this.field_214048_a, this.field_214049_b, p_createMenu_2_); // CraftBukkit
    }
 
    public ITextComponent func_145748_c_() {
