--- a/net/minecraft/tileentity/LockableLootTileEntity.java
+++ b/net/minecraft/tileentity/LockableLootTileEntity.java
@@ -23,8 +_,8 @@
 
 public abstract class LockableLootTileEntity extends LockableTileEntity {
    @Nullable
-   protected ResourceLocation field_184284_m;
-   protected long field_184285_n;
+   public ResourceLocation field_184284_m;
+   public long field_184285_n;
 
    protected LockableLootTileEntity(TileEntityType<?> p_i48284_1_) {
       super(p_i48284_1_);
