--- ../src-base/minecraft/net/minecraftforge/common/util/Constants.java
+++ ../src-work/minecraft/net/minecraftforge/common/util/Constants.java
@@ -29,7 +29,7 @@
      * NBT Tag type IDS, used when storing the nbt to disc, Should align with NBTBase.getId, 
      * table used in NBTBase.func_150283_g
      * 
-     * Main use is checking tag type in NBTTagCompound.func_150297_b(String, int)
+     * Main use is checking tag type in NBTTagCompound.hasKey(String, int)
      *
      */
     public static class NBT
