--- ../src-base/minecraft/net/minecraft/util/text/event/ClickEvent.java
+++ ../src-work/minecraft/net/minecraft/util/text/event/ClickEvent.java
@@ -5,16 +5,16 @@
 
 public class ClickEvent
 {
-    private final ClickEvent.Action action;
+    private final Action action;
     private final String value;
 
-    public ClickEvent(ClickEvent.Action theAction, String theValue)
+    public ClickEvent(Action theAction, String theValue)
     {
         this.action = theAction;
         this.value = theValue;
     }
 
-    public ClickEvent.Action getAction()
+    public Action getAction()
     {
         return this.action;
     }
@@ -81,7 +81,7 @@
         SUGGEST_COMMAND("suggest_command", true),
         CHANGE_PAGE("change_page", true);
 
-        private static final Map<String, ClickEvent.Action> NAME_MAPPING = Maps.<String, ClickEvent.Action>newHashMap();
+        private static final Map<String, Action> NAME_MAPPING = Maps.<String, Action>newHashMap();
         private final boolean allowedInChat;
         private final String canonicalName;
 
@@ -101,14 +101,14 @@
             return this.canonicalName;
         }
 
-        public static ClickEvent.Action getValueByCanonicalName(String canonicalNameIn)
+        public static Action getValueByCanonicalName(String canonicalNameIn)
         {
             return NAME_MAPPING.get(canonicalNameIn);
         }
 
         static
         {
-            for (ClickEvent.Action clickevent$action : values())
+            for (Action clickevent$action : values())
             {
                 NAME_MAPPING.put(clickevent$action.getCanonicalName(), clickevent$action);
             }
