--- ../src-base/minecraft/net/minecraft/util/text/event/HoverEvent.java
+++ ../src-work/minecraft/net/minecraft/util/text/event/HoverEvent.java
@@ -6,16 +6,16 @@
 
 public class HoverEvent
 {
-    private final HoverEvent.Action action;
+    private final Action action;
     private final ITextComponent value;
 
-    public HoverEvent(HoverEvent.Action actionIn, ITextComponent valueIn)
+    public HoverEvent(Action actionIn, ITextComponent valueIn)
     {
         this.action = actionIn;
         this.value = valueIn;
     }
 
-    public HoverEvent.Action getAction()
+    public Action getAction()
     {
         return this.action;
     }
@@ -80,7 +80,7 @@
         SHOW_ITEM("show_item", true),
         SHOW_ENTITY("show_entity", true);
 
-        private static final Map<String, HoverEvent.Action> NAME_MAPPING = Maps.<String, HoverEvent.Action>newHashMap();
+        private static final Map<String, Action> NAME_MAPPING = Maps.<String, Action>newHashMap();
         private final boolean allowedInChat;
         private final String canonicalName;
 
@@ -100,14 +100,14 @@
             return this.canonicalName;
         }
 
-        public static HoverEvent.Action getValueByCanonicalName(String canonicalNameIn)
+        public static Action getValueByCanonicalName(String canonicalNameIn)
         {
             return NAME_MAPPING.get(canonicalNameIn);
         }
 
         static
         {
-            for (HoverEvent.Action hoverevent$action : values())
+            for (Action hoverevent$action : values())
             {
                 NAME_MAPPING.put(hoverevent$action.getCanonicalName(), hoverevent$action);
             }
