--- ../src-base/minecraft/net/minecraft/client/network/LanServerDetector.java
+++ ../src-work/minecraft/net/minecraft/client/network/LanServerDetector.java
@@ -75,11 +75,11 @@
     @SideOnly(Side.CLIENT)
     public static class ThreadLanServerFind extends Thread
         {
-            private final LanServerDetector.LanServerList localServerList;
+            private final LanServerList localServerList;
             private final InetAddress broadcastAddress;
             private final MulticastSocket socket;
 
-            public ThreadLanServerFind(LanServerDetector.LanServerList list) throws IOException
+            public ThreadLanServerFind(LanServerList list) throws IOException
             {
                 super("LanServerDetector #" + LanServerDetector.ATOMIC_COUNTER.incrementAndGet());
                 this.localServerList = list;
