--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketCreativeInventoryAction.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketCreativeInventoryAction.java
@@ -33,6 +33,8 @@
     {
         this.slotId = buf.readShort();
         this.stack = buf.readItemStackFromBuffer();
+        if(stack != null)
+        	System.out.println(stack);
     }
 
     public void writePacketData(PacketBuffer buf) throws IOException
