--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketEntityAction.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketEntityAction.java
@@ -11,7 +11,7 @@
 public class CPacketEntityAction implements Packet<INetHandlerPlayServer>
 {
     private int entityID;
-    private CPacketEntityAction.Action action;
+    private Action action;
     private int auxData;
 
     public CPacketEntityAction()
@@ -19,13 +19,13 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public CPacketEntityAction(Entity entityIn, CPacketEntityAction.Action actionIn)
+    public CPacketEntityAction(Entity entityIn, Action actionIn)
     {
         this(entityIn, actionIn, 0);
     }
 
     @SideOnly(Side.CLIENT)
-    public CPacketEntityAction(Entity entityIn, CPacketEntityAction.Action actionIn, int auxDataIn)
+    public CPacketEntityAction(Entity entityIn, Action actionIn, int auxDataIn)
     {
         this.entityID = entityIn.getEntityId();
         this.action = actionIn;
@@ -35,7 +35,7 @@
     public void readPacketData(PacketBuffer buf) throws IOException
     {
         this.entityID = buf.readVarInt();
-        this.action = (CPacketEntityAction.Action)buf.readEnumValue(CPacketEntityAction.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
         this.auxData = buf.readVarInt();
     }
 
@@ -51,7 +51,7 @@
         handler.processEntityAction(this);
     }
 
-    public CPacketEntityAction.Action getAction()
+    public Action getAction()
     {
         return this.action;
     }
