--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketCloseWindow.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketCloseWindow.java
@@ -4,8 +4,6 @@
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayServer;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class CPacketCloseWindow implements Packet<INetHandlerPlayServer>
 {
@@ -15,7 +13,6 @@
     {
     }
 
-    @SideOnly(Side.CLIENT)
     public CPacketCloseWindow(int windowIdIn)
     {
         this.windowId = windowIdIn;
