--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketRecipeInfo.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketRecipeInfo.java
@@ -11,7 +11,7 @@
 
 public class CPacketRecipeInfo implements Packet<INetHandlerPlayServer>
 {
-    private CPacketRecipeInfo.Purpose purpose;
+    private Purpose purpose;
     private IRecipe recipe;
     private boolean isGuiOpen;
     private boolean filteringCraftable;
@@ -22,27 +22,27 @@
 
     public CPacketRecipeInfo(IRecipe p_i47518_1_)
     {
-        this.purpose = CPacketRecipeInfo.Purpose.SHOWN;
+        this.purpose = Purpose.SHOWN;
         this.recipe = p_i47518_1_;
     }
 
     @SideOnly(Side.CLIENT)
     public CPacketRecipeInfo(boolean p_i47424_1_, boolean p_i47424_2_)
     {
-        this.purpose = CPacketRecipeInfo.Purpose.SETTINGS;
+        this.purpose = Purpose.SETTINGS;
         this.isGuiOpen = p_i47424_1_;
         this.filteringCraftable = p_i47424_2_;
     }
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.purpose = (CPacketRecipeInfo.Purpose)buf.readEnumValue(CPacketRecipeInfo.Purpose.class);
+        this.purpose = (Purpose)buf.readEnumValue(Purpose.class);
 
-        if (this.purpose == CPacketRecipeInfo.Purpose.SHOWN)
+        if (this.purpose == Purpose.SHOWN)
         {
             this.recipe = CraftingManager.getRecipeById(buf.readInt());
         }
-        else if (this.purpose == CPacketRecipeInfo.Purpose.SETTINGS)
+        else if (this.purpose == Purpose.SETTINGS)
         {
             this.isGuiOpen = buf.readBoolean();
             this.filteringCraftable = buf.readBoolean();
@@ -53,11 +53,11 @@
     {
         buf.writeEnumValue(this.purpose);
 
-        if (this.purpose == CPacketRecipeInfo.Purpose.SHOWN)
+        if (this.purpose == Purpose.SHOWN)
         {
             buf.writeInt(CraftingManager.getIDForRecipe(this.recipe));
         }
-        else if (this.purpose == CPacketRecipeInfo.Purpose.SETTINGS)
+        else if (this.purpose == Purpose.SETTINGS)
         {
             buf.writeBoolean(this.isGuiOpen);
             buf.writeBoolean(this.filteringCraftable);
@@ -69,7 +69,7 @@
         handler.handleRecipeBookUpdate(this);
     }
 
-    public CPacketRecipeInfo.Purpose getPurpose()
+    public Purpose getPurpose()
     {
         return this.purpose;
     }
