--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketClickWindow.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketClickWindow.java
@@ -55,7 +55,7 @@
         buf.writeByte(this.packedClickData);
         buf.writeShort(this.actionNumber);
         buf.writeEnumValue(this.mode);
-        buf.writeItemStack(this.clickedItem);
+        net.minecraftforge.common.util.PacketUtil.writeItemStackFromClientToServer(buf, this.clickedItem);
     }
 
     public int getWindowId()
