--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketSeenAdvancements.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketSeenAdvancements.java
@@ -12,7 +12,7 @@
 
 public class CPacketSeenAdvancements implements Packet<INetHandlerPlayServer>
 {
-    private CPacketSeenAdvancements.Action action;
+    private Action action;
     private ResourceLocation tab;
 
     public CPacketSeenAdvancements()
@@ -20,7 +20,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public CPacketSeenAdvancements(CPacketSeenAdvancements.Action p_i47595_1_, @Nullable ResourceLocation p_i47595_2_)
+    public CPacketSeenAdvancements(Action p_i47595_1_, @Nullable ResourceLocation p_i47595_2_)
     {
         this.action = p_i47595_1_;
         this.tab = p_i47595_2_;
@@ -29,20 +29,20 @@
     @SideOnly(Side.CLIENT)
     public static CPacketSeenAdvancements openedTab(Advancement p_194163_0_)
     {
-        return new CPacketSeenAdvancements(CPacketSeenAdvancements.Action.OPENED_TAB, p_194163_0_.getId());
+        return new CPacketSeenAdvancements(Action.OPENED_TAB, p_194163_0_.getId());
     }
 
     @SideOnly(Side.CLIENT)
     public static CPacketSeenAdvancements closedScreen()
     {
-        return new CPacketSeenAdvancements(CPacketSeenAdvancements.Action.CLOSED_SCREEN, (ResourceLocation)null);
+        return new CPacketSeenAdvancements(Action.CLOSED_SCREEN, (ResourceLocation)null);
     }
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.action = (CPacketSeenAdvancements.Action)buf.readEnumValue(CPacketSeenAdvancements.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
 
-        if (this.action == CPacketSeenAdvancements.Action.OPENED_TAB)
+        if (this.action == Action.OPENED_TAB)
         {
             this.tab = buf.readResourceLocation();
         }
@@ -52,7 +52,7 @@
     {
         buf.writeEnumValue(this.action);
 
-        if (this.action == CPacketSeenAdvancements.Action.OPENED_TAB)
+        if (this.action == Action.OPENED_TAB)
         {
             buf.writeResourceLocation(this.tab);
         }
@@ -63,7 +63,7 @@
         handler.handleSeenAdvancements(this);
     }
 
-    public CPacketSeenAdvancements.Action getAction()
+    public Action getAction()
     {
         return this.action;
     }
