--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketClientStatus.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketClientStatus.java
@@ -7,20 +7,20 @@
 
 public class CPacketClientStatus implements Packet<INetHandlerPlayServer>
 {
-    private CPacketClientStatus.State status;
+    private State status;
 
     public CPacketClientStatus()
     {
     }
 
-    public CPacketClientStatus(CPacketClientStatus.State p_i46886_1_)
+    public CPacketClientStatus(State p_i46886_1_)
     {
         this.status = p_i46886_1_;
     }
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.status = (CPacketClientStatus.State)buf.readEnumValue(CPacketClientStatus.State.class);
+        this.status = (State)buf.readEnumValue(State.class);
     }
 
     public void writePacketData(PacketBuffer buf) throws IOException
@@ -33,7 +33,7 @@
         handler.processClientStatus(this);
     }
 
-    public CPacketClientStatus.State getStatus()
+    public State getStatus()
     {
         return this.status;
     }
