--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketResourcePackStatus.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketResourcePackStatus.java
@@ -7,7 +7,7 @@
 
 public class CPacketResourcePackStatus implements Packet<INetHandlerPlayServer>
 {
-    private CPacketResourcePackStatus.Action action;
+    public CPacketResourcePackStatus.Action action;
 
     public CPacketResourcePackStatus()
     {
