--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketResourcePackStatus.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketResourcePackStatus.java
@@ -7,20 +7,20 @@
 
 public class CPacketResourcePackStatus implements Packet<INetHandlerPlayServer>
 {
-    private CPacketResourcePackStatus.Action action;
+    public Action action;
 
     public CPacketResourcePackStatus()
     {
     }
 
-    public CPacketResourcePackStatus(CPacketResourcePackStatus.Action p_i47156_1_)
+    public CPacketResourcePackStatus(Action p_i47156_1_)
     {
         this.action = p_i47156_1_;
     }
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.action = (CPacketResourcePackStatus.Action)buf.readEnumValue(CPacketResourcePackStatus.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
     }
 
     public void writePacketData(PacketBuffer buf) throws IOException
