--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketPlayerDigging.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketPlayerDigging.java
@@ -13,14 +13,14 @@
 {
     private BlockPos position;
     private EnumFacing facing;
-    private CPacketPlayerDigging.Action action;
+    private Action action;
 
     public CPacketPlayerDigging()
     {
     }
 
     @SideOnly(Side.CLIENT)
-    public CPacketPlayerDigging(CPacketPlayerDigging.Action actionIn, BlockPos posIn, EnumFacing facingIn)
+    public CPacketPlayerDigging(Action actionIn, BlockPos posIn, EnumFacing facingIn)
     {
         this.action = actionIn;
         this.position = posIn;
@@ -29,7 +29,7 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.action = (CPacketPlayerDigging.Action)buf.readEnumValue(CPacketPlayerDigging.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
         this.position = buf.readBlockPos();
         this.facing = EnumFacing.getFront(buf.readUnsignedByte());
     }
@@ -56,7 +56,7 @@
         return this.facing;
     }
 
-    public CPacketPlayerDigging.Action getAction()
+    public Action getAction()
     {
         return this.action;
     }
