--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketPlayerTryUseItemOnBlock.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketPlayerTryUseItemOnBlock.java
@@ -18,6 +18,7 @@
     private float facingX;
     private float facingY;
     private float facingZ;
+    public long timestamp; // Spigot
 
     public CPacketPlayerTryUseItemOnBlock()
     {
@@ -37,6 +38,7 @@
     public void readPacketData(PacketBuffer buf) throws IOException
     {
         this.position = buf.readBlockPos();
+        this.timestamp = System.currentTimeMillis(); // Spigot
         this.placedBlockDirection = (EnumFacing)buf.readEnumValue(EnumFacing.class);
         this.hand = (EnumHand)buf.readEnumValue(EnumHand.class);
         this.facingX = (float)buf.readUnsignedByte() / 16.0F;
