--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketPlayerTryUseItem.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketPlayerTryUseItem.java
@@ -9,6 +9,7 @@
 public class CPacketPlayerTryUseItem implements Packet<INetHandlerPlayServer>
 {
     private EnumHand hand;
+    public long timestamp;
 
     public CPacketPlayerTryUseItem()
     {
@@ -21,6 +22,7 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
+        this.timestamp = System.currentTimeMillis(); // Spigot
         this.hand = (EnumHand)buf.readEnumValue(EnumHand.class);
     }
 
