--- ../src-base/minecraft/net/minecraft/world/storage/loot/conditions/LootConditionManager.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/conditions/LootConditionManager.java
@@ -42,6 +42,15 @@
         }
     }
 
+    public static boolean testAllConditions(Iterable<LootCondition> conditions, Random rand, LootContext context)
+    {
+        if (conditions == null) return true;
+        for (LootCondition cond : conditions)
+           if (!cond.testCondition(rand, context))
+                return false;
+        return true;
+    }
+
     public static boolean testAllConditions(@Nullable LootCondition[] conditions, Random rand, LootContext context)
     {
         if (conditions == null)
