--- ../src-base/minecraft/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
@@ -23,13 +23,8 @@
 
     public boolean testCondition(Random rand, LootContext context)
     {
-        int i = 0;
+        int i = context.getLootingModifier();
 
-        if (context.getKiller() instanceof EntityLivingBase)
-        {
-            i = EnchantmentHelper.getLootingModifier((EntityLivingBase)context.getKiller());
-        }
-
         return rand.nextFloat() < this.chance + (float)i * this.lootingMultiplier;
     }
 
