--- a/net/minecraft/world/level/portal/PortalForcer.java
+++ b/net/minecraft/world/level/portal/PortalForcer.java
@@ -19,7 +_,7 @@
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.levelgen.Heightmap;
 
-public class PortalForcer {
+public class PortalForcer implements net.minecraftforge.common.util.ITeleporter {
    private static final int f_164734_ = 3;
    private static final int f_164735_ = 128;
    private static final int f_164736_ = 16;
@@ -40,8 +_,13 @@
    }
 
    public Optional<BlockUtil.FoundRectangle> m_192985_(BlockPos p_192986_, boolean p_192987_, WorldBorder p_192988_) {
+      // CraftBukkit start
+      return findPortalAround(p_192986_, p_192988_, p_192987_ ? 16 : 128); // Search Radius
+   }
+
+   public Optional<BlockUtil.FoundRectangle> findPortalAround(BlockPos p_192986_, WorldBorder p_192988_, int i) {
       PoiManager poimanager = this.f_77648_.m_8904_();
-      int i = p_192987_ ? 16 : 128;
+      // CraftBukkit end
       poimanager.m_27056_(this.f_77648_, p_192986_, i);
       Optional<PoiRecord> optional = poimanager.m_27166_((p_77654_) -> {
          return p_77654_ == PoiType.f_27350_;
@@ -65,6 +_,12 @@
    }
 
    public Optional<BlockUtil.FoundRectangle> m_77666_(BlockPos p_77667_, Direction.Axis p_77668_) {
+      // CraftBukkit start
+      return createPortal(p_77667_, p_77668_, null, 16);
+   }
+
+   public Optional<BlockUtil.FoundRectangle> createPortal(BlockPos p_77667_, Direction.Axis p_77668_, net.minecraft.world.entity.Entity entity, int createRadius) {
+      // CraftBukkit end
       Direction direction = Direction.m_122390_(Direction.AxisDirection.POSITIVE, p_77668_);
       double d0 = -1.0D;
       BlockPos blockpos = null;
@@ -74,7 +_,7 @@
       int i = Math.min(this.f_77648_.m_151558_(), this.f_77648_.m_141937_() + this.f_77648_.m_143344_()) - 1;
       BlockPos.MutableBlockPos blockpos$mutableblockpos = p_77667_.m_122032_();
 
-      for(BlockPos.MutableBlockPos blockpos$mutableblockpos1 : BlockPos.m_121935_(p_77667_, 16, Direction.EAST, Direction.SOUTH)) {
+      for(BlockPos.MutableBlockPos blockpos$mutableblockpos1 : BlockPos.m_121935_(p_77667_, createRadius, Direction.EAST, Direction.SOUTH)) {
          int j = Math.min(i, this.f_77648_.m_6924_(Heightmap.Types.MOTION_BLOCKING, blockpos$mutableblockpos1.m_123341_(), blockpos$mutableblockpos1.m_123343_()));
          int k = 1;
          if (worldborder.m_61937_(blockpos$mutableblockpos1) && worldborder.m_61937_(blockpos$mutableblockpos1.m_122175_(direction, 1))) {
@@ -115,6 +_,7 @@
          d0 = d1;
       }
 
+      org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator(this.f_77648_); // CraftBukkit - Use BlockStateListPopulator
       if (d0 == -1.0D) {
          int k1 = Math.max(this.f_77648_.m_141937_() - -1, 70);
          int i2 = i - 9;
@@ -133,7 +_,7 @@
                for(int k3 = -1; k3 < 3; ++k3) {
                   BlockState blockstate1 = k3 < 0 ? Blocks.f_50080_.m_49966_() : Blocks.f_50016_.m_49966_();
                   blockpos$mutableblockpos.m_122154_(blockpos, j3 * direction.m_122429_() + i3 * direction1.m_122429_(), k3, j3 * direction.m_122431_() + i3 * direction1.m_122431_());
-                  this.f_77648_.m_46597_(blockpos$mutableblockpos, blockstate1);
+                  blockList.m_7731_(blockpos$mutableblockpos, blockstate1, 3); // CraftBukkit
                }
             }
          }
@@ -143,7 +_,7 @@
          for(int j2 = -1; j2 < 4; ++j2) {
             if (l1 == -1 || l1 == 2 || j2 == -1 || j2 == 3) {
                blockpos$mutableblockpos.m_122154_(blockpos, l1 * direction.m_122429_(), j2, l1 * direction.m_122431_());
-               this.f_77648_.m_7731_(blockpos$mutableblockpos, Blocks.f_50080_.m_49966_(), 3);
+               blockList.m_7731_(blockpos$mutableblockpos, Blocks.f_50080_.m_49966_(), 3); // CraftBukkit
             }
          }
       }
@@ -153,9 +_,19 @@
       for(int k2 = 0; k2 < 2; ++k2) {
          for(int l2 = 0; l2 < 3; ++l2) {
             blockpos$mutableblockpos.m_122154_(blockpos, k2 * direction.m_122429_(), l2, k2 * direction.m_122431_());
-            this.f_77648_.m_7731_(blockpos$mutableblockpos, blockstate, 18);
+            blockList.m_7731_(blockpos$mutableblockpos, blockstate, 18); // CraftBukkit
          }
       }
+
+      // CraftBukkit start
+      org.bukkit.World bworld = this.f_77648_.getWorld();
+      org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blockList.getList(), bworld, (entity == null) ? null : entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.NETHER_PAIR);
+      this.f_77648_.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return Optional.empty();
+      }
+      blockList.updateList();
+      // CraftBukkit end
 
       return Optional.of(new BlockUtil.FoundRectangle(blockpos.m_7949_(), 2, 3));
    }
