--- a/net/minecraft/world/level/portal/PortalInfo.java
+++ b/net/minecraft/world/level/portal/PortalInfo.java
@@ -1,17 +_,31 @@
 package net.minecraft.world.level.portal;
 
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.phys.Vec3;
+import net.minecraftforge.event.world.BlockEvent;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent;
+import org.bukkit.event.world.PortalCreateEvent;
 
+// FoxServer TODO
 public class PortalInfo {
    public final Vec3 f_77676_;
    public final Vec3 f_77677_;
    public final float f_77678_;
    public final float f_77679_;
+   public ServerLevel level;
+   public CraftPortalEvent portalEvent;
+
 
    public PortalInfo(Vec3 p_77681_, Vec3 p_77682_, float p_77683_, float p_77684_) {
       this.f_77676_ = p_77681_;
       this.f_77677_ = p_77682_;
       this.f_77678_ = p_77683_;
       this.f_77679_ = p_77684_;
+   }
+
+   public PortalInfo(Vec3 p_77681_, Vec3 p_77682_, float p_77683_, float p_77684_, CraftPortalEvent event, ServerLevel level){
+      this(p_77681_, p_77682_, p_77683_, p_77684_);
+      this.portalEvent = event;
+      this.level = level;
    }
 }
