--- a/net/minecraft/world/level/portal/PortalShape.java
+++ b/net/minecraft/world/level/portal/PortalShape.java
@@ -22,13 +_,18 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftPortalEvent;
+import org.bukkit.event.world.PortalCreateEvent;
+// CraftBukkit end
+
 public class PortalShape {
    private static final int f_164752_ = 2;
    public static final int f_164750_ = 21;
    private static final int f_164753_ = 3;
    public static final int f_164751_ = 21;
    private static final BlockBehaviour.StatePredicate f_77685_ = (p_77720_, p_77721_, p_77722_) -> {
-      return p_77720_.m_60713_(Blocks.f_50080_);
+      return p_77720_.isPortalFrame(p_77721_, p_77722_);
    };
    private static final float f_256985_ = 4.0F;
    private static final double f_256802_ = 1.0D;
@@ -40,6 +_,7 @@
    private BlockPos f_77690_;
    private int f_77691_;
    private final int f_77692_;
+   org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator blocks; // CraftBukkit - add field
 
    public static Optional<PortalShape> m_77708_(LevelAccessor p_77709_, BlockPos p_77710_, Direction.Axis p_77711_) {
       return m_77712_(p_77709_, p_77710_, (p_77727_) -> {
@@ -58,6 +_,7 @@
    }
 
    public PortalShape(LevelAccessor p_77695_, BlockPos p_77696_, Direction.Axis p_77697_) {
+      blocks = new org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator(p_77695_.getMinecraftWorld()); // CraftBukkit
       this.f_77686_ = p_77695_;
       this.f_77687_ = p_77697_;
       this.f_77688_ = p_77697_ == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
@@ -98,6 +_,7 @@
          BlockState blockstate = this.f_77686_.m_8055_(blockpos$mutableblockpos);
          if (!m_77717_(blockstate)) {
             if (f_77685_.m_61035_(blockstate, this.f_77686_, blockpos$mutableblockpos)) {
+               blocks.m_7731_(blockpos$mutableblockpos, blockstate, 18); // CraftBukkit - lower left / right
                return i;
             }
             break;
@@ -107,6 +_,7 @@
          if (!f_77685_.m_61035_(blockstate1, this.f_77686_, blockpos$mutableblockpos)) {
             break;
          }
+         blocks.m_7731_(blockpos$mutableblockpos, blockstate1, 18); // CraftBukkit - bottom row
       }
 
       return 0;
@@ -124,6 +_,7 @@
          if (!f_77685_.m_61035_(this.f_77686_.m_8055_(blockpos$mutableblockpos), this.f_77686_, blockpos$mutableblockpos)) {
             return false;
          }
+         blocks.m_7731_(blockpos$mutableblockpos, this.f_77686_.m_8055_(blockpos$mutableblockpos), 18); // CraftBukkit - upper row
       }
 
       return true;
@@ -152,6 +_,10 @@
                ++this.f_77689_;
             }
          }
+         // CraftBukkit start - left and right
+         blocks.m_7731_(p_77729_.m_122190_(this.f_77690_).m_122175_(Direction.UP, i).m_122175_(this.f_77688_, -1), this.f_77686_.m_8055_(p_77729_), 18);
+         blocks.m_7731_(p_77729_.m_122190_(this.f_77690_).m_122175_(Direction.UP, i).m_122175_(this.f_77688_, this.f_77692_), this.f_77686_.m_8055_(p_77729_), 18);
+         // CraftBukkit end
       }
 
       return 21;
@@ -165,11 +_,33 @@
       return this.f_77690_ != null && this.f_77692_ >= 2 && this.f_77692_ <= 21 && this.f_77691_ >= 3 && this.f_77691_ <= 21;
    }
 
+   // CatServer start
    public void m_77743_() {
+      this.createPortalBlockCB();
+   }
+   // CatServer end
+
+   // CraftBukkit start - return boolean
+   public boolean createPortalBlockCB() {
+      org.bukkit.World bworld = this.f_77686_.getMinecraftWorld().getWorld();
+
+      // Copy below for loop
       BlockState blockstate = Blocks.f_50142_.m_49966_().m_61124_(NetherPortalBlock.f_54904_, this.f_77687_);
       BlockPos.m_121940_(this.f_77690_, this.f_77690_.m_5484_(Direction.UP, this.f_77691_ - 1).m_5484_(this.f_77688_, this.f_77692_ - 1)).forEach((p_77725_) -> {
+         blocks.m_7731_(p_77725_, blockstate, 18);
+      } );
+
+      PortalCreateEvent event = new PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks.getList(), bworld, null, PortalCreateEvent.CreateReason.FIRE);
+      this.f_77686_.getMinecraftWorld().m_7654_().server.getPluginManager().callEvent(event);
+
+      if (event.isCancelled()) {
+         return false;
+      }
+      // CraftBukkit end
+      BlockPos.m_121940_(this.f_77690_, this.f_77690_.m_5484_(Direction.UP, this.f_77691_ - 1).m_5484_(this.f_77688_, this.f_77692_ - 1)).forEach((p_77725_) -> {
          this.f_77686_.m_7731_(p_77725_, blockstate, 18);
       });
+      return true; // CraftBukkit
    }
 
    public boolean m_77744_() {
@@ -199,6 +_,10 @@
       Direction.Axis direction$axis1 = p_77740_ == Direction.Axis.X ? Direction.Axis.Z : Direction.Axis.X;
       double d3 = p_77741_.m_82507_(direction$axis1) - ((double)blockpos.m_123304_(direction$axis1) + 0.5D);
       return new Vec3(d2, d4, d3);
+   }
+
+   public static PortalInfo createPortalInfo(ServerLevel p_259301_, BlockUtil.FoundRectangle p_259931_, Direction.Axis p_259901_, Vec3 p_259630_, Entity p_259166_, Vec3 p_260043_, float p_259853_, float p_259667_, CraftPortalEvent portalEventInfo) { // CraftBukkit
+      return PortalShape.m_257966_(p_259301_, p_259931_, p_259901_, p_259630_, p_259166_, p_260043_, p_259853_, p_259667_).putCB(p_259301_, portalEventInfo); // CatServer
    }
 
    public static PortalInfo m_257966_(ServerLevel p_259301_, BlockUtil.FoundRectangle p_259931_, Direction.Axis p_259901_, Vec3 p_259630_, Entity p_259166_, Vec3 p_260043_, float p_259853_, float p_259667_) {
