--- a/net/minecraft/world/level/portal/PortalForcer.java
+++ b/net/minecraft/world/level/portal/PortalForcer.java
@@ -2,6 +_,7 @@
 
 import java.util.Comparator;
 import java.util.Optional;
+
 import net.minecraft.BlockUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -19,163 +_,193 @@
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.levelgen.Heightmap;
 
-public class PortalForcer {
-   private static final int f_164734_ = 3;
-   private static final int f_164735_ = 128;
-   private static final int f_164736_ = 16;
-   private static final int f_164737_ = 5;
-   private static final int f_164738_ = 4;
-   private static final int f_164739_ = 3;
-   private static final int f_164740_ = -1;
-   private static final int f_164741_ = 4;
-   private static final int f_164742_ = -1;
-   private static final int f_164743_ = 3;
-   private static final int f_164744_ = -1;
-   private static final int f_164745_ = 2;
-   private static final int f_164746_ = -1;
-   protected final ServerLevel f_77648_;
-
-   public PortalForcer(ServerLevel p_77650_) {
-      this.f_77648_ = p_77650_;
-   }
-
-   public Optional<BlockUtil.FoundRectangle> m_192985_(BlockPos p_192986_, boolean p_192987_, WorldBorder p_192988_) {
-      PoiManager poimanager = this.f_77648_.m_8904_();
-      int i = p_192987_ ? 16 : 128;
-      poimanager.m_27056_(this.f_77648_, p_192986_, i);
-      Optional<PoiRecord> optional = poimanager.m_27166_((p_77654_) -> {
-         return p_77654_ == PoiType.f_27350_;
-      }, p_192986_, i, PoiManager.Occupancy.ANY).filter((p_192981_) -> {
-         return p_192988_.m_61937_(p_192981_.m_27257_());
-      }).sorted(Comparator.<PoiRecord>comparingDouble((p_192984_) -> {
-         return p_192984_.m_27257_().m_123331_(p_192986_);
-      }).thenComparingInt((p_192992_) -> {
-         return p_192992_.m_27257_().m_123342_();
-      })).filter((p_192990_) -> {
-         return this.f_77648_.m_8055_(p_192990_.m_27257_()).m_61138_(BlockStateProperties.f_61364_);
-      }).findFirst();
-      return optional.map((p_192975_) -> {
-         BlockPos blockpos = p_192975_.m_27257_();
-         this.f_77648_.m_7726_().m_8387_(TicketType.f_9447_, new ChunkPos(blockpos), 3, blockpos);
-         BlockState blockstate = this.f_77648_.m_8055_(blockpos);
-         return BlockUtil.m_124334_(blockpos, blockstate.m_61143_(BlockStateProperties.f_61364_), 21, Direction.Axis.Y, 21, (p_192978_) -> {
-            return this.f_77648_.m_8055_(p_192978_) == blockstate;
-         });
-      });
-   }
-
-   public Optional<BlockUtil.FoundRectangle> m_77666_(BlockPos p_77667_, Direction.Axis p_77668_) {
-      Direction direction = Direction.m_122390_(Direction.AxisDirection.POSITIVE, p_77668_);
-      double d0 = -1.0D;
-      BlockPos blockpos = null;
-      double d1 = -1.0D;
-      BlockPos blockpos1 = null;
-      WorldBorder worldborder = this.f_77648_.m_6857_();
-      int i = Math.min(this.f_77648_.m_151558_(), this.f_77648_.m_141937_() + this.f_77648_.m_143344_()) - 1;
-      BlockPos.MutableBlockPos blockpos$mutableblockpos = p_77667_.m_122032_();
-
-      for(BlockPos.MutableBlockPos blockpos$mutableblockpos1 : BlockPos.m_121935_(p_77667_, 16, Direction.EAST, Direction.SOUTH)) {
-         int j = Math.min(i, this.f_77648_.m_6924_(Heightmap.Types.MOTION_BLOCKING, blockpos$mutableblockpos1.m_123341_(), blockpos$mutableblockpos1.m_123343_()));
-         int k = 1;
-         if (worldborder.m_61937_(blockpos$mutableblockpos1) && worldborder.m_61937_(blockpos$mutableblockpos1.m_122175_(direction, 1))) {
-            blockpos$mutableblockpos1.m_122175_(direction.m_122424_(), 1);
-
-            for(int l = j; l >= this.f_77648_.m_141937_(); --l) {
-               blockpos$mutableblockpos1.m_142448_(l);
-               if (this.f_77648_.m_46859_(blockpos$mutableblockpos1)) {
-                  int i1;
-                  for(i1 = l; l > this.f_77648_.m_141937_() && this.f_77648_.m_46859_(blockpos$mutableblockpos1.m_122173_(Direction.DOWN)); --l) {
-                  }
-
-                  if (l + 4 <= i) {
-                     int j1 = i1 - l;
-                     if (j1 <= 0 || j1 >= 3) {
-                        blockpos$mutableblockpos1.m_142448_(l);
-                        if (this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, 0)) {
-                           double d2 = p_77667_.m_123331_(blockpos$mutableblockpos1);
-                           if (this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, -1) && this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, 1) && (d0 == -1.0D || d0 > d2)) {
-                              d0 = d2;
-                              blockpos = blockpos$mutableblockpos1.m_7949_();
-                           }
-
-                           if (d0 == -1.0D && (d1 == -1.0D || d1 > d2)) {
-                              d1 = d2;
-                              blockpos1 = blockpos$mutableblockpos1.m_7949_();
-                           }
-                        }
-                     }
-                  }
-               }
-            }
-         }
-      }
-
-      if (d0 == -1.0D && d1 != -1.0D) {
-         blockpos = blockpos1;
-         d0 = d1;
-      }
-
-      if (d0 == -1.0D) {
-         int k1 = Math.max(this.f_77648_.m_141937_() - -1, 70);
-         int i2 = i - 9;
-         if (i2 < k1) {
-            return Optional.empty();
-         }
-
-         blockpos = (new BlockPos(p_77667_.m_123341_(), Mth.m_14045_(p_77667_.m_123342_(), k1, i2), p_77667_.m_123343_())).m_7949_();
-         Direction direction1 = direction.m_122427_();
-         if (!worldborder.m_61937_(blockpos)) {
-            return Optional.empty();
-         }
-
-         for(int i3 = -1; i3 < 2; ++i3) {
-            for(int j3 = 0; j3 < 2; ++j3) {
-               for(int k3 = -1; k3 < 3; ++k3) {
-                  BlockState blockstate1 = k3 < 0 ? Blocks.f_50080_.m_49966_() : Blocks.f_50016_.m_49966_();
-                  blockpos$mutableblockpos.m_122154_(blockpos, j3 * direction.m_122429_() + i3 * direction1.m_122429_(), k3, j3 * direction.m_122431_() + i3 * direction1.m_122431_());
-                  this.f_77648_.m_46597_(blockpos$mutableblockpos, blockstate1);
-               }
-            }
-         }
-      }
-
-      for(int l1 = -1; l1 < 3; ++l1) {
-         for(int j2 = -1; j2 < 4; ++j2) {
-            if (l1 == -1 || l1 == 2 || j2 == -1 || j2 == 3) {
-               blockpos$mutableblockpos.m_122154_(blockpos, l1 * direction.m_122429_(), j2, l1 * direction.m_122431_());
-               this.f_77648_.m_7731_(blockpos$mutableblockpos, Blocks.f_50080_.m_49966_(), 3);
-            }
-         }
-      }
-
-      BlockState blockstate = Blocks.f_50142_.m_49966_().m_61124_(NetherPortalBlock.f_54904_, p_77668_);
-
-      for(int k2 = 0; k2 < 2; ++k2) {
-         for(int l2 = 0; l2 < 3; ++l2) {
-            blockpos$mutableblockpos.m_122154_(blockpos, k2 * direction.m_122429_(), l2, k2 * direction.m_122431_());
-            this.f_77648_.m_7731_(blockpos$mutableblockpos, blockstate, 18);
-         }
-      }
-
-      return Optional.of(new BlockUtil.FoundRectangle(blockpos.m_7949_(), 2, 3));
-   }
-
-   private boolean m_77661_(BlockPos p_77662_, BlockPos.MutableBlockPos p_77663_, Direction p_77664_, int p_77665_) {
-      Direction direction = p_77664_.m_122427_();
-
-      for(int i = -1; i < 3; ++i) {
-         for(int j = -1; j < 4; ++j) {
-            p_77663_.m_122154_(p_77662_, p_77664_.m_122429_() * i + direction.m_122429_() * p_77665_, j, p_77664_.m_122431_() * i + direction.m_122431_() * p_77665_);
-            if (j < 0 && !this.f_77648_.m_8055_(p_77663_).m_60767_().m_76333_()) {
-               return false;
-            }
-
-            if (j >= 0 && !this.f_77648_.m_46859_(p_77663_)) {
-               return false;
-            }
-         }
-      }
-
-      return true;
-   }
+public class PortalForcer implements net.minecraftforge.common.util.ITeleporter {
+    private static final int f_164734_ = 3;
+    private static final int f_164735_ = 128;
+    private static final int f_164736_ = 16;
+    private static final int f_164737_ = 5;
+    private static final int f_164738_ = 4;
+    private static final int f_164739_ = 3;
+    private static final int f_164740_ = -1;
+    private static final int f_164741_ = 4;
+    private static final int f_164742_ = -1;
+    private static final int f_164743_ = 3;
+    private static final int f_164744_ = -1;
+    private static final int f_164745_ = 2;
+    private static final int f_164746_ = -1;
+    protected final ServerLevel f_77648_;
+
+    public PortalForcer(ServerLevel p_77650_) {
+        this.f_77648_ = p_77650_;
+    }
+
+    public Optional<BlockUtil.FoundRectangle> m_192985_(BlockPos p_192986_, boolean p_192987_, WorldBorder p_192988_) {
+        //CraftBukkit start
+        return findPortalAround(p_192986_, p_192988_, p_192987_ ? 16 : 128);
+    }
+
+    public Optional<BlockUtil.FoundRectangle> findPortalAround(BlockPos p_192986_, WorldBorder p_192988_, int i) {
+        PoiManager poimanager = this.f_77648_.m_8904_();
+        //CraftBukkit end
+
+        poimanager.m_27056_(this.f_77648_, p_192986_, i);
+        Optional<PoiRecord> optional = poimanager.m_27166_((p_77654_) -> {
+            return p_77654_ == PoiType.f_27350_;
+        }, p_192986_, i, PoiManager.Occupancy.ANY).filter((p_192981_) -> {
+            return p_192988_.m_61937_(p_192981_.m_27257_());
+        }).sorted(Comparator.<PoiRecord>comparingDouble((p_192984_) -> {
+            return p_192984_.m_27257_().m_123331_(p_192986_);
+        }).thenComparingInt((p_192992_) -> {
+            return p_192992_.m_27257_().m_123342_();
+        })).filter((p_192990_) -> {
+            return this.f_77648_.m_8055_(p_192990_.m_27257_()).m_61138_(BlockStateProperties.f_61364_);
+        }).findFirst();
+        return optional.map((p_192975_) -> {
+            BlockPos blockpos = p_192975_.m_27257_();
+            this.f_77648_.m_7726_().m_8387_(TicketType.f_9447_, new ChunkPos(blockpos), 3, blockpos);
+            BlockState blockstate = this.f_77648_.m_8055_(blockpos);
+            return BlockUtil.m_124334_(blockpos, blockstate.m_61143_(BlockStateProperties.f_61364_), 21, Direction.Axis.Y, 21, (p_192978_) -> {
+                return this.f_77648_.m_8055_(p_192978_) == blockstate;
+            });
+        });
+    }
+
+    public Optional<BlockUtil.FoundRectangle> createPortal(BlockPos p_77667_, Direction.Axis p_77668_, net.minecraft.world.entity.Entity entity, int createRadius) {
+        // CatServer start
+        catserver.server.CatServerCaptures.getCatServerCaptures().captureEntity(entity);
+        catserver.server.CatServerCaptures.getCatServerCaptures().capturePortalRadius(createRadius);
+        try {
+            return m_77666_(p_77667_, p_77668_);
+        } finally {
+            // CatServer - Make sure reset
+            catserver.server.CatServerCaptures.getCatServerCaptures().getCaptureEntity();
+            catserver.server.CatServerCaptures.getCatServerCaptures().getCapturePortalRadius();
+        }
+        // CatServer end
+    }
+
+    public Optional<BlockUtil.FoundRectangle> m_77666_(BlockPos p_77667_, Direction.Axis p_77668_) {
+        Direction direction = Direction.m_122390_(Direction.AxisDirection.POSITIVE, p_77668_);
+        double d0 = -1.0D;
+        BlockPos blockpos = null;
+        double d1 = -1.0D;
+        BlockPos blockpos1 = null;
+        WorldBorder worldborder = this.f_77648_.m_6857_();
+        int i = Math.min(this.f_77648_.m_151558_(), this.f_77648_.m_141937_() + this.f_77648_.m_143344_()) - 1;
+        BlockPos.MutableBlockPos blockpos$mutableblockpos = p_77667_.m_122032_();
+
+        for (BlockPos.MutableBlockPos blockpos$mutableblockpos1 : BlockPos.m_121935_(p_77667_, catserver.server.CatServerCaptures.getCatServerCaptures().getCapturePortalRadius(), Direction.EAST, Direction.SOUTH)) { // CraftBukkit // CatServer
+            int j = Math.min(i, this.f_77648_.m_6924_(Heightmap.Types.MOTION_BLOCKING, blockpos$mutableblockpos1.m_123341_(), blockpos$mutableblockpos1.m_123343_()));
+            int k = 1;
+            if (worldborder.m_61937_(blockpos$mutableblockpos1) && worldborder.m_61937_(blockpos$mutableblockpos1.m_122175_(direction, 1))) {
+                blockpos$mutableblockpos1.m_122175_(direction.m_122424_(), 1);
+
+                for (int l = j; l >= this.f_77648_.m_141937_(); --l) {
+                    blockpos$mutableblockpos1.m_142448_(l);
+                    if (this.f_77648_.m_46859_(blockpos$mutableblockpos1)) {
+                        int i1;
+                        for (i1 = l; l > this.f_77648_.m_141937_() && this.f_77648_.m_46859_(blockpos$mutableblockpos1.m_122173_(Direction.DOWN)); --l) {
+                        }
+
+                        if (l + 4 <= i) {
+                            int j1 = i1 - l;
+                            if (j1 <= 0 || j1 >= 3) {
+                                blockpos$mutableblockpos1.m_142448_(l);
+                                if (this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, 0)) {
+                                    double d2 = p_77667_.m_123331_(blockpos$mutableblockpos1);
+                                    if (this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, -1) && this.m_77661_(blockpos$mutableblockpos1, blockpos$mutableblockpos, direction, 1) && (d0 == -1.0D || d0 > d2)) {
+                                        d0 = d2;
+                                        blockpos = blockpos$mutableblockpos1.m_7949_();
+                                    }
+
+                                    if (d0 == -1.0D && (d1 == -1.0D || d1 > d2)) {
+                                        d1 = d2;
+                                        blockpos1 = blockpos$mutableblockpos1.m_7949_();
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (d0 == -1.0D && d1 != -1.0D) {
+            blockpos = blockpos1;
+            d0 = d1;
+        }
+
+        org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator(this.f_77648_); // CraftBukkit - Use BlockStateListPopulator
+        if (d0 == -1.0D) {
+            int k1 = Math.max(this.f_77648_.m_141937_() - -1, 70);
+            int i2 = i - 9;
+            if (i2 < k1) {
+                return Optional.empty();
+            }
+
+            blockpos = (new BlockPos(p_77667_.m_123341_(), Mth.m_14045_(p_77667_.m_123342_(), k1, i2), p_77667_.m_123343_())).m_7949_();
+            Direction direction1 = direction.m_122427_();
+            if (!worldborder.m_61937_(blockpos)) {
+                return Optional.empty();
+            }
+
+            for (int i3 = -1; i3 < 2; ++i3) {
+                for (int j3 = 0; j3 < 2; ++j3) {
+                    for (int k3 = -1; k3 < 3; ++k3) {
+                        BlockState blockstate1 = k3 < 0 ? Blocks.f_50080_.m_49966_() : Blocks.f_50016_.m_49966_();
+                        blockpos$mutableblockpos.m_122154_(blockpos, j3 * direction.m_122429_() + i3 * direction1.m_122429_(), k3, j3 * direction.m_122431_() + i3 * direction1.m_122431_());
+                        blockList.m_7731_(blockpos$mutableblockpos, blockstate1, 3); // CraftBukkit
+                    }
+                }
+            }
+        }
+
+        for (int l1 = -1; l1 < 3; ++l1) {
+            for (int j2 = -1; j2 < 4; ++j2) {
+                if (l1 == -1 || l1 == 2 || j2 == -1 || j2 == 3) {
+                    blockpos$mutableblockpos.m_122154_(blockpos, l1 * direction.m_122429_(), j2, l1 * direction.m_122431_());
+                    blockList.m_7731_(blockpos$mutableblockpos, Blocks.f_50080_.m_49966_(), 3); // CraftBukkit
+                }
+            }
+        }
+
+        BlockState blockstate = Blocks.f_50142_.m_49966_().m_61124_(NetherPortalBlock.f_54904_, p_77668_);
+
+        for (int k2 = 0; k2 < 2; ++k2) {
+            for (int l2 = 0; l2 < 3; ++l2) {
+                blockpos$mutableblockpos.m_122154_(blockpos, k2 * direction.m_122429_(), l2, k2 * direction.m_122431_());
+                blockList.m_7731_(blockpos$mutableblockpos, blockstate, 18); // CraftBukkit
+            }
+        }
+        var entity = catserver.server.CatServerCaptures.getCatServerCaptures().getCaptureEntity(); // CatServer
+        // CraftBukkit start
+        org.bukkit.World bworld = this.f_77648_.getWorld();
+        org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blockList.getList(), bworld, (entity == null) ? null : entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.NETHER_PAIR);
+        this.f_77648_.getCraftServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return Optional.empty();
+        }
+        blockList.updateList();
+        // CraftBukkit end
+        return Optional.of(new BlockUtil.FoundRectangle(blockpos.m_7949_(), 2, 3));
+    }
+
+    private boolean m_77661_(BlockPos p_77662_, BlockPos.MutableBlockPos p_77663_, Direction p_77664_, int p_77665_) {
+        Direction direction = p_77664_.m_122427_();
+
+        for (int i = -1; i < 3; ++i) {
+            for (int j = -1; j < 4; ++j) {
+                p_77663_.m_122154_(p_77662_, p_77664_.m_122429_() * i + direction.m_122429_() * p_77665_, j, p_77664_.m_122431_() * i + direction.m_122431_() * p_77665_);
+                if (j < 0 && !this.f_77648_.m_8055_(p_77663_).m_60767_().m_76333_()) {
+                    return false;
+                }
+
+                if (j >= 0 && !this.f_77648_.m_46859_(p_77663_)) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
 }
