--- a/net/minecraft/world/level/portal/PortalInfo.java
+++ b/net/minecraft/world/level/portal/PortalInfo.java
@@ -1,12 +_,16 @@
 package net.minecraft.world.level.portal;
 
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent;
 
 public class PortalInfo {
    public final Vec3 f_77676_;
    public final Vec3 f_77677_;
    public final float f_77678_;
    public final float f_77679_;
+   public ServerLevel world;
+   public CraftPortalEvent portalEventInfo;
 
    public PortalInfo(Vec3 p_77681_, Vec3 p_77682_, float p_77683_, float p_77684_) {
       this.f_77676_ = p_77681_;
@@ -14,4 +_,12 @@
       this.f_77678_ = p_77683_;
       this.f_77679_ = p_77684_;
    }
+
+   // CatServer start
+   public PortalInfo putCB(ServerLevel level, CraftPortalEvent event) {
+      this.world = level;
+      this.portalEventInfo = event;
+      return this;
+   }
+   // CatServer end
 }
