--- a/net/minecraft/world/level/portal/PortalShape.java
+++ b/net/minecraft/world/level/portal/PortalShape.java
@@ -17,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent;
+import org.bukkit.event.world.PortalCreateEvent;
 
 public class PortalShape {
    private static final int f_164752_ = 2;
@@ -24,7 +_,7 @@
    private static final int f_164753_ = 3;
    public static final int f_164751_ = 21;
    private static final BlockBehaviour.StatePredicate f_77685_ = (p_77720_, p_77721_, p_77722_) -> {
-      return p_77720_.m_60713_(Blocks.f_50080_);
+      return p_77720_.isPortalFrame(p_77721_, p_77722_);
    };
    private final LevelAccessor f_77686_;
    private final Direction.Axis f_77687_;
@@ -34,6 +_,7 @@
    private BlockPos f_77690_;
    private int f_77691_;
    private final int f_77692_;
+   org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator blocks; // CraftBukkit - add field
 
    public static Optional<PortalShape> m_77708_(LevelAccessor p_77709_, BlockPos p_77710_, Direction.Axis p_77711_) {
       return m_77712_(p_77709_, p_77710_, (p_77727_) -> {
@@ -52,6 +_,7 @@
    }
 
    public PortalShape(LevelAccessor p_77695_, BlockPos p_77696_, Direction.Axis p_77697_) {
+      blocks = new org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator(p_77695_.getMinecraftWorld()); // CraftBukkit
       this.f_77686_ = p_77695_;
       this.f_77687_ = p_77697_;
       this.f_77688_ = p_77697_ == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
@@ -92,6 +_,7 @@
          BlockState blockstate = this.f_77686_.m_8055_(blockpos$mutableblockpos);
          if (!m_77717_(blockstate)) {
             if (f_77685_.m_61035_(blockstate, this.f_77686_, blockpos$mutableblockpos)) {
+               blocks.m_7731_(blockpos$mutableblockpos, blockstate, 18); // CraftBukkit - lower left / right
                return i;
             }
             break;
@@ -101,6 +_,7 @@
          if (!f_77685_.m_61035_(blockstate1, this.f_77686_, blockpos$mutableblockpos)) {
             break;
          }
+         blocks.m_7731_(blockpos$mutableblockpos, blockstate1, 18); // CraftBukkit - bottom row
       }
 
       return 0;
@@ -118,6 +_,7 @@
          if (!f_77685_.m_61035_(this.f_77686_.m_8055_(blockpos$mutableblockpos), this.f_77686_, blockpos$mutableblockpos)) {
             return false;
          }
+         blocks.m_7731_(blockpos$mutableblockpos, this.f_77686_.m_8055_(blockpos$mutableblockpos), 18); // CraftBukkit - upper row
       }
 
       return true;
@@ -146,6 +_,10 @@
                ++this.f_77689_;
             }
          }
+         // CraftBukkit start - left and right
+         blocks.m_7731_(p_77729_.m_122190_(this.f_77690_).m_122175_(Direction.UP, i).m_122175_(this.f_77688_, -1), this.f_77686_.m_8055_(p_77729_), 18);
+         blocks.m_7731_(p_77729_.m_122190_(this.f_77690_).m_122175_(Direction.UP, i).m_122175_(this.f_77688_, this.f_77692_), this.f_77686_.m_8055_(p_77729_), 18);
+         // CraftBukkit end
       }
 
       return 21;
@@ -159,11 +_,27 @@
       return this.f_77690_ != null && this.f_77692_ >= 2 && this.f_77692_ <= 21 && this.f_77691_ >= 3 && this.f_77691_ <= 21;
    }
 
-   public void m_77743_() {
+   // CraftBukkit start - return boolean
+   public boolean createPortalBlocks() {
+      org.bukkit.World bworld = this.f_77686_.getMinecraftWorld().getWorld();
+
+      // Copy below for loop
       BlockState blockstate = Blocks.f_50142_.m_49966_().m_61124_(NetherPortalBlock.f_54904_, this.f_77687_);
       BlockPos.m_121940_(this.f_77690_, this.f_77690_.m_5484_(Direction.UP, this.f_77691_ - 1).m_5484_(this.f_77688_, this.f_77692_ - 1)).forEach((p_77725_) -> {
-         this.f_77686_.m_7731_(p_77725_, blockstate, 18);
-      });
+         blocks.m_7731_(p_77725_, blockstate, 18);
+      });
+
+      PortalCreateEvent event = new PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks.getList(), bworld, null, PortalCreateEvent.CreateReason.FIRE);
+      this.f_77686_.getMinecraftWorld().m_142572_().server.getPluginManager().callEvent(event);
+
+      if (event.isCancelled()) {
+         return false;
+      }
+      // CraftBukkit end
+      BlockPos.m_121940_(this.f_77690_, this.f_77690_.m_5484_(Direction.UP, this.f_77691_ - 1).m_5484_(this.f_77688_, this.f_77692_ - 1)).forEach((blockposition) -> {
+            this.f_77686_.m_7731_(blockposition, blockstate, 18);
+      });
+      return true; // CraftBukkit
    }
 
    public boolean m_77744_() {
@@ -195,7 +_,12 @@
       return new Vec3(d2, d4, d3);
    }
 
+   // FoxServer TODO
    public static PortalInfo m_77699_(ServerLevel p_77700_, BlockUtil.FoundRectangle p_77701_, Direction.Axis p_77702_, Vec3 p_77703_, EntityDimensions p_77704_, Vec3 p_77705_, float p_77706_, float p_77707_) {
+      return createPortalInfo(p_77700_, p_77701_, p_77702_, p_77703_, p_77704_, p_77705_, p_77706_, p_77707_, null);
+   }
+   // FoxServer TODO
+   public static PortalInfo createPortalInfo(ServerLevel p_77700_, BlockUtil.FoundRectangle p_77701_, Direction.Axis p_77702_, Vec3 p_77703_, EntityDimensions p_77704_, Vec3 p_77705_, float p_77706_, float p_77707_, CraftPortalEvent event) {
       BlockPos blockpos = p_77701_.f_124348_;
       BlockState blockstate = p_77700_.m_8055_(blockpos);
       Direction.Axis direction$axis = blockstate.m_61145_(BlockStateProperties.f_61364_).orElse(Direction.Axis.X);
@@ -208,6 +_,6 @@
       double d4 = 0.5D + p_77703_.m_7094_();
       boolean flag = direction$axis == Direction.Axis.X;
       Vec3 vec31 = new Vec3((double)blockpos.m_123341_() + (flag ? d2 : d4), (double)blockpos.m_123342_() + d3, (double)blockpos.m_123343_() + (flag ? d4 : d2));
-      return new PortalInfo(vec31, vec3, p_77706_ + (float)i, p_77707_);
+      return new PortalInfo(vec31, vec3, p_77706_ + (float)i, p_77707_, event, p_77700_);
    }
 }
