--- a/net/minecraft/world/gen/surfacebuilders/ValleySurfaceBuilder.java
+++ b/net/minecraft/world/gen/surfacebuilders/ValleySurfaceBuilder.java
@@ -41,7 +_,7 @@
       BlockPos.Mutable blockpos$mutable = new BlockPos.Mutable();
       BlockState blockstate2 = p_205610_2_.func_180495_p(blockpos$mutable.func_181079_c(j, 128, k));
 
-      for(int j1 = 127; j1 >= 0; --j1) {
+      for(int j1 = p_205610_6_; j1 >= 0; --j1) {
          blockpos$mutable.func_181079_c(j, j1, k);
          BlockState blockstate3 = p_205610_2_.func_180495_p(blockpos$mutable);
          if (blockstate2.func_203425_a(p_205610_9_.func_177230_c()) && (blockstate3.func_196958_f() || blockstate3 == p_205610_10_)) {
