--- ../src-base/minecraft/net/minecraft/server/management/PlayerChunkMap.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerChunkMap.java
@@ -20,6 +20,9 @@
 import net.minecraft.world.WorldProvider;
 import net.minecraft.world.WorldServer;
 import net.minecraft.world.chunk.Chunk;
+//CraftBukkit start
+import java.util.LinkedList;
+// CraftBukkit end
 
 public class PlayerChunkMap
 {
@@ -48,12 +51,23 @@
     private long previousTotalWorldTime;
     private boolean sortMissingChunks = true;
     private boolean sortSendToPlayers = true;
+    private boolean wasNotEmpty; // CraftBukkit - add field
 
-    public PlayerChunkMap(WorldServer serverWorld)
+    public PlayerChunkMap(WorldServer serverWorld, int viewDistance)
     {
         this.theWorldServer = serverWorld;
-        this.setPlayerViewRadius(serverWorld.getMinecraftServer().getPlayerList().getViewDistance());
+        this.setPlayerViewRadius(viewDistance);
     }
+    
+    // CraftBukkit start - add method
+    public final boolean isChunkInUse(int x, int z) {
+    	PlayerChunkMapEntry pi = getEntry(x, z);
+    	if (pi != null) {
+    		return (pi.players.size() > 0);
+        }
+    	return false;
+    }
+    // CraftBukkit end
 
     public WorldServer getWorldServer()
     {
