--- ../src-base/minecraft/net/minecraft/server/management/UserList.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserList.java
@@ -59,7 +59,7 @@
     {
         this.saveFile = saveFile;
         GsonBuilder gsonbuilder = (new GsonBuilder()).setPrettyPrinting();
-        gsonbuilder.registerTypeHierarchyAdapter(UserListEntry.class, new UserList.Serializer());
+        gsonbuilder.registerTypeHierarchyAdapter(UserListEntry.class, new Serializer());
         this.gson = gsonbuilder.create();
     }
 
@@ -156,6 +156,10 @@
         return this.values;
     }
 
+    public Collection<V> getValuesCB() {
+        return this.values.values();
+    }
+
     public void writeChanges() throws IOException
     {
         Collection<V> collection = this.values.values();
