--- ../src-base/minecraft/net/minecraft/server/management/UserList.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserList.java
@@ -25,6 +25,8 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+
+import net.minecraft.server.MinecraftServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.commons.io.IOUtils;
@@ -56,7 +58,7 @@
 
     public UserList(File saveFile)
     {
-        this.saveFile = saveFile;
+        this.saveFile = new File(MinecraftServer.serverConfigDir,saveFile.getName());
         GsonBuilder gsonbuilder = (new GsonBuilder()).setPrettyPrinting();
         gsonbuilder.registerTypeHierarchyAdapter(UserListEntry.class, new UserList.Serializer());
         this.gson = gsonbuilder.create();
@@ -150,7 +152,7 @@
         return new UserListEntry((Object)null, entryData);
     }
 
-    protected Map<String, V> getValues()
+    public Map<String, V> getValues()
     {
         return this.values;
     }
@@ -234,4 +236,8 @@
             }
         }
     }
+    
+    public Collection<V> getValuesCB() {// Craftbukkit
+        return this.values.values();
+    }
 }
