--- ../src-base/minecraft/net/minecraft/server/management/UserList.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserList.java
@@ -26,6 +26,7 @@
 import java.util.List;
 import java.util.Map;
 import net.minecraft.util.JsonUtils;
+import net.minecraftforge.fml.common.FMLLog;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.commons.io.IOUtils;
@@ -59,7 +60,7 @@
     {
         this.saveFile = saveFile;
         GsonBuilder gsonbuilder = (new GsonBuilder()).setPrettyPrinting();
-        gsonbuilder.registerTypeHierarchyAdapter(UserListEntry.class, new UserList.Serializer());
+        gsonbuilder.registerTypeHierarchyAdapter(UserListEntry.class, new Serializer());
         this.gson = gsonbuilder.create();
     }
 
@@ -156,6 +157,10 @@
         return this.values;
     }
 
+    public Collection<V> getValuesCB() {
+        return this.values.values();
+    }
+
     public void writeChanges() throws IOException
     {
         Collection<V> collection = this.values.values();
@@ -191,6 +196,11 @@
             {
                 bufferedreader = Files.newReader(this.saveFile, StandardCharsets.UTF_8);
                 collection = (Collection)JsonUtils.fromJson(this.gson, bufferedreader, USER_LIST_ENTRY_TYPE);
+            }catch (Exception e) {
+                for (int i = 0; i < 100 ; i++) {
+                    FMLLog.log.info("您的服务端文件损坏，文件路径：" + this.saveFile.getCanonicalPath());
+                }
+                throw new RuntimeException("您的服务端文件损坏，文件路径：" + this.saveFile.getCanonicalPath());
             }
             finally
             {
