--- ../src-base/minecraft/net/minecraft/server/management/UserList.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserList.java
@@ -150,7 +150,7 @@
         return new UserListEntry((Object)null, entryData);
     }
 
-    protected Map<String, V> getValues()
+    public Map<String, V> getValues()
     {
         return this.values;
     }
@@ -234,4 +234,8 @@
             }
         }
     }
+    
+    public Collection<V> getValuesCB() {// Craftbukkit
+        return this.values.values();
+    }
 }
