--- ../src-base/minecraft/net/minecraft/server/management/UserListEntryBan.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserListEntryBan.java
@@ -24,7 +24,7 @@
 
     protected UserListEntryBan(T valueIn, JsonObject json)
     {
-        super(valueIn, json);
+        super(checkExpiry(valueIn, json), json);
         Date date;
 
         try
@@ -75,4 +75,28 @@
         data.addProperty("expires", this.banEndDate == null ? "forever" : DATE_FORMAT.format(this.banEndDate));
         data.addProperty("reason", this.reason);
     }
+
+    public String getSource() {
+        return this.bannedBy;
+    }
+
+    public Date getCreated() {
+        return this.banStartDate;
+    }
+
+    private static <T> T checkExpiry(T object, JsonObject jsonobject) {
+        Date expires = null;
+
+        try {
+            expires = jsonobject.has("expires") ? DATE_FORMAT.parse(jsonobject.get("expires").getAsString()) : null;
+        } catch (ParseException ex) {
+            // Guess we don't have a date
+        }
+
+        if (expires == null || expires.after(new Date())) {
+            return object;
+        } else {
+            return null;
+        }
+    }
 }
