--- ../src-base/minecraft/net/minecraft/server/management/UserListEntryBan.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserListEntryBan.java
@@ -24,7 +24,7 @@
 
     protected UserListEntryBan(T valueIn, JsonObject json)
     {
-        super(valueIn, json);
+        super(checkExpiry(valueIn, json), json);// Craftbukkit
         Date date;
 
         try
@@ -75,4 +75,24 @@
         data.addProperty("expires", this.banEndDate == null ? "forever" : DATE_FORMAT.format(this.banEndDate));
         data.addProperty("reason", this.reason);
     }
+    
+    public String getSource() {
+        return this.bannedBy;
+    }
+    
+    public Date getCreated() {
+        return this.banStartDate;
+    }
+    
+    private static <T> T checkExpiry(final T object, final JsonObject jsonobject) {
+        Date expires = null;
+        try {
+            expires = (jsonobject.has("expires") ? UserListEntryBan.DATE_FORMAT.parse(jsonobject.get("expires").getAsString()) : null);
+        }
+        catch (ParseException ex) {}
+        if (expires == null || expires.after(new Date())) {
+            return object;
+        }
+        return null;
+    }
 }
