--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -22,6 +_,7 @@
 import net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket;
 import net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.network.ServerPlayerConnection;
 import net.minecraft.util.Mth;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.entity.Entity;
@@ -36,6 +_,8 @@
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerVelocityEvent;
 
 public class ServerEntity {
    private static final Logger f_8508_ = LogUtils.getLogger();
@@ -57,8 +_,12 @@
    private List<Entity> f_8523_ = Collections.emptyList();
    private boolean f_8524_;
    private boolean f_8525_;
+   // CraftBukkit start
+   private final Set<ServerPlayerConnection> trackedPlayers;
 
-   public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_) {
+   public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_, Set< ServerPlayerConnection > trackedPlayers) {
+      this.trackedPlayers = trackedPlayers;
+      // CraftBukkit end
       this.f_8509_ = p_8528_;
       this.f_8513_ = p_8532_;
       this.f_8510_ = p_8529_;
@@ -75,24 +_,22 @@
       List<Entity> list = this.f_8510_.m_20197_();
       if (!list.equals(this.f_8523_)) {
          this.f_8523_ = list;
-         this.f_8513_.accept(new ClientboundSetPassengersPacket(this.f_8510_));
+         this.m_8538_(new ClientboundSetPassengersPacket(this.f_8510_)); // CraftBukkit
       }
 
-      if (this.f_8510_ instanceof ItemFrame && this.f_8521_ % 10 == 0) {
+      if (this.f_8510_ instanceof ItemFrame /*&& this.tickCounter % 10 == 0*/) { // CraftBukkit - Moved below, should always enter this block
          ItemFrame itemframe = (ItemFrame)this.f_8510_;
          ItemStack itemstack = itemframe.m_31822_();
-         if (itemstack.m_41720_() instanceof MapItem) {
-            Integer integer = MapItem.m_151131_(itemstack);
-            MapItemSavedData mapitemsaveddata = MapItem.m_151128_(integer, this.f_8509_);
-            if (mapitemsaveddata != null) {
-               for(ServerPlayer serverplayer : this.f_8509_.m_6907_()) {
-                  mapitemsaveddata.m_77918_(serverplayer, itemstack);
-                  Packet<?> packet = mapitemsaveddata.m_164796_(integer, serverplayer);
+         Integer integer = MapItem.m_151131_(itemstack);
+         MapItemSavedData mapitemsaveddata = MapItem.m_42853_(itemstack, this.f_8509_);
+         if (mapitemsaveddata != null) {
+               for(ServerPlayerConnection serverplayer : this.trackedPlayers) {
+                  mapitemsaveddata.m_77918_(serverplayer.m_142253_(), itemstack);
+                  Packet<?> packet = mapitemsaveddata.m_164796_(integer, serverplayer.m_142253_());
                   if (packet != null) {
-                     serverplayer.f_8906_.m_141995_(packet);
+                     serverplayer.m_142253_().f_8906_.m_141995_(packet);
                   }
                }
-            }
          }
 
          this.m_8543_();
@@ -121,6 +_,18 @@
             Packet<?> packet1 = null;
             boolean flag3 = flag2 || this.f_8521_ % 60 == 0;
             boolean flag4 = Math.abs(l - this.f_8517_) >= 1 || Math.abs(k1 - this.f_8518_) >= 1;
+
+            // CraftBukkit start - Code moved from below
+            if (flag3) {
+               this.m_8544_();
+            }
+
+            if (flag4) {
+               this.f_8517_ = l;
+               this.f_8518_ = k1;
+            }
+            // CraftBukkit end
+
             if (this.f_8521_ > 0 || this.f_8510_ instanceof AbstractArrow) {
                long i = ClientboundMoveEntityPacket.m_132511_(vec3.f_82479_);
                long j = ClientboundMoveEntityPacket.m_132511_(vec3.f_82480_);
@@ -157,14 +_,16 @@
             }
 
             this.m_8543_();
+            /* CraftBukkit start - Code moved up
             if (flag3) {
-               this.m_8544_();
+               this.updateSentPos();
             }
 
             if (flag4) {
-               this.f_8517_ = l;
-               this.f_8518_ = k1;
+               this.yRotp = l;
+               this.xRotp = k1;
             }
+            // CraftBukkit end */
 
             this.f_8524_ = false;
          }
@@ -180,7 +_,27 @@
 
       ++this.f_8521_;
       if (this.f_8510_.f_19864_) {
-         this.m_8538_(new ClientboundSetEntityMotionPacket(this.f_8510_));
+         // CraftBukkit start - Create PlayerVelocity event
+         boolean cancelled = false;
+
+         if (this.f_8510_ instanceof ServerPlayer) {
+            Player player = (Player) this.f_8510_.getBukkitEntity();
+            org.bukkit.util.Vector velocity = player.getVelocity();
+
+            PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+            this.f_8510_.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               cancelled = true;
+            } else if (!velocity.equals(event.getVelocity())) {
+               player.setVelocity(event.getVelocity());
+            }
+         }
+
+         if (!cancelled) {
+            this.m_8538_(new ClientboundSetEntityMotionPacket(this.f_8510_));
+         }
+         // CraftBukkit end
          this.f_8510_.f_19864_ = false;
       }
 
@@ -189,16 +_,21 @@
    public void m_8534_(ServerPlayer p_8535_) {
       this.f_8510_.m_6452_(p_8535_);
       p_8535_.f_8906_.m_141995_(new ClientboundRemoveEntitiesPacket(this.f_8510_.m_142049_()));
+      net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(this.f_8510_, p_8535_);
    }
 
    public void m_8541_(ServerPlayer p_8542_) {
-      this.m_8536_(p_8542_.f_8906_::m_141995_);
+      this.sendPairingData(p_8542_.f_8906_::m_141995_, p_8542_); // CraftBukkit - add player
       this.f_8510_.m_6457_(p_8542_);
+      net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(this.f_8510_, p_8542_);
    }
 
-   public void m_8536_(Consumer<Packet<?>> p_8537_) {
+    public void sendPairingData(Consumer<Packet<?>> p_8537_, ServerPlayer entityplayer) { // CraftBukkit - add player
       if (this.f_8510_.m_146910_()) {
-         f_8508_.warn("Fetching packet for removed entity {}", (Object)this.f_8510_);
+         // CraftBukkit start - Remove useless error spam, just return
+         // LOGGER.warn("Fetching packet for removed entity {}", (Object)this.entity);
+         return;
+         // CraftBukkit end
       }
 
       Packet<?> packet = this.f_8510_.m_5654_();
@@ -211,6 +_,12 @@
       boolean flag = this.f_8512_;
       if (this.f_8510_ instanceof LivingEntity) {
          Collection<AttributeInstance> collection = ((LivingEntity)this.f_8510_).m_21204_().m_22170_();
+
+         // CraftBukkit start - If sending own attributes send scaled health instead of current maximum health
+         if (this.f_8510_.m_142049_() == entityplayer.m_142049_()) {
+            ((ServerPlayer) this.f_8510_).getBukkitEntity().injectScaledMaxHealth(collection, false);
+         }
+         // CraftBukkit end
          if (!collection.isEmpty()) {
             p_8537_.accept(new ClientboundUpdateAttributesPacket(this.f_8510_.m_142049_(), collection));
          }
@@ -238,7 +_,13 @@
          if (!list.isEmpty()) {
             p_8537_.accept(new ClientboundSetEquipmentPacket(this.f_8510_.m_142049_(), list));
          }
+         ((LivingEntity) this.f_8510_).m_21315_(); // CraftBukkit - SPIGOT-3789: sync again immediately after sending
       }
+
+       // CraftBukkit start - Fix for nonsensical head yaw
+       this.f_8519_ = Mth.m_14143_(this.f_8510_.m_6080_() * 256.0F / 360.0F);
+       p_8537_.accept(new ClientboundRotateHeadPacket(this.f_8510_, (byte) f_8519_));
+       // CraftBukkit end
 
       if (this.f_8510_ instanceof LivingEntity) {
          LivingEntity livingentity = (LivingEntity)this.f_8510_;
@@ -274,6 +_,11 @@
       if (this.f_8510_ instanceof LivingEntity) {
          Set<AttributeInstance> set = ((LivingEntity)this.f_8510_).m_21204_().m_22145_();
          if (!set.isEmpty()) {
+            // CraftBukkit start - Send scaled max health
+            if (this.f_8510_ instanceof ServerPlayer) {
+               ((ServerPlayer) this.f_8510_).getBukkitEntity().injectScaledMaxHealth(set, false);
+            }
+            // CraftBukkit end
             this.m_8538_(new ClientboundUpdateAttributesPacket(this.f_8510_.m_142049_(), set));
          }
 
