--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -18,14 +_,7 @@
 import java.io.Writer;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.ArrayList;
-import java.util.Comparator;
-import java.util.List;
-import java.util.Locale;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 import java.util.concurrent.Executor;
 import java.util.function.BooleanSupplier;
 import java.util.function.Function;
@@ -157,6 +_,28 @@
 import net.minecraft.world.ticks.LevelTicks;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.world.level.biome.BiomeSource;
+import net.minecraft.world.level.dimension.LevelStem;
+import net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator;
+import net.minecraft.world.level.levelgen.FlatLevelSource;
+import net.minecraft.world.level.storage.PrimaryLevelData;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.v1_20_R1.SpigotTimings;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R1.generator.CustomWorldChunkManager;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+import org.bukkit.craftbukkit.v1_20_R1.util.WorldUUID;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.server.MapInitializeEvent;
+import org.bukkit.event.weather.LightningStrikeEvent;
+import org.bukkit.event.world.GenericGameEvent;
+import org.bukkit.event.world.TimeSkipEvent;
+// CraftBukkit end
+
+
 public class ServerLevel extends Level implements WorldGenLevel {
    public static final BlockPos f_8562_ = new BlockPos(100, 50, 0);
    public static final IntProvider f_263704_ = UniformInt.m_146622_(12000, 180000);
@@ -169,9 +_,9 @@
    final List<ServerPlayer> f_8546_ = Lists.newArrayList();
    private final ServerChunkCache f_8547_;
    private final MinecraftServer f_8548_;
-   private final ServerLevelData f_8549_;
+   public final ServerLevelData f_8549_; // CatServer - private -> public
    final EntityTickList f_143243_ = new EntityTickList();
-   private final PersistentEntitySectionManager<Entity> f_143244_;
+   public final PersistentEntitySectionManager<Entity> f_143244_; // CatServer - private -> public
    private final GameEventDispatcher f_243695_;
    public boolean f_8564_;
    private final SleepStatus f_143245_;
@@ -188,58 +_,212 @@
    private final List<CustomSpawner> f_8558_;
    @Nullable
    private EndDragonFight f_8559_;
-   final Int2ObjectMap<EnderDragonPart> f_143247_ = new Int2ObjectOpenHashMap<>();
+   final Int2ObjectMap<net.minecraftforge.entity.PartEntity<?>> f_143247_ = new Int2ObjectOpenHashMap<>();
    private final StructureManager f_214997_;
    private final StructureCheck f_196556_;
    private final boolean f_8561_;
    private final RandomSequences f_286969_;
-
-   public ServerLevel(MinecraftServer p_214999_, Executor p_215000_, LevelStorageSource.LevelStorageAccess p_215001_, ServerLevelData p_215002_, ResourceKey<Level> p_215003_, LevelStem p_215004_, ChunkProgressListener p_215005_, boolean p_215006_, long p_215007_, List<CustomSpawner> p_215008_, boolean p_215009_, @Nullable RandomSequences p_288977_) {
-      super(p_215002_, p_215003_, p_214999_.m_206579_(), p_215004_.f_63975_(), p_214999_::m_129905_, false, p_215006_, p_215007_, p_214999_.m_213994_());
-      this.f_8561_ = p_215009_;
-      this.f_8548_ = p_214999_;
-      this.f_8558_ = p_215008_;
-      this.f_8549_ = p_215002_;
-      ChunkGenerator chunkgenerator = p_215004_.f_63976_();
-      boolean flag = p_214999_.m_6365_();
-      DataFixer datafixer = p_214999_.m_129933_();
-      EntityPersistentStorage<Entity> entitypersistentstorage = new EntityStorage(this, p_215001_.m_197394_(p_215003_).resolve("entities"), datafixer, flag, p_214999_);
-      this.f_143244_ = new PersistentEntitySectionManager<>(Entity.class, new ServerLevel.EntityCallbacks(), entitypersistentstorage);
-      this.f_8547_ = new ServerChunkCache(this, p_215001_, datafixer, p_214999_.m_236738_(), p_215000_, chunkgenerator, p_214999_.m_6846_().m_11312_(), p_214999_.m_6846_().m_184213_(), flag, p_215005_, this.f_143244_::m_287207_, () -> {
-         return p_214999_.m_129783_().m_8895_();
-      });
-      this.f_8547_.m_255415_().m_254958_();
-      this.f_8552_ = new PortalForcer(this);
-      this.m_46465_();
-      this.m_46466_();
-      this.m_6857_().m_61923_(p_214999_.m_6329_());
-      this.f_8565_ = this.m_8895_().m_164861_((p_184095_) -> {
-         return Raids.m_150235_(this, p_184095_);
-      }, () -> {
-         return new Raids(this);
-      }, Raids.m_211596_(this.m_204156_()));
-      if (!p_214999_.m_129792_()) {
-         p_215002_.m_5458_(p_214999_.m_130008_());
-      }
-
-      long i = p_214999_.m_129910_().m_246337_().m_245499_();
-      this.f_196556_ = new StructureCheck(this.f_8547_.m_196555_(), this.m_9598_(), p_214999_.m_236738_(), p_215003_, chunkgenerator, this.f_8547_.m_214994_(), this, chunkgenerator.m_62218_(), i, datafixer);
-      this.f_214997_ = new StructureManager(this, p_214999_.m_129910_().m_246337_(), this.f_196556_);
-      if (this.m_46472_() == Level.f_46430_ && this.m_204156_().m_203565_(BuiltinDimensionTypes.f_223540_)) {
-         this.f_8559_ = new EndDragonFight(this, i, p_214999_.m_129910_().m_6564_());
-      } else {
-         this.f_8559_ = null;
-      }
-
-      this.f_143245_ = new SleepStatus();
-      this.f_243695_ = new GameEventDispatcher(this);
-      this.f_286969_ = Objects.requireNonNullElseGet(p_288977_, () -> {
-         return this.m_8895_().m_164861_((p_287374_) -> {
-            return RandomSequences.m_287187_(i, p_287374_);
-         }, () -> {
-            return new RandomSequences(i);
-         }, "random_sequences");
-      });
+   private net.minecraftforge.common.util.LevelCapabilityData capabilityData;
+   // CraftBukkit start
+   public net.minecraft.world.level.storage.PrimaryLevelData worldDataServer;
+   private int tickPosition;
+   public final LevelStorageSource.LevelStorageAccess convertable;
+   public final UUID uuid;
+   public boolean isBukkitWorld = false; // CatServer
+
+   public LevelChunk getChunkIfLoaded(int x, int z) {
+      return this.f_8547_.m_62227_(x, z, false);
+   }
+
+   public ServerLevel(MinecraftServer p_214999_, Executor p_215000_, LevelStorageSource.LevelStorageAccess p_215001_, ServerLevelData p_215002_, ResourceKey<Level> p_215003_, LevelStem p_215004_, ChunkProgressListener p_215005_, boolean p_215006_, long p_215007_, List<CustomSpawner> p_215008_, boolean p_215009_, @Nullable RandomSequences p_288977_) { // Paper // CatServer - recover
+      super((p_215002_ instanceof net.minecraft.world.level.storage.DerivedLevelData derivedLevelData) ? derivedLevelData.setProperName(p_215003_) : p_215002_, p_215003_, p_214999_.m_206579_(), p_215004_.f_63975_(), p_214999_::m_129905_, false, p_215006_, p_215007_, p_214999_.m_213994_());
+      this.pvpMode = p_214999_.m_129799_();
+      this.convertable = p_215001_;
+      this.uuid = WorldUUID.getUUID(p_214999_.f_129744_.m_197394_(p_215003_).toFile()); // CatServer
+      // CatServer start
+      if (p_215002_ instanceof net.minecraft.world.level.storage.PrimaryLevelData primaryLevelData) {
+         this.worldDataServer = primaryLevelData;
+      } else if (p_215002_ instanceof net.minecraft.world.level.storage.DerivedLevelData derivedLevelData) {
+         this.worldDataServer = catserver.server.level.CustomDerivedLevelData.wrap(derivedLevelData);
+      }
+      this.spigotConfig = new org.spigotmc.SpigotWorldConfig(MinecraftServer.getServer().f_129744_.m_197394_(p_215003_).toFile().getName()); // Spigot // CatServer - from net.minecraft.world.level.Level#<init>
+      // CatServer end
+      this.f_8561_ = p_215009_;
+      this.f_8548_ = p_214999_;
+      this.f_8558_ = p_215008_;
+      this.f_8549_ = p_215002_;
+      ChunkGenerator chunkgenerator = p_215004_.f_63976_();
+      this.world = new org.bukkit.craftbukkit.v1_20_R1.CraftWorld(this, this.generator, this.biomeProvider, this.environment); // CatServer - before Level#getWorld() called
+      // CraftBukkit start - 1.20.1
+      if (this.biomeProvider != null) {
+         BiomeSource worldChunkManager = new CustomWorldChunkManager(this.getWorld(), this.biomeProvider, f_8548_.m_206579_().m_175515_(Registries.f_256952_));
+         if (chunkgenerator instanceof NoiseBasedChunkGenerator cga) {
+            chunkgenerator = new NoiseBasedChunkGenerator(worldChunkManager, cga.f_64318_);
+         } else if (chunkgenerator instanceof FlatLevelSource cpf) {
+            chunkgenerator = new FlatLevelSource(cpf.m_64191_(), worldChunkManager);
+         }
+      }
+      // CraftBukkit end
+      // CraftBukkit start
+      if (this.generator != null) {
+         chunkgenerator = new org.bukkit.craftbukkit.v1_20_R1.generator.CustomChunkGenerator(this, chunkgenerator, this.generator);
+      }
+      // CraftBukkit end
+
+      boolean flag = p_214999_.m_6365_();
+      DataFixer datafixer = p_214999_.m_129933_();
+      EntityPersistentStorage<Entity> entitypersistentstorage = new EntityStorage(this, p_215001_.m_197394_(p_215003_).resolve("entities"), datafixer, flag, p_214999_);
+      this.f_143244_ = new PersistentEntitySectionManager<>(Entity.class, new ServerLevel.EntityCallbacks(), entitypersistentstorage);
+      this.f_8547_ = new ServerChunkCache(this, p_215001_, datafixer, p_214999_.m_236738_(), p_215000_, chunkgenerator, this.spigotConfig.viewDistance, this.spigotConfig.simulationDistance, flag, p_215005_, this.f_143244_::m_287207_, () -> { // Spigot
+         return p_214999_.m_129783_().m_8895_();
+      });
+      this.f_8547_.m_255415_().m_254958_();
+      this.f_8552_ = new PortalForcer(this);
+      this.m_46465_();
+      this.m_46466_();
+      this.m_6857_().m_61923_(p_214999_.m_6329_());
+      this.f_8565_ = this.m_8895_().m_164861_((p_184095_) -> {
+         return Raids.m_150235_(this, p_184095_);
+      }, () -> {
+         return new Raids(this);
+      }, Raids.m_211596_(this.m_204156_()));
+      if (!p_214999_.m_129792_()) {
+         p_215002_.m_5458_(p_214999_.m_130008_());
+      }
+
+      long i = p_214999_.m_129910_().m_246337_().m_245499_();
+      this.f_196556_ = new StructureCheck(this.f_8547_.m_196555_(), this.m_9598_(), p_214999_.m_236738_(), p_215003_, chunkgenerator, this.f_8547_.m_214994_(), this, chunkgenerator.m_62218_(), i, datafixer);
+      this.f_214997_ = new StructureManager(this, this.worldDataServer.m_246337_(), this.f_196556_); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
+      if ((this.m_46472_() == Level.f_46430_ && this.m_204156_().m_203565_(BuiltinDimensionTypes.f_223540_)) || this.environment == org.bukkit.World.Environment.THE_END) { // CraftBukkit - Allow to create EnderDragonBattle in default and custom END
+         this.f_8559_ = new EndDragonFight(this, this.worldDataServer.m_246337_().m_245499_(), this.worldDataServer.m_6564_()); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
+      } else {
+         this.f_8559_ = null;
+      }
+
+      this.f_143245_ = new SleepStatus();
+      this.f_243695_ = new GameEventDispatcher(this);
+      this.f_286969_ = Objects.requireNonNullElseGet(p_288977_, () -> {
+         return this.m_8895_().m_164861_((p_287374_) -> {
+            return RandomSequences.m_287187_(i, p_287374_);
+         }, () -> {
+            return new RandomSequences(i);
+         }, "random_sequences");
+      });
+      this.initCapabilities();
+
+      m_6857_().world = this;
+      this.getCraftServer().addWorld(this.world);
+      this.worldDataServer.setWorld(this); // CraftBukkit
+      // Spigot start // CatServer - from net.minecraft.world.level.Level#<init>
+      this.timings = new SpigotTimings.WorldTimingsHandler(this); // Spigot - code below can generate new world and access timings
+      this.entityLimiter = new org.spigotmc.TickLimiter(this.spigotConfig.entityMaxTickTime);
+      this.tileLimiter = new org.spigotmc.TickLimiter(this.spigotConfig.tileMaxTickTime);
+      // Spigot end
+
+      // CatServer start
+      if (!Level.f_46428_.equals(p_215003_)) {
+         this.worldDataServer = new PrimaryLevelData(
+                 new net.minecraft.world.level.LevelSettings(
+                         worldDataServer.m_5462_(),
+                         worldDataServer.m_5464_(),
+                         worldDataServer.m_5466_(),
+                         worldDataServer.m_5472_(),
+                         false,
+                         worldDataServer.m_5470_(),
+                         this.f_8548_.worldLoader.f_244127_()
+                 ),
+                 this.f_8548_.f_129749_.m_246337_(),
+                 ((PrimaryLevelData) this.f_8548_.f_129749_).f_244368_,
+                 this.f_8548_.f_129749_.m_5754_()
+         );
+      }
+      // CatServer end
+   }
+
+   public ServerLevel(MinecraftServer p_214999_, Executor p_215000_, LevelStorageSource.LevelStorageAccess p_215001_, ServerLevelData p_215002_, ResourceKey<Level> p_215003_, LevelStem p_215004_, ChunkProgressListener p_215005_, boolean p_215006_, long p_215007_, List<CustomSpawner> p_215008_, boolean p_215009_, @Nullable RandomSequences p_288977_, org.bukkit.World.Environment env, org.bukkit.generator.ChunkGenerator gen, org.bukkit.generator.BiomeProvider biomeProvider) { // Paper // CatServer - recover
+      super((p_215002_ instanceof net.minecraft.world.level.storage.DerivedLevelData derivedLevelData) ? derivedLevelData.setProperName(p_215003_) : p_215002_, p_215003_, p_214999_.m_206579_(), p_215004_.f_63975_(), p_214999_::m_129905_, false, p_215006_, p_215007_, p_214999_.m_213994_(), gen, biomeProvider, env);
+      this.pvpMode = p_214999_.m_129799_();
+      this.convertable = p_215001_;
+      this.uuid = WorldUUID.getUUID(p_214999_.f_129744_.m_197394_(p_215003_).toFile()); // CatServer
+      // CatServer start
+      if (p_215002_ instanceof net.minecraft.world.level.storage.PrimaryLevelData primaryLevelData) {
+         this.worldDataServer = primaryLevelData;
+      } else if (p_215002_ instanceof net.minecraft.world.level.storage.DerivedLevelData derivedLevelData) {
+         this.worldDataServer = catserver.server.level.CustomDerivedLevelData.wrap(derivedLevelData);
+      }
+      this.spigotConfig = new org.spigotmc.SpigotWorldConfig(MinecraftServer.getServer().f_129744_.m_197394_(p_215003_).toFile().getName()); // Spigot // CatServer - from net.minecraft.world.level.Level#<init>
+      // CatServer end
+      this.f_8561_ = p_215009_;
+      this.f_8548_ = p_214999_;
+      this.f_8558_ = p_215008_;
+      this.f_8549_ = p_215002_;
+      ChunkGenerator chunkgenerator = p_215004_.f_63976_();
+      this.world = new org.bukkit.craftbukkit.v1_20_R1.CraftWorld(this, this.generator, this.biomeProvider, this.environment); // CatServer - before Level#getWorld() called
+      // CraftBukkit start - 1.20.1
+      if (this.biomeProvider != null) {
+         BiomeSource worldChunkManager = new CustomWorldChunkManager(this.getWorld(), this.biomeProvider, f_8548_.m_206579_().m_175515_(Registries.f_256952_));
+         if (chunkgenerator instanceof NoiseBasedChunkGenerator cga) {
+            chunkgenerator = new NoiseBasedChunkGenerator(worldChunkManager, cga.f_64318_);
+         } else if (chunkgenerator instanceof FlatLevelSource cpf) {
+            chunkgenerator = new FlatLevelSource(cpf.m_64191_(), worldChunkManager);
+         }
+      }
+      // CraftBukkit end
+      // CraftBukkit start
+      if (this.generator != null) {
+         chunkgenerator = new org.bukkit.craftbukkit.v1_20_R1.generator.CustomChunkGenerator(this, chunkgenerator, this.generator);
+      }
+      // CraftBukkit end
+
+      boolean flag = p_214999_.m_6365_();
+      DataFixer datafixer = p_214999_.m_129933_();
+      EntityPersistentStorage<Entity> entitypersistentstorage = new EntityStorage(this, p_215001_.m_197394_(p_215003_).resolve("entities"), datafixer, flag, p_214999_);
+      this.f_143244_ = new PersistentEntitySectionManager<>(Entity.class, new ServerLevel.EntityCallbacks(), entitypersistentstorage);
+      this.f_8547_ = new ServerChunkCache(this, p_215001_, datafixer, p_214999_.m_236738_(), p_215000_, chunkgenerator, this.spigotConfig.viewDistance, this.spigotConfig.simulationDistance, flag, p_215005_, this.f_143244_::m_287207_, () -> { // Spigot
+         return p_214999_.m_129783_().m_8895_();
+      });
+      this.f_8547_.m_255415_().m_254958_();
+      this.f_8552_ = new PortalForcer(this);
+      this.m_46465_();
+      this.m_46466_();
+      this.m_6857_().m_61923_(p_214999_.m_6329_());
+      this.f_8565_ = this.m_8895_().m_164861_((p_184095_) -> {
+         return Raids.m_150235_(this, p_184095_);
+      }, () -> {
+         return new Raids(this);
+      }, Raids.m_211596_(this.m_204156_()));
+      if (!p_214999_.m_129792_()) {
+         p_215002_.m_5458_(p_214999_.m_130008_());
+      }
+
+      long i = p_214999_.m_129910_().m_246337_().m_245499_();
+      this.f_196556_ = new StructureCheck(this.f_8547_.m_196555_(), this.m_9598_(), p_214999_.m_236738_(), p_215003_, chunkgenerator, this.f_8547_.m_214994_(), this, chunkgenerator.m_62218_(), i, datafixer);
+      this.f_214997_ = new StructureManager(this, this.worldDataServer.m_246337_(), this.f_196556_); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
+      if ((this.m_46472_() == Level.f_46430_ && this.m_204156_().m_203565_(BuiltinDimensionTypes.f_223540_)) || this.environment == org.bukkit.World.Environment.THE_END) { // CraftBukkit - Allow to create EnderDragonBattle in default and custom END
+         this.f_8559_ = new EndDragonFight(this, this.worldDataServer.m_246337_().m_245499_(), this.worldDataServer.m_6564_()); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
+      } else {
+         this.f_8559_ = null;
+      }
+
+      this.f_143245_ = new SleepStatus();
+      this.f_243695_ = new GameEventDispatcher(this);
+      this.f_286969_ = Objects.requireNonNullElseGet(p_288977_, () -> {
+         return this.m_8895_().m_164861_((p_287374_) -> {
+            return RandomSequences.m_287187_(i, p_287374_);
+         }, () -> {
+            return new RandomSequences(i);
+         }, "random_sequences");
+      });
+      this.initCapabilities();
+
+      m_6857_().world = this;
+      this.getCraftServer().addWorld(this.world);
+      this.worldDataServer.setWorld(this); // CraftBukkit
+      // Spigot start // CatServer - from net.minecraft.world.level.Level#<init>
+      this.timings = new SpigotTimings.WorldTimingsHandler(this); // Spigot - code below can generate new world and access timings
+      this.entityLimiter = new org.spigotmc.TickLimiter(this.spigotConfig.entityMaxTickTime);
+      this.tileLimiter = new org.spigotmc.TickLimiter(this.spigotConfig.tileMaxTickTime);
+      // Spigot end
    }
 
    /** @deprecated */
@@ -274,12 +_,20 @@
       this.m_184096_();
       int i = this.m_46469_().m_46215_(GameRules.f_151486_);
       if (this.f_143245_.m_144002_(i) && this.f_143245_.m_144004_(i, this.f_8546_)) {
+         // CraftBukkit start
+         long j = this.f_46442_.m_6792_() + 24000L;
+         TimeSkipEvent event = new TimeSkipEvent(this.getWorld(), TimeSkipEvent.SkipReason.NIGHT_SKIP, (j - j % 24000L) - this.m_46468_());
          if (this.m_46469_().m_46207_(GameRules.f_46140_)) {
-            long j = this.f_46442_.m_6792_() + 24000L;
-            this.m_8615_(j - j % 24000L);
+            getCraftServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               this.m_8615_(net.minecraftforge.event.ForgeEventFactory.onSleepFinished(this, this.m_46468_() + event.getSkipAmount(), this.m_46468_()));
+            }
          }
 
-         this.m_8804_();
+         if (!event.isCancelled()) {
+            this.m_8804_();
+         }
+         // CraftBukkit end
          if (this.m_46469_().m_46207_(GameRules.f_46150_) && this.m_46471_()) {
             this.m_184097_();
          }
@@ -288,6 +_,7 @@
       this.m_46465_();
       this.m_8809_();
       profilerfiller.m_6182_("tickPending");
+      timings.doTickPending.startTiming(); // Spigot
       if (!this.m_46659_()) {
          long k = this.m_46467_();
          profilerfiller.m_6180_("blockTicks");
@@ -296,28 +_,34 @@
          this.f_184047_.m_193225_(k, 65536, this::m_184076_);
          profilerfiller.m_7238_();
       }
+      timings.doTickPending.stopTiming(); // Spigot
 
       profilerfiller.m_6182_("raid");
       this.f_8565_.m_37957_();
       profilerfiller.m_6182_("chunkSource");
       this.m_7726_().m_201698_(p_8794_, true);
       profilerfiller.m_6182_("blockEvents");
+      timings.doSounds.startTiming(); // Spigot
       this.m_8807_();
+      timings.doSounds.stopTiming(); // Spigot
       this.f_8557_ = false;
       profilerfiller.m_7238_();
-      boolean flag = !this.f_8546_.isEmpty() || !this.m_8902_().isEmpty();
+      boolean flag = !this.f_8546_.isEmpty() || net.minecraftforge.common.world.ForgeChunkManager.hasForcedChunks(this); //Forge: Replace vanilla's has forced chunk check with forge's that checks both the vanilla and forge added ones
       if (flag) {
          this.m_8886_();
       }
 
       if (flag || this.f_8551_++ < 300) {
          profilerfiller.m_6180_("entities");
+         timings.tickEntities.startTiming(); // Spigot
          if (this.f_8559_ != null) {
             profilerfiller.m_6180_("dragonFight");
             this.f_8559_.m_64095_();
             profilerfiller.m_7238_();
          }
 
+         org.spigotmc.ActivationRange.activateEntities(this); // Spigot
+         timings.entityTick.startTiming(); // Spigot
          this.f_143243_.m_156910_((p_184065_) -> {
             if (!p_184065_.m_213877_()) {
                if (this.m_143342_(p_184065_)) {
@@ -337,12 +_,16 @@
                      }
 
                      profilerfiller.m_6180_("tick");
-                     this.m_46653_(this::m_8647_, p_184065_);
+                     if (!p_184065_.m_213877_() && !(p_184065_ instanceof net.minecraftforge.entity.PartEntity)) {
+                        this.m_46653_(this::m_8647_, p_184065_);
+                     }
                      profilerfiller.m_7238_();
                   }
                }
             }
          });
+         timings.entityTick.stopTiming(); // Spigot
+         timings.tickEntities.stopTiming(); // Spigot
          profilerfiller.m_7238_();
          this.m_46463_();
       }
@@ -401,7 +_,7 @@
       int j = chunkpos.m_45605_();
       ProfilerFiller profilerfiller = this.m_46473_();
       profilerfiller.m_6180_("thunder");
-      if (flag && this.m_46470_() && this.f_46441_.m_188503_(100000) == 0) {
+      if (flag && this.m_46470_() && this.spigotConfig.thunderChance > 0 && this.f_46441_.m_188503_(this.spigotConfig.thunderChance) == 0) { // Spigot
          BlockPos blockpos = this.m_143288_(this.m_46496_(i, 0, j, 15));
          if (this.m_46758_(blockpos)) {
             DifficultyInstance difficultyinstance = this.m_6436_(blockpos);
@@ -412,7 +_,7 @@
                   skeletonhorse.m_30923_(true);
                   skeletonhorse.m_146762_(0);
                   skeletonhorse.m_6034_((double)blockpos.m_123341_(), (double)blockpos.m_123342_(), (double)blockpos.m_123343_());
-                  this.m_7967_(skeletonhorse);
+                  this.addFreshEntity(skeletonhorse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
                }
             }
 
@@ -420,7 +_,7 @@
             if (lightningbolt != null) {
                lightningbolt.m_20219_(Vec3.m_82539_(blockpos));
                lightningbolt.m_20874_(flag1);
-               this.m_7967_(lightningbolt);
+               this.strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.WEATHER); // CraftBukkit
             }
          }
       }
@@ -430,8 +_,9 @@
          BlockPos blockpos1 = this.m_5452_(Heightmap.Types.MOTION_BLOCKING, this.m_46496_(i, 0, j, 15));
          BlockPos blockpos2 = blockpos1.m_7495_();
          Biome biome = this.m_204166_(blockpos1).m_203334_();
+         if (this.isAreaLoaded(blockpos2, 1)) // Forge: check area to avoid loading neighbors in unloaded chunks
          if (biome.m_47477_(this, blockpos2)) {
-            this.m_46597_(blockpos2, Blocks.f_50126_.m_49966_());
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockFormEvent(this, blockpos2, Blocks.f_50126_.m_49966_(), null); // CraftBukkit
          }
 
          if (flag) {
@@ -443,10 +_,10 @@
                   if (k < Math.min(i1, 8)) {
                      BlockState blockstate1 = blockstate.m_61124_(SnowLayerBlock.f_56581_, Integer.valueOf(k + 1));
                      Block.m_49897_(blockstate, blockstate1, this, blockpos1);
-                     this.m_46597_(blockpos1, blockstate1);
+                     org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockFormEvent(this, blockpos1, blockstate1, null); // CraftBukkit - 1.20.1
                   }
                } else {
-                  this.m_46597_(blockpos1, Blocks.f_50125_.m_49966_());
+                  org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockFormEvent(this, blockpos1, Blocks.f_50125_.m_49966_(), null); // CraftBukkit
                }
             }
 
@@ -625,32 +_,69 @@
          this.f_46438_ = Mth.m_14036_(this.f_46438_, 0.0F, 1.0F);
       }
 
-      if (this.f_46437_ != this.f_46438_) {
+      /* CraftBukkit start
+      if (this.oRainLevel != this.rainLevel) {
+         this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, this.rainLevel), this.dimension());
+      }
+
+      if (this.oThunderLevel != this.thunderLevel) {
+         this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, this.thunderLevel), this.dimension());
+      }
+       */
+      for (int idx = 0; idx < this.f_8546_.size(); ++idx) {
+         if (((ServerPlayer) this.f_8546_.get(idx)).m_9236_() == this) {
+            ((ServerPlayer) this.f_8546_.get(idx)).tickWeather();
+         }
+      }
+      if (flag != this.m_46471_()) {
+         // Only send weather packets to those affected
+         for (int idx = 0; idx < this.f_8546_.size(); ++idx) {
+            if (((ServerPlayer) this.f_8546_.get(idx)).m_9236_() == this) {
+               ((ServerPlayer) this.f_8546_.get(idx)).setPlayerWeather((!flag ? WeatherType.DOWNFALL : WeatherType.CLEAR), false);
+            }
+         }
+      }
+      for (int idx = 0; idx < this.f_8546_.size(); ++idx) {
+         if (((ServerPlayer) this.f_8546_.get(idx)).m_9236_() == this) {
+            ((ServerPlayer) this.f_8546_.get(idx)).updateWeather(this.f_46437_, this.f_46438_, this.f_46439_, this.f_46440_);
+         }
+      }
+      // CraftBukkit end
+
+      /* The function in use here has been replaced in order to only send the weather info to players in the correct dimension,
+       * rather than to all players on the server. This is what causes the client-side rain, as the
+       * client believes that it has started raining locally, rather than in another dimension.
+       */
+      if (flag != this.m_46471_()) {
+         if (flag) {
+            this.f_8548_.m_6846_().m_11270_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132155_, 0.0F), this.m_46472_());
+         } else {
+            this.f_8548_.m_6846_().m_11270_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132154_, 0.0F), this.m_46472_());
+         }
+
          this.f_8548_.m_6846_().m_11270_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132160_, this.f_46438_), this.m_46472_());
-      }
-
-      if (this.f_46439_ != this.f_46440_) {
          this.f_8548_.m_6846_().m_11270_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132161_, this.f_46440_), this.m_46472_());
       }
 
-      if (flag != this.m_46471_()) {
-         if (flag) {
-            this.f_8548_.m_6846_().m_11268_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132155_, 0.0F));
-         } else {
-            this.f_8548_.m_6846_().m_11268_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132154_, 0.0F));
-         }
-
-         this.f_8548_.m_6846_().m_11268_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132160_, this.f_46438_));
-         this.f_8548_.m_6846_().m_11268_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132161_, this.f_46440_));
-      }
-
    }
 
    private void m_184097_() {
-      this.f_8549_.m_6399_(0);
+      // CraftBukkit start
       this.f_8549_.m_5565_(false);
-      this.f_8549_.m_6398_(0);
+      // If we stop due to everyone sleeping we should reset the weather duration to some other random value.
+      // Not that everyone ever manages to get the whole server to sleep at the same time....
+      if (!this.f_8549_.m_6533_()) {
+         this.f_8549_.m_6399_(0);
+      }
+      // CraftBukkit end
       this.f_8549_.m_5557_(false);
+      // CraftBukkit start
+      // If we stop due to everyone sleeping we should reset the weather duration to some other random value.
+      // Not that everyone ever manages to get the whole server to sleep at the same time....
+      if (!this.f_8549_.m_6534_()) {
+         this.f_8549_.m_6398_(0);
+      }
+      // CraftBukkit end
    }
 
    public void m_8886_() {
@@ -674,6 +_,14 @@
    }
 
    public void m_8647_(Entity p_8648_) {
+      // Spigot start
+      if (!org.spigotmc.ActivationRange.checkIfActive(p_8648_)) {
+         p_8648_.f_19797_++;
+         p_8648_.inactiveTick();
+         return;
+      }
+      // Spigot end
+      p_8648_.tickTimer.startTiming(); // Spigot
       p_8648_.m_146867_();
       ProfilerFiller profilerfiller = this.m_46473_();
       ++p_8648_.f_19797_;
@@ -682,12 +_,13 @@
       });
       profilerfiller.m_6174_("tickNonPassenger");
       p_8648_.m_8119_();
+      p_8648_.postTick(); // CraftBukkit
       this.m_46473_().m_7238_();
 
       for(Entity entity : p_8648_.m_20197_()) {
          this.m_8662_(p_8648_, entity);
       }
-
+      p_8648_.tickTimer.stopTiming(); // Spigot
    }
 
    private void m_8662_(Entity p_8663_, Entity p_8664_) {
@@ -700,7 +_,10 @@
                return BuiltInRegistries.f_256780_.m_7981_(p_8664_.m_6095_()).toString();
             });
             profilerfiller.m_6174_("tickPassenger");
-            p_8664_.m_6083_();
+            if (p_8664_.canUpdate()) {
+               p_8664_.m_6083_();
+               p_8664_.postTick(); // CraftBukkit
+            } // CatServer
             profilerfiller.m_7238_();
 
             for(Entity entity : p_8664_.m_20197_()) {
@@ -720,6 +_,7 @@
    public void m_8643_(@Nullable ProgressListener p_8644_, boolean p_8645_, boolean p_8646_) {
       ServerChunkCache serverchunkcache = this.m_7726_();
       if (!p_8646_) {
+         org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(getWorld())); // CraftBukkit
          if (p_8644_ != null) {
             p_8644_.m_6309_(Component.m_237115_("menu.savingLevel"));
          }
@@ -736,12 +_,13 @@
             this.f_143244_.m_157554_();
          }
 
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Save(this));
       }
    }
 
    private void m_8806_() {
       if (this.f_8559_ != null) {
-         this.f_8548_.m_129910_().m_5915_(this.f_8559_.m_289745_());
+         this.worldDataServer.m_5915_(this.f_8559_.m_289745_()); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
       }
 
       this.m_7726_().m_8483_().m_78151_();
@@ -799,17 +_,36 @@
       return list.isEmpty() ? null : list.get(this.f_46441_.m_188503_(list.size()));
    }
 
+   // CatServer start
    public boolean m_7967_(Entity p_8837_) {
-      return this.m_8872_(p_8837_);
+      CreatureSpawnEvent.SpawnReason reason = catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addFreshEntity$reason.get(this);
+      return this.addFreshEntity(p_8837_, reason);
+   }
+   // CatServer end
+
+   public boolean addFreshEntity(Entity p_8837_, CreatureSpawnEvent.SpawnReason reason) {
+      return this.addEntity(p_8837_, reason);
    }
 
    public boolean m_8847_(Entity p_8848_) {
-      return this.m_8872_(p_8848_);
+      return this.addEntity(p_8848_, CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   public boolean addWithUUID(Entity p_8848_, CreatureSpawnEvent.SpawnReason reason) {
+      return this.addEntity(p_8848_, reason);
    }
 
    public void m_143334_(Entity p_143335_) {
-      this.m_8872_(p_143335_);
-   }
+      // SPIGOT-6415: Don't call spawn event for entities which travel trough worlds,
+      // since it is only an implementation detail, that a new entity is created when
+      // they are traveling between worlds.
+      this.addDuringTeleport(p_143335_, CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   public void addDuringTeleport(Entity p_143335_, CreatureSpawnEvent.SpawnReason reason) {
+      this.addEntity(p_143335_, reason);
+   }
+   // CatServer end
 
    public void m_8622_(ServerPlayer p_8623_) {
       this.m_8853_(p_8623_);
@@ -828,6 +_,7 @@
    }
 
    private void m_8853_(ServerPlayer p_8854_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinLevelEvent(p_8854_, this))) return;
       Entity entity = this.m_142646_().m_142694_(p_8854_.m_20148_());
       if (entity != null) {
          f_8566_.warn("Force-added player with duplicate UUID {}", (Object)p_8854_.m_20148_().toString());
@@ -835,28 +_,76 @@
          this.m_143261_((ServerPlayer)entity, Entity.RemovalReason.DISCARDED);
       }
 
-      this.f_143244_.m_157533_(p_8854_);
+      this.f_143244_.addNewEntityWithoutEvent(p_8854_);
+      p_8854_.onAddedToWorld();
    }
 
+   // CatServer start
+   // Must use CatServerCaptures here to compatible with mixin inject
    private boolean m_8872_(Entity p_8873_) {
+      CreatureSpawnEvent.SpawnReason spawnReason = catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addEntity$spawnReason.get(this);
+      // CatServer start - Thread safety check
+      if (!f_8548_.m_18695_()) {
+         catserver.server.CatServer.LOGGER.debug("Async add entity captured!", new RuntimeException());
+         f_8548_.processQueue.add(() -> addEntity(p_8873_, spawnReason));
+         return true;
+      }
+      org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+      // CatServer end
       if (p_8873_.m_213877_()) {
-         f_8566_.warn("Tried to add entity {} but it was marked as removed already", (Object)EntityType.m_20613_(p_8873_.m_6095_()));
+         // LOGGER.warn("Tried to add entity {} but it was marked as removed already", (Object)EntityType.getKey(p_8873_.getType())); // CraftBukkit
          return false;
       } else {
-         return this.f_143244_.m_157533_(p_8873_);
-      }
-   }
+         // SPIGOT-6415: Don't call spawn event when reason is null. For example when an entity teleports to a new world.
+         if (spawnReason != null && !CraftEventFactory.doEntityAddEventCalling(this, p_8873_, spawnReason)) {
+            return false;
+         }
+         // CraftBukkit end
+         if (this.f_143244_.m_157533_(p_8873_)) {
+            p_8873_.onAddedToWorld();
+            return true;
+         } else {
+            return false;
+         }
+      }
+   }
+
+   private boolean addEntity(Entity p_8873_, CreatureSpawnEvent.SpawnReason spawnReason) {
+      catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addEntity$spawnReason.push(this, spawnReason);
+      try {
+         return m_8872_(p_8873_);
+      } finally {
+         // CatServer - Make sure reset
+         catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevel$addEntity$spawnReason.pop(this);
+      }
+   }
+   // CatServer end
 
    public boolean m_8860_(Entity p_8861_) {
+      // CraftBukkit start
+      return tryAddFreshEntityWithPassengers(p_8861_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   public boolean tryAddFreshEntityWithPassengers(Entity p_8861_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      // CraftBukkit end
       if (p_8861_.m_20199_().map(Entity::m_20148_).anyMatch(this.f_143244_::m_157550_)) {
          return false;
       } else {
-         this.m_47205_(p_8861_);
+         this.addFreshEntityWithPassengers(p_8861_, reason); // CraftBukkit
          return true;
       }
    }
 
    public void m_8712_(LevelChunk p_8713_) {
+      // Spigot Start
+      for (net.minecraft.world.level.block.entity.BlockEntity tileentity : p_8713_.m_62954_().values()) {
+         if (tileentity instanceof net.minecraft.world.Container) {
+            for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(catserver.server.inventory.CatInventoryUtils.getViewers((net.minecraft.world.Container) tileentity))) { // CatServer
+               h.closeInventory();
+            }
+         }
+      }
+      // Spigot End
       p_8713_.m_187957_();
       p_8713_.m_187979_(this);
    }
@@ -865,12 +_,36 @@
       p_143262_.m_142687_(p_143263_);
    }
 
+   // CraftBukkit start
+   public boolean strikeLightning(Entity entitylightning) {
+      return this.strikeLightning(entitylightning, LightningStrikeEvent.Cause.UNKNOWN);
+   }
+
+   public boolean strikeLightning(Entity entitylightning, LightningStrikeEvent.Cause cause) {
+      LightningStrikeEvent lightning = CraftEventFactory.callLightningStrikeEvent((org.bukkit.entity.LightningStrike) entitylightning.getBukkitEntity(), cause);
+      if (lightning.isCancelled()) {
+         return false;
+      }
+      return this.m_7967_(entitylightning);
+   }
+   // CraftBukkit end
+
    public void m_6801_(int p_8612_, BlockPos p_8613_, int p_8614_) {
+      // CraftBukkit start
+      Player player = null;
+      Entity entity = this.m_6815_(p_8612_);
+      if (entity instanceof Player) player = (Player) entity;
+      // CraftBukkit end
       for(ServerPlayer serverplayer : this.f_8548_.m_6846_().m_11314_()) {
          if (serverplayer != null && serverplayer.m_9236_() == this && serverplayer.m_19879_() != p_8612_) {
             double d0 = (double)p_8613_.m_123341_() - serverplayer.m_20185_();
             double d1 = (double)p_8613_.m_123342_() - serverplayer.m_20186_();
             double d2 = (double)p_8613_.m_123343_() - serverplayer.m_20189_();
+            // CraftBukkit start
+            if (player != null && !serverplayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
+               continue;
+            }
+            // CraftBukkit end
             if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0D) {
                serverplayer.f_8906_.m_9829_(new ClientboundBlockDestructionPacket(p_8612_, p_8613_, p_8614_));
             }
@@ -880,10 +_,22 @@
    }
 
    public void m_262808_(@Nullable Player p_263330_, double p_263393_, double p_263369_, double p_263354_, Holder<SoundEvent> p_263412_, SoundSource p_263338_, float p_263352_, float p_263390_, long p_263403_) {
+      net.minecraftforge.event.PlayLevelSoundEvent.AtPosition event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtPosition(this, p_263393_, p_263369_, p_263354_, p_263412_, p_263338_, p_263352_, p_263390_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_263412_ = event.getSound();
+      p_263338_ = event.getSource();
+      p_263352_ = event.getNewVolume();
+      p_263390_ = event.getNewPitch();
       this.f_8548_.m_6846_().m_11241_(p_263330_, p_263393_, p_263369_, p_263354_, (double)p_263412_.m_203334_().m_215668_(p_263352_), this.m_46472_(), new ClientboundSoundPacket(p_263412_, p_263338_, p_263393_, p_263369_, p_263354_, p_263352_, p_263390_, p_263403_));
    }
 
    public void m_213890_(@Nullable Player p_263545_, Entity p_263544_, Holder<SoundEvent> p_263491_, SoundSource p_263542_, float p_263530_, float p_263520_, long p_263490_) {
+      net.minecraftforge.event.PlayLevelSoundEvent.AtEntity event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(p_263544_, p_263491_, p_263542_, p_263530_, p_263520_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_263491_ = event.getSound();
+      p_263542_ = event.getSource();
+      p_263530_ = event.getNewVolume();
+      p_263520_ = event.getNewPitch();
       this.f_8548_.m_6846_().m_11241_(p_263545_, p_263544_.m_20185_(), p_263544_.m_20186_(), p_263544_.m_20189_(), (double)p_263491_.m_203334_().m_215668_(p_263530_), this.m_46472_(), new ClientboundSoundEntityPacket(p_263491_, p_263542_, p_263544_, p_263530_, p_263520_, p_263490_));
    }
 
@@ -905,9 +_,17 @@
    }
 
    public void m_214171_(GameEvent p_215041_, Vec3 p_215042_, GameEvent.Context p_215043_) {
+      if (!net.minecraftforge.common.ForgeHooks.onVanillaGameEvent(this, p_215041_, p_215042_, p_215043_)) return;
       this.f_243695_.m_245905_(p_215041_, p_215042_, p_215043_);
    }
 
+   // CatServer start
+   @Override
+   public ServerLevel getMinecraftWorld() {
+      return this.m_6018_();
+   }
+   // CatServer end
+
    public void m_7260_(BlockPos p_8755_, BlockState p_8756_, BlockState p_8757_, int p_8758_) {
       if (this.f_200893_) {
          String s = "recursive call to sendBlockUpdated";
@@ -920,7 +_,20 @@
       if (Shapes.m_83157_(voxelshape1, voxelshape, BooleanOp.f_82687_)) {
          List<PathNavigation> list = new ObjectArrayList<>();
 
-         for(Mob mob : this.f_143246_) {
+         // CraftBukkit start - fix SPIGOT-6362
+         java.util.Iterator<Mob> iterator = this.f_143246_.iterator();
+         while (iterator.hasNext()) {
+            Mob mob;
+            try {
+               mob = iterator.next();
+            } catch (java.util.ConcurrentModificationException ex) {
+               // This can happen because the pathfinder update below may trigger a chunk load, which in turn may cause more navigators to register
+               // In this case we just run the update again across all the iterators as the chunk will then be loaded
+               // As this is a relative edge case it is much faster than copying navigators (on either read or write)
+               m_7260_(p_8755_, p_8756_, p_8757_, p_8758_);
+               return;
+            }
+            // CraftBukkit end
             PathNavigation pathnavigation = mob.m_21573_();
             if (pathnavigation.m_200903_(p_8755_)) {
                list.add(pathnavigation);
@@ -941,10 +_,15 @@
    }
 
    public void m_46672_(BlockPos p_215045_, Block p_215046_) {
+      net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(this, p_215045_, this.m_8055_(p_215045_), java.util.EnumSet.allOf(Direction.class), false).isCanceled();
       this.f_220350_.m_214152_(p_215045_, p_215046_, (Direction)null);
    }
 
    public void m_46590_(BlockPos p_215052_, Block p_215053_, Direction p_215054_) {
+      java.util.EnumSet<Direction> directions = java.util.EnumSet.allOf(Direction.class);
+      directions.remove(p_215054_);
+      if (net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(this, p_215052_, this.m_8055_(p_215052_), directions, false).isCanceled())
+         return;
       this.f_220350_.m_214152_(p_215052_, p_215053_, p_215054_);
    }
 
@@ -970,6 +_,11 @@
 
    public Explosion m_254877_(@Nullable Entity p_256039_, @Nullable DamageSource p_255778_, @Nullable ExplosionDamageCalculator p_256002_, double p_256067_, double p_256370_, double p_256153_, float p_256045_, boolean p_255686_, Level.ExplosionInteraction p_255827_) {
       Explosion explosion = this.m_255278_(p_256039_, p_255778_, p_256002_, p_256067_, p_256370_, p_256153_, p_256045_, p_255686_, p_255827_, false);
+      // CraftBukkit start
+      if (explosion.wasCanceled) {
+         return explosion;
+      }
+      // CraftBukkit end
       if (!explosion.m_254884_()) {
          explosion.m_46080_();
       }
@@ -1031,12 +_,22 @@
    }
 
    public <T extends ParticleOptions> int m_8767_(T p_8768_, double p_8769_, double p_8770_, double p_8771_, int p_8772_, double p_8773_, double p_8774_, double p_8775_, double p_8776_) {
-      ClientboundLevelParticlesPacket clientboundlevelparticlespacket = new ClientboundLevelParticlesPacket(p_8768_, false, p_8769_, p_8770_, p_8771_, (float)p_8773_, (float)p_8774_, (float)p_8775_, (float)p_8776_, p_8772_);
+
+      // CraftBukkit - visibility api support
+      return sendParticles(null, p_8768_, p_8769_, p_8770_, p_8771_, p_8772_, p_8773_, p_8774_, p_8775_, p_8776_, false);
+   }
+
+   public <T extends ParticleOptions> int sendParticles(ServerPlayer sender, T p_8768_, double p_8769_, double p_8770_, double p_8771_, int p_8772_, double p_8773_, double p_8774_, double p_8775_, double p_8776_, boolean force) {
+      ClientboundLevelParticlesPacket clientboundlevelparticlespacket = new ClientboundLevelParticlesPacket(p_8768_, force, p_8769_, p_8770_, p_8771_, (float) p_8773_, (float) p_8774_, (float) p_8775_, (float) p_8776_, p_8772_);
+      // CraftBukkit end
       int i = 0;
 
-      for(int j = 0; j < this.f_8546_.size(); ++j) {
+      for (int j = 0; j < this.f_8546_.size(); ++j) {
          ServerPlayer serverplayer = this.f_8546_.get(j);
-         if (this.m_8636_(serverplayer, false, p_8769_, p_8770_, p_8771_, clientboundlevelparticlespacket)) {
+         if (sender != null && !serverplayer.getBukkitEntity().canSee(sender.getBukkitEntity()))
+            continue; // CraftBukkit
+
+         if (this.m_8636_(serverplayer, force, p_8769_, p_8770_, p_8771_, clientboundlevelparticlespacket)) {  // CraftBukkit
             ++i;
          }
       }
@@ -1083,7 +_,7 @@
 
    @Nullable
    public BlockPos m_215011_(TagKey<Structure> p_215012_, BlockPos p_215013_, int p_215014_, boolean p_215015_) {
-      if (!this.f_8548_.m_129910_().m_246337_().m_247749_()) {
+      if (!this.worldDataServer.m_246337_().m_247749_()) { // CraftBukkit // CatServer - serverLevelData -> worldDataServer
          return null;
       } else {
          Optional<HolderSet.Named<Structure>> optional = this.m_9598_().m_175515_(Registries.f_256944_).m_203431_(p_215012_);
@@ -1115,10 +_,20 @@
 
    @Nullable
    public MapItemSavedData m_7489_(String p_8785_) {
-      return this.m_7654_().m_129783_().m_8895_().m_164858_(MapItemSavedData::m_164807_, p_8785_);
+      // CraftBukkit start
+      return this.m_7654_().m_129783_().m_8895_().m_164858_((compound) -> {
+         // We only get here when the data file exists, but is not a valid map
+         MapItemSavedData newMap = MapItemSavedData.m_164807_(compound);
+         newMap.id = p_8785_;
+         MapInitializeEvent event = new MapInitializeEvent(newMap.mapView);
+         Bukkit.getServer().getPluginManager().callEvent(event);
+         return newMap;
+      }, p_8785_);
+      // CraftBukkit end
    }
 
    public void m_142325_(String p_143305_, MapItemSavedData p_143306_) {
+      p_143306_.id = p_143305_; // CraftBukkit
       this.m_7654_().m_129783_().m_8895_().m_164855_(p_143305_, p_143306_);
    }
 
@@ -1309,6 +_,11 @@
 
    public void m_6289_(BlockPos p_8743_, Block p_8744_) {
       if (!this.m_46659_()) {
+         // CraftBukkit start
+         if (populating) {
+            return;
+         }
+         // CraftBukkit end
          this.m_46672_(p_8743_, p_8744_);
       }
 
@@ -1327,11 +_,11 @@
    }
 
    public boolean m_8584_() {
-      return this.f_8548_.m_129910_().m_5961_();
+      return this.worldDataServer.m_5961_(); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
    }
 
    public long m_7328_() {
-      return this.f_8548_.m_129910_().m_246337_().m_245499_();
+      return this.worldDataServer.m_246337_().m_245499_(); // CraftBukkit // CatServer - serverLevelData -> worldDataServer
    }
 
    @Nullable
@@ -1368,19 +_,41 @@
    }
 
    public static void m_8617_(ServerLevel p_8618_) {
-      BlockPos blockpos = f_8562_;
+      // CraftBukkit start
+      ServerLevel.makeObsidianPlatform(p_8618_, null);
+   }
+
+   public static void makeObsidianPlatform(ServerLevel p_8618_, Entity entity) {
+      // CraftBukkit end
+      BlockPos blockpos = ServerLevel.f_8562_;
       int i = blockpos.m_123341_();
       int j = blockpos.m_123342_() - 2;
       int k = blockpos.m_123343_();
+      // CraftBukkit start
+      org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator(p_8618_);
       BlockPos.m_121976_(i - 2, j + 1, k - 2, i + 2, j + 3, k + 2).forEach((p_207578_) -> {
-         p_8618_.m_46597_(p_207578_, Blocks.f_50016_.m_49966_());
+         blockList.m_7731_(p_207578_, Blocks.f_50016_.m_49966_(), 3);
       });
       BlockPos.m_121976_(i - 2, j, k - 2, i + 2, j, k + 2).forEach((p_184101_) -> {
-         p_8618_.m_46597_(p_184101_, Blocks.f_50080_.m_49966_());
+         blockList.m_7731_(p_184101_, Blocks.f_50080_.m_49966_(), 3);
       });
+      org.bukkit.World bworld = p_8618_.getWorld();
+      org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), bworld, (entity == null) ? null : entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
+      p_8618_.getCraftServer().getPluginManager().callEvent(portalEvent);
+      if (!portalEvent.isCancelled()) {
+         blockList.updateList();
+      }
+      // CraftBukkit end
+   }
+
+   protected void initCapabilities() {
+      this.gatherCapabilities();
+      capabilityData = this.m_8895_().m_164861_(e -> net.minecraftforge.common.util.LevelCapabilityData.load(e, getCapabilities()), () -> new net.minecraftforge.common.util.LevelCapabilityData(getCapabilities()), net.minecraftforge.common.util.LevelCapabilityData.ID);
+      capabilityData.setCapabilities(getCapabilities());
    }
 
    public LevelEntityGetter<Entity> m_142646_() {
+      org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
       return this.f_143244_.m_157567_();
    }
 
@@ -1460,6 +_,7 @@
       }
 
       public void m_141985_(Entity p_143371_) {
+         org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot
          ServerLevel.this.m_7726_().m_8463_(p_143371_);
          if (p_143371_ instanceof ServerPlayer serverplayer) {
             ServerLevel.this.f_8546_.add(serverplayer);
@@ -1475,16 +_,48 @@
             ServerLevel.this.f_143246_.add(mob);
          }
 
-         if (p_143371_ instanceof EnderDragon enderdragon) {
-            for(EnderDragonPart enderdragonpart : enderdragon.m_31156_()) {
+         if (p_143371_.isMultipartEntity()) {
+            for(net.minecraftforge.entity.PartEntity<?> enderdragonpart : p_143371_.getParts()) {
                ServerLevel.this.f_143247_.put(enderdragonpart.m_19879_(), enderdragonpart);
             }
          }
 
          p_143371_.m_213651_(DynamicGameEventListener::m_223617_);
+         p_143371_.valid = true; // CraftBukkit
       }
 
       public void m_141981_(Entity p_143375_) {
+         org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot
+         // Spigot start
+         if ( p_143375_ instanceof Player )
+         {
+            com.google.common.collect.Streams.stream( m_7654_().m_129785_() ).map( ServerLevel::m_8895_ ).forEach( (worldData) ->
+            {
+               for (Object o : worldData.f_78144_.values() )
+               {
+                  if ( o instanceof MapItemSavedData )
+                  {
+                     MapItemSavedData map = (MapItemSavedData) o;
+                     map.f_77896_.remove( (Player) p_143375_ );
+                     for (Iterator<MapItemSavedData.HoldingPlayer> iter = (Iterator<MapItemSavedData.HoldingPlayer>) map.f_77893_.iterator(); iter.hasNext(); )
+                     {
+                        if ( iter.next().f_77959_ == p_143375_ )
+                        {
+                           iter.remove();
+                        }
+                     }
+                  }
+               }
+            } );
+         }
+         // Spigot end
+         // Spigot Start
+         if (p_143375_.getBukkitEntity() instanceof org.bukkit.inventory.InventoryHolder && (!(p_143375_ instanceof ServerPlayer) || p_143375_.m_146911_() != Entity.RemovalReason.KILLED)) { // SPIGOT-6876: closeInventory clears death message
+            for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((org.bukkit.inventory.InventoryHolder) p_143375_.getBukkitEntity()).getInventory().getViewers())) { // TODO : CatServer - wait support
+               h.closeInventory();
+            }
+         }
+         // Spigot End
          ServerLevel.this.m_7726_().m_8443_(p_143375_);
          if (p_143375_ instanceof ServerPlayer serverplayer) {
             ServerLevel.this.f_8546_.remove(serverplayer);
@@ -1500,17 +_,33 @@
             ServerLevel.this.f_143246_.remove(mob);
          }
 
-         if (p_143375_ instanceof EnderDragon enderdragon) {
-            for(EnderDragonPart enderdragonpart : enderdragon.m_31156_()) {
+         if (p_143375_.isMultipartEntity()) {
+            for(net.minecraftforge.entity.PartEntity<?> enderdragonpart : p_143375_.getParts()) {
                ServerLevel.this.f_143247_.remove(enderdragonpart.m_19879_());
             }
          }
 
          p_143375_.m_213651_(DynamicGameEventListener::m_223634_);
+         // CraftBukkit start
+         p_143375_.valid = false;
+         if (!(p_143375_ instanceof ServerPlayer)) {
+            for (ServerPlayer player : f_8546_) {
+               player.getBukkitEntity().onEntityRemove(p_143375_);
+            }
+         }
+         // CraftBukkit end
+
+         p_143375_.onRemovedFromWorld();
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityLeaveLevelEvent(p_143375_, ServerLevel.this));
       }
 
       public void m_214006_(Entity p_215086_) {
          p_215086_.m_213651_(DynamicGameEventListener::m_223641_);
       }
+   }
+
+   @Override
+   public java.util.Collection<net.minecraftforge.entity.PartEntity<?>> getPartEntities() {
+      return this.f_143247_.values();
    }
 }
