--- a/net/minecraft/server/level/ServerBossEvent.java
+++ b/net/minecraft/server/level/ServerBossEvent.java
@@ -80,7 +_,7 @@
 
    }
 
-   private void m_143224_(Function<BossEvent, ClientboundBossEventPacket> p_143225_) {
+   public void m_143224_(Function<BossEvent, ClientboundBossEventPacket> p_143225_) {
       if (this.f_8298_) {
          ClientboundBossEventPacket clientboundbosseventpacket = p_143225_.apply(this);
 
