--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -40,6 +_,12 @@
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.server.network.ServerPlayerConnection;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerVelocityEvent;
+// CraftBukkit end
+
 public class ServerEntity {
    private static final Logger f_8508_ = LogUtils.getLogger();
    private static final int f_143241_ = 1;
@@ -60,8 +_,12 @@
    private boolean f_8525_;
    @Nullable
    private List<SynchedEntityData.DataValue<?>> f_263120_;
+   // CraftBukkit start
+   private final Set<ServerPlayerConnection> trackedPlayers;
 
-   public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_) {
+   public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_, Set<ServerPlayerConnection> trackedPlayers) {
+      this.trackedPlayers = trackedPlayers;
+      // CraftBukkit end
       this.f_8509_ = p_8528_;
       this.f_8513_ = p_8532_;
       this.f_8510_ = p_8529_;
@@ -78,7 +_,7 @@
    public void m_8533_() {
       List<Entity> list = this.f_8510_.m_20197_();
       if (!list.equals(this.f_8523_)) {
-         this.f_8513_.accept(new ClientboundSetPassengersPacket(this.f_8510_));
+         this.m_8538_(new ClientboundSetPassengersPacket(this.f_8510_)); // CraftBukkit
          m_277180_(list, this.f_8523_).forEach((p_289307_) -> {
             if (p_289307_ instanceof ServerPlayer serverplayer1) {
                serverplayer1.f_8906_.m_9774_(serverplayer1.m_20185_(), serverplayer1.m_20186_(), serverplayer1.m_20189_(), serverplayer1.m_146908_(), serverplayer1.m_146909_());
@@ -92,11 +_,11 @@
       if ($$11 instanceof ItemFrame itemframe) {
          if (this.f_8521_ % 10 == 0) {
             ItemStack itemstack = itemframe.m_31822_();
-            if (itemstack.m_41720_() instanceof MapItem) {
-               Integer integer = MapItem.m_151131_(itemstack);
-               MapItemSavedData mapitemsaveddata = MapItem.m_151128_(integer, this.f_8509_);
-               if (mapitemsaveddata != null) {
-                  for(ServerPlayer serverplayer : this.f_8509_.m_6907_()) {
+            Integer integer = MapItem.m_151131_(itemstack);
+            MapItemSavedData mapitemsaveddata = MapItem.m_42853_(itemstack, this.f_8509_);
+            if (mapitemsaveddata != null) {
+               for(ServerPlayerConnection trackedPlayer : this.trackedPlayers) { // CraftBukkit
+                     ServerPlayer serverplayer = trackedPlayer.m_142253_(); // CatServer
                      mapitemsaveddata.m_77918_(serverplayer, itemstack);
                      Packet<?> packet = mapitemsaveddata.m_164796_(integer, serverplayer);
                      if (packet != null) {
@@ -104,11 +_,12 @@
                      }
                   }
                }
-            }
 
             this.m_8543_();
          }
       }
+
+
 
       if (this.f_8521_ % this.f_8511_ == 0 || this.f_8510_.f_19812_ || this.f_8510_.m_20088_().m_135352_()) {
          if (this.f_8510_.m_20159_()) {
@@ -200,7 +_,23 @@
 
       ++this.f_8521_;
       if (this.f_8510_.f_19864_) {
-         this.m_8538_(new ClientboundSetEntityMotionPacket(this.f_8510_));
+         // CraftBukkit start - Create PlayerVelocity event
+         boolean cancelled = false;
+         if (this.f_8510_ instanceof ServerPlayer) {
+            Player player = (Player) this.f_8510_.getBukkitEntity();
+            org.bukkit.util.Vector velocity = player.getVelocity();
+            PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+            this.f_8510_.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               cancelled = true;
+            } else if (!velocity.equals(event.getVelocity())) {
+               player.setVelocity(event.getVelocity());
+            }
+         }
+         if (!cancelled) {
+            this.m_8538_(new ClientboundSetEntityMotionPacket(this.f_8510_));
+         }
+         // CraftBukkit end
          this.f_8510_.f_19864_ = false;
       }
 
@@ -215,6 +_,7 @@
    public void m_8534_(ServerPlayer p_8535_) {
       this.f_8510_.m_6452_(p_8535_);
       p_8535_.f_8906_.m_9829_(new ClientboundRemoveEntitiesPacket(this.f_8510_.m_19879_()));
+      net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(this.f_8510_, p_8535_);
    }
 
    public void m_8541_(ServerPlayer p_8542_) {
@@ -222,11 +_,15 @@
       this.m_289200_(p_8542_, list::add);
       p_8542_.f_8906_.m_9829_(new ClientboundBundlePacket(list));
       this.f_8510_.m_6457_(p_8542_);
+      net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(this.f_8510_, p_8542_);
    }
 
    public void m_289200_(ServerPlayer p_289562_, Consumer<Packet<ClientGamePacketListener>> p_289563_) {
       if (this.f_8510_.m_213877_()) {
-         f_8508_.warn("Fetching packet for removed entity {}", (Object)this.f_8510_);
+         // CraftBukkit start - Remove useless error spam, just return
+         // EntityTrackerEntry.LOGGER.warn("Fetching packet for removed entity {}", this.entity);
+         return;
+         // CraftBukkit end
       }
 
       Packet<ClientGamePacketListener> packet = this.f_8510_.m_5654_();
@@ -239,6 +_,13 @@
       boolean flag = this.f_8512_;
       if (this.f_8510_ instanceof LivingEntity) {
          Collection<AttributeInstance> collection = ((LivingEntity)this.f_8510_).m_21204_().m_22170_();
+
+         // CraftBukkit start - If sending own attributes send scaled health instead of current maximum health
+         if (this.f_8510_.m_19879_() == p_289562_.m_19879_()) {
+            ((ServerPlayer) this.f_8510_).getBukkitEntity().injectScaledMaxHealth(collection, false);
+         }
+         // CraftBukkit end
+
          if (!collection.isEmpty()) {
             p_289563_.accept(new ClientboundUpdateAttributesPacket(this.f_8510_.m_19879_(), collection));
          }
@@ -266,7 +_,14 @@
          if (!list.isEmpty()) {
             p_289563_.accept(new ClientboundSetEquipmentPacket(this.f_8510_.m_19879_(), list));
          }
-      }
+         ((LivingEntity) this.f_8510_).m_21315_(); // CraftBukkit - SPIGOT-3789: sync again immediately after sending
+      }
+
+      // CraftBukkit start - MC-109346: Fix for nonsensical head yaw
+      if (this.f_8510_ instanceof ServerPlayer) {
+         p_289563_.accept(new ClientboundRotateHeadPacket(this.f_8510_, (byte) Mth.m_14143_(this.f_8510_.m_6080_() * 256.0F / 360.0F)));
+      }
+      // CraftBukkit end
 
       if (!this.f_8510_.m_20197_().isEmpty()) {
          p_289563_.accept(new ClientboundSetPassengersPacket(this.f_8510_));
@@ -296,6 +_,11 @@
       if (this.f_8510_ instanceof LivingEntity) {
          Set<AttributeInstance> set = ((LivingEntity)this.f_8510_).m_21204_().m_22145_();
          if (!set.isEmpty()) {
+            // CraftBukkit start - Send scaled max health
+            if (this.f_8510_ instanceof ServerPlayer) {
+               ((ServerPlayer) this.f_8510_).getBukkitEntity().injectScaledMaxHealth(set, false);
+            }
+            // CraftBukkit end
             this.m_8538_(new ClientboundUpdateAttributesPacket(this.f_8510_.m_19879_(), set));
          }
 
