--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -9,7 +_,7 @@
 public class TicketType<T> {
    private final String f_9450_;
    private final Comparator<T> f_9451_;
-   private final long f_9452_;
+   public long f_9452_;
    public static final TicketType<Unit> f_9442_ = m_9462_("start", (p_9471_, p_9472_) -> {
       return 0;
    });
@@ -22,6 +_,8 @@
    public static final TicketType<BlockPos> f_9447_ = m_9465_("portal", Vec3i::compareTo, 300);
    public static final TicketType<Integer> f_9448_ = m_9465_("post_teleport", Integer::compareTo, 5);
    public static final TicketType<ChunkPos> f_9449_ = m_9465_("unknown", Comparator.comparingLong(ChunkPos::m_45588_), 1);
+   public static final TicketType<Unit> PLUGIN = m_9462_("plugin", (a, b) -> 0); // CraftBukkit
+   public static final TicketType<org.bukkit.plugin.Plugin> PLUGIN_TICKET = m_9462_("plugin_ticket", Comparator.comparing(plugin -> plugin.getClass().getName())); // CraftBukkit
 
    public static <T> TicketType<T> m_9462_(String p_9463_, Comparator<T> p_9464_) {
       return new TicketType<>(p_9463_, p_9464_, 0L);
