--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -5,6 +_,7 @@
 import java.util.List;
 import java.util.Locale;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
@@ -190,7 +_,7 @@
       if (blockstate.m_60795_()) {
          return false;
       } else {
-         if (p_9551_) {
+         if (p_9551_) { // CraftBukkit - SPIGOT-6833: Do not drop during world generation // CatServer - recover
             BlockEntity blockentity = blockstate.m_155947_() ? this.m_7702_(p_9550_) : null;
             Block.m_49881_(blockstate, this.f_9479_, p_9550_, blockentity, p_9552_, ItemStack.f_41583_);
          }
@@ -226,7 +_,7 @@
          }
 
          if (blockstate.m_155947_()) {
-            f_9474_.warn("Tried to access a block entity before it was created. {}", (Object)p_9582_);
+            f_9474_.debug("Tried to access a block entity before it was created. {}", (Object)p_9582_); // CatServer - warn -> debug
          }
 
          return null;
@@ -297,6 +_,14 @@
    }
 
    public boolean m_7967_(Entity p_9580_) {
+      // CraftBukkit start
+      return addFreshEntity(p_9580_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   @Override
+   public boolean addFreshEntity(Entity p_9580_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      if (p_9580_ instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
+      // CraftBukkit end
       int i = SectionPos.m_123171_(p_9580_.m_146903_());
       int j = SectionPos.m_123171_(p_9580_.m_146907_());
       this.m_6325_(i, j).m_6286_(p_9580_);
