--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -5,6 +_,7 @@
 import java.util.List;
 import java.util.Random;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
@@ -50,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.ticks.LevelTickAccess;
 import net.minecraft.world.ticks.WorldGenTickAccess;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.slf4j.Logger;
 
 public class WorldGenRegion implements WorldGenLevel {
@@ -217,7 +_,7 @@
          }
 
          if (blockstate.m_155947_()) {
-            f_9474_.warn("Tried to access a block entity before it was created. {}", (Object)p_9582_);
+            f_9474_.debug("Tried to access a block entity before it was created. {}", (Object)p_9582_); // CatServer - warn -> debug
          }
 
          return null;
@@ -287,12 +_,18 @@
       this.m_46865_(p_9592_).m_8113_(p_9592_);
    }
 
+   // CraftBukkit start
    public boolean m_7967_(Entity p_9580_) {
+      return addFreshEntity(p_9580_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   public boolean addFreshEntity(Entity p_9580_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
       int i = SectionPos.m_123171_(p_9580_.m_146903_());
       int j = SectionPos.m_123171_(p_9580_.m_146907_());
       this.m_6325_(i, j).m_6286_(p_9580_);
       return true;
    }
+   // CraftBukkit end
 
    public boolean m_7471_(BlockPos p_9547_, boolean p_9548_) {
       return this.m_7731_(p_9547_, Blocks.f_50016_.m_49966_(), 3);
