--- a/net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ b/net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -58,7 +_,7 @@
       if (p_44639_ instanceof LivingEntity livingentity) {
          if (this.f_44621_ == 2 && p_44640_ > 0 && livingentity.m_6336_() == MobType.f_21642_) {
             int i = 20 + p_44638_.m_217043_().m_188503_(10 * p_44640_);
-            livingentity.m_7292_(new MobEffectInstance(MobEffects.f_19597_, i, 3));
+            livingentity.addEffect(new MobEffectInstance(MobEffects.f_19597_, i, 3), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
          }
       }
 
