--- a/net/minecraft/world/item/enchantment/Enchantments.java
+++ b/net/minecraft/world/item/enchantment/Enchantments.java
@@ -46,6 +_,7 @@
    public static final Enchantment f_44962_ = m_44992_("mending", new MendingEnchantment(Enchantment.Rarity.RARE, EquipmentSlot.values()));
    public static final Enchantment f_44963_ = m_44992_("vanishing_curse", new VanishingCurseEnchantment(Enchantment.Rarity.VERY_RARE, EquipmentSlot.values()));
 
+   // CatServer - bukkit patches moved to BukkitInjector#registerEnchantments
    private static Enchantment m_44992_(String p_44993_, Enchantment p_44994_) {
       return Registry.m_122961_(BuiltInRegistries.f_256876_, p_44993_, p_44994_);
    }
