--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -3,6 +_,10 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftCampfireRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey;
 
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
    public CampfireCookingRecipe(ResourceLocation p_43822_, String p_43823_, Ingredient p_43824_, ItemStack p_43825_, float p_43826_, int p_43827_) {
@@ -16,4 +_,16 @@
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44094_;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+
+      CraftCampfireRecipe recipe = new CraftCampfireRecipe(CraftNamespacedKey.fromMinecraft(this.f_43727_), result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
