--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -4,6 +_,14 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftCampfireRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
+// CraftBukkit end
+
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
    public CampfireCookingRecipe(ResourceLocation p_249468_, String p_250140_, CookingBookCategory p_251808_, Ingredient p_249826_, ItemStack p_251839_, float p_251432_, int p_251471_) {
       super(RecipeType.f_44111_, p_249468_, p_250140_, p_251808_, p_249826_, p_251839_, p_251432_, p_251471_);
@@ -16,4 +_,15 @@
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44094_;
    }
+
+   // CraftBukkit start
+   @Override
+   public Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+      CraftCampfireRecipe recipe = new CraftCampfireRecipe(CraftNamespacedKey.fromMinecraft(this.f_43727_), result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+      recipe.setCategory(CraftRecipe.getCategory(this.m_245534_()));
+      return recipe;
+   }
+   // CraftBukkit end
 }
