--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -21,4 +_,11 @@
    public ItemStack m_8043_() {
       return ItemStack.f_41583_;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      return new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftComplexRecipe(this);
+   }
+   // CraftBukkit end
 }
