--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -28,4 +_,11 @@
    public CraftingBookCategory m_245232_() {
       return this.f_244184_;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      return new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftComplexRecipe(this);
+   }
+   // CraftBukkit end
 }
