--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -4,6 +_,13 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftFurnaceRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+// CraftBukkit end
+
 public class SmeltingRecipe extends AbstractCookingRecipe {
    public SmeltingRecipe(ResourceLocation p_249157_, String p_250200_, CookingBookCategory p_251114_, Ingredient p_250340_, ItemStack p_250306_, float p_249577_, int p_250030_) {
       super(RecipeType.f_44108_, p_249157_, p_250200_, p_251114_, p_250340_, p_250306_, p_249577_, p_250030_);
@@ -16,4 +_,15 @@
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44091_;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+      CraftFurnaceRecipe recipe = new CraftFurnaceRecipe(CraftNamespacedKey.fromMinecraft(this.f_43727_), result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+      recipe.setCategory(CraftRecipe.getCategory(this.m_245534_()));
+      return recipe;
+   }
+   // CraftBukkit end
 }
