--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -3,6 +_,10 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftFurnaceRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey;
 
 public class SmeltingRecipe extends AbstractCookingRecipe {
    public SmeltingRecipe(ResourceLocation p_44460_, String p_44461_, Ingredient p_44462_, ItemStack p_44463_, float p_44464_, int p_44465_) {
@@ -15,5 +_,13 @@
 
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44091_;
+   }
+
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+      CraftFurnaceRecipe recipe = new CraftFurnaceRecipe(CraftNamespacedKey.fromMinecraft(this.f_43727_), result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+      return recipe;
    }
 }
