--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -19,6 +_,12 @@
 import net.minecraft.world.item.armortrim.TrimPatterns;
 import net.minecraft.world.level.Level;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftSmithingTrimRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+// CraftBukkit end
+
 public class SmithingTrimRecipe implements SmithingRecipe {
    private final ResourceLocation f_265885_;
    final Ingredient f_265958_;
@@ -94,8 +_,15 @@
    }
 
    public boolean m_142505_() {
-      return Stream.of(this.f_265958_, this.f_266040_, this.f_266053_).anyMatch(Ingredient::m_43947_);
-   }
+      return Stream.of(this.f_265958_, this.f_266040_, this.f_266053_).anyMatch(net.minecraftforge.common.ForgeHooks::hasNoElements);
+   }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      return new CraftSmithingTrimRecipe(CraftNamespacedKey.fromMinecraft(this.f_265885_), CraftRecipe.toBukkit(this.f_265958_), CraftRecipe.toBukkit(this.f_266040_), CraftRecipe.toBukkit(this.f_266053_));
+   }
+   // CraftBukkit end
 
    public static class Serializer implements RecipeSerializer<SmithingTrimRecipe> {
       public SmithingTrimRecipe m_6729_(ResourceLocation p_267037_, JsonObject p_267004_) {
