--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -11,6 +_,13 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftSmithingTransformRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+// CraftBukkit end
+
 public class SmithingTransformRecipe implements SmithingRecipe {
    private final ResourceLocation f_265924_;
    final Ingredient f_265949_;
@@ -65,8 +_,19 @@
    }
 
    public boolean m_142505_() {
-      return Stream.of(this.f_265949_, this.f_265888_, this.f_265907_).anyMatch(Ingredient::m_43947_);
-   }
+      return Stream.of(this.f_265949_, this.f_265888_, this.f_265907_).anyMatch(net.minecraftforge.common.ForgeHooks::hasNoElements);
+   }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_266098_);
+
+      CraftSmithingTransformRecipe recipe = new CraftSmithingTransformRecipe(CraftNamespacedKey.fromMinecraft(this.f_265924_), result, CraftRecipe.toBukkit(this.f_265949_), CraftRecipe.toBukkit(this.f_265888_), CraftRecipe.toBukkit(this.f_265907_));
+
+      return recipe;
+   }
+   // CraftBukkit end
 
    public static class Serializer implements RecipeSerializer<SmithingTransformRecipe> {
       public SmithingTransformRecipe m_6729_(ResourceLocation p_266953_, JsonObject p_266720_) {
