--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -5,6 +_,10 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftStonecuttingRecipe;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey;
 
 public class StonecutterRecipe extends SingleItemRecipe {
    public StonecutterRecipe(ResourceLocation p_44478_, String p_44479_, Ingredient p_44480_, ItemStack p_44481_) {
@@ -18,4 +_,16 @@
    public ItemStack m_8042_() {
       return new ItemStack(Blocks.f_50679_);
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_44410_);
+
+      CraftStonecuttingRecipe recipe = new CraftStonecuttingRecipe(CraftNamespacedKey.fromMinecraft(this.f_44411_), result, CraftRecipe.toBukkit(this.f_44409_));
+      recipe.setGroup(this.f_44412_);
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
