--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -6,6 +_,13 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftStonecuttingRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey;
+// CraftBukkit end
+
 public class StonecutterRecipe extends SingleItemRecipe {
    public StonecutterRecipe(ResourceLocation p_44478_, String p_44479_, Ingredient p_44480_, ItemStack p_44481_) {
       super(RecipeType.f_44112_, RecipeSerializer.f_44095_, p_44478_, p_44479_, p_44480_, p_44481_);
@@ -18,4 +_,14 @@
    public ItemStack m_8042_() {
       return new ItemStack(Blocks.f_50679_);
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_44410_);
+      CraftStonecuttingRecipe recipe = new CraftStonecuttingRecipe(CraftNamespacedKey.fromMinecraft(this.f_44411_), result, CraftRecipe.toBukkit(this.f_44409_));
+      recipe.setGroup(this.f_44412_);
+      return recipe;
+   }
+   // CraftBukkit end
 }
