--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -30,7 +_,7 @@
    }
 
    public Component m_6418_(SocketAddress p_120007_, GameProfile p_120008_) {
-      return (Component)(p_120008_.getName().equalsIgnoreCase(this.m_7873_().m_129791_()) && this.m_11255_(p_120008_.getName()) != null ? new TranslatableComponent("multiplayer.disconnect.name_taken") : super.m_6418_(p_120007_, p_120008_));
+      return null;
    }
 
    public IntegratedServer m_7873_() {
