--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -7,6 +_,7 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.network.ServerLoginPacketListenerImpl;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.world.level.storage.PlayerDataStorage;
 import net.minecraftforge.api.distmarker.Dist;
@@ -29,8 +_,8 @@
       super.m_6765_(p_120011_);
    }
 
-   public Component m_6418_(SocketAddress p_120007_, GameProfile p_120008_) {
-      return (Component)(p_120008_.getName().equalsIgnoreCase(this.m_7873_().m_129791_()) && this.m_11255_(p_120008_.getName()) != null ? new TranslatableComponent("multiplayer.disconnect.name_taken") : super.m_6418_(p_120007_, p_120008_));
+   public ServerPlayer canPlayerLogin(ServerLoginPacketListenerImpl p_120007_, GameProfile p_120008_) {
+      return (ServerPlayer)(p_120008_.getName().equalsIgnoreCase(this.m_7873_().m_129791_()) && this.m_11255_(p_120008_.getName()) != null ? new TranslatableComponent("multiplayer.disconnect.name_taken") : super.canPlayerLogin(p_120007_, p_120008_, null));
    }
 
    public IntegratedServer m_7873_() {
