--- a/net/minecraft/client/server/IntegratedServer.java
+++ b/net/minecraft/client/server/IntegratedServer.java
@@ -16,6 +_,7 @@
 import net.minecraft.SharedConstants;
 import net.minecraft.SystemReport;
 import net.minecraft.client.Minecraft;
+import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.WorldStem;
 import net.minecraft.server.level.ServerPlayer;
@@ -29,6 +_,7 @@
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -47,7 +_,7 @@
    private int f_194467_ = 0;
 
    public IntegratedServer(Thread p_205653_, Minecraft p_205654_, LevelStorageSource.LevelStorageAccess p_205655_, PackRepository p_205656_, WorldStem p_205657_, MinecraftSessionService p_205658_, GameProfileRepository p_205659_, GameProfileCache p_205660_, ChunkProgressListenerFactory p_205661_) {
-      super(p_205653_, p_205655_, p_205656_, p_205657_, p_205654_.m_91096_(), p_205654_.m_91295_(), p_205658_, p_205659_, p_205660_, p_205661_);
+      super(null, null, null, p_205653_, p_205655_, p_205656_, p_205657_, p_205654_.m_91096_(), p_205654_.m_91295_(), p_205658_, p_205659_, p_205660_, p_205661_);
       this.m_129981_(p_205654_.m_91094_().m_92546_());
       this.m_129975_(p_205654_.m_91402_());
       this.m_129823_(new IntegratedPlayerList(this, this.m_206579_(), this.f_129745_));
@@ -60,9 +_,10 @@
       this.m_129997_(true);
       this.m_129999_(true);
       this.m_129793_();
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.m_130006_();
       this.m_129989_(this.m_129791_() + " - " + this.m_129910_().m_5462_());
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void m_5705_(BooleanSupplier p_120049_) {
@@ -112,6 +_,11 @@
       return true;
    }
 
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return null;
+   }
+
    public File m_6237_() {
       return this.f_120015_.f_91069_;
    }
@@ -179,10 +_,11 @@
    }
 
    public void m_7570_(boolean p_120053_) {
+      if (m_130010_())
       this.m_18709_(() -> {
          for(ServerPlayer serverplayer : Lists.newArrayList(this.m_6846_().m_11314_())) {
             if (!serverplayer.m_142081_().equals(this.f_120019_)) {
-               this.m_6846_().m_11286_(serverplayer);
+               this.m_6846_().remove(serverplayer);
             }
          }
 
