--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -30,7 +_,7 @@
    }
 
    public Component m_6418_(SocketAddress p_120007_, GameProfile p_120008_) {
-      return (Component)(this.m_7873_().m_7779_(p_120008_) && this.m_11255_(p_120008_.getName()) != null ? Component.m_237115_("multiplayer.disconnect.name_taken") : super.m_6418_(p_120007_, p_120008_));
+      return null; // CatServer - null
    }
 
    public IntegratedServer m_7873_() {
