--- a/net/minecraft/client/server/LanServerDetection.java
+++ b/net/minecraft/client/server/LanServerDetection.java
@@ -2,6 +_,11 @@
 
 import com.google.common.collect.Lists;
 import com.mojang.logging.LogUtils;
+import net.minecraft.DefaultUncaughtExceptionHandler;
+import net.minecraftforge.api.distmarker.Dist;
+import net.minecraftforge.api.distmarker.OnlyIn;
+import org.slf4j.Logger;
+
 import java.io.IOException;
 import java.net.DatagramPacket;
 import java.net.InetAddress;
@@ -11,10 +_,6 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicInteger;
-import net.minecraft.DefaultUncaughtExceptionHandler;
-import net.minecraftforge.api.distmarker.Dist;
-import net.minecraftforge.api.distmarker.OnlyIn;
-import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
 public class LanServerDetection {
@@ -88,7 +_,17 @@
          String s = LanServerPinger.m_120111_(p_120097_);
          String s1 = LanServerPinger.m_120116_(p_120097_);
          if (s1 != null) {
-            s1 = p_120098_.getHostAddress() + ":" + s1;
+            if (net.minecraftforge.network.DualStackUtils.checkIPv6(p_120098_)) {
+               final String ip;
+               // compress to short form if enabled in config
+               if (net.minecraftforge.common.ForgeConfig.CLIENT.compressLanIPv6Addresses.get())
+                  ip = com.google.common.net.InetAddresses.toAddrString(p_120098_);
+               else
+                  ip = p_120098_.getHostAddress();
+               s1 = "[" + ip + "]:" + s1;
+            } else {
+               s1 = p_120098_.getHostAddress() + ":" + s1;
+            }
             boolean flag = false;
 
             for(LanServer lanserver : this.f_120092_) {
