--- ../src-base/minecraft/net/minecraft/server/network/NetHandlerHandshakeTCP.java
+++ ../src-work/minecraft/net/minecraft/server/network/NetHandlerHandshakeTCP.java
@@ -31,13 +31,13 @@
 
                 if (packetIn.getProtocolVersion() > 210)
                 {
-                    TextComponentString textcomponentstring = new TextComponentString("Outdated server! I\'m still on 1.10.2");
+                    TextComponentString textcomponentstring = new TextComponentString( java.text.MessageFormat.format( org.spigotmc.SpigotConfig.outdatedServerMessage.replaceAll("'", "''"), "1.10.2" ) ); // Spigot
                     this.networkManager.sendPacket(new SPacketDisconnect(textcomponentstring));
                     this.networkManager.closeChannel(textcomponentstring);
                 }
                 else if (packetIn.getProtocolVersion() < 210)
                 {
-                    TextComponentString textcomponentstring1 = new TextComponentString("Outdated client! Please use 1.10.2");
+                    TextComponentString textcomponentstring1 = new TextComponentString( java.text.MessageFormat.format( org.spigotmc.SpigotConfig.outdatedClientMessage.replaceAll("'", "''"), "1.10.2" ) ); // Spigot
                     this.networkManager.sendPacket(new SPacketDisconnect(textcomponentstring1));
                     this.networkManager.closeChannel(textcomponentstring1);
                 }
