--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
@@ -1,9 +1,16 @@
 package net.minecraft.inventory;
 
 import javax.annotation.Nullable;
+
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
+import org.bukkit.inventory.PlayerInventory;
+
 import net.minecraft.entity.passive.EntityHorse;
 import net.minecraft.entity.passive.HorseArmorType;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.init.Items;
 import net.minecraft.item.ItemStack;
 import net.minecraftforge.fml.relauncher.Side;
@@ -13,9 +20,24 @@
 {
     private final IInventory horseInventory;
     private final EntityHorse theHorse;
+    // CraftBukkit start
+    CraftInventoryView bukkitEntity;
+    InventoryPlayer player;
 
+    @Override
+    public InventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryHorse(this.horseInventory);
+        return bukkitEntity = new CraftInventoryView(player.player.getBukkitEntity(), inventory, this);
+    }
+
     public ContainerHorseInventory(IInventory playerInventory, final IInventory horseInventoryIn, final EntityHorse horse, EntityPlayer player)
     {
+        this.player = (InventoryPlayer) playerInventory;
+        // CraftBukkit end
         this.horseInventory = horseInventoryIn;
         this.theHorse = horse;
         int i = 3;
