--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
@@ -3,18 +3,25 @@
 import net.minecraft.entity.passive.AbstractChestHorse;
 import net.minecraft.entity.passive.AbstractHorse;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.init.Items;
 import net.minecraft.item.ItemStack;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class ContainerHorseInventory extends Container
 {
     private final IInventory horseInventory;
     private final AbstractHorse horse;
 
+    org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity;
+    InventoryPlayer player;
+
     public ContainerHorseInventory(IInventory playerInventory, IInventory horseInventoryIn, final AbstractHorse horse, EntityPlayer player)
     {
+        this.player = (InventoryPlayer) playerInventory;
         this.horseInventory = horseInventoryIn;
         this.horse = horse;
         int i = 3;
@@ -133,4 +140,14 @@
         super.onContainerClosed(playerIn);
         this.horseInventory.closeInventory(playerIn);
     }
+
+    @Override
+    public InventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        return bukkitEntity = (horseInventory.getOwner() != null ? new CraftInventoryView(player.player.getBukkitEntity(), horseInventory.getOwner().getInventory(), this) : null); // CatServer
+    }
+
 }
