--- ../src-base/minecraft/net/minecraft/inventory/EntityEquipmentSlot.java
+++ ../src-work/minecraft/net/minecraft/inventory/EntityEquipmentSlot.java
@@ -2,19 +2,19 @@
 
 public enum EntityEquipmentSlot
 {
-    MAINHAND(EntityEquipmentSlot.Type.HAND, 0, 0, "mainhand"),
-    OFFHAND(EntityEquipmentSlot.Type.HAND, 1, 5, "offhand"),
-    FEET(EntityEquipmentSlot.Type.ARMOR, 0, 1, "feet"),
-    LEGS(EntityEquipmentSlot.Type.ARMOR, 1, 2, "legs"),
-    CHEST(EntityEquipmentSlot.Type.ARMOR, 2, 3, "chest"),
-    HEAD(EntityEquipmentSlot.Type.ARMOR, 3, 4, "head");
+    MAINHAND(Type.HAND, 0, 0, "mainhand"),
+    OFFHAND(Type.HAND, 1, 5, "offhand"),
+    FEET(Type.ARMOR, 0, 1, "feet"),
+    LEGS(Type.ARMOR, 1, 2, "legs"),
+    CHEST(Type.ARMOR, 2, 3, "chest"),
+    HEAD(Type.ARMOR, 3, 4, "head");
 
-    private final EntityEquipmentSlot.Type slotType;
+    private final Type slotType;
     private final int index;
     private final int slotIndex;
     private final String name;
 
-    private EntityEquipmentSlot(EntityEquipmentSlot.Type slotTypeIn, int indexIn, int slotIndexIn, String nameIn)
+    private EntityEquipmentSlot(Type slotTypeIn, int indexIn, int slotIndexIn, String nameIn)
     {
         this.slotType = slotTypeIn;
         this.index = indexIn;
@@ -22,7 +22,7 @@
         this.name = nameIn;
     }
 
-    public EntityEquipmentSlot.Type getSlotType()
+    public Type getSlotType()
     {
         return this.slotType;
     }
