--- ../src-base/minecraft/net/minecraft/inventory/IInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/IInventory.java
@@ -1,11 +1,12 @@
 package net.minecraft.inventory;
 
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.IWorldNameable;
 
-public interface IInventory extends IWorldNameable
-{
+public interface IInventory extends IWorldNameable {
     int getSizeInventory();
 
     boolean isEmpty();
@@ -37,4 +38,29 @@
     int getFieldCount();
 
     void clear();
+
+    // CraftBukkit start
+    default java.util.List<ItemStack> getContents() { return java.util.Collections.emptyList(); }
+
+    default void onOpen(CraftHumanEntity who) {}
+
+    default void onClose(CraftHumanEntity who) {}
+
+    default java.util.List<org.bukkit.entity.HumanEntity> getViewers() { return new java.util.ArrayList<org.bukkit.entity.HumanEntity>(); }
+
+    default org.bukkit.inventory.InventoryHolder getOwner() { return (this instanceof net.minecraft.tileentity.TileEntity) ? catserver.server.inventory.CatInventoryUtils.getOwner((net.minecraft.tileentity.TileEntity) this) : null; }
+
+    default void setMaxStackSize(int size) {}
+
+    default org.bukkit.Location getLocation() {
+        if (this instanceof net.minecraft.tileentity.TileEntity) {
+            net.minecraft.util.math.BlockPos pos = ((net.minecraft.tileentity.TileEntity) this).getPos();
+            return new org.bukkit.Location(((net.minecraft.tileentity.TileEntity) this).getWorld().getWorld(), pos.getX(), pos.getY(), pos.getZ());
+        } else {
+            return null;
+        }
+    }
+
+    int MAX_STACK = 64;
+    // CraftBukkit end
 }
