--- ../src-base/minecraft/net/minecraft/inventory/InventoryCraftResult.java
+++ ../src-work/minecraft/net/minecraft/inventory/InventoryCraftResult.java
@@ -7,8 +7,47 @@
 import net.minecraft.util.text.TextComponentString;
 import net.minecraft.util.text.TextComponentTranslation;
 
-public class InventoryCraftResult implements IInventory
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+
+import luohuayu.CatServer.inventory.ICBInventory;
+
+//CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
+public class InventoryCraftResult implements ICBInventory
 {
+	// CraftBukkit start
+	private int maxStack = MAX_STACK;
+	
+	public ItemStack[] getContents() {
+		return this.stackResult;
+	}
+	
+	public org.bukkit.inventory.InventoryHolder getOwner() {
+		return null; // Result slots don't get an owner
+	}
+	
+	// Don't need a transaction; the InventoryCrafting keeps track of it for us
+	public void onOpen(CraftHumanEntity who) {}
+	public void onClose(CraftHumanEntity who) {}
+	public java.util.List<HumanEntity> getViewers() {
+		return new java.util.ArrayList<HumanEntity>();
+	}
+	
+	public void setMaxStackSize(int size) {
+		maxStack = size;
+	}
+	
+	@Override
+	public Location getLocation() {
+		return null;
+	}
+	// CraftBukkit end
     private final ItemStack[] stackResult = new ItemStack[1];
 
     public int getSizeInventory()
