--- ../src-base/minecraft/net/minecraft/village/MerchantRecipe.java
+++ ../src-work/minecraft/net/minecraft/village/MerchantRecipe.java
@@ -6,15 +6,30 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe; // CraftBukkit
 
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
+
 public class MerchantRecipe
 {
-    private ItemStack itemToBuy;
-    private ItemStack secondItemToBuy;
-    private ItemStack itemToSell;
-    private int toolUses;
-    private int maxTradeUses;
-    private boolean rewardsExp;
+    public ItemStack itemToBuy;
+    public ItemStack secondItemToBuy;
+    public ItemStack itemToSell;
+    public int toolUses;
+    public int maxTradeUses;
+    public boolean rewardsExp;
+    // CraftBukkit start
+    private CraftMerchantRecipe bukkitHandle;
+    
+    public CraftMerchantRecipe asBukkit() {
+    	return (bukkitHandle == null) ? bukkitHandle = new CraftMerchantRecipe(this) : bukkitHandle;
+    }
+    
+    public MerchantRecipe(ItemStack itemstack, ItemStack itemstack1, ItemStack itemstack2, int i, int j, CraftMerchantRecipe bukkit) {
+    	this(itemstack, itemstack1, itemstack2, i, j);
+    	this.bukkitHandle = bukkit;
+    }
+    // CraftBukkit end
 
     public MerchantRecipe(NBTTagCompound tagCompound)
     {
