--- ../src-base/minecraft/net/minecraft/village/VillageCollection.java
+++ ../src-work/minecraft/net/minecraft/village/VillageCollection.java
@@ -38,9 +38,11 @@
     public void setWorldsForAll(World worldIn)
     {
         this.worldObj = worldIn;
+        Iterator iterator = this.villageList.iterator();
 
-        for (Village village : this.villageList)
+        while (iterator.hasNext())
         {
+            Village village = (Village)iterator.next();
             village.setWorld(worldIn);
         }
     }
@@ -59,9 +61,11 @@
     public void tick()
     {
         ++this.tickCounter;
+        Iterator iterator = this.villageList.iterator();
 
-        for (Village village : this.villageList)
+        while (iterator.hasNext())
         {
+            Village village = (Village)iterator.next();
             village.tick(this.tickCounter);
         }
 
