--- a/net/minecraft/entity/passive/fish/AbstractFishEntity.java
+++ b/net/minecraft/entity/passive/fish/AbstractFishEntity.java
@@ -64,7 +_,7 @@
    }
 
    public boolean func_213397_c(double p_213397_1_) {
-      return !this.func_203705_dA() && !this.func_145818_k_();
+      return true; // CraftBukkit
    }
 
    public int func_70641_bl() {
@@ -82,6 +_,7 @@
 
    public void func_203706_r(boolean p_203706_1_) {
       this.field_70180_af.func_187227_b(field_203711_b, p_203706_1_);
+      this.field_82179_bU = this.func_104002_bU();  // CraftBukkit - SPIGOT-4106 update persistence
    }
 
    public void func_213281_b(CompoundNBT p_213281_1_) {
