--- a/net/minecraft/entity/passive/fish/PufferfishEntity.java
+++ b/net/minecraft/entity/passive/fish/PufferfishEntity.java
@@ -24,6 +_,7 @@
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class PufferfishEntity extends AbstractFishEntity {
    private static final DataParameter<Integer> field_203716_b = EntityDataManager.func_187226_a(PufferfishEntity.class, DataSerializers.field_187192_b);
@@ -126,7 +_,7 @@
    private void func_205719_a(MobEntity p_205719_1_) {
       int i = this.func_203715_dA();
       if (p_205719_1_.func_70097_a(DamageSource.func_76358_a(this), (float)(1 + i))) {
-         p_205719_1_.func_195064_c(new EffectInstance(Effects.field_76436_u, 60 * i, 0));
+         p_205719_1_.addEffect(new EffectInstance(Effects.field_76436_u, 60 * i, 0), Cause.ATTACK); // CraftBukkit
          this.func_184185_a(SoundEvents.field_203830_gs, 1.0F, 1.0F);
       }
 
@@ -139,7 +_,7 @@
             ((ServerPlayerEntity)p_70100_1_).field_71135_a.func_147359_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241773_j_, 0.0F));
          }
 
-         p_70100_1_.func_195064_c(new EffectInstance(Effects.field_76436_u, 60 * i, 0));
+         p_70100_1_.addEffect(new EffectInstance(Effects.field_76436_u, 60 * i, 0), Cause.ATTACK); // CraftBukkit
       }
 
    }
