--- ../src-base/minecraft/net/minecraftforge/fml/common/network/handshake/NetworkDispatcher.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/network/handshake/NetworkDispatcher.java
@@ -529,12 +529,10 @@
             else
             {
                 List<Packet<INetHandlerPlayClient>> parts = ((FMLProxyPacket)msg).toS3FPackets();
-                int sizeMinusOne = parts.size() - 1;
-                for (int i = 0; i < sizeMinusOne; i++)
+                for (Packet<INetHandlerPlayClient> pkt : parts)
                 {
-                    ctx.write(parts.get(i), ctx.voidPromise());
+                    ctx.write(pkt, promise);
                 }
-                ctx.write(parts.get(sizeMinusOne), promise);
             }
         }
         else
@@ -647,7 +645,7 @@
             {
                 throw new IOException("Received FML MultiPart packet out of order, Expected " + part_expected + " Got " + part);
             }
-            int len = input.readableBytes();
+            int len = input.readableBytes() - 1;
             input.readBytes(data, offset, len);
             part_expected++;
             offset += len;
