--- a/net/minecraft/world/level/block/state/StateHolder.java
+++ b/net/minecraft/world/level/block/state/StateHolder.java
@@ -20,7 +_,7 @@
 public abstract class StateHolder<O, S> {
    public static final String f_155962_ = "Name";
    public static final String f_155963_ = "Properties";
-   private static final Function<Map.Entry<Property<?>, Comparable<?>>, String> f_61110_ = new Function<Map.Entry<Property<?>, Comparable<?>>, String>() {
+   public static final Function<Map.Entry<Property<?>, Comparable<?>>, String> f_61110_ = new Function<Map.Entry<Property<?>, Comparable<?>>, String>() { // CatServer - private -> public
       public String apply(@Nullable Map.Entry<Property<?>, Comparable<?>> p_61155_) {
          if (p_61155_ == null) {
             return "<NULL>";
