--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -30,10 +_,10 @@
    private static final Logger f_156535_ = LogUtils.getLogger();
    private static final String f_156536_ = "Entities";
    private static final String f_156537_ = "Position";
-   private final ServerLevel f_156538_;
+   public final ServerLevel f_156538_;
    private final IOWorker f_156539_;
    private final LongSet f_156540_ = new LongOpenHashSet();
-   private final ProcessorMailbox<Runnable> f_182485_;
+   public final ProcessorMailbox<Runnable> f_182485_;
    protected final DataFixer f_156534_;
 
    public EntityStorage(ServerLevel p_196924_, Path p_196925_, DataFixer p_196926_, boolean p_196927_, Executor p_196928_) {
@@ -90,8 +_,12 @@
          ListTag listtag = new ListTag();
          p_156559_.m_156792_().forEach((p_156567_) -> {
             CompoundTag compoundtag1 = new CompoundTag();
+            try {
             if (p_156567_.m_20223_(compoundtag1)) {
                listtag.add(compoundtag1);
+            }
+            } catch (Exception e) {
+               f_156535_.error("An Entity type {} has thrown an exception trying to write state. It will not persist. Report this to the mod author", p_156567_.m_6095_(), e);
             }
 
          });
