--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -43,6 +_,7 @@
 import org.slf4j.Logger;
 
 public class DedicatedServerProperties extends Settings<DedicatedServerProperties> {
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
    static final Logger f_214805_ = LogUtils.getLogger();
    private static final Pattern f_214806_ = Pattern.compile("^[a-fA-F0-9]{40}$");
    private static final Splitter f_243988_ = Splitter.on(',').trimResults();
@@ -189,7 +_,7 @@
       return this.f_244504_.m_247373_(p_250466_);
    }
 
-   static record WorldDimensionData(JsonObject f_244404_, String f_243780_) {
+   public static record WorldDimensionData(JsonObject f_244404_, String f_243780_) { // CatServer - public
       private static final Map<String, ResourceKey<WorldPreset>> f_244358_ = Map.of("default", WorldPresets.f_226437_, "largebiomes", WorldPresets.f_226439_);
 
       public WorldDimensions m_247373_(RegistryAccess p_248812_) {
