--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -6,6 +_,8 @@
 import java.util.Properties;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
+
+import joptsimple.OptionSet;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.util.Mth;
@@ -77,6 +_,7 @@
    }, "default"));
    @Nullable
    private WorldGenSettings f_139727_;
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
 
    public DedicatedServerProperties(Properties p_180926_) {
       super(p_180926_);
