--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 
 public class DedicatedServerProperties extends Settings<DedicatedServerProperties> {
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
    public final boolean f_139728_ = this.m_139836_("online-mode", true);
    public final boolean f_139729_ = this.m_139836_("prevent-proxy-connections", false);
    public final String f_139730_ = this.m_139811_("server-ip", "");
