--- a/net/minecraft/server/dedicated/Settings.java
+++ b/net/minecraft/server/dedicated/Settings.java
@@ -25,7 +_,7 @@
 
 public abstract class Settings<T extends Settings<T>> {
    private static final Logger f_139797_ = LogUtils.getLogger();
-   protected final Properties f_139798_;
+   public final Properties f_139798_; // CatServer - protected -> public
 
    public Settings(Properties p_139801_) {
       this.f_139798_ = p_139801_;
@@ -33,6 +_,10 @@
 
    public static Properties m_139839_(Path p_139840_) {
       try {
+         // CraftBukkit start - 1.20.1 SPIGOT-7465, MC-264979: Don't load if file doesn't exist
+         if (!p_139840_.toFile().exists()) {
+            return new Properties();
+         }
          try (InputStream inputstream = Files.newInputStream(p_139840_)) {
             CharsetDecoder charsetdecoder = StandardCharsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(CodingErrorAction.REPORT);
             Properties properties1 = new Properties();
@@ -54,8 +_,12 @@
    }
 
    public void m_139876_(Path p_139877_) {
+      // CraftBukkit start - Don't attempt writing to file if it's read only
+      if (p_139877_.toFile().exists() && !p_139877_.toFile().canWrite()) {
+         return;
+      }
       try (Writer writer = Files.newBufferedWriter(p_139877_, StandardCharsets.UTF_8)) {
-         this.f_139798_.store(writer, "Minecraft server properties");
+         net.minecraftforge.common.util.SortedProperties.store(this.f_139798_, writer, "Minecraft server properties");
       } catch (IOException ioexception) {
          f_139797_.error("Failed to store properties to file: {}", (Object)p_139877_);
       }
