--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -17,11 +_,60 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 
+// CraftBukkit start
+import java.util.List;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
+
 public abstract class AbstractMinecartContainer extends AbstractMinecart implements ContainerEntity {
-   private NonNullList<ItemStack> f_38202_ = NonNullList.m_122780_(36, ItemStack.f_41583_);
+   private NonNullList<ItemStack> f_38202_ = NonNullList.m_122780_(this.m_6643_(), ItemStack.f_41583_); // CraftBukkit - SPIGOT-3513
    @Nullable
-   private ResourceLocation f_38204_;
-   private long f_38205_;
+   public ResourceLocation f_38204_; // CatServer - private -> public
+   public long f_38205_; // CatServer - private -> public
+
+   // CraftBukkit start
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private int maxStack = MAX_STACK;
+
+   public List<ItemStack> getContents() {
+      return this.f_38202_;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public InventoryHolder getOwner() {
+      org.bukkit.entity.Entity cart = getBukkitEntity();
+      if(cart instanceof InventoryHolder) return (InventoryHolder) cart;
+      return null;
+   }
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+   }
+
+   @Override
+   public Location getLocation() {
+      return getBukkitEntity().getLocation();
+   }
+   // CraftBukkit end
 
    protected AbstractMinecartContainer(EntityType<?> p_38213_, Level p_38214_) {
       super(p_38213_, p_38214_);
@@ -82,6 +_,8 @@
    }
 
    public InteractionResult m_6096_(Player p_38232_, InteractionHand p_38233_) {
+      InteractionResult ret = super.m_6096_(p_38232_, p_38233_);
+      if (ret.m_19077_()) return ret;
       return this.m_268996_(p_38232_);
    }
 
@@ -119,6 +_,28 @@
    }
 
    protected abstract AbstractContainerMenu m_7402_(int p_38222_, Inventory p_38223_);
+
+   // Forge Start
+   private net.minecraftforge.common.util.LazyOptional<?> itemHandler = net.minecraftforge.common.util.LazyOptional.of(() -> new net.minecraftforge.items.wrapper.InvWrapper(this));
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.core.Direction facing) {
+      if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && this.m_6084_())
+         return itemHandler.cast();
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void invalidateCaps() {
+      super.invalidateCaps();
+      itemHandler.invalidate();
+   }
+
+   @Override
+   public void reviveCaps() {
+      super.reviveCaps();
+      itemHandler = net.minecraftforge.common.util.LazyOptional.of(() -> new net.minecraftforge.items.wrapper.InvWrapper(this));
+   }
 
    @Nullable
    public ResourceLocation m_214142_() {
