--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -103,6 +_,9 @@
 
          this.m_214199_((ResourceLocation)null);
          LootParams.Builder lootparams$builder = (new LootParams.Builder((ServerLevel)this.m_9236_())).m_287286_(LootContextParams.f_81460_, this.m_20182_());
+         // Forge: set the chest to killer_entity for loot context.
+         if (this instanceof AbstractMinecartContainer entityContainer)
+            lootparams$builder.m_287286_(LootContextParams.f_81458_, entityContainer);
          if (p_219950_ != null) {
             lootparams$builder.m_287239_(p_219950_.m_36336_()).m_287286_(LootContextParams.f_81455_, p_219950_);
          }
@@ -171,6 +_,6 @@
    }
 
    default boolean m_219954_(Player p_219955_) {
-      return !this.m_213877_() && this.m_20182_().m_82509_(p_219955_.m_20182_(), 8.0D);
+      return !this.m_213877_() && this.m_20182_().m_82509_(p_219955_.m_20182_(), p_219955_.m_21133_(net.minecraftforge.common.ForgeMod.ENTITY_REACH.get()) + 5.0); // Vanilla uses 8 default reach is 3.0 so add 5.0 padding
    }
 }
