--- a/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
 
 public class MinecartCommandBlock extends AbstractMinecart {
    public static final EntityDataAccessor<String> f_38503_ = SynchedEntityData.m_135353_(MinecartCommandBlock.class, EntityDataSerializers.f_135030_);
@@ -72,6 +_,8 @@
    }
 
    public InteractionResult m_6096_(Player p_38522_, InteractionHand p_38523_) {
+      InteractionResult ret = super.m_6096_(p_38522_, p_38523_);
+      if (ret.m_19077_()) return ret;
       return this.f_38505_.m_45412_(p_38522_);
    }
 
@@ -112,6 +_,11 @@
 
       public CommandSourceStack m_6712_() {
          return new CommandSourceStack(this, MinecartCommandBlock.this.m_20182_(), MinecartCommandBlock.this.m_20155_(), this.m_5991_(), 2, this.m_45439_().getString(), MinecartCommandBlock.this.m_5446_(), this.m_5991_().m_142572_(), MinecartCommandBlock.this);
+      }
+
+      @Override
+      public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return (org.bukkit.craftbukkit.v1_18_R2.entity.CraftMinecartCommand) MinecartCommandBlock.this.getBukkitEntity();
       }
    }
 }
