--- a/net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -25,7 +_,7 @@
 
 public class MinecartTNT extends AbstractMinecart {
    private static final byte f_150345_ = 10;
-   private int f_38647_ = -1;
+   public int f_38647_ = -1; // CatServer - private -> public
 
    public MinecartTNT(EntityType<? extends MinecartTNT> p_38649_, Level p_38650_) {
       super(p_38649_, p_38650_);
@@ -90,7 +_,7 @@
       return Items.f_42693_;
    }
 
-   protected void m_38688_(double p_38689_) {
+   public void m_38688_(double p_38689_) { // CatServer - protected -> public
       this.m_257440_((DamageSource)null, p_38689_);
    }
 
@@ -101,7 +_,15 @@
             d0 = 5.0D;
          }
 
-         this.m_9236_().m_254877_(this, p_259539_, (ExplosionDamageCalculator)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), (float)(4.0D + this.f_19796_.m_188500_() * 1.5D * d0), false, Level.ExplosionInteraction.TNT);
+         // CraftBukkit start - 1.20.1
+         org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), (float) (4.0D + this.f_19796_.m_188500_() * 1.5D * d0), false);
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            f_38647_ = -1;
+            return;
+         }
+         this.m_9236_().m_254877_(this, p_259539_, (ExplosionDamageCalculator)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.TNT);
+         // CraftBukkit end
          this.m_146870_();
       }
 
