--- a/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
@@ -21,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 
 public class MinecartCommandBlock extends AbstractMinecart {
-   static final EntityDataAccessor<String> f_38503_ = SynchedEntityData.m_135353_(MinecartCommandBlock.class, EntityDataSerializers.f_135030_);
+   public static final EntityDataAccessor<String> f_38503_ = SynchedEntityData.m_135353_(MinecartCommandBlock.class, EntityDataSerializers.f_135030_); // CatServer - public
    static final EntityDataAccessor<Component> f_38504_ = SynchedEntityData.m_135353_(MinecartCommandBlock.class, EntityDataSerializers.f_135031_);
    private final BaseCommandBlock f_38505_ = new MinecartCommandBlock.MinecartCommandBase();
    private static final int f_150284_ = 4;
@@ -78,6 +_,8 @@
    }
 
    public InteractionResult m_6096_(Player p_38522_, InteractionHand p_38523_) {
+      InteractionResult ret = super.m_6096_(p_38522_, p_38523_);
+      if (ret.m_19077_()) return ret;
       return this.f_38505_.m_45412_(p_38522_);
    }
 
@@ -123,5 +_,12 @@
       public boolean m_288209_() {
          return !MinecartCommandBlock.this.m_213877_();
       }
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return (org.bukkit.craftbukkit.v1_20_R1.entity.CraftMinecartCommand) MinecartCommandBlock.this.getBukkitEntity();
+      }
+      // CraftBukkit end
    }
 }
