--- a/net/minecraft/entity/monster/AbstractSkeletonEntity.java
+++ b/net/minecraft/entity/monster/AbstractSkeletonEntity.java
@@ -33,10 +_,12 @@
 import net.minecraft.entity.projectile.AbstractArrowEntity;
 import net.minecraft.entity.projectile.ProjectileHelper;
 import net.minecraft.inventory.EquipmentSlotType;
+import net.minecraft.item.BowItem;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
 import net.minecraft.item.ShootableItem;
 import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.Hand;
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
@@ -45,6 +_,7 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public abstract class AbstractSkeletonEntity extends MonsterEntity implements IRangedAttackMob {
    private final RangedBowAttackGoal<AbstractSkeletonEntity> field_85037_d = new RangedBowAttackGoal<>(this, 1.0D, 20, 15.0F);
@@ -154,7 +_,7 @@
       if (this.field_70170_p != null && !this.field_70170_p.field_72995_K) {
          this.field_70714_bg.func_85156_a(this.field_85038_e);
          this.field_70714_bg.func_85156_a(this.field_85037_d);
-         ItemStack itemstack = this.func_184586_b(ProjectileHelper.func_221274_a(this, Items.field_151031_f));
+         ItemStack itemstack = this.func_184586_b(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.item.BowItem));
          if (itemstack.func_77973_b() == Items.field_151031_f) {
             int i = 20;
             if (this.field_70170_p.func_175659_aa() != Difficulty.HARD) {
@@ -171,15 +_,28 @@
    }
 
    public void func_82196_d(LivingEntity p_82196_1_, float p_82196_2_) {
-      ItemStack itemstack = this.func_213356_f(this.func_184586_b(ProjectileHelper.func_221274_a(this, Items.field_151031_f)));
+      ItemStack itemstack = this.func_213356_f(this.func_184586_b(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof BowItem)));
       AbstractArrowEntity abstractarrowentity = this.func_213624_b(itemstack, p_82196_2_);
+      if (this.func_184614_ca().func_77973_b() == Items.field_151031_f) {
+         abstractarrowentity = ((net.minecraft.item.BowItem) this.func_184614_ca().func_77973_b()).customArrow(abstractarrowentity);
+      }
       double d0 = p_82196_1_.func_226277_ct_() - this.func_226277_ct_();
       double d1 = p_82196_1_.func_226283_e_(0.3333333333333333D) - abstractarrowentity.func_226278_cu_();
       double d2 = p_82196_1_.func_226281_cx_() - this.func_226281_cx_();
-      double d3 = (double)MathHelper.func_76133_a(d0 * d0 + d2 * d2);
-      abstractarrowentity.func_70186_c(d0, d1 + d3 * (double)0.2F, d2, 1.6F, (float)(14 - this.field_70170_p.func_175659_aa().func_151525_a() * 4));
+      double d3 = (double) MathHelper.func_76133_a(d0 * d0 + d2 * d2);
+      abstractarrowentity.func_70186_c(d0, d1 + d3 * (double) 0.2F, d2, 1.6F, (float) (14 - this.field_70170_p.func_175659_aa().func_151525_a() * 4));
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityShootBowEvent event = CraftEventFactory.callEntityShootBowEvent(this, this.func_184614_ca(), null, abstractarrowentity, Hand.MAIN_HAND, 0.8F, true);
+      if (event.isCancelled()) {
+         event.getProjectile().remove();
+         return;
+      }
+      if (event.getProjectile() == abstractarrowentity.getBukkitEntity()) {
+         field_70170_p.func_217376_c(abstractarrowentity);
+      }
+      // CraftBukkit end
       this.func_184185_a(SoundEvents.field_187866_fi, 1.0F, 1.0F / (this.func_70681_au().nextFloat() * 0.4F + 0.8F));
-      this.field_70170_p.func_217376_c(abstractarrowentity);
+      // this.world.addEntity(abstractarrowentity); // CraftBukkit - moved up
    }
 
    protected AbstractArrowEntity func_213624_b(ItemStack p_213624_1_, float p_213624_2_) {
