--- a/net/minecraft/entity/monster/AbstractRaiderEntity.java
+++ b/net/minecraft/entity/monster/AbstractRaiderEntity.java
@@ -43,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public abstract class AbstractRaiderEntity extends PatrollerEntity {
    protected static final DataParameter<Boolean> field_213666_c = EntityDataManager.func_187226_a(AbstractRaiderEntity.class, DataSerializers.field_187198_h);
@@ -151,7 +_,7 @@
                i = MathHelper.func_76125_a(i, 0, 4);
                EffectInstance effectinstance = new EffectInstance(Effects.field_220309_E, 120000, i, false, false, true);
                if (!this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223621_x)) {
-                  playerentity.func_195064_c(effectinstance);
+                  playerentity.addEffect(effectinstance, Cause.PATROL_CAPTAIN); // CraftBukkit
                }
             }
          }
@@ -335,7 +_,7 @@
          this.field_220853_c.func_70661_as().func_75499_g();
 
          for(AbstractRaiderEntity abstractraiderentity : this.field_220853_c.field_70170_p.func_217374_a(AbstractRaiderEntity.class, this.field_220851_a, this.field_220853_c, this.field_220853_c.func_174813_aQ().func_72314_b(8.0D, 8.0D, 8.0D))) {
-            abstractraiderentity.func_70624_b(this.field_220853_c.func_70638_az());
+            abstractraiderentity.setGoalTarget(this.field_220853_c.func_70638_az(), TargetReason.FOLLOW_LEADER, true); // CraftBukkit
          }
 
       }
@@ -345,7 +_,7 @@
          LivingEntity livingentity = this.field_220853_c.func_70638_az();
          if (livingentity != null) {
             for(AbstractRaiderEntity abstractraiderentity : this.field_220853_c.field_70170_p.func_217374_a(AbstractRaiderEntity.class, this.field_220851_a, this.field_220853_c, this.field_220853_c.func_174813_aQ().func_72314_b(8.0D, 8.0D, 8.0D))) {
-               abstractraiderentity.func_70624_b(livingentity);
+               abstractraiderentity.setGoalTarget(livingentity, TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                abstractraiderentity.func_213395_q(true);
             }
 
