--- a/net/minecraft/entity/monster/DrownedEntity.java
+++ b/net/minecraft/entity/monster/DrownedEntity.java
@@ -56,8 +_,8 @@
 
 public class DrownedEntity extends ZombieEntity implements IRangedAttackMob {
    private boolean field_204718_bx;
-   protected final SwimmerPathNavigator field_204716_a;
-   protected final GroundPathNavigator field_204717_b;
+   public final SwimmerPathNavigator field_204716_a;
+   public final GroundPathNavigator field_204717_b;
 
    public DrownedEntity(EntityType<? extends DrownedEntity> p_i50212_1_, World p_i50212_2_) {
       super(p_i50212_1_, p_i50212_2_);
