--- a/net/minecraft/entity/monster/ElderGuardianEntity.java
+++ b/net/minecraft/entity/monster/ElderGuardianEntity.java
@@ -14,6 +_,7 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ElderGuardianEntity extends GuardianEntity {
    public static final float field_213629_b = EntityType.field_200800_n.func_220333_h() / EntityType.field_200761_A.func_220333_h();
@@ -66,7 +_,7 @@
          for(ServerPlayerEntity serverplayerentity : list) {
             if (!serverplayerentity.func_70644_a(effect) || serverplayerentity.func_70660_b(effect).func_76458_c() < 2 || serverplayerentity.func_70660_b(effect).func_76459_b() < 1200) {
                serverplayerentity.field_71135_a.func_147359_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241774_k_, this.func_174814_R() ? 0.0F : 1.0F));
-               serverplayerentity.func_195064_c(new EffectInstance(effect, 6000, 2));
+               serverplayerentity.addEffect(new EffectInstance(effect, 6000, 2), Cause.ATTACK); // CraftBukkit
             }
          }
       }
