--- a/net/minecraft/entity/monster/GhastEntity.java
+++ b/net/minecraft/entity/monster/GhastEntity.java
@@ -176,7 +_,8 @@
                }
 
                FireballEntity fireballentity = new FireballEntity(world, this.field_179470_b, d2, d3, d4);
-               fireballentity.field_92057_e = this.field_179470_b.func_175453_cd();
+               // CraftBukkit - set bukkitYield when setting explosionPower
+               fireballentity.bukkitYield = fireballentity.field_92057_e = this.field_179470_b.func_175453_cd();
                fireballentity.func_70107_b(this.field_179470_b.func_226277_ct_() + vector3d.field_72450_a * 4.0D, this.field_179470_b.func_226283_e_(0.5D) + 0.5D, fireballentity.func_226281_cx_() + vector3d.field_72449_c * 4.0D);
                world.func_217376_c(fireballentity);
                this.field_179471_a = -40;
