--- a/net/minecraft/entity/monster/HoglinEntity.java
+++ b/net/minecraft/entity/monster/HoglinEntity.java
@@ -48,8 +_,8 @@
 public class HoglinEntity extends AnimalEntity implements IMob, IFlinging {
    private static final DataParameter<Boolean> field_234356_bw_ = EntityDataManager.func_187226_a(HoglinEntity.class, DataSerializers.field_187198_h);
    private int field_234357_bx_;
-   private int field_234358_by_ = 0;
-   private boolean field_234359_bz_ = false;
+   public int field_234358_by_ = 0;
+   public boolean field_234359_bz_ = false;
    protected static final ImmutableList<? extends SensorType<? extends Sensor<? super HoglinEntity>>> field_234354_bu_ = ImmutableList.of(SensorType.field_220998_b, SensorType.field_220999_c, SensorType.field_234132_n_, SensorType.field_234131_m_);
    protected static final ImmutableList<? extends MemoryModuleType<?>> field_234355_bv_ = ImmutableList.of(MemoryModuleType.field_220953_n, MemoryModuleType.field_220945_f, MemoryModuleType.field_220946_g, MemoryModuleType.field_220949_j, MemoryModuleType.field_234102_l_, MemoryModuleType.field_220951_l, MemoryModuleType.field_220950_k, MemoryModuleType.field_223021_x, MemoryModuleType.field_220954_o, MemoryModuleType.field_234103_o_, MemoryModuleType.field_234104_p_, MemoryModuleType.field_234092_Z_, MemoryModuleType.field_234106_z_, MemoryModuleType.field_234094_ab_, MemoryModuleType.field_234095_ac_, MemoryModuleType.field_234091_Y_, MemoryModuleType.field_234075_I_, MemoryModuleType.field_234098_af_, MemoryModuleType.field_234099_ag_);
 
@@ -117,7 +_,7 @@
       HoglinTasks.func_234377_a_(this);
       if (this.func_234364_eK_()) {
          ++this.field_234358_by_;
-         if (this.field_234358_by_ > 300) {
+         if (this.field_234358_by_ > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.field_233590_aW_, (timer) -> this.field_234358_by_ = timer)) {
             this.func_241412_a_(SoundEvents.field_232715_fE_);
             this.func_234360_a_((ServerWorld)this.field_70170_p);
          }
@@ -212,6 +_,7 @@
       ZoglinEntity zoglinentity = this.func_233656_b_(EntityType.field_233590_aW_, true);
       if (zoglinentity != null) {
          zoglinentity.func_195064_c(new EffectInstance(Effects.field_76431_k, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zoglinentity);
       }
 
    }
@@ -253,7 +_,7 @@
       this.func_184212_Q().func_187227_b(field_234356_bw_, p_234370_1_);
    }
 
-   private boolean func_234368_eV_() {
+   public boolean func_234368_eV_() {
       return this.func_184212_Q().func_187225_a(field_234356_bw_);
    }
 
