--- a/net/minecraft/entity/monster/HuskEntity.java
+++ b/net/minecraft/entity/monster/HuskEntity.java
@@ -15,6 +_,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class HuskEntity extends ZombieEntity {
    public HuskEntity(EntityType<? extends HuskEntity> p_i50204_1_, World p_i50204_2_) {
@@ -49,7 +_,7 @@
       boolean flag = super.func_70652_k(p_70652_1_);
       if (flag && this.func_184614_ca().func_190926_b() && p_70652_1_ instanceof LivingEntity) {
          float f = this.field_70170_p.func_175649_E(this.func_233580_cy_()).func_180168_b();
-         ((LivingEntity)p_70652_1_).func_195064_c(new EffectInstance(Effects.field_76438_s, 140 * (int)f));
+         ((LivingEntity)p_70652_1_).addEffect(new EffectInstance(Effects.field_76438_s, 140 * (int)f), Cause.ATTACK); // CraftBukkit
       }
 
       return flag;
