--- a/net/minecraft/entity/monster/IllusionerEntity.java
+++ b/net/minecraft/entity/monster/IllusionerEntity.java
@@ -23,6 +_,7 @@
 import net.minecraft.entity.projectile.AbstractArrowEntity;
 import net.minecraft.entity.projectile.ProjectileHelper;
 import net.minecraft.inventory.EquipmentSlotType;
+import net.minecraft.item.BowItem;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
 import net.minecraft.nbt.CompoundNBT;
@@ -41,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class IllusionerEntity extends SpellcastingIllagerEntity implements IRangedAttackMob {
    private int field_193099_c;
@@ -180,8 +_,10 @@
    }
 
    public void func_82196_d(LivingEntity p_82196_1_, float p_82196_2_) {
-      ItemStack itemstack = this.func_213356_f(this.func_184586_b(ProjectileHelper.func_221274_a(this, Items.field_151031_f)));
+      ItemStack itemstack = this.func_213356_f(this.func_184586_b(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.item.BowItem)));
       AbstractArrowEntity abstractarrowentity = ProjectileHelper.func_221272_a(this, itemstack, p_82196_2_);
+      if (this.func_184614_ca().func_77973_b() instanceof net.minecraft.item.BowItem)
+         abstractarrowentity = ((net.minecraft.item.BowItem)this.func_184614_ca().func_77973_b()).customArrow(abstractarrowentity);
       double d0 = p_82196_1_.func_226277_ct_() - this.func_226277_ct_();
       double d1 = p_82196_1_.func_226283_e_(0.3333333333333333D) - abstractarrowentity.func_226278_cu_();
       double d2 = p_82196_1_.func_226281_cx_() - this.func_226281_cx_();
@@ -232,7 +_,7 @@
       }
 
       protected void func_190868_j() {
-         IllusionerEntity.this.func_70638_az().func_195064_c(new EffectInstance(Effects.field_76440_q, 400));
+         IllusionerEntity.this.func_70638_az().addEffect(new EffectInstance(Effects.field_76440_q, 400), Cause.ATTACK); // CraftBukkit
       }
 
       protected SoundEvent func_190871_k() {
@@ -265,7 +_,7 @@
       }
 
       protected void func_190868_j() {
-         IllusionerEntity.this.func_195064_c(new EffectInstance(Effects.field_76441_p, 1200));
+         IllusionerEntity.this.addEffect(new EffectInstance(Effects.field_76441_p, 1200), Cause.ILLUSION); // Craftbukkit
       }
 
       @Nullable
