--- a/net/minecraft/entity/monster/MagmaCubeEntity.java
+++ b/net/minecraft/entity/monster/MagmaCubeEntity.java
@@ -39,7 +_,7 @@
       return p_205019_1_.func_226668_i_(this) && !p_205019_1_.func_72953_d(this.func_174813_aQ());
    }
 
-   protected void func_70799_a(int p_70799_1_, boolean p_70799_2_) {
+   public void func_70799_a(int p_70799_1_, boolean p_70799_2_) {
       super.func_70799_a(p_70799_1_, p_70799_2_);
       this.func_110148_a(Attributes.field_233826_i_).func_111128_a((double)(p_70799_1_ * 3));
    }
@@ -52,7 +_,7 @@
       return ParticleTypes.field_197631_x;
    }
 
-   protected ResourceLocation func_184647_J() {
+   public ResourceLocation func_184647_J() {
       return this.func_189101_db() ? LootTables.field_186419_a : this.func_200600_R().func_220348_g();
    }
 
@@ -72,6 +_,7 @@
       Vector3d vector3d = this.func_213322_ci();
       this.func_213293_j(vector3d.field_72450_a, (double)(this.func_175134_bD() + (float)this.func_70809_q() * 0.1F), vector3d.field_72449_c);
       this.field_70160_al = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    protected void func_180466_bG(ITag<Fluid> p_180466_1_) {
