--- a/net/minecraft/entity/monster/PhantomEntity.java
+++ b/net/minecraft/entity/monster/PhantomEntity.java
@@ -41,6 +_,7 @@
 import net.minecraft.world.gen.Heightmap;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class PhantomEntity extends FlyingEntity implements IMob {
    private static final DataParameter<Integer> field_203035_a = EntityDataManager.func_187226_a(PhantomEntity.class, DataSerializers.field_187192_b);
@@ -218,7 +_,7 @@
 
                for(PlayerEntity playerentity : list) {
                   if (PhantomEntity.this.func_213344_a(playerentity, EntityPredicate.field_221016_a)) {
-                     PhantomEntity.this.func_70624_b(playerentity);
+                     PhantomEntity.this.setGoalTarget(playerentity, TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                      return true;
                   }
                }
