--- a/net/minecraft/entity/monster/PillagerEntity.java
+++ b/net/minecraft/entity/monster/PillagerEntity.java
@@ -56,7 +_,7 @@
 
 public class PillagerEntity extends AbstractIllagerEntity implements ICrossbowUser {
    private static final DataParameter<Boolean> field_213676_b = EntityDataManager.func_187226_a(PillagerEntity.class, DataSerializers.field_187198_h);
-   private final Inventory field_213677_bz = new Inventory(5);
+   public final Inventory field_213677_bz = new Inventory(5);
 
    public PillagerEntity(EntityType<? extends PillagerEntity> p_i50198_1_, World p_i50198_2_) {
       super(p_i50198_1_, p_i50198_2_);
@@ -120,7 +_,7 @@
    public AbstractIllagerEntity.ArmPose func_193077_p() {
       if (this.func_213675_dX()) {
          return AbstractIllagerEntity.ArmPose.CROSSBOW_CHARGE;
-      } else if (this.func_233631_a_(Items.field_222114_py)) {
+      } else if (this.func_233634_a_(item -> item instanceof net.minecraft.item.CrossbowItem)) {
          return AbstractIllagerEntity.ArmPose.CROSSBOW_HOLD;
       } else {
          return this.func_213398_dR() ? AbstractIllagerEntity.ArmPose.ATTACKING : AbstractIllagerEntity.ArmPose.NEUTRAL;
