--- a/net/minecraft/entity/monster/RavagerEntity.java
+++ b/net/minecraft/entity/monster/RavagerEntity.java
@@ -4,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.block.LeavesBlock;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
@@ -37,12 +_,12 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.vector.Vector3d;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class RavagerEntity extends AbstractRaiderEntity {
    private static final Predicate<Entity> field_213690_b = (p_213685_0_) -> {
@@ -134,14 +_,14 @@
             this.func_110148_a(Attributes.field_233821_d_).func_111128_a(MathHelper.func_219803_d(0.1D, d1, d0));
          }
 
-         if (this.field_70123_F && this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+         if (this.field_70123_F && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this)) {
             boolean flag = false;
             AxisAlignedBB axisalignedbb = this.func_174813_aQ().func_186662_g(0.2D);
 
             for(BlockPos blockpos : BlockPos.func_191531_b(MathHelper.func_76128_c(axisalignedbb.field_72340_a), MathHelper.func_76128_c(axisalignedbb.field_72338_b), MathHelper.func_76128_c(axisalignedbb.field_72339_c), MathHelper.func_76128_c(axisalignedbb.field_72336_d), MathHelper.func_76128_c(axisalignedbb.field_72337_e), MathHelper.func_76128_c(axisalignedbb.field_72334_f))) {
                BlockState blockstate = this.field_70170_p.func_180495_p(blockpos);
                Block block = blockstate.func_177230_c();
-               if (block instanceof LeavesBlock) {
+               if (block instanceof LeavesBlock && !CraftEventFactory.callEntityChangeBlockEvent(this,blockpos, Blocks.field_150350_a.func_176223_P()).isCancelled()) { // CraftBukkit
                   flag = this.field_70170_p.func_225521_a_(blockpos, true, this) || flag;
                }
             }
