--- a/net/minecraft/entity/monster/SilverfishEntity.java
+++ b/net/minecraft/entity/monster/SilverfishEntity.java
@@ -4,6 +_,7 @@
 import java.util.Random;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.block.SilverfishBlock;
 import net.minecraft.entity.CreatureAttribute;
 import net.minecraft.entity.EntitySize;
@@ -25,10 +_,10 @@
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class SilverfishEntity extends MonsterEntity {
    private SilverfishEntity.SummonSilverfishGoal field_175460_b;
@@ -134,7 +_,7 @@
             return false;
          } else {
             Random random = this.field_75457_a.func_70681_au();
-            if (this.field_75457_a.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b) && random.nextInt(10) == 0) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_75457_a.field_70170_p, this.field_75457_a) && random.nextInt(10) == 0) {
                this.field_179483_b = Direction.func_239631_a_(random);
                BlockPos blockpos = (new BlockPos(this.field_75457_a.func_226277_ct_(), this.field_75457_a.func_226278_cu_() + 0.5D, this.field_75457_a.func_226281_cx_())).func_177972_a(this.field_179483_b);
                BlockState blockstate = this.field_75457_a.field_70170_p.func_180495_p(blockpos);
@@ -161,6 +_,11 @@
             BlockPos blockpos = (new BlockPos(this.field_75457_a.func_226277_ct_(), this.field_75457_a.func_226278_cu_() + 0.5D, this.field_75457_a.func_226281_cx_())).func_177972_a(this.field_179483_b);
             BlockState blockstate = iworld.func_180495_p(blockpos);
             if (SilverfishBlock.func_196466_i(blockstate)) {
+               // CraftBukkit start
+               if (CraftEventFactory.callEntityChangeBlockEvent(this.field_75457_a, blockpos, SilverfishBlock.func_196467_h(blockstate.func_177230_c())).isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                iworld.func_180501_a(blockpos, SilverfishBlock.func_196467_h(blockstate.func_177230_c()), 3);
                this.field_75457_a.func_70656_aK();
                this.field_75457_a.func_70106_y();
@@ -203,7 +_,12 @@
                      BlockState blockstate = world.func_180495_p(blockpos1);
                      Block block = blockstate.func_177230_c();
                      if (block instanceof SilverfishBlock) {
-                        if (world.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+                        // CraftBukkit start
+                        if (CraftEventFactory.callEntityChangeBlockEvent(this.field_179464_a, blockpos1, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+                           continue;
+                        }
+                        // CraftBukkit end
+                        if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this.field_179464_a)) {
                            world.func_225521_a_(blockpos1, true, this.field_179464_a);
                         } else {
                            world.func_180501_a(blockpos1, ((SilverfishBlock)block).func_196468_d().func_176223_P(), 3);
