--- a/net/minecraft/entity/monster/SpellcastingIllagerEntity.java
+++ b/net/minecraft/entity/monster/SpellcastingIllagerEntity.java
@@ -62,7 +_,7 @@
       this.field_70180_af.func_187227_b(field_193088_c, (byte)p_193081_1_.field_193345_g);
    }
 
-   protected SpellcastingIllagerEntity.SpellType func_193083_dm() {
+   public SpellcastingIllagerEntity.SpellType func_193083_dm() {
       return !this.field_70170_p.field_72995_K ? this.field_193089_bx : SpellcastingIllagerEntity.SpellType.func_193337_a(this.field_70180_af.func_187225_a(field_193088_c));
    }
 
@@ -190,6 +_,11 @@
       public void func_75246_d() {
          --this.field_193321_c;
          if (this.field_193321_c == 0) {
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.handleEntitySpellCastEvent(SpellcastingIllagerEntity.this, func_193320_l())) { // PAIL rename getSpell
+               return;
+            }
+            // CraftBukkit end
             this.func_190868_j();
             SpellcastingIllagerEntity.this.func_184185_a(SpellcastingIllagerEntity.this.func_193086_dk(), 1.0F, 1.0F);
          }
