--- a/net/minecraft/entity/monster/SpiderEntity.java
+++ b/net/minecraft/entity/monster/SpiderEntity.java
@@ -41,6 +_,7 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class SpiderEntity extends MonsterEntity {
    private static final DataParameter<Byte> field_184729_a = EntityDataManager.func_187226_a(SpiderEntity.class, DataSerializers.field_187191_a);
@@ -118,7 +_,12 @@
    }
 
    public boolean func_70687_e(EffectInstance p_70687_1_) {
-      return p_70687_1_.func_188419_a() == Effects.field_76436_u ? false : super.func_70687_e(p_70687_1_);
+      if (p_70687_1_.func_188419_a() == Effects.field_76436_u) {
+         net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, p_70687_1_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
+      }
+      return super.func_70687_e(p_70687_1_);
    }
 
    public boolean func_70841_p() {
@@ -156,7 +_,7 @@
       if (p_213386_4_ instanceof SpiderEntity.GroupData) {
          Effect effect = ((SpiderEntity.GroupData)p_213386_4_).field_188478_a;
          if (effect != null) {
-            this.func_195064_c(new EffectInstance(effect, Integer.MAX_VALUE));
+            this.addEffect(new EffectInstance(effect, Integer.MAX_VALUE), Cause.SPIDER_SPAWN); // CraftBukkit
          }
       }
 
