--- a/net/minecraft/entity/monster/VexEntity.java
+++ b/net/minecraft/entity/monster/VexEntity.java
@@ -36,6 +_,7 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class VexEntity extends MonsterEntity {
    protected static final DataParameter<Byte> field_190664_a = EntityDataManager.func_187226_a(VexEntity.class, DataSerializers.field_187191_a);
@@ -246,7 +_,7 @@
       }
 
       public void func_75249_e() {
-         VexEntity.this.func_70624_b(VexEntity.this.field_190665_b.func_70638_az());
+         VexEntity.this.setGoalTarget(VexEntity.this.field_190665_b.func_70638_az(), TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
          super.func_75249_e();
       }
    }
