--- a/net/minecraft/entity/monster/WitchEntity.java
+++ b/net/minecraft/entity/monster/WitchEntity.java
@@ -43,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class WitchEntity extends AbstractRaiderEntity implements IRangedAttackMob {
    private static final UUID field_110184_bp = UUID.fromString("5CD17E52-A79A-43D3-A529-90FDE04B181E");
@@ -119,7 +_,7 @@
                   List<EffectInstance> list = PotionUtils.func_185189_a(itemstack);
                   if (list != null) {
                      for(EffectInstance effectinstance : list) {
-                        this.func_195064_c(new EffectInstance(effectinstance));
+                        this.addEffect(new EffectInstance(effectinstance), Cause.ATTACK); // CraftBukkit
                      }
                   }
                }
