--- a/net/minecraft/entity/monster/WitherSkeletonEntity.java
+++ b/net/minecraft/entity/monster/WitherSkeletonEntity.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class WitherSkeletonEntity extends AbstractSkeletonEntity {
    public WitherSkeletonEntity(EntityType<? extends WitherSkeletonEntity> p_i50187_1_, World p_i50187_2_) {
@@ -90,7 +_,7 @@
          return false;
       } else {
          if (p_70652_1_ instanceof LivingEntity) {
-            ((LivingEntity)p_70652_1_).func_195064_c(new EffectInstance(Effects.field_82731_v, 200));
+            ((LivingEntity)p_70652_1_).addEffect(new EffectInstance(Effects.field_82731_v, 200), Cause.ATTACK); // CraftBukkit
          }
 
          return true;
