--- a/net/minecraft/entity/monster/ZombieEntity.java
+++ b/net/minecraft/entity/monster/ZombieEntity.java
@@ -47,6 +_,7 @@
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.pathfinding.GroundPathNavigator;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EntityPredicates;
@@ -64,20 +_,24 @@
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.spawner.WorldEntitySpawner;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 
 public class ZombieEntity extends MonsterEntity {
    private static final UUID field_110187_bq = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
    private static final AttributeModifier field_110188_br = new AttributeModifier(field_110187_bq, "Baby speed boost", 0.5D, AttributeModifier.Operation.MULTIPLY_BASE);
    private static final DataParameter<Boolean> field_184737_bv = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187198_h);
    private static final DataParameter<Integer> field_184738_bw = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187192_b);
-   private static final DataParameter<Boolean> field_204709_bA = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187198_h);
+   public static final DataParameter<Boolean> field_204709_bA = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187198_h);
    private static final Predicate<Difficulty> field_213699_bC = (p_213697_0_) -> {
       return p_213697_0_ == Difficulty.HARD;
    };
    private final BreakDoorGoal field_146075_bs = new BreakDoorGoal(this, field_213699_bC);
    private boolean field_146076_bu;
    private int field_204707_bD;
-   private int field_204708_bE;
+   public int field_204708_bE;
+   private int lastTick = MinecraftServer.currentTick;
 
    public ZombieEntity(EntityType<? extends ZombieEntity> p_i48549_1_, World p_i48549_2_) {
       super(p_i48549_1_, p_i48549_2_);
@@ -100,7 +_,7 @@
       this.field_70714_bg.func_75776_a(7, new WaterAvoidingRandomWalkingGoal(this, 1.0D));
       this.field_70715_bh.func_75776_a(1, (new HurtByTargetGoal(this)).func_220794_a(ZombifiedPiglinEntity.class));
       this.field_70715_bh.func_75776_a(2, new NearestAttackableTargetGoal<>(this, PlayerEntity.class, true));
-      this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, AbstractVillagerEntity.class, false));
+      if ( field_70170_p.spigotConfig.zombieAggressiveTowardsVillager ) this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, AbstractVillagerEntity.class, false)); // Spigot
       this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, IronGolemEntity.class, true));
       this.field_70715_bh.func_75776_a(5, new NearestAttackableTargetGoal<>(this, TurtleEntity.class, 10, true, false, TurtleEntity.field_203029_bx));
    }
@@ -185,8 +_,12 @@
    public void func_70071_h_() {
       if (!this.field_70170_p.field_72995_K && this.func_70089_S() && !this.func_175446_cd()) {
          if (this.func_204706_dD()) {
-            --this.field_204708_bE;
-            if (this.field_204708_bE < 0) {
+            // CraftBukkit start - Use wall time instead of ticks for conversion
+            int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+            this.lastTick = MinecraftServer.currentTick;
+            this.field_204708_bE -= elapsedTicks;
+            // CraftBukkit end
+            if (this.field_204708_bE < 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.field_204724_o, (timer) -> this.field_204708_bE = timer)) {
                this.func_207302_dI();
             }
          } else if (this.func_204703_dA()) {
@@ -202,6 +_,7 @@
       }
 
       super.func_70071_h_();
+      this.lastTick = MinecraftServer.currentTick;
    }
 
    public void func_70636_d() {
@@ -230,7 +_,8 @@
       super.func_70636_d();
    }
 
-   private void func_204704_a(int p_204704_1_) {
+   public void func_204704_a(int p_204704_1_) {
+      this.lastTick = MinecraftServer.currentTick; // CraftBukkit
       this.field_204708_bE = p_204704_1_;
       this.func_184212_Q().func_187227_b(field_204709_bA, true);
    }
@@ -248,6 +_,7 @@
       if (zombieentity != null) {
          zombieentity.func_207304_a(zombieentity.field_70170_p.func_175649_E(zombieentity.func_233580_cy_()).func_180170_c());
          zombieentity.func_146070_a(zombieentity.func_204900_dz() && this.func_146072_bX());
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombieentity);
       }
 
    }
@@ -268,11 +_,15 @@
             livingentity = (LivingEntity)p_70097_1_.func_76346_g();
          }
 
-         if (livingentity != null && this.field_70170_p.func_175659_aa() == Difficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_233637_b_(Attributes.field_233829_l_) && this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223601_d)) {
             int i = MathHelper.func_76128_c(this.func_226277_ct_());
             int j = MathHelper.func_76128_c(this.func_226278_cu_());
             int k = MathHelper.func_76128_c(this.func_226281_cx_());
-            ZombieEntity zombieentity = new ZombieEntity(this.field_70170_p);
+
+         net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent event = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, field_70170_p, i, j, k, livingentity, this.func_110148_a(Attributes.field_233829_l_).func_111126_e());
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW  ||
+            livingentity != null && this.field_70170_p.func_175659_aa() == Difficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(Attributes.field_233829_l_).func_111126_e() && this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223601_d)) {
+            ZombieEntity zombieentity = event.getCustomSummonedAid() != null && event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW ? event.getCustomSummonedAid() : EntityType.field_200725_aD.func_200721_a(this.field_70170_p);
 
             for(int l = 0; l < 50; ++l) {
                int i1 = i + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
@@ -284,9 +_,10 @@
                if (WorldEntitySpawner.func_209382_a(entityspawnplacementregistry$placementtype, this.field_70170_p, blockpos, entitytype) && EntitySpawnPlacementRegistry.func_223515_a(entitytype, serverworld, SpawnReason.REINFORCEMENT, blockpos, this.field_70170_p.field_73012_v)) {
                   zombieentity.func_70107_b((double)i1, (double)j1, (double)k1);
                   if (!this.field_70170_p.func_217358_a((double)i1, (double)j1, (double)k1, 7.0D) && this.field_70170_p.func_226668_i_(zombieentity) && this.field_70170_p.func_226669_j_(zombieentity) && !this.field_70170_p.func_72953_d(zombieentity.func_174813_aQ())) {
+                     if (livingentity != null)
                      zombieentity.func_70624_b(livingentity);
                      zombieentity.func_213386_a(serverworld, this.field_70170_p.func_175649_E(zombieentity.func_233580_cy_()), SpawnReason.REINFORCEMENT, (ILivingEntityData)null, (CompoundNBT)null);
-                     serverworld.func_242417_l(zombieentity);
+                     serverworld.addAllEntities(zombieentity, CreatureSpawnEvent.SpawnReason.REINFORCEMENTS); // CraftBukkit
                      this.func_110148_a(Attributes.field_233829_l_).func_233769_c_(new AttributeModifier("Zombie reinforcement caller charge", (double)-0.05F, AttributeModifier.Operation.ADDITION));
                      zombieentity.func_110148_a(Attributes.field_233829_l_).func_233769_c_(new AttributeModifier("Zombie reinforcement callee charge", (double)-0.05F, AttributeModifier.Operation.ADDITION));
                      break;
@@ -304,7 +_,15 @@
       if (flag) {
          float f = this.field_70170_p.func_175649_E(this.func_233580_cy_()).func_180168_b();
          if (this.func_184614_ca().func_190926_b() && this.func_70027_ad() && this.field_70146_Z.nextFloat() < f * 0.3F) {
-            p_70652_1_.func_70015_d(2 * (int)f);
+            // CraftBukkit start
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent(this.getBukkitEntity(), p_70652_1_.getBukkitEntity(), 2 * (int) f); // PAIL: fixme
+            this.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               p_70652_1_.putCallCombustEvent(false);
+               p_70652_1_.func_70015_d(event.getDuration());
+               p_70652_1_.putCallCombustEvent(true);
+            }
+            // CraftBukkit end
          }
       }
 
@@ -369,18 +_,24 @@
 
    public void func_241847_a(ServerWorld p_241847_1_, LivingEntity p_241847_2_) {
       super.func_241847_a(p_241847_1_, p_241847_2_);
-      if ((p_241847_1_.func_175659_aa() == Difficulty.NORMAL || p_241847_1_.func_175659_aa() == Difficulty.HARD) && p_241847_2_ instanceof VillagerEntity) {
+      if ((p_241847_1_.func_175659_aa() == Difficulty.NORMAL || p_241847_1_.func_175659_aa() == Difficulty.HARD) && p_241847_2_ instanceof VillagerEntity && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(p_241847_2_, EntityType.field_200727_aF, (timer) -> {})) {
          if (p_241847_1_.func_175659_aa() != Difficulty.HARD && this.field_70146_Z.nextBoolean()) {
             return;
          }
 
          VillagerEntity villagerentity = (VillagerEntity)p_241847_2_;
-         ZombieVillagerEntity zombievillagerentity = villagerentity.func_233656_b_(EntityType.field_200727_aF, false);
+         // CraftBukkit start
+         ZombieVillagerEntity zombievillagerentity = villagerentity.convertTo(EntityType.field_200727_aF, false, EntityTransformEvent.TransformReason.INFECTION, CreatureSpawnEvent.SpawnReason.INFECTION);
+         if (zombievillagerentity == null) {
+            return;
+         }
+         // CraftBukkit end
          zombievillagerentity.func_213386_a(p_241847_1_, p_241847_1_.func_175649_E(zombievillagerentity.func_233580_cy_()), SpawnReason.CONVERSION, new ZombieEntity.GroupData(false, true), (CompoundNBT)null);
          zombievillagerentity.func_213792_a(villagerentity.func_213700_eh());
          zombievillagerentity.func_223727_a(villagerentity.func_223722_es().func_234058_a_(NBTDynamicOps.field_210820_a).getValue());
          zombievillagerentity.func_213790_g(villagerentity.func_213706_dY().func_222199_a());
          zombievillagerentity.func_213789_a(villagerentity.func_213708_dV());
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(p_241847_2_, zombievillagerentity);
          if (!this.func_174814_R()) {
             p_241847_1_.func_217378_a((PlayerEntity)null, 1026, this.func_233580_cy_(), 0);
          }
@@ -423,7 +_,7 @@
                   chickenentity1.func_213386_a(p_213386_1_, p_213386_2_, SpawnReason.JOCKEY, (ILivingEntityData)null, (CompoundNBT)null);
                   chickenentity1.func_152117_i(true);
                   this.func_184220_m(chickenentity1);
-                  p_213386_1_.func_217376_c(chickenentity1);
+                  p_213386_1_.addEntity(chickenentity1, CreatureSpawnEvent.SpawnReason.MOUNT);
                }
             }
          }
@@ -448,7 +_,7 @@
    }
 
    public static boolean func_241399_a_(Random p_241399_0_) {
-      return p_241399_0_.nextFloat() < 0.05F;
+      return p_241399_0_.nextFloat() < net.minecraftforge.common.ForgeConfig.SERVER.zombieBabyChance.get();
    }
 
    protected void func_207304_a(float p_207304_1_) {
@@ -468,7 +_,7 @@
    }
 
    protected void func_230291_eT_() {
-      this.func_110148_a(Attributes.field_233829_l_).func_111128_a(this.field_70146_Z.nextDouble() * (double)0.1F);
+      this.func_110148_a(Attributes.field_233829_l_).func_111128_a(this.field_70146_Z.nextDouble() * net.minecraftforge.common.ForgeConfig.SERVER.zombieBaseSummonChance.get());
    }
 
    public double func_70033_W() {
