--- a/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
+++ b/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
@@ -4,6 +_,7 @@
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.block.Blocks;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.IAngerable;
 import net.minecraft.entity.LivingEntity;
@@ -36,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class ZombifiedPiglinEntity extends ZombieEntity implements IAngerable {
    private static final UUID field_234344_b_ = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
@@ -133,7 +_,7 @@
       }).filter((p_241406_1_) -> {
          return !p_241406_1_.func_184191_r(this.func_70638_az());
       }).forEach((p_241405_1_) -> {
-         p_241405_1_.func_70624_b(this.func_70638_az());
+         p_241405_1_.setAttackTarget(this.func_70638_az(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -141,21 +_,30 @@
       this.func_184185_a(SoundEvents.field_232847_rN_, this.func_70599_aP() * 2.0F, this.func_70647_i() * 1.8F);
    }
 
-   public void func_70624_b(@Nullable LivingEntity p_70624_1_) {
-      if (this.func_70638_az() == null && p_70624_1_ != null) {
+   public boolean setAttackTarget(@Nullable LivingEntity entitylivingbaseIn, EntityTargetEvent.TargetReason reason, final boolean fireEvent) {
+      if (this.func_70638_az() == null && entitylivingbaseIn != null) {
          this.field_234345_bu_ = field_234350_d_.func_233018_a_(this.field_70146_Z);
          this.field_241401_bA_ = field_241403_bz_.func_233018_a_(this.field_70146_Z);
       }
 
-      if (p_70624_1_ instanceof PlayerEntity) {
-         this.func_230246_e_((PlayerEntity)p_70624_1_);
+      if (entitylivingbaseIn instanceof PlayerEntity) {
+         this.func_230246_e_((PlayerEntity)entitylivingbaseIn);
       }
 
-      super.func_70624_b(p_70624_1_);
+      return super.setGoalTarget(entitylivingbaseIn, reason, fireEvent);
    }
 
    public void func_230258_H__() {
-      this.func_230260_a__(field_234346_bv_.func_233018_a_(this.field_70146_Z));
+      // CraftBukkit start
+      Entity entity = ((ServerWorld) this.field_70170_p).func_217461_a(func_230257_G__());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), field_234346_bv_.func_233018_a_(this.field_70146_Z));
+      this.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.func_230259_a_(null);
+         return;
+      }
+      this.func_230260_a__(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean func_234351_b_(EntityType<ZombifiedPiglinEntity> p_234351_0_, IWorld p_234351_1_, SpawnReason p_234351_2_, BlockPos p_234351_3_, Random p_234351_4_) {
@@ -173,6 +_,7 @@
 
    public void func_70037_a(CompoundNBT p_70037_1_) {
       super.func_70037_a(p_70037_1_);
+      if(!field_70170_p.field_72995_K) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
       this.func_241358_a_((ServerWorld)this.field_70170_p, p_70037_1_);
    }
 
