--- a/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
@@ -48,7 +_,7 @@
    protected void m_6725_(ServerLevel p_22428_, Animal p_22429_, long p_22430_) {
       Animal animal = this.m_22409_(p_22429_);
       BehaviorUtils.m_22602_(p_22429_, animal, this.f_22388_);
-      if (p_22429_.m_19950_(animal, 3.0D)) {
+      if (p_22429_.closerToCenterThan(animal, 3.0D)) {
          if (p_22430_ >= this.f_22389_) {
             p_22429_.m_27563_(p_22428_, animal);
             p_22429_.m_6274_().m_21936_(MemoryModuleType.f_26375_);
