--- a/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
@@ -11,6 +_,9 @@
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.entity.npc.VillagerProfession;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class AssignProfessionFromJobSite extends Behavior<Villager> {
    public AssignProfessionFromJobSite() {
@@ -36,7 +_,14 @@
                return p_147412_.m_35622_() == p_22457_;
             }).findFirst();
          }).ifPresent((p_22464_) -> {
-            p_22454_.m_141967_(p_22454_.m_7141_().m_35565_(p_22464_));
+            // CraftBukkit start - Fire VillagerCareerChangeEvent where Villager gets employed
+            VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(p_22454_, CraftVillager.nmsToBukkitProfession(p_22464_), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+            if (event.isCancelled()) {
+               return;
+            }
+
+            p_22454_.m_141967_(p_22454_.m_7141_().m_35565_(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+            // CraftBukkit end
             p_22454_.m_35483_(p_22453_);
          });
       }
