--- a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
@@ -8,6 +_,13 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+// CraftBukkit end
+
 public class BabyFollowAdult {
    public static OneShot<AgeableMob> m_257685_(UniformInt p_260109_, float p_259621_) {
       return m_257631_(p_260109_, (p_147421_) -> {
@@ -22,8 +_,19 @@
                if (!p_258327_.m_6162_()) {
                   return false;
                } else {
-                  AgeableMob ageablemob = p_258331_.m_258051_(p_258317_);
+                  LivingEntity ageablemob = p_258331_.m_258051_(p_258317_); // CraftBukkit
                   if (p_258327_.m_19950_(ageablemob, (double)(p_259321_.m_142737_() + 1)) && !p_258327_.m_19950_(ageablemob, (double)p_259321_.m_142739_())) {
+                     // CraftBukkit start
+                     EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_258327_, ageablemob, EntityTargetEvent.TargetReason.FOLLOW_LEADER);
+                     if (event.isCancelled()) {
+                        return false;
+                     }
+                     if (event.getTarget() == null) {
+                        p_258317_.m_257971_();
+                        return true;
+                     }
+                     ageablemob = ((CraftLivingEntity) event.getTarget()).getHandle();
+                     // CraftBukkit end
                      WalkTarget walktarget = new WalkTarget(new EntityTracker(ageablemob, false), p_259190_.apply(p_258327_), p_259321_.m_142739_() - 1);
                      p_258318_.m_257512_(new EntityTracker(ageablemob, true));
                      p_258319_.m_257512_(walktarget);
