--- a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
@@ -30,7 +_,7 @@
          return false;
       } else {
          AgeableMob ageablemob = this.m_147429_(p_147424_);
-         return p_147424_.m_19950_(ageablemob, (double)(this.f_22468_.m_142737_() + 1)) && !p_147424_.m_19950_(ageablemob, (double)this.f_22468_.m_142739_());
+         return p_147424_.closerToCenterThan(ageablemob, (double)(this.f_22468_.m_142737_() + 1)) && !p_147424_.closerToCenterThan(ageablemob, (double)this.f_22468_.m_142739_());
       }
    }
 
