--- a/net/minecraft/world/entity/ai/behavior/BackUpIfTooClose.java
+++ b/net/minecraft/world/entity/ai/behavior/BackUpIfTooClose.java
@@ -33,7 +_,7 @@
    }
 
    private boolean m_22510_(E p_22511_) {
-      return this.m_22512_(p_22511_).m_19950_(p_22511_, (double)this.f_22489_);
+      return this.m_22512_(p_22511_).closerToCenterThan(p_22511_, (double)this.f_22489_);
    }
 
    private LivingEntity m_22512_(E p_22513_) {
