--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -81,6 +_,7 @@
    }
 
    public static void m_22613_(LivingEntity p_22614_, ItemStack p_22615_, Vec3 p_22616_) {
+      if (p_22615_.m_41619_()) return; // CraftBukkit - SPIGOT-4940: no empty loot
       double d0 = p_22614_.m_20188_() - (double)0.3F;
       ItemEntity itementity = new ItemEntity(p_22614_.f_19853_, p_22614_.m_20185_(), d0, p_22614_.m_20189_(), p_22615_);
       float f = 0.3F;
@@ -88,6 +_,13 @@
       vec3 = vec3.m_82541_().m_82490_((double)0.3F);
       itementity.m_20256_(vec3);
       itementity.m_32060_();
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(p_22614_.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+      itementity.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_22614_.f_19853_.m_7967_(itementity);
    }
 
@@ -104,7 +_,7 @@
          ProjectileWeaponItem projectileweaponitem = (ProjectileWeaponItem)item;
          if (p_22633_.m_5886_((ProjectileWeaponItem)item)) {
             int i = projectileweaponitem.m_6615_() - p_22635_;
-            return p_22633_.m_19950_(p_22634_, (double)i);
+            return p_22633_.closerToCenterThan(p_22634_, (double)i);
          }
       }
 
