--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -86,6 +_,7 @@
    }
 
    public static void m_217133_(LivingEntity p_217134_, ItemStack p_217135_, Vec3 p_217136_, Vec3 p_217137_, float p_217138_) {
+      if (p_217135_.m_41619_()) return; // CraftBukkit - SPIGOT-4940: no empty loot
       double d0 = p_217134_.m_20188_() - (double)p_217138_;
       ItemEntity itementity = new ItemEntity(p_217134_.m_9236_(), p_217134_.m_20185_(), d0, p_217134_.m_20189_(), p_217135_);
       itementity.m_32052_(p_217134_.m_20148_());
@@ -93,6 +_,13 @@
       vec3 = vec3.m_82541_().m_82542_(p_217137_.f_82479_, p_217137_.f_82480_, p_217137_.f_82481_);
       itementity.m_20256_(vec3);
       itementity.m_32060_();
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(p_217134_.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+      itementity.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_217134_.m_9236_().m_7967_(itementity);
    }
 
