--- a/net/minecraft/world/entity/ai/behavior/DismountOrSkipMounting.java
+++ b/net/minecraft/world/entity/ai/behavior/DismountOrSkipMounting.java
@@ -30,7 +_,7 @@
    }
 
    private boolean m_22836_(E p_22837_, Entity p_22838_) {
-      return p_22838_.m_6084_() && p_22838_.m_19950_(p_22837_, (double)this.f_22824_) && p_22838_.f_19853_ == p_22837_.f_19853_;
+      return p_22838_.m_6084_() && p_22838_.closerToCenterThan(p_22837_, (double)this.f_22824_) && p_22838_.f_19853_ == p_22837_.f_19853_;
    }
 
    protected void m_6735_(ServerLevel p_22833_, E p_22834_, long p_22835_) {
