--- a/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
@@ -23,6 +_,20 @@
             return (p_258380_, p_258381_, p_258382_) -> {
                ItemEntity itementity = p_258371_.m_258051_(p_258389_);
                if (p_258371_.m_257828_(p_258390_).isEmpty() && p_259490_.test(p_258381_) && itementity.m_19950_(p_258381_, (double)p_259054_) && p_258381_.m_9236_().m_6857_().m_61937_(itementity.m_20183_())) {
+                  // CraftBukkit start
+                  if (p_258381_ instanceof net.minecraft.world.entity.animal.allay.Allay) {
+                     org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTargetEvent(p_258381_, itementity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+
+                     if (event.isCancelled()) {
+                        return false;
+                     }
+                     if (!(event.getTarget() instanceof ItemEntity)) {
+                        p_258389_.m_257971_();
+                     }
+
+                     itementity = (ItemEntity) ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftEntity) event.getTarget()).getHandle();
+                  }
+                  // CraftBukkit end
                   WalkTarget walktarget = new WalkTarget(new EntityTracker(itementity, false), p_260346_, 0);
                   p_258387_.m_257512_(new EntityTracker(itementity, true));
                   p_258388_.m_257512_(walktarget);
