--- a/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
@@ -27,7 +_,7 @@
    }
 
    protected boolean m_6114_(ServerLevel p_23149_, E p_23150_) {
-      return this.f_23136_.test(p_23150_) && this.m_23155_(p_23150_).m_19950_(p_23150_, (double)this.f_23137_);
+      return this.f_23136_.test(p_23150_) && this.m_23155_(p_23150_).closerToCenterThan(p_23150_, (double)this.f_23137_);
    }
 
    protected void m_6735_(ServerLevel p_23152_, E p_23153_, long p_23154_) {
