--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -2,8 +_,10 @@
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
+
 import java.util.List;
 import javax.annotation.Nullable;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvents;
@@ -25,129 +_,148 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class HarvestFarmland extends Behavior<Villager> {
-   private static final int f_147559_ = 200;
-   public static final float f_147558_ = 0.5F;
-   @Nullable
-   private BlockPos f_23159_;
-   private long f_23160_;
-   private int f_23161_;
-   private final List<BlockPos> f_23162_ = Lists.newArrayList();
-
-   public HarvestFarmland() {
-      super(ImmutableMap.of(MemoryModuleType.f_26371_, MemoryStatus.VALUE_ABSENT, MemoryModuleType.f_26370_, MemoryStatus.VALUE_ABSENT, MemoryModuleType.f_26363_, MemoryStatus.VALUE_PRESENT));
-   }
-
-   protected boolean m_6114_(ServerLevel p_23174_, Villager p_23175_) {
-      if (!p_23174_.m_46469_().m_46207_(GameRules.f_46132_)) {
-         return false;
-      } else if (p_23175_.m_7141_().m_35571_() != VillagerProfession.f_35590_) {
-         return false;
-      } else {
-         BlockPos.MutableBlockPos blockpos$mutableblockpos = p_23175_.m_142538_().m_122032_();
-         this.f_23162_.clear();
-
-         for(int i = -1; i <= 1; ++i) {
-            for(int j = -1; j <= 1; ++j) {
-               for(int k = -1; k <= 1; ++k) {
-                  blockpos$mutableblockpos.m_122169_(p_23175_.m_20185_() + (double)i, p_23175_.m_20186_() + (double)j, p_23175_.m_20189_() + (double)k);
-                  if (this.m_23180_(blockpos$mutableblockpos, p_23174_)) {
-                     this.f_23162_.add(new BlockPos(blockpos$mutableblockpos));
-                  }
-               }
-            }
-         }
-
-         this.f_23159_ = this.m_23164_(p_23174_);
-         return this.f_23159_ != null;
-      }
-   }
-
-   @Nullable
-   private BlockPos m_23164_(ServerLevel p_23165_) {
-      return this.f_23162_.isEmpty() ? null : this.f_23162_.get(p_23165_.m_5822_().nextInt(this.f_23162_.size()));
-   }
-
-   private boolean m_23180_(BlockPos p_23181_, ServerLevel p_23182_) {
-      BlockState blockstate = p_23182_.m_8055_(p_23181_);
-      Block block = blockstate.m_60734_();
-      Block block1 = p_23182_.m_8055_(p_23181_.m_7495_()).m_60734_();
-      return block instanceof CropBlock && ((CropBlock)block).m_52307_(blockstate) || blockstate.m_60795_() && block1 instanceof FarmBlock;
-   }
-
-   protected void m_6735_(ServerLevel p_23177_, Villager p_23178_, long p_23179_) {
-      if (p_23179_ > this.f_23160_ && this.f_23159_ != null) {
-         p_23178_.m_6274_().m_21879_(MemoryModuleType.f_26371_, new BlockPosTracker(this.f_23159_));
-         p_23178_.m_6274_().m_21879_(MemoryModuleType.f_26370_, new WalkTarget(new BlockPosTracker(this.f_23159_), 0.5F, 1));
-      }
-
-   }
-
-   protected void m_6732_(ServerLevel p_23188_, Villager p_23189_, long p_23190_) {
-      p_23189_.m_6274_().m_21936_(MemoryModuleType.f_26371_);
-      p_23189_.m_6274_().m_21936_(MemoryModuleType.f_26370_);
-      this.f_23161_ = 0;
-      this.f_23160_ = p_23190_ + 40L;
-   }
-
-   protected void m_6725_(ServerLevel p_23196_, Villager p_23197_, long p_23198_) {
-      if (this.f_23159_ == null || this.f_23159_.m_203195_(p_23197_.m_20182_(), 1.0D)) {
-         if (this.f_23159_ != null && p_23198_ > this.f_23160_) {
-            BlockState blockstate = p_23196_.m_8055_(this.f_23159_);
-            Block block = blockstate.m_60734_();
-            Block block1 = p_23196_.m_8055_(this.f_23159_.m_7495_()).m_60734_();
-            if (block instanceof CropBlock && ((CropBlock)block).m_52307_(blockstate)) {
-               p_23196_.m_46953_(this.f_23159_, true, p_23197_);
-            }
-
-            if (blockstate.m_60795_() && block1 instanceof FarmBlock && p_23197_.m_35516_()) {
-               SimpleContainer simplecontainer = p_23197_.m_141944_();
-
-               for(int i = 0; i < simplecontainer.m_6643_(); ++i) {
-                  ItemStack itemstack = simplecontainer.m_8020_(i);
-                  boolean flag = false;
-                  if (!itemstack.m_41619_()) {
-                     if (itemstack.m_150930_(Items.f_42404_)) {
-                        p_23196_.m_7731_(this.f_23159_, Blocks.f_50092_.m_49966_(), 3);
-                        flag = true;
-                     } else if (itemstack.m_150930_(Items.f_42620_)) {
-                        p_23196_.m_7731_(this.f_23159_, Blocks.f_50250_.m_49966_(), 3);
-                        flag = true;
-                     } else if (itemstack.m_150930_(Items.f_42619_)) {
-                        p_23196_.m_7731_(this.f_23159_, Blocks.f_50249_.m_49966_(), 3);
-                        flag = true;
-                     } else if (itemstack.m_150930_(Items.f_42733_)) {
-                        p_23196_.m_7731_(this.f_23159_, Blocks.f_50444_.m_49966_(), 3);
-                        flag = true;
-                     }
-                  }
-
-                  if (flag) {
-                     p_23196_.m_6263_((Player)null, (double)this.f_23159_.m_123341_(), (double)this.f_23159_.m_123342_(), (double)this.f_23159_.m_123343_(), SoundEvents.f_11839_, SoundSource.BLOCKS, 1.0F, 1.0F);
-                     itemstack.m_41774_(1);
-                     if (itemstack.m_41619_()) {
-                        simplecontainer.m_6836_(i, ItemStack.f_41583_);
-                     }
-                     break;
-                  }
-               }
-            }
-
-            if (block instanceof CropBlock && !((CropBlock)block).m_52307_(blockstate)) {
-               this.f_23162_.remove(this.f_23159_);
-               this.f_23159_ = this.m_23164_(p_23196_);
-               if (this.f_23159_ != null) {
-                  this.f_23160_ = p_23198_ + 20L;
-                  p_23197_.m_6274_().m_21879_(MemoryModuleType.f_26370_, new WalkTarget(new BlockPosTracker(this.f_23159_), 0.5F, 1));
-                  p_23197_.m_6274_().m_21879_(MemoryModuleType.f_26371_, new BlockPosTracker(this.f_23159_));
-               }
-            }
-         }
-
-         ++this.f_23161_;
-      }
-   }
-
-   protected boolean m_6737_(ServerLevel p_23204_, Villager p_23205_, long p_23206_) {
-      return this.f_23161_ < 200;
-   }
+    private static final int f_147559_ = 200;
+    public static final float f_147558_ = 0.5F;
+    @Nullable
+    private BlockPos f_23159_;
+    private long f_23160_;
+    private int f_23161_;
+    private final List<BlockPos> f_23162_ = Lists.newArrayList();
+
+    public HarvestFarmland() {
+        super(ImmutableMap.of(MemoryModuleType.f_26371_, MemoryStatus.VALUE_ABSENT, MemoryModuleType.f_26370_, MemoryStatus.VALUE_ABSENT, MemoryModuleType.f_26363_, MemoryStatus.VALUE_PRESENT));
+    }
+
+    protected boolean m_6114_(ServerLevel p_23174_, Villager p_23175_) {
+        if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_23174_, p_23175_)) {
+            return false;
+        } else if (p_23175_.m_7141_().m_35571_() != VillagerProfession.f_35590_) {
+            return false;
+        } else {
+            BlockPos.MutableBlockPos blockpos$mutableblockpos = p_23175_.m_142538_().m_122032_();
+            this.f_23162_.clear();
+
+            for (int i = -1; i <= 1; ++i) {
+                for (int j = -1; j <= 1; ++j) {
+                    for (int k = -1; k <= 1; ++k) {
+                        blockpos$mutableblockpos.m_122169_(p_23175_.m_20185_() + (double) i, p_23175_.m_20186_() + (double) j, p_23175_.m_20189_() + (double) k);
+                        if (this.m_23180_(blockpos$mutableblockpos, p_23174_)) {
+                            this.f_23162_.add(new BlockPos(blockpos$mutableblockpos));
+                        }
+                    }
+                }
+            }
+
+            this.f_23159_ = this.m_23164_(p_23174_);
+            return this.f_23159_ != null;
+        }
+    }
+
+    @Nullable
+    private BlockPos m_23164_(ServerLevel p_23165_) {
+        return this.f_23162_.isEmpty() ? null : this.f_23162_.get(p_23165_.m_5822_().nextInt(this.f_23162_.size()));
+    }
+
+    private boolean m_23180_(BlockPos p_23181_, ServerLevel p_23182_) {
+        BlockState blockstate = p_23182_.m_8055_(p_23181_);
+        Block block = blockstate.m_60734_();
+        Block block1 = p_23182_.m_8055_(p_23181_.m_7495_()).m_60734_();
+        return block instanceof CropBlock && ((CropBlock) block).m_52307_(blockstate) || blockstate.m_60795_() && block1 instanceof FarmBlock;
+    }
+
+    protected void m_6735_(ServerLevel p_23177_, Villager p_23178_, long p_23179_) {
+        if (p_23179_ > this.f_23160_ && this.f_23159_ != null) {
+            p_23178_.m_6274_().m_21879_(MemoryModuleType.f_26371_, new BlockPosTracker(this.f_23159_));
+            p_23178_.m_6274_().m_21879_(MemoryModuleType.f_26370_, new WalkTarget(new BlockPosTracker(this.f_23159_), 0.5F, 1));
+        }
+
+    }
+
+    protected void m_6732_(ServerLevel p_23188_, Villager p_23189_, long p_23190_) {
+        p_23189_.m_6274_().m_21936_(MemoryModuleType.f_26371_);
+        p_23189_.m_6274_().m_21936_(MemoryModuleType.f_26370_);
+        this.f_23161_ = 0;
+        this.f_23160_ = p_23190_ + 40L;
+    }
+
+    protected void m_6725_(ServerLevel p_23196_, Villager p_23197_, long p_23198_) {
+        if (this.f_23159_ == null || this.f_23159_.m_203195_(p_23197_.m_20182_(), 1.0D)) {
+            if (this.f_23159_ != null && p_23198_ > this.f_23160_) {
+                BlockState blockstate = p_23196_.m_8055_(this.f_23159_);
+                Block block = blockstate.m_60734_();
+                Block block1 = p_23196_.m_8055_(this.f_23159_.m_7495_()).m_60734_();
+                if (block instanceof CropBlock && ((CropBlock) block).m_52307_(blockstate)) {
+                    // CraftBukkit start
+                    if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.f_23159_, Blocks.f_50016_.m_49966_()).isCancelled()) {
+                        p_23196_.m_46953_(this.f_23159_, true, p_23197_);
+                    }
+                    // CraftBukkit end
+                }
+
+                if (blockstate.m_60795_() && block1 instanceof FarmBlock && p_23197_.m_35516_()) {
+                    SimpleContainer simplecontainer = p_23197_.m_141944_();
+
+                    for (int i = 0; i < simplecontainer.m_6643_(); ++i) {
+                        ItemStack itemstack = simplecontainer.m_8020_(i);
+                        boolean flag = false;
+                        if (!itemstack.m_41619_()) {
+                            // CraftBukkit start
+                            Block planted = null;
+                            if (itemstack.m_150930_(Items.f_42404_)) {
+                                planted = Blocks.f_50092_;
+                                flag = true;
+                            } else if (itemstack.m_150930_(Items.f_42620_)) {
+                                planted = Blocks.f_50250_;
+                                flag = true;
+                            } else if (itemstack.m_150930_(Items.f_42619_)) {
+                                planted = Blocks.f_50249_;
+                                flag = true;
+                            } else if (itemstack.m_150930_(Items.f_42733_)) {
+                                planted = Blocks.f_50444_;
+                                flag = true;
+                            } else if (itemstack.m_41720_() instanceof net.minecraftforge.common.IPlantable) {
+                                if (((net.minecraftforge.common.IPlantable) itemstack.m_41720_()).getPlantType(p_23196_, f_23159_) == net.minecraftforge.common.PlantType.CROP) {
+                                    p_23196_.m_7731_(f_23159_, ((net.minecraftforge.common.IPlantable) itemstack.m_41720_()).getPlant(p_23196_, f_23159_), 3);
+                                    flag = true;
+                                }
+                            }
+
+                            if (planted != null && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.f_23159_, planted.m_49966_()).isCancelled()) {
+                                p_23196_.m_7731_(this.f_23159_, planted.m_49966_(), 3);
+                            } else {
+                                flag = false;
+                            }
+                            // CraftBukkit end
+                        }
+
+                        if (flag) {
+                            p_23196_.m_6263_((Player) null, (double) this.f_23159_.m_123341_(), (double) this.f_23159_.m_123342_(), (double) this.f_23159_.m_123343_(), SoundEvents.f_11839_, SoundSource.BLOCKS, 1.0F, 1.0F);
+                            itemstack.m_41774_(1);
+                            if (itemstack.m_41619_()) {
+                                simplecontainer.m_6836_(i, ItemStack.f_41583_);
+                            }
+                            break;
+                        }
+                    }
+                }
+
+                if (block instanceof CropBlock && !((CropBlock) block).m_52307_(blockstate)) {
+                    this.f_23162_.remove(this.f_23159_);
+                    this.f_23159_ = this.m_23164_(p_23196_);
+                    if (this.f_23159_ != null) {
+                        this.f_23160_ = p_23198_ + 20L;
+                        p_23197_.m_6274_().m_21879_(MemoryModuleType.f_26370_, new WalkTarget(new BlockPosTracker(this.f_23159_), 0.5F, 1));
+                        p_23197_.m_6274_().m_21879_(MemoryModuleType.f_26371_, new BlockPosTracker(this.f_23159_));
+                    }
+                }
+            }
+
+            ++this.f_23161_;
+        }
+    }
+
+
+    protected boolean m_6737_(ServerLevel p_23204_, Villager p_23205_, long p_23206_) {
+        return this.f_23161_ < 200;
+    }
 }
