--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -60,6 +_,13 @@
       })) {
          DoorBlock doorblock = (DoorBlock)blockstate.m_60734_();
          if (!doorblock.m_52815_(blockstate)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_23296_.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_23296_.f_19853_, blockpos));
+            p_23296_.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock.m_153165_(p_23296_, p_23295_, blockstate, blockpos, true);
          }
 
@@ -73,6 +_,13 @@
       })) {
          DoorBlock doorblock1 = (DoorBlock)blockstate1.m_60734_();
          if (!doorblock1.m_52815_(blockstate1)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_23296_.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_23296_.f_19853_, blockpos1));
+            p_23296_.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock1.m_153165_(p_23296_, p_23295_, blockstate1, blockpos1, true);
             this.m_23325_(p_23295_, p_23296_, blockpos1);
          }
