--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -26,6 +_,13 @@
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
+
 public class PrepareRamNearestTarget<E extends PathfinderMob> extends Behavior<E> {
    public static final int f_147713_ = 160;
    private final ToIntFunction<E> f_147714_;
@@ -56,6 +_,13 @@
             return this.f_147718_.m_26885_(p_147737_, p_147789_);
          });
       }).ifPresent((p_147778_) -> {
+         // CraftBukkit start
+         EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_147737_, p_147778_, (p_147778_ instanceof ServerPlayer) ? EntityTargetEvent.TargetReason.CLOSEST_PLAYER : EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+         if (event.isCancelled() || event.getTarget() == null) {
+            return;
+         }
+         p_147778_ = ((CraftLivingEntity) event.getTarget()).getHandle();
+         // CraftBukkit end
          this.m_147765_(p_147737_, p_147778_);
       });
    }
