--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -10,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.util.Mth;
@@ -25,6 +_,9 @@
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class PrepareRamNearestTarget<E extends PathfinderMob> extends Behavior<E> {
    public static final int f_147713_ = 160;
@@ -56,6 +_,13 @@
             return this.f_147718_.m_26885_(p_147737_, p_147789_);
          });
       }).ifPresent((p_147778_) -> {
+         // CraftBukkit start
+         EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_147737_, p_147778_, (p_147778_ instanceof ServerPlayer) ? EntityTargetEvent.TargetReason.CLOSEST_PLAYER : EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+         if (event.isCancelled() || event.getTarget() == null) {
+            return;
+         }
+         p_147778_ = ((CraftLivingEntity) event.getTarget()).getHandle();
+         // CraftBukkit end
          this.m_147765_(p_147737_, p_147778_);
       });
    }
