--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -7,6 +_,9 @@
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.entity.npc.VillagerData;
 import net.minecraft.world.entity.npc.VillagerProfession;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class ResetProfession extends Behavior<Villager> {
    public ResetProfession() {
@@ -19,7 +_,14 @@
    }
 
    protected void m_6735_(ServerLevel p_23776_, Villager p_23777_, long p_23778_) {
-      p_23777_.m_141967_(p_23777_.m_7141_().m_35565_(VillagerProfession.f_35585_));
+      // CraftBukkit start
+      VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(p_23777_, CraftVillager.nmsToBukkitProfession(VillagerProfession.f_35585_), VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+      if (event.isCancelled()) {
+         return;
+      }
+
+      p_23777_.m_141967_(p_23777_.m_7141_().m_35565_(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+      // CraftBukkit end
       p_23777_.m_35483_(p_23776_);
    }
 }
