--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -5,6 +_,9 @@
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.entity.npc.VillagerData;
 import net.minecraft.world.entity.npc.VillagerProfession;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftVillager;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class ResetProfession {
    public static BehaviorControl<Villager> m_257593_() {
@@ -13,7 +_,13 @@
             return (p_260244_, p_260084_, p_259597_) -> {
                VillagerData villagerdata = p_260084_.m_7141_();
                if (villagerdata.m_35571_() != VillagerProfession.f_35585_ && villagerdata.m_35571_() != VillagerProfession.f_35596_ && p_260084_.m_7809_() == 0 && villagerdata.m_35576_() <= 1) {
-                  p_260084_.m_34375_(p_260084_.m_7141_().m_35565_(VillagerProfession.f_35585_));
+                  // CraftBukkit start
+                  VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(p_260084_, CraftVillager.nmsToBukkitProfession(VillagerProfession.f_35585_), VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+                  if (event.isCancelled()) {
+                     return false;
+                  }
+                  p_260084_.m_34375_(p_260084_.m_7141_().m_35565_(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+                  // CraftBukkit end
                   p_260084_.m_35483_(p_260244_);
                   return true;
                } else {
