--- a/net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ b/net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -5,10 +_,14 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class StartAttacking<E extends Mob> extends Behavior<E> {
    private final Predicate<E> f_24190_;
@@ -42,6 +_,13 @@
    }
 
    private void m_24213_(E p_24214_, LivingEntity p_24215_) {
+      // CraftBukkit start
+      EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_24214_, p_24215_, (p_24215_ instanceof ServerPlayer) ? EntityTargetEvent.TargetReason.CLOSEST_PLAYER : EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+      if (event.isCancelled()) {
+         return;
+      }
+      p_24215_ = (event.getTarget() != null) ? ((CraftLivingEntity) event.getTarget()).getHandle() : null;
+      // CraftBukkit end
       p_24214_.m_6274_().m_21879_(MemoryModuleType.f_26372_, p_24215_);
       p_24214_.m_6274_().m_21936_(MemoryModuleType.f_26326_);
    }
