--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -106,11 +_,16 @@
       if (villager == null) {
          return Optional.empty();
       } else {
-         p_24657_.m_146762_(6000);
-         p_24658_.m_146762_(6000);
          villager.m_146762_(-24000);
          villager.m_7678_(p_24657_.m_20185_(), p_24657_.m_20186_(), p_24657_.m_20189_(), 0.0F, 0.0F);
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityBreedEvent(villager, p_24657_, p_24658_, null, null, 0).isCancelled()) {
+            return Optional.empty();
+         }
+         // CraftBukkit end
+         p_24657_.m_146762_(6000);
+         p_24658_.m_146762_(6000);
          p_24656_.m_47205_(villager);
+         //p_24656_.addFreshEntityWithPassengers(villager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason // FoxServer start
          p_24656_.m_7605_(villager, (byte)12);
          return Optional.of(villager);
       }
