--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -17,6 +_,11 @@
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.level.pathfinder.Path;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+// CraftBukkit end
+
 public class VillagerMakeLove extends Behavior<Villager> {
    private static final int f_148042_ = 5;
    private static final float f_148043_ = 0.5F;
@@ -110,11 +_,24 @@
       if (villager == null) {
          return Optional.empty();
       } else {
-         p_24657_.m_146762_(6000);
-         p_24658_.m_146762_(6000);
+         // CraftBukkit start - call EntityBreedEvent
          villager.m_146762_(-24000);
          villager.m_7678_(p_24657_.m_20185_(), p_24657_.m_20186_(), p_24657_.m_20189_(), 0.0F, 0.0F);
-         p_24656_.m_47205_(villager);
+         if (CraftEventFactory.callEntityBreedEvent(villager, p_24657_, p_24658_, null, null, 0).isCancelled()) {
+            return Optional.empty();
+         }
+         // Move age setting down
+         p_24657_.m_146762_(6000);
+         p_24658_.m_146762_(6000);
+         // CatServer start - compat:mod:Pehkui-3.8.2+1.20.1-forge:virtuoel.pehkui.mixin.compat116plus.VillagerBreedTaskMixin#pehkui$createChild
+         try {
+            catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevelAccessor$addFreshEntityWithPassengers$reason.push(p_24656_, CreatureSpawnEvent.SpawnReason.BREEDING);
+            p_24656_.m_47205_(villager);
+         } finally {
+            catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevelAccessor$addFreshEntityWithPassengers$reason.pop(p_24656_);
+         }
+         // CatServer end
+         // CraftBukkit end
          p_24656_.m_7605_(villager, (byte)12);
          return Optional.of(villager);
       }
