--- a/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
+++ b/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
@@ -7,6 +_,7 @@
 import net.minecraft.core.GlobalPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.SimpleContainer;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.item.Item;
@@ -26,16 +_,16 @@
          BlockState blockstate = p_24790_.m_8055_(globalpos.m_122646_());
          if (blockstate.m_60713_(Blocks.f_50715_)) {
             this.m_24802_(p_24791_);
-            this.m_24792_(p_24790_, p_24791_, globalpos, blockstate);
+            this.compostItems(p_24790_, p_24791_, globalpos, blockstate, p_24791_);
          }
 
       }
    }
 
-   private void m_24792_(ServerLevel p_24793_, Villager p_24794_, GlobalPos p_24795_, BlockState p_24796_) {
+   private void compostItems(ServerLevel p_24793_, Villager p_24794_, GlobalPos p_24795_, BlockState p_24796_, Entity p_25797_) {
       BlockPos blockpos = p_24795_.m_122646_();
       if (p_24796_.m_61143_(ComposterBlock.f_51913_) == 8) {
-         p_24796_ = ComposterBlock.m_51998_(p_24796_, p_24793_, blockpos);
+         p_24796_ = ComposterBlock.extractProduce(p_24796_, p_24793_, blockpos, p_25797_); // FoxServer - Fix for CraftBukkit nms patch
       }
 
       int i = 20;
@@ -57,7 +_,7 @@
                i -= l1;
 
                for(int i2 = 0; i2 < l1; ++i2) {
-                  blockstate = ComposterBlock.m_51929_(blockstate, p_24793_, itemstack, blockpos);
+                  blockstate = ComposterBlock.insertItem(blockstate, p_24793_, itemstack, blockpos, p_24794_); // FoxServer TODO
                   if (blockstate.m_61143_(ComposterBlock.f_51913_) == 7) {
                      this.m_24797_(p_24793_, p_24796_, blockpos, blockstate);
                      return;
