--- a/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
+++ b/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
@@ -35,7 +_,7 @@
    private void m_24792_(ServerLevel p_24793_, Villager p_24794_, GlobalPos p_24795_, BlockState p_24796_) {
       BlockPos blockpos = p_24795_.m_122646_();
       if (p_24796_.m_61143_(ComposterBlock.f_51913_) == 8) {
-         p_24796_ = ComposterBlock.m_51998_(p_24796_, p_24793_, blockpos);
+         p_24796_ = ComposterBlock.extractProduce(p_24796_, p_24793_, blockpos, p_24794_); // CraftBukkit
       }
 
       int i = 20;
@@ -57,7 +_,7 @@
                i -= l1;
 
                for(int i2 = 0; i2 < l1; ++i2) {
-                  blockstate = ComposterBlock.m_51929_(blockstate, p_24793_, itemstack, blockpos);
+                  blockstate = ComposterBlock.insertItem(blockstate, p_24793_, itemstack, blockpos, p_24794_); // CraftBukkit
                   if (blockstate.m_61143_(ComposterBlock.f_51913_) == 7) {
                      this.m_24797_(p_24793_, p_24796_, blockpos, blockstate);
                      return;
