--- a/net/minecraft/world/biome/provider/EndBiomeProvider.java
+++ b/net/minecraft/world/biome/provider/EndBiomeProvider.java
@@ -83,7 +_,9 @@
       int j = p_235317_2_ / 2;
       int k = p_235317_1_ % 2;
       int l = p_235317_2_ % 2;
-      float f = 100.0F - MathHelper.func_76129_c((float)(p_235317_1_ * p_235317_1_ + p_235317_2_ * p_235317_2_)) * 8.0F;
+      // Paper start - cast ints to long to avoid integer overflow
+      float f = 100.0F - MathHelper.func_76129_c((long) p_235317_1_ * (long) p_235317_1_ + (long) p_235317_2_ * (long) p_235317_2_) * 8.0F;
+      // Paper end
       f = MathHelper.func_76131_a(f, -100.0F, 80.0F);
 
       for(int i1 = -12; i1 <= 12; ++i1) {
