--- ../src-base/minecraft/net/minecraft/network/rcon/RConThreadClient.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConThreadClient.java
@@ -38,11 +38,12 @@
         this.logInfo("Rcon connection from: " + socket.getInetAddress());
     }
 
+    // CatSercer - fix decompile error
     public void run()
     {
-        while (true)
+        try
         {
-            try
+            while (true)
             {
                 if (!this.running)
                 {
@@ -85,11 +86,11 @@
                                     this.sendMultipacketResponse(l, "Error executing: " + s1 + " (" + exception.getMessage() + ")");
                                 }
 
-                                continue;
+                                break;
                             }
 
                             this.sendLoginFailedResponse();
-                            continue;
+                            break;
                         case 3:
                             String s = RConUtils.getBytesAsString(this.buffer, j, i);
                             int j1 = j + s.length();
@@ -98,38 +99,28 @@
                             {
                                 this.loggedIn = true;
                                 this.sendResponse(l, 2, "");
-                                continue;
+                                break;
                             }
 
                             this.loggedIn = false;
                             this.sendLoginFailedResponse();
-                            continue;
+                            break;
                         default:
                             this.sendMultipacketResponse(l, String.format("Unknown request %s", Integer.toHexString(i1)));
-                            continue;
                     }
                 }
             }
-            catch (SocketTimeoutException var17)
-            {
-                return;
-            }
-            catch (IOException var18)
-            {
-                return;
-            }
-            catch (Exception exception1)
-            {
-                LOGGER.error("Exception whilst parsing RCON input", (Throwable)exception1);
-                return;
-            }
-            finally
-            {
-                this.closeSocket();
-            }
-
-            return;
         }
+        catch (SocketTimeoutException var17) {}
+        catch (IOException var18) {}
+        catch (Exception exception1)
+        {
+            LOGGER.error("Exception whilst parsing RCON input", (Throwable)exception1);
+        }
+        finally
+        {
+            this.closeSocket();
+        }
     }
 
     private void sendResponse(int p_72654_1_, int p_72654_2_, String message) throws IOException
