--- ../src-base/minecraft/net/minecraft/pathfinding/PathFinder.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/PathFinder.java
@@ -33,7 +33,7 @@
     }
 
     @Nullable
-    private Path findPath(IBlockAccess worldIn, EntityLiving entitylivingIn, double x, double y, double z, float maxDistance)
+    public Path findPath(IBlockAccess worldIn, EntityLiving entitylivingIn, double x, double y, double z, float maxDistance)
     {
         this.path.clearPath();
         this.nodeProcessor.init(worldIn, entitylivingIn);
@@ -45,7 +45,7 @@
     }
 
     @Nullable
-    private Path findPath(PathPoint pathFrom, PathPoint pathTo, float maxDistance)
+    public Path findPath(PathPoint pathFrom, PathPoint pathTo, float maxDistance)
     {
         pathFrom.totalPathDistance = 0.0F;
         pathFrom.distanceToNext = pathFrom.distanceManhattan(pathTo);
@@ -119,7 +119,7 @@
         }
     }
 
-    private Path createPath(PathPoint start, PathPoint end)
+    public Path createPath(PathPoint start, PathPoint end)
     {
         int i = 1;
 
