--- a/net/minecraft/client/multiplayer/ClientSuggestionProvider.java
+++ b/net/minecraft/client/multiplayer/ClientSuggestionProvider.java
@@ -87,6 +_,12 @@
       return localplayer != null ? localplayer.m_20310_(p_105178_) : p_105178_ == 0;
    }
 
+   // CatServer start - 1.20.1
+   public boolean hasPermission(int p_105178_, String bukkitPermission) {
+      return this.m_6761_(p_105178_);
+   }
+   // CatServer end
+
    public CompletableFuture<Suggestions> m_212095_(ResourceKey<? extends Registry<?>> p_212429_, SharedSuggestionProvider.ElementSuggestionType p_212430_, SuggestionsBuilder p_212431_, CommandContext<?> p_212432_) {
       return this.m_5894_().m_6632_(p_212429_).map((p_212427_) -> {
          this.m_212335_(p_212427_, p_212430_, p_212431_);
