--- a/net/minecraft/util/datafix/fixes/BlockStateFlatteningMap.java
+++ b/net/minecraft/util/datafix/fixes/BlockStateFlatteningMap.java
@@ -20,7 +_,7 @@
       p_209419_0_.defaultReturnValue(-1);
    });
 
-   private static void func_199194_a(int p_199194_0_, String p_199194_1_, String... p_199194_2_) {
+   public static void func_199194_a(int p_199194_0_, String p_199194_1_, String... p_199194_2_) {
       Dynamic<?> dynamic = func_210048_b(p_199194_1_);
       field_199200_b[p_199194_0_] = dynamic;
       int i = p_199194_0_ >> 4;
