--- a/net/minecraft/util/datafix/fixes/BlockStateData.java
+++ b/net/minecraft/util/datafix/fixes/BlockStateData.java
@@ -21,7 +_,7 @@
    });
    static final String f_145177_ = "%%FILTER_ME%%";
 
-   private static void m_14942_(int p_14943_, String p_14944_, String... p_14945_) {
+   public static void m_14942_(int p_14943_, String p_14944_, String... p_14945_) {
       Dynamic<?> dynamic = m_14956_(p_14944_);
       f_14934_[p_14943_] = dynamic;
       int i = p_14943_ >> 4;
