--- a/net/minecraft/util/datafix/fixes/ItemFilledMapMetadata.java
+++ b/net/minecraft/util/datafix/fixes/ItemFilledMapMetadata.java
@@ -29,7 +_,7 @@
             Dynamic<?> dynamic = p_206360_2_.get(DSL.remainderFinder());
             Typed<?> typed = p_206360_2_.getOrCreateTyped(opticfinder1);
             Dynamic<?> dynamic1 = typed.get(DSL.remainderFinder());
-            dynamic1 = dynamic1.set("map", dynamic1.createInt(dynamic.get("Damage").asInt(0)));
+            if (!dynamic1.getElement("map").result().isPresent()) dynamic1 = dynamic1.set("map", dynamic1.createInt(dynamic.get("Damage").asInt(0)));
             return p_206360_2_.set(opticfinder1, typed.set(DSL.remainderFinder(), dynamic1));
          } else {
             return p_206360_2_;
