--- a/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
@@ -371,7 +_,7 @@
             if (f_16111_.contains(optional.get().getSecond())) {
                Typed<?> typed1 = p_16119_.getOrCreateTyped(opticfinder1);
                Dynamic<?> dynamic1 = typed1.get(DSL.remainderFinder());
-               dynamic1 = dynamic1.set("Damage", dynamic1.createInt(i));
+               if (i != 0) dynamic1 = dynamic1.set("Damage", dynamic1.createInt(i)); // CraftBukkit
                typed = typed.set(opticfinder1, typed1.set(DSL.remainderFinder(), dynamic1));
             }
 
