--- a/net/minecraft/util/datafix/fixes/ItemIdFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemIdFix.java
@@ -14,7 +_,7 @@
 import net.minecraft.util.datafix.schemas.NamespacedSchema;
 
 public class ItemIdFix extends DataFix {
-   private static final Int2ObjectMap<String> f_15937_ = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), (p_15953_) -> {
+   public static final Int2ObjectMap<String> f_15937_ = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), ( p_15953_) -> {
       p_15953_.put(1, "minecraft:stone");
       p_15953_.put(2, "minecraft:grass");
       p_15953_.put(3, "minecraft:dirt");
