--- a/net/minecraft/util/datafix/fixes/ItemStackMapIdFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemStackMapIdFix.java
@@ -28,7 +_,7 @@
             Dynamic<?> dynamic = p_16093_.get(DSL.remainderFinder());
             Typed<?> typed = p_16093_.getOrCreateTyped(opticfinder1);
             Dynamic<?> dynamic1 = typed.get(DSL.remainderFinder());
-            dynamic1 = dynamic1.set("map", dynamic1.createInt(dynamic.get("Damage").asInt(0)));
+            if (!dynamic1.getElement("map").result().isPresent()) dynamic1 = dynamic1.set("map", dynamic1.createInt(dynamic.get("Damage").asInt(0))); // CraftBukkit
             return p_16093_.set(opticfinder1, typed.set(DSL.remainderFinder(), dynamic1));
          } else {
             return p_16093_;
