--- a/net/minecraft/util/datafix/fixes/ItemIntIDToString.java
+++ b/net/minecraft/util/datafix/fixes/ItemIntIDToString.java
@@ -15,7 +_,7 @@
 import net.minecraft.util.datafix.TypeReferences;
 
 public class ItemIntIDToString extends DataFix {
-   private static final Int2ObjectMap<String> field_188221_a = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), (p_207470_0_) -> {
+   public static final Int2ObjectMap<String> field_188221_a = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), (p_207470_0_) -> {
       p_207470_0_.put(1, "minecraft:stone");
       p_207470_0_.put(2, "minecraft:grass");
       p_207470_0_.put(3, "minecraft:dirt");
