--- a/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
@@ -15,7 +_,7 @@
 import net.minecraft.util.datafix.schemas.NamespacedSchema;
 
 public class ItemSpawnEggFix extends DataFix {
-   private static final String[] f_16031_ = DataFixUtils.make(new String[256], (p_16054_) -> {
+   public static final String[] f_16031_ = DataFixUtils.make(new String[256], ( p_16054_) -> {
       p_16054_[1] = "Item";
       p_16054_[2] = "XPOrb";
       p_16054_[7] = "ThrownEgg";
