--- a/net/minecraft/util/datafix/fixes/SpawnEggNames.java
+++ b/net/minecraft/util/datafix/fixes/SpawnEggNames.java
@@ -16,7 +_,7 @@
 import net.minecraft.util.datafix.TypeReferences;
 
 public class SpawnEggNames extends DataFix {
-   private static final String[] field_188226_a = DataFixUtils.make(new String[256], (p_209278_0_) -> {
+   public static final String[] field_188226_a = DataFixUtils.make(new String[256], (p_209278_0_) -> {
       p_209278_0_[1] = "Item";
       p_209278_0_[2] = "XPOrb";
       p_209278_0_[7] = "ThrownEgg";
