--- ../src-base/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -68,7 +68,8 @@
 
     public String getWorldName()
     {
-        return this.theWorldInfo.getWorldName();
+        String name = super.getWorldName();
+        return name != null ? name : this.theWorldInfo.getWorldName();
     }
 
     public int getSaveVersion()
@@ -134,9 +135,6 @@
     {
     }
 
-    public void setWorldName(String worldName)
-    {
-    }
 
     public void setSaveVersion(int version)
     {
@@ -227,4 +225,11 @@
     {
         return this.theWorldInfo.getDimensionData(dimensionIn);
     }
+    public int getDimension() {
+        return this.theWorldInfo.getDimension();
+    }
+
+    public void setDimension(int dim) {
+        this.theWorldInfo.setDimension(dim);
+    }
 }
