--- ../src-base/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -218,13 +218,25 @@
     {
     }
 
+    @Deprecated
     public void setDimensionData(DimensionType dimensionIn, NBTTagCompound compound)
     {
         this.delegate.setDimensionData(dimensionIn, compound);
     }
 
+    @Deprecated
     public NBTTagCompound getDimensionData(DimensionType dimensionIn)
     {
         return this.delegate.getDimensionData(dimensionIn);
     }
+
+    public void setDimensionData(int dimensionID, NBTTagCompound compound)
+    {
+        this.delegate.setDimensionData(dimensionID, compound);
+    }
+
+    public NBTTagCompound getDimensionData(int dimensionID)
+    {
+        return this.delegate.getDimensionData(dimensionID);
+    }
 }
