--- ../src-base/minecraft/net/minecraft/world/storage/MapDecoration.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapDecoration.java
@@ -6,12 +6,12 @@
 
 public class MapDecoration
 {
-    private final MapDecoration.Type type;
+    private final Type type;
     private byte x;
     private byte y;
     private byte rotation;
 
-    public MapDecoration(MapDecoration.Type typeIn, byte xIn, byte yIn, byte rotationIn)
+    public MapDecoration(Type typeIn, byte xIn, byte yIn, byte rotationIn)
     {
         this.type = typeIn;
         this.x = xIn;
@@ -24,7 +24,7 @@
         return this.type.getIcon();
     }
 
-    public MapDecoration.Type getType()
+    public Type getType()
     {
         return this.type;
     }
@@ -92,6 +92,17 @@
         return i;
     }
 
+    /**
+     * Renders this decoration, useful for custom sprite sheets.
+     * @param index The index of this icon in the MapData's list. Used by vanilla to offset the Z-coordinate to prevent Z-fighting
+     * @return false to run vanilla logic for this decoration, true to skip it
+     */
+    @SideOnly(Side.CLIENT)
+    public boolean render(int index)
+    {
+        return false;
+    }
+
     public static enum Type
     {
         PLAYER(false),
@@ -142,7 +153,7 @@
             return this.mapColor;
         }
 
-        public static MapDecoration.Type byIcon(byte p_191159_0_)
+        public static Type byIcon(byte p_191159_0_)
         {
             return values()[MathHelper.clamp(p_191159_0_, 0, values().length - 1)];
         }
