--- ../src-base/minecraft/net/minecraft/world/storage/MapStorage.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapStorage.java
@@ -20,7 +20,7 @@
 {
     private final ISaveHandler saveHandler;
     protected Map<String, WorldSavedData> loadedDataMap = Maps.<String, WorldSavedData>newHashMap();
-    private final List<WorldSavedData> loadedDataList = Lists.<WorldSavedData>newArrayList();
+    public final List<WorldSavedData> loadedDataList = Lists.<WorldSavedData>newArrayList(); // Spigot
     private final Map<String, Short> idCounts = Maps.<String, Short>newHashMap();
 
     public MapStorage(ISaveHandler saveHandlerIn)
