--- ../src-base/minecraft/net/minecraft/world/storage/ThreadedFileIOBase.java
+++ ../src-work/minecraft/net/minecraft/world/storage/ThreadedFileIOBase.java
@@ -45,6 +45,7 @@
                 ++this.savedIOCounter;
             }
 
+            /*
             try
             {
                 Thread.sleep(this.isThreadWaiting ? 0L : 10L);
@@ -53,6 +54,7 @@
             {
                 interruptedexception1.printStackTrace();
             }
+            */
         }
 
         if (this.threadedIOQueue.isEmpty())
