--- ../src-base/minecraft/net/minecraft/world/storage/WorldSavedData.java
+++ ../src-work/minecraft/net/minecraft/world/storage/WorldSavedData.java
@@ -2,7 +2,7 @@
 
 import net.minecraft.nbt.NBTTagCompound;
 
-public abstract class WorldSavedData
+public abstract class WorldSavedData implements net.minecraftforge.common.util.INBTSerializable<NBTTagCompound>
 {
     public final String mapName;
     private boolean dirty;
@@ -30,4 +30,14 @@
     {
         return this.dirty;
     }
+
+    public void deserializeNBT(NBTTagCompound nbt)
+    {
+        this.readFromNBT(nbt);
+    }
+
+    public NBTTagCompound serializeNBT()
+    {
+        return this.writeToNBT(new NBTTagCompound());
+    }
 }
