--- ../src-base/minecraft/net/minecraft/world/storage/WorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/WorldInfo.java
@@ -162,6 +162,7 @@
         this.thunderTime = nbt.getInteger("thunderTime");
         this.thundering = nbt.getBoolean("thundering");
         this.hardcore = nbt.getBoolean("hardcore");
+        this.dimension = nbt.getInteger("dimension"); // Svarka
 
         if (nbt.hasKey("initialized", 99))
         {
@@ -376,6 +377,8 @@
 
         nbt.setBoolean("DifficultyLocked", this.difficultyLocked);
         nbt.setTag("GameRules", this.theGameRules.writeToNBT());
+        nbt.setInteger("dimension", this.dimension); // Svarka
+
         NBTTagCompound nbttagcompound1 = new NBTTagCompound();
 
         for (Entry<DimensionType, NBTTagCompound> entry : this.dimensionData.entrySet())
@@ -856,4 +859,25 @@
     {
         return this.versionName;
     }
+
+    public void checkName( String name ) {
+        if ( !this.levelName.equals( name ) ) {
+            this.levelName = name;
+        }
+    }
+
+    // Svarka start
+    /**
+     * Sets the Dimension.
+     */
+    public void setDimension(int dim)
+    {
+        this.dimension = dim;
+    }
+
+    public int getDimension()
+    {
+        return this.dimension;
+    }
+    // Svarka end
 }
