--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -52,6 +_,19 @@
    private boolean f_181119_;
    public final String f_19326_;
 
+   // CraftBukkit start
+   private boolean sweep;
+
+   public boolean isSweep() {
+      return sweep;
+   }
+
+   public DamageSource sweep() {
+      this.sweep = true;
+      return this;
+   }
+   // CraftBukkit end
+
    public static DamageSource m_19364_(LivingEntity p_19365_) {
       return new EntityDamageSource("sting", p_19365_);
    }
