--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -19,6 +_,38 @@
    private final Entity f_268595_;
    @Nullable
    private final Vec3 f_268454_;
+   // CraftBukkit start
+   private boolean sweep;
+   private boolean melting; // CraftBukkit - 1.20.1
+   private boolean poison; // CraftBukkit - 1.20.1
+
+   public boolean isSweep() {
+      return sweep;
+   }
+
+   public DamageSource sweep() {
+      this.sweep = true;
+      return this;
+   }
+
+   public boolean isMelting() { // CraftBukkit - 1.20.1
+      return melting;
+   }
+
+   public DamageSource melting() { // CraftBukkit - 1.20.1
+      this.melting = true;
+      return this;
+   }
+
+   public boolean isPoison() { // CraftBukkit - 1.20.1
+      return poison;
+   }
+
+   public DamageSource poison() { // CraftBukkit - 1.20.1
+      this.poison = true;
+      return this;
+   }
+   // CraftBukkit end
 
    public String toString() {
       return "DamageSource (" + this.m_269415_().f_268677_() + ")";
