--- a/net/minecraft/world/damagesource/DamageSources.java
+++ b/net/minecraft/world/damagesource/DamageSources.java
@@ -40,9 +_,15 @@
    private final DamageSource f_268610_;
    private final DamageSource f_287008_;
    private final DamageSource f_286964_;
+   // CraftBukkit start - 1.20.1
+   public final DamageSource melting;
+   public final DamageSource poison;
 
    public DamageSources(RegistryAccess p_270740_) {
       this.f_268645_ = p_270740_.m_175515_(Registries.f_268580_);
+      this.melting = this.m_269079_(DamageTypes.f_268468_).melting();
+      this.poison = this.m_269079_(DamageTypes.f_268515_).poison();
+      // CraftBukkit end
       this.f_268567_ = this.m_269079_(DamageTypes.f_268631_);
       this.f_268551_ = this.m_269079_(DamageTypes.f_268450_);
       this.f_268475_ = this.m_269079_(DamageTypes.f_268468_);
