--- a/net/minecraft/world/damagesource/IndirectEntityDamageSource.java
+++ b/net/minecraft/world/damagesource/IndirectEntityDamageSource.java
@@ -33,4 +_,10 @@
       String s1 = s + ".item";
       return !itemstack.m_41619_() && itemstack.m_41788_() ? new TranslatableComponent(s1, p_19410_.m_5446_(), component, itemstack.m_41611_()) : new TranslatableComponent(s, p_19410_.m_5446_(), component);
    }
+
+   // CraftBukkit start
+   public Entity getProximateDamageSource() {
+      return super.m_7639_();
+   }
+   // CraftBukkit end
 }
