--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIOwnerHurtTarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIOwnerHurtTarget.java
@@ -41,7 +41,7 @@
 
     public void startExecuting()
     {
-        this.taskOwner.setAttackTarget(this.theTarget);
+        this.taskOwner.setGoalTarget(this.theTarget, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true);
         EntityLivingBase entitylivingbase = this.theEntityTameable.getOwner();
 
         if (entitylivingbase != null)
