--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIHurtByTarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIHurtByTarget.java
@@ -28,7 +28,7 @@
 
     public void startExecuting()
     {
-        this.taskOwner.setAttackTarget(this.taskOwner.getRevengeTarget());
+        this.taskOwner.setGoalTarget(this.taskOwner.getRevengeTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true);
         this.target = this.taskOwner.getAttackTarget();
         this.revengeTimerOld = this.taskOwner.getRevengeTimer();
         this.unseenMemoryTicks = 300;
@@ -70,6 +70,6 @@
 
     protected void setEntityAttackTarget(EntityCreature creatureIn, EntityLivingBase entityLivingBaseIn)
     {
-        creatureIn.setAttackTarget(entityLivingBaseIn);
+        creatureIn.setGoalTarget(entityLivingBaseIn, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true);
     }
 }
