--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIBreakDoor.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIBreakDoor.java
@@ -84,6 +84,13 @@
 
         if (this.breakingTime == 240 && this.theEntity.worldObj.getDifficulty() == EnumDifficulty.HARD)
         {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.theEntity, this.doorPosition.getX(), this.doorPosition.getY(), this.doorPosition.getZ()).isCancelled())
+            {
+                this.updateTask();
+                return;
+            }
+            // CraftBukkit end
             this.theEntity.worldObj.setBlockToAir(this.doorPosition);
             this.theEntity.worldObj.playEvent(1021, this.doorPosition, 0);
             this.theEntity.worldObj.playEvent(2001, this.doorPosition, Block.getIdFromBlock(this.doorBlock));
