--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAITarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAITarget.java
@@ -91,7 +91,7 @@
                     }
                     else
                     {
-                        this.taskOwner.setAttackTarget(entitylivingbase);
+                        this.taskOwner.setGoalTarget(entitylivingbase, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true);
                         return true;
                     }
                 }
@@ -114,7 +114,7 @@
 
     public void resetTask()
     {
-        this.taskOwner.setAttackTarget((EntityLivingBase)null);
+        this.taskOwner.setGoalTarget((EntityLivingBase)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true);
         this.target = null;
     }
 
