--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIDefendVillage.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIDefendVillage.java
@@ -51,7 +51,7 @@
 
     public void startExecuting()
     {
-        this.irongolem.setAttackTarget(this.villageAgressorTarget);
+        this.irongolem.setGoalTarget(this.villageAgressorTarget, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true);
         super.startExecuting();
     }
 }
