--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIFindEntityNearest.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIFindEntityNearest.java
@@ -102,7 +102,7 @@
 
     public void startExecuting()
     {
-        this.mob.setAttackTarget(this.target);
+        this.mob.setGoalTarget(this.target, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true);
         super.startExecuting();
     }
 
