--- ../src-base/minecraft/net/minecraft/client/particle/ParticleFirework.java
+++ ../src-work/minecraft/net/minecraft/client/particle/ParticleFirework.java
@@ -23,7 +23,7 @@
         {
             public Particle createParticle(int particleID, World worldIn, double xCoordIn, double yCoordIn, double zCoordIn, double xSpeedIn, double ySpeedIn, double zSpeedIn, int... p_178902_15_)
             {
-                ParticleFirework.Spark particlefirework$spark = new ParticleFirework.Spark(worldIn, xCoordIn, yCoordIn, zCoordIn, xSpeedIn, ySpeedIn, zSpeedIn, Minecraft.getMinecraft().effectRenderer);
+                Spark particlefirework$spark = new Spark(worldIn, xCoordIn, yCoordIn, zCoordIn, xSpeedIn, ySpeedIn, zSpeedIn, Minecraft.getMinecraft().effectRenderer);
                 particlefirework$spark.setAlphaF(0.99F);
                 return particlefirework$spark;
             }
@@ -110,7 +110,7 @@
 
                 if (this.trail && this.particleAge < this.particleMaxAge / 2 && (this.particleAge + this.particleMaxAge) % 2 == 0)
                 {
-                    ParticleFirework.Spark particlefirework$spark = new ParticleFirework.Spark(this.world, this.posX, this.posY, this.posZ, 0.0D, 0.0D, 0.0D, this.effectRenderer);
+                    Spark particlefirework$spark = new Spark(this.world, this.posX, this.posY, this.posZ, 0.0D, 0.0D, 0.0D, this.effectRenderer);
                     particlefirework$spark.setAlphaF(0.99F);
                     particlefirework$spark.setRBGColorF(this.particleRed, this.particleGreen, this.particleBlue);
                     particlefirework$spark.particleAge = particlefirework$spark.particleMaxAge / 2;
@@ -256,7 +256,7 @@
                     float f = (float)((j & 16711680) >> 16) / 255.0F;
                     float f1 = (float)((j & 65280) >> 8) / 255.0F;
                     float f2 = (float)((j & 255) >> 0) / 255.0F;
-                    ParticleFirework.Overlay particlefirework$overlay = new ParticleFirework.Overlay(this.world, this.posX, this.posY, this.posZ);
+                    Overlay particlefirework$overlay = new Overlay(this.world, this.posX, this.posY, this.posZ);
                     particlefirework$overlay.setRBGColorF(f, f1, f2);
                     this.manager.addEffect(particlefirework$overlay);
                 }
@@ -284,7 +284,7 @@
 
             private void createParticle(double p_92034_1_, double p_92034_3_, double p_92034_5_, double p_92034_7_, double p_92034_9_, double p_92034_11_, int[] p_92034_13_, int[] p_92034_14_, boolean p_92034_15_, boolean p_92034_16_)
             {
-                ParticleFirework.Spark particlefirework$spark = new ParticleFirework.Spark(this.world, p_92034_1_, p_92034_3_, p_92034_5_, p_92034_7_, p_92034_9_, p_92034_11_, this.manager);
+                Spark particlefirework$spark = new Spark(this.world, p_92034_1_, p_92034_3_, p_92034_5_, p_92034_7_, p_92034_9_, p_92034_11_, this.manager);
                 particlefirework$spark.setAlphaF(0.99F);
                 particlefirework$spark.setTrail(p_92034_15_);
                 particlefirework$spark.setTwinkle(p_92034_16_);
