--- ../src-base/minecraft/net/minecraft/client/particle/Particle.java
+++ ../src-work/minecraft/net/minecraft/client/particle/Particle.java
@@ -265,8 +265,8 @@
         {
             this.width = p_187115_1_;
             this.height = p_187115_2_;
-            AxisAlignedBB axisalignedbb = this.getBoundingBox();
-            this.setBoundingBox(new AxisAlignedBB(axisalignedbb.minX, axisalignedbb.minY, axisalignedbb.minZ, axisalignedbb.minX + (double)this.width, axisalignedbb.minY + (double)this.height, axisalignedbb.minZ + (double)this.width));
+            // FORGE: Fix MC-12269 - Glitchy movement when setSize is called without setPosition
+            setPosition(posX, posY, posZ);
         }
     }
 
@@ -283,6 +283,8 @@
     public void move(double x, double y, double z)
     {
         double d0 = y;
+        double origX = x;
+        double origZ = z;
 
         if (this.canCollide)
         {
@@ -315,14 +317,14 @@
         }
 
         this.resetPositionToBB();
-        this.onGround = y != y && d0 < 0.0D;
+        this.onGround = d0 != y && d0 < 0.0D;
 
-        if (x != x)
+        if (origX != x)
         {
             this.motionX = 0.0D;
         }
 
-        if (z != z)
+        if (origZ != z)
         {
             this.motionZ = 0.0D;
         }
