--- a/net/minecraft/entity/ai/brain/task/AssignProfessionTask.java
+++ b/net/minecraft/entity/ai/brain/task/AssignProfessionTask.java
@@ -11,6 +_,9 @@
 import net.minecraft.util.math.GlobalPos;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftVillager;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class AssignProfessionTask extends Task<VillagerEntity> {
    public AssignProfessionTask() {
@@ -36,7 +_,13 @@
                return p_220389_1_.func_221149_b() == p_220390_0_;
             }).findFirst();
          }).ifPresent((p_220388_2_) -> {
-            p_212831_2_.func_213753_a(p_212831_2_.func_213700_eh().func_221126_a(p_220388_2_));
+            // CraftBukkit start - Fire VillagerCareerChangeEvent where Villager gets employed
+            VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(p_212831_2_, CraftVillager.nmsToBukkitProfession(p_220388_2_), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+            if (event.isCancelled()) {
+               return;
+            }
+            p_212831_2_.func_213753_a(p_212831_2_.func_213700_eh().func_221126_a(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+            // CraftBukkit end
             p_212831_2_.func_213770_a(p_212831_1_);
          });
       }
