--- a/net/minecraft/entity/ai/brain/task/ChangeJobTask.java
+++ b/net/minecraft/entity/ai/brain/task/ChangeJobTask.java
@@ -7,6 +_,9 @@
 import net.minecraft.entity.merchant.villager.VillagerEntity;
 import net.minecraft.entity.merchant.villager.VillagerProfession;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftVillager;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class ChangeJobTask extends Task<VillagerEntity> {
    public ChangeJobTask() {
@@ -19,7 +_,13 @@
    }
 
    protected void func_212831_a_(ServerWorld p_212831_1_, VillagerEntity p_212831_2_, long p_212831_3_) {
-      p_212831_2_.func_213753_a(p_212831_2_.func_213700_eh().func_221126_a(VillagerProfession.field_221151_a));
+      // CraftBukkit start
+      VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(p_212831_2_, CraftVillager.nmsToBukkitProfession(VillagerProfession.field_221151_a), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+      if (event.isCancelled()) {
+         return;
+      }
+      p_212831_2_.func_213753_a(p_212831_2_.func_213700_eh().func_221126_a(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+      // CraftBukkit end
       p_212831_2_.func_213770_a(p_212831_1_);
    }
 }
