--- a/net/minecraft/entity/ai/brain/task/CreateBabyVillagerTask.java
+++ b/net/minecraft/entity/ai/brain/task/CreateBabyVillagerTask.java
@@ -102,14 +_,28 @@
 
    private Optional<VillagerEntity> func_242307_b(ServerWorld p_242307_1_, VillagerEntity p_242307_2_, VillagerEntity p_242307_3_) {
       VillagerEntity villagerentity = p_242307_2_.func_241840_a(p_242307_1_, p_242307_3_);
+      // CraftBukkit start - call EntityBreedEvent
+      if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityBreedEvent(villagerentity, p_242307_2_, p_242307_3_, null, null, 0).isCancelled()) {
+         return Optional.empty();
+      }
+      // CraftBukkit end
       if (villagerentity == null) {
          return Optional.empty();
       } else {
-         p_242307_2_.func_70873_a(6000);
-         p_242307_3_.func_70873_a(6000);
+         // CraftBukkit start - call EntityBreedEvent
          villagerentity.func_70873_a(-24000);
          villagerentity.func_70012_b(p_242307_2_.func_226277_ct_(), p_242307_2_.func_226278_cu_(), p_242307_2_.func_226281_cx_(), 0.0F, 0.0F);
+         p_242307_1_.dummyFreshEntitySpawn = true;
          p_242307_1_.func_242417_l(villagerentity);
+         p_242307_1_.dummyFreshEntitySpawn = false;
+         // CraftBukkit start - call EntityBreedEvent
+         if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityBreedEvent(villagerentity, p_242307_2_, p_242307_3_, null, null, 0).isCancelled()) {
+            return Optional.empty();
+         }
+         // CraftBukkit end
+         p_242307_2_.func_70873_a(6000);
+         p_242307_3_.func_70873_a(6000);
+         p_242307_1_.addAllEntities(villagerentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
          p_242307_1_.func_72960_a(villagerentity, (byte)12);
          return Optional.of(villagerentity);
       }
