--- a/net/minecraft/entity/ai/brain/task/FarmTask.java
+++ b/net/minecraft/entity/ai/brain/task/FarmTask.java
@@ -22,7 +_,6 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.BlockPosWrapper;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.server.ServerWorld;
 
 public class FarmTask extends Task<VillagerEntity> {
@@ -37,7 +_,7 @@
    }
 
    protected boolean func_212832_a_(ServerWorld p_212832_1_, VillagerEntity p_212832_2_) {
-      if (!p_212832_1_.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_212832_1_, p_212832_2_)) {
          return false;
       } else if (p_212832_2_.func_213700_eh().func_221130_b() != VillagerProfession.field_221156_f) {
          return false;
@@ -95,7 +_,11 @@
             Block block = blockstate.func_177230_c();
             Block block1 = p_212833_1_.func_180495_p(this.field_220422_a.func_177977_b()).func_177230_c();
             if (block instanceof CropsBlock && ((CropsBlock)block).func_185525_y(blockstate)) {
-               p_212833_1_.func_225521_a_(this.field_220422_a, true, p_212833_2_);
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityChangeBlockEvent(p_212833_2_, this.field_220422_a, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+                  p_212833_1_.func_225521_a_(this.field_220422_a, true, p_212833_2_);
+               }
+               // CraftBukkit end
             }
 
             if (blockstate.func_196958_f() && block1 instanceof FarmlandBlock && p_212833_2_.func_175556_cs()) {
@@ -105,19 +_,33 @@
                   ItemStack itemstack = inventory.func_70301_a(i);
                   boolean flag = false;
                   if (!itemstack.func_190926_b()) {
+                     // CraftBukkit start
+                     Block planted = null;
                      if (itemstack.func_77973_b() == Items.field_151014_N) {
-                        p_212833_1_.func_180501_a(this.field_220422_a, Blocks.field_150464_aj.func_176223_P(), 3);
+                        planted = Blocks.field_150464_aj;
                         flag = true;
                      } else if (itemstack.func_77973_b() == Items.field_151174_bG) {
-                        p_212833_1_.func_180501_a(this.field_220422_a, Blocks.field_150469_bN.func_176223_P(), 3);
+                        planted = Blocks.field_150469_bN;
                         flag = true;
                      } else if (itemstack.func_77973_b() == Items.field_151172_bF) {
-                        p_212833_1_.func_180501_a(this.field_220422_a, Blocks.field_150459_bM.func_176223_P(), 3);
+                        planted = Blocks.field_150459_bM;
                         flag = true;
                      } else if (itemstack.func_77973_b() == Items.field_185163_cU) {
-                        p_212833_1_.func_180501_a(this.field_220422_a, Blocks.field_185773_cZ.func_176223_P(), 3);
+                        planted = Blocks.field_185773_cZ;
                         flag = true;
-                     }
+                     } else if (itemstack.func_77973_b() instanceof net.minecraftforge.common.IPlantable) {
+                        if (((net.minecraftforge.common.IPlantable)itemstack.func_77973_b()).getPlantType(p_212833_1_, field_220422_a) == net.minecraftforge.common.PlantType.CROP) {
+                           p_212833_1_.func_180501_a(field_220422_a, ((net.minecraftforge.common.IPlantable)itemstack.func_77973_b()).getPlant(p_212833_1_, field_220422_a), 3);
+                           flag = true;
+                        }
+                     }
+
+                     if (planted != null && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityChangeBlockEvent(p_212833_2_, this.field_220422_a, planted.func_176223_P()).isCancelled()) {
+                        p_212833_1_.func_180501_a(this.field_220422_a, planted.func_176223_P(), 3);
+                     } else {
+                        flag = false;
+                     }
+                     // CraftBukkit end
                   }
 
                   if (flag) {
