--- a/net/minecraft/entity/ai/brain/task/FindNewAttackTargetTask.java
+++ b/net/minecraft/entity/ai/brain/task/FindNewAttackTargetTask.java
@@ -9,6 +_,9 @@
 import net.minecraft.entity.ai.brain.memory.MemoryModuleType;
 import net.minecraft.util.EntityPredicates;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class FindNewAttackTargetTask<E extends MobEntity> extends Task<E> {
    private final Predicate<LivingEntity> field_233981_b_;
@@ -57,6 +_,17 @@
    }
 
    private void func_233987_d_(E p_233987_1_) {
+      // CraftBukkit start
+      LivingEntity old = p_233987_1_.func_213375_cj().func_218207_c(MemoryModuleType.field_234103_o_).orElse(null);
+      EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_233987_1_, null, (old != null && !old.func_70089_S()) ? EntityTargetEvent.TargetReason.TARGET_DIED : EntityTargetEvent.TargetReason.FORGOT_TARGET);
+      if (event.isCancelled()) {
+         return;
+      }
+      if (event.getTarget() != null) {
+         p_233987_1_.func_213375_cj().func_218205_a(MemoryModuleType.field_234103_o_, ((CraftLivingEntity) event.getTarget()).getHandle());
+         return;
+      }
+      // CraftBukkit end
       p_233987_1_.func_213375_cj().func_218189_b(MemoryModuleType.field_234103_o_);
    }
 }
