--- a/net/minecraft/entity/ai/brain/task/ForgetAttackTargetTask.java
+++ b/net/minecraft/entity/ai/brain/task/ForgetAttackTargetTask.java
@@ -8,7 +_,11 @@
 import net.minecraft.entity.MobEntity;
 import net.minecraft.entity.ai.brain.memory.MemoryModuleStatus;
 import net.minecraft.entity.ai.brain.memory.MemoryModuleType;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class ForgetAttackTargetTask<E extends MobEntity> extends Task<E> {
    private final Predicate<E> field_233973_b_;
@@ -42,6 +_,13 @@
    }
 
    private void func_233976_a_(E p_233976_1_, LivingEntity p_233976_2_) {
+      // CraftBukkit start
+      EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_233976_1_, p_233976_2_, (p_233976_2_ instanceof ServerPlayerEntity) ? EntityTargetEvent.TargetReason.CLOSEST_PLAYER : EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+      if (event.isCancelled()) {
+         return;
+      }
+      p_233976_2_ = (event.getTarget() != null) ? ((CraftLivingEntity) event.getTarget()).getHandle() : null;
+      // CraftBukkit end
       p_233976_1_.func_213375_cj().func_218205_a(MemoryModuleType.field_234103_o_, p_233976_2_);
       p_233976_1_.func_213375_cj().func_218189_b(MemoryModuleType.field_223021_x);
    }
