--- a/net/minecraft/entity/ai/brain/task/InteractWithDoorTask.java
+++ b/net/minecraft/entity/ai/brain/task/InteractWithDoorTask.java
@@ -55,6 +_,13 @@
       if (blockstate.func_235714_a_(BlockTags.field_200152_g)) {
          DoorBlock doorblock = (DoorBlock)blockstate.func_177230_c();
          if (!doorblock.func_242664_h(blockstate)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_212831_2_.getBukkitEntity(), org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock.at(p_212831_2_.field_70170_p, blockpos));
+            p_212831_2_.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock.func_242663_a(p_212831_1_, blockstate, blockpos, true);
          }
 
@@ -66,6 +_,13 @@
       if (blockstate1.func_235714_a_(BlockTags.field_200152_g)) {
          DoorBlock doorblock1 = (DoorBlock)blockstate1.func_177230_c();
          if (!doorblock1.func_242664_h(blockstate1)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_212831_2_.getBukkitEntity(), org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock.at(p_212831_2_.field_70170_p, blockpos));
+            p_212831_2_.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock1.func_242663_a(p_212831_1_, blockstate1, blockpos1, true);
             this.func_242301_c(p_212831_1_, p_212831_2_, blockpos1);
          }
