--- a/net/minecraft/entity/ai/brain/task/MultiTask.java
+++ b/net/minecraft/entity/ai/brain/task/MultiTask.java
@@ -16,7 +_,7 @@
    private final Set<MemoryModuleType<?>> field_220416_b;
    private final MultiTask.Ordering field_220417_c;
    private final MultiTask.RunType field_220418_d;
-   private final WeightedList<Task<? super E>> field_220419_e = new WeightedList<>();
+   private final WeightedList<Task<? super E>> field_220419_e = new WeightedList<>(false); // Paper - don't use a clone
 
    public MultiTask(Map<MemoryModuleType<?>, MemoryModuleStatus> p_i51503_1_, Set<MemoryModuleType<?>> p_i51503_2_, MultiTask.Ordering p_i51503_3_, MultiTask.RunType p_i51503_4_, List<Pair<Task<? super E>, Integer>> p_i51503_5_) {
       super(p_i51503_1_);
