--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -57,8 +_,8 @@
    private static final float f_149897_ = 0.5F;
    private static final int f_149898_ = 300;
    private int f_34483_;
-   private int f_34484_;
-   private boolean f_34485_;
+   public int f_34484_;
+   public boolean f_34485_;
    protected static final ImmutableList<? extends SensorType<? extends Sensor<? super Hoglin>>> f_34480_ = ImmutableList.of(SensorType.f_26811_, SensorType.f_26812_, SensorType.f_26822_, SensorType.f_26821_);
    protected static final ImmutableList<? extends MemoryModuleType<?>> f_34481_ = ImmutableList.of(MemoryModuleType.f_26375_, MemoryModuleType.f_148204_, MemoryModuleType.f_148205_, MemoryModuleType.f_26368_, MemoryModuleType.f_148206_, MemoryModuleType.f_26371_, MemoryModuleType.f_26370_, MemoryModuleType.f_26326_, MemoryModuleType.f_26377_, MemoryModuleType.f_26372_, MemoryModuleType.f_26373_, MemoryModuleType.f_26350_, MemoryModuleType.f_26383_, MemoryModuleType.f_26352_, MemoryModuleType.f_26353_, MemoryModuleType.f_26348_, MemoryModuleType.f_26331_, MemoryModuleType.f_26356_, MemoryModuleType.f_26357_);
 
@@ -126,7 +_,7 @@
       HoglinAi.m_34577_(this);
       if (this.m_34554_()) {
          ++this.f_34484_;
-         if (this.f_34484_ > 300) {
+         if (this.f_34484_ > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.f_20500_, (timer) -> this.f_34484_ = timer)) {
             this.m_34500_(SoundEvents.f_11959_);
             this.m_34531_((ServerLevel)this.f_19853_);
          }
@@ -219,6 +_,7 @@
       Zoglin zoglin = this.m_21406_(EntityType.f_20500_, true);
       if (zoglin != null) {
          zoglin.m_7292_(new MobEffectInstance(MobEffects.f_19604_, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zoglin);
       }
 
    }
@@ -260,7 +_,7 @@
       this.m_20088_().m_135381_(f_34482_, p_34565_);
    }
 
-   private boolean m_34557_() {
+   public boolean m_34557_() {
       return this.m_20088_().m_135370_(f_34482_);
    }
 
