--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -29,7 +_,7 @@
       Optional<ItemEntity> optional = list.stream().filter((p_26706_) -> {
          return p_26698_.m_7243_(p_26706_.m_32055_());
       }).filter((p_26701_) -> {
-         return p_26701_.m_19950_(p_26698_, 9.0D);
+         return p_26701_.closerToCenterThan(p_26698_, 9.0D);
       }).filter(p_26698_::m_142582_).findFirst();
       brain.m_21886_(MemoryModuleType.f_26332_, optional);
    }
