--- a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
@@ -20,7 +_,7 @@
 
    protected void m_5578_(ServerLevel p_26740_, LivingEntity p_26741_) {
       List<Player> list = p_26740_.m_6907_().stream().filter(EntitySelector.f_20408_).filter((p_26744_) -> {
-         return p_26741_.m_19950_(p_26744_, 16.0D);
+         return p_26741_.closerToCenterThan(p_26744_, 16.0D);
       }).sorted(Comparator.comparingDouble(p_26741_::m_20280_)).collect(Collectors.toList());
       Brain<?> brain = p_26741_.m_6274_();
       brain.m_21879_(MemoryModuleType.f_26367_, list);
