--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -15,6 +_,14 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Ingredient;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+// CraftBukkit end
+
 public class TemptingSensor extends Sensor<PathfinderMob> {
    public static final int f_148320_ = 10;
    private static final TargetingConditions f_148321_ = TargetingConditions.m_148353_().m_26883_(10.0D).m_148355_();
@@ -35,7 +_,18 @@
       }).sorted(Comparator.comparingDouble(p_148332_::m_20280_)).collect(Collectors.toList());
       if (!list.isEmpty()) {
          Player player = list.get(0);
-         brain.m_21879_(MemoryModuleType.f_148196_, player);
+
+         // CraftBukkit start
+         EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(p_148332_, player, EntityTargetEvent.TargetReason.TEMPT);
+         if (event.isCancelled()) {
+            return;
+         }
+         if (event.getTarget() instanceof HumanEntity) {
+            brain.m_21879_(MemoryModuleType.f_148196_, ((CraftHumanEntity) event.getTarget()).getHandle());
+         } else {
+            brain.m_21936_(MemoryModuleType.f_148196_);
+         }
+         // CraftBukkit end
       } else {
          brain.m_21936_(MemoryModuleType.f_148196_);
       }
