--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -29,7 +_,7 @@
       List<Player> list = p_148331_.m_6907_().stream().filter(EntitySelector.f_20408_).filter((p_148342_) -> {
          return f_148321_.m_26885_(p_148332_, p_148342_);
       }).filter((p_148335_) -> {
-         return p_148332_.m_19950_(p_148335_, 10.0D);
+         return p_148332_.closerToCenterThan(p_148335_, 10.0D);
       }).filter(this::m_148336_).sorted(Comparator.comparingDouble(p_148332_::m_20280_)).collect(Collectors.toList());
       if (!list.isEmpty()) {
          Player player = list.get(0);
