--- ../src-base/minecraft/net/minecraft/client/multiplayer/ServerData.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/ServerData.java
@@ -18,7 +18,7 @@
     public String gameVersion = "1.12.2";
     public boolean pinged;
     public String playerList;
-    private ServerData.ServerResourceMode resourceMode = ServerData.ServerResourceMode.PROMPT;
+    private ServerResourceMode resourceMode = ServerResourceMode.PROMPT;
     private String serverIcon;
     private boolean lanServer;
 
@@ -40,11 +40,11 @@
             nbttagcompound.setString("icon", this.serverIcon);
         }
 
-        if (this.resourceMode == ServerData.ServerResourceMode.ENABLED)
+        if (this.resourceMode == ServerResourceMode.ENABLED)
         {
             nbttagcompound.setBoolean("acceptTextures", true);
         }
-        else if (this.resourceMode == ServerData.ServerResourceMode.DISABLED)
+        else if (this.resourceMode == ServerResourceMode.DISABLED)
         {
             nbttagcompound.setBoolean("acceptTextures", false);
         }
@@ -52,12 +52,12 @@
         return nbttagcompound;
     }
 
-    public ServerData.ServerResourceMode getResourceMode()
+    public ServerResourceMode getResourceMode()
     {
         return this.resourceMode;
     }
 
-    public void setResourceMode(ServerData.ServerResourceMode mode)
+    public void setResourceMode(ServerResourceMode mode)
     {
         this.resourceMode = mode;
     }
@@ -75,16 +75,16 @@
         {
             if (nbtCompound.getBoolean("acceptTextures"))
             {
-                serverdata.setResourceMode(ServerData.ServerResourceMode.ENABLED);
+                serverdata.setResourceMode(ServerResourceMode.ENABLED);
             }
             else
             {
-                serverdata.setResourceMode(ServerData.ServerResourceMode.DISABLED);
+                serverdata.setResourceMode(ServerResourceMode.DISABLED);
             }
         }
         else
         {
-            serverdata.setResourceMode(ServerData.ServerResourceMode.PROMPT);
+            serverdata.setResourceMode(ServerResourceMode.PROMPT);
         }
 
         return serverdata;
