--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -43,13 +_,26 @@
       return p_19589_.m_21023_(MobEffects.f_19608_) || p_19589_.m_21023_(MobEffects.f_19592_);
    }
 
+   private static final ThreadLocal<org.bukkit.event.entity.EntityPotionEffectEvent.Cause> catserver$addEffectToPlayersAroundCause = ThreadLocal.withInitial(() -> org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+   public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel p_216947_, @Nullable Entity p_216948_, Vec3 p_216949_, double p_216950_, MobEffectInstance p_216951_, int p_216952_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause) {
+      // CraftBukkit start
+      try {
+         catserver$addEffectToPlayersAroundCause.set(cause);
+         return m_216946_(p_216947_, p_216948_, p_216949_, p_216950_, p_216951_, p_216952_);
+      } finally {
+         catserver$addEffectToPlayersAroundCause.remove();
+      }
+   }
+
    public static List<ServerPlayer> m_216946_(ServerLevel p_216947_, @Nullable Entity p_216948_, Vec3 p_216949_, double p_216950_, MobEffectInstance p_216951_, int p_216952_) {
       MobEffect mobeffect = p_216951_.m_19544_();
       List<ServerPlayer> list = p_216947_.m_8795_((p_267925_) -> {
          return p_267925_.f_8941_.m_9294_() && (p_216948_ == null || !p_216948_.m_7307_(p_267925_)) && p_216949_.m_82509_(p_267925_.m_20182_(), p_216950_) && (!p_267925_.m_21023_(mobeffect) || p_267925_.m_21124_(mobeffect).m_19564_() < p_216951_.m_19564_() || p_267925_.m_21124_(mobeffect).m_267633_(p_216952_ - 1));
       });
+      final org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause = catserver$addEffectToPlayersAroundCause.get();
+      catserver$addEffectToPlayersAroundCause.remove(); // CatServer
       list.forEach((p_238232_) -> {
-         p_238232_.m_147207_(new MobEffectInstance(p_216951_), p_216948_);
+         p_238232_.addEffect(new MobEffectInstance(p_216951_), p_216948_, cause); // CraftBukkit
       });
       return list;
    }
