--- ../src-base/minecraft/net/minecraftforge/fml/common/FMLCommonHandler.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/FMLCommonHandler.java
@@ -151,7 +151,6 @@
     {
         sidedDelegate = handler;
         MinecraftForge.initialize();
-//        MinecraftForge.registerCrashCallable();
         return ImmutableList.<String>of();
     }
 
@@ -400,7 +399,7 @@
         {
             return;
         }
-        if (handlerSet.contains(handler))
+        if (handlerSet.contains(handler) || worldInfo.getDimension() != 0)
         {
             return;
         }
