--- ../src-base/minecraft/net/minecraftforge/fml/common/FMLModContainer.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/FMLModContainer.java
@@ -138,10 +138,10 @@
         if (modid.length() > 64) {
             FMLLog.bigWarning("The modid %s is longer than the recommended maximum of 64 characters. Truncation will be enforced in 1.11", modid);
         }
-        if (!modid.equals(modid.toLowerCase(Locale.ENGLISH)))
-        {
-            FMLLog.bigWarning("The modid %s is not the same as it's lowercase version. Lowercasing will be enforced in 1.11", modid);
-        }
+        //if (!modid.equals(modid.toLowerCase(Locale.ENGLISH)))
+        //{
+        //    FMLLog.bigWarning("The modid %s is not the same as it's lowercase version. Lowercasing will be enforced in 1.11", modid);
+        //}
     }
     private ILanguageAdapter getLanguageAdapter()
     {
