--- ../src-base/minecraft/net/minecraftforge/fml/common/TracingPrintStream.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/TracingPrintStream.java
@@ -56,7 +56,7 @@
         } else if (elem.getClassName().startsWith("java.lang.Throwable")) {
             elem = elems[BASE_DEPTH + 4];
         }
-        return "[" + elem.getClassName() + ":" + elem.getMethodName() + ":" + elem.getLineNumber() + "]: ";
+        return ""; // CatServer
     }
 
 }
