--- ../src-base/minecraft/net/minecraft/block/state/BlockStateContainer.java
+++ ../src-work/minecraft/net/minecraft/block/state/BlockStateContainer.java
@@ -502,7 +502,7 @@
      *
      *   protected BlockStateContainer createBlockState()
      *   {
-     *       return (new BlockStateContainer.Builder(this)).add(FACING).add(SOME_UNLISTED).build();
+     *       return (new BlockStateContainer.Builder(this)).add(FACING).add(SOME_UNLISTED).build1();
      *   }
      *
      */
