--- ../src-base/minecraft/net/minecraft/block/state/IBlockProperties.java
+++ ../src-work/minecraft/net/minecraft/block/state/IBlockProperties.java
@@ -28,9 +28,13 @@
 
     boolean canEntitySpawn(Entity entityIn);
 
+    @Deprecated //Forge location aware version below
     int getLightOpacity();
+    int getLightOpacity(IBlockAccess world, BlockPos pos);
 
+    @Deprecated //Forge location aware version below
     int getLightValue();
+    int getLightValue(IBlockAccess world, BlockPos pos);
 
     @SideOnly(Side.CLIENT)
     boolean isTranslucent();
@@ -95,8 +99,14 @@
 
     RayTraceResult collisionRayTrace(World worldIn, BlockPos pos, Vec3d start, Vec3d end);
 
+    @Deprecated // Forge: Use isSideSolid(IBlockAccess, BlockPos, EnumFacing.UP) instead
     boolean isTopSolid();
 
+    //Forge added functions
+    boolean doesSideBlockRendering(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean isSideSolid(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean doesSideBlockChestOpening(IBlockAccess world, BlockPos pos, EnumFacing side);
+
     Vec3d getOffset(IBlockAccess access, BlockPos pos);
 
     boolean causesSuffocation();
