--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -4,13 +_,16 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
+import net.minecraft.data.worldgen.features.TreeFeatures;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.SaplingBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+import org.bukkit.TreeType;
 
 public abstract class AbstractTreeGrower {
    @Nullable
@@ -18,10 +_,12 @@
 
    public boolean m_6334_(ServerLevel p_60006_, ChunkGenerator p_60007_, BlockPos p_60008_, BlockState p_60009_, Random p_60010_) {
       Holder<? extends ConfiguredFeature<?, ?>> holder = this.m_203525_(p_60010_, this.m_60011_(p_60006_, p_60008_));
-      if (holder == null) {
+      net.minecraftforge.event.world.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_60006_, p_60010_, p_60008_, holder);
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || holder == null) {
          return false;
       } else {
-         ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
+         setTreeType(holder); // CraftBukkit
+         ConfiguredFeature<?, ?> configuredfeature = event.getFeature().m_203334_();
          p_60006_.m_7731_(p_60008_, Blocks.f_50016_.m_49966_(), 4);
          if (configuredfeature.m_65385_(p_60006_, p_60007_, p_60010_, p_60008_)) {
             return true;
@@ -41,4 +_,48 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<? extends ConfiguredFeature<?, ?>> worldgentreeabstract) {
+      if (worldgentreeabstract == TreeFeatures.f_195123_ || worldgentreeabstract == TreeFeatures.f_195142_) {
+         SaplingBlock.treeType = TreeType.TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195122_) {
+         SaplingBlock.treeType = TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == TreeFeatures.f_195121_) {
+         SaplingBlock.treeType = TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == TreeFeatures.f_195129_) {
+         SaplingBlock.treeType = TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195131_) {
+         SaplingBlock.treeType = TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195128_) {
+         SaplingBlock.treeType = TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195127_) {
+         SaplingBlock.treeType = TreeType.REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195126_) {
+         SaplingBlock.treeType = TreeType.ACACIA;
+      } else if (worldgentreeabstract == TreeFeatures.f_195125_ || worldgentreeabstract == TreeFeatures.f_195108_) {
+         SaplingBlock.treeType = TreeType.BIRCH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195135_) {
+         SaplingBlock.treeType = TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195137_) {
+         SaplingBlock.treeType = TreeType.SWAMP;
+      } else if (worldgentreeabstract == TreeFeatures.f_195130_ || worldgentreeabstract == TreeFeatures.f_195111_) {
+         SaplingBlock.treeType = TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195138_) {
+         SaplingBlock.treeType = TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195124_) {
+         SaplingBlock.treeType = TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == TreeFeatures.f_195133_) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195134_) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195132_) {
+         SaplingBlock.treeType = TreeType.JUNGLE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195139_) {
+         SaplingBlock.treeType = TreeType.AZALEA;
+      } else {
+         SaplingBlock.treeType = TreeType.MOD_CUSTOM; // CatServer
+      }
+   }
+   // CraftBukkit end
 }
