--- ../src-base/minecraft/net/minecraft/entity/item/EntityItem.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItem.java
@@ -35,7 +35,7 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final DataParameter<Optional<ItemStack>> ITEM = EntityDataManager.<Optional<ItemStack>>createKey(EntityItem.class, DataSerializers.OPTIONAL_ITEM_STACK);
     private int age;
-    private int delayBeforeCanPickup;
+    public int delayBeforeCanPickup;
     private int health;
     private String thrower;
     private String owner;
@@ -226,6 +226,7 @@
                     }
                     else
                     {
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemMergeEvent(this, other).isCancelled()) return false; // CraftBukkit
                         itemstack1.stackSize += itemstack.stackSize;
                         other.delayBeforeCanPickup = Math.max(other.delayBeforeCanPickup, this.delayBeforeCanPickup);
                         other.age = Math.min(other.age, this.age);
@@ -291,6 +292,11 @@
         }
         else
         {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
+                return false;
+            }
+            // CraftBukkit end
             this.setBeenAttacked();
             this.health = (int)((float)this.health - amount);
 
