--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartCommandBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartCommandBlock.java
@@ -28,7 +28,7 @@
 
 public class EntityMinecartCommandBlock extends EntityMinecart
 {
-    private static final DataParameter<String> COMMAND = EntityDataManager.<String>createKey(EntityMinecartCommandBlock.class, DataSerializers.STRING);
+    public static final DataParameter<String> COMMAND = EntityDataManager.<String>createKey(EntityMinecartCommandBlock.class, DataSerializers.STRING);
     private static final DataParameter<ITextComponent> LAST_OUTPUT = EntityDataManager.<ITextComponent>createKey(EntityMinecartCommandBlock.class, DataSerializers.TEXT_COMPONENT);
     private final CommandBlockBaseLogic commandBlockLogic = new CommandBlockBaseLogic()
     {
