--- ../src-base/minecraft/net/minecraft/entity/item/EntityEnderEye.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityEnderEye.java
@@ -14,11 +14,11 @@
 
 public class EntityEnderEye extends Entity
 {
-    private double targetX;
-    private double targetY;
-    private double targetZ;
-    private int despawnTimer;
-    private boolean shatterOrDrop;
+    public double targetX;
+    public double targetY;
+    public double targetZ;
+    public int despawnTimer;
+    public boolean shatterOrDrop;
 
     public EntityEnderEye(World worldIn)
     {
