--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartCommandBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartCommandBlock.java
@@ -29,10 +29,13 @@
 
 public class EntityMinecartCommandBlock extends EntityMinecart
 {
-    private static final DataParameter<String> COMMAND = EntityDataManager.<String>createKey(EntityMinecartCommandBlock.class, DataSerializers.STRING);
+    public static final DataParameter<String> COMMAND = EntityDataManager.<String>createKey(EntityMinecartCommandBlock.class, DataSerializers.STRING);
     private static final DataParameter<ITextComponent> LAST_OUTPUT = EntityDataManager.<ITextComponent>createKey(EntityMinecartCommandBlock.class, DataSerializers.TEXT_COMPONENT);
     private final CommandBlockBaseLogic commandBlockLogic = new CommandBlockBaseLogic()
     {
+        {
+            this.sender = EntityMinecartCommandBlock.this.getBukkitEntity(); // CraftBukkit - Set the sender
+        }
         public void updateCommand()
         {
             EntityMinecartCommandBlock.this.getDataManager().set(EntityMinecartCommandBlock.COMMAND, this.getCommand());
@@ -121,9 +124,9 @@
         this.commandBlockLogic.writeToNBT(compound);
     }
 
-    public EntityMinecart.Type getType()
+    public Type getType()
     {
-        return EntityMinecart.Type.COMMAND_BLOCK;
+        return Type.COMMAND_BLOCK;
     }
 
     public IBlockState getDefaultDisplayTile()
@@ -147,6 +150,7 @@
 
     public boolean processInitialInteract(EntityPlayer player, EnumHand hand)
     {
+        if (super.processInitialInteract(player, hand)) return true;
         this.commandBlockLogic.tryOpenEditCommandBlock(player);
         return false;
     }
