--- ../src-base/minecraft/net/minecraft/entity/item/EntityItem.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItem.java
@@ -35,7 +35,7 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final DataParameter<Optional<ItemStack>> ITEM = EntityDataManager.<Optional<ItemStack>>createKey(EntityItem.class, DataSerializers.OPTIONAL_ITEM_STACK);
     private int age;
-    private int delayBeforeCanPickup;
+    public int delayBeforeCanPickup;
     private int health;
     private String thrower;
     private String owner;
