--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartFurnace.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartFurnace.java
@@ -41,9 +41,9 @@
         EntityMinecart.registerFixesMinecart(fixer, EntityMinecartFurnace.class);
     }
 
-    public EntityMinecart.Type getType()
+    public Type getType()
     {
-        return EntityMinecart.Type.FURNACE;
+        return Type.FURNACE;
     }
 
     protected void entityInit()
@@ -145,6 +145,8 @@
     {
         ItemStack itemstack = player.getHeldItem(hand);
 
+        if (super.processInitialInteract(player, hand)) return true;
+
         if (itemstack.getItem() == Items.COAL && this.fuel + 3600 <= 32000)
         {
             if (!player.capabilities.isCreativeMode)
