--- ../src-base/minecraft/net/minecraft/entity/item/EntityPainting.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityPainting.java
@@ -19,20 +19,21 @@
 
 public class EntityPainting extends EntityHanging
 {
-    public EntityPainting.EnumArt art;
+    public EnumArt art;
 
     public EntityPainting(World worldIn)
     {
         super(worldIn);
+        this.art = EnumArt.values()[this.rand.nextInt(EnumArt.values().length)]; // CraftBukkit - generate a non-null painting
     }
 
     public EntityPainting(World worldIn, BlockPos pos, EnumFacing facing)
     {
         super(worldIn, pos);
-        List<EntityPainting.EnumArt> list = Lists.<EntityPainting.EnumArt>newArrayList();
+        List<EnumArt> list = Lists.<EnumArt>newArrayList();
         int i = 0;
 
-        for (EntityPainting.EnumArt entitypainting$enumart : EntityPainting.EnumArt.values())
+        for (EnumArt entitypainting$enumart : EnumArt.values())
         {
             this.art = entitypainting$enumart;
             this.updateFacingWithBoundingBox(facing);
@@ -51,11 +52,11 @@
 
         if (!list.isEmpty())
         {
-            Iterator<EntityPainting.EnumArt> iterator = list.iterator();
+            Iterator<EnumArt> iterator = list.iterator();
 
             while (iterator.hasNext())
             {
-                EntityPainting.EnumArt entitypainting$enumart1 = iterator.next();
+                EnumArt entitypainting$enumart1 = iterator.next();
 
                 if (entitypainting$enumart1.sizeX * entitypainting$enumart1.sizeY < i)
                 {
@@ -74,7 +75,7 @@
     {
         this(worldIn, pos, facing);
 
-        for (EntityPainting.EnumArt entitypainting$enumart : EntityPainting.EnumArt.values())
+        for (EnumArt entitypainting$enumart : EnumArt.values())
         {
             if (entitypainting$enumart.title.equals(title))
             {
@@ -96,7 +97,7 @@
     {
         String s = compound.getString("Motive");
 
-        for (EntityPainting.EnumArt entitypainting$enumart : EntityPainting.EnumArt.values())
+        for (EnumArt entitypainting$enumart : EnumArt.values())
         {
             if (entitypainting$enumart.title.equals(s))
             {
@@ -106,7 +107,7 @@
 
         if (this.art == null)
         {
-            this.art = EntityPainting.EnumArt.KEBAB;
+            this.art = EnumArt.KEBAB;
         }
 
         super.readEntityFromNBT(compound);
