--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartContainer.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartContainer.java
@@ -1,7 +1,15 @@
 package net.minecraft.entity.item;
 
+import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
+
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+
+import luohuayu.CatServer.inventory.ICBLockableContainer;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.inventory.Container;
@@ -24,12 +32,47 @@
 import net.minecraft.world.storage.loot.LootContext;
 import net.minecraft.world.storage.loot.LootTable;
 
-public abstract class EntityMinecartContainer extends EntityMinecart implements ILockableContainer, ILootContainer
+public abstract class EntityMinecartContainer extends EntityMinecart implements ICBLockableContainer, ILootContainer
 {
-    private ItemStack[] minecartContainerItems = new ItemStack[36];
+    private ItemStack[] minecartContainerItems = new ItemStack[27]; // CB - 36 -> 27
     public boolean dropContentsWhenDead = true;
     private ResourceLocation lootTable;
     private long lootTableSeed;
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+    
+    public ItemStack[] getContents() {
+    	return this.minecartContainerItems;
+    }
+    
+    public void onOpen(CraftHumanEntity who) {
+    	transaction.add(who);
+    }
+    
+    public void onClose(CraftHumanEntity who) {
+    	transaction.remove(who);
+    }
+    
+    public List<HumanEntity> getViewers() {
+    	return transaction;
+    }
+    
+    public InventoryHolder getOwner() {
+    	org.bukkit.entity.Entity cart = getBukkitEntity();
+    	if(cart instanceof InventoryHolder) return (InventoryHolder) cart;
+    	return null;
+    }
+    
+    public void setMaxStackSize(int size) {
+    	maxStack = size;
+    }
+    
+    @Override
+    public Location getLocation() {
+    	return getBukkitEntity().getLocation();
+    }
+    // CraftBukkit end
 
     public EntityMinecartContainer(World worldIn)
     {
