--- ../src-base/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -22,9 +22,9 @@
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<Optional<ItemStack>> FIREWORK_ITEM = EntityDataManager.<Optional<ItemStack>>createKey(EntityFireworkRocket.class, DataSerializers.OPTIONAL_ITEM_STACK);
+    public static final DataParameter<Optional<ItemStack>> FIREWORK_ITEM = EntityDataManager.<Optional<ItemStack>>createKey(EntityFireworkRocket.class, DataSerializers.OPTIONAL_ITEM_STACK);
     private int fireworkAge;
-    private int lifetime;
+    public int lifetime;
 
     public EntityFireworkRocket(World worldIn)
     {
