--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
@@ -57,6 +57,7 @@
 
     public boolean processInitialInteract(EntityPlayer player, EnumHand hand)
     {
+        if (super.processInitialInteract(player, hand)) return true;
         if (!this.world.isRemote)
         {
             player.displayGUIChest(this);
