--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartTNT.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartTNT.java
@@ -38,9 +38,9 @@
         EntityMinecart.registerFixesMinecart(fixer, EntityMinecartTNT.class);
     }
 
-    public EntityMinecart.Type getType()
+    public Type getType()
     {
-        return EntityMinecart.Type.TNT;
+        return Type.TNT;
     }
 
     public IBlockState getDefaultDisplayTile()
