--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecart.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecart.java
@@ -34,7 +34,17 @@
 import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.util.Vector;
 
+//CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.vehicle.VehicleDamageEvent;
+import org.bukkit.event.vehicle.VehicleDestroyEvent;
+import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
+import org.bukkit.util.Vector;
+// CraftBukkit end
+
 public abstract class EntityMinecart extends Entity implements IWorldNameable
 {
     private static final DataParameter<Integer> ROLLING_AMPLITUDE = EntityDataManager.<Integer>createKey(EntityMinecart.class, DataSerializers.VARINT);
@@ -71,6 +81,16 @@
     protected float maxSpeedAirLateral = defaultMaxSpeedAirLateral;
     protected float maxSpeedAirVertical = defaultMaxSpeedAirVertical;
     protected double dragAir = defaultDragAir;
+    // CraftBukkit start
+    public boolean slowWhenEmpty = true;
+    private double derailedX = 0.5;
+    private double derailedY = 0.5;
+    private double derailedZ = 0.5;
+    private double flyingX = 0.95;
+    private double flyingY = 0.95;
+    private double flyingZ = 0.95;
+    public double maxSpeed = 0.4D;
+    // CraftBukkit end
 
     public EntityMinecart(World worldIn)
     {
@@ -1269,7 +1289,27 @@
     {
         return -1;
     }
-
+    // CraftBukkit start - Methods for getting and setting flying and derailed velocity modifiers
+    public Vector getFlyingVelocityMod() {
+    	return new Vector(flyingX, flyingY, flyingZ);
+    }
+    
+    public void setFlyingVelocityMod(Vector flying) {
+    	flyingX = flying.getX();
+    	flyingY = flying.getY();
+    	flyingZ = flying.getZ();
+    }
+    
+    public Vector getDerailedVelocityMod() {
+    	return new Vector(derailedX, derailedY, derailedZ);
+    }
+    
+    public void setDerailedVelocityMod(Vector derailed) {
+    	derailedX = derailed.getX();
+    	derailedY = derailed.getY();
+    	derailedZ = derailed.getZ();
+    }
+    // CraftBukkit end
     /* =================================== FORGE END ===========================================*/
 
     public static enum Type
