--- ../src-base/minecraft/net/minecraft/world/border/WorldBorder.java
+++ ../src-work/minecraft/net/minecraft/world/border/WorldBorder.java
@@ -6,6 +6,7 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
@@ -24,6 +25,9 @@
     private int warningTime;
     private int warningDistance;
 
+    // TODO: Check if it's initialized properly
+    public WorldServer world;
+
     public WorldBorder()
     {
         this.endDiameter = this.startDiameter;
@@ -206,6 +210,7 @@
 
     public void addListener(IBorderListener listener)
     {
+        if (listeners.contains(listener)) return;
         this.listeners.add(listener);
     }
 
@@ -284,4 +289,9 @@
             iborderlistener.onWarningDistanceChanged(this, warningDistance);
         }
     }
+
+    public void removeListener(IBorderListener listener)
+    {
+        this.listeners.remove(listener);
+    }
 }
