--- a/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
@@ -194,8 +_,21 @@
       };
    }
 
+   // CatServer start
+   private final ThreadLocal<Boolean> catserver$isOverrideParseSelector = ThreadLocal.withInitial(() -> false);
+   protected void parseSelector(boolean override) throws CommandSyntaxException {
+      try {
+         this.catserver$isOverrideParseSelector.set(override);
+         this.m_121281_();
+      } finally {
+         this.catserver$isOverrideParseSelector.remove();
+      }
+   }
+
    protected void m_121281_() throws CommandSyntaxException {
-      this.f_121189_ = true;
+      var overridePermissions = this.catserver$isOverrideParseSelector.get(); // CatServer
+      this.catserver$isOverrideParseSelector.remove();
+      this.f_121189_ = !overridePermissions; // CraftBukkit
       this.f_121176_ = this::m_121322_;
       if (!this.f_121201_.canRead()) {
          throw f_121193_.createWithContext(this.f_121201_);
@@ -242,6 +_,7 @@
 
       }
    }
+   // CatServer end
 
    protected void m_121304_() throws CommandSyntaxException {
       if (this.f_121201_.canRead()) {
@@ -450,7 +_,20 @@
       this.f_121171_ = p_121269_;
    }
 
+   // CatServer start
+   public final ThreadLocal<Boolean> catserver$captureParseOverride = ThreadLocal.withInitial(() -> false);
+   public EntitySelector parse(boolean override) throws CommandSyntaxException {
+      try {
+         this.catserver$captureParseOverride.set(override);
+         return this.m_121377_();
+      } finally {
+         this.catserver$captureParseOverride.remove();
+      }
+   }
+
    public EntitySelector m_121377_() throws CommandSyntaxException {
+      var overridePermissions = this.catserver$captureParseOverride.get();
+      this.catserver$captureParseOverride.remove();
       this.f_121174_ = this.f_121201_.getCursor();
       this.f_121176_ = this::m_121286_;
       if (this.f_121201_.canRead() && this.f_121201_.peek() == '@') {
@@ -459,7 +_,10 @@
          }
 
          this.f_121201_.skip();
-         this.m_121281_();
+         EntitySelector forgeSelector = net.minecraftforge.common.command.EntitySelectorManager.parseSelector(this);
+         if (forgeSelector != null)
+            return forgeSelector;
+         this.parseSelector(overridePermissions); // CraftBukkit
       } else {
          this.m_121304_();
       }
@@ -467,6 +_,7 @@
       this.m_121229_();
       return this.m_121230_();
    }
+   // CatServer end
 
    private static void m_121247_(SuggestionsBuilder p_121248_) {
       p_121248_.suggest("@p", Component.m_237115_("argument.entity.selector.nearestPlayer"));
@@ -474,6 +_,7 @@
       p_121248_.suggest("@r", Component.m_237115_("argument.entity.selector.randomPlayer"));
       p_121248_.suggest("@s", Component.m_237115_("argument.entity.selector.self"));
       p_121248_.suggest("@e", Component.m_237115_("argument.entity.selector.allEntities"));
+      net.minecraftforge.common.command.EntitySelectorManager.fillSelectorSuggestions(p_121248_);
    }
 
    private CompletableFuture<Suggestions> m_121286_(SuggestionsBuilder p_121287_, Consumer<SuggestionsBuilder> p_121288_) {
