--- a/net/minecraft/world/level/block/state/properties/IntegerProperty.java
+++ b/net/minecraft/world/level/block/state/properties/IntegerProperty.java
@@ -8,9 +_,15 @@
 
 public class IntegerProperty extends Property<Integer> {
    private final ImmutableSet<Integer> f_61621_;
+   // CraftBukkit start
+   public final int min;
+   public final int max;
 
-   protected IntegerProperty(String p_61623_, int p_61624_, int p_61625_) {
+    protected IntegerProperty(String p_61623_, int p_61624_, int p_61625_) {
       super(p_61623_, Integer.class);
+       this.min = p_61624_;
+       this.max = p_61625_;
+       // CraftBukkit end
       if (p_61624_ < 0) {
          throw new IllegalArgumentException("Min value of " + p_61623_ + " must be 0 or greater");
       } else if (p_61625_ <= p_61624_) {
