--- a/net/minecraft/world/level/block/state/properties/IntegerProperty.java
+++ b/net/minecraft/world/level/block/state/properties/IntegerProperty.java
@@ -8,8 +_,8 @@
 
 public class IntegerProperty extends Property<Integer> {
    private final ImmutableSet<Integer> f_61621_;
-   private final int f_223000_;
-   private final int f_223001_;
+   public final int f_223000_; // CatServer - private -> public
+   public final int f_223001_; // CatServer - private -> public
 
    protected IntegerProperty(String p_61623_, int p_61624_, int p_61625_) {
       super(p_61623_, Integer.class);
