--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -30,6 +_,7 @@
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.profiling.ProfileResults;
 import org.slf4j.Logger;
+import org.bukkit.command.CommandSender;
 
 public class DebugCommand {
    private static final Logger f_136900_ = LogUtils.getLogger();
@@ -216,6 +_,11 @@
 
       public boolean m_142559_() {
          return true;
+      }
+
+      @Override
+      public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return null;
       }
    }
 }
