--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -25,6 +_,7 @@
 import net.minecraft.server.ServerFunctionManager;
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.profiling.ProfileResults;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 public class DebugCommand {
@@ -202,5 +_,12 @@
       public boolean m_142559_() {
          return true;
       }
+
+      // CraftBukkit start
+      @Override
+      public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return null;
+      }
+      // CraftBukkit end
    }
 }
