--- a/net/minecraft/server/commands/DifficultyCommand.java
+++ b/net/minecraft/server/commands/DifficultyCommand.java
@@ -35,7 +_,8 @@
 
    public static int m_136944_(CommandSourceStack p_136945_, Difficulty p_136946_) throws CommandSyntaxException {
       MinecraftServer minecraftserver = p_136945_.m_81377_();
-      if (minecraftserver.m_129910_().m_5472_() == p_136946_) {
+      net.minecraft.server.level.ServerLevel worldServer = p_136945_.m_81372_(); // CraftBukkit
+      if (worldServer.m_46791_() == p_136946_) {
          throw f_136933_.create(p_136946_.m_19036_());
       } else {
          minecraftserver.m_129827_(p_136946_, true);
