--- a/net/minecraft/server/commands/EffectCommands.java
+++ b/net/minecraft/server/commands/EffectCommands.java
@@ -61,7 +_,7 @@
       for(Entity entity : p_136968_) {
          if (entity instanceof LivingEntity) {
             MobEffectInstance mobeffectinstance = new MobEffectInstance(p_136969_, j, p_136971_, false, p_136972_);
-            if (((LivingEntity)entity).m_147207_(mobeffectinstance, p_136967_.m_81373_())) {
+            if (((LivingEntity)entity).addEffect(mobeffectinstance, p_136967_.m_81373_(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                ++i;
             }
          }
@@ -84,7 +_,7 @@
       int i = 0;
 
       for(Entity entity : p_136961_) {
-         if (entity instanceof LivingEntity && ((LivingEntity)entity).m_21219_()) {
+         if (entity instanceof LivingEntity && ((LivingEntity)entity).removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
             ++i;
          }
       }
@@ -106,7 +_,7 @@
       int i = 0;
 
       for(Entity entity : p_136964_) {
-         if (entity instanceof LivingEntity && ((LivingEntity)entity).m_21195_(p_136965_)) {
+         if (entity instanceof LivingEntity && ((LivingEntity)entity).removeEffect(p_136965_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
             ++i;
          }
       }
