--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -27,7 +_,7 @@
 
    static <T extends GameRules.Value<T>> int m_137754_(CommandContext<CommandSourceStack> p_137755_, GameRules.Key<T> p_137756_) {
       CommandSourceStack commandsourcestack = p_137755_.getSource();
-      T t = commandsourcestack.m_81377_().m_129900_().m_46170_(p_137756_);
+      T t = commandsourcestack.m_81372_().m_46469_().m_46170_(p_137756_); // CraftBukkit
       t.m_46370_(p_137755_, "value");
       commandsourcestack.m_288197_(() -> {
          return Component.m_237110_("commands.gamerule.set", p_137756_.m_46328_(), t.toString());
@@ -36,7 +_,7 @@
    }
 
    static <T extends GameRules.Value<T>> int m_137757_(CommandSourceStack p_137758_, GameRules.Key<T> p_137759_) {
-      T t = p_137758_.m_81377_().m_129900_().m_46170_(p_137759_);
+      T t = p_137758_.m_81372_().m_46469_().m_46170_(p_137759_); // CraftBukkit
       p_137758_.m_288197_(() -> {
          return Component.m_237110_("commands.gamerule.query", p_137759_.m_46328_(), t.toString());
       }, false);
