--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -49,7 +_,7 @@
                boolean flag = serverplayer.m_150109_().m_36054_(itemstack1);
                if (flag && itemstack1.m_41619_()) {
                   itemstack1.m_41764_(1);
-                  ItemEntity itementity1 = serverplayer.m_36176_(itemstack1, false);
+                  ItemEntity itementity1 = serverplayer.drop(itemstack1, false, false, false); // SPIGOT-2942: Add boolean to call event
                   if (itementity1 != null) {
                      itementity1.m_32065_();
                   }
