--- a/net/minecraft/server/commands/ListPlayersCommand.java
+++ b/net/minecraft/server/commands/ListPlayersCommand.java
@@ -34,6 +_,12 @@
    private static int m_137826_(CommandSourceStack p_137827_, Function<ServerPlayer, Component> p_137828_) {
       PlayerList playerlist = p_137827_.m_81377_().m_6846_();
       List<ServerPlayer> list = playerlist.m_11314_();
+      // CraftBukkit start
+      if (p_137827_.getBukkitSender() instanceof org.bukkit.entity.Player) {
+         org.bukkit.entity.Player sender = (org.bukkit.entity.Player) p_137827_.getBukkitSender();
+         list = list.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList());
+      }
+      // CraftBukkit end
       Component component = ComponentUtils.m_178440_(list, p_137828_);
       p_137827_.m_81354_(new TranslatableComponent("commands.list.players", list.size(), playerlist.m_11310_(), component), false);
       return list.size();
