--- a/net/minecraft/server/commands/ListPlayersCommand.java
+++ b/net/minecraft/server/commands/ListPlayersCommand.java
@@ -32,7 +_,13 @@
 
    private static int m_137826_(CommandSourceStack p_137827_, Function<ServerPlayer, Component> p_137828_) {
       PlayerList playerlist = p_137827_.m_81377_().m_6846_();
-      List<ServerPlayer> list = playerlist.m_11314_();
+      // CraftBukkit start
+      List<ServerPlayer> players = playerlist.m_11314_();
+      if (p_137827_.getBukkitSender() instanceof org.bukkit.entity.Player sender) {
+         players = players.stream().filter((ep) -> sender.canSee(ep.getBukkitEntity())).collect(java.util.stream.Collectors.toList());
+      }
+      List<ServerPlayer> list = players;
+      // CraftBukkit end
       Component component = ComponentUtils.m_178440_(list, p_137828_);
       p_137827_.m_288197_(() -> {
          return Component.m_237110_("commands.list.players", list.size(), playerlist.m_11310_(), component);
