--- a/net/minecraft/server/commands/LootCommand.java
+++ b/net/minecraft/server/commands/LootCommand.java
@@ -220,6 +_,7 @@
 
    private static int m_137945_(CommandSourceStack p_137946_, Vec3 p_137947_, List<ItemStack> p_137948_, LootCommand.Callback p_137949_) throws CommandSyntaxException {
       ServerLevel serverlevel = p_137946_.m_81372_();
+      p_137948_.removeIf(ItemStack::m_41619_); // CraftBukkit - SPIGOT-6959 Remove empty items for avoid throw an error in new EntityItem
       p_137948_.forEach((p_137884_) -> {
          ItemEntity itementity = new ItemEntity(serverlevel, p_137947_.f_82479_, p_137947_.f_82480_, p_137947_.f_82481_, p_137884_.m_41777_());
          itementity.m_32060_();
