--- a/net/minecraft/server/commands/ReloadCommand.java
+++ b/net/minecraft/server/commands/ReloadCommand.java
@@ -23,6 +_,16 @@
       });
    }
 
+   // CraftBukkit start
+   public static void reload(MinecraftServer minecraftserver) {
+      PackRepository resourcepackrepository = minecraftserver.m_129891_();
+      WorldData savedata = minecraftserver.m_129910_();
+      Collection<String> collection = resourcepackrepository.m_10523_();
+      Collection<String> collection1 = m_138222_(resourcepackrepository, savedata, collection);
+      minecraftserver.m_129861_(collection1);
+   }
+   // CraftBukkit end
+
    private static Collection<String> m_138222_(PackRepository p_138223_, WorldData p_138224_, Collection<String> p_138225_) {
       p_138223_.m_10506_();
       Collection<String> collection = Lists.newArrayList(p_138225_);
