--- a/net/minecraft/server/commands/ReloadCommand.java
+++ b/net/minecraft/server/commands/ReloadCommand.java
@@ -52,4 +_,14 @@
          return 0;
       }));
    }
+
+   // CraftBukkit start
+   public static void reload(MinecraftServer minecraftserver) {
+      PackRepository resourcepackrepository = minecraftserver.m_129891_();
+      WorldData savedata = minecraftserver.m_129910_();
+      Collection<String> collection = resourcepackrepository.m_10523_();
+      Collection<String> collection1 = m_138222_(resourcepackrepository, savedata, collection);
+      minecraftserver.m_129861_(collection1);
+   }
+   // CraftBukkit end
 }
