--- a/net/minecraft/server/commands/ScheduleCommand.java
+++ b/net/minecraft/server/commands/ScheduleCommand.java
@@ -52,7 +_,7 @@
       } else {
          long i = p_138429_.m_81372_().m_46467_() + (long)p_138431_;
          ResourceLocation resourcelocation = p_138430_.getFirst();
-         TimerQueue<MinecraftServer> timerqueue = p_138429_.m_81377_().m_129910_().m_5996_().m_7540_();
+         TimerQueue<MinecraftServer> timerqueue = p_138429_.m_81372_().f_8549_.m_5996_().m_7540_(); // CraftBukkit - SPIGOT-6667: Use world specific function timer
          p_138430_.getSecond().ifLeft((p_138453_) -> {
             String s = resourcelocation.toString();
             if (p_138432_) {
