--- a/net/minecraft/server/commands/SetSpawnCommand.java
+++ b/net/minecraft/server/commands/SetSpawnCommand.java
@@ -33,7 +_,7 @@
       ResourceKey<Level> resourcekey = p_138650_.m_81372_().m_46472_();
 
       for(ServerPlayer serverplayer : p_138651_) {
-         serverplayer.m_9158_(resourcekey, p_138652_, p_138653_, true, false);
+         serverplayer.setRespawnPosition(resourcekey, p_138652_, p_138653_, true, false, org.bukkit.event.player.PlayerSpawnChangeEvent.Cause.COMMAND); // CraftBukkit - 1.20.1
       }
 
       String s = resourcekey.m_135782_().toString();
