--- a/net/minecraft/server/commands/TimeCommand.java
+++ b/net/minecraft/server/commands/TimeCommand.java
@@ -7,6 +_,10 @@
 import net.minecraft.commands.arguments.TimeArgument;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.level.ServerLevel;
+import org.bukkit.Bukkit;
+import org.bukkit.event.world.TimeSkipEvent;
+
+import java.util.Iterator;
 
 public class TimeCommand {
    public static void m_139071_(CommandDispatcher<CommandSourceStack> p_139072_) {
@@ -43,8 +_,16 @@
    }
 
    public static int m_139077_(CommandSourceStack p_139078_, int p_139079_) {
-      for(ServerLevel serverlevel : p_139078_.m_81377_().m_129785_()) {
-         serverlevel.m_8615_((long)p_139079_);
+      Iterator<ServerLevel> iterator = com.google.common.collect.Iterators.singletonIterator(p_139078_.m_81372_()); // CraftBukkit - SPIGOT-6496: Only set the time for the world the command originates in
+      while(iterator.hasNext()) {
+         ServerLevel serverLevel = iterator.next();
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverLevel.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_139079_ - serverLevel.m_46468_());
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverLevel.m_8615_((long) serverLevel.m_46468_() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       p_139078_.m_81354_(new TranslatableComponent("commands.time.set", p_139079_), true);
@@ -52,8 +_,16 @@
    }
 
    public static int m_139082_(CommandSourceStack p_139083_, int p_139084_) {
-      for(ServerLevel serverlevel : p_139083_.m_81377_().m_129785_()) {
-         serverlevel.m_8615_(serverlevel.m_46468_() + (long)p_139084_);
+      Iterator iterator = com.google.common.collect.Iterators.singletonIterator(p_139083_.m_81372_()); // CraftBukkit - SPIGOT-6496: Only set the time for the world the command originates in
+      while (iterator.hasNext()) {
+         ServerLevel serverLevel = (ServerLevel) iterator.next();
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverLevel.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_139084_);
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverLevel.m_8615_(serverLevel.m_46468_() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       int i = m_139069_(p_139083_.m_81372_());
