--- a/net/minecraft/server/commands/TimeCommand.java
+++ b/net/minecraft/server/commands/TimeCommand.java
@@ -7,6 +_,8 @@
 import net.minecraft.commands.arguments.TimeArgument;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.level.ServerLevel;
+import org.bukkit.Bukkit;
+import org.bukkit.event.world.TimeSkipEvent;
 
 public class TimeCommand {
    public static void m_139071_(CommandDispatcher<CommandSourceStack> p_139072_) {
@@ -44,7 +_,13 @@
 
    public static int m_139077_(CommandSourceStack p_139078_, int p_139079_) {
       for(ServerLevel serverlevel : p_139078_.m_81377_().m_129785_()) {
-         serverlevel.m_8615_((long)p_139079_);
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverlevel.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_139079_ - serverlevel.m_46468_());
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverlevel.m_8615_((long) serverlevel.m_46468_() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       p_139078_.m_81354_(new TranslatableComponent("commands.time.set", p_139079_), true);
@@ -53,7 +_,13 @@
 
    public static int m_139082_(CommandSourceStack p_139083_, int p_139084_) {
       for(ServerLevel serverlevel : p_139083_.m_81377_().m_129785_()) {
-         serverlevel.m_8615_(serverlevel.m_46468_() + (long)p_139084_);
+         // CraftBukkit start
+         TimeSkipEvent event = new TimeSkipEvent(serverlevel.getWorld(), TimeSkipEvent.SkipReason.COMMAND, p_139084_);
+         Bukkit.getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            serverlevel.m_8615_(serverlevel.m_46468_() + event.getSkipAmount());
+         }
+         // CraftBukkit end
       }
 
       int i = m_139069_(p_139083_.m_81372_());
