--- a/net/minecraft/server/commands/TriggerCommand.java
+++ b/net/minecraft/server/commands/TriggerCommand.java
@@ -85,7 +_,7 @@
       if (p_139140_.m_83321_() != ObjectiveCriteria.f_83589_) {
          throw f_139136_.create();
       } else {
-         Scoreboard scoreboard = p_139139_.m_36329_();
+         Scoreboard scoreboard = p_139139_.m_20194_().m_129896_(); // CraftBukkit - SPIGOT-6917: use main scoreboard
          String s = p_139139_.m_6302_();
          if (!scoreboard.m_83461_(s, p_139140_)) {
             throw f_139135_.create();
