--- a/net/minecraft/server/commands/WorldBorderCommand.java
+++ b/net/minecraft/server/commands/WorldBorderCommand.java
@@ -53,7 +_,7 @@
    }
 
    private static int m_139256_(CommandSourceStack p_139257_, float p_139258_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139257_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139257_.m_81372_().m_6857_(); // CraftBukkit
       if (worldborder.m_61964_() == (double)p_139258_) {
          throw f_139243_.create();
       } else {
@@ -66,7 +_,7 @@
    }
 
    private static int m_139269_(CommandSourceStack p_139270_, float p_139271_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139270_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139270_.m_81372_().m_6857_(); // CraftBukkit
       if (worldborder.m_61965_() == (double)p_139271_) {
          throw f_139244_.create();
       } else {
@@ -79,7 +_,7 @@
    }
 
    private static int m_139259_(CommandSourceStack p_139260_, int p_139261_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139260_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139260_.m_81372_().m_6857_(); // CraftBukkit
       if (worldborder.m_61967_() == p_139261_) {
          throw f_139241_.create();
       } else {
@@ -92,7 +_,7 @@
    }
 
    private static int m_139272_(CommandSourceStack p_139273_, int p_139274_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139273_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139273_.m_81372_().m_6857_(); // CraftBukkit
       if (worldborder.m_61968_() == p_139274_) {
          throw f_139242_.create();
       } else {
@@ -105,7 +_,7 @@
    }
 
    private static int m_139250_(CommandSourceStack p_139251_) {
-      double d0 = p_139251_.m_81377_().m_129783_().m_6857_().m_61959_();
+      double d0 = p_139251_.m_81372_().m_6857_().m_61959_(); // CraftBukkit
       p_139251_.m_288197_(() -> {
          return Component.m_237110_("commands.worldborder.get", String.format(Locale.ROOT, "%.0f", d0));
       }, false);
@@ -113,7 +_,7 @@
    }
 
    private static int m_139262_(CommandSourceStack p_139263_, Vec2 p_139264_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139263_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139263_.m_81372_().m_6857_(); // CraftBukkit
       if (worldborder.m_6347_() == (double)p_139264_.f_82470_ && worldborder.m_6345_() == (double)p_139264_.f_82471_) {
          throw f_139237_.create();
       } else if (!((double)Math.abs(p_139264_.f_82470_) > 2.9999984E7D) && !((double)Math.abs(p_139264_.f_82471_) > 2.9999984E7D)) {
@@ -128,7 +_,7 @@
    }
 
    private static int m_139252_(CommandSourceStack p_139253_, double p_139254_, long p_139255_) throws CommandSyntaxException {
-      WorldBorder worldborder = p_139253_.m_81377_().m_129783_().m_6857_();
+      WorldBorder worldborder = p_139253_.m_81372_().m_6857_(); // CraftBukkit
       double d0 = worldborder.m_61959_();
       if (d0 == p_139254_) {
          throw f_139238_.create();
