--- a/net/minecraft/network/protocol/PacketUtils.java
+++ b/net/minecraft/network/protocol/PacketUtils.java
@@ -2,8 +_,10 @@
 
 import com.mojang.logging.LogUtils;
 import net.minecraft.network.PacketListener;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.RunningOnDifferentThreadException;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.network.ServerGamePacketListenerImpl;
 import net.minecraft.util.thread.BlockableEventLoop;
 import org.slf4j.Logger;
 
@@ -16,23 +_,20 @@
 
    public static <T extends PacketListener> void m_131363_(Packet<T> p_131364_, T p_131365_, BlockableEventLoop<?> p_131366_) throws RunningOnDifferentThreadException {
       if (!p_131366_.m_18695_()) {
-         p_131366_.m_201446_(() -> {
+         p_131366_.execute(() -> {
+            if (MinecraftServer.getServer().hasStopped() || (p_131365_ instanceof ServerGamePacketListenerImpl && ((ServerGamePacketListenerImpl) p_131365_).processedDisconnect)) return; // CraftBukkit, MC-142590
             if (p_131365_.m_6198_().m_129536_()) {
-               try {
-                  p_131364_.m_5797_(p_131365_);
-               } catch (Exception exception) {
-                  if (p_131365_.m_201767_()) {
-                     throw exception;
-                  }
-
-                  f_131354_.error("Failed to handle packet {}, suppressing error", p_131364_, exception);
-               }
+               p_131364_.m_5797_(p_131365_);
             } else {
                f_131354_.debug("Ignoring packet due to disconnection: {}", (Object)p_131364_);
             }
 
          });
          throw RunningOnDifferentThreadException.f_136017_;
+         // CraftBukkit start - SPIGOT-5477, MC-142590
+      } else if (MinecraftServer.getServer().hasStopped() || (p_131365_ instanceof ServerGamePacketListenerImpl && ((ServerGamePacketListenerImpl) p_131365_).processedDisconnect)) {
+         throw RunningOnDifferentThreadException.f_136017_;
+         // CraftBukkit end
       }
    }
 }
