--- ../src-base/minecraft/net/minecraft/advancements/critereon/BredAnimalsTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/BredAnimalsTrigger.java
@@ -18,29 +18,29 @@
 public class BredAnimalsTrigger implements ICriterionTrigger<BredAnimalsTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("bred_animals");
-    private final Map<PlayerAdvancements, BredAnimalsTrigger.Listeners> listeners = Maps.<PlayerAdvancements, BredAnimalsTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        BredAnimalsTrigger.Listeners bredanimalstrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners bredanimalstrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (bredanimalstrigger$listeners == null)
         {
-            bredanimalstrigger$listeners = new BredAnimalsTrigger.Listeners(playerAdvancementsIn);
+            bredanimalstrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, bredanimalstrigger$listeners);
         }
 
         bredanimalstrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        BredAnimalsTrigger.Listeners bredanimalstrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners bredanimalstrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (bredanimalstrigger$listeners != null)
         {
@@ -58,17 +58,17 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public BredAnimalsTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         EntityPredicate entitypredicate = EntityPredicate.deserialize(json.get("parent"));
         EntityPredicate entitypredicate1 = EntityPredicate.deserialize(json.get("partner"));
         EntityPredicate entitypredicate2 = EntityPredicate.deserialize(json.get("child"));
-        return new BredAnimalsTrigger.Instance(entitypredicate, entitypredicate1, entitypredicate2);
+        return new Instance(entitypredicate, entitypredicate1, entitypredicate2);
     }
 
     public void trigger(EntityPlayerMP player, EntityAnimal parent1, EntityAnimal parent2, EntityAgeable child)
     {
-        BredAnimalsTrigger.Listeners bredanimalstrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners bredanimalstrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (bredanimalstrigger$listeners != null)
         {
@@ -106,7 +106,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<BredAnimalsTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<BredAnimalsTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -118,27 +118,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(EntityPlayerMP player, EntityAnimal parent1, EntityAnimal parent2, EntityAgeable child)
             {
-                List<ICriterionTrigger.Listener<BredAnimalsTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((BredAnimalsTrigger.Instance)listener.getCriterionInstance()).test(player, parent1, parent2, child))
+                    if (((Instance)listener.getCriterionInstance()).test(player, parent1, parent2, child))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<BredAnimalsTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -147,7 +147,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<BredAnimalsTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
