--- ../src-base/minecraft/net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
@@ -19,29 +19,29 @@
 public class ChangeDimensionTrigger implements ICriterionTrigger<ChangeDimensionTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("changed_dimension");
-    private final Map<PlayerAdvancements, ChangeDimensionTrigger.Listeners> listeners = Maps.<PlayerAdvancements, ChangeDimensionTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ChangeDimensionTrigger.Listeners changedimensiontrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners changedimensiontrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (changedimensiontrigger$listeners == null)
         {
-            changedimensiontrigger$listeners = new ChangeDimensionTrigger.Listeners(playerAdvancementsIn);
+            changedimensiontrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, changedimensiontrigger$listeners);
         }
 
         changedimensiontrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ChangeDimensionTrigger.Listeners changedimensiontrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners changedimensiontrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (changedimensiontrigger$listeners != null)
         {
@@ -59,16 +59,16 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public ChangeDimensionTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         DimensionType dimensiontype = json.has("from") ? DimensionType.byName(JsonUtils.getString(json, "from")) : null;
         DimensionType dimensiontype1 = json.has("to") ? DimensionType.byName(JsonUtils.getString(json, "to")) : null;
-        return new ChangeDimensionTrigger.Instance(dimensiontype, dimensiontype1);
+        return new Instance(dimensiontype, dimensiontype1);
     }
 
     public void trigger(EntityPlayerMP player, DimensionType from, DimensionType to)
     {
-        ChangeDimensionTrigger.Listeners changedimensiontrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners changedimensiontrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (changedimensiontrigger$listeners != null)
         {
@@ -106,7 +106,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -118,27 +118,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(DimensionType from, DimensionType to)
             {
-                List<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((ChangeDimensionTrigger.Instance)listener.getCriterionInstance()).test(from, to))
+                    if (((Instance)listener.getCriterionInstance()).test(from, to))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -147,7 +147,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
