--- ../src-base/minecraft/net/minecraft/advancements/critereon/ConstructBeaconTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/ConstructBeaconTrigger.java
@@ -17,29 +17,29 @@
 public class ConstructBeaconTrigger implements ICriterionTrigger<ConstructBeaconTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("construct_beacon");
-    private final Map<PlayerAdvancements, ConstructBeaconTrigger.Listeners> listeners = Maps.<PlayerAdvancements, ConstructBeaconTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ConstructBeaconTrigger.Listeners constructbeacontrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners constructbeacontrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (constructbeacontrigger$listeners == null)
         {
-            constructbeacontrigger$listeners = new ConstructBeaconTrigger.Listeners(playerAdvancementsIn);
+            constructbeacontrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, constructbeacontrigger$listeners);
         }
 
         constructbeacontrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ConstructBeaconTrigger.Listeners constructbeacontrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners constructbeacontrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (constructbeacontrigger$listeners != null)
         {
@@ -57,15 +57,15 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public ConstructBeaconTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         MinMaxBounds minmaxbounds = MinMaxBounds.deserialize(json.get("level"));
-        return new ConstructBeaconTrigger.Instance(minmaxbounds);
+        return new Instance(minmaxbounds);
     }
 
     public void trigger(EntityPlayerMP player, TileEntityBeacon beacon)
     {
-        ConstructBeaconTrigger.Listeners constructbeacontrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners constructbeacontrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (constructbeacontrigger$listeners != null)
         {
@@ -92,7 +92,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -104,27 +104,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(TileEntityBeacon beacon)
             {
-                List<ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((ConstructBeaconTrigger.Instance)listener.getCriterionInstance()).test(beacon))
+                    if (((Instance)listener.getCriterionInstance()).test(beacon))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -133,7 +133,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<ConstructBeaconTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
