--- ../src-base/minecraft/net/minecraft/advancements/critereon/ConsumeItemTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/ConsumeItemTrigger.java
@@ -17,29 +17,29 @@
 public class ConsumeItemTrigger implements ICriterionTrigger<ConsumeItemTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("consume_item");
-    private final Map<PlayerAdvancements, ConsumeItemTrigger.Listeners> listeners = Maps.<PlayerAdvancements, ConsumeItemTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ConsumeItemTrigger.Listeners consumeitemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners consumeitemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (consumeitemtrigger$listeners == null)
         {
-            consumeitemtrigger$listeners = new ConsumeItemTrigger.Listeners(playerAdvancementsIn);
+            consumeitemtrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, consumeitemtrigger$listeners);
         }
 
         consumeitemtrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        ConsumeItemTrigger.Listeners consumeitemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners consumeitemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (consumeitemtrigger$listeners != null)
         {
@@ -57,15 +57,15 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public ConsumeItemTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         ItemPredicate itempredicate = ItemPredicate.deserialize(json.get("item"));
-        return new ConsumeItemTrigger.Instance(itempredicate);
+        return new Instance(itempredicate);
     }
 
     public void trigger(EntityPlayerMP player, ItemStack item)
     {
-        ConsumeItemTrigger.Listeners consumeitemtrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners consumeitemtrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (consumeitemtrigger$listeners != null)
         {
@@ -92,7 +92,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<ConsumeItemTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<ConsumeItemTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -104,27 +104,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(ItemStack item)
             {
-                List<ICriterionTrigger.Listener<ConsumeItemTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((ConsumeItemTrigger.Instance)listener.getCriterionInstance()).test(item))
+                    if (((Instance)listener.getCriterionInstance()).test(item))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<ConsumeItemTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -133,7 +133,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<ConsumeItemTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
