--- ../src-base/minecraft/net/minecraft/advancements/critereon/EnchantedItemTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/EnchantedItemTrigger.java
@@ -17,29 +17,29 @@
 public class EnchantedItemTrigger implements ICriterionTrigger<EnchantedItemTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("enchanted_item");
-    private final Map<PlayerAdvancements, EnchantedItemTrigger.Listeners> listeners = Maps.<PlayerAdvancements, EnchantedItemTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EnchantedItemTrigger.Listeners enchanteditemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners enchanteditemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (enchanteditemtrigger$listeners == null)
         {
-            enchanteditemtrigger$listeners = new EnchantedItemTrigger.Listeners(playerAdvancementsIn);
+            enchanteditemtrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, enchanteditemtrigger$listeners);
         }
 
         enchanteditemtrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EnchantedItemTrigger.Listeners enchanteditemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners enchanteditemtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (enchanteditemtrigger$listeners != null)
         {
@@ -57,16 +57,16 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public EnchantedItemTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         ItemPredicate itempredicate = ItemPredicate.deserialize(json.get("item"));
         MinMaxBounds minmaxbounds = MinMaxBounds.deserialize(json.get("levels"));
-        return new EnchantedItemTrigger.Instance(itempredicate, minmaxbounds);
+        return new Instance(itempredicate, minmaxbounds);
     }
 
     public void trigger(EntityPlayerMP player, ItemStack item, int levelsSpent)
     {
-        EnchantedItemTrigger.Listeners enchanteditemtrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners enchanteditemtrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (enchanteditemtrigger$listeners != null)
         {
@@ -102,7 +102,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<EnchantedItemTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<EnchantedItemTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -114,27 +114,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(ItemStack item, int levelsIn)
             {
-                List<ICriterionTrigger.Listener<EnchantedItemTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((EnchantedItemTrigger.Instance)listener.getCriterionInstance()).test(item, levelsIn))
+                    if (((Instance)listener.getCriterionInstance()).test(item, levelsIn))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<EnchantedItemTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -143,7 +143,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<EnchantedItemTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
