--- ../src-base/minecraft/net/minecraft/advancements/critereon/EnterBlockTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/EnterBlockTrigger.java
@@ -26,29 +26,29 @@
 public class EnterBlockTrigger implements ICriterionTrigger<EnterBlockTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("enter_block");
-    private final Map<PlayerAdvancements, EnterBlockTrigger.Listeners> listeners = Maps.<PlayerAdvancements, EnterBlockTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EnterBlockTrigger.Listeners enterblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners enterblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (enterblocktrigger$listeners == null)
         {
-            enterblocktrigger$listeners = new EnterBlockTrigger.Listeners(playerAdvancementsIn);
+            enterblocktrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, enterblocktrigger$listeners);
         }
 
         enterblocktrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EnterBlockTrigger.Listeners enterblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners enterblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (enterblocktrigger$listeners != null)
         {
@@ -66,7 +66,7 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public EnterBlockTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         Block block = null;
 
@@ -119,12 +119,12 @@
             }
         }
 
-        return new EnterBlockTrigger.Instance(block, map);
+        return new Instance(block, map);
     }
 
     public void trigger(EntityPlayerMP player, IBlockState state)
     {
-        EnterBlockTrigger.Listeners enterblocktrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners enterblocktrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (enterblocktrigger$listeners != null)
         {
@@ -171,7 +171,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<EnterBlockTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<EnterBlockTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -183,27 +183,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(IBlockState state)
             {
-                List<ICriterionTrigger.Listener<EnterBlockTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((EnterBlockTrigger.Instance)listener.getCriterionInstance()).test(state))
+                    if (((Instance)listener.getCriterionInstance()).test(state))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<EnterBlockTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -212,7 +212,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<EnterBlockTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
