--- ../src-base/minecraft/net/minecraft/advancements/critereon/EntityHurtPlayerTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/EntityHurtPlayerTrigger.java
@@ -17,29 +17,29 @@
 public class EntityHurtPlayerTrigger implements ICriterionTrigger<EntityHurtPlayerTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("entity_hurt_player");
-    private final Map<PlayerAdvancements, EntityHurtPlayerTrigger.Listeners> listeners = Maps.<PlayerAdvancements, EntityHurtPlayerTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EntityHurtPlayerTrigger.Listeners entityhurtplayertrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners entityhurtplayertrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (entityhurtplayertrigger$listeners == null)
         {
-            entityhurtplayertrigger$listeners = new EntityHurtPlayerTrigger.Listeners(playerAdvancementsIn);
+            entityhurtplayertrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, entityhurtplayertrigger$listeners);
         }
 
         entityhurtplayertrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        EntityHurtPlayerTrigger.Listeners entityhurtplayertrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners entityhurtplayertrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (entityhurtplayertrigger$listeners != null)
         {
@@ -57,15 +57,15 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public EntityHurtPlayerTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         DamagePredicate damagepredicate = DamagePredicate.deserialize(json.get("damage"));
-        return new EntityHurtPlayerTrigger.Instance(damagepredicate);
+        return new Instance(damagepredicate);
     }
 
     public void trigger(EntityPlayerMP player, DamageSource source, float amountDealt, float amountTaken, boolean wasBlocked)
     {
-        EntityHurtPlayerTrigger.Listeners entityhurtplayertrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners entityhurtplayertrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (entityhurtplayertrigger$listeners != null)
         {
@@ -92,7 +92,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47439_1_)
             {
@@ -104,27 +104,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(EntityPlayerMP player, DamageSource source, float amountDealt, float amountTaken, boolean wasBlocked)
             {
-                List<ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((EntityHurtPlayerTrigger.Instance)listener.getCriterionInstance()).test(player, source, amountDealt, amountTaken, wasBlocked))
+                    if (((Instance)listener.getCriterionInstance()).test(player, source, amountDealt, amountTaken, wasBlocked))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -133,7 +133,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<EntityHurtPlayerTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
