--- ../src-base/minecraft/net/minecraft/advancements/critereon/ImpossibleTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/ImpossibleTrigger.java
@@ -15,11 +15,11 @@
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ImpossibleTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<ImpossibleTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
     }
 
@@ -27,9 +27,9 @@
     {
     }
 
-    public ImpossibleTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
-        return new ImpossibleTrigger.Instance();
+        return new Instance();
     }
 
     public static class Instance extends AbstractCriterionInstance
