--- ../src-base/minecraft/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
@@ -20,29 +20,29 @@
 public class InventoryChangeTrigger implements ICriterionTrigger<InventoryChangeTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("inventory_changed");
-    private final Map<PlayerAdvancements, InventoryChangeTrigger.Listeners> listeners = Maps.<PlayerAdvancements, InventoryChangeTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        InventoryChangeTrigger.Listeners inventorychangetrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners inventorychangetrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (inventorychangetrigger$listeners == null)
         {
-            inventorychangetrigger$listeners = new InventoryChangeTrigger.Listeners(playerAdvancementsIn);
+            inventorychangetrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, inventorychangetrigger$listeners);
         }
 
         inventorychangetrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        InventoryChangeTrigger.Listeners inventorychangetrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners inventorychangetrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (inventorychangetrigger$listeners != null)
         {
@@ -60,19 +60,19 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public InventoryChangeTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         JsonObject jsonobject = JsonUtils.getJsonObject(json, "slots", new JsonObject());
         MinMaxBounds minmaxbounds = MinMaxBounds.deserialize(jsonobject.get("occupied"));
         MinMaxBounds minmaxbounds1 = MinMaxBounds.deserialize(jsonobject.get("full"));
         MinMaxBounds minmaxbounds2 = MinMaxBounds.deserialize(jsonobject.get("empty"));
         ItemPredicate[] aitempredicate = ItemPredicate.deserializeArray(json.get("items"));
-        return new InventoryChangeTrigger.Instance(minmaxbounds, minmaxbounds1, minmaxbounds2, aitempredicate);
+        return new Instance(minmaxbounds, minmaxbounds1, minmaxbounds2, aitempredicate);
     }
 
     public void trigger(EntityPlayerMP player, InventoryPlayer inventory)
     {
-        InventoryChangeTrigger.Listeners inventorychangetrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners inventorychangetrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (inventorychangetrigger$listeners != null)
         {
@@ -160,7 +160,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<InventoryChangeTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<InventoryChangeTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -172,27 +172,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(InventoryPlayer inventory)
             {
-                List<ICriterionTrigger.Listener<InventoryChangeTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((InventoryChangeTrigger.Instance)listener.getCriterionInstance()).test(inventory))
+                    if (((Instance)listener.getCriterionInstance()).test(inventory))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<InventoryChangeTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -201,7 +201,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<InventoryChangeTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
