--- ../src-base/minecraft/net/minecraft/advancements/critereon/LevitationTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/LevitationTrigger.java
@@ -17,29 +17,29 @@
 public class LevitationTrigger implements ICriterionTrigger<LevitationTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("levitation");
-    private final Map<PlayerAdvancements, LevitationTrigger.Listeners> listeners = Maps.<PlayerAdvancements, LevitationTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<LevitationTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        LevitationTrigger.Listeners levitationtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners levitationtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (levitationtrigger$listeners == null)
         {
-            levitationtrigger$listeners = new LevitationTrigger.Listeners(playerAdvancementsIn);
+            levitationtrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, levitationtrigger$listeners);
         }
 
         levitationtrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<LevitationTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        LevitationTrigger.Listeners levitationtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners levitationtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (levitationtrigger$listeners != null)
         {
@@ -57,16 +57,16 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public LevitationTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         DistancePredicate distancepredicate = DistancePredicate.deserialize(json.get("distance"));
         MinMaxBounds minmaxbounds = MinMaxBounds.deserialize(json.get("duration"));
-        return new LevitationTrigger.Instance(distancepredicate, minmaxbounds);
+        return new Instance(distancepredicate, minmaxbounds);
     }
 
     public void trigger(EntityPlayerMP player, Vec3d startPos, int duration)
     {
-        LevitationTrigger.Listeners levitationtrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners levitationtrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (levitationtrigger$listeners != null)
         {
@@ -102,7 +102,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<LevitationTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<LevitationTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -114,27 +114,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<LevitationTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<LevitationTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(EntityPlayerMP player, Vec3d startPos, int durationIn)
             {
-                List<ICriterionTrigger.Listener<LevitationTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<LevitationTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((LevitationTrigger.Instance)listener.getCriterionInstance()).test(player, startPos, durationIn))
+                    if (((Instance)listener.getCriterionInstance()).test(player, startPos, durationIn))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<LevitationTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -143,7 +143,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<LevitationTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
