--- ../src-base/minecraft/net/minecraft/advancements/critereon/PlacedBlockTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/PlacedBlockTrigger.java
@@ -29,29 +29,29 @@
 public class PlacedBlockTrigger implements ICriterionTrigger<PlacedBlockTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("placed_block");
-    private final Map<PlayerAdvancements, PlacedBlockTrigger.Listeners> listeners = Maps.<PlayerAdvancements, PlacedBlockTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        PlacedBlockTrigger.Listeners placedblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners placedblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (placedblocktrigger$listeners == null)
         {
-            placedblocktrigger$listeners = new PlacedBlockTrigger.Listeners(playerAdvancementsIn);
+            placedblocktrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, placedblocktrigger$listeners);
         }
 
         placedblocktrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        PlacedBlockTrigger.Listeners placedblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners placedblocktrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (placedblocktrigger$listeners != null)
         {
@@ -69,7 +69,7 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public PlacedBlockTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         Block block = null;
 
@@ -124,13 +124,13 @@
 
         LocationPredicate locationpredicate = LocationPredicate.deserialize(json.get("location"));
         ItemPredicate itempredicate = ItemPredicate.deserialize(json.get("item"));
-        return new PlacedBlockTrigger.Instance(block, map, locationpredicate, itempredicate);
+        return new Instance(block, map, locationpredicate, itempredicate);
     }
 
     public void trigger(EntityPlayerMP player, BlockPos pos, ItemStack item)
     {
         IBlockState iblockstate = player.world.getBlockState(pos);
-        PlacedBlockTrigger.Listeners placedblocktrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners placedblocktrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (placedblocktrigger$listeners != null)
         {
@@ -188,7 +188,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -200,27 +200,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(IBlockState state, BlockPos pos, WorldServer world, ItemStack item)
             {
-                List<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((PlacedBlockTrigger.Instance)listener.getCriterionInstance()).test(state, pos, world, item))
+                    if (((Instance)listener.getCriterionInstance()).test(state, pos, world, item))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -229,7 +229,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
