--- ../src-base/minecraft/net/minecraft/advancements/critereon/RecipeUnlockedTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/RecipeUnlockedTrigger.java
@@ -20,29 +20,29 @@
 public class RecipeUnlockedTrigger implements ICriterionTrigger<RecipeUnlockedTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("recipe_unlocked");
-    private final Map<PlayerAdvancements, RecipeUnlockedTrigger.Listeners> listeners = Maps.<PlayerAdvancements, RecipeUnlockedTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        RecipeUnlockedTrigger.Listeners recipeunlockedtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners recipeunlockedtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (recipeunlockedtrigger$listeners == null)
         {
-            recipeunlockedtrigger$listeners = new RecipeUnlockedTrigger.Listeners(playerAdvancementsIn);
+            recipeunlockedtrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, recipeunlockedtrigger$listeners);
         }
 
         recipeunlockedtrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        RecipeUnlockedTrigger.Listeners recipeunlockedtrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners recipeunlockedtrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (recipeunlockedtrigger$listeners != null)
         {
@@ -60,7 +60,7 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public RecipeUnlockedTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         ResourceLocation resourcelocation = new ResourceLocation(JsonUtils.getString(json, "recipe"));
         IRecipe irecipe = CraftingManager.getRecipe(resourcelocation);
@@ -71,13 +71,13 @@
         }
         else
         {
-            return new RecipeUnlockedTrigger.Instance(irecipe);
+            return new Instance(irecipe);
         }
     }
 
     public void trigger(EntityPlayerMP player, IRecipe recipe)
     {
-        RecipeUnlockedTrigger.Listeners recipeunlockedtrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners recipeunlockedtrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (recipeunlockedtrigger$listeners != null)
         {
@@ -104,7 +104,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -116,27 +116,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(IRecipe recipe)
             {
-                List<ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((RecipeUnlockedTrigger.Instance)listener.getCriterionInstance()).test(recipe))
+                    if (((Instance)listener.getCriterionInstance()).test(recipe))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -145,7 +145,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<RecipeUnlockedTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
