--- ../src-base/minecraft/net/minecraft/advancements/critereon/TameAnimalTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/TameAnimalTrigger.java
@@ -17,29 +17,29 @@
 public class TameAnimalTrigger implements ICriterionTrigger<TameAnimalTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("tame_animal");
-    private final Map<PlayerAdvancements, TameAnimalTrigger.Listeners> listeners = Maps.<PlayerAdvancements, TameAnimalTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        TameAnimalTrigger.Listeners tameanimaltrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners tameanimaltrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (tameanimaltrigger$listeners == null)
         {
-            tameanimaltrigger$listeners = new TameAnimalTrigger.Listeners(playerAdvancementsIn);
+            tameanimaltrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, tameanimaltrigger$listeners);
         }
 
         tameanimaltrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        TameAnimalTrigger.Listeners tameanimaltrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners tameanimaltrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (tameanimaltrigger$listeners != null)
         {
@@ -57,15 +57,15 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public TameAnimalTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         EntityPredicate entitypredicate = EntityPredicate.deserialize(json.get("entity"));
-        return new TameAnimalTrigger.Instance(entitypredicate);
+        return new Instance(entitypredicate);
     }
 
     public void trigger(EntityPlayerMP player, EntityAnimal entity)
     {
-        TameAnimalTrigger.Listeners tameanimaltrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners tameanimaltrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (tameanimaltrigger$listeners != null)
         {
@@ -92,7 +92,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<TameAnimalTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<TameAnimalTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -104,27 +104,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(EntityPlayerMP player, EntityAnimal entity)
             {
-                List<ICriterionTrigger.Listener<TameAnimalTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((TameAnimalTrigger.Instance)listener.getCriterionInstance()).test(player, entity))
+                    if (((Instance)listener.getCriterionInstance()).test(player, entity))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<TameAnimalTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -133,7 +133,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<TameAnimalTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
