--- ../src-base/minecraft/net/minecraft/advancements/critereon/VillagerTradeTrigger.java
+++ ../src-work/minecraft/net/minecraft/advancements/critereon/VillagerTradeTrigger.java
@@ -18,29 +18,29 @@
 public class VillagerTradeTrigger implements ICriterionTrigger<VillagerTradeTrigger.Instance>
 {
     private static final ResourceLocation ID = new ResourceLocation("villager_trade");
-    private final Map<PlayerAdvancements, VillagerTradeTrigger.Listeners> listeners = Maps.<PlayerAdvancements, VillagerTradeTrigger.Listeners>newHashMap();
+    private final Map<PlayerAdvancements, Listeners> listeners = Maps.<PlayerAdvancements, Listeners>newHashMap();
 
     public ResourceLocation getId()
     {
         return ID;
     }
 
-    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener)
+    public void addListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        VillagerTradeTrigger.Listeners villagertradetrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners villagertradetrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (villagertradetrigger$listeners == null)
         {
-            villagertradetrigger$listeners = new VillagerTradeTrigger.Listeners(playerAdvancementsIn);
+            villagertradetrigger$listeners = new Listeners(playerAdvancementsIn);
             this.listeners.put(playerAdvancementsIn, villagertradetrigger$listeners);
         }
 
         villagertradetrigger$listeners.add(listener);
     }
 
-    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener)
+    public void removeListener(PlayerAdvancements playerAdvancementsIn, ICriterionTrigger.Listener<Instance> listener)
     {
-        VillagerTradeTrigger.Listeners villagertradetrigger$listeners = this.listeners.get(playerAdvancementsIn);
+        Listeners villagertradetrigger$listeners = this.listeners.get(playerAdvancementsIn);
 
         if (villagertradetrigger$listeners != null)
         {
@@ -58,16 +58,16 @@
         this.listeners.remove(playerAdvancementsIn);
     }
 
-    public VillagerTradeTrigger.Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
+    public Instance deserializeInstance(JsonObject json, JsonDeserializationContext context)
     {
         EntityPredicate entitypredicate = EntityPredicate.deserialize(json.get("villager"));
         ItemPredicate itempredicate = ItemPredicate.deserialize(json.get("item"));
-        return new VillagerTradeTrigger.Instance(entitypredicate, itempredicate);
+        return new Instance(entitypredicate, itempredicate);
     }
 
     public void trigger(EntityPlayerMP player, EntityVillager villager, ItemStack item)
     {
-        VillagerTradeTrigger.Listeners villagertradetrigger$listeners = this.listeners.get(player.getAdvancements());
+        Listeners villagertradetrigger$listeners = this.listeners.get(player.getAdvancements());
 
         if (villagertradetrigger$listeners != null)
         {
@@ -103,7 +103,7 @@
     static class Listeners
         {
             private final PlayerAdvancements playerAdvancements;
-            private final Set<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>> listeners = Sets.<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>>newHashSet();
+            private final Set<ICriterionTrigger.Listener<Instance>> listeners = Sets.<ICriterionTrigger.Listener<Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements playerAdvancementsIn)
             {
@@ -115,27 +115,27 @@
                 return this.listeners.isEmpty();
             }
 
-            public void add(ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener)
+            public void add(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.add(listener);
             }
 
-            public void remove(ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener)
+            public void remove(ICriterionTrigger.Listener<Instance> listener)
             {
                 this.listeners.remove(listener);
             }
 
             public void trigger(EntityPlayerMP player, EntityVillager villager, ItemStack item)
             {
-                List<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>> list = null;
+                List<ICriterionTrigger.Listener<Instance>> list = null;
 
-                for (ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener : this.listeners)
+                for (ICriterionTrigger.Listener<Instance> listener : this.listeners)
                 {
-                    if (((VillagerTradeTrigger.Instance)listener.getCriterionInstance()).test(player, villager, item))
+                    if (((Instance)listener.getCriterionInstance()).test(player, villager, item))
                     {
                         if (list == null)
                         {
-                            list = Lists.<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>>newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<Instance>>newArrayList();
                         }
 
                         list.add(listener);
@@ -144,7 +144,7 @@
 
                 if (list != null)
                 {
-                    for (ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener1 : list)
+                    for (ICriterionTrigger.Listener<Instance> listener1 : list)
                     {
                         listener1.grantCriterion(this.playerAdvancements);
                     }
