--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -28,6 +_,7 @@
 import net.minecraft.world.level.block.entity.BannerBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public interface CauldronInteraction {
    Map<Item, CauldronInteraction> f_175606_ = m_175617_();
@@ -49,6 +_,11 @@
          return InteractionResult.PASS;
       } else {
          if (!p_175663_.f_46443_) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.lowerFillLevel(p_175662_, p_175663_, p_175664_, p_175665_, CauldronLevelChangeEvent.ChangeReason.SHULKER_WASH)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             ItemStack itemstack = new ItemStack(Blocks.f_50456_);
             if (p_175667_.m_41782_()) {
                itemstack.m_41751_(p_175667_.m_41783_().m_6426_());
@@ -56,7 +_,7 @@
 
             p_175665_.m_21008_(p_175666_, itemstack);
             p_175665_.m_36220_(Stats.f_12947_);
-            LayeredCauldronBlock.m_153559_(p_175662_, p_175663_, p_175664_);
+            //LayeredCauldronBlock.lowerFillLevel(p_175662_, p_175663_, p_175664_); // CraftBukkit
          }
 
          return InteractionResult.m_19078_(p_175663_.f_46443_);
@@ -67,6 +_,11 @@
          return InteractionResult.PASS;
       } else {
          if (!p_175654_.f_46443_) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.lowerFillLevel(p_175653_, p_175654_, p_175655_, p_175656_, CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             ItemStack itemstack = p_175658_.m_41777_();
             itemstack.m_41764_(1);
             BannerBlockEntity.m_58509_(itemstack);
@@ -83,7 +_,7 @@
             }
 
             p_175656_.m_36220_(Stats.f_12946_);
-            LayeredCauldronBlock.m_153559_(p_175653_, p_175654_, p_175655_);
+            //LayeredCauldronBlock.lowerFillLevel(p_175653_, p_175654_, p_175655_); // CraftBukkit
          }
 
          return InteractionResult.m_19078_(p_175654_.f_46443_);
@@ -99,9 +_,14 @@
             return InteractionResult.PASS;
          } else {
             if (!p_175630_.f_46443_) {
+               // CraftBukkit start
+               if (!LayeredCauldronBlock.lowerFillLevel(p_175629_, p_175630_, p_175631_, p_175632_, CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                  return InteractionResult.SUCCESS;
+               }
+               // CraftBukkit end
                dyeableleatheritem.m_41123_(p_175634_);
                p_175632_.m_36220_(Stats.f_12945_);
-               LayeredCauldronBlock.m_153559_(p_175629_, p_175630_, p_175631_);
+               //LayeredCauldronBlock.lowerFillLevel(p_175629_, p_175630_, p_175631_); // CraftBukkit
             }
 
             return InteractionResult.m_19078_(p_175630_.f_46443_);
@@ -126,11 +_,16 @@
             return InteractionResult.PASS;
          } else {
             if (!p_175733_.f_46443_) {
+               // CraftBukkit start
+               if (!LayeredCauldronBlock.changeLevel(p_175732_, p_175733_, p_175734_, Blocks.f_152476_.m_49966_(), p_175735_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                  return InteractionResult.SUCCESS;
+               }
+               // CraftBukkit end
                Item item = p_175737_.m_41720_();
                p_175735_.m_21008_(p_175736_, ItemUtils.m_41813_(p_175737_, p_175735_, new ItemStack(Items.f_42590_)));
                p_175735_.m_36220_(Stats.f_12944_);
                p_175735_.m_36246_(Stats.f_12982_.m_12902_(item));
-               p_175733_.m_46597_(p_175734_, Blocks.f_152476_.m_49966_());
+               //p_175733_.setBlockAndUpdate(p_175734_, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
                p_175733_.m_5594_((Player)null, p_175734_, SoundEvents.f_11769_, SoundSource.BLOCKS, 1.0F, 1.0F);
                p_175733_.m_142346_((Entity)null, GameEvent.f_157769_, p_175734_);
             }
@@ -146,11 +_,16 @@
       });
       f_175607_.put(Items.f_42590_, (p_175718_, p_175719_, p_175720_, p_175721_, p_175722_, p_175723_) -> {
          if (!p_175719_.f_46443_) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_, p_175721_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = p_175723_.m_41720_();
             p_175721_.m_21008_(p_175722_, ItemUtils.m_41813_(p_175723_, p_175721_, PotionUtils.m_43549_(new ItemStack(Items.f_42589_), Potions.f_43599_)));
             p_175721_.m_36220_(Stats.f_12944_);
             p_175721_.m_36246_(Stats.f_12982_.m_12902_(item));
-            LayeredCauldronBlock.m_153559_(p_175718_, p_175719_, p_175720_);
+            // LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_); // CraftBukkit
             p_175719_.m_5594_((Player)null, p_175720_, SoundEvents.f_11770_, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175719_.m_142346_((Entity)null, GameEvent.f_157816_, p_175720_);
          }
@@ -160,10 +_,15 @@
       f_175607_.put(Items.f_42589_, (p_175704_, p_175705_, p_175706_, p_175707_, p_175708_, p_175709_) -> {
          if (p_175704_.m_61143_(LayeredCauldronBlock.f_153514_) != 3 && PotionUtils.m_43579_(p_175709_) == Potions.f_43599_) {
             if (!p_175705_.f_46443_) {
+               // CraftBukkit start
+               if (!LayeredCauldronBlock.changeLevel(p_175704_, p_175705_, p_175706_, p_175704_.m_61122_(LayeredCauldronBlock.f_153514_), p_175707_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                  return InteractionResult.SUCCESS;
+               }
+               // CraftBukkit end
                p_175707_.m_21008_(p_175708_, ItemUtils.m_41813_(p_175709_, p_175707_, new ItemStack(Items.f_42590_)));
                p_175707_.m_36220_(Stats.f_12944_);
                p_175707_.m_36246_(Stats.f_12982_.m_12902_(p_175709_.m_41720_()));
-               p_175705_.m_46597_(p_175706_, p_175704_.m_61122_(LayeredCauldronBlock.f_153514_));
+               //p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL)); // CraftBukkit
                p_175705_.m_5594_((Player)null, p_175706_, SoundEvents.f_11769_, SoundSource.BLOCKS, 1.0F, 1.0F);
                p_175705_.m_142346_((Entity)null, GameEvent.f_157769_, p_175706_);
             }
@@ -235,11 +_,16 @@
          return InteractionResult.PASS;
       } else {
          if (!p_175637_.f_46443_) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.changeLevel(p_175636_, p_175637_, p_175638_, Blocks.f_50256_.m_49966_(), p_175639_, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = p_175641_.m_41720_();
             p_175639_.m_21008_(p_175640_, ItemUtils.m_41813_(p_175641_, p_175639_, p_175642_));
             p_175639_.m_36220_(Stats.f_12944_);
             p_175639_.m_36246_(Stats.f_12982_.m_12902_(item));
-            p_175637_.m_46597_(p_175638_, Blocks.f_50256_.m_49966_());
+            // p_175637_.setBlockAndUpdate(p_175638_, Blocks.CAULDRON.defaultBlockState()); // CraftBukkit
             p_175637_.m_5594_((Player)null, p_175638_, p_175644_, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175637_.m_142346_((Entity)null, GameEvent.f_157816_, p_175638_);
          }
@@ -250,11 +_,16 @@
 
    static InteractionResult m_175618_(Level p_175619_, BlockPos p_175620_, Player p_175621_, InteractionHand p_175622_, ItemStack p_175623_, BlockState p_175624_, SoundEvent p_175625_) {
       if (!p_175619_.f_46443_) {
+         // CraftBukkit start
+         if (!LayeredCauldronBlock.changeLevel(p_175624_, p_175619_, p_175620_, p_175624_, p_175621_, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          Item item = p_175623_.m_41720_();
          p_175621_.m_21008_(p_175622_, ItemUtils.m_41813_(p_175623_, p_175621_, new ItemStack(Items.f_42446_)));
          p_175621_.m_36220_(Stats.f_12943_);
          p_175621_.m_36246_(Stats.f_12982_.m_12902_(item));
-         p_175619_.m_46597_(p_175620_, p_175624_);
+         //p_175619_.setBlockAndUpdate(p_175620_, p_175624_); // CraftBukkit
          p_175619_.m_5594_((Player)null, p_175620_, p_175625_, SoundSource.BLOCKS, 1.0F, 1.0F);
          p_175619_.m_142346_((Entity)null, GameEvent.f_157769_, p_175620_);
       }
