--- a/net/minecraft/entity/merchant/IMerchant.java
+++ b/net/minecraft/entity/merchant/IMerchant.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftMerchant;
 
 public interface IMerchant {
    void func_70932_a_(@Nullable PlayerEntity p_70932_1_);
@@ -55,4 +_,6 @@
       }
 
    }
+
+   CraftMerchant getCraftMerchant(); // CraftBukkit
 }
