--- ../src-base/minecraft/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -30,6 +30,11 @@
 
     public static void updateRenderInfo(EntityPlayer entityplayerIn, boolean p_74583_1_)
     {
+        updateRenderInfo((Entity) entityplayerIn, p_74583_1_);
+    }
+
+    public static void updateRenderInfo(Entity entityplayerIn, boolean p_74583_1_)
+    {
         GlStateManager.getFloat(2982, MODELVIEW);
         GlStateManager.getFloat(2983, PROJECTION);
         GlStateManager.glGetInteger(2978, VIEWPORT);
@@ -108,4 +113,14 @@
     {
         return rotationXY;
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    /**
+     * Vector from render view entity position (corrected for partialTickTime) to the middle of screen
+     */
+    public static Vec3d getCameraPosition()
+    {
+        return position;
+    }
 }
