--- ../src-base/minecraft/net/minecraft/client/renderer/ChestRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ChestRenderer.java
@@ -13,6 +13,7 @@
     {
         GlStateManager.color(color, color, color, 1.0F);
         GlStateManager.rotate(90.0F, 0.0F, 1.0F, 0.0F);
-        TileEntityItemStackRenderer.instance.renderByItem(new ItemStack(blockIn));
+        ItemStack stack = new ItemStack(blockIn);
+        stack.getItem().getTileEntityItemStackRenderer().renderByItem(stack);
     }
 }
