--- ../src-base/minecraft/net/minecraft/client/renderer/GlStateManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/GlStateManager.java
@@ -20,26 +20,26 @@
 {
     private static final FloatBuffer BUF_FLOAT_16 = BufferUtils.createFloatBuffer(16);
     private static final FloatBuffer BUF_FLOAT_4 = BufferUtils.createFloatBuffer(4);
-    private static final GlStateManager.AlphaState alphaState = new GlStateManager.AlphaState();
-    private static final GlStateManager.BooleanState lightingState = new GlStateManager.BooleanState(2896);
-    private static final GlStateManager.BooleanState[] lightState = new GlStateManager.BooleanState[8];
-    private static final GlStateManager.ColorMaterialState colorMaterialState;
-    private static final GlStateManager.BlendState blendState;
-    private static final GlStateManager.DepthState depthState;
-    private static final GlStateManager.FogState fogState;
-    private static final GlStateManager.CullState cullState;
-    private static final GlStateManager.PolygonOffsetState polygonOffsetState;
-    private static final GlStateManager.ColorLogicState colorLogicState;
-    private static final GlStateManager.TexGenState texGenState;
-    private static final GlStateManager.ClearState clearState;
-    private static final GlStateManager.StencilState stencilState;
-    private static final GlStateManager.BooleanState normalizeState;
+    private static final AlphaState alphaState = new AlphaState();
+    private static final BooleanState lightingState = new BooleanState(2896);
+    private static final BooleanState[] lightState = new BooleanState[8];
+    private static final ColorMaterialState colorMaterialState;
+    private static final BlendState blendState;
+    private static final DepthState depthState;
+    private static final FogState fogState;
+    private static final CullState cullState;
+    private static final PolygonOffsetState polygonOffsetState;
+    private static final ColorLogicState colorLogicState;
+    private static final TexGenState texGenState;
+    private static final ClearState clearState;
+    private static final StencilState stencilState;
+    private static final BooleanState normalizeState;
     private static int activeTextureUnit;
-    private static final GlStateManager.TextureState[] textureState;
+    private static final TextureState[] textureState;
     private static int activeShadeModel;
-    private static final GlStateManager.BooleanState rescaleNormalState;
-    private static final GlStateManager.ColorMask colorMaskState;
-    private static final GlStateManager.Color colorState;
+    private static final BooleanState rescaleNormalState;
+    private static final ColorMask colorMaskState;
+    private static final Color colorState;
 
     public static void pushAttrib()
     {
@@ -164,7 +164,7 @@
         blendState.blend.setEnabled();
     }
 
-    public static void blendFunc(GlStateManager.SourceFactor srcFactor, GlStateManager.DestFactor dstFactor)
+    public static void blendFunc(SourceFactor srcFactor, DestFactor dstFactor)
     {
         blendFunc(srcFactor.factor, dstFactor.factor);
     }
@@ -179,7 +179,7 @@
         }
     }
 
-    public static void tryBlendFuncSeparate(GlStateManager.SourceFactor srcFactor, GlStateManager.DestFactor dstFactor, GlStateManager.SourceFactor srcFactorAlpha, GlStateManager.DestFactor dstFactorAlpha)
+    public static void tryBlendFuncSeparate(SourceFactor srcFactor, DestFactor dstFactor, SourceFactor srcFactorAlpha, DestFactor dstFactorAlpha)
     {
         tryBlendFuncSeparate(srcFactor.factor, dstFactor.factor, srcFactorAlpha.factor, dstFactorAlpha.factor);
     }
@@ -238,7 +238,7 @@
         fogState.fog.setDisabled();
     }
 
-    public static void setFog(GlStateManager.FogMode fogMode)
+    public static void setFog(FogMode fogMode)
     {
         setFog(fogMode.capabilityId);
     }
@@ -299,7 +299,7 @@
         cullState.cullFace.setDisabled();
     }
 
-    public static void cullFace(GlStateManager.CullFace cullFace)
+    public static void cullFace(CullFace cullFace)
     {
         cullFace(cullFace.mode);
     }
@@ -348,7 +348,7 @@
         colorLogicState.colorLogicOp.setDisabled();
     }
 
-    public static void colorLogicOp(GlStateManager.LogicOp logicOperation)
+    public static void colorLogicOp(LogicOp logicOperation)
     {
         colorLogicOp(logicOperation.opcode);
     }
@@ -362,19 +362,19 @@
         }
     }
 
-    public static void enableTexGenCoord(GlStateManager.TexGen texGen)
+    public static void enableTexGenCoord(TexGen texGen)
     {
         texGenCoord(texGen).textureGen.setEnabled();
     }
 
-    public static void disableTexGenCoord(GlStateManager.TexGen texGen)
+    public static void disableTexGenCoord(TexGen texGen)
     {
         texGenCoord(texGen).textureGen.setDisabled();
     }
 
-    public static void texGen(GlStateManager.TexGen texGen, int param)
+    public static void texGen(TexGen texGen, int param)
     {
-        GlStateManager.TexGenCoord glstatemanager$texgencoord = texGenCoord(texGen);
+        TexGenCoord glstatemanager$texgencoord = texGenCoord(texGen);
 
         if (param != glstatemanager$texgencoord.param)
         {
@@ -383,12 +383,12 @@
         }
     }
 
-    public static void texGen(GlStateManager.TexGen texGen, int pname, FloatBuffer params)
+    public static void texGen(TexGen texGen, int pname, FloatBuffer params)
     {
         GL11.glTexGen(texGenCoord(texGen).coord, pname, params);
     }
 
-    private static GlStateManager.TexGenCoord texGenCoord(GlStateManager.TexGen texGen)
+    private static TexGenCoord texGenCoord(TexGen texGen)
     {
         switch (texGen)
         {
@@ -463,7 +463,7 @@
     {
         GL11.glDeleteTextures(texture);
 
-        for (GlStateManager.TextureState glstatemanager$texturestate : textureState)
+        for (TextureState glstatemanager$texturestate : textureState)
         {
             if (glstatemanager$texturestate.textureName == texture)
             {
@@ -825,12 +825,12 @@
         return GL11.glGetInteger(parameterName);
     }
 
-    public static void enableBlendProfile(GlStateManager.Profile p_187408_0_)
+    public static void enableBlendProfile(Profile p_187408_0_)
     {
         p_187408_0_.apply();
     }
 
-    public static void disableBlendProfile(GlStateManager.Profile p_187440_0_)
+    public static void disableBlendProfile(Profile p_187440_0_)
     {
         p_187440_0_.clean();
     }
@@ -839,43 +839,43 @@
     {
         for (int i = 0; i < 8; ++i)
         {
-            lightState[i] = new GlStateManager.BooleanState(16384 + i);
+            lightState[i] = new BooleanState(16384 + i);
         }
 
-        colorMaterialState = new GlStateManager.ColorMaterialState();
-        blendState = new GlStateManager.BlendState();
-        depthState = new GlStateManager.DepthState();
-        fogState = new GlStateManager.FogState();
-        cullState = new GlStateManager.CullState();
-        polygonOffsetState = new GlStateManager.PolygonOffsetState();
-        colorLogicState = new GlStateManager.ColorLogicState();
-        texGenState = new GlStateManager.TexGenState();
-        clearState = new GlStateManager.ClearState();
-        stencilState = new GlStateManager.StencilState();
-        normalizeState = new GlStateManager.BooleanState(2977);
-        textureState = new GlStateManager.TextureState[8];
+        colorMaterialState = new ColorMaterialState();
+        blendState = new BlendState();
+        depthState = new DepthState();
+        fogState = new FogState();
+        cullState = new CullState();
+        polygonOffsetState = new PolygonOffsetState();
+        colorLogicState = new ColorLogicState();
+        texGenState = new TexGenState();
+        clearState = new ClearState();
+        stencilState = new StencilState();
+        normalizeState = new BooleanState(2977);
+        textureState = new TextureState[8];
 
         for (int j = 0; j < 8; ++j)
         {
-            textureState[j] = new GlStateManager.TextureState();
+            textureState[j] = new TextureState();
         }
 
         activeShadeModel = 7425;
-        rescaleNormalState = new GlStateManager.BooleanState(32826);
-        colorMaskState = new GlStateManager.ColorMask();
-        colorState = new GlStateManager.Color();
+        rescaleNormalState = new BooleanState(32826);
+        colorMaskState = new ColorMask();
+        colorState = new Color();
     }
 
     @SideOnly(Side.CLIENT)
     static class AlphaState
         {
-            public GlStateManager.BooleanState alphaTest;
+            public BooleanState alphaTest;
             public int func;
             public float ref;
 
             private AlphaState()
             {
-                this.alphaTest = new GlStateManager.BooleanState(3008);
+                this.alphaTest = new BooleanState(3008);
                 this.func = 519;
                 this.ref = -1.0F;
             }
@@ -884,7 +884,7 @@
     @SideOnly(Side.CLIENT)
     static class BlendState
         {
-            public GlStateManager.BooleanState blend;
+            public BooleanState blend;
             public int srcFactor;
             public int dstFactor;
             public int srcFactorAlpha;
@@ -892,7 +892,7 @@
 
             private BlendState()
             {
-                this.blend = new GlStateManager.BooleanState(3042);
+                this.blend = new BooleanState(3042);
                 this.srcFactor = 1;
                 this.dstFactor = 0;
                 this.srcFactorAlpha = 1;
@@ -943,12 +943,12 @@
     static class ClearState
         {
             public double depth;
-            public GlStateManager.Color color;
+            public Color color;
 
             private ClearState()
             {
                 this.depth = 1.0D;
-                this.color = new GlStateManager.Color(0.0F, 0.0F, 0.0F, 0.0F);
+                this.color = new Color(0.0F, 0.0F, 0.0F, 0.0F);
             }
         }
 
@@ -981,12 +981,12 @@
     @SideOnly(Side.CLIENT)
     static class ColorLogicState
         {
-            public GlStateManager.BooleanState colorLogicOp;
+            public BooleanState colorLogicOp;
             public int opcode;
 
             private ColorLogicState()
             {
-                this.colorLogicOp = new GlStateManager.BooleanState(3058);
+                this.colorLogicOp = new BooleanState(3058);
                 this.opcode = 5379;
             }
         }
@@ -1011,13 +1011,13 @@
     @SideOnly(Side.CLIENT)
     static class ColorMaterialState
         {
-            public GlStateManager.BooleanState colorMaterial;
+            public BooleanState colorMaterial;
             public int face;
             public int mode;
 
             private ColorMaterialState()
             {
-                this.colorMaterial = new GlStateManager.BooleanState(2903);
+                this.colorMaterial = new BooleanState(2903);
                 this.face = 1032;
                 this.mode = 5634;
             }
@@ -1041,12 +1041,12 @@
     @SideOnly(Side.CLIENT)
     static class CullState
         {
-            public GlStateManager.BooleanState cullFace;
+            public BooleanState cullFace;
             public int mode;
 
             private CullState()
             {
-                this.cullFace = new GlStateManager.BooleanState(2884);
+                this.cullFace = new BooleanState(2884);
                 this.mode = 1029;
             }
         }
@@ -1054,13 +1054,13 @@
     @SideOnly(Side.CLIENT)
     static class DepthState
         {
-            public GlStateManager.BooleanState depthTest;
+            public BooleanState depthTest;
             public boolean maskEnabled;
             public int depthFunc;
 
             private DepthState()
             {
-                this.depthTest = new GlStateManager.BooleanState(2929);
+                this.depthTest = new BooleanState(2929);
                 this.maskEnabled = true;
                 this.depthFunc = 513;
             }
@@ -1110,7 +1110,7 @@
     @SideOnly(Side.CLIENT)
     static class FogState
         {
-            public GlStateManager.BooleanState fog;
+            public BooleanState fog;
             public int mode;
             public float density;
             public float start;
@@ -1118,7 +1118,7 @@
 
             private FogState()
             {
-                this.fog = new GlStateManager.BooleanState(2912);
+                this.fog = new BooleanState(2912);
                 this.mode = 2048;
                 this.density = 1.0F;
                 this.end = 1.0F;
@@ -1156,15 +1156,15 @@
     @SideOnly(Side.CLIENT)
     static class PolygonOffsetState
         {
-            public GlStateManager.BooleanState polygonOffsetFill;
-            public GlStateManager.BooleanState polygonOffsetLine;
+            public BooleanState polygonOffsetFill;
+            public BooleanState polygonOffsetLine;
             public float factor;
             public float units;
 
             private PolygonOffsetState()
             {
-                this.polygonOffsetFill = new GlStateManager.BooleanState(32823);
-                this.polygonOffsetLine = new GlStateManager.BooleanState(10754);
+                this.polygonOffsetFill = new BooleanState(32823);
+                this.polygonOffsetLine = new BooleanState(10754);
             }
         }
 
@@ -1203,8 +1203,8 @@
                 GlStateManager.depthFunc(513);
                 GlStateManager.depthMask(true);
                 GlStateManager.disableBlend();
-                GlStateManager.blendFunc(GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
-                GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
+                GlStateManager.blendFunc(SourceFactor.ONE, DestFactor.ZERO);
+                GlStateManager.tryBlendFuncSeparate(SourceFactor.ONE, DestFactor.ZERO, SourceFactor.ONE, DestFactor.ZERO);
                 GL14.glBlendEquation(GL14.GL_FUNC_ADD);
                 GlStateManager.disableFog();
                 GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
@@ -1221,22 +1221,22 @@
                 GlStateManager.doPolygonOffset(0.0F, 0.0F);
                 GlStateManager.disableColorLogic();
                 GlStateManager.colorLogicOp(5379);
-                GlStateManager.disableTexGenCoord(GlStateManager.TexGen.S);
-                GlStateManager.texGen(GlStateManager.TexGen.S, 9216);
-                GlStateManager.texGen(GlStateManager.TexGen.S, 9474, RenderHelper.setColorBuffer(1.0F, 0.0F, 0.0F, 0.0F));
-                GlStateManager.texGen(GlStateManager.TexGen.S, 9217, RenderHelper.setColorBuffer(1.0F, 0.0F, 0.0F, 0.0F));
-                GlStateManager.disableTexGenCoord(GlStateManager.TexGen.T);
-                GlStateManager.texGen(GlStateManager.TexGen.T, 9216);
-                GlStateManager.texGen(GlStateManager.TexGen.T, 9474, RenderHelper.setColorBuffer(0.0F, 1.0F, 0.0F, 0.0F));
-                GlStateManager.texGen(GlStateManager.TexGen.T, 9217, RenderHelper.setColorBuffer(0.0F, 1.0F, 0.0F, 0.0F));
-                GlStateManager.disableTexGenCoord(GlStateManager.TexGen.R);
-                GlStateManager.texGen(GlStateManager.TexGen.R, 9216);
-                GlStateManager.texGen(GlStateManager.TexGen.R, 9474, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
-                GlStateManager.texGen(GlStateManager.TexGen.R, 9217, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
-                GlStateManager.disableTexGenCoord(GlStateManager.TexGen.Q);
-                GlStateManager.texGen(GlStateManager.TexGen.Q, 9216);
-                GlStateManager.texGen(GlStateManager.TexGen.Q, 9474, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
-                GlStateManager.texGen(GlStateManager.TexGen.Q, 9217, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.disableTexGenCoord(TexGen.S);
+                GlStateManager.texGen(TexGen.S, 9216);
+                GlStateManager.texGen(TexGen.S, 9474, RenderHelper.setColorBuffer(1.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.texGen(TexGen.S, 9217, RenderHelper.setColorBuffer(1.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.disableTexGenCoord(TexGen.T);
+                GlStateManager.texGen(TexGen.T, 9216);
+                GlStateManager.texGen(TexGen.T, 9474, RenderHelper.setColorBuffer(0.0F, 1.0F, 0.0F, 0.0F));
+                GlStateManager.texGen(TexGen.T, 9217, RenderHelper.setColorBuffer(0.0F, 1.0F, 0.0F, 0.0F));
+                GlStateManager.disableTexGenCoord(TexGen.R);
+                GlStateManager.texGen(TexGen.R, 9216);
+                GlStateManager.texGen(TexGen.R, 9474, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.texGen(TexGen.R, 9217, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.disableTexGenCoord(TexGen.Q);
+                GlStateManager.texGen(TexGen.Q, 9216);
+                GlStateManager.texGen(TexGen.Q, 9474, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
+                GlStateManager.texGen(TexGen.Q, 9217, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 0.0F));
                 GlStateManager.setActiveTexture(0);
                 GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR);
                 GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_NEAREST_MIPMAP_LINEAR);
@@ -1297,7 +1297,7 @@
                 GlStateManager.color(1.0F, 1.0F, 1.0F, 0.15F);
                 GlStateManager.depthMask(false);
                 GlStateManager.enableBlend();
-                GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
+                GlStateManager.blendFunc(SourceFactor.SRC_ALPHA, DestFactor.ONE_MINUS_SRC_ALPHA);
                 GlStateManager.alphaFunc(516, 0.003921569F);
             }
 
@@ -1361,7 +1361,7 @@
     @SideOnly(Side.CLIENT)
     static class StencilState
         {
-            public GlStateManager.StencilFunc func;
+            public StencilFunc func;
             public int mask;
             public int fail;
             public int zfail;
@@ -1369,7 +1369,7 @@
 
             private StencilState()
             {
-                this.func = new GlStateManager.StencilFunc();
+                this.func = new StencilFunc();
                 this.mask = -1;
                 this.fail = 7680;
                 this.zfail = 7680;
@@ -1389,43 +1389,43 @@
     @SideOnly(Side.CLIENT)
     static class TexGenCoord
         {
-            public GlStateManager.BooleanState textureGen;
+            public BooleanState textureGen;
             public int coord;
             public int param = -1;
 
             public TexGenCoord(int coordIn, int capabilityIn)
             {
                 this.coord = coordIn;
-                this.textureGen = new GlStateManager.BooleanState(capabilityIn);
+                this.textureGen = new BooleanState(capabilityIn);
             }
         }
 
     @SideOnly(Side.CLIENT)
     static class TexGenState
         {
-            public GlStateManager.TexGenCoord s;
-            public GlStateManager.TexGenCoord t;
-            public GlStateManager.TexGenCoord r;
-            public GlStateManager.TexGenCoord q;
+            public TexGenCoord s;
+            public TexGenCoord t;
+            public TexGenCoord r;
+            public TexGenCoord q;
 
             private TexGenState()
             {
-                this.s = new GlStateManager.TexGenCoord(8192, 3168);
-                this.t = new GlStateManager.TexGenCoord(8193, 3169);
-                this.r = new GlStateManager.TexGenCoord(8194, 3170);
-                this.q = new GlStateManager.TexGenCoord(8195, 3171);
+                this.s = new TexGenCoord(8192, 3168);
+                this.t = new TexGenCoord(8193, 3169);
+                this.r = new TexGenCoord(8194, 3170);
+                this.q = new TexGenCoord(8195, 3171);
             }
         }
 
     @SideOnly(Side.CLIENT)
     static class TextureState
         {
-            public GlStateManager.BooleanState texture2DState;
+            public BooleanState texture2DState;
             public int textureName;
 
             private TextureState()
             {
-                this.texture2DState = new GlStateManager.BooleanState(3553);
+                this.texture2DState = new BooleanState(3553);
             }
         }
 }
