--- ../src-base/minecraft/net/minecraft/client/renderer/ItemModelMesher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ItemModelMesher.java
@@ -18,7 +18,7 @@
 {
     private final Map<Integer, ModelResourceLocation> simpleShapes = Maps.<Integer, ModelResourceLocation>newHashMap();
     private final Map<Integer, IBakedModel> simpleShapesCache = Maps.<Integer, IBakedModel>newHashMap();
-    protected final Map<Item, ItemMeshDefinition> shapers = Maps.<Item, ItemMeshDefinition>newHashMap();
+    private final Map<Item, ItemMeshDefinition> shapers = Maps.<Item, ItemMeshDefinition>newHashMap();
     private final ModelManager modelManager;
 
     public ItemModelMesher(ModelManager modelManager)
@@ -33,7 +33,9 @@
 
     public TextureAtlasSprite getParticleIcon(Item item, int meta)
     {
-        return this.getItemModel(new ItemStack(item, 1, meta)).getParticleTexture();
+        ItemStack stack = new ItemStack(item, 1, meta);
+        IBakedModel model = this.getItemModel(stack);
+        return model.getOverrides().handleItemState(model, stack, null, null).getParticleTexture();
     }
 
     public IBakedModel getItemModel(ItemStack stack)
