--- ../src-base/minecraft/net/minecraft/client/renderer/OpenGlHelper.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/OpenGlHelper.java
@@ -49,7 +49,7 @@
     public static int GL_FB_INCOMPLETE_MISS_ATTACH;
     public static int GL_FB_INCOMPLETE_DRAW_BUFFER;
     public static int GL_FB_INCOMPLETE_READ_BUFFER;
-    private static OpenGlHelper.FboMode framebufferType;
+    private static FboMode framebufferType;
     public static boolean framebufferSupported;
     private static boolean shadersAvailable;
     private static boolean arbShaders;
@@ -93,6 +93,10 @@
     public static int GL_ARRAY_BUFFER;
     public static int GL_STATIC_DRAW;
 
+    /* Stores the last values sent into setLightmapTextureCoords */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void initializeTextures()
     {
         ContextCapabilities contextcapabilities = GLContext.getCapabilities();
@@ -172,7 +176,7 @@
             if (contextcapabilities.OpenGL30)
             {
                 logText = logText + "OpenGL 3.0 is supported and separate blending is supported.\n";
-                framebufferType = OpenGlHelper.FboMode.BASE;
+                framebufferType = FboMode.BASE;
                 GL_FRAMEBUFFER = 36160;
                 GL_RENDERBUFFER = 36161;
                 GL_COLOR_ATTACHMENT0 = 36064;
@@ -186,7 +190,7 @@
             else if (contextcapabilities.GL_ARB_framebuffer_object)
             {
                 logText = logText + "ARB_framebuffer_object is supported and separate blending is supported.\n";
-                framebufferType = OpenGlHelper.FboMode.ARB;
+                framebufferType = FboMode.ARB;
                 GL_FRAMEBUFFER = 36160;
                 GL_RENDERBUFFER = 36161;
                 GL_COLOR_ATTACHMENT0 = 36064;
@@ -200,7 +204,7 @@
             else if (contextcapabilities.GL_EXT_framebuffer_object)
             {
                 logText = logText + "EXT_framebuffer_object is supported.\n";
-                framebufferType = OpenGlHelper.FboMode.EXT;
+                framebufferType = FboMode.EXT;
                 GL_FRAMEBUFFER = 36160;
                 GL_RENDERBUFFER = 36161;
                 GL_COLOR_ATTACHMENT0 = 36064;
@@ -853,6 +857,12 @@
         {
             GL13.glMultiTexCoord2f(target, p_77475_1_, t);
         }
+
+        if (target == lightmapTexUnit)
+        {
+            lastBrightnessX = p_77475_1_;
+            lastBrightnessY = t;
+        }
     }
 
     public static void glBlendFunc(int sFactorRGB, int dFactorRGB, int sfactorAlpha, int dfactorAlpha)
