--- ../src-base/minecraft/net/minecraft/client/renderer/RenderGlobal.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderGlobal.java
@@ -948,7 +948,7 @@
                 }
                 else
                 {
-                    this.mc.mcProfiler.startSection("build near");
+                    this.mc.mcProfiler.startSection("build1 near");
                     this.renderDispatcher.updateChunkNow(renderchunk4);
                     renderchunk4.clearNeedsUpdate();
                     this.mc.mcProfiler.endSection();
