--- ../src-base/minecraft/net/minecraft/client/renderer/WorldVertexBufferUploader.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/WorldVertexBufferUploader.java
@@ -27,26 +27,8 @@
                 int l = vertexformatelement.getIndex();
                 bytebuffer.position(vertexformat.getOffset(j));
 
-                switch (vertexformatelement$enumusage)
-                {
-                    case POSITION:
-                        GlStateManager.glVertexPointer(vertexformatelement.getElementCount(), k, i, bytebuffer);
-                        GlStateManager.glEnableClientState(32884);
-                        break;
-                    case UV:
-                        OpenGlHelper.setClientActiveTexture(OpenGlHelper.defaultTexUnit + l);
-                        GlStateManager.glTexCoordPointer(vertexformatelement.getElementCount(), k, i, bytebuffer);
-                        GlStateManager.glEnableClientState(32888);
-                        OpenGlHelper.setClientActiveTexture(OpenGlHelper.defaultTexUnit);
-                        break;
-                    case COLOR:
-                        GlStateManager.glColorPointer(vertexformatelement.getElementCount(), k, i, bytebuffer);
-                        GlStateManager.glEnableClientState(32886);
-                        break;
-                    case NORMAL:
-                        GlStateManager.glNormalPointer(k, i, bytebuffer);
-                        GlStateManager.glEnableClientState(32885);
-                }
+                // moved to VertexFormatElement.preDraw
+                vertexformatelement.getUsage().preDraw(vertexformat, j, i, bytebuffer);
             }
 
             GlStateManager.glDrawArrays(bufferBuilderIn.getDrawMode(), 0, bufferBuilderIn.getVertexCount());
@@ -58,23 +40,8 @@
                 VertexFormatElement.EnumUsage vertexformatelement$enumusage1 = vertexformatelement1.getUsage();
                 int k1 = vertexformatelement1.getIndex();
 
-                switch (vertexformatelement$enumusage1)
-                {
-                    case POSITION:
-                        GlStateManager.glDisableClientState(32884);
-                        break;
-                    case UV:
-                        OpenGlHelper.setClientActiveTexture(OpenGlHelper.defaultTexUnit + k1);
-                        GlStateManager.glDisableClientState(32888);
-                        OpenGlHelper.setClientActiveTexture(OpenGlHelper.defaultTexUnit);
-                        break;
-                    case COLOR:
-                        GlStateManager.glDisableClientState(32886);
-                        GlStateManager.resetColor();
-                        break;
-                    case NORMAL:
-                        GlStateManager.glDisableClientState(32885);
-                }
+                // moved to VertexFormatElement.postDraw
+                vertexformatelement1.getUsage().postDraw(vertexformat, i1, i, bytebuffer);
             }
         }
 
