--- ../src-base/minecraft/net/minecraft/world/BossInfoServer.java
+++ ../src-work/minecraft/net/minecraft/world/BossInfoServer.java
@@ -14,7 +14,7 @@
 {
     private final Set<EntityPlayerMP> players = Sets.<EntityPlayerMP>newHashSet();
     private final Set<EntityPlayerMP> readOnlyPlayers;
-    private boolean visible;
+    public boolean visible;
 
     public BossInfoServer(ITextComponent nameIn, BossInfo.Color colorIn, BossInfo.Overlay overlayIn)
     {
@@ -92,7 +92,7 @@
         }
     }
 
-    private void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
+    public void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
     {
         if (this.visible)
         {
