--- ../src-base/minecraft/net/minecraft/world/BossInfoServer.java
+++ ../src-work/minecraft/net/minecraft/world/BossInfoServer.java
@@ -9,16 +9,14 @@
 import net.minecraft.network.play.server.SPacketUpdateBossInfo;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class BossInfoServer extends BossInfo
 {
     private final Set<EntityPlayerMP> players = Sets.<EntityPlayerMP>newHashSet();
     private final Set<EntityPlayerMP> readOnlyPlayers;
-    private boolean visible;
+    public boolean visible;
 
-    public BossInfoServer(ITextComponent nameIn, BossInfo.Color colorIn, BossInfo.Overlay overlayIn)
+    public BossInfoServer(ITextComponent nameIn, Color colorIn, Overlay overlayIn)
     {
         super(MathHelper.getRandomUUID(), nameIn, colorIn, overlayIn);
         this.readOnlyPlayers = Collections.<EntityPlayerMP>unmodifiableSet(this.players);
@@ -34,8 +32,7 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    public void setColor(BossInfo.Color colorIn)
+    public void setColor(Color colorIn)
     {
         if (colorIn != this.color)
         {
@@ -44,8 +41,7 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    public void setOverlay(BossInfo.Overlay overlayIn)
+    public void setOverlay(Overlay overlayIn)
     {
         if (overlayIn != this.overlay)
         {
@@ -96,7 +92,7 @@
         }
     }
 
-    private void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
+    public void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
     {
         if (this.visible)
         {
