--- ../src-base/minecraft/net/minecraft/world/BossInfo.java
+++ ../src-work/minecraft/net/minecraft/world/BossInfo.java
@@ -6,10 +6,10 @@
 public abstract class BossInfo
 {
     private final UUID uniqueId;
-    protected ITextComponent name;
+    public ITextComponent name;
     protected float percent;
-    protected BossInfo.Color color;
-    protected BossInfo.Overlay overlay;
+    public BossInfo.Color color;
+    public BossInfo.Overlay overlay;
     protected boolean darkenSky;
     protected boolean playEndBossMusic;
     protected boolean createFog;
