--- ../src-base/minecraft/net/minecraft/world/BossInfo.java
+++ ../src-work/minecraft/net/minecraft/world/BossInfo.java
@@ -2,21 +2,19 @@
 
 import java.util.UUID;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public abstract class BossInfo
 {
     private final UUID uniqueId;
-    protected ITextComponent name;
-    protected float percent;
-    protected BossInfo.Color color;
-    protected BossInfo.Overlay overlay;
+    public ITextComponent name;
+    public float percent;
+    public Color color;
+    public Overlay overlay;
     protected boolean darkenSky;
     protected boolean playEndBossMusic;
     protected boolean createFog;
 
-    public BossInfo(UUID uniqueIdIn, ITextComponent nameIn, BossInfo.Color colorIn, BossInfo.Overlay overlayIn)
+    public BossInfo(UUID uniqueIdIn, ITextComponent nameIn, Color colorIn, Overlay overlayIn)
     {
         this.uniqueId = uniqueIdIn;
         this.name = nameIn;
@@ -50,24 +48,22 @@
         this.percent = percentIn;
     }
 
-    public BossInfo.Color getColor()
+    public Color getColor()
     {
         return this.color;
     }
 
-    @SideOnly(Side.CLIENT)
-    public void setColor(BossInfo.Color colorIn)
+    public void setColor(Color colorIn)
     {
         this.color = colorIn;
     }
 
-    public BossInfo.Overlay getOverlay()
+    public Overlay getOverlay()
     {
         return this.overlay;
     }
 
-    @SideOnly(Side.CLIENT)
-    public void setOverlay(BossInfo.Overlay overlayIn)
+    public void setOverlay(Overlay overlayIn)
     {
         this.overlay = overlayIn;
     }
