--- ../src-base/minecraft/net/minecraft/world/Explosion.java
+++ ../src-work/minecraft/net/minecraft/world/Explosion.java
@@ -3,18 +3,22 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+
 import java.util.List;
 import java.util.Map;
 import java.util.Random;
 import java.util.Set;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.enchantment.EnchantmentProtection;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.item.EntityFallingBlock;
 import net.minecraft.entity.item.EntityTNTPrimed;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.projectile.EntityFireball;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.util.DamageSource;
@@ -26,9 +30,19 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockExplodeEvent;
+import org.bukkit.event.entity.EntityExplodeEvent;
 
-public class Explosion
-{
+//CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityExplodeEvent;
+import org.bukkit.Location;
+import org.bukkit.event.block.BlockExplodeEvent;
+// CraftBukkit end
+
+public class Explosion {
     private final boolean isFlaming;
     private final boolean isSmoking;
     private final Random explosionRNG;
@@ -36,33 +50,31 @@
     private final double explosionX;
     private final double explosionY;
     private final double explosionZ;
-    private final Entity exploder;
+    public final Entity exploder;
     private final float explosionSize;
     private final List<BlockPos> affectedBlockPositions;
     private final Map<EntityPlayer, Vec3d> playerKnockbackMap;
     private final Vec3d position;
+    public boolean wasCanceled = false; // CraftBukkit - add field
 
     @SideOnly(Side.CLIENT)
-    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, List<BlockPos> affectedPositions)
-    {
+    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, List<BlockPos> affectedPositions) {
         this(worldIn, entityIn, x, y, z, size, false, true, affectedPositions);
     }
 
     @SideOnly(Side.CLIENT)
-    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, boolean flaming, boolean smoking, List<BlockPos> affectedPositions)
-    {
+    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, boolean flaming, boolean smoking, List<BlockPos> affectedPositions) {
         this(worldIn, entityIn, x, y, z, size, flaming, smoking);
         this.affectedBlockPositions.addAll(affectedPositions);
     }
 
-    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, boolean flaming, boolean smoking)
-    {
+    public Explosion(World worldIn, Entity entityIn, double x, double y, double z, float size, boolean flaming, boolean smoking) {
         this.explosionRNG = new Random();
         this.affectedBlockPositions = Lists.<BlockPos>newArrayList();
         this.playerKnockbackMap = Maps.<EntityPlayer, Vec3d>newHashMap();
         this.worldObj = worldIn;
         this.exploder = entityIn;
-        this.explosionSize = size;
+        this.explosionSize = (float) Math.max(size, 0.0); // CraftBukkit - clamp bad values
         this.explosionX = x;
         this.explosionY = y;
         this.explosionZ = z;
@@ -71,22 +83,22 @@
         this.position = new Vec3d(explosionX, explosionY, explosionZ);
     }
 
-    public void doExplosionA()
-    {
+    public void doExplosionA() {
+        // CraftBukkit start
+        if (this.explosionSize < 0.1F) {
+            return;
+        }
+        // CraftBukkit end
         Set<BlockPos> set = Sets.<BlockPos>newHashSet();
         int i = 16;
 
-        for (int j = 0; j < 16; ++j)
-        {
-            for (int k = 0; k < 16; ++k)
-            {
-                for (int l = 0; l < 16; ++l)
-                {
-                    if (j == 0 || j == 15 || k == 0 || k == 15 || l == 0 || l == 15)
-                    {
-                        double d0 = (double)((float)j / 15.0F * 2.0F - 1.0F);
-                        double d1 = (double)((float)k / 15.0F * 2.0F - 1.0F);
-                        double d2 = (double)((float)l / 15.0F * 2.0F - 1.0F);
+        for (int j = 0; j < 16; ++j) {
+            for (int k = 0; k < 16; ++k) {
+                for (int l = 0; l < 16; ++l) {
+                    if (j == 0 || j == 15 || k == 0 || k == 15 || l == 0 || l == 15) {
+                        double d0 = (double) ((float) j / 15.0F * 2.0F - 1.0F);
+                        double d1 = (double) ((float) k / 15.0F * 2.0F - 1.0F);
+                        double d2 = (double) ((float) l / 15.0F * 2.0F - 1.0F);
                         double d3 = Math.sqrt(d0 * d0 + d1 * d1 + d2 * d2);
                         d0 = d0 / d3;
                         d1 = d1 / d3;
@@ -96,19 +108,16 @@
                         double d6 = this.explosionY;
                         double d8 = this.explosionZ;
 
-                        for (float f1 = 0.3F; f > 0.0F; f -= 0.22500001F)
-                        {
+                        for (float f1 = 0.3F; f > 0.0F; f -= 0.22500001F) {
                             BlockPos blockpos = new BlockPos(d4, d6, d8);
                             IBlockState iblockstate = this.worldObj.getBlockState(blockpos);
 
-                            if (iblockstate.getMaterial() != Material.AIR)
-                            {
-                                float f2 = this.exploder != null ? this.exploder.getExplosionResistance(this, this.worldObj, blockpos, iblockstate) : iblockstate.getBlock().getExplosionResistance(worldObj, blockpos, (Entity)null, this);
+                            if (iblockstate.getMaterial() != Material.AIR) {
+                                float f2 = this.exploder != null ? this.exploder.getExplosionResistance(this, this.worldObj, blockpos, iblockstate) : iblockstate.getBlock().getExplosionResistance(worldObj, blockpos, (Entity) null, this);
                                 f -= (f2 + 0.3F) * 0.3F;
                             }
 
-                            if (f > 0.0F && (this.exploder == null || this.exploder.verifyExplosion(this, this.worldObj, blockpos, iblockstate, f)))
-                            {
+                            if (f > 0.0F && (this.exploder == null || this.exploder.verifyExplosion(this, this.worldObj, blockpos, iblockstate, f)) && blockpos.getY() < 256 && blockpos.getY() >= 0) {
                                 set.add(blockpos);
                             }
 
@@ -123,56 +132,57 @@
 
         this.affectedBlockPositions.addAll(set);
         float f3 = this.explosionSize * 2.0F;
-        int k1 = MathHelper.floor_double(this.explosionX - (double)f3 - 1.0D);
-        int l1 = MathHelper.floor_double(this.explosionX + (double)f3 + 1.0D);
-        int i2 = MathHelper.floor_double(this.explosionY - (double)f3 - 1.0D);
-        int i1 = MathHelper.floor_double(this.explosionY + (double)f3 + 1.0D);
-        int j2 = MathHelper.floor_double(this.explosionZ - (double)f3 - 1.0D);
-        int j1 = MathHelper.floor_double(this.explosionZ + (double)f3 + 1.0D);
-        List<Entity> list = this.worldObj.getEntitiesWithinAABBExcludingEntity(this.exploder, new AxisAlignedBB((double)k1, (double)i2, (double)j2, (double)l1, (double)i1, (double)j1));
+        int k1 = MathHelper.floor_double(this.explosionX - (double) f3 - 1.0D);
+        int l1 = MathHelper.floor_double(this.explosionX + (double) f3 + 1.0D);
+        int i2 = MathHelper.floor_double(this.explosionY - (double) f3 - 1.0D);
+        int i1 = MathHelper.floor_double(this.explosionY + (double) f3 + 1.0D);
+        int j2 = MathHelper.floor_double(this.explosionZ - (double) f3 - 1.0D);
+        int j1 = MathHelper.floor_double(this.explosionZ + (double) f3 + 1.0D);
+        List<Entity> list = this.worldObj.getEntitiesWithinAABBExcludingEntity(this.exploder, new AxisAlignedBB((double) k1, (double) i2, (double) j2, (double) l1, (double) i1, (double) j1));
         net.minecraftforge.event.ForgeEventFactory.onExplosionDetonate(this.worldObj, this, list, f3);
         Vec3d vec3d = new Vec3d(this.explosionX, this.explosionY, this.explosionZ);
 
-        for (int k2 = 0; k2 < list.size(); ++k2)
-        {
-            Entity entity = (Entity)list.get(k2);
+        for (int k2 = 0; k2 < list.size(); ++k2) {
+            Entity entity = (Entity) list.get(k2);
 
-            if (!entity.isImmuneToExplosions())
-            {
-                double d12 = entity.getDistance(this.explosionX, this.explosionY, this.explosionZ) / (double)f3;
+            if (!entity.isImmuneToExplosions()) {
+                double d12 = entity.getDistance(this.explosionX, this.explosionY, this.explosionZ) / (double) f3;
 
-                if (d12 <= 1.0D)
-                {
+                if (d12 <= 1.0D) {
                     double d5 = entity.posX - this.explosionX;
-                    double d7 = entity.posY + (double)entity.getEyeHeight() - this.explosionY;
+                    double d7 = entity.posY + (double) entity.getEyeHeight() - this.explosionY;
                     double d9 = entity.posZ - this.explosionZ;
-                    double d13 = (double)MathHelper.sqrt_double(d5 * d5 + d7 * d7 + d9 * d9);
+                    double d13 = (double) MathHelper.sqrt_double(d5 * d5 + d7 * d7 + d9 * d9);
 
-                    if (d13 != 0.0D)
-                    {
+                    if (d13 != 0.0D) {
                         d5 = d5 / d13;
                         d7 = d7 / d13;
                         d9 = d9 / d13;
-                        double d14 = (double)this.worldObj.getBlockDensity(vec3d, entity.getEntityBoundingBox());
+                        double d14 = (double) this.worldObj.getBlockDensity(vec3d, entity.getEntityBoundingBox());
                         double d10 = (1.0D - d12) * d14;
-                        entity.attackEntityFrom(DamageSource.causeExplosionDamage(this), (float)((int)((d10 * d10 + d10) / 2.0D * 7.0D * (double)f3 + 1.0D)));
+                        // CraftBukkit start
+                        CraftEventFactory.entityDamage = exploder;
+                        entity.forceExplosionKnockback = false;
+                        boolean wasDamaged = entity.attackEntityFrom(DamageSource.causeExplosionDamage(this), (float) ((int) ((d13 * d13 + d13) / 2.0D * 7.0D * (double) f3 + 1.0D)));
+                        CraftEventFactory.entityDamage = null;
+                        if (!wasDamaged && !(entity instanceof EntityTNTPrimed || entity instanceof EntityFallingBlock) && !entity.forceExplosionKnockback) {
+                            continue;
+                        }
+                        // CraftBukkit end
                         double d11 = 1.0D;
 
-                        if (entity instanceof EntityLivingBase)
-                        {
-                            d11 = EnchantmentProtection.getBlastDamageReduction((EntityLivingBase)entity, d10);
+                        if (entity instanceof EntityLivingBase) {
+                            d11 = EnchantmentProtection.getBlastDamageReduction((EntityLivingBase) entity, d10);
                         }
 
                         entity.motionX += d5 * d11;
                         entity.motionY += d7 * d11;
                         entity.motionZ += d9 * d11;
 
-                        if (entity instanceof EntityPlayer)
-                        {
-                            EntityPlayer entityplayer = (EntityPlayer)entity;
+                        if (entity instanceof EntityPlayer) {
+                            EntityPlayer entityplayer = (EntityPlayer) entity;
 
-                            if (!entityplayer.isSpectator() && (!entityplayer.isCreative() || !entityplayer.capabilities.isFlying))
-                            {
+                            if (!entityplayer.isSpectator() && (!entityplayer.isCreative() || !entityplayer.capabilities.isFlying)) {
                                 this.playerKnockbackMap.put(entityplayer, new Vec3d(d5 * d10, d7 * d10, d9 * d10));
                             }
                         }
@@ -182,40 +192,71 @@
         }
     }
 
-    public void doExplosionB(boolean spawnParticles)
-    {
-        this.worldObj.playSound((EntityPlayer)null, this.explosionX, this.explosionY, this.explosionZ, SoundEvents.ENTITY_GENERIC_EXPLODE, SoundCategory.BLOCKS, 4.0F, (1.0F + (this.worldObj.rand.nextFloat() - this.worldObj.rand.nextFloat()) * 0.2F) * 0.7F);
+    public void doExplosionB(boolean spawnParticles) {
+        this.worldObj.playSound((EntityPlayer) null, this.explosionX, this.explosionY, this.explosionZ, SoundEvents.ENTITY_GENERIC_EXPLODE, SoundCategory.BLOCKS, 4.0F, (1.0F + (this.worldObj.rand.nextFloat() - this.worldObj.rand.nextFloat()) * 0.2F) * 0.7F);
 
-        if (this.explosionSize >= 2.0F && this.isSmoking)
-        {
+        if (this.explosionSize >= 2.0F && this.isSmoking) {
             this.worldObj.spawnParticle(EnumParticleTypes.EXPLOSION_HUGE, this.explosionX, this.explosionY, this.explosionZ, 1.0D, 0.0D, 0.0D, new int[0]);
-        }
-        else
-        {
+        } else {
             this.worldObj.spawnParticle(EnumParticleTypes.EXPLOSION_LARGE, this.explosionX, this.explosionY, this.explosionZ, 1.0D, 0.0D, 0.0D, new int[0]);
         }
 
-        if (this.isSmoking)
-        {
-            for (BlockPos blockpos : this.affectedBlockPositions)
-            {
+        if (this.isSmoking) {
+            // CraftBukkit start
+            final org.bukkit.World bworld = this.worldObj.getWorld();
+            final org.bukkit.entity.Entity explode = (this.exploder == null) ? null : this.exploder.getBukkitEntity();
+            final Location location = new Location(bworld, this.explosionX, this.explosionY, this.explosionZ);
+            final List<org.bukkit.block.Block> blockList = /*(List<Block>)*/Lists.newArrayList();
+            for (int i1 = this.affectedBlockPositions.size() - 1; i1 >= 0; --i1) {
+                final BlockPos cpos = this.affectedBlockPositions.get(i1);
+                final org.bukkit.block.Block bblock = bworld.getBlockAt(cpos.getX(), cpos.getY(), cpos.getZ());
+                if (bblock.getType() != org.bukkit.Material.AIR) {
+                    blockList.add(bblock);
+                }
+            }
+            boolean cancelled;
+            List<org.bukkit.block.Block> bukkitBlocks;
+            float yield;
+            if (explode != null) {
+                final EntityExplodeEvent event = new EntityExplodeEvent(explode, location, blockList, 1.0f / this.explosionSize);
+                this.worldObj.getServer().getPluginManager().callEvent(event);
+                cancelled = event.isCancelled();
+                bukkitBlocks = event.blockList();
+                yield = event.getYield();
+            } else {
+                final BlockExplodeEvent event2 = new BlockExplodeEvent(location.getBlock(), blockList, 1.0f / this.explosionSize);
+                this.worldObj.getServer().getPluginManager().callEvent(event2);
+                cancelled = event2.isCancelled();
+                bukkitBlocks = event2.blockList();
+                yield = event2.getYield();
+            }
+            this.affectedBlockPositions.clear();
+            for (final org.bukkit.block.Block bblock2 : bukkitBlocks) {
+                final BlockPos coords = new BlockPos(bblock2.getX(), bblock2.getY(), bblock2.getZ());
+                this.affectedBlockPositions.add(coords);
+            }
+            if (cancelled) {
+                this.wasCanceled = true;
+                return;
+            }
+            // CraftBukkit end
+            for (BlockPos blockpos : this.affectedBlockPositions) {
                 IBlockState iblockstate = this.worldObj.getBlockState(blockpos);
                 Block block = iblockstate.getBlock();
 
-                if (spawnParticles)
-                {
-                    double d0 = (double)((float)blockpos.getX() + this.worldObj.rand.nextFloat());
-                    double d1 = (double)((float)blockpos.getY() + this.worldObj.rand.nextFloat());
-                    double d2 = (double)((float)blockpos.getZ() + this.worldObj.rand.nextFloat());
+                if (spawnParticles) {
+                    double d0 = (double) ((float) blockpos.getX() + this.worldObj.rand.nextFloat());
+                    double d1 = (double) ((float) blockpos.getY() + this.worldObj.rand.nextFloat());
+                    double d2 = (double) ((float) blockpos.getZ() + this.worldObj.rand.nextFloat());
                     double d3 = d0 - this.explosionX;
                     double d4 = d1 - this.explosionY;
                     double d5 = d2 - this.explosionZ;
-                    double d6 = (double)MathHelper.sqrt_double(d3 * d3 + d4 * d4 + d5 * d5);
+                    double d6 = (double) MathHelper.sqrt_double(d3 * d3 + d4 * d4 + d5 * d5);
                     d3 = d3 / d6;
                     d4 = d4 / d6;
                     d5 = d5 / d6;
-                    double d7 = 0.5D / (d6 / (double)this.explosionSize + 0.1D);
-                    d7 = d7 * (double)(this.worldObj.rand.nextFloat() * this.worldObj.rand.nextFloat() + 0.3F);
+                    double d7 = 0.5D / (d6 / (double) this.explosionSize + 0.1D);
+                    d7 = d7 * (double) (this.worldObj.rand.nextFloat() * this.worldObj.rand.nextFloat() + 0.3F);
                     d3 = d3 * d7;
                     d4 = d4 * d7;
                     d5 = d5 * d7;
@@ -223,11 +264,10 @@
                     this.worldObj.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, d0, d1, d2, d3, d4, d5, new int[0]);
                 }
 
-                if (iblockstate.getMaterial() != Material.AIR)
-                {
-                    if (block.canDropFromExplosion(this))
-                    {
-                        block.dropBlockAsItemWithChance(this.worldObj, blockpos, this.worldObj.getBlockState(blockpos), 1.0F / this.explosionSize, 0);
+                if (iblockstate.getMaterial() != Material.AIR) {
+                    if (block.canDropFromExplosion(this)) {
+                        // CraftBukkit - add yield
+                        block.dropBlockAsItemWithChance(this.worldObj, blockpos, this.worldObj.getBlockState(blockpos), yield, 0);
                     }
 
                     block.onBlockExploded(this.worldObj, blockpos, this);
@@ -235,37 +275,34 @@
             }
         }
 
-        if (this.isFlaming)
-        {
-            for (BlockPos blockpos1 : this.affectedBlockPositions)
-            {
-                if (this.worldObj.getBlockState(blockpos1).getMaterial() == Material.AIR && this.worldObj.getBlockState(blockpos1.down()).isFullBlock() && this.explosionRNG.nextInt(3) == 0)
-                {
-                    this.worldObj.setBlockState(blockpos1, Blocks.FIRE.getDefaultState());
+        if (this.isFlaming) {
+            for (BlockPos blockpos1 : this.affectedBlockPositions) {
+                if (this.worldObj.getBlockState(blockpos1).getMaterial() == Material.AIR && this.worldObj.getBlockState(blockpos1.down()).isFullBlock() && this.explosionRNG.nextInt(3) == 0) {
+                    if (!CraftEventFactory.callBlockIgniteEvent(this.worldObj, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ(), this).isCancelled()) {
+                        this.worldObj.setBlockState(blockpos1, Blocks.FIRE.getDefaultState());
+                    }
                 }
             }
         }
     }
 
-    public Map<EntityPlayer, Vec3d> getPlayerKnockbackMap()
-    {
+    public Map<EntityPlayer, Vec3d> getPlayerKnockbackMap() {
         return this.playerKnockbackMap;
     }
 
-    public EntityLivingBase getExplosivePlacedBy()
-    {
-        return this.exploder == null ? null : (this.exploder instanceof EntityTNTPrimed ? ((EntityTNTPrimed)this.exploder).getTntPlacedBy() : (this.exploder instanceof EntityLivingBase ? (EntityLivingBase)this.exploder : null));
+    public EntityLivingBase getExplosivePlacedBy() {
+        return this.exploder == null ? null : (this.exploder instanceof EntityTNTPrimed ? ((EntityTNTPrimed) this.exploder).getTntPlacedBy() : (this.exploder instanceof EntityLivingBase ? (EntityLivingBase) this.exploder : (this.exploder instanceof EntityFireball ? ((EntityFireball) this.exploder).shootingEntity : null)));
     }
 
-    public void clearAffectedBlockPositions()
-    {
+    public void clearAffectedBlockPositions() {
         this.affectedBlockPositions.clear();
     }
 
-    public List<BlockPos> getAffectedBlockPositions()
-    {
+    public List<BlockPos> getAffectedBlockPositions() {
         return this.affectedBlockPositions;
     }
 
-    public Vec3d getPosition(){ return this.position; }
+    public Vec3d getPosition() {
+        return this.position;
+    }
 }
