--- ../src-base/minecraft/net/minecraft/world/WorldProvider.java
+++ ../src-work/minecraft/net/minecraft/world/WorldProvider.java
@@ -16,6 +16,7 @@
 import net.minecraft.world.gen.ChunkProviderFlat;
 import net.minecraft.world.gen.ChunkProviderOverworld;
 import net.minecraft.world.gen.FlatGeneratorInfo;
+import net.minecraftforge.common.DimensionManager;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
@@ -141,6 +142,10 @@
         return true;
     }
 
+    public static WorldProvider getProviderForDimension(int dimensionId){
+        return DimensionManager.createProviderFor(dimensionId);
+    }
+
     @SideOnly(Side.CLIENT)
     public float getCloudHeight()
     {
