--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -26,13 +_,20 @@
 import org.apache.commons.lang3.Validate;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.tags.DamageTypeTags;
+import org.bukkit.entity.Hanging;
+import org.bukkit.event.hanging.HangingBreakByEntityEvent;
+import org.bukkit.event.hanging.HangingBreakEvent;
+// CraftBukkit end
+
 public abstract class HangingEntity extends Entity {
    private static final Logger f_238173_ = LogUtils.getLogger();
    protected static final Predicate<Entity> f_31697_ = (p_31734_) -> {
       return p_31734_ instanceof HangingEntity;
    };
    private int f_31700_;
-   protected BlockPos f_31698_;
+   public BlockPos f_31698_; // CatServer - protected -> public
    protected Direction f_31699_ = Direction.SOUTH;
 
    protected HangingEntity(EntityType<? extends HangingEntity> p_31703_, Level p_31704_) {
@@ -47,7 +_,7 @@
    protected void m_8097_() {
    }
 
-   protected void m_6022_(Direction p_31728_) {
+   public void m_6022_(Direction p_31728_) { // CatServer - protected -> public
       Validate.notNull(p_31728_);
       Validate.isTrue(p_31728_.m_122434_().m_122479_());
       this.f_31699_ = p_31728_;
@@ -58,45 +_,71 @@
 
    protected void m_7087_() {
       if (this.f_31699_ != null) {
-         double d0 = (double)this.f_31698_.m_123341_() + 0.5D;
-         double d1 = (double)this.f_31698_.m_123342_() + 0.5D;
-         double d2 = (double)this.f_31698_.m_123343_() + 0.5D;
-         double d3 = 0.46875D;
-         double d4 = this.m_31709_(this.m_7076_());
-         double d5 = this.m_31709_(this.m_7068_());
-         d0 -= (double)this.f_31699_.m_122429_() * 0.46875D;
-         d2 -= (double)this.f_31699_.m_122431_() * 0.46875D;
-         d1 += d5;
-         Direction direction = this.f_31699_.m_122428_();
-         d0 += d4 * (double)direction.m_122429_();
-         d2 += d4 * (double)direction.m_122431_();
-         this.m_20343_(d0, d1, d2);
-         double d6 = (double)this.m_7076_();
-         double d7 = (double)this.m_7068_();
-         double d8 = (double)this.m_7076_();
-         if (this.f_31699_.m_122434_() == Direction.Axis.Z) {
-            d8 = 1.0D;
-         } else {
-            d6 = 1.0D;
-         }
-
-         d6 /= 32.0D;
-         d7 /= 32.0D;
-         d8 /= 32.0D;
-         this.m_20011_(new AABB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8));
-      }
-   }
-
-   private double m_31709_(int p_31710_) {
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.m_20011_(calculateBoundingBox(this, this.f_31698_, this.f_31699_, this.m_7076_(), this.m_7068_()));
+         // CraftBukkit end
+      }
+   }
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AABB calculateBoundingBox(Entity entity, BlockPos pos, Direction opposite, int width, int height) {
+      double d0 = (double) pos.m_123341_() + 0.5D;
+      double d1 = (double) pos.m_123342_() + 0.5D;
+      double d2 = (double) pos.m_123343_() + 0.5D;
+      double d3 = 0.46875D;
+      double d4 = m_31709_(width);
+      double d5 = m_31709_(height);
+      d0 -= (double) opposite.m_122429_() * 0.46875D;
+      d2 -= (double) opposite.m_122431_() * 0.46875D;
+      d1 += d5;
+      Direction direction = opposite.m_122428_();
+      d0 += d4 * (double) direction.m_122429_();
+      d2 += d4 * (double) direction.m_122431_();
+      if (entity != null) {
+         entity.m_20343_(d0, d1, d2);
+      }
+      double d6 = width;
+      double d7 = height;
+      double d8 = width;
+      if (opposite.m_122434_() == Direction.Axis.Z) {
+         d8 = 1.0D;
+      } else {
+         d6 = 1.0D;
+      }
+
+      d6 /= 32.0D;
+      d7 /= 32.0D;
+      d8 /= 32.0D;
+
+      return new AABB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8);
+   }
+   // CraftBukkit end
+
+   private static double m_31709_(int p_31710_) { // CraftBukkit - static
       return p_31710_ % 32 == 0 ? 0.5D : 0.0D;
    }
 
    public void m_8119_() {
       if (!this.m_9236_().f_46443_) {
          this.m_146871_();
-         if (this.f_31700_++ == 100) {
+         if (this.f_31700_++ == this.m_9236_().spigotConfig.hangingTickFrequency) { // Spigot
             this.f_31700_ = 0;
             if (!this.m_213877_() && !this.m_7088_()) {
+               // CraftBukkit start - fire break events
+               BlockState material = this.m_9236_().m_8055_(this.m_20183_());
+               HangingBreakEvent.RemoveCause cause;
+               if (!material.m_60795_()) {
+                  // TODO: This feels insufficient to catch 100% of suffocation cases
+                  cause = HangingBreakEvent.RemoveCause.OBSTRUCTION;
+               } else {
+                  cause = HangingBreakEvent.RemoveCause.PHYSICS;
+               }
+               HangingBreakEvent event = new HangingBreakEvent((Hanging) this.getBukkitEntity(), cause);
+               this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+               if (this.m_213877_() || event.isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                this.m_146870_();
                this.m_5553_((Entity)null);
             }
@@ -121,6 +_,8 @@
                int j1 = (j - 1) / -2;
                blockpos$mutableblockpos.m_122190_(blockpos).m_122175_(direction, k + i1).m_122175_(Direction.UP, l + j1);
                BlockState blockstate = this.m_9236_().m_8055_(blockpos$mutableblockpos);
+               if (net.minecraft.world.level.block.Block.m_49863_(this.m_9236_(), blockpos$mutableblockpos, this.f_31699_))
+                  continue;
                if (!blockstate.m_280296_() && !DiodeBlock.m_52586_(blockstate)) {
                   return false;
                }
@@ -152,6 +_,19 @@
          return false;
       } else {
          if (!this.m_213877_() && !this.m_9236_().f_46443_) {
+            // CraftBukkit start - fire break events
+            Entity damager = (p_31715_.m_269014_()) ? p_31715_.m_7639_() : p_31715_.m_7640_();
+            HangingBreakEvent event;
+            if (damager != null) {
+               event = new HangingBreakByEntityEvent((Hanging) this.getBukkitEntity(), damager.getBukkitEntity(), p_31715_.m_269533_(DamageTypeTags.f_268415_) ? HangingBreakEvent.RemoveCause.EXPLOSION : HangingBreakEvent.RemoveCause.ENTITY);
+            } else {
+               event = new HangingBreakEvent((Hanging) this.getBukkitEntity(), p_31715_.m_269533_(DamageTypeTags.f_268415_) ? HangingBreakEvent.RemoveCause.EXPLOSION : HangingBreakEvent.RemoveCause.DEFAULT);
+            }
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+            if (this.m_213877_() || event.isCancelled()) {
+               return true;
+            }
+            // CraftBukkit end
             this.m_6074_();
             this.m_5834_();
             this.m_5553_(p_31715_.m_7639_());
@@ -163,6 +_,15 @@
 
    public void m_6478_(MoverType p_31719_, Vec3 p_31720_) {
       if (!this.m_9236_().f_46443_ && !this.m_213877_() && p_31720_.m_82556_() > 0.0D) {
+         if (this.m_213877_()) return; // CraftBukkit
+         // CraftBukkit start - fire break events
+         // TODO - Does this need its own cause? Seems to only be triggered by pistons
+         HangingBreakEvent event = new HangingBreakEvent((Hanging) this.getBukkitEntity(), HangingBreakEvent.RemoveCause.PHYSICS);
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (this.m_213877_() || event.isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          this.m_6074_();
          this.m_5553_((Entity)null);
       }
@@ -170,7 +_,7 @@
    }
 
    public void m_5997_(double p_31744_, double p_31745_, double p_31746_) {
-      if (!this.m_9236_().f_46443_ && !this.m_213877_() && p_31744_ * p_31744_ + p_31745_ * p_31745_ + p_31746_ * p_31746_ > 0.0D) {
+      if (false && !this.m_9236_().f_46443_ && !this.m_213877_() && p_31744_ * p_31744_ + p_31745_ * p_31745_ + p_31746_ * p_31746_ > 0.0D) { // CraftBukkit - not needed
          this.m_6074_();
          this.m_5553_((Entity)null);
       }
