--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -8,6 +_,7 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.damagesource.IndirectEntityDamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LightningBolt;
@@ -20,9 +_,13 @@
 import net.minecraft.world.level.block.Mirror;
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.Material;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.Validate;
+import org.bukkit.entity.Hanging;
+import org.bukkit.event.hanging.HangingBreakByEntityEvent;
+import org.bukkit.event.hanging.HangingBreakEvent;
 
 public abstract class HangingEntity extends Entity {
    protected static final Predicate<Entity> f_31697_ = (p_31734_) -> {
@@ -41,10 +_,10 @@
       this.f_31698_ = p_31708_;
    }
 
-   protected void m_8097_() {
+    protected void m_8097_() {
    }
 
-   protected void m_6022_(Direction p_31728_) {
+   public void m_6022_(Direction p_31728_) {
       Validate.notNull(p_31728_);
       Validate.isTrue(p_31728_.m_122434_().m_122479_());
       this.f_31699_ = p_31728_;
@@ -55,23 +_,34 @@
 
    protected void m_7087_() {
       if (this.f_31699_ != null) {
-         double d0 = (double)this.f_31698_.m_123341_() + 0.5D;
-         double d1 = (double)this.f_31698_.m_123342_() + 0.5D;
-         double d2 = (double)this.f_31698_.m_123343_() + 0.5D;
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.m_20011_(calculateBoundingBox(this, this.f_31698_, this.f_31699_, this.m_7076_(), this.m_7068_()));
+         // CraftBukkit end
+      }
+   }
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AABB calculateBoundingBox(Entity entity, BlockPos pos, Direction opposite, int width, int height) {
+         double d0 = (double) pos.m_123341_() + 0.5D;
+         double d1 = (double) pos.m_123342_() + 0.5D;
+         double d2 = (double) pos.m_123343_() + 0.5D;
          double d3 = 0.46875D;
-         double d4 = this.m_31709_(this.m_7076_());
-         double d5 = this.m_31709_(this.m_7068_());
-         d0 -= (double)this.f_31699_.m_122429_() * 0.46875D;
-         d2 -= (double)this.f_31699_.m_122431_() * 0.46875D;
+         double d4 = m_31709_(width);
+         double d5 = m_31709_(height);
+         d0 -= (double) opposite.m_122429_() * 0.46875D;
+         d2 -= (double) opposite.m_122431_() * 0.46875D;
          d1 += d5;
-         Direction direction = this.f_31699_.m_122428_();
+         Direction direction = opposite.m_122428_();
          d0 += d4 * (double)direction.m_122429_();
          d2 += d4 * (double)direction.m_122431_();
-         this.m_20343_(d0, d1, d2);
-         double d6 = (double)this.m_7076_();
-         double d7 = (double)this.m_7068_();
-         double d8 = (double)this.m_7076_();
-         if (this.f_31699_.m_122434_() == Direction.Axis.Z) {
+         if(entity != null)
+         {
+            entity.m_20343_( d0, d1, d2 );
+         }
+         double d6 = (double)width;
+         double d7 = (double)height;
+         double d8 = (double)width;
+         if (opposite.m_122434_() == Direction.Axis.Z) {
             d8 = 1.0D;
          } else {
             d6 = 1.0D;
@@ -80,20 +_,40 @@
          d6 /= 32.0D;
          d7 /= 32.0D;
          d8 /= 32.0D;
-         this.m_20011_(new AABB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8));
+
+         return new AABB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8);
       }
-   }
-
-   private double m_31709_(int p_31710_) {
+
+   // CraftBukkit end
+
+   private static double m_31709_(int p_31710_) {// CraftBukkit - static
       return p_31710_ % 32 == 0 ? 0.5D : 0.0D;
    }
 
    public void m_8119_() {
       if (!this.f_19853_.f_46443_) {
          this.m_146871_();
-         if (this.f_31700_++ == 100) {
+         if (this.f_31700_++ == this.f_19853_.spigotConfig.hangingTickFrequency) { // Spigot
             this.f_31700_ = 0;
             if (!this.m_146910_() && !this.m_7088_()) {
+               // CraftBukkit start - fire break events
+               Material material = this.f_19853_.m_8055_(this.m_142538_()).m_60767_();
+               HangingBreakEvent.RemoveCause cause;
+
+               if (!material.equals(Material.f_76296_)) {
+                  // TODO: This feels insufficient to catch 100% of suffocation cases
+                  cause = HangingBreakEvent.RemoveCause.OBSTRUCTION;
+               } else {
+                  cause = HangingBreakEvent.RemoveCause.PHYSICS;
+               }
+
+               HangingBreakEvent event = new HangingBreakEvent(( Hanging ) this.getBukkitEntity(), cause);
+               this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+               if (this.m_146910_() || event.isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                this.m_146870_();
                this.m_5553_((Entity)null);
             }
@@ -118,6 +_,8 @@
                int j1 = (j - 1) / -2;
                blockpos$mutableblockpos.m_122190_(blockpos).m_122175_(direction, k + i1).m_122175_(Direction.UP, l + j1);
                BlockState blockstate = this.f_19853_.m_8055_(blockpos$mutableblockpos);
+               if (net.minecraft.world.level.block.Block.m_49863_(this.f_19853_, blockpos$mutableblockpos, this.f_31699_))
+                  continue;
                if (!blockstate.m_60767_().m_76333_() && !DiodeBlock.m_52586_(blockstate)) {
                   return false;
                }
@@ -150,6 +_,21 @@
          return false;
       } else {
          if (!this.m_146910_() && !this.f_19853_.f_46443_) {
+            // CraftBukkit start - fire break events
+            Entity damager = (p_31715_ instanceof IndirectEntityDamageSource ) ? ((IndirectEntityDamageSource) p_31715_).getProximateDamageSource() : p_31715_.m_7639_();
+            HangingBreakEvent event;
+            if (damager != null) {
+               event = new HangingBreakByEntityEvent((Hanging) this.getBukkitEntity(), damager.getBukkitEntity(), p_31715_.m_19372_() ? HangingBreakEvent.RemoveCause.EXPLOSION : HangingBreakEvent.RemoveCause.ENTITY);
+            } else {
+               event = new HangingBreakEvent((Hanging) this.getBukkitEntity(), p_31715_.m_19372_() ? HangingBreakEvent.RemoveCause.EXPLOSION : HangingBreakEvent.RemoveCause.DEFAULT);
+            }
+
+            this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (this.m_146910_() || event.isCancelled()) {
+               return true;
+            }
+            // CraftBukkit end
             this.m_6074_();
             this.m_5834_();
             this.m_5553_(p_31715_.m_7639_());
@@ -161,6 +_,17 @@
 
    public void m_6478_(MoverType p_31719_, Vec3 p_31720_) {
       if (!this.f_19853_.f_46443_ && !this.m_146910_() && p_31720_.m_82556_() > 0.0D) {
+         if (this.m_146910_()) return; // CraftBukkit
+
+         // CraftBukkit start - fire break events
+         // TODO - Does this need its own cause? Seems to only be triggered by pistons
+         HangingBreakEvent event = new HangingBreakEvent((Hanging) this.getBukkitEntity(), HangingBreakEvent.RemoveCause.PHYSICS);
+         this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (this.m_146910_() || event.isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          this.m_6074_();
          this.m_5553_((Entity)null);
       }
