--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -57,7 +_,11 @@
       this.m_6022_(p_149624_);
    }
 
-   protected float m_6380_(Pose p_31784_, EntityDimensions p_31785_) {
+    public static AABB calculateBoundingBox(Object o, BlockPos pos, Direction opposite, int width, int height) {
+      return  null;
+    }
+
+    protected float m_6380_(Pose p_31784_, EntityDimensions p_31785_) {
       return 0.0F;
    }
 
@@ -66,7 +_,7 @@
       this.m_20088_().m_135372_(f_31758_, 0);
    }
 
-   protected void m_6022_(Direction p_31793_) {
+   public void m_6022_(Direction p_31793_) {
       Validate.notNull(p_31793_);
       this.f_31699_ = p_31793_;
       if (p_31793_.m_122434_().m_122479_()) {
@@ -111,6 +_,35 @@
       }
    }
 
+   public static AABB calculateBoundingBox(@Nullable Entity entity, BlockPos blockPosition, Direction direction, int width, int height) {
+      double d0 = 0.46875D;
+      double d1 = (double) blockPosition.m_123341_() + 0.5D - (double) direction.m_122429_() * 0.46875D;
+      double d2 = (double) blockPosition.m_123342_() + 0.5D - (double) direction.m_122430_() * 0.46875D;
+      double d3 = (double) blockPosition.m_123343_() + 0.5D - (double) direction.m_122431_() * 0.46875D;
+      if (entity != null) {
+         entity.m_20343_(d1, d2, d3);
+      }
+      double d4 = (double) width;
+      double d5 = (double) height;
+      double d6 = (double) width;
+      Direction.Axis direction$axis = direction.m_122434_();
+      switch (direction$axis) {
+         case X:
+            d4 = 1.0D;
+            break;
+         case Y:
+            d5 = 1.0D;
+            break;
+         case Z:
+            d6 = 1.0D;
+      }
+
+      d4 /= 32.0D;
+      d5 /= 32.0D;
+      d6 /= 32.0D;
+      return new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6);
+   }
+
    public boolean m_7088_() {
       if (this.f_31755_) {
          return true;
@@ -152,6 +_,11 @@
          return false;
       } else if (!p_31776_.m_19372_() && !this.m_31822_().m_41619_()) {
          if (!this.f_19853_.f_46443_) {
+            // CraftBukkit start - fire EntityDamageEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31776_, p_31777_, false) || this.m_146910_()) {
+               return true;
+            }
+            // CraftBukkit end
             this.m_31802_(p_31776_.m_7639_(), false);
             this.m_5496_(this.m_142544_(), 1.0F, 1.0F);
          }
@@ -260,6 +_,24 @@
 
       this.m_20088_().m_135381_(f_31757_, p_31790_);
       if (!p_31790_.m_41619_()) {
+         this.m_5496_(this.m_142546_(), 1.0F, 1.0F);
+      }
+
+      if (p_31791_ && this.f_31698_ != null) {
+         this.f_19853_.m_46717_(this.f_31698_, Blocks.f_50016_);
+      }
+
+   }
+
+   public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
+      if (!p_31790_.m_41619_()) {
+         p_31790_ = p_31790_.m_41777_();
+         p_31790_.m_41764_(1);
+         p_31790_.m_41636_(this);
+      }
+
+      this.m_20088_().m_135381_(f_31757_, p_31790_);
+      if (!p_31790_.m_41619_() && playSound) { // CraftBukkit
          this.m_5496_(this.m_142546_(), 1.0F, 1.0F);
       }
 
