--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -41,8 +_,8 @@
    private static final EntityDataAccessor<ItemStack> f_31757_ = SynchedEntityData.m_135353_(ItemFrame.class, EntityDataSerializers.f_135033_);
    private static final EntityDataAccessor<Integer> f_31758_ = SynchedEntityData.m_135353_(ItemFrame.class, EntityDataSerializers.f_135028_);
    public static final int f_149619_ = 8;
-   private float f_31754_ = 1.0F;
-   private boolean f_31755_;
+   public float f_31754_ = 1.0F;
+   public boolean f_31755_;
 
    public ItemFrame(EntityType<? extends ItemFrame> p_31761_, Level p_31762_) {
       super(p_31761_, p_31762_);
@@ -66,7 +_,7 @@
       this.m_20088_().m_135372_(f_31758_, 0);
    }
 
-   protected void m_6022_(Direction p_31793_) {
+   public void m_6022_(Direction p_31793_) {
       Validate.notNull(p_31793_);
       this.f_31699_ = p_31793_;
       if (p_31793_.m_122434_().m_122479_()) {
@@ -84,16 +_,26 @@
 
    protected void m_7087_() {
       if (this.f_31699_ != null) {
-         double d0 = 0.46875D;
-         double d1 = (double)this.f_31698_.m_123341_() + 0.5D - (double)this.f_31699_.m_122429_() * 0.46875D;
-         double d2 = (double)this.f_31698_.m_123342_() + 0.5D - (double)this.f_31699_.m_122430_() * 0.46875D;
-         double d3 = (double)this.f_31698_.m_123343_() + 0.5D - (double)this.f_31699_.m_122431_() * 0.46875D;
-         this.m_20343_(d1, d2, d3);
-         double d4 = (double)this.m_7076_();
-         double d5 = (double)this.m_7068_();
-         double d6 = (double)this.m_7076_();
-         Direction.Axis direction$axis = this.f_31699_.m_122434_();
-         switch(direction$axis) {
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.m_20011_(calculateBoundingBox(this, this.f_31698_, this.f_31699_, this.m_7076_(), this.m_7068_()));
+         // CraftBukkit end
+      }
+   }
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AABB calculateBoundingBox(@Nullable Entity entity, BlockPos blockPosition, Direction direction, int width, int height) {
+      double d0 = 0.46875D;
+      double d1 = (double) blockPosition.m_123341_() + 0.5D - (double) direction.m_122429_() * 0.46875D;
+      double d2 = (double) blockPosition.m_123342_() + 0.5D - (double) direction.m_122430_() * 0.46875D;
+      double d3 = (double) blockPosition.m_123343_() + 0.5D - (double) direction.m_122431_() * 0.46875D;
+      if (entity != null) {
+         entity.m_20343_(d1, d2, d3);
+      }
+      double d4 = width;
+      double d5 = height;
+      double d6 = width;
+      Direction.Axis direction$axis = direction.m_122434_();
+      switch (direction$axis) {
          case X:
             d4 = 1.0D;
             break;
@@ -102,14 +_,14 @@
             break;
          case Z:
             d6 = 1.0D;
-         }
-
-         d4 /= 32.0D;
-         d5 /= 32.0D;
-         d6 /= 32.0D;
-         this.m_20011_(new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6));
       }
+
+      d4 /= 32.0D;
+      d5 /= 32.0D;
+      d6 /= 32.0D;
+      return new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6);
    }
+   // CraftBukkit end
 
    public boolean m_7088_() {
       if (this.f_31755_) {
@@ -152,6 +_,11 @@
          return false;
       } else if (!p_31776_.m_19372_() && !this.m_31822_().m_41619_()) {
          if (!this.f_19853_.f_46443_) {
+            // CraftBukkit start - fire EntityDamageEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31776_, p_31777_, false) || this.m_146910_()) {
+               return true;
+            }
+            // CraftBukkit end
             this.m_31802_(p_31776_.m_7639_(), false);
             this.m_5496_(this.m_142544_(), 1.0F, 1.0F);
          }
@@ -252,6 +_,10 @@
    }
 
    public void m_31789_(ItemStack p_31790_, boolean p_31791_) {
+      this.setItem(p_31790_, p_31791_, true);
+   }
+
+   public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
       if (!p_31790_.m_41619_()) {
          p_31790_ = p_31790_.m_41777_();
          p_31790_.m_41764_(1);
@@ -259,7 +_,7 @@
       }
 
       this.m_20088_().m_135381_(f_31757_, p_31790_);
-      if (!p_31790_.m_41619_()) {
+      if (!p_31790_.m_41619_() && playSound) { // CraftBukkit
          this.m_5496_(this.m_142546_(), 1.0F, 1.0F);
       }
 
