--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -25,6 +_,12 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+// CraftBukkit end
+
 public class LeashFenceKnotEntity extends HangingEntity {
    public static final double f_149638_ = 0.375D;
 
@@ -83,6 +_,12 @@
 
          for(Mob mob : list) {
             if (mob.m_21524_() == p_31842_) {
+               // CraftBukkit start
+               if (CraftEventFactory.callPlayerLeashEntityEvent(mob, this, p_31842_, p_31843_).isCancelled()) {
+                  ((ServerPlayer) p_31842_).f_8906_.m_9829_(new ClientboundSetEntityLinkPacket(mob, mob.m_21524_()));
+                  continue;
+               }
+               // CraftBukkit end
                mob.m_21463_(this, true);
                flag = true;
             }
@@ -90,14 +_,28 @@
 
          boolean flag1 = false;
          if (!flag) {
-            this.m_146870_();
-            if (p_31842_.m_150110_().f_35937_) {
+            // CraftBukkit start - Move below
+            // this.discard();
+            boolean die = true;
+            // CraftBukkit end
+            if (true || p_31842_.m_150110_().f_35937_) { // CraftBukkit - Process for non-creative as well
                for(Mob mob1 : list) {
                   if (mob1.m_21523_() && mob1.m_21524_() == this) {
-                     mob1.m_21455_(true, false);
+                     // CraftBukkit start
+                     if (CraftEventFactory.callPlayerUnleashEntityEvent(mob1, p_31842_, p_31843_).isCancelled()) {
+                        die = false;
+                        continue;
+                     }
+                     mob1.m_21455_(true, !p_31842_.m_150110_().f_35937_); // false -> survival mode boolean
+                     // CraftBukkit end
                      flag1 = true;
                   }
                }
+               // CraftBukkit start
+               if (die) {
+                  this.m_146870_();
+               }
+               // CraftBukkit end
             }
          }
 
