--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -7,6 +_,8 @@
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
+import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.world.InteractionHand;
@@ -22,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
 
 public class LeashFenceKnotEntity extends HangingEntity {
    public static final double f_149638_ = 0.375D;
@@ -81,19 +_,38 @@
 
          for(Mob mob : list) {
             if (mob.m_21524_() == p_31842_) {
+               // CraftBukkit start
+               if (CraftEventFactory.callPlayerLeashEntityEvent(mob, this, p_31842_).isCancelled()) {
+                  ((ServerPlayer) p_31842_).f_8906_.m_141995_(new ClientboundSetEntityLinkPacket(mob, mob.m_21524_()));
+                  continue;
+               }
+               // CraftBukkit end
                mob.m_21463_(this, true);
                flag = true;
             }
          }
 
          if (!flag) {
+            boolean die = true;
             this.m_146870_();
-            if (p_31842_.m_150110_().f_35937_) {
+            if (true || p_31842_.m_150110_().f_35937_) {
                for(Mob mob1 : list) {
                   if (mob1.m_21523_() && mob1.m_21524_() == this) {
                      mob1.m_21455_(true, false);
+                     // CraftBukkit start
+                     if (CraftEventFactory.callPlayerUnleashEntityEvent(mob1, p_31842_).isCancelled()) {
+                        die = false;
+                        continue;
+                     }
+                     mob1.m_21455_(true, !p_31842_.m_150110_().f_35937_); // false -> survival mode boolean
+                     // CraftBukkit end
                   }
                }
+               // CraftBukkit start
+               if (die) {
+                  this.m_146870_();
+               }
+               // CraftBukkit end
             }
          }
 
