--- a/net/minecraft/world/biome/BiomeContainer.java
+++ b/net/minecraft/world/biome/BiomeContainer.java
@@ -17,7 +_,7 @@
    public static final int field_227049_a_ = 1 << field_227052_d_ + field_227052_d_ + field_227053_e_;
    public static final int field_227050_b_ = (1 << field_227052_d_) - 1;
    public static final int field_227051_c_ = (1 << field_227053_e_) - 1;
-   private final IObjectIntIterable<Biome> field_242704_g;
+   public final IObjectIntIterable<Biome> field_242704_g; // PAIL
    private final Biome[] field_227054_f_;
 
    public BiomeContainer(IObjectIntIterable<Biome> p_i241971_1_, Biome[] p_i241971_2_) {
@@ -101,4 +_,13 @@
       int k = p_225526_3_ & field_227050_b_;
       return this.field_227054_f_[j << field_227052_d_ + field_227052_d_ | k << field_227052_d_ | i];
    }
+
+   // CraftBukkit start
+   public void setBiome(int x, int y, int z, Biome biome) {
+      int i = x & field_227050_b_;
+      int j = MathHelper.func_76125_a(y, 0, field_227051_c_);
+      int k = z & field_227050_b_;
+      this.field_227054_f_[j << field_227052_d_ + field_227052_d_ | k << field_227052_d_ | i] = biome;
+   }
+   // CraftBukkit end
 }
