--- ../src-base/minecraft/net/minecraft/client/gui/spectator/categories/TeleportToTeam.java
+++ ../src-work/minecraft/net/minecraft/client/gui/spectator/categories/TeleportToTeam.java
@@ -34,7 +34,7 @@
 
         for (ScorePlayerTeam scoreplayerteam : minecraft.world.getScoreboard().getTeams())
         {
-            this.items.add(new TeleportToTeam.TeamSelectionObject(scoreplayerteam));
+            this.items.add(new TeamSelectionObject(scoreplayerteam));
         }
     }
 
