--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.border;
 
+import catserver.server.CatServer;
 import com.google.common.collect.Lists;
 import com.mojang.serialization.DynamicLike;
 import java.util.List;
@@ -9,6 +_,7 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
@@ -26,6 +_,7 @@
    private double f_61912_;
    int f_61913_ = 29999984;
    private WorldBorder.BorderExtent f_61914_ = new WorldBorder.StaticBorderExtent(5.9999968E7D);
+   public Level world; // CraftBukkit // CatServer - ServerLevel -> Level
    public static final WorldBorder.Settings f_61907_ = new WorldBorder.Settings(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 5.9999968E7D, 0L, 0.0D);
 
    public boolean m_61937_(BlockPos p_61938_) {
@@ -149,6 +_,12 @@
    }
 
    public void m_61929_(BorderChangeListener p_61930_) {
+      // CraftBukkit start
+      if (f_61905_.contains(p_61930_)) {
+         CatServer.LOGGER.log(org.apache.logging.log4j.Level.INFO, "Debug Information: WorldBorder listener already exists.");
+         return;
+      }
+      // CraftBukkit end
       this.f_61905_.add(p_61930_);
    }
 
