--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -6,6 +_,7 @@
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
@@ -27,6 +_,7 @@
    int f_61913_ = 29999984;
    private WorldBorder.BorderExtent f_61914_ = new WorldBorder.StaticBorderExtent(5.9999968E7D);
    public static final WorldBorder.Settings f_61907_ = new WorldBorder.Settings(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 5.9999968E7D, 0L, 0.0D);
+   public ServerLevel world; // CraftBukkit
 
    public boolean m_61937_(BlockPos p_61938_) {
       return (double)(p_61938_.m_123341_() + 1) > this.m_61955_() && (double)p_61938_.m_123341_() < this.m_61957_() && (double)(p_61938_.m_123343_() + 1) > this.m_61956_() && (double)p_61938_.m_123343_() < this.m_61958_();
@@ -149,6 +_,7 @@
    }
 
    public void m_61929_(BorderChangeListener p_61930_) {
+      if (f_61905_.contains(p_61930_)) return; // CraftBukkit
       this.f_61905_.add(p_61930_);
    }
 
