--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -31,6 +_,20 @@
 import net.minecraft.world.level.block.entity.BlockEntity;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.Map;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.event.Event.Result;
+import org.bukkit.event.inventory.InventoryDragEvent;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.inventory.InventoryView;
+// CraftBukkit end
+
 public abstract class AbstractContainerMenu {
    private static final Logger f_207773_ = LogUtils.getLogger();
    public static final int f_150385_ = -999;
@@ -41,11 +_,11 @@
    public static final int f_150390_ = 1;
    public static final int f_150391_ = 2;
    public static final int f_150392_ = Integer.MAX_VALUE;
-   private final NonNullList<ItemStack> f_38841_ = NonNullList.m_122779_();
-   public final NonNullList<Slot> f_38839_ = NonNullList.m_122779_();
+   public NonNullList<ItemStack> f_38841_ = NonNullList.m_122779_(); // CatServer - private -> public  remove final
+   public NonNullList<Slot> f_38839_ = NonNullList.m_122779_(); // CatServer - remove final
    private final List<DataSlot> f_38842_ = Lists.newArrayList();
    private ItemStack f_150393_ = ItemStack.f_41583_;
-   private final NonNullList<ItemStack> f_150394_ = NonNullList.m_122779_();
+   public NonNullList<ItemStack> f_150394_ = NonNullList.m_122779_(); // CatServer - private -> public  remove final
    private final IntList f_150395_ = new IntArrayList();
    private ItemStack f_150396_ = ItemStack.f_41583_;
    private int f_182405_;
@@ -59,6 +_,95 @@
    @Nullable
    private ContainerSynchronizer f_150397_;
    private boolean f_150398_;
+   // CraftBukkit start
+   public boolean checkReachable = true;
+
+   // CatServer start
+   private boolean isBukkitViewCreated;
+
+   public InventoryView bukkitView;
+
+   @Nullable
+   public InventoryView getBukkitView() {
+      if (!isBukkitViewCreated) {
+         isBukkitViewCreated = true;
+         bukkitView = computeBukkitView();
+         return bukkitView;
+      }
+      return bukkitView;
+   }
+
+   public void setBukkitView(InventoryView iv) {
+      bukkitView = iv;
+      isBukkitViewCreated = true;
+   }
+
+   @Nullable
+   public InventoryView computeBukkitView() {
+      Set<Container> uniqueInventorySet = new java.util.HashSet<>();
+      for(Slot slot : f_38839_) {
+         if (slot.f_40218_ != null) {
+            uniqueInventorySet.add(slot.f_40218_);
+         }
+      }
+      List<Container> inventories = new java.util.ArrayList<>(uniqueInventorySet);
+      Inventory playerInv = null;
+
+      for(java.util.Iterator<Container> it = inventories.iterator(); it.hasNext();) {
+         Container inv = it.next();
+         if(inv instanceof Inventory) {
+            playerInv = (Inventory) inv;
+            it.remove();
+            break;
+         } else if (inv instanceof net.minecraftforge.items.wrapper.PlayerInvWrapper) {
+            playerInv = catserver.server.inventory.CatCustomInventory.getPlayerInv((net.minecraftforge.items.wrapper.PlayerInvWrapper) inv);
+            it.remove();
+            break;
+         }
+      }
+      if(playerInv == null) return null;
+      org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer bukkitPlayer = (org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer) playerInv.f_35978_.getBukkitEntity();
+      org.bukkit.inventory.Inventory craftInv;
+      if(inventories.size() != 1)
+         craftInv = org.bukkit.Bukkit.getServer().createInventory(bukkitPlayer, InventoryType.CHEST);
+      else
+         craftInv = new CraftInventory(inventories.get(0));
+
+      return new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView(bukkitPlayer, craftInv, this);
+   }
+
+   public void transferTo(AbstractContainerMenu other, org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity player) {
+      InventoryView source = this.getBukkitView(), destination = other.getBukkitView();
+      if (source != null) {
+         ((CraftInventory) source.getTopInventory()).getInventory().onClose(player);
+         ((CraftInventory) source.getBottomInventory()).getInventory().onClose(player);
+      }
+      if (destination != null) {
+         ((CraftInventory) destination.getTopInventory()).getInventory().onOpen(player);
+         ((CraftInventory) destination.getBottomInventory()).getInventory().onOpen(player);
+      }
+   }
+   // CatServer end
+
+   private Component title;
+
+   public final Component getTitle() {
+      return this.title == null ? Component.m_237119_() : this.title; // CatServer
+   }
+
+   public final void setTitle(Component title) {
+      Preconditions.checkState(this.title == null, "Title already set");
+      this.title = title;
+   }
+
+   // CraftBukkit start
+   public void broadcastCarriedItem() {
+      this.f_150396_ = this.m_142621_().m_41777_();
+      if (this.f_150397_ != null) {
+         this.f_150397_.m_142529_(this, this.f_150396_);
+      }
+   }
+   // CraftBukkit end
 
    protected AbstractContainerMenu(@Nullable MenuType<?> p_38851_, int p_38852_) {
       this.f_38843_ = p_38851_;
@@ -67,7 +_,11 @@
 
    protected static boolean m_38889_(ContainerLevelAccess p_38890_, Player p_38891_, Block p_38892_) {
       return p_38890_.m_39299_((p_38916_, p_38917_) -> {
-         return !p_38916_.m_8055_(p_38917_).m_60713_(p_38892_) ? false : p_38891_.m_20275_((double)p_38917_.m_123341_() + 0.5D, (double)p_38917_.m_123342_() + 0.5D, (double)p_38917_.m_123343_() + 0.5D) <= 64.0D;
+         // Vanilla uses 64 which is 8*8. Default reach is 4.5 so add 3.5 padding
+         // magic number so required Mixins can still target here
+         double reachOld = 64.0;
+         double reach = p_38891_.m_21133_(net.minecraftforge.common.ForgeMod.BLOCK_REACH.get()) + 3.5;
+         return !p_38916_.m_8055_(p_38917_).m_60713_(p_38892_) ? false : p_38891_.m_20275_((double)p_38917_.m_123341_() + 0.5D, (double)p_38917_.m_123342_() + 0.5D, (double)p_38917_.m_123343_() + 0.5D) <= reach * reach;
       }, true);
    }
 
@@ -229,6 +_,8 @@
             ItemStack itemstack1 = p_150438_.get();
             this.f_150394_.set(p_150436_, itemstack1);
             if (this.f_150397_ != null) {
+               // Forge: Only synchronize a slot change if the itemstack actually changed in a way that is relevant to the client (i.e. share tag changed)
+               if (!p_150437_.equals(itemstack, true))
                this.f_150397_.m_142074_(this, p_150436_, itemstack1);
             }
          }
@@ -332,7 +_,7 @@
             }
          } else if (this.f_38846_ == 2) {
             if (!this.f_38847_.isEmpty()) {
-               if (this.f_38847_.size() == 1) {
+               if (false && this.f_38847_.size() == 1) { // CraftBukkit - treat everything as a drag since we are unable to easily call InventoryClickEvent instead
                   int i1 = (this.f_38847_.iterator().next()).f_40219_;
                   this.m_38951_();
                   this.m_150430_(i1, this.f_38845_, ClickType.PICKUP, p_150434_);
@@ -347,6 +_,7 @@
 
                int k1 = this.m_142621_().m_41613_();
 
+               Map<Integer, ItemStack> draggedSlots = new HashMap<Integer, ItemStack>(); // CraftBukkit - Store slots from drag in map (raw slot id -> new stack)
                for(Slot slot1 : this.f_38847_) {
                   ItemStack itemstack1 = this.m_142621_();
                   if (slot1 != null && m_38899_(slot1, itemstack1, true) && slot1.m_5857_(itemstack1) && (this.f_38845_ == 2 || itemstack1.m_41613_() >= this.f_38847_.size()) && this.m_5622_(slot1)) {
@@ -354,12 +_,45 @@
                      int k = Math.min(itemstack2.m_41741_(), slot1.m_5866_(itemstack2));
                      int l = Math.min(m_278794_(this.f_38847_, this.f_38845_, itemstack2) + j, k);
                      k1 -= l - j;
-                     slot1.m_269060_(itemstack2.m_255036_(l));
+                     // slot1.setByPlayer(itemstack2.copyWithCount(l));
+                     draggedSlots.put(slot1.f_40219_, itemstack2.m_255036_(l)); // CraftBukkit - Put in map instead of setting
                   }
                }
 
-               itemstack2.m_41764_(k1);
-               this.m_142503_(itemstack2);
+               // itemstack2.setCount(k1);
+               // this.setCarried(itemstack2);
+               // CraftBukkit start - InventoryDragEvent
+               InventoryView view = getBukkitView();
+               org.bukkit.inventory.ItemStack newcursor = CraftItemStack.asCraftMirror(itemstack2);
+               newcursor.setAmount(k1);
+               Map<Integer, org.bukkit.inventory.ItemStack> eventmap = new HashMap<>();
+               for (Map.Entry<Integer, ItemStack> ditem : draggedSlots.entrySet()) {
+                  eventmap.put(ditem.getKey(), CraftItemStack.asBukkitCopy(ditem.getValue()));
+               }
+               // It's essential that we set the cursor to the new value here to prevent item duplication if a plugin closes the inventory.
+               ItemStack oldCursor = this.m_142621_();
+               this.m_142503_(CraftItemStack.asNMSCopy(newcursor));
+               InventoryDragEvent event = new InventoryDragEvent(view, (newcursor.getType() != org.bukkit.Material.AIR ? newcursor : null), CraftItemStack.asBukkitCopy(oldCursor), this.f_38845_ == 1, eventmap);
+               p_150434_.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+               // Whether or not a change was made to the inventory that requires an update.
+               boolean needsUpdate = event.getResult() != org.bukkit.event.Event.Result.DEFAULT;
+               if (event.getResult() != org.bukkit.event.Event.Result.DENY) {
+                  for (Map.Entry<Integer, ItemStack> dslot : draggedSlots.entrySet()) {
+                     view.setItem(dslot.getKey(), CraftItemStack.asBukkitCopy(dslot.getValue()));
+                  }
+                  // The only time the carried item will be set to null is if the inventory is closed by the server.
+                  // If the inventory is closed by the server, then the cursor items are dropped.  This is why we change the cursor early.
+                  if (this.m_142621_() != null) {
+                     this.m_142503_(CraftItemStack.asNMSCopy(event.getCursor()));
+                     needsUpdate = true;
+                  }
+               } else {
+                  this.m_142503_(oldCursor);
+               }
+               if (needsUpdate && p_150434_ instanceof ServerPlayer) {
+                  this.m_150429_();
+               }
+               // CraftBukkit end
             }
 
             this.m_38951_();
@@ -373,8 +_,11 @@
          if (p_150431_ == -999) {
             if (!this.m_142621_().m_41619_()) {
                if (clickaction == ClickAction.PRIMARY) {
-                  p_150434_.m_36176_(this.m_142621_(), true);
+                  // CraftBukkit start
+                  ItemStack carried = this.m_142621_();
                   this.m_142503_(ItemStack.f_41583_);
+                  p_150434_.m_36176_(carried, true);
+                  // CraftBukkit end
                } else {
                   p_150434_.m_36176_(this.m_142621_().m_41620_(1), true);
                }
@@ -401,6 +_,7 @@
             ItemStack itemstack10 = this.m_142621_();
             p_150434_.m_141945_(itemstack10, slot7.m_7993_(), clickaction);
             if (!this.m_246200_(p_150434_, clickaction, slot7, itemstack9, itemstack10)) {
+            if (!net.minecraftforge.common.ForgeHooks.onItemStackedOn(itemstack9, itemstack10, slot7, clickaction, p_150434_, m_150446_()))
                if (itemstack9.m_41619_()) {
                   if (!itemstack10.m_41619_()) {
                      int i3 = clickaction == ClickAction.PRIMARY ? itemstack10.m_41613_() : 1;
@@ -433,6 +_,16 @@
             }
 
             slot7.m_6654_();
+            // CraftBukkit start - Make sure the client has the right slot contents
+            if (p_150434_ instanceof ServerPlayer && slot7.m_6641_() != 64) {
+               ((ServerPlayer) p_150434_).f_8906_.m_9829_(new ClientboundContainerSetSlotPacket(this.f_38840_, this.m_182425_(), slot7.f_40219_, slot7.m_7993_()));
+               // Updating a crafting inventory makes the client reset the result slot, have to send it again
+               if (getBukkitView() != null) // CatServer - Bypass send packet if Bukkit view is null (mod custom container)
+                  if (this.getBukkitView().getType() == InventoryType.WORKBENCH || this.getBukkitView().getType() == InventoryType.CRAFTING) {
+                     ((ServerPlayer) p_150434_).f_8906_.m_9829_(new ClientboundContainerSetSlotPacket(this.f_38840_, this.m_182425_(), 0, this.m_38853_(0).m_7993_()));
+                  }
+            }
+            // CraftBukkit end
          }
       } else if (p_150433_ == ClickType.SWAP) {
          Slot slot2 = this.f_38839_.get(p_150431_);
@@ -536,13 +_,14 @@
       if (p_38940_ instanceof ServerPlayer) {
          ItemStack itemstack = this.m_142621_();
          if (!itemstack.m_41619_()) {
+            this.m_142503_(ItemStack.f_41583_); // CraftBukkit - SPIGOT-4556 - from below
             if (p_38940_.m_6084_() && !((ServerPlayer)p_38940_).m_9232_()) {
                p_38940_.m_150109_().m_150079_(itemstack);
             } else {
                p_38940_.m_36176_(itemstack, false);
             }
 
-            this.m_142503_(ItemStack.f_41583_);
+            // this.setCarried(ItemStack.EMPTY); // CraftBukkit - moved up
          }
       }
 
@@ -610,14 +_,15 @@
             ItemStack itemstack = slot.m_7993_();
             if (!itemstack.m_41619_() && ItemStack.m_150942_(p_38904_, itemstack)) {
                int j = itemstack.m_41613_() + p_38904_.m_41613_();
-               if (j <= p_38904_.m_41741_()) {
+               int maxSize = Math.min(slot.m_6641_(), p_38904_.m_41741_());
+               if (j <= maxSize) {
                   p_38904_.m_41764_(0);
                   itemstack.m_41764_(j);
                   slot.m_6654_();
                   flag = true;
-               } else if (itemstack.m_41613_() < p_38904_.m_41741_()) {
-                  p_38904_.m_41774_(p_38904_.m_41741_() - itemstack.m_41613_());
-                  itemstack.m_41764_(p_38904_.m_41741_());
+               } else if (itemstack.m_41613_() < maxSize) {
+                  p_38904_.m_41774_(maxSize - itemstack.m_41613_());
+                  itemstack.m_41764_(maxSize);
                   slot.m_6654_();
                   flag = true;
                }
@@ -718,7 +_,7 @@
             i = 1;
             break;
          case 2:
-            i = p_279172_.m_41720_().m_41459_();
+            i = p_279172_.m_41741_();
             break;
          default:
             i = p_279172_.m_41613_();
@@ -760,6 +_,11 @@
    }
 
    public ItemStack m_142621_() {
+      // CraftBukkit start
+      if (this.f_150393_.m_41619_()) {
+         this.m_142503_(ItemStack.f_41583_);
+      }
+      // CraftBukkit end
       return this.f_150393_;
    }
 
@@ -809,4 +_,15 @@
       this.f_182405_ = this.f_182405_ + 1 & 32767;
       return this.f_182405_;
    }
+
+   // Paper start - add missing BlockInventoryHolder to inventories
+   // The reason this is a supplier, is that the createHolder method uses the bukkit InventoryView#getTopInventory to get the inventory in question
+   // and that can't be obtained safely until the AbstractContainerMenu has been fully constructed. Using a supplier lazily
+   // initializes the InventoryHolder safely.
+   protected final Supplier<org.bukkit.inventory.BlockInventoryHolder> createBlockHolder(final ContainerLevelAccess context) {
+      //noinspection ConstantValue
+      Preconditions.checkArgument(context != null, "context was null");
+      return () -> context.createBlockHolder(this);
+   }
+   // Paper end
 }
