--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -11,6 +_,8 @@
 import net.minecraft.world.item.crafting.RecipeType;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryFurnace;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public abstract class AbstractFurnaceMenu extends RecipeBookMenu<Container> {
    public static final int f_150453_ = 0;
@@ -28,6 +_,22 @@
    private final RecipeType<? extends AbstractCookingRecipe> f_38957_;
    private final RecipeBookType f_38958_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryFurnace inventory = new CraftInventoryFurnace((AbstractFurnaceBlockEntity) this.f_38955_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    protected AbstractFurnaceMenu(MenuType<?> p_38960_, RecipeType<? extends AbstractCookingRecipe> p_38961_, RecipeBookType p_38962_, int p_38963_, Inventory p_38964_) {
       this(p_38960_, p_38961_, p_38962_, p_38963_, p_38964_, new SimpleContainer(3), new SimpleContainerData(4));
    }
@@ -44,7 +_,7 @@
       this.m_38897_(new Slot(p_38971_, 0, 56, 17));
       this.m_38897_(new FurnaceFuelSlot(this, p_38971_, 1, 56, 53));
       this.m_38897_(new FurnaceResultSlot(p_38970_.f_35978_, p_38971_, 2, 116, 35));
-
+      this.player = p_38970_; // CraftBukkit - save player
       for(int i = 0; i < 3; ++i) {
          for(int j = 0; j < 9; ++j) {
             this.m_38897_(new Slot(p_38970_, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
@@ -91,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_38974_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_38955_.m_6542_(p_38974_);
    }
 
@@ -147,7 +_,7 @@
    }
 
    protected boolean m_38988_(ItemStack p_38989_) {
-      return AbstractFurnaceBlockEntity.m_58399_(p_38989_);
+      return net.minecraftforge.common.ForgeHooks.getBurnTime(p_38989_, this.f_38957_) > 0;
    }
 
    public int m_38995_() {
