--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class AnvilMenu extends ItemCombinerMenu {
    private static final Logger f_38999_ = LogUtils.getLogger();
@@ -31,6 +_,22 @@
    private static final int f_150466_ = 1;
    private static final int f_150467_ = 1;
 
+   // CraftBukkit start
+   public int maximumRepairCost = 40;
+   private CraftInventoryView bukkitEntity;
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryAnvil(
+              f_39770_.getLocation(), this.f_39769_, this.f_39768_, this);
+      bukkitEntity = new CraftInventoryView(this.f_39771_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public AnvilMenu(int p_39005_, Inventory p_39006_) {
       this(p_39005_, p_39006_, ContainerLevelAccess.f_39287_);
    }
@@ -53,6 +_,8 @@
          p_150474_.m_6749_(-this.f_39002_.m_6501_());
       }
 
+      float breakChance = net.minecraftforge.common.ForgeHooks.onAnvilRepair(p_150474_, p_150475_, AnvilMenu.this.f_39769_.m_8020_(0), AnvilMenu.this.f_39769_.m_8020_(1));
+
       this.f_39769_.m_6836_(0, ItemStack.f_41583_);
       if (this.f_39000_ > 0) {
          ItemStack itemstack = this.f_39769_.m_8020_(1);
@@ -69,7 +_,7 @@
       this.f_39002_.m_6422_(0);
       this.f_39770_.m_39292_((p_150479_, p_150480_) -> {
          BlockState blockstate = p_150479_.m_8055_(p_150480_);
-         if (!p_150474_.m_150110_().f_35937_ && blockstate.m_204336_(BlockTags.f_13033_) && p_150474_.m_21187_().nextFloat() < 0.12F) {
+         if (!p_150474_.m_150110_().f_35937_ && blockstate.m_204336_(BlockTags.f_13033_) && p_150474_.m_21187_().nextFloat() < breakChance) {
             BlockState blockstate1 = AnvilBlock.m_48824_(blockstate);
             if (blockstate1 == null) {
                p_150479_.m_7471_(p_150480_, false);
@@ -92,7 +_,7 @@
       int j = 0;
       int k = 0;
       if (itemstack.m_41619_()) {
-         this.f_39768_.m_6836_(0, ItemStack.f_41583_);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.f_41583_); // CraftBukkit
          this.f_39002_.m_6422_(0);
       } else {
          ItemStack itemstack1 = itemstack.m_41777_();
@@ -100,12 +_,15 @@
          Map<Enchantment, Integer> map = EnchantmentHelper.m_44831_(itemstack1);
          j += itemstack.m_41610_() + (itemstack2.m_41619_() ? 0 : itemstack2.m_41610_());
          this.f_39000_ = 0;
+         boolean flag = false;
+
          if (!itemstack2.m_41619_()) {
-            boolean flag = itemstack2.m_150930_(Items.f_42690_) && !EnchantedBookItem.m_41163_(itemstack2).isEmpty();
+            if (!net.minecraftforge.common.ForgeHooks.onAnvilChange(this, itemstack, itemstack2, f_39768_, f_39001_, j, this.f_39771_)) return;
+            flag = itemstack2.m_41720_() == Items.f_42690_ && !EnchantedBookItem.m_41163_(itemstack2).isEmpty();
             if (itemstack1.m_41763_() && itemstack1.m_41720_().m_6832_(itemstack, itemstack2)) {
                int l2 = Math.min(itemstack1.m_41773_(), itemstack1.m_41776_() / 4);
                if (l2 <= 0) {
-                  this.f_39768_.m_6836_(0, ItemStack.f_41583_);
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.f_41583_); // CraftBukkit
                   this.f_39002_.m_6422_(0);
                   return;
                }
@@ -121,7 +_,7 @@
                this.f_39000_ = i3;
             } else {
                if (!flag && (!itemstack1.m_150930_(itemstack2.m_41720_()) || !itemstack1.m_41763_())) {
-                  this.f_39768_.m_6836_(0, ItemStack.f_41583_);
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.f_41583_); // CraftBukkit
                   this.f_39002_.m_6422_(0);
                   return;
                }
@@ -200,7 +_,7 @@
                }
 
                if (flag3 && !flag2) {
-                  this.f_39768_.m_6836_(0, ItemStack.f_41583_);
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.f_41583_); // CraftBukkit
                   this.f_39002_.m_6422_(0);
                   return;
                }
@@ -218,17 +_,18 @@
             i += k;
             itemstack1.m_41714_(new TextComponent(this.f_39001_));
          }
+         if (flag && !itemstack1.isBookEnchantable(itemstack2)) itemstack1 = ItemStack.f_41583_;
 
          this.f_39002_.m_6422_(j + i);
          if (i <= 0) {
             itemstack1 = ItemStack.f_41583_;
          }
 
-         if (k == i && k > 0 && this.f_39002_.m_6501_() >= 40) {
-            this.f_39002_.m_6422_(39);
+         if (k == i && k > 0 && this.f_39002_.m_6501_() >= maximumRepairCost) {
+            this.f_39002_.m_6422_(maximumRepairCost - 1);
          }
 
-         if (this.f_39002_.m_6501_() >= 40 && !this.f_39771_.m_150110_().f_35937_) {
+         if (this.f_39002_.m_6501_() >= maximumRepairCost && !this.f_39771_.m_150110_().f_35937_) {
             itemstack1 = ItemStack.f_41583_;
          }
 
@@ -246,7 +_,8 @@
             EnchantmentHelper.m_44865_(map, itemstack1);
          }
 
-         this.f_39768_.m_6836_(0, itemstack1);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), itemstack1); // CraftBukkit
+         m_150429_(); // CraftBukkit - SPIGOT-6686: Always send completed inventory to stay in sync with client
          this.m_38946_();
       }
    }
@@ -271,5 +_,9 @@
 
    public int m_39028_() {
       return this.f_39002_.m_6501_();
+   }
+
+   public void setMaximumCost(int value) {
+      this.f_39002_.m_6422_(value);
    }
 }
