--- a/net/minecraft/world/inventory/BeaconMenu.java
+++ b/net/minecraft/world/inventory/BeaconMenu.java
@@ -11,6 +_,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import net.minecraft.world.entity.player.Inventory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class BeaconMenu extends AbstractContainerMenu {
    private static final int f_150481_ = 0;
    private static final int f_150482_ = 1;
@@ -19,18 +_,14 @@
    private static final int f_150485_ = 28;
    private static final int f_150486_ = 28;
    private static final int f_150487_ = 37;
-   private final Container f_39031_ = new SimpleContainer(1) {
-      public boolean m_7013_(int p_39066_, ItemStack p_39067_) {
-         return p_39067_.m_204117_(ItemTags.f_13164_);
-      }
-
-      public int m_6893_() {
-         return 1;
-      }
-   };
+   private final Container f_39031_; // CatServer - moved into constructor
    private final BeaconMenu.PaymentSlot f_39032_;
    private final ContainerLevelAccess f_39033_;
    private final ContainerData f_39034_;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+   // CraftBukkit end
 
    public BeaconMenu(int p_39036_, Container p_39037_) {
       this(p_39036_, p_39037_, new SimpleContainerData(3), ContainerLevelAccess.f_39287_);
@@ -38,6 +_,20 @@
 
    public BeaconMenu(int p_39039_, Container p_39040_, ContainerData p_39041_, ContainerLevelAccess p_39042_) {
       super(MenuType.f_39965_, p_39039_);
+      player = (Inventory) p_39040_; // CraftBukkit - TODO: check this
+
+      // CatServer start
+      this.f_39031_ = new SimpleContainer(1) {
+         public boolean m_7013_(int p_39066_, ItemStack p_39067_) {
+            return p_39067_.m_204117_(ItemTags.f_13164_);
+         }
+
+         public int m_6893_() {
+            return 1;
+         }
+      }.setBukkitOwnerCreator(this.createBlockHolder(p_39042_)); // Paper
+      // CatServer end
+
       m_38886_(p_39041_, 3);
       this.f_39034_ = p_39041_;
       this.f_39033_ = p_39042_;
@@ -71,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39047_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39033_, p_39047_, Blocks.f_50273_);
    }
 
@@ -91,10 +_,8 @@
             }
 
             slot.m_40234_(itemstack1, itemstack);
-         } else if (!this.f_39032_.m_6657_() && this.f_39032_.m_5857_(itemstack1) && itemstack1.m_41613_() == 1) {
-            if (!this.m_38903_(itemstack1, 0, 1, false)) {
-               return ItemStack.f_41583_;
-            }
+         } else if (this.m_38903_(itemstack1, 0, 1, false)) { //Forge Fix Shift Clicking in beacons with stacks larger then 1.
+            return ItemStack.f_41583_;
          } else if (p_39052_ >= 1 && p_39052_ < 28) {
             if (!this.m_38903_(itemstack1, 28, 37, false)) {
                return ItemStack.f_41583_;
@@ -164,4 +_,16 @@
          return 1;
       }
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryBeacon(this.f_39031_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
