--- a/net/minecraft/world/inventory/BeaconMenu.java
+++ b/net/minecraft/world/inventory/BeaconMenu.java
@@ -5,10 +_,12 @@
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.effect.MobEffect;
+import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class BeaconMenu extends AbstractContainerMenu {
    private static final int f_150481_ = 0;
@@ -31,6 +_,22 @@
    private final ContainerLevelAccess f_39033_;
    private final ContainerData f_39034_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryBeacon(this.f_39031_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public BeaconMenu(int p_39036_, Container p_39037_) {
       this(p_39036_, p_39037_, new SimpleContainerData(3), ContainerLevelAccess.f_39287_);
    }
@@ -38,6 +_,7 @@
    public BeaconMenu(int p_39039_, Container p_39040_, ContainerData p_39041_, ContainerLevelAccess p_39042_) {
       super(MenuType.f_39965_, p_39039_);
       m_38886_(p_39041_, 3);
+      player = (Inventory) p_39040_; // CraftBukkit - TODO: check this
       this.f_39034_ = p_39041_;
       this.f_39033_ = p_39042_;
       this.f_39032_ = new BeaconMenu.PaymentSlot(this.f_39031_, 0, 136, 110);
@@ -70,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39047_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39033_, p_39047_, Blocks.f_50273_);
    }
 
@@ -90,10 +_,8 @@
             }
 
             slot.m_40234_(itemstack1, itemstack);
-         } else if (!this.f_39032_.m_6657_() && this.f_39032_.m_5857_(itemstack1) && itemstack1.m_41613_() == 1) {
-            if (!this.m_38903_(itemstack1, 0, 1, false)) {
-               return ItemStack.f_41583_;
-            }
+         } else if (this.m_38903_(itemstack1, 0, 1, false)) { //Forge Fix Shift Clicking in beacons with stacks larger then 1.
+            return ItemStack.f_41583_;
          } else if (p_39052_ >= 1 && p_39052_ < 28) {
             if (!this.m_38903_(itemstack1, 28, 37, false)) {
                return ItemStack.f_41583_;
