--- a/net/minecraft/world/inventory/BlastFurnaceMenu.java
+++ b/net/minecraft/world/inventory/BlastFurnaceMenu.java
@@ -3,6 +_,7 @@
 import net.minecraft.world.Container;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.item.crafting.RecipeType;
+import org.bukkit.inventory.InventoryView;
 
 public class BlastFurnaceMenu extends AbstractFurnaceMenu {
    public BlastFurnaceMenu(int p_39079_, Inventory p_39080_) {
