--- a/net/minecraft/world/inventory/CartographyTableMenu.java
+++ b/net/minecraft/world/inventory/CartographyTableMenu.java
@@ -11,6 +_,9 @@
 import net.minecraft.world.item.MapItem;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryCartography;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class CartographyTableMenu extends AbstractContainerMenu {
    public static final int f_150501_ = 0;
@@ -27,13 +_,39 @@
          CartographyTableMenu.this.m_6199_(this);
          super.m_6596_();
       }
+
+      @Override
+      public Location getLocation() {
+         return f_39135_.getLocation();
+      }
    };
    private final ResultContainer f_39138_ = new ResultContainer() {
       public void m_6596_() {
          CartographyTableMenu.this.m_6199_(this);
          super.m_6596_();
       }
+
+      @Override
+      public Location getLocation() {
+         return f_39138_.getLocation();
+      }
    };
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryCartography inventory = new CraftInventoryCartography(this.f_39135_, this.f_39138_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public CartographyTableMenu(int p_39140_, Inventory p_39141_) {
       this(p_39140_, p_39141_, ContainerLevelAccess.f_39287_);
@@ -82,10 +_,11 @@
       for(int k = 0; k < 9; ++k) {
          this.m_38897_(new Slot(p_39144_, k, 8 + k * 18, 142));
       }
-
+      player = (org.bukkit.entity.Player) p_39144_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public boolean m_6875_(Player p_39149_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39136_, p_39149_, Blocks.f_50621_);
    }
 
