--- a/net/minecraft/world/inventory/ContainerLevelAccess.java
+++ b/net/minecraft/world/inventory/ContainerLevelAccess.java
@@ -18,6 +_,18 @@
          public <T> Optional<T> m_6721_(BiFunction<Level, BlockPos, T> p_39311_) {
             return Optional.of(p_39311_.apply(p_39290_, p_39291_));
          }
+
+         // CraftBukkit start
+         @Override
+         public Level getWorld() {
+            return p_39290_;
+         }
+
+         @Override
+         public BlockPos getPosition() {
+            return p_39291_;
+         }
+         // CraftBukkit end
       };
    }
 
@@ -33,4 +_,18 @@
          return Optional.empty();
       });
    }
+
+   // CraftBukkit start
+   default Level getWorld() {
+      throw new UnsupportedOperationException("Not supported yet.");
+   }
+
+   default BlockPos getPosition() {
+      throw new UnsupportedOperationException("Not supported yet.");
+   }
+
+   default org.bukkit.Location getLocation() {
+      return new org.bukkit.Location(getWorld().getWorld(), getPosition().m_123341_(), getPosition().m_123342_(), getPosition().m_123343_());
+   }
+   // CraftBukkit end
 }
