--- a/net/minecraft/world/inventory/CraftingMenu.java
+++ b/net/minecraft/world/inventory/CraftingMenu.java
@@ -11,8 +_,13 @@
 import net.minecraft.world.item.crafting.CraftingRecipe;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.crafting.RecipeType;
+import net.minecraft.world.item.crafting.RepairItemRecipe;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class CraftingMenu extends RecipeBookMenu<CraftingContainer> {
    public static final int f_150539_ = 0;
@@ -22,10 +_,11 @@
    private static final int f_150543_ = 37;
    private static final int f_150544_ = 37;
    private static final int f_150545_ = 46;
-   private final CraftingContainer f_39348_ = new CraftingContainer(this, 3, 3);
-   private final ResultContainer f_39349_ = new ResultContainer();
-   private final ContainerLevelAccess f_39350_;
+   public final CraftingContainer f_39348_ = new CraftingContainer(this, 3, 3);
+   public final ResultContainer f_39349_ = new ResultContainer();
+   final ContainerLevelAccess f_39350_;
    private final Player f_39351_;
+   private CraftInventoryView bukkitEntity = null;
 
    public CraftingMenu(int p_39353_, Inventory p_39354_) {
       this(p_39353_, p_39354_, ContainerLevelAccess.f_39287_);
@@ -35,6 +_,10 @@
       super(MenuType.f_39968_, p_39356_);
       this.f_39350_ = p_39358_;
       this.f_39351_ = p_39357_.f_35978_;
+      // CraftBukkit start - Switched order of IInventory construction and stored player
+      this.f_39348_.setOwner(this.f_39351_);
+      this.f_39348_.resultInventory = this.f_39349_;
+      // CraftBukkit end
       this.m_38897_(new ResultSlot(p_39357_.f_35978_, this.f_39348_, this.f_39349_, 0, 124, 35));
 
       for(int i = 0; i < 3; ++i) {
@@ -66,6 +_,8 @@
                itemstack = craftingrecipe.m_5874_(p_150550_);
             }
          }
+         if (p_150547_.getBukkitView() != null) // CatServer - Bypass call event if Bukkit view is null (mod custom container)
+         itemstack = CraftEventFactory.callPreCraftEvent(p_150550_, p_150551_, itemstack, p_150547_.getBukkitView(), optional.orElse(null) instanceof RepairItemRecipe); // CraftBukkit
 
          p_150551_.m_6836_(0, itemstack);
          p_150547_.m_150404_(0, itemstack);
@@ -100,7 +_,18 @@
    }
 
    public boolean m_6875_(Player p_39368_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39350_, p_39368_, Blocks.f_50091_);
+   }
+
+   @Override
+   public InventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.f_39348_, this.f_39349_);
+      bukkitEntity = new CraftInventoryView(this.f_39351_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
    }
 
    public ItemStack m_7648_(Player p_39391_, int p_39392_) {
