--- a/net/minecraft/world/inventory/CraftingContainer.java
+++ b/net/minecraft/world/inventory/CraftingContainer.java
@@ -6,12 +_,86 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.player.StackedContents;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Recipe;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.inventory.InventoryType;
+
+import java.util.List;
 
 public class CraftingContainer implements Container, StackedContentsCompatible {
    private final NonNullList<ItemStack> f_39320_;
    private final int f_39321_;
    private final int f_39322_;
    private final AbstractContainerMenu f_39323_;
+
+   // CraftBukkit start - add fields
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private Recipe currentRecipe;
+   public Container resultInventory;
+   public Player owner;
+   private int maxStack = MAX_STACK;
+
+   public List<ItemStack> getContents() {
+      return this.f_39320_;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public InventoryType getInvType() {
+      return f_39320_.size() == 4 ? InventoryType.CRAFTING : InventoryType.WORKBENCH;
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (owner == null) ? null : owner.getBukkitEntity();
+   }
+
+   public void setOwner(Player owner) {
+      this.owner = owner;
+   }
+
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+      resultInventory.setMaxStackSize(size);
+   }
+
+   @Override
+   public Location getLocation() {
+      return f_39323_ instanceof CraftingMenu ? ((CraftingMenu) f_39323_).f_39350_.getLocation() : owner.getBukkitEntity().getLocation();
+   }
+
+   @Override
+   public Recipe getCurrentRecipe() {
+      return currentRecipe;
+   }
+
+   @Override
+   public void setCurrentRecipe(Recipe currentRecipe) {
+      this.currentRecipe = currentRecipe;
+   }
+
+   public CraftingContainer(AbstractContainerMenu container, int i, int j, Player player) {
+      this(container, i, j);
+      this.owner = player;
+   }
+   // CraftBukkit end
 
    public CraftingContainer(AbstractContainerMenu p_39325_, int p_39326_, int p_39327_) {
       this.f_39320_ = NonNullList.m_122780_(p_39326_ * p_39327_, ItemStack.f_41583_);
