--- a/net/minecraft/world/inventory/CraftingMenu.java
+++ b/net/minecraft/world/inventory/CraftingMenu.java
@@ -14,6 +_,12 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import net.minecraft.world.item.crafting.RepairItemRecipe;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class CraftingMenu extends RecipeBookMenu<CraftingContainer> {
    public static final int f_150539_ = 0;
    private static final int f_150540_ = 1;
@@ -22,10 +_,13 @@
    private static final int f_150543_ = 37;
    private static final int f_150544_ = 37;
    private static final int f_150545_ = 46;
-   private final CraftingContainer f_39348_ = new TransientCraftingContainer(this, 3, 3);
-   private final ResultContainer f_39349_ = new ResultContainer();
-   private final ContainerLevelAccess f_39350_;
+   public final CraftingContainer f_39348_ = new TransientCraftingContainer(this, 3, 3); // CraftBukkit - 1.20.1 InventoryCrafting -> TransientCraftingContainer // CatServer - recover // CatServer - private -> public
+   public final ResultContainer f_39349_ = new ResultContainer(); // CatServer - private -> public
+   public final ContainerLevelAccess f_39350_; // CatServer - private -> public
    private final Player f_39351_;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   // CraftBukkit end
 
    public CraftingMenu(int p_39353_, Inventory p_39354_) {
       this(p_39353_, p_39354_, ContainerLevelAccess.f_39287_);
@@ -35,6 +_,10 @@
       super(MenuType.f_39968_, p_39356_);
       this.f_39350_ = p_39358_;
       this.f_39351_ = p_39357_.f_35978_;
+      // CraftBukkit start - Switched order of IInventory construction and stored player
+      ((TransientCraftingContainer) this.f_39348_).setOwner(this.f_39351_); // CraftBukkit - pass player
+      ((TransientCraftingContainer) this.f_39348_).resultInventory = this.f_39349_;
+      // CraftBukkit end
       this.m_38897_(new ResultSlot(p_39357_.f_35978_, this.f_39348_, this.f_39349_, 0, 124, 35));
 
       for(int i = 0; i < 3; ++i) {
@@ -69,6 +_,8 @@
                }
             }
          }
+         if (p_150547_.getBukkitView() != null) // CatServer - Bypass call event if Bukkit view is null (mod custom container)
+            itemstack = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPreCraftEvent(p_150550_, p_150551_, itemstack, p_150547_.getBukkitView(), optional.orElse(null) instanceof RepairItemRecipe); // CraftBukkit
 
          p_150551_.m_6836_(0, itemstack);
          p_150547_.m_150404_(0, itemstack);
@@ -103,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39368_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39350_, p_39368_, Blocks.f_50091_);
    }
 
@@ -181,4 +_,16 @@
    public boolean m_142157_(int p_150553_) {
       return p_150553_ != this.m_6636_();
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.f_39348_, this.f_39349_);
+      bukkitEntity = new CraftInventoryView(this.f_39351_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
