--- a/net/minecraft/world/inventory/DispenserMenu.java
+++ b/net/minecraft/world/inventory/DispenserMenu.java
@@ -5,6 +_,8 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class DispenserMenu extends AbstractContainerMenu {
    private static final int f_150557_ = 9;
@@ -14,6 +_,22 @@
    private static final int f_150561_ = 45;
    private final Container f_39431_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventory inventory = new CraftInventory(this.f_39431_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public DispenserMenu(int p_39433_, Inventory p_39434_) {
       this(p_39433_, p_39434_, new SimpleContainer(9));
    }
@@ -22,6 +_,7 @@
       super(MenuType.f_39963_, p_39436_);
       m_38869_(p_39438_, 9);
       this.f_39431_ = p_39438_;
+      this.player = p_39437_;
       p_39438_.m_5856_(p_39437_.f_35978_);
 
       for(int i = 0; i < 3; ++i) {
@@ -43,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39440_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_39431_.m_6542_(p_39440_);
    }
 
