--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.inventory;
 
 import java.util.List;
+import java.util.Map;
 import java.util.Random;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -22,6 +_,17 @@
 import net.minecraft.world.item.enchantment.EnchantmentInstance;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.EnchantmentTableBlock;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryEnchanting;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.enchantments.EnchantmentOffer;
+import org.bukkit.event.enchantment.EnchantItemEvent;
+import org.bukkit.event.enchantment.PrepareItemEnchantEvent;
 
 public class EnchantmentMenu extends AbstractContainerMenu {
    private final Container f_39449_ = new SimpleContainer(2) {
@@ -29,6 +_,11 @@
          super.m_6596_();
          EnchantmentMenu.this.m_6199_(this);
       }
+
+      @Override
+      public Location getLocation() {
+         return f_39449_.getLocation();
+      }
    };
    private final ContainerLevelAccess f_39450_;
    private final Random f_39451_ = new Random();
@@ -37,6 +_,23 @@
    public final int[] f_39447_ = new int[]{-1, -1, -1};
    public final int[] f_39448_ = new int[]{-1, -1, -1};
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryEnchanting inventory = new CraftInventoryEnchanting(this.f_39449_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
+
    public EnchantmentMenu(int p_39454_, Inventory p_39455_) {
       this(p_39454_, p_39455_, ContainerLevelAccess.f_39287_);
    }
@@ -55,7 +_,7 @@
       });
       this.m_38897_(new Slot(this.f_39449_, 1, 35, 47) {
          public boolean m_5857_(ItemStack p_39517_) {
-            return p_39517_.m_150930_(Items.f_42534_);
+            return p_39517_.m_204117_(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS);
          }
       });
 
@@ -79,30 +_,38 @@
       this.m_38895_(DataSlot.m_39406_(this.f_39448_, 0));
       this.m_38895_(DataSlot.m_39406_(this.f_39448_, 1));
       this.m_38895_(DataSlot.m_39406_(this.f_39448_, 2));
+      // CraftBukkit start
+      player = (org.bukkit.entity.Player) p_39458_.f_35978_.getBukkitEntity();
+      // CraftBukkit end
+   }
+
+   private float getPower(net.minecraft.world.level.Level world, net.minecraft.core.BlockPos pos) {
+      return world.m_8055_(pos).getEnchantPowerBonus(world, pos);
    }
 
    public void m_6199_(Container p_39461_) {
       if (p_39461_ == this.f_39449_) {
          ItemStack itemstack = p_39461_.m_8020_(0);
-         if (!itemstack.m_41619_() && itemstack.m_41792_()) {
+         if (!itemstack.m_41619_()) { // CraftBukkit - relax condition
             this.f_39450_.m_39292_((p_39485_, p_39486_) -> {
-               int j = 0;
+               float j = 0;
 
                for(BlockPos blockpos : EnchantmentTableBlock.f_207902_) {
                   if (EnchantmentTableBlock.m_207909_(p_39485_, p_39486_, blockpos)) {
-                     ++j;
+                     j += p_39485_.m_8055_(p_39486_.m_141952_(blockpos)).getEnchantPowerBonus(p_39485_, p_39486_.m_141952_(blockpos));
                   }
                }
 
                this.f_39451_.setSeed((long)this.f_39452_.m_6501_());
 
                for(int k = 0; k < 3; ++k) {
-                  this.f_39446_[k] = EnchantmentHelper.m_44872_(this.f_39451_, k, j, itemstack);
+                  this.f_39446_[k] = EnchantmentHelper.m_44872_(this.f_39451_, k, (int)j, itemstack);
                   this.f_39447_[k] = -1;
                   this.f_39448_[k] = -1;
                   if (this.f_39446_[k] < k + 1) {
                      this.f_39446_[k] = 0;
                   }
+                  this.f_39446_[k] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(p_39485_, p_39486_, k, (int)j, itemstack, f_39446_[k]);
                }
 
                for(int l = 0; l < 3; ++l) {
@@ -218,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39463_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39450_, p_39463_, Blocks.f_50201_);
    }
 
@@ -235,7 +_,7 @@
             if (!this.m_38903_(itemstack1, 2, 38, true)) {
                return ItemStack.f_41583_;
             }
-         } else if (itemstack1.m_150930_(Items.f_42534_)) {
+         } else if (itemstack1.m_204117_(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS)) {
             if (!this.m_38903_(itemstack1, 1, 2, true)) {
                return ItemStack.f_41583_;
             }
