--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -40,9 +_,10 @@
    protected void m_5845_(ItemStack p_39558_) {
       p_39558_.m_41678_(this.f_39539_.f_19853_, this.f_39539_, this.f_39540_);
       if (this.f_39539_ instanceof ServerPlayer && this.f_40218_ instanceof AbstractFurnaceBlockEntity) {
-         ((AbstractFurnaceBlockEntity)this.f_40218_).m_155003_((ServerPlayer)this.f_39539_);
+         ((AbstractFurnaceBlockEntity)this.f_40218_).awardUsedRecipesAndPopExperience((ServerPlayer)this.f_39539_, p_39558_, this.f_39540_); // CraftBukkit
       }
 
       this.f_39540_ = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.f_39539_, p_39558_);
    }
 }
