--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -17,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryGrindstone;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
    public static final int f_150565_ = 35;
@@ -33,8 +_,18 @@
          super.m_6596_();
          GrindstoneMenu.this.m_6199_(this);
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return f_39561_.getLocation();
+      }
+      // CraftBukkit end
    };
    private final ContainerLevelAccess f_39561_;
+   private int xp = -1;
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
 
    public GrindstoneMenu(int p_39563_, Inventory p_39564_) {
       this(p_39563_, p_39564_, ContainerLevelAccess.f_39287_);
@@ -45,12 +_,12 @@
       this.f_39561_ = p_39568_;
       this.m_38897_(new Slot(this.f_39560_, 0, 49, 19) {
          public boolean m_5857_(ItemStack p_39607_) {
-            return p_39607_.m_41763_() || p_39607_.m_150930_(Items.f_42690_) || p_39607_.m_41793_();
+            return true; //Allow all items in the slot, not just repairable
          }
       });
       this.m_38897_(new Slot(this.f_39560_, 1, 49, 40) {
          public boolean m_5857_(ItemStack p_39616_) {
-            return p_39616_.m_41763_() || p_39616_.m_150930_(Items.f_42690_) || p_39616_.m_41793_();
+            return true; //Allow all items in the slot, not just repairable
          }
       });
       this.m_38897_(new Slot(this.f_39559_, 2, 129, 34) {
@@ -59,6 +_,7 @@
          }
 
          public void m_142406_(Player p_150574_, ItemStack p_150575_) {
+            if (net.minecraftforge.common.ForgeHooks.onGrindstoneTake(GrindstoneMenu.this.f_39560_, p_39568_, this::m_39631_)) return;
             p_39568_.m_39292_((p_39634_, p_39635_) -> {
                if (p_39634_ instanceof ServerLevel) {
                   ExperienceOrb.m_147082_((ServerLevel)p_39634_, Vec3.m_82512_(p_39635_), this.m_39631_(p_39634_));
@@ -71,6 +_,7 @@
          }
 
          private int m_39631_(Level p_39632_) {
+            if (xp > -1) return xp;
             int l = 0;
             l += this.m_39636_(GrindstoneMenu.this.f_39560_.m_8020_(0));
             l += this.m_39636_(GrindstoneMenu.this.f_39560_.m_8020_(1));
@@ -107,7 +_,7 @@
       for(int k = 0; k < 9; ++k) {
          this.m_38897_(new Slot(p_39567_, k, 8 + k * 18, 142));
       }
-
+      player = (org.bukkit.entity.Player) p_39567_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public void m_6199_(Container p_39570_) {
@@ -123,6 +_,8 @@
       ItemStack itemstack1 = this.f_39560_.m_8020_(1);
       boolean flag = !itemstack.m_41619_() || !itemstack1.m_41619_();
       boolean flag1 = !itemstack.m_41619_() && !itemstack1.m_41619_();
+      this.xp = net.minecraftforge.common.ForgeHooks.onGrindstoneChange(itemstack, itemstack1, this.f_39559_, -1);
+      if (this.xp == Integer.MIN_VALUE)
       if (!flag) {
          this.f_39559_.m_6836_(0, ItemStack.f_41583_);
       } else {
@@ -144,12 +_,13 @@
             }
 
             Item item = itemstack.m_41720_();
-            int k = item.m_41462_() - itemstack.m_41773_();
-            int l = item.m_41462_() - itemstack1.m_41773_();
-            int i1 = k + l + item.m_41462_() * 5 / 100;
-            i = Math.max(item.m_41462_() - i1, 0);
+            int k = itemstack.m_41776_() - itemstack.m_41773_();
+            int l = itemstack.m_41776_() - itemstack1.m_41773_();
+            int i1 = k + l + itemstack.m_41776_() * 5 / 100;
+            i = Math.max(itemstack.m_41776_() - i1, 0);
             itemstack2 = this.m_39590_(itemstack, itemstack1);
-            if (!itemstack2.m_41763_()) {
+            if (!itemstack2.isRepairable()) i = itemstack.m_41773_();
+            if (!itemstack2.m_41763_() || !itemstack2.isRepairable()) {
                if (!ItemStack.m_41728_(itemstack, itemstack1)) {
                   this.f_39559_.m_6836_(0, ItemStack.f_41583_);
                   this.m_38946_();
@@ -164,6 +_,10 @@
             itemstack2 = flag3 ? itemstack : itemstack1;
          }
 
+         // Forge: Skip the repair if the result would give an item stack with a count not normally obtainable
+         if (j > itemstack2.m_41741_())
+            this.f_39559_.m_6836_(0, ItemStack.f_41583_);
+         else
          this.f_39559_.m_6836_(0, this.m_39579_(itemstack2, i, j));
       }
 
@@ -223,6 +_,16 @@
 
    public boolean m_6875_(Player p_39572_) {
       return m_38889_(this.f_39561_, p_39572_, Blocks.f_50623_);
+   }
+
+   @Override
+   public InventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryGrindstone inventory = new CraftInventoryGrindstone(this.f_39560_, this.f_39559_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
    }
 
    public ItemStack m_7648_(Player p_39588_, int p_39589_) {
