--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -17,6 +_,9 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryGrindstone;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
    public static final int f_150565_ = 35;
@@ -33,9 +_,30 @@
          super.m_6596_();
          GrindstoneMenu.this.m_6199_(this);
       }
+
+      @Override
+      public Location getLocation() {
+         return f_39560_.getLocation();
+      }
    };
    private final ContainerLevelAccess f_39561_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryGrindstone inventory = new CraftInventoryGrindstone(this.f_39560_, this.f_39559_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public GrindstoneMenu(int p_39563_, Inventory p_39564_) {
       this(p_39563_, p_39564_, ContainerLevelAccess.f_39287_);
    }
@@ -107,7 +_,7 @@
       for(int k = 0; k < 9; ++k) {
          this.m_38897_(new Slot(p_39567_, k, 8 + k * 18, 142));
       }
-
+      player = (org.bukkit.entity.Player) p_39567_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public void m_6199_(Container p_39570_) {
@@ -144,12 +_,13 @@
             }
 
             Item item = itemstack.m_41720_();
-            int k = item.m_41462_() - itemstack.m_41773_();
-            int l = item.m_41462_() - itemstack1.m_41773_();
-            int i1 = k + l + item.m_41462_() * 5 / 100;
-            i = Math.max(item.m_41462_() - i1, 0);
+            int k = itemstack.m_41776_() - itemstack.m_41773_();
+            int l = itemstack.m_41776_() - itemstack1.m_41773_();
+            int i1 = k + l + itemstack.m_41776_() * 5 / 100;
+            i = Math.max(itemstack.m_41776_() - i1, 0);
             itemstack2 = this.m_39590_(itemstack, itemstack1);
-            if (!itemstack2.m_41763_()) {
+            if (!itemstack2.isRepairable()) i = itemstack.m_41773_();
+            if (!itemstack2.m_41763_() || !itemstack2.isRepairable()) {
                if (!ItemStack.m_41728_(itemstack, itemstack1)) {
                   this.f_39559_.m_6836_(0, ItemStack.f_41583_);
                   this.m_38946_();
@@ -222,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39572_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39561_, p_39572_, Blocks.f_50623_);
    }
 
