--- a/net/minecraft/world/inventory/HopperMenu.java
+++ b/net/minecraft/world/inventory/HopperMenu.java
@@ -5,11 +_,29 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class HopperMenu extends AbstractContainerMenu {
    public static final int f_150576_ = 5;
    private final Container f_39638_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventory inventory = new CraftInventory(this.f_39638_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public HopperMenu(int p_39640_, Inventory p_39641_) {
       this(p_39640_, p_39641_, new SimpleContainer(5));
    }
@@ -17,6 +_,7 @@
    public HopperMenu(int p_39643_, Inventory p_39644_, Container p_39645_) {
       super(MenuType.f_39972_, p_39643_);
       this.f_39638_ = p_39645_;
+      this.player = p_39644_; // CraftBukkit - save player
       m_38869_(p_39645_, 5);
       p_39645_.m_5856_(p_39644_.f_35978_);
       int i = 51;
@@ -38,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39647_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_39638_.m_6542_(p_39647_);
    }
 
