--- a/net/minecraft/world/inventory/HorseInventoryMenu.java
+++ b/net/minecraft/world/inventory/HorseInventoryMenu.java
@@ -7,13 +_,30 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class HorseInventoryMenu extends AbstractContainerMenu {
    private final Container f_39653_;
    private final AbstractHorse f_39654_;
 
+   // CraftBukkit start
+   org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView bukkitEntity;
+   Inventory player;
+
+   @Override
+   public InventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      return bukkitEntity = new CraftInventoryView(player.f_35978_.getBukkitEntity(), f_39653_.getOwner().getInventory(), this);
+   }
+
    public HorseInventoryMenu(int p_39656_, Inventory p_39657_, Container p_39658_, final AbstractHorse p_39659_) {
       super((MenuType<?>)null, p_39656_);
+      player = p_39657_;
+      // CraftBukkit end
       this.f_39653_ = p_39658_;
       this.f_39654_ = p_39659_;
       int i = 3;
