--- a/net/minecraft/world/inventory/HorseInventoryMenu.java
+++ b/net/minecraft/world/inventory/HorseInventoryMenu.java
@@ -8,12 +_,30 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
+// CraftBukkit end
+
 public class HorseInventoryMenu extends AbstractContainerMenu {
    private final Container f_39653_;
    private final AbstractHorse f_39654_;
+   // CraftBukkit start
+   org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView bukkitEntity;
+   Inventory player;
+
+   @Override
+   public InventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      return bukkitEntity = new CraftInventoryView(player.f_35978_.getBukkitEntity(), f_39653_.getOwner().getInventory(), this);
+   }
 
    public HorseInventoryMenu(int p_39656_, Inventory p_39657_, Container p_39658_, final AbstractHorse p_39659_) {
       super((MenuType<?>)null, p_39656_);
+      player = p_39657_;
+      // CraftBukkit end
       this.f_39653_ = p_39658_;
       this.f_39654_ = p_39659_;
       int i = 3;
@@ -62,7 +_,11 @@
    }
 
    public boolean m_6875_(Player p_39661_) {
-      return !this.f_39654_.m_149511_(this.f_39653_) && this.f_39653_.m_6542_(p_39661_) && this.f_39654_.m_6084_() && this.f_39654_.m_20270_(p_39661_) < 8.0F;
+      // Make pehkui's mixin happy so we stop getting crash reports. It requires the magic float 8.0F
+      float magic_number = 8.0F;
+      // Default reach is 3 vanilla uses 8 here so add 5 padding
+      double reach = p_39661_.m_21133_(net.minecraftforge.common.ForgeMod.ENTITY_REACH.get()) + 5;
+      return !this.f_39654_.m_149511_(this.f_39653_) && this.f_39653_.m_6542_(p_39661_) && this.f_39654_.m_6084_() && this.f_39654_.m_20270_(p_39661_) < reach;
    }
 
    private boolean m_150577_(AbstractHorse p_150578_) {
