--- a/net/minecraft/world/inventory/InventoryMenu.java
+++ b/net/minecraft/world/inventory/InventoryMenu.java
@@ -11,6 +_,8 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class InventoryMenu extends RecipeBookMenu<CraftingContainer> {
    public static final int f_150579_ = 0;
@@ -37,10 +_,28 @@
    public final boolean f_39698_;
    private final Player f_39703_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.f_39701_, this.f_39702_);
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public InventoryMenu(Inventory p_39706_, boolean p_39707_, Player p_39708_) {
       super((MenuType<?>)null, 0);
       this.f_39698_ = p_39707_;
       this.f_39703_ = p_39708_;
+      this.f_39701_.resultInventory = this.f_39702_; // CraftBukkit - let InventoryCrafting know about its result slot
+      this.f_39701_.setOwner(p_39706_.f_35978_);
+      this.player = p_39706_; // CraftBukkit - save player
       this.m_38897_(new ResultSlot(p_39706_.f_35978_, this.f_39701_, this.f_39702_, 0, 154, 28));
 
       for(int i = 0; i < 2; ++i) {
@@ -57,7 +_,7 @@
             }
 
             public boolean m_5857_(ItemStack p_39746_) {
-               return equipmentslot == Mob.m_147233_(p_39746_);
+               return p_39746_.canEquip(equipmentslot, f_39703_);
             }
 
             public boolean m_8010_(Player p_39744_) {
