--- a/net/minecraft/world/inventory/ItemCombinerMenu.java
+++ b/net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -82,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39780_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_39770_.m_39299_((p_39785_, p_39786_) -> {
          return !this.m_8039_(p_39785_.m_8055_(p_39786_)) ? false : p_39780_.m_20275_((double)p_39786_.m_123341_() + 0.5D, (double)p_39786_.m_123342_() + 0.5D, (double)p_39786_.m_123343_() + 0.5D) <= 64.0D;
       }, true);
