--- a/net/minecraft/world/inventory/ItemCombinerMenu.java
+++ b/net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -16,7 +_,7 @@
    protected final Player f_39771_;
    protected final Container f_39769_;
    private final List<Integer> f_266110_;
-   protected final ResultContainer f_39768_ = new ResultContainer();
+   protected final ResultContainer f_39768_; // Paper - delay field init
    private final int f_266048_;
 
    protected abstract boolean m_6560_(Player p_39798_, boolean p_39799_);
@@ -28,6 +_,7 @@
    public ItemCombinerMenu(@Nullable MenuType<?> p_39773_, int p_39774_, Inventory p_39775_, ContainerLevelAccess p_39776_) {
       super(p_39773_, p_39774_);
       this.f_39770_ = p_39776_;
+      this.f_39768_ = new ResultContainer().setHolderCreator(this.createBlockHolder(this.f_39770_)); // Paper - delay field init // CatServer
       this.f_39771_ = p_39775_.f_35978_;
       ItemCombinerMenuSlotDefinition itemcombinermenuslotdefinition = this.m_266183_();
       this.f_39769_ = this.m_266190_(itemcombinermenuslotdefinition.m_266578_());
@@ -88,7 +_,7 @@
             super.m_6596_();
             ItemCombinerMenu.this.m_6199_(this);
          }
-      };
+      }.setBukkitOwnerCreator(this.createBlockHolder(this.f_39770_)); // Paper
    }
 
    public void m_6199_(Container p_39778_) {
@@ -107,8 +_,13 @@
    }
 
    public boolean m_6875_(Player p_39780_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_39770_.m_39299_((p_39785_, p_39786_) -> {
-         return !this.m_8039_(p_39785_.m_8055_(p_39786_)) ? false : p_39780_.m_20275_((double)p_39786_.m_123341_() + 0.5D, (double)p_39786_.m_123342_() + 0.5D, (double)p_39786_.m_123343_() + 0.5D) <= 64.0D;
+         // Vanilla uses 64 which is 8*8. Default reach is 4.5 so add 3.5 padding
+         // magic number so required Mixins can still target here
+         double reachOld = 64.0;
+         double reach = p_39780_.m_21133_(net.minecraftforge.common.ForgeMod.BLOCK_REACH.get()) + 3.5;
+         return !this.m_8039_(p_39785_.m_8055_(p_39786_)) ? false : p_39780_.m_20275_((double)p_39786_.m_123341_() + 0.5D, (double)p_39786_.m_123342_() + 0.5D, (double)p_39786_.m_123343_() + 0.5D) <= reach * reach;
       }, true);
    }
 
