--- a/net/minecraft/world/inventory/LecternMenu.java
+++ b/net/minecraft/world/inventory/LecternMenu.java
@@ -2,8 +_,13 @@
 
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
+import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryLectern;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
+import org.bukkit.event.player.PlayerTakeLecternBookEvent;
 
 public class LecternMenu extends AbstractContainerMenu {
    private static final int f_150610_ = 1;
@@ -15,11 +_,27 @@
    private final Container f_39819_;
    private final ContainerData f_39820_;
 
-   public LecternMenu(int p_39822_) {
-      this(p_39822_, new SimpleContainer(1), new SimpleContainerData(1));
-   }
-
-   public LecternMenu(int p_39824_, Container p_39825_, ContainerData p_39826_) {
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryLectern inventory = new CraftInventoryLectern(this.f_39819_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
+   public LecternMenu(int p_39822_, Inventory playerinventory) {
+      this(p_39822_, new SimpleContainer(1), new SimpleContainerData(1), playerinventory);
+   }
+
+   public LecternMenu(int p_39824_, Container p_39825_, ContainerData p_39826_, Inventory playerinventory) {
       super(MenuType.f_39973_, p_39824_);
       m_38869_(p_39825_, 1);
       m_38886_(p_39826_, 1);
@@ -32,6 +_,7 @@
          }
       });
       this.m_38884_(p_39826_);
+      player = (org.bukkit.entity.Player) playerinventory.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public boolean m_6366_(Player p_39833_, int p_39834_) {
@@ -53,7 +_,13 @@
             if (!p_39833_.m_36326_()) {
                return false;
             }
-
+            // CraftBukkit start - Event for taking the book
+            PlayerTakeLecternBookEvent event = new PlayerTakeLecternBookEvent(player, ((CraftInventoryLectern) getBukkitView().getTopInventory()).getHolder());
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return false;
+                                }
+                                // CraftBukkit end
             ItemStack itemstack = this.f_39819_.m_8016_(0);
             this.f_39819_.m_6596_();
             if (!p_39833_.m_150109_().m_36054_(itemstack)) {
@@ -73,6 +_,8 @@
    }
 
    public boolean m_6875_(Player p_39831_) {
+      // if (lectern instanceof LecternBlockEntity.LecternInventory && !((LecternBlockEntity.LecternInventory) lectern).getLectern().hasBook()) return false; // CraftBukkit FoxServer TODO
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.f_39819_.m_6542_(p_39831_);
    }
 
