--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -24,7 +_,27 @@
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.entity.BlockEntityType;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class LoomMenu extends AbstractContainerMenu {
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryLoom inventory = new CraftInventoryLoom(this.f_39853_, this.f_39854_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
    private static final int f_219989_ = -1;
    private static final int f_150612_ = 4;
    private static final int f_150613_ = 31;
@@ -40,19 +_,8 @@
    private final Slot f_39850_;
    private final Slot f_39851_;
    long f_39852_;
-   private final Container f_39853_ = new SimpleContainer(3) {
-      public void m_6596_() {
-         super.m_6596_();
-         LoomMenu.this.m_6199_(this);
-         LoomMenu.this.f_39847_.run();
-      }
-   };
-   private final Container f_39854_ = new SimpleContainer(1) {
-      public void m_6596_() {
-         super.m_6596_();
-         LoomMenu.this.f_39847_.run();
-      }
-   };
+   private final Container f_39853_; // CatServer - moved into constructor
+   private final Container f_39854_; // CatServer - moved into constructor
 
    public LoomMenu(int p_39856_, Inventory p_39857_) {
       this(p_39856_, p_39857_, ContainerLevelAccess.f_39287_);
@@ -60,6 +_,37 @@
 
    public LoomMenu(int p_39859_, Inventory p_39860_, final ContainerLevelAccess p_39861_) {
       super(MenuType.f_39974_, p_39859_);
+
+      // CatServer start
+      this.f_39853_ = new SimpleContainer(3) {
+         public void m_6596_() {
+            super.m_6596_();
+            LoomMenu.this.m_6199_(this);
+            LoomMenu.this.f_39847_.run();
+         }
+
+         // CraftBukkit start
+         @Override
+         public Location getLocation() {
+            return f_39845_.getLocation();
+         }
+         // CraftBukkit end
+      }.setBukkitOwnerCreator(this.createBlockHolder(p_39861_)); // Paper
+      this.f_39854_ = new SimpleContainer(1) {
+         public void m_6596_() {
+            super.m_6596_();
+            LoomMenu.this.f_39847_.run();
+         }
+
+         // CraftBukkit start
+         @Override
+         public Location getLocation() {
+            return f_39845_.getLocation();
+         }
+         // CraftBukkit end
+      }.setBukkitOwnerCreator(this.createBlockHolder(p_39861_)); // Paper
+      // CatServer end
+
       this.f_39845_ = p_39861_;
       this.f_39848_ = this.m_38897_(new Slot(this.f_39853_, 0, 13, 26) {
          public boolean m_5857_(ItemStack p_39918_) {
@@ -111,9 +_,11 @@
       }
 
       this.m_38895_(this.f_39846_);
+      player = (org.bukkit.entity.Player) p_39860_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public boolean m_6875_(Player p_39865_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39845_, p_39865_, Blocks.f_50617_);
    }
 
@@ -276,6 +_,11 @@
          ListTag listtag;
          if (compoundtag != null && compoundtag.m_128425_("Patterns", 9)) {
             listtag = compoundtag.m_128437_("Patterns", 10);
+            // CraftBukkit start
+            while (listtag.size() > 20) {
+               listtag.remove(20);
+            }
+            // CraftBukkit end
          } else {
             listtag = new ListTag();
             if (compoundtag == null) {
