--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -17,6 +_,9 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.entity.BlockEntityType;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class LoomMenu extends AbstractContainerMenu {
    private static final int f_150612_ = 4;
@@ -38,13 +_,43 @@
          LoomMenu.this.m_6199_(this);
          LoomMenu.this.f_39847_.run();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return f_39853_.getLocation();
+      }
+      // CraftBukkit end
    };
    private final Container f_39854_ = new SimpleContainer(1) {
       public void m_6596_() {
          super.m_6596_();
          LoomMenu.this.f_39847_.run();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return f_39854_.getLocation();
+      }
+      // CraftBukkit end
    };
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryLoom inventory = new CraftInventoryLoom(this.f_39853_, this.f_39854_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public LoomMenu(int p_39856_, Inventory p_39857_) {
       this(p_39856_, p_39857_, ContainerLevelAccess.f_39287_);
@@ -103,6 +_,7 @@
       }
 
       this.m_38895_(this.f_39846_);
+      player = (org.bukkit.entity.Player) p_39860_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public int m_39891_() {
@@ -110,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_39865_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39845_, p_39865_, Blocks.f_50617_);
    }
 
@@ -224,6 +_,12 @@
             ListTag listtag;
             if (compoundtag != null && compoundtag.m_128425_("Patterns", 9)) {
                listtag = compoundtag.m_128437_("Patterns", 10);
+               /*/ CraftBukkit start // FoxServer TODO
+               while (listtag.size() > 20) {
+                  listtag.remove(20);
+               }
+               // CraftBukkit end
+               */
             } else {
                listtag = new ListTag();
                if (compoundtag == null) {
