--- a/net/minecraft/world/inventory/MerchantContainer.java
+++ b/net/minecraft/world/inventory/MerchantContainer.java
@@ -4,11 +_,18 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.world.Container;
 import net.minecraft.world.ContainerHelper;
+import net.minecraft.world.entity.npc.AbstractVillager;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftAbstractVillager;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+
+import java.util.List;
 
 public class MerchantContainer implements Container {
    private final Merchant f_39997_;
@@ -17,6 +_,46 @@
    private MerchantOffer f_39999_;
    public int f_40000_;
    private int f_40001_;
+
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private int maxStack = MAX_STACK;
+
+   public List<ItemStack> getContents() {
+      return this.f_39998_;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+      f_39997_.m_7189_((Player) null); // SPIGOT-4860
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int i) {
+      maxStack = i;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (f_39997_ instanceof AbstractVillager) ? (CraftAbstractVillager) ((AbstractVillager) this.f_39997_).getBukkitEntity() : null;
+   }
+
+   @Override
+   public Location getLocation() {
+      return (f_39997_ instanceof net.minecraft.world.entity.npc.Villager) ? ((net.minecraft.world.entity.npc.Villager) this.f_39997_).getBukkitEntity().getLocation() : null;
+   }
+   // CraftBukkit end
 
    public MerchantContainer(Merchant p_40003_) {
       this.f_39997_ = p_40003_;
