--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class MerchantMenu extends AbstractContainerMenu {
    protected static final int f_150619_ = 0;
@@ -29,6 +_,19 @@
    private boolean f_40030_;
    private boolean f_40031_;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity == null) {
+         bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryMerchant(f_40027_, f_40028_), this);
+      }
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public MerchantMenu(int p_40033_, Inventory p_40034_) {
       this(p_40033_, p_40034_, new ClientSideMerchant(p_40034_.f_35978_));
    }
@@ -40,7 +_,7 @@
       this.m_38897_(new Slot(this.f_40028_, 0, 136, 37));
       this.m_38897_(new Slot(this.f_40028_, 1, 162, 37));
       this.m_38897_(new MerchantResultSlot(p_40037_.f_35978_, p_40038_, this.f_40028_, 2, 220, 37));
-
+      this.player = p_40037_; // CraftBukkit - save player
       for(int i = 0; i < 3; ++i) {
          for(int j = 0; j < 9; ++j) {
             this.m_38897_(new Slot(p_40037_, j + i * 9 + 9, 108 + j * 18, 84 + i * 18));
@@ -144,7 +_,7 @@
    }
 
    private void m_40077_() {
-      if (!this.f_40027_.m_183595_()) {
+      if (!this.f_40027_.m_183595_() && this.f_40027_ instanceof Entity) { // CraftBukkit - SPIGOT-5035
          Entity entity = (Entity)this.f_40027_;
          entity.m_183503_().m_7785_(entity.m_20185_(), entity.m_20186_(), entity.m_20189_(), this.f_40027_.m_7596_(), SoundSource.NEUTRAL, 1.0F, 1.0F, false);
       }
