--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -8,12 +_,31 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.entity.EnderChestBlockEntity;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
+
 public class PlayerEnderChestContainer extends SimpleContainer {
    @Nullable
    private EnderChestBlockEntity f_40101_;
-
-   public PlayerEnderChestContainer() {
+   // CraftBukkit start
+   private final Player owner;
+
+   public InventoryHolder getBukkitOwner() {
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public Location getLocation() {
+      return this.f_40101_ != null ? CraftLocation.toBukkit(this.f_40101_.m_58899_(), this.f_40101_.m_58904_().getWorld()) : null;
+   }
+
+   public PlayerEnderChestContainer(Player owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void m_40105_(EnderChestBlockEntity p_40106_) {
