--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -7,13 +_,28 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.entity.EnderChestBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.inventory.InventoryHolder;
 
 public class PlayerEnderChestContainer extends SimpleContainer {
    @Nullable
    private EnderChestBlockEntity f_40101_;
-
-   public PlayerEnderChestContainer() {
+   // CraftBukkit start
+   private final Player owner;
+
+   public InventoryHolder getBukkitOwner() {
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public Location getLocation() {
+      return this.f_40101_ != null ? new Location(this.f_40101_.m_58904_().getWorld(), this.f_40101_.m_58899_().m_123341_(), this.f_40101_.m_58899_().m_123342_(), this.f_40101_.m_58899_().m_123343_()) : null;
+   }
+
+   public PlayerEnderChestContainer(Player owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void m_40105_(EnderChestBlockEntity p_40106_) {
