--- a/net/minecraft/world/inventory/ResultContainer.java
+++ b/net/minecraft/world/inventory/ResultContainer.java
@@ -8,10 +_,73 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Recipe;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
 public class ResultContainer implements Container, RecipeHolder {
    private final NonNullList<ItemStack> f_40140_ = NonNullList.m_122780_(1, ItemStack.f_41583_);
    @Nullable
    private Recipe<?> f_40141_;
+
+   // CraftBukkit start
+   private int maxStack = MAX_STACK;
+
+   public java.util.List<ItemStack> getContents() {
+      return this.f_40140_;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      // return null; // Result slots don't get an owner // Paper - remove
+      // Paper start
+      if (this.holder == null && this.holderCreator != null) {
+         this.holder = this.holderCreator.get();
+      }
+      return this.holder; // Result slots don't get an owner
+      // Paper end - yes they do
+   }
+
+   // Don't need a transaction; the InventoryCrafting keeps track of it for us
+   public void onOpen(CraftHumanEntity who) {}
+
+   public void onClose(CraftHumanEntity who) {}
+
+   public java.util.List<HumanEntity> getViewers() {
+      return new java.util.ArrayList<HumanEntity>();
+   }
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+   }
+   @Override
+   public Location getLocation() {
+      return null;
+   }
+   // CraftBukkit end
+   // Paper start
+   private @Nullable java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> holderCreator;
+   private @Nullable org.bukkit.inventory.InventoryHolder holder;
+   public ResultContainer(java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> holderCreator) {
+      this.holderCreator = holderCreator;
+   }
+
+   // CatServer start
+   public ResultContainer setHolderCreator(java.util.function.Supplier<? extends org.bukkit.inventory.InventoryHolder> holderCreator) {
+      this.holderCreator = holderCreator;
+      return this;
+   }
+   // CatServer end
+   // Paper end
+
+   // CatServer start - Just keep an empty constructor
+   public ResultContainer() {
+   }
+   // CatServer end
 
    public int m_6643_() {
       return 1;
