--- a/net/minecraft/world/inventory/ShulkerBoxMenu.java
+++ b/net/minecraft/world/inventory/ShulkerBoxMenu.java
@@ -6,9 +_,27 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class ShulkerBoxMenu extends AbstractContainerMenu {
    private static final int f_150640_ = 27;
    private final Container f_40186_;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), new CraftInventory(this.f_40186_), this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public ShulkerBoxMenu(int p_40188_, Inventory p_40189_) {
       this(p_40188_, p_40189_, new SimpleContainer(27));
@@ -18,6 +_,7 @@
       super(MenuType.f_39976_, p_40191_);
       m_38869_(p_40193_, 27);
       this.f_40186_ = p_40193_;
+      this.player = p_40192_; // CraftBukkit - save player
       p_40193_.m_5856_(p_40192_.f_35978_);
       int i = 3;
       int j = 9;
@@ -41,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_40195_) {
+      if (!this.checkReachable) return true; // CraftBukkit - 1.20.1
       return this.f_40186_.m_6542_(p_40195_);
    }
 
