--- a/net/minecraft/world/inventory/SmithingMenu.java
+++ b/net/minecraft/world/inventory/SmithingMenu.java
@@ -12,6 +_,8 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView; // CraftBukkit
+
 public class SmithingMenu extends ItemCombinerMenu {
    public static final int f_266015_ = 0;
    public static final int f_266035_ = 1;
@@ -26,6 +_,9 @@
    @Nullable
    private SmithingRecipe f_40242_;
    private final List<SmithingRecipe> f_40243_;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   // CraftBukkit end
 
    public SmithingMenu(int p_40245_, Inventory p_40246_) {
       this(p_40245_, p_40246_, ContainerLevelAccess.f_39287_);
@@ -88,14 +_,16 @@
    public void m_6640_() {
       List<SmithingRecipe> list = this.f_40241_.m_7465_().m_44056_(RecipeType.f_44113_, this.f_39769_, this.f_40241_);
       if (list.isEmpty()) {
-         this.f_39768_.m_6836_(0, ItemStack.f_41583_);
+         if (getBukkitView() != null) // CatServer - Bypass call event if Bukkit view is null (mod custom container)
+            org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), ItemStack.f_41583_); // CraftBukkit
       } else {
          SmithingRecipe smithingrecipe = list.get(0);
          ItemStack itemstack = smithingrecipe.m_5874_(this.f_39769_, this.f_40241_.m_9598_());
          if (itemstack.m_246617_(this.f_40241_.m_246046_())) {
             this.f_40242_ = smithingrecipe;
             this.f_39768_.m_6029_(smithingrecipe);
-            this.f_39768_.m_6836_(0, itemstack);
+            if (getBukkitView() != null) // CatServer - Bypass call event if Bukkit view is null (mod custom container)
+               org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), itemstack); // CraftBukkit
          }
       }
 
@@ -126,4 +_,17 @@
          return m_266320_(p_266647_, p_266846_);
       }).anyMatch(Optional::isPresent);
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventorySmithing(
+              f_39770_.getLocation(), this.f_39769_, this.f_39768_);
+      bukkitEntity = new CraftInventoryView(this.f_39771_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
