--- a/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/net/minecraft/world/inventory/StonecutterMenu.java
@@ -15,6 +_,12 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryStonecutter;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class StonecutterMenu extends AbstractContainerMenu {
    public static final int f_150665_ = 0;
    public static final int f_150666_ = 1;
@@ -32,14 +_,22 @@
    final Slot f_40283_;
    Runnable f_40291_ = () -> {
    };
-   public final Container f_40284_ = new SimpleContainer(1) {
-      public void m_6596_() {
-         super.m_6596_();
-         StonecutterMenu.this.m_6199_(this);
-         StonecutterMenu.this.f_40291_.run();
+   public final Container f_40284_; // CatServer - moved into constructor
+   final ResultContainer f_40292_; // CatServer - moved into constructor
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
       }
-   };
-   final ResultContainer f_40292_ = new ResultContainer();
+      CraftInventoryStonecutter inventory = new CraftInventoryStonecutter(this.f_40284_, this.f_40292_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public StonecutterMenu(int p_40294_, Inventory p_40295_) {
       this(p_40294_, p_40295_, ContainerLevelAccess.f_39287_);
@@ -47,6 +_,23 @@
 
    public StonecutterMenu(int p_40297_, Inventory p_40298_, final ContainerLevelAccess p_40299_) {
       super(MenuType.f_39980_, p_40297_);
+
+      // CatServer start
+      this.f_40284_ = new SimpleContainer(1) {
+         public void m_6596_() {
+            super.m_6596_();
+            StonecutterMenu.this.m_6199_(this);
+            StonecutterMenu.this.f_40291_.run();
+         }
+
+         @Override
+         public Location getLocation() {
+            return f_40285_.getLocation();
+         }
+      }.setBukkitOwnerCreator(this.createBlockHolder(p_40299_)); // Paper
+      this.f_40292_ = new ResultContainer().setHolderCreator(this.createBlockHolder(p_40299_)); // Paper
+      // CatServer end
+
       this.f_40285_ = p_40299_;
       this.f_40287_ = p_40298_.f_35978_.m_9236_();
       this.f_40282_ = this.m_38897_(new Slot(this.f_40284_, 0, 20, 33));
@@ -90,6 +_,7 @@
       }
 
       this.m_38895_(this.f_40286_);
+      player = (org.bukkit.entity.Player) p_40298_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public int m_40338_() {
@@ -109,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_40307_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_40285_, p_40307_, Blocks.f_50679_);
    }
 
