--- a/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/net/minecraft/world/inventory/StonecutterMenu.java
@@ -14,6 +_,9 @@
 import net.minecraft.world.item.crafting.StonecutterRecipe;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryStonecutter;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView;
 
 public class StonecutterMenu extends AbstractContainerMenu {
    public static final int f_150665_ = 0;
@@ -38,9 +_,32 @@
          StonecutterMenu.this.m_6199_(this);
          StonecutterMenu.this.f_40291_.run();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return f_40284_.getLocation();
+      }
+      // CraftBukkit end
    };
    final ResultContainer f_40292_ = new ResultContainer();
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryStonecutter inventory = new CraftInventoryStonecutter(this.f_40284_, this.f_40292_);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public StonecutterMenu(int p_40294_, Inventory p_40295_) {
       this(p_40294_, p_40295_, ContainerLevelAccess.f_39287_);
    }
@@ -86,6 +_,7 @@
       }
 
       this.m_38895_(this.f_40286_);
+      player = (org.bukkit.entity.Player) p_40298_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public int m_40338_() {
@@ -105,6 +_,7 @@
    }
 
    public boolean m_6875_(Player p_40307_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_40285_, p_40307_, Blocks.f_50679_);
    }
 
