--- ../src-base/minecraft/net/minecraft/network/datasync/DataSerializers.java
+++ ../src-work/minecraft/net/minecraft/network/datasync/DataSerializers.java
@@ -307,18 +307,18 @@
 
     public static void registerSerializer(DataSerializer<?> serializer)
     {
-        REGISTRY.add(serializer);
+        if (REGISTRY.add(serializer) >= 256) throw new RuntimeException("Vanilla DataSerializer ID limit exceeded");
     }
 
     @Nullable
     public static DataSerializer<?> getSerializer(int id)
     {
-        return (DataSerializer)REGISTRY.get(id);
+        return net.minecraftforge.common.ForgeHooks.getSerializer(id, REGISTRY);
     }
 
     public static int getSerializerId(DataSerializer<?> serializer)
     {
-        return REGISTRY.getId(serializer);
+        return net.minecraftforge.common.ForgeHooks.getSerializerId(serializer, REGISTRY);
     }
 
     static
