--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -15,6 +_,7 @@
 import net.minecraft.CrashReport;
 import net.minecraft.SystemReport;
 import net.minecraft.Util;
+import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
@@ -42,6 +_,7 @@
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 public class GameTestServer extends MinecraftServer {
@@ -85,7 +_,7 @@
    }
 
    public GameTestServer(Thread p_206597_, LevelStorageSource.LevelStorageAccess p_206598_, PackRepository p_206599_, WorldStem p_206600_, Collection<GameTestBatch> p_206601_, BlockPos p_206602_) {
-      super(p_206597_, p_206598_, p_206599_, p_206600_, Proxy.NO_PROXY, DataFixers.m_14512_(), (MinecraftSessionService)null, (GameProfileRepository)null, (GameProfileCache)null, LoggerChunkProgressListener::new);
+      super(null, null, null, p_206597_, p_206598_, p_206599_, p_206600_, Proxy.NO_PROXY, DataFixers.m_14512_(), (MinecraftSessionService)null, (GameProfileRepository)null, (GameProfileCache)null, LoggerChunkProgressListener::new);
       this.f_177587_ = Lists.newArrayList(p_206601_);
       this.f_177588_ = p_206602_;
    }
@@ -93,12 +_,13 @@
    public boolean m_7038_() {
       this.m_129823_(new PlayerList(this, this.m_206579_(), this.f_129745_, 1) {
       });
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.m_130006_();
       ServerLevel serverlevel = this.m_129783_();
       serverlevel.m_8733_(this.f_177588_, 0.0F);
       int i = 20000000;
       serverlevel.m_8606_(20000000, 20000000, false, false);
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void m_5705_(BooleanSupplier p_177619_) {
@@ -200,6 +_,11 @@
 
    public boolean m_6102_() {
       return false;
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return null;
    }
 
    public boolean m_7779_(GameProfile p_177617_) {
