--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -18,6 +_,7 @@
 import net.minecraft.CrashReport;
 import net.minecraft.SystemReport;
 import net.minecraft.Util;
+import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.LayeredRegistryAccess;
@@ -49,6 +_,7 @@
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.storage.PrimaryLevelData;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 public class GameTestServer extends MinecraftServer {
@@ -105,13 +_,14 @@
    public boolean m_7038_() {
       this.m_129823_(new PlayerList(this, this.m_247573_(), this.f_129745_, 1) {
       });
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.m_130006_();
       ServerLevel serverlevel = this.m_129783_();
       serverlevel.m_8733_(this.f_177588_, 0.0F);
       int i = 20000000;
       serverlevel.m_8606_(20000000, 20000000, false, false);
       f_177585_.info("Started game test server");
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void m_5705_(BooleanSupplier p_177619_) {
@@ -221,6 +_,13 @@
    public boolean m_6102_() {
       return false;
    }
+
+   // CatServer start
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return console;
+   }
+   // CatServer end
 
    public boolean m_7779_(GameProfile p_177617_) {
       return false;
