--- a/net/minecraft/advancements/AdvancementList.java
+++ b/net/minecraft/advancements/AdvancementList.java
@@ -13,7 +_,7 @@
 
 public class AdvancementList {
    private static final Logger f_139325_ = LogUtils.getLogger();
-   private final Map<ResourceLocation, Advancement> f_139326_ = Maps.newHashMap();
+   public final Map<ResourceLocation, Advancement> f_139326_ = Maps.newHashMap(); // CatServer - private -> public
    private final Set<Advancement> f_139327_ = Sets.newLinkedHashSet();
    private final Set<Advancement> f_139328_ = Sets.newLinkedHashSet();
    @Nullable
@@ -90,7 +_,7 @@
          }
       }
 
-      f_139325_.info("Loaded {} advancements", (int)this.f_139326_.size());
+      // LOGGER.info("Loaded {} advancements", (int)this.advancements.size()); // CraftBukkit - moved to AdvancementDataWorld#reload
    }
 
    public void m_139332_() {
