--- a/net/minecraft/advancements/AdvancementList.java
+++ b/net/minecraft/advancements/AdvancementList.java
@@ -3,14 +_,15 @@
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 import com.mojang.logging.LogUtils;
+import net.minecraft.resources.ResourceLocation;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.Map;
-import java.util.Set;
 import java.util.Map.Entry;
-import javax.annotation.Nullable;
-import net.minecraft.resources.ResourceLocation;
-import org.slf4j.Logger;
+import java.util.Set;
 
 public class AdvancementList {
    private static final Logger f_139325_ = LogUtils.getLogger();
@@ -91,6 +_,7 @@
          }
       }
 
+      net.minecraftforge.common.AdvancementLoadFix.buildSortedTrees(this.f_139327_);
       f_139325_.info("Loaded {} advancements", (int)this.f_139326_.size());
    }
 
