--- a/net/minecraft/advancements/AdvancementList.java
+++ b/net/minecraft/advancements/AdvancementList.java
@@ -91,7 +_,8 @@
          }
       }
 
-      f_139325_.info("Loaded {} advancements", (int)this.f_139326_.size());
+      net.minecraftforge.common.AdvancementLoadFix.buildSortedTrees(this.f_139327_);
+      // LOGGER.info("Loaded {} advancements", (int)this.advancements.size()); // CraftBukkit - moved to AdvancementDataWorld#reload
    }
 
    public void m_139332_() {
