--- a/net/minecraft/advancements/AdvancementList.java
+++ b/net/minecraft/advancements/AdvancementList.java
@@ -14,7 +_,7 @@
 
 public class AdvancementList {
    private static final Logger f_139325_ = LogUtils.getLogger();
-   private final Map<ResourceLocation, Advancement> f_139326_ = Maps.newHashMap();
+   public final Map<ResourceLocation, Advancement> f_139326_ = Maps.newHashMap(); // CatServer - private -> public
    private final Set<Advancement> f_139327_ = Sets.newLinkedHashSet();
    private final Set<Advancement> f_139328_ = Sets.newLinkedHashSet();
    @Nullable
@@ -91,6 +_,7 @@
          }
       }
 
+      net.minecraftforge.common.AdvancementLoadFix.buildSortedTrees(this.f_139327_);
       f_139325_.info("Loaded {} advancements", (int)this.f_139326_.size());
    }
 
