--- ../src-base/minecraft/net/minecraft/client/audio/MovingSoundMinecartRiding.java
+++ ../src-work/minecraft/net/minecraft/client/audio/MovingSoundMinecartRiding.java
@@ -19,7 +19,7 @@
         super(SoundEvents.ENTITY_MINECART_INSIDE, SoundCategory.NEUTRAL);
         this.player = playerRiding;
         this.minecart = minecart;
-        this.attenuationType = ISound.AttenuationType.NONE;
+        this.attenuationType = AttenuationType.NONE;
         this.repeat = true;
         this.repeatDelay = 0;
     }
