--- ../src-base/minecraft/net/minecraft/client/audio/MusicTicker.java
+++ ../src-work/minecraft/net/minecraft/client/audio/MusicTicker.java
@@ -24,7 +24,7 @@
 
     public void update()
     {
-        MusicTicker.MusicType musicticker$musictype = this.mc.getAmbientMusicType();
+        MusicType musicticker$musictype = this.mc.getAmbientMusicType();
 
         if (this.currentMusic != null)
         {
@@ -49,7 +49,7 @@
         }
     }
 
-    public void playMusic(MusicTicker.MusicType requestedMusicType)
+    public void playMusic(MusicType requestedMusicType)
     {
         this.currentMusic = PositionedSoundRecord.getMusicRecord(requestedMusicType.getMusicLocation());
         this.mc.getSoundHandler().playSound(this.currentMusic);
