--- ../src-base/minecraft/net/minecraft/client/audio/PositionedSound.java
+++ ../src-work/minecraft/net/minecraft/client/audio/PositionedSound.java
@@ -22,7 +22,7 @@
     protected float zPosF;
     protected boolean repeat;
     protected int repeatDelay;
-    protected ISound.AttenuationType attenuationType;
+    protected AttenuationType attenuationType;
 
     protected PositionedSound(SoundEvent soundIn, SoundCategory categoryIn)
     {
@@ -33,7 +33,7 @@
     {
         this.volume = 1.0F;
         this.pitch = 1.0F;
-        this.attenuationType = ISound.AttenuationType.LINEAR;
+        this.attenuationType = AttenuationType.LINEAR;
         this.positionedSoundLocation = soundId;
         this.category = categoryIn;
     }
@@ -104,7 +104,7 @@
         return this.zPosF;
     }
 
-    public ISound.AttenuationType getAttenuationType()
+    public AttenuationType getAttenuationType()
     {
         return this.attenuationType;
     }
