--- ../src-base/minecraft/net/minecraft/client/audio/Sound.java
+++ ../src-work/minecraft/net/minecraft/client/audio/Sound.java
@@ -11,10 +11,10 @@
     private final float volume;
     private final float pitch;
     private final int weight;
-    private final Sound.Type type;
+    private final Type type;
     private final boolean streaming;
 
-    public Sound(String nameIn, float volumeIn, float pitchIn, int weightIn, Sound.Type typeIn, boolean p_i46526_6_)
+    public Sound(String nameIn, float volumeIn, float pitchIn, int weightIn, Type typeIn, boolean p_i46526_6_)
     {
         this.name = new ResourceLocation(nameIn);
         this.volume = volumeIn;
@@ -54,7 +54,7 @@
         return this;
     }
 
-    public Sound.Type getType()
+    public Type getType()
     {
         return this.type;
     }
@@ -77,9 +77,9 @@
             this.name = nameIn;
         }
 
-        public static Sound.Type getByName(String nameIn)
+        public static Type getByName(String nameIn)
         {
-            for (Sound.Type sound$type : values())
+            for (Type sound$type : values())
             {
                 if (sound$type.name.equals(nameIn))
                 {
