--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -26,9 +_,9 @@
 public abstract class BaseSpawner {
    private static final Logger f_45441_ = LogUtils.getLogger();
    private static final int f_151303_ = 1;
-   private int f_45442_ = 20;
+   public int f_45442_ = 20;
    private SimpleWeightedRandomList<SpawnData> f_45443_ = SimpleWeightedRandomList.m_185864_();
-   private SpawnData f_45444_ = new SpawnData();
+   public SpawnData f_45444_ = new SpawnData();
    private double f_45445_;
    private double f_45446_;
    public int f_45447_ = 200;
@@ -43,6 +_,7 @@
 
    public void m_45462_(EntityType<?> p_45463_) {
       this.f_45444_.m_186567_().m_128359_("id", Registry.f_122826_.m_7981_(p_45463_).toString());
+      this.f_45443_ = SimpleWeightedRandomList.m_185864_(); // CraftBukkit - SPIGOT-3496, MC-92282
    }
 
    private boolean m_151343_(Level p_151344_, BlockPos p_151345_) {
@@ -125,16 +_,37 @@
                   entity.m_7678_(entity.m_20185_(), entity.m_20186_(), entity.m_20189_(), p_151312_.f_46441_.nextFloat() * 360.0F, 0.0F);
                   if (entity instanceof Mob) {
                      Mob mob = (Mob)entity;
+                     net.minecraftforge.eventbus.api.Event.Result res = net.minecraftforge.event.ForgeEventFactory.canEntitySpawn(mob, p_151312_, (float)entity.m_20185_(), (float)entity.m_20186_(), (float)entity.m_20189_(), this, MobSpawnType.SPAWNER);
+                     if (res == net.minecraftforge.eventbus.api.Event.Result.DENY) continue;
+                     if (res == net.minecraftforge.eventbus.api.Event.Result.DEFAULT)
                      if (this.f_45444_.m_186574_().isEmpty() && !mob.m_5545_(p_151312_, MobSpawnType.SPAWNER) || !mob.m_6914_(p_151312_)) {
                         continue;
                      }
 
                      if (this.f_45444_.m_186567_().m_128440_() == 1 && this.f_45444_.m_186567_().m_128425_("id", 8)) {
+                        if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mob, (LevelAccessor)p_151312_, (float)entity.m_20185_(), (float)entity.m_20186_(), (float)entity.m_20189_(), this, MobSpawnType.SPAWNER))
                         ((Mob)entity).m_6518_(p_151312_, p_151312_.m_6436_(entity.m_142538_()), MobSpawnType.SPAWNER, (SpawnGroupData)null, (CompoundTag)null);
                      }
-                  }
+                     // Spigot Start
+                     if ( mob.f_19853_.spigotConfig.nerfSpawnerMobs ) {
+                        mob.aware = false;
+                     }
+                     // Spigot End
+                  }
+                  // Spigot Start
+                  if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                     Entity vehicle = entity.m_20202_();
+                     if (vehicle != null) {
+                        vehicle.m_146870_();
+                     }
+                     for (Entity passenger : entity.m_146897_()) {
+                        passenger.m_146870_();
+                     }
+                     continue;
+                  }
+                  // Spigot End
 
-                  if (!p_151312_.m_8860_(entity)) {
+                  if (!p_151312_.m_8860_(entity)) { // FoxServer TODO
                      this.m_151350_(p_151312_, p_151313_);
                      return;
                   }
@@ -271,4 +_,12 @@
    public double m_45474_() {
       return this.f_45446_;
    }
+
+   @Nullable
+   public Entity getSpawnerEntity() {
+      return null;
+   }
+
+   @Nullable
+   public net.minecraft.world.level.block.entity.BlockEntity getSpawnerBlockEntity(){ return null; }
 }
