--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -30,6 +_,11 @@
    private String f_45404_ = "";
    private Component f_45405_ = f_45398_;
 
+   // CraftBukkit start
+   @Override
+   public abstract org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper);
+   // CraftBukkit end
+
    public int m_45436_() {
       return this.f_45401_;
    }
@@ -119,7 +_,7 @@
                      }
 
                   });
-                  minecraftserver.m_129892_().m_230957_(commandsourcestack, this.f_45404_);
+                  minecraftserver.m_129892_().dispatchServerCommand(commandsourcestack, this.f_45404_); // CraftBukkit
                } catch (Throwable throwable) {
                   CrashReport crashreport = CrashReport.m_127521_(throwable, "Executing command block");
                   CrashReportCategory crashreportcategory = crashreport.m_127514_("Command to be executed");
