--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -1,10 +_,6 @@
 package net.minecraft.world.level;
 
 import com.mojang.logging.LogUtils;
-import java.util.Optional;
-import java.util.Random;
-import java.util.function.Function;
-import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
 import net.minecraft.core.particles.ParticleTypes;
@@ -14,261 +_,294 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.random.SimpleWeightedRandomList;
 import net.minecraft.world.Difficulty;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.Mob;
-import net.minecraft.world.entity.MobSpawnType;
-import net.minecraft.world.entity.SpawnGroupData;
-import net.minecraft.world.entity.SpawnPlacements;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.phys.AABB;
 import org.slf4j.Logger;
 
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.Random;
+import java.util.function.Function;
+
 public abstract class BaseSpawner {
-   private static final Logger f_45441_ = LogUtils.getLogger();
-   private static final int f_151303_ = 1;
-   private int f_45442_ = 20;
-   private SimpleWeightedRandomList<SpawnData> f_45443_ = SimpleWeightedRandomList.m_185864_();
-   private SpawnData f_45444_ = new SpawnData();
-   private double f_45445_;
-   private double f_45446_;
-   public int f_45447_ = 200;
-   public int f_45448_ = 800;
-   public int f_45449_ = 4;
-   @Nullable
-   private Entity f_45450_;
-   public int f_45451_ = 6;
-   public int f_45452_ = 16;
-   public int f_45453_ = 4;
-   private final Random f_151305_ = new Random();
-
-   public void m_45462_(EntityType<?> p_45463_) {
-      this.f_45444_.m_186567_().m_128359_("id", Registry.f_122826_.m_7981_(p_45463_).toString());
-   }
-
-   private boolean m_151343_(Level p_151344_, BlockPos p_151345_) {
-      return p_151344_.m_45914_((double)p_151345_.m_123341_() + 0.5D, (double)p_151345_.m_123342_() + 0.5D, (double)p_151345_.m_123343_() + 0.5D, (double)this.f_45452_);
-   }
-
-   public void m_151319_(Level p_151320_, BlockPos p_151321_) {
-      if (!this.m_151343_(p_151320_, p_151321_)) {
-         this.f_45446_ = this.f_45445_;
-      } else {
-         double d0 = (double)p_151321_.m_123341_() + p_151320_.f_46441_.nextDouble();
-         double d1 = (double)p_151321_.m_123342_() + p_151320_.f_46441_.nextDouble();
-         double d2 = (double)p_151321_.m_123343_() + p_151320_.f_46441_.nextDouble();
-         p_151320_.m_7106_(ParticleTypes.f_123762_, d0, d1, d2, 0.0D, 0.0D, 0.0D);
-         p_151320_.m_7106_(ParticleTypes.f_123744_, d0, d1, d2, 0.0D, 0.0D, 0.0D);
-         if (this.f_45442_ > 0) {
-            --this.f_45442_;
-         }
-
-         this.f_45446_ = this.f_45445_;
-         this.f_45445_ = (this.f_45445_ + (double)(1000.0F / ((float)this.f_45442_ + 200.0F))) % 360.0D;
-      }
-
-   }
-
-   public void m_151311_(ServerLevel p_151312_, BlockPos p_151313_) {
-      if (this.m_151343_(p_151312_, p_151313_)) {
-         if (this.f_45442_ == -1) {
-            this.m_151350_(p_151312_, p_151313_);
-         }
-
-         if (this.f_45442_ > 0) {
-            --this.f_45442_;
-         } else {
-            boolean flag = false;
-
-            for(int i = 0; i < this.f_45449_; ++i) {
-               CompoundTag compoundtag = this.f_45444_.m_186567_();
-               Optional<EntityType<?>> optional = EntityType.m_20637_(compoundtag);
-               if (optional.isEmpty()) {
-                  this.m_151350_(p_151312_, p_151313_);
-                  return;
-               }
-
-               ListTag listtag = compoundtag.m_128437_("Pos", 6);
-               int j = listtag.size();
-               double d0 = j >= 1 ? listtag.m_128772_(0) : (double)p_151313_.m_123341_() + (p_151312_.f_46441_.nextDouble() - p_151312_.f_46441_.nextDouble()) * (double)this.f_45453_ + 0.5D;
-               double d1 = j >= 2 ? listtag.m_128772_(1) : (double)(p_151313_.m_123342_() + p_151312_.f_46441_.nextInt(3) - 1);
-               double d2 = j >= 3 ? listtag.m_128772_(2) : (double)p_151313_.m_123343_() + (p_151312_.f_46441_.nextDouble() - p_151312_.f_46441_.nextDouble()) * (double)this.f_45453_ + 0.5D;
-               if (p_151312_.m_45772_(optional.get().m_20585_(d0, d1, d2))) {
-                  BlockPos blockpos = new BlockPos(d0, d1, d2);
-                  if (this.f_45444_.m_186574_().isPresent()) {
-                     if (!optional.get().m_20674_().m_21609_() && p_151312_.m_46791_() == Difficulty.PEACEFUL) {
-                        continue;
-                     }
-
-                     SpawnData.CustomSpawnRules spawndata$customspawnrules = this.f_45444_.m_186574_().get();
-                     if (!spawndata$customspawnrules.f_186584_().m_184578_(p_151312_.m_45517_(LightLayer.BLOCK, blockpos)) || !spawndata$customspawnrules.f_186585_().m_184578_(p_151312_.m_45517_(LightLayer.SKY, blockpos))) {
-                        continue;
-                     }
-                  } else if (!SpawnPlacements.m_21759_(optional.get(), p_151312_, MobSpawnType.SPAWNER, blockpos, p_151312_.m_5822_())) {
-                     continue;
-                  }
-
-                  Entity entity = EntityType.m_20645_(compoundtag, p_151312_, (p_151310_) -> {
-                     p_151310_.m_7678_(d0, d1, d2, p_151310_.m_146908_(), p_151310_.m_146909_());
-                     return p_151310_;
-                  });
-                  if (entity == null) {
-                     this.m_151350_(p_151312_, p_151313_);
-                     return;
-                  }
-
-                  int k = p_151312_.m_45976_(entity.getClass(), (new AABB((double)p_151313_.m_123341_(), (double)p_151313_.m_123342_(), (double)p_151313_.m_123343_(), (double)(p_151313_.m_123341_() + 1), (double)(p_151313_.m_123342_() + 1), (double)(p_151313_.m_123343_() + 1))).m_82400_((double)this.f_45453_)).size();
-                  if (k >= this.f_45451_) {
-                     this.m_151350_(p_151312_, p_151313_);
-                     return;
-                  }
-
-                  entity.m_7678_(entity.m_20185_(), entity.m_20186_(), entity.m_20189_(), p_151312_.f_46441_.nextFloat() * 360.0F, 0.0F);
-                  if (entity instanceof Mob) {
-                     Mob mob = (Mob)entity;
-                     if (this.f_45444_.m_186574_().isEmpty() && !mob.m_5545_(p_151312_, MobSpawnType.SPAWNER) || !mob.m_6914_(p_151312_)) {
-                        continue;
-                     }
-
-                     if (this.f_45444_.m_186567_().m_128440_() == 1 && this.f_45444_.m_186567_().m_128425_("id", 8)) {
-                        ((Mob)entity).m_6518_(p_151312_, p_151312_.m_6436_(entity.m_142538_()), MobSpawnType.SPAWNER, (SpawnGroupData)null, (CompoundTag)null);
-                     }
-                  }
-
-                  if (!p_151312_.m_8860_(entity)) {
-                     this.m_151350_(p_151312_, p_151313_);
-                     return;
-                  }
-
-                  p_151312_.m_46796_(2004, p_151313_, 0);
-                  if (entity instanceof Mob) {
-                     ((Mob)entity).m_21373_();
-                  }
-
-                  flag = true;
-               }
-            }
-
-            if (flag) {
-               this.m_151350_(p_151312_, p_151313_);
-            }
-
-         }
-      }
-   }
-
-   private void m_151350_(Level p_151351_, BlockPos p_151352_) {
-      if (this.f_45448_ <= this.f_45447_) {
-         this.f_45442_ = this.f_45447_;
-      } else {
-         this.f_45442_ = this.f_45447_ + this.f_151305_.nextInt(this.f_45448_ - this.f_45447_);
-      }
-
-      this.f_45443_.m_146335_(this.f_151305_).ifPresent((p_186386_) -> {
-         this.m_142667_(p_151351_, p_151352_, p_186386_.m_146310_());
-      });
-      this.m_142523_(p_151351_, p_151352_, 1);
-   }
-
-   public void m_151328_(@Nullable Level p_151329_, BlockPos p_151330_, CompoundTag p_151331_) {
-      this.f_45442_ = p_151331_.m_128448_("Delay");
-      boolean flag = p_151331_.m_128425_("SpawnPotentials", 9);
-      boolean flag1 = p_151331_.m_128425_("SpawnData", 10);
-      if (!flag) {
-         SpawnData spawndata;
-         if (flag1) {
-            spawndata = SpawnData.f_186559_.parse(NbtOps.f_128958_, p_151331_.m_128469_("SpawnData")).resultOrPartial((p_186391_) -> {
-               f_45441_.warn("Invalid SpawnData: {}", (Object)p_186391_);
-            }).orElseGet(SpawnData::new);
-         } else {
-            spawndata = new SpawnData();
-         }
-
-         this.f_45443_ = SimpleWeightedRandomList.m_185862_(spawndata);
-         this.m_142667_(p_151329_, p_151330_, spawndata);
-      } else {
-         ListTag listtag = p_151331_.m_128437_("SpawnPotentials", 10);
-         this.f_45443_ = SpawnData.f_186560_.parse(NbtOps.f_128958_, listtag).resultOrPartial((p_186388_) -> {
-            f_45441_.warn("Invalid SpawnPotentials list: {}", (Object)p_186388_);
-         }).orElseGet(SimpleWeightedRandomList::m_185864_);
-         if (flag1) {
-            SpawnData spawndata1 = SpawnData.f_186559_.parse(NbtOps.f_128958_, p_151331_.m_128469_("SpawnData")).resultOrPartial((p_186380_) -> {
-               f_45441_.warn("Invalid SpawnData: {}", (Object)p_186380_);
-            }).orElseGet(SpawnData::new);
-            this.m_142667_(p_151329_, p_151330_, spawndata1);
-         } else {
-            this.f_45443_.m_146335_(this.f_151305_).ifPresent((p_186378_) -> {
-               this.m_142667_(p_151329_, p_151330_, p_186378_.m_146310_());
-            });
-         }
-      }
-
-      if (p_151331_.m_128425_("MinSpawnDelay", 99)) {
-         this.f_45447_ = p_151331_.m_128448_("MinSpawnDelay");
-         this.f_45448_ = p_151331_.m_128448_("MaxSpawnDelay");
-         this.f_45449_ = p_151331_.m_128448_("SpawnCount");
-      }
-
-      if (p_151331_.m_128425_("MaxNearbyEntities", 99)) {
-         this.f_45451_ = p_151331_.m_128448_("MaxNearbyEntities");
-         this.f_45452_ = p_151331_.m_128448_("RequiredPlayerRange");
-      }
-
-      if (p_151331_.m_128425_("SpawnRange", 99)) {
-         this.f_45453_ = p_151331_.m_128448_("SpawnRange");
-      }
-
-      this.f_45450_ = null;
-   }
-
-   public CompoundTag m_186381_(CompoundTag p_186382_) {
-      p_186382_.m_128376_("Delay", (short)this.f_45442_);
-      p_186382_.m_128376_("MinSpawnDelay", (short)this.f_45447_);
-      p_186382_.m_128376_("MaxSpawnDelay", (short)this.f_45448_);
-      p_186382_.m_128376_("SpawnCount", (short)this.f_45449_);
-      p_186382_.m_128376_("MaxNearbyEntities", (short)this.f_45451_);
-      p_186382_.m_128376_("RequiredPlayerRange", (short)this.f_45452_);
-      p_186382_.m_128376_("SpawnRange", (short)this.f_45453_);
-      p_186382_.m_128365_("SpawnData", SpawnData.f_186559_.encodeStart(NbtOps.f_128958_, this.f_45444_).result().orElseThrow(() -> {
-         return new IllegalStateException("Invalid SpawnData");
-      }));
-      p_186382_.m_128365_("SpawnPotentials", SpawnData.f_186560_.encodeStart(NbtOps.f_128958_, this.f_45443_).result().orElseThrow());
-      return p_186382_;
-   }
-
-   @Nullable
-   public Entity m_151314_(Level p_151315_) {
-      if (this.f_45450_ == null) {
-         this.f_45450_ = EntityType.m_20645_(this.f_45444_.m_186567_(), p_151315_, Function.identity());
-         if (this.f_45444_.m_186567_().m_128440_() == 1 && this.f_45444_.m_186567_().m_128425_("id", 8) && this.f_45450_ instanceof Mob) {
-         }
-      }
-
-      return this.f_45450_;
-   }
-
-   public boolean m_151316_(Level p_151317_, int p_151318_) {
-      if (p_151318_ == 1) {
-         if (p_151317_.f_46443_) {
+    private static final Logger f_45441_ = LogUtils.getLogger();
+    private static final int f_151303_ = 1;
+    public int f_45442_ = 20;
+    private SimpleWeightedRandomList<SpawnData> f_45443_ = SimpleWeightedRandomList.m_185864_();
+    public SpawnData f_45444_ = new SpawnData();
+    private double f_45445_;
+    private double f_45446_;
+    public int f_45447_ = 200;
+    public int f_45448_ = 800;
+    public int f_45449_ = 4;
+    @Nullable
+    private Entity f_45450_;
+    public int f_45451_ = 6;
+    public int f_45452_ = 16;
+    public int f_45453_ = 4;
+    private final Random f_151305_ = new Random();
+
+    public void m_45462_(EntityType<?> p_45463_) {
+        this.f_45444_.m_186567_().m_128359_("id", Registry.f_122826_.m_7981_(p_45463_).toString());
+        this.f_45443_ = SimpleWeightedRandomList.m_185864_(); // CraftBukkit - SPIGOT-3496, MC-92282
+    }
+
+    private boolean m_151343_(Level p_151344_, BlockPos p_151345_) {
+        return p_151344_.m_45914_((double) p_151345_.m_123341_() + 0.5D, (double) p_151345_.m_123342_() + 0.5D, (double) p_151345_.m_123343_() + 0.5D, (double) this.f_45452_);
+    }
+
+    public void m_151319_(Level p_151320_, BlockPos p_151321_) {
+        if (!this.m_151343_(p_151320_, p_151321_)) {
+            this.f_45446_ = this.f_45445_;
+        } else {
+            double d0 = (double) p_151321_.m_123341_() + p_151320_.f_46441_.nextDouble();
+            double d1 = (double) p_151321_.m_123342_() + p_151320_.f_46441_.nextDouble();
+            double d2 = (double) p_151321_.m_123343_() + p_151320_.f_46441_.nextDouble();
+            p_151320_.m_7106_(ParticleTypes.f_123762_, d0, d1, d2, 0.0D, 0.0D, 0.0D);
+            p_151320_.m_7106_(ParticleTypes.f_123744_, d0, d1, d2, 0.0D, 0.0D, 0.0D);
+            if (this.f_45442_ > 0) {
+                --this.f_45442_;
+            }
+
+            this.f_45446_ = this.f_45445_;
+            this.f_45445_ = (this.f_45445_ + (double) (1000.0F / ((float) this.f_45442_ + 200.0F))) % 360.0D;
+        }
+
+    }
+
+    public void m_151311_(ServerLevel p_151312_, BlockPos p_151313_) {
+        if (this.m_151343_(p_151312_, p_151313_)) {
+            if (this.f_45442_ == -1) {
+                this.m_151350_(p_151312_, p_151313_);
+            }
+
+            if (this.f_45442_ > 0) {
+                --this.f_45442_;
+            } else {
+                boolean flag = false;
+
+                for (int i = 0; i < this.f_45449_; ++i) {
+                    CompoundTag compoundtag = this.f_45444_.m_186567_();
+                    Optional<EntityType<?>> optional = EntityType.m_20637_(compoundtag);
+                    if (optional.isEmpty()) {
+                        this.m_151350_(p_151312_, p_151313_);
+                        return;
+                    }
+
+                    ListTag listtag = compoundtag.m_128437_("Pos", 6);
+                    int j = listtag.size();
+                    double d0 = j >= 1 ? listtag.m_128772_(0) : (double) p_151313_.m_123341_() + (p_151312_.f_46441_.nextDouble() - p_151312_.f_46441_.nextDouble()) * (double) this.f_45453_ + 0.5D;
+                    double d1 = j >= 2 ? listtag.m_128772_(1) : (double) (p_151313_.m_123342_() + p_151312_.f_46441_.nextInt(3) - 1);
+                    double d2 = j >= 3 ? listtag.m_128772_(2) : (double) p_151313_.m_123343_() + (p_151312_.f_46441_.nextDouble() - p_151312_.f_46441_.nextDouble()) * (double) this.f_45453_ + 0.5D;
+                    if (p_151312_.m_45772_(optional.get().m_20585_(d0, d1, d2))) {
+                        BlockPos blockpos = new BlockPos(d0, d1, d2);
+                        if (this.f_45444_.m_186574_().isPresent()) {
+                            if (!optional.get().m_20674_().m_21609_() && p_151312_.m_46791_() == Difficulty.PEACEFUL) {
+                                continue;
+                            }
+
+                            SpawnData.CustomSpawnRules spawndata$customspawnrules = this.f_45444_.m_186574_().get();
+                            if (!spawndata$customspawnrules.f_186584_().m_184578_(p_151312_.m_45517_(LightLayer.BLOCK, blockpos)) || !spawndata$customspawnrules.f_186585_().m_184578_(p_151312_.m_45517_(LightLayer.SKY, blockpos))) {
+                                continue;
+                            }
+                        } else if (!SpawnPlacements.m_21759_(optional.get(), p_151312_, MobSpawnType.SPAWNER, blockpos, p_151312_.m_5822_())) {
+                            continue;
+                        }
+
+                        Entity entity = EntityType.m_20645_(compoundtag, p_151312_, (p_151310_) -> {
+                            p_151310_.m_7678_(d0, d1, d2, p_151310_.m_146908_(), p_151310_.m_146909_());
+                            return p_151310_;
+                        });
+                        if (entity == null) {
+                            this.m_151350_(p_151312_, p_151313_);
+                            return;
+                        }
+
+                        int k = p_151312_.m_45976_(entity.getClass(), (new AABB((double) p_151313_.m_123341_(), (double) p_151313_.m_123342_(), (double) p_151313_.m_123343_(), (double) (p_151313_.m_123341_() + 1), (double) (p_151313_.m_123342_() + 1), (double) (p_151313_.m_123343_() + 1))).m_82400_((double) this.f_45453_)).size();
+                        if (k >= this.f_45451_) {
+                            this.m_151350_(p_151312_, p_151313_);
+                            return;
+                        }
+
+                        entity.m_7678_(entity.m_20185_(), entity.m_20186_(), entity.m_20189_(), p_151312_.f_46441_.nextFloat() * 360.0F, 0.0F);
+                        if (entity instanceof Mob) {
+                            Mob mob = (Mob) entity;
+                            net.minecraftforge.eventbus.api.Event.Result res = net.minecraftforge.event.ForgeEventFactory.canEntitySpawn(mob, p_151312_, (float) entity.m_20185_(), (float) entity.m_20186_(), (float) entity.m_20189_(), this, MobSpawnType.SPAWNER);
+                            if (res == net.minecraftforge.eventbus.api.Event.Result.DENY) continue;
+                            if (res == net.minecraftforge.eventbus.api.Event.Result.DEFAULT)
+                                if (this.f_45444_.m_186574_().isEmpty() && !mob.m_5545_(p_151312_, MobSpawnType.SPAWNER) || !mob.m_6914_(p_151312_)) {
+                                    continue;
+                                }
+
+                            if (this.f_45444_.m_186567_().m_128440_() == 1 && this.f_45444_.m_186567_().m_128425_("id", 8)) {
+                                if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mob, (LevelAccessor) p_151312_, (float) entity.m_20185_(), (float) entity.m_20186_(), (float) entity.m_20189_(), this, MobSpawnType.SPAWNER))
+                                    ((Mob) entity).m_6518_(p_151312_, p_151312_.m_6436_(entity.m_142538_()), MobSpawnType.SPAWNER, (SpawnGroupData) null, (CompoundTag) null);
+                            }
+                            // Spigot Start
+                            if (mob.f_19853_.spigotConfig.nerfSpawnerMobs) {
+                                mob.aware = false;
+                            }
+                            // Spigot End
+                        }
+
+                        // Spigot Start
+                        if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                            Entity vehicle = entity.m_20202_();
+                            if (vehicle != null) {
+                                vehicle.m_146870_();
+                            }
+                            for (Entity passenger : entity.m_146897_()) {
+                                passenger.m_146870_();
+                            }
+                            continue;
+                        }
+                        // Spigot End
+
+                        if (!p_151312_.m_8860_(entity)) { // FoxServer TODO
+                            this.m_151350_(p_151312_, p_151313_);
+                            return;
+                        }
+
+                        p_151312_.m_46796_(2004, p_151313_, 0);
+                        if (entity instanceof Mob) {
+                            ((Mob) entity).m_21373_();
+                        }
+
+                        flag = true;
+                    }
+                }
+
+                if (flag) {
+                    this.m_151350_(p_151312_, p_151313_);
+                }
+
+            }
+        }
+    }
+
+    private void m_151350_(Level p_151351_, BlockPos p_151352_) {
+        if (this.f_45448_ <= this.f_45447_) {
             this.f_45442_ = this.f_45447_;
-         }
-
-         return true;
-      } else {
-         return false;
-      }
-   }
-
-   public void m_142667_(@Nullable Level p_151325_, BlockPos p_151326_, SpawnData p_151327_) {
-      this.f_45444_ = p_151327_;
-   }
-
-   public abstract void m_142523_(Level p_151322_, BlockPos p_151323_, int p_151324_);
-
-   public double m_45473_() {
-      return this.f_45445_;
-   }
-
-   public double m_45474_() {
-      return this.f_45446_;
-   }
+        } else {
+            this.f_45442_ = this.f_45447_ + this.f_151305_.nextInt(this.f_45448_ - this.f_45447_);
+        }
+
+        this.f_45443_.m_146335_(this.f_151305_).ifPresent((p_186386_) -> {
+            this.m_142667_(p_151351_, p_151352_, p_186386_.m_146310_());
+        });
+        this.m_142523_(p_151351_, p_151352_, 1);
+    }
+
+    public void m_151328_(@Nullable Level p_151329_, BlockPos p_151330_, CompoundTag p_151331_) {
+        this.f_45442_ = p_151331_.m_128448_("Delay");
+        boolean flag = p_151331_.m_128425_("SpawnPotentials", 9);
+        boolean flag1 = p_151331_.m_128425_("SpawnData", 10);
+        if (!flag) {
+            SpawnData spawndata;
+            if (flag1) {
+                spawndata = SpawnData.f_186559_.parse(NbtOps.f_128958_, p_151331_.m_128469_("SpawnData")).resultOrPartial((p_186391_) -> {
+                    f_45441_.warn("Invalid SpawnData: {}", (Object) p_186391_);
+                }).orElseGet(SpawnData::new);
+            } else {
+                spawndata = new SpawnData();
+            }
+
+            this.f_45443_ = SimpleWeightedRandomList.m_185862_(spawndata);
+            this.m_142667_(p_151329_, p_151330_, spawndata);
+        } else {
+            ListTag listtag = p_151331_.m_128437_("SpawnPotentials", 10);
+            this.f_45443_ = SpawnData.f_186560_.parse(NbtOps.f_128958_, listtag).resultOrPartial((p_186388_) -> {
+                f_45441_.warn("Invalid SpawnPotentials list: {}", (Object) p_186388_);
+            }).orElseGet(SimpleWeightedRandomList::m_185864_);
+            if (flag1) {
+                SpawnData spawndata1 = SpawnData.f_186559_.parse(NbtOps.f_128958_, p_151331_.m_128469_("SpawnData")).resultOrPartial((p_186380_) -> {
+                    f_45441_.warn("Invalid SpawnData: {}", (Object) p_186380_);
+                }).orElseGet(SpawnData::new);
+                this.m_142667_(p_151329_, p_151330_, spawndata1);
+            } else {
+                this.f_45443_.m_146335_(this.f_151305_).ifPresent((p_186378_) -> {
+                    this.m_142667_(p_151329_, p_151330_, p_186378_.m_146310_());
+                });
+            }
+        }
+
+        if (p_151331_.m_128425_("MinSpawnDelay", 99)) {
+            this.f_45447_ = p_151331_.m_128448_("MinSpawnDelay");
+            this.f_45448_ = p_151331_.m_128448_("MaxSpawnDelay");
+            this.f_45449_ = p_151331_.m_128448_("SpawnCount");
+        }
+
+        if (p_151331_.m_128425_("MaxNearbyEntities", 99)) {
+            this.f_45451_ = p_151331_.m_128448_("MaxNearbyEntities");
+            this.f_45452_ = p_151331_.m_128448_("RequiredPlayerRange");
+        }
+
+        if (p_151331_.m_128425_("SpawnRange", 99)) {
+            this.f_45453_ = p_151331_.m_128448_("SpawnRange");
+        }
+
+        this.f_45450_ = null;
+    }
+
+    public CompoundTag m_186381_(CompoundTag p_186382_) {
+        p_186382_.m_128376_("Delay", (short) this.f_45442_);
+        p_186382_.m_128376_("MinSpawnDelay", (short) this.f_45447_);
+        p_186382_.m_128376_("MaxSpawnDelay", (short) this.f_45448_);
+        p_186382_.m_128376_("SpawnCount", (short) this.f_45449_);
+        p_186382_.m_128376_("MaxNearbyEntities", (short) this.f_45451_);
+        p_186382_.m_128376_("RequiredPlayerRange", (short) this.f_45452_);
+        p_186382_.m_128376_("SpawnRange", (short) this.f_45453_);
+        p_186382_.m_128365_("SpawnData", SpawnData.f_186559_.encodeStart(NbtOps.f_128958_, this.f_45444_).result().orElseThrow(() -> {
+            return new IllegalStateException("Invalid SpawnData");
+        }));
+        p_186382_.m_128365_("SpawnPotentials", SpawnData.f_186560_.encodeStart(NbtOps.f_128958_, this.f_45443_).result().orElseThrow());
+        return p_186382_;
+    }
+
+    @Nullable
+    public Entity m_151314_(Level p_151315_) {
+        if (this.f_45450_ == null) {
+            this.f_45450_ = EntityType.m_20645_(this.f_45444_.m_186567_(), p_151315_, Function.identity());
+            if (this.f_45444_.m_186567_().m_128440_() == 1 && this.f_45444_.m_186567_().m_128425_("id", 8) && this.f_45450_ instanceof Mob) {
+            }
+        }
+
+        return this.f_45450_;
+    }
+
+    public boolean m_151316_(Level p_151317_, int p_151318_) {
+        if (p_151318_ == 1) {
+            if (p_151317_.f_46443_) {
+                this.f_45442_ = this.f_45447_;
+            }
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public void m_142667_(@Nullable Level p_151325_, BlockPos p_151326_, SpawnData p_151327_) {
+        this.f_45444_ = p_151327_;
+    }
+
+    public abstract void m_142523_(Level p_151322_, BlockPos p_151323_, int p_151324_);
+
+    public double m_45473_() {
+        return this.f_45445_;
+    }
+
+    public double m_45474_() {
+        return this.f_45446_;
+    }
+
+    @Nullable
+    public Entity getSpawnerEntity() {
+        return null;
+    }
+
+    @Nullable
+    public net.minecraft.world.level.block.entity.BlockEntity getSpawnerBlockEntity() {
+        return null;
+    }
 }
