--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -120,7 +_,7 @@
                      }
 
                   });
-                  minecraftserver.m_129892_().m_82117_(commandsourcestack, this.f_45404_);
+                  minecraftserver.m_129892_().dispatchServerCommand(commandsourcestack, this.f_45404_); // CraftBukkit
                } catch (Throwable throwable) {
                   CrashReport crashreport = CrashReport.m_127521_(throwable, "Executing command block");
                   CrashReportCategory crashreportcategory = crashreport.m_127514_("Command to be executed");
