--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -43,6 +_,9 @@
       return this.f_45403_ == null ? TextComponent.f_131282_ : this.f_45403_;
    }
 
+   @Override
+   public abstract org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper);
+
    public CompoundTag m_45421_(CompoundTag p_45422_) {
       p_45422_.m_128359_("Command", this.f_45404_);
       p_45422_.m_128405_("SuccessCount", this.f_45401_);
@@ -120,7 +_,7 @@
                      }
 
                   });
-                  minecraftserver.m_129892_().m_82117_(commandsourcestack, this.f_45404_);
+                  minecraftserver.m_129892_().dispatchServerCommand(commandsourcestack, this.f_45404_);
                } catch (Throwable throwable) {
                   CrashReport crashreport = CrashReport.m_127521_(throwable, "Executing command block");
                   CrashReportCategory crashreportcategory = crashreport.m_127514_("Command to be executed");
