--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public interface BlockGetter extends LevelHeightAccessor {
+public interface BlockGetter extends LevelHeightAccessor, net.minecraftforge.common.extensions.IForgeBlockGetter {
    @Nullable
    BlockEntity m_7702_(BlockPos p_45570_);
 
@@ -32,7 +_,7 @@
    FluidState m_6425_(BlockPos p_45569_);
 
    default int m_7146_(BlockPos p_45572_) {
-      return this.m_8055_(p_45572_).m_60791_();
+      return this.m_8055_(p_45572_).getLightEmission(this, p_45572_);
    }
 
    default int m_7469_() {
@@ -53,6 +_,22 @@
          return BlockHitResult.m_82426_(p_151370_.m_151404_(), Direction.m_122366_(vec3.f_82479_, vec3.f_82480_, vec3.f_82481_), new BlockPos(p_151370_.m_151404_()));
       });
    }
+
+   // CraftBukkit start
+   default BlockHitResult clip(ClipContext p_45548_, BlockPos p_217297_2_) {
+      BlockState blockstate = this.m_8055_(p_217297_2_);
+      FluidState fluidstate = this.m_6425_(p_217297_2_);
+      Vec3 vec3 = p_45548_.m_45702_();
+      Vec3 vec31 = p_45548_.m_45693_();
+      VoxelShape voxelshape = p_45548_.m_45694_(blockstate, this, p_217297_2_);
+      BlockHitResult blockraytraceresult = this.m_45558_(vec3, vec31, p_217297_2_, voxelshape, blockstate);
+      VoxelShape voxelshape1 = p_45548_.m_45698_(fluidstate, this, p_217297_2_);
+      BlockHitResult blockraytraceresult1 = voxelshape1.m_83220_(vec3, vec31, p_217297_2_);
+      double d0 = blockraytraceresult == null ? Double.MAX_VALUE : p_45548_.m_45702_().m_82557_(blockraytraceresult.m_82450_());
+      double d1 = blockraytraceresult1 == null ? Double.MAX_VALUE : p_45548_.m_45702_().m_82557_(blockraytraceresult1.m_82450_());
+      return d0 <= d1 ? blockraytraceresult : blockraytraceresult1;
+   }
+   // CraftBukkit end
 
    default BlockHitResult m_45547_(ClipContext p_45548_) {
       return m_151361_(p_45548_.m_45702_(), p_45548_.m_45693_(), p_45548_, (p_151359_, p_151360_) -> {
