--- a/net/minecraft/world/level/GameRules.java
+++ b/net/minecraft/world/level/GameRules.java
@@ -195,7 +_,7 @@
          return Boolean.toString(this.f_46219_);
       }
 
-      protected void m_7377_(String p_46234_) {
+      public void m_7377_(String p_46234_) {
          this.f_46219_ = Boolean.parseBoolean(p_46234_);
       }
 
@@ -284,7 +_,7 @@
          return Integer.toString(this.f_46284_);
       }
 
-      protected void m_7377_(String p_46307_) {
+      public void m_7377_(String p_46307_) {
          this.f_46284_ = m_46317_(p_46307_);
       }
 
@@ -405,14 +_,14 @@
          this.m_46368_(p_46371_.getSource().m_81377_());
       }
 
-      protected void m_46368_(@Nullable MinecraftServer p_46369_) {
+      public void m_46368_(@Nullable MinecraftServer p_46369_) {
          if (p_46369_ != null) {
             this.f_46360_.f_46339_.accept(p_46369_, this.m_5589_());
          }
 
       }
 
-      protected abstract void m_7377_(String p_46367_);
+      public abstract void m_7377_(String p_46367_);
 
       public abstract String m_5831_();
 
