--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -112,4 +_,10 @@
    default void m_220407_(GameEvent p_220408_, BlockPos p_220409_, GameEvent.Context p_220410_) {
       this.m_214171_(p_220408_, Vec3.m_82512_(p_220409_), p_220410_);
    }
+
+   // CatServer start - Can't be solved, at least don't throw AbstractMethodError
+   default net.minecraft.server.level.ServerLevel getMinecraftWorld() { // CraftBukkit
+      throw new UnsupportedOperationException(String.format("Unsupported Operation! Please report to https://github.com/Luohuayu/CatServer/issues (World class: %s)", getClass().getName()));
+   }
+   // CatServer end
 }
