--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -5,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.Difficulty;
@@ -99,5 +_,10 @@
 
    default void m_151545_(@Nullable Entity p_151546_, GameEvent p_151547_, Entity p_151548_) {
       this.m_142346_(p_151546_, p_151547_, p_151548_.m_142538_());
+   }
+
+   // CatServer - Can't be solved, at least don't throw AbstractMethodError
+   default ServerLevel getMinecraftWorld() {
+      throw new UnsupportedOperationException(String.format("Unsupported Operation! Please report to https://github.com/Luohuayu/CatServer/issues (World class: %s)", getClass().getName()));
    }
 }
