--- a/net/minecraft/world/level/LevelWriter.java
+++ b/net/minecraft/world/level/LevelWriter.java
@@ -4,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public interface LevelWriter {
    boolean m_6933_(BlockPos p_46947_, BlockState p_46948_, int p_46949_, int p_46950_);
@@ -25,6 +_,10 @@
    boolean m_7740_(BlockPos p_46957_, boolean p_46958_, @Nullable Entity p_46959_, int p_46960_);
 
    default boolean m_7967_(Entity p_46964_) {
+      return false;
+   }
+
+   default boolean addFreshEntity(Entity p_46964_, CreatureSpawnEvent.SpawnReason reason) {
       return false;
    }
 }
