--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -7,6 +_,17 @@
    ServerLevel m_6018_();
 
    default void m_47205_(Entity p_47206_) {
-      p_47206_.m_142428_().forEach(this::m_7967_);
-   }
+      // CraftBukkit start
+      this.addFreshEntityWithPassengers(p_47206_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   default void addFreshEntityWithPassengers(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      entity.m_142428_().forEach((e) -> this.addFreshEntity(e, reason));
+   }
+
+   @Override
+   default ServerLevel getMinecraftWorld() {
+      return m_6018_();
+   }
+   // CraftBukkit end
 }
