--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -7,6 +_,18 @@
    ServerLevel m_6018_();
 
    default void m_47205_(Entity p_47206_) {
-      p_47206_.m_20199_().forEach(this::m_7967_);
-   }
+      org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason = catserver.server.CatServerCaptures.getCatServerCaptures().ServerLevelAccessor$addFreshEntityWithPassengers$reason.get(this);
+      // CraftBukkit start
+      this.addFreshEntityWithPassengers(p_47206_, reason);
+   }
+
+   default void addFreshEntityWithPassengers(Entity p_47206_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      p_47206_.m_20199_().forEach((e) -> this.addFreshEntity(e, reason));
+   }
+
+   @Override
+   default ServerLevel getMinecraftWorld() {
+      return m_6018_();
+   }
+   // CraftBukkit end
 }
