--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -2,6 +_,7 @@
 
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public interface ServerLevelAccessor extends LevelAccessor {
    ServerLevel m_6018_();
@@ -9,4 +_,14 @@
    default void m_47205_(Entity p_47206_) {
       p_47206_.m_142428_().forEach(this::m_7967_);
    }
+
+   default void addFreshEntityWithPassengers(Entity p_47206_, CreatureSpawnEvent.SpawnReason spawnReason) {
+      p_47206_.m_142428_().forEach(this::m_7967_);
+   }
+
+   @Override
+   default ServerLevel getMinecraftWorld() {
+      return m_6018_();
+   }
+   // CraftBukkit end
 }
