--- a/net/minecraft/server/management/BanEntry.java
+++ b/net/minecraft/server/management/BanEntry.java
@@ -23,7 +_,7 @@
    }
 
    protected BanEntry(T p_i1174_1_, JsonObject p_i1174_2_) {
-      super(p_i1174_1_);
+      super(checkExpiry(p_i1174_1_, p_i1174_2_)); // CraftBukkit
 
       Date date;
       try {
@@ -70,4 +_,23 @@
       p_152641_1_.addProperty("expires", this.field_73692_f == null ? "forever" : field_73698_a.format(this.field_73692_f));
       p_152641_1_.addProperty("reason", this.field_73693_g);
    }
+
+   // CraftBukkit start
+   public Date getCreated() {
+      return this.field_73694_d;
+   }
+
+   private static <T> T checkExpiry(final T object, final JsonObject jsonobject) {
+      Date expires = null;
+      try {
+         expires = (jsonobject.has("expires") ? BanEntry.field_73698_a.parse(jsonobject.get("expires").getAsString()) : null);
+      } catch (ParseException ex) {
+      }
+      if (expires == null || expires.after(new Date())) {
+         return object;
+      }
+      return null;
+   }
+   // CraftBukkit end
+
 }
