--- a/net/minecraft/server/management/PreYggdrasilConverter.java
+++ b/net/minecraft/server/management/PreYggdrasilConverter.java
@@ -19,6 +_,8 @@
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.CompressedStreamTools;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.util.StringUtils;
@@ -53,7 +_,7 @@
       }).toArray((p_201149_0_) -> {
          return new String[p_201149_0_];
       });
-      if (p_152717_0_.func_71266_T()) {
+      if (p_152717_0_.func_71266_T() || org.spigotmc.SpigotConfig.bungee) { // Spigot: bungee = online mode, for now.
          p_152717_0_.func_152359_aw().findProfilesByNames(astring, Agent.MINECRAFT, p_152717_2_);
       } else {
          for(String s : astring) {
@@ -72,7 +_,7 @@
             try {
                banlist.func_152679_g();
             } catch (IOException ioexception1) {
-               field_152732_e.warn("Could not load existing file {}", banlist.func_152691_c().getName(), ioexception1);
+               field_152732_e.warn("Could not load existing file {}", banlist.func_152691_c().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -125,7 +_,7 @@
             try {
                ipbanlist.func_152679_g();
             } catch (IOException ioexception1) {
-               field_152732_e.warn("Could not load existing file {}", ipbanlist.func_152691_c().getName(), ioexception1);
+               field_152732_e.warn("Could not load existing file {}", ipbanlist.func_152691_c().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -161,7 +_,7 @@
             try {
                oplist.func_152679_g();
             } catch (IOException ioexception1) {
-               field_152732_e.warn("Could not load existing file {}", oplist.func_152691_c().getName(), ioexception1);
+               field_152732_e.warn("Could not load existing file {}", oplist.func_152691_c().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -203,7 +_,7 @@
             try {
                whitelist.func_152679_g();
             } catch (IOException ioexception1) {
-               field_152732_e.warn("Could not load existing file {}", whitelist.func_152691_c().getName(), ioexception1);
+               field_152732_e.warn("Could not load existing file {}", whitelist.func_152691_c().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -314,6 +_,28 @@
                private void func_152743_a(File p_152743_1_, String p_152743_2_, String p_152743_3_) {
                   File file5 = new File(file1, p_152743_2_ + ".dat");
                   File file6 = new File(p_152743_1_, p_152743_3_ + ".dat");
+                  // CraftBukkit start - Use old file name to seed lastKnownName
+                  CompoundNBT root = null;
+                  try {
+                     root = CompressedStreamTools.func_74796_a(new java.io.FileInputStream(file5));
+                  }
+                  catch (Exception exception) {
+                     exception.printStackTrace();
+                  }
+                  if (root != null) {
+                     if (!root.func_74764_b("bukkit")) {
+                        root.func_218657_a("bukkit", new CompoundNBT());
+                     }
+                     CompoundNBT data = root.func_74775_l("bukkit");
+                     data.func_74778_a("lastKnownName", p_152743_2_);
+                     try {
+                        CompressedStreamTools.func_74799_a(root, new java.io.FileOutputStream(file3));
+                     }
+                     catch (Exception exception2) {
+                        exception2.printStackTrace();
+                     }
+                  }
+                  // CraftBukkit end
                   PreYggdrasilConverter.func_152711_b(p_152743_1_);
                   if (!file5.renameTo(file6)) {
                      throw new PreYggdrasilConverter.ConversionError("Could not convert file for " + p_152743_2_);
