--- a/net/minecraft/server/management/ProfileBanEntry.java
+++ b/net/minecraft/server/management/ProfileBanEntry.java
@@ -2,6 +_,7 @@
 
 import com.google.gson.JsonObject;
 import com.mojang.authlib.GameProfile;
+import java.text.ParseException;
 import java.util.Date;
 import java.util.Objects;
 import java.util.UUID;
@@ -36,19 +_,45 @@
    }
 
    private static GameProfile func_152648_b(JsonObject p_152648_0_) {
-      if (p_152648_0_.has("uuid") && p_152648_0_.has("name")) {
+      // Spigot start
+      // this whole method has to be reworked to account for the fact Bukkit only accepts UUID bans and gives no way for usernames to be stored!
+      UUID uuid = null;
+      String name = null;
+      if (p_152648_0_.has("uuid")) {
          String s = p_152648_0_.get("uuid").getAsString();
 
-         UUID uuid;
          try {
             uuid = UUID.fromString(s);
          } catch (Throwable throwable) {
-            return null;
          }
 
-         return new GameProfile(uuid, p_152648_0_.get("name").getAsString());
+      }
+      if (p_152648_0_.has("name")) {
+         name = p_152648_0_.get("name").getAsString();
+      }
+      if (uuid != null || name != null) {
+         return new GameProfile(uuid, name);
       } else {
          return null;
       }
-   }
+      // Spigot End
+   }
+
+   // CraftBukkit start
+   public Date getCreated() {
+      return this.field_73694_d;
+   }
+
+   private static <T> T checkExpiry(final T object, final JsonObject jsonobject) {
+      Date expires = null;
+      try {
+         expires = (jsonobject.has("expires") ? BanEntry.field_73698_a.parse(jsonobject.get("expires").getAsString()) : null);
+      } catch (ParseException ex) {
+      }
+      if (expires == null || expires.after(new Date())) {
+         return object;
+      }
+      return null;
+   }
+   // CraftBukkit end
 }
