--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -40,7 +_,7 @@
    public static int f_218678_ = Math.abs(-24000);
    public static float f_218681_ = 0.4F;
    public static float f_218682_ = 0.3F;
-   private int f_218680_;
+   public int f_218680_; // CatServer - private -> public
    protected static final ImmutableList<SensorType<? extends Sensor<? super Tadpole>>> f_218683_ = ImmutableList.of(SensorType.f_26811_, SensorType.f_26812_, SensorType.f_26814_, SensorType.f_217822_);
    protected static final ImmutableList<MemoryModuleType<?>> f_218679_ = ImmutableList.of(MemoryModuleType.f_26371_, MemoryModuleType.f_148205_, MemoryModuleType.f_26370_, MemoryModuleType.f_26326_, MemoryModuleType.f_26377_, MemoryModuleType.f_26331_, MemoryModuleType.f_148197_, MemoryModuleType.f_148198_, MemoryModuleType.f_148196_, MemoryModuleType.f_26375_, MemoryModuleType.f_217768_);
 
@@ -208,8 +_,14 @@
             }
 
             frog.m_21530_();
+            // CraftBukkit start - 1.20.1
+            if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTransformEvent(this, frog, org.bukkit.event.entity.EntityTransformEvent.TransformReason.METAMORPHOSIS).isCancelled()) {
+               this.m_218710_(0); // Sets the age to 0 for avoid a loop if the event is canceled
+               return;
+            }
+            // CraftBukkit end
             this.m_5496_(SoundEvents.f_215760_, 0.15F, 1.0F);
-            serverlevel.m_47205_(frog);
+            serverlevel.addFreshEntityWithPassengers(frog, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.METAMORPHOSIS); // CraftBukkit - 1.20.1 add SpawnReason
             this.m_146870_();
          }
       }
