--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -14,6 +_,13 @@
 public class DerivedLevelData implements ServerLevelData {
    private final WorldData f_78076_;
    private final ServerLevelData f_78077_;
+   private String properName;
+
+   public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_, String properName) {
+      this.f_78076_ = p_78079_;
+      this.f_78077_ = p_78080_;
+      this.properName = properName;
+   }
 
    public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
       this.f_78076_ = p_78079_;
@@ -45,7 +_,7 @@
    }
 
    public String m_5462_() {
-      return this.f_78076_.m_5462_();
+      return properName != null ? properName : this.f_78076_.m_5462_();
    }
 
    public int m_6537_() {
@@ -174,4 +_,10 @@
       p_164852_.m_128159_("Derived", true);
       this.f_78077_.m_142471_(p_164852_, p_164853_);
    }
+
+   // CatServer start
+   public ServerLevelData getWrapped() {
+      return f_78077_;
+   }
+   // CatServer end
 }
