--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -14,12 +_,38 @@
 public class DerivedLevelData implements ServerLevelData {
    private final WorldData f_78076_;
    private final ServerLevelData f_78077_;
+   // CatServer start
+   private String properName;
+
+   // CatServer - 1.20.1 set properName in ServerLevel#<init> to ensure properName is not null
+   @Deprecated
+   public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_, String properName) {
+      this.f_78076_ = p_78079_;
+      this.f_78077_ = p_78080_;
+      this.properName = properName;
+   }
+   // CatServer end
 
    public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
       this.f_78076_ = p_78079_;
       this.f_78077_ = p_78080_;
    }
 
+   // CatServer start - [compat-mod-AE2:appeng.mixins.spatial.MinecraftServerMixin#injectSpatialLevel] If a mod creates a DerivedLevelData with vanilla constructor, properName will be [world], and it will replace overworld in CraftServer#worlds.
+   public DerivedLevelData setProperName(net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> levelResourceKey) {
+      if (net.minecraft.world.level.Level.f_46428_.equals(levelResourceKey)) {
+         this.properName = "world";
+      } else if (net.minecraft.world.level.Level.f_46430_.equals(levelResourceKey)) {
+         this.properName = "DIM1";
+      } else if (net.minecraft.world.level.Level.f_46429_.equals(levelResourceKey)){
+         this.properName = "DIM-1";
+      } else {
+         this.properName = levelResourceKey.m_135782_().toString().replace(":", "_");
+      }
+      return this;
+   }
+   // CatServer end
+
    public int m_6789_() {
       return this.f_78077_.m_6789_();
    }
@@ -45,7 +_,7 @@
    }
 
    public String m_5462_() {
-      return this.f_78076_.m_5462_();
+      return (!net.minecraft.util.StringUtil.m_14408_(properName)) ? properName : this.f_78076_.m_5462_(); // CatServer
    }
 
    public int m_6537_() {
@@ -174,4 +_,10 @@
       p_164852_.m_128159_("Derived", true);
       this.f_78077_.m_142471_(p_164852_, p_164853_);
    }
+
+   // CatServer start
+   public ServerLevelData getWrapped() {
+      return f_78077_;
+   }
+   // CatServer end
 }
