--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -12,6 +_,13 @@
 import net.minecraft.world.entity.player.Player;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.io.FileInputStream;
+import java.io.InputStream;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer;
+// CraftBukkit end
+
 public class PlayerDataStorage {
    private static final Logger f_78426_ = LogUtils.getLogger();
    private final File f_78427_;
@@ -24,6 +_,7 @@
    }
 
    public void m_78433_(Player p_78434_) {
+      if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
       try {
          CompoundTag compoundtag = p_78434_.m_20240_(new CompoundTag());
          File file1 = File.createTempFile(p_78434_.m_20149_() + "-", ".dat", this.f_78427_);
@@ -31,8 +_,10 @@
          File file2 = new File(this.f_78427_, p_78434_.m_20149_() + ".dat");
          File file3 = new File(this.f_78427_, p_78434_.m_20149_() + ".dat_old");
          Util.m_137462_(file2, file1, file3);
+         net.minecraftforge.event.ForgeEventFactory.firePlayerSavingEvent(p_78434_, f_78427_, p_78434_.m_20149_());
       } catch (Exception exception) {
          f_78426_.warn("Failed to save player data for {}", (Object)p_78434_.m_7755_().getString());
+         exception.printStackTrace(); // CatServer
       }
 
    }
@@ -51,13 +_,38 @@
       }
 
       if (compoundtag != null) {
+         // CraftBukkit start
+         if (p_78436_ instanceof ServerPlayer) {
+            CraftPlayer player = (CraftPlayer) p_78436_.getBukkitEntity();
+            // Only update first played if it is older than the one we have
+            long modified = new File(this.f_78427_, p_78436_.m_20148_().toString() + ".dat").lastModified();
+            if (modified < player.getFirstPlayed()) {
+               player.setFirstPlayed(modified);
+            }
+         }
+         // CraftBukkit end
          int i = NbtUtils.m_264487_(compoundtag, -1);
          p_78436_.m_20258_(DataFixTypes.PLAYER.m_264218_(this.f_78425_, compoundtag, i));
       }
+      net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(p_78436_, f_78427_, p_78436_.m_20149_());
 
       return compoundtag;
    }
 
+   // CraftBukkit start
+   public CompoundTag getPlayerData(String s) {
+      try {
+         File file1 = new File(this.f_78427_, s + ".dat");
+         if (file1.exists()) {
+            return NbtIo.m_128939_((InputStream) (new FileInputStream(file1)));
+         }
+      } catch (Exception exception) {
+         f_78426_.warn("Failed to load player data for " + s);
+      }
+      return null;
+   }
+   // CraftBukkit end
+
    public String[] m_78432_() {
       String[] astring = this.f_78427_.list();
       if (astring == null) {
@@ -71,5 +_,15 @@
       }
 
       return astring;
+   }
+
+   // CraftBukkit start
+   public File getPlayerDir() {
+      return f_78427_;
+   }
+   // CraftBukkit end
+
+   public File getPlayerDataFolder() {
+      return f_78427_;
    }
 }
