--- a/net/minecraft/state/IntegerProperty.java
+++ b/net/minecraft/state/IntegerProperty.java
@@ -9,8 +_,15 @@
 public class IntegerProperty extends Property<Integer> {
    private final ImmutableSet<Integer> field_177720_a;
 
+   // CraftBukkit start
+   public final int min;
+   public final int max;
+
    protected IntegerProperty(String p_i45648_1_, int p_i45648_2_, int p_i45648_3_) {
       super(p_i45648_1_, Integer.class);
+      this.min = p_i45648_2_;
+      this.max = p_i45648_3_;
+      // CraftBukkit end
       if (p_i45648_2_ < 0) {
          throw new IllegalArgumentException("Min value of " + p_i45648_1_ + " must be 0 or greater");
       } else if (p_i45648_3_ <= p_i45648_2_) {
