--- a/net/minecraft/state/StateHolder.java
+++ b/net/minecraft/state/StateHolder.java
@@ -18,7 +_,7 @@
 import javax.annotation.Nullable;
 
 public abstract class StateHolder<O, S> {
-   private static final Function<Entry<Property<?>, Comparable<?>>, String> field_235890_a_ = new Function<Entry<Property<?>, Comparable<?>>, String>() {
+   public static final Function<Entry<Property<?>, Comparable<?>>, String> field_235890_a_ = new Function<Entry<Property<?>, Comparable<?>>, String>() {
       public String apply(@Nullable Entry<Property<?>, Comparable<?>> p_apply_1_) {
          if (p_apply_1_ == null) {
             return "<NULL>";
