--- ../src-base/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
@@ -137,17 +137,20 @@
 
         if (itextureobject != null)
         {
+            this.mapTextureObjects.remove(textureLocation); // Forge: fix MC-98707
             TextureUtil.deleteTexture(itextureobject.getGlTextureId());
         }
     }
 
     public void onResourceManagerReload(IResourceManager resourceManager)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Reloading Texture Manager", this.mapTextureObjects.keySet().size(), true);
         Iterator<Entry<ResourceLocation, ITextureObject>> iterator = this.mapTextureObjects.entrySet().iterator();
 
         while (iterator.hasNext())
         {
             Entry<ResourceLocation, ITextureObject> entry = (Entry)iterator.next();
+            bar.step(entry.getKey().toString());
             ITextureObject itextureobject = entry.getValue();
 
             if (itextureobject == TextureUtil.MISSING_TEXTURE)
@@ -159,5 +162,6 @@
                 this.loadTexture(entry.getKey(), itextureobject);
             }
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 }
