--- ../src-base/minecraft/net/minecraft/world/demo/DemoWorldServer.java
+++ ../src-work/minecraft/net/minecraft/world/demo/DemoWorldServer.java
@@ -16,7 +16,7 @@
 
     public DemoWorldServer(MinecraftServer server, ISaveHandler saveHandlerIn, WorldInfo worldInfoIn, int dimensionId, Profiler profilerIn)
     {
-        super(server, saveHandlerIn, worldInfoIn, dimensionId, profilerIn);
+        super(server, saveHandlerIn, worldInfoIn, dimensionId, profilerIn, null, null);
         this.worldInfo.populateFromWorldSettings(DEMO_WORLD_SETTINGS);
     }
 }
