--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -88,13 +_,14 @@
    private final VibrationSystem.User f_279563_;
    private final DynamicGameEventListener<Allay.JukeboxListener> f_238563_;
    private final SimpleContainer f_218303_ = new SimpleContainer(1);
-   private @Nullable BlockPos f_238682_;
-   private long f_238791_;
+   public @Nullable BlockPos f_238682_; // PAIL - 1.20.1 private -> public
+   public long f_238791_; // PAIL - 1.20.1 private -> public
    private float f_218304_;
    private float f_218305_;
    private float f_238687_;
    private float f_238541_;
    private float f_238552_;
+   public boolean forceDancing = false; // CraftBukkit - 1.20.1
 
    public Allay(EntityType<? extends Allay> p_218310_, Level p_218311_) {
       super(p_218310_, p_218311_);
@@ -106,6 +_,12 @@
       this.f_238563_ = new DynamicGameEventListener<>(new Allay.JukeboxListener(this.f_279563_.m_280010_(), GameEvent.f_238690_.m_157827_()));
    }
 
+   // CraftBukkit start
+   public void setCanDuplicate(boolean canDuplicate) {
+      this.f_19804_.m_135381_(Allay.f_238802_, canDuplicate);
+   }
+   // CraftBukkit end
+
    protected Brain.Provider<Allay> m_5490_() {
       return Brain.m_21923_(f_218301_, f_218297_);
    }
@@ -207,7 +_,7 @@
    public void m_8107_() {
       super.m_8107_();
       if (!this.m_9236_().f_46443_ && this.m_6084_() && this.f_19797_ % 10 == 0) {
-         this.m_5634_(1.0F);
+         this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
       }
 
       if (this.m_239559_() && this.m_239812_() && this.f_19797_ % 20 == 0) {
@@ -272,7 +_,12 @@
       ItemStack itemstack = p_218361_.m_21120_(p_218362_);
       ItemStack itemstack1 = this.m_21120_(InteractionHand.MAIN_HAND);
       if (this.m_239559_() && this.m_239735_(itemstack) && this.m_218324_()) {
-         this.m_218376_();
+         // CraftBukkit start - handle cancel duplication
+         Allay allay = this.duplicateAllayCB(); // CatServer - this.duplicateAllay() -> this.duplicateAllayCB()
+         if (allay == null) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          this.m_9236_().m_7605_(this, (byte)18);
          this.m_9236_().m_6269_(p_218361_, this, SoundEvents.f_144243_, SoundSource.NEUTRAL, 2.0F, 1.0F);
          this.m_239358_(p_218361_, itemstack);
@@ -324,7 +_,7 @@
 
    public boolean m_7243_(ItemStack p_218387_) {
       ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-      return !itemstack.m_41619_() && this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_) && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_);
+      return !itemstack.m_41619_() && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_) && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this);
    }
 
    private boolean m_245613_(ItemStack p_252278_, ItemStack p_250405_) {
@@ -386,6 +_,7 @@
    }
 
    private boolean m_239812_() {
+      if (this.forceDancing) {return false;} // CraftBukkit
       return this.f_238682_ == null || !this.f_238682_.m_203195_(this.m_20182_(), (double)GameEvent.f_238690_.m_157827_()) || !this.m_9236_().m_8055_(this.f_238682_).m_60713_(Blocks.f_50131_);
    }
 
@@ -463,6 +_,14 @@
       return f_238776_.test(p_239736_);
    }
 
+   // CatServer start - 1.20.1
+   private Allay catserver$duplicateAllayAllay = null;
+   public Allay duplicateAllayCB() { // CraftBukkit - return allay and private -> public
+      this.m_218376_();
+      return this.catserver$duplicateAllayAllay;
+   }
+   // CatServer end
+
    private void m_218376_() {
       Allay allay = EntityType.f_217014_.m_20615_(this.m_9236_());
       if (allay != null) {
@@ -470,17 +_,18 @@
          allay.m_21530_();
          allay.m_239811_();
          this.m_239811_();
-         this.m_9236_().m_7967_(allay);
+         this.m_9236_().addFreshEntity(allay, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DUPLICATION); // CraftBukkit - reason for duplicated allay
       }
 
+      this.catserver$duplicateAllayAllay = allay; // CatServer - 1.20.1
    }
 
-   private void m_239811_() {
+   public void m_239811_() { // PAIL - 1.20.1 private -> public
       this.f_238791_ = 6000L;
       this.f_19804_.m_135381_(f_238802_, false);
    }
 
-   private boolean m_218324_() {
+   public boolean m_218324_() { // PAIL - 1.20.1 private -> public
       return this.f_19804_.m_135370_(f_238802_);
    }
 
