--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderVillager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderVillager.java
@@ -31,22 +31,7 @@
 
     protected ResourceLocation getEntityTexture(EntityVillager entity)
     {
-        switch (entity.getProfession())
-        {
-            case 0:
-                return FARMER_VILLAGER_TEXTURES;
-            case 1:
-                return LIBRARIAN_VILLAGER_TEXTURES;
-            case 2:
-                return PRIEST_VILLAGER_TEXTURES;
-            case 3:
-                return SMITH_VILLAGER_TEXTURES;
-            case 4:
-                return BUTCHER_VILLAGER_TEXTURES;
-            case 5:
-            default:
-                return VILLAGER_TEXTURES;
-        }
+        return entity.getProfessionForge().getSkin();
     }
 
     protected void preRenderCallback(EntityVillager entitylivingbaseIn, float partialTickTime)
