--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderShulker.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderShulker.java
@@ -21,7 +21,7 @@
     public RenderShulker(RenderManager p_i47194_1_)
     {
         super(p_i47194_1_, new ModelShulker(), 0.0F);
-        this.addLayer(new RenderShulker.HeadLayer());
+        this.addLayer(new HeadLayer());
     }
 
     public ModelShulker getMainModel()
