--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderFish.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderFish.java
@@ -75,7 +75,7 @@
             int k = entityplayer.getPrimaryHand() == EnumHandSide.RIGHT ? 1 : -1;
             ItemStack itemstack = entityplayer.getHeldItemMainhand();
 
-            if (itemstack.getItem() != Items.FISHING_ROD)
+            if (!(itemstack.getItem() instanceof net.minecraft.item.ItemFishingRod))
             {
                 k = -k;
             }
