--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderZombieVillager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderZombieVillager.java
@@ -25,6 +25,7 @@
 
     protected ResourceLocation getEntityTexture(EntityZombieVillager entity)
     {
+        if (true) return entity.getForgeProfession().getZombieSkin();
         switch (entity.getProfession())
         {
             case 0:
