--- a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
+++ b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
@@ -29,6 +_,12 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_20_R1.CraftGameEvent;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.event.block.BlockReceiveGameEvent;
+// CraftBukkit end
+
 public interface VibrationSystem {
    GameEvent[] f_279664_ = new GameEvent[]{GameEvent.f_276553_, GameEvent.f_276530_, GameEvent.f_276691_, GameEvent.f_276533_, GameEvent.f_276695_, GameEvent.f_276518_, GameEvent.f_276655_, GameEvent.f_276494_, GameEvent.f_276419_, GameEvent.f_276431_, GameEvent.f_276621_, GameEvent.f_276548_, GameEvent.f_276569_, GameEvent.f_276454_, GameEvent.f_276436_};
    ToIntFunction<GameEvent> f_279561_ = Util.m_137469_(new Object2IntOpenHashMap<>(), (p_282267_) -> {
@@ -185,7 +_,14 @@
                return false;
             } else {
                Vec3 vec3 = optional.get();
-               if (!vibrationsystem$user.m_280080_(p_282254_, BlockPos.m_274446_(p_282426_), p_283599_, p_283664_)) {
+               // CraftBukkit start
+               boolean defaultCancel = !vibrationsystem$user.m_280080_(p_282254_, BlockPos.m_274446_(p_282426_), p_283599_, p_283664_);
+               Entity entity = p_283664_.f_223711_();
+               BlockReceiveGameEvent event = new BlockReceiveGameEvent(CraftGameEvent.minecraftToBukkit(p_283599_), CraftBlock.at(p_282254_, BlockPos.m_274446_(vec3)), (entity == null) ? null : entity.getBukkitEntity());
+               event.setCancelled(defaultCancel);
+               p_282254_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  // CraftBukkit end
                   return false;
                } else if (m_280258_(p_282254_, p_282426_, vec3)) {
                   return false;
