--- a/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java
+++ b/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java
@@ -3,6 +_,7 @@
 import java.util.Optional;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Registry;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.GameEventTags;
@@ -15,6 +_,9 @@
 import net.minecraft.world.level.gameevent.PositionSource;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey;
+import org.bukkit.event.block.BlockReceiveGameEvent;
 
 public class VibrationListener implements GameEventListener {
    protected final PositionSource f_157887_;
@@ -59,7 +_,14 @@
             return false;
          } else {
             BlockPos blockpos = optional.get();
-            if (!this.f_157889_.m_142008_(p_157901_, this, p_157904_, p_157902_, p_157903_)) {
+            //if (!this.config.shouldListen(p_157901_, this, p_157904_, p_157902_, p_157903_)) {
+            // CraftBukkit start
+            boolean defaultCancel = !this.f_157889_.m_142008_(p_157901_, this, p_157904_, p_157902_, p_157903_);
+            BlockReceiveGameEvent event = new BlockReceiveGameEvent(org.bukkit.GameEvent.getByKey(CraftNamespacedKey.fromMinecraft(Registry.f_175412_.m_7981_(p_157902_))), CraftBlock.at(p_157901_, blockpos), (p_157903_ == null) ? null : p_157903_.getBukkitEntity());
+            event.setCancelled(defaultCancel);
+            p_157901_.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               // CraftBukkit end
                return false;
             } else if (this.m_157910_(p_157901_, p_157904_, blockpos)) {
                return false;
