--- ../src-base/minecraft/net/minecraft/util/datafix/walkers/BlockEntityTag.java
+++ ../src-work/minecraft/net/minecraft/util/datafix/walkers/BlockEntityTag.java
@@ -42,7 +42,8 @@
 
                 if (s1 == null)
                 {
-                    LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", (Object)s);
+                    // CraftBukkit - Remove unnecessary warning (occurs when deserializing a Shulker Box item)
+                    // LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", (Object)s);
                     flag = false;
                 }
                 else
