--- a/net/minecraft/entity/AgeableEntity.java
+++ b/net/minecraft/entity/AgeableEntity.java
@@ -16,11 +_,36 @@
    protected int field_175504_a;
    protected int field_175502_b;
    protected int field_175503_c;
+   public boolean ageLocked; // CraftBukkit
 
    protected AgeableEntity(EntityType<? extends AgeableEntity> p_i48581_1_, World p_i48581_2_) {
       super(p_i48581_1_, p_i48581_2_);
    }
 
+   // Spigot start
+   @Override
+   public void inactiveTick()
+   {
+      super.inactiveTick();
+      if ( this.field_70170_p.field_72995_K || this.ageLocked )
+      { // CraftBukkit
+         this.func_213323_x_();
+      } else
+      {
+         int i = this.func_70874_b();
+         if ( i < 0 )
+         {
+            ++i;
+            this.func_70873_a( i );
+         } else if ( i > 0 )
+         {
+            --i;
+            this.func_70873_a( i );
+         }
+      }
+   }
+   // Spigot end
+
    public ILivingEntityData func_213386_a(IServerWorld p_213386_1_, DifficultyInstance p_213386_2_, SpawnReason p_213386_3_, @Nullable ILivingEntityData p_213386_4_, @Nullable CompoundNBT p_213386_5_) {
       if (p_213386_4_ == null) {
          p_213386_4_ = new AgeableEntity.AgeableData(true);
@@ -95,12 +_,14 @@
       super.func_213281_b(p_213281_1_);
       p_213281_1_.func_74768_a("Age", this.func_70874_b());
       p_213281_1_.func_74768_a("ForcedAge", this.field_175502_b);
+      p_213281_1_.func_74757_a("AgeLocked", this.ageLocked); // CraftBukkit
    }
 
    public void func_70037_a(CompoundNBT p_70037_1_) {
       super.func_70037_a(p_70037_1_);
       this.func_70873_a(p_70037_1_.func_74762_e("Age"));
       this.field_175502_b = p_70037_1_.func_74762_e("ForcedAge");
+      this.ageLocked = p_70037_1_.func_74767_n("AgeLocked"); // CraftBukkit
    }
 
    public void func_184206_a(DataParameter<?> p_184206_1_) {
@@ -113,7 +_,7 @@
 
    public void func_70636_d() {
       super.func_70636_d();
-      if (this.field_70170_p.field_72995_K) {
+      if (this.field_70170_p.field_72995_K || ageLocked) { // CraftBukkit
          if (this.field_175503_c > 0) {
             if (this.field_175503_c % 4 == 0) {
                this.field_70170_p.func_195594_a(ParticleTypes.field_197632_y, this.func_226282_d_(1.0D), this.func_226279_cv_() + 0.5D, this.func_226287_g_(1.0D), 0.0D, 0.0D, 0.0D);
