--- a/net/minecraft/entity/BoostHelper.java
+++ b/net/minecraft/entity/BoostHelper.java
@@ -37,6 +_,15 @@
       }
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.field_233610_a_ = true;
+      this.field_233611_b_ = 0;
+      this.field_233612_c_ = ticks;
+      this.field_233613_d_.func_187227_b(this.field_233614_e_, this.field_233612_c_);
+   }
+   // CraftBukkit end
+
    public void func_233618_a_(CompoundNBT p_233618_1_) {
       p_233618_1_.func_74757_a("Saddle", this.func_233620_b_());
    }
