--- a/net/minecraft/entity/CreatureEntity.java
+++ b/net/minecraft/entity/CreatureEntity.java
@@ -7,6 +_,8 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
 
 public abstract class CreatureEntity extends MobEntity {
    protected CreatureEntity(EntityType<? extends CreatureEntity> p_i48575_1_, World p_i48575_2_) {
@@ -37,6 +_,7 @@
          float f = this.func_70032_d(entity);
          if (this instanceof TameableEntity && ((TameableEntity)this).func_233684_eK_()) {
             if (f > 10.0F) {
+               this.field_70170_p.getCBServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.DISTANCE)); // CraftBukkit
                this.func_110160_i(true, true);
             }
 
@@ -45,6 +_,7 @@
 
          this.func_142017_o(f);
          if (f > 10.0F) {
+            this.field_70170_p.getCBServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.DISTANCE)); // CraftBukkit
             this.func_110160_i(true, true);
             this.field_70714_bg.func_220880_a(Goal.Flag.MOVE);
          } else if (f > 6.0F) {
