--- a/net/minecraft/entity/IAngerable.java
+++ b/net/minecraft/entity/IAngerable.java
@@ -104,7 +_,7 @@
    default void func_241356_K__() {
       this.func_70604_c((LivingEntity)null);
       this.func_230259_a_((UUID)null);
-      this.func_70624_b((LivingEntity)null);
+      this.setGoalTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.func_230260_a__(0);
    }
 
@@ -113,6 +_,8 @@
    void func_230246_e_(@Nullable PlayerEntity p_230246_1_);
 
    void func_70624_b(@Nullable LivingEntity p_70624_1_);
+
+   boolean setGoalTarget(LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent); // CraftBukkit
 
    @Nullable
    LivingEntity func_70638_az();
