--- a/net/minecraft/entity/NPCMerchant.java
+++ b/net/minecraft/entity/NPCMerchant.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftMerchant;
 
 public class NPCMerchant implements IMerchant {
    private final MerchantInventory field_70937_a;
@@ -66,5 +_,10 @@
 
    public SoundEvent func_213714_ea() {
       return SoundEvents.field_187915_go;
+   }
+
+   @Override
+   public CraftMerchant getCraftMerchant() {
+      return null;
    }
 }
