--- ../src-base/minecraft/net/minecraft/client/gui/recipebook/GhostRecipe.java
+++ ../src-work/minecraft/net/minecraft/client/gui/recipebook/GhostRecipe.java
@@ -20,7 +20,7 @@
 public class GhostRecipe
 {
     private IRecipe recipe;
-    private final List<GhostRecipe.GhostIngredient> ingredients = Lists.<GhostRecipe.GhostIngredient>newArrayList();
+    private final List<GhostIngredient> ingredients = Lists.<GhostIngredient>newArrayList();
     private float time;
 
     public void clear()
@@ -32,10 +32,10 @@
 
     public void addIngredient(Ingredient p_194187_1_, int p_194187_2_, int p_194187_3_)
     {
-        this.ingredients.add(new GhostRecipe.GhostIngredient(p_194187_1_, p_194187_2_, p_194187_3_));
+        this.ingredients.add(new GhostIngredient(p_194187_1_, p_194187_2_, p_194187_3_));
     }
 
-    public GhostRecipe.GhostIngredient get(int p_192681_1_)
+    public GhostIngredient get(int p_192681_1_)
     {
         return this.ingredients.get(p_192681_1_);
     }
@@ -68,7 +68,7 @@
 
         for (int i = 0; i < this.ingredients.size(); ++i)
         {
-            GhostRecipe.GhostIngredient ghostrecipe$ghostingredient = this.ingredients.get(i);
+            GhostIngredient ghostrecipe$ghostingredient = this.ingredients.get(i);
             int j = ghostrecipe$ghostingredient.getX() + p_194188_2_;
             int k = ghostrecipe$ghostingredient.getY() + p_194188_3_;
 
