--- ../src-base/minecraft/net/minecraft/client/util/JsonException.java
+++ ../src-work/minecraft/net/minecraft/client/util/JsonException.java
@@ -11,31 +11,31 @@
 @SideOnly(Side.CLIENT)
 public class JsonException extends IOException
 {
-    private final List<JsonException.Entry> entries = Lists.<JsonException.Entry>newArrayList();
+    private final List<Entry> entries = Lists.<Entry>newArrayList();
     private final String message;
 
     public JsonException(String messageIn)
     {
-        this.entries.add(new JsonException.Entry());
+        this.entries.add(new Entry());
         this.message = messageIn;
     }
 
     public JsonException(String messageIn, Throwable cause)
     {
         super(cause);
-        this.entries.add(new JsonException.Entry());
+        this.entries.add(new Entry());
         this.message = messageIn;
     }
 
     public void prependJsonKey(String key)
     {
-        ((JsonException.Entry)this.entries.get(0)).addJsonKey(key);
+        ((Entry)this.entries.get(0)).addJsonKey(key);
     }
 
     public void setFilenameAndFlush(String filenameIn)
     {
         (this.entries.get(0)).filename = filenameIn;
-        this.entries.add(0, new JsonException.Entry());
+        this.entries.add(0, new Entry());
     }
 
     public String getMessage()
