--- ../src-base/minecraft/net/minecraft/client/util/RecipeBookClient.java
+++ ../src-work/minecraft/net/minecraft/client/util/RecipeBookClient.java
@@ -54,6 +54,14 @@
 
     static
     {
+        rebuildTable();
+    }
+
+    public static void rebuildTable()
+    {
+        RECIPES_BY_TAB.clear();
+        ALL_RECIPES.clear();
+
         Table<CreativeTabs, String, RecipeList> table = HashBasedTable.<CreativeTabs, String, RecipeList>create();
 
         for (IRecipe irecipe : CraftingManager.REGISTRY)
