--- ../src-base/minecraft/net/minecraft/client/util/RecipeItemHelper.java
+++ ../src-work/minecraft/net/minecraft/client/util/RecipeItemHelper.java
@@ -24,10 +24,15 @@
 
     public void accountStack(ItemStack stack)
     {
+        this.accountStack(stack, -1);
+    }
+
+    public void accountStack(ItemStack stack, int forceCount)
+    {
         if (!stack.isEmpty() && !stack.isItemDamaged() && !stack.isItemEnchanted() && !stack.hasDisplayName())
         {
             int i = pack(stack);
-            int j = stack.getCount();
+            int j = forceCount == -1 ? stack.getCount() : forceCount;
             this.increment(i, j);
         }
     }
@@ -71,7 +76,7 @@
 
     public boolean canCraft(IRecipe recipe, @Nullable IntList p_194118_2_, int p_194118_3_)
     {
-        return (new RecipeItemHelper.RecipePicker(recipe)).tryPick(p_194118_3_, p_194118_2_);
+        return (new RecipePicker(recipe)).tryPick(p_194118_3_, p_194118_2_);
     }
 
     public int getBiggestCraftableStack(IRecipe recipe, @Nullable IntList p_194114_2_)
@@ -81,7 +86,7 @@
 
     public int getBiggestCraftableStack(IRecipe recipe, int p_194121_2_, @Nullable IntList p_194121_3_)
     {
-        return (new RecipeItemHelper.RecipePicker(recipe)).tryPickAll(p_194121_2_, p_194121_3_);
+        return (new RecipePicker(recipe)).tryPickAll(p_194121_2_, p_194121_3_);
     }
 
     public static ItemStack unpack(int p_194115_0_)
