--- ../src-base/minecraft/net/minecraft/client/util/SearchTree.java
+++ ../src-work/minecraft/net/minecraft/client/util/SearchTree.java
@@ -73,7 +73,7 @@
         else
         {
             List<T> list1 = this.byName.search(searchText);
-            return (List<T>)(list1.isEmpty() ? list : Lists.newArrayList(new SearchTree.MergingIterator(list.iterator(), list1.iterator(), this.numericContents)));
+            return (List<T>)(list1.isEmpty() ? list : Lists.newArrayList(new MergingIterator(list.iterator(), list1.iterator(), this.numericContents)));
         }
     }
 
