--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java
@@ -39,7 +39,10 @@
 import net.minecraftforge.event.RegistryEvent;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.Level;
+import org.bukkit.Material;
 
+import net.minecraft.block.Block;
+import net.minecraft.item.Item;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.RegistryNamespacedDefaultedByKey;
 import net.minecraftforge.fml.common.FMLLog;
@@ -480,6 +483,9 @@
 
 
         addObjectRaw(idToUse, name, thing);
+        if((thing instanceof Block || thing instanceof Item) && !name.getResourceDomain().equals("minecraft")) {
+        	Material.addEnum(name.toString(), idToUse);
+        }
         if (isDelegated)
         {
             getExistingDelegate(thing).setName(name);
@@ -798,12 +804,12 @@
     }
 
     @Override
-    public ResourceLocation getNameForObject(I p_177774_1_)
+    public ResourceLocation getNameForObject(I value)
     {
-        ResourceLocation rl = super.getNameForObjectBypass(p_177774_1_);
+        ResourceLocation rl = super.getNameForObjectBypass(value);
         if (rl == null)
         {
-            rl = activeSubstitutions.inverse().get(p_177774_1_);
+            rl = activeSubstitutions.inverse().get(value);
         }
         return rl;
     }
