--- a/net/minecraft/network/play/client/CCustomPayloadPacket.java
+++ b/net/minecraft/network/play/client/CCustomPayloadPacket.java
@@ -1,5 +_,6 @@
 package net.minecraft.network.play.client;
 
+import catserver.server.CatServer;
 import io.netty.buffer.ByteBuf;
 import java.io.IOException;
 import net.minecraft.network.IPacket;
@@ -9,10 +_,10 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class CCustomPayloadPacket implements IPacket<IServerPlayNetHandler> {
+public class CCustomPayloadPacket implements IPacket<IServerPlayNetHandler>, net.minecraftforge.fml.network.ICustomPacket<CCustomPayloadPacket> {
    public static final ResourceLocation field_210344_a = new ResourceLocation("brand");
-   private ResourceLocation field_149562_a;
-   private PacketBuffer field_149561_c;
+   public ResourceLocation field_149562_a;
+   public PacketBuffer field_149561_c;
 
    public CCustomPayloadPacket() {
    }
@@ -35,10 +_,17 @@
 
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
       p_148840_1_.func_192572_a(this.field_149562_a);
-      p_148840_1_.writeBytes((ByteBuf)this.field_149561_c);
+      p_148840_1_.writeBytes((ByteBuf)this.field_149561_c.copy()); //This may be access multiple times, from multiple threads, lets be safe like the S->C packet
    }
 
    public void func_148833_a(IServerPlayNetHandler p_148833_1_) {
+      // CatServer start
+      if ("FML|HS".equalsIgnoreCase(field_149562_a.func_110623_a()) || "FML|MP".equalsIgnoreCase(field_149562_a.func_110623_a())) {
+         CatServer.log.warn("Unsupported Forge client or proxy detected! (Player: {} Channel: {})", ((net.minecraft.network.play.ServerPlayNetHandler)p_148833_1_).field_147369_b.func_200200_C_().getString(), field_149562_a.func_110623_a());
+         if (this.field_149561_c != null) this.field_149561_c.release();
+         return;
+      }
+      // CatServer end
       p_148833_1_.func_147349_a(this);
       if (this.field_149561_c != null) {
          this.field_149561_c.release();
