--- a/net/minecraft/network/play/client/CPlayerTryUseItemOnBlockPacket.java
+++ b/net/minecraft/network/play/client/CPlayerTryUseItemOnBlockPacket.java
@@ -12,6 +_,7 @@
 public class CPlayerTryUseItemOnBlockPacket implements IPacket<IServerPlayNetHandler> {
    private BlockRayTraceResult field_218795_a;
    private Hand field_187027_c;
+   public long timestamp;
 
    public CPlayerTryUseItemOnBlockPacket() {
    }
@@ -23,6 +_,7 @@
    }
 
    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.field_187027_c = p_148837_1_.func_179257_a(Hand.class);
       this.field_218795_a = p_148837_1_.func_218669_q();
    }
