--- a/net/minecraft/network/play/client/CPlayerTryUseItemPacket.java
+++ b/net/minecraft/network/play/client/CPlayerTryUseItemPacket.java
@@ -8,6 +_,7 @@
 
 public class CPlayerTryUseItemPacket implements IPacket<IServerPlayNetHandler> {
    private Hand field_187029_a;
+   public long timestamp; // Spigot
 
    public CPlayerTryUseItemPacket() {
    }
@@ -17,6 +_,7 @@
    }
 
    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.field_187029_a = p_148837_1_.func_179257_a(Hand.class);
    }
 
