--- a/net/minecraft/network/play/client/CResourcePackStatusPacket.java
+++ b/net/minecraft/network/play/client/CResourcePackStatusPacket.java
@@ -6,7 +_,7 @@
 import net.minecraft.network.play.IServerPlayNetHandler;
 
 public class CResourcePackStatusPacket implements IPacket<IServerPlayNetHandler> {
-   private CResourcePackStatusPacket.Action field_179719_b;
+   public CResourcePackStatusPacket.Action field_179719_b;
 
    public CResourcePackStatusPacket() {
    }
