--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -4,19 +_,37 @@
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchantRecipe;
 
 public class MerchantOffer {
    public ItemStack f_45310_;
    public ItemStack f_45311_;
    public final ItemStack f_45312_;
    public int f_45313_;
-   public final int f_45314_;
+   public int f_45314_;
    public boolean f_45315_ = true;
-   private int f_45316_;
-   private int f_45317_;
+   public int f_45316_;
+   public int f_45317_;
    public float f_45318_;
    public int f_45319_ = 1;
 
+   // CraftBukkit start
+   private CraftMerchantRecipe bukkitHandle; // CraftBukkit
+
+   public CraftMerchantRecipe asBukkit() {
+      return (this.bukkitHandle == null) ? (this.bukkitHandle = new CraftMerchantRecipe(this)) : this.bukkitHandle;
+   }
+
+   public MerchantOffer(ItemStack itemstack, ItemStack itemstack1, ItemStack itemstack2, int uses, int maxUses, int experience, float priceMultiplier, CraftMerchantRecipe bukkit) {
+      this(itemstack, itemstack1, itemstack2, uses, maxUses, experience, priceMultiplier, 0, bukkit);
+   }
+
+   public MerchantOffer(ItemStack itemstack, ItemStack itemstack1, ItemStack itemstack2, int uses, int maxUses, int experience, float priceMultiplier, int demand, CraftMerchantRecipe bukkit) {
+      this(itemstack, itemstack1, itemstack2, uses, maxUses, experience, priceMultiplier, demand);
+      this.bukkitHandle = bukkit;
+   }
+   // CraftBukkit end
+
    public MerchantOffer(CompoundTag p_45351_) {
       this.f_45310_ = ItemStack.m_41712_(p_45351_.m_128469_("buy"));
       this.f_45311_ = ItemStack.m_41712_(p_45351_.m_128469_("buyB"));
@@ -73,9 +_,10 @@
 
    public ItemStack m_45358_() {
       int i = this.f_45310_.m_41613_();
+      if (i <= 0) return ItemStack.f_41583_; // CraftBukkit - SPIGOT-5476
       ItemStack itemstack = this.f_45310_.m_41777_();
       int j = Math.max(0, Mth.m_14143_((float)(i * this.f_45317_) * this.f_45318_));
-      itemstack.m_41764_(Mth.m_14045_(i + j + this.f_45316_, 1, this.f_45310_.m_41720_().m_41459_()));
+      itemstack.m_41764_(Mth.m_14045_(i + j + this.f_45316_, 1, this.f_45310_.m_41741_()));
       return itemstack;
    }
 
@@ -179,7 +_,7 @@
          return true;
       } else {
          ItemStack itemstack = p_45366_.m_41777_();
-         if (itemstack.m_41720_().m_41465_()) {
+         if (itemstack.m_41720_().isDamageable(itemstack)) {
             itemstack.m_41721_(itemstack.m_41773_());
          }
 
@@ -191,7 +_,11 @@
       if (!this.m_45355_(p_45362_, p_45363_)) {
          return false;
       } else {
-         p_45362_.m_41774_(this.m_45358_().m_41613_());
+         // CraftBukkit start
+         if (!this.m_45358_().m_41619_()) {
+            p_45362_.m_41774_(this.m_45358_().m_41613_());
+         }
+         // CraftBukkit end
          if (!this.m_45364_().m_41619_()) {
             p_45363_.m_41774_(this.m_45364_().m_41613_());
          }
