--- ../src-base/minecraft/net/minecraft/server/integrated/IntegratedPlayerList.java
+++ ../src-work/minecraft/net/minecraft/server/integrated/IntegratedPlayerList.java
@@ -31,7 +31,7 @@
 
     public String allowUserToConnect(SocketAddress address, GameProfile profile)
     {
-        return profile.getName().equalsIgnoreCase(this.getServerInstance().getServerOwner()) && this.getPlayerByUsername(profile.getName()) != null ? "That name is already taken." : super.allowUserToConnect(address, profile);
+        return (String) (profile.getName().equalsIgnoreCase(this.getServerInstance().getServerOwner()) && this.getPlayerByUsername(profile.getName()) != null ? "That name is already taken." : super.allowUserToConnect(null, /*address*/ null, /*profile*/ null));
     }
 
     public IntegratedServer getServerInstance()
