--- ../src-base/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
@@ -44,7 +44,7 @@
         return this.world.checkNoEntityCollision(this.getEntityBoundingBox(), this) && this.world.getCollisionBoxes(this, this.getEntityBoundingBox()).isEmpty() && !this.world.containsAnyLiquid(this.getEntityBoundingBox());
     }
 
-    protected void setSlimeSize(int size, boolean resetHealth)
+    public void setSlimeSize(int size, boolean resetHealth)
     {
         super.setSlimeSize(size, resetHealth);
         this.getEntityAttribute(SharedMonsterAttributes.ARMOR).setBaseValue((double)(size * 3));
@@ -96,6 +96,7 @@
     {
         this.motionY = (double)(0.42F + (float)this.getSlimeSize() * 0.1F);
         this.isAirBorne = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     protected void handleJumpLava()
