--- ../src-base/minecraft/net/minecraft/entity/monster/EntityBlaze.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityBlaze.java
@@ -55,7 +55,7 @@
 
     protected void initEntityAI()
     {
-        this.tasks.addTask(4, new EntityBlaze.AIFireballAttack(this));
+        this.tasks.addTask(4, new AIFireballAttack(this));
         this.tasks.addTask(5, new EntityAIMoveTowardsRestriction(this, 1.0D));
         this.tasks.addTask(7, new EntityAIWanderAvoidWater(this, 1.0D, 0.0F));
         this.tasks.addTask(8, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
