--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGhast.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGhast.java
@@ -42,14 +42,14 @@
         this.setSize(4.0F, 4.0F);
         this.isImmuneToFire = true;
         this.experienceValue = 5;
-        this.moveHelper = new EntityGhast.GhastMoveHelper(this);
+        this.moveHelper = new GhastMoveHelper(this);
     }
 
     protected void initEntityAI()
     {
-        this.tasks.addTask(5, new EntityGhast.AIRandomFly(this));
-        this.tasks.addTask(7, new EntityGhast.AILookAround(this));
-        this.tasks.addTask(7, new EntityGhast.AIFireballAttack(this));
+        this.tasks.addTask(5, new AIRandomFly(this));
+        this.tasks.addTask(7, new AILookAround(this));
+        this.tasks.addTask(7, new AIFireballAttack(this));
         this.targetTasks.addTask(1, new EntityAIFindEntityNearestPlayer(this));
     }
 
@@ -225,7 +225,8 @@
                         double d4 = entitylivingbase.posZ - (this.parentEntity.posZ + vec3d.z * 4.0D);
                         world.playEvent((EntityPlayer)null, 1016, new BlockPos(this.parentEntity), 0);
                         EntityLargeFireball entitylargefireball = new EntityLargeFireball(world, this.parentEntity, d2, d3, d4);
-                        entitylargefireball.explosionPower = this.parentEntity.getFireballStrength();
+                        // CraftBukkit - set bukkitYield when setting explosionpower
+                        entitylargefireball.bukkitYield = entitylargefireball.explosionPower = this.parentEntity.getFireballStrength();
                         entitylargefireball.posX = this.parentEntity.posX + vec3d.x * 4.0D;
                         entitylargefireball.posY = this.parentEntity.posY + (double)(this.parentEntity.height / 2.0F) + 0.5D;
                         entitylargefireball.posZ = this.parentEntity.posZ + vec3d.z * 4.0D;
@@ -336,7 +337,7 @@
 
             public void onUpdateMoveHelper()
             {
-                if (this.action == EntityMoveHelper.Action.MOVE_TO)
+                if (this.action == Action.MOVE_TO)
                 {
                     double d0 = this.posX - this.parentEntity.posX;
                     double d1 = this.posY - this.parentEntity.posY;
@@ -356,7 +357,7 @@
                         }
                         else
                         {
-                            this.action = EntityMoveHelper.Action.WAIT;
+                            this.action = Action.WAIT;
                         }
                     }
                 }
