--- ../src-base/minecraft/net/minecraft/entity/monster/EntityPigZombie.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityPigZombie.java
@@ -31,7 +31,7 @@
 {
     private static final UUID ATTACK_SPEED_BOOST_MODIFIER_UUID = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
     private static final AttributeModifier ATTACK_SPEED_BOOST_MODIFIER = (new AttributeModifier(ATTACK_SPEED_BOOST_MODIFIER_UUID, "Attacking speed boost", 0.05D, 0)).setSaved(false);
-    private int angerLevel;
+    public int angerLevel;
     private int randomSoundDelay;
     private UUID angerTargetUUID;
 
@@ -53,8 +53,8 @@
 
     protected void applyEntityAI()
     {
-        this.targetTasks.addTask(1, new EntityPigZombie.AIHurtByAggressor(this));
-        this.targetTasks.addTask(2, new EntityPigZombie.AITargetAggressor(this));
+        this.targetTasks.addTask(1, new AIHurtByAggressor(this));
+        this.targetTasks.addTask(2, new AITargetAggressor(this));
     }
 
     protected void applyEntityAttributes()
