--- ../src-base/minecraft/net/minecraft/entity/monster/EntityZombie.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityZombie.java
@@ -687,7 +687,7 @@
         }
     }
 
-    protected boolean canDespawn()
+    public boolean canDespawn()
     {
         return !this.isConverting();
     }
@@ -766,7 +766,7 @@
         this.multiplySize(isChild ? 0.5F : 1.0F);
     }
 
-    protected final void setSize(float width, float height)
+    public final void setSize(float width, float height)
     {
         boolean flag = this.zombieWidth > 0.0F && this.zombieHeight > 0.0F;
         this.zombieWidth = width;
