--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGuardian.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGuardian.java
@@ -57,7 +57,7 @@
         super(worldIn);
         this.experienceValue = 10;
         this.setSize(0.85F, 0.85F);
-        this.moveHelper = new EntityGuardian.GuardianMoveHelper(this);
+        this.moveHelper = new GuardianMoveHelper(this);
         this.clientSideTailAnimation = this.rand.nextFloat();
         this.clientSideTailAnimationO = this.clientSideTailAnimation;
     }
@@ -66,7 +66,7 @@
     {
         EntityAIMoveTowardsRestriction entityaimovetowardsrestriction = new EntityAIMoveTowardsRestriction(this, 1.0D);
         this.wander = new EntityAIWander(this, 1.0D, 80);
-        this.tasks.addTask(4, new EntityGuardian.AIGuardianAttack(this));
+        this.tasks.addTask(4, new AIGuardianAttack(this));
         this.tasks.addTask(5, entityaimovetowardsrestriction);
         this.tasks.addTask(7, this.wander);
         this.tasks.addTask(8, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
@@ -74,7 +74,7 @@
         this.tasks.addTask(9, new EntityAILookIdle(this));
         this.wander.setMutexBits(3);
         entityaimovetowardsrestriction.setMutexBits(3);
-        this.targetTasks.addTask(1, new EntityAINearestAttackableTarget(this, EntityLivingBase.class, 10, true, false, new EntityGuardian.GuardianTargetSelector(this)));
+        this.targetTasks.addTask(1, new EntityAINearestAttackableTarget(this, EntityLivingBase.class, 10, true, false, new GuardianTargetSelector(this)));
     }
 
     protected void applyEntityAttributes()
@@ -504,7 +504,7 @@
 
             public void onUpdateMoveHelper()
             {
-                if (this.action == EntityMoveHelper.Action.MOVE_TO && !this.entityGuardian.getNavigator().noPath())
+                if (this.action == Action.MOVE_TO && !this.entityGuardian.getNavigator().noPath())
                 {
                     double d0 = this.posX - this.entityGuardian.posX;
                     double d1 = this.posY - this.entityGuardian.posY;
