--- ../src-base/minecraft/net/minecraft/entity/monster/EntityVindicator.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityVindicator.java
@@ -65,7 +65,7 @@
         this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
         this.targetTasks.addTask(3, new EntityAINearestAttackableTarget(this, EntityVillager.class, true));
         this.targetTasks.addTask(3, new EntityAINearestAttackableTarget(this, EntityIronGolem.class, true));
-        this.targetTasks.addTask(4, new EntityVindicator.AIJohnnyAttack(this));
+        this.targetTasks.addTask(4, new AIJohnnyAttack(this));
     }
 
     protected void applyEntityAttributes()
@@ -111,7 +111,7 @@
     @SideOnly(Side.CLIENT)
     public AbstractIllager.IllagerArmPose getArmPose()
     {
-        return this.isAggressive() ? AbstractIllager.IllagerArmPose.ATTACKING : AbstractIllager.IllagerArmPose.CROSSED;
+        return this.isAggressive() ? IllagerArmPose.ATTACKING : IllagerArmPose.CROSSED;
     }
 
     public void readEntityFromNBT(NBTTagCompound compound)
