--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGuardian.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGuardian.java
@@ -55,7 +55,7 @@
     private EntityLivingBase targetedEntity;
     private int clientSideAttackTime;
     private boolean clientSideTouchedGround;
-    private EntityAIWander wander;
+    public EntityAIWander wander;
 
     public EntityGuardian(World worldIn)
     {
@@ -82,7 +82,7 @@
         this.targetTasks.addTask(1, new EntityAINearestAttackableTarget(this, EntityLivingBase.class, 10, true, false, new EntityGuardian.GuardianTargetSelector(this)));
     }
 
-    protected void applyEntityAttributes()
+    public void applyEntityAttributes()
     {
         super.applyEntityAttributes();
         this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).setBaseValue(6.0D);
