--- a/net/minecraft/loot/LootParameters.java
+++ b/net/minecraft/loot/LootParameters.java
@@ -7,6 +_,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.vector.Vector3d;
 
 public class LootParameters {
@@ -15,11 +_,15 @@
    public static final LootParameter<DamageSource> field_216283_c = func_216280_a("damage_source");
    public static final LootParameter<Entity> field_216284_d = func_216280_a("killer_entity");
    public static final LootParameter<Entity> field_216285_e = func_216280_a("direct_killer_entity");
+   public static final LootParameter<BlockPos> POSITION = func_216280_a("position");
    public static final LootParameter<Vector3d> field_237457_g_ = func_216280_a("origin");
    public static final LootParameter<BlockState> field_216287_g = func_216280_a("block_state");
    public static final LootParameter<TileEntity> field_216288_h = func_216280_a("block_entity");
    public static final LootParameter<ItemStack> field_216289_i = func_216280_a("tool");
    public static final LootParameter<Float> field_216290_j = func_216280_a("explosion_radius");
+
+   public static final LootParameter<Integer> LOOTING_MOD = new LootParameter<>(new ResourceLocation("bukkit:looting_mod")); // CraftBukkit
+
 
    private static <T> LootParameter<T> func_216280_a(String p_216280_0_) {
       return new LootParameter<>(new ResourceLocation(p_216280_0_));
