--- a/net/minecraft/loot/LootTableManager.java
+++ b/net/minecraft/loot/LootTableManager.java
@@ -24,6 +_,18 @@
       this.field_227507_d_ = p_i225887_1_;
    }
 
+   // LoliServer start
+   public ResourceLocation getKey(LootTable loot) {
+      ResourceLocation resourceLocation = ResourceLocation.func_208304_a("");
+      for (Map.Entry<ResourceLocation, LootTable> tableEntry : field_186527_c.entrySet()) {
+         if (loot.equals(tableEntry.getValue())) {
+            resourceLocation = tableEntry.getKey();
+         }
+      }
+      return resourceLocation;
+   }
+   // LoliServer end
+
    public LootTable func_186521_a(ResourceLocation p_186521_1_) {
       return this.field_186527_c.getOrDefault(p_186521_1_, LootTable.field_186464_a);
    }
@@ -36,8 +_,8 @@
       }
 
       p_212853_1_.forEach((p_237403_1_, p_237403_2_) -> {
-         try {
-            LootTable loottable = field_186526_b.fromJson(p_237403_2_, LootTable.class);
+         try (net.minecraft.resources.IResource res = p_212853_2_.func_199002_a(getPreparedPath(p_237403_1_));){
+            LootTable loottable = net.minecraftforge.common.ForgeHooks.loadLootTable(field_186526_b, p_237403_1_, p_237403_2_, res == null || !res.func_199026_d().equals("Default"), this);
             builder.put(p_237403_1_, loottable);
          } catch (Exception exception) {
             field_186525_a.error("Couldn't parse loot table {}", p_237403_1_, exception);
