--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentArrowDamage.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentArrowDamage.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentArrowDamage extends Enchantment
 {
-    public EnchantmentArrowDamage(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentArrowDamage(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BOW, slots);
         this.setName("arrowDamage");
