--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentSweepingEdge.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentSweepingEdge.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentSweepingEdge extends Enchantment
 {
-    public EnchantmentSweepingEdge(Enchantment.Rarity p_i47366_1_, EntityEquipmentSlot... p_i47366_2_)
+    public EnchantmentSweepingEdge(Rarity p_i47366_1_, EntityEquipmentSlot... p_i47366_2_)
     {
         super(p_i47366_1_, EnumEnchantmentType.WEAPON, p_i47366_2_);
     }
