--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentFishingSpeed.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentFishingSpeed.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentFishingSpeed extends Enchantment
 {
-    protected EnchantmentFishingSpeed(Enchantment.Rarity rarityIn, EnumEnchantmentType typeIn, EntityEquipmentSlot... slots)
+    protected EnchantmentFishingSpeed(Rarity rarityIn, EnumEnchantmentType typeIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, typeIn, slots);
         this.setName("fishingSpeed");
