--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentDigging.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentDigging.java
@@ -6,7 +6,7 @@
 
 public class EnchantmentDigging extends Enchantment
 {
-    protected EnchantmentDigging(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    protected EnchantmentDigging(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.DIGGER, slots);
         this.setName("digging");
