--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentBindingCurse.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentBindingCurse.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentBindingCurse extends Enchantment
 {
-    public EnchantmentBindingCurse(Enchantment.Rarity p_i47254_1_, EntityEquipmentSlot... p_i47254_2_)
+    public EnchantmentBindingCurse(Rarity p_i47254_1_, EntityEquipmentSlot... p_i47254_2_)
     {
         super(p_i47254_1_, EnumEnchantmentType.WEARABLE, p_i47254_2_);
         this.setName("binding_curse");
