--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentDamage.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentDamage.java
@@ -17,7 +17,7 @@
     private static final int[] LEVEL_COST_SPAN = new int[] {20, 20, 20};
     public final int damageType;
 
-    public EnchantmentDamage(Enchantment.Rarity rarityIn, int damageTypeIn, EntityEquipmentSlot... slots)
+    public EnchantmentDamage(Rarity rarityIn, int damageTypeIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.WEAPON, slots);
         this.damageType = damageTypeIn;
