--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentArrowKnockback.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentArrowKnockback.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentArrowKnockback extends Enchantment
 {
-    public EnchantmentArrowKnockback(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentArrowKnockback(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BOW, slots);
         this.setName("arrowKnockback");
