--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentFireAspect.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentFireAspect.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentFireAspect extends Enchantment
 {
-    protected EnchantmentFireAspect(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    protected EnchantmentFireAspect(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.WEAPON, slots);
         this.setName("fire");
