--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentProtection.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentProtection.java
@@ -8,14 +8,14 @@
 
 public class EnchantmentProtection extends Enchantment
 {
-    public final EnchantmentProtection.Type protectionType;
+    public final Type protectionType;
 
-    public EnchantmentProtection(Enchantment.Rarity rarityIn, EnchantmentProtection.Type protectionTypeIn, EntityEquipmentSlot... slots)
+    public EnchantmentProtection(Rarity rarityIn, Type protectionTypeIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.ARMOR, slots);
         this.protectionType = protectionTypeIn;
 
-        if (protectionTypeIn == EnchantmentProtection.Type.FALL)
+        if (protectionTypeIn == Type.FALL)
         {
             this.type = EnumEnchantmentType.ARMOR_FEET;
         }
@@ -42,25 +42,25 @@
         {
             return 0;
         }
-        else if (this.protectionType == EnchantmentProtection.Type.ALL)
+        else if (this.protectionType == Type.ALL)
         {
             return level;
         }
-        else if (this.protectionType == EnchantmentProtection.Type.FIRE && source.isFireDamage())
+        else if (this.protectionType == Type.FIRE && source.isFireDamage())
         {
             return level * 2;
         }
-        else if (this.protectionType == EnchantmentProtection.Type.FALL && source == DamageSource.FALL)
+        else if (this.protectionType == Type.FALL && source == DamageSource.FALL)
         {
             return level * 3;
         }
-        else if (this.protectionType == EnchantmentProtection.Type.EXPLOSION && source.isExplosion())
+        else if (this.protectionType == Type.EXPLOSION && source.isExplosion())
         {
             return level * 2;
         }
         else
         {
-            return this.protectionType == EnchantmentProtection.Type.PROJECTILE && source.isProjectile() ? level * 2 : 0;
+            return this.protectionType == Type.PROJECTILE && source.isProjectile() ? level * 2 : 0;
         }
     }
 
@@ -81,7 +81,7 @@
             }
             else
             {
-                return this.protectionType == EnchantmentProtection.Type.FALL || enchantmentprotection.protectionType == EnchantmentProtection.Type.FALL;
+                return this.protectionType == Type.FALL || enchantmentprotection.protectionType == Type.FALL;
             }
         }
         else
