--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentDurability.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentDurability.java
@@ -7,7 +7,7 @@
 
 public class EnchantmentDurability extends Enchantment
 {
-    protected EnchantmentDurability(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    protected EnchantmentDurability(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BREAKABLE, slots);
         this.setName("durability");
