--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentArrowInfinite.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentArrowInfinite.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentArrowInfinite extends Enchantment
 {
-    public EnchantmentArrowInfinite(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentArrowInfinite(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BOW, slots);
         this.setName("arrowInfinite");
