--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentArrowFire.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentArrowFire.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentArrowFire extends Enchantment
 {
-    public EnchantmentArrowFire(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentArrowFire(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BOW, slots);
         this.setName("arrowFire");
