--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentWaterWorker.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentWaterWorker.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentWaterWorker extends Enchantment
 {
-    public EnchantmentWaterWorker(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentWaterWorker(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.ARMOR_HEAD, slots);
         this.setName("waterWorker");
