--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentMending.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentMending.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentMending extends Enchantment
 {
-    public EnchantmentMending(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentMending(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.BREAKABLE, slots);
         this.setName("mending");
