--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentKnockback.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentKnockback.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentKnockback extends Enchantment
 {
-    protected EnchantmentKnockback(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    protected EnchantmentKnockback(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.WEAPON, slots);
         this.setName("knockback");
