--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentWaterWalker.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentWaterWalker.java
@@ -5,7 +5,7 @@
 
 public class EnchantmentWaterWalker extends Enchantment
 {
-    public EnchantmentWaterWalker(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentWaterWalker(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.ARMOR_FEET, slots);
         this.setName("waterWalker");
