--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentUntouching.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentUntouching.java
@@ -5,7 +5,7 @@
 
 public class EnchantmentUntouching extends Enchantment
 {
-    protected EnchantmentUntouching(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    protected EnchantmentUntouching(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.DIGGER, slots);
         this.setName("untouching");
