--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentLootBonus.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentLootBonus.java
@@ -5,7 +5,7 @@
 
 public class EnchantmentLootBonus extends Enchantment
 {
-    protected EnchantmentLootBonus(Enchantment.Rarity rarityIn, EnumEnchantmentType typeIn, EntityEquipmentSlot... slots)
+    protected EnchantmentLootBonus(Rarity rarityIn, EnumEnchantmentType typeIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, typeIn, slots);
 
