--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentVanishingCurse.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentVanishingCurse.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentVanishingCurse extends Enchantment
 {
-    public EnchantmentVanishingCurse(Enchantment.Rarity p_i47252_1_, EntityEquipmentSlot... p_i47252_2_)
+    public EnchantmentVanishingCurse(Rarity p_i47252_1_, EntityEquipmentSlot... p_i47252_2_)
     {
         super(p_i47252_1_, EnumEnchantmentType.ALL, p_i47252_2_);
         this.setName("vanishing_curse");
