--- ../src-base/minecraft/net/minecraft/enchantment/EnchantmentOxygen.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnchantmentOxygen.java
@@ -4,7 +4,7 @@
 
 public class EnchantmentOxygen extends Enchantment
 {
-    public EnchantmentOxygen(Enchantment.Rarity rarityIn, EntityEquipmentSlot... slots)
+    public EnchantmentOxygen(Rarity rarityIn, EntityEquipmentSlot... slots)
     {
         super(rarityIn, EnumEnchantmentType.ARMOR_HEAD, slots);
         this.setName("oxygen");
