--- ../src-base/minecraft/net/minecraft/enchantment/EnumEnchantmentType.java
+++ ../src-work/minecraft/net/minecraft/enchantment/EnumEnchantmentType.java
@@ -100,5 +100,13 @@
     {
     }
 
-    public abstract boolean canEnchantItem(Item itemIn);
+    private com.google.common.base.Predicate<Item> delegate = null;
+    private EnumEnchantmentType(com.google.common.base.Predicate<Item> delegate)
+    {
+        this.delegate = delegate;
+    }
+    public boolean canEnchantItem(Item itemIn)
+    {
+        return this.delegate == null ? false : this.delegate.apply(itemIn);
+    }
 }
