--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -33,8 +_,8 @@
    protected double f_26497_;
    protected int f_26498_;
    protected int f_26499_;
-   protected Vec3 f_26500_ = Vec3.f_82478_;
-   protected Vec3i f_26501_ = Vec3i.f_123288_;
+   public Vec3 f_26500_ = Vec3.f_82478_; // CatServer - protected -> public
+   public Vec3i f_26501_ = Vec3i.f_123288_; // CatServer - protected -> public
    protected long f_26502_;
    protected long f_26503_;
    protected double f_26504_;
@@ -225,10 +_,10 @@
       Vec3 vec3 = this.m_7475_();
       this.f_26505_ = this.f_26494_.m_20205_() > 0.75F ? this.f_26494_.m_20205_() / 2.0F : 0.75F - this.f_26494_.m_20205_() / 2.0F;
       Vec3i vec3i = this.f_26496_.m_77400_();
-      double d0 = Math.abs(this.f_26494_.m_20185_() - ((double)vec3i.m_123341_() + 0.5D));
+      double d0 = Math.abs(this.f_26494_.m_20185_() - ((double)vec3i.m_123341_() + (this.f_26494_.m_20205_() + 1) / 2D)); //Forge: Fix MC-94054
       double d1 = Math.abs(this.f_26494_.m_20186_() - (double)vec3i.m_123342_());
-      double d2 = Math.abs(this.f_26494_.m_20189_() - ((double)vec3i.m_123343_() + 0.5D));
-      boolean flag = d0 < (double)this.f_26505_ && d2 < (double)this.f_26505_ && d1 < 1.0D;
+      double d2 = Math.abs(this.f_26494_.m_20189_() - ((double)vec3i.m_123343_() + (this.f_26494_.m_20205_() + 1) / 2D)); //Forge: Fix MC-94054
+      boolean flag = d0 <= (double)this.f_26505_ && d2 <= (double)this.f_26505_ && d1 < 1.0D; //Forge: Fix MC-94054
       if (flag || this.f_26494_.m_21481_(this.f_26496_.m_77401_().f_77282_) && this.m_26559_(vec3)) {
          this.f_26496_.m_77374_();
       }
