--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -108,19 +_,33 @@
             Raid raid = this.m_37960_(p_37964_.m_183503_(), blockpos1);
             boolean flag = false;
             if (!raid.m_37770_()) {
-               if (!this.f_37951_.containsKey(raid.m_37781_())) {
-                  this.f_37951_.put(raid.m_37781_(), raid);
+               /* CraftBukkit - moved down
+               if (!this.raidMap.containsKey(raid.getId())) {
+                  this.raidMap.put(raid.getId(), raid);
                }
+               */
 
                flag = true;
-            } else if (raid.m_37773_() < raid.m_37772_()) {
+               // CraftBukkit start - fixed a bug with raid: players could add up Bad Omen level even when the raid had finished
+            } else if (raid.isInProgress() && raid.m_37773_() < raid.m_37772_()) {
                flag = true;
+               // CraftBukkit end
             } else {
                p_37964_.m_21195_(MobEffects.f_19594_);
                p_37964_.f_8906_.m_141995_(new ClientboundEntityEventPacket(p_37964_, (byte)43));
             }
 
             if (flag) {
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidTriggerEvent(raid, p_37964_)) {
+                  p_37964_.m_21195_(MobEffects.f_19594_);
+                  return null;
+               }
+
+               if (!this.f_37951_.containsKey(raid.m_37781_())) {
+                  this.f_37951_.put(raid.m_37781_(), raid);
+               }
+               // CraftBukkit end
                raid.m_37728_(p_37964_);
                p_37964_.f_8906_.m_141995_(new ClientboundEntityEventPacket(p_37964_, (byte)43));
                if (!raid.m_37757_()) {
