--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -149,7 +_,7 @@
                i = Mth.m_14045_(i, 0, 4);
                MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.f_19594_, 120000, i, false, false, true);
                if (!this.f_19853_.m_46469_().m_46207_(GameRules.f_46154_)) {
-                  player.m_7292_(mobeffectinstance);
+                  player.addEffect(mobeffectinstance, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                }
             }
          }
@@ -296,7 +_,7 @@
          this.f_37903_.m_21573_().m_26573_();
 
          for(Raider raider : this.f_37903_.f_19853_.m_45971_(Raider.class, this.f_37901_, this.f_37903_, this.f_37903_.m_142469_().m_82377_(8.0D, 8.0D, 8.0D))) {
-            raider.m_6710_(this.f_37903_.m_5448_());
+            raider.setTarget(this.f_37903_.m_5448_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
          }
 
       }
@@ -306,7 +_,7 @@
          LivingEntity livingentity = this.f_37903_.m_5448_();
          if (livingentity != null) {
             for(Raider raider : this.f_37903_.f_19853_.m_45971_(Raider.class, this.f_37901_, this.f_37903_, this.f_37903_.m_142469_().m_82377_(8.0D, 8.0D, 8.0D))) {
-               raider.m_6710_(livingentity);
+               raider.setTarget(livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                raider.m_21561_(true);
             }
 
