--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -165,6 +_,12 @@
       return this.f_37687_ == Raid.RaidStatus.LOSS;
    }
 
+   // CraftBukkit start
+   public boolean isInProgress() {
+      return this.f_37687_ == Raid.RaidStatus.ONGOING;
+   }
+   // CraftBukkit end
+
    public float m_150220_() {
       return this.f_37678_;
    }
@@ -249,6 +_,7 @@
             boolean flag = this.f_37680_;
             this.f_37680_ = this.f_37675_.m_46805_(this.f_37674_);
             if (this.f_37675_.m_46791_() == Difficulty.PEACEFUL) {
+               org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                this.m_37774_();
                return;
             }
@@ -267,14 +_,17 @@
 
             if (!this.f_37675_.m_8802_(this.f_37674_)) {
                if (this.f_37681_ > 0) {
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidFinishEvent(this, new java.util.ArrayList<>()); // CraftBukkit
                   this.f_37687_ = Raid.RaidStatus.LOSS;
                } else {
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.NOT_IN_VILLAGE); // CraftBukkit
                   this.m_37774_();
                }
             }
 
             ++this.f_37673_;
             if (this.f_37673_ >= 48000L) {
+               org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.TIMEOUT); // CraftBukkit
                this.m_37774_();
                return;
             }
@@ -345,6 +_,7 @@
                }
 
                if (k > 3) {
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.UNSPAWNABLE);  // CraftBukkit
                   this.m_37774_();
                   break;
                }
@@ -355,7 +_,7 @@
                   ++this.f_37683_;
                } else {
                   this.f_37687_ = Raid.RaidStatus.VICTORY;
-
+                  List<org.bukkit.entity.Player> winners = new java.util.ArrayList<>(); // CraftBukkit
                   for(UUID uuid : this.f_37672_) {
                      Entity entity = this.f_37675_.m_8791_(uuid);
                      if (entity instanceof LivingEntity && !entity.m_5833_()) {
@@ -365,9 +_,11 @@
                            ServerPlayer serverplayer = (ServerPlayer)livingentity;
                            serverplayer.m_36220_(Stats.f_12950_);
                            CriteriaTriggers.f_10557_.m_53645_(serverplayer);
+                           winners.add(serverplayer.getBukkitEntity()); // CraftBukkit
                         }
                      }
                   }
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidFinishEvent(this, winners); // CraftBukkit
                }
             }
 
@@ -375,6 +_,7 @@
          } else if (this.m_37706_()) {
             ++this.f_37688_;
             if (this.f_37688_ >= 600) {
+               org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.FINISHED); // CraftBukkit
                this.m_37774_();
                return;
             }
@@ -495,7 +_,10 @@
       this.f_37678_ = 0.0F;
       DifficultyInstance difficultyinstance = this.f_37675_.m_6436_(p_37756_);
       boolean flag1 = this.m_37702_();
-
+      // CraftBukkit start
+      Raider leader = null;
+      List<Raider> raiders = new java.util.ArrayList<>();
+      // CraftBukkit end
       for(Raid.RaiderType raid$raidertype : Raid.RaiderType.f_37813_) {
          int j = this.m_37730_(raid$raidertype, i, flag1) + this.m_37734_(raid$raidertype, this.f_37685_, i, difficultyinstance, flag1);
          int k = 0;
@@ -506,9 +_,11 @@
                raider.m_33075_(true);
                this.m_37710_(i, raider);
                flag = true;
+               leader = raider; // CraftBukkit
             }
 
             this.m_37713_(i, raider, p_37756_, false);
+            raiders.add(raider); // CraftBukkit
             if (raid$raidertype.f_37814_ == EntityType.f_20518_) {
                Raider raider1 = null;
                if (i == this.m_37724_(Difficulty.NORMAL)) {
@@ -526,6 +_,7 @@
                   this.m_37713_(i, raider1, p_37756_, false);
                   raider1.m_20035_(p_37756_, 0.0F, 0.0F);
                   raider1.m_20329_(raider);
+                  raiders.add(raider); // CraftBukkit
                }
             }
          }
@@ -535,6 +_,7 @@
       ++this.f_37681_;
       this.m_37776_();
       this.m_37705_();
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callRaidSpawnWaveEvent(this, leader, raiders); // CraftBukkit
    }
 
    public void m_37713_(int p_37714_, Raider p_37715_, @Nullable BlockPos p_37716_, boolean p_37717_) {
@@ -550,6 +_,7 @@
             p_37715_.m_7895_(p_37714_, false);
             p_37715_.m_6853_(true);
             this.f_37675_.m_47205_(p_37715_);
+            // this.level.addFreshEntityWithPassengers(entityraider, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.RAID); // CraftBukkit TODO
          }
       }
 
@@ -789,6 +_,12 @@
       this.f_37672_.add(p_37727_.m_142081_());
    }
 
+   // CraftBukkit start - a method to get all raiders
+   public java.util.Collection<Raider> getRaiders() {
+      return this.f_37671_.values().stream().flatMap(Set::stream).collect(java.util.stream.Collectors.toSet());
+   }
+   // CraftBukkit end
+
    static enum RaidStatus {
       ONGOING,
       VICTORY,
@@ -812,7 +_,7 @@
       }
    }
 
-   static enum RaiderType {
+   public static enum RaiderType implements net.minecraftforge.common.IExtensibleEnum {
       VINDICATOR(EntityType.f_20493_, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
       EVOKER(EntityType.f_20568_, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
       PILLAGER(EntityType.f_20513_, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
@@ -826,6 +_,20 @@
       private RaiderType(EntityType<? extends Raider> p_37821_, int[] p_37822_) {
          this.f_37814_ = p_37821_;
          this.f_37815_ = p_37822_;
+      }
+      
+      /**
+       * The waveCountsIn integer decides how many entities of the EntityType defined in typeIn will spawn in each wave.
+       * For example, one ravager will always spawn in wave 3.
+       */
+      public static RaiderType create(String name, EntityType<? extends Raider> typeIn, int[] waveCountsIn) {
+         throw new IllegalStateException("Enum not extended");
+      }
+      
+      @Override
+      @Deprecated
+      public void init() {
+         f_37813_ = values();
       }
    }
 }
