--- ../src-base/minecraft/net/minecraft/block/BlockSand.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSand.java
@@ -16,21 +16,21 @@
 
 public class BlockSand extends BlockFalling
 {
-    public static final PropertyEnum<BlockSand.EnumType> VARIANT = PropertyEnum.<BlockSand.EnumType>create("variant", BlockSand.EnumType.class);
+    public static final PropertyEnum<EnumType> VARIANT = PropertyEnum.<EnumType>create("variant", EnumType.class);
 
     public BlockSand()
     {
-        this.setDefaultState(this.blockState.getBaseState().withProperty(VARIANT, BlockSand.EnumType.SAND));
+        this.setDefaultState(this.blockState.getBaseState().withProperty(VARIANT, EnumType.SAND));
     }
 
     public int damageDropped(IBlockState state)
     {
-        return ((BlockSand.EnumType)state.getValue(VARIANT)).getMetadata();
+        return ((EnumType)state.getValue(VARIANT)).getMetadata();
     }
 
     public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items)
     {
-        for (BlockSand.EnumType blocksand$enumtype : BlockSand.EnumType.values())
+        for (EnumType blocksand$enumtype : EnumType.values())
         {
             items.add(new ItemStack(this, 1, blocksand$enumtype.getMetadata()));
         }
@@ -38,17 +38,17 @@
 
     public MapColor getMapColor(IBlockState state, IBlockAccess worldIn, BlockPos pos)
     {
-        return ((BlockSand.EnumType)state.getValue(VARIANT)).getMapColor();
+        return ((EnumType)state.getValue(VARIANT)).getMapColor();
     }
 
     public IBlockState getStateFromMeta(int meta)
     {
-        return this.getDefaultState().withProperty(VARIANT, BlockSand.EnumType.byMetadata(meta));
+        return this.getDefaultState().withProperty(VARIANT, EnumType.byMetadata(meta));
     }
 
     public int getMetaFromState(IBlockState state)
     {
-        return ((BlockSand.EnumType)state.getValue(VARIANT)).getMetadata();
+        return ((EnumType)state.getValue(VARIANT)).getMetadata();
     }
 
     protected BlockStateContainer createBlockState()
@@ -59,7 +59,7 @@
     @SideOnly(Side.CLIENT)
     public int getDustColor(IBlockState state)
     {
-        BlockSand.EnumType blocksand$enumtype = (BlockSand.EnumType)state.getValue(VARIANT);
+        EnumType blocksand$enumtype = (EnumType)state.getValue(VARIANT);
         return blocksand$enumtype.getDustColor();
     }
 
@@ -68,7 +68,7 @@
         SAND(0, "sand", "default", MapColor.SAND, -2370656),
         RED_SAND(1, "red_sand", "red", MapColor.ADOBE, -5679071);
 
-        private static final BlockSand.EnumType[] META_LOOKUP = new BlockSand.EnumType[values().length];
+        private static final EnumType[] META_LOOKUP = new EnumType[values().length];
         private final int meta;
         private final String name;
         private final MapColor mapColor;
@@ -105,7 +105,7 @@
             return this.mapColor;
         }
 
-        public static BlockSand.EnumType byMetadata(int meta)
+        public static EnumType byMetadata(int meta)
         {
             if (meta < 0 || meta >= META_LOOKUP.length)
             {
@@ -127,7 +127,7 @@
 
         static
         {
-            for (BlockSand.EnumType blocksand$enumtype : values())
+            for (EnumType blocksand$enumtype : values())
             {
                 META_LOOKUP[blocksand$enumtype.getMetadata()] = blocksand$enumtype;
             }
