--- ../src-base/minecraft/net/minecraft/block/BlockTorch.java
+++ ../src-work/minecraft/net/minecraft/block/BlockTorch.java
@@ -82,18 +82,8 @@
 
     private boolean canPlaceOn(World worldIn, BlockPos pos)
     {
-        Block block = worldIn.getBlockState(pos).getBlock();
-        boolean flag = block == Blocks.END_GATEWAY || block == Blocks.LIT_PUMPKIN;
-
-        if (worldIn.getBlockState(pos).isTopSolid())
-        {
-            return !flag;
-        }
-        else
-        {
-            boolean flag1 = block instanceof BlockFence || block == Blocks.GLASS || block == Blocks.COBBLESTONE_WALL || block == Blocks.STAINED_GLASS;
-            return flag1 && !flag;
-        }
+        IBlockState state = worldIn.getBlockState(pos);
+        return state.getBlock().canPlaceTorchOnTop(state, worldIn, pos);
     }
 
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
