--- ../src-base/minecraft/net/minecraft/block/BlockShulkerBox.java
+++ ../src-work/minecraft/net/minecraft/block/BlockShulkerBox.java
@@ -44,7 +44,7 @@
 public class BlockShulkerBox extends BlockContainer
 {
     public static final PropertyEnum<EnumFacing> FACING = PropertyDirection.create("facing");
-    private final EnumDyeColor color;
+    public final EnumDyeColor color;
 
     public BlockShulkerBox(EnumDyeColor colorIn)
     {
