--- ../src-base/minecraft/net/minecraft/block/BlockEnderChest.java
+++ ../src-work/minecraft/net/minecraft/block/BlockEnderChest.java
@@ -100,7 +100,7 @@
 
         if (inventoryenderchest != null && tileentity instanceof TileEntityEnderChest)
         {
-            if (worldIn.getBlockState(pos.up()).isNormalCube())
+            if (worldIn.getBlockState(pos.up()).doesSideBlockChestOpening(worldIn, pos.up(), EnumFacing.DOWN))
             {
                 return true;
             }
