--- ../src-base/minecraft/net/minecraft/block/BlockPurpurSlab.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPurpurSlab.java
@@ -17,7 +17,7 @@
 
 public abstract class BlockPurpurSlab extends BlockSlab
 {
-    public static final PropertyEnum<BlockPurpurSlab.Variant> VARIANT = PropertyEnum.<BlockPurpurSlab.Variant>create("variant", BlockPurpurSlab.Variant.class);
+    public static final PropertyEnum<Variant> VARIANT = PropertyEnum.<Variant>create("variant", Variant.class);
 
     public BlockPurpurSlab()
     {
@@ -26,10 +26,10 @@
 
         if (!this.isDouble())
         {
-            iblockstate = iblockstate.withProperty(HALF, BlockSlab.EnumBlockHalf.BOTTOM);
+            iblockstate = iblockstate.withProperty(HALF, EnumBlockHalf.BOTTOM);
         }
 
-        this.setDefaultState(iblockstate.withProperty(VARIANT, BlockPurpurSlab.Variant.DEFAULT));
+        this.setDefaultState(iblockstate.withProperty(VARIANT, Variant.DEFAULT));
         this.setCreativeTab(CreativeTabs.BUILDING_BLOCKS);
     }
 
@@ -45,11 +45,11 @@
 
     public IBlockState getStateFromMeta(int meta)
     {
-        IBlockState iblockstate = this.getDefaultState().withProperty(VARIANT, BlockPurpurSlab.Variant.DEFAULT);
+        IBlockState iblockstate = this.getDefaultState().withProperty(VARIANT, Variant.DEFAULT);
 
         if (!this.isDouble())
         {
-            iblockstate = iblockstate.withProperty(HALF, (meta & 8) == 0 ? BlockSlab.EnumBlockHalf.BOTTOM : BlockSlab.EnumBlockHalf.TOP);
+            iblockstate = iblockstate.withProperty(HALF, (meta & 8) == 0 ? EnumBlockHalf.BOTTOM : EnumBlockHalf.TOP);
         }
 
         return iblockstate;
@@ -59,7 +59,7 @@
     {
         int i = 0;
 
-        if (!this.isDouble() && state.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP)
+        if (!this.isDouble() && state.getValue(HALF) == EnumBlockHalf.TOP)
         {
             i |= 8;
         }
@@ -84,7 +84,7 @@
 
     public Comparable<?> getTypeForItem(ItemStack stack)
     {
-        return BlockPurpurSlab.Variant.DEFAULT;
+        return Variant.DEFAULT;
     }
 
     public static class Double extends BlockPurpurSlab
