--- ../src-base/minecraft/net/minecraft/block/BlockWoodSlab.java
+++ ../src-work/minecraft/net/minecraft/block/BlockWoodSlab.java
@@ -27,7 +27,7 @@
 
         if (!this.isDouble())
         {
-            iblockstate = iblockstate.withProperty(HALF, BlockSlab.EnumBlockHalf.BOTTOM);
+            iblockstate = iblockstate.withProperty(HALF, EnumBlockHalf.BOTTOM);
         }
 
         this.setDefaultState(iblockstate.withProperty(VARIANT, BlockPlanks.EnumType.OAK));
@@ -78,7 +78,7 @@
 
         if (!this.isDouble())
         {
-            iblockstate = iblockstate.withProperty(HALF, (meta & 8) == 0 ? BlockSlab.EnumBlockHalf.BOTTOM : BlockSlab.EnumBlockHalf.TOP);
+            iblockstate = iblockstate.withProperty(HALF, (meta & 8) == 0 ? EnumBlockHalf.BOTTOM : EnumBlockHalf.TOP);
         }
 
         return iblockstate;
@@ -89,7 +89,7 @@
         int i = 0;
         i = i | ((BlockPlanks.EnumType)state.getValue(VARIANT)).getMetadata();
 
-        if (!this.isDouble() && state.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP)
+        if (!this.isDouble() && state.getValue(HALF) == EnumBlockHalf.TOP)
         {
             i |= 8;
         }
