--- ../src-base/minecraft/net/minecraft/block/BlockChest.java
+++ ../src-work/minecraft/net/minecraft/block/BlockChest.java
@@ -431,7 +431,7 @@
         }
         else
         {
-            ILockableContainer ilockablecontainer = (TileEntityChest)tileentity;
+            ILockableContainer ilockablecontainer = (ILockableContainer)tileentity;
 
             if (!p_189418_3_ && this.isBlocked(p_189418_1_, p_189418_2_))
             {
@@ -457,11 +457,11 @@
                         {
                             if (enumfacing != EnumFacing.WEST && enumfacing != EnumFacing.NORTH)
                             {
-                                ilockablecontainer = new InventoryLargeChest("container.chestDouble", ilockablecontainer, (TileEntityChest)tileentity1);
+                                ilockablecontainer = (ILockableContainer) new InventoryLargeChest("container.chestDouble", ilockablecontainer, (ILockableContainer)tileentity1);
                             }
                             else
                             {
-                                ilockablecontainer = new InventoryLargeChest("container.chestDouble", (TileEntityChest)tileentity1, ilockablecontainer);
+                                ilockablecontainer = (ILockableContainer) new InventoryLargeChest("container.chestDouble", (ILockableContainer)tileentity1, ilockablecontainer);
                             }
                         }
                     }
