--- ../src-base/minecraft/net/minecraft/block/BlockOldLog.java
+++ ../src-work/minecraft/net/minecraft/block/BlockOldLog.java
@@ -26,14 +26,14 @@
 
     public BlockOldLog()
     {
-        this.setDefaultState(this.blockState.getBaseState().withProperty(VARIANT, BlockPlanks.EnumType.OAK).withProperty(LOG_AXIS, BlockLog.EnumAxis.Y));
+        this.setDefaultState(this.blockState.getBaseState().withProperty(VARIANT, BlockPlanks.EnumType.OAK).withProperty(LOG_AXIS, EnumAxis.Y));
     }
 
     public MapColor getMapColor(IBlockState state, IBlockAccess worldIn, BlockPos pos)
     {
         BlockPlanks.EnumType blockplanks$enumtype = (BlockPlanks.EnumType)state.getValue(VARIANT);
 
-        switch ((BlockLog.EnumAxis)state.getValue(LOG_AXIS))
+        switch ((EnumAxis)state.getValue(LOG_AXIS))
         {
             case X:
             case Z:
@@ -73,16 +73,16 @@
         switch (meta & 12)
         {
             case 0:
-                iblockstate = iblockstate.withProperty(LOG_AXIS, BlockLog.EnumAxis.Y);
+                iblockstate = iblockstate.withProperty(LOG_AXIS, EnumAxis.Y);
                 break;
             case 4:
-                iblockstate = iblockstate.withProperty(LOG_AXIS, BlockLog.EnumAxis.X);
+                iblockstate = iblockstate.withProperty(LOG_AXIS, EnumAxis.X);
                 break;
             case 8:
-                iblockstate = iblockstate.withProperty(LOG_AXIS, BlockLog.EnumAxis.Z);
+                iblockstate = iblockstate.withProperty(LOG_AXIS, EnumAxis.Z);
                 break;
             default:
-                iblockstate = iblockstate.withProperty(LOG_AXIS, BlockLog.EnumAxis.NONE);
+                iblockstate = iblockstate.withProperty(LOG_AXIS, EnumAxis.NONE);
         }
 
         return iblockstate;
@@ -94,7 +94,7 @@
         int i = 0;
         i = i | ((BlockPlanks.EnumType)state.getValue(VARIANT)).getMetadata();
 
-        switch ((BlockLog.EnumAxis)state.getValue(LOG_AXIS))
+        switch ((EnumAxis)state.getValue(LOG_AXIS))
         {
             case X:
                 i |= 4;
