--- ../src-base/minecraft/net/minecraft/block/BlockEnchantmentTable.java
+++ ../src-work/minecraft/net/minecraft/block/BlockEnchantmentTable.java
@@ -64,7 +64,7 @@
                     {
                         BlockPos blockpos = pos.add(i, k, j);
 
-                        if (worldIn.getBlockState(blockpos).getBlock() == Blocks.BOOKSHELF)
+                        if (net.minecraftforge.common.ForgeHooks.getEnchantPower(worldIn, blockpos) > 0)
                         {
                             if (!worldIn.isAirBlock(pos.add(i / 2, 0, j / 2)))
                             {
