--- ../src-base/minecraft/net/minecraft/block/BlockLilyPad.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLilyPad.java
@@ -34,7 +34,7 @@
     {
         super.onEntityCollidedWithBlock(worldIn, pos, state, entityIn);
 
-        if (entityIn instanceof EntityBoat)
+        if (entityIn instanceof EntityBoat && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entityIn, pos, Blocks.AIR, 0).isCancelled())
         {
             worldIn.destroyBlock(new BlockPos(pos), true);
         }
