--- ../src-base/minecraft/net/minecraft/block/BlockYellowFlower.java
+++ ../src-work/minecraft/net/minecraft/block/BlockYellowFlower.java
@@ -2,8 +2,8 @@
 
 public class BlockYellowFlower extends BlockFlower
 {
-    public BlockFlower.EnumFlowerColor getBlockType()
+    public EnumFlowerColor getBlockType()
     {
-        return BlockFlower.EnumFlowerColor.YELLOW;
+        return EnumFlowerColor.YELLOW;
     }
 }
