--- ../src-base/minecraft/net/minecraft/block/BlockRedstoneWire.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedstoneWire.java
@@ -30,7 +30,10 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
+import org.bukkit.event.block.BlockRedstoneEvent;
+
 public class BlockRedstoneWire extends Block
 {
     public static final PropertyEnum<BlockRedstoneWire.EnumAttachPosition> NORTH = PropertyEnum.<BlockRedstoneWire.EnumAttachPosition>create("north", BlockRedstoneWire.EnumAttachPosition.class);
@@ -217,6 +220,14 @@
         {
             j = k;
         }
+        // CraftBukkit start
+        if (i != j) {
+        	BlockRedstoneEvent event = new BlockRedstoneEvent(worldIn.getWorld().getBlockAt(pos1.getX(), pos1.getY(),pos1.getZ()), i, j);
+        	worldIn.getServer().getPluginManager().callEvent(event);
+        	
+        	j = event.getNewCurrent();
+        }
+        // CraftBukkit end
 
         if (i != j)
         {
@@ -317,7 +328,7 @@
         }
     }
 
-    private int getMaxCurrentStrength(World worldIn, BlockPos pos, int strength)
+    public int getMaxCurrentStrength(World worldIn, BlockPos pos, int strength)
     {
         if (worldIn.getBlockState(pos).getBlock() != this)
         {
