--- ../src-base/minecraft/net/minecraft/block/BlockWorkbench.java
+++ ../src-work/minecraft/net/minecraft/block/BlockWorkbench.java
@@ -33,7 +33,7 @@
         }
         else
         {
-            playerIn.displayGui(new BlockWorkbench.InterfaceCraftingTable(worldIn, pos));
+            playerIn.displayGui(new InterfaceCraftingTable(worldIn, pos));
             playerIn.addStat(StatList.CRAFTING_TABLE_INTERACTION);
             return true;
         }
