--- ../src-base/minecraft/net/minecraft/block/BlockRedSandstone.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedSandstone.java
@@ -12,23 +12,23 @@
 
 public class BlockRedSandstone extends Block
 {
-    public static final PropertyEnum<BlockRedSandstone.EnumType> TYPE = PropertyEnum.<BlockRedSandstone.EnumType>create("type", BlockRedSandstone.EnumType.class);
+    public static final PropertyEnum<EnumType> TYPE = PropertyEnum.<EnumType>create("type", EnumType.class);
 
     public BlockRedSandstone()
     {
         super(Material.ROCK, BlockSand.EnumType.RED_SAND.getMapColor());
-        this.setDefaultState(this.blockState.getBaseState().withProperty(TYPE, BlockRedSandstone.EnumType.DEFAULT));
+        this.setDefaultState(this.blockState.getBaseState().withProperty(TYPE, EnumType.DEFAULT));
         this.setCreativeTab(CreativeTabs.BUILDING_BLOCKS);
     }
 
     public int damageDropped(IBlockState state)
     {
-        return ((BlockRedSandstone.EnumType)state.getValue(TYPE)).getMetadata();
+        return ((EnumType)state.getValue(TYPE)).getMetadata();
     }
 
     public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items)
     {
-        for (BlockRedSandstone.EnumType blockredsandstone$enumtype : BlockRedSandstone.EnumType.values())
+        for (EnumType blockredsandstone$enumtype : EnumType.values())
         {
             items.add(new ItemStack(this, 1, blockredsandstone$enumtype.getMetadata()));
         }
@@ -36,12 +36,12 @@
 
     public IBlockState getStateFromMeta(int meta)
     {
-        return this.getDefaultState().withProperty(TYPE, BlockRedSandstone.EnumType.byMetadata(meta));
+        return this.getDefaultState().withProperty(TYPE, EnumType.byMetadata(meta));
     }
 
     public int getMetaFromState(IBlockState state)
     {
-        return ((BlockRedSandstone.EnumType)state.getValue(TYPE)).getMetadata();
+        return ((EnumType)state.getValue(TYPE)).getMetadata();
     }
 
     protected BlockStateContainer createBlockState()
@@ -55,7 +55,7 @@
         CHISELED(1, "chiseled_red_sandstone", "chiseled"),
         SMOOTH(2, "smooth_red_sandstone", "smooth");
 
-        private static final BlockRedSandstone.EnumType[] META_LOOKUP = new BlockRedSandstone.EnumType[values().length];
+        private static final EnumType[] META_LOOKUP = new EnumType[values().length];
         private final int meta;
         private final String name;
         private final String unlocalizedName;
@@ -77,7 +77,7 @@
             return this.name;
         }
 
-        public static BlockRedSandstone.EnumType byMetadata(int meta)
+        public static EnumType byMetadata(int meta)
         {
             if (meta < 0 || meta >= META_LOOKUP.length)
             {
@@ -99,7 +99,7 @@
 
         static
         {
-            for (BlockRedSandstone.EnumType blockredsandstone$enumtype : values())
+            for (EnumType blockredsandstone$enumtype : values())
             {
                 META_LOOKUP[blockredsandstone$enumtype.getMetadata()] = blockredsandstone$enumtype;
             }
