--- ../src-base/minecraft/net/minecraft/block/BlockRedFlower.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedFlower.java
@@ -2,8 +2,8 @@
 
 public class BlockRedFlower extends BlockFlower
 {
-    public BlockFlower.EnumFlowerColor getBlockType()
+    public EnumFlowerColor getBlockType()
     {
-        return BlockFlower.EnumFlowerColor.RED;
+        return EnumFlowerColor.RED;
     }
 }
