--- ../src-base/minecraft/net/minecraft/block/BlockVine.java
+++ ../src-work/minecraft/net/minecraft/block/BlockVine.java
@@ -182,7 +182,7 @@
     {
         if (!worldIn.isRemote)
         {
-            if (worldIn.rand.nextInt(4) == 0)
+            if (worldIn.rand.nextInt(Math.max(1, (int) (100.0F / worldIn.spigotConfig.vineModifier) * 4)) == 0) //Spigot
             {
                 int i = 4;
                 int j = 5;
