--- ../src-base/minecraft/net/minecraft/potion/PotionType.java
+++ ../src-work/minecraft/net/minecraft/potion/PotionType.java
@@ -8,10 +8,11 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.RegistryNamespacedDefaultedByKey;
 
-public class PotionType
+public class PotionType extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<PotionType>
 {
+    @Deprecated // unused
     private static final ResourceLocation EMPTY = new ResourceLocation("empty");
-    public static final RegistryNamespacedDefaultedByKey<ResourceLocation, PotionType> REGISTRY = new RegistryNamespacedDefaultedByKey<ResourceLocation, PotionType>(EMPTY);
+    public static final RegistryNamespacedDefaultedByKey<ResourceLocation, PotionType> REGISTRY = net.minecraftforge.registries.GameData.getWrapperDefaulted(PotionType.class);
     private static int nextPotionTypeId;
     private final String baseName;
     private final ImmutableList<PotionEffect> effects;
