--- a/net/minecraft/server/CustomServerBossInfo.java
+++ b/net/minecraft/server/CustomServerBossInfo.java
@@ -16,12 +_,25 @@
 import net.minecraft.util.text.event.HoverEvent;
 import net.minecraft.world.BossInfo;
 import net.minecraft.world.server.ServerBossInfo;
+import org.bukkit.boss.KeyedBossBar;
+import org.bukkit.craftbukkit.v1_16_R3.boss.CraftKeyedBossbar;
 
 public class CustomServerBossInfo extends ServerBossInfo {
    private final ResourceLocation field_201373_h;
    private final Set<UUID> field_201374_i = Sets.newHashSet();
    private int field_201375_j;
    private int field_201376_k = 100;
+
+   // CraftBukkit start
+   private KeyedBossBar bossBar;
+
+   public KeyedBossBar getBukkitEntity() {
+      if (bossBar == null) {
+         bossBar = new CraftKeyedBossbar(this);
+      }
+      return bossBar;
+   }
+   // CraftBukkit end
 
    public CustomServerBossInfo(ResourceLocation p_i48620_1_, ITextComponent p_i48620_2_) {
       super(p_i48620_2_, BossInfo.Color.WHITE, BossInfo.Overlay.PROGRESS);
