--- a/net/minecraft/server/DebugLoggedPrintStream.java
+++ b/net/minecraft/server/DebugLoggedPrintStream.java
@@ -1,12 +_,8 @@
 package net.minecraft.server;
 
-import com.mojang.logging.LogUtils;
 import java.io.OutputStream;
-import org.slf4j.Logger;
 
 public class DebugLoggedPrintStream extends LoggedPrintStream {
-   private static final Logger f_202580_ = LogUtils.getLogger();
-
    public DebugLoggedPrintStream(String p_135934_, OutputStream p_135935_) {
       super(p_135934_, p_135935_);
    }
@@ -14,6 +_,6 @@
    protected void m_6812_(String p_135937_) {
       StackTraceElement[] astacktraceelement = Thread.currentThread().getStackTrace();
       StackTraceElement stacktraceelement = astacktraceelement[Math.min(3, astacktraceelement.length)];
-      f_202580_.info("[{}]@.({}:{}): {}", this.f_135948_, stacktraceelement.getFileName(), stacktraceelement.getLineNumber(), p_135937_);
+      f_135947_.info("[{}]@.({}:{}): {}", this.f_135948_, stacktraceelement.getFileName(), stacktraceelement.getLineNumber(), p_135937_);
    }
 }
