--- a/net/minecraft/server/LoggedPrintStream.java
+++ b/net/minecraft/server/LoggedPrintStream.java
@@ -7,7 +_,7 @@
 import org.slf4j.Logger;
 
 public class LoggedPrintStream extends PrintStream {
-   private static final Logger f_135947_ = LogUtils.getLogger();
+   public static final Logger f_135947_ = LogUtils.getLogger();
    protected final String f_135948_;
 
    public LoggedPrintStream(String p_135951_, OutputStream p_135952_) {
