--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1,5 +_,7 @@
 package net.minecraft.server;
 
+import catserver.server.BukkitInjector;
+import catserver.server.utils.MinecraftServerHelper;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
@@ -138,6 +_,7 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.border.BorderChangeListener;
 import net.minecraft.world.level.border.WorldBorder;
+import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.PatrolSpawner;
@@ -156,10 +_,41 @@
 import net.minecraft.world.level.storage.loot.LootDataManager;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import net.minecraftforge.common.MinecraftForge;
+import org.bukkit.event.world.WorldInitEvent;
+import org.bukkit.event.world.WorldLoadEvent;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.Lifecycle;
+import java.util.Random;
+import jline.console.ConsoleReader;
+import joptsimple.OptionSet;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.server.dedicated.DedicatedServer;
+import net.minecraft.server.dedicated.DedicatedServerProperties;
+import net.minecraft.util.datafix.DataFixers;
+import net.minecraft.world.level.levelgen.WorldDimensions;
+import net.minecraft.world.level.levelgen.presets.WorldPresets;
+import net.minecraft.world.level.storage.PrimaryLevelData;
+import net.minecraft.world.level.validation.ContentValidationException;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R1.CraftServer;
+import org.bukkit.craftbukkit.Main;
+import org.bukkit.craftbukkit.v1_20_R1.scoreboard.CraftScoreboardManager;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftChatMessage;
+import org.bukkit.craftbukkit.v1_20_R1.util.LazyPlayerSet;
+import org.bukkit.event.player.AsyncPlayerChatPreviewEvent;
+import org.bukkit.event.server.ServerLoadEvent;
+// CraftBukkit end
+
 public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements CommandSource, AutoCloseable {
-   private static final Logger f_129750_ = LogUtils.getLogger();
+   public static final Logger f_129750_ = LogUtils.getLogger(); // CatServer - private -> public
    public static final String f_195495_ = "vanilla";
    private static final float f_177884_ = 0.8F;
    private static final int f_177885_ = 100;
@@ -176,8 +_,8 @@
    public static final LevelSettings f_129743_ = new LevelSettings("Demo World", GameType.SURVIVAL, false, Difficulty.NORMAL, false, new GameRules(), WorldDataConfiguration.f_244649_);
    private static final long f_177871_ = 50L;
    public static final GameProfile f_195496_ = new GameProfile(Util.f_137441_, "Anonymous Player");
-   protected final LevelStorageSource.LevelStorageAccess f_129744_;
-   protected final PlayerDataStorage f_129745_;
+   public final LevelStorageSource.LevelStorageAccess f_129744_; // CatServer - protected -> public
+   public final PlayerDataStorage f_129745_; // CatServer - protected -> public
    private final List<Runnable> f_129752_ = Lists.newArrayList();
    private MetricsRecorder f_177872_ = InactiveMetricsRecorder.f_146153_;
    private ProfilerFiller f_129754_ = this.f_177872_.m_142610_();
@@ -191,17 +_,17 @@
    private MinecraftServer.TimeProfiler f_177876_;
    private boolean f_177877_;
    private final ServerConnectionListener f_129755_;
-   private final ChunkProgressListenerFactory f_129756_;
+   public final ChunkProgressListenerFactory f_129756_; // CatServer - private -> public
    @Nullable
    private ServerStatus f_129757_;
    @Nullable
    private ServerStatus.Favicon f_271173_;
    private final RandomSource f_129758_ = RandomSource.m_216327_();
-   private final DataFixer f_129759_;
+   public final DataFixer f_129759_; // CatServer - private -> public
    private String f_129760_;
    private int f_129761_ = -1;
-   private final LayeredRegistryAccess<RegistryLayer> f_244176_;
-   private final Map<ResourceKey<Level>, ServerLevel> f_129762_ = Maps.newLinkedHashMap();
+   public final LayeredRegistryAccess<RegistryLayer> f_244176_; // CatServer - private -> public
+   public Map<ResourceKey<Level>, ServerLevel> f_129762_ = Maps.newLinkedHashMap(); // CatServer - private -> public // CatServer - 1.20.1 remove final
    private PlayerList f_129763_;
    private volatile boolean f_129764_ = true;
    private boolean f_129765_;
@@ -215,6 +_,11 @@
    private String f_129709_;
    private int f_129711_;
    public final long[] f_129748_ = new long[100];
+   // Paper start
+   public final TickTimes tickTimes5s = new TickTimes(100);
+   public final TickTimes tickTimes10s = new TickTimes(200);
+   public final TickTimes tickTimes60s = new TickTimes(1200);
+   // Paper end
    @Nullable
    private KeyPair f_129712_;
    @Nullable
@@ -224,7 +_,7 @@
    private long f_129718_;
    protected final Services f_236721_;
    private long f_129724_;
-   private final Thread f_129725_;
+   public final Thread f_129725_; // CatServer - private -> public
    protected long f_129726_ = Util.m_137550_();
    private long f_129727_;
    private boolean f_129728_;
@@ -237,17 +_,74 @@
    private final FrameTimer f_129735_ = new FrameTimer();
    private boolean f_129736_;
    private float f_129737_;
-   private final Executor f_129738_;
+   public final Executor f_129738_; // CatServer - private -> public
    @Nullable
    private String f_129739_;
-   private MinecraftServer.ReloadableResources f_129740_;
+   public MinecraftServer.ReloadableResources f_129740_; // CatServer - private -> public
    private final StructureTemplateManager f_236720_;
-   protected final WorldData f_129749_;
+   public final WorldData f_129749_; // CatServer - protected -> public
    private volatile boolean f_195494_;
+   // CraftBukkit start
+   public WorldLoader.DataLoadContext worldLoader; // CatServer - remove final
+   public org.bukkit.craftbukkit.v1_20_R1.CraftServer server;
+   public OptionSet options;
+   public org.bukkit.command.ConsoleCommandSender console;
+   // public ConsoleReader reader; // CatServer - remove
+   public static int currentTick = (int) (System.currentTimeMillis() / 50);
+   public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+   public int autosavePeriod;
+   public Commands vanillaCommandDispatcher;
+   private boolean forceTicks;
+   // CraftBukkit end
+   // Spigot start
+   public static final int TPS = 20;
+   public static final int TICK_TIME = 1000000000 / TPS;
+   private static final int SAMPLE_INTERVAL = 100;
+   public final double[] recentTps = new double[3];
+   // Spigot end
+
+   // Spigot Start
+   private static double calcTps(double avg, double exp, double tps) {
+      return (avg * exp) + (tps * (1 - exp));
+   }
+   // Spigot End
+
+   // CatServer start
+   public void prepareMinecraftServer(OptionSet options, WorldLoader.DataLoadContext worldLoader) {
+      this.options = options;
+      this.worldLoader = worldLoader;
+   }
+   // CatServer end
+
+   // Paper start
+   public static class TickTimes {
+      private final long[] times;
+
+      public TickTimes(int length) {
+         times = new long[length];
+      }
+
+      void add(int index, long time) {
+         times[index % times.length] = time;
+      }
+
+      public long[] getTimes() {
+         return times.clone();
+      }
+
+      public double getAverage() {
+         long total = 0L;
+         for (long value : times) {
+            total += value;
+         }
+         return ((double) total / (double) times.length) * 1.0E-6D;
+      }
+   }
+   // Paper end
 
    public static <S extends MinecraftServer> S m_129872_(Function<Thread, S> p_129873_) {
       AtomicReference<S> atomicreference = new AtomicReference<>();
-      Thread thread = new Thread(() -> {
+      Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, () -> {
          atomicreference.get().m_130011_();
       }, "Server thread");
       thread.setUncaughtExceptionHandler((p_177909_, p_177910_) -> {
@@ -265,9 +_,10 @@
 
    public MinecraftServer(Thread p_236723_, LevelStorageSource.LevelStorageAccess p_236724_, PackRepository p_236725_, WorldStem p_236726_, Proxy p_236727_, DataFixer p_236728_, Services p_236729_, ChunkProgressListenerFactory p_236730_) {
       super("Server");
+      this.prepareMinecraftServer(MinecraftServerHelper.getInstance().getOptionSet(), MinecraftServerHelper.getInstance().getWorldLoader()); // CatServer - fire
       this.f_244176_ = p_236726_.f_244542_();
       this.f_129749_ = p_236726_.f_206895_();
-      if (!this.f_244176_.m_247579_().m_175515_(Registries.f_256862_).m_142003_(LevelStem.f_63971_)) {
+      if (false && !this.f_244176_.m_247579_().m_175515_(Registries.f_256862_).m_142003_(LevelStem.f_63971_)) { // CraftBukkit - 1.20.1 initialised later
          throw new IllegalStateException("Missing Overworld dimension data");
       } else {
          this.f_129747_ = p_236727_;
@@ -289,7 +_,10 @@
          this.f_129725_ = p_236723_;
          this.f_129738_ = Util.m_183991_();
       }
+      this.vanillaCommandDispatcher = p_236726_.f_206893_().f_206847_; // CraftBukkit
+      Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.v1_20_R1.util.ServerShutdownThread(this));
    }
+   // CraftBukkit end
 
    private void m_129841_(DimensionDataStorage p_129842_) {
       p_129842_.m_164861_(this.m_129896_()::m_180013_, this.m_129896_()::m_180015_, "scoreboard");
@@ -297,7 +_,7 @@
 
    protected abstract boolean m_7038_() throws IOException;
 
-   protected void m_130006_() {
+   protected void m_130006_() { // CraftBukkit // CatServer - recover
       if (!JvmProfiler.f_185340_.m_183608_()) {
       }
 
@@ -325,6 +_,20 @@
    protected void m_7044_() {
    }
 
+   private ServerLevel worldInit(ServerLevel serverLevel) {
+      var scoreManager = ((CraftServer) Bukkit.getServer()).scoreboardManager;
+      if (scoreManager == null) {
+         scoreManager = new CraftScoreboardManager(MinecraftServer.getServer(), serverLevel.m_6188_());
+         ((CraftServer) Bukkit.getServer()).scoreboardManager = scoreManager;
+      }
+      var generator = serverLevel.getWorld().getGenerator();
+      if (generator != null) {
+         serverLevel.getWorld().getPopulators().addAll(generator.getDefaultPopulators(serverLevel.getWorld()));
+      }
+      Bukkit.getPluginManager().callEvent(new WorldInitEvent(serverLevel.getWorld()));
+      return serverLevel;
+   }
+
    protected void m_129815_(ChunkProgressListener p_129816_) {
       ServerLevelData serverleveldata = this.f_129749_.m_5996_();
       boolean flag = this.f_129749_.m_7513_();
@@ -334,12 +_,15 @@
       long j = BiomeManager.m_47877_(i);
       List<CustomSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(serverleveldata));
       LevelStem levelstem = registry.m_6246_(LevelStem.f_63971_);
+
       ServerLevel serverlevel = new ServerLevel(this, this.f_129738_, this.f_129744_, serverleveldata, Level.f_46428_, levelstem, p_129816_, flag, j, list, true, (RandomSequences)null);
-      this.f_129762_.put(Level.f_46428_, serverlevel);
+      BukkitInjector.registerEnvironments(this.f_244176_.m_247579_().m_175515_(Registries.f_256862_)); // CatServer
+      this.f_129762_.put(Level.f_46428_, this.worldInit(serverlevel)); // CatServer
       DimensionDataStorage dimensiondatastorage = serverlevel.m_8895_();
       this.m_129841_(dimensiondatastorage);
       this.f_129732_ = new CommandStorage(dimensiondatastorage);
       WorldBorder worldborder = serverlevel.m_6857_();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Load(f_129762_.get(Level.f_46428_)));
       if (!serverleveldata.m_6535_()) {
          try {
             m_177896_(serverlevel, serverleveldata, worldoptions.m_245100_(), flag);
@@ -368,18 +_,140 @@
 
       RandomSequences randomsequences = serverlevel.m_288231_();
 
-      for(Map.Entry<ResourceKey<LevelStem>, LevelStem> entry : registry.m_6579_()) {
+      for (Map.Entry<ResourceKey<LevelStem>, LevelStem> entry : registry.m_6579_()) {
          ResourceKey<LevelStem> resourcekey = entry.getKey();
          if (resourcekey != LevelStem.f_63971_) {
             ResourceKey<Level> resourcekey1 = ResourceKey.m_135785_(Registries.f_256858_, resourcekey.m_135782_());
             DerivedLevelData derivedleveldata = new DerivedLevelData(this.f_129749_, serverleveldata);
             ServerLevel serverlevel1 = new ServerLevel(this, this.f_129738_, this.f_129744_, derivedleveldata, resourcekey1, entry.getValue(), p_129816_, flag, j, ImmutableList.of(), false, randomsequences);
             worldborder.m_61929_(new BorderChangeListener.DelegateBorderChangeListener(serverlevel1.m_6857_()));
-            this.f_129762_.put(resourcekey1, serverlevel1);
+            this.f_129762_.put(resourcekey1, this.worldInit(serverlevel1)); // CatServer
+            // CatServer start - Fix Multiverse-Core plugin
+            if (!"minecraft".equals(resourcekey.m_135782_().m_135827_()) ) {
+               try {
+                  File bukkitWorldDir = new File(m_129783_().convertable.getWorldDir().toFile(), resourcekey.m_135782_().m_135815_());
+                  if (!bukkitWorldDir.exists()) {
+                     bukkitWorldDir.mkdir(); // Create empty dir to bypass Multiverse-Core check
+                  }
+               } catch (Exception e) {
+                  System.out.println(e.toString());
+               }
+            }
+            // CatServer end
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Load(f_129762_.get(resourcekey)));
          }
       }
 
       worldborder.m_61931_(serverleveldata.m_5813_());
+      // CraftBukkit start
+      this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+      this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
+      this.f_129755_.acceptConnections();
+      // CraftBukkit end
+   }
+
+   // CatServer start - bukkit method
+   public void initWorld(ServerLevel serverWorld, ServerLevelData worldInfo, WorldData saveData, WorldOptions worldoptions) {
+      boolean flag = saveData.m_7513_();
+      var generator = serverWorld.getWorld().getGenerator();
+      if (generator != null) {
+         serverWorld.getWorld().getPopulators().addAll(generator.getDefaultPopulators(serverWorld.getWorld()));
+      }
+      WorldBorder worldborder = serverWorld.m_6857_();
+      worldborder.m_61931_(worldInfo.m_5813_());
+      if (!worldInfo.m_6535_()) {
+         try {
+            m_177896_(serverWorld, worldInfo, worldoptions.m_245100_(), flag);
+            worldInfo.m_5555_(true);
+            if (flag) {
+               this.m_129847_(this.f_129749_);
+            }
+         } catch (Throwable throwable) {
+            CrashReport crashreport = CrashReport.m_127521_(throwable, "Exception initializing level");
+            try {
+               serverWorld.m_6026_(crashreport);
+            } catch (Throwable throwable2) {
+               // empty catch block
+            }
+            throw new ReportedException(crashreport);
+         }
+         worldInfo.m_5555_(true);
+      }
+   }
+
+   public void prepareLevels(ChunkProgressListener listener, ServerLevel serverWorld) {
+      this.markWorldsDirty();
+      // CatServer start - Fix Multiverse-Core plugin
+      if (!serverWorld.isBukkitWorld) {
+         try {
+            File bukkitWorldDir = new File(m_129783_().convertable.getWorldDir().toFile(), f_129744_.m_78277_());
+            if (!bukkitWorldDir.exists()) {
+               bukkitWorldDir.mkdir(); // Create empty dir to bypass Multiverse-Core check
+            }
+         } catch (Exception e) {
+            System.out.println(e.toString());
+         }
+      }
+      // CatServer end
+      Bukkit.getPluginManager().callEvent(new WorldInitEvent(serverWorld.getWorld()));
+      Bukkit.getPluginManager().callEvent(new WorldLoadEvent(serverWorld.getWorld()));
+      MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Load(serverWorld));
+      if (!serverWorld.getWorld().getKeepSpawnInMemory()) {
+         return;
+      }
+      this.forceTicks = true;
+      f_129750_.info("Preparing start region for dimension {}", serverWorld.m_46472_().m_135782_());
+      BlockPos blockpos = serverWorld.m_220360_();
+      listener.m_7647_(new ChunkPos(blockpos));
+      ServerChunkCache serverchunkprovider = serverWorld.m_7726_();
+      this.f_129726_ = Util.m_137550_();
+      serverchunkprovider.m_8387_(TicketType.f_9442_, new ChunkPos(blockpos), 11, Unit.INSTANCE);
+
+      while (serverchunkprovider.m_8427_() < 441) {
+         // this.nextTickTime = SystemUtils.getMillis() + 10L; // CraftBukkit
+         this.executeModerately();
+      }
+
+      // this.nextTickTime = SystemUtils.getMillis() + 10L; // CraftBukkit
+      this.executeModerately();
+      // Iterator iterator = this.levels.values().iterator(); // CraftBukkit
+
+      ForcedChunksSavedData forcedchunkssavedata = serverWorld.m_8895_().m_164858_(ForcedChunksSavedData::m_151483_, "chunks");
+      if (forcedchunkssavedata != null) {
+         LongIterator longiterator = forcedchunkssavedata.m_46116_().iterator();
+
+         while (longiterator.hasNext()) {
+            long i = longiterator.nextLong();
+            ChunkPos chunkpos = new ChunkPos(i);
+            serverWorld.m_7726_().m_6692_(chunkpos, true);
+         }
+         net.minecraftforge.common.world.ForgeChunkManager.reinstatePersistentChunks(serverWorld, forcedchunkssavedata);
+      }
+      // CraftBukkit start
+      // this.nextTickTime = SystemUtils.getMillis() + 10L;
+      this.executeModerately();
+      // CraftBukkit end
+      listener.m_7646_();
+      // CraftBukkit start
+      // this.updateMobSpawningFlags();
+      serverWorld.m_46703_(this.m_7004_(), this.m_6998_());
+      this.forceTicks = false;
+      // CraftBukkit end
+   }
+   // CatServer end
+
+   // CraftBukkit start
+   private void executeModerately() {
+      this.m_18699_();
+      this.drainQueuedTasks();
+      java.util.concurrent.locks.LockSupport.parkNanos("executing tasks", 1000L);
+   }
+   // CraftBukkit end
+
+   public void drainQueuedTasks() {
+      while (!processQueue.isEmpty()) {
+         processQueue.remove().run();
+      }
    }
 
    private static void m_177896_(ServerLevel p_177897_, ServerLevelData p_177898_, boolean p_177899_, boolean p_177900_) {
@@ -387,7 +_,22 @@
          p_177898_.m_7250_(BlockPos.f_121853_.m_6630_(80), 0.0F);
       } else {
          ServerChunkCache serverchunkcache = p_177897_.m_7726_();
+         if (net.minecraftforge.event.ForgeEventFactory.onCreateWorldSpawn(p_177897_, p_177898_)) return;
          ChunkPos chunkpos = new ChunkPos(serverchunkcache.m_214994_().m_224579_().m_183230_());
+         // CraftBukkit start
+         if (p_177897_.generator != null) {
+            Random rand = new Random(p_177897_.m_7328_());
+            org.bukkit.Location spawn = p_177897_.generator.getFixedSpawnLocation(p_177897_.getWorld(), rand);
+            if (spawn != null) {
+               if (spawn.getWorld() != p_177897_.getWorld()) {
+                  throw new IllegalStateException("Cannot set spawn point for " + p_177898_.m_5462_() + " to be in another world (" + spawn.getWorld().getName() + ")");
+               } else {
+                  p_177898_.m_7250_(new BlockPos(spawn.getBlockX(), spawn.getBlockY(), spawn.getBlockZ()), spawn.getYaw());
+                  return;
+               }
+            }
+         }
+         // CraftBukkit end
          int i = serverchunkcache.m_8481_().m_142051_(p_177897_);
          if (i < p_177897_.m_141937_()) {
             BlockPos blockpos = chunkpos.m_45615_();
@@ -444,38 +_,49 @@
 
    private void m_129940_(ChunkProgressListener p_129941_) {
       ServerLevel serverlevel = this.m_129783_();
-      f_129750_.info("Preparing start region for dimension {}", (Object)serverlevel.m_46472_().m_135782_());
+      this.forceTicks = true;
+      f_129750_.info("Preparing start region for dimension {}", (Object) serverlevel.m_46472_().m_135782_());
       BlockPos blockpos = serverlevel.m_220360_();
       p_129941_.m_7647_(new ChunkPos(blockpos));
       ServerChunkCache serverchunkcache = serverlevel.m_7726_();
       this.f_129726_ = Util.m_137550_();
       serverchunkcache.m_8387_(TicketType.f_9442_, new ChunkPos(blockpos), 11, Unit.INSTANCE);
 
-      while(serverchunkcache.m_8427_() != 441) {
-         this.f_129726_ = Util.m_137550_() + 10L;
-         this.m_130012_();
+      while (serverchunkcache.m_8427_() != 441) {
+         this.executeModerately();
       }
 
-      this.f_129726_ = Util.m_137550_() + 10L;
-      this.m_130012_();
-
-      for(ServerLevel serverlevel1 : this.f_129762_.values()) {
-         ForcedChunksSavedData forcedchunkssaveddata = serverlevel1.m_8895_().m_164858_(ForcedChunksSavedData::m_151483_, "chunks");
-         if (forcedchunkssaveddata != null) {
-            LongIterator longiterator = forcedchunkssaveddata.m_46116_().iterator();
-
-            while(longiterator.hasNext()) {
-               long i = longiterator.nextLong();
-               ChunkPos chunkpos = new ChunkPos(i);
-               serverlevel1.m_7726_().m_6692_(chunkpos, true);
+      // this.nextTickTime = SystemUtils.getMillis() + 10L;
+      // this.waitUntilNextTick();
+      this.executeModerately();
+
+      for (ServerLevel serverlevel1 : this.f_129762_.values()) {
+         if (serverlevel1.getWorld().getKeepSpawnInMemory()) {
+            ForcedChunksSavedData forcedchunkssaveddata = serverlevel1.m_8895_().m_164858_(ForcedChunksSavedData::m_151483_, "chunks");
+            if (forcedchunkssaveddata != null) {
+               LongIterator longiterator = forcedchunkssaveddata.m_46116_().iterator();
+
+               while (longiterator.hasNext()) {
+                  long i = longiterator.nextLong();
+                  ChunkPos chunkpos = new ChunkPos(i);
+                  serverlevel1.m_7726_().m_6692_(chunkpos, true);
+               }
+               net.minecraftforge.common.world.ForgeChunkManager.reinstatePersistentChunks(serverlevel1, forcedchunkssaveddata);
             }
          }
+         Bukkit.getPluginManager().callEvent(new WorldLoadEvent(serverlevel1.getWorld()));
       }
 
-      this.f_129726_ = Util.m_137550_() + 10L;
-      this.m_130012_();
+      // CraftBukkit start
+      // this.nextTickTime = SystemUtils.getMillis() + 10L;
+      // this.waitUntilNextTick();
+      this.executeModerately();
+      // CraftBukkit end
       p_129941_.m_7646_();
-      this.m_129962_();
+      // CraftBukkit start
+      this.m_129962_(); // CatServer - recover
+      this.forceTicks = false;
+      // CraftBukkit end
    }
 
    public GameType m_130008_() {
@@ -504,11 +_,26 @@
          flag = true;
       }
 
+      // CraftBukkit start - moved to WorldServer.save // CatServer - recover
       ServerLevel serverlevel2 = this.m_129783_();
-      ServerLevelData serverleveldata = this.f_129749_.m_5996_();
-      serverleveldata.m_7831_(serverlevel2.m_6857_().m_61970_());
+      // CatServer start - 1.20.1 add if (serverlevel2 != null) to prevent NPE when creating crashreport if server crashes when worlds not created
+      if (serverlevel2 != null) {
+         ServerLevelData serverleveldata = this.f_129749_.m_5996_();
+         serverleveldata.m_7831_(serverlevel2.m_6857_().m_61970_());
+      }
+      // CatServer end
       this.f_129749_.m_5917_(this.m_129901_().m_136307_());
       this.f_129744_.m_78290_(this.m_206579_(), this.f_129749_, this.m_6846_().m_6960_());
+      // CraftBukkit end
+      // CatServer start - Save level.dat to all plugin world
+      for (ServerLevel world : this.f_129762_.values()) {
+         if (world.convertable != this.f_129744_) {
+            world.worldDataServer.m_7831_(world.m_6857_().m_61970_());
+            world.worldDataServer.m_5917_(this.m_129901_().m_136307_());
+            world.convertable.m_78287_(this.m_206579_(), world.worldDataServer);
+         }
+      }
+      // CatServer end
       if (p_129887_) {
          for(ServerLevel serverlevel1 : this.m_129785_()) {
             f_129750_.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", (Object)serverlevel1.m_7726_().f_8325_.m_182285_());
@@ -537,12 +_,34 @@
       this.m_7041_();
    }
 
+   // CraftBukkit start
+   private boolean hasStopped = false;
+   private final Object stopLock = new Object();
+
+   public final boolean hasStopped() {
+      synchronized (stopLock) {
+         return hasStopped;
+      }
+   }
+   // CraftBukkit end
+
    public void m_7041_() {
+      // CraftBukkit start - prevent double stopping on multiple threads
+      synchronized (stopLock) {
+         if (hasStopped) return;
+         hasStopped = true;
+      }
+      // CraftBukkit end
       if (this.f_177872_.m_142763_()) {
          this.m_236737_();
       }
 
       f_129750_.info("Stopping server");
+      // CraftBukkit start
+      if (this.server != null) {
+         this.server.disablePlugins();
+      }
+      // CraftBukkit end
       if (this.m_129919_() != null) {
          this.m_129919_().m_9718_();
       }
@@ -552,6 +_,10 @@
          f_129750_.info("Saving players");
          this.f_129763_.m_11302_();
          this.f_129763_.m_11313_();
+         try {
+            Thread.sleep(100);
+         } catch (InterruptedException ex) {
+         } // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
       }
 
       f_129750_.info("Saving worlds");
@@ -582,6 +_,7 @@
       for(ServerLevel serverlevel2 : this.m_129785_()) {
          if (serverlevel2 != null) {
             try {
+               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Unload(serverlevel2));
                serverlevel2.close();
             } catch (IOException ioexception1) {
                f_129750_.error("Exception closing the level", (Throwable)ioexception1);
@@ -597,7 +_,12 @@
       } catch (IOException ioexception) {
          f_129750_.error("Failed to unlock level {}", this.f_129744_.m_78277_(), ioexception);
       }
-
+      // Spigot start
+      if (org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) {
+         f_129750_.info("Saving usercache.json");
+         this.m_129927_().m_11006_();
+      }
+      // Spigot end
    }
 
    public String m_130009_() {
@@ -630,24 +_,42 @@
             throw new IllegalStateException("Failed to initialize server");
          }
 
+         net.minecraftforge.server.ServerLifecycleHooks.handleServerStarted(this);
          this.f_129726_ = Util.m_137550_();
          this.f_271173_ = this.m_272273_().orElse((ServerStatus.Favicon)null);
          this.f_129757_ = this.m_271988_();
+         resetStatusCache(f_129757_);
 
+         // Spigot start
+         Arrays.fill(recentTps, 20);
+         long curTime, tickSection = Util.m_137550_(), tickCount = 1;
          while(this.f_129764_) {
-            long i = Util.m_137550_() - this.f_129726_;
-            if (i > 2000L && this.f_129726_ - this.f_129718_ >= 15000L) {
+            long i = (curTime = Util.m_137550_()) - this.f_129726_;
+            if (i > 5000L && this.f_129726_ - this.f_129718_ >= 30000L) { // CraftBukkit
                long j = i / 50L;
-               f_129750_.warn("Can't keep up! Is the server overloaded? Running {}ms or {} ticks behind", i, j);
+               if (server.getWarnOnOverload()) { // CraftBukkit
+                  // LOGGER.warn("Can't keep up! Is the server overloaded? Running {}ms or {} ticks behind", i, j); // CatServer - disable overloaded warn
+               }
                this.f_129726_ += j * 50L;
                this.f_129718_ = this.f_129726_;
             }
 
+            if (tickCount++ % SAMPLE_INTERVAL == 0) {
+               double currentTps = 1E3 / (curTime - tickSection) * SAMPLE_INTERVAL;
+               recentTps[0] = calcTps(recentTps[0], 0.92, currentTps); // 1/exp(5sec/1min)
+               recentTps[1] = calcTps(recentTps[1], 0.9835, currentTps); // 1/exp(5sec/5min)
+               recentTps[2] = calcTps(recentTps[2], 0.9945, currentTps); // 1/exp(5sec/15min)
+               tickSection = curTime;
+            }
+            // Spigot end
+
+
             if (this.f_177877_) {
                this.f_177877_ = false;
                this.f_177876_ = new MinecraftServer.TimeProfiler(Util.m_137569_(), this.f_129766_);
             }
 
+            MinecraftServer.currentTick = (int) (System.currentTimeMillis() / 50); // CraftBukkit
             this.f_129726_ += 50L;
             this.m_177945_();
             this.f_129754_.m_6180_("tick");
@@ -661,6 +_,8 @@
             this.f_129717_ = true;
             JvmProfiler.f_185340_.m_183597_(this.f_129737_);
          }
+         net.minecraftforge.server.ServerLifecycleHooks.handleServerStopping(this);
+         net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
       } catch (Throwable throwable1) {
          f_129750_.error("Encountered an unexpected exception", throwable1);
          CrashReport crashreport = m_206568_(throwable1);
@@ -672,6 +_,7 @@
             f_129750_.error("We were unable to save this crash report to disk.");
          }
 
+         net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
          this.m_7268_(crashreport);
       } finally {
          try {
@@ -684,7 +_,10 @@
                this.f_236721_.f_214336_().m_196559_();
             }
 
+            org.spigotmc.WatchdogThread.doStop(); // Spigot
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStopped(this);
             this.m_6988_();
+            Runtime.getRuntime().halt(0); // CatServer
          }
 
       }
@@ -714,7 +_,8 @@
    }
 
    private boolean m_129960_() {
-      return this.m_18767_() || Util.m_137550_() < (this.f_129728_ ? this.f_129727_ : this.f_129726_);
+      // CraftBukkit start
+      return this.forceTicks || this.m_18767_() || Util.m_137550_() < (this.f_129728_ ? this.f_129727_ : this.f_129726_);
    }
 
    protected void m_130012_() {
@@ -724,7 +_,7 @@
       });
    }
 
-   protected TickTask m_6681_(Runnable p_129852_) {
+   public TickTask m_6681_(Runnable p_129852_) { // CatServer - protected -> public
       return new TickTask(this.f_129766_, p_129852_);
    }
 
@@ -797,33 +_,59 @@
    }
 
    public void m_5705_(BooleanSupplier p_129871_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.serverTickTimer.startTiming(); // Spigot
       long i = Util.m_137569_();
+      net.minecraftforge.event.ForgeEventFactory.onPreServerTick(p_129871_, this);
       ++this.f_129766_;
       this.m_5703_(p_129871_);
       if (i - this.f_129724_ >= 5000000000L) {
          this.f_129724_ = i;
          this.f_129757_ = this.m_271988_();
+         resetStatusCache(f_129757_);
       }
 
-      if (this.f_129766_ % 6000 == 0) {
+      if (autosavePeriod > 0 && this.f_129766_ % autosavePeriod == 0) { // CraftBukkit
+         org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.worldSaveTimer.startTiming(); // Spigot
          f_129750_.debug("Autosave started");
          this.f_129754_.m_6180_("save");
          this.m_195514_(true, false, false);
          this.f_129754_.m_7238_();
          f_129750_.debug("Autosave finished");
+         org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.worldSaveTimer.stopTiming(); // Spigot
       }
 
       this.f_129754_.m_6180_("tallying");
       long j = this.f_129748_[this.f_129766_ % 100] = Util.m_137569_() - i;
       this.f_129737_ = this.f_129737_ * 0.8F + (float)j / 1000000.0F * 0.19999999F;
       long k = Util.m_137569_();
+
+      // Paper start
+      tickTimes5s.add(this.f_129766_, j);
+      tickTimes10s.add(this.f_129766_, j);
+      tickTimes60s.add(this.f_129766_, j);
+      // Paper end
+
       this.f_129735_.m_13755_(k - i);
       this.f_129754_.m_7238_();
+      org.spigotmc.WatchdogThread.tick(); // Spigot
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.serverTickTimer.stopTiming(); // Spigot
+      org.spigotmc.CustomTimingsHandler.tick(); // Spigot
+      net.minecraftforge.event.ForgeEventFactory.onPostServerTick(p_129871_, this);
+   }
+
+   private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+   private String cachedServerStatus; // FORGE: cache the server status json in case a client spams requests
+   private void resetStatusCache(ServerStatus status) {
+      this.cachedServerStatus = GSON.toJson(ServerStatus.f_271163_.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+              .result().orElseThrow());
+   }
+   public String getStatusJson() {
+      return cachedServerStatus;
    }
 
    private ServerStatus m_271988_() {
       ServerStatus.Players serverstatus$players = this.m_271961_();
-      return new ServerStatus(Component.m_130674_(this.f_129709_), Optional.of(serverstatus$players), Optional.of(ServerStatus.Version.m_272202_()), Optional.ofNullable(this.f_271173_), this.m_214005_());
+      return new ServerStatus(Component.m_130674_(this.f_129709_), Optional.of(serverstatus$players), Optional.of(ServerStatus.Version.m_272202_()), Optional.ofNullable(this.f_271173_), this.m_214005_(), Optional.of(new net.minecraftforge.network.ServerStatusPing()));
    }
 
    private ServerStatus.Players m_271961_() {
@@ -847,47 +_,83 @@
    }
 
    public void m_5703_(BooleanSupplier p_129954_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.schedulerTimer.startTiming(); // Spigot
+      this.server.getScheduler().mainThreadHeartbeat(this.f_129766_); // CraftBukkit
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.schedulerTimer.stopTiming(); // Spigot
       this.f_129754_.m_6180_("commandFunctions");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.commandFunctionsTimer.startTiming(); // Spigot
       this.m_129890_().m_136128_();
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.commandFunctionsTimer.stopTiming(); // Spigot
       this.f_129754_.m_6182_("levels");
 
-      for(ServerLevel serverlevel : this.m_129785_()) {
+      // CraftBukkit start
+      // Run tasks that are waiting on processing
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.processQueueTimer.startTiming(); // Spigot
+      this.drainQueuedTasks(); // CatServer
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.processQueueTimer.stopTiming(); // Spigot
+
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.timeUpdateTimer.startTiming(); // Spigot
+      // Send time updates to everyone, it will get the right time from the world the player is in.
+      if (this.f_129766_ % 20 == 0) {
+         this.f_129754_.m_6180_("timeSync"); // CatServer
+         for (ServerPlayer entityplayer : this.m_6846_().f_11196_) {
+            entityplayer.f_8906_.m_9829_(new ClientboundSetTimePacket(entityplayer.m_9236_().m_46467_(), entityplayer.getPlayerTime(), entityplayer.m_9236_().m_46469_().m_46207_(GameRules.f_46140_))); // Add support for per player time
+         }
+         this.f_129754_.m_7238_(); // CatServer
+      }
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.timeUpdateTimer.stopTiming(); // Spigot
+
+      for(ServerLevel serverlevel : this.getWorldArray()) {
+         long tickStart = Util.m_137569_();
          this.f_129754_.m_6521_(() -> {
             return serverlevel + " " + serverlevel.m_46472_().m_135782_();
          });
-         if (this.f_129766_ % 20 == 0) {
-            this.f_129754_.m_6180_("timeSync");
-            this.m_276346_(serverlevel);
-            this.f_129754_.m_7238_();
+         /* Drop global time updates
+         if (this.tickCount % 20 == 0) {
+            this.profiler.push("timeSync");
+            this.synchronizeTime(serverlevel);
+            this.profiler.pop();
          }
+         // CraftBukkit end */
 
          this.f_129754_.m_6180_("tick");
+         net.minecraftforge.event.ForgeEventFactory.onPreLevelTick(serverlevel, p_129954_);
 
          try {
+            serverlevel.timings.doTick.startTiming(); // Spigot
             serverlevel.m_8793_(p_129954_);
+            serverlevel.timings.doTick.stopTiming(); // Spigot
          } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.m_127521_(throwable, "Exception ticking world");
             serverlevel.m_6026_(crashreport);
             throw new ReportedException(crashreport);
          }
+         net.minecraftforge.event.ForgeEventFactory.onPostLevelTick(serverlevel, p_129954_);
 
          this.f_129754_.m_7238_();
          this.f_129754_.m_7238_();
+         perWorldTickTimes.computeIfAbsent(serverlevel.m_46472_(), k -> new long[100])[this.f_129766_ % 100] = Util.m_137569_() - tickStart;
       }
 
       this.f_129754_.m_6182_("connection");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.connectionTimer.startTiming(); // Spigot
       this.m_129919_().m_9721_();
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.connectionTimer.stopTiming(); // Spigot
       this.f_129754_.m_6182_("players");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.playerListTimer.startTiming(); // Spigot
       this.f_129763_.m_11288_();
-      if (SharedConstants.f_136183_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.playerListTimer.stopTiming(); // Spigot
+      if (net.minecraftforge.gametest.ForgeGameTestHooks.isGametestEnabled()) {
          GameTestTicker.f_177648_.m_127790_();
       }
 
       this.f_129754_.m_6182_("server gui refresh");
 
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.tickablesTimer.startTiming(); // Spigot
       for(int i = 0; i < this.f_129752_.size(); ++i) {
          this.f_129752_.get(i).run();
       }
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.tickablesTimer.stopTiming(); // Spigot
 
       this.f_129754_.m_7238_();
    }
@@ -935,6 +_,22 @@
       return this.f_129762_.get(p_129881_);
    }
 
+   // CraftBukkit start - 1.20.1
+   public void addLevel(ServerLevel level) {
+      Map<ResourceKey<Level>, ServerLevel> oldLevels = this.f_129762_;
+      Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+      newLevels.put(level.m_46472_(), level);
+      this.f_129762_ = Collections.unmodifiableMap(newLevels);
+   }
+
+   public void removeLevel(ServerLevel level) {
+      Map<ResourceKey<Level>, ServerLevel> oldLevels = this.f_129762_;
+      Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+      newLevels.remove(level.m_46472_());
+      this.f_129762_ = Collections.unmodifiableMap(newLevels);
+   }
+   // CraftBukkit end
+
    public Set<ResourceKey<Level>> m_129784_() {
       return this.f_129762_.keySet();
    }
@@ -961,7 +_,7 @@
 
    @DontObfuscate
    public String getServerModName() {
-      return "vanilla";
+      return net.minecraftforge.internal.BrandingControl.getServerBranding(); // CraftBukkit - cb > vanilla! // CatServer - forge patched
    }
 
    public SystemReport m_177935_(SystemReport p_177936_) {
@@ -1303,6 +_,7 @@
       }).thenAcceptAsync((p_248090_) -> {
          this.f_129740_.close();
          this.f_129740_ = p_248090_;
+         this.server.syncCommands(); // SPIGOT-5884: Lost on reload
          this.f_129730_.m_10509_(p_129862_);
          WorldDataConfiguration worlddataconfiguration = new WorldDataConfiguration(m_129817_(this.f_129730_), this.f_129749_.m_247623_());
          this.f_129749_.m_245843_(worlddataconfiguration);
@@ -1311,6 +_,7 @@
          this.m_6846_().m_11315_();
          this.f_129734_.m_136120_(this.f_129740_.f_206585_.m_206860_());
          this.f_236720_.m_230370_(this.f_129740_.f_206584_);
+         this.m_6846_().m_11314_().forEach(this.m_6846_()::m_11289_); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
       }, this);
       if (this.m_18695_()) {
          this.m_18701_(completablefuture::isDone);
@@ -1320,10 +_,13 @@
    }
 
    public static WorldDataConfiguration m_246048_(PackRepository p_248681_, DataPackConfig p_248920_, boolean p_249869_, FeatureFlagSet p_251243_) {
+      net.minecraftforge.resource.ResourcePackLoader.loadResourcePacks(p_248681_, net.minecraftforge.server.ServerLifecycleHooks::buildPackFinder);
       p_248681_.m_10506_();
+      DataPackConfig.f_45842_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
+      p_248920_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
       if (p_249869_) {
-         p_248681_.m_10509_(Collections.singleton("vanilla"));
-         return WorldDataConfiguration.f_244649_;
+         p_248681_.m_10509_(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla());
+         return new WorldDataConfiguration(new DataPackConfig(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla(), ImmutableList.of()), FeatureFlags.f_244332_);
       } else {
          Set<String> set = Sets.newLinkedHashSet();
 
@@ -1481,6 +_,31 @@
 
    public abstract boolean m_7779_(GameProfile p_129840_);
 
+   private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+   @Nullable
+   public long[] getTickTime(ResourceKey<Level> dim) {
+      return perWorldTickTimes.get(dim);
+   }
+
+   @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+   public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+      return this.f_129762_;
+   }
+   private int worldArrayMarker = 0;
+   private int worldArrayLast = -1;
+   private ServerLevel[] worldArray;
+   @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+   public synchronized void markWorldsDirty() {
+      worldArrayMarker++;
+   }
+   private ServerLevel[] getWorldArray() {
+      if (worldArrayMarker == worldArrayLast && worldArray != null)
+         return worldArray;
+      worldArray = this.f_129762_.values().stream().toArray(x -> new ServerLevel[x]);
+      worldArrayLast = worldArrayMarker;
+      return worldArray;
+   }
+
    public void m_142116_(Path p_177911_) throws IOException {
    }
 
@@ -1583,6 +_,22 @@
       }
    }
 
+   // CraftBukkit start
+   @Override
+   public boolean m_18695_() {
+      return super.m_18695_() || this.m_129918_(); // CraftBukkit - MC-142590
+   }
+
+   public boolean isDebugging() {
+      return false;
+   }
+
+   @Deprecated
+   public static MinecraftServer getServer() {
+      return (Bukkit.getServer() instanceof CraftServer) ? ((CraftServer) Bukkit.getServer()).getServer() : null;
+   }
+   // CraftBukkit end
+
    private void m_177945_() {
       if (this.f_177875_) {
          this.f_177872_ = ActiveMetricsRecorder.m_146132_(new ServerMetricsSamplersProvider(Util.f_137440_, this.m_6982_()), Util.f_137440_, Util.m_183992_(), new MetricsPersister("server"), this.f_177873_, (p_212927_) -> {
@@ -1646,6 +_,10 @@
       return this.f_129749_;
    }
 
+   public MinecraftServer.ReloadableResources getServerResources() {
+       return f_129740_;
+   }
+
    public RegistryAccess.Frozen m_206579_() {
       return this.f_244176_.m_247579_();
    }
@@ -1707,9 +_,41 @@
 
    }
 
+   // CraftBukkit start - 1.20.1
+   public final java.util.concurrent.ExecutorService chatExecutor = java.util.concurrent.Executors.newCachedThreadPool(
+           new com.google.common.util.concurrent.ThreadFactoryBuilder()
+                   .setDaemon(true)
+                   .setNameFormat("Async Chat Thread - #%d")
+                   .build()
+   );
+
    public ChatDecorator m_236742_() {
-      return ChatDecorator.f_236947_;
-   }
+      return (entityplayer, ichatbasecomponent) -> {
+         // SPIGOT-7127: Console /say and similar
+         if (entityplayer == null) {
+            return CompletableFuture.completedFuture(ichatbasecomponent);
+         }
+
+         return CompletableFuture.supplyAsync(() -> {
+            return this.decorateCB(entityplayer, ichatbasecomponent); // CatServer - 1.20.1
+         }, chatExecutor);
+      };
+      // CraftBukkit end
+   }
+
+   // CatServer start - 1.20.1
+   public Component decorateCB(@javax.annotation.Nonnull ServerPlayer entityplayer, @javax.annotation.Nullable Component ichatbasecomponent) {
+      AsyncPlayerChatPreviewEvent event = new AsyncPlayerChatPreviewEvent(true, entityplayer.getBukkitEntity(), CraftChatMessage.fromComponent(ichatbasecomponent), new LazyPlayerSet(this));
+      String originalFormat = event.getFormat(), originalMessage = event.getMessage();
+      this.server.getPluginManager().callEvent(event);
+
+      if (originalFormat.equals(event.getFormat()) && originalMessage.equals(event.getMessage()) && event.getPlayer().getName().equalsIgnoreCase(event.getPlayer().getDisplayName())) {
+         return ichatbasecomponent;
+      }
+
+      return CraftChatMessage.fromStringOrNull(String.format(event.getFormat(), event.getPlayer().getDisplayName(), event.getMessage()));
+   }
+   // CatServer end
 
    public static record ReloadableResources(CloseableResourceManager f_206584_, ReloadableServerResources f_206585_) implements AutoCloseable {
       public void close() {
