--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -157,7 +_,11 @@
                for(Entry<ResourceLocation, AdvancementProgress> entry : stream.collect(Collectors.toList())) {
                   Advancement advancement = p_136007_.m_136041_(entry.getKey());
                   if (advancement == null) {
-                     f_135958_.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), this.f_135963_);
+                     // CraftBukkit start
+                     if (entry.getKey().m_135827_().equals("minecraft")) {
+                        f_135958_.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), this.f_135963_);
+                     }
+                     // CraftBukkit end
                   } else {
                      this.m_135985_(advancement, entry.getValue());
                   }
@@ -181,11 +_,16 @@
       }
 
       this.m_136002_(p_136007_);
+
+      if (net.minecraftforge.common.ForgeConfig.SERVER.fixAdvancementLoading.get())
+         net.minecraftforge.common.AdvancementLoadFix.loadVisibility(this, this.f_135965_, this.f_135966_, this.f_135964_, this.f_135967_, this::m_136012_);
+      else
       this.m_136001_();
       this.m_135994_(p_136007_);
    }
 
    public void m_135991_() {
+      if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return; // Spigot
       Map<ResourceLocation, AdvancementProgress> map = Maps.newHashMap();
 
       for(Entry<Advancement, AdvancementProgress> entry : this.f_135964_.entrySet()) {
@@ -239,6 +_,8 @@
    }
 
    public boolean m_135988_(Advancement p_135989_, String p_135990_) {
+      // Forge: don't grant advancements for fake players
+      if (this.f_135968_ instanceof net.minecraftforge.common.util.FakePlayer) return false;
       boolean flag = false;
       AdvancementProgress advancementprogress = this.m_135996_(p_135989_);
       boolean flag1 = advancementprogress.m_8193_();
@@ -247,10 +_,12 @@
          this.f_135967_.add(p_135989_);
          flag = true;
          if (!flag1 && advancementprogress.m_8193_()) {
+            this.f_135968_.f_19853_.getCraftServer().getPluginManager().callEvent(new org.bukkit.event.player.PlayerAdvancementDoneEvent(this.f_135968_.getBukkitEntity(), p_135989_.bukkit)); // CraftBukkit
             p_135989_.m_138321_().m_9989_(this.f_135968_);
             if (p_135989_.m_138320_() != null && p_135989_.m_138320_().m_14996_() && this.f_135968_.f_19853_.m_46469_().m_46207_(GameRules.f_46153_)) {
                this.f_135962_.m_11264_(new TranslatableComponent("chat.type.advancement." + p_135989_.m_138320_().m_14992_().m_15548_(), this.f_135968_.m_5446_(), p_135989_.m_138330_()), ChatType.SYSTEM, Util.f_137441_);
             }
+            net.minecraftforge.common.ForgeHooks.onAdvancement(this.f_135968_, p_135989_);
          }
       }
 
