--- a/net/minecraft/server/ServerScoreboard.java
+++ b/net/minecraft/server/ServerScoreboard.java
@@ -30,7 +_,7 @@
    public void m_5734_(Score p_136206_) {
       super.m_5734_(p_136206_);
       if (this.f_136194_.contains(p_136206_.m_83404_())) {
-         this.f_136193_.m_6846_().m_11268_(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, p_136206_.m_83404_().m_83320_(), p_136206_.m_83405_(), p_136206_.m_83400_()));
+         this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, p_136206_.m_83404_().m_83320_(), p_136206_.m_83405_(), p_136206_.m_83400_()));
       }
 
       this.m_136217_();
@@ -38,14 +_,14 @@
 
    public void m_7182_(String p_136210_) {
       super.m_7182_(p_136210_);
-      this.f_136193_.m_6846_().m_11268_(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, (String)null, p_136210_, 0));
+      this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, (String) null, p_136210_, 0));
       this.m_136217_();
    }
 
    public void m_5973_(String p_136212_, Objective p_136213_) {
       super.m_5973_(p_136212_, p_136213_);
       if (this.f_136194_.contains(p_136213_)) {
-         this.f_136193_.m_6846_().m_11268_(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, p_136213_.m_83320_(), p_136212_, 0));
+         this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, p_136213_.m_83320_(), p_136212_, 0));
       }
 
       this.m_136217_();
@@ -56,7 +_,7 @@
       super.m_7136_(p_136199_, p_136200_);
       if (objective != p_136200_ && objective != null) {
          if (this.m_136237_(objective) > 0) {
-            this.f_136193_.m_6846_().m_11268_(new ClientboundSetDisplayObjectivePacket(p_136199_, p_136200_));
+            this.broadcastAll(new ClientboundSetDisplayObjectivePacket(p_136199_, p_136200_));
          } else {
             this.m_136235_(objective);
          }
@@ -64,7 +_,7 @@
 
       if (p_136200_ != null) {
          if (this.f_136194_.contains(p_136200_)) {
-            this.f_136193_.m_6846_().m_11268_(new ClientboundSetDisplayObjectivePacket(p_136199_, p_136200_));
+            this.broadcastAll(new ClientboundSetDisplayObjectivePacket(p_136199_, p_136200_));
          } else {
             this.m_136231_(p_136200_);
          }
@@ -75,7 +_,7 @@
 
    public boolean m_6546_(String p_136215_, PlayerTeam p_136216_) {
       if (super.m_6546_(p_136215_, p_136216_)) {
-         this.f_136193_.m_6846_().m_11268_(ClientboundSetPlayerTeamPacket.m_179328_(p_136216_, p_136215_, ClientboundSetPlayerTeamPacket.Action.ADD));
+         this.broadcastAll(ClientboundSetPlayerTeamPacket.m_179328_(p_136216_, p_136215_, ClientboundSetPlayerTeamPacket.Action.ADD));
          this.m_136217_();
          return true;
       } else {
@@ -85,7 +_,7 @@
 
    public void m_6519_(String p_136223_, PlayerTeam p_136224_) {
       super.m_6519_(p_136223_, p_136224_);
-      this.f_136193_.m_6846_().m_11268_(ClientboundSetPlayerTeamPacket.m_179328_(p_136224_, p_136223_, ClientboundSetPlayerTeamPacket.Action.REMOVE));
+      this.broadcastAll(ClientboundSetPlayerTeamPacket.m_179328_(p_136224_, p_136223_, ClientboundSetPlayerTeamPacket.Action.REMOVE));
       this.m_136217_();
    }
 
@@ -97,7 +_,7 @@
    public void m_7091_(Objective p_136219_) {
       super.m_7091_(p_136219_);
       if (this.f_136194_.contains(p_136219_)) {
-         this.f_136193_.m_6846_().m_11268_(new ClientboundSetObjectivePacket(p_136219_, 2));
+         this.broadcastAll(new ClientboundSetObjectivePacket(p_136219_, 2));
       }
 
       this.m_136217_();
@@ -114,19 +_,19 @@
 
    public void m_7650_(PlayerTeam p_136204_) {
       super.m_7650_(p_136204_);
-      this.f_136193_.m_6846_().m_11268_(ClientboundSetPlayerTeamPacket.m_179332_(p_136204_, true));
+      this.broadcastAll(ClientboundSetPlayerTeamPacket.m_179332_(p_136204_, true));
       this.m_136217_();
    }
 
    public void m_7645_(PlayerTeam p_136221_) {
       super.m_7645_(p_136221_);
-      this.f_136193_.m_6846_().m_11268_(ClientboundSetPlayerTeamPacket.m_179332_(p_136221_, false));
+      this.broadcastAll(ClientboundSetPlayerTeamPacket.m_179332_(p_136221_, false));
       this.m_136217_();
    }
 
    public void m_7644_(PlayerTeam p_136228_) {
       super.m_7644_(p_136228_);
-      this.f_136193_.m_6846_().m_11268_(ClientboundSetPlayerTeamPacket.m_179326_(p_136228_));
+      this.broadcastAll(ClientboundSetPlayerTeamPacket.m_179326_(p_136228_));
       this.m_136217_();
    }
 
@@ -162,6 +_,7 @@
       List<Packet<?>> list = this.m_136229_(p_136232_);
 
       for(ServerPlayer serverplayer : this.f_136193_.m_6846_().m_11314_()) {
+         if (serverplayer.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(Packet<?> packet : list) {
             serverplayer.f_8906_.m_141995_(packet);
          }
@@ -187,6 +_,7 @@
       List<Packet<?>> list = this.m_136233_(p_136236_);
 
       for(ServerPlayer serverplayer : this.f_136193_.m_6846_().m_11314_()) {
+         if (serverplayer.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(Packet<?> packet : list) {
             serverplayer.f_8906_.m_141995_(packet);
          }
@@ -206,6 +_,16 @@
 
       return i;
    }
+
+   // CraftBukkit start - Send to players
+   private void broadcastAll(Packet packet) {
+      for (ServerPlayer entityplayer : (List<ServerPlayer>) this.f_136193_.m_6846_().f_11196_) {
+         if (entityplayer.getBukkitEntity().getScoreboard().getHandle() == this) {
+            entityplayer.f_8906_.m_141995_(packet);
+         }
+      }
+   }
+   // CraftBukkit end
 
    public ScoreboardSaveData m_180015_() {
       ScoreboardSaveData scoreboardsavedata = new ScoreboardSaveData(this);
