From 38ada0d9c173cb956a5423c3ebd57752dcefbcae Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Sun, 22 Sep 2013 19:10:53 +1000
Subject: [PATCH] Item Despawn Rate


diff --git a/src/main/java/net/minecraft/world/entity/item/EntityItem.java b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
index 2e93c1e67..97d6a3df0 100644
--- a/src/main/java/net/minecraft/world/entity/item/EntityItem.java
+++ b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
@@ -172,7 +172,7 @@ public class EntityItem extends Entity {
                 }
             }
 
-            if (!this.level.isClientSide && this.age >= 6000) {
+            if (!this.level.isClientSide && this.age >= level.spigotConfig.itemDespawnRate) { // Spigot
                 // CraftBukkit start - fire ItemDespawnEvent
                 if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
                     this.age = 0;
@@ -504,7 +504,7 @@ public class EntityItem extends Entity {
 
     public void makeFakeItem() {
         this.setNeverPickUp();
-        this.age = 5999;
+        this.age = level.spigotConfig.itemDespawnRate - 1; // Spigot
     }
 
     public float getSpin(float f) {
-- 
2.25.1

