From 1faf0ec615da3d799a7c5c29d9a23e6f729d777c Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Sun, 19 May 2013 18:29:48 +1000
Subject: [PATCH] Remove -o Option

Serves no purpose other than to confuse users.

diff --git a/src/main/java/net/minecraft/server/dedicated/PropertyManager.java b/src/main/java/net/minecraft/server/dedicated/PropertyManager.java
index 4592e9b05..08085ff66 100644
--- a/src/main/java/net/minecraft/server/dedicated/PropertyManager.java
+++ b/src/main/java/net/minecraft/server/dedicated/PropertyManager.java
@@ -33,7 +33,7 @@ public abstract class PropertyManager<T extends PropertyManager<T>> {
     }
 
     private String getOverride(String name, String value) {
-        if ((this.options != null) && (this.options.has(name))) {
+        if ((this.options != null) && (this.options.has(name)) && !name.equals( "online-mode")) { // Spigot
             return String.valueOf(this.options.valueOf(name));
         }
 
-- 
2.25.1

