From 398c5f44c051fdbbea808d99013695023661e21d Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Mon, 20 Jan 2014 13:44:07 +1100
Subject: [PATCH] Catch stalling on corrupted map data / NBT arrays.


diff --git a/src/main/java/net/minecraft/nbt/NBTTagByteArray.java b/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
index d1569b1e8..b4fe8ddb5 100644
--- a/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
+++ b/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
@@ -16,6 +16,7 @@ public class NBTTagByteArray extends NBTList<NBTTagByte> {
         public NBTTagByteArray load(DataInput datainput, int i, NBTReadLimiter nbtreadlimiter) throws IOException {
             nbtreadlimiter.accountBits(192L);
             int j = datainput.readInt();
+            com.google.common.base.Preconditions.checkArgument( j < 1 << 24); // Spigot
 
             nbtreadlimiter.accountBits(8L * (long) j);
             byte[] abyte = new byte[j];
diff --git a/src/main/java/net/minecraft/nbt/NBTTagIntArray.java b/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
index d3f3537d0..76c2dcd52 100644
--- a/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
+++ b/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
@@ -16,6 +16,7 @@ public class NBTTagIntArray extends NBTList<NBTTagInt> {
         public NBTTagIntArray load(DataInput datainput, int i, NBTReadLimiter nbtreadlimiter) throws IOException {
             nbtreadlimiter.accountBits(192L);
             int j = datainput.readInt();
+            com.google.common.base.Preconditions.checkArgument( j < 1 << 24); // Spigot
 
             nbtreadlimiter.accountBits(32L * (long) j);
             int[] aint = new int[j];
-- 
2.25.1

