From cce8155530c3e2162c1fc766ab05a7ebf5f32616 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sun, 26 Jan 2014 21:48:34 +1100
Subject: [PATCH] Configurable Ping Sample Size


diff --git a/src/main/java/net/minecraft/server/network/PacketStatusListener.java b/src/main/java/net/minecraft/server/network/PacketStatusListener.java
index d8cbae670..868237a37 100644
--- a/src/main/java/net/minecraft/server/network/PacketStatusListener.java
+++ b/src/main/java/net/minecraft/server/network/PacketStatusListener.java
@@ -132,6 +132,13 @@ public class PacketStatusListener implements PacketStatusInListener {
 
             ServerPing.ServerPingPlayerSample playerSample = new ServerPing.ServerPingPlayerSample(event.getMaxPlayers(), profiles.size());
             if (!this.server.hidesOnlinePlayers()) {
+                // Spigot Start
+                if ( !profiles.isEmpty() )
+                {
+                    java.util.Collections.shuffle( profiles ); // This sucks, its inefficient but we have no simple way of doing it differently
+                    profiles = profiles.subList( 0, Math.min( profiles.size(), org.spigotmc.SpigotConfig.playerSample ) ); // Cap the sample to n (or less) displayed players, ie: Vanilla behaviour
+                }
+                // Spigot End
                 playerSample.setSample(profiles.toArray(new GameProfile[profiles.size()]));
             }
-- 
2.25.1

