From e14f4fbc8e45968cd133d751e99601a07e3759bb Mon Sep 17 00:00:00 2001
From: David <dmck2b@gmail.com>
Date: Mon, 21 Apr 2014 12:43:08 +0100
Subject: [PATCH] Prevent NoClassDefError crash and notify on crash


diff --git a/src/main/java/net/minecraft/world/level/World.java b/src/main/java/net/minecraft/world/level/World.java
index d4dd846db..c57e2d40b 100644
--- a/src/main/java/net/minecraft/world/level/World.java
+++ b/src/main/java/net/minecraft/world/level/World.java
@@ -145,6 +145,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     public final org.spigotmc.SpigotWorldConfig spigotConfig; // Spigot
 
     public final SpigotTimings.WorldTimingsHandler timings; // Spigot
 
     public CraftWorld getWorld() {
         return this.world;
@@ -366,7 +367,13 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
                 // CraftBukkit start
                 if (!this.captureBlockStates) { // Don't notify clients or update physics while capturing blockstates
                     // Modularize client and physic updates
-                    notifyAndUpdatePhysics(blockposition, chunk, iblockdata1, iblockdata, iblockdata2, i, j);
+                    // Spigot start
+                    try {
+                        notifyAndUpdatePhysics(blockposition, chunk, iblockdata1, iblockdata, iblockdata2, i, j);
+                    } catch (StackOverflowError ex) {
+                        lastPhysicsProblem = new BlockPosition(blockposition);
+                    }
+                    // Spigot end
                 }
                 // CraftBukkit end
 
@@ -525,6 +532,10 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
                 }
                 // CraftBukkit end
                 iblockdata.neighborChanged(this, blockposition, block, blockposition1, false);
+            // Spigot Start
+            } catch (StackOverflowError ex) {
+                lastPhysicsProblem = new BlockPosition(blockposition);
+                // Spigot End
             } catch (Throwable throwable) {
                 CrashReport crashreport = CrashReport.forThrowable(throwable, "Exception while updating neighbours");
                 CrashReportSystemDetails crashreportsystemdetails = crashreport.addCategory("Block being updated");
-- 
2.25.1

