From 2717452042ff819bb909f3132d2ea21692374712 Mon Sep 17 00:00:00 2001
From: drXor <mcyoungsota@gmail.com>
Date: Fri, 25 Apr 2014 18:17:30 -0400
Subject: [PATCH] Configurable UserCache cap


diff --git a/src/main/java/net/minecraft/server/players/UserCache.java b/src/main/java/net/minecraft/server/players/UserCache.java
index c9a9e6266..961db661e 100644
--- a/src/main/java/net/minecraft/server/players/UserCache.java
+++ b/src/main/java/net/minecraft/server/players/UserCache.java
@@ -273,7 +273,7 @@ public class UserCache {
         JsonArray jsonarray = new JsonArray();
         DateFormat dateformat = createDateFormat();
 
-        this.getTopMRUProfiles(1000).forEach((usercache_usercacheentry) -> {
+        this.getTopMRUProfiles(org.spigotmc.SpigotConfig.userCacheCap).forEach((usercache_usercacheentry) -> { // Spigot
             jsonarray.add(writeGameProfile(usercache_usercacheentry, dateformat));
         });
         String s = this.gson.toJson(jsonarray);
-- 
2.25.1

