From 776875de851246efad55456beef9d57282321ff9 Mon Sep 17 00:00:00 2001
From: drXor <mcyoung@mit.edu>
Date: Sat, 9 Aug 2014 13:56:51 -0400
Subject: [PATCH] Configurable Hanging Tick


diff --git a/src/main/java/net/minecraft/world/entity/decoration/EntityHanging.java b/src/main/java/net/minecraft/world/entity/decoration/EntityHanging.java
index 0a79a2eee..b47fd1c13 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/EntityHanging.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/EntityHanging.java
@@ -117,7 +117,7 @@ public abstract class EntityHanging extends Entity {
     public void tick() {
         if (!this.level.isClientSide) {
             this.checkOutOfWorld();
-            if (this.checkInterval++ == 100) {
+            if (this.checkInterval++ == this.level.spigotConfig.hangingTickFrequency) { // Spigot
                 this.checkInterval = 0;
                 if (!this.isRemoved() && !this.survives()) {
                     // CraftBukkit start - fire break events
-- 
2.25.1

