--- a/net/minecraft/entity/merchant/villager/WanderingTraderEntity.java
+++ b/net/minecraft/entity/merchant/villager/WanderingTraderEntity.java
@@ -44,6 +_,10 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftMerchantRecipe;
+import org.bukkit.entity.AbstractVillager;
+import org.bukkit.event.entity.VillagerAcquireTradeEvent;
 
 public class WanderingTraderEntity extends AbstractVillagerEntity {
    @Nullable
@@ -121,7 +_,15 @@
          VillagerTrades.ITrade villagertrades$itrade = avillagertrades$itrade1[i];
          MerchantOffer merchantoffer = villagertrades$itrade.func_221182_a(this, this.field_70146_Z);
          if (merchantoffer != null) {
-            merchantoffers.add(merchantoffer);
+            // CraftBukkit start
+            final VillagerAcquireTradeEvent event = new VillagerAcquireTradeEvent((AbstractVillager) this.getBukkitEntity(), merchantoffer.asBukkit());
+            if (this.valid) {
+               Bukkit.getPluginManager().callEvent(event);
+            }
+            if (!event.isCancelled()) {
+               merchantoffers.add(CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+            }
+            // CraftBukkit end
          }
 
       }
