--- ../src-base/minecraft/net/minecraft/command/server/CommandTeleport.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandTeleport.java
@@ -91,7 +91,7 @@
             }
 
             teleportingEntity.dismountRidingEntity();
-            ((EntityPlayerMP)teleportingEntity).connection.setPlayerLocation(argX.getResult(), argY.getResult(), argZ.getResult(), f, f1, set);
+            ((EntityPlayerMP)teleportingEntity).connection.setPlayerLocation(argX.getResult(), argY.getResult(), argZ.getResult(), f, f1, set, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND);
             teleportingEntity.setRotationYawHead(f);
         }
         else
