--- ../src-base/minecraft/net/minecraft/command/server/CommandSetBlock.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandSetBlock.java
@@ -70,7 +70,7 @@
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 boolean flag = false;
 
-                if (args.length >= 7 && block.hasTileEntity())
+                if (args.length >= 7 && block.hasTileEntity(iblockstate))
                 {
                     String s = buildString(args, 6);
 
