--- a/net/minecraft/entity/ai/brain/BrainUtil.java
+++ b/net/minecraft/entity/ai/brain/BrainUtil.java
@@ -75,6 +_,7 @@
    }
 
    public static void func_233865_a_(LivingEntity p_233865_0_, ItemStack p_233865_1_, Vector3d p_233865_2_) {
+      if (p_233865_1_.func_190926_b()) return; // CraftBukkit - SPIGOT-4940: no empty loot
       double d0 = p_233865_0_.func_226280_cw_() - (double)0.3F;
       ItemEntity itementity = new ItemEntity(p_233865_0_.field_70170_p, p_233865_0_.func_226277_ct_(), d0, p_233865_0_.func_226281_cx_(), p_233865_1_);
       float f = 0.3F;
@@ -82,6 +_,13 @@
       vector3d = vector3d.func_72432_b().func_186678_a((double)0.3F);
       itementity.func_213317_d(vector3d);
       itementity.func_174869_p();
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(p_233865_0_.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+      itementity.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_233865_0_.field_70170_p.func_217376_c(itementity);
    }
 
