--- ../src-base/minecraft/net/minecraft/entity/effect/EntityLightningBolt.java
+++ ../src-work/minecraft/net/minecraft/entity/effect/EntityLightningBolt.java
@@ -12,13 +12,15 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.EnumDifficulty;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public class EntityLightningBolt extends EntityWeatherEffect
 {
     private int lightningState;
     public long boltVertex;
     private int boltLivingTime;
-    private final boolean effectOnly;
+    public final boolean effectOnly;
+    //public boolean isEffect; // CraftBukkit
 
     public EntityLightningBolt(World worldIn, double x, double y, double z, boolean effectOnlyIn)
     {
