--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -1,5 +1,7 @@
 package net.minecraft.item.crafting;
 
+import java.util.List;
+
 import javax.annotation.Nullable;
 import net.minecraft.init.Items;
 import net.minecraft.inventory.InventoryCrafting;
@@ -7,9 +9,13 @@
 import net.minecraft.item.ItemWrittenBook;
 import net.minecraft.world.World;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends ShapelessRecipes implements IRecipe
 {
-    public boolean matches(InventoryCrafting inv, World worldIn)
+    public RecipeBookCloning() {
+		super(new ItemStack(Items.WRITTEN_BOOK, 0, -1), java.util.Arrays.asList(new ItemStack(Items.WRITABLE_BOOK, 0, 0)));
+	}
+
+	public boolean matches(InventoryCrafting inv, World worldIn)
     {
         int i = 0;
         ItemStack itemstack = null;
