--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
@@ -8,10 +8,18 @@
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.world.World;
 
-public class RecipeTippedArrow implements IRecipe
+public class RecipeTippedArrow extends ShapedRecipes implements IRecipe
 {
-    private static final ItemStack[] EMPTY_ITEMS = new ItemStack[9];
+    public RecipeTippedArrow() {
+    	super(3, 3, new ItemStack[]{
+    			new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0),
+    			new ItemStack(Items.ARROW, 0), new ItemStack(Items.LINGERING_POTION, 0), new ItemStack(Items.ARROW, 0),
+    			new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0)
+    	}, new ItemStack(Items.TIPPED_ARROW, 8));
+	}
 
+	private static final ItemStack[] EMPTY_ITEMS = new ItemStack[9];
+
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
         if (inv.getWidth() == 3 && inv.getHeight() == 3)
