--- ../src-base/minecraft/net/minecraft/item/crafting/RecipesMapCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipesMapCloning.java
@@ -1,14 +1,20 @@
 package net.minecraft.item.crafting;
 
+import java.util.List;
+
 import javax.annotation.Nullable;
 import net.minecraft.init.Items;
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
 
-public class RecipesMapCloning implements IRecipe
+public class RecipesMapCloning extends ShapelessRecipes implements IRecipe
 {
-    public boolean matches(InventoryCrafting inv, World worldIn)
+    public RecipesMapCloning() {
+		super(new ItemStack(Items.MAP, 0, -1), java.util.Arrays.asList(new ItemStack(Items.MAP, 0, 0)));
+	}
+
+	public boolean matches(InventoryCrafting inv, World worldIn)
     {
         int i = 0;
         ItemStack itemstack = null;
