--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -3,14 +3,18 @@
 import com.google.common.collect.Lists;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import net.minecraft.init.Items;
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import org.bukkit.inventory.Recipe;
 
 public class RecipeRepairItem implements IRecipe
 {
-    public boolean matches(InventoryCrafting inv, World worldIn)
+
+    public boolean matches(InventoryCrafting inv, World worldIn)
     {
         List<ItemStack> list = Lists.<ItemStack>newArrayList();
 
@@ -110,4 +114,9 @@
 
         return aitemstack;
     }
+
+    @Override
+    public Recipe toBukkitRecipe() {
+        return null;
+    }
 }
