--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -1,16 +1,25 @@
 package net.minecraft.item.crafting;
 
-import javax.annotation.Nullable;
 import net.minecraft.init.Items;
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.ItemWrittenBook;
 import net.minecraft.world.World;
+import org.bukkit.inventory.Recipe;
 
-public class RecipeBookCloning implements IRecipe
+import javax.annotation.Nullable;
+
+public class RecipeBookCloning /*extends ShapelessRecipes*/ implements IRecipe
 {
+    /*public RecipeBookCloning() {
+        super(new ItemStack(Items.WRITTEN_BOOK, 0, -1), java.util.Arrays.asList(new ItemStack(Items.WRITABLE_BOOK, 0, 0)));
+    }*/
+
+    private ItemStack resultItem = new ItemStack(Items.WRITTEN_BOOK, 0, -1);
+
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
+        this.resultItem = null;
         int i = 0;
         ItemStack itemstack = null;
 
@@ -27,7 +36,7 @@
                         return false;
                     }
 
-                    itemstack = itemstack1;
+                    this.resultItem = itemstack = itemstack1;
                 }
                 else
                 {
@@ -104,7 +113,7 @@
     @Nullable
     public ItemStack getRecipeOutput()
     {
-        return null;
+        return this.resultItem;
     }
 
     public ItemStack[] getRemainingItems(InventoryCrafting inv)
@@ -125,4 +134,10 @@
 
         return aitemstack;
     }
+
+    @Override
+    public Recipe toBukkitRecipe() {
+        return null;
+    }
+
 }
