--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
@@ -7,11 +7,17 @@
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeTippedArrow implements IRecipe
+public class RecipeTippedArrow extends ShapedRecipes implements IRecipe
 {
+    public RecipeTippedArrow() {
+        super("", 3, 3, NonNullList.from(Ingredient.EMPTY,
+                Ingredient.fromItem(Items.ARROW), Ingredient.fromItem(Items.ARROW), Ingredient.fromItem(Items.ARROW),
+                Ingredient.fromItem(Items.ARROW), Ingredient.fromItem(Items.LINGERING_POTION), Ingredient.fromItem(Items.ARROW),
+                Ingredient.fromItem(Items.ARROW), Ingredient.fromItem(Items.ARROW), Ingredient.fromItem(Items.ARROW)),
+                new ItemStack(Items.TIPPED_ARROW, 8));
+    }
+
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
         if (inv.getWidth() == 3 && inv.getHeight() == 3)
@@ -83,7 +89,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     public boolean canFit(int width, int height)
     {
         return width >= 2 && height >= 2;
