--- ../src-base/minecraft/net/minecraft/item/crafting/RecipesBanners.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipesBanners.java
@@ -10,6 +10,7 @@
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.tileentity.TileEntityBanner;
 import net.minecraft.world.World;
+import org.bukkit.inventory.Recipe;
 
 public class RecipesBanners
 {
@@ -145,6 +146,11 @@
                 return aitemstack;
             }
 
+            @Override //TODO
+            public Recipe toBukkitRecipe() {
+                return null;
+            }
+
             @Nullable
             private TileEntityBanner.EnumBannerPattern matchPatterns(InventoryCrafting invCrafting)
             {
@@ -405,5 +411,10 @@
 
                 return aitemstack;
             }
+
+            @Override //TODO
+            public Recipe toBukkitRecipe() {
+                return null;
+            }
         }
 }
