--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -6,11 +6,13 @@
 import net.minecraft.item.ItemWrittenBook;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends ShapelessRecipes implements IRecipe
 {
+    public RecipeBookCloning() {
+        super("", new ItemStack(Items.WRITTEN_BOOK, 0, -1), NonNullList.from(Ingredient.EMPTY, Ingredient.fromItem(Items.WRITABLE_BOOK)));
+    }
+
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
         int i = 0;
@@ -127,7 +129,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     public boolean canFit(int width, int height)
     {
         return width >= 3 && height >= 3;
