--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
@@ -11,10 +11,14 @@
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.world.World;
 
-public class RecipeFireworks implements IRecipe
+public class RecipeFireworks extends ShapelessRecipes implements IRecipe
 {
-    private ItemStack resultItem;
+    public RecipeFireworks() {
+		super(new ItemStack(Items.FIREWORKS, 0, 0), java.util.Arrays.asList(new ItemStack(Items.GUNPOWDER, 0, 5)));
+	}
 
+	private ItemStack resultItem;
+
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
         this.resultItem = null;
