--- ../src-base/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
@@ -5,7 +5,17 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
 
-public class ShapedRecipes implements IRecipe
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftShapedRecipe;
+
+import luohuayu.CatServer.crafting.ICBRecipe;
+
+//CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftShapedRecipe;
+// CraftBukkit end
+
+public class ShapedRecipes implements ICBRecipe
 {
     public final int recipeWidth;
     public final int recipeHeight;
@@ -20,7 +30,61 @@
         this.recipeItems = p_i1917_3_;
         this.recipeOutput = output;
     }
-
+    // CraftBukkit start
+    public org.bukkit.inventory.ShapedRecipe toBukkitRecipe() {
+    	CraftItemStack result = CraftItemStack.asCraftMirror(this.recipeOutput);
+    	CraftShapedRecipe recipe = new CraftShapedRecipe(result, this);
+    	switch (this.recipeHeight) {
+    	case 1:
+    		switch (this.recipeWidth) {
+    		case 1:
+    			recipe.shape("a");
+    			break;
+    		case 2:
+    			recipe.shape("ab");
+    			break;
+    		case 3:
+    			recipe.shape("abc");
+    			break;
+    		}
+    		break;
+    	case 2:
+    		switch (this.recipeWidth) {
+    		case 1:
+    			recipe.shape("a","b");
+    			break;
+    		case 2:
+    			recipe.shape("ab","cd");
+    			break;
+    		case 3:
+    			recipe.shape("abc","def");
+    			break;
+    		}
+                break;
+    	case 3:
+    		switch (this.recipeWidth) {
+    		case 1:
+    			recipe.shape("a","b","c");
+    			break;
+    		case 2:
+    			recipe.shape("ab","cd","ef");
+    			break;
+    		case 3:
+    			recipe.shape("abc","def","ghi");
+    			break;
+    		}
+    		break;
+    	}
+    	char c = 'a';
+    	for (ItemStack stack : this.recipeItems) {
+    		if (stack != null) {
+    			recipe.setIngredient(c, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.getItem()), stack.getMetadata());
+    		}
+    		c++;
+    	}
+    	return recipe;
+    }
+    // CraftBukkit end
     @Nullable
     public ItemStack getRecipeOutput()
     {
