--- ../src-base/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
@@ -9,6 +9,7 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntityBanner;
 import net.minecraft.world.World;
+import org.bukkit.inventory.Recipe;
 
 public class ShieldRecipes
 {
@@ -138,5 +139,10 @@
 
                 return aitemstack;
             }
+
+            @Override //TODO
+            public Recipe toBukkitRecipe() {
+                return null;
+            }
         }
 }
