--- ../src-base/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
@@ -6,13 +6,17 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class ShieldRecipes
 {
-    public static class Decoration implements IRecipe
+    public static class Decoration extends ShapelessRecipes implements IRecipe
         {
+            // CraftBukkit start - Delegate to new parent class with bogus info
+            public Decoration() {
+                super("", new ItemStack(Items.SHIELD, 0, 0), NonNullList.from(Ingredient.EMPTY, Ingredient.fromItem(Items.BANNER)));
+            }
+            // CraftBukkit end
+
             public boolean matches(InventoryCrafting inv, World worldIn)
             {
                 ItemStack itemstack = ItemStack.EMPTY;
@@ -128,7 +132,6 @@
                 return true;
             }
 
-            @SideOnly(Side.CLIENT)
             public boolean canFit(int width, int height)
             {
                 return width * height >= 2;
