--- ../src-base/minecraft/net/minecraft/util/CooldownTracker.java
+++ ../src-work/minecraft/net/minecraft/util/CooldownTracker.java
@@ -11,8 +11,8 @@
 
 public class CooldownTracker
 {
-    private final Map<Item, CooldownTracker.Cooldown> cooldowns = Maps.<Item, CooldownTracker.Cooldown>newHashMap();
-    private int ticks;
+    public final Map<Item, Cooldown> cooldowns = Maps.<Item, Cooldown>newHashMap();
+    public int ticks;
 
     public boolean hasCooldown(Item itemIn)
     {
@@ -77,10 +77,10 @@
     {
     }
 
-    class Cooldown
+    public class Cooldown
     {
-        final int createTicks;
-        final int expireTicks;
+        public final int createTicks;
+        public final int expireTicks;
 
         private Cooldown(int createTicksIn, int expireTicksIn)
         {
