--- ../src-base/minecraft/net/minecraft/util/DamageSource.java
+++ ../src-work/minecraft/net/minecraft/util/DamageSource.java
@@ -45,6 +45,17 @@
     private boolean explosion;
     public String damageType;
 
+    private boolean sweep;
+
+    public boolean isSweep() {
+        return sweep;
+    }
+
+    public DamageSource sweep() {
+        this.sweep = true;
+        return this;
+    }
+
     public static DamageSource causeMobDamage(EntityLivingBase mob)
     {
         return new EntityDamageSource("mob", mob);
