--- ../src-base/minecraft/net/minecraft/util/Util.java
+++ ../src-work/minecraft/net/minecraft/util/Util.java
@@ -18,27 +18,27 @@
 
         if (s.contains("win"))
         {
-            return Util.EnumOS.WINDOWS;
+            return EnumOS.WINDOWS;
         }
         else if (s.contains("mac"))
         {
-            return Util.EnumOS.OSX;
+            return EnumOS.OSX;
         }
         else if (s.contains("solaris"))
         {
-            return Util.EnumOS.SOLARIS;
+            return EnumOS.SOLARIS;
         }
         else if (s.contains("sunos"))
         {
-            return Util.EnumOS.SOLARIS;
+            return EnumOS.SOLARIS;
         }
         else if (s.contains("linux"))
         {
-            return Util.EnumOS.LINUX;
+            return EnumOS.LINUX;
         }
         else
         {
-            return s.contains("unix") ? Util.EnumOS.LINUX : Util.EnumOS.UNKNOWN;
+            return s.contains("unix") ? EnumOS.LINUX : EnumOS.UNKNOWN;
         }
     }
 
