--- ../src-base/minecraft/net/minecraft/util/WeightedRandom.java
+++ ../src-work/minecraft/net/minecraft/util/WeightedRandom.java
@@ -5,21 +5,21 @@
 
 public class WeightedRandom
 {
-    public static int getTotalWeight(List <? extends WeightedRandom.Item > collection)
+    public static int getTotalWeight(List <? extends Item > collection)
     {
         int i = 0;
         int j = 0;
 
         for (int k = collection.size(); j < k; ++j)
         {
-            WeightedRandom.Item weightedrandom$item = collection.get(j);
+            Item weightedrandom$item = collection.get(j);
             i += weightedrandom$item.itemWeight;
         }
 
         return i;
     }
 
-    public static <T extends WeightedRandom.Item> T getRandomItem(Random random, List<T> collection, int totalWeight)
+    public static <T extends Item> T getRandomItem(Random random, List<T> collection, int totalWeight)
     {
         if (totalWeight <= 0)
         {
@@ -32,7 +32,7 @@
         }
     }
 
-    public static <T extends WeightedRandom.Item> T getRandomItem(List<T> collection, int weight)
+    public static <T extends Item> T getRandomItem(List<T> collection, int weight)
     {
         int i = 0;
 
@@ -50,7 +50,7 @@
         return (T)null;
     }
 
-    public static <T extends WeightedRandom.Item> T getRandomItem(Random random, List<T> collection)
+    public static <T extends Item> T getRandomItem(Random random, List<T> collection)
     {
         return (T)getRandomItem(random, collection, getTotalWeight(collection));
     }
