--- a/net/minecraft/util/text/Color.java
+++ b/net/minecraft/util/text/Color.java
@@ -11,7 +_,7 @@
 
 public final class Color {
    private static final Map<TextFormatting, Color> field_240738_a_ = Stream.of(TextFormatting.values()).filter(TextFormatting::func_96302_c).collect(ImmutableMap.toImmutableMap(Function.identity(), (p_240748_0_) -> {
-      return new Color(p_240748_0_.func_211163_e(), p_240748_0_.func_96297_d());
+      return new Color(p_240748_0_.func_211163_e(), p_240748_0_.func_96297_d(), p_240748_0_); // CraftBukkit
    }));
    private static final Map<String, Color> field_240739_b_ = field_240738_a_.values().stream().collect(ImmutableMap.toImmutableMap((p_240746_0_) -> {
       return p_240746_0_.field_240741_d_;
@@ -20,15 +_,30 @@
    @Nullable
    private final String field_240741_d_;
 
+   // CraftBukkit start
+   @Nullable
+   public final TextFormatting format;
+
+   private Color(int color, String name, TextFormatting format) {
+      this.field_240740_c_ = color;
+      this.field_240741_d_ = name;
+      this.format = format;
+   }
+
+   // LoliServer start
    private Color(int p_i232573_1_, String p_i232573_2_) {
       this.field_240740_c_ = p_i232573_1_;
       this.field_240741_d_ = p_i232573_2_;
+      this.format = null;
    }
+   // LoliServer end
 
    private Color(int p_i232572_1_) {
       this.field_240740_c_ = p_i232572_1_;
       this.field_240741_d_ = null;
+      this.format = null;
    }
+   // CraftBukkit end
 
    @OnlyIn(Dist.CLIENT)
    public int func_240742_a_() {
