--- a/net/minecraft/util/text/TextFormatting.java
+++ b/net/minecraft/util/text/TextFormatting.java
@@ -43,7 +_,7 @@
    }));
    private static final Pattern field_96330_y = Pattern.compile("(?i)\u00a7[0-9A-FK-OR]");
    private final String field_175748_y;
-   private final char field_96329_z;
+   public final char field_96329_z;
    private final boolean field_96303_A;
    private final String field_96304_B;
    private final int field_175747_C;
