--- a/net/minecraft/client/gui/screen/OptimizeWorldScreen.java
+++ b/net/minecraft/client/gui/screen/OptimizeWorldScreen.java
@@ -55,7 +_,7 @@
    private OptimizeWorldScreen(BooleanConsumer p_i232319_1_, DataFixer p_i232319_2_, SaveFormat.LevelSave p_i232319_3_, WorldSettings p_i232319_4_, boolean p_i232319_5_, ImmutableSet<RegistryKey<World>> p_i232319_6_) {
       super(new TranslationTextComponent("optimizeWorld.title", p_i232319_4_.func_234947_a_()));
       this.field_214332_b = p_i232319_1_;
-      this.field_212203_f = new WorldOptimizer(p_i232319_3_, p_i232319_2_, p_i232319_6_, p_i232319_5_);
+      this.field_212203_f = new WorldOptimizer(p_i232319_3_, p_i232319_2_, null, p_i232319_5_);
    }
 
    protected void func_231160_c_() {
