--- a/net/minecraft/network/play/server/SChatPacket.java
+++ b/net/minecraft/network/play/server/SChatPacket.java
@@ -12,6 +_,7 @@
 
 public class SChatPacket implements IPacket<IClientPlayNetHandler> {
    private ITextComponent field_148919_a;
+   public net.md_5.bungee.api.chat.BaseComponent[] components; // Spigot
    private ChatType field_179842_b;
    private UUID field_240809_c_;
 
@@ -31,7 +_,13 @@
    }
 
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
-      p_148840_1_.func_179256_a(this.field_148919_a);
+      // Spigot start
+      if (components != null) {
+         p_148840_1_.func_180714_a(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+      } else {
+         p_148840_1_.func_179256_a(this.field_148919_a);
+      }
+      // Spigot end
       p_148840_1_.writeByte(this.field_179842_b.func_192583_a());
       p_148840_1_.func_179252_a(this.field_240809_c_);
    }
