--- a/net/minecraft/network/play/server/SMultiBlockChangePacket.java
+++ b/net/minecraft/network/play/server/SMultiBlockChangePacket.java
@@ -5,6 +_,7 @@
 import java.util.function.BiConsumer;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.client.network.play.IClientPlayNetHandler;
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
@@ -31,7 +_,7 @@
 
       for(short short1 : p_i242085_2_) {
          this.field_244306_b[i] = short1;
-         this.field_244307_c[i] = p_i242085_3_.func_177485_a(SectionPos.func_243641_a(short1), SectionPos.func_243642_b(short1), SectionPos.func_243643_c(short1));
+         this.field_244307_c[i] = (p_i242085_3_ != null) ? p_i242085_3_.func_177485_a(SectionPos.func_243641_a(short1), SectionPos.func_243642_b(short1), SectionPos.func_243643_c(short1)): Blocks.field_150350_a.func_176223_P(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
          ++i;
       }
 
