--- a/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
+++ b/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
@@ -9,8 +_,8 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class SPlayerListHeaderFooterPacket implements IPacket<IClientPlayNetHandler> {
-   private ITextComponent field_179703_a;
-   private ITextComponent field_179702_b;
+   public ITextComponent field_179703_a;
+   public ITextComponent field_179702_b;
 
    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
       this.field_179703_a = p_148837_1_.func_179258_d();
