--- a/net/minecraft/network/play/server/SWorldBorderPacket.java
+++ b/net/minecraft/network/play/server/SWorldBorderPacket.java
@@ -24,8 +_,10 @@
 
    public SWorldBorderPacket(WorldBorder p_i46921_1_, SWorldBorderPacket.Action p_i46921_2_) {
       this.field_179795_a = p_i46921_2_;
-      this.field_179794_c = p_i46921_1_.func_230316_a_();
-      this.field_179791_d = p_i46921_1_.func_230317_b_();
+      // CraftBukkit start - multiply out nether border
+      this.field_179794_c = p_i46921_1_.func_230316_a_() * p_i46921_1_.world.func_230315_m_().func_242724_f();
+      this.field_179791_d = p_i46921_1_.func_230317_b_() * p_i46921_1_.world.func_230315_m_().func_242724_f();
+      // CraftBukkit end
       this.field_179789_f = p_i46921_1_.func_177741_h();
       this.field_179792_e = p_i46921_1_.func_177751_j();
       this.field_179790_g = p_i46921_1_.func_177732_i();
