--- ../src-base/minecraft/net/minecraft/network/NetworkManager.java
+++ ../src-work/minecraft/net/minecraft/network/NetworkManager.java
@@ -157,7 +157,7 @@
 
     public void setNetHandler(INetHandler handler)
     {
-        Validate.notNull(handler, "packetListener", new Object[0]);
+        Validate.notNull(handler, "packetListener");
         LOGGER.debug("Set listener of {} to {}", new Object[] {this, handler});
         this.packetListener = handler;
     }
