--- ../src-base/minecraft/net/minecraft/network/PacketBuffer.java
+++ ../src-work/minecraft/net/minecraft/network/PacketBuffer.java
@@ -180,7 +180,7 @@
 
     public <T extends Enum<T>> T readEnumValue(Class<T> enumClass)
     {
-        return (T)((Enum[])enumClass.getEnumConstants())[this.readVarIntFromBuffer()];
+        return (T)(enumClass.getEnumConstants())[this.readVarIntFromBuffer()];
     }
 
     public PacketBuffer writeEnumValue(Enum<?> value)
