--- ../src-base/minecraft/net/minecraft/item/EnumRarity.java
+++ ../src-work/minecraft/net/minecraft/item/EnumRarity.java
@@ -2,7 +2,7 @@
 
 import net.minecraft.util.text.TextFormatting;
 
-public enum EnumRarity
+public enum EnumRarity implements net.minecraftforge.common.IRarity
 {
     COMMON(TextFormatting.WHITE, "Common"),
     UNCOMMON(TextFormatting.YELLOW, "Uncommon"),
@@ -17,4 +17,17 @@
         this.rarityColor = color;
         this.rarityName = name;
     }
+
+    @Override
+    public TextFormatting getColor()
+    {
+        return this.rarityColor;
+    }
+
+    @Override
+    public String getName()
+    {
+        return this.rarityName;
+    }
+
 }
