--- ../src-base/minecraft/net/minecraft/item/ItemDye.java
+++ ../src-work/minecraft/net/minecraft/item/ItemDye.java
@@ -1,6 +1,9 @@
 package net.minecraft.item;
 
 import java.util.List;
+
+import org.bukkit.event.entity.SheepDyeWoolEvent;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockOldLog;
 import net.minecraft.block.BlockPlanks;
@@ -173,6 +176,17 @@
 
             if (!entitysheep.getSheared() && entitysheep.getFleeceColor() != enumdyecolor)
             {
+                // CraftBukkit start
+                byte bColor = (byte) enumdyecolor.getMetadata();
+                SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) entitysheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor));
+                entitysheep.worldObj.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return false;
+                }
+
+                enumdyecolor = EnumDyeColor.byMetadata(event.getColor().getWoolData());
+                // CraftBukkit end
                 entitysheep.setFleeceColor(enumdyecolor);
                 --stack.stackSize;
             }
