--- ../src-base/minecraft/net/minecraft/item/ItemMultiTexture.java
+++ ../src-work/minecraft/net/minecraft/item/ItemMultiTexture.java
@@ -5,9 +5,9 @@
 public class ItemMultiTexture extends ItemBlock
 {
     protected final Block unused;
-    protected final ItemMultiTexture.Mapper nameFunction;
+    protected final Mapper nameFunction;
 
-    public ItemMultiTexture(Block p_i47262_1_, Block p_i47262_2_, ItemMultiTexture.Mapper p_i47262_3_)
+    public ItemMultiTexture(Block p_i47262_1_, Block p_i47262_2_, Mapper p_i47262_3_)
     {
         super(p_i47262_1_);
         this.unused = p_i47262_2_;
@@ -18,7 +18,7 @@
 
     public ItemMultiTexture(Block block, Block block2, final String[] namesByMeta)
     {
-        this(block, block2, new ItemMultiTexture.Mapper()
+        this(block, block2, new Mapper()
         {
             public String apply(ItemStack p_apply_1_)
             {
