--- ../src-base/minecraft/net/minecraft/item/ItemStack.java
+++ ../src-work/minecraft/net/minecraft/item/ItemStack.java
@@ -260,8 +260,10 @@
         return this.item == null ? 0 : this.item.getMaxDamage(this);
     }
 
-    public boolean attemptDamageItem(int amount, Random rand)
-    {
+    public boolean attemptDamageItem(int amount, Random rand) {
+        return attemptDamageItem(amount, rand, null);
+    }
+    public boolean attemptDamageItem(int amount, Random rand, EntityLivingBase entityLivingBase){
         if (!this.isItemStackDamageable())
         {
             return false;
@@ -282,7 +284,18 @@
                 }
 
                 amount -= j;
-
+                // Spigot start
+                /*if (entityLivingBase instanceof EntityPlayer) {
+                    org.bukkit.craftbukkit.inventory.CraftItemStack item = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this);
+                    org.bukkit.event.player.PlayerItemDamageEvent event = new org.bukkit.event.player.PlayerItemDamageEvent((org.bukkit.entity.Player) entityLivingBase.getBukkitEntity(), item, i);
+                    org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+                    if (i != event.getDamage() || event.isCancelled()) {
+                        event.getPlayer().updateInventory();
+                        }
+                        if (event.isCancelled()) return false;
+                    amount = event.getDamage();
+                    }*/
+                    // Spigot end
                 if (amount <= 0)
                 {
                     return false;
@@ -300,7 +313,7 @@
         {
             if (this.isItemStackDamageable())
             {
-                if (this.attemptDamageItem(amount, entityIn.getRNG()))
+                if (this.attemptDamageItem(amount, entityIn.getRNG(), entityIn))
                 {
                     entityIn.renderBrokenItemStack(this);
                     --this.stackSize;
