--- ../src-base/minecraft/net/minecraft/item/ItemChorusFruit.java
+++ ../src-work/minecraft/net/minecraft/item/ItemChorusFruit.java
@@ -1,8 +1,14 @@
 package net.minecraft.item;
 
 import javax.annotation.Nullable;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerTeleportEvent;
+
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.math.MathHelper;
@@ -32,6 +38,20 @@
                 double d4 = MathHelper.clamp_double(entityLiving.posY + (double)(entityLiving.getRNG().nextInt(16) - 8), 0.0D, (double)(worldIn.getActualHeight() - 1));
                 double d5 = entityLiving.posZ + (entityLiving.getRNG().nextDouble() - 0.5D) * 16.0D;
 
+                // CraftBukkit start
+                if (entityLiving instanceof EntityPlayerMP) {
+                    Player player = ((EntityPlayerMP) entityLiving).getBukkitEntity();
+                    PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), new Location(player.getWorld(), d3, d4, d5), PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+                    worldIn.getServer().getPluginManager().callEvent(teleEvent);
+                    if (teleEvent.isCancelled()) {
+                        break;
+                    }
+                    d3 = teleEvent.getTo().getX();
+                    d4 = teleEvent.getTo().getY();
+                    d5 = teleEvent.getTo().getZ();
+                }
+                // CraftBukkit end
+
                 if (entityLiving.isRiding())
                 {
                     entityLiving.dismountRidingEntity();
