--- ../src-base/minecraft/net/minecraft/item/ItemAxe.java
+++ ../src-work/minecraft/net/minecraft/item/ItemAxe.java
@@ -20,6 +20,13 @@
         this.attackSpeed = ATTACK_SPEEDS[material.ordinal()];
     }
 
+    protected ItemAxe(Item.ToolMaterial material, float damage, float speed)
+    {
+        super(material, EFFECTIVE_ON);
+        this.attackDamage = damage;
+        this.attackSpeed = speed;
+    }
+
     public float getDestroySpeed(ItemStack stack, IBlockState state)
     {
         Material material = state.getMaterial();
