--- ../src-base/minecraft/net/minecraft/item/ItemSpade.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSpade.java
@@ -19,7 +19,7 @@
 {
     private static final Set<Block> EFFECTIVE_ON = Sets.newHashSet(Blocks.CLAY, Blocks.DIRT, Blocks.FARMLAND, Blocks.GRASS, Blocks.GRAVEL, Blocks.MYCELIUM, Blocks.SAND, Blocks.SNOW, Blocks.SNOW_LAYER, Blocks.SOUL_SAND, Blocks.GRASS_PATH, Blocks.CONCRETE_POWDER);
 
-    public ItemSpade(Item.ToolMaterial material)
+    public ItemSpade(ToolMaterial material)
     {
         super(1.5F, -3.0F, material, EFFECTIVE_ON);
     }
