--- a/net/minecraft/dispenser/BeehiveDispenseBehavior.java
+++ b/net/minecraft/dispenser/BeehiveDispenseBehavior.java
@@ -17,13 +_,36 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class BeehiveDispenseBehavior extends OptionalDispenseBehavior {
    protected ItemStack func_82487_b(IBlockSource p_82487_1_, ItemStack p_82487_2_) {
       World world = p_82487_1_.func_197524_h();
+      // CraftBukkit start
+      org.bukkit.block.Block bukkitBlock = world.getWorld().getBlockAt(p_82487_1_.func_180699_d().func_177958_n(), p_82487_1_.func_180699_d().func_177956_o(), p_82487_1_.func_180699_d().func_177952_p());
+      CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_82487_2_);
+      BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+      if (!DispenserBlock.eventFired) {
+         world.getCBServer().getPluginManager().callEvent(event);
+      }
+      if (event.isCancelled()) {
+         return p_82487_2_;
+      }
+      if (!event.getItem().equals(craftItem)) {
+         // Chain to handler for new item
+         ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+         IDispenseItemBehavior idispensebehavior = DispenserBlock.field_149943_a.get(eventStack.func_77973_b());
+         if (idispensebehavior != IDispenseItemBehavior.field_223216_a_ && idispensebehavior != this) {
+            idispensebehavior.dispense(p_82487_1_, eventStack);
+            return p_82487_2_;
+         }
+      }
+      // CraftBukkit end
       if (!world.func_201670_d()) {
          BlockPos blockpos = p_82487_1_.func_180699_d().func_177972_a(p_82487_1_.func_189992_e().func_177229_b(DispenserBlock.field_176441_a));
-         this.func_239796_a_(func_239797_a_((ServerWorld)world, blockpos) || func_239798_b_((ServerWorld)world, blockpos));
+         this.func_239796_a_(func_239797_a_((ServerWorld)world, blockpos) || shear((ServerWorld)world, blockpos, bukkitBlock, craftItem));  // CraftBukkit
          if (this.func_239795_a_() && p_82487_2_.func_96631_a(1, world.func_201674_k(), (ServerPlayerEntity)null)) {
             p_82487_2_.func_190920_e(0);
          }
@@ -47,11 +_,16 @@
       return false;
    }
 
-   private static boolean func_239798_b_(ServerWorld p_239798_0_, BlockPos p_239798_1_) {
-      for(LivingEntity livingentity : p_239798_0_.func_175647_a(LivingEntity.class, new AxisAlignedBB(p_239798_1_), EntityPredicates.field_180132_d)) {
+   private static boolean shear(ServerWorld world, BlockPos pos, org.bukkit.block.Block bukkitBlock, CraftItemStack craftItem) { // CraftBukkit - add args
+      for(LivingEntity livingentity : world.func_175647_a(LivingEntity.class, new AxisAlignedBB(pos), EntityPredicates.field_180132_d)) {
          if (livingentity instanceof IShearable) {
             IShearable ishearable = (IShearable)livingentity;
             if (ishearable.func_230262_K__()) {
+               // CraftBukkit start
+               if (CraftEventFactory.callBlockShearEntityEvent(livingentity, bukkitBlock, craftItem).isCancelled()) {
+                  continue;
+               }
+               // CraftBukkit end
                ishearable.func_230263_a_(SoundCategory.BLOCKS);
                return true;
             }
