--- a/net/minecraft/dispenser/DispenseBoatBehavior.java
+++ b/net/minecraft/dispenser/DispenseBoatBehavior.java
@@ -7,6 +_,9 @@
 import net.minecraft.util.Direction;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.util.Vector;
 
 public class DispenseBoatBehavior extends DefaultDispenseItemBehavior {
    private final DefaultDispenseItemBehavior field_218402_b = new DefaultDispenseItemBehavior();
@@ -33,12 +_,34 @@
 
          d3 = 0.0D;
       }
+      // CraftBukkit start
+      ItemStack itemstack2 = p_82487_2_.func_77979_a(1);
+      org.bukkit.block.Block block = world.getWorld().getBlockAt(p_82487_1_.func_180699_d().func_177958_n(), p_82487_1_.func_180699_d().func_177956_o(), p_82487_1_.func_180699_d().func_177952_p());
+      CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack2);
+      BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new Vector(d0, d1 + d3, d2));
+      if (!DispenserBlock.eventFired) {
+         world.getCBServer().getPluginManager().callEvent(event);
+      }
+      if (event.isCancelled()) {
+         p_82487_2_.func_190917_f(1);
+         return p_82487_2_;
+      }
+      if (!event.getItem().equals(craftItem)) {
+         p_82487_2_.func_190917_f(1);
+         // Chain to handler for new item
+         final ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+         final IDispenseItemBehavior idispensebehavior = DispenserBlock.field_149943_a.get(eventStack.func_77973_b());
+         if (idispensebehavior != IDispenseItemBehavior.field_223216_a_ && idispensebehavior != this) {
+            idispensebehavior.dispense(p_82487_1_, eventStack);
+            return p_82487_2_;
+         }
+      }
+      BoatEntity boatentity = new BoatEntity(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+      // CraftBukkit end
 
-      BoatEntity boatentity = new BoatEntity(world, d0, d1 + d3, d2);
       boatentity.func_184458_a(this.field_218403_c);
       boatentity.field_70177_z = direction.func_185119_l();
-      world.func_217376_c(boatentity);
-      p_82487_2_.func_190918_g(1);
+      if (!world.func_217376_c(boatentity)) p_82487_2_.func_190918_g(1); // CraftBukkit
       return p_82487_2_;
    }
 
