--- a/net/minecraft/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/net/minecraft/dispenser/ShulkerBoxDispenseBehavior.java
@@ -7,6 +_,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.Direction;
 import net.minecraft.util.math.BlockPos;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class ShulkerBoxDispenseBehavior extends OptionalDispenseBehavior {
    protected ItemStack func_82487_b(IBlockSource p_82487_1_, ItemStack p_82487_2_) {
@@ -16,6 +_,32 @@
          Direction direction = p_82487_1_.func_189992_e().func_177229_b(DispenserBlock.field_176441_a);
          BlockPos blockpos = p_82487_1_.func_180699_d().func_177972_a(direction);
          Direction direction1 = p_82487_1_.func_197524_h().func_175623_d(blockpos.func_177977_b()) ? direction : Direction.UP;
+
+
+         // CraftBukkit start
+         org.bukkit.block.Block bukkitBlock = p_82487_1_.func_197524_h().getWorld().getBlockAt(p_82487_1_.func_180699_d().func_177958_n(), p_82487_1_.func_180699_d().func_177956_o(), p_82487_1_.func_180699_d().func_177952_p());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_82487_2_);
+
+         BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p()));
+         if (!DispenserBlock.eventFired) {
+            p_82487_1_.func_197524_h().getCBServer().getPluginManager().callEvent(event);
+         }
+
+         if (event.isCancelled()) {
+            return p_82487_2_;
+         }
+
+         if (!event.getItem().equals(craftItem)) {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            IDispenseItemBehavior idispensebehavior = (IDispenseItemBehavior) DispenserBlock.field_149943_a.get(eventStack);
+            if (idispensebehavior != IDispenseItemBehavior.field_223216_a_ && idispensebehavior != this) {
+               idispensebehavior.dispense(p_82487_1_, eventStack);
+               return p_82487_2_;
+            }
+         }
+         // CraftBukkit end
+
          this.func_239796_a_(((BlockItem)item).func_195942_a(new DirectionalPlaceContext(p_82487_1_.func_197524_h(), blockpos, direction, p_82487_2_, direction1)).func_226246_a_());
       }
 
