--- a/net/minecraft/util/math/BlockPos.java
+++ b/net/minecraft/util/math/BlockPos.java
@@ -10,6 +_,7 @@
 import java.util.stream.StreamSupport;
 import javax.annotation.concurrent.Immutable;
 import net.minecraft.dispenser.IPosition;
+import net.minecraft.entity.Entity;
 import net.minecraft.util.AxisRotation;
 import net.minecraft.util.Direction;
 import net.minecraft.util.Rotation;
@@ -39,6 +_,11 @@
    private static final long field_177993_j = (1L << field_177991_c) - 1L;
    private static final int field_218292_j = field_177989_d;
    private static final int field_218293_k = field_177989_d + field_177991_c;
+
+   // CraftBukkit
+   public BlockPos(Entity p_i46032_1_) {
+      this(p_i46032_1_.func_226277_ct_(), p_i46032_1_.func_226278_cu_(), p_i46032_1_.func_226281_cx_());
+   }
 
    public BlockPos(int p_i46030_1_, int p_i46030_2_, int p_i46030_3_) {
       super(p_i46030_1_, p_i46030_2_, p_i46030_3_);
