--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -116,6 +_,12 @@
       return i;
    }
 
+   // Normal compare sorts by path first, this compares namespace first.
+   public int compareNamespaced(ResourceLocation o) {
+      int ret = this.f_135804_.compareTo(o.f_135804_);
+      return ret != 0 ? ret : this.f_135805_.compareTo(o.f_135805_);
+   }
+
    public String m_179910_() {
       return this.toString().replace('/', '_').replace(':', '_');
    }
@@ -151,7 +_,7 @@
       return true;
    }
 
-   public static boolean m_135843_(String p_135844_) {
+   private static boolean m_135843_(String p_135844_) {
       for(int i = 0; i < p_135844_.length(); ++i) {
          if (!m_135835_(p_135844_.charAt(i))) {
             return false;
@@ -165,7 +_,7 @@
       return p_135829_ == '_' || p_135829_ == '-' || p_135829_ >= 'a' && p_135829_ <= 'z' || p_135829_ >= '0' && p_135829_ <= '9' || p_135829_ == '/' || p_135829_ == '.';
    }
 
-   public static boolean m_135835_(char p_135836_) {
+   private static boolean m_135835_(char p_135836_) {
       return p_135836_ == '_' || p_135836_ == '-' || p_135836_ >= 'a' && p_135836_ <= 'z' || p_135836_ >= '0' && p_135836_ <= '9' || p_135836_ == '.';
    }
 
