--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -79,7 +_,7 @@
    public static final Codec<ChunkGenerator> f_62136_ = BuiltInRegistries.f_256914_.m_194605_().dispatchStable(ChunkGenerator::m_6909_, Function.identity());
    protected final BiomeSource f_62137_;
    private final Supplier<List<FeatureSorter.StepFeatureData>> f_223020_;
-   private final Function<Holder<Biome>, BiomeGenerationSettings> f_223021_;
+   public final Function<Holder<Biome>, BiomeGenerationSettings> f_223021_; // CatServer - private -> public
 
    public ChunkGenerator(BiomeSource p_256133_) {
       this(p_256133_, (p_223234_) -> {
@@ -365,6 +_,29 @@
       }
    }
 
+   // CraftBukkit start
+   public void applyBiomeDecoration(WorldGenLevel generatoraccessseed, ChunkAccess ichunkaccess, StructureManager structuremanager, boolean vanilla) {
+      if (vanilla) {
+         m_213609_(generatoraccessseed, ichunkaccess, structuremanager);
+      }
+
+      org.bukkit.World world = generatoraccessseed.getMinecraftWorld().getWorld();
+      // only call when a populator is present (prevents unnecessary entity conversion)
+      if (!world.getPopulators().isEmpty()) {
+         org.bukkit.craftbukkit.v1_20_R1.generator.CraftLimitedRegion limitedRegion = new org.bukkit.craftbukkit.v1_20_R1.generator.CraftLimitedRegion(generatoraccessseed, ichunkaccess.m_7697_());
+         int x = ichunkaccess.m_7697_().f_45578_;
+         int z = ichunkaccess.m_7697_().f_45579_;
+         for (org.bukkit.generator.BlockPopulator populator : world.getPopulators()) {
+            WorldgenRandom seededrandom = new net.minecraft.world.level.levelgen.WorldgenRandom(new LegacyRandomSource(generatoraccessseed.m_7328_()));
+            seededrandom.m_64690_(generatoraccessseed.m_7328_(), x, z);
+            populator.populate(world, new org.bukkit.craftbukkit.v1_20_R1.util.RandomSourceWrapper.RandomWrapper(seededrandom), x, z, limitedRegion);
+         }
+         limitedRegion.saveEntities();
+         limitedRegion.breakLink();
+      }
+   }
+   // CraftBukkit end
+
    private static BoundingBox m_187717_(ChunkAccess p_187718_) {
       ChunkPos chunkpos = p_187718_.m_7697_();
       int i = chunkpos.m_45604_();
@@ -480,6 +_,14 @@
       Predicate<Holder<Biome>> predicate = holderset::m_203333_;
       StructureStart structurestart = structure.m_226596_(p_223107_, this, this.f_62137_, p_223108_, p_223109_, p_223110_, p_223112_, i, p_223111_, predicate);
       if (structurestart.m_73603_()) {
+         // CraftBukkit start - 1.20.1
+         BoundingBox box = structurestart.m_73601_();
+         org.bukkit.event.world.AsyncStructureSpawnEvent event = new org.bukkit.event.world.AsyncStructureSpawnEvent(p_223106_.f_220460_.getMinecraftWorld().getWorld(), org.bukkit.craftbukkit.v1_20_R1.generator.structure.CraftStructure.minecraftToBukkit(structure, p_223107_), new org.bukkit.util.BoundingBox(box.m_162395_(), box.m_162396_(), box.m_162398_(), box.m_162399_(), box.m_162400_(), box.m_162401_()), p_223112_.f_45578_, p_223112_.f_45579_);
+         org.bukkit.Bukkit.getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return true;
+         }
+         // CraftBukkit end
          p_223106_.m_220516_(p_223113_, structure, structurestart, p_223111_);
          return true;
       } else {
