--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -35,6 +_,16 @@
       this.f_187995_ = new PalettedContainer<>(p_282873_.m_206115_(), p_282873_.m_246971_(Biomes.f_48202_), PalettedContainer.Strategy.f_188138_);
    }
 
+   // CraftBukkit start
+   public void setBiome(int i, int j, int k, Holder<Biome> biome) {
+      // CatServer start - 1.20.1
+      if(this.f_187995_ instanceof PalettedContainer) {
+         ((PalettedContainer)this.f_187995_).m_156470_(i, j, k, biome);
+      }
+      // CatServer end
+   }
+   // CraftBukkit end
+
    public BlockState m_62982_(int p_62983_, int p_62984_, int p_62985_) {
       return this.f_62972_.m_63087_(p_62983_, p_62984_, p_62985_);
    }
