--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -39,6 +_,12 @@
       this.f_187995_ = new PalettedContainer<>(p_188002_.m_206115_(), p_188002_.m_206081_(Biomes.f_48202_), PalettedContainer.Strategy.f_188138_);
    }
 
+   // CraftBukkit start
+   public void setBiome(int i, int j, int k, Holder<Biome> biome) {
+      this.f_187995_.m_156470_(i, j, k, biome);
+   }
+   // CraftBukkit end
+
    public static int m_156458_(int p_156459_) {
       return p_156459_ << 4;
    }
