--- ../src-base/minecraft/net/minecraft/client/resources/data/MetadataSerializer.java
+++ ../src-work/minecraft/net/minecraft/client/resources/data/MetadataSerializer.java
@@ -14,7 +14,7 @@
 @SideOnly(Side.CLIENT)
 public class MetadataSerializer
 {
-    private final IRegistry < String, MetadataSerializer.Registration <? extends IMetadataSection >> metadataSectionSerializerRegistry = new RegistrySimple < String, MetadataSerializer.Registration <? extends IMetadataSection >> ();
+    private final IRegistry < String, Registration <? extends IMetadataSection >> metadataSectionSerializerRegistry = new RegistrySimple < String, Registration <? extends IMetadataSection >> ();
     private final GsonBuilder gsonBuilder = new GsonBuilder();
     private Gson gson;
 
@@ -27,7 +27,7 @@
 
     public <T extends IMetadataSection> void registerMetadataSectionType(IMetadataSectionSerializer<T> metadataSectionSerializer, Class<T> clazz)
     {
-        this.metadataSectionSerializerRegistry.putObject(metadataSectionSerializer.getSectionName(), new MetadataSerializer.Registration(metadataSectionSerializer, clazz));
+        this.metadataSectionSerializerRegistry.putObject(metadataSectionSerializer.getSectionName(), new Registration(metadataSectionSerializer, clazz));
         this.gsonBuilder.registerTypeAdapter(clazz, metadataSectionSerializer);
         this.gson = null;
     }
@@ -48,7 +48,7 @@
         }
         else
         {
-            MetadataSerializer.Registration<?> registration = (MetadataSerializer.Registration)this.metadataSectionSerializerRegistry.getObject(sectionName);
+            Registration<?> registration = (Registration)this.metadataSectionSerializerRegistry.getObject(sectionName);
 
             if (registration == null)
             {
