--- a/net/minecraft/entity/passive/BatEntity.java
+++ b/net/minecraft/entity/passive/BatEntity.java
@@ -27,6 +_,7 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class BatEntity extends AmbientEntity {
    private static final DataParameter<Byte> field_184660_a = EntityDataManager.func_187226_a(BatEntity.class, DataSerializers.field_187191_a);
@@ -115,16 +_,24 @@
             }
 
             if (this.field_70170_p.func_217370_a(field_213813_c, this) != null) {
+               // CraftBukkit Start - Call BatToggleSleepEvent
+               if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+                  this.func_82236_f(false);
+                  if (!flag) {
+                     this.field_70170_p.func_217378_a((PlayerEntity) null, 1025, blockpos, 0);
+                  }
+               }
+               // CraftBukkit end
+            }
+         } else {
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if(CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
                this.func_82236_f(false);
                if (!flag) {
-                  this.field_70170_p.func_217378_a((PlayerEntity)null, 1025, blockpos, 0);
+                  this.field_70170_p.func_217378_a((PlayerEntity) null, 1025, blockpos, 0);
                }
             }
-         } else {
-            this.func_82236_f(false);
-            if (!flag) {
-               this.field_70170_p.func_217378_a((PlayerEntity)null, 1025, blockpos, 0);
-            }
+            // CraftBukkit End - Call BatToggleSleepEvent
          }
       } else {
          if (this.field_82237_a != null && (!this.field_70170_p.func_175623_d(this.field_82237_a) || this.field_82237_a.func_177956_o() < 1)) {
@@ -146,7 +_,11 @@
          this.field_191988_bg = 0.5F;
          this.field_70177_z += f1;
          if (this.field_70146_Z.nextInt(100) == 0 && this.field_70170_p.func_180495_p(blockpos1).func_215686_e(this.field_70170_p, blockpos1)) {
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if(CraftEventFactory.handleBatToggleSleepEvent(this, false)){
             this.func_82236_f(true);
+         }
+            // CraftBukkit End
          }
       }
 
