--- a/net/minecraft/entity/passive/CatEntity.java
+++ b/net/minecraft/entity/passive/CatEntity.java
@@ -67,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class CatEntity extends TameableEntity {
    private static final Ingredient field_213426_bE = Ingredient.func_199804_a(Items.field_196086_aW, Items.field_196087_aX);
@@ -402,7 +_,7 @@
             }
          } else if (this.func_70877_b(itemstack)) {
             this.func_175505_a(p_230254_1_, itemstack);
-            if (this.field_70146_Z.nextInt(3) == 0) {
+            if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) { // CraftBukkit
                this.func_193101_c(p_230254_1_);
                this.func_233687_w_(true);
                this.field_70170_p.func_72960_a(this, (byte)7);
@@ -551,7 +_,13 @@
          LootContext.Builder lootcontext$builder = (new LootContext.Builder((ServerWorld)this.field_220806_a.field_70170_p)).func_216015_a(LootParameters.field_237457_g_, this.field_220806_a.func_213303_ch()).func_216015_a(LootParameters.field_216281_a, this.field_220806_a).func_216023_a(random);
 
          for(ItemStack itemstack : loottable.func_216113_a(lootcontext$builder.func_216022_a(LootParameterSets.field_216264_e))) {
-            this.field_220806_a.field_70170_p.func_217376_c(new ItemEntity(this.field_220806_a.field_70170_p, (double)blockpos$mutable.func_177958_n() - (double)MathHelper.func_76126_a(this.field_220806_a.field_70761_aq * ((float)Math.PI / 180F)), (double)blockpos$mutable.func_177956_o(), (double)blockpos$mutable.func_177952_p() + (double)MathHelper.func_76134_b(this.field_220806_a.field_70761_aq * ((float)Math.PI / 180F)), itemstack));
+            ItemEntity itemEntity = new ItemEntity(this.field_220806_a.field_70170_p, (double)blockpos$mutable.func_177958_n() - (double)MathHelper.func_76126_a(this.field_220806_a.field_70761_aq * ((float)Math.PI / 180F)), (double)blockpos$mutable.func_177956_o(), (double)blockpos$mutable.func_177952_p() + (double)MathHelper.func_76134_b(this.field_220806_a.field_70761_aq * ((float)Math.PI / 180F)), itemstack);
+            org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.field_220806_a.getBukkitEntity(), (org.bukkit.entity.Item) itemEntity.getBukkitEntity());
+            itemEntity.field_70170_p.getCBServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               continue;
+            }
+            this.field_220806_a.field_70170_p.func_217376_c(itemEntity);
          }
 
       }
@@ -579,7 +_,7 @@
 
    static class TemptGoal extends net.minecraft.entity.ai.goal.TemptGoal {
       @Nullable
-      private PlayerEntity field_220763_c;
+      private LivingEntity field_220763_c; // CraftBukkit
       private final CatEntity field_220764_d;
 
       public TemptGoal(CatEntity p_i50438_1_, double p_i50438_2_, Ingredient p_i50438_4_, boolean p_i50438_5_) {
