--- a/net/minecraft/entity/passive/ChickenEntity.java
+++ b/net/minecraft/entity/passive/ChickenEntity.java
@@ -68,6 +_,11 @@
    }
 
    public void func_70636_d() {
+      // CraftBukkit start
+      if (this.func_152116_bZ()) {
+         this.field_82179_bU = !this.func_213397_c(0);
+      }
+      // CraftBukkit end
       super.func_70636_d();
       this.field_70888_h = this.field_70886_e;
       this.field_70884_g = this.field_70883_f;
@@ -86,7 +_,9 @@
       this.field_70886_e += this.field_70889_i * 2.0F;
       if (!this.field_70170_p.field_72995_K && this.func_70089_S() && !this.func_70631_g_() && !this.func_152116_bZ() && --this.field_70887_j <= 0) {
          this.func_184185_a(SoundEvents.field_187665_Y, 1.0F, (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F);
+         this.forceDrops = true; // CraftBukkit
          this.func_199703_a(Items.field_151110_aK);
+         this.forceDrops = false; // CraftBukkit
          this.field_70887_j = this.field_70146_Z.nextInt(6000) + 6000;
       }
 
