--- a/net/minecraft/entity/passive/CowEntity.java
+++ b/net/minecraft/entity/passive/CowEntity.java
@@ -29,6 +_,9 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class CowEntity extends AnimalEntity {
    public CowEntity(EntityType<? extends CowEntity> p_i48567_1_, World p_i48567_2_) {
@@ -73,8 +_,14 @@
    public ActionResultType func_230254_b_(PlayerEntity p_230254_1_, Hand p_230254_2_) {
       ItemStack itemstack = p_230254_1_.func_184586_b(p_230254_2_);
       if (itemstack.func_77973_b() == Items.field_151133_ar && !this.func_70631_g_()) {
+         // CraftBukkit start - Got milk?
+         PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_230254_1_.field_70170_p, p_230254_1_, this.func_233580_cy_(), this.func_233580_cy_(), null, itemstack, Items.field_151117_aB);
+         if (event.isCancelled()) {
+            return ActionResultType.PASS;
+         }
+         // CraftBukkit end
          p_230254_1_.func_184185_a(SoundEvents.field_187564_an, 1.0F, 1.0F);
-         ItemStack itemstack1 = DrinkHelper.func_242398_a(itemstack, p_230254_1_, Items.field_151117_aB.func_190903_i());
+         ItemStack itemstack1 = DrinkHelper.func_242398_a(itemstack, p_230254_1_, CraftItemStack.asNMSCopy(event.getItemStack()));
          p_230254_1_.func_184611_a(p_230254_2_, itemstack1);
          return ActionResultType.func_233537_a_(this.field_70170_p.field_72995_K);
       } else {
