--- a/net/minecraft/entity/passive/DolphinEntity.java
+++ b/net/minecraft/entity/passive/DolphinEntity.java
@@ -69,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class DolphinEntity extends WaterMobEntity {
    private static final DataParameter<BlockPos> field_208014_b = EntityDataManager.func_187226_a(DolphinEntity.class, DataSerializers.field_187200_j);
@@ -221,6 +_,12 @@
       if (this.func_184582_a(EquipmentSlotType.MAINHAND).func_190926_b()) {
          ItemStack itemstack = p_175445_1_.func_92059_d();
          if (this.func_175448_a(itemstack)) {
+            // CraftBukkit start - call EntityPickupItemEvent
+            if(CraftEventFactory.callEntityPickupItemEvent(this,p_175445_1_, 0, false).isCancelled()){
+               return;
+            }
+            itemstack = p_175445_1_.func_92059_d(); // update ItemStack from event
+            // CraftBukkit end
             this.func_233630_a_(p_175445_1_);
             this.func_184201_a(EquipmentSlotType.MAINHAND, itemstack);
             this.field_82174_bp[EquipmentSlotType.MAINHAND.func_188454_b()] = 2.0F;
@@ -485,7 +_,7 @@
       }
 
       public boolean func_75250_a() {
-         return this.field_208057_a.func_208011_dD() && this.field_208057_a.func_70086_ai() >= 100;
+         return this.field_208057_a.func_208011_dD() && this.field_208057_a.func_70086_ai() >= 100 && this.field_208057_a.field_70170_p.getWorld().canGenerateStructures(); // MC-151364, SPIGOT-5494: hangs if generate-structures=false
       }
 
       public boolean func_75253_b() {
@@ -582,7 +_,7 @@
       }
 
       public void func_75249_e() {
-         this.field_206836_c.func_195064_c(new EffectInstance(Effects.field_206827_D, 100));
+         this.field_206836_c.addEffect(new EffectInstance(Effects.field_206827_D, 100), Cause.DOLPHIN); // CraftBukkit - reason
       }
 
       public void func_75251_c() {
@@ -599,7 +_,7 @@
          }
 
          if (this.field_206836_c.func_203007_ba() && this.field_206836_c.field_70170_p.field_73012_v.nextInt(6) == 0) {
-            this.field_206836_c.func_195064_c(new EffectInstance(Effects.field_206827_D, 100));
+            this.field_206836_c.addEffect(new EffectInstance(Effects.field_206827_D, 100), Cause.DOLPHIN); // CraftBukkit - reason
          }
 
       }
