--- a/net/minecraft/entity/passive/OcelotEntity.java
+++ b/net/minecraft/entity/passive/OcelotEntity.java
@@ -48,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class OcelotEntity extends AnimalEntity {
    private static final Ingredient field_195402_bB = Ingredient.func_199804_a(Items.field_196086_aW, Items.field_196087_aX);
@@ -118,7 +_,7 @@
    }
 
    public boolean func_213397_c(double p_213397_1_) {
-      return !this.func_213530_dX() && this.field_70173_aa > 2400;
+      return !this.func_213530_dX() /*&& this.ticksExisted > 2400; */; // CraftBukkit
    }
 
    public static AttributeModifierMap.MutableAttribute func_234201_eI_() {
@@ -163,7 +_,8 @@
       if ((this.field_70914_e == null || this.field_70914_e.func_75277_f()) && !this.func_213530_dX() && this.func_70877_b(itemstack) && p_230254_1_.func_70068_e(this) < 9.0D) {
          this.func_175505_a(p_230254_1_, itemstack);
          if (!this.field_70170_p.field_72995_K) {
-            if (this.field_70146_Z.nextInt(3) == 0) {
+            // CraftBukkit - added event call and isCancelled check
+            if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) {
                this.func_213528_r(true);
                this.func_213527_s(true);
                this.field_70170_p.func_72960_a(this, (byte)41);
