--- a/net/minecraft/entity/passive/PandaEntity.java
+++ b/net/minecraft/entity/passive/PandaEntity.java
@@ -61,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class PandaEntity extends AnimalEntity {
    private static final DataParameter<Integer> field_213609_bA = EntityDataManager.func_187226_a(PandaEntity.class, DataSerializers.field_187192_b);
@@ -492,7 +_,7 @@
    }
 
    protected void func_175445_a(ItemEntity p_175445_1_) {
-      if (this.func_184582_a(EquipmentSlotType.MAINHAND).func_190926_b() && field_213607_bQ.test(p_175445_1_)) {
+      if(!CraftEventFactory.callEntityPickupItemEvent(this, p_175445_1_, 0, !(this.func_184582_a(EquipmentSlotType.MAINHAND).func_190926_b() && field_213607_bQ.test(p_175445_1_))).isCancelled()) {  // CraftBukkit
          this.func_233630_a_(p_175445_1_);
          ItemStack itemstack = p_175445_1_.func_92059_d();
          this.func_184201_a(EquipmentSlotType.MAINHAND, itemstack);
@@ -928,7 +_,7 @@
 
       protected void func_220793_a(MobEntity p_220793_1_, LivingEntity p_220793_2_) {
          if (p_220793_1_ instanceof PandaEntity && ((PandaEntity)p_220793_1_).func_213398_dR()) {
-            p_220793_1_.func_70624_b(p_220793_2_);
+            p_220793_1_.setGoalTarget(p_220793_2_, TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit
          }
 
       }
