--- a/net/minecraft/entity/passive/RabbitEntity.java
+++ b/net/minecraft/entity/passive/RabbitEntity.java
@@ -48,7 +_,6 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.DifficultyInstance;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
@@ -57,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class RabbitEntity extends AnimalEntity {
    private static final DataParameter<Integer> field_184773_bv = EntityDataManager.func_187226_a(RabbitEntity.class, DataSerializers.field_187192_b);
@@ -71,10 +_,16 @@
       super(p_i50247_1_, p_i50247_2_);
       this.field_70767_i = new RabbitEntity.JumpHelperController(this);
       this.field_70765_h = new RabbitEntity.MoveHelperController(this);
+      initializePathFinderGoals(); // CraftBukkit - moved code
+   }
+
+   // CraftBukkit start - code from constructor
+   public void initializePathFinderGoals(){
       this.func_175515_b(0.0D);
    }
+   // CraftBukkit end
 
-   protected void func_184651_r() {
+   public void func_184651_r() { // protected->public CraftBukkit
       this.field_70714_bg.func_75776_a(1, new SwimGoal(this));
       this.field_70714_bg.func_75776_a(1, new RabbitEntity.PanicGoal(this, 2.2D));
       this.field_70714_bg.func_75776_a(2, new BreedGoal(this, 0.8D));
@@ -500,7 +_,7 @@
 
       public boolean func_75250_a() {
          if (this.field_179496_a <= 0) {
-            if (!this.field_179500_c.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_179500_c.field_70170_p, this.field_179500_c)) {
                return false;
             }
 
@@ -527,9 +_,19 @@
             if (this.field_179499_e && block instanceof CarrotBlock) {
                Integer integer = blockstate.func_177229_b(CarrotBlock.field_176488_a);
                if (integer == 0) {
+                  // CraftBukkit start
+                  if (CraftEventFactory.callEntityChangeBlockEvent(this.field_179500_c, blockpos, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+                     return;
+                  }
+                  // CraftBukkit end
                   world.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 2);
                   world.func_225521_a_(blockpos, true, this.field_179500_c);
                } else {
+                  // CraftBukkit start
+                  if (CraftEventFactory.callEntityChangeBlockEvent(this.field_179500_c, blockpos, blockstate.func_206870_a(CarrotBlock.field_176488_a, integer -1)).isCancelled()) {
+                     return;
+                  }
+                  // CraftBukkit end
                   world.func_180501_a(blockpos, blockstate.func_206870_a(CarrotBlock.field_176488_a, Integer.valueOf(integer - 1)), 2);
                   world.func_217379_c(2001, blockpos, Block.func_196246_j(blockstate));
                }
