--- a/net/minecraft/entity/passive/SquidEntity.java
+++ b/net/minecraft/entity/passive/SquidEntity.java
@@ -178,7 +_,7 @@
    }
 
    public static boolean func_223365_b(EntityType<SquidEntity> p_223365_0_, IWorld p_223365_1_, SpawnReason p_223365_2_, BlockPos p_223365_3_, Random p_223365_4_) {
-      return p_223365_3_.func_177956_o() > 45 && p_223365_3_.func_177956_o() < p_223365_1_.func_181545_F();
+      return p_223365_3_.func_177956_o() > p_223365_1_.getMinecraftWorld().spigotConfig.squidSpawnRangeMin && p_223365_3_.func_177956_o() < p_223365_1_.func_181545_F(); // Spigot
    }
 
    @OnlyIn(Dist.CLIENT)
