--- a/net/minecraft/entity/passive/TurtleEntity.java
+++ b/net/minecraft/entity/passive/TurtleEntity.java
@@ -61,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class TurtleEntity extends AnimalEntity {
    private static final DataParameter<BlockPos> field_203030_by = EntityDataManager.func_187226_a(TurtleEntity.class, DataSerializers.field_187200_j);
@@ -282,7 +_,9 @@
    protected void func_175500_n() {
       super.func_175500_n();
       if (!this.func_70631_g_() && this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223602_e)) {
+         this.forceDrops = true; // CraftBukkit
          this.func_199702_a(Items.field_203183_eM, 1);
+         this.forceDrops = false; // CraftBukkit
       }
 
    }
@@ -306,7 +_,9 @@
    }
 
    public void func_241841_a(ServerWorld p_241841_1_, LightningBoltEntity p_241841_2_) {
+      CraftEventFactory.entityDamage = p_241841_2_; // CraftBukkit
       this.func_70097_a(DamageSource.field_180137_b, Float.MAX_VALUE);
+      CraftEventFactory.entityDamage = null; // CraftBukkit
    }
 
    static class GoHomeGoal extends Goal {
@@ -429,8 +_,12 @@
                this.field_203122_f.func_203015_s(true);
             } else if (this.field_203122_f.field_203028_bF > 200) {
                World world = this.field_203122_f.field_70170_p;
-               world.func_184133_a((PlayerEntity)null, blockpos, SoundEvents.field_203278_iw, SoundCategory.BLOCKS, 0.3F, 0.9F + world.field_73012_v.nextFloat() * 0.2F);
-               world.func_180501_a(this.field_179494_b.func_177984_a(), Blocks.field_203213_jA.func_176223_P().func_206870_a(TurtleEggBlock.field_203171_b, Integer.valueOf(this.field_203122_f.field_70146_Z.nextInt(4) + 1)), 3);
+               // CraftBukkit start
+               if (!CraftEventFactory.callEntityChangeBlockEvent(this.field_203122_f, this.field_179494_b.func_177984_a(), (BlockState) Blocks.field_203213_jA.func_176223_P().func_206870_a(TurtleEggBlock.field_203171_b, this.field_203122_f.field_70146_Z.nextInt(4) + 1)).isCancelled()) {
+                  world.func_184133_a((PlayerEntity) null, blockpos, SoundEvents.field_203278_iw, SoundCategory.BLOCKS, 0.3F, 0.9F + world.field_73012_v.nextFloat() * 0.2F);
+                  world.func_180501_a(this.field_179494_b.func_177984_a(), Blocks.field_203213_jA.func_176223_P().func_206870_a(TurtleEggBlock.field_203171_b, Integer.valueOf(this.field_203122_f.field_70146_Z.nextInt(4) + 1)), 3);
+               }
+               // CraftBukkit end
                this.field_203122_f.func_203017_r(false);
                this.field_203122_f.func_203015_s(false);
                this.field_203122_f.func_204700_e(600);
