--- a/net/minecraft/entity/passive/WolfEntity.java
+++ b/net/minecraft/entity/passive/WolfEntity.java
@@ -61,6 +_,9 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class WolfEntity extends TameableEntity implements IAngerable {
    private static final DataParameter<Boolean> field_184760_bA = EntityDataManager.func_187226_a(WolfEntity.class, DataSerializers.field_187198_h);
@@ -110,6 +_,24 @@
       return MobEntity.func_233666_p_().func_233815_a_(Attributes.field_233821_d_, (double)0.3F).func_233815_a_(Attributes.field_233818_a_, 8.0D).func_233815_a_(Attributes.field_233823_f_, 2.0D);
    }
 
+   // CraftBukkit - add overriden version
+   @Override
+   public boolean setGoalTarget(LivingEntity entitylivingbaseIn, TargetReason reason, boolean fireEvent) {
+      if (!super.setGoalTarget(entitylivingbaseIn, reason, fireEvent)) {
+         return false;
+      }
+      entitylivingbaseIn = func_70638_az();
+      /*
+      if (entitylivingbaseIn == null) {
+         this.setAngry(false);
+      } else if (!this.isTamed()) {
+         this.setAngry(true);
+      }
+       */
+      return true;
+   }
+   // CraftBukkit end
+
    protected void func_70088_a() {
       super.func_70088_a();
       this.field_70180_af.func_187214_a(field_184760_bA, false);
@@ -133,7 +_,8 @@
          this.func_175547_a(DyeColor.func_196056_a(p_70037_1_.func_74762_e("CollarColor")));
       }
 
-      this.func_241358_a_((ServerWorld)this.field_70170_p, p_70037_1_);
+      if(!field_70170_p.field_72995_K) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
+         this.func_241358_a_((ServerWorld)this.field_70170_p, p_70037_1_);
    }
 
    protected SoundEvent func_184639_G() {
@@ -273,7 +_,7 @@
          return false;
       } else {
          Entity entity = p_70097_1_.func_76346_g();
-         this.func_233687_w_(false);
+         // this.setOrderedToSit(false);   // CraftBukkit - moved into EntityLiving.damageEntity(DamageSource, float)
          if (entity != null && !(entity instanceof PlayerEntity) && !(entity instanceof AbstractArrowEntity)) {
             p_70097_2_ = (p_70097_2_ + 1.0F) / 2.0F;
          }
@@ -295,7 +_,7 @@
       super.func_70903_f(p_70903_1_);
       if (p_70903_1_) {
          this.func_110148_a(Attributes.field_233818_a_).func_111128_a(20.0D);
-         this.func_70606_j(20.0F);
+         this.func_70606_j(this.func_110138_aP()); // CraftBukkit - 20.0 -> getMaxHealth()
       } else {
          this.func_110148_a(Attributes.field_233818_a_).func_111128_a(8.0D);
       }
@@ -316,7 +_,7 @@
                   itemstack.func_190918_g(1);
                }
 
-               this.func_70691_i((float)item.func_219967_s().func_221466_a());
+               this.heal((float)item.func_219967_s().func_221466_a(), RegainReason.EATING); // CraftBukkit
                return ActionResultType.SUCCESS;
             }
 
@@ -326,7 +_,7 @@
                   this.func_233687_w_(!this.func_233685_eM_());
                   this.field_70703_bu = false;
                   this.field_70699_by.func_75499_g();
-                  this.func_70624_b((LivingEntity)null);
+                  this.setGoalTarget((LivingEntity)null, TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
                   return ActionResultType.SUCCESS;
                }
 
@@ -347,7 +_,8 @@
                itemstack.func_190918_g(1);
             }
 
-            if (this.field_70146_Z.nextInt(3) == 0) {
+            // CraftBukkit - added event call and isCancelled check.
+            if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) {
                this.func_193101_c(p_230254_1_);
                this.field_70699_by.func_75499_g();
                this.func_70624_b((LivingEntity)null);
