--- a/net/minecraft/world/spawner/AbstractSpawner.java
+++ b/net/minecraft/world/spawner/AbstractSpawner.java
@@ -31,22 +_,22 @@
 
 public abstract class AbstractSpawner {
    private static final Logger field_209160_a = LogManager.getLogger();
-   private int field_98286_b = 20;
+   public int field_98286_b = 20;
    private final List<WeightedSpawnerEntity> field_98285_e = Lists.newArrayList();
    private WeightedSpawnerEntity field_98282_f = new WeightedSpawnerEntity();
    private double field_98287_c;
    private double field_98284_d;
-   private int field_98283_g = 200;
-   private int field_98293_h = 800;
-   private int field_98294_i = 4;
+   public int field_98283_g = 200;
+   public int field_98293_h = 800;
+   public int field_98294_i = 4;
    @Nullable
    private Entity field_98291_j;
-   private int field_98292_k = 6;
-   private int field_98289_l = 16;
-   private int field_98290_m = 4;
+   public int field_98292_k = 6;
+   public int field_98289_l = 16;
+   public int field_98290_m = 4;
 
    @Nullable
-   private ResourceLocation func_190895_g() {
+   public ResourceLocation func_190895_g() {
       String s = this.field_98282_f.func_185277_b().func_74779_i("id");
 
       try {
@@ -60,6 +_,7 @@
 
    public void func_200876_a(EntityType<?> p_200876_1_) {
       this.field_98282_f.func_185277_b().func_74778_a("id", Registry.field_212629_r.func_177774_c(p_200876_1_).toString());
+      this.field_98285_e.clear(); // CraftBukkit - SPIGOT-3496, MC-92282
    }
 
    private boolean func_98279_f() {
@@ -131,16 +_,36 @@
                      entity.func_70012_b(entity.func_226277_ct_(), entity.func_226278_cu_(), entity.func_226281_cx_(), world.field_73012_v.nextFloat() * 360.0F, 0.0F);
                      if (entity instanceof MobEntity) {
                         MobEntity mobentity = (MobEntity)entity;
-                        if (!mobentity.func_213380_a(world, SpawnReason.SPAWNER) || !mobentity.func_205019_a(world)) {
+                        if (!net.minecraftforge.event.ForgeEventFactory.canEntitySpawnSpawner(mobentity, world, (float)entity.func_226277_ct_(), (float)entity.func_226278_cu_(), (float)entity.func_226281_cx_(), this)) {
                            continue;
                         }
 
                         if (this.field_98282_f.func_185277_b().func_186856_d() == 1 && this.field_98282_f.func_185277_b().func_150297_b("id", 8)) {
+                        if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mobentity, world, (float)entity.func_226277_ct_(), (float)entity.func_226278_cu_(), (float)entity.func_226281_cx_(), this, SpawnReason.SPAWNER))
                            ((MobEntity)entity).func_213386_a(serverworld, world.func_175649_E(entity.func_233580_cy_()), SpawnReason.SPAWNER, (ILivingEntityData)null, (CompoundNBT)null);
                         }
-                     }
-
-                     if (!serverworld.func_242106_g(entity)) {
+                        // Spigot Start
+                        if ( mobentity.field_70170_p.spigotConfig.nerfSpawnerMobs )
+                        {
+                           mobentity.aware = false;
+                        }
+                        // Spigot End
+                     }
+
+                     // Spigot Start
+                     if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callSpawnerSpawnEvent(entity, blockpos).isCancelled()) {
+                        Entity vehicle = entity.func_184187_bx();
+                        if (vehicle != null) {
+                           vehicle.field_70128_L = true;
+                        }
+                        for (Entity passenger : entity.func_184182_bu()) {
+                           passenger.field_70128_L = true;
+                        }
+                        continue;
+                     }
+                     // Spigot End
+
+                     if (!serverworld.addAllEntitiesSafely(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER)) { // CraftBukkit
                         this.func_98273_j();
                         return;
                      }
@@ -282,5 +_,10 @@
    @OnlyIn(Dist.CLIENT)
    public double func_177223_e() {
       return this.field_98284_d;
+   }
+
+   @Nullable
+   public Entity getSpawnerEntity() {
+      return null;
    }
 }
