--- a/net/minecraft/world/spawner/WanderingTraderSpawner.java
+++ b/net/minecraft/world/spawner/WanderingTraderSpawner.java
@@ -96,7 +_,7 @@
                return false;
             }
 
-            WanderingTraderEntity wanderingtraderentity = EntityType.field_220351_aK.func_220342_a(p_234562_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos2, SpawnReason.EVENT, false, false);
+            WanderingTraderEntity wanderingtraderentity = EntityType.field_220351_aK.spawn(p_234562_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos2, SpawnReason.EVENT, false, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
             if (wanderingtraderentity != null) {
                for(int j = 0; j < 2; ++j) {
                   this.func_242373_a(p_234562_1_, wanderingtraderentity, 4);
@@ -117,7 +_,7 @@
    private void func_242373_a(ServerWorld p_242373_1_, WanderingTraderEntity p_242373_2_, int p_242373_3_) {
       BlockPos blockpos = this.func_234561_a_(p_242373_1_, p_242373_2_.func_233580_cy_(), p_242373_3_);
       if (blockpos != null) {
-         TraderLlamaEntity traderllamaentity = EntityType.field_220354_ax.func_220342_a(p_242373_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos, SpawnReason.EVENT, false, false);
+         TraderLlamaEntity traderllamaentity = EntityType.field_220354_ax.spawn(p_242373_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos, SpawnReason.EVENT, false, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
          if (traderllamaentity != null) {
             traderllamaentity.func_110162_b(p_242373_2_, true);
          }
