--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -137,8 +_,10 @@
       return jpanel;
    }
 
+   private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
    public void func_206931_a() {
       this.field_206932_d.start();
+      latch.countDown();
    }
 
    public void func_219050_b() {
@@ -152,7 +_,11 @@
       this.field_219051_e.forEach(Runnable::run);
    }
 
+   private static final java.util.regex.Pattern ANSI = java.util.regex.Pattern.compile("\\x1B\\[([0-9]{1,2}(;[0-9]{1,2})*)?[m|K]"); // CraftBukkit
    public void func_164247_a(JTextArea p_164247_1_, JScrollPane p_164247_2_, String p_164247_3_) {
+      try {
+         latch.await();
+      } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
       if (!SwingUtilities.isEventDispatchThread()) {
          SwingUtilities.invokeLater(() -> {
             this.func_164247_a(p_164247_1_, p_164247_2_, p_164247_3_);
@@ -166,7 +_,7 @@
          }
 
          try {
-            document.insertString(document.getLength(), p_164247_3_, (AttributeSet)null);
+            document.insertString(document.getLength(), ANSI.matcher(p_164247_3_).replaceAll(""), (AttributeSet)null); // CraftBukkit
          } catch (BadLocationException badlocationexception) {
          }
 
