--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -139,8 +_,10 @@
       return jpanel;
    }
 
+   private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
    public void m_139908_() {
       this.f_139902_.start();
+      latch.countDown();
    }
 
    public void m_139925_() {
@@ -154,7 +_,11 @@
       this.f_139903_.forEach(Runnable::run);
    }
 
+   private static final java.util.regex.Pattern ANSI = java.util.regex.Pattern.compile("\\x1B\\[([0-9]{1,2}(;[0-9]{1,2})*)?[m|K]"); // CraftBukkit
    public void m_139914_(JTextArea p_139915_, JScrollPane p_139916_, String p_139917_) {
+      try {
+         latch.await();
+      } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
       if (!SwingUtilities.isEventDispatchThread()) {
          SwingUtilities.invokeLater(() -> {
             this.m_139914_(p_139915_, p_139916_, p_139917_);
@@ -168,7 +_,7 @@
          }
 
          try {
-            document.insertString(document.getLength(), p_139917_, (AttributeSet)null);
+            document.insertString(document.getLength(), ANSI.matcher(p_139917_).replaceAll(""), (AttributeSet) null); // CraftBukkit
          } catch (BadLocationException badlocationexception) {
          }
 
