--- ../src-base/minecraft/net/minecraft/client/model/ModelParrot.java
+++ ../src-work/minecraft/net/minecraft/client/model/ModelParrot.java
@@ -21,7 +21,7 @@
     ModelRenderer feather;
     ModelRenderer legLeft;
     ModelRenderer legRight;
-    private ModelParrot.State state = ModelParrot.State.STANDING;
+    private State state = State.STANDING;
 
     public ModelParrot()
     {
@@ -89,14 +89,14 @@
         this.wingRight.rotationPointX = -1.5F;
         this.wingLeft.rotationPointX = 1.5F;
 
-        if (this.state != ModelParrot.State.FLYING)
+        if (this.state != State.FLYING)
         {
-            if (this.state == ModelParrot.State.SITTING)
+            if (this.state == State.SITTING)
             {
                 return;
             }
 
-            if (this.state == ModelParrot.State.PARTY)
+            if (this.state == State.PARTY)
             {
                 float f1 = MathHelper.cos((float)entityIn.ticksExisted);
                 float f2 = MathHelper.sin((float)entityIn.ticksExisted);
@@ -155,7 +155,7 @@
             {
                 this.legLeft.rotateAngleZ = -0.34906584F;
                 this.legRight.rotateAngleZ = 0.34906584F;
-                this.state = ModelParrot.State.PARTY;
+                this.state = State.PARTY;
                 return;
             }
 
@@ -174,17 +174,17 @@
                 ++this.legRight.rotationPointY;
                 ++this.legLeft.rotateAngleX;
                 ++this.legRight.rotateAngleX;
-                this.state = ModelParrot.State.SITTING;
+                this.state = State.SITTING;
             }
             else if (entityparrot.isFlying())
             {
                 this.legLeft.rotateAngleX += ((float)Math.PI * 2F / 9F);
                 this.legRight.rotateAngleX += ((float)Math.PI * 2F / 9F);
-                this.state = ModelParrot.State.FLYING;
+                this.state = State.FLYING;
             }
             else
             {
-                this.state = ModelParrot.State.STANDING;
+                this.state = State.STANDING;
             }
 
             this.legLeft.rotateAngleZ = 0.0F;
