--- ../src-base/minecraft/net/minecraft/client/model/ModelBiped.java
+++ ../src-work/minecraft/net/minecraft/client/model/ModelBiped.java
@@ -19,8 +19,8 @@
     public ModelRenderer bipedLeftArm;
     public ModelRenderer bipedRightLeg;
     public ModelRenderer bipedLeftLeg;
-    public ModelBiped.ArmPose leftArmPose;
-    public ModelBiped.ArmPose rightArmPose;
+    public ArmPose leftArmPose;
+    public ArmPose rightArmPose;
     public boolean isSneak;
 
     public ModelBiped()
@@ -35,8 +35,8 @@
 
     public ModelBiped(float modelSize, float p_i1149_2_, int textureWidthIn, int textureHeightIn)
     {
-        this.leftArmPose = ModelBiped.ArmPose.EMPTY;
-        this.rightArmPose = ModelBiped.ArmPose.EMPTY;
+        this.leftArmPose = ArmPose.EMPTY;
+        this.rightArmPose = ArmPose.EMPTY;
         this.textureWidth = textureWidthIn;
         this.textureHeight = textureHeightIn;
         this.bipedHead = new ModelRenderer(this, 0, 0);
@@ -249,14 +249,14 @@
         this.bipedRightArm.rotateAngleX += MathHelper.sin(ageInTicks * 0.067F) * 0.05F;
         this.bipedLeftArm.rotateAngleX -= MathHelper.sin(ageInTicks * 0.067F) * 0.05F;
 
-        if (this.rightArmPose == ModelBiped.ArmPose.BOW_AND_ARROW)
+        if (this.rightArmPose == ArmPose.BOW_AND_ARROW)
         {
             this.bipedRightArm.rotateAngleY = -0.1F + this.bipedHead.rotateAngleY;
             this.bipedLeftArm.rotateAngleY = 0.1F + this.bipedHead.rotateAngleY + 0.4F;
             this.bipedRightArm.rotateAngleX = -((float)Math.PI / 2F) + this.bipedHead.rotateAngleX;
             this.bipedLeftArm.rotateAngleX = -((float)Math.PI / 2F) + this.bipedHead.rotateAngleX;
         }
-        else if (this.leftArmPose == ModelBiped.ArmPose.BOW_AND_ARROW)
+        else if (this.leftArmPose == ArmPose.BOW_AND_ARROW)
         {
             this.bipedRightArm.rotateAngleY = -0.1F + this.bipedHead.rotateAngleY - 0.4F;
             this.bipedLeftArm.rotateAngleY = 0.1F + this.bipedHead.rotateAngleY;
