--- ../src-base/minecraft/net/minecraft/client/model/ModelBox.java
+++ ../src-work/minecraft/net/minecraft/client/model/ModelBox.java
@@ -4,7 +4,6 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public class ModelBox
 {
     private final PositionTextureVertex[] vertexPositions;
@@ -81,6 +80,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void render(BufferBuilder renderer, float scale)
     {
         for (TexturedQuad texturedquad : this.quadList)
