--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity;
 
 import javax.annotation.Nullable;
+
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.syncher.EntityDataAccessor;
@@ -12,173 +_,195 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 
 public abstract class AgeableMob extends PathfinderMob {
-   private static final EntityDataAccessor<Boolean> f_146731_ = SynchedEntityData.m_135353_(AgeableMob.class, EntityDataSerializers.f_135035_);
-   public static final int f_146730_ = -24000;
-   private static final int f_146732_ = 40;
-   protected int f_146733_;
-   protected int f_146734_;
-   protected int f_146735_;
-
-   protected AgeableMob(EntityType<? extends AgeableMob> p_146738_, Level p_146739_) {
-      super(p_146738_, p_146739_);
-   }
-
-   public SpawnGroupData m_6518_(ServerLevelAccessor p_146746_, DifficultyInstance p_146747_, MobSpawnType p_146748_, @Nullable SpawnGroupData p_146749_, @Nullable CompoundTag p_146750_) {
-      if (p_146749_ == null) {
-         p_146749_ = new AgeableMob.AgeableMobGroupData(true);
-      }
-
-      AgeableMob.AgeableMobGroupData ageablemob$ageablemobgroupdata = (AgeableMob.AgeableMobGroupData)p_146749_;
-      if (ageablemob$ageablemobgroupdata.m_146779_() && ageablemob$ageablemobgroupdata.m_146777_() > 0 && this.f_19796_.nextFloat() <= ageablemob$ageablemobgroupdata.m_146780_()) {
-         this.m_146762_(-24000);
-      }
-
-      ageablemob$ageablemobgroupdata.m_146778_();
-      return super.m_6518_(p_146746_, p_146747_, p_146748_, p_146749_, p_146750_);
-   }
-
-   @Nullable
-   public abstract AgeableMob m_142606_(ServerLevel p_146743_, AgeableMob p_146744_);
-
-   protected void m_8097_() {
-      super.m_8097_();
-      this.f_19804_.m_135372_(f_146731_, false);
-   }
-
-   public boolean m_142668_() {
-      return false;
-   }
-
-   public int m_146764_() {
-      if (this.f_19853_.f_46443_) {
-         return this.f_19804_.m_135370_(f_146731_) ? -1 : 1;
-      } else {
-         return this.f_146733_;
-      }
-   }
-
-   public void m_146740_(int p_146741_, boolean p_146742_) {
-      int i = this.m_146764_();
-      i += p_146741_ * 20;
-      if (i > 0) {
-         i = 0;
-      }
-
-      int j = i - i;
-      this.m_146762_(i);
-      if (p_146742_) {
-         this.f_146734_ += j;
-         if (this.f_146735_ == 0) {
-            this.f_146735_ = 40;
-         }
-      }
-
-      if (this.m_146764_() == 0) {
-         this.m_146762_(this.f_146734_);
-      }
-
-   }
-
-   public void m_146758_(int p_146759_) {
-      this.m_146740_(p_146759_, false);
-   }
-
-   public void m_146762_(int p_146763_) {
-      int i = this.f_146733_;
-      this.f_146733_ = p_146763_;
-      if (i < 0 && p_146763_ >= 0 || i >= 0 && p_146763_ < 0) {
-         this.f_19804_.m_135381_(f_146731_, p_146763_ < 0);
-         this.m_142669_();
-      }
-
-   }
-
-   public void m_7380_(CompoundTag p_146761_) {
-      super.m_7380_(p_146761_);
-      p_146761_.m_128405_("Age", this.m_146764_());
-      p_146761_.m_128405_("ForcedAge", this.f_146734_);
-   }
-
-   public void m_7378_(CompoundTag p_146752_) {
-      super.m_7378_(p_146752_);
-      this.m_146762_(p_146752_.m_128451_("Age"));
-      this.f_146734_ = p_146752_.m_128451_("ForcedAge");
-   }
-
-   public void m_7350_(EntityDataAccessor<?> p_146754_) {
-      if (f_146731_.equals(p_146754_)) {
-         this.m_6210_();
-      }
-
-      super.m_7350_(p_146754_);
-   }
-
-   public void m_8107_() {
-      super.m_8107_();
-      if (this.f_19853_.f_46443_) {
-         if (this.f_146735_ > 0) {
-            if (this.f_146735_ % 4 == 0) {
-               this.f_19853_.m_7106_(ParticleTypes.f_123748_, this.m_20208_(1.0D), this.m_20187_() + 0.5D, this.m_20262_(1.0D), 0.0D, 0.0D, 0.0D);
-            }
-
-            --this.f_146735_;
-         }
-      } else if (this.m_6084_()) {
-         int i = this.m_146764_();
-         if (i < 0) {
-            ++i;
-            this.m_146762_(i);
-         } else if (i > 0) {
-            --i;
-            this.m_146762_(i);
-         }
-      }
-
-   }
-
-   protected void m_142669_() {
-   }
-
-   public boolean m_6162_() {
-      return this.m_146764_() < 0;
-   }
-
-   public void m_6863_(boolean p_146756_) {
-      this.m_146762_(p_146756_ ? -24000 : 0);
-   }
-
-   public static class AgeableMobGroupData implements SpawnGroupData {
-      private int f_146767_;
-      private final boolean f_146768_;
-      private final float f_146769_;
-
-      private AgeableMobGroupData(boolean p_146775_, float p_146776_) {
-         this.f_146768_ = p_146775_;
-         this.f_146769_ = p_146776_;
-      }
-
-      public AgeableMobGroupData(boolean p_146773_) {
-         this(p_146773_, 0.05F);
-      }
-
-      public AgeableMobGroupData(float p_146771_) {
-         this(true, p_146771_);
-      }
-
-      public int m_146777_() {
-         return this.f_146767_;
-      }
-
-      public void m_146778_() {
-         ++this.f_146767_;
-      }
-
-      public boolean m_146779_() {
-         return this.f_146768_;
-      }
-
-      public float m_146780_() {
-         return this.f_146769_;
-      }
-   }
+    private static final EntityDataAccessor<Boolean> f_146731_ = SynchedEntityData.m_135353_(AgeableMob.class, EntityDataSerializers.f_135035_);
+    public static final int f_146730_ = -24000;
+    private static final int f_146732_ = 40;
+    protected int f_146733_;
+    protected int f_146734_;
+    protected int f_146735_;
+    public boolean ageLocked; // CraftBukkit
+
+    protected AgeableMob(EntityType<? extends AgeableMob> p_146738_, Level p_146739_) {
+        super(p_146738_, p_146739_);
+    }
+
+    // Spigot start
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        if (this.f_19853_.f_46443_ || this.ageLocked) { // CraftBukkit
+            this.m_6210_();
+        } else {
+            int i = this.m_146764_();
+            if (i < 0) {
+                ++i;
+                this.m_146762_(i);
+            } else if (i > 0) {
+                --i;
+                this.m_146762_(i);
+            }
+        }
+    }
+    // Spigot end
+
+    public SpawnGroupData m_6518_(ServerLevelAccessor p_146746_, DifficultyInstance p_146747_, MobSpawnType p_146748_, @Nullable SpawnGroupData p_146749_, @Nullable CompoundTag p_146750_) {
+        if (p_146749_ == null) {
+            p_146749_ = new AgeableMob.AgeableMobGroupData(true);
+        }
+
+        AgeableMob.AgeableMobGroupData ageablemob$ageablemobgroupdata = (AgeableMob.AgeableMobGroupData) p_146749_;
+        if (ageablemob$ageablemobgroupdata.m_146779_() && ageablemob$ageablemobgroupdata.m_146777_() > 0 && this.f_19796_.nextFloat() <= ageablemob$ageablemobgroupdata.m_146780_()) {
+            this.m_146762_(-24000);
+        }
+
+        ageablemob$ageablemobgroupdata.m_146778_();
+        return super.m_6518_(p_146746_, p_146747_, p_146748_, p_146749_, p_146750_);
+    }
+
+    @Nullable
+    public abstract AgeableMob m_142606_(ServerLevel p_146743_, AgeableMob p_146744_);
+
+    protected void m_8097_() {
+        super.m_8097_();
+        this.f_19804_.m_135372_(f_146731_, false);
+    }
+
+    public boolean m_142668_() {
+        return false;
+    }
+
+    public int m_146764_() {
+        if (this.f_19853_.f_46443_) {
+            return this.f_19804_.m_135370_(f_146731_) ? -1 : 1;
+        } else {
+            return this.f_146733_;
+        }
+    }
+
+    public void m_146740_(int p_146741_, boolean p_146742_) {
+        int i = this.m_146764_();
+        i += p_146741_ * 20;
+        if (i > 0) {
+            i = 0;
+        }
+
+        int j = i - i;
+        this.m_146762_(i);
+        if (p_146742_) {
+            this.f_146734_ += j;
+            if (this.f_146735_ == 0) {
+                this.f_146735_ = 40;
+            }
+        }
+
+        if (this.m_146764_() == 0) {
+            this.m_146762_(this.f_146734_);
+        }
+
+    }
+
+    public void m_146758_(int p_146759_) {
+        this.m_146740_(p_146759_, false);
+    }
+
+    public void m_146762_(int p_146763_) {
+        int i = this.f_146733_;
+        this.f_146733_ = p_146763_;
+        if (i < 0 && p_146763_ >= 0 || i >= 0 && p_146763_ < 0) {
+            this.f_19804_.m_135381_(f_146731_, p_146763_ < 0);
+            this.m_142669_();
+        }
+
+    }
+
+    public void m_7380_(CompoundTag p_146761_) {
+        super.m_7380_(p_146761_);
+        p_146761_.m_128405_("Age", this.m_146764_());
+        p_146761_.m_128405_("ForcedAge", this.f_146734_);
+        p_146761_.m_128379_("AgeLocked", this.ageLocked); // CraftBukkit
+    }
+
+    public void m_7378_(CompoundTag p_146752_) {
+        super.m_7378_(p_146752_);
+        this.m_146762_(p_146752_.m_128451_("Age"));
+        this.f_146734_ = p_146752_.m_128451_("ForcedAge");
+        this.ageLocked = p_146752_.m_128471_("AgeLocked"); // CraftBukkit
+    }
+
+    public void m_7350_(EntityDataAccessor<?> p_146754_) {
+        if (f_146731_.equals(p_146754_)) {
+            this.m_6210_();
+        }
+
+        super.m_7350_(p_146754_);
+    }
+
+    public void m_8107_() {
+        super.m_8107_();
+        if (this.f_19853_.f_46443_ || ageLocked) { // CraftBukkit
+            if (this.f_146735_ > 0) {
+                if (this.f_146735_ % 4 == 0) {
+                    this.f_19853_.m_7106_(ParticleTypes.f_123748_, this.m_20208_(1.0D), this.m_20187_() + 0.5D, this.m_20262_(1.0D), 0.0D, 0.0D, 0.0D);
+                }
+
+                --this.f_146735_;
+            }
+        } else if (this.m_6084_()) {
+            int i = this.m_146764_();
+            if (i < 0) {
+                ++i;
+                this.m_146762_(i);
+            } else if (i > 0) {
+                --i;
+                this.m_146762_(i);
+            }
+        }
+
+    }
+
+    protected void m_142669_() {
+    }
+
+    public boolean m_6162_() {
+        return this.m_146764_() < 0;
+    }
+
+    public void m_6863_(boolean p_146756_) {
+        this.m_146762_(p_146756_ ? -24000 : 0);
+    }
+
+    public static class AgeableMobGroupData implements SpawnGroupData {
+        private int f_146767_;
+        private final boolean f_146768_;
+        private final float f_146769_;
+
+        private AgeableMobGroupData(boolean p_146775_, float p_146776_) {
+            this.f_146768_ = p_146775_;
+            this.f_146769_ = p_146776_;
+        }
+
+        public AgeableMobGroupData(boolean p_146773_) {
+            this(p_146773_, 0.05F);
+        }
+
+        public AgeableMobGroupData(float p_146771_) {
+            this(true, p_146771_);
+        }
+
+        public int m_146777_() {
+            return this.f_146767_;
+        }
+
+        public void m_146778_() {
+            ++this.f_146767_;
+        }
+
+        public boolean m_146779_() {
+            return this.f_146768_;
+        }
+
+        public float m_146780_() {
+            return this.f_146769_;
+        }
+    }
 }
