--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -18,6 +_,7 @@
    protected int f_146733_;
    protected int f_146734_;
    protected int f_146735_;
+   public boolean ageLocked; // CraftBukkit
 
    protected AgeableMob(EntityType<? extends AgeableMob> p_146738_, Level p_146739_) {
       super(p_146738_, p_146739_);
@@ -97,12 +_,14 @@
       super.m_7380_(p_146761_);
       p_146761_.m_128405_("Age", this.m_146764_());
       p_146761_.m_128405_("ForcedAge", this.f_146734_);
+      p_146761_.m_128379_("AgeLocked", this.ageLocked); // CraftBukkit
    }
 
    public void m_7378_(CompoundTag p_146752_) {
       super.m_7378_(p_146752_);
       this.m_146762_(p_146752_.m_128451_("Age"));
       this.f_146734_ = p_146752_.m_128451_("ForcedAge");
+      this.ageLocked = p_146752_.m_128471_("AgeLocked"); // CraftBukkit
    }
 
    public void m_7350_(EntityDataAccessor<?> p_146754_) {
@@ -115,7 +_,7 @@
 
    public void m_8107_() {
       super.m_8107_();
-      if (this.f_19853_.f_46443_) {
+      if (this.f_19853_.f_46443_ || ageLocked) { // CraftBukkit
          if (this.f_146735_ > 0) {
             if (this.f_146735_ % 4 == 0) {
                this.f_19853_.m_7106_(ParticleTypes.f_123748_, this.m_20208_(1.0D), this.m_20187_() + 0.5D, this.m_20262_(1.0D), 0.0D, 0.0D, 0.0D);
