--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -40,12 +_,12 @@
    private static final float f_252519_ = 3.0F;
    public static final float f_252512_ = 6.0F;
    public static final float f_252504_ = 0.5F;
-   private Potion f_19701_ = Potions.f_43598_;
-   private final List<MobEffectInstance> f_19685_ = Lists.newArrayList();
+   public Potion f_19701_ = Potions.f_43598_; // CraftBukkit - 1.20.1 private -> public
+   public final List<MobEffectInstance> f_19685_ = Lists.newArrayList(); // CatServer - private -> public
    private final Map<Entity, Integer> f_19686_ = Maps.newHashMap();
    private int f_19687_ = 600;
    private int f_19688_ = 20;
-   private int f_19689_ = 20;
+   public int f_19689_ = 20; // CatServer - private -> public
    private boolean f_19690_;
    private int f_19691_;
    private float f_19692_;
@@ -99,7 +_,7 @@
 
    }
 
-   private void m_19750_() {
+   public void m_19750_() { // CraftBukkit - 1.20.1 private -> public
       if (this.f_19701_ == Potions.f_43598_ && this.f_19685_.isEmpty()) {
          this.m_20088_().m_135381_(f_19698_, 0);
       } else {
@@ -149,6 +_,17 @@
       this.f_19687_ = p_19735_;
    }
 
+   // Spigot start - copied from below
+   @Override
+   public void inactiveTick() {
+      super.inactiveTick();
+      if (this.f_19797_ >= this.f_19688_ + this.f_19687_) {
+         this.m_146870_();
+         return;
+      }
+   }
+   // Spigot end
+
    public void m_8119_() {
       super.m_8119_();
       boolean flag = this.m_19747_();
@@ -238,19 +_,31 @@
             } else {
                List<LivingEntity> list1 = this.m_9236_().m_45976_(LivingEntity.class, this.m_20191_());
                if (!list1.isEmpty()) {
+                  List<org.bukkit.entity.LivingEntity> entities = new java.util.ArrayList<org.bukkit.entity.LivingEntity>(); // CraftBukkit
                   for(LivingEntity livingentity : list1) {
                      if (!this.f_19686_.containsKey(livingentity) && livingentity.m_5801_()) {
                         double d8 = livingentity.m_20185_() - this.m_20185_();
                         double d1 = livingentity.m_20189_() - this.m_20189_();
                         double d3 = d8 * d8 + d1 * d1;
                         if (d3 <= (double)(f * f)) {
+                           // CraftBukkit start
+                           entities.add((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity());
+                        }
+                     }
+                  }
+                  org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, entities);
+                  if (!event.isCancelled()) {
+                     for (org.bukkit.entity.LivingEntity entity : event.getAffectedEntities()) {
+                        if (entity instanceof org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) {
+                           LivingEntity livingentity = ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) entity).getHandle();
+                           // CraftBukkit end
                            this.f_19686_.put(livingentity, this.f_19797_ + this.f_19689_);
 
                            for(MobEffectInstance mobeffectinstance1 : list) {
                               if (mobeffectinstance1.m_19544_().m_8093_()) {
                                  mobeffectinstance1.m_19544_().m_19461_(this, this.m_19749_(), livingentity, mobeffectinstance1.m_19564_(), 0.5D);
                               } else {
-                                 livingentity.m_147207_(new MobEffectInstance(mobeffectinstance1), this);
+                                 livingentity.addEffect(new MobEffectInstance(mobeffectinstance1), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AREA_EFFECT_CLOUD); // CraftBukkit
                               }
                            }
 
