--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -5,6 +_,8 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.logging.LogUtils;
+
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
@@ -20,6 +_,7 @@
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.world.effect.MobEffectInstance;
@@ -28,6 +_,7 @@
 import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.material.PushReaction;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity;
 import org.slf4j.Logger;
 
 public class AreaEffectCloud extends Entity {
@@ -39,11 +_,11 @@
    private static final EntityDataAccessor<ParticleOptions> f_19700_ = SynchedEntityData.m_135353_(AreaEffectCloud.class, EntityDataSerializers.f_135036_);
    private static final float f_146781_ = 32.0F;
    private Potion f_19701_ = Potions.f_43598_;
-   private final List<MobEffectInstance> f_19685_ = Lists.newArrayList();
+   public final List<MobEffectInstance> f_19685_ = Lists.newArrayList();
    private final Map<Entity, Integer> f_19686_ = Maps.newHashMap();
    private int f_19687_ = 600;
    private int f_19688_ = 20;
-   private int f_19689_ = 20;
+   public int f_19689_ = 20;
    private boolean f_19690_;
    private int f_19691_;
    private float f_19692_;
@@ -112,8 +_,21 @@
       if (!this.f_19690_) {
          this.m_19750_();
       }
+   }
 
-   }
+   // CraftBukkit start accessor methods
+   public void refreshEffects() {
+      if (!this.f_19690_) {
+         this.m_20088_().m_135381_(AreaEffectCloud.f_19698_, PotionUtils.m_43564_(PotionUtils.m_43561_(this.f_19701_, this.f_19685_)));
+      }
+   }
+   public String getPotionType() {
+      return Registry.f_122828_.m_7981_(this.f_19701_).toString();
+   }
+   public void setPotionType(String string) {
+      m_19722_(Registry.f_122828_.m_7745_(new ResourceLocation(string)));
+   }
+   // CraftBukkit end
 
    public int m_19744_() {
       return this.m_20088_().m_135370_(f_19698_);
@@ -237,19 +_,31 @@
             } else {
                List<LivingEntity> list1 = this.f_19853_.m_45976_(LivingEntity.class, this.m_142469_());
                if (!list1.isEmpty()) {
+                  List<org.bukkit.entity.LivingEntity> entities = new java.util.ArrayList<org.bukkit.entity.LivingEntity>(); // CraftBukkit
                   for(LivingEntity livingentity : list1) {
                      if (!this.f_19686_.containsKey(livingentity) && livingentity.m_5801_()) {
                         double d8 = livingentity.m_20185_() - this.m_20185_();
                         double d1 = livingentity.m_20189_() - this.m_20189_();
                         double d3 = d8 * d8 + d1 * d1;
-                        if (d3 <= (double)(f * f)) {
+                        if (d3 <= (double) (f * f)) {
+                           // CraftBukkit start
+                           entities.add((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity());
+                        }
+                     }
+                  }
+                  org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, entities);
+                  if (!event.isCancelled()) {
+                     for (org.bukkit.entity.LivingEntity entity : event.getAffectedEntities()) {
+                        if (entity instanceof CraftLivingEntity) {
+                           LivingEntity livingentity = ((CraftLivingEntity) entity).getHandle();
+                           // CraftBukkit end
                            this.f_19686_.put(livingentity, this.f_19797_ + this.f_19689_);
 
                            for(MobEffectInstance mobeffectinstance1 : list) {
                               if (mobeffectinstance1.m_19544_().m_8093_()) {
                                  mobeffectinstance1.m_19544_().m_19461_(this, this.m_19749_(), livingentity, mobeffectinstance1.m_19564_(), 0.5D);
                               } else {
-                                 livingentity.m_147207_(new MobEffectInstance(mobeffectinstance1), this);
+                                 livingentity.addEffect(new MobEffectInstance(mobeffectinstance1), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AREA_EFFECT_CLOUD); // CraftBukkit
                               }
                            }
 
