--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -110,7 +_,7 @@
 
    }
 
-   private static Transformation m_269448_(SynchedEntityData p_270278_) {
+   public static Transformation m_269448_(SynchedEntityData p_270278_) { // CatServer - private -> public
       Vector3f vector3f = p_270278_.m_135370_(f_268598_);
       Quaternionf quaternionf = p_270278_.m_135370_(f_268693_);
       Vector3f vector3f1 = p_270278_.m_135370_(f_268489_);
@@ -227,7 +_,7 @@
 
    }
 
-   private void m_269214_(Transformation p_270186_) {
+   public void m_269214_(Transformation p_270186_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268598_, p_270186_.m_252829_());
       this.f_19804_.m_135381_(f_268693_, p_270186_.m_253244_());
       this.f_19804_.m_135381_(f_268489_, p_270186_.m_252900_());
@@ -282,36 +_,36 @@
       return this.f_276646_;
    }
 
-   private void m_269317_(int p_270803_) {
+   public void m_269317_(int p_270803_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268449_, p_270803_);
    }
 
-   private int m_269272_() {
+   public int m_269272_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268449_);
    }
 
-   private void m_276345_(int p_276366_) {
+   public void m_276345_(int p_276366_) { // CatServer - private -> public
       this.f_19804_.m_276349_(f_276329_, p_276366_, true);
    }
 
-   private int m_276347_() {
+   public int m_276347_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_276329_);
    }
 
-   private void m_269423_(Display.BillboardConstraints p_270345_) {
+   public void m_269423_(Display.BillboardConstraints p_270345_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268593_, p_270345_.m_269016_());
    }
 
-   private Display.BillboardConstraints m_269218_() {
+   public Display.BillboardConstraints m_269218_() { // CatServer - private -> public
       return Display.BillboardConstraints.f_268544_.apply(this.f_19804_.m_135370_(f_268593_));
    }
 
-   private void m_269586_(@Nullable Brightness p_270461_) {
+   public void m_269586_(@Nullable Brightness p_270461_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268599_, p_270461_ != null ? p_270461_.m_269595_() : -1);
    }
 
    @Nullable
-   private Brightness m_269102_() {
+   public Brightness m_269102_() { // CatServer - private -> public
       int i = this.f_19804_.m_135370_(f_268599_);
       return i != -1 ? Brightness.m_269373_(i) : null;
    }
@@ -320,47 +_,47 @@
       return this.f_19804_.m_135370_(f_268599_);
    }
 
-   private void m_269215_(float p_270907_) {
+   public void m_269215_(float p_270907_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268421_, p_270907_);
    }
 
-   private float m_269081_() {
+   public float m_269081_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268421_);
    }
 
-   private void m_269526_(float p_270122_) {
+   public void m_269526_(float p_270122_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268527_, p_270122_);
    }
 
-   private float m_269459_() {
+   public float m_269459_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268527_);
    }
 
-   private void m_269228_(float p_270866_) {
+   public void m_269228_(float p_270866_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268708_, p_270866_);
    }
 
-   private float m_269155_() {
+   public float m_269155_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268708_);
    }
 
-   private void m_269441_(float p_270741_) {
+   public void m_269441_(float p_270741_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268532_, p_270741_);
    }
 
-   private float m_269558_() {
+   public float m_269558_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268532_);
    }
 
-   private void m_269087_(float p_270716_) {
+   public void m_269087_(float p_270716_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268639_, p_270716_);
    }
 
-   private int m_269034_() {
+   public int m_269034_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268687_);
    }
 
-   private void m_269026_(int p_270784_) {
+   public void m_269026_(int p_270784_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_268687_, p_270784_);
    }
 
@@ -377,7 +_,7 @@
       }
    }
 
-   private float m_269410_() {
+   public float m_269410_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_268639_);
    }
 
@@ -484,11 +_,11 @@
 
       }
 
-      private BlockState m_269134_() {
+      public BlockState m_269134_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268543_);
       }
 
-      private void m_269329_(BlockState p_270267_) {
+      public void m_269329_(BlockState p_270267_) { // CatServer - private -> public
          this.f_19804_.m_135381_(f_268543_, p_270267_);
       }
 
@@ -590,19 +_,19 @@
 
       }
 
-      ItemStack m_269568_() {
+      public ItemStack m_269568_() { // CatServer - public
          return this.f_19804_.m_135370_(f_268455_);
       }
 
-      void m_269362_(ItemStack p_270310_) {
+      public void m_269362_(ItemStack p_270310_) { // CatServer - public
          this.f_19804_.m_135381_(f_268455_, p_270310_);
       }
 
-      private void m_269028_(ItemDisplayContext p_270370_) {
+      public void m_269028_(ItemDisplayContext p_270370_) { // CatServer - private -> public
          this.f_19804_.m_135381_(f_268601_, p_270370_.m_269462_());
       }
 
-      private ItemDisplayContext m_269386_() {
+      public ItemDisplayContext m_269386_() { // CatServer - private -> public
          return ItemDisplayContext.f_268648_.apply(this.f_19804_.m_135370_(f_268601_));
       }
 
@@ -674,8 +_,8 @@
       private static final byte f_268541_ = -1;
       public static final int f_268744_ = 1073741824;
       private static final EntityDataAccessor<Component> f_268542_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135031_);
-      private static final EntityDataAccessor<Integer> f_268476_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135028_);
-      private static final EntityDataAccessor<Integer> f_268494_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135028_);
+      public static final EntityDataAccessor<Integer> f_268476_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135028_); // CraftBukkit - 1.20.1 private -> public
+      public static final EntityDataAccessor<Integer> f_268494_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135028_); // CraftBukkit - 1.20.1 private -> public
       private static final EntityDataAccessor<Byte> f_268481_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135027_);
       private static final EntityDataAccessor<Byte> f_268636_ = SynchedEntityData.m_135353_(Display.TextDisplay.class, EntityDataSerializers.f_135027_);
       private static final IntSet f_276462_ = IntSet.of(f_268542_.m_135015_(), f_268476_.m_135015_(), f_268494_.m_135015_(), f_268481_.m_135015_(), f_268636_.m_135015_());
@@ -705,15 +_,15 @@
 
       }
 
-      private Component m_269000_() {
+      public Component m_269000_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268542_);
       }
 
-      private void m_269037_(Component p_270902_) {
+      public void m_269037_(Component p_270902_) { // CatServer - private -> public
          this.f_19804_.m_135381_(f_268542_, p_270902_);
       }
 
-      private int m_269517_() {
+      public int m_269517_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268476_);
       }
 
@@ -721,15 +_,15 @@
          this.f_19804_.m_135381_(f_268476_, p_270545_);
       }
 
-      private byte m_269180_() {
+      public byte m_269180_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268481_);
       }
 
-      private void m_269007_(byte p_270583_) {
+      public void m_269007_(byte p_270583_) { // CatServer - private -> public
          this.f_19804_.m_135381_(f_268481_, p_270583_);
       }
 
-      private int m_269375_() {
+      public int m_269375_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268494_);
       }
 
@@ -737,11 +_,11 @@
          this.f_19804_.m_135381_(f_268494_, p_270241_);
       }
 
-      private byte m_269327_() {
+      public byte m_269327_() { // CatServer - private -> public
          return this.f_19804_.m_135370_(f_268636_);
       }
 
-      private void m_269559_(byte p_270855_) {
+      public void m_269559_(byte p_270855_) { // CatServer - private -> public
          this.f_19804_.m_135381_(f_268636_, p_270855_);
       }
 
