--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -41,7 +_,7 @@
       Team team = p_20422_.m_5647_();
       Team.CollisionRule team$collisionrule = team == null ? Team.CollisionRule.ALWAYS : team.m_7156_();
       return (Predicate<Entity>)(team$collisionrule == Team.CollisionRule.NEVER ? Predicates.alwaysFalse() : f_20408_.and((p_20430_) -> {
-         if (!p_20430_.m_6094_()) {
+         if (!p_20430_.canCollideWithBukkit(p_20422_) || !p_20422_.canCollideWithBukkit(p_20430_)) { // CraftBukkit - collidable API
             return false;
          } else if (!p_20422_.m_9236_().f_46443_ || p_20430_ instanceof Player && ((Player)p_20430_).m_7578_()) {
             Team team1 = p_20430_.m_5647_();
