--- a/net/minecraft/world/entity/GlowSquid.java
+++ b/net/minecraft/world/entity/GlowSquid.java
@@ -77,7 +_,7 @@
       return flag;
    }
 
-   private void m_147119_(int p_147120_) {
+   public void m_147119_(int p_147120_) {
       this.f_19804_.m_135381_(f_147108_, p_147120_);
    }
 
@@ -86,6 +_,6 @@
    }
 
    public static boolean m_186001_(EntityType<? extends LivingEntity> p_186002_, ServerLevelAccessor p_186003_, MobSpawnType p_186004_, BlockPos p_186005_, Random p_186006_) {
-      return p_186005_.m_123342_() <= p_186003_.m_5736_() - 33 && p_186003_.m_45524_(p_186005_, 0) == 0 && p_186003_.m_8055_(p_186005_).m_60713_(Blocks.f_49990_);
+      return p_186003_.m_8055_(p_186005_).m_60713_(Blocks.f_49990_) && p_186005_.m_123342_() <= p_186003_.m_5736_() - 33;
    }
 }
