--- a/net/minecraft/world/entity/Interaction.java
+++ b/net/minecraft/world/entity/Interaction.java
@@ -33,9 +_,9 @@
    private static final String f_271476_ = "interaction";
    private static final String f_271508_ = "response";
    @Nullable
-   private Interaction.PlayerAction f_271404_;
+   public Interaction.PlayerAction f_271404_; // CatServer - private -> public
    @Nullable
-   private Interaction.PlayerAction f_271193_;
+   public Interaction.PlayerAction f_271193_; // CatServer - private -> public
 
    public Interaction(EntityType<?> p_273319_, Level p_272713_) {
       super(p_273319_, p_272713_);
@@ -121,9 +_,16 @@
 
    public boolean m_7313_(Entity p_273553_) {
       if (p_273553_ instanceof Player player) {
+         // CraftBukkit start - 1.20.1
+         net.minecraft.world.damagesource.DamageSource source = player.m_269291_().m_269075_(player);
+         org.bukkit.event.entity.EntityDamageEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callNonLivingEntityDamageEvent(this, source, 1.0F, false);
+         if (event.isCancelled()) {
+            return true;
+         }
+         // CraftBukkit end
          this.f_271404_ = new Interaction.PlayerAction(player.m_20148_(), this.m_9236_().m_46467_());
          if (player instanceof ServerPlayer serverplayer) {
-            CriteriaTriggers.f_10573_.m_60112_(serverplayer, this, player.m_269291_().m_269264_(), 1.0F, 1.0F, false);
+            CriteriaTriggers.f_10573_.m_60112_(serverplayer, this, source, (float) event.getFinalDamage(), 1.0F, false); // CraftBukkit - 1.20.1
          }
 
          return !this.m_271819_();
@@ -154,27 +_,27 @@
       return this.f_271193_ != null ? this.m_9236_().m_46003_(this.f_271193_.f_271379_()) : null;
    }
 
-   private void m_272058_(float p_273385_) {
+   public void m_272058_(float p_273385_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_271210_, p_273385_);
    }
 
-   private float m_272023_() {
+   public float m_272023_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_271210_);
    }
 
-   private void m_271774_(float p_273733_) {
+   public void m_271774_(float p_273733_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_271237_, p_273733_);
    }
 
-   private float m_271858_() {
+   public float m_271858_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_271237_);
    }
 
-   private void m_271717_(boolean p_273657_) {
+   public void m_271717_(boolean p_273657_) { // CatServer - private -> public
       this.f_19804_.m_135381_(f_271389_, p_273657_);
    }
 
-   private boolean m_271819_() {
+   public boolean m_271819_() { // CatServer - private -> public
       return this.f_19804_.m_135370_(f_271389_);
    }
 
@@ -190,7 +_,7 @@
       return this.m_271772_().m_20393_(this.m_20182_());
    }
 
-   static record PlayerAction(UUID f_271379_, long f_271492_) {
+   public static record PlayerAction(UUID f_271379_, long f_271492_) { // CatServer - public
       public static final Codec<Interaction.PlayerAction> f_271433_ = RecordCodecBuilder.create((p_273237_) -> {
          return p_273237_.group(UUIDUtil.f_235867_.fieldOf("player").forGetter(Interaction.PlayerAction::f_271379_), Codec.LONG.fieldOf("timestamp").forGetter(Interaction.PlayerAction::f_271492_)).apply(p_273237_, Interaction.PlayerAction::new);
       });
