--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -12,8 +_,8 @@
    private final SynchedEntityData f_20837_;
    private final EntityDataAccessor<Integer> f_20838_;
    private final EntityDataAccessor<Boolean> f_20839_;
-   private boolean f_20834_;
-   private int f_20835_;
+   public boolean f_20834_; // CatServer - private -> public
+   public int f_20835_; // CatServer - private -> public
 
    public ItemBasedSteering(SynchedEntityData p_20841_, EntityDataAccessor<Integer> p_20842_, EntityDataAccessor<Boolean> p_20843_) {
       this.f_20837_ = p_20841_;
@@ -21,6 +_,14 @@
       this.f_20839_ = p_20843_;
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.f_20834_ = true;
+      this.f_20835_ = 0;
+      this.f_20837_.m_135381_(this.f_20838_, ticks);
+   }
+   // CraftBukkit end
+
    public void m_20844_() {
       this.f_20834_ = true;
       this.f_20835_ = 0;
@@ -48,7 +_,7 @@
       return this.f_20834_ ? 1.0F + 1.15F * Mth.m_14031_((float)this.f_20835_ / (float)this.m_274397_() * (float)Math.PI) : 1.0F;
    }
 
-   private int m_274397_() {
+   public int m_274397_() { // CatServer - private -> public
       return this.f_20837_.m_135370_(this.f_20838_);
    }
 
