--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -21,6 +_,15 @@
       this.f_20839_ = p_20843_;
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.f_20834_ = true;
+      this.f_20835_ = 0;
+      this.f_20836_ = ticks;
+      this.f_20837_.m_135381_(this.f_20838_, this.f_20836_);
+   }
+   // CraftBukkit end
+
    public void m_20844_() {
       this.f_20834_ = true;
       this.f_20835_ = 0;
