--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -54,4 +_,13 @@
    public boolean m_20851_() {
       return this.f_20837_.m_135370_(this.f_20839_);
    }
+
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.f_20834_ = true;
+      this.f_20835_ = 0;
+      this.f_20836_ = ticks;
+      this.f_20837_.m_135381_(this.f_20838_, this.f_20836_);
+   }
+   // CraftBukkit end
 }
