--- a/net/minecraft/world/entity/NeutralMob.java
+++ b/net/minecraft/world/entity/NeutralMob.java
@@ -108,7 +_,7 @@
    default void m_21662_() {
       this.m_6703_((LivingEntity)null);
       this.m_6925_((UUID)null);
-      this.m_6710_((LivingEntity)null);
+      this.setTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.m_7870_(0);
    }
 
@@ -120,6 +_,8 @@
    void m_6598_(@Nullable Player p_21680_);
 
    void m_6710_(@Nullable LivingEntity p_21681_);
+
+   boolean setTarget(@Nullable LivingEntity p_21681_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent); // CraftBukkit
 
    boolean m_6779_(LivingEntity p_181126_);
 
