--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -6,6 +_,7 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityUnleashEvent;
 
 public abstract class PathfinderMob extends Mob {
    protected static final float f_186010_ = 0.0F;
@@ -38,6 +_,7 @@
          float f = this.m_20270_(entity);
          if (this instanceof TamableAnimal && ((TamableAnimal)this).m_21825_()) {
             if (f > 10.0F) {
+               this.f_19853_.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                this.m_21455_(true, true);
             }
 
@@ -46,6 +_,7 @@
 
          this.m_7880_(f);
          if (f > 10.0F) {
+            this.f_19853_.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
             this.m_21455_(true, true);
             this.f_21345_.m_25355_(Goal.Flag.MOVE);
          } else if (f > 6.0F) {
