--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -38,6 +_,7 @@
          float f = this.m_20270_(entity);
          if (this instanceof TamableAnimal && ((TamableAnimal)this).m_21825_()) {
             if (f > 10.0F) {
+               this.m_9236_().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                this.m_21455_(true, true);
             }
 
@@ -46,6 +_,7 @@
 
          this.m_7880_(f);
          if (f > 10.0F) {
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
             this.m_21455_(true, true);
             this.f_21345_.m_25355_(Goal.Flag.MOVE);
          } else if (f > 6.0F) {
