--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/multipart/Selector.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/multipart/Selector.java
@@ -88,7 +88,7 @@
                 @Nullable
                 public ICondition apply(@Nullable JsonElement p_apply_1_)
                 {
-                    return p_apply_1_ == null ? null : Selector.Deserializer.getOrAndCondition(p_apply_1_.getAsJsonObject());
+                    return p_apply_1_ == null ? null : Deserializer.getOrAndCondition(p_apply_1_.getAsJsonObject());
                 }
             };
             private static final Function<Entry<String, JsonElement>, ICondition> FUNCTION_PROPERTY_VALUE = new Function<Entry<String, JsonElement>, ICondition>()
@@ -96,7 +96,7 @@
                 @Nullable
                 public ICondition apply(@Nullable Entry<String, JsonElement> p_apply_1_)
                 {
-                    return p_apply_1_ == null ? null : Selector.Deserializer.makePropertyValue(p_apply_1_);
+                    return p_apply_1_ == null ? null : Deserializer.makePropertyValue(p_apply_1_);
                 }
             };
 
