--- ../src-base/minecraft/net/minecraft/command/AdvancementCommand.java
+++ ../src-work/minecraft/net/minecraft/command/AdvancementCommand.java
@@ -38,7 +38,7 @@
         }
         else
         {
-            AdvancementCommand.ActionType advancementcommand$actiontype = AdvancementCommand.ActionType.byName(args[0]);
+            ActionType advancementcommand$actiontype = ActionType.byName(args[0]);
 
             if (advancementcommand$actiontype != null)
             {
@@ -48,7 +48,7 @@
                 }
 
                 EntityPlayerMP entityplayermp = getPlayer(server, sender, args[1]);
-                AdvancementCommand.Mode advancementcommand$mode = AdvancementCommand.Mode.byName(args[2]);
+                Mode advancementcommand$mode = Mode.byName(args[2]);
 
                 if (advancementcommand$mode == null)
                 {
@@ -81,9 +81,9 @@
         }
     }
 
-    private void perform(MinecraftServer server, ICommandSender sender, String[] args, EntityPlayerMP player, AdvancementCommand.ActionType p_193516_5_, AdvancementCommand.Mode p_193516_6_) throws CommandException
+    private void perform(MinecraftServer server, ICommandSender sender, String[] args, EntityPlayerMP player, ActionType p_193516_5_, Mode p_193516_6_) throws CommandException
     {
-        if (p_193516_6_ == AdvancementCommand.Mode.EVERYTHING)
+        if (p_193516_6_ == Mode.EVERYTHING)
         {
             if (args.length == 3)
             {
@@ -111,7 +111,7 @@
         {
             Advancement advancement = findAdvancement(server, args[3]);
 
-            if (p_193516_6_ == AdvancementCommand.Mode.ONLY && args.length == 5)
+            if (p_193516_6_ == Mode.ONLY && args.length == 5)
             {
                 String s = args[4];
 
@@ -156,7 +156,7 @@
         }
     }
 
-    private List<Advancement> getAdvancements(Advancement p_193514_1_, AdvancementCommand.Mode p_193514_2_)
+    private List<Advancement> getAdvancements(Advancement p_193514_1_, Mode p_193514_2_)
     {
         List<Advancement> list = Lists.<Advancement>newArrayList();
 
@@ -219,7 +219,7 @@
         }
         else
         {
-            AdvancementCommand.ActionType advancementcommand$actiontype = AdvancementCommand.ActionType.byName(args[0]);
+            ActionType advancementcommand$actiontype = ActionType.byName(args[0]);
 
             if (advancementcommand$actiontype != null)
             {
@@ -230,19 +230,19 @@
 
                 if (args.length == 3)
                 {
-                    return getListOfStringsMatchingLastWord(args, AdvancementCommand.Mode.NAMES);
+                    return getListOfStringsMatchingLastWord(args, Mode.NAMES);
                 }
 
-                AdvancementCommand.Mode advancementcommand$mode = AdvancementCommand.Mode.byName(args[2]);
+                Mode advancementcommand$mode = Mode.byName(args[2]);
 
-                if (advancementcommand$mode != null && advancementcommand$mode != AdvancementCommand.Mode.EVERYTHING)
+                if (advancementcommand$mode != null && advancementcommand$mode != Mode.EVERYTHING)
                 {
                     if (args.length == 4)
                     {
                         return getListOfStringsMatchingLastWord(args, this.getAdvancementNames(server));
                     }
 
-                    if (args.length == 5 && advancementcommand$mode == AdvancementCommand.Mode.ONLY)
+                    if (args.length == 5 && advancementcommand$mode == Mode.ONLY)
                     {
                         Advancement advancement = server.getAdvancementManager().getAdvancement(new ResourceLocation(args[3]));
 
@@ -377,7 +377,7 @@
         @Nullable
         static AdvancementCommand.ActionType byName(String nameIn)
         {
-            for (AdvancementCommand.ActionType advancementcommand$actiontype : values())
+            for (ActionType advancementcommand$actiontype : values())
             {
                 if (advancementcommand$actiontype.name.equals(nameIn))
                 {
@@ -433,17 +433,17 @@
             this.children = p_i47556_5_;
         }
 
-        CommandException fail(AdvancementCommand.ActionType p_193543_1_, Object... p_193543_2_)
+        CommandException fail(ActionType p_193543_1_, Object... p_193543_2_)
         {
             return new CommandException(p_193543_1_.baseTranslationKey + "." + this.name + ".failed", p_193543_2_);
         }
 
-        CommandException usage(AdvancementCommand.ActionType p_193544_1_)
+        CommandException usage(ActionType p_193544_1_)
         {
             return new CommandException(p_193544_1_.baseTranslationKey + "." + this.name + ".usage", new Object[0]);
         }
 
-        void success(ICommandSender sender, AdvancementCommand p_193546_2_, AdvancementCommand.ActionType type, Object... args)
+        void success(ICommandSender sender, AdvancementCommand p_193546_2_, ActionType type, Object... args)
         {
             CommandBase.notifyCommandListener(sender, p_193546_2_, type.baseTranslationKey + "." + this.name + ".success", args);
         }
@@ -451,7 +451,7 @@
         @Nullable
         static AdvancementCommand.Mode byName(String nameIn)
         {
-            for (AdvancementCommand.Mode advancementcommand$mode : values())
+            for (Mode advancementcommand$mode : values())
             {
                 if (advancementcommand$mode.name.equals(nameIn))
                 {
