--- ../src-base/minecraft/net/minecraft/command/CommandTP.java
+++ ../src-work/minecraft/net/minecraft/command/CommandTP.java
@@ -74,7 +74,7 @@
             {
                 Entity entity1 = getEntity(server, sender, args[args.length - 1]);
 
-                if (entity1.worldObj != entity.worldObj)
+                /*if (entity1.worldObj != entity.worldObj)
                 {
                     throw new CommandException("commands.tp.notSameDimension", new Object[0]);
                 }
@@ -92,6 +92,12 @@
                     }
 
                     notifyCommandListener(sender, this, "commands.tp.success", new Object[] {entity.getName(), entity1.getName()});
+                }*/
+                // CraftBukkit Start
+                // Use Bukkit teleport method in all cases. It has cross dimensional handling, events
+                if ((entity).getBukkitEntity().teleport(entity.getBukkitEntity(), org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) {
+                	notifyCommandListener(sender, (ICommand) this, "commands.tp.success", new Object[] { (entity).getName(), entity.getName()});
+                    // CraftBukkit End
                 }
             }
         }
