--- ../src-base/minecraft/net/minecraft/command/CommandResultStats.java
+++ ../src-work/minecraft/net/minecraft/command/CommandResultStats.java
@@ -14,7 +14,7 @@
 
 public class CommandResultStats
 {
-    private static final int NUM_RESULT_TYPES = CommandResultStats.Type.values().length;
+    private static final int NUM_RESULT_TYPES = Type.values().length;
     private static final String[] STRING_RESULT_TYPES = new String[NUM_RESULT_TYPES];
     private String[] entitiesID;
     private String[] objectives;
@@ -25,7 +25,7 @@
         this.objectives = STRING_RESULT_TYPES;
     }
 
-    public void setCommandStatForSender(MinecraftServer server, final ICommandSender sender, CommandResultStats.Type typeIn, int p_184932_4_)
+    public void setCommandStatForSender(MinecraftServer server, final ICommandSender sender, Type typeIn, int p_184932_4_)
     {
         String s = this.entitiesID[typeIn.getTypeID()];
 
@@ -69,7 +69,7 @@
                 {
                     return sender.sendCommandFeedback();
                 }
-                public void setCommandStat(CommandResultStats.Type type, int amount)
+                public void setCommandStat(Type type, int amount)
                 {
                     sender.setCommandStat(type, amount);
                 }
@@ -114,7 +114,7 @@
         {
             NBTTagCompound nbttagcompound = tagcompound.getCompoundTag("CommandStats");
 
-            for (CommandResultStats.Type commandresultstats$type : CommandResultStats.Type.values())
+            for (Type commandresultstats$type : Type.values())
             {
                 String s = commandresultstats$type.getTypeName() + "Name";
                 String s1 = commandresultstats$type.getTypeName() + "Objective";
@@ -133,7 +133,7 @@
     {
         NBTTagCompound nbttagcompound = new NBTTagCompound();
 
-        for (CommandResultStats.Type commandresultstats$type : CommandResultStats.Type.values())
+        for (Type commandresultstats$type : Type.values())
         {
             String s = this.entitiesID[commandresultstats$type.getTypeID()];
             String s1 = this.objectives[commandresultstats$type.getTypeID()];
@@ -151,7 +151,7 @@
         }
     }
 
-    public static void setScoreBoardStat(CommandResultStats stats, CommandResultStats.Type resultType, @Nullable String entityID, @Nullable String objectiveName)
+    public static void setScoreBoardStat(CommandResultStats stats, Type resultType, @Nullable String entityID, @Nullable String objectiveName)
     {
         if (entityID != null && !entityID.isEmpty() && objectiveName != null && !objectiveName.isEmpty())
         {
@@ -170,7 +170,7 @@
         }
     }
 
-    private static void removeScoreBoardStat(CommandResultStats resultStatsIn, CommandResultStats.Type resultTypeIn)
+    private static void removeScoreBoardStat(CommandResultStats resultStatsIn, Type resultTypeIn)
     {
         if (resultStatsIn.entitiesID != STRING_RESULT_TYPES && resultStatsIn.objectives != STRING_RESULT_TYPES)
         {
@@ -178,7 +178,7 @@
             resultStatsIn.objectives[resultTypeIn.getTypeID()] = null;
             boolean flag = true;
 
-            for (CommandResultStats.Type commandresultstats$type : CommandResultStats.Type.values())
+            for (Type commandresultstats$type : Type.values())
             {
                 if (resultStatsIn.entitiesID[commandresultstats$type.getTypeID()] != null && resultStatsIn.objectives[commandresultstats$type.getTypeID()] != null)
                 {
@@ -197,7 +197,7 @@
 
     public void addAllStats(CommandResultStats resultStatsIn)
     {
-        for (CommandResultStats.Type commandresultstats$type : CommandResultStats.Type.values())
+        for (Type commandresultstats$type : Type.values())
         {
             setScoreBoardStat(this, commandresultstats$type, resultStatsIn.entitiesID[commandresultstats$type.getTypeID()], resultStatsIn.objectives[commandresultstats$type.getTypeID()]);
         }
@@ -235,7 +235,7 @@
             String[] astring = new String[values().length];
             int i = 0;
 
-            for (CommandResultStats.Type commandresultstats$type : values())
+            for (Type commandresultstats$type : values())
             {
                 astring[i++] = commandresultstats$type.getTypeName();
             }
@@ -246,7 +246,7 @@
         @Nullable
         public static CommandResultStats.Type getTypeByName(String name)
         {
-            for (CommandResultStats.Type commandresultstats$type : values())
+            for (Type commandresultstats$type : values())
             {
                 if (commandresultstats$type.getTypeName().equals(name))
                 {
