--- ../src-base/minecraft/net/minecraft/command/CommandClone.java
+++ ../src-work/minecraft/net/minecraft/command/CommandClone.java
@@ -102,9 +102,9 @@
                                 }
                             }
 
-                            List<CommandClone.StaticCloneData> list = Lists.<CommandClone.StaticCloneData>newArrayList();
-                            List<CommandClone.StaticCloneData> list1 = Lists.<CommandClone.StaticCloneData>newArrayList();
-                            List<CommandClone.StaticCloneData> list2 = Lists.<CommandClone.StaticCloneData>newArrayList();
+                            List<StaticCloneData> list = Lists.<StaticCloneData>newArrayList();
+                            List<StaticCloneData> list1 = Lists.<StaticCloneData>newArrayList();
+                            List<StaticCloneData> list2 = Lists.<StaticCloneData>newArrayList();
                             Deque<BlockPos> deque = Lists.<BlockPos>newLinkedList();
                             BlockPos blockpos3 = new BlockPos(structureboundingbox1.minX - structureboundingbox.minX, structureboundingbox1.minY - structureboundingbox.minY, structureboundingbox1.minZ - structureboundingbox.minZ);
 
@@ -125,17 +125,17 @@
                                             if (tileentity != null)
                                             {
                                                 NBTTagCompound nbttagcompound = tileentity.writeToNBT(new NBTTagCompound());
-                                                list1.add(new CommandClone.StaticCloneData(blockpos5, iblockstate, nbttagcompound));
+                                                list1.add(new StaticCloneData(blockpos5, iblockstate, nbttagcompound));
                                                 deque.addLast(blockpos4);
                                             }
                                             else if (!iblockstate.isFullBlock() && !iblockstate.isFullCube())
                                             {
-                                                list2.add(new CommandClone.StaticCloneData(blockpos5, iblockstate, (NBTTagCompound)null));
+                                                list2.add(new StaticCloneData(blockpos5, iblockstate, (NBTTagCompound)null));
                                                 deque.addFirst(blockpos4);
                                             }
                                             else
                                             {
-                                                list.add(new CommandClone.StaticCloneData(blockpos5, iblockstate, (NBTTagCompound)null));
+                                                list.add(new StaticCloneData(blockpos5, iblockstate, (NBTTagCompound)null));
                                                 deque.addLast(blockpos4);
                                             }
                                         }
@@ -163,13 +163,13 @@
                                 }
                             }
 
-                            List<CommandClone.StaticCloneData> list3 = Lists.<CommandClone.StaticCloneData>newArrayList();
+                            List<StaticCloneData> list3 = Lists.<StaticCloneData>newArrayList();
                             list3.addAll(list);
                             list3.addAll(list1);
                             list3.addAll(list2);
-                            List<CommandClone.StaticCloneData> list4 = Lists.<CommandClone.StaticCloneData>reverse(list3);
+                            List<StaticCloneData> list4 = Lists.<StaticCloneData>reverse(list3);
 
-                            for (CommandClone.StaticCloneData commandclone$staticclonedata : list4)
+                            for (StaticCloneData commandclone$staticclonedata : list4)
                             {
                                 TileEntity tileentity2 = world.getTileEntity(commandclone$staticclonedata.pos);
 
@@ -183,7 +183,7 @@
 
                             i = 0;
 
-                            for (CommandClone.StaticCloneData commandclone$staticclonedata1 : list3)
+                            for (StaticCloneData commandclone$staticclonedata1 : list3)
                             {
                                 if (world.setBlockState(commandclone$staticclonedata1.pos, commandclone$staticclonedata1.blockState, 2))
                                 {
@@ -191,7 +191,7 @@
                                 }
                             }
 
-                            for (CommandClone.StaticCloneData commandclone$staticclonedata2 : list1)
+                            for (StaticCloneData commandclone$staticclonedata2 : list1)
                             {
                                 TileEntity tileentity3 = world.getTileEntity(commandclone$staticclonedata2.pos);
 
@@ -207,7 +207,7 @@
                                 world.setBlockState(commandclone$staticclonedata2.pos, commandclone$staticclonedata2.blockState, 2);
                             }
 
-                            for (CommandClone.StaticCloneData commandclone$staticclonedata3 : list4)
+                            for (StaticCloneData commandclone$staticclonedata3 : list4)
                             {
                                 world.notifyNeighborsRespectDebug(commandclone$staticclonedata3.pos, commandclone$staticclonedata3.blockState.getBlock(), false);
                             }
