--- ../src-base/minecraft/net/minecraft/command/CommandGameRule.java
+++ ../src-work/minecraft/net/minecraft/command/CommandGameRule.java
@@ -29,7 +29,7 @@
 
     public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException
     {
-        GameRules gamerules = this.getOverWorldGameRules(server);
+        GameRules gamerules = sender.getEntityWorld().getGameRules(); // CraftBukkit - Use current world
         String s = args.length > 0 ? args[0] : "";
         String s1 = args.length > 1 ? buildString(args, 1) : "";
 
@@ -73,6 +73,7 @@
                 entityplayermp.connection.sendPacket(new SPacketEntityStatus(entityplayermp, b0));
             }
         }
+        net.minecraftforge.event.ForgeEventFactory.onGameRuleChange(rules, p_184898_1_, server);
     }
 
     public List<String> getTabCompletions(MinecraftServer server, ICommandSender sender, String[] args, @Nullable BlockPos targetPos)
