--- ../src-base/minecraft/net/minecraft/command/CommandFill.java
+++ ../src-work/minecraft/net/minecraft/command/CommandFill.java
@@ -82,7 +82,7 @@
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 boolean flag = false;
 
-                if (args.length >= 10 && block.hasTileEntity())
+                if (args.length >= 10 && block.hasTileEntity(iblockstate))
                 {
                     String s = buildString(args, 9);
 
@@ -123,7 +123,7 @@
                                             continue;
                                         }
                                     }
-                                    else if ("replace".equals(args[8]) && !block.hasTileEntity() && args.length > 9)
+                                    else if ("replace".equals(args[8]) && !block.hasTileEntity(iblockstate) && args.length > 9)
                                     {
                                         Block block1 = CommandBase.getBlockByText(sender, args[9]);
 
