--- ../src-base/minecraft/net/minecraft/command/CommandGameRule.java
+++ ../src-work/minecraft/net/minecraft/command/CommandGameRule.java
@@ -29,7 +29,7 @@
 
     public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException
     {
-        GameRules gamerules = this.getOverWorldGameRules(server);
+        GameRules gamerules = sender.getEntityWorld().getGameRules();
         String s = args.length > 0 ? args[0] : "";
         String s1 = args.length > 1 ? buildString(args, 1) : "";
 
