--- ../src-base/minecraft/net/minecraft/command/CommandBase.java
+++ ../src-work/minecraft/net/minecraft/command/CommandBase.java
@@ -393,7 +393,7 @@
                 itextcomponent.appendText(" ");
             }
 
-            ITextComponent itextcomponent1 = new TextComponentString(args[i]);
+            ITextComponent itextcomponent1 = net.minecraftforge.common.ForgeHooks.newChatWithLinks(args[i]); // Forge: links for messages
 
             if (p_147176_3_)
             {
@@ -436,12 +436,12 @@
         return stringbuilder.toString();
     }
 
-    public static CommandBase.CoordinateArg parseCoordinate(double base, String selectorArg, boolean centerBlock) throws NumberInvalidException
+    public static CoordinateArg parseCoordinate(double base, String selectorArg, boolean centerBlock) throws NumberInvalidException
     {
         return parseCoordinate(base, selectorArg, -30000000, 30000000, centerBlock);
     }
 
-    public static CommandBase.CoordinateArg parseCoordinate(double base, String selectorArg, int min, int max, boolean centerBlock) throws NumberInvalidException
+    public static CoordinateArg parseCoordinate(double base, String selectorArg, int min, int max, boolean centerBlock) throws NumberInvalidException
     {
         boolean flag = selectorArg.startsWith("~");
 
@@ -485,7 +485,7 @@
                 }
             }
 
-            return new CommandBase.CoordinateArg(d1, d0, flag);
+            return new CoordinateArg(d1, d0, flag);
         }
     }
 
