--- ../src-base/minecraft/net/minecraft/command/EntitySelector.java
+++ ../src-work/minecraft/net/minecraft/command/EntitySelector.java
@@ -24,6 +24,7 @@
 import net.minecraft.scoreboard.ScoreObjective;
 import net.minecraft.scoreboard.Scoreboard;
 import net.minecraft.scoreboard.Team;
+import net.minecraft.tileentity.CommandBlockBaseLogic;
 import net.minecraft.util.EntitySelectors;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
@@ -79,6 +80,12 @@
 
     public static <T extends Entity> List<T> matchEntities(ICommandSender sender, String token, Class <? extends T > targetClass)
     {
+        //Svarka CommandBlockLogic
+        if(!(sender instanceof CommandBlockBaseLogic)){
+            return Collections.<T>emptyList();
+        }
+        //Svarka end
+
         Matcher matcher = TOKEN_PATTERN.matcher(token);
 
         if (matcher.matches() && sender.canCommandSenderUseCommand(1, "@"))
