--- ../src-base/minecraft/net/minecraft/command/EntitySelector.java
+++ ../src-work/minecraft/net/minecraft/command/EntitySelector.java
@@ -121,6 +121,11 @@
 
     public static <T extends Entity> List<T> matchEntities(ICommandSender sender, String token, Class <? extends T > targetClass) throws CommandException
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.matchEntities(sender, token, targetClass);
+    }
+
+    public static <T extends Entity> List<T> matchEntitiesDefault(ICommandSender sender, String token, Class <? extends T > targetClass) throws CommandException
+    {
         Matcher matcher = TOKEN_PATTERN.matcher(token);
 
         if (matcher.matches() && sender.canUseCommand(1, "@"))
@@ -153,6 +158,7 @@
                         list2.addAll(getTagPredicates(map));
                         list2.addAll(getRadiusPredicates(map, vec3d));
                         list2.addAll(getRotationsPredicates(map));
+                        list2.addAll(net.minecraftforge.fml.common.registry.GameRegistry.createEntitySelectors(map, s, sender, vec3d));
 
                         if ("s".equalsIgnoreCase(s))
                         {
@@ -786,6 +792,11 @@
 
     public static boolean matchesMultiplePlayers(String selectorStr) throws CommandException
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.matchesMultiplePlayers(selectorStr);
+    }
+
+    public static boolean matchesMultiplePlayersDefault(String selectorStr) throws CommandException
+    {
         Matcher matcher = TOKEN_PATTERN.matcher(selectorStr);
 
         if (!matcher.matches())
@@ -803,6 +814,11 @@
 
     public static boolean isSelector(String selectorStr)
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.isSelector(selectorStr);
+    }
+
+    public static boolean isSelectorDefault(String selectorStr)
+    {
         return TOKEN_PATTERN.matcher(selectorStr).matches();
     }
 
