--- ../src-base/minecraft/net/minecraft/command/ICommandSender.java
+++ ../src-work/minecraft/net/minecraft/command/ICommandSender.java
@@ -9,46 +9,38 @@
 import net.minecraft.util.text.TextComponentString;
 import net.minecraft.world.World;
 
-public interface ICommandSender
-{
+public interface ICommandSender {
     String getName();
 
-default ITextComponent getDisplayName()
-    {
+    default ITextComponent getDisplayName() {
         return new TextComponentString(this.getName());
     }
 
-default void sendMessage(ITextComponent component)
-    {
+    default void sendMessage(ITextComponent component) {
     }
 
     boolean canUseCommand(int permLevel, String commandName);
 
-default BlockPos getPosition()
-    {
+    default BlockPos getPosition() {
         return BlockPos.ORIGIN;
     }
 
-default Vec3d getPositionVector()
-    {
+    default Vec3d getPositionVector() {
         return Vec3d.ZERO;
     }
 
     World getEntityWorld();
 
     @Nullable
-default Entity getCommandSenderEntity()
-    {
+    default Entity getCommandSenderEntity() {
         return null;
     }
 
-default boolean sendCommandFeedback()
-    {
+    default boolean sendCommandFeedback() {
         return false;
     }
 
-default void setCommandStat(CommandResultStats.Type type, int amount)
-    {
+    default void setCommandStat(CommandResultStats.Type type, int amount) {
     }
 
     @Nullable
