--- a/net/minecraft/entity/ai/brain/sensor/SecondaryPositionSensor.java
+++ b/net/minecraft/entity/ai/brain/sensor/SecondaryPositionSensor.java
@@ -28,6 +_,9 @@
          for(int k = -2; k <= 2; ++k) {
             for(int l = -4; l <= 4; ++l) {
                BlockPos blockpos1 = blockpos.func_177982_a(j, k, l);
+               if (!p_212872_1_.isPosLoaded(blockpos1)){
+                  continue;
+               }
                if (p_212872_2_.func_213700_eh().func_221130_b().func_221150_d().contains(p_212872_1_.func_180495_p(blockpos1).func_177230_c())) {
                   list.add(GlobalPos.func_239648_a_(registrykey, blockpos1));
                }
