--- a/net/minecraft/inventory/CraftingInventory.java
+++ b/net/minecraft/inventory/CraftingInventory.java
@@ -1,16 +_,91 @@
 package net.minecraft.inventory;
 
+import java.util.List;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.inventory.container.Container;
+import net.minecraft.inventory.container.WorkbenchContainer;
 import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
 import net.minecraft.item.crafting.RecipeItemHelper;
 import net.minecraft.util.NonNullList;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.inventory.InventoryType;
 
 public class CraftingInventory implements IInventory, IRecipeHelperPopulator {
    private final NonNullList<ItemStack> field_70466_a;
    private final int field_70464_b;
    private final int field_174924_c;
    private final Container field_70465_c;
+
+   // CraftBukkit start - add fields
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+   private IRecipe currentRecipe;
+   public IInventory resultInventory;
+   private PlayerEntity owner;
+   private int maxStack = MAX_STACK;
+
+   public List<ItemStack> getContents() {
+      return this.field_70466_a;
+   }
+
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public InventoryType getInvType() {
+      return field_70466_a.size() == 4 ? InventoryType.CRAFTING : InventoryType.WORKBENCH;
+   }
+
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (owner == null) ? null : owner.getBukkitEntity();
+   }
+
+   // LoliServer start
+   public void setOwner(PlayerEntity player) {
+      this.owner = player;
+   }
+   // LoliServer end
+
+   @Override
+   public int func_70297_j_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+      resultInventory.setMaxStackSize(size);
+   }
+
+   @Override
+   public Location getLocation() {
+      return field_70465_c instanceof WorkbenchContainer ? ((WorkbenchContainer) field_70465_c).field_75160_f.getLocation() : owner.getBukkitEntity().getLocation();
+   }
+
+   @Override
+   public IRecipe getCurrentRecipe() {
+      return currentRecipe;
+   }
+
+   @Override
+   public void setCurrentRecipe(IRecipe currentRecipe) {
+      this.currentRecipe = currentRecipe;
+   }
+
+   public CraftingInventory(Container container, int i, int j, PlayerEntity player) {
+      this(container, i, j);
+      this.owner = player;
+   }
+   // CraftBukkit end
 
    public CraftingInventory(Container p_i1807_1_, int p_i1807_2_, int p_i1807_3_) {
       this.field_70466_a = NonNullList.func_191197_a(p_i1807_2_ * p_i1807_3_, ItemStack.field_190927_a);
