--- a/net/minecraft/inventory/EnderChestInventory.java
+++ b/net/minecraft/inventory/EnderChestInventory.java
@@ -5,12 +_,28 @@
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.ListNBT;
 import net.minecraft.tileentity.EnderChestTileEntity;
+import org.bukkit.Location;
+import org.bukkit.inventory.InventoryHolder;
 
 public class EnderChestInventory extends Inventory {
    private EnderChestTileEntity field_70488_a;
 
-   public EnderChestInventory() {
+   // CraftBukkit start
+   private final PlayerEntity owner;
+
+   public InventoryHolder getBukkitOwner(){
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public Location getLocation() {
+      return this.field_70488_a != null ?new Location(this.field_70488_a.func_145831_w().getWorld(), this.field_70488_a.func_174877_v().func_177958_n(), this.field_70488_a.func_174877_v().func_177956_o(), this.field_70488_a.func_174877_v().func_177952_p()) : null;
+   }
+
+   public EnderChestInventory(PlayerEntity owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void func_146031_a(EnderChestTileEntity p_146031_1_) {
