--- a/net/minecraft/inventory/IInventory.java
+++ b/net/minecraft/inventory/IInventory.java
@@ -59,4 +_,41 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   // CatServer - Add default, prevent throw AbstractMethodError
+   default java.util.List<ItemStack> getContents() {
+       return new net.minecraft.util.NonNullList<ItemStack>();
+   }
+
+   default void onOpen(org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity who) {};
+
+   default void onClose(org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity who) {};
+
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers() { return new java.util.ArrayList<org.bukkit.entity.HumanEntity>(); }
+
+   default org.bukkit.inventory.InventoryHolder getOwner() {
+      return (this instanceof net.minecraft.tileentity.TileEntity) ? catserver.server.inventory.CatInventoryUtils.getOwner((net.minecraft.tileentity.TileEntity) this) : null;
+   }
+
+   default void setMaxStackSize(int size) {};
+
+   default org.bukkit.Location getLocation() {
+      if (this instanceof net.minecraft.tileentity.TileEntity) {
+         net.minecraft.util.math.BlockPos pos = ((net.minecraft.tileentity.TileEntity) this).func_174877_v();
+         return new org.bukkit.Location(((net.minecraft.tileentity.TileEntity) this).func_145831_w().getWorld(), pos.func_177958_n(), pos.func_177956_o(), pos.func_177952_p());
+      } else {
+         return null;
+      }
+   };
+
+   default net.minecraft.item.crafting.IRecipe getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(net.minecraft.item.crafting.IRecipe recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
