--- a/net/minecraft/inventory/MerchantInventory.java
+++ b/net/minecraft/inventory/MerchantInventory.java
@@ -1,7 +_,11 @@
 package net.minecraft.inventory;
 
+import java.util.ArrayList;
+import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.entity.merchant.IMerchant;
+import net.minecraft.entity.merchant.villager.AbstractVillagerEntity;
+import net.minecraft.entity.merchant.villager.VillagerEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.MerchantOffer;
@@ -9,14 +_,66 @@
 import net.minecraft.util.NonNullList;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftAbstractVillager;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public class MerchantInventory implements IInventory {
    private final IMerchant field_70476_a;
    private final NonNullList<ItemStack> field_70474_b = NonNullList.func_191197_a(3, ItemStack.field_190927_a);
    @Nullable
    private MerchantOffer field_214026_c;
-   private int field_70473_e;
+   public int field_70473_e; // private->public CraftBukkit
    private int field_214027_e;
+
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new ArrayList<>();
+   private int maxStack = MAX_STACK;
+
+   @Override
+   public List<ItemStack> getContents() {
+      return this.field_70474_b;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+      field_70476_a.func_70932_a_((PlayerEntity) null); // SPIGOT-4860
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public InventoryHolder getOwner() {
+      return (this.field_70476_a instanceof AbstractVillagerEntity) ? ((CraftAbstractVillager) ((AbstractVillagerEntity) this.field_70476_a).getBukkitEntity()) : null;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+   }
+
+   @Override
+   public Location getLocation() {
+      return (this.field_70476_a instanceof VillagerEntity) ? ((VillagerEntity) this.field_70476_a).getBukkitEntity().getLocation() : null;
+   }
+
+   @Override
+   public int func_70297_j_() {
+      return maxStack;
+   }
+
+   // CraftBukkit end
 
    public MerchantInventory(IMerchant p_i50071_1_) {
       this.field_70476_a = p_i50071_1_;
