--- ../src-base/minecraft/net/minecraft/client/renderer/block/statemap/StateMap.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/statemap/StateMap.java
@@ -39,7 +39,7 @@
         }
         else
         {
-            s = this.removeName(this.name, map);
+            s = String.format("%s:%s", Block.REGISTRY.getNameForObject(state.getBlock()).getResourceDomain(), this.removeName(this.name, map));
         }
 
         if (this.suffix != null)
@@ -67,19 +67,19 @@
             private String suffix;
             private final List < IProperty<? >> ignored = Lists. < IProperty<? >> newArrayList();
 
-            public StateMap.Builder withName(IProperty<?> builderPropertyIn)
+            public Builder withName(IProperty<?> builderPropertyIn)
             {
                 this.name = builderPropertyIn;
                 return this;
             }
 
-            public StateMap.Builder withSuffix(String builderSuffixIn)
+            public Builder withSuffix(String builderSuffixIn)
             {
                 this.suffix = builderSuffixIn;
                 return this;
             }
 
-            public StateMap.Builder ignore(IProperty<?>... ignores)
+            public Builder ignore(IProperty<?>... ignores)
             {
                 Collections.addAll(this.ignored, ignores);
                 return this;
