--- a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -47,7 +_,7 @@
    }
 
    public void m_8056_() {
-      this.f_26025_.m_6710_(this.f_26026_);
+      this.f_26025_.setTarget(this.f_26026_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
       super.m_8056_();
    }
 }
