--- a/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
@@ -57,7 +_,7 @@
    }
 
    public void m_8056_() {
-      this.f_26135_.m_6710_(this.f_26135_.m_142581_());
+      this.f_26135_.setTarget(this.f_26135_.m_142581_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
       this.f_26137_ = this.f_26135_.m_5448_();
       this.f_26034_ = this.f_26135_.m_21213_();
       this.f_26138_ = 300;
@@ -107,6 +_,6 @@
    }
 
    protected void m_5766_(Mob p_26042_, LivingEntity p_26043_) {
-      p_26042_.m_6710_(p_26043_);
+      p_26042_.setTarget(p_26043_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit - reason
    }
 }
