--- a/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
@@ -64,7 +_,7 @@
    }
 
    public void m_8056_() {
-      this.f_26135_.m_6710_(this.f_26050_);
+      this.f_26135_.setTarget(this.f_26050_, f_26050_ instanceof ServerPlayer ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit - reason
       super.m_8056_();
    }
 
