--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
@@ -33,7 +_,7 @@
    }
 
    public void m_8056_() {
-      this.f_26135_.m_6710_(this.f_26111_);
+      this.f_26135_.setTarget(this.f_26111_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit - reason
       LivingEntity livingentity = this.f_26110_.m_269323_();
       if (livingentity != null) {
          this.f_26112_ = livingentity.m_21215_();
