--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
@@ -33,7 +_,7 @@
    }
 
    public void m_8056_() {
-      this.f_26135_.m_6710_(this.f_26104_);
+      this.f_26135_.setTarget(this.f_26104_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit - reason
       LivingEntity livingentity = this.f_26103_.m_269323_();
       if (livingentity != null) {
          this.f_26105_ = livingentity.m_21213_();
