--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.scores.Team;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public abstract class TargetGoal extends Goal {
    private static final int f_148155_ = 0;
@@ -62,7 +_,7 @@
                   }
                }
 
-               this.f_26135_.m_6710_(livingentity);
+               this.f_26135_.setTarget(livingentity, EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit
                return true;
             }
          }
@@ -80,7 +_,7 @@
    }
 
    public void m_8041_() {
-      this.f_26135_.m_6710_((LivingEntity)null);
+      this.f_26135_.setTarget((LivingEntity)null, EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.f_26137_ = null;
    }
 
