--- a/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -61,7 +_,7 @@
 
    public WorldUpgrader(LevelStorageSource.LevelStorageAccess p_249922_, DataFixer p_250273_, Registry<LevelStem> p_252191_, boolean p_250738_) {
       this.f_243889_ = p_252191_;
-      this.f_243666_ = p_252191_.m_214010_().stream().map(Registries::m_257551_).collect(Collectors.toUnmodifiableSet());
+      this.f_243666_ = (Set) java.util.stream.Stream.of(p_249922_.dimensionType).map(Registries::m_257551_).collect(Collectors.toUnmodifiableSet()); // CraftBukkit - 1.20.1
       this.f_18800_ = p_250738_;
       this.f_18803_ = p_250273_;
       this.f_18801_ = p_249922_;
@@ -127,9 +_,9 @@
                      if (compoundtag != null) {
                         int k = ChunkStorage.m_63505_(compoundtag);
                         ChunkGenerator chunkgenerator = this.f_243889_.m_123013_(Registries.m_257452_(resourcekey2)).f_63976_();
-                        CompoundTag compoundtag1 = chunkstorage.m_188288_(resourcekey2, () -> {
+                        CompoundTag compoundtag1 = chunkstorage.m_188288_(resourcekey2, () -> { // CraftBukkit // CatServer - recover
                            return this.f_18813_;
-                        }, compoundtag, chunkgenerator.m_187743_());
+                        }, compoundtag, chunkgenerator.m_187743_()); // CraftBukkit // CatServer - recover
                         ChunkPos chunkpos1 = new ChunkPos(compoundtag1.m_128451_("xPos"), compoundtag1.m_128451_("zPos"));
                         if (!chunkpos1.equals(chunkpos)) {
                            f_18797_.warn("Chunk {} has invalid position {}", chunkpos, chunkpos1);
