--- a/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -123,9 +_,9 @@
                      if (compoundtag != null) {
                         int j = ChunkStorage.m_63505_(compoundtag);
                         ChunkGenerator chunkgenerator = this.f_185925_.m_204655_().m_6246_(WorldGenSettings.m_190052_(resourcekey2)).m_63990_();
-                        CompoundTag compoundtag1 = chunkstorage.m_188288_(resourcekey2, () -> {
+                        CompoundTag compoundtag1 = chunkstorage.upgradeChunkTag(resourcekey2, () -> {
                            return this.f_18813_;
-                        }, compoundtag, chunkgenerator.m_187743_());
+                        }, compoundtag, chunkgenerator.m_187743_(), chunkpos, null);
                         ChunkPos chunkpos1 = new ChunkPos(compoundtag1.m_128451_("xPos"), compoundtag1.m_128451_("zPos"));
                         if (!chunkpos1.equals(chunkpos)) {
                            f_18797_.warn("Chunk {} has invalid position {}", chunkpos, chunkpos1);
