--- ../src-base/minecraft/net/minecraftforge/fml/client/FMLFolderResourcePack.java
+++ ../src-work/minecraft/net/minecraftforge/fml/client/FMLFolderResourcePack.java
@@ -46,9 +46,9 @@
     }
 
     @Override
-    protected boolean hasResourceName(String p_110593_1_)
+    protected boolean hasResourceName(String name)
     {
-        return super.hasResourceName(p_110593_1_);
+        return super.hasResourceName(name);
     }
     @Override
     public String getPackName()
