--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketCombatEvent.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketCombatEvent.java
@@ -11,7 +11,7 @@
 
 public class SPacketCombatEvent implements Packet<INetHandlerPlayClient>
 {
-    public SPacketCombatEvent.Event eventType;
+    public Event eventType;
     public int playerId;
     public int entityId;
     public int duration;
@@ -21,13 +21,13 @@
     {
     }
 
-    public SPacketCombatEvent(CombatTracker tracker, SPacketCombatEvent.Event eventIn)
+    public SPacketCombatEvent(CombatTracker tracker, Event eventIn)
     {
         this(tracker, eventIn, true);
     }
 
     @SuppressWarnings("incomplete-switch")
-    public SPacketCombatEvent(CombatTracker tracker, SPacketCombatEvent.Event eventIn, boolean showDeathMessage)
+    public SPacketCombatEvent(CombatTracker tracker, Event eventIn, boolean showDeathMessage)
     {
         this.eventType = eventIn;
         EntityLivingBase entitylivingbase = tracker.getBestAttacker();
@@ -55,14 +55,14 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.eventType = (SPacketCombatEvent.Event)buf.readEnumValue(SPacketCombatEvent.Event.class);
+        this.eventType = (Event)buf.readEnumValue(Event.class);
 
-        if (this.eventType == SPacketCombatEvent.Event.END_COMBAT)
+        if (this.eventType == Event.END_COMBAT)
         {
             this.duration = buf.readVarInt();
             this.entityId = buf.readInt();
         }
-        else if (this.eventType == SPacketCombatEvent.Event.ENTITY_DIED)
+        else if (this.eventType == Event.ENTITY_DIED)
         {
             this.playerId = buf.readVarInt();
             this.entityId = buf.readInt();
@@ -74,12 +74,12 @@
     {
         buf.writeEnumValue(this.eventType);
 
-        if (this.eventType == SPacketCombatEvent.Event.END_COMBAT)
+        if (this.eventType == Event.END_COMBAT)
         {
             buf.writeVarInt(this.duration);
             buf.writeInt(this.entityId);
         }
-        else if (this.eventType == SPacketCombatEvent.Event.ENTITY_DIED)
+        else if (this.eventType == Event.ENTITY_DIED)
         {
             buf.writeVarInt(this.playerId);
             buf.writeInt(this.entityId);
