--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketUpdateScore.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketUpdateScore.java
@@ -14,7 +14,7 @@
     private String name = "";
     private String objective = "";
     private int value;
-    private SPacketUpdateScore.Action action;
+    private Action action;
 
     public SPacketUpdateScore()
     {
@@ -25,7 +25,7 @@
         this.name = scoreIn.getPlayerName();
         this.objective = scoreIn.getObjective().getName();
         this.value = scoreIn.getScorePoints();
-        this.action = SPacketUpdateScore.Action.CHANGE;
+        this.action = Action.CHANGE;
     }
 
     public SPacketUpdateScore(String nameIn)
@@ -33,7 +33,7 @@
         this.name = nameIn;
         this.objective = "";
         this.value = 0;
-        this.action = SPacketUpdateScore.Action.REMOVE;
+        this.action = Action.REMOVE;
     }
 
     public SPacketUpdateScore(String nameIn, ScoreObjective objectiveIn)
@@ -41,16 +41,16 @@
         this.name = nameIn;
         this.objective = objectiveIn.getName();
         this.value = 0;
-        this.action = SPacketUpdateScore.Action.REMOVE;
+        this.action = Action.REMOVE;
     }
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
         this.name = buf.readString(40);
-        this.action = (SPacketUpdateScore.Action)buf.readEnumValue(SPacketUpdateScore.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
         this.objective = buf.readString(16);
 
-        if (this.action != SPacketUpdateScore.Action.REMOVE)
+        if (this.action != Action.REMOVE)
         {
             this.value = buf.readVarInt();
         }
@@ -62,7 +62,7 @@
         buf.writeEnumValue(this.action);
         buf.writeString(this.objective);
 
-        if (this.action != SPacketUpdateScore.Action.REMOVE)
+        if (this.action != Action.REMOVE)
         {
             buf.writeVarInt(this.value);
         }
