--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketUpdateBossInfo.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketUpdateBossInfo.java
@@ -13,7 +13,7 @@
 public class SPacketUpdateBossInfo implements Packet<INetHandlerPlayClient>
 {
     private UUID uniqueId;
-    private SPacketUpdateBossInfo.Operation operation;
+    private Operation operation;
     private ITextComponent name;
     private float percent;
     private BossInfo.Color color;
@@ -26,7 +26,7 @@
     {
     }
 
-    public SPacketUpdateBossInfo(SPacketUpdateBossInfo.Operation operationIn, BossInfo data)
+    public SPacketUpdateBossInfo(Operation operationIn, BossInfo data)
     {
         this.operation = operationIn;
         this.uniqueId = data.getUniqueId();
@@ -42,7 +42,7 @@
     public void readPacketData(PacketBuffer buf) throws IOException
     {
         this.uniqueId = buf.readUniqueId();
-        this.operation = (SPacketUpdateBossInfo.Operation)buf.readEnumValue(SPacketUpdateBossInfo.Operation.class);
+        this.operation = (Operation)buf.readEnumValue(Operation.class);
 
         switch (this.operation)
         {
