--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketRecipeBook.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketRecipeBook.java
@@ -13,7 +13,7 @@
 
 public class SPacketRecipeBook implements Packet<INetHandlerPlayClient>
 {
-    private SPacketRecipeBook.State state;
+    private State state;
     private List<IRecipe> recipes;
     private List<IRecipe> displayedRecipes;
     private boolean guiOpen;
@@ -23,7 +23,7 @@
     {
     }
 
-    public SPacketRecipeBook(SPacketRecipeBook.State stateIn, List<IRecipe> recipesIn, List<IRecipe> displayedRecipesIn, boolean isGuiOpen, boolean p_i47597_5_)
+    public SPacketRecipeBook(State stateIn, List<IRecipe> recipesIn, List<IRecipe> displayedRecipesIn, boolean isGuiOpen, boolean p_i47597_5_)
     {
         this.state = stateIn;
         this.recipes = recipesIn;
@@ -39,7 +39,7 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.state = (SPacketRecipeBook.State)buf.readEnumValue(SPacketRecipeBook.State.class);
+        this.state = (State)buf.readEnumValue(State.class);
         this.guiOpen = buf.readBoolean();
         this.filteringCraftable = buf.readBoolean();
         int i = buf.readVarInt();
@@ -50,7 +50,7 @@
             this.recipes.add(CraftingManager.getRecipeById(buf.readVarInt()));
         }
 
-        if (this.state == SPacketRecipeBook.State.INIT)
+        if (this.state == State.INIT)
         {
             i = buf.readVarInt();
             this.displayedRecipes = Lists.<IRecipe>newArrayList();
@@ -74,7 +74,7 @@
             buf.writeVarInt(CraftingManager.getIDForRecipe(irecipe));
         }
 
-        if (this.state == SPacketRecipeBook.State.INIT)
+        if (this.state == State.INIT)
         {
             buf.writeVarInt(this.displayedRecipes.size());
 
