--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketWorldBorder.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketWorldBorder.java
@@ -4,13 +4,14 @@
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
+import net.minecraft.world.WorldProviderHell;
 import net.minecraft.world.border.WorldBorder;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class SPacketWorldBorder implements Packet<INetHandlerPlayClient>
 {
-    private SPacketWorldBorder.Action action;
+    private Action action;
     private int size;
     private double centerX;
     private double centerZ;
@@ -24,11 +25,15 @@
     {
     }
 
-    public SPacketWorldBorder(WorldBorder border, SPacketWorldBorder.Action actionIn)
+    public SPacketWorldBorder(WorldBorder border, Action actionIn)
     {
         this.action = actionIn;
-        this.centerX = border.getCenterX();
-        this.centerZ = border.getCenterZ();
+        // CraftBukkit start - multiply out nether border
+        // this.centerX = border.getCenterX();
+        // this.centerZ = border.getCenterZ();
+        this.centerX = border.getCenterX() * (border.world != null && border.world.provider instanceof WorldProviderHell ? 8 : 1);
+        this.centerZ = border.getCenterZ() * (border.world != null && border.world.provider instanceof WorldProviderHell ? 8 : 1);
+        // CraftBukkit end
         this.diameter = border.getDiameter();
         this.targetSize = border.getTargetSize();
         this.timeUntilTarget = border.getTimeUntilTarget();
@@ -39,7 +44,7 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.action = (SPacketWorldBorder.Action)buf.readEnumValue(SPacketWorldBorder.Action.class);
+        this.action = (Action)buf.readEnumValue(Action.class);
 
         switch (this.action)
         {
