--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketPlayerPosLook.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketPlayerPosLook.java
@@ -16,14 +16,14 @@
     private double z;
     private float yaw;
     private float pitch;
-    private Set<SPacketPlayerPosLook.EnumFlags> flags;
+    private Set<EnumFlags> flags;
     private int teleportId;
 
     public SPacketPlayerPosLook()
     {
     }
 
-    public SPacketPlayerPosLook(double xIn, double yIn, double zIn, float yawIn, float pitchIn, Set<SPacketPlayerPosLook.EnumFlags> flagsIn, int teleportIdIn)
+    public SPacketPlayerPosLook(double xIn, double yIn, double zIn, float yawIn, float pitchIn, Set<EnumFlags> flagsIn, int teleportIdIn)
     {
         this.x = xIn;
         this.y = yIn;
@@ -41,7 +41,7 @@
         this.z = buf.readDouble();
         this.yaw = buf.readFloat();
         this.pitch = buf.readFloat();
-        this.flags = SPacketPlayerPosLook.EnumFlags.unpack(buf.readUnsignedByte());
+        this.flags = EnumFlags.unpack(buf.readUnsignedByte());
         this.teleportId = buf.readVarInt();
     }
 
@@ -52,7 +52,7 @@
         buf.writeDouble(this.z);
         buf.writeFloat(this.yaw);
         buf.writeFloat(this.pitch);
-        buf.writeByte(SPacketPlayerPosLook.EnumFlags.pack(this.flags));
+        buf.writeByte(EnumFlags.pack(this.flags));
         buf.writeVarInt(this.teleportId);
     }
 
@@ -98,7 +98,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public Set<SPacketPlayerPosLook.EnumFlags> getFlags()
+    public Set<EnumFlags> getFlags()
     {
         return this.flags;
     }
@@ -128,11 +128,11 @@
             return (flags & this.getMask()) == this.getMask();
         }
 
-        public static Set<SPacketPlayerPosLook.EnumFlags> unpack(int flags)
+        public static Set<EnumFlags> unpack(int flags)
         {
-            Set<SPacketPlayerPosLook.EnumFlags> set = EnumSet.<SPacketPlayerPosLook.EnumFlags>noneOf(SPacketPlayerPosLook.EnumFlags.class);
+            Set<EnumFlags> set = EnumSet.<EnumFlags>noneOf(EnumFlags.class);
 
-            for (SPacketPlayerPosLook.EnumFlags spacketplayerposlook$enumflags : values())
+            for (EnumFlags spacketplayerposlook$enumflags : values())
             {
                 if (spacketplayerposlook$enumflags.isSet(flags))
                 {
@@ -143,11 +143,11 @@
             return set;
         }
 
-        public static int pack(Set<SPacketPlayerPosLook.EnumFlags> flags)
+        public static int pack(Set<EnumFlags> flags)
         {
             int i = 0;
 
-            for (SPacketPlayerPosLook.EnumFlags spacketplayerposlook$enumflags : flags)
+            for (EnumFlags spacketplayerposlook$enumflags : flags)
             {
                 i |= spacketplayerposlook$enumflags.getMask();
             }
