--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketTitle.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketTitle.java
@@ -12,7 +12,7 @@
 
 public class SPacketTitle implements Packet<INetHandlerPlayClient>
 {
-    private SPacketTitle.Type type;
+    private Type type;
     private ITextComponent message;
     private int fadeInTime;
     private int displayTime;
@@ -22,17 +22,17 @@
     {
     }
 
-    public SPacketTitle(SPacketTitle.Type typeIn, ITextComponent messageIn)
+    public SPacketTitle(Type typeIn, ITextComponent messageIn)
     {
         this(typeIn, messageIn, -1, -1, -1);
     }
 
     public SPacketTitle(int fadeInTimeIn, int displayTimeIn, int fadeOutTimeIn)
     {
-        this(SPacketTitle.Type.TIMES, (ITextComponent)null, fadeInTimeIn, displayTimeIn, fadeOutTimeIn);
+        this(Type.TIMES, (ITextComponent)null, fadeInTimeIn, displayTimeIn, fadeOutTimeIn);
     }
 
-    public SPacketTitle(SPacketTitle.Type typeIn, @Nullable ITextComponent messageIn, int fadeInTimeIn, int displayTimeIn, int fadeOutTimeIn)
+    public SPacketTitle(Type typeIn, @Nullable ITextComponent messageIn, int fadeInTimeIn, int displayTimeIn, int fadeOutTimeIn)
     {
         this.type = typeIn;
         this.message = messageIn;
@@ -43,14 +43,14 @@
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
-        this.type = (SPacketTitle.Type)buf.readEnumValue(SPacketTitle.Type.class);
+        this.type = (Type)buf.readEnumValue(Type.class);
 
-        if (this.type == SPacketTitle.Type.TITLE || this.type == SPacketTitle.Type.SUBTITLE || this.type == SPacketTitle.Type.ACTIONBAR)
+        if (this.type == Type.TITLE || this.type == Type.SUBTITLE || this.type == Type.ACTIONBAR)
         {
             this.message = buf.readTextComponent();
         }
 
-        if (this.type == SPacketTitle.Type.TIMES)
+        if (this.type == Type.TIMES)
         {
             this.fadeInTime = buf.readInt();
             this.displayTime = buf.readInt();
@@ -62,12 +62,12 @@
     {
         buf.writeEnumValue(this.type);
 
-        if (this.type == SPacketTitle.Type.TITLE || this.type == SPacketTitle.Type.SUBTITLE || this.type == SPacketTitle.Type.ACTIONBAR)
+        if (this.type == Type.TITLE || this.type == Type.SUBTITLE || this.type == Type.ACTIONBAR)
         {
             buf.writeTextComponent(this.message);
         }
 
-        if (this.type == SPacketTitle.Type.TIMES)
+        if (this.type == Type.TIMES)
         {
             buf.writeInt(this.fadeInTime);
             buf.writeInt(this.displayTime);
@@ -119,9 +119,9 @@
         CLEAR,
         RESET;
 
-        public static SPacketTitle.Type byName(String name)
+        public static Type byName(String name)
         {
-            for (SPacketTitle.Type spackettitle$type : values())
+            for (Type spackettitle$type : values())
             {
                 if (spackettitle$type.name().equalsIgnoreCase(name))
                 {
@@ -137,7 +137,7 @@
             String[] astring = new String[values().length];
             int i = 0;
 
-            for (SPacketTitle.Type spackettitle$type : values())
+            for (Type spackettitle$type : values())
             {
                 astring[i++] = spackettitle$type.name().toLowerCase(Locale.ROOT);
             }
