--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketChunkData.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketChunkData.java
@@ -4,6 +4,7 @@
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map.Entry;
 import net.minecraft.nbt.NBTTagCompound;
@@ -36,8 +37,10 @@
         this.chunkZ = chunkIn.z;
         this.fullChunk = changedSectionFilter == 65535;
         boolean flag = chunkIn.getWorld().provider.hasSkyLight();
-        this.buffer = new byte[this.calculateChunkSize(chunkIn, flag, changedSectionFilter)];
-        this.availableSections = this.extractChunkData(new PacketBuffer(this.getWriteBuffer()), chunkIn, flag, changedSectionFilter);
+        this.buffer = new byte[this.calculateChunkSize(chunkIn, flag, changedSectionFilter) + 20480]; // CatServer 缓解缓冲区溢出
+        PacketBuffer buf = new PacketBuffer(this.getWriteBuffer());
+        this.availableSections = this.extractChunkData(buf, chunkIn, flag, changedSectionFilter);
+        buffer = Arrays.copyOfRange(buffer, 0, buf.writerIndex());
         this.tileEntityTags = Lists.<NBTTagCompound>newArrayList();
 
         for (Entry<BlockPos, TileEntity> entry : chunkIn.getTileEntityMap().entrySet())
