--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketSpawnMob.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketSpawnMob.java
@@ -38,7 +38,7 @@
     {
         this.entityId = entityIn.getEntityId();
         this.uniqueId = entityIn.getUniqueID();
-        this.type = EntityList.REGISTRY.getIDForObject(entityIn.getClass());
+        this.type = EntityList.getID(entityIn.getClass());
         this.x = entityIn.posX;
         this.y = entityIn.posY;
         this.z = entityIn.posZ;
