--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketEntityProperties.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketEntityProperties.java
@@ -16,7 +16,7 @@
 public class SPacketEntityProperties implements Packet<INetHandlerPlayClient>
 {
     private int entityId;
-    private final List<SPacketEntityProperties.Snapshot> snapshots = Lists.<SPacketEntityProperties.Snapshot>newArrayList();
+    private final List<Snapshot> snapshots = Lists.<Snapshot>newArrayList();
 
     public SPacketEntityProperties()
     {
@@ -28,7 +28,7 @@
 
         for (IAttributeInstance iattributeinstance : instances)
         {
-            this.snapshots.add(new SPacketEntityProperties.Snapshot(iattributeinstance.getAttribute().getName(), iattributeinstance.getBaseValue(), iattributeinstance.getModifiers()));
+            this.snapshots.add(new Snapshot(iattributeinstance.getAttribute().getName(), iattributeinstance.getBaseValue(), iattributeinstance.getModifiers()));
         }
     }
 
@@ -50,7 +50,7 @@
                 list.add(new AttributeModifier(uuid, "Unknown synced attribute modifier", buf.readDouble(), buf.readByte()));
             }
 
-            this.snapshots.add(new SPacketEntityProperties.Snapshot(s, d0, list));
+            this.snapshots.add(new Snapshot(s, d0, list));
         }
     }
 
@@ -59,7 +59,7 @@
         buf.writeVarInt(this.entityId);
         buf.writeInt(this.snapshots.size());
 
-        for (SPacketEntityProperties.Snapshot spacketentityproperties$snapshot : this.snapshots)
+        for (Snapshot spacketentityproperties$snapshot : this.snapshots)
         {
             buf.writeString(spacketentityproperties$snapshot.getName());
             buf.writeDouble(spacketentityproperties$snapshot.getBaseValue());
@@ -86,7 +86,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public List<SPacketEntityProperties.Snapshot> getSnapshots()
+    public List<Snapshot> getSnapshots()
     {
         return this.snapshots;
     }
