--- ../src-base/minecraft/net/minecraft/client/gui/GuiControls.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiControls.java
@@ -65,7 +65,7 @@
         {
             for (KeyBinding keybinding : this.mc.gameSettings.keyBindings)
             {
-                keybinding.setKeyCode(keybinding.getKeyCodeDefault());
+                keybinding.setToDefault();
             }
 
             KeyBinding.resetKeyBindingArrayAndHash();
@@ -81,6 +81,7 @@
     {
         if (this.buttonId != null)
         {
+            this.buttonId.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), -100 + mouseButton);
             this.options.setOptionKeyBinding(this.buttonId, -100 + mouseButton);
             this.buttonId = null;
             KeyBinding.resetKeyBindingArrayAndHash();
@@ -105,17 +106,21 @@
         {
             if (keyCode == 1)
             {
+                this.buttonId.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.NONE, 0);
                 this.options.setOptionKeyBinding(this.buttonId, 0);
             }
             else if (keyCode != 0)
             {
+                this.buttonId.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), keyCode);
                 this.options.setOptionKeyBinding(this.buttonId, keyCode);
             }
             else if (typedChar > 0)
             {
+                this.buttonId.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), typedChar + 256);
                 this.options.setOptionKeyBinding(this.buttonId, typedChar + 256);
             }
 
+            if (!net.minecraftforge.client.settings.KeyModifier.isKeyCodeModifier(keyCode))
             this.buttonId = null;
             this.time = Minecraft.getSystemTime();
             KeyBinding.resetKeyBindingArrayAndHash();
@@ -135,7 +140,7 @@
 
         for (KeyBinding keybinding : this.options.keyBindings)
         {
-            if (keybinding.getKeyCode() != keybinding.getKeyCodeDefault())
+            if (!keybinding.isSetToDefaultValue())
             {
                 flag = true;
                 break;
