--- ../src-base/minecraft/net/minecraft/client/gui/GuiCreateFlatWorld.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiCreateFlatWorld.java
@@ -26,7 +26,7 @@
     private String flatWorldTitle;
     private String materialText;
     private String heightText;
-    private GuiCreateFlatWorld.Details createFlatWorldListSlotGui;
+    private Details createFlatWorldListSlotGui;
     private GuiButton addLayerButton;
     private GuiButton editLayerButton;
     private GuiButton removeLayerButton;
@@ -53,7 +53,7 @@
         this.flatWorldTitle = I18n.format("createWorld.customize.flat.title");
         this.materialText = I18n.format("createWorld.customize.flat.tile");
         this.heightText = I18n.format("createWorld.customize.flat.height");
-        this.createFlatWorldListSlotGui = new GuiCreateFlatWorld.Details();
+        this.createFlatWorldListSlotGui = new Details();
         this.addLayerButton = this.addButton(new GuiButton(2, this.width / 2 - 154, this.height - 52, 100, 20, I18n.format("createWorld.customize.flat.addLayer") + " (NYI)"));
         this.editLayerButton = this.addButton(new GuiButton(3, this.width / 2 - 50, this.height - 52, 100, 20, I18n.format("createWorld.customize.flat.editLayer") + " (NYI)"));
         this.removeLayerButton = this.addButton(new GuiButton(4, this.width / 2 - 155, this.height - 52, 150, 20, I18n.format("createWorld.customize.flat.removeLayer")));
