--- ../src-base/minecraft/net/minecraft/client/gui/GuiChat.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiChat.java
@@ -46,7 +46,7 @@
         this.inputField.setFocused(true);
         this.inputField.setText(this.defaultInputFieldText);
         this.inputField.setCanLoseFocus(false);
-        this.tabCompleter = new GuiChat.ChatTabCompleter(this.inputField);
+        this.tabCompleter = new ChatTabCompleter(this.inputField);
     }
 
     public void onGuiClosed()
