--- ../src-base/minecraft/net/minecraft/client/gui/GuiCustomizeSkin.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiCustomizeSkin.java
@@ -25,7 +25,7 @@
 
         for (EnumPlayerModelParts enumplayermodelparts : EnumPlayerModelParts.values())
         {
-            this.buttonList.add(new GuiCustomizeSkin.ButtonPart(enumplayermodelparts.getPartId(), this.width / 2 - 155 + i % 2 * 160, this.height / 6 + 24 * (i >> 1), 150, 20, enumplayermodelparts));
+            this.buttonList.add(new ButtonPart(enumplayermodelparts.getPartId(), this.width / 2 - 155 + i % 2 * 160, this.height / 6 + 24 * (i >> 1), 150, 20, enumplayermodelparts));
             ++i;
         }
 
@@ -65,9 +65,9 @@
                 button.displayString = this.mc.gameSettings.getKeyBinding(GameSettings.Options.MAIN_HAND);
                 this.mc.gameSettings.sendSettingsToServer();
             }
-            else if (button instanceof GuiCustomizeSkin.ButtonPart)
+            else if (button instanceof ButtonPart)
             {
-                EnumPlayerModelParts enumplayermodelparts = ((GuiCustomizeSkin.ButtonPart)button).playerModelParts;
+                EnumPlayerModelParts enumplayermodelparts = ((ButtonPart)button).playerModelParts;
                 this.mc.gameSettings.switchModelPartEnabled(enumplayermodelparts);
                 button.displayString = this.getMessage(enumplayermodelparts);
             }
