--- ../src-base/minecraft/net/minecraft/client/gui/GuiFlatPresets.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiFlatPresets.java
@@ -28,12 +28,12 @@
 @SideOnly(Side.CLIENT)
 public class GuiFlatPresets extends GuiScreen
 {
-    private static final List<GuiFlatPresets.LayerItem> FLAT_WORLD_PRESETS = Lists.<GuiFlatPresets.LayerItem>newArrayList();
+    private static final List<LayerItem> FLAT_WORLD_PRESETS = Lists.<LayerItem>newArrayList();
     private final GuiCreateFlatWorld parentScreen;
     private String presetsTitle;
     private String presetsShare;
     private String listText;
-    private GuiFlatPresets.ListSlot list;
+    private ListSlot list;
     private GuiButton btnSelect;
     private GuiTextField export;
 
@@ -50,7 +50,7 @@
         this.presetsShare = I18n.format("createWorld.customize.presets.share");
         this.listText = I18n.format("createWorld.customize.presets.list");
         this.export = new GuiTextField(2, this.fontRenderer, 50, 40, this.width - 100, 20);
-        this.list = new GuiFlatPresets.ListSlot();
+        this.list = new ListSlot();
         this.export.setMaxStringLength(1230);
         this.export.setText(this.parentScreen.getPreset());
         this.btnSelect = this.addButton(new GuiButton(0, this.width / 2 - 155, this.height - 28, 150, 20, I18n.format("createWorld.customize.presets.select")));
@@ -145,7 +145,7 @@
             flatgeneratorinfo.getWorldFeatures().put(s, Maps.newHashMap());
         }
 
-        FLAT_WORLD_PRESETS.add(new GuiFlatPresets.LayerItem(icon, iconMetadata, name, flatgeneratorinfo.toString()));
+        FLAT_WORLD_PRESETS.add(new LayerItem(icon, iconMetadata, name, flatgeneratorinfo.toString()));
     }
 
     static
@@ -244,7 +244,7 @@
 
         protected void drawSlot(int slotIndex, int xPos, int yPos, int heightIn, int mouseXIn, int mouseYIn, float partialTicks)
         {
-            GuiFlatPresets.LayerItem guiflatpresets$layeritem = GuiFlatPresets.FLAT_WORLD_PRESETS.get(slotIndex);
+            LayerItem guiflatpresets$layeritem = GuiFlatPresets.FLAT_WORLD_PRESETS.get(slotIndex);
             this.renderIcon(xPos, yPos, guiflatpresets$layeritem.icon, guiflatpresets$layeritem.iconMetadata);
             GuiFlatPresets.this.fontRenderer.drawString(guiflatpresets$layeritem.name, xPos + 18 + 5, yPos + 6, 16777215);
         }
