--- ../src-base/minecraft/net/minecraft/client/gui/GuiLockIconButton.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiLockIconButton.java
@@ -32,34 +32,34 @@
             mc.getTextureManager().bindTexture(GuiButton.BUTTON_TEXTURES);
             GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
             boolean flag = mouseX >= this.x && mouseY >= this.y && mouseX < this.x + this.width && mouseY < this.y + this.height;
-            GuiLockIconButton.Icon guilockiconbutton$icon;
+            Icon guilockiconbutton$icon;
 
             if (this.locked)
             {
                 if (!this.enabled)
                 {
-                    guilockiconbutton$icon = GuiLockIconButton.Icon.LOCKED_DISABLED;
+                    guilockiconbutton$icon = Icon.LOCKED_DISABLED;
                 }
                 else if (flag)
                 {
-                    guilockiconbutton$icon = GuiLockIconButton.Icon.LOCKED_HOVER;
+                    guilockiconbutton$icon = Icon.LOCKED_HOVER;
                 }
                 else
                 {
-                    guilockiconbutton$icon = GuiLockIconButton.Icon.LOCKED;
+                    guilockiconbutton$icon = Icon.LOCKED;
                 }
             }
             else if (!this.enabled)
             {
-                guilockiconbutton$icon = GuiLockIconButton.Icon.UNLOCKED_DISABLED;
+                guilockiconbutton$icon = Icon.UNLOCKED_DISABLED;
             }
             else if (flag)
             {
-                guilockiconbutton$icon = GuiLockIconButton.Icon.UNLOCKED_HOVER;
+                guilockiconbutton$icon = Icon.UNLOCKED_HOVER;
             }
             else
             {
-                guilockiconbutton$icon = GuiLockIconButton.Icon.UNLOCKED;
+                guilockiconbutton$icon = Icon.UNLOCKED;
             }
 
             this.drawTexturedModalRect(this.x, this.y, guilockiconbutton$icon.getX(), guilockiconbutton$icon.getY(), this.width, this.height);
