--- ../src-base/minecraft/net/minecraft/client/gui/GuiLanguage.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiLanguage.java
@@ -16,7 +16,7 @@
 public class GuiLanguage extends GuiScreen
 {
     protected GuiScreen parentScreen;
-    private GuiLanguage.List list;
+    private List list;
     private final GameSettings game_settings_3;
     private final LanguageManager languageManager;
     private GuiOptionButton forceUnicodeFontBtn;
@@ -33,7 +33,7 @@
     {
         this.forceUnicodeFontBtn = (GuiOptionButton)this.addButton(new GuiOptionButton(100, this.width / 2 - 155, this.height - 38, GameSettings.Options.FORCE_UNICODE_FONT, this.game_settings_3.getKeyBinding(GameSettings.Options.FORCE_UNICODE_FONT)));
         this.confirmSettingsBtn = (GuiOptionButton)this.addButton(new GuiOptionButton(6, this.width / 2 - 155 + 160, this.height - 38, I18n.format("gui.done")));
-        this.list = new GuiLanguage.List(this.mc);
+        this.list = new List(this.mc);
         this.list.registerScrollButtons(7, 8);
     }
 
