--- ../src-base/minecraft/net/minecraft/client/gui/GuiMerchant.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiMerchant.java
@@ -29,8 +29,8 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final ResourceLocation MERCHANT_GUI_TEXTURE = new ResourceLocation("textures/gui/container/villager.png");
     private final IMerchant merchant;
-    private GuiMerchant.MerchantButton nextButton;
-    private GuiMerchant.MerchantButton previousButton;
+    private MerchantButton nextButton;
+    private MerchantButton previousButton;
     private int selectedMerchantRecipe;
     private final ITextComponent chatComponent;
 
@@ -46,8 +46,8 @@
         super.initGui();
         int i = (this.width - this.xSize) / 2;
         int j = (this.height - this.ySize) / 2;
-        this.nextButton = (GuiMerchant.MerchantButton)this.addButton(new GuiMerchant.MerchantButton(1, i + 120 + 27, j + 24 - 1, true));
-        this.previousButton = (GuiMerchant.MerchantButton)this.addButton(new GuiMerchant.MerchantButton(2, i + 36 - 19, j + 24 - 1, false));
+        this.nextButton = (MerchantButton)this.addButton(new MerchantButton(1, i + 120 + 27, j + 24 - 1, true));
+        this.previousButton = (MerchantButton)this.addButton(new MerchantButton(2, i + 36 - 19, j + 24 - 1, false));
         this.nextButton.enabled = false;
         this.previousButton.enabled = false;
     }
