--- ../src-base/minecraft/net/minecraft/client/gui/GuiOptionsRowList.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiOptionsRowList.java
@@ -10,7 +10,7 @@
 @SideOnly(Side.CLIENT)
 public class GuiOptionsRowList extends GuiListExtended
 {
-    private final List<GuiOptionsRowList.Row> options = Lists.<GuiOptionsRowList.Row>newArrayList();
+    private final List<Row> options = Lists.<Row>newArrayList();
 
     public GuiOptionsRowList(Minecraft mcIn, int p_i45015_2_, int p_i45015_3_, int p_i45015_4_, int p_i45015_5_, int p_i45015_6_, GameSettings.Options... p_i45015_7_)
     {
@@ -23,7 +23,7 @@
             GameSettings.Options gamesettings$options1 = i < p_i45015_7_.length - 1 ? p_i45015_7_[i + 1] : null;
             GuiButton guibutton = this.createButton(mcIn, p_i45015_2_ / 2 - 155, 0, gamesettings$options);
             GuiButton guibutton1 = this.createButton(mcIn, p_i45015_2_ / 2 - 155 + 160, 0, gamesettings$options1);
-            this.options.add(new GuiOptionsRowList.Row(guibutton, guibutton1));
+            this.options.add(new Row(guibutton, guibutton1));
         }
     }
 
@@ -40,7 +40,7 @@
         }
     }
 
-    public GuiOptionsRowList.Row getListEntry(int index)
+    public Row getListEntry(int index)
     {
         return this.options.get(index);
     }
@@ -61,7 +61,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public static class Row implements GuiListExtended.IGuiListEntry
+    public static class Row implements IGuiListEntry
         {
             private final Minecraft client = Minecraft.getMinecraft();
             private final GuiButton buttonA;
