--- ../src-base/minecraft/net/minecraft/client/gui/GuiPageButtonList.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiPageButtonList.java
@@ -14,15 +14,15 @@
 @SideOnly(Side.CLIENT)
 public class GuiPageButtonList extends GuiListExtended
 {
-    private final List<GuiPageButtonList.GuiEntry> entries = Lists.<GuiPageButtonList.GuiEntry>newArrayList();
+    private final List<GuiEntry> entries = Lists.<GuiEntry>newArrayList();
     private final IntHashMap<Gui> componentMap = new IntHashMap<Gui>();
     private final List<GuiTextField> editBoxes = Lists.<GuiTextField>newArrayList();
-    private final GuiPageButtonList.GuiListEntry[][] pages;
+    private final GuiListEntry[][] pages;
     private int page;
-    private final GuiPageButtonList.GuiResponder responder;
+    private final GuiResponder responder;
     private Gui focusedControl;
 
-    public GuiPageButtonList(Minecraft mcIn, int widthIn, int heightIn, int topIn, int bottomIn, int slotHeightIn, GuiPageButtonList.GuiResponder p_i45536_7_, GuiPageButtonList.GuiListEntry[]... p_i45536_8_)
+    public GuiPageButtonList(Minecraft mcIn, int widthIn, int heightIn, int topIn, int bottomIn, int slotHeightIn, GuiResponder p_i45536_7_, GuiListEntry[]... p_i45536_8_)
     {
         super(mcIn, widthIn, heightIn, topIn, bottomIn, slotHeightIn);
         this.responder = p_i45536_7_;
@@ -34,15 +34,15 @@
 
     private void populateComponents()
     {
-        for (GuiPageButtonList.GuiListEntry[] aguipagebuttonlist$guilistentry : this.pages)
+        for (GuiListEntry[] aguipagebuttonlist$guilistentry : this.pages)
         {
             for (int i = 0; i < aguipagebuttonlist$guilistentry.length; i += 2)
             {
-                GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry = aguipagebuttonlist$guilistentry[i];
-                GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry1 = i < aguipagebuttonlist$guilistentry.length - 1 ? aguipagebuttonlist$guilistentry[i + 1] : null;
+                GuiListEntry guipagebuttonlist$guilistentry = aguipagebuttonlist$guilistentry[i];
+                GuiListEntry guipagebuttonlist$guilistentry1 = i < aguipagebuttonlist$guilistentry.length - 1 ? aguipagebuttonlist$guilistentry[i + 1] : null;
                 Gui gui = this.createEntry(guipagebuttonlist$guilistentry, 0, guipagebuttonlist$guilistentry1 == null);
                 Gui gui1 = this.createEntry(guipagebuttonlist$guilistentry1, 160, guipagebuttonlist$guilistentry == null);
-                GuiPageButtonList.GuiEntry guipagebuttonlist$guientry = new GuiPageButtonList.GuiEntry(gui, gui1);
+                GuiEntry guipagebuttonlist$guientry = new GuiEntry(gui, gui1);
                 this.entries.add(guipagebuttonlist$guientry);
 
                 if (guipagebuttonlist$guilistentry != null && gui != null)
@@ -74,11 +74,11 @@
 
         for (int i = 0; i < this.pages[this.page].length; i += 2)
         {
-            GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry = this.pages[this.page][i];
-            GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry1 = i < this.pages[this.page].length - 1 ? this.pages[this.page][i + 1] : null;
+            GuiListEntry guipagebuttonlist$guilistentry = this.pages[this.page][i];
+            GuiListEntry guipagebuttonlist$guilistentry1 = i < this.pages[this.page].length - 1 ? this.pages[this.page][i + 1] : null;
             Gui gui = this.componentMap.lookup(guipagebuttonlist$guilistentry.getId());
             Gui gui1 = guipagebuttonlist$guilistentry1 != null ? (Gui)this.componentMap.lookup(guipagebuttonlist$guilistentry1.getId()) : null;
-            GuiPageButtonList.GuiEntry guipagebuttonlist$guientry = new GuiPageButtonList.GuiEntry(gui, gui1);
+            GuiEntry guipagebuttonlist$guientry = new GuiEntry(gui, gui1);
             this.entries.add(guipagebuttonlist$guientry);
         }
     }
@@ -133,7 +133,7 @@
 
     private void markVisibility(int p_178060_1_, int p_178060_2_)
     {
-        for (GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry : this.pages[p_178060_1_])
+        for (GuiListEntry guipagebuttonlist$guilistentry : this.pages[p_178060_1_])
         {
             if (guipagebuttonlist$guilistentry != null)
             {
@@ -141,7 +141,7 @@
             }
         }
 
-        for (GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry1 : this.pages[p_178060_2_])
+        for (GuiListEntry guipagebuttonlist$guilistentry1 : this.pages[p_178060_2_])
         {
             if (guipagebuttonlist$guilistentry1 != null)
             {
@@ -169,27 +169,27 @@
     @Nullable
     private Gui createEntry(@Nullable GuiPageButtonList.GuiListEntry p_178058_1_, int p_178058_2_, boolean p_178058_3_)
     {
-        if (p_178058_1_ instanceof GuiPageButtonList.GuiSlideEntry)
+        if (p_178058_1_ instanceof GuiSlideEntry)
         {
-            return this.createSlider(this.width / 2 - 155 + p_178058_2_, 0, (GuiPageButtonList.GuiSlideEntry)p_178058_1_);
+            return this.createSlider(this.width / 2 - 155 + p_178058_2_, 0, (GuiSlideEntry)p_178058_1_);
         }
-        else if (p_178058_1_ instanceof GuiPageButtonList.GuiButtonEntry)
+        else if (p_178058_1_ instanceof GuiButtonEntry)
         {
-            return this.createButton(this.width / 2 - 155 + p_178058_2_, 0, (GuiPageButtonList.GuiButtonEntry)p_178058_1_);
+            return this.createButton(this.width / 2 - 155 + p_178058_2_, 0, (GuiButtonEntry)p_178058_1_);
         }
-        else if (p_178058_1_ instanceof GuiPageButtonList.EditBoxEntry)
+        else if (p_178058_1_ instanceof EditBoxEntry)
         {
-            return this.createTextField(this.width / 2 - 155 + p_178058_2_, 0, (GuiPageButtonList.EditBoxEntry)p_178058_1_);
+            return this.createTextField(this.width / 2 - 155 + p_178058_2_, 0, (EditBoxEntry)p_178058_1_);
         }
         else
         {
-            return p_178058_1_ instanceof GuiPageButtonList.GuiLabelEntry ? this.createLabel(this.width / 2 - 155 + p_178058_2_, 0, (GuiPageButtonList.GuiLabelEntry)p_178058_1_, p_178058_3_) : null;
+            return p_178058_1_ instanceof GuiLabelEntry ? this.createLabel(this.width / 2 - 155 + p_178058_2_, 0, (GuiLabelEntry)p_178058_1_, p_178058_3_) : null;
         }
     }
 
     public void setActive(boolean p_181155_1_)
     {
-        for (GuiPageButtonList.GuiEntry guipagebuttonlist$guientry : this.entries)
+        for (GuiEntry guipagebuttonlist$guientry : this.entries)
         {
             if (guipagebuttonlist$guientry.component1 instanceof GuiButton)
             {
@@ -210,7 +210,7 @@
 
         if (i >= 0)
         {
-            GuiPageButtonList.GuiEntry guipagebuttonlist$guientry = this.getListEntry(i);
+            GuiEntry guipagebuttonlist$guientry = this.getListEntry(i);
 
             if (this.focusedControl != guipagebuttonlist$guientry.focusedControl && this.focusedControl != null && this.focusedControl instanceof GuiTextField)
             {
@@ -223,21 +223,21 @@
         return flag;
     }
 
-    private GuiSlider createSlider(int p_178067_1_, int p_178067_2_, GuiPageButtonList.GuiSlideEntry p_178067_3_)
+    private GuiSlider createSlider(int p_178067_1_, int p_178067_2_, GuiSlideEntry p_178067_3_)
     {
         GuiSlider guislider = new GuiSlider(this.responder, p_178067_3_.getId(), p_178067_1_, p_178067_2_, p_178067_3_.getCaption(), p_178067_3_.getMinValue(), p_178067_3_.getMaxValue(), p_178067_3_.getInitalValue(), p_178067_3_.getFormatter());
         guislider.visible = p_178067_3_.shouldStartVisible();
         return guislider;
     }
 
-    private GuiListButton createButton(int p_178065_1_, int p_178065_2_, GuiPageButtonList.GuiButtonEntry p_178065_3_)
+    private GuiListButton createButton(int p_178065_1_, int p_178065_2_, GuiButtonEntry p_178065_3_)
     {
         GuiListButton guilistbutton = new GuiListButton(this.responder, p_178065_3_.getId(), p_178065_1_, p_178065_2_, p_178065_3_.getCaption(), p_178065_3_.getInitialValue());
         guilistbutton.visible = p_178065_3_.shouldStartVisible();
         return guilistbutton;
     }
 
-    private GuiTextField createTextField(int p_178068_1_, int p_178068_2_, GuiPageButtonList.EditBoxEntry p_178068_3_)
+    private GuiTextField createTextField(int p_178068_1_, int p_178068_2_, EditBoxEntry p_178068_3_)
     {
         GuiTextField guitextfield = new GuiTextField(p_178068_3_.getId(), this.mc.fontRenderer, p_178068_1_, p_178068_2_, 150, 20);
         guitextfield.setText(p_178068_3_.getCaption());
@@ -247,7 +247,7 @@
         return guitextfield;
     }
 
-    private GuiLabel createLabel(int p_178063_1_, int p_178063_2_, GuiPageButtonList.GuiLabelEntry p_178063_3_, boolean p_178063_4_)
+    private GuiLabel createLabel(int p_178063_1_, int p_178063_2_, GuiLabelEntry p_178063_3_, boolean p_178063_4_)
     {
         GuiLabel guilabel;
 
@@ -350,7 +350,7 @@
         }
     }
 
-    public GuiPageButtonList.GuiEntry getListEntry(int index)
+    public GuiEntry getListEntry(int index)
     {
         return this.entries.get(index);
     }
@@ -371,7 +371,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public static class EditBoxEntry extends GuiPageButtonList.GuiListEntry
+    public static class EditBoxEntry extends GuiListEntry
         {
             private final Predicate<String> filter;
 
@@ -388,7 +388,7 @@
         }
 
     @SideOnly(Side.CLIENT)
-    public static class GuiButtonEntry extends GuiPageButtonList.GuiListEntry
+    public static class GuiButtonEntry extends GuiListEntry
         {
             private final boolean initialValue;
 
@@ -405,7 +405,7 @@
         }
 
     @SideOnly(Side.CLIENT)
-    public static class GuiEntry implements GuiListExtended.IGuiListEntry
+    public static class GuiEntry implements IGuiListEntry
         {
             private final Minecraft client = Minecraft.getMinecraft();
             private final Gui component1;
@@ -563,7 +563,7 @@
         }
 
     @SideOnly(Side.CLIENT)
-    public static class GuiLabelEntry extends GuiPageButtonList.GuiListEntry
+    public static class GuiLabelEntry extends GuiListEntry
         {
             public GuiLabelEntry(int p_i45532_1_, String p_i45532_2_, boolean p_i45532_3_)
             {
@@ -612,7 +612,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public static class GuiSlideEntry extends GuiPageButtonList.GuiListEntry
+    public static class GuiSlideEntry extends GuiListEntry
         {
             private final GuiSlider.FormatHelper formatter;
             private final float minValue;
