--- ../src-base/minecraft/net/minecraft/client/gui/GuiScreenBook.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiScreenBook.java
@@ -48,8 +48,8 @@
     private String bookTitle = "";
     private List<ITextComponent> cachedComponents;
     private int cachedPage = -1;
-    private GuiScreenBook.NextPageButton buttonNextPage;
-    private GuiScreenBook.NextPageButton buttonPreviousPage;
+    private NextPageButton buttonNextPage;
+    private NextPageButton buttonPreviousPage;
     private GuiButton buttonDone;
     private GuiButton buttonSign;
     private GuiButton buttonFinalize;
@@ -69,6 +69,7 @@
 
             if (this.bookTotalPages < 1)
             {
+                this.bookPages.appendTag(new NBTTagString("")); // Forge: fix MC-1685
                 this.bookTotalPages = 1;
             }
         }
@@ -106,8 +107,8 @@
 
         int i = (this.width - 192) / 2;
         int j = 2;
-        this.buttonNextPage = (GuiScreenBook.NextPageButton)this.addButton(new GuiScreenBook.NextPageButton(1, i + 120, 156, true));
-        this.buttonPreviousPage = (GuiScreenBook.NextPageButton)this.addButton(new GuiScreenBook.NextPageButton(2, i + 38, 156, false));
+        this.buttonNextPage = (NextPageButton)this.addButton(new NextPageButton(1, i + 120, 156, true));
+        this.buttonPreviousPage = (NextPageButton)this.addButton(new NextPageButton(2, i + 38, 156, false));
         this.updateButtons();
     }
 
