--- ../src-base/minecraft/net/minecraft/client/gui/GuiSubtitleOverlay.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiSubtitleOverlay.java
@@ -17,7 +17,7 @@
 public class GuiSubtitleOverlay extends Gui implements ISoundEventListener
 {
     private final Minecraft client;
-    private final List<GuiSubtitleOverlay.Subtitle> subtitles = Lists.<GuiSubtitleOverlay.Subtitle>newArrayList();
+    private final List<Subtitle> subtitles = Lists.<Subtitle>newArrayList();
     private boolean enabled;
 
     public GuiSubtitleOverlay(Minecraft clientIn)
@@ -49,11 +49,11 @@
             Vec3d vec3d3 = vec3d1.crossProduct(vec3d2);
             int i = 0;
             int j = 0;
-            Iterator<GuiSubtitleOverlay.Subtitle> iterator = this.subtitles.iterator();
+            Iterator<Subtitle> iterator = this.subtitles.iterator();
 
             while (iterator.hasNext())
             {
-                GuiSubtitleOverlay.Subtitle guisubtitleoverlay$subtitle = iterator.next();
+                Subtitle guisubtitleoverlay$subtitle = iterator.next();
 
                 if (guisubtitleoverlay$subtitle.getStartTime() + 3000L <= Minecraft.getSystemTime())
                 {
@@ -67,7 +67,7 @@
 
             j = j + this.client.fontRenderer.getStringWidth("<") + this.client.fontRenderer.getStringWidth(" ") + this.client.fontRenderer.getStringWidth(">") + this.client.fontRenderer.getStringWidth(" ");
 
-            for (GuiSubtitleOverlay.Subtitle guisubtitleoverlay$subtitle1 : this.subtitles)
+            for (Subtitle guisubtitleoverlay$subtitle1 : this.subtitles)
             {
                 int k = 255;
                 String s = guisubtitleoverlay$subtitle1.getString();
@@ -118,7 +118,7 @@
 
             if (!this.subtitles.isEmpty())
             {
-                for (GuiSubtitleOverlay.Subtitle guisubtitleoverlay$subtitle : this.subtitles)
+                for (Subtitle guisubtitleoverlay$subtitle : this.subtitles)
                 {
                     if (guisubtitleoverlay$subtitle.getString().equals(s))
                     {
@@ -128,7 +128,7 @@
                 }
             }
 
-            this.subtitles.add(new GuiSubtitleOverlay.Subtitle(s, new Vec3d((double)soundIn.getXPosF(), (double)soundIn.getYPosF(), (double)soundIn.getZPosF())));
+            this.subtitles.add(new Subtitle(s, new Vec3d((double)soundIn.getXPosF(), (double)soundIn.getYPosF(), (double)soundIn.getZPosF())));
         }
     }
 
