--- ../src-base/minecraft/net/minecraft/client/gui/GuiSnooper.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiSnooper.java
@@ -18,7 +18,7 @@
     private final java.util.List<String> values = Lists.<String>newArrayList();
     private String title;
     private String[] desc;
-    private GuiSnooper.List list;
+    private List list;
     private GuiButton toggleButton;
 
     public GuiSnooper(GuiScreen p_i1061_1_, GameSettings p_i1061_2_)
@@ -60,7 +60,7 @@
             }
         }
 
-        this.list = new GuiSnooper.List();
+        this.list = new List();
     }
 
     public void handleMouseInput() throws IOException
