--- ../src-base/minecraft/net/minecraft/client/gui/GuiSlider.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiSlider.java
@@ -15,9 +15,9 @@
     private final float min;
     private final float max;
     private final GuiPageButtonList.GuiResponder responder;
-    private GuiSlider.FormatHelper formatHelper;
+    private FormatHelper formatHelper;
 
-    public GuiSlider(GuiPageButtonList.GuiResponder guiResponder, int idIn, int x, int y, String nameIn, float minIn, float maxIn, float defaultValue, GuiSlider.FormatHelper formatter)
+    public GuiSlider(GuiPageButtonList.GuiResponder guiResponder, int idIn, int x, int y, String nameIn, float minIn, float maxIn, float defaultValue, FormatHelper formatter)
     {
         super(idIn, x, y, 150, 20, "");
         this.name = nameIn;
