--- ../src-base/minecraft/net/minecraft/client/gui/GuiVideoSettings.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiVideoSettings.java
@@ -117,4 +117,12 @@
         this.drawCenteredString(this.fontRenderer, this.screenTitle, this.width / 2, 5, 16777215);
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
+
+    // FORGE: fix for MC-64581 very laggy mipmap slider
+    @Override
+    public void onGuiClosed()
+    {
+        super.onGuiClosed();
+        this.mc.gameSettings.onGuiClosed();
+    }
 }
