--- ../src-base/minecraft/net/minecraft/client/gui/MapItemRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/gui/MapItemRenderer.java
@@ -21,7 +21,7 @@
 {
     private static final ResourceLocation TEXTURE_MAP_ICONS = new ResourceLocation("textures/map/map_icons.png");
     private final TextureManager textureManager;
-    private final Map<String, MapItemRenderer.Instance> loadedMaps = Maps.<String, MapItemRenderer.Instance>newHashMap();
+    private final Map<String, Instance> loadedMaps = Maps.<String, Instance>newHashMap();
 
     public MapItemRenderer(TextureManager textureManagerIn)
     {
@@ -38,13 +38,13 @@
         this.getMapRendererInstance(mapdataIn).render(noOverlayRendering);
     }
 
-    private MapItemRenderer.Instance getMapRendererInstance(MapData mapdataIn)
+    private Instance getMapRendererInstance(MapData mapdataIn)
     {
-        MapItemRenderer.Instance mapitemrenderer$instance = this.loadedMaps.get(mapdataIn.mapName);
+        Instance mapitemrenderer$instance = this.loadedMaps.get(mapdataIn.mapName);
 
         if (mapitemrenderer$instance == null)
         {
-            mapitemrenderer$instance = new MapItemRenderer.Instance(mapdataIn);
+            mapitemrenderer$instance = new Instance(mapdataIn);
             this.loadedMaps.put(mapdataIn.mapName, mapitemrenderer$instance);
         }
 
@@ -59,7 +59,7 @@
 
     public void clearLoadedMaps()
     {
-        for (MapItemRenderer.Instance mapitemrenderer$instance : this.loadedMaps.values())
+        for (Instance mapitemrenderer$instance : this.loadedMaps.values())
         {
             this.textureManager.deleteTexture(mapitemrenderer$instance.location);
         }
@@ -139,6 +139,8 @@
             {
                 if (!noOverlayRendering || mapdecoration.renderOnFrame())
                 {
+                    if (mapdecoration.render(k)) { k++; continue; }
+                    MapItemRenderer.this.textureManager.bindTexture(MapItemRenderer.TEXTURE_MAP_ICONS); // Rebind in case custom render changes it
                     GlStateManager.pushMatrix();
                     GlStateManager.translate(0.0F + (float)mapdecoration.getX() / 2.0F + 64.0F, 0.0F + (float)mapdecoration.getY() / 2.0F + 64.0F, -0.02F);
                     GlStateManager.rotate((float)(mapdecoration.getRotation() * 360) / 16.0F, 0.0F, 0.0F, 1.0F);
