--- ../src-base/minecraft/net/minecraft/client/gui/ServerSelectionList.java
+++ ../src-work/minecraft/net/minecraft/client/gui/ServerSelectionList.java
@@ -14,7 +14,7 @@
     private final GuiMultiplayer owner;
     private final List<ServerListEntryNormal> serverListInternet = Lists.<ServerListEntryNormal>newArrayList();
     private final List<ServerListEntryLanDetected> serverListLan = Lists.<ServerListEntryLanDetected>newArrayList();
-    private final GuiListExtended.IGuiListEntry lanScanEntry = new ServerListEntryLanScan();
+    private final IGuiListEntry lanScanEntry = new ServerListEntryLanScan();
     private int selectedSlotIndex = -1;
 
     public ServerSelectionList(GuiMultiplayer ownerIn, Minecraft mcIn, int widthIn, int heightIn, int topIn, int bottomIn, int slotHeightIn)
@@ -23,7 +23,7 @@
         this.owner = ownerIn;
     }
 
-    public GuiListExtended.IGuiListEntry getListEntry(int index)
+    public IGuiListEntry getListEntry(int index)
     {
         if (index < this.serverListInternet.size())
         {
