--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -67,6 +_,12 @@
 import org.joml.Vector3f;
 
 public class Axolotl extends Animal implements LerpingModel, VariantHolder<Axolotl.Variant>, Bucketable {
+   // CraftBukkit start - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
+   @Override
+   public int getDefaultMaxAirSupply() {
+      return f_149099_;
+   }
+   // CraftBukkit end
    public static final int f_149090_ = 200;
    protected static final ImmutableList<? extends SensorType<? extends Sensor<? super Axolotl>>> f_149091_ = ImmutableList.of(SensorType.f_26811_, SensorType.f_26822_, SensorType.f_26814_, SensorType.f_148315_, SensorType.f_148316_);
    protected static final ImmutableList<? extends MemoryModuleType<?>> f_149092_ = ImmutableList.of(MemoryModuleType.f_26375_, MemoryModuleType.f_148204_, MemoryModuleType.f_148205_, MemoryModuleType.f_26368_, MemoryModuleType.f_148206_, MemoryModuleType.f_26371_, MemoryModuleType.f_26370_, MemoryModuleType.f_26326_, MemoryModuleType.f_26377_, MemoryModuleType.f_26372_, MemoryModuleType.f_26373_, MemoryModuleType.f_26331_, MemoryModuleType.f_26382_, MemoryModuleType.f_148195_, MemoryModuleType.f_148194_, MemoryModuleType.f_148196_, MemoryModuleType.f_148197_, MemoryModuleType.f_148198_, MemoryModuleType.f_148201_, MemoryModuleType.f_217768_);
@@ -174,7 +_,7 @@
    }
 
    public int m_6062_() {
-      return 6000;
+      return maxAirTicks; // CraftBukkit - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
    }
 
    public Axolotl.Variant m_28554_() {
@@ -368,7 +_,7 @@
       if (mobeffectinstance == null || mobeffectinstance.m_267633_(2399)) {
          int i = mobeffectinstance != null ? mobeffectinstance.m_19557_() : 0;
          int j = Math.min(2400, 100 + i);
-         p_149174_.m_147207_(new MobEffectInstance(MobEffects.f_19605_, j, 0), this);
+         p_149174_.addEffect(new MobEffectInstance(MobEffects.f_19605_, i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AXOLOTL); // CraftBukkit
       }
 
       p_149174_.m_21195_(MobEffects.f_19599_);
