--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -83,6 +_,13 @@
    private final Map<String, Vector3f> f_149101_ = Maps.newHashMap();
    private static final int f_149102_ = 100;
 
+   // CraftBukkit start - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
+   @Override
+   public int getDefaultMaxAirSupply() {
+      return f_149099_;
+   }
+   // CraftBukkit end
+
    public Axolotl(EntityType<? extends Axolotl> p_149105_, Level p_149106_) {
       super(p_149105_, p_149106_);
       this.m_21441_(BlockPathTypes.WATER, 0.0F);
@@ -174,14 +_,14 @@
    }
 
    public int m_6062_() {
-      return 6000;
+      return maxAirTicks; // CraftBukkit - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
    }
 
    public Axolotl.Variant m_149179_() {
       return Axolotl.Variant.f_149230_[this.f_19804_.m_135370_(f_149096_)];
    }
 
-   private void m_149117_(Axolotl.Variant p_149118_) {
+   public void m_149117_(Axolotl.Variant p_149118_) {
       this.f_19804_.m_135381_(f_149096_, p_149118_.m_149242_());
    }
 
@@ -378,7 +_,7 @@
       int i = mobeffectinstance != null ? mobeffectinstance.m_19557_() : 0;
       if (i < 2400) {
          i = Math.min(2400, 100 + i);
-         p_149174_.m_147207_(new MobEffectInstance(MobEffects.f_19605_, i, 0), this);
+         p_149174_.addEffect(new MobEffectInstance(MobEffects.f_19605_, i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AXOLOTL); // CraftBukkit
       }
 
       p_149174_.m_21195_(MobEffects.f_19599_);
