--- a/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
@@ -4,6 +_,8 @@
 import javax.annotation.Nullable;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import org.slf4j.Logger;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftEnderDragon;
+import org.bukkit.event.entity.EnderDragonChangePhaseEvent;
 
 public class EnderDragonPhaseManager {
    private static final Logger f_31408_ = LogUtils.getLogger();
@@ -22,6 +_,19 @@
          if (this.f_31411_ != null) {
             this.f_31411_.m_7081_();
          }
+
+         // CraftBukkit start - Call EnderDragonChangePhaseEvent
+         EnderDragonChangePhaseEvent event = new EnderDragonChangePhaseEvent(
+                 (CraftEnderDragon) this.f_31409_.getBukkitEntity(),
+                 (this.f_31411_ == null) ? null : CraftEnderDragon.getBukkitPhase(this.f_31411_.m_7309_()),
+                 CraftEnderDragon.getBukkitPhase(p_31417_)
+         );
+         this.f_31409_.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return;
+         }
+         p_31417_ = CraftEnderDragon.getMinecraftPhase(event.getNewPhase());
+         // CraftBukkit end
 
          this.f_31411_ = this.m_31418_(p_31417_);
          if (!this.f_31409_.f_19853_.f_46443_) {
