--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ItemCameraTransforms.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ItemCameraTransforms.java
@@ -12,6 +12,9 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.lwjgl.util.vector.Quaternion;
 
+/*
+ * @deprecated use {@link net.minecraftforge.client.model.IPerspectiveAwareModel} instead
+ */
 @SideOnly(Side.CLIENT)
 public class ItemCameraTransforms
 {
@@ -39,6 +42,7 @@
         this(ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT);
     }
 
+    @Deprecated
     public ItemCameraTransforms(ItemCameraTransforms transforms)
     {
         this.thirdperson_left = transforms.thirdperson_left;
@@ -51,6 +55,7 @@
         this.fixed = transforms.fixed;
     }
 
+    @Deprecated
     public ItemCameraTransforms(ItemTransformVec3f thirdperson_leftIn, ItemTransformVec3f thirdperson_rightIn, ItemTransformVec3f firstperson_leftIn, ItemTransformVec3f firstperson_rightIn, ItemTransformVec3f headIn, ItemTransformVec3f guiIn, ItemTransformVec3f groundIn, ItemTransformVec3f fixedIn)
     {
         this.thirdperson_left = thirdperson_leftIn;
@@ -63,7 +68,7 @@
         this.fixed = fixedIn;
     }
 
-    public void applyTransform(ItemCameraTransforms.TransformType type)
+    public void applyTransform(TransformType type)
     {
         applyTransformSide(this.getTransform(type), false);
     }
@@ -103,7 +108,8 @@
         return new Quaternion(f3 * f6 * f8 + f4 * f5 * f7, f4 * f5 * f8 - f3 * f6 * f7, f3 * f5 * f8 + f4 * f6 * f7, f4 * f6 * f8 - f3 * f5 * f7);
     }
 
-    public ItemTransformVec3f getTransform(ItemCameraTransforms.TransformType type)
+    @Deprecated
+    public ItemTransformVec3f getTransform(TransformType type)
     {
         switch (type)
         {
@@ -128,7 +134,7 @@
         }
     }
 
-    public boolean hasCustomTransform(ItemCameraTransforms.TransformType type)
+    public boolean hasCustomTransform(TransformType type)
     {
         return this.getTransform(type) != ItemTransformVec3f.DEFAULT;
     }
@@ -169,7 +175,7 @@
         }
 
     @SideOnly(Side.CLIENT)
-    public static enum TransformType
+    public static enum TransformType implements net.minecraftforge.common.model.IModelPart
     {
         NONE,
         THIRD_PERSON_LEFT_HAND,
