--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ModelBlock.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ModelBlock.java
@@ -32,7 +32,7 @@
 {
     private static final Logger LOGGER = LogManager.getLogger();
     @VisibleForTesting
-    static final Gson SERIALIZER = (new GsonBuilder()).registerTypeAdapter(ModelBlock.class, new ModelBlock.Deserializer()).registerTypeAdapter(BlockPart.class, new BlockPart.Deserializer()).registerTypeAdapter(BlockPartFace.class, new BlockPartFace.Deserializer()).registerTypeAdapter(BlockFaceUV.class, new BlockFaceUV.Deserializer()).registerTypeAdapter(ItemTransformVec3f.class, new ItemTransformVec3f.Deserializer()).registerTypeAdapter(ItemCameraTransforms.class, new ItemCameraTransforms.Deserializer()).registerTypeAdapter(ItemOverride.class, new ItemOverride.Deserializer()).create();
+    static final Gson SERIALIZER = (new GsonBuilder()).registerTypeAdapter(ModelBlock.class, new Deserializer()).registerTypeAdapter(BlockPart.class, new BlockPart.Deserializer()).registerTypeAdapter(BlockPartFace.class, new BlockPartFace.Deserializer()).registerTypeAdapter(BlockFaceUV.class, new BlockFaceUV.Deserializer()).registerTypeAdapter(ItemTransformVec3f.class, new ItemTransformVec3f.Deserializer()).registerTypeAdapter(ItemCameraTransforms.class, new ItemCameraTransforms.Deserializer()).registerTypeAdapter(ItemOverride.class, new ItemOverride.Deserializer()).create();
     private final List<BlockPart> elements;
     private final boolean gui3d;
     public final boolean ambientOcclusion;
@@ -134,10 +134,10 @@
             textureName = '#' + textureName;
         }
 
-        return this.resolveTextureName(textureName, new ModelBlock.Bookkeep(this));
+        return this.resolveTextureName(textureName, new Bookkeep(this));
     }
 
-    private String resolveTextureName(String textureName, ModelBlock.Bookkeep p_178302_2_)
+    private String resolveTextureName(String textureName, Bookkeep p_178302_2_)
     {
         if (this.startsWithHash(textureName))
         {
@@ -218,7 +218,7 @@
                     modelblock1 = modelblock1.parent;
                 }
 
-                throw new ModelBlock.LoopException();
+                throw new LoopException();
             }
             catch (NullPointerException var5)
             {
