--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
@@ -88,7 +88,7 @@
 
             public Builder(IBlockState state, IBakedModel model, TextureAtlasSprite texture, BlockPos pos)
             {
-                this(model.isAmbientOcclusion(), model.isGui3d(), model.getItemCameraTransforms(), model.getOverrides());
+                this(model.isAmbientOcclusion(state), model.isGui3d(), model.getItemCameraTransforms(), model.getOverrides());
                 this.builderTexture = model.getParticleTexture();
                 long i = MathHelper.getPositionRandom(pos);
 
@@ -132,19 +132,19 @@
                 }
             }
 
-            public SimpleBakedModel.Builder addFaceQuad(EnumFacing facing, BakedQuad quad)
+            public Builder addFaceQuad(EnumFacing facing, BakedQuad quad)
             {
                 (this.builderFaceQuads.get(facing)).add(quad);
                 return this;
             }
 
-            public SimpleBakedModel.Builder addGeneralQuad(BakedQuad quad)
+            public Builder addGeneralQuad(BakedQuad quad)
             {
                 this.builderGeneralQuads.add(quad);
                 return this;
             }
 
-            public SimpleBakedModel.Builder setTexture(TextureAtlasSprite texture)
+            public Builder setTexture(TextureAtlasSprite texture)
             {
                 this.builderTexture = texture;
                 return this;
