--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/Variant.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/Variant.java
@@ -12,7 +12,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class Variant
+public class Variant implements net.minecraftforge.client.model.ISmartVariant
 {
     private final ResourceLocation modelLocation;
     private final ModelRotation rotation;
@@ -32,11 +32,17 @@
         return this.modelLocation;
     }
 
+    @Deprecated
     public ModelRotation getRotation()
     {
         return this.rotation;
     }
 
+    public net.minecraftforge.common.model.IModelState getState()
+    {
+        return this.rotation;
+    }
+
     public boolean isUvLock()
     {
         return this.uvLock;
@@ -138,4 +144,9 @@
                 }
             }
         }
+
+    public net.minecraftforge.client.model.IModel process(net.minecraftforge.client.model.IModel base)
+    {
+        return base.uvlock(isUvLock());
+    }
 }
