--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenIceSpike.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenIceSpike.java
@@ -51,7 +51,7 @@
                             IBlockState iblockstate = worldIn.getBlockState(position.add(i1, k, j1));
                             Block block = iblockstate.getBlock();
 
-                            if (iblockstate.getMaterial() == Material.AIR || block == Blocks.DIRT || block == Blocks.SNOW || block == Blocks.ICE)
+                            if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, k, j1)) || block == Blocks.DIRT || block == Blocks.SNOW || block == Blocks.ICE)
                             {
                                 this.setBlockAndNotifyAdequately(worldIn, position.add(i1, k, j1), Blocks.PACKED_ICE.getDefaultState());
                             }
@@ -61,7 +61,7 @@
                                 iblockstate = worldIn.getBlockState(position.add(i1, -k, j1));
                                 block = iblockstate.getBlock();
 
-                                if (iblockstate.getMaterial() == Material.AIR || block == Blocks.DIRT || block == Blocks.SNOW || block == Blocks.ICE)
+                                if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, -k, j1)) || block == Blocks.DIRT || block == Blocks.SNOW || block == Blocks.ICE)
                                 {
                                     this.setBlockAndNotifyAdequately(worldIn, position.add(i1, -k, j1), Blocks.PACKED_ICE.getDefaultState());
                                 }
@@ -99,7 +99,7 @@
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos);
                         Block block1 = iblockstate1.getBlock();
 
-                        if (iblockstate1.getMaterial() != Material.AIR && block1 != Blocks.DIRT && block1 != Blocks.SNOW && block1 != Blocks.ICE && block1 != Blocks.PACKED_ICE)
+                        if (!iblockstate1.getBlock().isAir(iblockstate1, worldIn, blockpos) && block1 != Blocks.DIRT && block1 != Blocks.SNOW && block1 != Blocks.ICE && block1 != Blocks.PACKED_ICE)
                         {
                             break;
                         }
