--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenBush.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenBush.java
@@ -20,7 +20,7 @@
         {
             BlockPos blockpos = position.add(rand.nextInt(8) - rand.nextInt(8), rand.nextInt(4) - rand.nextInt(4), rand.nextInt(8) - rand.nextInt(8));
 
-            if (worldIn.isAirBlock(blockpos) && (!worldIn.provider.isNether() || blockpos.getY() < 255) && this.block.canBlockStay(worldIn, blockpos, this.block.getDefaultState()))
+            if (worldIn.isAirBlock(blockpos) && (!worldIn.provider.isNether() || blockpos.getY() < worldIn.getHeight() - 1) && this.block.canBlockStay(worldIn, blockpos, this.block.getDefaultState()))
             {
                 worldIn.setBlockState(blockpos, this.block.getDefaultState(), 2);
             }
