--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenSpikes.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenSpikes.java
@@ -12,10 +12,10 @@
 public class WorldGenSpikes extends WorldGenerator
 {
     private boolean crystalInvulnerable;
-    private WorldGenSpikes.EndSpike spike;
+    private EndSpike spike;
     private BlockPos beamTarget;
 
-    public void setSpike(WorldGenSpikes.EndSpike p_186143_1_)
+    public void setSpike(EndSpike p_186143_1_)
     {
         this.spike = p_186143_1_;
     }
