--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenHellLava.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenHellLava.java
@@ -25,7 +25,7 @@
         {
             return false;
         }
-        else if (worldIn.getBlockState(position).getMaterial() != Material.AIR && worldIn.getBlockState(position).getBlock() != Blocks.NETHERRACK)
+        else if (!worldIn.isAirBlock(position) && worldIn.getBlockState(position).getBlock() != Blocks.NETHERRACK)
         {
             return false;
         }
