--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenLiquids.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenLiquids.java
@@ -31,7 +31,7 @@
         {
             IBlockState iblockstate = worldIn.getBlockState(position);
 
-            if (iblockstate.getMaterial() != Material.AIR && iblockstate.getBlock() != Blocks.STONE)
+            if (!iblockstate.getBlock().isAir(iblockstate, worldIn, position) && iblockstate.getBlock() != Blocks.STONE)
             {
                 return false;
             }
