--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenLakes.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenLakes.java
@@ -161,7 +161,8 @@
 
                         if (worldIn.canBlockFreezeWater(position.add(k2, 4, l3)))
                         {
-                            worldIn.setBlockState(position.add(k2, 4, l3), Blocks.ICE.getDefaultState(), 2);
+                            int flag = net.minecraftforge.common.ForgeModContainer.fixVanillaCascading ? 2| 16 : 2; //Forge: With bit 5 unset, it will notify neighbors and load adjacent chunks.
+                            worldIn.setBlockState(position.add(k2, 4, l3), Blocks.ICE.getDefaultState(), flag); //Forge
                         }
                     }
                 }
