--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
@@ -17,7 +17,7 @@
 
     public WorldGenMinable(IBlockState state, int blockCount)
     {
-        this(state, blockCount, new WorldGenMinable.StonePredicate());
+        this(state, blockCount, new StonePredicate());
     }
 
     public WorldGenMinable(IBlockState state, int blockCount, Predicate<IBlockState> p_i45631_3_)
@@ -73,7 +73,8 @@
                                 {
                                     BlockPos blockpos = new BlockPos(l1, i2, j2);
 
-                                    if (this.predicate.apply(worldIn.getBlockState(blockpos)))
+                                    IBlockState state = worldIn.getBlockState(blockpos);
+                                    if (state.getBlock().isReplaceableOreGen(state, worldIn, blockpos, this.predicate))
                                     {
                                         worldIn.setBlockState(blockpos, this.oreBlock, 2);
                                     }
