--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -49,7 +_,7 @@
    private static final String f_230342_ = "gameteststructures";
    private static final String f_230343_ = ".nbt";
    private static final String f_230344_ = ".snbt";
-   private final Map<ResourceLocation, Optional<StructureTemplate>> f_230345_ = Maps.newConcurrentMap();
+   public final Map<ResourceLocation, Optional<StructureTemplate>> f_230345_ = Maps.newConcurrentMap(); // CatServer - private -> public
    private final DataFixer f_230346_;
    private ResourceManager f_230347_;
    private final Path f_230348_;
@@ -112,7 +_,7 @@
       this.f_230345_.clear();
    }
 
-   private Optional<StructureTemplate> m_230427_(ResourceLocation p_230428_) {
+   public Optional<StructureTemplate> m_230427_(ResourceLocation p_230428_) { // CatServer - private -> public
       ResourceLocation resourcelocation = f_244413_.m_245698_(p_230428_);
       return this.m_230372_(() -> {
          return this.f_230347_.m_215595_(resourcelocation);
@@ -133,7 +_,7 @@
       return this.m_230394_(Paths.get("gameteststructures"), "minecraft", ".snbt");
    }
 
-   private Optional<StructureTemplate> m_230431_(ResourceLocation p_230432_) {
+   public Optional<StructureTemplate> m_230431_(ResourceLocation p_230432_) { // CatServer - private -> public
       if (!Files.isDirectory(this.f_230348_)) {
          return Optional.empty();
       } else {
@@ -227,7 +_,7 @@
       }
    }
 
-   private StructureTemplate m_230377_(InputStream p_230378_) throws IOException {
+   public StructureTemplate m_230377_(InputStream p_230378_) throws IOException { // CatServer - private -> public
       CompoundTag compoundtag = NbtIo.m_128939_(p_230378_);
       return this.m_230404_(compoundtag);
    }
@@ -286,7 +_,7 @@
       }
    }
 
-   private static Path m_230411_(Path p_230412_, ResourceLocation p_230413_, String p_230414_) {
+   public static Path m_230411_(Path p_230412_, ResourceLocation p_230413_, String p_230414_) { // CatServer - private -> public
       if (p_230413_.m_135815_().contains("//")) {
          throw new ResourceLocationException("Invalid resource path: " + p_230413_);
       } else {
