--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureManager.java
@@ -32,7 +_,7 @@
    private static final String f_163771_ = "structures";
    private static final String f_163772_ = ".nbt";
    private static final String f_163773_ = ".snbt";
-   private final Map<ResourceLocation, Optional<StructureTemplate>> f_74326_ = Maps.newConcurrentMap();
+   public final Map<ResourceLocation, Optional<StructureTemplate>> f_74326_ = Maps.newConcurrentMap();
    private final DataFixer f_74327_;
    private ResourceManager f_74328_;
    private final Path f_74329_;
@@ -66,7 +_,7 @@
       this.f_74326_.clear();
    }
 
-   private Optional<StructureTemplate> m_163776_(ResourceLocation p_163777_) {
+   public Optional<StructureTemplate> m_163776_(ResourceLocation p_163777_) {
       ResourceLocation resourcelocation = new ResourceLocation(p_163777_.m_135827_(), "structures/" + p_163777_.m_135815_() + ".nbt");
 
       try {
@@ -100,7 +_,7 @@
       }
    }
 
-   private Optional<StructureTemplate> m_163778_(ResourceLocation p_163779_) {
+   public Optional<StructureTemplate> m_163778_(ResourceLocation p_163779_) {
       if (!this.f_74329_.toFile().isDirectory()) {
          return Optional.empty();
       } else {
@@ -133,7 +_,7 @@
       }
    }
 
-   private StructureTemplate m_74337_(InputStream p_74338_) throws IOException {
+   public StructureTemplate m_74337_(InputStream p_74338_) throws IOException {
       CompoundTag compoundtag = NbtIo.m_128939_(p_74338_);
       return this.m_74339_(compoundtag);
    }
@@ -202,7 +_,7 @@
       }
    }
 
-   private Path m_74348_(ResourceLocation p_74349_, String p_74350_) {
+   public Path m_74348_(ResourceLocation p_74349_, String p_74350_) {
       if (p_74349_.m_135815_().contains("//")) {
          throw new ResourceLocationException("Invalid resource path: " + p_74349_);
       } else {
