--- a/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
+++ b/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
@@ -68,7 +_,7 @@
       return this.field_233757_c_.get(p_180374_1_.func_111167_a()) != null;
    }
 
-   private void func_111121_a(AttributeModifier p_111121_1_) {
+   public void func_111121_a(AttributeModifier p_111121_1_) {
       AttributeModifier attributemodifier = this.field_233757_c_.putIfAbsent(p_111121_1_.func_111167_a(), p_111121_1_);
       if (attributemodifier != null) {
          throw new IllegalArgumentException("Modifier is already applied on this attribute!");
