--- a/net/minecraft/entity/ai/attributes/RangedAttribute.java
+++ b/net/minecraft/entity/ai/attributes/RangedAttribute.java
@@ -4,7 +_,7 @@
 
 public class RangedAttribute extends Attribute {
    private final double field_111120_a;
-   private final double field_111118_b;
+   public double field_111118_b; // Spigot
 
    public RangedAttribute(String p_i231504_1_, double p_i231504_2_, double p_i231504_4_, double p_i231504_6_) {
       super(p_i231504_1_, p_i231504_2_);
@@ -20,6 +_,7 @@
    }
 
    public double func_111109_a(double p_111109_1_) {
+      if (p_111109_1_ != p_111109_1_) return func_111110_b(); // CraftBukkit
       return MathHelper.func_151237_a(p_111109_1_, this.field_111120_a, this.field_111118_b);
    }
 }
