--- ../src-base/minecraft/net/minecraftforge/common/ForgeHooks.java
+++ ../src-work/minecraft/net/minecraftforge/common/ForgeHooks.java
@@ -312,7 +312,7 @@
                     return;
                 }
 
-                itemstack = block.func_185473_a(this.theWorld, blockpos, iblockstate);
+                itemstack = block.getItem(this.theWorld, blockpos, iblockstate);
 
                 if (itemstack == null)
                 {
@@ -358,7 +358,7 @@
                     EntityMinecart entityminecart = (EntityMinecart)this.objectMouseOver.entityHit;
                     Item item;
 
-                    switch (entityminecart.func_184264_v())
+                    switch (entityminecart.getType())
                     {
                         case FURNACE:
                             item = Items.furnace_minecart;
@@ -383,7 +383,7 @@
                 }
                 else if (this.objectMouseOver.entityHit instanceof EntityBoat)
                 {
-                    itemstack = new ItemStack(((EntityBoat)this.objectMouseOver.entityHit).func_184455_j());
+                    itemstack = new ItemStack(((EntityBoat)this.objectMouseOver.entityHit).getItemBoat());
                 }
                 else if (this.objectMouseOver.entityHit instanceof EntityArmorStand)
                 {
@@ -391,7 +391,7 @@
                 }
                 else if (this.objectMouseOver.entityHit instanceof EntityEnderCrystal)
                 {
-                    itemstack = new ItemStack(Items.field_185158_cP);
+                    itemstack = new ItemStack(Items.END_CRYSTAL);
                 }
                 else
                 {
@@ -403,7 +403,7 @@
                     }
 
                     itemstack = new ItemStack(Items.spawn_egg);
-                    ItemMonsterPlacer.func_185078_a(itemstack, s);
+                    ItemMonsterPlacer.applyEntityIdToItemStack(itemstack, s);
                 }
             }
          */
@@ -884,7 +884,7 @@
     }
 
     /**
-     * Default implementation of IRecipe.func_179532_b {getRemainingItems} because
+     * Default implementation of IRecipe.getRemainingItems {getRemainingItems} because
      * this is just copy pasted over a lot of recipes.
      *
      * Another use case for java 8 but sadly we can't use it!
