--- ../src-base/minecraft/net/minecraftforge/common/IMinecartCollisionHandler.java
+++ ../src-work/minecraft/net/minecraftforge/common/IMinecartCollisionHandler.java
@@ -57,7 +57,6 @@
      * @return The collision box, cannot be null.
      */
     public AxisAlignedBB getMinecartCollisionBox(EntityMinecart cart);
-
     /**
      * This function replaces the function of the same name in EntityMinecart.
      * It defines whether minecarts are solid to the player.
@@ -66,4 +65,3 @@
      */
     public AxisAlignedBB getBoundingBox(EntityMinecart cart);
 }
-
