--- ../src-base/minecraft/net/minecraftforge/common/UsernameCache.java
+++ ../src-work/minecraft/net/minecraftforge/common/UsernameCache.java
@@ -42,6 +42,7 @@
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonSyntaxException;
 
+import net.minecraft.server.MinecraftServer;
 import net.minecraftforge.fml.relauncher.FMLInjectionData;
 
 /**
@@ -58,7 +59,7 @@
 
     private static final Charset charset = Charsets.UTF_8;
 
-    private static final File saveFile = new File( /* The minecraft dir */(File) FMLInjectionData.data()[6], "usernamecache.json");
+    private static final File saveFile = new File( /* The minecraft dir */MinecraftServer.serverConfigDir, "usernamecache.json");
     private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
 
     private static final Logger log = LogManager.getLogger(UsernameCache.class);
