--- a/net/minecraft/util/datafix/DataFixers.java
+++ b/net/minecraft/util/datafix/DataFixers.java
@@ -1,10 +_,7 @@
 package net.minecraft.util.datafix;
 
 import com.google.common.collect.ImmutableMap;
-import com.mojang.datafixers.DSL;
-import com.mojang.datafixers.DataFixer;
-import com.mojang.datafixers.DataFixerBuilder;
-import com.mojang.datafixers.Typed;
+import com.mojang.datafixers.*;
 import com.mojang.datafixers.schemas.Schema;
 import java.util.Map;
 import java.util.Objects;
@@ -344,6 +_,14 @@
       Schema schema46 = p_14514_.addSchema(1456, f_14509_);
       p_14514_.addFixer(new EntityItemFrameDirectionFix(schema46, false));
       Schema schema47 = p_14514_.addSchema(1458, f_14509_);
+      // CraftBukkit start
+      p_14514_.addFixer(new DataFix(schema47, false) {
+         @Override
+         protected TypeRewriteRule makeRule() {
+            return this.fixTypeEverywhereTyped("Player CustomName", this.getInputSchema().getType(References.f_16772_), (typed) -> typed.update(DSL.remainderFinder(), EntityCustomNameToComponentFix::m_15407_));
+         }
+      });
+      // CraftBukkit end
       p_14514_.addFixer(new EntityCustomNameToComponentFix(schema47, false));
       p_14514_.addFixer(new ItemCustomNameToComponentFix(schema47, false));
       p_14514_.addFixer(new BlockEntityCustomNameToComponentFix(schema47, false));
@@ -385,8 +_,8 @@
       p_14514_.addFixer(BlockRenameFix.m_14914_(schema59, "Rename kelp/kelptop", m_14524_(ImmutableMap.of("minecraft:kelp_top", "minecraft:kelp", "minecraft:kelp", "minecraft:kelp_plant"))));
       p_14514_.addFixer(ItemRenameFix.m_16003_(schema59, "Rename kelptop", m_14517_("minecraft:kelp_top", "minecraft:kelp")));
       p_14514_.addFixer(new NamedEntityFix(schema59, false, "Command block block entity custom name fix", References.f_16781_, "minecraft:command_block") {
-         protected Typed<?> m_7504_(Typed<?> p_14541_) {
-            return p_14541_.update(DSL.remainderFinder(), EntityCustomNameToComponentFix::m_15407_);
+         protected Typed<?> m_7504_(Typed<?> p_14549_) {
+            return p_14549_.update(DSL.remainderFinder(), EntityCustomNameToComponentFix::m_15407_);
          }
       });
       p_14514_.addFixer(new NamedEntityFix(schema59, false, "Command block minecart custom name fix", References.f_16786_, "minecraft:commandblock_minecart") {
