--- a/net/minecraft/util/datafix/DataFixesManager.java
+++ b/net/minecraft/util/datafix/DataFixesManager.java
@@ -315,6 +_,14 @@
       Schema schema46 = p_210891_0_.addSchema(1456, field_207593_f);
       p_210891_0_.addFixer(new EntityItemFrameFacing(schema46, false));
       Schema schema47 = p_210891_0_.addSchema(1458, field_207593_f);
+      // CraftBukkit start
+      p_210891_0_.addFixer(new com.mojang.datafixers.DataFix(schema47, false) {
+         @Override
+         protected com.mojang.datafixers.TypeRewriteRule makeRule() {
+            return this.fixTypeEverywhereTyped("Player CustomName", this.getInputSchema().getType(TypeReferences.field_211286_b), (typed) -> typed.update(DSL.remainderFinder(), CustomNameStringToComponentEntity::func_209740_a));
+         }
+      });
+      // CraftBukkit end
       p_210891_0_.addFixer(new CustomNameStringToComponentEntity(schema47, false));
       p_210891_0_.addFixer(new CustomNameStringToComponentItem(schema47, false));
       p_210891_0_.addFixer(new CustomNameStringToComponentFixTileEntity(schema47, false));
